{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Objectives\n",
    "\n",
    "1. Convert newick tree to PhyloXML format.\n",
    "1. Fixes internal node names from an IQTREE tree. \n",
    "   - internal nodes by default get named the confidence (ex. \"99/100\")\n",
    "   - this notebook instead gives them the name NODEi where i is an integer starting from 0.\n",
    "   - NODE0 therefore becomes the root.\n",
    "1. Convert tree to data frame.\n",
    "1. Combine metadata with tree data frame.\n",
    "1. Export:\n",
    "   - Tree (newick, nexus, phyloxml)\n",
    "   - Tree Dataframe (tsv)\n",
    "   - Color mapping (tsv)\n",
    "   - Lat Lon Mapping (tsv)\n",
    "   - Augur JSON\n",
    "   - Auspice JSON"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Module Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# stats and plotting\n",
    "import pandas as pd\n",
    "from matplotlib import colors, gridspec\n",
    "import math\n",
    "\n",
    "# Phylogenetics\n",
    "from Bio import Phylo\n",
    "import treetime # Clock filter tips\n",
    "\n",
    "# System misc\n",
    "\n",
    "import time # Augur export\n",
    "import os\n",
    "import copy\n",
    "from augur import utils, export_v2\n",
    "import json\n",
    "\n",
    "# Logging to file\n",
    "import sys\n",
    "import io\n",
    "\n",
    "# Writing python objects to file\n",
    "import dill"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input File Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_path = \"../../docs/results/latest/iqtree/iqtree-core_chromosome.filter5_post.cf.tree\"\n",
    "metadata_path = \"../../docs/results/latest/metadata/metadata.tsv\"\n",
    "aln_path = \"../../docs/results/latest/snippy_multi/snippy-core_chromosome.snps.filter5.aln\"\n",
    "auspice_config_path = \"../../config/auspice_config.json\"\n",
    "auspice_remote_dir_path = \"../../auspice/\"\n",
    "outdir = \"../../results/parse_tree\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "\n",
    "# Custom script variables\n",
    "SCRIPT_NAME = \"parse_tree\"\n",
    "NAME_COL = \"Sample\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Metadata Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669605.1_ASM966960v1_genomic</th>\n",
       "      <td>42117</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722922</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP283</th>\n",
       "      <td>IP283</td>\n",
       "      <td>1994</td>\n",
       "      <td>-27</td>\n",
       "      <td>India</td>\n",
       "      <td>Gujarat</td>\n",
       "      <td>22.3511</td>\n",
       "      <td>78.6677</td>\n",
       "      <td>22.385</td>\n",
       "      <td>71.7453</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>IP283</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP562</th>\n",
       "      <td>IP562</td>\n",
       "      <td>1947</td>\n",
       "      <td>-74</td>\n",
       "      <td>Iran</td>\n",
       "      <td>Kurdistan</td>\n",
       "      <td>32.6475</td>\n",
       "      <td>54.5644</td>\n",
       "      <td>35.6728</td>\n",
       "      <td>47.0124</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>IP562</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP542</th>\n",
       "      <td>IP542</td>\n",
       "      <td>1952</td>\n",
       "      <td>-69</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.44197</td>\n",
       "      <td>38.4314</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>1.ANT1</td>\n",
       "      <td>IP542</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP543</th>\n",
       "      <td>IP543</td>\n",
       "      <td>1953</td>\n",
       "      <td>-68</td>\n",
       "      <td>Democratic Republic of the Congo</td>\n",
       "      <td>NA</td>\n",
       "      <td>-2.98143</td>\n",
       "      <td>23.8223</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>1.ANT1</td>\n",
       "      <td>IP543</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP557</th>\n",
       "      <td>IP557</td>\n",
       "      <td>1963</td>\n",
       "      <td>-58</td>\n",
       "      <td>Iran</td>\n",
       "      <td>Kurdistan</td>\n",
       "      <td>32.6475</td>\n",
       "      <td>54.5644</td>\n",
       "      <td>35.6728</td>\n",
       "      <td>47.0124</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>IP557</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>633 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Strain  Date DateBP  \\\n",
       "Sample                                                    \n",
       "GCA_009909635.1_ASM990963v1_genomic   9_10  1923    -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic  42126  2006    -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic  42123  2005    -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic  42118  2005    -16   \n",
       "GCA_009669605.1_ASM966960v1_genomic  42117  2005    -16   \n",
       "...                                    ...   ...    ...   \n",
       "IP283                                IP283  1994    -27   \n",
       "IP562                                IP562  1947    -74   \n",
       "IP542                                IP542  1952    -69   \n",
       "IP543                                IP543  1953    -68   \n",
       "IP557                                IP557  1963    -58   \n",
       "\n",
       "                                                              Country  \\\n",
       "Sample                                                                  \n",
       "GCA_009909635.1_ASM990963v1_genomic                            Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                             China   \n",
       "GCA_009669605.1_ASM966960v1_genomic                             China   \n",
       "...                                                               ...   \n",
       "IP283                                                           India   \n",
       "IP562                                                            Iran   \n",
       "IP542                                                           Kenya   \n",
       "IP543                                Democratic Republic of the Congo   \n",
       "IP557                                                            Iran   \n",
       "\n",
       "                                          Province CountryLat CountryLon  \\\n",
       "Sample                                                                     \n",
       "GCA_009909635.1_ASM990963v1_genomic  Rostov Oblast    64.6863    97.7453   \n",
       "GCA_009669545.1_ASM966954v1_genomic       Xinjiang    35.0001        105   \n",
       "GCA_009669555.1_ASM966955v1_genomic       Xinjiang    35.0001        105   \n",
       "GCA_009669565.1_ASM966956v1_genomic       Xinjiang    35.0001        105   \n",
       "GCA_009669605.1_ASM966960v1_genomic       Xinjiang    35.0001        105   \n",
       "...                                            ...        ...        ...   \n",
       "IP283                                      Gujarat    22.3511    78.6677   \n",
       "IP562                                    Kurdistan    32.6475    54.5644   \n",
       "IP542                                           NA    1.44197    38.4314   \n",
       "IP543                                           NA   -2.98143    23.8223   \n",
       "IP557                                    Kurdistan    32.6475    54.5644   \n",
       "\n",
       "                                    ProvinceLat ProvinceLon      Biovar  \\\n",
       "Sample                                                                    \n",
       "GCA_009909635.1_ASM990963v1_genomic     47.6222     40.7958  Medievalis   \n",
       "GCA_009669545.1_ASM966954v1_genomic     42.4805     85.4633     Antiqua   \n",
       "GCA_009669555.1_ASM966955v1_genomic     42.4805     85.4633     Antiqua   \n",
       "GCA_009669565.1_ASM966956v1_genomic     42.4805     85.4633     Antiqua   \n",
       "GCA_009669605.1_ASM966960v1_genomic     42.4805     85.4633     Antiqua   \n",
       "...                                         ...         ...         ...   \n",
       "IP283                                    22.385     71.7453  Orientalis   \n",
       "IP562                                   35.6728     47.0124  Medievalis   \n",
       "IP542                                        NA          NA     Antiqua   \n",
       "IP543                                        NA          NA     Antiqua   \n",
       "IP557                                   35.6728     47.0124  Medievalis   \n",
       "\n",
       "                                    Branch_Major Branch_Minor     BioSample  \\\n",
       "Sample                                                                        \n",
       "GCA_009909635.1_ASM990963v1_genomic        2.MED       2.MED1  SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic        0.ANT       0.ANT1  SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic        0.ANT       0.ANT1  SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic        0.ANT       0.ANT1  SAMN07722923   \n",
       "GCA_009669605.1_ASM966960v1_genomic        0.ANT       0.ANT1  SAMN07722922   \n",
       "...                                          ...          ...           ...   \n",
       "IP283                                      1.ORI       1.ORI3         IP283   \n",
       "IP562                                      2.MED       2.MED1         IP562   \n",
       "IP542                                      1.ANT       1.ANT1         IP542   \n",
       "IP543                                      1.ANT       1.ANT1         IP543   \n",
       "IP557                                      2.MED       2.MED1         IP557   \n",
       "\n",
       "                                          BioSampleComment  Branch_Number  \n",
       "Sample                                                                     \n",
       "GCA_009909635.1_ASM990963v1_genomic  KEEP: Assembly Modern              2  \n",
       "GCA_009669545.1_ASM966954v1_genomic  KEEP: Assembly Modern              0  \n",
       "GCA_009669555.1_ASM966955v1_genomic  KEEP: Assembly Modern              0  \n",
       "GCA_009669565.1_ASM966956v1_genomic  KEEP: Assembly Modern              0  \n",
       "GCA_009669605.1_ASM966960v1_genomic  KEEP: Assembly Modern              0  \n",
       "...                                                    ...            ...  \n",
       "IP283                                   KEEP: Local Modern              1  \n",
       "IP562                                   KEEP: Local Modern              2  \n",
       "IP542                                   KEEP: Local Modern              1  \n",
       "IP543                                   KEEP: Local Modern              1  \n",
       "IP557                                   KEEP: Local Modern              2  \n",
       "\n",
       "[633 rows x 15 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse the metadata\n",
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "# set the metadata index\n",
    "metadata_df.set_index(NAME_COL, inplace=True)\n",
    "\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Convert tree to PhyloXML format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_xml = os.path.join(outdir, \"temp.xml\")\n",
    "Phylo.convert(tree_path, \"newick\", tmp_xml, \"phyloxml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. Fix internal node names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in XML tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the new tree\n",
    "tree = Phylo.read(tmp_xml, \"phyloxml\")\n",
    "\n",
    "# Deepest nodes first\n",
    "tree.ladderize(reverse=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fix node names and confidences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Counter for integer node names\n",
    "node_i = 0\n",
    "\n",
    "# Iterate through the nodes in the tree\n",
    "for c in tree.find_clades():\n",
    "    # Check if the name has 'accidentally' become the confidence vals\n",
    "    ufboot_val = NO_DATA_CHAR    \n",
    "    scf_val = NO_DATA_CHAR\n",
    "    # If it's a named node, parse it\n",
    "    if c.name:\n",
    "        name_split = c.name.split(\"/\")\n",
    "        \n",
    "        # Check if the name has 'accidentally' become the confidence vals\n",
    "        if len(name_split) > 1:\n",
    "            # Name the internal node\n",
    "            c.name = \"NODE\" + str(node_i)    \n",
    "\n",
    "            # Assign confidence values     \n",
    "            ufboot_val = float(name_split[0])\n",
    "            scf_val = float(name_split[1])\n",
    "    # If it's not a named node, give it a name\n",
    "    else:\n",
    "        c.name = \"NODE\" + str(node_i) \n",
    "        # The one and only confidence is \"unknown\" and if UFboot\n",
    "        ufboot_val = c.confidences[0].value\n",
    "        c.confidences = []\n",
    "        \n",
    "    \n",
    "    ufboot_conf = Phylo.PhyloXML.Confidence(ufboot_val, type=\"ufboot\")\n",
    "    scf_conf = Phylo.PhyloXML.Confidence(scf_val, type=\"scf\")\n",
    "    \n",
    "    c.confidences.append(ufboot_conf)\n",
    "    c.confidences.append(scf_conf)\n",
    "        \n",
    "    # Increment node counter if not a terminal\n",
    "    if not c.is_terminal():\n",
    "        node_i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Convert tree to dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize the tree dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.850400e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>1.327080e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.248400e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>94.6</td>\n",
       "      <td>1.363640e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE630</th>\n",
       "      <td>19</td>\n",
       "      <td>48.8</td>\n",
       "      <td>6.380000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269245.1_ASM26924v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.230000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE631</th>\n",
       "      <td>37</td>\n",
       "      <td>49.6</td>\n",
       "      <td>2.936000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269125.1_ASM26912v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.990000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269365.1_ASM26936v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.841000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1266 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot   sCF  Branch_Length  \\\n",
       "Name                                                             \n",
       "NODE0                                 100    NA   0.000000e+00   \n",
       "SAMEA3541827                           NA    NA   3.850400e-06   \n",
       "NODE1                                 100    85   1.327080e-05   \n",
       "SAMEA104233046                         NA    NA   2.248400e-06   \n",
       "NODE2                                 100  94.6   1.363640e-05   \n",
       "...                                   ...   ...            ...   \n",
       "NODE630                                19  48.8   6.380000e-08   \n",
       "GCA_000269245.1_ASM26924v1_genomic     NA    NA   2.230000e-07   \n",
       "NODE631                                37  49.6   2.936000e-07   \n",
       "GCA_000269125.1_ASM26912v1_genomic     NA    NA   4.990000e-08   \n",
       "GCA_000269365.1_ASM26936v1_genomic     NA    NA   8.841000e-07   \n",
       "\n",
       "                                   Branch_Support_Color coord_x coord_y  \n",
       "Name                                                                     \n",
       "NODE0                                                NA      NA      NA  \n",
       "SAMEA3541827                                         NA      NA      NA  \n",
       "NODE1                                                NA      NA      NA  \n",
       "SAMEA104233046                                       NA      NA      NA  \n",
       "NODE2                                                NA      NA      NA  \n",
       "...                                                 ...     ...     ...  \n",
       "NODE630                                              NA      NA      NA  \n",
       "GCA_000269245.1_ASM26924v1_genomic                   NA      NA      NA  \n",
       "NODE631                                              NA      NA      NA  \n",
       "GCA_000269125.1_ASM26912v1_genomic                   NA      NA      NA  \n",
       "GCA_000269365.1_ASM26936v1_genomic                   NA      NA      NA  \n",
       "\n",
       "[1266 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# initialize empty dataframe\n",
    "tree_df = pd.DataFrame(columns = [\n",
    "    \"Name\",\n",
    "    \"UFboot\", \n",
    "    \"sCF\", \n",
    "    \"Branch_Length\", \n",
    "    \"Branch_Support_Color\", \n",
    "    \"coord_x\", \n",
    "    \"coord_y\",]\n",
    ")\n",
    "\n",
    "for c in tree.find_clades():\n",
    "    if not c.branch_length: c.branch_length = 0.0\n",
    "    \n",
    "    node_data = {\n",
    "        \"Name\" : c.name,\n",
    "        \"UFboot\" : [conf.value for conf in c.confidences if conf.type==\"ufboot\"][0],\n",
    "        \"sCF\" :  [conf.value for conf in c.confidences if conf.type==\"scf\"][0],\n",
    "        \"Branch_Support_Color\" :  NO_DATA_CHAR,\n",
    "        \"Branch_Length\" :  c.branch_length,\n",
    "        \"coord_x\" :  NO_DATA_CHAR,\n",
    "        \"coord_y\" :  NO_DATA_CHAR,\n",
    "    }\n",
    "    \n",
    "    tree_df = tree_df.append(node_data, ignore_index=True)\n",
    "\n",
    "# Set the index to the node name\n",
    "tree_df.set_index(\"Name\", inplace=True)\n",
    "\n",
    "# Visualize data frame\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add plotting x and y coordinates of each node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>17.8205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.850400e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.8504e-06</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>1.327080e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.32708e-05</td>\n",
       "      <td>3.15625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.248400e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.55192e-05</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>94.6</td>\n",
       "      <td>1.363640e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.69072e-05</td>\n",
       "      <td>4.3125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE630</th>\n",
       "      <td>19</td>\n",
       "      <td>48.8</td>\n",
       "      <td>6.380000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.58989e-05</td>\n",
       "      <td>632.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269245.1_ASM26924v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.230000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.61219e-05</td>\n",
       "      <td>632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE631</th>\n",
       "      <td>37</td>\n",
       "      <td>49.6</td>\n",
       "      <td>2.936000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.61925e-05</td>\n",
       "      <td>633.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269125.1_ASM26912v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.990000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.62424e-05</td>\n",
       "      <td>633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269365.1_ASM26936v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.841000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.70766e-05</td>\n",
       "      <td>634</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1266 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot   sCF  Branch_Length  \\\n",
       "Name                                                             \n",
       "NODE0                                 100    NA   0.000000e+00   \n",
       "SAMEA3541827                           NA    NA   3.850400e-06   \n",
       "NODE1                                 100    85   1.327080e-05   \n",
       "SAMEA104233046                         NA    NA   2.248400e-06   \n",
       "NODE2                                 100  94.6   1.363640e-05   \n",
       "...                                   ...   ...            ...   \n",
       "NODE630                                19  48.8   6.380000e-08   \n",
       "GCA_000269245.1_ASM26924v1_genomic     NA    NA   2.230000e-07   \n",
       "NODE631                                37  49.6   2.936000e-07   \n",
       "GCA_000269125.1_ASM26912v1_genomic     NA    NA   4.990000e-08   \n",
       "GCA_000269365.1_ASM26936v1_genomic     NA    NA   8.841000e-07   \n",
       "\n",
       "                                   Branch_Support_Color      coord_x  coord_y  \n",
       "Name                                                                           \n",
       "NODE0                                                NA            0  17.8205  \n",
       "SAMEA3541827                                         NA   3.8504e-06        1  \n",
       "NODE1                                                NA  1.32708e-05  3.15625  \n",
       "SAMEA104233046                                       NA  1.55192e-05        2  \n",
       "NODE2                                                NA  2.69072e-05   4.3125  \n",
       "...                                                 ...          ...      ...  \n",
       "NODE630                                              NA  7.58989e-05   632.75  \n",
       "GCA_000269245.1_ASM26924v1_genomic                   NA  7.61219e-05      632  \n",
       "NODE631                                              NA  7.61925e-05    633.5  \n",
       "GCA_000269125.1_ASM26912v1_genomic                   NA  7.62424e-05      633  \n",
       "GCA_000269365.1_ASM26936v1_genomic                   NA  7.70766e-05      634  \n",
       "\n",
       "[1266 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_posns = get_x_positions(tree)\n",
    "y_posns = get_y_positions(tree)\n",
    "\n",
    "# Add x and y coordinates as other attributes\n",
    "for c in tree.find_clades():\n",
    "    # x coordinates will be of branch length units\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    # y coordinates will be arbitrary, based on number of tips\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    # Add data to tree dataframe\n",
    "    tree_df.at[c.name, 'coord_x'] = coord_x\n",
    "    tree_df.at[c.name, 'coord_y'] = coord_y    \n",
    "\n",
    "# Visualize dataframe\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add branch support color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>black</td>\n",
       "      <td>0</td>\n",
       "      <td>17.8205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.850400e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>3.8504e-06</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>1.327080e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>1.32708e-05</td>\n",
       "      <td>3.15625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.248400e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>1.55192e-05</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>94.6</td>\n",
       "      <td>1.363640e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>2.69072e-05</td>\n",
       "      <td>4.3125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE630</th>\n",
       "      <td>19</td>\n",
       "      <td>48.8</td>\n",
       "      <td>6.380000e-08</td>\n",
       "      <td>black</td>\n",
       "      <td>7.58989e-05</td>\n",
       "      <td>632.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269245.1_ASM26924v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.230000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.61219e-05</td>\n",
       "      <td>632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE631</th>\n",
       "      <td>37</td>\n",
       "      <td>49.6</td>\n",
       "      <td>2.936000e-07</td>\n",
       "      <td>black</td>\n",
       "      <td>7.61925e-05</td>\n",
       "      <td>633.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269125.1_ASM26912v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.990000e-08</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.62424e-05</td>\n",
       "      <td>633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269365.1_ASM26936v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.841000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.70766e-05</td>\n",
       "      <td>634</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1266 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot   sCF  Branch_Length  \\\n",
       "Name                                                             \n",
       "NODE0                                 100    NA   0.000000e+00   \n",
       "SAMEA3541827                           NA    NA   3.850400e-06   \n",
       "NODE1                                 100    85   1.327080e-05   \n",
       "SAMEA104233046                         NA    NA   2.248400e-06   \n",
       "NODE2                                 100  94.6   1.363640e-05   \n",
       "...                                   ...   ...            ...   \n",
       "NODE630                                19  48.8   6.380000e-08   \n",
       "GCA_000269245.1_ASM26924v1_genomic     NA    NA   2.230000e-07   \n",
       "NODE631                                37  49.6   2.936000e-07   \n",
       "GCA_000269125.1_ASM26912v1_genomic     NA    NA   4.990000e-08   \n",
       "GCA_000269365.1_ASM26936v1_genomic     NA    NA   8.841000e-07   \n",
       "\n",
       "                                   Branch_Support_Color      coord_x  coord_y  \n",
       "Name                                                                           \n",
       "NODE0                                             black            0  17.8205  \n",
       "SAMEA3541827                                       grey   3.8504e-06        1  \n",
       "NODE1                                             black  1.32708e-05  3.15625  \n",
       "SAMEA104233046                                     grey  1.55192e-05        2  \n",
       "NODE2                                             black  2.69072e-05   4.3125  \n",
       "...                                                 ...          ...      ...  \n",
       "NODE630                                           black  7.58989e-05   632.75  \n",
       "GCA_000269245.1_ASM26924v1_genomic                 grey  7.61219e-05      632  \n",
       "NODE631                                           black  7.61925e-05    633.5  \n",
       "GCA_000269125.1_ASM26912v1_genomic                 grey  7.62424e-05      633  \n",
       "GCA_000269365.1_ASM26936v1_genomic                 grey  7.70766e-05      634  \n",
       "\n",
       "[1266 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for c in tree.find_clades():   \n",
    "    # Default Color\n",
    "    branch_color = LOW_COL\n",
    "    # Terminal branches will be grey\n",
    "    if c.is_terminal():\n",
    "        branch_color = TERM_COL\n",
    "    # High support branches\n",
    "    ufboot_val = tree_df[\"UFboot\"][c.name]\n",
    "    scf_val = tree_df[\"sCF\"][c.name]\n",
    "    if scf_val != NO_DATA_CHAR and ufboot_val != NO_DATA_CHAR:\n",
    "        if scf_val >= SCF_THRESH and ufboot_val >= UFBOOT_THRESH:\n",
    "            branch_color = HIGH_COL\n",
    "    tree_df.at[c.name, \"Branch_Support_Color\"] = branch_color            \n",
    "\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 5. Combine metadata into tree dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse metadata into df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669605.1_ASM966960v1_genomic</th>\n",
       "      <td>42117</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722922</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP283</th>\n",
       "      <td>IP283</td>\n",
       "      <td>1994</td>\n",
       "      <td>-27</td>\n",
       "      <td>India</td>\n",
       "      <td>Gujarat</td>\n",
       "      <td>22.3511</td>\n",
       "      <td>78.6677</td>\n",
       "      <td>22.385</td>\n",
       "      <td>71.7453</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>IP283</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP562</th>\n",
       "      <td>IP562</td>\n",
       "      <td>1947</td>\n",
       "      <td>-74</td>\n",
       "      <td>Iran</td>\n",
       "      <td>Kurdistan</td>\n",
       "      <td>32.6475</td>\n",
       "      <td>54.5644</td>\n",
       "      <td>35.6728</td>\n",
       "      <td>47.0124</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>IP562</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP542</th>\n",
       "      <td>IP542</td>\n",
       "      <td>1952</td>\n",
       "      <td>-69</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.44197</td>\n",
       "      <td>38.4314</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>1.ANT1</td>\n",
       "      <td>IP542</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP543</th>\n",
       "      <td>IP543</td>\n",
       "      <td>1953</td>\n",
       "      <td>-68</td>\n",
       "      <td>Democratic Republic of the Congo</td>\n",
       "      <td>NA</td>\n",
       "      <td>-2.98143</td>\n",
       "      <td>23.8223</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>1.ANT1</td>\n",
       "      <td>IP543</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP557</th>\n",
       "      <td>IP557</td>\n",
       "      <td>1963</td>\n",
       "      <td>-58</td>\n",
       "      <td>Iran</td>\n",
       "      <td>Kurdistan</td>\n",
       "      <td>32.6475</td>\n",
       "      <td>54.5644</td>\n",
       "      <td>35.6728</td>\n",
       "      <td>47.0124</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>IP557</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>633 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Strain  Date DateBP  \\\n",
       "Sample                                                    \n",
       "GCA_009909635.1_ASM990963v1_genomic   9_10  1923    -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic  42126  2006    -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic  42123  2005    -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic  42118  2005    -16   \n",
       "GCA_009669605.1_ASM966960v1_genomic  42117  2005    -16   \n",
       "...                                    ...   ...    ...   \n",
       "IP283                                IP283  1994    -27   \n",
       "IP562                                IP562  1947    -74   \n",
       "IP542                                IP542  1952    -69   \n",
       "IP543                                IP543  1953    -68   \n",
       "IP557                                IP557  1963    -58   \n",
       "\n",
       "                                                              Country  \\\n",
       "Sample                                                                  \n",
       "GCA_009909635.1_ASM990963v1_genomic                            Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                             China   \n",
       "GCA_009669605.1_ASM966960v1_genomic                             China   \n",
       "...                                                               ...   \n",
       "IP283                                                           India   \n",
       "IP562                                                            Iran   \n",
       "IP542                                                           Kenya   \n",
       "IP543                                Democratic Republic of the Congo   \n",
       "IP557                                                            Iran   \n",
       "\n",
       "                                          Province CountryLat CountryLon  \\\n",
       "Sample                                                                     \n",
       "GCA_009909635.1_ASM990963v1_genomic  Rostov Oblast    64.6863    97.7453   \n",
       "GCA_009669545.1_ASM966954v1_genomic       Xinjiang    35.0001        105   \n",
       "GCA_009669555.1_ASM966955v1_genomic       Xinjiang    35.0001        105   \n",
       "GCA_009669565.1_ASM966956v1_genomic       Xinjiang    35.0001        105   \n",
       "GCA_009669605.1_ASM966960v1_genomic       Xinjiang    35.0001        105   \n",
       "...                                            ...        ...        ...   \n",
       "IP283                                      Gujarat    22.3511    78.6677   \n",
       "IP562                                    Kurdistan    32.6475    54.5644   \n",
       "IP542                                           NA    1.44197    38.4314   \n",
       "IP543                                           NA   -2.98143    23.8223   \n",
       "IP557                                    Kurdistan    32.6475    54.5644   \n",
       "\n",
       "                                    ProvinceLat ProvinceLon      Biovar  \\\n",
       "Sample                                                                    \n",
       "GCA_009909635.1_ASM990963v1_genomic     47.6222     40.7958  Medievalis   \n",
       "GCA_009669545.1_ASM966954v1_genomic     42.4805     85.4633     Antiqua   \n",
       "GCA_009669555.1_ASM966955v1_genomic     42.4805     85.4633     Antiqua   \n",
       "GCA_009669565.1_ASM966956v1_genomic     42.4805     85.4633     Antiqua   \n",
       "GCA_009669605.1_ASM966960v1_genomic     42.4805     85.4633     Antiqua   \n",
       "...                                         ...         ...         ...   \n",
       "IP283                                    22.385     71.7453  Orientalis   \n",
       "IP562                                   35.6728     47.0124  Medievalis   \n",
       "IP542                                        NA          NA     Antiqua   \n",
       "IP543                                        NA          NA     Antiqua   \n",
       "IP557                                   35.6728     47.0124  Medievalis   \n",
       "\n",
       "                                    Branch_Major Branch_Minor     BioSample  \\\n",
       "Sample                                                                        \n",
       "GCA_009909635.1_ASM990963v1_genomic        2.MED       2.MED1  SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic        0.ANT       0.ANT1  SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic        0.ANT       0.ANT1  SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic        0.ANT       0.ANT1  SAMN07722923   \n",
       "GCA_009669605.1_ASM966960v1_genomic        0.ANT       0.ANT1  SAMN07722922   \n",
       "...                                          ...          ...           ...   \n",
       "IP283                                      1.ORI       1.ORI3         IP283   \n",
       "IP562                                      2.MED       2.MED1         IP562   \n",
       "IP542                                      1.ANT       1.ANT1         IP542   \n",
       "IP543                                      1.ANT       1.ANT1         IP543   \n",
       "IP557                                      2.MED       2.MED1         IP557   \n",
       "\n",
       "                                          BioSampleComment  Branch_Number  \n",
       "Sample                                                                     \n",
       "GCA_009909635.1_ASM990963v1_genomic  KEEP: Assembly Modern              2  \n",
       "GCA_009669545.1_ASM966954v1_genomic  KEEP: Assembly Modern              0  \n",
       "GCA_009669555.1_ASM966955v1_genomic  KEEP: Assembly Modern              0  \n",
       "GCA_009669565.1_ASM966956v1_genomic  KEEP: Assembly Modern              0  \n",
       "GCA_009669605.1_ASM966960v1_genomic  KEEP: Assembly Modern              0  \n",
       "...                                                    ...            ...  \n",
       "IP283                                   KEEP: Local Modern              1  \n",
       "IP562                                   KEEP: Local Modern              2  \n",
       "IP542                                   KEEP: Local Modern              1  \n",
       "IP543                                   KEEP: Local Modern              1  \n",
       "IP557                                   KEEP: Local Modern              2  \n",
       "\n",
       "[633 rows x 15 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse the metadata\n",
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "# set the metadata index\n",
    "metadata_df.set_index(NAME_COL, inplace=True)\n",
    "\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add metadata to tree and tree dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>black</td>\n",
       "      <td>0</td>\n",
       "      <td>17.8205</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.850400e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>3.8504e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>RISE509</td>\n",
       "      <td>[-2876:-2677]</td>\n",
       "      <td>[-4897:-4698]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE2</td>\n",
       "      <td>SAMEA3541827</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>1.327080e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>1.32708e-05</td>\n",
       "      <td>3.15625</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.248400e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>1.55192e-05</td>\n",
       "      <td>2</td>\n",
       "      <td>Gyvakarai1</td>\n",
       "      <td>[-2621:-2472]</td>\n",
       "      <td>[-4642:-4493]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>55.35</td>\n",
       "      <td>23.75</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>SAMEA104233046</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>94.6</td>\n",
       "      <td>1.363640e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>2.69072e-05</td>\n",
       "      <td>4.3125</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE630</th>\n",
       "      <td>19</td>\n",
       "      <td>48.8</td>\n",
       "      <td>6.380000e-08</td>\n",
       "      <td>black</td>\n",
       "      <td>7.58989e-05</td>\n",
       "      <td>632.75</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269245.1_ASM26924v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.230000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.61219e-05</td>\n",
       "      <td>632</td>\n",
       "      <td>PY-113</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141506</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE631</th>\n",
       "      <td>37</td>\n",
       "      <td>49.6</td>\n",
       "      <td>2.936000e-07</td>\n",
       "      <td>black</td>\n",
       "      <td>7.61925e-05</td>\n",
       "      <td>633.5</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269125.1_ASM26912v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.990000e-08</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.62424e-05</td>\n",
       "      <td>633</td>\n",
       "      <td>PY-93</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-6.25</td>\n",
       "      <td>-78.8333</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141496</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269365.1_ASM26936v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.841000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.70766e-05</td>\n",
       "      <td>634</td>\n",
       "      <td>PY-61</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141483</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1266 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot   sCF  Branch_Length  \\\n",
       "Name                                                             \n",
       "NODE0                                 100    NA   0.000000e+00   \n",
       "SAMEA3541827                           NA    NA   3.850400e-06   \n",
       "NODE1                                 100    85   1.327080e-05   \n",
       "SAMEA104233046                         NA    NA   2.248400e-06   \n",
       "NODE2                                 100  94.6   1.363640e-05   \n",
       "...                                   ...   ...            ...   \n",
       "NODE630                                19  48.8   6.380000e-08   \n",
       "GCA_000269245.1_ASM26924v1_genomic     NA    NA   2.230000e-07   \n",
       "NODE631                                37  49.6   2.936000e-07   \n",
       "GCA_000269125.1_ASM26912v1_genomic     NA    NA   4.990000e-08   \n",
       "GCA_000269365.1_ASM26936v1_genomic     NA    NA   8.841000e-07   \n",
       "\n",
       "                                   Branch_Support_Color      coord_x  coord_y  \\\n",
       "Name                                                                            \n",
       "NODE0                                             black            0  17.8205   \n",
       "SAMEA3541827                                       grey   3.8504e-06        1   \n",
       "NODE1                                             black  1.32708e-05  3.15625   \n",
       "SAMEA104233046                                     grey  1.55192e-05        2   \n",
       "NODE2                                             black  2.69072e-05   4.3125   \n",
       "...                                                 ...          ...      ...   \n",
       "NODE630                                           black  7.58989e-05   632.75   \n",
       "GCA_000269245.1_ASM26924v1_genomic                 grey  7.61219e-05      632   \n",
       "NODE631                                           black  7.61925e-05    633.5   \n",
       "GCA_000269125.1_ASM26912v1_genomic                 grey  7.62424e-05      633   \n",
       "GCA_000269365.1_ASM26936v1_genomic                 grey  7.70766e-05      634   \n",
       "\n",
       "                                        Strain           Date         DateBP  \\\n",
       "Name                                                                           \n",
       "NODE0                                       NA             NA             NA   \n",
       "SAMEA3541827                           RISE509  [-2876:-2677]  [-4897:-4698]   \n",
       "NODE1                                       NA             NA             NA   \n",
       "SAMEA104233046                      Gyvakarai1  [-2621:-2472]  [-4642:-4493]   \n",
       "NODE2                                       NA             NA             NA   \n",
       "...                                        ...            ...            ...   \n",
       "NODE630                                     NA             NA             NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic      PY-113           2010            -11   \n",
       "NODE631                                     NA             NA             NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic       PY-93           2010            -11   \n",
       "GCA_000269365.1_ASM26936v1_genomic       PY-61           2010            -11   \n",
       "\n",
       "                                      Country          Province CountryLat  \\\n",
       "Name                                                                         \n",
       "NODE0                                      NA                NA         NA   \n",
       "SAMEA3541827                           Russia  Krasnoyarsk Krai    64.6863   \n",
       "NODE1                                      NA                NA         NA   \n",
       "SAMEA104233046                      Lithuania  Panevezys County      55.35   \n",
       "NODE2                                      NA                NA         NA   \n",
       "...                                       ...               ...        ...   \n",
       "NODE630                                    NA                NA         NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic       Peru       La Libertad   -6.86997   \n",
       "NODE631                                    NA                NA         NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic       Peru         Cajamarca   -6.86997   \n",
       "GCA_000269365.1_ASM26936v1_genomic       Peru       La Libertad   -6.86997   \n",
       "\n",
       "                                   CountryLon ProvinceLat ProvinceLon  \\\n",
       "Name                                                                    \n",
       "NODE0                                      NA          NA          NA   \n",
       "SAMEA3541827                          97.7453     63.3234      97.098   \n",
       "NODE1                                      NA          NA          NA   \n",
       "SAMEA104233046                          23.75     55.9156     25.0312   \n",
       "NODE2                                      NA          NA          NA   \n",
       "...                                       ...         ...         ...   \n",
       "NODE630                                    NA          NA          NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic   -75.0459          -8       -78.5   \n",
       "NODE631                                    NA          NA          NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic   -75.0459       -6.25    -78.8333   \n",
       "GCA_000269365.1_ASM26936v1_genomic   -75.0459          -8       -78.5   \n",
       "\n",
       "                                        Biovar Branch_Major Branch_Minor  \\\n",
       "Name                                                                       \n",
       "NODE0                                       NA           NA           NA   \n",
       "SAMEA3541827                        Bronze Age        0.PRE       0.PRE2   \n",
       "NODE1                                       NA           NA           NA   \n",
       "SAMEA104233046                      Bronze Age        0.PRE       0.PRE1   \n",
       "NODE2                                       NA           NA           NA   \n",
       "...                                        ...          ...          ...   \n",
       "NODE630                                     NA           NA           NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "NODE631                                     NA           NA           NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "GCA_000269365.1_ASM26936v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "\n",
       "                                         BioSample       BioSampleComment  \\\n",
       "Name                                                                        \n",
       "NODE0                                           NA                     NA   \n",
       "SAMEA3541827                          SAMEA3541827      KEEP: SRA Ancient   \n",
       "NODE1                                           NA                     NA   \n",
       "SAMEA104233046                      SAMEA104233046      KEEP: SRA Ancient   \n",
       "NODE2                                           NA                     NA   \n",
       "...                                            ...                    ...   \n",
       "NODE630                                         NA                     NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic    SAMN02141506  KEEP: Assembly Modern   \n",
       "NODE631                                         NA                     NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic    SAMN02141496  KEEP: Assembly Modern   \n",
       "GCA_000269365.1_ASM26936v1_genomic    SAMN02141483  KEEP: Assembly Modern   \n",
       "\n",
       "                                   Branch_Number  \n",
       "Name                                              \n",
       "NODE0                                         NA  \n",
       "SAMEA3541827                                   0  \n",
       "NODE1                                         NA  \n",
       "SAMEA104233046                                 0  \n",
       "NODE2                                         NA  \n",
       "...                                          ...  \n",
       "NODE630                                       NA  \n",
       "GCA_000269245.1_ASM26924v1_genomic             1  \n",
       "NODE631                                       NA  \n",
       "GCA_000269125.1_ASM26912v1_genomic             1  \n",
       "GCA_000269365.1_ASM26936v1_genomic             1  \n",
       "\n",
       "[1266 rows x 21 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Iterate through the different metadata attributes\n",
    "for attr in metadata_df.columns:\n",
    "    # Initialize an empty column for the attribute\n",
    "    tree_df[attr] = [NO_DATA_CHAR for row in range(0,len(tree_df))]\n",
    "    # Iterate over the nodes in the tree\n",
    "    for c in tree.find_clades():\n",
    "        # Initialize to empty\n",
    "        attr_val = NO_DATA_CHAR\n",
    "        # If the node shows up in the metadata\n",
    "        if c.name in metadata_df.index:\n",
    "            tree_df.at[c.name, attr] = metadata_df[attr][c.name]\n",
    "            attr_val = metadata_df[attr][c.name]         \n",
    "\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add Reference Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in tree_df.columns:\n",
    "    if col in REF_META:\n",
    "        tree_df.at[\"Reference\", col] = REF_META[col]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make NA values consistent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "tree_df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ?. Add Metadata As Node Comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in tree.find_clades():\n",
    "    c.comment = \"\"\n",
    "    for col in tree_df.columns:\n",
    "        col_val = tree_df[col][c.name]   \n",
    "        # Test for non ascii\n",
    "        if type(col_val) == str and not col_val.isascii():\n",
    "            col_val = col_val.encode(\"unicode_escape\")     \n",
    "        # Add comment\n",
    "        if not hasattr(c, \"comment\") or not c.comment:\n",
    "            c.comment = \"&{}={}\".format(col, col_val)\n",
    "        else:\n",
    "            c.comment += \",{}={}\".format(col, col_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 6. Export"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Objects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save tree and tree dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe\n",
    "out_path_df = os.path.join(outdir, SCRIPT_NAME + \".tsv\" )\n",
    "out_path_pickle_df = os.path.join(outdir,  SCRIPT_NAME + \".df.obj\" )\n",
    "\n",
    "tree_df.to_csv(out_path_df, sep=\"\\t\")\n",
    "with open(out_path_pickle_df,\"wb\") as outfile:\n",
    "    dill.dump(tree_df, outfile)\n",
    "\n",
    "# Phyloxml\n",
    "out_path_xml = os.path.join(outdir,  SCRIPT_NAME + \".xml\" )\n",
    "Phylo.write(tree, out_path_xml, 'phyloxml')\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \".nexus\" )\n",
    "Phylo.write(tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \".phylo.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tree, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \".nwk\" )\n",
    "Phylo.write(tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save lat and lon\n",
    "\n",
    "Write an output tsv that is:  \n",
    "resolution   name   lat    lon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "latlon_dict = {\"Country\": {}, \"Province\": {}}\n",
    "out_path_lat_lon = os.path.join(outdir, SCRIPT_NAME + \"_latlon.tsv\")\n",
    "\n",
    "# Columns in data frame will be ex: CountryLat, CountryLon\n",
    "for c in tree.find_clades():\n",
    "    # Iterate through each geographic level\n",
    "    for level in latlon_dict:\n",
    "        # Store the location name\n",
    "        loc = tree_df[level][c.name]\n",
    "        # If the node has location data and is new\n",
    "        if loc != NO_DATA_CHAR and loc not in latlon_dict[level]:\n",
    "            latlon_dict[level][loc] = {}\n",
    "            latlon_dict[level][loc][\"lat\"] = tree_df[level + \"Lat\"][c.name]\n",
    "            latlon_dict[level][loc][\"lon\"] = tree_df[level + \"Lon\"][c.name]\n",
    "\n",
    "with open(out_path_lat_lon, \"w\") as outfile:\n",
    "    for level in latlon_dict:\n",
    "        for loc in latlon_dict[level]:\n",
    "            outfile.write(level.lower() + \"\\t\"\n",
    "                         + loc + \"\\t\"\n",
    "                         + str(latlon_dict[level][loc][\"lat\"]) + \"\\t\"\n",
    "                         + str(latlon_dict[level][loc][\"lon\"]) + \"\\n\"\n",
    "                 )\n",
    "\n",
    "# For clarity with auspice        \n",
    "auspice_latlons_path = out_path_lat_lon         "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save colors\n",
    "\n",
    "Write an output tsv that is:\n",
    "attribute   attribute_val   hex_color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_colors = os.path.join(outdir, SCRIPT_NAME + \"_colors.tsv\")\n",
    "file_colors = open(out_path_colors, \"w\")\n",
    "\n",
    "hex_dict = {}\n",
    "\n",
    "for attr in ATTRIBUTE_LIST:\n",
    "    # Create the color map\n",
    "    attr_key = attr.lower()\n",
    "    hex_dict[attr_key] = {}\n",
    "    for t in tree.get_terminals():\n",
    "        attr_val = tree_df[attr][t.name]\n",
    "        if attr_val not in hex_dict[attr_key] and attr_val != NO_DATA_CHAR:\n",
    "            hex_dict[attr_key][attr_val] = {}\n",
    "            \n",
    "    # Create the custom color map (pyplot)\n",
    "    cmap = plt.get_cmap(CONT_COLOR_PAL, len(hex_dict[attr_key]))\n",
    "    # Convert the color map to a list of RGB values\n",
    "    cmaplist = [cmap(i) for i in range(cmap.N)]\n",
    "    # Convert RGB values to hex colors\n",
    "    attr_hex = [colors.to_hex(col) for col in cmaplist]\n",
    "    \n",
    "    # Assign colors to value\n",
    "    for attr_val, attr_col in zip(hex_dict[attr_key], attr_hex):\n",
    "        hex_dict[attr_key][attr_val] = attr_col   \n",
    "\n",
    "for attr_key in hex_dict:\n",
    "    for attr_val in hex_dict[attr_key]:\n",
    "        file_colors.write(str(attr_key) + \"\\t\" + str(attr_val) + \"\\t\" + str(hex_dict[attr_key][attr_val]) + \"\\n\")\n",
    "\n",
    "file_colors.close()        \n",
    "\n",
    "# For clarity with auspice\n",
    "auspice_colors_path = out_path_colors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Augur JSON\n",
    "\n",
    "  - alignment (empty)\n",
    "  - input_tree (tree_path)\n",
    "  - nodes (node_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ufboot': 'NA', 'scf': 'NA', 'branch_length': 1.8768e-06, 'strain': 'NA', 'date': 1992.0, 'datebp': -29.0, 'country': 'United States of America', 'province': 'Colorado', 'biovar': 'Orientalis', 'branch_major': '1.ORI', 'branch_minor': '1.ORI1', 'biosample': 'NA', 'biosamplecomment': 'KEEP: Assembly Modern Reference', 'branch_number': '1'}\n"
     ]
    }
   ],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=out_path_nwk, \n",
    "    aln_path=aln_path, \n",
    "    tree=tree, \n",
    "    tree_df=tree_df, \n",
    "    color_keyword_exclude=[\"color\", \"coord\", \"lat\", \"lon\"],\n",
    "    type_convert = {\n",
    "        \"Branch_Number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "out_path_augur_json = os.path.join(outdir, SCRIPT_NAME + \"_augur.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Auspice JSON\n",
    "\n",
    "Manual edits of https://github.com/nextstrain/augur/blob/master/augur/export_v2.py\n",
    "\n",
    "This can then be used for auspice via:\n",
    "\n",
    "```\n",
    "HOST=\"localhost\" auspice view --datasetDir .\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n",
      "Validating schema of '../../config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'rate', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_fold_change', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'raw_date', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait branch_minor were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait branch_number were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '../../results/parse_tree/parse_tree_auspice.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '../../results/parse_tree/parse_tree_auspice.json' succeeded.\n",
      "Validation successful for local JSON.\n",
      "Validating produced JSON\n",
      "Validating schema of '../../auspice/plague-phylogeography_parse-tree.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '../../auspice/plague-phylogeography_parse-tree.json' succeeded.\n",
      "Validation successful for remote JSON.\n"
     ]
    }
   ],
   "source": [
    "auspice_dict = auspice_export(\n",
    "    tree=tree, \n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=auspice_colors_path,\n",
    "    auspice_latlons_path=auspice_latlons_path, \n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(outdir, SCRIPT_NAME + \"_auspice.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "out_path_auspice_remote_json = os.path.join(auspice_remote_dir_path, AUSPICE_PREFIX + SCRIPT_NAME.replace(\"_\",\"-\") + \".json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "print(\"Validation successful for remote JSON.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEPRECATED: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "os.remove(tmp_xml)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
