{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "careful-broadcasting",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "1. Estimate Clock Model\n",
    "1. Filter Outliers\n",
    "1. Add Clock Model to Dataframe\n",
    "\n",
    "NOTE: Filter outliers after clock model estimated?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "resistant-charlotte",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "satisfied-front",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "union-excellence",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bio\n",
    "import treetime\n",
    "from Bio import Phylo\n",
    "\n",
    "# Plotting\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "import seaborn as sns\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "# Stats\n",
    "import pandas as pd\n",
    "\n",
    "# System IO\n",
    "import dill\n",
    "import copy\n",
    "import os\n",
    "import io\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "involved-korea",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "digital-lemon",
   "metadata": {},
   "outputs": [],
   "source": [
    "# snakemake output\n",
    "aln_path = \"../../docs/results/latest/snippy_multi/snippy-core_chromosome.snps.filter5.aln\"\n",
    "\n",
    "# jupyter notebook output\n",
    "tree_df_path = \"../../results/parse_tree/parse_tree.tsv\"\n",
    "tree_df_dill = \"../../results/parse_tree/parse_tree.df.obj\"\n",
    "tree_path = \"../../results/parse_tree/parse_tree.nwk\"\n",
    "auspice_latlon_path = \"../../results/parse_tree/parse_tree_latlon.tsv\"\n",
    "auspice_colors_path = \"../../results/parse_tree/parse_tree_colors.tsv\"\n",
    "\n",
    "# static dir\n",
    "auspice_config_path = \"../../config/auspice_config.json\"\n",
    "auspice_remote_dir_path = \"../../auspice/\"\n",
    "\n",
    "\n",
    "outdir = \"../../results/clock_model\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "weighted-punishment",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "municipal-accordance",
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "\n",
    "NAME_COL = \"Name\"\n",
    "SCRIPT_NAME = \"clock_model\"\n",
    "\n",
    "# Clock model params\n",
    "\n",
    "N_IQD = 3\n",
    "MAX_ITER = 3\n",
    "TC = \"skyline\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "polished-conviction",
   "metadata": {},
   "source": [
    "## Import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "colonial-symposium",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = Phylo.read(tree_path, \"newick\")\n",
    "tree.ladderize(reverse=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "tight-handy",
   "metadata": {},
   "source": [
    "## Import Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "approximate-waterproof",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.820516</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.850400e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>RISE509</td>\n",
       "      <td>[-2876:-2677]</td>\n",
       "      <td>[-4897:-4698]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE2</td>\n",
       "      <td>SAMEA3541827</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>1.327080e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>3.156250</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.248400e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>Gyvakarai1</td>\n",
       "      <td>[-2621:-2472]</td>\n",
       "      <td>[-4642:-4493]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>55.35</td>\n",
       "      <td>23.75</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>SAMEA104233046</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>94.6</td>\n",
       "      <td>1.363640e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>4.312500</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE630</th>\n",
       "      <td>19</td>\n",
       "      <td>48.8</td>\n",
       "      <td>6.380000e-08</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>632.750000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269245.1_ASM26924v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.230000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>632.000000</td>\n",
       "      <td>PY-113</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141506</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE631</th>\n",
       "      <td>37</td>\n",
       "      <td>49.6</td>\n",
       "      <td>2.936000e-07</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>633.500000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269125.1_ASM26912v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.990000e-08</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>633.000000</td>\n",
       "      <td>PY-93</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-6.25</td>\n",
       "      <td>-78.8333</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141496</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269365.1_ASM26936v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.841000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>634.000000</td>\n",
       "      <td>PY-61</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141483</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1266 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot   sCF  Branch_Length  \\\n",
       "Name                                                             \n",
       "NODE0                                  NA    NA   0.000000e+00   \n",
       "SAMEA3541827                           NA    NA   3.850400e-06   \n",
       "NODE1                                 100    85   1.327080e-05   \n",
       "SAMEA104233046                         NA    NA   2.248400e-06   \n",
       "NODE2                                 100  94.6   1.363640e-05   \n",
       "...                                   ...   ...            ...   \n",
       "NODE630                                19  48.8   6.380000e-08   \n",
       "GCA_000269245.1_ASM26924v1_genomic     NA    NA   2.230000e-07   \n",
       "NODE631                                37  49.6   2.936000e-07   \n",
       "GCA_000269125.1_ASM26912v1_genomic     NA    NA   4.990000e-08   \n",
       "GCA_000269365.1_ASM26936v1_genomic     NA    NA   8.841000e-07   \n",
       "\n",
       "                                   Branch_Support_Color   coord_x     coord_y  \\\n",
       "Name                                                                            \n",
       "NODE0                                             black  0.000000   17.820516   \n",
       "SAMEA3541827                                       grey  0.000004    1.000000   \n",
       "NODE1                                             black  0.000013    3.156250   \n",
       "SAMEA104233046                                     grey  0.000016    2.000000   \n",
       "NODE2                                             black  0.000027    4.312500   \n",
       "...                                                 ...       ...         ...   \n",
       "NODE630                                           black  0.000076  632.750000   \n",
       "GCA_000269245.1_ASM26924v1_genomic                 grey  0.000076  632.000000   \n",
       "NODE631                                           black  0.000076  633.500000   \n",
       "GCA_000269125.1_ASM26912v1_genomic                 grey  0.000076  633.000000   \n",
       "GCA_000269365.1_ASM26936v1_genomic                 grey  0.000077  634.000000   \n",
       "\n",
       "                                        Strain           Date         DateBP  \\\n",
       "Name                                                                           \n",
       "NODE0                                       NA             NA             NA   \n",
       "SAMEA3541827                           RISE509  [-2876:-2677]  [-4897:-4698]   \n",
       "NODE1                                       NA             NA             NA   \n",
       "SAMEA104233046                      Gyvakarai1  [-2621:-2472]  [-4642:-4493]   \n",
       "NODE2                                       NA             NA             NA   \n",
       "...                                        ...            ...            ...   \n",
       "NODE630                                     NA             NA             NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic      PY-113           2010            -11   \n",
       "NODE631                                     NA             NA             NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic       PY-93           2010            -11   \n",
       "GCA_000269365.1_ASM26936v1_genomic       PY-61           2010            -11   \n",
       "\n",
       "                                      Country          Province CountryLat  \\\n",
       "Name                                                                         \n",
       "NODE0                                      NA                NA         NA   \n",
       "SAMEA3541827                           Russia  Krasnoyarsk Krai    64.6863   \n",
       "NODE1                                      NA                NA         NA   \n",
       "SAMEA104233046                      Lithuania  Panevezys County      55.35   \n",
       "NODE2                                      NA                NA         NA   \n",
       "...                                       ...               ...        ...   \n",
       "NODE630                                    NA                NA         NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic       Peru       La Libertad   -6.86997   \n",
       "NODE631                                    NA                NA         NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic       Peru         Cajamarca   -6.86997   \n",
       "GCA_000269365.1_ASM26936v1_genomic       Peru       La Libertad   -6.86997   \n",
       "\n",
       "                                   CountryLon ProvinceLat ProvinceLon  \\\n",
       "Name                                                                    \n",
       "NODE0                                      NA          NA          NA   \n",
       "SAMEA3541827                          97.7453     63.3234      97.098   \n",
       "NODE1                                      NA          NA          NA   \n",
       "SAMEA104233046                          23.75     55.9156     25.0312   \n",
       "NODE2                                      NA          NA          NA   \n",
       "...                                       ...         ...         ...   \n",
       "NODE630                                    NA          NA          NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic   -75.0459          -8       -78.5   \n",
       "NODE631                                    NA          NA          NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic   -75.0459       -6.25    -78.8333   \n",
       "GCA_000269365.1_ASM26936v1_genomic   -75.0459          -8       -78.5   \n",
       "\n",
       "                                        Biovar Branch_Major Branch_Minor  \\\n",
       "Name                                                                       \n",
       "NODE0                                       NA           NA           NA   \n",
       "SAMEA3541827                        Bronze Age        0.PRE       0.PRE2   \n",
       "NODE1                                       NA           NA           NA   \n",
       "SAMEA104233046                      Bronze Age        0.PRE       0.PRE1   \n",
       "NODE2                                       NA           NA           NA   \n",
       "...                                        ...          ...          ...   \n",
       "NODE630                                     NA           NA           NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "NODE631                                     NA           NA           NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "GCA_000269365.1_ASM26936v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "\n",
       "                                         BioSample       BioSampleComment  \\\n",
       "Name                                                                        \n",
       "NODE0                                           NA                     NA   \n",
       "SAMEA3541827                          SAMEA3541827      KEEP: SRA Ancient   \n",
       "NODE1                                           NA                     NA   \n",
       "SAMEA104233046                      SAMEA104233046      KEEP: SRA Ancient   \n",
       "NODE2                                           NA                     NA   \n",
       "...                                            ...                    ...   \n",
       "NODE630                                         NA                     NA   \n",
       "GCA_000269245.1_ASM26924v1_genomic    SAMN02141506  KEEP: Assembly Modern   \n",
       "NODE631                                         NA                     NA   \n",
       "GCA_000269125.1_ASM26912v1_genomic    SAMN02141496  KEEP: Assembly Modern   \n",
       "GCA_000269365.1_ASM26936v1_genomic    SAMN02141483  KEEP: Assembly Modern   \n",
       "\n",
       "                                   Branch_Number  \n",
       "Name                                              \n",
       "NODE0                                         NA  \n",
       "SAMEA3541827                                   0  \n",
       "NODE1                                         NA  \n",
       "SAMEA104233046                                 0  \n",
       "NODE2                                         NA  \n",
       "...                                          ...  \n",
       "NODE630                                       NA  \n",
       "GCA_000269245.1_ASM26924v1_genomic             1  \n",
       "NODE631                                       NA  \n",
       "GCA_000269125.1_ASM26912v1_genomic             1  \n",
       "GCA_000269365.1_ASM26936v1_genomic             1  \n",
       "\n",
       "[1266 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(tree_df_dill, \"rb\") as infile:\n",
    "    tree_df = dill.load(infile)\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "missing-delay",
   "metadata": {},
   "source": [
    "# 1. Filter Outlier Tips"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "indonesian-religious",
   "metadata": {},
   "source": [
    "## Parse Tip Dates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "formal-timeline",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Attempting to parse dates...\n",
      "\tUsing column 'Name' as name. This needs match the taxon names in the tree!!\n",
      "\tUsing column 'Date' as date.\n"
     ]
    }
   ],
   "source": [
    "# Use the utils function to parse the metadata dates\n",
    "dates_raw = treetime.utils.parse_dates(tree_df_path, \n",
    "                                   date_col=DATE_COL, \n",
    "                                   name_col = NAME_COL)\n",
    "\n",
    "# Remove nan elements (internal nodes)\n",
    "dates = {}\n",
    "for k,v in dates_raw.items():\n",
    "    if type(v) == list:\n",
    "        dates[k] = v\n",
    "    elif not pd.isnull(v):\n",
    "        dates[k] = v\n",
    "        \n",
    "# Add Reference\n",
    "dates[\"Reference\"] = REF_DATE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "verbal-carroll",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Estimate Clock Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "substantial-interface",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "0.00\t-TreeAnc: set-up\n",
      "\n",
      "1.26\t-SequenceData: loaded alignment.\n",
      "1.26\t--SequenceData.aln: specified sequence length doesn't match alignment\n",
      "    \t  length. Treating difference as constant sites.\n",
      "\n",
      "1.26\t-SeqData: making compressed alignment...\n",
      "\n",
      "4.17\t-SequenceData: constructed compressed alignment...\n",
      "4.46\t--ClockTree: Setting precision to level 2\n",
      "\n",
      "4.54\t-ClockTree._assign_dates: assigned date contraints to 634 out of 634 tips.\n",
      "\n",
      "4.54\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, marginal\n",
      "\n",
      "4.55\tWARNING: Previous versions of TreeTime (<0.7.0) RECONSTRUCTED sequences of\n",
      "    \ttips at positions with AMBIGUOUS bases. This resulted in unexpected\n",
      "    \tbehavior is some cases and is no longer done by default. If you want to\n",
      "    \treplace those ambiguous sites with their most likely state, rerun with\n",
      "    \t`reconstruct_tip_states=True` or `--reconstruct-tip-states`.\n",
      "4.55\t--TreeAnc._ml_anc_marginal: type of reconstruction: Marginal\n",
      "4.55\t---Attaching sequence profiles to leafs...\n",
      "7.21\t---Postorder: computing likelihoods...\n",
      "8.89\t---Computing root node sequence and total tree likelihood...\n",
      "8.89\t---Preorder: computing marginal profiles...\n",
      "11.35\t---TreeAnc._ml_anc_marginal: ...done\n",
      "11.45\t--TreeAnc.optimize_tree_marginal: initial, LH=-7675704.08, total\n",
      "     \t  branch_length 0.0014\n",
      "11.45\t--TreeAnc.infer_gtr: counting mutations...\n",
      "17.13\t---TreeAnc.infer_gtr: counting mutations...done\n",
      "\n",
      "17.13\t-GTR: with alphabet: ['A', 'C', 'G', 'T', '-']\n",
      "17.13\t--GTR: ambiguous character: N\n",
      "17.13\t---GTR: init with dummy values!\n",
      "\n",
      "17.13\t-GTR: model inference\n",
      "17.13\t---GTR inference iteration 0 change: 0.447213595499958\n",
      "17.13\t---GTR inference iteration 1 change: 0.24175240879077176\n",
      "17.13\t---GTR inference iteration 2 change: 0.0007845699836583344\n",
      "17.13\t---GTR inference iteration 3 change: 6.842151191933392e-05\n",
      "17.13\t---GTR inference iteration 4 change: 6.807244861649466e-05\n",
      "17.13\t---GTR inference iteration 5 change: 3.057879174227687e-05\n",
      "17.13\t--The model allows for gaps which are estimated to occur at a low fraction\n",
      "     \t  of 6.090e-04 this can potentially result in artificats. gap fraction\n",
      "     \t  will be set to 0.0100\n",
      "17.14\t--TreeAnc.infer_gtr: setting overall rate to 1.0...\n",
      "\n",
      "17.14\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, marginal\n",
      "17.14\t--TreeAnc._ml_anc_marginal: type of reconstruction: Marginal\n",
      "17.14\t---Attaching sequence profiles to leafs...\n",
      "17.15\t---Postorder: computing likelihoods...\n",
      "18.58\t---Computing root node sequence and total tree likelihood...\n",
      "18.59\t---Preorder: computing marginal profiles...\n",
      "20.92\t---TreeAnc._ml_anc_marginal: ...done\n",
      "\n",
      "38.31\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, marginal\n",
      "38.31\t--TreeAnc._ml_anc_marginal: type of reconstruction: Marginal\n",
      "38.31\t---Attaching sequence profiles to leafs...\n",
      "38.32\t---Postorder: computing likelihoods...\n",
      "39.76\t---Computing root node sequence and total tree likelihood...\n",
      "39.76\t---Preorder: computing marginal profiles...\n",
      "42.05\t---TreeAnc._ml_anc_marginal: ...done\n",
      "42.18\t--TreeAnc.optimize_tree_marginal: iteration 0, LH=-6599788.54\n",
      "     \t  (1075915.54), delta branch_length=-0.0000, total branch_length 0.0014\n",
      "\n",
      "42.18\t-TreeAnc.prune_short_branches: pruning short branches (max prob at\n",
      "     \t zero)...\n",
      "\n",
      "42.58\tTreeTime.reroot: with method or node: NODE0\n",
      "42.98\t--ClockTree.date2dist: Setting new molecular clock. rate=9.460e-09,\n",
      "     \t  R^2=0.2280\n",
      "42.98\t--TreeTime.reroot: Tree was re-rooted to node NODE0\n",
      "43.54\t--ClockTree.date2dist: Setting new molecular clock. rate=8.819e-09,\n",
      "     \t  R^2=0.2280\n",
      "43.55\t---TreeTime.ClockFilter: marking GCA_001601675.1_ASM160167v1_genomic as\n",
      "     \t   outlier, residual 3.632611 interquartile distances\n",
      "43.55\t---TreeTime.ClockFilter: marking GCA_002127365.1_ASM212736v1_genomic as\n",
      "     \t   outlier, residual 3.060968 interquartile distances\n",
      "43.55\t---TreeTime.ClockFilter: marking GCA_001972405.1_ASM197240v1_genomic as\n",
      "     \t   outlier, residual 3.586080 interquartile distances\n",
      "43.55\t---TreeTime.ClockFilter: marking GCA_001972385.1_ASM197238v1_genomic as\n",
      "     \t   outlier, residual 3.678037 interquartile distances\n",
      "43.55\t---TreeTime.ClockFilter: marking GCA_001972465.1_ASM197246v1_genomic as\n",
      "     \t   outlier, residual 3.081157 interquartile distances\n",
      "43.55\t---TreeTime.ClockFilter: marking GCA_001972395.1_ASM197239v1_genomic as\n",
      "     \t   outlier, residual 3.330505 interquartile distances\n",
      "43.55\t---TreeTime.ClockFilter: marking GCA_002165475.1_ASM216547v1_genomic as\n",
      "     \t   outlier, residual 3.349444 interquartile distances\n",
      "43.55\t---TreeTime.ClockFilter: marking GCA_001972415.1_ASM197241v1_genomic as\n",
      "     \t   outlier, residual 4.192421 interquartile distances\n",
      "43.55\t---TreeTime.ClockFilter: marking GCA_006376505.1_ASM637650v1_genomic as\n",
      "     \t   outlier, residual 3.022217 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking\n",
      "     \t   GCA_000382525.1_Yersinia_pestis_G3768_assembly_genomic as outlier,\n",
      "     \t   residual 3.003447 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_001188695.1_ASM118869v1_genomic as\n",
      "     \t   outlier, residual 3.105333 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_001188935.1_ASM118893v1_genomic as\n",
      "     \t   outlier, residual 3.105333 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_000324925.1_ASM32492v1_genomic as\n",
      "     \t   outlier, residual 3.003447 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_001294815.1_ASM129481v1_genomic as\n",
      "     \t   outlier, residual 3.189823 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking\n",
      "     \t   GCA_000382505.1_Yersinia_pestis_G1670E_assembly_genomic as outlier,\n",
      "     \t   residual 3.195225 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking\n",
      "     \t   GCA_000382545.1_Yersinia_pestis_assembly_genomic as outlier, residual\n",
      "     \t   3.377378 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_001188335.1_DYA_1_genomic as outlier,\n",
      "     \t   residual 3.377378 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_006376545.1_ASM637654v1_genomic as\n",
      "     \t   outlier, residual 3.048982 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_000834275.1_ASM83427v1_genomic as\n",
      "     \t   outlier, residual 4.204628 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_000834885.1_ASM83488v1_genomic as\n",
      "     \t   outlier, residual 4.149859 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_008630375.1_ASM863037v1_genomic as\n",
      "     \t   outlier, residual 5.414677 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_008630375.2_ASM863037v2_genomic as\n",
      "     \t   outlier, residual 5.701158 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_002412305.1_Y.pestis_A-1809_genomic\n",
      "     \t   as outlier, residual 3.058341 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_003086075.1_ASM308607v1_genomic as\n",
      "     \t   outlier, residual 4.434421 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_003086115.1_ASM308611v1_genomic as\n",
      "     \t   outlier, residual 3.800364 interquartile distances\n",
      "43.56\t---TreeTime.ClockFilter: marking GCA_001613865.1_ASM161386v1_genomic as\n",
      "     \t   outlier, residual 4.600769 interquartile distances\n",
      "\n",
      "43.72\tTreeTime.reroot: with method or node: NODE0\n",
      "44.15\t--ClockTree.date2dist: Setting new molecular clock. rate=8.742e-09,\n",
      "     \t  R^2=0.2837\n",
      "44.15\t--TreeTime.reroot: Tree was re-rooted to node NODE0\n",
      "44.75\t--ClockTree.date2dist: Setting new molecular clock. rate=8.742e-09,\n",
      "     \t  R^2=0.2837\n",
      "\n",
      "44.75\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, marginal\n",
      "44.75\t--TreeAnc._ml_anc_marginal: type of reconstruction: Marginal\n",
      "44.75\t---Attaching sequence profiles to leafs...\n",
      "44.77\t---Postorder: computing likelihoods...\n",
      "45.76\t---Computing root node sequence and total tree likelihood...\n",
      "45.76\t---Preorder: computing marginal profiles...\n",
      "47.45\t---TreeAnc._ml_anc_marginal: ...done\n",
      "47.54\t--TreeAnc.optimize_tree_marginal: initial, LH=-6601245.25, total\n",
      "     \t  branch_length 0.0014\n",
      "\n",
      "61.37\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, marginal\n",
      "61.37\t--TreeAnc._ml_anc_marginal: type of reconstruction: Marginal\n",
      "61.37\t---Attaching sequence profiles to leafs...\n",
      "61.38\t---Postorder: computing likelihoods...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "62.38\t---Computing root node sequence and total tree likelihood...\n",
      "62.38\t---Preorder: computing marginal profiles...\n",
      "64.08\t---TreeAnc._ml_anc_marginal: ...done\n",
      "64.17\t--TreeAnc.optimize_tree_marginal: iteration 0, LH=-6602064.25 (-819.01),\n",
      "     \t  delta branch_length=-0.0000, total branch_length 0.0014\n",
      "\n",
      "64.17\t-TreeAnc.optimize_tree_marginal: deltaLH=-819.006407, stopping iteration.\n",
      "\n",
      "64.18\t###TreeTime.run: INITIAL ROUND\n",
      "\n",
      "64.18\t-ClockTree: Maximum likelihood tree optimization with temporal constraints\n",
      "64.18\t--ClockTree.init_date_constraints...\n",
      "64.18\t---ClockTree.init_date_constraints: Initializing branch length\n",
      "     \t   interpolation objects...\n",
      "276.94\t--ClockTree.date2dist: Setting new molecular clock. rate=8.742e-09,\n",
      "      \t  R^2=0.2837\n",
      "276.94\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.94\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.94\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.94\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.94\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.95\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.95\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.95\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.95\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.95\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.95\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.95\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.95\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.96\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.96\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.96\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.96\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.96\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.97\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.97\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "276.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "277.01\t--ClockTree - Joint reconstruction:  Propagating leaves -> root...\n",
      "325.20\t--ClockTree - Joint reconstruction:  Propagating root -> leaves...\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.36\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.38\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.38\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.38\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.39\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.39\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.39\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.39\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.39\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.39\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.39\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.39\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "325.39\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "\n",
      "325.41\tTreeTime.reroot: with method or node: NODE0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "325.84\t--ClockTree.date2dist: Setting new molecular clock. rate=1.055e-08,\n",
      "      \t  R^2=0.2837\n",
      "325.84\t--TreeTime.reroot: Tree was re-rooted to node NODE0\n",
      "326.47\t--ClockTree.date2dist: Setting new molecular clock. rate=1.055e-08,\n",
      "      \t  R^2=0.2837\n",
      "\n",
      "326.48\t###TreeTime.run: rerunning timetree after rerooting\n",
      "\n",
      "326.48\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "      \t constraints\n",
      "326.48\t--ClockTree.init_date_constraints...\n",
      "\n",
      "326.48\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, marginal\n",
      "326.48\t--TreeAnc._ml_anc_marginal: type of reconstruction: Marginal\n",
      "326.48\t---Attaching sequence profiles to leafs...\n",
      "326.50\t---Postorder: computing likelihoods...\n",
      "327.66\t---Computing root node sequence and total tree likelihood...\n",
      "327.66\t---Preorder: computing marginal profiles...\n",
      "329.53\t---TreeAnc._ml_anc_marginal: ...done\n",
      "329.65\t---ClockTree.init_date_constraints: Initializing branch length\n",
      "      \t   interpolation objects...\n",
      "531.97\t--ClockTree.date2dist: Setting new molecular clock. rate=1.055e-08,\n",
      "      \t  R^2=0.2837\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.98\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.99\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.99\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.99\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.99\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.99\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.99\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "531.99\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "532.01\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "532.01\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "532.01\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "532.01\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "532.01\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "532.02\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "532.02\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "532.02\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "532.04\t--ClockTree - Joint reconstruction:  Propagating leaves -> root...\n",
      "579.05\t--ClockTree - Joint reconstruction:  Propagating root -> leaves...\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.28\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.30\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.30\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.30\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.30\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.30\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.31\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.31\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.31\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.31\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.31\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.31\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "579.31\t----ClockTree.convert_dates -- WARNING: node which is marked as \"BAD\"\n",
      "      \t    optimized later than present day\n",
      "\n",
      "579.33\t###TreeTime.run: ITERATION 1 out of 3 iterations\n",
      "\n",
      "579.33\t-TreeTime.run: adding coalescent prior with Tc=const\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "589.16\t--optimized Tc to 0.000224\n",
      "relaxed_clock {'slack': 0.1, 'coupling': 0}\n",
      "589.95\t--TreeTime.relaxed_clock: slack=0.100000, coupling=0.000000\n",
      "\n",
      "589.99\t-TreeTime.resolve_polytomies: resolving multiple mergers...\n",
      "590.21\t---TreeTime._poly.merge_nodes: creating new node as child of NODE107\n",
      "590.22\t---TreeTime._poly.merge_nodes: Delta-LH = 0.081\n",
      "591.05\t---TreeTime._poly.merge_nodes: creating new node as child of NODE216\n",
      "591.05\t---TreeTime._poly.merge_nodes: Delta-LH = 0.08\n",
      "591.26\t---TreeTime._poly.merge_nodes: creating new node as child of NODE241\n",
      "591.26\t---TreeTime._poly.merge_nodes: Delta-LH = 0.086\n",
      "591.46\t---TreeTime._poly.merge_nodes: creating new node as child of NODE241\n",
      "591.46\t---TreeTime._poly.merge_nodes: Delta-LH = 0.065\n",
      "591.93\t---TreeTime._poly.merge_nodes: creating new node as child of NODE469\n",
      "591.93\t---TreeTime._poly.merge_nodes: Delta-LH = 0.007\n",
      "592.13\t---TreeTime._poly.merge_nodes: creating new node as child of NODE497\n",
      "592.13\t---TreeTime._poly.merge_nodes: Delta-LH = 0.11\n",
      "592.42\t---TreeTime._poly.merge_nodes: creating new node as child of NODE537\n",
      "592.42\t---TreeTime._poly.merge_nodes: Delta-LH = 0.171\n",
      "592.58\t---TreeTime.resolve_polytomies: introduces 7 new nodes\n",
      "\n",
      "592.72\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, marginal\n",
      "592.72\t--TreeAnc._ml_anc_marginal: type of reconstruction: Marginal\n",
      "592.72\t---Attaching sequence profiles to leafs...\n",
      "592.74\t---Postorder: computing likelihoods...\n",
      "593.69\t---Computing root node sequence and total tree likelihood...\n",
      "593.70\t---Preorder: computing marginal profiles...\n",
      "595.31\t---TreeAnc._ml_anc_marginal: ...done\n",
      "595.41\t--TreeAnc.optimize_tree_marginal: initial, LH=-6598306.65, total\n",
      "      \t  branch_length 0.0020\n",
      "\n",
      "595.41\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "      \t constraints\n",
      "595.41\t--ClockTree.init_date_constraints...\n",
      "595.41\t---ClockTree.init_date_constraints: Initializing branch length\n",
      "      \t   interpolation objects...\n",
      "800.78\t--ClockTree.date2dist: Setting new molecular clock. rate=1.054e-08,\n",
      "      \t  R^2=0.2837\n",
      "800.79\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.79\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.79\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.79\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.79\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.79\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.79\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.79\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.79\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.80\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.80\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.80\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.80\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.80\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.80\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.80\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.80\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.80\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.82\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.82\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.82\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.82\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.82\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.82\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.82\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.83\t----ClockTree.init_date_constraints -- WARNING: Branch is marked as bad,\n",
      "      \t    excluding it from the optimization process. Date constraint will be\n",
      "      \t    ignored!\n",
      "800.85\t--ClockTree - Joint reconstruction:  Propagating leaves -> root...\n"
     ]
    }
   ],
   "source": [
    "# Initialize stdout capture\n",
    "# print(\"Estimating clock model...\")\n",
    "# old_stdout = sys.stdout\n",
    "# new_stdout = io.StringIO()\n",
    "# sys.stdout = new_stdout\n",
    "\n",
    "# branch_length_mode:input --> tMRCA ~20,000 YBP\n",
    "\n",
    "tt = treetime.TreeTime(dates=dates, \n",
    "                       aln=aln_path,                     \n",
    "                       tree=tree, \n",
    "                       verbose=4, \n",
    "                       fill_overhangs=False,\n",
    "                       seq_len=REF_LEN,\n",
    "                      )\n",
    "\n",
    "tt.run(\n",
    "    max_iter=MAX_ITER,    \n",
    "    n_iqd=N_IQD,   \n",
    "    verbose=4,    \n",
    "    infer_gtr=True,  \n",
    "    use_covariation=False,     \n",
    "    relaxed_clock={\"slack\" : 0.1, \"coupling\": 0},\n",
    "    root=tt.tree.root,       \n",
    "    resolve_polytomies=True,     \n",
    "    vary_rate=True,    \n",
    "    # Variable Parameters\n",
    "    branch_length_mode = \"marginal\", # joint or marginal  \n",
    "    Tc=TC,    \n",
    "    time_marginal=\"assign\",\n",
    "    )\n",
    "\n",
    "tt.tree.ladderize(reverse=False)\n",
    "\n",
    "# Save stdout to file\n",
    "# output = new_stdout.getvalue()\n",
    "# out_path = os.path.join(outdir, SCRIPT_NAME + \"_estimate.log\") \n",
    "# with open(out_path, \"w\") as file:\n",
    "#     file.write(output)\n",
    "# # Restore stdout\n",
    "# sys.stdout = old_stdout\n",
    "# print(\"Standard output restored, logging to file disabled.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "limited-stereo",
   "metadata": {},
   "source": [
    "## Quick Checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "christian-collar",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common Ancestor\n",
    "tt.tree.common_ancestor(\"NODE0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "immediate-florida",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RTT Regression\n",
    "tt.plot_root_to_tip(add_internal=True, label=True)\n",
    "\n",
    "# Save\n",
    "out_path = os.path.join(outdir, SCRIPT_NAME + \"_rtt-pre.\" + FMT) \n",
    "plt.savefig(out_path, \n",
    "            dpi=dpi, \n",
    "            bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "differential-raising",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Timetree\n",
    "tt_copy = copy.deepcopy(tt)\n",
    "fig, ax1 = treetime.treetime.plot_vs_years(tt_copy,\n",
    "                         #confidence=(0.05,0.95), \n",
    "                         label_func = lambda x:\"\",\n",
    "                         do_show=False)\n",
    "ax1.set_title(\"Time Tree\")\n",
    "\n",
    "# Save\n",
    "out_path = os.path.join(outdir, SCRIPT_NAME + \"_timetree-pre.\" + FMT) \n",
    "plt.savefig(out_path, \n",
    "            dpi=dpi, \n",
    "            bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "infinite-voluntary",
   "metadata": {},
   "source": [
    "# 2. Filter Outliers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dominican-staff",
   "metadata": {},
   "source": [
    "## Prune 'Bad' Branches from Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aware-priest",
   "metadata": {},
   "outputs": [],
   "source": [
    "tt_pre = copy.deepcopy(tt)\n",
    "tt_prune = copy.deepcopy(tt)\n",
    "\n",
    "# Color branches and set to divtree\n",
    "for n in tt_pre.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "    if n.bad_branch:\n",
    "        n.color = \"red\"\n",
    "for n in tt_prune.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "\n",
    "bad_nodes = [c for c in tt_prune.tree.find_clades() if c.bad_branch]\n",
    "\n",
    "while len(bad_nodes) > 0:\n",
    "    for node in bad_nodes:\n",
    "        if node.is_terminal():\n",
    "            print(\"PRUNING:\", node.name)\n",
    "            tt_prune.tree.prune(node)\n",
    "    bad_nodes = [c for c in tt_prune.tree.find_clades() if c.bad_branch]\n",
    "\n",
    "fig, (ax1,ax2) = plt.subplots(2, sharex=True, dpi=dpi)\n",
    "\n",
    "Phylo.draw(tt_pre.tree, \n",
    "        show_confidence=False, \n",
    "        label_func = lambda x: '',\n",
    "        do_show=False,\n",
    "        axes=ax1,\n",
    "        )  \n",
    "Phylo.draw(tt_prune.tree, \n",
    "        show_confidence=False, \n",
    "        label_func = lambda x: '',\n",
    "        do_show=False,\n",
    "        axes=ax2,\n",
    "        )\n",
    "\n",
    "ax1.set_xlabel(\"\")\n",
    "ax1.set_ylabel(\"\")\n",
    "ax2.set_xlabel(\"Branch Length\")\n",
    "ax2.set_ylabel(\"\")\n",
    "\n",
    "# Save\n",
    "out_path = os.path.join(outdir, SCRIPT_NAME + \"_divtree-pre.\" + FMT) \n",
    "plt.savefig(out_path, \n",
    "            dpi=dpi, \n",
    "            bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "meaningful-schedule",
   "metadata": {},
   "source": [
    "## Update tt variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eleven-complex",
   "metadata": {},
   "outputs": [],
   "source": [
    "tt = tt_prune"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "conscious-formula",
   "metadata": {},
   "source": [
    "## Timetree After Pruning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "working-grade",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Timetree\n",
    "fig, ax1 = treetime.treetime.plot_vs_years(tt,\n",
    "                         #confidence=(0.05,0.95), \n",
    "                         label_func = lambda x:\"\",\n",
    "                         do_show=False)\n",
    "ax1.set_title(\"Time Tree\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "neutral-reception",
   "metadata": {},
   "source": [
    "## Remove Collapsed Nodes from Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "otherwise-proposition",
   "metadata": {},
   "outputs": [],
   "source": [
    "tt_nodes = [c.name for c in tt.tree.find_clades()]\n",
    "for rec in tree_df.iterrows():\n",
    "    node = rec[0]\n",
    "    if node not in tt_nodes:\n",
    "        tree_df.drop(node, inplace=True)\n",
    "        \n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "shaped-monitor",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Add Clock Model to Dataframe\n",
    "\n",
    "- Rates\n",
    "- Dates\n",
    "- RTT Regression\n",
    "- Skyline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "departmental-things",
   "metadata": {},
   "source": [
    "## Rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "wrapped-bunch",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_df[\"timetree_rate\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_rate_fold_change\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_mutation_length\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "\n",
    "# The mean rate is the slope\n",
    "mean_rate = tt.clock_model[\"slope\"]\n",
    "\n",
    "for c in tt.tree.find_clades():\n",
    "    tree_df.at[c.name, \"timetree_mutation_length\"] = c.mutation_length\n",
    "    \n",
    "    # Relaxed Clock\n",
    "    if hasattr(c, \"branch_length_interpolator\") and c.branch_length_interpolator:\n",
    "        g = c.branch_length_interpolator.gamma\n",
    "        tree_df.at[c.name, \"timetree_rate_fold_change\"] = g\n",
    "        tree_df.at[c.name, \"timetree_rate\"] = mean_rate * g\n",
    "        \n",
    "    # Strict Clock\n",
    "    else:\n",
    "        tree_df.at[c.name, \"timetree_rate_fold_change\"] = 1\n",
    "        tree_df.at[c.name, \"timetree_rate\"] = mean_rate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "underlying-local",
   "metadata": {},
   "source": [
    "## Dates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "insured-reference",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create new columns\n",
    "tree_df[\"timetree_raw_date\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_num_date\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "\n",
    "if hasattr(tt.tree.root, \"marginal_inverse_cdf\"):\n",
    "    tree_df[\"timetree_num_date_confidence\"] = [[NO_DATA_CHAR,NO_DATA_CHAR] for row in range(0,len(tree_df))]\n",
    "    \n",
    "HAS_MARGINAL = False\n",
    "# clock_length is the same as branch_length until running branch_length_to_years()\n",
    "tree_df[\"timetree_clock_length\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "\n",
    "# Make a copy to change branch_length\n",
    "tt_copy = copy.deepcopy(tt)\n",
    "tt_copy.branch_length_to_years()\n",
    "\n",
    "for c in tt_copy.tree.find_clades():\n",
    "    # Marginal Probability\n",
    "    if hasattr(c, \"marginal_inverse_cdf\"):   \n",
    "        # Set the boolean flag on the first time encountered\n",
    "        if not HAS_MARGINAL:\n",
    "            HAS_MARGINAL = True\n",
    "        # Retrieve the region containing the confidence interval\n",
    "        conf = tt.get_max_posterior_region(c, fraction=CONFIDENCE)\n",
    "        \n",
    "        # Set as lower and upper bounds on date\n",
    "        tree_df.at[c.name, \"timetree_num_date_confidence\"][0] = conf[0]\n",
    "        tree_df.at[c.name, \"timetree_num_date_confidence\"][1] = conf[1]\n",
    "        \n",
    "    tree_df.at[c.name, \"timetree_raw_date\"] = c.date  \n",
    "    tree_df.at[c.name, \"timetree_num_date\"] = c.numdate\n",
    "    tree_df.at[c.name, \"timetree_clock_length\"] = c.branch_length"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "right-teddy",
   "metadata": {},
   "source": [
    "## RTT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "broad-participant",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a copy of the tree\n",
    "tt_copy = copy.deepcopy(tt)\n",
    "tt_copy.branch_length_to_years()\n",
    "\n",
    "# Plotting the tree\n",
    "tree_df[\"timetree_coord_x\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_coord_y\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))] \n",
    "\n",
    "# Plotting the regression\n",
    "tree_df[\"timetree_reg_x\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_reg_y\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_reg_bad\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "\n",
    "x_posns = get_x_positions(tt_copy.tree)\n",
    "y_posns = get_y_positions(tt_copy.tree)\n",
    "tt_reg = tt_copy.setup_TreeRegression()\n",
    "\n",
    "# Add x and y coordinates\n",
    "for c in tt_copy.tree.find_clades():     \n",
    "        \n",
    "    # Tree Node Coordinates\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    tree_df.at[c.name, 'timetree_coord_x'] = coord_x\n",
    "    tree_df.at[c.name, 'timetree_coord_y'] = coord_y\n",
    "    \n",
    "    # Regression Node Coordinates\n",
    "    reg_y = c._v\n",
    "    if c.is_terminal():\n",
    "        reg_x = tt_reg.tip_value(c)\n",
    "    else:\n",
    "        reg_x = c.numdate\n",
    "    reg_bad = c.bad_branch  if hasattr(c, 'bad_branch') else False\n",
    "    tree_df.at[c.name, 'timetree_reg_x'] = reg_x\n",
    "    tree_df.at[c.name, 'timetree_reg_y'] = reg_y    \n",
    "    tree_df.at[c.name, 'timetree_reg_bad'] = reg_bad\n",
    "\n",
    "# Fix up new values that could be none\n",
    "tree_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "departmental-allen",
   "metadata": {},
   "source": [
    "## Skyline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "overhead-extent",
   "metadata": {},
   "outputs": [],
   "source": [
    "if hasattr(tt, \"merger_model\"):\n",
    "    #skyline, conf = tt.merger_model.skyline_inferred(gen=50, confidence=N_STD)\n",
    "    #print(skyline)\n",
    "    #print(conf)\n",
    "    out_path_skyline_pdf = os.path.join(outdir, SCRIPT_NAME + \"_skyline.\" + FMT )\n",
    "    out_path_skyline_txt = os.path.join(outdir, SCRIPT_NAME + \"_skyline.tsv\" )\n",
    "    treetime.wrappers.print_save_plot_skyline(tt, \n",
    "                                              plot=out_path_skyline_pdf, \n",
    "                                              save=out_path_skyline_txt,\n",
    "                                              screen=True,\n",
    "                                              n_std=2.0, )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "middle-afternoon",
   "metadata": {},
   "source": [
    "---\n",
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "thousand-austin",
   "metadata": {},
   "source": [
    "## Tree and Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "buried-tackle",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save tree dataframe with clock info\n",
    "out_path_df = os.path.join(outdir, SCRIPT_NAME + \".tsv\" )\n",
    "out_path_pickle_df = os.path.join(outdir,  SCRIPT_NAME + \".df.obj\" )\n",
    "\n",
    "tree_df.to_csv(out_path_df, sep=\"\\t\")\n",
    "with open(out_path_pickle_df,\"wb\") as outfile:\n",
    "    dill.dump(tree_df, outfile)\n",
    "    \n",
    "# Save timetree trees\n",
    "out_path_xml = os.path.join(outdir, SCRIPT_NAME + \".xml\" )\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \".nwk\" )\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \".nexus\" )\n",
    "out_path_pickle_tree = os.path.join(outdir,  SCRIPT_NAME + \".treetime.obj\" )\n",
    "\n",
    "\n",
    "Phylo.write(tt.tree, out_path_xml, 'phyloxml')\n",
    "Phylo.write(tt.tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "Phylo.write(tt.tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "with open(out_path_pickle_tree,\"wb\") as outfile:\n",
    "    dill.dump(tt, outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "competent-sugar",
   "metadata": {},
   "source": [
    "## Augur JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "structural-amount",
   "metadata": {},
   "outputs": [],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=tree_path, \n",
    "    aln_path=aln_path, \n",
    "    tree=tt.tree,     \n",
    "    #tree=tree_div, \n",
    "    tree_df=tree_df, \n",
    "    color_keyword_exclude=[\"color\", \"coord\", \"reg\", \"lat\", \"lon\"],\n",
    "    type_convert = {\n",
    "        \"Branch_Number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])\n",
    "\n",
    "if hasattr(tt, \"merger_model\"):\n",
    "    skyline, conf = tt.merger_model.skyline_inferred(gen=50, confidence=2)\n",
    "    augur_dict['skyline'] = [[float(x) for x in skyline.x], [float(y) for y in conf[0]],\n",
    "                           [float(y) for y in skyline.y], [float(y) for y in conf[1]]]\n",
    "\n",
    "out_path_augur_json = os.path.join(outdir, SCRIPT_NAME + \"_augur.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "experimental-ideal",
   "metadata": {},
   "source": [
    "## Auspice JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "recent-silicon",
   "metadata": {},
   "outputs": [],
   "source": [
    "auspice_dict = auspice_export(\n",
    "    tree=tt.tree, \n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=auspice_colors_path,\n",
    "    auspice_latlons_path=auspice_latlon_path, \n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(outdir, SCRIPT_NAME + \"_auspice.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "out_path_auspice_remote_json = os.path.join(auspice_remote_dir_path, AUSPICE_PREFIX + SCRIPT_NAME.replace(\"_\",\"-\") + \".json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "print(\"Validation successful for remote JSON.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
