{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "august-width",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "1. Reduce tree with treemmer.\n",
    "1. Add treemmer info to dataframe.\n",
    "1. Plot reduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eight-evidence",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "activated-culture",
   "metadata": {},
   "source": [
    "## Module Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "appropriate-management",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from Bio import Phylo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nutritional-overall",
   "metadata": {},
   "source": [
    "## Input Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "adequate-permit",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEPRECATED: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "tree_path = \"../../docs/results/latest/parse_tree/parse_tree.nwk\"\n",
    "aln_path = \"../../docs/results/latest/snippy_multi/snippy-core_chromosome.snps.filter5.aln\"\n",
    "tree_df_path = \"../../docs/results/latest/parse_tree/parse_tree.tsv\"\n",
    "treemmer_path = \"../scripts/Treemmer.py\"\n",
    "\n",
    "auspice_config_path = \"../../config/auspice_config.json\" \n",
    "auspice_colors_path=\"../../docs/results/latest/parse_tree/parse_tree_colors.tsv\"\n",
    "auspice_latlons_path=\"../../docs/results/latest/parse_tree/parse_tree_latlon.tsv\"\n",
    "auspice_remote_dir_path = \"../../auspice/\"\n",
    "\n",
    "outdir = \"../../docs/results/latest/treemmer\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "racial-grade",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "wanted-hartford",
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "\n",
    "TREEMMER_SIZES = [200]\n",
    "TREEMMER_SIZES = [str(n) for n in TREEMMER_SIZES]\n",
    "\n",
    "SCRIPT_NAME = \"treemmer\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "requested-right",
   "metadata": {},
   "source": [
    "## Import Tree and Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "alert-programming",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_pre = Phylo.read(tree_path, \"newick\")\n",
    "tree_pre.ladderize(reverse=False)\n",
    "\n",
    "tree_df = pd.read_csv(tree_df_path, sep='\\t')\n",
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "tree_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "tree_df.set_index(\"Name\", inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "organic-final",
   "metadata": {},
   "source": [
    "# 1. Reduce tree with treemmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "opening-discount",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "living-reasoning",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# Remove old treemmer results\\nold_treemmer_files = os.popen(\"ls {}*\".format(os.path.join(outdir,\"treemmer\"))).read().strip().split(\"\\n\")\\nfor file in old_treemmer_files:\\n    if not file: continue\\n    print(\"Deleting old treemmer file:\", file)\\n    os.remove(file)\\n    \\n# Copy over the tree to the new directory\\nout_path_tree = os.path.join(outdir, \"treemmer.nwk\")\\nos.system(\"cp {old_tree} {new_tree}\".format(old_tree = tree_path, new_tree = out_path_tree))\\n\\n#x = os.popen(\"ls \" + outdir).read().strip().split(\"\\n\")\\n# Run treemmer\\n# os.popen(\"python3 {treemmer} {tree}\".format(treemmer=treemmer_path, tree=out_path_tree)).read()\\nos.system(\"python3 {treemmer} {tree} --stop_at_X_leaves {stop} --verbose 0 --plot_complete\".format(treemmer=treemmer_path, \\n                                                                                                    tree=out_path_tree,\\n                                                                                                    stop=\" \".join(TREEMMER_SIZES)\\n                                                                                                   )\\n          )'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"# Remove old treemmer results\n",
    "old_treemmer_files = os.popen(\"ls {}*\".format(os.path.join(outdir,\"treemmer\"))).read().strip().split(\"\\n\")\n",
    "for file in old_treemmer_files:\n",
    "    if not file: continue\n",
    "    print(\"Deleting old treemmer file:\", file)\n",
    "    os.remove(file)\n",
    "    \n",
    "# Copy over the tree to the new directory\n",
    "out_path_tree = os.path.join(outdir, \"treemmer.nwk\")\n",
    "os.system(\"cp {old_tree} {new_tree}\".format(old_tree = tree_path, new_tree = out_path_tree))\n",
    "\n",
    "#x = os.popen(\"ls \" + outdir).read().strip().split(\"\\n\")\n",
    "# Run treemmer\n",
    "# os.popen(\"python3 {treemmer} {tree}\".format(treemmer=treemmer_path, tree=out_path_tree)).read()\n",
    "os.system(\"python3 {treemmer} {tree} --stop_at_X_leaves {stop} --verbose 0 --plot_complete\".format(treemmer=treemmer_path, \n",
    "                                                                                                    tree=out_path_tree,\n",
    "                                                                                                    stop=\" \".join(TREEMMER_SIZES)\n",
    "                                                                                                   )\n",
    "          )\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mineral-lounge",
   "metadata": {},
   "source": [
    "# 2. Edit Tree and Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "armed-petersburg",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAECCAYAAADw0Rw8AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAZ+0lEQVR4nO3df4wjZ33H8c83CU24XsEKS1hatBi4opaWnotW6eUHsPwqx4YU1F5VlQQVCboNNGpMiFD4KQsQuRYquVUk4Aj0BOEEVJVOilgifgQ3ou0puoCv0FLaDXUqQNdgErekyeXnt3/Ys5n1+sfYnhnPjN+vfy67tneeudx+nme+8zzPmLsLAFBcZ827AQCAZBH0AFBwBD0AFBxBDwAFR9ADQMER9ABQcAQ9ABTcOfNuwCBLS0teLpfn3QwAyI0777yz7e5PH/RaJoO+XC7r5MmTkd9fr9fV6XTGvm9tbU1ra2vTNwwAMsrM7h72WiaDflKdTke1Wm3ezRjY4dC5AJi3QgR9q9XKRNC3Wi0dPXp03s1QvV5XpVKhgwEgqSBBXy6XMxH09Xo9E+1otVqqVquSdl5lcHUBLKZCBH1WBOE6b+EOJytXGQDmh6AvoHCHk+ZVRrvd1qFDh7hqADKGoC+4NK8y6vW6Go2GGo1GascMa7fbWlpaokQF9CHoEZt5l66C+xHDOhs6ACyqhQ36qHPvEa8kw3beHQ2QVQsb9FmZew8kiam2kBY46IFF0Ol0tLa2lpsrWMprySDogQjyEpT9Njc3JXGjetER9EAEeS31HT9+PJftRrwIemCBjLoyYbRfXAQ9UGClUmlHuAclnEHiWgNBh5E9BD1QYA888ICazaYqlYo6nc7QkI/TtB0GHURyCHqgwPbs2aNKpaJms7lrz6Os3WAOOggCP34EPVBg5XJZ1Wp14J5HbHi3OAh6oMCCR3IOWjWc1IZ3jMizh6DPoaxdci+CVqs17ybMZNi/GUJ5MRD0OZTXOd15lte/71KppFqtps3NTa2vr29/P5h9k/Zuo3Qs80HQAwUWlGz6H7c5r6vCoFMh7NNViKAPRi2TyPulODCJSqWiWq22PaJmp8/FUoign+YfbV4vxYFpxBnswRRI5Echgh5AssKlnv56f1ZQ/x8u8aA3s9dLqkj6d0kvkHSbu9+W9HEBxFeLD8+5b7fbhGrOJB707n7czL4u6W2STNJ5g95nZhuSNiRpZWUl6WYBCyGuGVrhn8ED4PMnjRH92ZLeKekj7v6/ZnaDpM3+97n7EUlHJGl1ddWTbhcwiWlu+GdBXJMOwuffbrd5Jm/OpFGj/2DvOG83s8ck3ZPCMYFY5XWWSlyd07Dz58ZsPqRRunl30scAMFjSVyKtVkvNZpM97jOOWTdAgSV9JVKv1wfujIlsIegBTK1/Z8z+B5swos8Ggh7ATMJXDf3TOQftpUP4p4+gx0jcbMMk8nrTuugIeozUaDRG3mzLG0aTWEQEPUZqtVoql8u5nEOehGlXmtLBYJ4WNujzugBGSnfnzSDo0TXJStN6va5KpULAY+4WNujzXEvMaweVliT3Wg862SjHaLfb6nQ6O/ZgJ/QxDwsb9IjmvPPOy93e/Uk+gSuYSjjuwdrcxEaWEPQY6cCBA/NuQqYEV4LDOpJgpD9sP5gkccWAYQh6YArD7vG0Wi1VKpXU2yMNnrM+Dp3DYiDogSkMu8dTr9clcR8F2ULQAzHK801+FBdBj7HyNhU16s3jJGfnpInyC8Yh6DFW3kap/Z3SsEAfN3Mmbkl1LONq83QEIOhReMOmW4Z3XUzDqVOntH//fkmEL9JF0GNhpX2lcvXVVxPwmAuCHkjJvn37ppoCmQd0YNlG0KNw+m8eJ7WyN+mbuYQn4kLQo3D6SzJJ1OHr9bqOHTum9fX17e8RzMgqgh6IKDyCb7VauvDCC7dfI+SRZQQ9EFF49s7Bgwd1/fXXE+7IhcSD3syqks6S9E1Jl0m6z93rSR8XSNLy8vJ2yM9z4RVXEogijRH9vZKeI+kVkj4k6dpBbzKzDUkbkrSyspJCs/IrzZWqeduieJC4/r7CfxeVSmX7Z7bbbS0tLc3886cxbBYPHQDCzN3TOZDZg5KeIulad//zUe9dXV31kydPptIujFar1XK1/UGSovxdpD26J9ARMLM73X110GtplG4uk7Qq6bCk90i6L+ljAmnqv0mb5rYKQBSJB727f0nSl5I+DpC0UXvQB+Ee57YKjNYRF2bdABEN2zLh4MGD248OzNsGcFgMBD0wo+XlZTWbTTUajVRvzDLiR1QEPTCjSqWiZrOpo0ePpnozdtp9c+ggFg9BD8yoWq2qVqvFHvIEMuJC0AMxKJVK2tra2lG2mbWMM2jETvhjGgQ9EINBN2GTKOMQ/pgGQQ8kZNgMnHq9rkqlQjgjNQQ9MKVpR+ytVkvVanXs5xmpIy4EPTClYc+iHSdYVJXmnkVYbAQ9JjbP3RqzJLzJWf/fyajROIuqkDaCHhObdiRbNOHtDvq3QQhumFJ6QRYQ9MCUwiPzYaEPZAFBD8RgWOhL3FTF/BH0yJ007hGUSqWpa+nU4JE1BD1GGjQzJI2nTo0K8zRKI5NsN8yIHVlH0GOkQaPTNG7EjrrhO0kITzsyZ1SOIiHokTuThHDUToFROYqMoEduBeUd5qwDoxH0yJW0n8/KvjQoAoIeuRKu3cf5fNZhgn1pguMR+sgjgh4TS2OPligze9Ioy4Q7k3a7rU6nE+mpTtT8kSUEPSaWRsDG3ZFEmXs/KJyp8aMIEg96M3u5pN+Q9KeSviDpNne/LenjAmFx7s8TdcEWo3pkReJB7+63mdkDku6SdEDSeYPeZ2YbkjYkaWVlJelmYcHEscgrCHi2F0bepFW6+W1JN7j7LWZ2g6TN/je4+xFJRyRpdXXVU2oXFsTp06dnDmc2K0NepVG6OVvSWZJ+y8zWJN2T9DGBfgcOHJg56IMbs7M+9LuIKFNlWxqlm8ck1XpffjPp46EYhpVH0thnZ5jwNMugRk/AIQ+YdYNMGjbbJa3aeBDmweg9HOjMxEHeEPRYOFFmzQT1+OC9jUZjx1OjGMUjTwh6LJwoUy2DxwEyekcRnDXvBgBZ1Ww2590EIBaM6IEBxs2Xp3yDPCHosfAG1eyjzplvNBoEPjKPoEeuTLsqddS0zEE1+6g7Y7Zare2gT+NZtoNwdYFxCHrkyrQ3RyftHKIeJ7xl8jTHAdJA0GOh1ev1mRZhBVcY/dMxB2HkjXkh6LHQOp2OyuXy1J8Pr5YNtkc4dOgQgY5MIeiBGDDfHllG0AMjUIpBERD0WHijZvIMm2YZ3HwF8oCgx8IbVXYZNs0y/NDwQZhfjywh6IERouyiOai8c+rUKUki7JEJBD0whXC5Z1B5J9gULdjxclLU/xEngh6YQnikHy7vDHr6FKGNeYsU9GZ2rqQLJJkkuft/JdkoIG79I/A49Yd+pVIh2JEpY4PezK5X9+HevyLpLkkPSXplwu0CYhUO4yS3KYgyn54pm0hblBH977j7xWbWcPc1M/tC4q0CCiYc7uHyDsGONEQJ+od7fz5gZi+R9IIE2wMUQv+oPbwXDuGOtEUJ+mt6Nfp3SHqrpOuSbRKQf/1bHwf/3el01Gw2B87GoQNAUqIE/dnu/pCk70n6MzN7xbgPmNmapKskvV/SFZLuk/RZdTuJB939A9M2GMij8C6X5XJZpVKJm7ZITZRnxr7dzA6b2TPN7CZJrx33AXdvSGpKukzShySdK2lN0k2Sfmxm50/bYCCPqtWqarWaKpWKWq2Wms0mIY/UjB3Ru/sbzezjku6WdI27f2zKY5kkH/qi2YakDUlaWVmZ8hBAtgWBv7m5OZeHlFAeWkxRpld+TtIPJD1X0mEz+yV3f++Yz+yXdImkr0l6j7qlm2+oW+d/0N3v7f+Mux+RdESSVldXh3YIwLxFeWTgqLn6pVJp4PcJYSQlSo3+I+7e7P33lWb2unEfcPdTki4f8NK7J2gbkEmDnjHbb9Tr1WpVW1tbu74/y5YJoKMcJUrQn+rdgP1F9VbGApjNvn37uBmL1ES5GftFdW+kvkvS8yW9KskGAUXUP1KvVquEPFITZUT/dHf/fTN7sbu/18xuSbxVQMEcPnx4R9hTZkCaogT9I70FUz8ys/dLembCbQIK58yZM9t1+yhbGNMRIE6Rdq9094d60x9fLenFyTYJyL/+xxP+8Ic/3J6ts7m5qfX19ZGf7+8ICH7MYmjQm9nLJL1c0j4z+0Do/T+XRsOAPOvfxfLmm29Ws9nU0aNHVSqVIu1yCcRl1Ij+B5IeV3f+/Nd733tE0uGkGwUkKTza3tzcnPjz0+xnf+mll+r06dNqNBqEPFI3NOjd/W51V8P+fXrNAZIXDtpGozHxCtVpVrSWy2Xt3buXkgzmgkcJAik4ceKEDhw4IIlwR/oIeiAFy8vLc9nbBpAIeiAVlUplV9AzskdaCHogYcM2QYtzb5vw4wnD6EwgEfRA4qJsgjarer3O3jkYiqAH5mDcVseTjsSZsolRCHpgQv2rXscJz7sPAj54WHgYo3IkhaAHJjTp6DncKQRlnHq9vquz2NzcVKfTYZ49YkfQA3MwqLNot9s7vibkEReCHkhYuNQTlHH6a/Rra2u68cYb028cFgJBDyQsPHoPAn/SmTjhjoGRPiZF0AMZMWomTv/N2+C9hD6iIOiBFAVlnEE7YI4a5fffvG21WqpUKts3bgl7jELQAykKyjiTLqDqv3lbr9e1tbWlpaWl7RW2jO4xTCJBb2Zrkq6S9AlJF0v6maTbJW1I+oq7H0/iuMCiYIEUJpFI0Lt7w8wOuPs3zOx2Se+X9LCkM5L2JnFMIE8GLbqK+kCTYbV8RvQYJo3SzTslfdLdfyjpWjO7YdCbes+k3ZCklZWVFJoFzM+gEfmo+nw42NvtNlMxMZGkSjf7JV1iZteo+yjCi8zs3yS9Vt3HEe7i7kckHZGk1dVVT6JdQB7136QdtKo2qmG7XMaFq4psSqp0c0rS5QNe+k4SxwMWySz1+XGbqc2qf+vlUqk08/HoPGbHrBsgI4ZtltZfu59l8zNu4i4mgh7IiGEh3B/+t956647NzxjxYpyz5t0AAJM5ffq0SqXS9tdxPqkKxcSIHgutXC7Puwlj9Zd0fvKTn2wvlpIY0WM8gh4LLQ9B31/SaTQaO6ZX1ut1VsZiJIIeyLnww0yo22MQgh7IuWBWTricE9Ttmd4IiaAHcqe/3HT69GnVajXdf//92rdvH8+dxS4EPZAzQbAHlpeXt0s3zWaTufLYhaAHcubMmTM7gj7472q1qnq9rquvvlqHDh1iVI9tBD1QIEHYTzK3nhp88RH0QM701+jb7bZqtdp2YFO6QT9WxgI50x/0+/btk9SdaVOr1Vgli10Y0QM5c+LEie26PCN4RMGIHsiZM2fObO91wz43iIKgB3KmXC5ra2tr+2vCHuNQugFyplKp6NixY1pfX2fGDCIh6IGcqVarOn78+K596md9ehSdRnER9EAOlcvlXcE+7AlVAEEP5FClUlGz2dSb3vQmRuEYi5uxQA5Vq1WVy2UdPnyYG7EYi6AHcqpUKmlra4sRPcZKpHRjZmuSrpL0eUmrkm6TdErSdZIedPcPJHFcYJFUq1UdO3ZsYF2eG6sISyTo3b1hZgck/Z8kk3SepDVJN0l6mZmd7+73JnFsYJG84Q1v0NbWFrtVYqREb8a6+1clfdXMbpB0pyQf9l4z25C0IUkrKytJNgsojGG7VTKiR1hSpZv9ki4xs/epG+73SPqGpHeoW7rZNZp39yOSjkjS6urq0A4BiFvwYO28Yq8bjJNU6eaUpMsHvPTuJI4HzGJU0Oe9EwAk5tEDkoavKm21WgODng4AeULQY6GVSiU1m02Vy+WBs1fq9frA74c7gFFbD1ArRxaYe/bK4aurq37y5Ml5NwMLIgjySbYPCId7q9XS0aNHY2/XOFxVIMzM7nT31UGvMaLHwgtG9ZMI3wAdNupP2rCykvRER8QVBSRG9ICk7mg+bxuCjSoZtdttLS0tbX9N4BcfI3pgjFarNe8mTIxplYiKoAfUDfr+ETKjYBQFQQ+ou797p9PJXfkGiILdKwF193cPRvVs+4uiYUQP6Ik9Y5rN5tDa96yP6ouq/0ZqWihVFRdBD/RUq9WRpZu0SjtpdSj9+jdGC6MTyDeCHggZ9dzVUTNz4ryRy2waxI159EBEo0bam5ubWl9fl8ToF/Mxah49QQ/EYG1tbbvsEe4QCH2khQVTQMLK5fJ2wIdvpgZ1bwIf80TQAzEI5uGXSqVdrxHymDeCHojBuO2Oo6jX66pUKnQKiB1BD8QgPA9/Wrfeeqs6nc7AKY5cFWAWBD0Qk3Hz8Mc5ffr09ufZax5xIuiBjLj//vu3g77dbrN4CbEh6IEYjVpwNc7evXvZVA2JYB49kBH79u3TlVdeOfT1drutQ4cOMZrHQKkvmDKzNUlXSfqopEslXSzpw5I2JH3F3Y+P+jxBj0VULpcn2mYh6uZnlHoWQ+oLpty9YWYH3P2kmf1Y0kOSHpZ0RtLeJI4J5N2gOfhh/XvgsAIXUaVRoz8k6fPufo+ka83shkFvMrMNdUf8WllZSaFZQL70z8Rh8zNElUjQm9l+SZeY2WskXeDu95jZCyW9VtIjgz7j7kckHZG6pZsk2gVkWf+GaYNKNY1GY+b96hn9L56kSjenJF3e+/LLve99R9J3kjgeUATnnHPOjlk3rVZLR48e3fU+VtBiUkyvBDLiWc961o6gP3jw4MCFU5RsMCmeGQtk1PLy8o7tj4FpMaIHMqJcLu/4ulKpqFarqdVqaW1tjVk2mBpBD2REf9AHJZp6va5arTa3h4Yj/wh6IOPCgd/pdBjNY2IEPZARJ06cGLnXzbBZOMA4BD2QEcvLyyODPijhhDG6RxQEPZARQY2+f6FUEOZMq8S0CHogI4Itjjc3N7W+vs5oHbEh6IGMCEbsrVZrYAmHm7GYFkEPZEz/NMtAp9PZ0QEwrx5REfRAxgx7SlV4r/p6va6trS1J0tLS0vbqWcIegxD0QMYMu+kaDv9Op6Mbb7xxx6iesMcwBD2QE+GR/ubm5q7XKd9gGIIeyInwSD8IesIdUbB7JZBDe/bsUalUUqPRYHdLjEXQAzn0wAMPqNlsbod9rVYj8DEUQQ/k0J49e1SpVLZn3khidI+hqNEDOVQul9kSAZER9EAOlctlVsoiMoIeyKFSqaRjx47pjjvumHdTkAPm7vNuwy6rq6t+8uTJeTcDyLQLL7xQ6+vrO77H6H5xmdmd7r466LVERvRm9npJFUmPSzpb0n2SPivpOkkPuvsHkjgusEjOP/98SYQ7xktk1o27H5f0l5IekvQhSedKWpN0k6Qfm9n5SRwXWCRPfvKTJYnplRgrqRH92ZLeKenR8LclDa0TmdmGpA1JWllZSaJZQKG89KUv5WYsIkmkRm9mH1a3E7lL0jPVLd3cLOkd6pZuPjjm8z+RdPcEh1yS1J6utZlTlHPhPLKF88ieuM/l2e7+9EEvZPJm7KTM7OSwmxB5U5Rz4TyyhfPInjTPhZWxAFBwBD0AFFxRgv7IvBsQo6KcC+eRLZxH9qR2LoWo0QMAhivKiB4AMETm97oxs5dIuljSPe7+aTN7s6QLJN0u6RmSnidpS9JPg/dJ+qakK5ShFbkxnMftkl4t6Vx3r6V+Aj2znoe7183soKQ1d79+HucgxfL/4xPqrhW5x90/lv4ZdMVwHrdIerOk8yW91ed4iT/luXxL0mF3P2hmT1N+f9fD5/Eixfy7nocR/UXuflhSMD90yd1vkHSppF92949Ien7f+y5T9lbkznQe7v6t3vufNIe2h810Hmb2DEk/L6mTest3mvXf1avU/f15LPWW7zTreTyqbvj4PEO+Z+JzcfempEbv/WvK6e96+DyS+F3PQ9D3/+PzCP8dNnJFbopmPQ+Z2Z9I+mKcjZrCrOfxEkkrki4xs6fE2bAJzXoeT5L0D+p2XgMXqaRk1vN4tqSPSvpXMyvF2K5p8LveE/fveuZvxvYugy5St8z0GXVHUs/Qzsug/5B0b+9990j6R0l/qAlX5CYphvP4vqS3SPqyu38q7fYHZj0Pd/+r3s+5vjeamYsY/n98RtL71A2X69x9LiP7GM7jNvW2HpFUndd5SFOfS0PSX6v7e/415fd3vaEnzuN/FPPveuaDHgAwmzyUbgAAMyDoAaDgCHoAKDiCHgAKjqBHrpnZmpl9aB4/28waZjbTokMzq/QWyCR6LlhsBD0Kzcyy/m+8IulF824Eii3zWyAAEew3sy+ru9LzkKSnSPobdZeYb/ZW4x6UdJ6kq9z922bWkPRPkl4p6ePu/ikzu0TSX0h6WNLHJf23pBea2S3qbhHwane/v//gvQVTn5L0C5K+5+5vM7OapGdJKktquftbzOx5ko5JOi1pr7pbD2xIepqZvUzSJ6McD5hU1kc7QBTnuftr1N1/Jlj8c4GkP3D3T0v6K3d/qbr7u1wX+twX1V2W/ke9rw9Lep27v0zS3wZvcvfLJW1KesWQ418v6Ybe535mZhf1vv8v7v5KSSu9VafXSbpG0u/22id1t6r9iLtfMcHxgIkwokcRfLv3Z1PdVYiSdCq0yvONZnaFpMe1c9n5d939ETN7PPiGu7d7fz5uZpL03d5LP5JUGnL8X5V02Mxc3ZH6HcHP7/35Y0lPlfQcSf/s7o+Z2Xd3/5gdnxl1PGAiBD2KYH/oz7t6//146PW3SfpNdZeefzL0/V17kpjZ09z9p6Hafvg9NuT435d0s7vfKUm9G7QvHPDZ/1S3NHNS0q/1vv+IuiWnQW0adjxgIgQ9iuARM7tV3Rr876lbKw+7Q919Rm4f83PeJekWM3tIT9Too/iwpCNm9lR1O5g/HvK+j0r6nLp7m9ynbsifkHTUzH5d0t9FPB4wEfa6AVJiZue4+6Nmdra6O19e6u6PzrtdKD5uxgLpea6Z3a7uFcZnCHmkhRE9ABQcI3oAKDiCHgAKjqAHgIIj6AGg4Ah6ACg4gh4ACu7/AfbBpGjQzCi/AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "tree = Phylo.read(tree_path, \"newick\")\n",
    "tree.ladderize(reverse=False)\n",
    "\n",
    "tree_df = pd.read_csv(tree_df_path, sep='\\t')\n",
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "tree_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "tree_df.set_index(\"Name\", inplace=True)\n",
    "\n",
    "treemmer_path = os.path.join(outdir, \"treemmer_trimmed_tree_X_200.nwk\")\n",
    "treemmer_tree = Phylo.read(treemmer_path, \"newick\")\n",
    "treemmer_tips = [t.name for t in treemmer_tree.get_terminals()]\n",
    "\n",
    "for t in tree.get_terminals():\n",
    "    \n",
    "    if t.name not in treemmer_tips:\n",
    "        # Remove tip from tree\n",
    "        tree.prune(t)\n",
    "        \n",
    "        # Remove sample from dataframe\n",
    "        tree_df.drop(t.name, inplace=True)\n",
    "\n",
    "Phylo.draw(tree, \n",
    "        show_confidence=False, \n",
    "        label_func = lambda x: '',\n",
    "        do_show=False,\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "behavioral-visiting",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "affected-voltage",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'size = \"200\"\\ntree_post_path = os.path.join(outdir, \"treemmer_trimmed_tree_X_{}.nwk\".format(size))\\ntree_post = Phylo.read(tree_post_path, \"newick\")\\ntree_post.ladderize(reverse=False)\\n\\n#tree_post_tips_path = \\n\\n# Setup the figure\\nfig = plt.figure(dpi=dpi,\\n                 figsize=(figsize[0] * 1, figsize[1] * 1),\\n                 )\\n\\n# Setup the grid\\ngs = gridspec.GridSpec(2, 2, figure=fig, wspace=0.1)\\nax1 = fig.add_subplot(gs[0, 0])\\nax2 = fig.add_subplot(gs[0, 1])\\n\\n\\n# -----------------------------------------------------\\n# Axis 1: Phylogeny Pre-Filter\\n\\nfor c in tree_pre.get_terminals():\\n    if c.name not in tree_dict[size][\"tips\"]:\\n        c.color = \"red\"\\n\\nPhylo.draw(tree_pre, \\n        show_confidence=False, \\n        label_func = lambda x: \\'\\',\\n        do_show=False,\\n        axes=ax1)\\n\\n# Labels\\nax1.set_xlabel(\"Branch Length\")\\nax1.set_yticks([])\\nax1.set_ylabel(\\'\\')\\nnum_tips = len(tree_pre.get_terminals())\\nax1.set_title(\"Phylogeny Pre Filter (N={})\".format(num_tips))\\n\\n\\nPhylo.draw(tree_post, \\n        show_confidence=False, \\n        label_func = lambda x: \\'\\',\\n        do_show=False,\\n        axes=ax2)\\n\\n# Labels\\nax2.set_xlabel(\"Branch Length\")\\nax2.set_yticks([])\\nax2.set_ylabel(\\'\\')\\nnum_tips = len(tree_post.get_terminals())\\nax2.set_title(\"Phylogeny Post Filter (N={})\".format(num_tips))'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"size = \"200\"\n",
    "tree_post_path = os.path.join(outdir, \"treemmer_trimmed_tree_X_{}.nwk\".format(size))\n",
    "tree_post = Phylo.read(tree_post_path, \"newick\")\n",
    "tree_post.ladderize(reverse=False)\n",
    "\n",
    "#tree_post_tips_path = \n",
    "\n",
    "# Setup the figure\n",
    "fig = plt.figure(dpi=dpi,\n",
    "                 figsize=(figsize[0] * 1, figsize[1] * 1),\n",
    "                 )\n",
    "\n",
    "# Setup the grid\n",
    "gs = gridspec.GridSpec(2, 2, figure=fig, wspace=0.1)\n",
    "ax1 = fig.add_subplot(gs[0, 0])\n",
    "ax2 = fig.add_subplot(gs[0, 1])\n",
    "\n",
    "\n",
    "# -----------------------------------------------------\n",
    "# Axis 1: Phylogeny Pre-Filter\n",
    "\n",
    "for c in tree_pre.get_terminals():\n",
    "    if c.name not in tree_dict[size][\"tips\"]:\n",
    "        c.color = \"red\"\n",
    "\n",
    "Phylo.draw(tree_pre, \n",
    "        show_confidence=False, \n",
    "        label_func = lambda x: '',\n",
    "        do_show=False,\n",
    "        axes=ax1)\n",
    "\n",
    "# Labels\n",
    "ax1.set_xlabel(\"Branch Length\")\n",
    "ax1.set_yticks([])\n",
    "ax1.set_ylabel('')\n",
    "num_tips = len(tree_pre.get_terminals())\n",
    "ax1.set_title(\"Phylogeny Pre Filter (N={})\".format(num_tips))\n",
    "\n",
    "\n",
    "Phylo.draw(tree_post, \n",
    "        show_confidence=False, \n",
    "        label_func = lambda x: '',\n",
    "        do_show=False,\n",
    "        axes=ax2)\n",
    "\n",
    "# Labels\n",
    "ax2.set_xlabel(\"Branch Length\")\n",
    "ax2.set_yticks([])\n",
    "ax2.set_ylabel('')\n",
    "num_tips = len(tree_post.get_terminals())\n",
    "ax2.set_title(\"Phylogeny Post Filter (N={})\".format(num_tips))\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "timely-economy",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. Add treemmer info to dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "objective-poster",
   "metadata": {},
   "source": [
    "---\n",
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "apart-scale",
   "metadata": {},
   "source": [
    "## Save Tree and Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "daily-keeping",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_path_df = os.path.join(outdir, SCRIPT_NAME + \".tsv\" )\n",
    "tree_df.to_csv(out_path_df, sep=\"\\t\")\n",
    "\n",
    "out_path_xml = os.path.join(outdir,  SCRIPT_NAME + \".xml\" )\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \".nwk\" )\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \".nexus\" )\n",
    "\n",
    "Phylo.write(tree, out_path_xml, 'phyloxml')\n",
    "Phylo.write(tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "Phylo.write(tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "suitable-edmonton",
   "metadata": {},
   "source": [
    "## Augur JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "gross-norman",
   "metadata": {},
   "outputs": [],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=out_path_nwk, \n",
    "    aln_path=aln_path, \n",
    "    tree=tree, \n",
    "    tree_df=tree_df, \n",
    "    color_keyword_exclude=[\"color\", \"coord\"],\n",
    "    type_convert = {\n",
    "        \"Branch_Number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "out_path_augur_json = os.path.join(outdir, SCRIPT_NAME + \"_augur.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sized-slave",
   "metadata": {},
   "source": [
    "## Auspice JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "controversial-connectivity",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '../../config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'mugration_biovar', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'mugration_branch_major', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'mugration_branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'mugration_country', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'mugration_province', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'branch_number' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait branch_minor were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: You've specified a color scale for branch_number but none of the values found on the tree had associated colors. Auspice will generate its own color scale for this trait.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '../../docs/results/latest/treemmer/treemmer_auspice.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '../../docs/results/latest/treemmer/treemmer_auspice.json' succeeded.\n",
      "Validation successful for local JSON.\n",
      "Validating produced JSON\n",
      "Validating schema of '../../auspice/plague-phylogeography_treemmer.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '../../auspice/plague-phylogeography_treemmer.json' succeeded.\n",
      "Validation successful for remote JSON.\n"
     ]
    }
   ],
   "source": [
    "auspice_dict = auspice_export(\n",
    "    tree=tree, \n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=auspice_colors_path,\n",
    "    auspice_latlons_path=auspice_latlons_path, \n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(outdir, SCRIPT_NAME + \"_auspice.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "out_path_auspice_remote_json = os.path.join(auspice_remote_dir_path, AUSPICE_PREFIX + SCRIPT_NAME.replace(\"_\",\"-\") + \".json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "print(\"Validation successful for remote JSON.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "comparative-particular",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
