{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "abstract-ribbon",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. SETUP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "diverse-genealogy",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bronze-allowance",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import Phylo, AlignIO\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors, gridspec, lines\n",
    "import scipy\n",
    "import math\n",
    "import numpy as np\n",
    "import geopandas\n",
    "import datetime\n",
    "import shapely\n",
    "import copy\n",
    "from geopy import distance\n",
    "import geopy\n",
    "import cartopy.crs as ccrs\n",
    "from augur import utils, export_v2\n",
    "import subprocess\n",
    "import itertools\n",
    "\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "spare-semester",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "female-personality",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"prune\", \"5\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "    \n",
    "results_dir = os.path.join(project_dir, \"results/\")\n",
    "\n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "PRUNE = WILDCARDS[2]\n",
    "MISSING_DATA = WILDCARDS[3]\n",
    "\n",
    "AUSPICE_PREFIX = \"plague-phylogeography-projects_main_{}\".format(PRUNE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "resistant-silicon",
   "metadata": {},
   "outputs": [],
   "source": [
    "iqtree_dir    = os.path.join(results_dir, \"iqtree/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "lsd_dir       = os.path.join(results_dir, \"lsd/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "mugration_dir = os.path.join(results_dir, \"mugration/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "snippy_dir    = os.path.join(results_dir, \"snippy_multi/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "\n",
    "# ------------------------------------------\n",
    "# Trees\n",
    "divtree_path              = iqtree_dir    + \"filter-taxa/iqtree.treefile\"\n",
    "timetree_path             = lsd_dir       + \"lsd.timetree.nex\"\n",
    "#timetree_path             = lsd_dir       + \"lsd.nex\"\n",
    "mugtree_country_path      = mugration_dir + \"country.nex\".format(MISSING_DATA)\n",
    "mugtree_province_path     = mugration_dir + \"province.nex\".format(MISSING_DATA)\n",
    "mugtree_branch_major_path = mugration_dir + \"branch_major.nex\".format(MISSING_DATA)\n",
    "mugtree_branch_minor_path = mugration_dir + \"branch_minor.nex\".format(MISSING_DATA)\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "constant_sites_path   = results_dir + \"snippy_multi/all/chromosome/full/snippy-multi.constant_sites.txt\"\n",
    "aln_path              = iqtree_dir + \"filter-sites/snippy-multi.snps.aln\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Metadata\n",
    "auspice_config_path          = results_dir + \"config/auspice_config.json\"\n",
    "metadata_path                = iqtree_dir    + \"filter-taxa/metadata.tsv\"\n",
    "mugconf_country_path         = mugration_dir + \"country_confidence.csv\"\n",
    "mugstates_country_path       = mugration_dir + \"country_states.csv\"\n",
    "mugconf_province_path        = mugration_dir + \"province_confidence.csv\"\n",
    "mugstates_province_path      = mugration_dir + \"province_states.csv\"\n",
    "mugconf_branch_major_path    = mugration_dir + \"branch_major_confidence.csv\"\n",
    "mugstates_branch_major_path  = mugration_dir + \"branch_major_states.csv\"\n",
    "mugconf_branch_minor_path    = mugration_dir + \"branch_minor_confidence.csv\"\n",
    "mugstates_branch_minor_path  = mugration_dir + \"branch_minor_states.csv\"\n",
    "\n",
    "\n",
    "if PRUNE == \"prune\":\n",
    "    full_divtree_path        = divtree_path.replace(\"prune\", \"full\")\n",
    "    full_metadata_path       = metadata_path.replace(\"prune\", \"full\")\n",
    "else:\n",
    "    full_divtree_path        = divtree_path\n",
    "    full_metadata_path       = metadata_path\n",
    "    \n",
    "# ------------------------------------------\n",
    "# Output\n",
    "auspice_dir = os.path.join(results_dir, \"auspice/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "if not os.path.exists(auspice_dir):\n",
    "    subprocess.run([\"mkdir\", \"-p\", auspice_dir]) \n",
    "    #os.mkdir(out_dir)\n",
    "auspice_remote_dir = os.path.join(project_dir, \"auspice\")\n",
    "    \n",
    "out_dir = os.path.join(results_dir, \"augur/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "if not os.path.exists(out_dir):\n",
    "    subprocess.run([\"mkdir\", \"-p\", out_dir]) \n",
    "    #os.mkdir(out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "spanish-center",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "pressing-biography",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", None)\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "UNKNOWN_CHAR = \"?\"\n",
    "JSON_INDENT=2\n",
    "CURRENT_YEAR = datetime.datetime.utcnow().year\n",
    "MUG_CONF = 0.95\n",
    "ALPHA = 0.05\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "with open(constant_sites_path) as infile:\n",
    "    data = infile.read().strip().split(\",\")\n",
    "    constant_sites = sum([int(count) for count in data])\n",
    "\n",
    "aln = AlignIO.read(aln_path, \"fasta\")\n",
    "variant_sites = len(aln[0].seq)\n",
    "SEQ_LEN = constant_sites + variant_sites\n",
    "\n",
    "# ------------------------------------------\n",
    "# Geo\n",
    "CRS = \"epsg:3857\"\n",
    "CRS_EPSG = ccrs.epsg('3857')\n",
    "world_polygons = geopandas.read_file(geopandas.datasets.get_path(\"naturalearth_lowres\"))\n",
    "\n",
    "# ------------------------------------------\n",
    "# Plotting\n",
    "SM_FONT = 4\n",
    "MED_FONT = 6\n",
    "LG_FONT = 8\n",
    "D3_PAL = [\"#1f77b4\", \"#ff7f0e\", \"#2ca02c\", \"#d62728\", \"#9467bd\", \"#8c564b\", \"#e377c2\", \"#7f7f7f\", \"#bcbd22\", \"#17becf\" ]\n",
    "\n",
    "# ------------------------------------------\n",
    "BRANCH_LIST = {\n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"], \n",
    "    \"0.ANT4\" : [\"0.ANT4\"], \n",
    "    \"0.PE\": [\"0.PE2\", \"0.PE4m\", \"0.PE4m\", \"0.PE4t\", \"0.PE4a\", \"0.PE5\", \"0.PE7\", \"0.PE8\", \"0.PE10\"],   \n",
    "    \"0.ANT\": [\"0.ANT1\", \"0.ANT2\",\"0.ANT3\",\"0.ANT5\"],     \n",
    "    \"1.PRE\" : [\"1.PRE0\",\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],    \n",
    "    \"1.ANT\": [\"1.ANT1\"], \n",
    "    \"1.IN\": [\"1.IN1\",\"1.IN2\",\"1.IN3\"],   \n",
    "    \"1.ORI\" : [\"1.ORI1\", \"1.ORI2\", \"1.ORI3\"],\n",
    "    \"2.ANT\": [\"2.ANT1\",\"2.ANT2\",\"2.ANT3\" ], \n",
    "    \"2.MED\": [\"2.MED0\", \"2.MED1\",\"2.MED2\",\"2.MED3\" ], \n",
    "    \"3.ANT\": [\"3.ANT1\", \"3.ANT2\" ],     \n",
    "    \"4.ANT\": [\"4.ANT1\" ],           \n",
    "}\n",
    "\n",
    "ANCIENT_BRANCH_LIST = {\n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"], \n",
    "    \"0.ANT4\" : [\"0.ANT4\"],  \n",
    "    \"1.PRE\" : [\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],     \n",
    "}\n",
    "\n",
    "MUG_ATTRIBUTE_LIST = [\n",
    "    \"branch_major\",\n",
    "    \"branch_minor\",\n",
    "    \"country\",\n",
    "    \"province\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nuclear-final",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 2. IMPORT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "charming-depression",
   "metadata": {},
   "source": [
    "## Mugration Confidence Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "representative-norway",
   "metadata": {},
   "outputs": [],
   "source": [
    "states_files = [mugstates_country_path, mugstates_province_path, mugstates_branch_major_path, mugstates_branch_minor_path]\n",
    "conf_files = [mugconf_country_path, mugconf_province_path, mugconf_branch_major_path, mugconf_branch_minor_path]\n",
    "\n",
    "# ----------------------------------------\n",
    "# Convert character states to numbers\n",
    "for mugstates,mugconf in zip(states_files, conf_files):\n",
    "    node_i = 0\n",
    "    conf_basename = os.path.splitext(mugconf)[0]\n",
    "    conf_outname = conf_basename + \"_tidy.csv\"\n",
    "    i = 0\n",
    "    state_dict = {}\n",
    "    with open(mugstates, \"r\") as infile:\n",
    "        read_line = infile.readline()\n",
    "        while read_line:\n",
    "            split_line = read_line.strip().split(\",\")\n",
    "            state = split_line[1]\n",
    "            if state == \"nan\":\n",
    "                state = \"NA\"\n",
    "            if state == UNKNOWN_CHAR:\n",
    "                state_dict[-1] = state\n",
    "            else:\n",
    "                state_dict[i] = state\n",
    "                i += 1\n",
    "            read_line = infile.readline()\n",
    "    # Swap first to last place\n",
    "    final_i = len(state_dict) - 1   \n",
    "    state_dict[final_i] = \"?\"\n",
    "    # Remove old first\n",
    "    state_dict.pop(-1)\n",
    "\n",
    "    with open(mugconf, \"r\") as infile:            \n",
    "        i = 0\n",
    "        read_line = infile.readline()\n",
    "        header = read_line.strip().split(\", \")\n",
    "        header[0] = \"name\"\n",
    "\n",
    "        data = \"\"\n",
    "        for i in range(0, len(header) - 1):\n",
    "            char = header[i + 1]\n",
    "            header[i + 1] = state_dict[i]\n",
    "        header = \",\".join(header)\n",
    "        read_line = infile.readline()            \n",
    "        while read_line:\n",
    "            # Rename internal nodes\n",
    "            split_line = read_line.split(\", \")\n",
    "            node_name = split_line[0]\n",
    "            if \"NODE\" in node_name:\n",
    "                node_name = \"NODE{}\".format(node_i)\n",
    "                node_i += 1\n",
    "            new_line = \",\".join([node_name] + split_line[1:])\n",
    "            data += new_line\n",
    "            read_line = infile.readline()\n",
    "    \n",
    "    with open(conf_outname, \"w\") as outfile:\n",
    "        outfile.write(header + \"\\n\")\n",
    "        outfile.write(data)\n",
    "\n",
    "# ----------------------------------------\n",
    "# Import the tidy files as dataframes\n",
    "mugconf_country_tidy_path = os.path.splitext(mugconf_country_path)[0] + \"_tidy.csv\"\n",
    "mugconf_country_df = pd.read_csv(mugconf_country_tidy_path, sep=',', index_col = 0)\n",
    "\n",
    "mugconf_province_tidy_path = os.path.splitext(mugconf_province_path)[0] + \"_tidy.csv\"\n",
    "mugconf_province_df = pd.read_csv(mugconf_province_tidy_path, sep=',', index_col = 0)\n",
    "\n",
    "mugconf_branch_major_tidy_path = os.path.splitext(mugconf_branch_major_path)[0] + \"_tidy.csv\"\n",
    "mugconf_branch_major_df = pd.read_csv(mugconf_branch_major_tidy_path, sep=',', index_col = 0)\n",
    "\n",
    "mugconf_branch_minor_tidy_path = os.path.splitext(mugconf_branch_minor_path)[0] + \"_tidy.csv\"\n",
    "mugconf_branch_minor_df = pd.read_csv(mugconf_branch_minor_tidy_path, sep=',', index_col = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "orange-aside",
   "metadata": {},
   "source": [
    "## Trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "induced-development",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "divtree: iqtree.treefile\n",
      "timetree: lsd.timetree.nex\n",
      "mugtree_country: country.nex\n",
      "mugtree_province: province.nex\n",
      "mugtree_branch_major: branch_major.nex\n",
      "mugtree_branch_minor: branch_minor.nex\n"
     ]
    }
   ],
   "source": [
    "tree_paths = [divtree_path, timetree_path, mugtree_country_path, mugtree_province_path, mugtree_branch_major_path, mugtree_branch_minor_path  ]\n",
    "\n",
    "for tree_path in tree_paths:\n",
    "    tree_ext = os.path.splitext(tree_path)[1]\n",
    "    if tree_ext == \".nwk\" or tree_ext == \".newick\" or tree_ext == \".treefile\":\n",
    "        tree = Phylo.read(tree_path, \"newick\")\n",
    "    elif tree_ext == \".nex\" or tree_ext == \".nexus\":\n",
    "        parse_trees = Phylo.parse(tree_path, \"nexus\")\n",
    "        for parse_tree in parse_trees:\n",
    "            clades = [c for c in parse_tree.find_clades()]\n",
    "            if len(clades) > 1:\n",
    "                break\n",
    "        tree = parse_tree\n",
    "    tree.ladderize(reverse=False)\n",
    "    \n",
    "    if tree_path == divtree_path:\n",
    "        print(\"divtree:\", os.path.basename(divtree_path))\n",
    "        divtree = tree\n",
    "    elif tree_path == timetree_path:\n",
    "        print(\"timetree:\", os.path.basename(timetree_path))        \n",
    "        timetree = tree    \n",
    "    elif tree_path == mugtree_country_path:\n",
    "        print(\"mugtree_country:\", os.path.basename(mugtree_country_path))       \n",
    "        mugtree_country = tree    \n",
    "    elif tree_path == mugtree_province_path:\n",
    "        print(\"mugtree_province:\",os.path.basename(mugtree_province_path))\n",
    "        mugtree_province = tree  \n",
    "    elif tree_path == mugtree_branch_major_path:\n",
    "        print(\"mugtree_branch_major:\",os.path.basename(mugtree_branch_major_path))        \n",
    "        mugtree_branch_major = tree            \n",
    "    elif tree_path == mugtree_branch_minor_path:\n",
    "        print(\"mugtree_branch_minor:\",os.path.basename(mugtree_branch_minor_path))        \n",
    "        mugtree_branch_minor = tree                \n",
    "    \n",
    "full_divtree = Phylo.read(full_divtree_path, \"newick\")\n",
    "full_divtree.ladderize(reverse=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "interested-piano",
   "metadata": {},
   "source": [
    "## Genomic Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "interested-corrections",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.set_index(metadata_df.columns[0], inplace=True)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "full_metadata_df = pd.read_csv(full_metadata_path, sep='\\t')\n",
    "full_metadata_df.set_index(full_metadata_df.columns[0], inplace=True)\n",
    "full_metadata_df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "entitled-refund",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. MERGE METADATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "brilliant-produce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if they're synchronized\n",
    "node_i = 0\n",
    "\n",
    "metadata_df[\"branch_support_char\"] = [\"\"] * len(metadata_df)\n",
    "\n",
    "for cd,ct,cmc,cmp,cmbmaj, cmbmin in zip(\n",
    "        divtree.find_clades(), \n",
    "        timetree.find_clades(), \n",
    "        mugtree_country.find_clades(),\n",
    "        mugtree_province.find_clades(),\n",
    "        mugtree_branch_major.find_clades(),    \n",
    "        mugtree_branch_minor.find_clades(),      \n",
    "    ):\n",
    "    # Rename internal nodes\n",
    "    if not cd.name or \"NODE\" in cmc.name:\n",
    "        cd.name = \"NODE{}\".format(node_i)\n",
    "        ct.name = \"NODE{}\".format(node_i)\n",
    "        cmc.name = \"NODE{}\".format(node_i) \n",
    "        cmp.name = \"NODE{}\".format(node_i) \n",
    "        cmbmaj.name = \"NODE{}\".format(node_i) \n",
    "        cmbmin.name = \"NODE{}\".format(node_i) \n",
    "        node_i += 1\n",
    "        \n",
    "    if cd.name != ct.name or cd.name != cmc.name or cd.name != cmp.name:\n",
    "        print(\"ERROR. Trees are not synchronized.\")\n",
    "        print(cd.name, ct.name, cmc.name, cmp.name, cmbmaj.name, cmbmin.name)\n",
    "        break\n",
    "        \n",
    "    node_name = cd.name\n",
    "    \n",
    "    # node type\n",
    "    node_type = \"internal\"\n",
    "    if cd.is_terminal():\n",
    "        node_type = \"terminal\"\n",
    "    metadata_df.at[node_name, \"node_type\"] = node_type\n",
    "\n",
    "    \n",
    "    # Branch Length\n",
    "    branch_length = cd.branch_length\n",
    "    if not branch_length:\n",
    "        branch_length = 0\n",
    "    metadata_df.at[node_name, \"branch_length\"] = branch_length\n",
    "    # Branch Support\n",
    "    # Give terminal nodes 100% confidence\n",
    "    branch_support = cd.confidence\n",
    "    if cd.is_terminal():\n",
    "        branch_support = 100\n",
    "        \n",
    "    # Internal nodes without confidence can be 0\n",
    "    elif not cd.confidence:\n",
    "        branch_support = 0\n",
    "    metadata_df.at[node_name, \"branch_support\"] = branch_support\n",
    "    \n",
    "    if branch_support >= 95 and not cd.is_terminal():\n",
    "        branch_support_char = \"*\"\n",
    "        metadata_df.at[node_name, \"branch_support_char\"] = branch_support_char\n",
    "\n",
    "    if branch_support >= 95:\n",
    "        metadata_df.at[node_name, \"branch_support_conf_category\"] = \"HIGH\"\n",
    "    else:\n",
    "        metadata_df.at[node_name, \"branch_support_conf_category\"] = \"LOW\"        \n",
    "    \n",
    "    \n",
    "    comment_dict = {}\n",
    "    # Mugration  Country\n",
    "    if cmc.comment:\n",
    "        cmc_comment_dict = parse_comment(cmc.comment)\n",
    "        for attr in cmc_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmc_comment_dict[attr]\n",
    "            \n",
    "    # Mugration  Province\n",
    "    if cmp.comment:\n",
    "        cmp_comment_dict = parse_comment(cmp.comment)\n",
    "        for attr in cmp_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmp_comment_dict[attr]            \n",
    "            \n",
    "    # Mugration Branch Major\n",
    "    if cmbmaj.comment:\n",
    "        cmbmaj_comment_dict = parse_comment(cmbmaj.comment)\n",
    "        for attr in cmbmaj_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmbmaj_comment_dict[attr]            \n",
    "            \n",
    "    # Mugration Branch Minor\n",
    "    if cmbmin.comment:\n",
    "        cmbmin_comment_dict = parse_comment(cmbmin.comment)\n",
    "        for attr in cmbmin_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmbmin_comment_dict[attr]              \n",
    "\n",
    "    # Time Tree         \n",
    "    if ct.comment:\n",
    "        ct_comment_dict = parse_comment(ct.comment)\n",
    "        for attr in ct_comment_dict:\n",
    "            comment_dict[\"timetree_\" + attr] = ct_comment_dict[attr]\n",
    "            \n",
    "    for attr in comment_dict:\n",
    "        metadata_df.at[node_name, attr] = comment_dict[attr]\n",
    "           \n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "#display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "compound-kuwait",
   "metadata": {},
   "source": [
    "## Add Mugration Confidence and Handle Root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "overall-amber",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "strain                                     NA\n",
       "date                                       NA\n",
       "date_bp                                    NA\n",
       "country                                    NA\n",
       "province                                   NA\n",
       "                                       ...   \n",
       "mugration_province_confidence            0.37\n",
       "mugration_province_lat_confidence    0.369092\n",
       "mugration_province_lon_confidence    0.369092\n",
       "mugration_branch_major_confidence         0.5\n",
       "mugration_branch_minor_confidence        0.29\n",
       "Name: NODE0, Length: 41, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_list = [mugconf_country_df, mugconf_province_df, mugconf_branch_major_df, mugconf_branch_minor_df]\n",
    "attr_list = [\"country\", \"province\", \"branch_major\", \"branch_minor\"]\n",
    "\n",
    "for df,attr in zip(df_list, attr_list):\n",
    "    for sample in df.index:\n",
    "        max_state = \"\"\n",
    "        max_val = 0        \n",
    "        for state,val in zip(df.columns, df.loc[sample]):\n",
    "            if val >= max_val:\n",
    "                max_val = val\n",
    "                max_state = state\n",
    "        metadata_df.at[sample,\"mugration_\" + attr] = max_state      \n",
    "        metadata_df.at[sample,\"mugration_\" + attr + \"_confidence\"] = round(max_val,2)\n",
    "        \n",
    "        # if the attr is country or province, extend this confidence to lat and lon\n",
    "        if attr == \"country\" or attr == \"province\":\n",
    "            metadata_df.at[sample,\"mugration_\" + attr + \"_lat_confidence\"] = max_val\n",
    "            metadata_df.at[sample,\"mugration_\" + attr + \"_lon_confidence\"] = max_val\n",
    "\n",
    "display(metadata_df.loc[\"NODE0\"])\n",
    "#print(list(metadata_df[\"mugration_branch_minor_confidence\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "meaningful-athletics",
   "metadata": {},
   "source": [
    "## Add Mugration P-Value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "extreme-inspector",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'divtree_copy = copy.deepcopy(divtree)\\n\\nfor attr in MUG_ATTRIBUTE_LIST:\\n    #if attr != \"country\" and attr != \"province\": continue\\n    if attr != \"country\": continue\\n    \\n    metadata_df[\"mugration_\" + attr + \"_transition_pvalue\"] = [NO_DATA_CHAR] * len(metadata_df)\\n    metadata_df[\"mugration_\" + attr + \"_transition_conf_category\"] = [NO_DATA_CHAR] * len(metadata_df)\\n    \\n    # There are too many states in the full tree to do permutations\\n    # But mugration is estimated from the full tree\\n    # If a subtree is not monophyletic, it\\'s root could be a state not observed\\n    # Leave the subtree root as NA?\\n        \\n    # Sub Trees\\n    for branch in BRANCH_LIST:\\n        \\n        print(attr, branch)\\n        \\n        # Extract subtree\\n        # Create the subtree df\\n        subtree_df = metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])]\\n        if len(subtree_df) < 2: continue\\n        subtree_tips = subtree_df[subtree_df[\"node_type\"] == \"terminal\"]\\n        subtree_all = metadata_df[metadata_df[\"mugration_branch_minor\"].isin(BRANCH_LIST[branch])]\\n    \\n        subtree = extract_subtree(\\n            tree=divtree_copy, \\n            tips=list(subtree_tips.index), \\n            df=metadata_df, \\n            color_branches=False)  \\n        \\n        states = list(subtree_tips[\"mugration_\" + attr])\\n        states_uniq = set(states)\\n        # Static transitions in place\\n        transitions = [(s,s) for s in states_uniq]\\n        # Add movements\\n        transitions += list(itertools.permutations(states, 2))\\n        transitions_uniq = set(transitions)\\n        transitions_dict = {t: transitions.count(t) / len(transitions) for t in transitions_uniq}\\n\\n        for rec in subtree_all.iterrows():\\n            sample = rec[0]\\n            # Check that this rec is in the subtree\\n            if sample not in [c.name for c in subtree.find_clades()]:\\n                continue\\n                \\n            # Get parent\\n            walk_path = [subtree.root] + subtree.get_path(sample)\\n            # Root\\n            if len(walk_path) == 1:\\n                parent = walk_path[0]\\n            else:\\n                parent = walk_path[-2]\\n                        \\n                \\n            # Get transition based on parent and node\\n            state = metadata_df[\"mugration_\" + attr][sample]\\n            parent_state = metadata_df[\"mugration_\" + attr][parent.name]\\n            transition = (parent_state, state) \\n            \\n            # Check that the parent state is in the dictionary\\n            if parent_state not in states: continue\\n                \\n            pvalue = transitions_dict[transition]\\n            \\n            # Get mugration confidence\\n            conf = metadata_df[\"mugration_\" + attr + \"_confidence\"][sample]\\n            \\n            # If conf >= 0.95 and pvalue < 0.05\\n            conf_category = \"LOW\"\\n            if conf >= MUG_CONF and pvalue < ALPHA:\\n                conf_category = \"HIGH\"                \\n            \\n            # Add to database\\n            metadata_df.at[sample, \"mugration_\" + attr + \"_transition_pvalue\"] = pvalue\\n            metadata_df.at[sample, \"mugration_\" + attr + \"_transition_conf_category\"] = conf_category\\n    \\ndisplay(metadata_df)'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"divtree_copy = copy.deepcopy(divtree)\n",
    "\n",
    "for attr in MUG_ATTRIBUTE_LIST:\n",
    "    #if attr != \"country\" and attr != \"province\": continue\n",
    "    if attr != \"country\": continue\n",
    "    \n",
    "    metadata_df[\"mugration_\" + attr + \"_transition_pvalue\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "    metadata_df[\"mugration_\" + attr + \"_transition_conf_category\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "    \n",
    "    # There are too many states in the full tree to do permutations\n",
    "    # But mugration is estimated from the full tree\n",
    "    # If a subtree is not monophyletic, it's root could be a state not observed\n",
    "    # Leave the subtree root as NA?\n",
    "        \n",
    "    # Sub Trees\n",
    "    for branch in BRANCH_LIST:\n",
    "        \n",
    "        print(attr, branch)\n",
    "        \n",
    "        # Extract subtree\n",
    "        # Create the subtree df\n",
    "        subtree_df = metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])]\n",
    "        if len(subtree_df) < 2: continue\n",
    "        subtree_tips = subtree_df[subtree_df[\"node_type\"] == \"terminal\"]\n",
    "        subtree_all = metadata_df[metadata_df[\"mugration_branch_minor\"].isin(BRANCH_LIST[branch])]\n",
    "    \n",
    "        subtree = extract_subtree(\n",
    "            tree=divtree_copy, \n",
    "            tips=list(subtree_tips.index), \n",
    "            df=metadata_df, \n",
    "            color_branches=False)  \n",
    "        \n",
    "        states = list(subtree_tips[\"mugration_\" + attr])\n",
    "        states_uniq = set(states)\n",
    "        # Static transitions in place\n",
    "        transitions = [(s,s) for s in states_uniq]\n",
    "        # Add movements\n",
    "        transitions += list(itertools.permutations(states, 2))\n",
    "        transitions_uniq = set(transitions)\n",
    "        transitions_dict = {t: transitions.count(t) / len(transitions) for t in transitions_uniq}\n",
    "\n",
    "        for rec in subtree_all.iterrows():\n",
    "            sample = rec[0]\n",
    "            # Check that this rec is in the subtree\n",
    "            if sample not in [c.name for c in subtree.find_clades()]:\n",
    "                continue\n",
    "                \n",
    "            # Get parent\n",
    "            walk_path = [subtree.root] + subtree.get_path(sample)\n",
    "            # Root\n",
    "            if len(walk_path) == 1:\n",
    "                parent = walk_path[0]\n",
    "            else:\n",
    "                parent = walk_path[-2]\n",
    "                        \n",
    "                \n",
    "            # Get transition based on parent and node\n",
    "            state = metadata_df[\"mugration_\" + attr][sample]\n",
    "            parent_state = metadata_df[\"mugration_\" + attr][parent.name]\n",
    "            transition = (parent_state, state) \n",
    "            \n",
    "            # Check that the parent state is in the dictionary\n",
    "            if parent_state not in states: continue\n",
    "                \n",
    "            pvalue = transitions_dict[transition]\n",
    "            \n",
    "            # Get mugration confidence\n",
    "            conf = metadata_df[\"mugration_\" + attr + \"_confidence\"][sample]\n",
    "            \n",
    "            # If conf >= 0.95 and pvalue < 0.05\n",
    "            conf_category = \"LOW\"\n",
    "            if conf >= MUG_CONF and pvalue < ALPHA:\n",
    "                conf_category = \"HIGH\"                \n",
    "            \n",
    "            # Add to database\n",
    "            metadata_df.at[sample, \"mugration_\" + attr + \"_transition_pvalue\"] = pvalue\n",
    "            metadata_df.at[sample, \"mugration_\" + attr + \"_transition_conf_category\"] = conf_category\n",
    "    \n",
    "display(metadata_df)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "administrative-gross",
   "metadata": {},
   "source": [
    "## Latitude and Longtitude"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "mexican-outside",
   "metadata": {},
   "outputs": [],
   "source": [
    "latlon_country_df = pd.DataFrame()\n",
    "latlon_province_df = pd.DataFrame()\n",
    "\n",
    "df_list = [latlon_country_df, latlon_province_df]\n",
    "attr_list = [\"country\", \"province\"]\n",
    "conf_df_list = [mugconf_country_df, mugconf_province_df]\n",
    "\n",
    "# Create a mapping of geo name to lat,lon\n",
    "for df,attr in zip(df_list, attr_list):\n",
    "    # Get data\n",
    "    for rec in metadata_df.iterrows():\n",
    "        node_name = rec[0]\n",
    "        node_type = rec[1][\"node_type\"]\n",
    "        name = rec[1][attr]\n",
    "        country = rec[1][\"country\"]\n",
    "        \n",
    "        if node_type == \"internal\":\n",
    "            continue\n",
    "        \n",
    "        if attr == \"province\" and name == NO_DATA_CHAR and node_type == \"terminal\" and country != \"Russia\":\n",
    "            # Use country instead\n",
    "            name = rec[1][\"country\"]\n",
    "            lat = rec[1][\"country_lat\"]\n",
    "            lon = rec[1][\"country_lon\"]\n",
    "        else:\n",
    "            lat = rec[1][attr + \"_lat\"]\n",
    "            lon = rec[1][attr + \"_lon\"]             \n",
    "\n",
    "\n",
    "        if name not in df.index:\n",
    "            df.at[name, \"lat\"] = lat\n",
    "            df.at[name, \"lon\"] = lon\n",
    "            df.at[name, \"size\"] = 1\n",
    "        else:\n",
    "            df[\"size\"][name] += 1\n",
    "\n",
    "#display(latlon_country_df)\n",
    "#display(latlon_province_df)\n",
    "\n",
    "\n",
    "# Update lat lon for mugration\n",
    "for latlon_df, conf_df, attr in zip(df_list, conf_df_list, attr_list):\n",
    "    for rec in metadata_df.iterrows():\n",
    "        sample = rec[0]\n",
    "        name = rec[1][\"mugration_\" + attr]  \n",
    "        node_type = rec[1][\"node_type\"]  \n",
    "        country = rec[1][\"country\"]  \n",
    "        \n",
    "        if attr == \"province\" and rec[1][attr]  == NO_DATA_CHAR and node_type == \"terminal\" and country != \"Russia\":\n",
    "            # Use country instead\n",
    "            name = rec[1][\"country\"]\n",
    "            metadata_df.at[sample,\"mugration_\" + attr] = name\n",
    "        \n",
    "        lat = latlon_df[\"lat\"][name]\n",
    "        lon = latlon_df[\"lon\"][name]\n",
    "        \n",
    "                \n",
    "        metadata_df.at[sample, \"mugration_\" + attr +\"_lat\"] = lat\n",
    "        metadata_df.at[sample, \"mugration_\" + attr +\"_lon\"] = lon\n",
    "        \n",
    "        \n",
    "# Mapping file for auspice\n",
    "out_path_latlon = os.path.join(out_dir, \"latlon.tsv\")\n",
    "\n",
    "# Countries\n",
    "with open(out_path_latlon, \"w\") as outfile:\n",
    "    for country in latlon_country_df.index:\n",
    "        lat = str(latlon_country_df[\"lat\"][country])\n",
    "        lon = str(latlon_country_df[\"lon\"][country])\n",
    "        outfile.write(\"country\" + \"\\t\" + country + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "# Provinces\n",
    "with open(out_path_latlon, \"a\") as outfile:\n",
    "    for province in latlon_province_df.index:\n",
    "        lat = str(latlon_province_df[\"lat\"][province])\n",
    "        lon = str(latlon_province_df[\"lon\"][province])\n",
    "        outfile.write(\"province\" + \"\\t\" + province + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "        \n",
    "#display(metadata_df[metadata_df[\"continent\"] == \"Africa\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "desirable-miracle",
   "metadata": {},
   "source": [
    "## Convert Dataframe to Geopandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "worse-germany",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<GeoAxesSubplot:>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjwAAAEbCAYAAADeTl6JAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAACYRUlEQVR4nOz9d5Rs6V3fC393rpxT5xxO6hNnRkgIGAmQeDH2i20MCCO4RsjGJGMs7mVhLsj4BRwwy/haBoNtDFzhCxZcca9IRkJG0mg0c0LPCX3O6T6dU3V15Rx27f3+0fM8U9VdVV1VXZ2qn89aZ810hV27qnbt57t/4fvjdF0Hg8FgMBgMRjfDn/YOMBgMBoPBYBw3TPAwGAwGg8HoepjgYTAYDAaD0fUwwcNgMBgMBqPrYYKHwWAwGAxG18MED4PBYDAYjK5HbHTnBz/4QT0cDp/UvjAYDAaDwWC0zb179/5M1/UP1rqvoeAJh8O4e/fu8ewVg8FgMBgMRgfhOM5T7z6W0mIwGAwGg9H1MMHDYDAYDAaj62GCh8FgMBgMRtfDBA+DwWAwGIyuhwkeBoPBYDAYXQ8TPAwGg8FgMLoeJngYDAaDwWB0PUzwMBgMBoPB6HqY4GEwGAwGg9H1NHRaZjAYp4uu69B1HZqm0X8cx9X9B4D+l8FgMBjvwAQP41ygqiry+TyKxSJUVUW5XKZigPxTFAW6rkNVVfoYjuPA8zz9t/9v8k8URQiCAEmSIIrH/7MolUpIp9PIZDIoFosHhEs2m0U0GkWxWGx525XbEUURRqMRBoPhwD9FUZg4YjAYFwYmeBinjqZpKJVKKBaLSCaTSCQS0DQNoijCbrcjkUggGAye2P4YDAZ4vV6Mjo4eEAS6rlOxVYkoivSxpVIJqVQKxWIRJpMJJpOJiqj5+XmEw+G2hEw9BEGAIAhUzO0XT5qmIZfLIZ/PAwAVQna7HR6PBwaDoWP7wmAwGGcVJngYJ87Gxga2t7dRKBRomqYexyl0TCYTzGYzFEWB2WyGJEn07/X1dTx9+hSFQqEqqlRvXyVJgs1mQzabRS6Xq3m/JElVESVVVTv2XgRBgKZpUFWVfqaCIMBgMECWZRgMBho5ymaz8Hg86Ovrg6ZpyGQyKJVKNf/pug5RFGnky+FwwGq1dmy/GQwG46RggodxbBQKBSSTSYRCIaiqCl3XYbVa0dvbC6/Xi1AohK2trZoC4TgRBAEWiwV2ux1OpxOKoiCbzSKdTiMYDKJYLCKbzbYkSEqlEiKRSMP7S6VSJ3b/AOVyueZnqKoq0uk00ul01e08zyOZTOLLX/7yoZEmjuNgNpvpv06KtIuOpmk0WiiKIsrlMrLZLE1D8nz395Touo50Oo1EIoFCoUAvgmw2G5xOJywWC0u7MjoGEzyMjqPrOhYXF7G5uXkg9ROPx7G+vg5JkmAwGGCxWOB0OiGKIiwWC+LxOARBqIqEpFKptutZCIIgoLe3F729vTAYDMjn81hZWcHTp087ml46D2iahmQyWXUbz/M0CmWz2aAoCjweD0wm07lacEiBNwDs7OzQ1GKpVALP8zCZTNA0DYlEAoIgQJZlyLIMo9EIi8UCi8UCRVGOdR81TUOxWEQ+n8fy8jJSqRScTiesViskSYIsy+B5Hkaj8Vj34zTRNA3hcBgbGxsHjkUACIfDAPbSy729vejp6YEkSSe9m4wug9u/IFVy584d/e7duye4O4xuIZlMYnNzE/F4nBYVHxbh8Pv94DgO+XweiUTigFhqRKVIEgQBRqORLiCSJMHpdB44YWqahnw+j0gkgsXFxbbeZzdhMBgwMTEBt9t92rtSF13Xsbq6ilQqhVwuh0KhAL/fD0mSUCgUaAqU1CjJsgyLxYJCoUCjdpUF7aT+SRAEuFwu+Hy+E19YdV0/V6Kyk2iahvn5eezs7DT8vfM8jxs3bsBms9HzyUWIgDFah+O4e7qu36l1H4vwXGBIODmfz8NgMMBkMkEQhAOPSyQS2NzcRDqdhiiKMBgMtPPH5XLRK+JIJAJVVSFJEqxWKy5dugRgL7V17969hvvidDohyzKtF0kmky0JHgCYnp4+sFiTrq18Po9YLIZ8Po98Pk+LePP5fMuv0404HA5cuXLlzF9FcxwHj8eD3d1dZLNZAMDW1taBx6mqSlOMPM/DYrEgEAjA5/OdSBdeK1xUsQPsfTfT09OYmJhAKpVCJpOhdXOFQoGm+8h54dmzZ1QcKYoCm81GU9Nms/m03w7jjHO2fvmME0PXdYTDYRpJMZlMUBSlpuCx2WwQRRHZbBa6rqNYLCKdTiMSiWB9fR3j4+NwuVwol8tYWlpCsViE0WjE1NQU7HY7FEXBSy+9hGg0ilKpRAtrK6+sSb1PLpdDKBRqWMhcCc/zMJvNsFqtVVd8uq7jyZMniMViKJfLHfvcuhW3233mxQ7BYrHgzp07iMfj2NzcbPgde71eXL58GRzHoVQqYWVlhdbJDAwM1BX5jJNFEAQ4HA44HI6a92ezWczPz6NUKsHhcFQV0jd7rmAwmODpMkjUhqSSAKBYLMLtdsPpdNLHZTIZZDIZqKoKg8HQ0H+msnC1ET6fD16vFwCQTqeRTCaRTCYhyzJ8Ph/8fn/N52mahqdPnyKdTsNoNMJmsyGZTDY8kZGi4+Hh4ZoLNcdxuHr1KkqlEjKZDHK5HFRVRSaTQTabpW3wTAztEYlE4PF4zk3dCMdxcDqdcDqdUFWViluCwWDA9PQ0bDYbjaDoug6DwUCjBvl8nkYsL3KU5TxgMplw/fp1AHtF+slkkp7DstksrUGTZfmU95RxlmGCp0vIZrPY2trC7u4uCoUCvZ10JPX29lbVCpACTQIpIiyXy1VpqsMolUrIZrPU88VqtSIajWJ+fh6qqoLnedhsNhQKBfraJpMJLpcLAPDWW2/RWgzyPkiqoh4cx9HCzmKx2DAyQdJrqVQK6+vrF65AuREcx8Hn89EiZZLaPG+LvyiKuHbtGpLJJGKxGDRNg9vtPhAtkGUZ/f39p7OTp4iqqkgmk0in09RiQVEUDA0Nnbs6mO3tbSwuLh7oFkylUtje3oYkSRgfH0cul6N1WsDesU5sKM7b8c3oHKxouQuIxWJYX1+HpmnUc8Vut8Nms0GSJORyOSwuLiIajVJvFlKH43a7YbPZUCwW8fz5c0SjUVgsFkxOTiKfz4PjOLjd7qoTYzwepyed/e3Q4+Pj6O/vp7Uz5XIZqqoiGo1ifX2dFi6Pjo7SiALprFFVlboOC4IAURTB8zzK5TLttCHpsv0prGYg3UlkUax0X678VznKIZvNIplMUtO+boPneTidToyOjrIaiHOIruvI5/NIpVLUubuyMJv8l/zeyT+bzQabzXbauw9d1xGNRpHNZqv8o8xmM21Xr4z2HjV9JQgCrFYr/H4/AoHAqYqfcrmMcDiMXC5XNTKG53n6fbGUa+uwouUuh4T2a5FOp/HkyRPkcjnqupvP51EoFJBKpbCxsQFJkmCxWGhePJ1O4/79+wBACwMrIz5msxnj4+MQBAEPHjyoSgtZLBbouo6dnR1Eo1Hk83l4PB4MDAxA13UsLy8DAJaWlrC0tASTyUQN+YjAIcXE5GRN4HmeprHauTLleb5hnUAjisUiNjY2sLa21vJzzzLEhTmVSjHBcw4g6Zx4PI5EIoF0Ol0V7TAajbDb7dTh22QywWAwnFokR9d16q2TSqUQi8WQSqVQKpVQLpdPPKVcLpdpKuy0Iz2bm5tYWlo69HGKouDy5cuw2+3Y3NxELpeDLMuQJAmKokBRFBgMBiaMmoBFeM4xJBJBThykm6HWD7kynVUsFqn/BwBan1MpPkhBMfm7HqQDiggWWZahaRoikQhSqRRNaw0PD4PjOITDYSpo8vk8NeQjVzgAqmZbkaJEkroymUzw+/2w2+3H8Ik2Rtd1ZDIZbG1t1ewMOq+YTCZaYM44u+i6jlAohKdPn1bdTkaEuN1umEymU9q7g6TTaTx//pyeZ04LUu9ns9lgtVphNBpPXewQ8vk8dnd3aeS68sKSrM2SJNHfJhGQxF9KVVVaIsAEzx6NIjxM8JxTEokEnj17diClRKI0JDRKuh/8fn/bJ8NGPiHExddoNNIQOjmhJJNJ7Ozs0OgSx3HUcNDr9VbtD0mBFQoFmEymKpFVKpXo/KdCoQBVVakgqyfwjpvNzU0sLCyc+Ot2GqfTiZGRERiNxnPTpXXRKJfL2NrawubmZlVq1W63Y2pq6tREjqZp1B1ZVdWqtDBJT6+urp5q3Zwsy3A4HAfSQx6Ph15IVXaMkpl0tSAz/w67CDwKxC5D13XY7XYmYtqApbS6ECIu9kPy9wRZlulsKGLC19PTA1EUDwgZ8ncul0MsFqOt67FYDKOjoxgcHKx6re3tbTx//rzqNo/Hg8nJSQDAgwcPDnjcyLKM0dFRWruztbWF7e1tmsPv7+/H2NgYNjY2EA6Hkc1mUSwWIYoiRkdHoaoqVldXaShcFEV4PB4EAoG2UlXt0tfXB0VRMDc3d+7aYp1OJ1wuF0RRhKIosFqtZ+aKl1FNuVzGzs4Otre3qdgRRREjIyPo7e098vdGmgVkWW55W9FoFI8fPz7S6x83xWIRoVDowO0Wi+XAyBWC1WrFyMjIATPTfD6P2dlZAO8YnQLvDBv2+/1HcunOZDIIh8M0lR+JRBAIBNjsug7CIjznkHA4jCdPntQ1zJNlmXZlRaNRpNNp+lhStEc6o6xWK7xeLxKJBOLxOAKBAAYHB/Hmm29S4eR0OjEzM0NPiIVCAUtLS3Qb+7vCnE4npqam8OTJE8TjcXAch+HhYQQCgaoT6/LyMlZXV6v2nUSk6s2lGhoaQl9fH1KpFPL5PMLhMK0z8ng8J96WGovF8NZbb53oa7aLxWKBw+FAKpWCxWLBxMTEae8S4xDeeOONqq5FckHRieN8YWEBm5ub9G9ZlhEIBDA6OgpgL4qpqiqGhoYAoGqALs/z0DQNd+/eZdYOb2OxWHD79u22RGg0GsXm5iYymUyVx1Dlf2vdRuoeGe9wpiM8pDuHpGEYjSGV/aSNeD8ejwdTU1NYXV3FxsZGzefH43H6dyqVqsqxkxOgLMtU8MRiMWxtbaGvrw/AXhHdpUuX6FiGYDBIt0GKAkVRRF9fHzUGrHX1k0gkau5foyGcW1tb2NjYOHCSDQaDmJ+fh91uR39/P7xe77Fb0OdyuQP1FGeZdDqNXC4Hl8uFQCBw2rvDaEA2m8Xs7OyBdJDVaj2y2FFVFevr61ViB9g7F7tcLpRKJTrY1263Y319veaQX5PJBJvNVuV/dJExm80HouaRSIRaYwCg6T+j0QhBEKCqKkqlEkRRRDweb0s8DgwMYGho6Mw5iO/nLIxQOfUIz9bWFlZXV6EoCvr6+uqa03WScrlM26PD4TAikQhthezr66MFtKRG5KxSKBSwsrKCYrEIs9lMr+BJ4XAsFsPOzg415zrKCAVFUXD9+vWa9QK6rtMCZbPZDEVRsLKygpWVlarH9fb2Ynx8nAoQMkk9GAx2dAo3GQqZz+chCAJGRkbg9XqPJfpD9rtQKCAcDmNnZ+dQH6HTwO/3V41bAICbN29e+EJlIgIrC/oLhQJkWYbdbj813xZSnJrNZrG7u4vt7W0A74zW6O/vrzJVbJb19XWsrq7W/L0ZDAZcuXIF5XIZz58/Ry6Xo92blRdJjPoYjUb09fXB4XAgk8nQCyJSQlApaAwGA+1WOyocx8HhcMDlctECZtLMQv4Rw01JkqjgOilIk8zMzMyxv9aFL1omkYhcLodkMolUKlV38SehWmAvykFmPGmaBl3XqY8FaQMknUykUI/kfI8KCR1LkoRoNIpcLodAIEAP0mbUcqlUQjQapT+0yuGdmUwG8Xj80IGe+yHePU6nE3a7vUoQVnpJPHz4ENFo9MDz7XY7ZmZm6PtIJpPULLFYLFaZEDaC+OW0Aunycjgc6Ovr65gAevDgARKJBDiOg6IoB+qoThur1YrBwUGYTCZkMhmk02nauXLck8HPEuT4Ir+bjY2NA/UdJCJIji2j0Yjh4WF4vd5Tbe0mixbpytR1HWazueV9qpVGrsTj8eDq1auIx+O0XoXRfYiiiKtXr7Zd99hqtOYkoztnOqV13GxubuLFixdNL46VBajFYhE7OzuHPoeY15XLZZrCIW3esizDaDTCbDYfiBYRB1QibHK5HCKRCJLJJAqFAq2HIbOCVlZWMDw8DGDvSu3GjRvI5/P06o+0J1osFtoR5fP5sLKy0vAk1wq5XA47OzvUt2VychKqqiIcDmNlZQWapsHpdKK3txdutxvr6+tVqTdZlhGLxWA2m+saoEWjUSwvL9NOMzL8kbSqV4aI4/E4dVk97DsulUpIJBJIJBLY3t7G9evXO+I9QwSvrusQBAE+nw/pdLqm4DsNUqkUnjx5AkEQ4Ha74fV6L0Qbq67rVSnXRkKadAGqqgpd1yHLMgqFAvUoKhaLSCaTcLlc8Hq9Jxr5JfvW6DU1TcPi4iKAdxoVFEWBxWKp6r7zeDzUpLQWJBJxnlK1jNZRVRW7u7v0/JrL5VAqlagwIccQWdc0TYPdbgfP81BVFW+88QY1uiUjPYjPm6ZpUBSFroEk6n8WOPUITyaTAc/zHZnho6oq7S4iX+hp5pcVRaEnWY7j6FRfRVEQjUaxu7t7LJO6eZ4/YDxW6biaz+extrZ2wNSPnFTJQUoiIIIgIJvN0h+Fqqq4du0aPYh1XUc8HkcsFqOPzefzuHz5MiRJwoMHDw54cfT09MDn82F9fZ2KI3IFUKv7C9hLiZEOsEp0Xcfq6irW19ebCg8rikIXLjLi4qiQ1yXdHPfv3z8TYywMBgPcbjcdtXARChw1TcPOzg7W19dpepGcvEldF4FEZclwUYLBYKD1hfvFAXHCJSd60vp8HCKSeD/pun6gW0fXdWSzWeqeHo/H8fDhwwP7azKZ4Ha74fF4aBpsa2sLy8vLByK8Xq8X09PT2N3dRTgcRjgc7uj7YZxfDAYDhoaGaF0XudBuBqPRiJ6engOdvsfBmY7wxGIxvHjxAmazGR6PBz6fr+2r7ng8jq2tLSQSiTPXKkxEwUnkwjVNQzqdrtl2abPZMDk5iUAgQCMSpMWSLAQLCwtVESFJkjAwMIDR0VGUy+UD+V8yyDGdTmN7exsmkwk9PT3gOI5a3u8nFArRH0wikcC73/1uCIJAvTv2I4oienp66r5fg8GA8fFxOjdrd3e3boSFpDbS6TRNcx11oRIEARsbGwgGg3XbXY8LMmaDDMI0GAz06v4smawdN6VSifrVlEolyLJMu1iIuSUAehsRpJUzlwjkoqnWeYSMHKlkc3OT/g6IwDQYDFWPyefzWFxcBM/zUBSlpkN65fBf4qYM7I1iyefzCIVCNKWVzWZRKBTA8zx8Ph+i0WjN/SXz6dbX1yHLMjweD5xOJ4xG4wHBQ0zwbty4cWYilIyzQT6fr3kh2gy5XA7BYPBEBE8jTj3Cs7GxgRcvXlTdZrVaEQgE4PP52jJDK5VKdAzAcURQmoVEU84KHMdRN09JkqpqAkgkanh4GNlsFpubm3VFI8dxsNls9IRJ5l/l83m6iFgsFpTL5QOdHbXgeR6Tk5MwGo01zRQJQ0NDGBkZOXD76uoqlpeXwfM8rl27RhcRcnszKIqCwcHBtr1NgsEgnj171vLzjoLP58Po6OiBhfWioes6tre38eLFCxpmbya6VhmBrQWJhrV78eTz+TA9PU23o6oq7t69i3w+D7fbDZfLRUX86uoqkskkksnkqbZ5K4qCq1evwmQyYWVlBevr66e2L4zuwmQy4eWXXz721znTEZ5akFbpRCIBk8mEdDpN6zccDgc8Hk/DRYlEJLxeLyKRSJVp10lymmKrFuSqsFEXUTMKXtd1WgtTj1aiHJqmNRQLsizD7/fX7eDr7e2Fx+OBoii0ziGXy7UUji8UClhYWMDOzg7sdjs9hpoVPyeZo+6k8dx5R9M0LCws0Boug8HQVNF7M0iSVHNbJC1EBnXWIx6PQ1XVqtTwwMAAFhYWoGkajEYjwuEwEonEgRbx08Dr9WJycpIOEq5l2MdgnGdOXfCQaEFldIAQCoUgSVJV2HVzcxMmkwnj4+NV9Re5XA7RaBTJZBKJRIIOrXS73ejt7a1y52WcXWRZxs2bN2kbPUnNHCZwKyOBmUwGDx48aLgYDQwMwOVyIZfLYWlpiT6WXGWT2V/NcpwFwCRVRUwij1KLE41GsbW1hUKhALPZjP7+fhiNRkQiESQSCfpZk/o3juPQ29uLsbGxDr+ro1MoFPDkyRMkk8mWxQ6pc6vnWE62v//8Q27v6+vDzMwMVFWlwygzmQxKpRJMJhMsFgtsNhu2trYQDAZpuotcbMRisTPlX0MsOIhZKIPRjZwJwXP58mUkEglqWJfP5+lwxsqTTX9/P50MW9lOl06nsbCwcCDiwIruzhdGoxFXr17FkydP0NvbC4vFcqitOumgIYNUVVU9MEG6Fvl8HkajEU6nEx6PB6urqwiFQhAEAVNTU3Wnz9fDZrPh5s2bDVNyzaIoCvr7+2m3Xzu2/7UgdSlWqxWiKCIYDCIYDFZZMdRiZ2eHiq2zElGKx+OYm5ujF0nNih2LxYLp6Wnqr6NpGsLhMObm5mo+vpawJE7jBoMBPp+P1uNUfjabm5t44403qqK8Rz0ujpNyudxSESqDcR45dcETCARo10Mlbrcb8Xgc4XCYniiGh4drtmZaLBbcvHkT29vbCIVCyGaztJUzm83CarXCYrHQgjzG2cNkMuHq1au0mHh+fp7eTuodSFskIZ1O4/79+1WLCrFbPwzSheJ0OuH3+zE+Po6xsTFayN0OdrsdQ0NDR6rlEUURd+7cOZZBnhzHweVyUesD0rV0WI3KYUMVTxJd17G2tlZVm3VYLU4lU1NTsFgsAN4ZykkurmrR6LMhF2OlUgn379+nKXen03mmDUsZjJOCDCU2GAxnwt391H+V9ULzpNNhdHS0SsBUmtvtp6en50AnT+XcFzKsknH68DyPQCBA/UJ8Ph9tL69ElmWkUimsr6+jr6+PjqkQBAEWiwUvvfQSyuUyrfEi09PT6TR2d3cRCoXq1m+RWWPRaBSFQqEjHQROp7NmGqRZrFbrsU4tF0URfr+fDm/NZDIHxIIoirDb7bDb7XA4HLBYLKfezk4G2S4vL9O0EM/zkCSpqaJiRVEwNDQEs9kMVVWxsbGBjY2NQyOB5LxR63zz4sUL7OzswOl0wmw2IxwOI51O1xzpwmgNUgNIxCnjfJLP59Hf30/HEp02p96l1SpbW1u0dZMMyaxVMEpSHaIoUn+Wk25Xb+XK86IxPT19QPEXi0VomoatrS2USiUEAgFqSpjL5ZDNZuli3Gy0IZlM4unTp1XpBFEU4XK5wHEcIpEIVFWFKIp4z3ve05EoRiaTwdzcHDKZTNPP4TgOfr8fo6OjJz4AVdM0OlKB2M6fhWgOADogtla7fzPC0mKxoL+/Hz6fDzzPIx6P4+nTp03/Lkl0i0WGT5bP/r2/B5RKeP9v//Zp7wrjiHAchytXrsDj8ZzU63XnaIl6V16HTRM/KZjgqUaWZYyPj8Nms1EjuEaQ7q10Oo3BwUEqkIiQ5Xke6XQa2WyWmiwSJ1BVVWmalAjkaDSKWCxG3a2dTidNY6mq2nLdzmH7vri42FT3DcdxeOmll2rOKbuIkEhs5VDaWhw2XuTSpUvw+Xz0ONM0Da+99lrLYz9I8TwTPSdHPB6Hrusd/U0yDsfhcEAQhIYDnJvFbDbT2rpSqYRXXnnlRCLF564tvVlqLZhra2tYWlo6hb1hNMLhcODKlSstpWt4nsfIyEhVXY6u63j27BkGBwfhcDjw4sWLqq4SSZLoeACHw4HBwUE4nU5YrVZYrVYMDQ3RwYwkauTz+Tr9dsHzPMbHx1EsFrG7u9vwsbquI5fLMcHzNs+fP0cwGDz0cTzP1+28dDgcB2wMeJ6Hx+NpatsEkiJlYudkaXfGE6MxHMfROZD7I6ZerxeXL1+mI0pa3S4ZYE0MNU86Ut0M51rw1ILMrOrU8Eae56mfz1kaCHmesNlsuHr1aluFnPtHjnAcVzVx12azVQmeyhQHcav1+/1wOp0QBIHa/xNH4uOE4zhMTk7Sjqh60QiDwcCuZDuMyWSqGQFuNlVH5gmxCC2jGxgYGEBfX1/ViJX79+8jlUrB5XKhv78fTqcT0Wi0JU8oURTR29tLt33W6RrBk8/nkU6naUtvMwWJh2Gz2dDX14dnz55VLVYDAwMYHh6mVu+xWKzuBPZOiq/zyvT09LF1rXg8HqytrTV8zM7ODh0C+653vatmFxZJa3W6YFiSJExNTWFqaoreFo1Gsbi4SGt8hoeHT70o+CzRaK5esVjE5sOHGLlzB4Ig1I3wbG1tQdO0AzVRgUCAztkjCIIAq9VKR4xwHIdoNIp0Oo0nf/qnMHk8GLlTM0LOYJxpTCYTpqenDwxoJnYwuq7DaDRiZ2cHd+/erVt3KMsy3G43BEHA9vY2RFHE0NAQbSA5L3SN4CGFhaqqYnh4GH19fQiFQpBlGaVSibY5NwNx2e3r68Pc3BzN4Xs8HmooJggCzGYzRkZGMDIyQscokC8/nU5Tx+hkMglRFKlfDJnGDKDtbp7zAqmtOS4MBkNLojKXyx2I7JTLZTx8+BCqquLmzZvH3lLscrngdDqRz+fp9HfGOzQSPFtPniDyW7+F8VdeOdRINBgM0unwBLvdDpvNhkgkQmdakVlbZBBoqVQCz/PIZDJI/t7vIeHzMcHDOJMIggCbzYZSqYRcLlf1mzCZTLh9+3ZdQWI0GpFIJDA3N0fTW263GzabDeVyGYFAAKlUCkajEVarlUZHh4aG6CDq88a5LlpulmKxiNdee422OJIiKgA15+643W5cuXIFPM8jmUyC53lqVNbu66+trSEYDNKrUrJAd3ths9FoxNTU1LGkkPL5PGZnZ1saG+L3+zE4OEgH1Oq6jsePH9MiPbvdjmvXrjEflVOkVCrhtddeOxAxdbvdGBsbw/r6OnZ2dhp2XBoMBvT29ja0Gpibm8P6+vqBUHy3/yYZZx+fzweLxQJJkqgQFwQBqqpWrR8kRQ8Ai4uLiEQi1JldURT4/f6arf3lchlLS0tV6Suz2YyZmZlzkZpqRNd2aTULSVcQH59yuUyv6J49e1ZVkS6KIiYnJ4+lkLVYLCIUCiGRSNBJzJqmIZ/PX4jCyJdffrmj0R5N07CxsYGVlZWW7QZ6e3tpbc3+z91ms2FmZoaJniMQj8exsLAAQRDg8/nQ39/f0vPJDLxsNguz2YxAIAC32w1g73t/44036gpdi8WCO01EZP7sj/4If/St34pv+vSn6aJwWOfXcXL34x+H5eWXMf1N33Qqr884faxWKyYmJg6koDoF6Vh9+vRp1UxFh8OBa9eunav0VD26tkurWTiOo2kD4qshyzIymQyi0Sg4joPH44HX6wXP83Ty8tDQUEcNk2RZRn9//4GT/87ODp4+fXokw7rzwMrKCi5fvtyx7fE8T6M1jx49aum5jQwok8kkHj16hJmZma44AZw0qqri0aNHNOytqioWFxfh8/kOHRVCIMajlZTLZYRCIXi9XoyMjODp06c1n2u32+tut1Ao0FTXN/y1v4bYr/86bDZblbg9rQsPPRpFgY3CuZCIoojR0VH09PQciweWrut49OgRkslkzfR/Pp/H0tISRkZGuvpCr3vfWQNyuRzW19extbVFBygSh+ZQKASe5zE1NVU1nPQ48fl8KBQKWFtb6+oi51AohEKhQK/MiZuv1+s9Uguj2+2G0+ns6DDGRCKB9fV1DA8Pd2ybF4VoNEprCfL5PFZWViAIQstRHkKxWMSzZ8/orL1AIACfz0fHg+xnv1AiaJqG2dlZ+vv3+XwYmJpCqVSqEjmnldJ66d/+2xN/zfNOOBzGyrd9G6JuN77xv//3096dtggEAhgbGzu2Wr5yuYxnz54hGo3WfUw+n8fm5mbXn+8ujOApl8u0lmZ7exs8z+PGjRsH/B58Pt+xpLMawXEcBgcH4Xa7cffuXSiKQt1vT9vxNpVKHTqtvBUqB7yGQiGEQiG8ePECPT09GB4ebkv4FIvFllyNm0VVVRSLxTPpJ3GW8Xg8MJvN9DvhOA7T09Ntf46RSISerAVBQDweh8ViQV9fH2KxGI0kkbEe9S5UiJM3qVGIRCII7+7i3g/+IL7qN36jqlj6tKKtCwsL+L9/4Rfwsf/8n0/8tc8jBoMBRY4D3q7JO0+YzWZMTk42jEh2gt3d3UO9wIxG44FIZzfS1e+O5CvX19cRj8fh8Xhgs9ngcDjgdrvP3JdrNpvR29uL7e3tQztQToqv/MAPYOwnfgIjFd43nYY4625tbcFgMMDr9cLr9VbNx6qXWopGo5ifnz+WNMTGxgY2NzcxMDCAoaEhlt5qEp7nMTAwgFAohGg0Cl3X8eTJE2ps1qp4JgJFEATcvHkT2WwWr732GmRZRl9fHzKZDIaGhg6te1AUhRZ9EnQAz3d2kPkv/wV/58d/HIVCgUZ3SIu6IAgnFvF59NGP4icB/OIv/iL+yf/2vx3b65BOUV3XUS6XT92Vvl0sFgve/bnPnfZutATP8xgeHkZ/f/+J2FHUioLuZ2pq6kKYPXZl0XK5XMb29jZisRgNgTscjlOPljRLsVjE48ePkUwmT3tXkE6nz8QAvzt37lRNuV5bW0MkEjngFnpccBxH2zOHhoaYK3IFuq4jEolgdXUVmUwGfX19GB4exsbGBpLJJGKxGC0qHx0dRV9fX0viUdd1PHz4ENlsFtPT03A6nchmsyiVSjWvjnO5HGKxGD7z7/4dvvGjH8Xg8DAWFhYQjUaRz+dp84KmaVhfX4fzwx/GfQBf+/nP11z4JUk6sWjr7/70T0P/4hfxN//kTzra1cjzPB3HQpo4KoXfYan0emN8GK0hCAJGR0dhtVqPrTB5P/Pz8w1rFq1WK8bGxrpG8Fy4omWe59Hb29t2zcBpEI1Gkc1mYbFY4HA4MDAwgPn5+VMvYj4LYgdA1QJJnJJPSuwAeyf8bDaLbDaLfD6PGzdusAUAe+Jifn6eXlxIkoRMJgOe5zE0NARgL4354MEDAMDS0hK2t7fx8ssvt+R6rCgKYrEY3nrrLToqJBwOU6dXs9mMUCiESCRCu0+Cn/gE/s+NDXz1938/gHcWfRK1IYvPrwUC6P/AB+pGOUqlEgwGA+2mPE6+8+d+7kjPF0WxagyLruvQNA2lUqlhJ2NlYwfP8ygUCtR/jNx/Eu+/2ymXy1hYWACwJ/4DgQCdGWez2dDf3w+O45DL5ZDJZDoycNPhcNQVPK+88kpD36tuoysFD8dx5+6HKYoiVldXcenSJQB7c00cDgdef/31M5PeOk3C4TAGBgbo336/H4uLi6dS4J1IJPDFL34R4+PjtNj9IkNGe5ChrWQSfTKZRDAYPPBbzOVy2NraqtkBqWka4vE4HeUSi8XAcVxVSikWi9EC9Yd/+IdwTk5i9MaNAxcH7/30p+k2K/9LIMXJf+93f/fQ91gqlaAoCkql0rGnf57/+38PsacHY3/zb7b8XEEQkMvlWh6pUfnZ8TwPWZZpDRuh0ftPp9NH8io7q5CU+n4OE5DNsLS0VDX3cXd3F7FYDOPj41haWkK5XO6Y4NkPz/Nwu90XSuwAAPOzPyPYbDbcuHGjakSFJEkYGRmhJnkXmdXV1aqTL8/zpzp/qlwud2Si8HlHkqSq7hISXQX2BND29nbN2Ty/NjGBnxkcPFBMubq6iocPH2JzcxM7OzsoFosN62fSf/AH2PjUp469FqLSKf24sb773bDfuNHy8zrlIUQaJvYv6MVisW4n0cMf/3E8/5//88ivfdJomoa/+NjHEA2FDtxH0n+krqvyX7lchqIoHT8eyCyrcDhMC/KPiizLVZH6oaEhvPe978WVK1eOvO3zBhM8ZwhStFx5kPf39+NGGye/bkNVVTx+/JimK8jV/2mhKErVyIKLytbWFnUiHx4exp07d2C1WlEul7GxsVHzOXd/+qeh5HLgtrer6lRIu/i9n/s5PP3932/K8fXdv/3b+Kr//X9va2HQNA2iKFIn23qLOekuk2UZPM/TwaLHRe/Nm/BPTLRU58RxHBVlxxllKRaLNRf6mV/6JUx97de2tc3PfvKT+KP3vQ9v1agXJWk1SZKORWxqmgZhZQWx7e2ar13vOyBRx067EvM837RXVSsQqwZJkuD3+7suEtcsXZnSOs/UOumKogi73V7V0n0RSSaTuHv3Lnp7exGPx088nWW1WmG322GxWBCLxRAMBuHz+S7syQMABgcHa45vEASBhuYjkUjVYqXNzUFyufD+z3wGkiQhGo0iHo8jGo3uFdbmctCy2aY7o0itSqut5DzP16xLITU+JGpEXNH3P6bWWJqjQtJGRMCR19F1vSqdRGqRiHM8uf8kPIQKhUKVUC0UCrBYLAeGuf7h+9+Pr9U03AXwjX/5l3W3t/LJT+JHdR2/8Eu/hOv70osGg4G+H57nae1Vp96jKIp49fd/v+Z9ZMZaI0qlUsMhtq0yOjoKv9+PnZ0dmrrtRFqrt7cXxWIRo6OjF3p2HxM854C1tTU6hLNYLCKdTl/YWT9knESnIFfGgiBUFXhWIkkSxsbG4Pf7AQBvvfUWcrkcbt++faHFTiNisRhevHiBUqkEt9tNhw+azWbcXFqCruuwWq2Ym5tDJBKBJEl08b758z/f1CJCxADpOCqXyzQSwPP8oWKkXg0G6WJqBPn9EYHCcVxbix4RakS07f9d77+NvF+yf6S9nLz+STU5EPNQUi9J2tsrO+CGNA1/E8D+1gIStSHP+8hnPoP/+F//K374wx+u+3oP/tW/AngeN3/8x6GqKhRFoSlGIlCJAFRVtanfZaPOs2aFo67rEEWxY4KHeJ45HA7EYjGsr6/D7XYf+TyjKAqmpqY6so/nGSZ4zgF+vx/hcBhbW1vIZrPw+XwwmUzY2NjoWlfmk8BgMGBmZoa2mKfTaUQiEdp+ThYTcjIFgM3NTcTjcZhMJrx48QKXLl2iCyzjHYhnTmXBZz6fhyzL0DQNkUgE6+vrtA6KiAayeAOou4j87quvog/Aytd+LT78sz9bdV/lgi9JUs0Zdbqud6zrqnJRrPSL2t/2XQ+yH5IkoVwu46++4RugCQK+7k//tObj978XjuNOpZOz8nOrjN6pqko/38X3vQ//7HOfw4rfjw/jHYG639kaAD704Q9XddERMUSjO5IEVBxLlVGf/T5CJCJW69xIBCYRxOR4q4yWkdRgsxQKBciyTId6HuWYIpEzIsYTiQTC4TBLn3eIrvTh6VbK5TJ2dnawsLAAo9GI/v5+lEolbG9vH7lj4LjY32FzFrDb7RgYGIDJZEI4HKYnN2IEFovFUCgUDhiDpVIpzM7OVi1kRqMRhUIBPT09mJiYOPH3ctYhi5Ysy4jH43j69CnK5fKhx2tlFKBcLmPj8WNIJhPeePIE3/gv/yXeD+DnAXzdl75ExQWJfpBzGllcgT0xQiJ4xFfnOAuQm40Q7E/Dfe5bvgWQJLzvD/7g2PbtpFEUBeVyGcFgEI++/duRtdnwzX/4h8f6mpVt9pUCupYIEgQBmqZ15HioTLm1KnxMJhNefvllAHtR5Hw+j4GBARSLRQwNDbFocpNcOB+ebkUQBPT29sLr9WJrawuRSASXLl1CsVis2QlzFqi8oqoMfZMrU3JiOqmrVK/XC7/fj83NzaqiZ1mW6UwukrqqpHIgZiW5XA7AO90VnRw2e95JpVK4f/8+XUhI+qcZKiMzPM8j8dnPgnc6cfW978VDADYAwbcfS0TM/tZqImpI1AAAPRY1TTsQpZAkiaY2j9p+ns1m8Ylv+AZ43v9+fOdP/VTNx5B9rNyP9/0//w9yuRzK5XLDouWdnR289aEPoedf/Stcu3Wr7f08ToizNSnu/fPf+z38I03Da/E4ln7v92C+cgX+Gp1C83NzePqjP4pXP/Wpts35dF1vKkpDorNHSUk9ee01qB//OPIf+hBe+Z7voaNOiOCtPLYaRf4qLS5UVUV/fz/teGR0BhaHP4dIkgSfz4dIJII33njj2GexHAWO46ifR6FQQLFYrAr9VgofRVGOddxHX18fDAYDHj9+fKDDy+PxNGxzJ8ZgZIq3z+eDzWYDx3Fwu90oFovY2Ng4txb9x4HZbMalS5fQ29sLURRrRjw+/9f/Ou7/8R/X3YaiKBAEAdd+7Mdw5Xu/F8PDwxB/5VfwM9/0TfjO116jYqHyyp2kskh3D5mjVyqVUCgUaHcWSaHIslzlNExM99o5FhVFgSzL+Dcf+AB+U9cx+Rd/Ufexmqbh/37f+/Dg1Vfxu7/0S/T2L3zLt+DTP/RDDV/H4/EgMT2N8bd9uxqxOz+PzTffbP5NdAAS2SOLe6FQwHf94A/i34yMYPk7vgN6Pg+tjiAZn56G/0d+5MSciEmqjRxrJLLYLOGlJXiKRaSWl+ltpM6IHINE6NQrQSCjJgiSJLEZfscAS2mdY9566y3EYrFTm+58XJDFqJMdMKTwj5jk7e/0UBQF165dgyRJVa2m+Xy+rsV/Op1GKpWCz+fDa6+9BpfLdSG9LZohn8/j8ePHB9yx/+ev/Apufu/31lzcSBfUo//4H6HF47j1kz9JIzDEI6XyGCG3kyvpWsiyfGi6gfyeSDEqKW6td67kOA4ulws2mw35fB7RaBR/+Fu/Bfuv/iqWZRl/98/+rOrx9z/7WSR2d/Hqd3wHvvLqq/gOAN8ny/inbz/u0ec/j5c++EEYDAbout5Sd+bypz8NdWcHEx/9aMPbjgOyQJN6mEbRvMNGWZwGpA2dRKM7GXUm266Vyh0cHMTo6GjVbXfv3kUul8PMzMyZvqA9izRKaTHBc45ZXFzE+vr6ae/GsUFSIJ2qT6psI67VUkxmPKmqCq/XC7PZjPv37+Py5cuHXm2SFAnLs9enVCphdna2qcn2pBaiWCwik0hALZXg6+urmaowGAx0gSXRncqCV3JbrQLmWpB6H7IoX7p0CQ6HAysrK8hms0in0yiXy7Db7ZicnKSCmIy+OIy3Xn0VVgCjf/mX+MwnP4kHn/wk/vc//mOMjY3B7XYfENhf/vKXm76g2V1YQDEeR99LLzX1+E4hiiKN1B5Go8W/Ec/+63+FHo/j0o/+KLbefBOC0Qj/1avt7nLV/hgMho6ea/ZTL21vMplw+/btA+nL1dVV+Hw+FAqFrplxdVKwGp4updOmV2cNcpXdatdELfYLnHK5TGs5yBXd1tYWBEGAxWLBixcvUCgUYDQam7oSZV1ahyNJElwuV1OChxQ2i6IIi8NRs2VbURQMDg4ilUohFovR6A2BXK2Tjh1imHfYokxcdMvlMiYmJqgPytjYGNbW1qjg0TSNDjEVBAHLFSmNWkSjUbhcLoTe8x5sRiIYBfDNH/oQfuinfgpOp7NKVJNtEyGxn8q6kMqF2ltROH/v938f8U9/Gu//nd9puF/N8PA//2eoW1u49U//6YH7OI6jv6FmaPRb+aMPfxjX19ex8NGP4uu/8zur7vN/4zdCfbsdPv/0KXibDTiC4CFjI2odW52CiO9agQWn04krV67UrNXyer0wGo0XbvTDccMEzznmtAeLngREbDRjm19p0kb+BkBrOCopl8sIBAKYmJhAPp/H2toaQqHQnvOqIMDr9cLpdMJoNLKoTQdpRbgS47d6n7/b7UZfXx8SiQSGhoYwNzcH4J3WaLKQkKJUYph3WIcWWaAGBwcRCATo7RzHoa+vD9FoFOVyGWazmf43HA5TF/BazH7xi3jxMz+Dv/4nf4Jv+Of/nN4+PT1d9Rpk/9fW1rC2tlZzWyRaUOnHQwqEKwtiS4kEhAb71BLlMrAv+lHZit3K90q65UgUj3y/hUIBlqkpbG1uYrCGkHFWFPWONvDsqeSvvu3bwPl8eO+///cHBG9ljVGnIRFFcmG1/6KJdHXWEn/5fB53797F9PQ0fD7fsezfRYUJnnMMMf+6CNS70iWQluTDrtQURYHH44Hf74fNZoOu63jzzTfBcRyGhobQ19d3oZ1Ij5v9x6yqqvjLX/5lvPpjP1ZVJDz7Ez8B8zd9E65+8IMQBAEjIyNIJpMIhUL0O47H48jn87TGYXJyEvfv36dt5yTqURkJLRQKdet8iCcLMUusHFZLEEURV65coQXPhL6+PvT09CAajWJxcZF2733xQx+CDmDi3/wbfGloiC5wv/NzP4f/74/9GFKp1AGr/42NjbpiB6h9oUMKrSv9hd71kY8AH/lI3e20wszbE+crIQt6OxcEpM6nMoUkSRK+6Z/9MxTqdLW1gxyNolhxTiDC7FOvvopRAH8K4CcbuEC3CxFWtS7URkdHMTAwUPdzSyQS0DQNc3NzCAaDCAQCyGQyGBwcbGncCOMgTPCcU1otZjzv7D85kAWHnEwOu8IUBAEzMzO0s6pyu1euXIHZbKYGhIzWyGQytL7FaDTS9IbBYKDOu0RI7D9mI5EI+M99DuHv/m6Mjo7CYrFAVVX0ft/3wT81hfHxcbjdbvA8D6/XC4/HgwcPHgDYa/1+44036JWw1WrFjRs3sLGxgVQqhXw+TyMIlXUjJLoAgP63sjbIYDA0tBeo1z3D8zzt9ltZWUEoFAL3djfgl/7+38e/SCbxi7/8y0jcvYufC4Xw//urv8L/R1XxX8xm/HxFMffw8DBSqRQikQju/6N/BNeHP4zhJlrP9893qkylCILQktFiM9409S5Afu3VV/EeAM8A/O0WxASJUJFUczN85aMfBXp68MrHP17z/nd99rMAQI9DAPhrX//1+CKAGwDaLQrI5/N4+NGPwvD+92Pme76n7uP21w1WOrbvR1VVRCIR7Ozs0Nui0SgdueLz+dgg6SPCBM85JZPJXJgID7liJ/A8D0mSWsq7Dw0N1e12YC6m7aPrOp2XBaBhfc5+rxuO4xAIBOD74z/GyMgIhoeH6WNnZmaqnhuPx5HL5RAIBOD1erG7u0unei8sLMBsNsNsNsNut8NutyOdTmN7exvRaBT5fB4mkwn5fJ6OJCDRnVopjZGRkSMtLIIgYGxsDAMDA7Dfu4dQKITFp0/x87/wC/jun/gJfOU//Af8mz/7M5SsVsjr69Bq1IhduXIFyWQS29/3fXDvGwlQGTUg0chKP5/K3wop5FZVtcqIEQCW797Fzmc+g3f9zM9AlmX8z6/7OqguF97/B39AO9kqh3b+5vd+L5yLixj9xCdw9epVWlS+n28B8CEADwC8/33vwx987nNNf3ZEmFZGWUkKjETHSIE6x3EwfeM3wliR6iKPJ9E3IngrzxVXbt/G8zfewCiADQCXm967d0in07AEg0g+e3bgPtKK3t/fD7/fvyd8OQ4ej6dh/dLS0hK2trbo80ulEkwmE65duwaDwcBS6x2ACZ5zitls7rp29HqQExipxSD1GK2QSqWQSqWOZRLxRYbjOFitVlrI26idl3gyEbHx/N49bPzsz+K9n/rUoZ4jW1tbtNvq8uXL2NjYwK9993dDePAAr37607h37x6GhoZgMpng9XphsVgwMTGBYrGIZ8+eIRqNwmAwQFVV8Dx/YDBkpbNzMpmseRWuaRqSySRSqRSNFrhcrrrHlCzLGBkZQSgUAp/JYOj2bSz89m/j5X/wD/CeH/5hapPwQZOJpi4IPM/D4XBg5n3vg+nt+yujUKTglkRtiJCrZO3pUzz7+Z/HN/zWb9HC4soalvTKCriVFVr3VpYk6BbLnpHp/Dy2/vW/hvUHfgCX3o4uXVpcxI8A+I4f+iHc+OIX637PxOFqF4BrX7t1M1QeI/Ugwvn2hz5Eo3ZkHlvl8NVagvZf/It/gR//4AfxbwsF/J1f/uWW9w/Y80Hy/Pmf079JPVKpVILP50NPTw+NGNeL6OyHfH8TExN7xe2hEBwOR11bDEbrMMFzTiFXXxdB8JAi1FaucHieh8VigaIocDqd1ASM0XnMZjM9DitTSIcxcu0aQq++isuXL1e5zNZiaGioaoL4wMAA/sHv/A5e/x//AyaTCdlsFsvLyzQa6Pf7qamg0+lENBqFqqo04kEE9P7jqlAoYGdnBzzPw+12U7GdTqexsrJy4Pe2vLwMq9UKv98Pl8tFR41wHEeL4UVRRPizn0X2L/8SS729QD6PraUlpHQdnlu38N7v+i7Mz8/TYvlKBgcHYbVaYTKZ8Pz586p0VaWfUKlUqqpB0jQNA9PTUH7xF6uKdBVFoaLgznd9Fwp/+2/T6Mn7/+IvaFpp7QtfwCsvXuC3fuqn0PO7vwun04mn167hHz96hA/+6q82PO/8ptGIZ7kc1gH8xm/8xqHHQTvsH6Da6gibX6ozq6xVSOE28Q0TBAGjo6NtRWNGRkZokwTP8weK2RlHh/nwnFM0TcMXv/jFMztD6zSRZZm6KttsNtbaeczE43HMzs4CaG2G1K1bt5r6bgqFAt544w0YDAbcuXPnwGJC5jQtLi7SeiGz2Qyv14uBgQEsLy9jfX29asI2SXlcvnwZZrMZuVwOc3Nz6O3thcPhQLlcxpMnTzr6+/rS938/NFmGMDyMzBe+gL5UCjLPY/ztOpPluTmsf/zj+NjKCv1cIpEIgsG9IRp333gDQyMjh75OLY+pWo8hnxXP81WeRX/8Ez+Br/rJn0SxWITNZoPJZKJTyC8SPM9jYGAAHo8Hjx49op9pPd8eQRAwOTnZdERnPwsLC9jc3ITZbK55nDOao5EPDzMPOadwHHfoVTGwtxi89Wu/dmHqfYC9gZ5DQ0Pw+/1M7JwApH4HwKEC4d6//td44wd+AFevXm36uyHtxPUckgVBQF9fH27fvk3Hrty+fRtWqxWqqmJwcBBOp7NqPtbk5CSGh4dhMBigKAocDgempqawsbGB2dlZPHr0qOMXE+/59V/H+3/912Hp70dUFDHrdOKh14u1//E/AACevj6Ybt6sSu8pioJMJoP19XW88ZGP4PHrrx/Y7vbDh4hubNC/KwuXa6HrOpafPsW9//7faU0JSfMlEgkIT54gEQrB5XJRD6yLJnZsNhtefvlljIyMIJFIoFgswmAwwOfzobe3t0rscByHyclJvPLKK22LHbId4GLVZ540TPCcU0jB52FomgYtHIbapnFfLpvF3DFPNu4EJIfe19eHS03MF2J0hlwuR92+SY1OI6w3bkC5dQsWi6Wl13E6nVVFzbUwm80YHx/Hzs4OYrEYXC4XZFmGJEkYHx+vEkvr6+vo6+ur6szLZDLHvrBzHAfvV381rOk0ApqGV77/+yEajRAEARNTU/gnf/qnVa3HFouFpspe/b/+L1x917sObDPxuc8h+sYbVa9Radq5H4PBgNVPfQrZ//bfUCgUkM/naSOAw+HABz7zGQxWGBh2Ow9/6qew9fQpgL16m7GxMczMzNDRHqVSiQqay5cvY2pqCteuXaPPlyQJTqfzyLOvSE0UOZcxOg9LaZ1jdF3H4uIiNiqu7jpNaHUVG7/6q7j58z9/pkOsHo8H4XC4ppEb4/golUp46623DszIaoTZbMbMzEzHa6pUVcXKygpsNltNw7ZIJIJHjx4B2BMFo6Oj6O/vp8f1+vo6FhcXO7pP+6lMN1XWOimKglu3btX8THRdxxe+8IW2I06kbqeyk6sy9UgcqSuLkEmBeCfn2XWCh/fugfsn/wQJAK/8+Z93xDPr+Z/8CQbe+15YbDa89NJLh0YedV3HkydPEA6HAYB2lTUzgqYexGKB4zhcu3YNLperre0wWEqra+E4DmNjY02lttrFNzSEW7/wC2da7AB7J+iJiYkjhZQZrSNJEqanpwG8M2KgEbIsY3x8vCNiJ5vN4rXXXsPrr7+OL3zhC3jrrbfg9/vr2gy43W4a0SGLeaWI8Hg8x+7FVOmNU3mxWSgUEAwGa4oa0oXE8zyd7M7zPBRFqfKiqkehUICqqgdavYlNADEQrNzOYW7Up8XT117D1wG4ChzwdGp3f6e+6ZtgsligaRru3btHRW+xWMTq6irS6fSB78XtdmN0dBRmsxmqqiKfz+Pp06dtRQjD4TD1ljIajUzsHCOsS+ucQxyCd3Z2qn6Uf/Xqq8gJAj7wF39xint3ciSTSQwPD595YdaN6LoOl8tFUwCkhXo/Ho8HV65cOfAdkVlXpENIlmUkEgkkk0kIggCbzYZkMolwOAyv1wur1QqXywWTyYSXXnoJ4XAY4XAYgiAcajvQ39+P+fl59Pf3Y2RkpGpfjEYjpqencf/+/abed+WcJI7jqL9QuVymbdPkccQ3hud5GAyGmoXdy8vLWF1dxfDwMLxeLx0XQQwd8/l8VcSFbIOko+pFY8h2VFWtGudQ6ctDuj6Jtw6J+jQzXf4k+fYf/mH8uydPIHs8+NtvzzgD9iKNn/1f/hfc+pf/Er7e3ra3r6oq1tfXkUwmaYpzdXUVgUAAk5OTAN6pnyTF7/Pz8wD20rtbW1sYHBw89HXC4TBKpRLi8XiV0SBzeT9emODpAgwGA4aGhqqGF3IAuDN4hXZcZDIZJJNJZiJ4Clit1iqjQFmWsb6+TgdvEkEgimJVYebm5iZisRgKhULDdI0gCLQlPZ1O4+bNm/Q+SZLQ09ODnp6eplI+vb29MJvNVemdSmw2G97znvdA0zRsbW0hkUhgfHwckiRhc3MT4XCYFkYHAgFEIhHMz89TsUaEHhEYiqJUFaAS0QEAa2tr+NN/9I/w7f/lv1BTTE3TsLS0hKWlpUPfC4EUdJPi7srUFElVkc+G+Fjth4gxUthd+bhaHj/HwcbGBp5+93fjOoAvA/gbdVyaP/Srv3rgNkmScOWnfxreimj34u/+LizXr8N/uXVrQRI9GhoaanghtX8ESKN5aoRyuYyFhYWaopfNzjpemODpEgYHB5HL5WgL63uPYT7MWees1RtcVNxuN3K5HEwmE3Rdp2NASqUSvXpOJBIHokCzn/kMks+f42v+8T+uup2MOCCLTiKRqOma3ezEevLceDxOC01Jqgh45yp7ZF8L+OjoKAYHB6uGkzrenuROaCYSUi6XIYoi/uIf/kP8ciaDH/jRH8WP/OZvNrXv9SApOoPBQIUXMcJrNjpTL5JDhBt5zHHx29/93fgUAAuAAQB/60Mfwic/+cmmnz+wz5FaL5Wgtzkc1Gg04tKlSw1rcjiOg9frpUX7AA4UiZMRQNlsFrlcDqIoIpfL1f0cWYTneGGCp0vgOA5TU1P0irkZ8vk8Pv+3/hbkr/1avO8nfqLmY+7+7M/CeO0arvytv9XJ3e0o5MRznLVMjObQdR0vXryggz4JpBCUDNWsReSP/gjS9jawT/DsHxzbqUXBYrEgm80iEolAURT4fL6aXU2EfD6PJ0+ewGQyYertxbVexITQ6L5v/k//CT/8oz+Kj3ziE+2/iQpIdIe8ZitmnZWzpoDqad/EWI+4Ubcyk6sVCgC2AUxgz6X5ao2J6a0w3uRE9VoMDAw0VYBMond2ux0ej6eqYaJcLuP58+dVv4P9kNShxWJBT08Pi/AcM6xLq8vI5XLIZDIIBoOIxWINT7izs7Po/bEfwz2TCd/0mc/UfMzdj30MyvXruPZ3/+5x7fKR8Hg8mJ6ebrhQMU4GXdextrZWlVqtpBlDvP2QQaTkOPZ6vbhy5cqR97Vd9ouItbU1bG5uHrjIIHU19cRBO5/FYZD5S61CUo2Vz220f8ex74TfefVVTAJ4A8APnXKUure3l9bt1IPMjjOZTFXf8+7uLhYWFqDrOm1ZL5fLMBqNWFpagtVqxZUrV9jYiGOgUZcWEzxdTCqVwsLCApLJZN3HPHr0CBMTE+fuh2c0GnH9+vVzt9/diK7rCIfDWFtbQyqVqvs4Mgut1REhlZEHnufh8/kwNTV16oW0pVIJd+/erRI1ZMbVYVHW4ygGJoXLrWyTPHb/OtDIMfuizPDjOA7vete7WuooLBQK2N7extraGoaGhuDxeHDv3j243W4Ae3OyQqEQenp6mNfOMdFI8LDL4i7GarXi+vXreP3112ktxf4QfKWB1nnB6/VibGyMiZ0zQrNeUJqm1e1QavQcURRpu6+maQgGg3QQrCzLUBQFbrf70OMhHo8jHo9DURTY7fYjt6CT90xqh2RZbtoht1gsUn+cTpkdkjRUK48n54T9EOPCWvt2FtvVjwNd17GwsFCzs7ARm5ub0DQNq6ur8Hq91O+JUPn/jJOFCZ4uRxAE+Hw+bG5u0tv2Dxk8Kvd/9mcBRcGtn/zJI2/rMCYmJtDX13fsr8NojmKxeKjYIbOHALS1uBMPGRK9kGWZpm4JCwsLcLvduHr1as3FibR8EyRJQiAQgN1uh81ma9klV9d1xONxWlAN7NVzkMGkpGC4kTggXjeV7eFHgbxes4uzIAgNv4/K7bz1O7+Dwpe/jFc+8Ykz1aZ+3ITDYayvrzfVag7sRb9u376N119/HT6fDyaT6di9nRjNwwTPBWBgYADb29vQNA3T09O0xTYSiVT5hVRCrlSbuSIXAwHwJzCJ3O/3M7Fzxtje3j70Mc2keA6DTDkXBKHutiKRCJaXl+HxeGC1WqsW5f1pXdIxtr6+DovFQutY7HY7XC4XrFYrNE3DxsYGUqkUZFlGf38/7eba2dlBNpul0RFJkmgrOhERHMfR30894dOMWWOz1BtqWYkkSbT9vFF9H4AqEWbw+1EKBOhrXISUFmFlZaVpwQPsFSubTCYMDAwc414x2oHV8FwQyBXu0NAQRkZGkMlk8Oabbzb1XCKIDjtBtktlncb+mo1KxsfHWTj4jLGysoKVlZWa95HFtVNdPbIsH/CZqYckSXC73ejv7wfP83j48GHddFOtIlxSZFp5zBOjQeJ0rKoqPVYbCQGe52nEpxakC60Tn1GjNFUrr3NYNxb5DNqJ9vzWd3wHxJ0dvO93f7ftMTD3P/YxBL7/+9F7SFFxJ2gnqtxKpI3RWVgNDwODg4MIBoOIRCIYHByE2WyG2+2umnRdD+J+S06ixFW22ZMnz/MwGo3I5XIHxMydO3dgsViqilk3NzextbWFQCCAcDhMC2ETiQQTPGeMWqmg/YW7nTrxNyt2gL3oRDAYRDAYrJpZVYtaaZ1aXUhESBAfHuJ+zHEcFEWpu2+apqFcLtcURCQF1t/fD6fTiUwmgxcvXrSVaibpsXodVIelsCr36TAho6oq7UQj54dmoz7v39nBtwH4e9/1XfjIZz/b1HP2Y/76r4e7hajLUQiHw/D5fC3ZITCxczZhgueCIAgCbty4gfn5eTx8+BDXrl3D1NQUvvzlLzdVhFgoFKipGc/zKBaLdI5PvZMzGdDY29sLQRDo41KpFNbX16sWy8q02sDAAA0H9/b24sGDB8hmswiHwygWi0eeSszoDNFoFC9evABQLXIO86Zpl2KxSP1gWuGwx4ui2FSbta7rsNvtSCQS4DgORqORjpY4bLGvnGBemcLSdR3ZbBZ9fX2QJAl2ux2FQqGq3qhZ9neKkfdNvo9mBAnHcU2LLSKMgL3zQ7Mdao8ApAEY7rxzEU5qs0hk5LBtTH3gA03tYyeIxWIXplC722EprQuGpml48OABJEnC1atXsb29jYWFhba3V68V1uPxYHBwsO3pwZUEg0E8e/YMANDT00NN3xinA+le2draqrrtvF7VttJm7fV6EQ6HYXl72GRl4fRRXn9wcJCmTcrlMu7evdvQpPGw7bXbkNCKx87+FKOmafj83/gb8H/sY7j1Dd/QdCqTjK6oXItkWa5KGZ4mZrMZL7300mnvBqNJWEqLQYlEIsjlckilUtjd3UVfXx+MRiOWl5cbeqjUo1QqHQjVd7rWpvIETEYSnNfF9TyTz+cxPz+PdDpd9Z2c1HfRqL7rpIhGoxgbG4PRaATP83j8+PGRo1mFQgEbGxs0bSIIAi5fvoyVlRVEo9GWowtkrlYzIo7UHpFIVSuGgkSkVI6d4N/3PgzdukUjPsRJuNISQ9M0eh+AmsKoWCzSdN9pjoypHBrKOP90pj2AcW4wGo2wWCwAgPn5eaRSKbhcLoyOjrbdLZLP56tSUp3upPL7/XT72WyWip5isYh8Ps/CzSeArut49uwZotEoTWeSBUmSpJbM2dqlU91MR6FcLiOXy8HtdsPpdHYs2pjL5XDv3j0a1bFarbh27Rpu375Nx3IAe79fMgC1EaVSqSn3cXKxUiwW22qNJzO8yDHxjf/r/0pN9jRNQ6lUosNhiWM2+a+qqjVrhRRFoVEqInxOgng0int/9EdV+zE+Pn4mjjtGZ2Df5AXDYrHg+vXruHXrFkwmEx48eIBSqQSn04nbt2+3dXIhRczkxNDpq3BFUaoGORKRs7Ozg2KxWJVaYXQeVVXx/PlzxONxAKAF66VSCaqqolgs0iv641yczkpUb3t7m0ZPfD4f/H5/R7abz+cxOzsL4J26I4vFgpdffhm3bt3C1atX8fLLL2NgYADj4+N1t0OEaDORp051hwF7YpAcCxzHUSGsKAqN7JBJ7OS/+yHFz6fR9v7wd34Hmf/0n+jfExMTbGRNl8EEzwWE4zjYbDbcunULPM9jZ2cHwN48mHbdi4vFIjRNgyRJyGazndxdAHtRI7KwmEwm8DxPB/x1asFh1GZ1dRXBYJD+LYoiZFmGJEmQJIkWkZMFT1GUMyNOjgNN07C4uEj/Hh0d7Zjrd7lcRiqVwr179+ht5Pfq8XgQiUSwvb0Nh8OBwcFBam5HIE0E+2tigNoRMk3TjqUJgNT2EPFCjE4bHRvEH6iSWsLt/3z1Vfz2+9/f8X3+mh/5EXzNpz8NYK9Wy+PxdPw1GKcLk68XGHIVtry8DJ/PB1mWcevWLWxubtLum1ZRFIWmzDoJx3GYmJigpnLAnsja2dlhBl/HTCKRoP9/mBcOWczIVX0nHIQJxHywk9tsl1AohEAgAJfLBUVR0N/f3/ZvphJVVes2EWxtbWF+fh5+v592QBLK5TKy2Syy2SzteCJpYEEQIMsyeJ7H06dPsbu7W7VdIkiOMzVM/ItIioqMwVBVtcq8sTKyQ+p+Kvfr9ddfx7cDiBxjLZfRaKSpymw2i6WlpZbHSzDOJizCc8G5ceMGnE4nlpeXaXi7v78f4+Pjbf3AHQ7HsZ0YRFGE1+ul2+d5HrFY7Fhei7G3SC0tLVGX4mY7eMrlMvL5PLUQ6GQNBIkinoVUw9zcHK258fv9cDqdHdluMpmErutIJBIIh8OIRqNYWVnB/Pw8jcjuFy2CIMBqtcLv98Pn88HhcMBqtcJqtcJkMtF2+OnpaeoqTdJNJxGRq5x4TyI/uq5DVVX62vs9gmodN+9617vwKQCfa+KYeuMHfxCP/tt/a2k/7XY7ZmZmaGTprbfeqtpHxvnm9M8ajFNFURRcuXIFKysrePLkCa5fv05Fj8lkwqNHj1q68jupAkNgTwBNTEyc2OtdNILBINbW1gC847zbzIRnslCRglae5w917m0W0ulD6oX2R3vamcjeLqqqIpfLwWg0QpIkzMzMYGlpCevr60fedjqdxsOHDw+kc0gU5MmTJzCbzbDZbBgbG2taAAqCgJdeegmbm5v0uz0JyHFAkCQJmqZR88JisXjgvZLvdn8X6Lf/5V829ZqcxwPZ623qsYIgYGRkBH19ffTY2draQqFQqKofZJxvmOBhgOM4DA8PHyj+dblcmJ6extOnT5veVicFDylILhaLyGazKBaLEAQBbrcbAwMDEAShqoOF0Vkq63Z0XacC47Ci9P1iQ9M0FAoF6tZ91KJ24ttChA3xbBFFkQ7wbHfsQasEg0HY7XYIggCO4zA0NITt7e0jT0AnIyKIgWAt88BMJoNMJoNCoYArV640JUaBvc9vdHQUDocDjx8/hqZp2FldxeKnPoV3/+N/fKT9bvR+SEGyJElwOp2IRqM0vVXvOcA7LeqV4paM6yCpMfKYynTrSx//eMN94nkeNpsNJpMJQ0NDBzoNyRBYb5OiiXH2uZCCJ5lMdsQQr5vgOK5mO7nf70cwGGw6ddSpOgDi+eJ0OjE+Po5sNoudnR3kcjmsrKxga2sLmqahv78fQ0NDLOTcYWKxGE1lka4sssC0C1mI2on2EL+X/Ys+cTnWdR08z9O6EACHDhslkFq2dojFYnj48CGuXr1KU21Go7EtT6v97I9e1ZvXFY1GEY1GW16YzWYz9TbafOMNcK+/fuR9bgQZvGqz2TA1NYWNjQ2srq4eKoArzynkMwBw4HMgBfOH+XTZ7XYaFWs0yZzYYTQrJBlnnwspeIjYCYfDsNvt9GqBUZvBwcGmBc/6+jp6e3ubWsxKpRJisRh1qxUEgU4lVhQF165do9uxWCy0GLpYLCKRSGB7exsrKyuIxWKYmJiA2WxmwqdDrK+vV13xthKVOUz0kmgPqeE4DFJz0qh+qNbUcTK5/LCRB63M6KpFIpHA7u4uent7Aewdu50QPPupV1js8Xja6igikZ75+Xnc+rZvA/d3/g4Vo53CYDBgdHQUJpMJsixXdYT5fD6sr6831YJO6sEEQaAiVxTFA8cPGYEzPDwMo9GIaDSKtbU1eL1eOBwOuFyupqPCPM/D5/O19oYZZ5oLKXgIHo8Huq4jHo/DZDKdiHnaeaSVE2A+n0cikYDD4Wj4uHQ6jXv37lWdwCtPLo2EiyzL8Hq9sFqt2NjYwO7uLu7evQubzYaenh74fD4mYI+Iy+VCNBpt67nNRPkqoz2NhEi9CeS1yOfzNduYSTqI53nIstzRBb0WIyMjiEajx+IKvV+YWSwWXL58uW2h39vbC5PJhHQ6Da/Xi3Q6jUePHnViVwHsdTzVEw2yLMNqtTY944vjOORyOfpeSeE1MTgE9s7pU1NTNGInCAI2NjYwNTV1JgrdGafLhe/S4jgOTqcTkUiE+tEwqmnVBOzp06eHdvNkMpmqhVEUxZZda8mVnNvths/ngyiKiEajiMVi0DQNkUiEOjOnUinmyNwC/f39h7r51qOVhT6fz0NRlAPRGdJK3cqxRwwwG+1X5XRzQRBgMBg6IkwqRZbZbMbt27fb/vxqQfa58hgOBAK4c+fOkbvgHA4H+vv7acdWJ+A4DoFAAJcuXWr4uOnp6UMvjsixQNyaCaTbi9zmcDhw+fJlKnY0TUMul4PX6z2WiBvj/MEk79vkcjmoqgqfz8fSIvsYHBxELpdrWhAWCgWsra3VdYMtl8vY3t6uuo0UCLaKKIo1Z91omoZUKoVHjx7Rgkez2YyRkRG43W72HTdBuwsp8U5p5jMmk7HJgkaOATJ1vZ3XPozKlEinoj3xeLzKD8psNuP69ev48pe/3LbQrpwgXst9+CQiFn/1Ld8CbmgI7/0//o+mn2OxWHD16tWmzBhFUcSNGzfwla98peawVBIBJMfU/ohfpVnh/vE48/PzCAaDtOuUwWCC521GR0fZIlgHnudx6dIl9PX14dGjR03VO2xsbCCfz8Pr9cLn8yGVSiGTySCRSNB5TARJktrKlZMojizLsNls9PsLhUIol8uwWCwQBAHlchmiKKJUKmF5eRm7u7u4dOkS7XIxm80dvRrvFkwmU9tXxoIgtBQ5IccDceRtB1Kw3OzvuJO/91gshnQ6XWW6SVKvoVDo0OeT9mxg77gmrfdkca/cV57nce3aNdjt9o7tP4G4mJPvTspmUdzn+XPY82dmZlp2bx4fHz+QSttvRFirhqkyVRkOh6uaUci4m+HhYWpW2irxeJy6iZ+k5QbjeOAaXX3cuXNHv3v37gnuDuOsUy6XEY1GEY/Hsbm52dRz9hcXVk5oJifyV155peWIgqZpWF9fRzKZrJppRIYZ8jwPXdextraGyclJWthKXjeRSCCdTsNkMnXMNK6beP78+YFIXLPUKihthqM4Kbf63Hb3sR4cx+HSpUtV4r1QKOD+/fu0k6yWyKq3H8QKoPI+j8cDp9PZ8QG9lTx48KDKXbtZ7HY77VZrh3v37iGVStFp7+QipfIzq0y5VXoyEeFz9erVqt8y8fpph3K5jC996UvQNA2jo6O0oYJxtuE47p6u63dq3Xfha3gYrUF8KSYmJnDr1i24XK5Dn7P/ZC5JEnXiBfauWNs5KfE8j6GhIVy7dq1qnpYsy7BYLDCZTDCbzbh06VLVcFNyArXb7ejr62Nipw5HmYl2GhOmTzNCS2qClpaWqiJbiqJQM08ye0wURVov0ygSUtl5ZjKZ4HA4aKT1OGknGjI6OoobN24cKQpiMBgQjUTwZ9/93bTEYH9HVT6fp+k9VVWpiNR1HYIg4OnTpwf8etolm83S75IMzmWcb1hKi9E2NpsNMzMzyOfziMfj2N7ePnBlWBkeJ6iqWpWLZ4Lj7EFchMkVdblcbika0o74OOk5WZ0qYic1NsDe5xaLxeB2u+n9JpMJvb29VRHRZj5LjuPQ29sLp9NZlbI9blrxKOM4Dh6PBwMDA0faPzJKw+V2472f+ARu3bqFZDKJjY0NajRJXm8/pG6L+PA8fPgQly9fPrIpaWV9VyO/Hsb5gQkexpExGAwIBALw+/3I5XKIRCLIZDIIBAKw2+3I5XK0doa44pKFTdM0KIqCFy9e0NA0iSy4XK4qq3fGyUEcfMln36hFvBPRHNJ23O5zBUFoas4XgXjztFMYvR9ZlquE2vr6OhwOR1URfjufkdvtxvDw8JH3r1WcTmdT4nN4eBg9PT0d6+waHR1FNpvFV33VV0FRFBiNRmxubh56XJCJ78VikR6f9+7dw+3bt48kepaXlwHs2WWMjY21vR3G2YEJHkbH4DgOJpPpwNWQyWTC5cuXsbi4iFAoVBXx4TiOnlj2Qxxkr169eiopkotKsVjE8+fPqxYajuOQz+dp7RWpnVBVtWZxci1hpOs6HSZKUhEkLVEul1sSLJW02r7eaB/bgRybZI5XLBbDvXv3cOnSJZoealbMKYoCl8sFr9d7apFPSZIwMTGBubm5ho8bHBzs2O+StLFXQmaUNePKXSnONE2DqqpYXFzE1atX294nq9WKbDbLDE27CCZ4GCcCx3EYGxuDLMtYXl6mowD2t6JX3q6qKjKZDBM7J0y99BURPUSwtBodqUxN7BcpJ7mgHEfqjBjgkWNVVVU8evQIIyMj6OnpgdfrPXRY5+jo6JFTQ53C6/XCYrEgnU7XfUyhUDjWWXak5gk4aLhI7q+syyP2BqRVPR6PN22PUIuJiQmUy2WEw2EYDAY6A+wsfD+M9mCCh3FicByHwcFB9Pb2Ip/PY2dnB+FwuMoBl6S8yILLBvedPEajEf39/VU1J7quo1gs0uLbZqMxpCBd07Sq5xSLxZrdSZXCgQwqrbVokdvaES+dFtDks6l1++rqKjiOg9/vp2kX4J35XcQbymQynSknYI7jYLfbGwoeMin+OPF6vdjY2AAAOktNVVUqmPdHIclw0vW/+AtwgoAbN25UWQW0AsdxsFgssNvtiEQiiMVisNlsuHbtWkfeG+PkOTu/MMaFQRRFOhuL5O01TcOTJ08OGMGx6M7pUWnZXxm9aFZgkKLgegW6HMdViZ79vivEm6ZcLtMrfVVVaQuyKIooFoswGAw0+tTKfnUCnufrij8yCmF5eZn6QiUSCdhsNkxOTh67WDgqh3VrnUSkY3BwEKurq9B1HVevXsXW1hZ2dnZq1nxVDRktlZB/u5OrXXieRywWQ39/P/r7+1EsFjE/P0/nejHOH2w1YZwqHMfBbDbDarViYmLigMA56qIQj8cxPz8PoLWZYBcdUj9CFhFJkqjzryiK9IQvy3LdotXDxjaUSiXasVeZniAQwaVpGm1FJmk0MvCTXNVXujQfRruFyjzPH3ivh70mWZSXlpYwNjaGr/7qr8b169cbHtfFYrEjxdRHxev11v1ujUbjoSMhOgE5fkgb/9TUFN7znvfQQa374XkeHMdhcXERy5/4BH7zH/7Dtl+b4zhcvnyZdsgpioKrV68ikUjgjTfeOBPfEaM1mOBhnBncbjcuX75ctfC14wlSKpWQz+fx4sULzM3NwWq1olwu48mTJ9jc3KxpYc+oxmq1Ynx8nE65JnU7xWIRpVKJRjVIZ8z+hV8QhEMFpq7rdDp3sVhsO/JCntdsNJA4b7cKqSsjKIrSVLSrUChQB+NaURFd1xGLxRAMBvH8+XO88cYbR/JA6hSCIGBiYqLqtjc/85mqqNpxI0kSbDYbrFYrFdmCIGBwcPDAuUFRFKTTaWSzWah/8icwfuQj+Hu/+qttvW6l0JJlGXNzc9SEkxw77ZgzMk4XltJinCk8Hg/cbjdisRgEQThg198IVVUxNzeHRCJBu4h0XUc2m6WOqalUCqIo4ubNm7ROxGKxgOd55PN5ZDIZlMtlmM1mpNNpGAyGY7HwPw/4/X643W48efIEsVis4WOJpQCwt4Dvb/kmM6HI/YIgVKWhjtKWTtrlm0kzkCgV0JwXTiU8z0NVVVqX1Gxqj6Tenjx5ghs3bkDTNCQSCepfRYbdVn6GZyVl4vF4qoqXQ7/yK3gLwDf+3b97Iq9Ponz7Izocx2FychJzc3MolUoQBAGJRAL/77d8C27+23+Lv//0adsF4LquY3FxkYo9EkXMZrNYW1uDw+GA3++/sOeF8wwTPIwzh8FgoHOEFhcXYbVam5p1FYvFEI1G6d+iKILjOLjdblr4COwtdG+++Sb922q1wul0Ynd3F7lcjs4Oi8fjx+5qe9YRRRFjY2N49OgRra8hU7sra1cqp3nrug5N02j9DYAD7euVizuhnQgPEU2KojS1uJFiV9IiXyqVar4fsm3yGDJolNQdkZb6ZiAiKZ1O40tf+tKh79Nms3XM26YT+P1+Kni++c/+DGazGZOTk0fqgGoFTdNqNi9YrVa88sorSCaTePjwIQwGA+78u3+H7/q+7zvS5xcMBquc2wHg6tWruHfvHhwOB4aGhgAcns5knD1YSotx5hgbG8Pt27cxPj5Oi5kzmQyAdyatv/XWW1heXqYRgkKhUDX3Sdd1WuBKIj31SKVSWFtbo6ku8po7Ozss/YW96de3bt2CLMtUMJBZZQRSZ0NSXiTtVenZUwmpwamsBWrHF0eWZXAch3K5XNf7hyzMJH0GvFNbQ67eSdcYGRRJBBSpDyIdagCqjDObhcyHqnccchxHBeFZcx7fLza8Xi8kScLs7OyxO2PncjkUi0WkUqm6n53ZbIamabDZbEcSO7quY3NzE4qiHHCbJgKZGBBaLBY8efKkpQG5jNOHCR7GmUSWZfT398Pv96NcLuPBgweYnZ3FV77yFWxubmJychK5XA7379/H6uoqHjx4UBXdISmSUqmEe/futbUPZDhpJBLpaGfPeURRFAwODh6Yck9ET7lcPnC1T2YcVc6EIpDoCHlOrecfBkkxAaARpVr7Dbwz/mF/G3Nl6ogIGVKX1Ggxa/V4IOKJFD6LoliV4qtM+e034DttDAYDjXROT09jaGgI8XgciUQCL168ONbXJl46s7OzdedZCYKAqakp3Lhx40iRnWw2C6fTWXM+oCzLuHLlCgqFApLJJL7yla9gd3eXXogxzgcspcU400xMTGBwcBClUgnz8/PQNA0zMzOQZRlDQ0MIhUJYX1+vmV5QVfXIE7GTySSeP38Oo9GIK1eunJnaitOAmK6RhblSEJA6lWYhreWEdq6UBUGgEYZKAUJETS2vlkoqTRCb3XdBEI50VU8iYcDeZ0YKYMltHo/nTLarj4+Pw+FwwOPxgOM4WlS9s7ODoaGhY5s1ZbfbMT4+jhcvXiAUCtWNfu1PQbVDo7T59vY2Xrx4AVEUqyKJmUymrcYKxunAIjyMMw3HcTAYDLBarbh9+zYGBwfp4mYymdDT09NQ0AiCQOcmtVNvQIpTE4kEnj592vb76AbMZjMuX74MSZJoh1K5XK6KTtSilkcO6fpqFfK8WrOzSH0NUD3Q8zCIMG4GUnfUifoNsq3Kz4bUh5w1OI6D1+ulvyGDwUDvO85uJYvFgv7+fvT29iISiXRsHMh+SOqz3n3RaBSXL1/GSy+9RCOJZrO57XEojNOBCR7GuWJwcJCGrTmOw8TERMPFh3i3kMLRdtIm5EQYj8fx5MmTC+2/4fV68a53vYtOA+c47tCTfi2xSTq5CM2IUlmWaeqqVsEs+Z5bFVKaptWsM6pHq23w9SA1ReR9eDyecxMtsNlstFutXSfjVhgeHobBYMDdu3dppLeTPH36FA8fPqx5H/HjcbvdtNaKNDbUSn8xzi4spcU415A6jGaKi/fXidTqFKqkskYE2Fvodnd3USwWYbPZaMu72WyGy+WCw+G4EHN2iD+LxWLB4uJild1/LfFTLpcPCBSTyYRAIACXy0U9agAgnU7jrbfeOlAMS1IJ9bqjWklL1YJEjcj+HgZxf2534SWjVMhnwvP8uZrIbTAYcOvWLSiKUrN2qtPIsgyHw4FcLodIJIJ0Oo2ZmZkjj+Mg4z9CoRAAYGVlBYFAoCqCBbyTJhXeHlexuroKQRDOZPqRUR8meBjnnlYiLuTqvJnaHk3TqFtw5QykfD5fFcbPZDJYX1+HLMu4fPnyiTjQnjYcx6G3txcOhwOPHj1CLperqp1xOp0wGo3gOA6pVAqFQgGpVAp2ux0jIyOw2+01xaHFYsHIyAh1xya0E50jSJJUN/JDarLItitb6RtR2fnVbJqFDMUlnk+V72d4ePjcLZ7HEdnZ2trCysoKDAYDbDYbxsbGqo4rIoQLhQJevHiB6enptl+L2F4QsQPsCZ54PI6rV6/WFVNWq/VIU9gZpwcTPIxzDcdxbdVT7E9H7I9AkDlOld1EpJum3gJXLBbx1ltv4cqVK/B4PC3v03nEZDLhpZdeQjAYRDabhcPhgMvlqpnuaTZt1NPTg3Q6jVwuB7vdDrvdjmAwiJ2dnbrPISnL/REXkjYi3VGV6atyuVwzKkRa3ck+1xLGlZPfyXb3R6XI/pB0HWnZB6rnULndbgwMDBz6uVwEenp6EIvFsLu7i2QyiXK5jMnJSXAch56enqqusEgk0vbrFItFPHr0iHoskdS1LMuYnJw8U4NcGZ2DfauMc4/BYGjZL4e4s5KreZIa0zSN1qVULp7khFjp61K5LQLHcXjy5Alu3759IrUNZwGe5+vONqqkWWFKXHQrsVgs0HW96mq8Ek3TqqItleKGiFVyHzFGrEctQ8VK9qdGiZGhwWCo8h0i3julUqmuSDabzbh06dKFSIU2A8dxuHTpEgRBQDAYxPb2NlRVxfT0NARBgN/vp8K33cJxMmYmlUrRQvdisQhFUXDz5k0mdroY9s0yzjW7u7uHjj2oRWV7deVt+/1hgL00BxE1ZI7Q/qneZAElE79DodCFETwngSRJGB8fryt4gL1uJ1J7QURIrUWxFXFBBO7+SND+tBfx2am8nxS41iuiVhSlI3Uo3QbP8zRVFQwGqQP6jRs3MDk5id3dXWiahnw+31KxeTKZRDweRyQSQSKRoBc5pDaMOLwzuhfWpcU4tywtLeHJkydtPbfWAEkyZmA/+1MVhUIBoijSgk3y/5WFs7u7uxferLDT5HI5OjV7/z+ShtJ1nbZ5kyjPUSHfY6XIIYNPG9Fo/ITdbsft27fP1AiJs8bY2Bj13Umn03j8+DEymQyuXr1KLyaeP3/ecBvk+yfmpUtLS7T+jkTjiABOJBIIBoNt7etRvL4YJwcTPIxzSblcbni1Xwue5+nYAEVRDtSZSJJUd8K3oigQBIE+jxjIkcnsZDwBWRxzudyF9+3pNNFoFIVCgX7uJIqjqioVmpULWCfTRJqm4X98z/dg84jOwhzHYXR0FDdu3LjQJpbNIEkSZmZmqOiJx+O4f/8+Hj9+jIGBAfT19SEcDtcVG5qm4dGjR5idncXs7CyN+EmSRFOV+yN1kUik6rZagrlcLldFlUk9UCqV6sTbZhwjLJbKOHdEIhEsLi7SvHurZmSVtRwkBUIW0HqLJHlOKx1hJ9Gue5HQdZ0uVuS7IiKTTGc/bLFqFXJ88TyP6//8nyPwtjHg/rbyZjAajbh69WpTg3AZe3Ach4GBgSqBoWkanj9/Tuft1fsOMpkMYrEYjf6R75LYJNQqWOd5HtFoFF6vF4lEAo8fP4bJZILZbKb+T/F4HJcuXaLPIV2b4XD43PgoXVSY4GGcK3Rdx8bGBrW2z+fzVV01pVKpbiqJ1NcQ9tdddBJBENDX10dNyhhHJ5FI1Oxyquymq6QTKcVK0dQzPEz/X1XVlsSO1WrFzMwME8Ft4HQ6YTabq+ZWaZqGBw8e4OrVq3VtIMh3V0vY1PruTCYTLSDXNA07OzvUWZx4cuVyOfT29lb59BAfLp/Pd8R3yjhumOBhnCvS6XTV1R4ZEAq8Yx4nimLNKc6lUqlm6/JxUC6X8cYbbwDYcyceGhqCqqp1/WcYjclms3VHGJCp6/u75Uql0pE/60bpEgJpb6/no2M0GnHt2jUmdtqE4zjcuHEDs7OzVaKHmF3Ww2w2H2ouWkml03c8HockSXA6nbTDT1XVmjO7BEGA2+2umiemaRqy2SyCwSA4joPT6aSz6BinBxM8jHPF0tJS3fvIYEfip7H/yk7TtCM78rZDIpHA/fv3oWkaRFGEz+eD3++H3W4/0f04zzSKxO2fxi4IQs3RE61C5rBxHEcXTdLSTgSWpmn4f7/zO2FIJvGBP/3TqucT48rp6WlWr3NEJEnCxMQEZmdnq26fnZ3FpUuXag4VFUUR4+PjmJ+fPzTaZ7FYqtJUu7u72N7eBgAqWq5du1b3+b29vVXH28bGBtbW1qhgXl9fx82bN9lv/pRhgodxrqi3iJFFjwgeknIgCx854Z1G51SlwFJVFVtbW9ja2sL09DQCgcCJ7895xOl0wu12V5nNcRx3oPWbfN+dmHdGxlkoinIg0kNqiBRFwa1f+AUka5giKoqCyclJtsh1CIfDceAYKBaL2NraqjtFnRgZHtbgEAgEaIRO13Ukk0kAwOrqKlwuF65cudKwZX3/eWlwcBADAwOYm5tDKpVCuVxGOp2mx0I2m8XKygpSqRRMJhP6+vpYBOgEYIKHca4YHh6G0+nE4uIigHcMA/c74hIDwcphf82MkzhJtre36VUkCXmLoghN02CxWNjJbx9TU1O4d+8eCoUC/c5J1IwIkP2prUYs/Pqvw/LKK+iZmal5f2UxdL37C4UChiYmIF66VFX0LkkSrl+/zgqUO8zk5CTefPPNqt/x7u4uXnvtNUxNTdGhtpVUpprqURmBi8ViyGQyEAQBVqsVsVisSqzUolQq0S4tm81G02M9PT00vRWPx9Hb24tMJkMjvgDofDCj0Qir1Qq32w2v19tS7V82m0UymYTf72fnjQYwwcM4V8iyjK2tLfo3cbStBbnqJ9EfskCelRNCZU1KIpHAysoK/dtkMmFsbKzmCfyiIssyrl+/TtMaRIhUFhC38t3yNhvEBvOriI9OrQhPJWTWWqUhpdfrZWLnGCBmjbOzs1V1VMViEbu7uzV/L8PDwxgcHESpVMLm5iZcLhei0ShisRh1W64sfDabzbBareB5Hj6fD/F4HNvb2wcET7FYxJMnTzD8djF7NBoFz/PY2trCwMAAHYuyuLgIVVWxu7uLu3fv0vPQfnK5HDUtffHiBXw+HzweD+x2+6Hix2QyQRAELC0tYXBwkNWL1YFrFOK/c+eOfvfu3RPcHQajPvl8Hg8ePKC1E620pIuiCEmSWh5Bcdr09/dXDVBk7F3N3r9//4AIaWXaeTMQU8Bm289JFLFYLOJd73rXgYnbjM4RDAbx/PnzqhT1+Pg4+vv7m95GqVRCNBqF0+k8UGNFrC/u3LmDubk5xGIx3L59+0C0qDKSvP92IlIikQgsFguSySRevHjR8JxVq6lCFEWMjY2hp6en4fspFotIJBKIRqOYmpo69P13KxzH3dN1/U6t+1iEh3FuIMZzxD+nmdQFeVypVDpT6axm2djYQD6fx/DwMEwmE2txx97V7PT0NBYWFmg3FunM62QHHklZNSs2iRnl5OQkEzvHTCAQgNlsxvPnz5FOpwGg5cJwSZJqdl0BoFEYMifOZrNhcXHxQOFyvd9j5e0k6uT1euF2u2k0anNzs67RaSWqquL58+coFAo0mrR/XwVBQDQaxebmJi2+JmNR2DnjHdgnwTg3+Hw+2jWTyWToia4eZFL2/mGP541wOIy7d+/ir/7qr7C6unrau3Mm8Hg8MJvN9PstFAodtxtop9OLDLhkHD9WqxU3b96k6ah639Uvut34nZ/8yZa2zfM8FSoulws2m+3IPjvJZJLO7BoYGMArr7yCsbGxpp+/sbGBUChEj/Pt7W1sbGxgbm4O8/Pz2NrawtjYWFUUam1tjY24qYAJHsa5gaSxAOAvP/YxfOXDH274+GaLV88TrY7T6GYcDkfNzplOtYCTK+RWICNIGCeDIAi4fv06vuZrvgahUAivvfYaHjx4UCUMJn/qp/ANP/IjLW3X4/FgYmKC/u1wOI4sZPeLD47j0N/f33Stl6qqmJubw2uvvYZIJAKv1wuO42hx9Y0bN6pqkUiqjYmed2A1PIxzBbGVX1hYQGJ3FwN1rpDaGTlxXviqr/oqNnQS75g77v+eJUlCsVisKmSu9M1phXaOI5/Ph8uXL7f0HMbRKBaLCAaDWF5epou73W6nE9CDwSCMRiOcTidGR0dPJc1DLA72s7i4iPX1dQC1a3j2Mzg4iMHBwQPDj2uh6zp2d3dht9uRSCSoSOpmGtXwsAgP41zB8zympqbgcrkQ+5M/QSoeP/AY4sPTrXTze2sFQRDg9XoP3E6m3pP6BTJShLSwtwKZo9UKrH7n5JFlGYODg7hz5w4mJycRCARgtVpRKpUQDAahaRoymQw2Njbw7NkzbG5uIhaLnWjko5bYacYjyOVyIRAI0GN3Z2eHjtY5DI7j4PP5oCgKXC4XFUDhcPhCRn1Y0TLj3MHzPKanp3GvWIRe42qI47iG/innGUmSmvIVuSi4XC5sbGwcuJ1EePa3AFcaUjaCCCVg73hqpriUwFJap4fZbIbZbEZvby+9TVVVJBIJZLNZmuaOx+NYXFyE3+/HxMQEOI5DOBymfjiDg4MIBoPI5/Ow2+1IJpNwu910OGipVEI8HkehUIDNZoPNZmtrfzOZDLxeLzweD1KpFFZXV6uO1+vXr1NTRU3TMDs7i2QyiQcPHuCll15q6VxABJPX60UymUQqlYKu67BYLBfmmGWCh3Eu8Xg8+Lqf+7mquVqV5PP5rhQ9FyEk3Qr1vn/C/s+qGbED1B40W8vgshY7OzsYHBxk39MZQRRFuN3uAx49la3jpVIJuVwOXq8XFosFuq7DarXSwcR9fX2QJAmapiEYDGJnZwdGoxG9vb1HmpBe2UbvcDhgtVoRDAZRLBahaVrVfDYS3X7w4AFUVcWzZ88wMjJS12W6ETabjUY9d3d3L4zjOxM8jHPL4OBg3QWv0vGWDBncX4tBZm7xPH8u6n1sNltLXR0XgVavrOvNWWvmeaRFnef5huMrstksotEoM40841SmKiVJwuDgIIC973p1dRXhcBhXrlypEh1LS0sIhUK4devWsaQuHQ5H3envwF4E6+bNm3j8+DGSySSePn2K0dHRthyWQ6EQBEGAzWbryOy58wATPIxzi8PhgM1mo3Nv9lMul+mitH9xItbvpVLpXPzY7XY7rl27dmFCz83i8XhgtVqprX8zHOW71nUdkiQdOq9rcXERZrOZ1fOcQziOw/DwMHp6eg783nw+H4aHh1uuBWsVktaqVT9mNpvx0ksvIZ/P07EUHMe13EU2MDAAYE+gb2xsoL+//8yfB48KEzyMcwvHcRgdHT0wQbkW5XIZsixDEASUy+WqK/zKwaNnEYPBgJmZGSZ2asBxHK1/aBZVVWl6QhRF6Lp+IOJTL+1FHiuK4oFjhqQ/SIEsGTfBOJ/UKjJuJaKo6zpCoRAURWkYtanF2toaDAZD3VQTz/MwmUwwmUxHjiSS7VyEImbWpcU415AJyodBFqpCoVAzndHO7Jnl3/s9bHzhCy0/r1XGxsaY2GlAq+nIcrlMW9Tz+TyKxSIkSYKiKHSRq3elW9nqTsQOiRYWCgUUCgWUSqW2rpS70TfqokI8nJxOJ2ZnZzE3N4dgMIh0Oo1oNIpQKNRQVO/u7p6451a3R3cAFuFhdAFjY2OIxWJNRWg0TYMkSQcWl0KhAEmS6JDRZlq/ebMZ4jF3TPX29sLj8Rzra5xnEokEnTjfLLVmH+0/HsiMpP3DZklUpzLVQI6V/dvM5/MtRQTu3buHmZkZ1oV3ztE0DY8fP65qew+FQlUChhRS17qQUVUVmUwGmUwGyWSy7Q4wxkFYhIdx7jGZTLTgsBnq1V6USiUUCgVa5EwKmusx9M3fjMDt2y3vb7P09PTQllnGQUqlEh4/ftxyKJ4Imf1ippJyuQxFUQ7cTyKBzYwraaVDMB6PI5/PUwM6xvklFAohGo3WPS5FUcTVq1frRm0raxIjkcix7ONFhUV4GF1Bs/bsAJqKBJG0F2ljPg3R4XQ6mdhpwM7OzrGlgUi6i6DrOk1dkUGlZK5bveOp2ZqweDyOhw8fAkDdAnzG2aNQKGBtbQ2apmF0dBSSJCGbzeL58+cNn2e32xvW9FQKpWw2S1Ou7FxwdJjgYXQFHo8Hbrf70Cui/WMHDoO0MZ9GfcVFKCI8Cod58NSjlc+VGBCWSqWqrj9FUVAsFhuOnsjlcoduv1gsYmlpiYqjXC5Hp18zzi7BYBAvXryg6cxkMomRkRFkMplDj6/R0dGG91d2gO3u7iIajcLn82FycvLERY+u611lscBSWoyugOM4XL169dACX+Kh0gqnNcqh20wTO4mqqm0LnmbZP5aikmaOoUZtwrquI5PJ4O7du1VRHYPB0FLHGePkCYfDeP78edV5IZPJ4PHjx1heXj70+U+ePEE8Hq8rjPY3VZTLZezs7Bxtp9uE47iuSquduuBp5iqIwWgGjuMwMDCAl19+ueaMJWAvzfDkD/4Ab37sY01vl9TzyLJ86rN3GHskEom2bQSaiZ6Qx+TzeQiCAFmWaQs78E5Rcz1R2tfXV9OBt1gsYmVlBW+++SbefPPNA4vbwMAA7HZ7q2+JcQLouo5YLIb19fUjnQey2SxmZ2cxOztbc/3b3d09cJumaae2Vg4NDSFeY2ZhK5RKpZrv66Q5dcHDfCoYnUZRFFy5cgW3bt2Cy+Wquo/jOIDjgBYGQpJW9lbdeY+KxWI50dc7T6iqCkEQoChKy+kfMlS0HmSERKWoKRaLUFWVnq9KpRJ4nj8Q/TOZTJiYmMD4+HjNbe/u7mJlZaXu8EeDwcBqNc4ghUIBDx8+RCQSQSKR6Mg2E4kE7t+/j+XlZZoyL5VKdSMqp1XQrijKkTrFyuUy9RU6bU69hof9uBnHhc1mw8zMDNbX17G4uEhvv/Kt3wp867e2tU1Ss3Hcx60oipBl+Vhf4zyjKApUVaURlsq6h8qrbzJPrbIri+d5CIJAbyNmlOVymdZ41aPye6+M7phMJoyPjx9aaN6ohV6W5bbmIjGOn2AwiFgs1vE0aqlUwurqKiKRCHp6ehCJROpGLre3t2lx9EnT6AKhEfPz8yiXy5ienj4Ta/2pR3gYjOOmv78f165dg8FgOHJKqlgsnkhB6UWweT8KFoulasaRqqr0HxEv5XKZ2gwQV2WyWJAOLFVV6f9rmnaoiWGt48diseDGjRtwuVyHfmfk9VVVxec//nFsVghxFtE7m2iaduSUzmGk02ksLCwgGo0eui/nCavVCrfbfWbOZUzwMLoejuPgdrvB8zwdC9AuJBpw3BxlAvNFQBRFXLt2raUTKfn+Sfqg0nqAUPn/T//8z/Ha935v1Tb2f/e9vb24efNm09E40u3y5Ctfwd/8/Ofx4j/8B3qfz+dr+r0wTo7FxcVjL5BvlqWlpRNPrR+Fnp6eM3Vcn3pKi8E4KYxGI7LZLARBOLTzStd1GAyGKtdlMivpuAWPKIpM8DSB2WyG0+k89KqY0Egcqap6IF1p9PsR3TfLqDK0Pzg4iJGRkZZEF6mFuP6e9+AzP/IjeOkDH6DbZY7aZwNd11EoFJDP5xEKhbC1tXXau0TZ2dnB7u4uhoeHMTAwcGYiJ+cFJngYF4ZWIjsGg+FAeoP4rhw3ly5dYvU7TdLf39+04DksnUm+X5ICG75+HcPXr1c9plIo+/3+lhecyvbir6qoI3O73cc+gZtRHzJfjcy5Ou4U1lHQNA1LS0uIx+OYmZnp+Pa3t7eRyWQwOjradu3OWaW73g2D0QBSe1Mul8+sqZ/L5eoak6+TwOl0Nt1B0ow4Ia62tVAUpaqGYm1trbmdfJt0Ol1VtPz5T34ST+/eBcDSWaeFrutYW1vDa6+9htdffx3z8/NnWuxUEo1Gj8Wfh0xon52dbXkw71mHCR7GhYFcnZNZWcBeEaksy1QMSZIEQRAaOisfp1ja30bPaAzHcZienj40OqIoStPfGxkkS+z8idv2/pN/KBRqyZRyYWGhSjDlf+M38PSf/TPwPM++9xOmVCohFovh3r17WFpaOrcmn6urqx0/H3Ech/HxcfT29uKtt97qWBv+WYDFUBkXhkrjLuKiq6oqdF2nNTuFQgG6rtPRAfspFArgeZ4+ppNXQFeuXKlrmMioj8lkgs/na1hroWla0+NBOI6DIAj0WABqi1xd15FIJNqOyH39n/85RFGEy+VioyROmO3tbayvr5/KyJhOks1m8dprr2F0dBQ9PT0d3XYgEIDH4zk1p/njgAkexoWhUvAUCoUq52SO46h4If9vMBiQz+cPpELIc45ysuQ4Dn6/n6ZkeJ5ndTtHoFPeJIIgQNf1qsGhHMfVTYe1UtPl9XqrrpZJVIpFd04eXdfPvdghlEolbG5uIhAIdLyIWRTFrqot6553wmA0oFQqVV2pkChNPYjoURSlSgxVUi6XqbdPs62iJBrh8/lgMplafyOMmnQyQlLL66Te99uKL0q9kRFM8Jw8Z6nzqhOk02k8f/4ck5OTXVdo3EmY4GFcCCqt/EVRpJ04h0EWOkEQIAhC1cJHhFCzkZmhoSEMDQ2xE9IxcFgb/2FXvqIoUp+eelR6MHk8HvT397dkuW80GumxR3C73WfCcv8iUZm+7iaCwSAKhQKuXr3KUqR1YIKHcSGoN7uoWYhz7/4FC2iu+2dkZARDQ0NH2gdGfWw2Gy0wrgWZn0VELvFUImmNRmJH13VwHAdVVeH1ejE2Nlbl8twsoiji5ZdfxubmJnK5HPr7+5nf0inAcRxGRkYwOzt72rvScWKxGF68eIGpqanT3pUzCbvUZFwIKmsyjkKtK6dSqdSwlsNsNmNwcLAjr8+ojSAIh86hInU+siyjWCyiUCjQeVqNIjvkuxVFEYFAoC2xQ5BlGSMjI7h8+TIVaYyTRdd1BIPBrq2ZC4VCZ9Z247RhgodxIahc0NpNKdWKApDJ2o0g7c2M48XhcDS8/yiddaQ1/TzZ+jNqQyKB3fpdnrd5WycJEzyMCwFZ5Hieb9tzo1wuQxAEKIoCRVHoAki6uerRTW2dZxm/399QzJKoTqvikwyMzWazrMC4SxgfHz/tXThWOhXR7jaY4GF0Pfl8HrFYDLquH2n4J6njKBQKNB3C8/yhi2grha2M9lEUBZcvX4aiKFVt6sRn6ShomgZJkk5ktAjj+JEkqWvFq67rXdeF1imY4GF0PTs7OzQl0SnvDUmSUC6XDw0f22w2DA8Pd+Q1GYfj8Xjg9/tRLpdpJI7Mx2oGkvYijtsEURRx6dIllprsIsbHx7v2+2RdWrVhgofR9ZAff6fawSsdmhshyzJmZma6tjjyrDI0NARRFGkkrpVaDdKuXCwW6WIoiiJu3brFZpx1GSaTqa430nmHHau1YYKH0fV0soaGLILNiJjx8fGucik9LwiCgNHR0baeWy6X6RwtctyMjo4yk8gupRtTlKIowmKxnPZunEmY4GF0PfF4vMpz5ShIkoRSqYR8Pg9RFOsOpQwEAmwu1ilCxna0iqqqVa7cfr+/4zOKGGeH/v7+096FjsO6QuvDLj8ZXY/RaEQikUC5XD7yiYA8nxQwq6pKa0RIPc/g4CBGRkbYSecUIROf33zzzba30dvbi4mJCfY9djFWqxV2u/3IE8EVRYHL5YLT6YTBYKDDZ9fX15HNZk/U1dnj8ZzYa503mOBhdD2SJHWsaLlWNIe0LXMch7Gxsa68ajyPHMVnZWRkBIODg0zsXADaSTubzWYMDAzAYrGA4ziYTKYDx4rZbIbL5YKu64jFYnj+/PmxCx9Zltn5pwFM8DC6nng8DqAzRcv1Wtp5noeiKCz9cYZoZ4K6y+XC8PAwsxK4IOi63rJnze3bt1saCcJxHFwuFyYnJ/HkyRMaCVYUBRaLBZFIpKXXr4fBYMCdO3dY3WAD2CfD6GqKxSKSySSA+mKlWUikqN59t27dYu2gZwiDwXDAHZnjOHg8HphMJsiyjFwuR4uUvV5vWyKJcX6Zm5tDJpM5cDvP8/D7/bSoOZ/PQ5ZlWCyWtuefud1uTExM4Pnz53A4HJicnMSzZ8+OtP+VDA8PM7FzCOzTYXQ129vb9P9rDf5sBV3XMTY2BlmWEQqFkMvlkM/n4XA4MDIywk42ZwxRFHH79m1sbGwgm81iYGAAVquViVIGZWpqChaLBaurq9A0DRzHob+/H36//1g6nZxOJ9xuN6anp5FOpyEIwpHPS8BenaLf7+/QXrYGaQo5Dy3+XCMvkTt37uh37949wd1hMDqHpml4/fXX6RV+o2narTA1NcVSVwxGF6GqKm1qOCnfrHg8DpPJhPv37x95FMTk5CR6e3s7tGetoWkaZmdnYTAYzoQ5J8dx93Rdv1PrPtaWzuhaotEoFTuCIHRsgnAymWTTiBmMLoJYTJykSajD4YAsy3j55ZcxNTXV9naMRuOpXoDxPI8rV64gmUwinU6f2n40AxM8jK6lMp3VyQnC29vbNfP+DAaD0So8z6Onp6ft2iCn03nqURVFUTAzM4OFhQVsbm6e2QtCJngYXcnGxkZV90MnBU8gEGBOpgwGo6MEAoGWn8Pz/JmpnSGNAAsLC7RRRFXVM3VxyKosGV3J/sLUTnjw8DwPj8eDiYmJI22HwWAw9tPX1wdZlrG7u4t0Oo1sNgsA1LF9d3cXwF4Ka3R0FG63u2PzATsFadx48OABDAYDyuUyfD7fmTlnMsHD6EocDgf9f1mWW+6CcDqdMJvN4DiOTtx2Op2sE4vBYBwbXq+XCpxCoYBUKkUHgUajUUSjUQwPD59J+4R4PI5oNEr/zufzkCTp1LrHasHO3oyuJJfLVf3dakrLZDJhfHy8k7vEYDAYTUMutAhut/tMT0GPx+NVtTuBQADT09OnuEcHOVvxMAajQ2xsbND/byeV1Y1TlBkMBqPTEEPX4eFh3LlzhxZQn6XIDoFFeBhdCSmU43m+qeiO2+2G0+lELBaDoihn+kqKwWBcLB4+fAhZluFwOGh9z+TkJAwGw6nuVyKRwJMnT3D58mVomoZ8Pg9d1zE4OAin03mq+1YLJngYXUepVKJD+pqdkG4wGNDf34++vr5Tb/FkMBiMSrxeL54/f45gMEhvKxQKpyp4SqUSXrx4gWKxiNnZWXi9XkxPT9PJ8WcRltJidB3JZBIulwsDAwNQFKUpT4jt7W08evQIa2trJ7CHDAaD0TyBQKCq89RkMp2qNYau61hcXEQqlQKwVwIwPj4OQRDgdrvP7EUjEzyMroO0QpJREs38+BRFwdDQEIaGhk5gDxkMBqN5OI7D8PAwAMBisWBsbOzUZsIVi0XMzc3RaJPBYMD169fPRd0jS2kxug6PxwOe53H//n0Ui8WmPHjGx8dhs9lOaA8ZDAajNfr7++HxeCBJ0qnYY+Tzeezs7GB9fZ3afNhsNrjdbhiNxhPfn3ZggofRdczNzaFYLCKVSoHjuKbMuZhzMoPBOMtwHHcqwkLXdWxtbeHFixdV5QF+vx+XLl068f05CkzwMLqOZDJJh4YCqFvDYzKZMDY2dsDvgsFgMBh7LC8vH6ht5DjuXKb/meBhdBXBYLBK7AD1Bc+VK1dgNptPYrcYDAbj3KGqKoLBIK2HBPasPkZHR2EymU5571qHCR5G11AsFhGLxWAwGJDP5wHs/TiLxWLNwuX19XVMTEycWvEfg8FgnGVEUcS73/1ulMtlhEIhqKoKr9d76v4/7cIED6NrkGUZfX19SCQS9DZJkuoaDwaDQezu7mJ4eBgDAwMntZsMBoNxrhAEAT09Pae9G0eGCR5G11Aul/Hw4cOqQaHE9rweoihWDRplMBgMRnfCBA+ja4jFYlViR5Kkhu3o/f39GB0dbaqLi8HoNlZXV2G1WmGxWCDL8mnvDoNx7LAzPaNrqKzFEUXxQPFyJYqiYGRkhIkdxoUlFAphcXER0Wj0tHeFwTgRWISH0TVUtpY3GifR39+PgYEBVqzMuNC89NJLp70LDMaJwi5vGV2DyWSCz+cDsFe7I0kSJEmqeowsy9R7p1mIeCoUCk3N5WIwGAzG2YNFeBhdRWXNjqqqB4QNcWBuZoyEruvY2NjA1tYWRFFEJpPBwMAARkZGOr7fDAaDwThemOBhdBXDw8MolUpIp9MA9ua/8DwPnudhMplgtVoRj8dhtVoPHSrKcRxsNhsCgQAKhQJkWWbFnQwGg3FOYSktRldhs9kwPDyMnp4eGAwG6hBKJg1vbW1hd3e3qpurEXa7HZIksU4WBoPBOOcwwcPoKjiOg9vthtlsBsdxcDqdcDqdyGQySKVSAPbSWs+fPz90gjqDwWAwugeW0mJ0HRzHob+/H16vF6urqwgGgwfclsPhMGKxGPx+P2w2G5xOJxsgymAwGF0MEzyMrkVRFExOTmJkZASRSATZbBahUAj5fB6iKGJgYAC6riOdTsPlcp327jIYDAbjGGGCh9H1SJKEQCAAABgdHYWqquA4jvnwMBgMxgWCCR7GhUMU2WHPYDAYFw1WtMxgMBgMBqPrYYKHwWAwGAxG18MED4PBYDAYjK6HCR4Gg8FgMBhdDxM8DAaDwWAwuh4meBgMBoPBYHQ9TPAwGAwGg8HoepjgYTAYDAaD0fUwwcNgtEC5XD7tXWAwGAxGGzDLWQajAbquo1QqIRKJYH19HeVyGVeuXIHNZjvtXWMwGAxGCzDBw2DsI5fLoVAoIJPJgOd5KIoCQRAwPT0Nq9UKjuNOexcZDAaD0SJM8DAYb5NKpbCwsIBkMonr16+jr6/vtHeJwWB0EaqqQhAEdtF0SjDBw7jwZLNZbG1tIRgM0knqiqKc9m4xLjiapiGTySCbzaJQKEDXdUiSBEVR4HA4IAjCae/iqVAul7G2toZYLAan04ne3t6mfq+apqFUKqFQKCCXy8HpdEKW5RPY43dYWFiAoigYHR090ddl7MEED+PCUigUsLKygu3tbQCAJEkYGRmBwWBgE9UZp0axWMT8/Dyi0Sg0Tav5GJPJhBs3bpz4gt0OqVQKq6uruHTp0pFEmqZpWF9fx+bmJorFIgAgmUxifX0dPT09GBgYgMFgqHp8oVBANptFJBLBzs5OVdMBz/Po7++H3W4HgLoislwuY2dnB7lcDpIk0fNDPp+nERtFUSBJEk2F53I5lMtl6LoOXddRLBZRLBZRKpXAcRzMZjO8Xi94fq9vSNd1aJpGn0Nu4zgOPM9DEAT6WEb7cOTDrcWdO3f0u3fvnuDuMBgnQ6FQwN27d1EqlQAA/f39GB4eZkKHceKoqoqdnR0qcJLJZFPdgBzHwev1wul00qhP5aKo6zrS6TSMRuOxHde6riOZTCIej0PXdQiCgEKhQCNSuq4jGo1C13UoigKbzQZFUWi0RRRFCIIAVVWhaRpMJhOMRiPy+TxtGCiVSpBlGeFwGPl8vuHnUZkuIr/tZuF5HhaLBRaLBZIkAdg7T4TDYaiq2v6H1GB/eZ5vuvPTbDbj5s2b7Bx1CBzH3dN1/U6t+9gnx7iQSJJEr6TMZjPGxsZYXp1x7BQKBcRiMUSjUcRiMXAcRxf7VtF1HaFQCKFQCMDeMe3xeGA0GqGqKiKRCDKZDABAURQYDAYoigJZlsHzPJxOJ5xOZ1uvm06nkUgkEAqFkEwmm3peoVDA7u5uy6/Xyn4dRZgQsdns+zkquq63ZHORyWTw+PFjeDwe2Gw2WCwWFvVpESZ4GBeCYrEIjuPolVsymaQnx0wmg4cPH6Kvrw9Op/PC1kYwDkfXdaytrSGTycDtdsPv99d9rKqqVNxks1lks9mWow6tUCqVaHp2PyTqUsnm5iZu374Nk8nU9GskEgk8e/YMuVzuSPvKaI94PI54PA5gLyJlNBphNBqpmDWZTDTNubu7C5PJBK/Xy85pb8MED6Pref78Oba3tyHLMm7duoVisYiHDx9WPSYWiyEWi0EURYyPjyMQCJzS3jLOIqSAeG1tjUYpQqEQVlZWYDQaoWkaenp64PF4UCwWsbi4iHA4fMp73ZhyuYxsNttQ8JCoSS6XQzqdxuLiIjPfPCOQY5JE8eqxsLAAl8sFq9VKi95NJlNVvdNFgQkeRtcTjUYB7EV5Hj16hJGRkbopBFVV8fz5c1itVpjN5pPcTcY+isUiQqEQUqkUzGYzVFVFX18f7cjJZrMIhULI5XLgeR6BQADxeBzpdBqlUokuzIIgQJZlSJIEk8mE3t7euulLXddRKBSQz+dpsWsqlaJFsvvJ5XI02hGPx8FxHBrVRZ4FBEHA0NAQfD5fzUUvm83i6dOnyGaz0DTtzL8fRmPK5TJ2d3er0okcx6G3txc2mw0ej+fCRICY4GF0NaQ7gpDJZBCNRtHX14fNzc2az9F1HQ8ePMDMzAxzVG6RVCoFXdebNmgkHSz5fB75fB6ZTAbJZJKaP+5na2sLvb29KBaL2N3drYo21Evn7GdjYwNGoxHZbLaqkwbAkRf48yAOSA2O1WqtEjykwHhhYaFhcTDj/KPrOjY3N7G5uQlRFGkxOfBO8TdBFEVaA2a1Ws+1OGKCh9EWxWIR5XIZiqKcqcK5dDqN9fV1xGIx2gK6fxFKJpO4ffs2stksYrFYze2oqort7W0meOoQjUaxtraGcrkMs9kMn8+HYrGIZ8+eAQCcTidN9RDxQr4Pl8sFk8mEaDSKdDrdUsGuqqpYW1s70r5XRmUuIpqmwWAwwOFwVN329OnTYy0qZpxNVFWlUfDD4DgOJpMJNpsNg4ODMBqNx7x3nYUJHkZb7O7uYmdnB/l8HrIsQxRFSJIESZKgqioGBwdhsVgO3Q7xmugUZFsulwsGgwG5XI7W5pjNZhiNRni9XvrDrSd4gL1FXVVV1gZaQT6fx+bmJjY2NqiQTKVSCAaDVY8jNVG1IF1FjNMjHo9jeXkZoiiiWCwiFosdWgvCYOi6jkwmcy4imbVgZ/JzBDnITrN9+sWLF8hkMrQuQhRFiKIIk8kEp9MJm81Wc/80TUOxWIQkSTQkmkql8PTpUwQCAdhsNsiyjFKpBJPJRLup9kOKKOvdbzabcenSparbyuUySqUSisUiCoUCotEoFhcXabdDPQqFAnZ2di70iAmSYkqlUkgkEmxR7BJOsv2a0V1YrVbMzMzUPQefZZjgOSfouo5wOIxsNguDwQCz2QyTyQSe5xGLxbC2tgZBEGh7osPhgNVqPbCdcrmMeDzetjV9IBDA7OxsTb+LlZUVmq7gOI4KjMp6DI7jYDAYqAjRdR1LS0tV2xEEAXa7HTzP0/yxyWSCKIpYX19HMpmEw+Gg7qRWqxWBQACKoqBUKiGdTiOZTCKRSCCXyx2LadhFgHS3MRiMi40oinA4HDAajRgaGjq3UW/mtHzC6LqObDaLdDqNbDYLjuMgyzKNgIiiCIPBQGtPiB25IAiQJIlGVMhiz3EcjaiUy2Woqgqe55FIJJBKpVAoFCDLMgwGA7U9Jzn8y5cvQ5IkZDIZ2O32phV7qVSidRBLS0s1i0vPKxzHwWKxQFEU9PT0wO12n/YunRr5fB6PHj1iUR0G4wLCcRz6+/thsVjgcrnOTUSHOS0fI2TxNxqNkCQJ5XIZmUwGqVSKdoCoqopSqUQHAZ60j0WxWEQ6na66LZ/P4/79+/RvjuNgNBohyzItSuM4DuVyGfl8HoVCgVqhy7JMq/bHxsYwNzd3ou/nuLDZbBgbG6OzdS4aRIzH43Hs7OywlAeDcUHxeDwYHh5uqg7zPMEEzxFQVRWvv/46yuUyBEGAIAh1/TrOOmSxIwve1tbWae/SsaEoCk0JGgwGGI1G2O32cxumbQfSPZXJZJBOp5FKpZBKpVj6j8G4wHAch+vXr1d18HUTF+cM30Gy2SwdTSDLMp2MyxxITwfSHQaAThYmKUFN08DzPHieR09PD1wuV9cLG5IuJZG5QqGAYrFIZ/eQgYwMBoNRid/v71qxA1xwwTM3Nwej0QiTyQRVVaGqKorFIjRNQ29vL2RZhqqqtPYllUohmUwyU64zhiiK8Hg8UBQFFosFVqv1THkDHTdk6GEikUA4HEYqlTrtXWIwGOeQTCaDYrFI53F1G10veBr5vJjNZiwvL9e8j3WnnB/y+Tx2d3ehaRpUVYXD4cD09PSZKbLLZDK0yLtYLNKuM4fD0bIwy2azSKVSyGazdJvpdPrc+mIwGIyzQyqVwuPHj3H58mUoinKqFijHQVcLnnw+T1uoJUmC1WqFz+eD2+0Gx3F0nkw0GkUkEmnabZJxdhgfH0dvb++ZjOioqorl5eW6IyxEUcTY2Bh6enqa2l40Gj0w9JTBYDA6STKZxFe+8hXawDIxMdFwwOx5oqsFj8FgwO3bt7G5uYloNIpoNIpkMol0Oo3h4WEAgNFoRF9fH/r6+pDL5ejQQdJZVSqVoGkaNE1DuVxGOp1u2Q6fcXzEYjEoigKPx3MsVyPtOkGnUinMzc01HGGgqioWFhbg8/ma8kSSZRkul4v+TdytRVGkdUpk2CaDwWA0g9PpRH9/P3ieB8dxUBTlzI0M6hTMh6cNiL329vZ23at3xsni9/vh9XphNpupm3OzwytzuRz1MyJFz8TLaGFhAYIg4Nq1aw1TZPF4nKaYksnkoS7OAGAymWC1WjE+Pt6x9Fsul8P8/HzDkRkMBoMB7Lkm37p1q6tSV8yHpw1IPYgkSVUHQ6lUqioQZZwNdnZ2sLOzQ/8mVyomk4n6CpGJwJXfJ8dxUFWVRkaI6CG3A2g46gLYO1ai0Sji8TiNCMqyTE0lK32LDAYDTCYTjEZjy07XlZPFS6USLbQnkUjiCZXNZlv89BgMxkXDYrHg2rVrXSV2DoNFeLCXWshkMkgkEggGgygUCrTFnOd5OjE7m82eW58dxh6CIMBsNsNgMFDXajK+ovJfJ8K5RDST35iu6xAEAaIoHjjJlEolRKNR6oVDnqNp2oHWcgaDwegEsiyjt7cXg4ODXZPCahThYYJnH7quI5/P0xoMxsWDTFI3m82wWCwwm811jQnJ8ZJOp5HP5+k/4oNT7/dFRoUQgRWLxZg3DoPBOBUURYHf76cDoNuZs3hWYCmtFiAjFkKh0GnvCuOUIDVamUyGHgeiKGJwcBB+v58Wv5PHtGM4SYwqmacTg8E4bQqFAtbW1rC2tkbnCdpsNlitVpjNZhiNxq4wbD2X70DXdei6fmwhOFVVEY/HYbFYUCqVWCqBAVVVsbS0dGCyO4PBYHQTuq7TUTOVSJKE3t5e9PX10c7Q88apCx5d17G5uYlsNguz2QxZlhGPx1EoFOB0OmG1Wv//7d3NqtpAGAbgUUM4gj9x7UW48CZ68b0DV1m4EEpEcJUY56ySammlh3oaz/A8kEVIMIPZvGS+mS8cj8dQVVWo67ov1Iwxhre3t1AURVgul2GxWITpdPqUl5BlWdhsNv1527bhcDiE8/l8VyTaNQe9rdHo6jYAIBVN04SyLENZliGEn9PyRVGE1WrVdyroPhB0fQq74xU2gh28hme/34fdbve035tMJv3KmNs/PM/zuyLV7vNc27b9S2qapl+W3B3j8fju/l9dr9dQ13VfVFpVlV2aAeDGfD4P2+3205/z0jU8z54qatu23w/ldDo9vLdrMPk3sizrlxXHGPuVMwpNAeCxV2iuPXjgGdJHwtblcul3WQYAvpavV3UEAPBBAg8AkDyBBwBInsADACRv8KLl2WwW1uv10MMAAD5JnudDD2H4wFMURSiKYuhhAAAJM6UFACRP4AEAkifwAADJE3gAgOQJPABA8gQeACB5Ag8AkDyBBwBInsADACRP4AEAkifwAADJG8UY/3xxNPr+H8cCAPAvfsQYv/3uwsPAAwCQAlNaAEDyBB4AIHkCDwCQPIEHAEiewAMAJO8d08F+bV0vVroAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "geometry = []\n",
    "\n",
    "for rec in metadata_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    # Terminal nodes use the \"lat/lon\" attribute        \n",
    "    if rec[1][\"node_type\"] == \"terminal\":  \n",
    "        province = rec[1][\"province\"]\n",
    "        country = rec[1][\"country\"]\n",
    "        if province != NO_DATA_CHAR:\n",
    "            attr = \"province\"\n",
    "            name = province\n",
    "        else:\n",
    "            attr = \"country\"\n",
    "            name = country\n",
    "        lat = rec[1][\"lat\"]\n",
    "        lon = rec[1][\"lon\"]\n",
    "    # Internal nodes use mugration province\n",
    "    else:\n",
    "        name = rec[1][\"mugration_province\"]\n",
    "        lat  = rec[1][\"mugration_province_lat\"]\n",
    "        lon  = rec[1][\"mugration_province_lon\"]     \n",
    "    \n",
    "    if attr == \"province\":\n",
    "        size = latlon_province_df[\"size\"][name]\n",
    "    elif attr == \"country\":\n",
    "        size = latlon_country_df[\"size\"][name]\n",
    "    point = shapely.geometry.Point(lon,lat)\n",
    "    geometry.append(point)\n",
    "\n",
    "\n",
    "metadata_gdf = geopandas.GeoDataFrame(metadata_df, crs=\"epsg:3857\", geometry=geometry)\n",
    "\n",
    "# Add geometry size\n",
    "for rec in metadata_gdf.iterrows():\n",
    "    sample = rec[0]\n",
    "    name = rec[1][\"mugration_\" + attr]\n",
    "    size = latlon_df[\"size\"][name]  \n",
    "    metadata_gdf.at[sample, \"geometry_size\"] = size\n",
    "    \n",
    "\n",
    "\n",
    "# Test map\n",
    "\n",
    "# Generate a figure with two axes, one for CartoPy, one for GeoPandas\n",
    "fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=(10, 5))\n",
    "world_polygons.plot(ax=ax, color=\"grey\", alpha=0.5)\n",
    "metadata_gdf.plot(ax=ax, c=\"red\", markersize=\"geometry_size\", ec=\"black\", lw=0.25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "compound-fifteen",
   "metadata": {},
   "source": [
    "## Add Sub  and Spread Rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "chinese-anxiety",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.         0.18930392]\n"
     ]
    }
   ],
   "source": [
    "# 1. Get Branch Lengths in Num Substitutions\n",
    "\n",
    "metadata_gdf[\"branch_length_sub\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"branch_length_time\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"sub_rate\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"geo_rate_sub\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"geo_rate_time\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "\n",
    "for c in divtree.find_clades():\n",
    "    branch_length_sub = c.branch_length * SEQ_LEN\n",
    "    branch_length_time = timetree.common_ancestor(c.name).branch_length\n",
    "    walk_path = [divtree.root] + divtree.get_path(c)\n",
    "    # Root\n",
    "    if len(walk_path) == 1:\n",
    "        parent = walk_path[0]\n",
    "    else:\n",
    "        parent = walk_path[-2]\n",
    "    coord1 = metadata_gdf[\"geometry\"][parent.name]\n",
    "    coord2 = metadata_gdf[\"geometry\"][c.name]\n",
    "    geo_dist = geopy.distance.great_circle([coord1.y, coord1.x], [coord2.y, coord2.x]).km\n",
    "    # 1. Rate as KM / subs / site\n",
    "    geo_rate_sub = geo_dist / branch_length_sub if branch_length_sub != 0 else 0\n",
    "    # 2. Rate as KM / time\n",
    "    geo_rate_time = geo_dist / branch_length_time if branch_length_time != 0 else 0\n",
    "    # 3. Rate as subs/year\n",
    "    rate_sub = branch_length_sub / branch_length_time  if branch_length_time != 0 else 0\n",
    "    metadata_gdf.at[c.name,\"branch_length_sub\"] = branch_length_sub\n",
    "    metadata_gdf.at[c.name,\"branch_length_time\"] = branch_length_time\n",
    "    metadata_gdf.at[c.name,\"rate_sub\"] = rate_sub\n",
    "    metadata_gdf.at[c.name,\"rate_geo_sub\"] = geo_rate_sub\n",
    "    metadata_gdf.at[c.name,\"rate_geo_time\"] = geo_rate_time\n",
    "     \n",
    "#display(metadata_gdf)\n",
    "\n",
    "rates = list(metadata_gdf[\"rate_sub\"])\n",
    "rates_ci = np.array(\n",
    "    np.percentile(\n",
    "        np.array(rates), (0.05, 95), axis=0\n",
    "    )\n",
    ")\n",
    "print(rates_ci)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "initial-scheduling",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n\\n# ----------------\\nfig, axes = plt.subplots(1,4, figsize=(12,3))\\nfig.subplots_adjust(wspace=0.5)\\n\\nfor ax in axes:\\n    ax.set_xlabel(\"Rate (sub/year)\")\\n    ax.set_ylabel(\"Density\")\\n\\nax_i = -1\\n\\n# Empirical Data\\nax_i += 1\\nax = axes[ax_i]\\nstddev = 0.01\\nmean = 0\\nloc = 0\\n\\nrates = list(metadata_gdf[\"rate_sub\"])\\n#sns.distplot(rates, ax=ax)\\nsns.kdeplot(rates, ax=ax)\\n#sns.histplot(rates, ax=ax)\\nax.set_xlim(0,0.2)\\n\\nshape, loc, scale = scipy.stats.lognorm._fitstart(rates) \\nprint(loc, shape, scale)\\nshape, loc, scale = scipy.stats.lognorm.fit(rates) \\nprint(loc, shape, scale)\\n\\n# Strict clock (mean 0, stdev ~0)\\nx=np.linspace(0.95,1.05,100)\\nax_i += 1\\nax = axes[ax_i]\\nstddev = 0.01\\nmean = 0\\ndist = scipy.stats.lognorm([stddev],loc=mean)\\nax.plot(x, dist.pdf(x))\\nax.set_xlim(0.95,1.05)\\nax.set_title(\"Strict Clock\")\\n\\n# Relaxed Clock (mean 0, stdev 0.2)\\nax_i += 1\\nax = axes[ax_i]\\nx=np.linspace(0,2,100)\\nstddev = 0.2\\nmean = 0\\ndist = scipy.stats.lognorm([stddev],loc=mean)\\nax.plot(x, dist.pdf(x))\\nax.set_xlim(0,2)\\nax.set_title(\"Relaxed Clock (stddev=0.2)\")\\n\\n# Relaxed Clock (mean 0, stdev 1)\\nax_i += 1\\nax = axes[ax_i]\\nx=np.linspace(0,5,100)\\nstddev = 1\\nmean = 0\\ndist = scipy.stats.lognorm([stddev],loc=mean)\\nax.plot(x, dist.pdf(x))\\nax.set_xlim(0,5)\\nax.set_title(\"Relaxed Clock (stddev=1)\")'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "\n",
    "# ----------------\n",
    "fig, axes = plt.subplots(1,4, figsize=(12,3))\n",
    "fig.subplots_adjust(wspace=0.5)\n",
    "\n",
    "for ax in axes:\n",
    "    ax.set_xlabel(\"Rate (sub/year)\")\n",
    "    ax.set_ylabel(\"Density\")\n",
    "\n",
    "ax_i = -1\n",
    "\n",
    "# Empirical Data\n",
    "ax_i += 1\n",
    "ax = axes[ax_i]\n",
    "stddev = 0.01\n",
    "mean = 0\n",
    "loc = 0\n",
    "\n",
    "rates = list(metadata_gdf[\"rate_sub\"])\n",
    "#sns.distplot(rates, ax=ax)\n",
    "sns.kdeplot(rates, ax=ax)\n",
    "#sns.histplot(rates, ax=ax)\n",
    "ax.set_xlim(0,0.2)\n",
    "\n",
    "shape, loc, scale = scipy.stats.lognorm._fitstart(rates) \n",
    "print(loc, shape, scale)\n",
    "shape, loc, scale = scipy.stats.lognorm.fit(rates) \n",
    "print(loc, shape, scale)\n",
    "\n",
    "# Strict clock (mean 0, stdev ~0)\n",
    "x=np.linspace(0.95,1.05,100)\n",
    "ax_i += 1\n",
    "ax = axes[ax_i]\n",
    "stddev = 0.01\n",
    "mean = 0\n",
    "dist = scipy.stats.lognorm([stddev],loc=mean)\n",
    "ax.plot(x, dist.pdf(x))\n",
    "ax.set_xlim(0.95,1.05)\n",
    "ax.set_title(\"Strict Clock\")\n",
    "\n",
    "# Relaxed Clock (mean 0, stdev 0.2)\n",
    "ax_i += 1\n",
    "ax = axes[ax_i]\n",
    "x=np.linspace(0,2,100)\n",
    "stddev = 0.2\n",
    "mean = 0\n",
    "dist = scipy.stats.lognorm([stddev],loc=mean)\n",
    "ax.plot(x, dist.pdf(x))\n",
    "ax.set_xlim(0,2)\n",
    "ax.set_title(\"Relaxed Clock (stddev=0.2)\")\n",
    "\n",
    "# Relaxed Clock (mean 0, stdev 1)\n",
    "ax_i += 1\n",
    "ax = axes[ax_i]\n",
    "x=np.linspace(0,5,100)\n",
    "stddev = 1\n",
    "mean = 0\n",
    "dist = scipy.stats.lognorm([stddev],loc=mean)\n",
    "ax.plot(x, dist.pdf(x))\n",
    "ax.set_xlim(0,5)\n",
    "ax.set_title(\"Relaxed Clock (stddev=1)\")\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "religious-community",
   "metadata": {},
   "source": [
    "## Colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "passing-christopher",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'branch_major': {'0.PRE': '#8000ff', '0.PE': '#4c4ffc', '0.ANT': '#1996f3', '3.ANT': '#1acee3', '4.ANT': '#4df3ce', '2.ANT': '#80ffb4', '2.MED': '#b3f396', '1.PRE': '#e6ce74', '1.ANT': '#ff964f', '1.IN': '#ff4f28', '1.ORI': '#ff0000', '?': '#969696'}, 'branch_minor': {'0.PRE2': '#8000ff', '0.PRE1': '#7018ff', '0.PE7': '#6130fe', '0.PE2': '#5148fc', '0.PE8': '#425ffa', '0.PE10': '#3275f8', '0.PE4': '#238af5', '0.PE5': '#139ef1', '0.ANT1': '#04b0ed', '0.ANT4': '#0cc1e8', '0.ANT5': '#1bd0e3', '0.ANT2': '#2bdddd', '0.ANT3': '#3ae8d7', '3.ANT1': '#49f1d0', '3.ANT2': '#59f8c8', '4.ANT1': '#68fcc1', '2.ANT3': '#78ffb9', '2.ANT2': '#87ffb0', '2.ANT1': '#97fca7', '2.MED0': '#a6f89e', '2.MED3': '#b6f194', '2.MED2': '#c5e88a', '2.MED1': '#d5dd7f', '1.PRE0': '#e4d075', '1.PRE1': '#f3c16a', '1.PRE2': '#ffb05f', '1.PRE3': '#ff9e53', '1.ANT1': '#ff8a48', '1.IN1': '#ff753c', '1.IN2': '#ff5f30', '1.IN3': '#ff4824', '1.ORI2': '#ff3018', '1.ORI1': '#ff180c', '1.ORI3': '#ff0000', '?': '#969696'}, 'country': {'Russia': '#8000ff', 'Lithuania': '#7215ff', 'Estonia': '#652afe', 'Germany': '#573ffd', 'China': '#4a53fc', 'Azerbaijan': '#3c66fa', 'Armenia': '#2f79f7', 'Georgia': '#228bf4', 'Mongolia': '#149df1', 'Tajikistan': '#07adee', 'Kyrgyzstan': '#07bcea', 'England': '#14c9e5', 'Spain': '#22d5e0', 'France': '#2fe0db', 'Nepal': '#3cead5', 'India': '#4af1d0', 'Kazakhstan': '#57f7c9', 'Turkmenistan': '#65fcc3', 'Uzbekistan': '#72febc', 'Iran': '#80ffb4', 'Italy': '#8dfead', 'Poland': '#9afca5', 'Switzerland': '#a8f79d', 'Norway': '#b5f194', 'The Netherlands': '#c3ea8b', 'Kenya': '#d0e083', 'Uganda': '#ddd579', 'Democratic Republic of the Congo': '#ebc970', 'Vietnam': '#f8bc66', 'Myanmar': '#ffad5d', 'United States of America': '#ff9d53', 'Peru': '#ff8b49', 'Canada': '#ff793f', 'Indonesia': '#ff6634', 'Madagascar': '#ff532a', 'Zimbabwe': '#ff3f20', 'Algeria': '#ff2a15', 'Brazil': '#ff150b', 'Bolivia': '#ff0000', '?': '#969696'}, 'province': {'Krasnoyarsk Krai': '#8000ff', 'Panevezys County': '#7b07ff', 'Prnu maakond': '#760eff', 'Irkutsk Oblast': '#7215ff', 'Bavaria': '#6d1dff', 'Altai Krai': '#6924fe', 'Qinghai': '#642bfe', 'Republic of Dagestan': '#6032fe', 'Goranboy District': '#5b39fd', 'Syunik Province': '#5740fd', 'Shirak Province': '#5247fc', 'Gegharkunik Province': '#4d4dfc', 'Samtskhe-Javakheti': '#4954fb', 'Shahbuz Rayon': '#445bfb', 'Samara Oblast': '#4062fa', 'Seun': '#3b68f9', 'Bayankhongor': '#376ff9', 'Inner Mongolia': '#3275f8', 'Sughd Province': '#2e7bf7', 'Talas Region': '#2982f6', 'Altai Republic': '#2488f5', 'Govi-Altai': '#208ef4', 'vrkhangai': '#1b94f3', 'mngovi': '#1799f2', 'Bayan-lgii': '#129ff1', 'Khovd': '#0ea5ef', 'Xinjiang': '#09aaee', 'Issyk-Kul Region': '#05afed', 'East of England': '#00b4ec', 'Valencia Community': '#05b9ea', 'Centre-Loire Valley': '#09bee9', 'Osh Region': '#0ec3e7', 'Naryn Region': '#12c7e6', 'Gansu': '#17cce4', 'Hovsgel': '#1bd0e3', 'Zavkhan Province': '#20d4e1', 'Dornogovi': '#24d8df', 'Tuva Republic': '#29dcdd', 'Jilin': '#2edfdc', 'St. Petersberg': '#32e3da', 'Zabaykalsky Krai': '#37e6d8', 'Heilongjiang': '#3be9d6', 'Tibet': '#40ecd4', 'Gandaki Pradesh': '#44eed2', 'Maharashtra': '#49f1d0', 'Himachal Pradesh': '#4df3ce', 'Kabardino-Balkaria': '#52f5cc', 'Karachay-Cherkessia': '#57f7ca', 'Ningxia': '#5bf9c7', 'Shaanxi': '#60fac5', 'Hebei': '#64fbc3', 'West Kazakhstan Region': '#69fcc1', 'Nagorno-Karabakh Republic': '#6dfdbe', 'Kyzylorda Region': '#72febc', 'Rostov Oblast': '#76ffb9', 'Astrakhan Oblast': '#7bffb7', 'Balkan': '#80ffb4', 'Kyzylorda Province': '#84ffb2', 'Shymkent': '#89ffaf', 'Navoiy Region': '#8dfead', 'Almaty Region': '#92fdaa', 'West Kazakhstan region': '#96fca7', 'Atyrau Region': '#9bfba5', 'Fizuli District': '#9ffaa2', 'Baku': '#a4f99f', 'Kurdistan': '#a8f79c', 'Shamkir District': '#adf599', 'Stavropol Krai': '#b2f396', 'Ingushetia': '#b6f194', 'Mangystau Province': '#bbee91', 'Chechnya': '#bfec8e', 'Aktobe Region': '#c4e98b', 'Atyrau': '#c8e688', 'Qazakh District': '#cde385', 'Republic of Kalmykia': '#d1df82', 'Tatarstan': '#d6dc7e', 'Lazio': '#dbd87b', 'Greater London': '#dfd478', \"Provence-Alpes-Cte d'Azur\": '#e4d075', 'Baden-Wrttemberg': '#e8cc72', 'Vilnius County': '#edc76f', 'Pomeranian Voivodeship': '#f1c36b', 'Brandenburg': '#f6be68', 'Nidwalden': '#fab965', 'Oslo': '#ffb462', 'Catalonia': '#ffaf5e', 'Occitanie': '#ffaa5b', 'North Brabant': '#ffa558', 'Nairobi': '#ff9f54', 'Yunnan': '#ff9951', 'Khnh Ho Province': '#ff944d', 'Lm ng Province': '#ff8e4a', 'Ho Chi Minh City': '#ff8847', 'Guangxi Zhuang': '#ff8243', 'California': '#ff7b40', 'La Libertad': '#ff753c', 'Tamil Nadu': '#ff6f39', 'Saskatchewan': '#ff6835', 'Arizona': '#ff6232', 'New Mexico': '#ff5b2e', 'Colorado': '#ff542b', 'Texas': '#ff4d27', 'Nevada': '#ff4724', 'Central Java': '#ff4020', 'Province de Fianarantsoa': '#ff391d', 'Kirov Oblast': '#ff3219', 'Gujarat': '#ff2b15', 'Fujian': '#ff2412', 'Midlands': '#ff1d0e', 'MSila': '#ff150b', 'State of Mato Grosso': '#ff0e07', 'La Paz': '#ff0704', 'Cajamarca': '#ff0000', '?': '#969696'}}\n"
     ]
    }
   ],
   "source": [
    "out_path_colors = os.path.join(out_dir, \"colors.tsv\")\n",
    "\n",
    "colors_dict = {}\n",
    "\n",
    "for attr in MUG_ATTRIBUTE_LIST:\n",
    "    # Create the color map\n",
    "    attr_key = attr.lower()\n",
    "    colors_dict[attr_key] = {}\n",
    "    for t in full_divtree.get_terminals():\n",
    "        attr_val = full_metadata_df[attr][t.name]\n",
    "        # Remove the letter suffix from branch_minor\n",
    "        if attr == \"branch_minor\":\n",
    "            while attr_val[-1].isalpha():\n",
    "                attr_val = attr_val[:-1] \n",
    "        if attr_val not in colors_dict[attr_key] and attr_val != NO_DATA_CHAR:\n",
    "            colors_dict[attr_key][attr_val] = {}\n",
    "            \n",
    "    # Create the custom color map (pyplot)\n",
    "    cmap = plt.get_cmap(\"rainbow\", len(colors_dict[attr_key]))\n",
    "    # Convert the color map to a list of RGB values\n",
    "    cmaplist = [cmap(i) for i in range(cmap.N)]\n",
    "    # Convert RGB values to hex colors\n",
    "    attr_hex = [colors.to_hex(col) for col in cmaplist]\n",
    "    \n",
    "    # Assign colors to value\n",
    "    for attr_val, attr_col in zip(colors_dict[attr_key], attr_hex):\n",
    "        colors_dict[attr_key][attr_val] = attr_col   \n",
    "        \n",
    "    # Add unknown\n",
    "    colors_dict[attr_key][UNKNOWN_CHAR] = \"#969696\"\n",
    "    \n",
    "print(colors_dict)\n",
    "\n",
    "with open(out_path_colors, \"w\") as outfile:\n",
    "    for attr_key in colors_dict:\n",
    "        for attr_val in colors_dict[attr_key]:\n",
    "            outfile.write(str(attr_key) + \"\\t\" + str(attr_val) + \"\\t\" + str(colors_dict[attr_key][attr_val]) + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "imperial-flooring",
   "metadata": {},
   "source": [
    "## Parse Tree Into Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "regulated-comment",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_posns = get_x_positions(divtree)\n",
    "y_posns = get_y_positions(divtree)\n",
    "\n",
    "# Initialize \n",
    "# Date lower and upper error bars for tips\n",
    "metadata_gdf[\"date_lower\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"date_upper\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "    \n",
    "for c in divtree.find_clades():\n",
    "    # ------------------------------------------------------------\n",
    "    # Coordinates for the divergence tree\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    metadata_gdf.at[c.name, 'divtree_coord_x'] = coord_x\n",
    "    metadata_gdf.at[c.name, 'divtree_coord_y'] = coord_y\n",
    "    \n",
    "    # ------------------------------------------------------------\n",
    "    # Root to tip distance (same as div x)\n",
    "    metadata_gdf.at[c.name, \"rtt_dist\"] = coord_x\n",
    "    \n",
    "    # ------------------------------------------------------------\n",
    "    # Branch Major and Branch Minor and Colors\n",
    "    branch_minor = metadata_gdf[\"mugration_branch_minor\"][c.name]\n",
    "    if branch_minor != NO_DATA_CHAR:\n",
    "        # Remove letters at end\n",
    "        while branch_minor[-1].isalpha():\n",
    "            branch_minor = branch_minor[:-1]    \n",
    "    metadata_gdf.at[c.name, \"mugration_branch_minor\"] = branch_minor\n",
    "    branch_major = metadata_gdf[\"mugration_branch_major\"][c.name]\n",
    "    branch_major_color = colors_dict[\"branch_major\"][branch_major]\n",
    "    metadata_gdf.at[c.name, \"branch_major_color\"] = branch_major_color    \n",
    "\n",
    "    # ------------------------------------------------------------\n",
    "    # Date lower and upper error bars for tips\n",
    "    date_mean = metadata_gdf[\"date_mean\"][c.name] \n",
    "    if date_mean != NO_DATA_CHAR:\n",
    "        metadata_gdf.at[c.name, \"date_lower\"] = date_mean - metadata_gdf[\"date_err\"][c.name]\n",
    "        metadata_gdf.at[c.name, \"date_upper\"] = date_mean + metadata_gdf[\"date_err\"][c.name]\n",
    "\n",
    "x_posns = get_x_positions(timetree)\n",
    "y_posns = get_y_positions(timetree)\n",
    "\n",
    "\n",
    "for c in timetree.find_clades():\n",
    "\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    metadata_gdf.at[c.name, 'timetree_coord_x'] = coord_x\n",
    "    metadata_gdf.at[c.name, 'timetree_coord_y'] = coord_y\n",
    "    \n",
    "    timetree_date = int(float(metadata_gdf[\"timetree_date\"][c.name]))\n",
    "    timetree_date_calendar = CURRENT_YEAR + timetree_date\n",
    "    metadata_gdf.at[c.name, 'timetree_date_calendar'] = timetree_date_calendar\n",
    "    \n",
    "    timetree_CI_date = metadata_gdf[\"timetree_CI_date\"][c.name]\n",
    "    if timetree_CI_date == NO_DATA_CHAR:\n",
    "        timetree_CI_lower_err = 0\n",
    "        timetree_CI_upper_err = 0\n",
    "    else:\n",
    "        timetree_CI_split = [int(float(d)) for d in timetree_CI_date.strip(\"{}\").split(\",\")]\n",
    "        timetree_CI_calendar = [CURRENT_YEAR + d for d in timetree_CI_split]\n",
    "        timetree_CI_lower_err = timetree_date_calendar - timetree_CI_calendar[0]\n",
    "        timetree_CI_upper_err = timetree_CI_calendar[1] - timetree_date_calendar\n",
    "    metadata_gdf.at[c.name,\"timetree_CI_lower_err\"] = timetree_CI_lower_err\n",
    "    metadata_gdf.at[c.name,\"timetree_CI_upper_err\"] = timetree_CI_upper_err   \n",
    "\n",
    "#display(metadata_gdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bibliographic-learning",
   "metadata": {},
   "source": [
    "## Create Custom Columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "former-bouquet",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_gdf[\"country_date_strain\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"province_date_strain\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "\n",
    "for c in divtree.get_terminals():\n",
    "    strain = metadata_gdf[\"strain\"][c.name]\n",
    "    country = metadata_gdf[\"country\"][c.name]\n",
    "    province = metadata_gdf[\"country\"][c.name]\n",
    "    date = metadata_gdf[\"date\"][c.name]\n",
    "    if type(date) == str:\n",
    "        date = date.strip(\"[]\")\n",
    "    metadata_gdf.at[c.name,\"country_date_strain\"] = \"{} ({}) {}\".format(country, date, strain)\n",
    "    metadata_gdf.at[c.name,\"province_date_strain\"] = \"{} ({}) {}\".format(province, date, strain)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "editorial-collins",
   "metadata": {},
   "source": [
    "## Calculate Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "prime-analyst",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_gdf[\"root_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"clade_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "\n",
    "branch_mrca_dict = {}\n",
    "\n",
    "for c in divtree.find_clades():\n",
    "    sample = c.name    \n",
    "    # Clade MRCA Dist    \n",
    "    branch_major = metadata_gdf[\"mugration_branch_major\"][sample]\n",
    "    if branch_major in branch_mrca_dict: continue\n",
    "\n",
    "    branch_df = metadata_gdf[metadata_gdf[\"mugration_branch_major\"] == branch_major]\n",
    "    branch_mrca = branch_df[branch_df[\"node_type\"] == \"internal\"].iloc[0].name\n",
    "    branch_mrca_dict[branch_major] = branch_mrca\n",
    "    \n",
    "\n",
    "for c in divtree.find_clades():\n",
    "    sample = c.name\n",
    "    # Root Dist\n",
    "    root_dist = divtree.distance(divtree.root, c)\n",
    "    \n",
    "    # Clade MRCA Dist    \n",
    "    branch_major = metadata_gdf[\"mugration_branch_major\"][sample]\n",
    "    branch_mrca = branch_mrca_dict[branch_major]\n",
    "    branch_mrca_dist = divtree.distance(branch_mrca, c)\n",
    "    \n",
    "    metadata_gdf.at[sample, \"root_rtt_dist\"] = root_dist\n",
    "    metadata_gdf.at[sample, \"clade_rtt_dist\"] = branch_mrca_dist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brutal-oracle",
   "metadata": {},
   "source": [
    "---\n",
    "# 4. EXPORT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "otherwise-screw",
   "metadata": {},
   "source": [
    "## Time Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "caring-affiliate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_timetree = copy.deepcopy(timetree)\n",
    "\n",
    "for c in out_timetree.find_clades():\n",
    "    if not c.confidence:\n",
    "        c.confidence = 0\n",
    "\n",
    "metadata_to_comment(out_timetree, metadata_gdf)    \n",
    "out_timetree_nex_path = os.path.join(out_dir, \"all.timetree.nex\")\n",
    "Phylo.write(out_timetree, out_timetree_nex_path, \"nexus\")\n",
    "\n",
    "for c in out_timetree.find_clades():\n",
    "    c.comment = None\n",
    "\n",
    "out_timetree_nwk_path = os.path.join(out_dir, \"all.timetree.nwk\")\n",
    "Phylo.write(out_timetree, out_timetree_nwk_path, \"newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "running-swimming",
   "metadata": {},
   "source": [
    "# Divergence Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "traditional-sheep",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_divtree = copy.deepcopy(divtree)\n",
    "\n",
    "for c in out_divtree.find_clades():\n",
    "    if not c.confidence:\n",
    "        c.confidence = 0\n",
    "\n",
    "metadata_to_comment(out_divtree, metadata_gdf)    \n",
    "out_divtree_nex_path = os.path.join(out_dir, \"all.divtree.nex\")\n",
    "Phylo.write(out_divtree, out_divtree_nex_path, \"nexus\", format_branch_length=\"%1.10f\")\n",
    "\n",
    "for c in out_divtree.find_clades():\n",
    "    c.comment = None\n",
    "\n",
    "out_divtree_nwk_path = os.path.join(out_dir, \"all.divtree.nwk\")\n",
    "Phylo.write(out_divtree, out_divtree_nwk_path, \"newick\", format_branch_length=\"%1.10f\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "assumed-missouri",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "guilty-trading",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_metadata = out_divtree_nwk_path = os.path.join(out_dir, \"metadata.tsv\")\n",
    "metadata_gdf.to_csv(out_path_metadata, sep=\"\\t\", index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "given-wallpaper",
   "metadata": {},
   "source": [
    "## Augur"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "convenient-calculator",
   "metadata": {},
   "source": [
    "### Reduced Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "unusual-indication",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>branch_length_time</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>mugration_country</th>\n",
       "      <th>mugration_country_confidence</th>\n",
       "      <th>mugration_country_lat</th>\n",
       "      <th>mugration_country_lat_confidence</th>\n",
       "      <th>mugration_country_lon</th>\n",
       "      <th>mugration_country_lon_confidence</th>\n",
       "      <th>mugration_province</th>\n",
       "      <th>mugration_province_confidence</th>\n",
       "      <th>mugration_province_lat</th>\n",
       "      <th>mugration_province_lat_confidence</th>\n",
       "      <th>mugration_province_lon</th>\n",
       "      <th>mugration_province_lon_confidence</th>\n",
       "      <th>mugration_branch_major</th>\n",
       "      <th>mugration_branch_major_confidence</th>\n",
       "      <th>mugration_branch_minor</th>\n",
       "      <th>mugration_branch_minor_confidence</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>continent</th>\n",
       "      <th>node_type</th>\n",
       "      <th>geometry_size</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>strain</th>\n",
       "      <th>country_date_strain</th>\n",
       "      <th>province_date_strain</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_geo_sub</th>\n",
       "      <th>rate_geo_time</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>node_name</th>\n",
       "      <th>blank</th>\n",
       "      <th>branch_major_conf_category</th>\n",
       "      <th>branch_minor_conf_category</th>\n",
       "      <th>country_conf_category</th>\n",
       "      <th>province_conf_category</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>3.774600e-06</td>\n",
       "      <td>16</td>\n",
       "      <td>183</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>1.00</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>1.00</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>North America</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>CO92</td>\n",
       "      <td>United States of America (1992) CO92</td>\n",
       "      <td>United States of America (1992) CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>6.63842e-05</td>\n",
       "      <td>5.7539e-06</td>\n",
       "      <td>0.087238</td>\n",
       "      <td>45.820070</td>\n",
       "      <td>3.997273</td>\n",
       "      <td>[1992.0, 1992.0]</td>\n",
       "      <td>Reference</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>1.817000e-06</td>\n",
       "      <td>8</td>\n",
       "      <td>504</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>Russia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>1.00</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Europe</td>\n",
       "      <td>terminal</td>\n",
       "      <td>3.0</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>9_10</td>\n",
       "      <td>Russia (1923.0) 9_10</td>\n",
       "      <td>Russia (1923.0) 9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>6.77518e-05</td>\n",
       "      <td>9.4529e-06</td>\n",
       "      <td>0.015222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[1923.0, 1923.0]</td>\n",
       "      <td>GCA_009909635.1_ASM990963v1_genomic</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>2.360000e-08</td>\n",
       "      <td>0</td>\n",
       "      <td>99</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>China</td>\n",
       "      <td>1.00</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>1.00</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Asia</td>\n",
       "      <td>terminal</td>\n",
       "      <td>8.0</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>42126</td>\n",
       "      <td>China (2006.0) 42126</td>\n",
       "      <td>China (2006.0) 42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>5.00169e-05</td>\n",
       "      <td>1.11117e-05</td>\n",
       "      <td>0.001002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[2006.0, 2006.0]</td>\n",
       "      <td>GCA_009669545.1_ASM966954v1_genomic</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669805.1_ASM966980v1_genomic</th>\n",
       "      <td>2.360000e-08</td>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "      <td>1980.0</td>\n",
       "      <td>China</td>\n",
       "      <td>1.00</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>1.00</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Asia</td>\n",
       "      <td>terminal</td>\n",
       "      <td>8.0</td>\n",
       "      <td>SAMN07722912</td>\n",
       "      <td>42092</td>\n",
       "      <td>China (1980.0) 42092</td>\n",
       "      <td>China (1980.0) 42092</td>\n",
       "      <td>1980</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>5.00169e-05</td>\n",
       "      <td>1.11117e-05</td>\n",
       "      <td>0.001357</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[1980.0, 1980.0]</td>\n",
       "      <td>GCA_009669805.1_ASM966980v1_genomic</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009296005.1_ASM929600v1_genomic</th>\n",
       "      <td>2.359000e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>186</td>\n",
       "      <td>1953.0</td>\n",
       "      <td>Russia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Chechnya</td>\n",
       "      <td>1.00</td>\n",
       "      <td>43.397615</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>45.698501</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Europe</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.0</td>\n",
       "      <td>SAMN12991209</td>\n",
       "      <td>C-25</td>\n",
       "      <td>Russia (1953.0) C-25</td>\n",
       "      <td>Russia (1953.0) C-25</td>\n",
       "      <td>1953</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>6.58604e-05</td>\n",
       "      <td>7.5615e-06</td>\n",
       "      <td>0.005345</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[1953.0, 1953.0]</td>\n",
       "      <td>GCA_009296005.1_ASM929600v1_genomic</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE188</th>\n",
       "      <td>1.179600e-06</td>\n",
       "      <td>5</td>\n",
       "      <td>68</td>\n",
       "      <td>1990.0</td>\n",
       "      <td>Kazakhstan</td>\n",
       "      <td>0.66</td>\n",
       "      <td>47.228609</td>\n",
       "      <td>0.662908</td>\n",
       "      <td>65.209320</td>\n",
       "      <td>0.662908</td>\n",
       "      <td>West Kazakhstan Region</td>\n",
       "      <td>0.36</td>\n",
       "      <td>49.556848</td>\n",
       "      <td>0.359162</td>\n",
       "      <td>50.222741</td>\n",
       "      <td>0.359162</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.85262e-05</td>\n",
       "      <td>1.02273e-05</td>\n",
       "      <td>0.073696</td>\n",
       "      <td>43.189632</td>\n",
       "      <td>3.182907</td>\n",
       "      <td>[1961.0, 1990.0]</td>\n",
       "      <td>NODE188</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE189</th>\n",
       "      <td>7.077000e-07</td>\n",
       "      <td>3</td>\n",
       "      <td>74</td>\n",
       "      <td>1941.0</td>\n",
       "      <td>Russia</td>\n",
       "      <td>0.98</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>0.984523</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>0.984523</td>\n",
       "      <td>Astrakhan Oblast</td>\n",
       "      <td>0.54</td>\n",
       "      <td>47.187819</td>\n",
       "      <td>0.535277</td>\n",
       "      <td>47.608851</td>\n",
       "      <td>0.535277</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.7323e-05</td>\n",
       "      <td>9.0241e-06</td>\n",
       "      <td>0.040189</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[1892.0, 1972.0]</td>\n",
       "      <td>NODE189</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE190</th>\n",
       "      <td>9.437000e-07</td>\n",
       "      <td>4</td>\n",
       "      <td>66</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>Russia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>0.999691</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>0.999691</td>\n",
       "      <td>Astrakhan Oblast</td>\n",
       "      <td>0.56</td>\n",
       "      <td>47.187819</td>\n",
       "      <td>0.560352</td>\n",
       "      <td>47.608851</td>\n",
       "      <td>0.560352</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.82667e-05</td>\n",
       "      <td>9.9678e-06</td>\n",
       "      <td>0.060226</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[1985.0, 2014.0]</td>\n",
       "      <td>NODE190</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE191</th>\n",
       "      <td>2.360000e-08</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>1964.0</td>\n",
       "      <td>Russia</td>\n",
       "      <td>0.98</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>0.980724</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>0.980724</td>\n",
       "      <td>Republic of Kalmykia</td>\n",
       "      <td>0.43</td>\n",
       "      <td>46.231302</td>\n",
       "      <td>0.428943</td>\n",
       "      <td>45.327574</td>\n",
       "      <td>0.428943</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>80.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.73466e-05</td>\n",
       "      <td>9.0477e-06</td>\n",
       "      <td>0.004298</td>\n",
       "      <td>2045.324858</td>\n",
       "      <td>8.791511</td>\n",
       "      <td>[1932.0, 1984.0]</td>\n",
       "      <td>NODE191</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE192</th>\n",
       "      <td>2.359000e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1984.0</td>\n",
       "      <td>Russia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>0.999722</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>0.999722</td>\n",
       "      <td>Republic of Kalmykia</td>\n",
       "      <td>0.59</td>\n",
       "      <td>46.231302</td>\n",
       "      <td>0.591156</td>\n",
       "      <td>45.327574</td>\n",
       "      <td>0.591156</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.75825e-05</td>\n",
       "      <td>9.2836e-06</td>\n",
       "      <td>0.049904</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[1963.0, 1986.0]</td>\n",
       "      <td>NODE192</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>387 rows  45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     branch_length branch_length_sub  \\\n",
       "sample                                                                 \n",
       "Reference                             3.774600e-06                16   \n",
       "GCA_009909635.1_ASM990963v1_genomic   1.817000e-06                 8   \n",
       "GCA_009669545.1_ASM966954v1_genomic   2.360000e-08                 0   \n",
       "GCA_009669805.1_ASM966980v1_genomic   2.360000e-08                 0   \n",
       "GCA_009296005.1_ASM929600v1_genomic   2.359000e-07                 1   \n",
       "...                                            ...               ...   \n",
       "NODE188                               1.179600e-06                 5   \n",
       "NODE189                               7.077000e-07                 3   \n",
       "NODE190                               9.437000e-07                 4   \n",
       "NODE191                               2.360000e-08                 0   \n",
       "NODE192                               2.359000e-07                 1   \n",
       "\n",
       "                                    branch_length_time  timetree_num_date  \\\n",
       "sample                                                                      \n",
       "Reference                                          183             1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                504             1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                 99             2006.0   \n",
       "GCA_009669805.1_ASM966980v1_genomic                 73             1980.0   \n",
       "GCA_009296005.1_ASM929600v1_genomic                186             1953.0   \n",
       "...                                                ...                ...   \n",
       "NODE188                                             68             1990.0   \n",
       "NODE189                                             74             1941.0   \n",
       "NODE190                                             66             2007.0   \n",
       "NODE191                                             23             1964.0   \n",
       "NODE192                                             20             1984.0   \n",
       "\n",
       "                                            mugration_country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669805.1_ASM966980v1_genomic                     China   \n",
       "GCA_009296005.1_ASM929600v1_genomic                    Russia   \n",
       "...                                                       ...   \n",
       "NODE188                                            Kazakhstan   \n",
       "NODE189                                                Russia   \n",
       "NODE190                                                Russia   \n",
       "NODE191                                                Russia   \n",
       "NODE192                                                Russia   \n",
       "\n",
       "                                     mugration_country_confidence  \\\n",
       "sample                                                              \n",
       "Reference                                                    1.00   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          1.00   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          1.00   \n",
       "GCA_009669805.1_ASM966980v1_genomic                          1.00   \n",
       "GCA_009296005.1_ASM929600v1_genomic                          1.00   \n",
       "...                                                           ...   \n",
       "NODE188                                                      0.66   \n",
       "NODE189                                                      0.98   \n",
       "NODE190                                                      1.00   \n",
       "NODE191                                                      0.98   \n",
       "NODE192                                                      1.00   \n",
       "\n",
       "                                     mugration_country_lat  \\\n",
       "sample                                                       \n",
       "Reference                                        39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic              64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic              35.000074   \n",
       "GCA_009669805.1_ASM966980v1_genomic              35.000074   \n",
       "GCA_009296005.1_ASM929600v1_genomic              64.686314   \n",
       "...                                                    ...   \n",
       "NODE188                                          47.228609   \n",
       "NODE189                                          64.686314   \n",
       "NODE190                                          64.686314   \n",
       "NODE191                                          64.686314   \n",
       "NODE192                                          64.686314   \n",
       "\n",
       "                                     mugration_country_lat_confidence  \\\n",
       "sample                                                                  \n",
       "Reference                                                    1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          1.000000   \n",
       "GCA_009669805.1_ASM966980v1_genomic                          1.000000   \n",
       "GCA_009296005.1_ASM929600v1_genomic                          1.000000   \n",
       "...                                                               ...   \n",
       "NODE188                                                      0.662908   \n",
       "NODE189                                                      0.984523   \n",
       "NODE190                                                      0.999691   \n",
       "NODE191                                                      0.980724   \n",
       "NODE192                                                      0.999722   \n",
       "\n",
       "                                     mugration_country_lon  \\\n",
       "sample                                                       \n",
       "Reference                                      -100.445882   \n",
       "GCA_009909635.1_ASM990963v1_genomic              97.745306   \n",
       "GCA_009669545.1_ASM966954v1_genomic             104.999927   \n",
       "GCA_009669805.1_ASM966980v1_genomic             104.999927   \n",
       "GCA_009296005.1_ASM929600v1_genomic              97.745306   \n",
       "...                                                    ...   \n",
       "NODE188                                          65.209320   \n",
       "NODE189                                          97.745306   \n",
       "NODE190                                          97.745306   \n",
       "NODE191                                          97.745306   \n",
       "NODE192                                          97.745306   \n",
       "\n",
       "                                     mugration_country_lon_confidence  \\\n",
       "sample                                                                  \n",
       "Reference                                                    1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          1.000000   \n",
       "GCA_009669805.1_ASM966980v1_genomic                          1.000000   \n",
       "GCA_009296005.1_ASM929600v1_genomic                          1.000000   \n",
       "...                                                               ...   \n",
       "NODE188                                                      0.662908   \n",
       "NODE189                                                      0.984523   \n",
       "NODE190                                                      0.999691   \n",
       "NODE191                                                      0.980724   \n",
       "NODE192                                                      0.999722   \n",
       "\n",
       "                                         mugration_province  \\\n",
       "sample                                                        \n",
       "Reference                                          Colorado   \n",
       "GCA_009909635.1_ASM990963v1_genomic           Rostov Oblast   \n",
       "GCA_009669545.1_ASM966954v1_genomic                Xinjiang   \n",
       "GCA_009669805.1_ASM966980v1_genomic                Xinjiang   \n",
       "GCA_009296005.1_ASM929600v1_genomic                Chechnya   \n",
       "...                                                     ...   \n",
       "NODE188                              West Kazakhstan Region   \n",
       "NODE189                                    Astrakhan Oblast   \n",
       "NODE190                                    Astrakhan Oblast   \n",
       "NODE191                                Republic of Kalmykia   \n",
       "NODE192                                Republic of Kalmykia   \n",
       "\n",
       "                                     mugration_province_confidence  \\\n",
       "sample                                                               \n",
       "Reference                                                     1.00   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.00   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.00   \n",
       "GCA_009669805.1_ASM966980v1_genomic                           1.00   \n",
       "GCA_009296005.1_ASM929600v1_genomic                           1.00   \n",
       "...                                                            ...   \n",
       "NODE188                                                       0.36   \n",
       "NODE189                                                       0.54   \n",
       "NODE190                                                       0.56   \n",
       "NODE191                                                       0.43   \n",
       "NODE192                                                       0.59   \n",
       "\n",
       "                                     mugration_province_lat  \\\n",
       "sample                                                        \n",
       "Reference                                         38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic               47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic               42.480495   \n",
       "GCA_009669805.1_ASM966980v1_genomic               42.480495   \n",
       "GCA_009296005.1_ASM929600v1_genomic               43.397615   \n",
       "...                                                     ...   \n",
       "NODE188                                           49.556848   \n",
       "NODE189                                           47.187819   \n",
       "NODE190                                           47.187819   \n",
       "NODE191                                           46.231302   \n",
       "NODE192                                           46.231302   \n",
       "\n",
       "                                     mugration_province_lat_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                     1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.000000   \n",
       "GCA_009669805.1_ASM966980v1_genomic                           1.000000   \n",
       "GCA_009296005.1_ASM929600v1_genomic                           1.000000   \n",
       "...                                                                ...   \n",
       "NODE188                                                       0.359162   \n",
       "NODE189                                                       0.535277   \n",
       "NODE190                                                       0.560352   \n",
       "NODE191                                                       0.428943   \n",
       "NODE192                                                       0.591156   \n",
       "\n",
       "                                     mugration_province_lon  \\\n",
       "sample                                                        \n",
       "Reference                                       -105.607716   \n",
       "GCA_009909635.1_ASM990963v1_genomic               40.795794   \n",
       "GCA_009669545.1_ASM966954v1_genomic               85.463346   \n",
       "GCA_009669805.1_ASM966980v1_genomic               85.463346   \n",
       "GCA_009296005.1_ASM929600v1_genomic               45.698501   \n",
       "...                                                     ...   \n",
       "NODE188                                           50.222741   \n",
       "NODE189                                           47.608851   \n",
       "NODE190                                           47.608851   \n",
       "NODE191                                           45.327574   \n",
       "NODE192                                           45.327574   \n",
       "\n",
       "                                     mugration_province_lon_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                     1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.000000   \n",
       "GCA_009669805.1_ASM966980v1_genomic                           1.000000   \n",
       "GCA_009296005.1_ASM929600v1_genomic                           1.000000   \n",
       "...                                                                ...   \n",
       "NODE188                                                       0.359162   \n",
       "NODE189                                                       0.535277   \n",
       "NODE190                                                       0.560352   \n",
       "NODE191                                                       0.428943   \n",
       "NODE192                                                       0.591156   \n",
       "\n",
       "                                    mugration_branch_major  \\\n",
       "sample                                                       \n",
       "Reference                                            1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic                  2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic                  0.ANT   \n",
       "GCA_009669805.1_ASM966980v1_genomic                  0.ANT   \n",
       "GCA_009296005.1_ASM929600v1_genomic                  2.MED   \n",
       "...                                                    ...   \n",
       "NODE188                                              2.MED   \n",
       "NODE189                                              2.MED   \n",
       "NODE190                                              2.MED   \n",
       "NODE191                                              2.MED   \n",
       "NODE192                                              2.MED   \n",
       "\n",
       "                                     mugration_branch_major_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                          1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                                1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                                1.0   \n",
       "GCA_009669805.1_ASM966980v1_genomic                                1.0   \n",
       "GCA_009296005.1_ASM929600v1_genomic                                1.0   \n",
       "...                                                                ...   \n",
       "NODE188                                                            1.0   \n",
       "NODE189                                                            1.0   \n",
       "NODE190                                                            1.0   \n",
       "NODE191                                                            1.0   \n",
       "NODE192                                                            1.0   \n",
       "\n",
       "                                    mugration_branch_minor  \\\n",
       "sample                                                       \n",
       "Reference                                           1.ORI1   \n",
       "GCA_009909635.1_ASM990963v1_genomic                 2.MED1   \n",
       "GCA_009669545.1_ASM966954v1_genomic                 0.ANT1   \n",
       "GCA_009669805.1_ASM966980v1_genomic                 0.ANT1   \n",
       "GCA_009296005.1_ASM929600v1_genomic                 2.MED1   \n",
       "...                                                    ...   \n",
       "NODE188                                             2.MED1   \n",
       "NODE189                                             2.MED1   \n",
       "NODE190                                             2.MED1   \n",
       "NODE191                                             2.MED1   \n",
       "NODE192                                             2.MED1   \n",
       "\n",
       "                                     mugration_branch_minor_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                          1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                                1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                                1.0   \n",
       "GCA_009669805.1_ASM966980v1_genomic                                1.0   \n",
       "GCA_009296005.1_ASM929600v1_genomic                                1.0   \n",
       "...                                                                ...   \n",
       "NODE188                                                            1.0   \n",
       "NODE189                                                            1.0   \n",
       "NODE190                                                            1.0   \n",
       "NODE191                                                            1.0   \n",
       "NODE192                                                            1.0   \n",
       "\n",
       "                                    branch_number  branch_support  \\\n",
       "sample                                                              \n",
       "Reference                                       1           100.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic             2           100.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0           100.0   \n",
       "GCA_009669805.1_ASM966980v1_genomic             0           100.0   \n",
       "GCA_009296005.1_ASM929600v1_genomic             2           100.0   \n",
       "...                                           ...             ...   \n",
       "NODE188                                        NA           100.0   \n",
       "NODE189                                        NA           100.0   \n",
       "NODE190                                        NA           100.0   \n",
       "NODE191                                        NA            80.0   \n",
       "NODE192                                        NA           100.0   \n",
       "\n",
       "                                    branch_support_conf_category  \\\n",
       "sample                                                             \n",
       "Reference                                                   HIGH   \n",
       "GCA_009909635.1_ASM990963v1_genomic                         HIGH   \n",
       "GCA_009669545.1_ASM966954v1_genomic                         HIGH   \n",
       "GCA_009669805.1_ASM966980v1_genomic                         HIGH   \n",
       "GCA_009296005.1_ASM929600v1_genomic                         HIGH   \n",
       "...                                                          ...   \n",
       "NODE188                                                     HIGH   \n",
       "NODE189                                                     HIGH   \n",
       "NODE190                                                     HIGH   \n",
       "NODE191                                                      LOW   \n",
       "NODE192                                                     HIGH   \n",
       "\n",
       "                                         continent node_type  geometry_size  \\\n",
       "sample                                                                        \n",
       "Reference                            North America  terminal            1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic         Europe  terminal            3.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic           Asia  terminal            8.0   \n",
       "GCA_009669805.1_ASM966980v1_genomic           Asia  terminal            8.0   \n",
       "GCA_009296005.1_ASM929600v1_genomic         Europe  terminal            2.0   \n",
       "...                                            ...       ...            ...   \n",
       "NODE188                                         NA  internal            2.0   \n",
       "NODE189                                         NA  internal            3.0   \n",
       "NODE190                                         NA  internal            3.0   \n",
       "NODE191                                         NA  internal            2.0   \n",
       "NODE192                                         NA  internal            2.0   \n",
       "\n",
       "                                    biosample_accession strain  \\\n",
       "sample                                                           \n",
       "Reference                                  SAMEA1705942   CO92   \n",
       "GCA_009909635.1_ASM990963v1_genomic        SAMN13632815   9_10   \n",
       "GCA_009669545.1_ASM966954v1_genomic        SAMN07722925  42126   \n",
       "GCA_009669805.1_ASM966980v1_genomic        SAMN07722912  42092   \n",
       "GCA_009296005.1_ASM929600v1_genomic        SAMN12991209   C-25   \n",
       "...                                                 ...    ...   \n",
       "NODE188                                              NA     NA   \n",
       "NODE189                                              NA     NA   \n",
       "NODE190                                              NA     NA   \n",
       "NODE191                                              NA     NA   \n",
       "NODE192                                              NA     NA   \n",
       "\n",
       "                                                      country_date_strain  \\\n",
       "sample                                                                      \n",
       "Reference                            United States of America (1992) CO92   \n",
       "GCA_009909635.1_ASM990963v1_genomic                  Russia (1923.0) 9_10   \n",
       "GCA_009669545.1_ASM966954v1_genomic                  China (2006.0) 42126   \n",
       "GCA_009669805.1_ASM966980v1_genomic                  China (1980.0) 42092   \n",
       "GCA_009296005.1_ASM929600v1_genomic                  Russia (1953.0) C-25   \n",
       "...                                                                   ...   \n",
       "NODE188                                                                NA   \n",
       "NODE189                                                                NA   \n",
       "NODE190                                                                NA   \n",
       "NODE191                                                                NA   \n",
       "NODE192                                                                NA   \n",
       "\n",
       "                                                     province_date_strain  \\\n",
       "sample                                                                      \n",
       "Reference                            United States of America (1992) CO92   \n",
       "GCA_009909635.1_ASM990963v1_genomic                  Russia (1923.0) 9_10   \n",
       "GCA_009669545.1_ASM966954v1_genomic                  China (2006.0) 42126   \n",
       "GCA_009669805.1_ASM966980v1_genomic                  China (1980.0) 42092   \n",
       "GCA_009296005.1_ASM929600v1_genomic                  Russia (1953.0) C-25   \n",
       "...                                                                   ...   \n",
       "NODE188                                                                NA   \n",
       "NODE189                                                                NA   \n",
       "NODE190                                                                NA   \n",
       "NODE191                                                                NA   \n",
       "NODE192                                                                NA   \n",
       "\n",
       "                                    date_mean date_err date_bp_mean  \\\n",
       "sample                                                                \n",
       "Reference                                1992        0           29   \n",
       "GCA_009909635.1_ASM990963v1_genomic      1923        0           98   \n",
       "GCA_009669545.1_ASM966954v1_genomic      2006        0           15   \n",
       "GCA_009669805.1_ASM966980v1_genomic      1980        0           41   \n",
       "GCA_009296005.1_ASM929600v1_genomic      1953        0           68   \n",
       "...                                       ...      ...          ...   \n",
       "NODE188                                    NA       NA           NA   \n",
       "NODE189                                    NA       NA           NA   \n",
       "NODE190                                    NA       NA           NA   \n",
       "NODE191                                    NA       NA           NA   \n",
       "NODE192                                    NA       NA           NA   \n",
       "\n",
       "                                    root_rtt_dist clade_rtt_dist  rate_sub  \\\n",
       "sample                                                                       \n",
       "Reference                             6.63842e-05     5.7539e-06  0.087238   \n",
       "GCA_009909635.1_ASM990963v1_genomic   6.77518e-05     9.4529e-06  0.015222   \n",
       "GCA_009669545.1_ASM966954v1_genomic   5.00169e-05    1.11117e-05  0.001002   \n",
       "GCA_009669805.1_ASM966980v1_genomic   5.00169e-05    1.11117e-05  0.001357   \n",
       "GCA_009296005.1_ASM929600v1_genomic   6.58604e-05     7.5615e-06  0.005345   \n",
       "...                                           ...            ...       ...   \n",
       "NODE188                               6.85262e-05    1.02273e-05  0.073696   \n",
       "NODE189                                6.7323e-05     9.0241e-06  0.040189   \n",
       "NODE190                               6.82667e-05     9.9678e-06  0.060226   \n",
       "NODE191                               6.73466e-05     9.0477e-06  0.004298   \n",
       "NODE192                               6.75825e-05     9.2836e-06  0.049904   \n",
       "\n",
       "                                     rate_geo_sub  rate_geo_time  \\\n",
       "sample                                                             \n",
       "Reference                               45.820070       3.997273   \n",
       "GCA_009909635.1_ASM990963v1_genomic      0.000000       0.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic      0.000000       0.000000   \n",
       "GCA_009669805.1_ASM966980v1_genomic      0.000000       0.000000   \n",
       "GCA_009296005.1_ASM929600v1_genomic      0.000000       0.000000   \n",
       "...                                           ...            ...   \n",
       "NODE188                                 43.189632       3.182907   \n",
       "NODE189                                  0.000000       0.000000   \n",
       "NODE190                                  0.000000       0.000000   \n",
       "NODE191                               2045.324858       8.791511   \n",
       "NODE192                                  0.000000       0.000000   \n",
       "\n",
       "                                    timetree_num_date_confidence  \\\n",
       "sample                                                             \n",
       "Reference                                       [1992.0, 1992.0]   \n",
       "GCA_009909635.1_ASM990963v1_genomic             [1923.0, 1923.0]   \n",
       "GCA_009669545.1_ASM966954v1_genomic             [2006.0, 2006.0]   \n",
       "GCA_009669805.1_ASM966980v1_genomic             [1980.0, 1980.0]   \n",
       "GCA_009296005.1_ASM929600v1_genomic             [1953.0, 1953.0]   \n",
       "...                                                          ...   \n",
       "NODE188                                         [1961.0, 1990.0]   \n",
       "NODE189                                         [1892.0, 1972.0]   \n",
       "NODE190                                         [1985.0, 2014.0]   \n",
       "NODE191                                         [1932.0, 1984.0]   \n",
       "NODE192                                         [1963.0, 1986.0]   \n",
       "\n",
       "                                                               node_name  \\\n",
       "sample                                                                     \n",
       "Reference                                                      Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic  GCA_009909635.1_ASM990963v1_genomic   \n",
       "GCA_009669545.1_ASM966954v1_genomic  GCA_009669545.1_ASM966954v1_genomic   \n",
       "GCA_009669805.1_ASM966980v1_genomic  GCA_009669805.1_ASM966980v1_genomic   \n",
       "GCA_009296005.1_ASM929600v1_genomic  GCA_009296005.1_ASM929600v1_genomic   \n",
       "...                                                                  ...   \n",
       "NODE188                                                          NODE188   \n",
       "NODE189                                                          NODE189   \n",
       "NODE190                                                          NODE190   \n",
       "NODE191                                                          NODE191   \n",
       "NODE192                                                          NODE192   \n",
       "\n",
       "                                    blank branch_major_conf_category  \\\n",
       "sample                                                                 \n",
       "Reference                                                       HIGH   \n",
       "GCA_009909635.1_ASM990963v1_genomic                             HIGH   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             HIGH   \n",
       "GCA_009669805.1_ASM966980v1_genomic                             HIGH   \n",
       "GCA_009296005.1_ASM929600v1_genomic                             HIGH   \n",
       "...                                   ...                        ...   \n",
       "NODE188                                                         HIGH   \n",
       "NODE189                                                         HIGH   \n",
       "NODE190                                                         HIGH   \n",
       "NODE191                                                         HIGH   \n",
       "NODE192                                                         HIGH   \n",
       "\n",
       "                                    branch_minor_conf_category  \\\n",
       "sample                                                           \n",
       "Reference                                                 HIGH   \n",
       "GCA_009909635.1_ASM990963v1_genomic                       HIGH   \n",
       "GCA_009669545.1_ASM966954v1_genomic                       HIGH   \n",
       "GCA_009669805.1_ASM966980v1_genomic                       HIGH   \n",
       "GCA_009296005.1_ASM929600v1_genomic                       HIGH   \n",
       "...                                                        ...   \n",
       "NODE188                                                   HIGH   \n",
       "NODE189                                                   HIGH   \n",
       "NODE190                                                   HIGH   \n",
       "NODE191                                                   HIGH   \n",
       "NODE192                                                   HIGH   \n",
       "\n",
       "                                    country_conf_category  \\\n",
       "sample                                                      \n",
       "Reference                                            HIGH   \n",
       "GCA_009909635.1_ASM990963v1_genomic                  HIGH   \n",
       "GCA_009669545.1_ASM966954v1_genomic                  HIGH   \n",
       "GCA_009669805.1_ASM966980v1_genomic                  HIGH   \n",
       "GCA_009296005.1_ASM929600v1_genomic                  HIGH   \n",
       "...                                                   ...   \n",
       "NODE188                                               LOW   \n",
       "NODE189                                              HIGH   \n",
       "NODE190                                              HIGH   \n",
       "NODE191                                              HIGH   \n",
       "NODE192                                              HIGH   \n",
       "\n",
       "                                    province_conf_category  \n",
       "sample                                                      \n",
       "Reference                                             HIGH  \n",
       "GCA_009909635.1_ASM990963v1_genomic                   HIGH  \n",
       "GCA_009669545.1_ASM966954v1_genomic                   HIGH  \n",
       "GCA_009669805.1_ASM966980v1_genomic                   HIGH  \n",
       "GCA_009296005.1_ASM929600v1_genomic                   HIGH  \n",
       "...                                                    ...  \n",
       "NODE188                                                LOW  \n",
       "NODE189                                                LOW  \n",
       "NODE190                                                LOW  \n",
       "NODE191                                                LOW  \n",
       "NODE192                                                LOW  \n",
       "\n",
       "[387 rows x 45 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remember, order atters when dealing with confidence!\n",
    "\n",
    "columns = [\n",
    "    # Draw Divergence Tree\n",
    "    \"branch_length\",\n",
    "    \"branch_length_sub\",\n",
    "    # Draw Time Tree\n",
    "    \"branch_length_time\",\n",
    "    \"timetree_date_calendar\",\n",
    "    # Draw Country Map\n",
    "    \"mugration_country\",\n",
    "    \"mugration_country_confidence\",    \n",
    "    \"mugration_country_lat\",   \n",
    "    \"mugration_country_lat_confidence\",\n",
    "    \"mugration_country_lon\",  \n",
    "    \"mugration_country_lon_confidence\", \n",
    "    #\"mugration_country_transition_pvalue\",\n",
    "    #\"mugration_country_transition_conf_category\",      \n",
    "    # Draw Province Map\n",
    "    \"mugration_province\",\n",
    "    \"mugration_province_confidence\",    \n",
    "    \"mugration_province_lat\",   \n",
    "    \"mugration_province_lat_confidence\",\n",
    "    \"mugration_province_lon\",  \n",
    "    \"mugration_province_lon_confidence\",    \n",
    "    #\"mugration_province_transition_pvalue\",\n",
    "    #\"mugration_province_transition_conf_category\",    \n",
    "    # Colors and Filters\n",
    "    \"mugration_branch_major\",\n",
    "    \"mugration_branch_major_confidence\",\n",
    "    \"mugration_branch_minor\",\n",
    "    \"mugration_branch_minor_confidence\",   \n",
    "    \"branch_number\",\n",
    "    \"branch_support\",\n",
    "    \"branch_support_conf_category\",    \n",
    "    \"continent\",\n",
    "    \"node_type\",\n",
    "    \"geometry_size\",\n",
    "    # Text Description\n",
    "    \"biosample_accession\",\n",
    "    \"strain\",\n",
    "    \"country_date_strain\",\n",
    "    \"province_date_strain\",\n",
    "    # Tip Dates\n",
    "    \"date_mean\",\n",
    "    \"date_err\",\n",
    "    \"date_bp_mean\",\n",
    "    # Stats\n",
    "    \"root_rtt_dist\",\n",
    "    \"clade_rtt_dist\",\n",
    "    \"rate_sub\",\n",
    "    \"rate_geo_sub\",\n",
    "    \"rate_geo_time\" \n",
    "]\n",
    "\n",
    "# Edit df\n",
    "auspice_gdf = copy.copy(metadata_gdf[columns])\n",
    "auspice_gdf[\"timetree_num_date_confidence\"] = [NO_DATA_CHAR] * len(auspice_gdf)\n",
    "auspice_gdf[\"node_name\"] = list(auspice_gdf.index)\n",
    "auspice_gdf[\"blank\"] = [\" \"] * len(auspice_gdf)\n",
    "\n",
    "for attr in MUG_ATTRIBUTE_LIST:\n",
    "    col = attr + \"_conf_category\"\n",
    "    auspice_gdf[col] = [\"NA\"] * len(auspice_gdf)\n",
    "    for rec in auspice_gdf.iterrows():\n",
    "        sample = rec[0]\n",
    "        conf = auspice_gdf[\"mugration_\" + attr + \"_confidence\"][sample]\n",
    "        if conf >= 0.95:\n",
    "            category = \"HIGH\"\n",
    "        elif conf < 0.95:\n",
    "            category = \"LOW\"\n",
    "        auspice_gdf.at[sample, col] = category\n",
    "\n",
    "for rec in metadata_gdf.iterrows():\n",
    "    sample = rec[0]\n",
    "    date = rec[1][\"timetree_date_calendar\"]\n",
    "    date_lower = float(date) - rec[1][\"timetree_CI_lower_err\"]\n",
    "    date_upper = float(date) + rec[1][\"timetree_CI_upper_err\"]\n",
    "    #date_confidence = \"{}:{}\".format(date_lower, date_upper)\n",
    "    date_confidence = [date_lower, date_upper]\n",
    "    auspice_gdf.at[sample,\"timetree_num_date_confidence\"] = date_confidence\n",
    "    # Rounding\n",
    "    auspice_gdf.at[sample, \"branch_length_sub\"] = round(auspice_gdf[\"branch_length_sub\"][sample])\n",
    "    auspice_gdf.at[sample, \"branch_length_time\"] = round(auspice_gdf[\"branch_length_time\"][sample])\n",
    "    #auspice_gdf.at[sample, \"mugration_country_transition_pvalue\"] = round(auspice_gdf[\"mugration_country_transition_pvalue\"][sample],2)\n",
    "    #auspice_gdf.at[sample, \"mugration_province_transition_pvalue\"] = round(auspice_gdf[\"mugration_province_transition_pvalue\"][sample],2)\n",
    "\n",
    "auspice_gdf.rename(columns={\"timetree_date_calendar\": \"timetree_num_date\"}, inplace=True)\n",
    "#auspice_gdf.rename(columns={\"mugration_country_transition_pvalue\": \"country_transition_pvalue\"}, inplace=True)\n",
    "#auspice_gdf.rename(columns={\"mugration_province_transition_pvalue\": \"province_transition_pvalue\"}, inplace=True)\n",
    "#auspice_gdf.rename(columns={\"mugration_country_transition_conf_category\": \"country_transition_conf_category\"}, inplace=True)\n",
    "#auspice_gdf.rename(columns={\"mugration_province_transition_conf_category\": \"province_transition_conf_category\"}, inplace=True)\n",
    "display(auspice_gdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "silent-tuesday",
   "metadata": {},
   "source": [
    "### JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "novel-enough",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'branch_length': 3.7746e-06, 'branch_length_sub': 16, 'branch_length_time': 183, 'num_date': 1992.0, 'country': 'United States of America', 'country_confidence': {'United States of America': 1.0}, 'country_lat': 39.7837304, 'country_lat_confidence': {39.7837304: 1.0}, 'country_lon': -100.4458825, 'country_lon_confidence': {-100.4458825: 1.0}, 'province': 'Colorado', 'province_confidence': {'Colorado': 1.0}, 'province_lat': 38.7251776, 'province_lat_confidence': {38.7251776: 1.0}, 'province_lon': -105.607716, 'province_lon_confidence': {-105.607716: 1.0}, 'branch_major': '1.ORI', 'branch_major_confidence': {'1.ORI': 1.0}, 'branch_minor': '1.ORI1', 'branch_minor_confidence': {'1.ORI1': 1.0}, 'branch_number': '1.0', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'continent': 'North America', 'node_type': 'terminal', 'biosample_accession': 'SAMEA1705942', 'strain': 'CO92', 'country_date_strain': 'United States of America (1992) CO92', 'province_date_strain': 'United States of America (1992) CO92', 'date_mean': 1992.0, 'date_err': 0.0, 'date_bp_mean': 29.0, 'root_rtt_dist': 6.638419999999997e-05, 'clade_rtt_dist': 5.753899999999999e-06, 'rate_sub': 0.08723847274339092, 'rate_geo_sub': 45.82006986697771, 'rate_geo_time': 3.9972729161906027, 'num_date_confidence': [1992.0, 1992.0], 'node_name': 'Reference', 'blank': ' ', 'branch_major_conf_category': 'HIGH', 'branch_minor_conf_category': 'HIGH', 'country_conf_category': 'HIGH', 'province_conf_category': 'HIGH'}\n"
     ]
    }
   ],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=None, \n",
    "    aln_path=None,  \n",
    "    tree=divtree, \n",
    "    tree_df=auspice_gdf, \n",
    "    color_keyword_exclude=[\"geometry\"],\n",
    "    type_convert = {\n",
    "        \"branch_number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])\n",
    "\n",
    "out_path_augur_json = os.path.join(out_dir, \"all.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hybrid-artwork",
   "metadata": {},
   "source": [
    "## Auspice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "parental-subdivision",
   "metadata": {},
   "source": [
    "### JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "buried-business",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: kyrgyzstan, madagascar, myanmar, uganda, turkmenistan, georgia, kazakhstan, democratic republic of the congo, vietnam. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/all.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/all.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n"
     ]
    }
   ],
   "source": [
    "#import sys, importlib\n",
    "#importlib.reload(sys.modules['functions'])\n",
    "#from functions import auspice_export, branch_attributes\n",
    "\n",
    "auspice_dict = auspice_export(\n",
    "    tree=divtree,\n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=out_path_colors,\n",
    "    auspice_latlons_path=out_path_latlon, \n",
    "    )\n",
    "\n",
    "label_col = list(auspice_gdf.columns)\n",
    "\n",
    "# Recursively add branch attrs\n",
    "branch_attributes(\n",
    "    tree_dict=auspice_dict[\"tree\"], \n",
    "    sub_dict=auspice_dict[\"tree\"], \n",
    "    df=auspice_gdf,\n",
    "    label_col=label_col,\n",
    "    )\n",
    "\n",
    "# Last manual changes\n",
    "auspice_dict_copy = copy.deepcopy(auspice_dict)\n",
    "for i in range(0, len(auspice_dict_copy[\"meta\"][\"colorings\"])):\n",
    "    coloring = auspice_dict_copy[\"meta\"][\"colorings\"][i]\n",
    "    for key in coloring:\n",
    "        # Node type as internal or terminal\n",
    "        if coloring[key] == \"node_type\":\n",
    "            auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['internal', '#FFFFFF'], ['terminal', '#000000']]\n",
    "            #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "        # Confidence category\n",
    "        if \"conf_category\" in coloring[key]:\n",
    "            auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['LOW', '#FFFFFF'], ['HIGH', '#000000']]\n",
    "            #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            \n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(auspice_dir, \"all.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\\n\")\n",
    "\n",
    "#out_path_auspice_remote_json = os.path.join(auspice_dir, \"{}_all.json\".format(AUSPICE_PREFIX))\n",
    "#utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "#export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "#print(\"Validation successful for local JSON.\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "amended-danish",
   "metadata": {},
   "source": [
    "## Subtrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "neutral-clause",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.PRE #8000ff\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEPRECATED: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/0.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/0.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "0.ANT4 #1996f3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_support_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_support_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/0.ANT4.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/0.ANT4.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "0.PE #4c4ffc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: georgia. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/0.PE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/0.PE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "0.ANT #1996f3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/0.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/0.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: kyrgyzstan. Auspice will create colors for them.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.PRE #e6ce74\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/1.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/1.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "1.ANT #ff964f\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "WARNING: These values for trait province were not specified in your provided color scale: uganda, democratic republic of the congo. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/1.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/1.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "1.IN #ff4f28\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_support_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_support_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/1.IN.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/1.IN.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "1.ORI #ff0000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: madagascar, myanmar, vietnam. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/1.ORI.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/1.ORI.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "2.ANT #80ffb4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/2.ANT.json'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/2.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "2.MED #b3f396\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: kazakhstan, turkmenistan. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/2.MED.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/2.MED.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "3.ANT #1acee3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/3.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/3.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "4.ANT #4df3ce\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_support_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_support_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/4.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/prune/filter5/4.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_support_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_support_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_major_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n"
     ]
    }
   ],
   "source": [
    "import sys, importlib\n",
    "importlib.reload(sys.modules['functions'])\n",
    "from functions import extract_subtree\n",
    "\n",
    "for branch in BRANCH_LIST:\n",
    "    timetree_copy = copy.deepcopy(timetree)\n",
    "    \n",
    "    # Create the subtree df\n",
    "    subtree_df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[branch])]\n",
    "    branch_major_color = subtree_df[\"branch_major_color\"][0]\n",
    "    \n",
    "    if len(subtree_df) < 2: continue\n",
    "    print(branch, branch_major_color)\n",
    "\n",
    "    subtree_df.sort_values(\"timetree_coord_y\", inplace=True)\n",
    "    subtree_tips = subtree_df[subtree_df[\"node_type\"] == \"terminal\"]\n",
    "    subtree_internals = subtree_df[subtree_df[\"node_type\"] == \"internal\"]\n",
    "    \n",
    "    # ----------------------------------------------\n",
    "    # Subtree Timetree\n",
    "    subtree = extract_subtree(\n",
    "        tree=timetree_copy, \n",
    "        tips=list(subtree_tips.index), \n",
    "        df=metadata_gdf, \n",
    "        color_branches=False)\n",
    "    \n",
    "    metadata_to_comment(subtree, metadata_gdf)   \n",
    "\n",
    "\n",
    "    out_subtree_nex_path = os.path.join(out_dir, \"{}.timetree.nex\".format(branch))\n",
    "    Phylo.write(subtree, out_subtree_nex_path, \"nexus\")\n",
    "    out_subtree_nwk_path = os.path.join(out_dir, \"{}.timetree.nwk\".format(branch))\n",
    "    Phylo.write(subtree, out_subtree_nwk_path, \"newick\")\n",
    "    \n",
    "    # ----------------------------------------------\n",
    "    # Subtree divtree\n",
    "    divtree_copy = copy.deepcopy(divtree)\n",
    "    subtree_div = extract_subtree(\n",
    "        tree=divtree_copy, \n",
    "        tips=list(subtree_tips.index), \n",
    "        df=metadata_gdf, \n",
    "        color_branches=False)\n",
    "    \n",
    "    metadata_to_comment(subtree_div, metadata_gdf)   \n",
    "    out_subtree_path = os.path.join(out_dir, \"{}.divtree.nex\".format(branch))\n",
    "    Phylo.write(subtree, out_subtree_path, \"nexus\", format_branch_length=\"%1.10f\")\n",
    "\n",
    "    #fig, ax = plt.subplots(1)\n",
    "    #Phylo.draw(subtree, axes=ax, show_confidence=False, label_func = lambda x: '', do_show=False)\n",
    "    \n",
    "    augur_dict = augur_export(\n",
    "        tree_path=None, \n",
    "        aln_path=None,  \n",
    "        tree=subtree, \n",
    "        tree_df=auspice_gdf, \n",
    "        color_keyword_exclude=[\"geometry\"],\n",
    "        type_convert = {\n",
    "            \"branch_number\" : (lambda x : str(x))\n",
    "        },\n",
    "    ) \n",
    "        \n",
    "    out_path_augur_json = os.path.join(out_dir, \"{}.json\".format(branch) )\n",
    "    utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "\n",
    "    auspice_dict = auspice_export(\n",
    "        tree=subtree,\n",
    "        augur_json_paths=[out_path_augur_json], \n",
    "        auspice_config_path=auspice_config_path, \n",
    "        auspice_colors_path=out_path_colors,\n",
    "        auspice_latlons_path=out_path_latlon, \n",
    "        )    \n",
    "    \n",
    "    label_col = list(auspice_gdf.columns)\n",
    "\n",
    "    # Recursively add branch attrs\n",
    "    branch_attributes(\n",
    "        tree_dict=auspice_dict[\"tree\"], \n",
    "        sub_dict=auspice_dict[\"tree\"], \n",
    "        df=auspice_gdf,\n",
    "        label_col=label_col,\n",
    "        )    \n",
    "    \n",
    "    # Last manual changes\n",
    "    auspice_dict_copy = copy.deepcopy(auspice_dict)\n",
    "    for i in range(0, len(auspice_dict_copy[\"meta\"][\"colorings\"])):\n",
    "        coloring = auspice_dict_copy[\"meta\"][\"colorings\"][i]\n",
    "        for key in coloring:\n",
    "            # Node type as internal or terminal\n",
    "            if coloring[key] == \"node_type\":\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['internal', '#FFFFFF'], ['terminal', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            # Confidence category\n",
    "            if \"conf_category\" in coloring[key]:\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['LOW', '#FFFFFF'], ['HIGH', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i]) \n",
    "    \n",
    "    # Write outputs - For Local Rendering\n",
    "    out_path_auspice_local_json = os.path.join(auspice_dir, \"{}.json\".format(branch) )\n",
    "    utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "    export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "    print(\"Validation successful for local JSON.\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acting-karma",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
