{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "abstract-ribbon",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. SETUP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "diverse-genealogy",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bronze-allowance",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import Phylo, AlignIO\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors, gridspec, lines\n",
    "import scipy\n",
    "import math\n",
    "import numpy as np\n",
    "import geopandas\n",
    "import datetime\n",
    "import shapely\n",
    "import copy\n",
    "from geopy import distance\n",
    "import geopy\n",
    "import cartopy.crs as ccrs\n",
    "from augur import utils, export_v2\n",
    "import subprocess\n",
    "import itertools\n",
    "\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "spare-semester",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "female-personality",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"full\", \"5\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "    \n",
    "results_dir = os.path.join(project_dir, \"results/\")\n",
    "\n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "PRUNE = WILDCARDS[2]\n",
    "MISSING_DATA = WILDCARDS[3]\n",
    "\n",
    "AUSPICE_PREFIX = \"plague-phylogeography-projects_main_{}\".format(PRUNE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "automatic-kentucky",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manual override\n",
    "# results_dir = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "resistant-silicon",
   "metadata": {},
   "outputs": [],
   "source": [
    "iqtree_dir    = os.path.join(results_dir, \"iqtree/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "lsd_dir       = os.path.join(results_dir, \"lsd/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "mugration_dir = os.path.join(results_dir, \"mugration/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "snippy_dir    = os.path.join(results_dir, \"snippy_multi/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "\n",
    "# ------------------------------------------\n",
    "# Trees\n",
    "divtree_path              = iqtree_dir    + \"filter-taxa/iqtree.treefile\"\n",
    "timetree_path             = lsd_dir       + \"lsd.timetree.nex\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "constant_sites_path   = results_dir + \"snippy_multi/all/chromosome/full/snippy-multi.constant_sites.txt\"\n",
    "aln_path              = iqtree_dir + \"filter-sites/snippy-multi.snps.aln\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Metadata\n",
    "\n",
    "auspice_config_path          = results_dir + \"config/auspice_config.json\"\n",
    "metadata_path                = iqtree_dir    + \"filter-taxa/metadata.tsv\"\n",
    "\n",
    "    \n",
    "# ------------------------------------------\n",
    "# Output\n",
    "augur_dir   = os.path.join(results_dir, \"augur/all/chromosome/{}/filter{}/ml\".format(PRUNE, MISSING_DATA))\n",
    "auspice_dir = os.path.join(results_dir, \"auspice/all/chromosome/{}/filter{}/ml\".format(PRUNE, MISSING_DATA))\n",
    "\n",
    "if not os.path.exists(augur_dir):\n",
    "    print(augur_dir)\n",
    "    subprocess.run([\"mkdir\", \"-p\", augur_dir]) \n",
    "    \n",
    "if not os.path.exists(auspice_dir):\n",
    "    print(auspice_dir)\n",
    "    subprocess.run([\"mkdir\", \"-p\", auspice_dir]) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "spanish-center",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "pressing-biography",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", None)\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "UNKNOWN_CHAR = \"?\"\n",
    "JSON_INDENT=2\n",
    "CURRENT_YEAR = datetime.datetime.utcnow().year\n",
    "MUG_CONF = 0.95\n",
    "ALPHA = 0.05\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "with open(constant_sites_path) as infile:\n",
    "    data = infile.read().strip().split(\",\")\n",
    "    constant_sites = sum([int(count) for count in data])\n",
    "\n",
    "aln = AlignIO.read(aln_path, \"fasta\")\n",
    "variant_sites = len(aln[0].seq)\n",
    "SEQ_LEN = constant_sites + variant_sites\n",
    "\n",
    "# ------------------------------------------\n",
    "# Geo\n",
    "CRS = \"epsg:3857\"\n",
    "CRS_EPSG = ccrs.epsg('3857')\n",
    "world_polygons = geopandas.read_file(geopandas.datasets.get_path(\"naturalearth_lowres\"))\n",
    "\n",
    "\n",
    "MUG_ATTRIBUTE_LIST = [\n",
    "    \"country\",\n",
    "    \"province\",\n",
    "    \"continent\",\n",
    "    \"host_order\",\n",
    "    \"population\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "technical-cheese",
   "metadata": {},
   "outputs": [],
   "source": [
    "mug_dict = {}\n",
    "\n",
    "for attr in MUG_ATTRIBUTE_LIST:\n",
    "    mug_dict[attr] = {}\n",
    "    mug_dict[attr][\"tree_path\"] = mugration_dir + attr + \".nex\".format(MISSING_DATA)\n",
    "    mug_dict[attr][\"conf_path\"] = mugration_dir + attr + \"_confidence.csv\"\n",
    "    mug_dict[attr][\"states_path\"] = mugration_dir + attr + \"_states.csv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nuclear-final",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 2. IMPORT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "charming-depression",
   "metadata": {},
   "source": [
    "## Mugration Confidence Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "representative-norway",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "states_files = [mug_dict[attr][\"states_path\"] for attr in mug_dict]\n",
    "conf_files = [mug_dict[attr][\"conf_path\"] for attr in mug_dict]\n",
    "\n",
    "# ----------------------------------------\n",
    "# Convert character states to numbers\n",
    "for mugstates,mugconf in zip(states_files, conf_files):\n",
    "    node_i = 0\n",
    "    conf_basename = os.path.splitext(mugconf)[0]\n",
    "    conf_outname = conf_basename + \"_tidy.csv\"\n",
    "    i = 0\n",
    "    state_dict = {}\n",
    "    with open(mugstates, \"r\") as infile:\n",
    "        read_line = infile.readline()\n",
    "        while read_line:\n",
    "            split_line = read_line.strip().split(\",\")\n",
    "            state = split_line[1]\n",
    "            if state == \"nan\":\n",
    "                state = \"NA\"\n",
    "            if state == UNKNOWN_CHAR:\n",
    "                state_dict[-1] = state\n",
    "            else:\n",
    "                state_dict[i] = state\n",
    "                i += 1\n",
    "            read_line = infile.readline()\n",
    "    # Swap first to last place\n",
    "    final_i = len(state_dict) - 1   \n",
    "    state_dict[final_i] = \"?\"\n",
    "    # Remove old first\n",
    "    state_dict.pop(-1)\n",
    "\n",
    "    with open(mugconf, \"r\") as infile:            \n",
    "        i = 0\n",
    "        read_line = infile.readline()\n",
    "        header = read_line.strip().split(\", \")\n",
    "        header[0] = \"name\"\n",
    "\n",
    "        data = \"\"\n",
    "        for i in range(0, len(header) - 1):\n",
    "            char = header[i + 1]\n",
    "            header[i + 1] = state_dict[i]\n",
    "        header = \",\".join(header)\n",
    "        read_line = infile.readline()            \n",
    "        while read_line:\n",
    "            # Rename internal nodes\n",
    "            split_line = read_line.split(\", \")\n",
    "            node_name = split_line[0]\n",
    "            if \"NODE\" in node_name:\n",
    "                node_name = \"NODE{}\".format(node_i)\n",
    "                node_i += 1\n",
    "            new_line = \",\".join([node_name] + split_line[1:])\n",
    "            data += new_line\n",
    "            read_line = infile.readline()\n",
    "    \n",
    "    with open(conf_outname, \"w\") as outfile:\n",
    "        outfile.write(header + \"\\n\")\n",
    "        outfile.write(data)\n",
    "\n",
    "# ----------------------------------------\n",
    "# Import the tidy files as dataframes\n",
    "\n",
    "for attr in mug_dict:\n",
    "    conf_path = mug_dict[attr][\"conf_path\"]\n",
    "    tidy_path = os.path.splitext(conf_path)[0] + \"_tidy.csv\"\n",
    "    df = pd.read_csv(tidy_path, sep=',', index_col = 0)\n",
    "    mug_dict[attr][\"conf_df\"] = df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "orange-aside",
   "metadata": {},
   "source": [
    "## Trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "induced-development",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "divtree: iqtree.treefile\n",
      "timetree: lsd.timetree.nex\n",
      "country: country.nex\n",
      "province: province.nex\n",
      "continent: continent.nex\n",
      "host_order: host_order.nex\n",
      "population: population.nex\n"
     ]
    }
   ],
   "source": [
    "tree_paths = [divtree_path, timetree_path] + [mug_dict[attr][\"tree_path\"] for attr in mug_dict]\n",
    "\n",
    "for tree_path in tree_paths:\n",
    "    tree_ext = os.path.splitext(tree_path)[1]\n",
    "    if tree_ext == \".nwk\" or tree_ext == \".newick\" or tree_ext == \".treefile\":\n",
    "        tree = Phylo.read(tree_path, \"newick\")\n",
    "    elif tree_ext == \".nex\" or tree_ext == \".nexus\":\n",
    "        parse_trees = Phylo.parse(tree_path, \"nexus\")\n",
    "        for parse_tree in parse_trees:\n",
    "            clades = [c for c in parse_tree.find_clades()]\n",
    "            if len(clades) > 1:\n",
    "                break\n",
    "        tree = parse_tree\n",
    "    tree.ladderize(reverse=False)\n",
    "    \n",
    "    if tree_path == divtree_path:\n",
    "        print(\"divtree:\", os.path.basename(tree_path))\n",
    "        divtree = tree\n",
    "    elif tree_path == timetree_path:\n",
    "        print(\"timetree:\", os.path.basename(tree_path))        \n",
    "        timetree = tree   \n",
    "        \n",
    "    # Check against mugration attributes\n",
    "    for attr in mug_dict:\n",
    "        if tree_path == mug_dict[attr][\"tree_path\"]:\n",
    "            print(attr + \":\", os.path.basename(tree_path))\n",
    "            mug_dict[attr][\"tree\"] = tree         "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "interested-piano",
   "metadata": {},
   "source": [
    "## Genomic Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "interested-corrections",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.set_index(metadata_df.columns[0], inplace=True)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "green-strike",
   "metadata": {},
   "source": [
    "## Populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "meaning-manufacturer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['0.PRE', '0.PE', '0.ANT', '0.ANT4', '3.ANT', '4.ANT', '2.ANT', '2.MED', '1.PRE', '1.ANT', '1.IN', '1.ORI']\n"
     ]
    }
   ],
   "source": [
    "population_list = []\n",
    "\n",
    "for t in divtree.get_terminals():\n",
    "    population = metadata_df[\"population\"][t.name]\n",
    "    if population not in population_list:\n",
    "        population_list.append(population)        \n",
    "        \n",
    "print(population_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "possible-fiber",
   "metadata": {},
   "source": [
    "## Most Recent Sampling Date for BEAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "essential-factory",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.PRE -1686.0\n",
      "0.PE 2014.0\n",
      "0.ANT 2019.0\n",
      "0.ANT4 765.0\n",
      "3.ANT 2017.0\n",
      "4.ANT 2015.0\n",
      "2.ANT 2008.0\n",
      "2.MED 2018.0\n",
      "1.PRE 1767.5\n",
      "1.ANT 2004.0\n",
      "1.IN 2008.0\n",
      "1.ORI 2016.0\n"
     ]
    }
   ],
   "source": [
    "out_path = os.path.join(augur_dir, \"most_recent_sampling_dates.tsv\")\n",
    "mrsd_dict = {}\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    for population in population_list:\n",
    "        df = metadata_df[metadata_df[\"population\"] == population]\n",
    "        max_date = max(df[\"date_mean\"])\n",
    "        print(population, max_date)\n",
    "        outfile.write(\"{}\\t{}\\n\".format(population, max_date))\n",
    "        mrsd_dict[population] = max_date   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "entitled-refund",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. MERGE METADATA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "meaningful-thumb",
   "metadata": {},
   "source": [
    "## Trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "weekly-lambda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "      <th>sequencing_technology</th>\n",
       "      <th>assembly_method</th>\n",
       "      <th>host_raw</th>\n",
       "      <th>host_order</th>\n",
       "      <th>population_color</th>\n",
       "      <th>population</th>\n",
       "      <th>geometry</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>population_rtt_dist</th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>timetree_date</th>\n",
       "      <th>timetree_CI_height</th>\n",
       "      <th>timetree_CI_date</th>\n",
       "      <th>mugration_country</th>\n",
       "      <th>mugration_province</th>\n",
       "      <th>mugration_continent</th>\n",
       "      <th>mugration_host_order</th>\n",
       "      <th>mugration_population</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Human</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Human</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>POINT (-105.607716 38.7251776)</td>\n",
       "      <td>7.31686e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.3815e-06</td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.004600e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-29</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>North America</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.ORI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Human</td>\n",
       "      <td>IonTorrent</td>\n",
       "      <td>Newbler v. 2.6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#c5e88a</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>POINT (40.7957942 47.6222451)</td>\n",
       "      <td>7.30501e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.6582e-06</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.120100e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-98</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Human</td>\n",
       "      <td>2.MED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>5.41847e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.15566e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-15</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>0.ANT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Frontopsylla elatoides</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>5.47035e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.20754e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.356000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>0.ANT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>5.4939e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.23109e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.711000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>0.ANT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE595</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.207000e-07</td>\n",
       "      <td>13.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-111.576</td>\n",
       "      <td>{89.1935,137.589}</td>\n",
       "      <td>{-137.589,-89.1935}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.ORI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE596</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.356000e-07</td>\n",
       "      <td>13.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-111.576</td>\n",
       "      <td>{89.1935,137.589}</td>\n",
       "      <td>{-137.589,-89.1935}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.ORI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE597</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.921000e-07</td>\n",
       "      <td>46.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-78.3728</td>\n",
       "      <td>{50.8548,116.815}</td>\n",
       "      <td>{-116.815,-50.8548}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.ORI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE598</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>5.010000e-08</td>\n",
       "      <td>34.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-70.382</td>\n",
       "      <td>{40.845,106.6}</td>\n",
       "      <td>{-106.6,-40.845}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.ORI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE599</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.248000e-07</td>\n",
       "      <td>58.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-54.3879</td>\n",
       "      <td>{27.1416,88.513}</td>\n",
       "      <td>{-88.513,-27.1416}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.ORI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1201 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    strain    date date_bp  \\\n",
       "sample                                                       \n",
       "Reference                             CO92    1992     -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic   9_10  1923.0     -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic  42126  2006.0     -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic  42123  2005.0     -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic  42118  2005.0     -16   \n",
       "...                                    ...     ...     ...   \n",
       "NODE595                                 NA      NA      NA   \n",
       "NODE596                                 NA      NA      NA   \n",
       "NODE597                                 NA      NA      NA   \n",
       "NODE598                                 NA      NA      NA   \n",
       "NODE599                                 NA      NA      NA   \n",
       "\n",
       "                                                      country       province  \\\n",
       "sample                                                                         \n",
       "Reference                            United States of America       Colorado   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia  Rostov Oblast   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China       Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China       Xinjiang   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China       Xinjiang   \n",
       "...                                                       ...            ...   \n",
       "NODE595                                                    NA             NA   \n",
       "NODE596                                                    NA             NA   \n",
       "NODE597                                                    NA             NA   \n",
       "NODE598                                                    NA             NA   \n",
       "NODE599                                                    NA             NA   \n",
       "\n",
       "                                    country_lat country_lon province_lat  \\\n",
       "sample                                                                     \n",
       "Reference                               39.7837    -100.446      38.7252   \n",
       "GCA_009909635.1_ASM990963v1_genomic     64.6863     97.7453      47.6222   \n",
       "GCA_009669545.1_ASM966954v1_genomic     35.0001         105      42.4805   \n",
       "GCA_009669555.1_ASM966955v1_genomic     35.0001         105      42.4805   \n",
       "GCA_009669565.1_ASM966956v1_genomic     35.0001         105      42.4805   \n",
       "...                                         ...         ...          ...   \n",
       "NODE595                                      NA          NA           NA   \n",
       "NODE596                                      NA          NA           NA   \n",
       "NODE597                                      NA          NA           NA   \n",
       "NODE598                                      NA          NA           NA   \n",
       "NODE599                                      NA          NA           NA   \n",
       "\n",
       "                                    province_lon      biovar branch_major  \\\n",
       "sample                                                                      \n",
       "Reference                               -105.608  Orientalis        1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic      40.7958  Medievalis        2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic      85.4633     Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic      85.4633     Antiqua        0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic      85.4633     Antiqua        0.ANT   \n",
       "...                                          ...         ...          ...   \n",
       "NODE595                                       NA          NA           NA   \n",
       "NODE596                                       NA          NA           NA   \n",
       "NODE597                                       NA          NA           NA   \n",
       "NODE598                                       NA          NA           NA   \n",
       "NODE599                                       NA          NA           NA   \n",
       "\n",
       "                                    branch_minor biosample_accession  \\\n",
       "sample                                                                 \n",
       "Reference                                 1.ORI1        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.ANT1        SAMN07722923   \n",
       "...                                          ...                 ...   \n",
       "NODE595                                       NA                  NA   \n",
       "NODE596                                       NA                  NA   \n",
       "NODE597                                       NA                  NA   \n",
       "NODE598                                       NA                  NA   \n",
       "NODE599                                       NA                  NA   \n",
       "\n",
       "                                                   biosample_comment  \\\n",
       "sample                                                                 \n",
       "Reference                            KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic            KEEP: Assembly Modern   \n",
       "...                                                              ...   \n",
       "NODE595                                                           NA   \n",
       "NODE596                                                           NA   \n",
       "NODE597                                                           NA   \n",
       "NODE598                                                           NA   \n",
       "NODE599                                                           NA   \n",
       "\n",
       "                                    branch_number      continent date_mean  \\\n",
       "sample                                                                       \n",
       "Reference                                       1  North America      1992   \n",
       "GCA_009909635.1_ASM990963v1_genomic             2         Europe      1923   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0           Asia      2006   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0           Asia      2005   \n",
       "GCA_009669565.1_ASM966956v1_genomic             0           Asia      2005   \n",
       "...                                           ...            ...       ...   \n",
       "NODE595                                        NA             NA        NA   \n",
       "NODE596                                        NA             NA        NA   \n",
       "NODE597                                        NA             NA        NA   \n",
       "NODE598                                        NA             NA        NA   \n",
       "NODE599                                        NA             NA        NA   \n",
       "\n",
       "                                    date_bp_mean date_err      lat      lon  \\\n",
       "sample                                                                        \n",
       "Reference                                     29        0  38.7252 -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic           98        0  47.6222  40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic           15        0  42.4805  85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic           16        0  42.4805  85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic           16        0  42.4805  85.4633   \n",
       "...                                          ...      ...      ...      ...   \n",
       "NODE595                                       NA       NA       NA       NA   \n",
       "NODE596                                       NA       NA       NA       NA   \n",
       "NODE597                                       NA       NA       NA       NA   \n",
       "NODE598                                       NA       NA       NA       NA   \n",
       "NODE599                                       NA       NA       NA       NA   \n",
       "\n",
       "                                    host_human sequencing_technology  \\\n",
       "sample                                                                 \n",
       "Reference                                Human                    NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Human            IonTorrent   \n",
       "GCA_009669545.1_ASM966954v1_genomic  Non-Human   Illumina Hiseq 2000   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Non-Human   Illumina Hiseq 2000   \n",
       "GCA_009669565.1_ASM966956v1_genomic  Non-Human   Illumina Hiseq 2000   \n",
       "...                                        ...                   ...   \n",
       "NODE595                                     NA                    NA   \n",
       "NODE596                                     NA                    NA   \n",
       "NODE597                                     NA                    NA   \n",
       "NODE598                                     NA                    NA   \n",
       "NODE599                                     NA                    NA   \n",
       "\n",
       "                                        assembly_method  \\\n",
       "sample                                                    \n",
       "Reference                                            NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Newbler v. 2.6   \n",
       "GCA_009669545.1_ASM966954v1_genomic  SOAPdenovo v. 2.04   \n",
       "GCA_009669555.1_ASM966955v1_genomic  SOAPdenovo v. 2.04   \n",
       "GCA_009669565.1_ASM966956v1_genomic  SOAPdenovo v. 2.04   \n",
       "...                                                 ...   \n",
       "NODE595                                              NA   \n",
       "NODE596                                              NA   \n",
       "NODE597                                              NA   \n",
       "NODE598                                              NA   \n",
       "NODE599                                              NA   \n",
       "\n",
       "                                                   host_raw    host_order  \\\n",
       "sample                                                                      \n",
       "Reference                                             Human         Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic            Homo sapiens         Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic      Citellus undulatus      Rodentia   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Frontopsylla elatoides  Siphonaptera   \n",
       "GCA_009669565.1_ASM966956v1_genomic      Citellus undulatus      Rodentia   \n",
       "...                                                     ...           ...   \n",
       "NODE595                                                  NA            NA   \n",
       "NODE596                                                  NA            NA   \n",
       "NODE597                                                  NA            NA   \n",
       "NODE598                                                  NA            NA   \n",
       "NODE599                                                  NA            NA   \n",
       "\n",
       "                                    population_color population  \\\n",
       "sample                                                            \n",
       "Reference                                    #ff0000      1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic          #c5e88a      2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          #238af5      0.ANT   \n",
       "...                                              ...        ...   \n",
       "NODE595                                           NA         NA   \n",
       "NODE596                                           NA         NA   \n",
       "NODE597                                           NA         NA   \n",
       "NODE598                                           NA         NA   \n",
       "NODE599                                           NA         NA   \n",
       "\n",
       "                                                                 geometry  \\\n",
       "sample                                                                      \n",
       "Reference                                  POINT (-105.607716 38.7251776)   \n",
       "GCA_009909635.1_ASM990963v1_genomic         POINT (40.7957942 47.6222451)   \n",
       "GCA_009669545.1_ASM966954v1_genomic  POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669555.1_ASM966955v1_genomic  POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669565.1_ASM966956v1_genomic  POINT (85.46334640000001 42.4804953)   \n",
       "...                                                                   ...   \n",
       "NODE595                                                                NA   \n",
       "NODE596                                                                NA   \n",
       "NODE597                                                                NA   \n",
       "NODE598                                                                NA   \n",
       "NODE599                                                                NA   \n",
       "\n",
       "                                    root_rtt_dist clade_rtt_dist  \\\n",
       "sample                                                             \n",
       "Reference                             7.31686e-05             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic   7.30501e-05             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic   5.41847e-05             NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic   5.47035e-05             NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic    5.4939e-05             NA   \n",
       "...                                           ...            ...   \n",
       "NODE595                                        NA             NA   \n",
       "NODE596                                        NA             NA   \n",
       "NODE597                                        NA             NA   \n",
       "NODE598                                        NA             NA   \n",
       "NODE599                                        NA             NA   \n",
       "\n",
       "                                    population_rtt_dist node_type  \\\n",
       "sample                                                              \n",
       "Reference                                    6.3815e-06  terminal   \n",
       "GCA_009909635.1_ASM990963v1_genomic          9.6582e-06  terminal   \n",
       "GCA_009669545.1_ASM966954v1_genomic         1.15566e-05  terminal   \n",
       "GCA_009669555.1_ASM966955v1_genomic         1.20754e-05  terminal   \n",
       "GCA_009669565.1_ASM966956v1_genomic         1.23109e-05  terminal   \n",
       "...                                                 ...       ...   \n",
       "NODE595                                              NA  internal   \n",
       "NODE596                                              NA  internal   \n",
       "NODE597                                              NA  internal   \n",
       "NODE598                                              NA  internal   \n",
       "NODE599                                              NA  internal   \n",
       "\n",
       "                                     branch_length  branch_support  \\\n",
       "sample                                                               \n",
       "Reference                             4.004600e-06             0.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic   2.120100e-06             0.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic   0.000000e+00             0.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic   2.356000e-07             0.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic   4.711000e-07             0.0   \n",
       "...                                            ...             ...   \n",
       "NODE595                               2.207000e-07            13.0   \n",
       "NODE596                               2.356000e-07            13.0   \n",
       "NODE597                               2.921000e-07            46.0   \n",
       "NODE598                               5.010000e-08            34.0   \n",
       "NODE599                               2.248000e-07            58.0   \n",
       "\n",
       "                                    branch_support_conf_char  \\\n",
       "sample                                                         \n",
       "Reference                                                      \n",
       "GCA_009909635.1_ASM990963v1_genomic                            \n",
       "GCA_009669545.1_ASM966954v1_genomic                            \n",
       "GCA_009669555.1_ASM966955v1_genomic                            \n",
       "GCA_009669565.1_ASM966956v1_genomic                            \n",
       "...                                                      ...   \n",
       "NODE595                                                        \n",
       "NODE596                                                        \n",
       "NODE597                                                        \n",
       "NODE598                                                        \n",
       "NODE599                                                        \n",
       "\n",
       "                                    branch_support_conf_category  \\\n",
       "sample                                                             \n",
       "Reference                                                    LOW   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          LOW   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          LOW   \n",
       "GCA_009669555.1_ASM966955v1_genomic                          LOW   \n",
       "GCA_009669565.1_ASM966956v1_genomic                          LOW   \n",
       "...                                                          ...   \n",
       "NODE595                                                      LOW   \n",
       "NODE596                                                      LOW   \n",
       "NODE597                                                      LOW   \n",
       "NODE598                                                      LOW   \n",
       "NODE599                                                      LOW   \n",
       "\n",
       "                                    timetree_date timetree_CI_height  \\\n",
       "sample                                                                 \n",
       "Reference                                     -29                 NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic           -98                 NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic           -15                 NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic           -16                 NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic           -16                 NA   \n",
       "...                                           ...                ...   \n",
       "NODE595                                  -111.576  {89.1935,137.589}   \n",
       "NODE596                                  -111.576  {89.1935,137.589}   \n",
       "NODE597                                  -78.3728  {50.8548,116.815}   \n",
       "NODE598                                   -70.382     {40.845,106.6}   \n",
       "NODE599                                  -54.3879   {27.1416,88.513}   \n",
       "\n",
       "                                        timetree_CI_date  \\\n",
       "sample                                                     \n",
       "Reference                                             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic                   NA   \n",
       "...                                                  ...   \n",
       "NODE595                              {-137.589,-89.1935}   \n",
       "NODE596                              {-137.589,-89.1935}   \n",
       "NODE597                              {-116.815,-50.8548}   \n",
       "NODE598                                 {-106.6,-40.845}   \n",
       "NODE599                               {-88.513,-27.1416}   \n",
       "\n",
       "                                            mugration_country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "NODE595                                                  Peru   \n",
       "NODE596                                                  Peru   \n",
       "NODE597                                                  Peru   \n",
       "NODE598                                                  Peru   \n",
       "NODE599                                                  Peru   \n",
       "\n",
       "                                    mugration_province mugration_continent  \\\n",
       "sample                                                                       \n",
       "Reference                                     Colorado       North America   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Rostov Oblast              Europe   \n",
       "GCA_009669545.1_ASM966954v1_genomic           Xinjiang                Asia   \n",
       "GCA_009669555.1_ASM966955v1_genomic           Xinjiang                Asia   \n",
       "GCA_009669565.1_ASM966956v1_genomic           Xinjiang                Asia   \n",
       "...                                                ...                 ...   \n",
       "NODE595                                      Cajamarca       South America   \n",
       "NODE596                                      Cajamarca       South America   \n",
       "NODE597                                      Cajamarca       South America   \n",
       "NODE598                                      Cajamarca       South America   \n",
       "NODE599                                      Cajamarca       South America   \n",
       "\n",
       "                                    mugration_host_order mugration_population  \n",
       "sample                                                                         \n",
       "Reference                                          Human                1.ORI  \n",
       "GCA_009909635.1_ASM990963v1_genomic                Human                2.MED  \n",
       "GCA_009669545.1_ASM966954v1_genomic             Rodentia                0.ANT  \n",
       "GCA_009669555.1_ASM966955v1_genomic         Siphonaptera                0.ANT  \n",
       "GCA_009669565.1_ASM966956v1_genomic             Rodentia                0.ANT  \n",
       "...                                                  ...                  ...  \n",
       "NODE595                                            Human                1.ORI  \n",
       "NODE596                                         Rodentia                1.ORI  \n",
       "NODE597                                         Rodentia                1.ORI  \n",
       "NODE598                                            Human                1.ORI  \n",
       "NODE599                                         Rodentia                1.ORI  \n",
       "\n",
       "[1201 rows x 45 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check if they're synchronized\n",
    "node_i = 0\n",
    "\n",
    "trees = [divtree, timetree] + [mug_dict[attr][\"tree\"] for attr in mug_dict]\n",
    "divtree_num_clades = len([c for c in divtree.find_clades()])\n",
    "divtree_tip_names = [t.name for t in divtree.get_terminals()]\n",
    "\n",
    "# Check for synchrony\n",
    "# Note, no checking for internal order yet\n",
    "for tree in trees:\n",
    "    clades = [c for c in tree.find_clades()]\n",
    "    tip_names = [t.name for t in tree.get_terminals()]\n",
    "    if (tip_names != divtree_tip_names) or (len(clades) != divtree_num_clades):\n",
    "        print(\"ERROR. Trees are not synchronized.\")               \n",
    "\n",
    "\n",
    "# Rename internal nodes\n",
    "for tree in trees:        \n",
    "    node_i = 0\n",
    "    for c in tree.find_clades():\n",
    "        if not c.name or \"NODE\" in c.name:\n",
    "            c.name = \"NODE{}\".format(node_i)\n",
    "            node_i += 1\n",
    "\n",
    "# Extract metadata from divtree\n",
    "for c in divtree.find_clades():\n",
    "    \n",
    "    # ---------------\n",
    "    # Node type\n",
    "    node_type = \"internal\"\n",
    "    if c.is_terminal():\n",
    "        node_type = \"terminal\"\n",
    "    metadata_df.at[c.name, \"node_type\"] = node_type\n",
    "\n",
    "    # ---------------\n",
    "    # Branch Length\n",
    "    branch_length = c.branch_length\n",
    "    if not branch_length:\n",
    "        branch_length = 0\n",
    "    metadata_df.at[c.name, \"branch_length\"] = branch_length    \n",
    "\n",
    "    # ---------------\n",
    "    # Branch Support\n",
    "    # Give terminal nodes 0% confidence\n",
    "    branch_support = c.confidence\n",
    "    if c.is_terminal():\n",
    "        branch_support = 0\n",
    "    # Internal nodes without confidence can be 0\n",
    "    elif not c.confidence:\n",
    "        branch_support = 0\n",
    "    metadata_df.at[c.name, \"branch_support\"] = branch_support\n",
    "\n",
    "    # Confidence category\n",
    "    branch_support_conf_char = \"\"\n",
    "    if branch_support >= 95 and not c.is_terminal():\n",
    "        branch_support_conf_char = \"*\"\n",
    "    metadata_df.at[c.name, \"branch_support_conf_char\"] = branch_support_conf_char\n",
    "    if branch_support >= 95:\n",
    "        metadata_df.at[c.name, \"branch_support_conf_category\"] = \"HIGH\"\n",
    "    else:\n",
    "        metadata_df.at[c.name, \"branch_support_conf_category\"] = \"LOW\"\n",
    "\n",
    "# Extract metadata from timetree\n",
    "for c in timetree.find_clades():                    \n",
    "    if c.comment:\n",
    "        comment_dict = parse_comment(c.comment)\n",
    "        for param,val in comment_dict.items():\n",
    "            rename_param = \"timetree_\" + param\n",
    "            metadata_df.at[c.name, rename_param] = val\n",
    "\n",
    "# Extract metadata from timetree\n",
    "for attr in mug_dict:\n",
    "    tree = mug_dict[attr][\"tree\"]\n",
    "    \n",
    "    for c in tree.find_clades():\n",
    "        if c.comment:\n",
    "            comment_dict = parse_comment(c.comment)\n",
    "            for param,val in comment_dict.items():\n",
    "                rename_param = \"mugration_\" + param\n",
    "                metadata_df.at[c.name, rename_param] = val \n",
    "            \n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "display(metadata_df)       "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "compound-kuwait",
   "metadata": {},
   "source": [
    "## Add Mugration Confidence and Handle Root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "overall-amber",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "strain                                     NA\n",
       "date                                       NA\n",
       "date_bp                                    NA\n",
       "country                                    NA\n",
       "province                                   NA\n",
       "                                       ...   \n",
       "mugration_province_lat_confidence    0.561953\n",
       "mugration_province_lon_confidence    0.561953\n",
       "mugration_continent_confidence           0.52\n",
       "mugration_host_order_confidence          0.93\n",
       "mugration_population_confidence           0.5\n",
       "Name: NODE0, Length: 54, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_list = [mug_dict[attr][\"conf_df\"] for attr in mug_dict]\n",
    "\n",
    "for df,attr in zip(df_list, MUG_ATTRIBUTE_LIST):\n",
    "    for sample in df.index:\n",
    "        max_state = \"\"\n",
    "        max_val = 0        \n",
    "        for state,val in zip(df.columns, df.loc[sample]):\n",
    "            if val >= max_val:\n",
    "                max_val = val\n",
    "                max_state = state\n",
    "                \n",
    "        metadata_df.at[sample,\"mugration_\" + attr] = max_state      \n",
    "        metadata_df.at[sample,\"mugration_\" + attr + \"_confidence\"] = round(max_val,2)\n",
    "        \n",
    "        #if the attr is country or province, extend this confidence to lat and lon\n",
    "        if attr == \"country\" or attr == \"province\":\n",
    "            metadata_df.at[sample,\"mugration_\" + attr + \"_lat_confidence\"] = max_val\n",
    "            metadata_df.at[sample,\"mugration_\" + attr + \"_lon_confidence\"] = max_val\n",
    "\n",
    "display(metadata_df.loc[\"NODE0\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "administrative-gross",
   "metadata": {},
   "source": [
    "## Latitude and Longtitude"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "mexican-outside",
   "metadata": {},
   "outputs": [],
   "source": [
    "latlon_country_df = pd.DataFrame()\n",
    "latlon_province_df = pd.DataFrame()\n",
    "\n",
    "df_list = [latlon_country_df, latlon_province_df]\n",
    "attr_list = [\"country\", \"province\"]\n",
    "conf_df_list = [mug_dict[\"country\"][\"conf_df\"], mug_dict[\"province\"][\"conf_df\"]]\n",
    "\n",
    "# Create a mapping of geo name to lat,lon\n",
    "for df,attr in zip(df_list, attr_list):\n",
    "    # Get data\n",
    "    for rec in metadata_df.iterrows():\n",
    "        node_name = rec[0]\n",
    "        node_type = rec[1][\"node_type\"]\n",
    "        name = rec[1][attr]\n",
    "        country = rec[1][\"country\"]\n",
    "        \n",
    "        if node_type == \"internal\":\n",
    "            continue\n",
    "        \n",
    "        if attr == \"province\" and name == NO_DATA_CHAR and node_type == \"terminal\" and country != \"Russia\":\n",
    "            # Use country instead\n",
    "            name = rec[1][\"country\"]\n",
    "            lat = rec[1][\"country_lat\"]\n",
    "            lon = rec[1][\"country_lon\"]\n",
    "        else:\n",
    "            lat = rec[1][attr + \"_lat\"]\n",
    "            lon = rec[1][attr + \"_lon\"]             \n",
    "\n",
    "\n",
    "        if name not in df.index:\n",
    "            df.at[name, \"lat\"] = lat\n",
    "            df.at[name, \"lon\"] = lon\n",
    "            df.at[name, \"size\"] = 1\n",
    "        else:\n",
    "            df[\"size\"][name] += 1\n",
    "\n",
    "#display(latlon_country_df)\n",
    "#display(latlon_province_df)\n",
    "\n",
    "\n",
    "# Update lat lon for mugration\n",
    "for latlon_df, conf_df, attr in zip(df_list, conf_df_list, attr_list):\n",
    "    for rec in metadata_df.iterrows():\n",
    "        sample = rec[0]\n",
    "        name = rec[1][\"mugration_\" + attr]  \n",
    "        node_type = rec[1][\"node_type\"]  \n",
    "        country = rec[1][\"country\"]  \n",
    "        \n",
    "        if attr == \"province\" and rec[1][attr]  == NO_DATA_CHAR and node_type == \"terminal\" and country != \"Russia\":\n",
    "            # Use country instead\n",
    "            name = rec[1][\"country\"]\n",
    "            metadata_df.at[sample,\"mugration_\" + attr] = name\n",
    "        \n",
    "        lat = latlon_df[\"lat\"][name]\n",
    "        lon = latlon_df[\"lon\"][name]\n",
    "        \n",
    "                \n",
    "        metadata_df.at[sample, \"mugration_\" + attr +\"_lat\"] = lat\n",
    "        metadata_df.at[sample, \"mugration_\" + attr +\"_lon\"] = lon\n",
    "        \n",
    "        \n",
    "# Mapping file for auspice\n",
    "out_path_latlon = os.path.join(augur_dir, \"latlon.tsv\")\n",
    "\n",
    "# Countries\n",
    "with open(out_path_latlon, \"w\") as outfile:\n",
    "    for country in latlon_country_df.index:\n",
    "        lat = str(latlon_country_df[\"lat\"][country])\n",
    "        lon = str(latlon_country_df[\"lon\"][country])\n",
    "        outfile.write(\"country\" + \"\\t\" + country + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "# Provinces\n",
    "with open(out_path_latlon, \"a\") as outfile:\n",
    "    for province in latlon_province_df.index:\n",
    "        lat = str(latlon_province_df[\"lat\"][province])\n",
    "        lon = str(latlon_province_df[\"lon\"][province])\n",
    "        outfile.write(\"province\" + \"\\t\" + province + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "        \n",
    "#display(metadata_df[metadata_df[\"continent\"] == \"Africa\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "compound-fifteen",
   "metadata": {},
   "source": [
    "## Add Sub  and Spread Rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "chinese-anxiety",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "      <th>sequencing_technology</th>\n",
       "      <th>assembly_method</th>\n",
       "      <th>host_raw</th>\n",
       "      <th>host_order</th>\n",
       "      <th>population_color</th>\n",
       "      <th>population</th>\n",
       "      <th>geometry</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>population_rtt_dist</th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>timetree_date</th>\n",
       "      <th>timetree_CI_height</th>\n",
       "      <th>timetree_CI_date</th>\n",
       "      <th>mugration_country</th>\n",
       "      <th>mugration_province</th>\n",
       "      <th>mugration_continent</th>\n",
       "      <th>mugration_host_order</th>\n",
       "      <th>mugration_population</th>\n",
       "      <th>mugration_country_confidence</th>\n",
       "      <th>mugration_country_lat_confidence</th>\n",
       "      <th>mugration_country_lon_confidence</th>\n",
       "      <th>mugration_province_confidence</th>\n",
       "      <th>mugration_province_lat_confidence</th>\n",
       "      <th>mugration_province_lon_confidence</th>\n",
       "      <th>mugration_continent_confidence</th>\n",
       "      <th>mugration_host_order_confidence</th>\n",
       "      <th>mugration_population_confidence</th>\n",
       "      <th>mugration_country_lat</th>\n",
       "      <th>mugration_country_lon</th>\n",
       "      <th>mugration_province_lat</th>\n",
       "      <th>mugration_province_lon</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>branch_length_time</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Human</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Human</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>POINT (-105.607716 38.7251776)</td>\n",
       "      <td>7.31686e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.3815e-06</td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.004600e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-29</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>North America</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>16.935846</td>\n",
       "      <td>162.52800</td>\n",
       "      <td>2.463945e-08</td>\n",
       "      <td>0.104203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Human</td>\n",
       "      <td>IonTorrent</td>\n",
       "      <td>Newbler v. 2.6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#c5e88a</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>POINT (40.7957942 47.6222451)</td>\n",
       "      <td>7.30501e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.6582e-06</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.120100e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-98</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Human</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>8.966111</td>\n",
       "      <td>878.89200</td>\n",
       "      <td>2.412242e-09</td>\n",
       "      <td>0.010202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>5.41847e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.15566e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-15</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.23246</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Frontopsylla elatoides</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>5.47035e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.20754e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.356000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>0.996375</td>\n",
       "      <td>24.38400</td>\n",
       "      <td>9.662073e-09</td>\n",
       "      <td>0.040862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>5.4939e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.23109e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.711000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>1.992328</td>\n",
       "      <td>35.90130</td>\n",
       "      <td>1.312209e-08</td>\n",
       "      <td>0.055495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE595</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.207000e-07</td>\n",
       "      <td>13.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-111.576</td>\n",
       "      <td>{89.1935,137.589}</td>\n",
       "      <td>{-137.589,-89.1935}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.999873</td>\n",
       "      <td>0.999873</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.55</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>0.933362</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE596</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.356000e-07</td>\n",
       "      <td>13.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-111.576</td>\n",
       "      <td>{89.1935,137.589}</td>\n",
       "      <td>{-137.589,-89.1935}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.61</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>0.996375</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE597</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.921000e-07</td>\n",
       "      <td>46.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-78.3728</td>\n",
       "      <td>{50.8548,116.815}</td>\n",
       "      <td>{-116.815,-50.8548}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>1.235320</td>\n",
       "      <td>33.20350</td>\n",
       "      <td>8.797265e-09</td>\n",
       "      <td>0.037204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE598</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>5.010000e-08</td>\n",
       "      <td>34.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-70.382</td>\n",
       "      <td>{40.845,106.6}</td>\n",
       "      <td>{-106.6,-40.845}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.999874</td>\n",
       "      <td>0.999874</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.60</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>0.211878</td>\n",
       "      <td>41.19420</td>\n",
       "      <td>1.216191e-09</td>\n",
       "      <td>0.005143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE599</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.248000e-07</td>\n",
       "      <td>58.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-54.3879</td>\n",
       "      <td>{27.1416,88.513}</td>\n",
       "      <td>{-88.513,-27.1416}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.991635</td>\n",
       "      <td>0.991635</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.82</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>0.950701</td>\n",
       "      <td>15.99410</td>\n",
       "      <td>1.405518e-08</td>\n",
       "      <td>0.059441</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1201 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    strain    date date_bp  \\\n",
       "sample                                                       \n",
       "Reference                             CO92    1992     -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic   9_10  1923.0     -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic  42126  2006.0     -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic  42123  2005.0     -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic  42118  2005.0     -16   \n",
       "...                                    ...     ...     ...   \n",
       "NODE595                                 NA      NA      NA   \n",
       "NODE596                                 NA      NA      NA   \n",
       "NODE597                                 NA      NA      NA   \n",
       "NODE598                                 NA      NA      NA   \n",
       "NODE599                                 NA      NA      NA   \n",
       "\n",
       "                                                      country       province  \\\n",
       "sample                                                                         \n",
       "Reference                            United States of America       Colorado   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia  Rostov Oblast   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China       Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China       Xinjiang   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China       Xinjiang   \n",
       "...                                                       ...            ...   \n",
       "NODE595                                                    NA             NA   \n",
       "NODE596                                                    NA             NA   \n",
       "NODE597                                                    NA             NA   \n",
       "NODE598                                                    NA             NA   \n",
       "NODE599                                                    NA             NA   \n",
       "\n",
       "                                    country_lat country_lon province_lat  \\\n",
       "sample                                                                     \n",
       "Reference                               39.7837    -100.446      38.7252   \n",
       "GCA_009909635.1_ASM990963v1_genomic     64.6863     97.7453      47.6222   \n",
       "GCA_009669545.1_ASM966954v1_genomic     35.0001         105      42.4805   \n",
       "GCA_009669555.1_ASM966955v1_genomic     35.0001         105      42.4805   \n",
       "GCA_009669565.1_ASM966956v1_genomic     35.0001         105      42.4805   \n",
       "...                                         ...         ...          ...   \n",
       "NODE595                                      NA          NA           NA   \n",
       "NODE596                                      NA          NA           NA   \n",
       "NODE597                                      NA          NA           NA   \n",
       "NODE598                                      NA          NA           NA   \n",
       "NODE599                                      NA          NA           NA   \n",
       "\n",
       "                                    province_lon      biovar branch_major  \\\n",
       "sample                                                                      \n",
       "Reference                               -105.608  Orientalis        1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic      40.7958  Medievalis        2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic      85.4633     Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic      85.4633     Antiqua        0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic      85.4633     Antiqua        0.ANT   \n",
       "...                                          ...         ...          ...   \n",
       "NODE595                                       NA          NA           NA   \n",
       "NODE596                                       NA          NA           NA   \n",
       "NODE597                                       NA          NA           NA   \n",
       "NODE598                                       NA          NA           NA   \n",
       "NODE599                                       NA          NA           NA   \n",
       "\n",
       "                                    branch_minor biosample_accession  \\\n",
       "sample                                                                 \n",
       "Reference                                 1.ORI1        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.ANT1        SAMN07722923   \n",
       "...                                          ...                 ...   \n",
       "NODE595                                       NA                  NA   \n",
       "NODE596                                       NA                  NA   \n",
       "NODE597                                       NA                  NA   \n",
       "NODE598                                       NA                  NA   \n",
       "NODE599                                       NA                  NA   \n",
       "\n",
       "                                                   biosample_comment  \\\n",
       "sample                                                                 \n",
       "Reference                            KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic            KEEP: Assembly Modern   \n",
       "...                                                              ...   \n",
       "NODE595                                                           NA   \n",
       "NODE596                                                           NA   \n",
       "NODE597                                                           NA   \n",
       "NODE598                                                           NA   \n",
       "NODE599                                                           NA   \n",
       "\n",
       "                                    branch_number      continent date_mean  \\\n",
       "sample                                                                       \n",
       "Reference                                       1  North America      1992   \n",
       "GCA_009909635.1_ASM990963v1_genomic             2         Europe      1923   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0           Asia      2006   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0           Asia      2005   \n",
       "GCA_009669565.1_ASM966956v1_genomic             0           Asia      2005   \n",
       "...                                           ...            ...       ...   \n",
       "NODE595                                        NA             NA        NA   \n",
       "NODE596                                        NA             NA        NA   \n",
       "NODE597                                        NA             NA        NA   \n",
       "NODE598                                        NA             NA        NA   \n",
       "NODE599                                        NA             NA        NA   \n",
       "\n",
       "                                    date_bp_mean date_err      lat      lon  \\\n",
       "sample                                                                        \n",
       "Reference                                     29        0  38.7252 -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic           98        0  47.6222  40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic           15        0  42.4805  85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic           16        0  42.4805  85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic           16        0  42.4805  85.4633   \n",
       "...                                          ...      ...      ...      ...   \n",
       "NODE595                                       NA       NA       NA       NA   \n",
       "NODE596                                       NA       NA       NA       NA   \n",
       "NODE597                                       NA       NA       NA       NA   \n",
       "NODE598                                       NA       NA       NA       NA   \n",
       "NODE599                                       NA       NA       NA       NA   \n",
       "\n",
       "                                    host_human sequencing_technology  \\\n",
       "sample                                                                 \n",
       "Reference                                Human                    NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Human            IonTorrent   \n",
       "GCA_009669545.1_ASM966954v1_genomic  Non-Human   Illumina Hiseq 2000   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Non-Human   Illumina Hiseq 2000   \n",
       "GCA_009669565.1_ASM966956v1_genomic  Non-Human   Illumina Hiseq 2000   \n",
       "...                                        ...                   ...   \n",
       "NODE595                                     NA                    NA   \n",
       "NODE596                                     NA                    NA   \n",
       "NODE597                                     NA                    NA   \n",
       "NODE598                                     NA                    NA   \n",
       "NODE599                                     NA                    NA   \n",
       "\n",
       "                                        assembly_method  \\\n",
       "sample                                                    \n",
       "Reference                                            NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Newbler v. 2.6   \n",
       "GCA_009669545.1_ASM966954v1_genomic  SOAPdenovo v. 2.04   \n",
       "GCA_009669555.1_ASM966955v1_genomic  SOAPdenovo v. 2.04   \n",
       "GCA_009669565.1_ASM966956v1_genomic  SOAPdenovo v. 2.04   \n",
       "...                                                 ...   \n",
       "NODE595                                              NA   \n",
       "NODE596                                              NA   \n",
       "NODE597                                              NA   \n",
       "NODE598                                              NA   \n",
       "NODE599                                              NA   \n",
       "\n",
       "                                                   host_raw    host_order  \\\n",
       "sample                                                                      \n",
       "Reference                                             Human         Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic            Homo sapiens         Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic      Citellus undulatus      Rodentia   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Frontopsylla elatoides  Siphonaptera   \n",
       "GCA_009669565.1_ASM966956v1_genomic      Citellus undulatus      Rodentia   \n",
       "...                                                     ...           ...   \n",
       "NODE595                                                  NA            NA   \n",
       "NODE596                                                  NA            NA   \n",
       "NODE597                                                  NA            NA   \n",
       "NODE598                                                  NA            NA   \n",
       "NODE599                                                  NA            NA   \n",
       "\n",
       "                                    population_color population  \\\n",
       "sample                                                            \n",
       "Reference                                    #ff0000      1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic          #c5e88a      2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          #238af5      0.ANT   \n",
       "...                                              ...        ...   \n",
       "NODE595                                           NA         NA   \n",
       "NODE596                                           NA         NA   \n",
       "NODE597                                           NA         NA   \n",
       "NODE598                                           NA         NA   \n",
       "NODE599                                           NA         NA   \n",
       "\n",
       "                                                                 geometry  \\\n",
       "sample                                                                      \n",
       "Reference                                  POINT (-105.607716 38.7251776)   \n",
       "GCA_009909635.1_ASM990963v1_genomic         POINT (40.7957942 47.6222451)   \n",
       "GCA_009669545.1_ASM966954v1_genomic  POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669555.1_ASM966955v1_genomic  POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669565.1_ASM966956v1_genomic  POINT (85.46334640000001 42.4804953)   \n",
       "...                                                                   ...   \n",
       "NODE595                                                                NA   \n",
       "NODE596                                                                NA   \n",
       "NODE597                                                                NA   \n",
       "NODE598                                                                NA   \n",
       "NODE599                                                                NA   \n",
       "\n",
       "                                    root_rtt_dist clade_rtt_dist  \\\n",
       "sample                                                             \n",
       "Reference                             7.31686e-05             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic   7.30501e-05             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic   5.41847e-05             NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic   5.47035e-05             NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic    5.4939e-05             NA   \n",
       "...                                           ...            ...   \n",
       "NODE595                                        NA             NA   \n",
       "NODE596                                        NA             NA   \n",
       "NODE597                                        NA             NA   \n",
       "NODE598                                        NA             NA   \n",
       "NODE599                                        NA             NA   \n",
       "\n",
       "                                    population_rtt_dist node_type  \\\n",
       "sample                                                              \n",
       "Reference                                    6.3815e-06  terminal   \n",
       "GCA_009909635.1_ASM990963v1_genomic          9.6582e-06  terminal   \n",
       "GCA_009669545.1_ASM966954v1_genomic         1.15566e-05  terminal   \n",
       "GCA_009669555.1_ASM966955v1_genomic         1.20754e-05  terminal   \n",
       "GCA_009669565.1_ASM966956v1_genomic         1.23109e-05  terminal   \n",
       "...                                                 ...       ...   \n",
       "NODE595                                              NA  internal   \n",
       "NODE596                                              NA  internal   \n",
       "NODE597                                              NA  internal   \n",
       "NODE598                                              NA  internal   \n",
       "NODE599                                              NA  internal   \n",
       "\n",
       "                                     branch_length  branch_support  \\\n",
       "sample                                                               \n",
       "Reference                             4.004600e-06             0.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic   2.120100e-06             0.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic   0.000000e+00             0.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic   2.356000e-07             0.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic   4.711000e-07             0.0   \n",
       "...                                            ...             ...   \n",
       "NODE595                               2.207000e-07            13.0   \n",
       "NODE596                               2.356000e-07            13.0   \n",
       "NODE597                               2.921000e-07            46.0   \n",
       "NODE598                               5.010000e-08            34.0   \n",
       "NODE599                               2.248000e-07            58.0   \n",
       "\n",
       "                                    branch_support_conf_char  \\\n",
       "sample                                                         \n",
       "Reference                                                      \n",
       "GCA_009909635.1_ASM990963v1_genomic                            \n",
       "GCA_009669545.1_ASM966954v1_genomic                            \n",
       "GCA_009669555.1_ASM966955v1_genomic                            \n",
       "GCA_009669565.1_ASM966956v1_genomic                            \n",
       "...                                                      ...   \n",
       "NODE595                                                        \n",
       "NODE596                                                        \n",
       "NODE597                                                        \n",
       "NODE598                                                        \n",
       "NODE599                                                        \n",
       "\n",
       "                                    branch_support_conf_category  \\\n",
       "sample                                                             \n",
       "Reference                                                    LOW   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          LOW   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          LOW   \n",
       "GCA_009669555.1_ASM966955v1_genomic                          LOW   \n",
       "GCA_009669565.1_ASM966956v1_genomic                          LOW   \n",
       "...                                                          ...   \n",
       "NODE595                                                      LOW   \n",
       "NODE596                                                      LOW   \n",
       "NODE597                                                      LOW   \n",
       "NODE598                                                      LOW   \n",
       "NODE599                                                      LOW   \n",
       "\n",
       "                                    timetree_date timetree_CI_height  \\\n",
       "sample                                                                 \n",
       "Reference                                     -29                 NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic           -98                 NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic           -15                 NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic           -16                 NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic           -16                 NA   \n",
       "...                                           ...                ...   \n",
       "NODE595                                  -111.576  {89.1935,137.589}   \n",
       "NODE596                                  -111.576  {89.1935,137.589}   \n",
       "NODE597                                  -78.3728  {50.8548,116.815}   \n",
       "NODE598                                   -70.382     {40.845,106.6}   \n",
       "NODE599                                  -54.3879   {27.1416,88.513}   \n",
       "\n",
       "                                        timetree_CI_date  \\\n",
       "sample                                                     \n",
       "Reference                                             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic                   NA   \n",
       "...                                                  ...   \n",
       "NODE595                              {-137.589,-89.1935}   \n",
       "NODE596                              {-137.589,-89.1935}   \n",
       "NODE597                              {-116.815,-50.8548}   \n",
       "NODE598                                 {-106.6,-40.845}   \n",
       "NODE599                               {-88.513,-27.1416}   \n",
       "\n",
       "                                            mugration_country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "NODE595                                                  Peru   \n",
       "NODE596                                                  Peru   \n",
       "NODE597                                                  Peru   \n",
       "NODE598                                                  Peru   \n",
       "NODE599                                                  Peru   \n",
       "\n",
       "                                    mugration_province mugration_continent  \\\n",
       "sample                                                                       \n",
       "Reference                                     Colorado       North America   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Rostov Oblast              Europe   \n",
       "GCA_009669545.1_ASM966954v1_genomic           Xinjiang                Asia   \n",
       "GCA_009669555.1_ASM966955v1_genomic           Xinjiang                Asia   \n",
       "GCA_009669565.1_ASM966956v1_genomic           Xinjiang                Asia   \n",
       "...                                                ...                 ...   \n",
       "NODE595                                      Cajamarca       South America   \n",
       "NODE596                                      Cajamarca       South America   \n",
       "NODE597                                      Cajamarca       South America   \n",
       "NODE598                                      Cajamarca       South America   \n",
       "NODE599                                      Cajamarca       South America   \n",
       "\n",
       "                                    mugration_host_order mugration_population  \\\n",
       "sample                                                                          \n",
       "Reference                                          Human                1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic                Human                2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic             Rodentia                0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic         Siphonaptera                0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic             Rodentia                0.ANT   \n",
       "...                                                  ...                  ...   \n",
       "NODE595                                            Human                1.ORI   \n",
       "NODE596                                         Rodentia                1.ORI   \n",
       "NODE597                                         Rodentia                1.ORI   \n",
       "NODE598                                            Human                1.ORI   \n",
       "NODE599                                         Rodentia                1.ORI   \n",
       "\n",
       "                                     mugration_country_confidence  \\\n",
       "sample                                                              \n",
       "Reference                                                     1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.0   \n",
       "...                                                           ...   \n",
       "NODE595                                                       1.0   \n",
       "NODE596                                                       1.0   \n",
       "NODE597                                                       1.0   \n",
       "NODE598                                                       1.0   \n",
       "NODE599                                                       1.0   \n",
       "\n",
       "                                     mugration_country_lat_confidence  \\\n",
       "sample                                                                  \n",
       "Reference                                                    1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          1.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic                          1.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic                          1.000000   \n",
       "...                                                               ...   \n",
       "NODE595                                                      0.999999   \n",
       "NODE596                                                      0.999999   \n",
       "NODE597                                                      0.999999   \n",
       "NODE598                                                      0.999999   \n",
       "NODE599                                                      0.999999   \n",
       "\n",
       "                                     mugration_country_lon_confidence  \\\n",
       "sample                                                                  \n",
       "Reference                                                    1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          1.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic                          1.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic                          1.000000   \n",
       "...                                                               ...   \n",
       "NODE595                                                      0.999999   \n",
       "NODE596                                                      0.999999   \n",
       "NODE597                                                      0.999999   \n",
       "NODE598                                                      0.999999   \n",
       "NODE599                                                      0.999999   \n",
       "\n",
       "                                     mugration_province_confidence  \\\n",
       "sample                                                               \n",
       "Reference                                                     1.00   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.00   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.00   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.00   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.00   \n",
       "...                                                            ...   \n",
       "NODE595                                                       1.00   \n",
       "NODE596                                                       1.00   \n",
       "NODE597                                                       1.00   \n",
       "NODE598                                                       1.00   \n",
       "NODE599                                                       0.99   \n",
       "\n",
       "                                     mugration_province_lat_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                     1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.000000   \n",
       "...                                                                ...   \n",
       "NODE595                                                       0.999873   \n",
       "NODE596                                                       0.999998   \n",
       "NODE597                                                       0.999999   \n",
       "NODE598                                                       0.999874   \n",
       "NODE599                                                       0.991635   \n",
       "\n",
       "                                     mugration_province_lon_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                     1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.000000   \n",
       "...                                                                ...   \n",
       "NODE595                                                       0.999873   \n",
       "NODE596                                                       0.999998   \n",
       "NODE597                                                       0.999999   \n",
       "NODE598                                                       0.999874   \n",
       "NODE599                                                       0.991635   \n",
       "\n",
       "                                     mugration_continent_confidence  \\\n",
       "sample                                                                \n",
       "Reference                                                       1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                             1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                             1.0   \n",
       "...                                                             ...   \n",
       "NODE595                                                         1.0   \n",
       "NODE596                                                         1.0   \n",
       "NODE597                                                         1.0   \n",
       "NODE598                                                         1.0   \n",
       "NODE599                                                         1.0   \n",
       "\n",
       "                                     mugration_host_order_confidence  \\\n",
       "sample                                                                 \n",
       "Reference                                                       1.00   \n",
       "GCA_009909635.1_ASM990963v1_genomic                             1.00   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             1.00   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             1.00   \n",
       "GCA_009669565.1_ASM966956v1_genomic                             1.00   \n",
       "...                                                              ...   \n",
       "NODE595                                                         0.55   \n",
       "NODE596                                                         0.61   \n",
       "NODE597                                                         0.56   \n",
       "NODE598                                                         0.60   \n",
       "NODE599                                                         0.82   \n",
       "\n",
       "                                     mugration_population_confidence  \\\n",
       "sample                                                                 \n",
       "Reference                                                        1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                              1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                              1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                              1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                              1.0   \n",
       "...                                                              ...   \n",
       "NODE595                                                          1.0   \n",
       "NODE596                                                          1.0   \n",
       "NODE597                                                          1.0   \n",
       "NODE598                                                          1.0   \n",
       "NODE599                                                          1.0   \n",
       "\n",
       "                                     mugration_country_lat  \\\n",
       "sample                                                       \n",
       "Reference                                        39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic              64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic              35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic              35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic              35.000074   \n",
       "...                                                    ...   \n",
       "NODE595                                          -6.869970   \n",
       "NODE596                                          -6.869970   \n",
       "NODE597                                          -6.869970   \n",
       "NODE598                                          -6.869970   \n",
       "NODE599                                          -6.869970   \n",
       "\n",
       "                                     mugration_country_lon  \\\n",
       "sample                                                       \n",
       "Reference                                      -100.445882   \n",
       "GCA_009909635.1_ASM990963v1_genomic              97.745306   \n",
       "GCA_009669545.1_ASM966954v1_genomic             104.999927   \n",
       "GCA_009669555.1_ASM966955v1_genomic             104.999927   \n",
       "GCA_009669565.1_ASM966956v1_genomic             104.999927   \n",
       "...                                                    ...   \n",
       "NODE595                                         -75.045851   \n",
       "NODE596                                         -75.045851   \n",
       "NODE597                                         -75.045851   \n",
       "NODE598                                         -75.045851   \n",
       "NODE599                                         -75.045851   \n",
       "\n",
       "                                     mugration_province_lat  \\\n",
       "sample                                                        \n",
       "Reference                                         38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic               47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic               42.480495   \n",
       "GCA_009669555.1_ASM966955v1_genomic               42.480495   \n",
       "GCA_009669565.1_ASM966956v1_genomic               42.480495   \n",
       "...                                                     ...   \n",
       "NODE595                                           -6.250000   \n",
       "NODE596                                           -6.250000   \n",
       "NODE597                                           -6.250000   \n",
       "NODE598                                           -6.250000   \n",
       "NODE599                                           -6.250000   \n",
       "\n",
       "                                     mugration_province_lon  \\\n",
       "sample                                                        \n",
       "Reference                                       -105.607716   \n",
       "GCA_009909635.1_ASM990963v1_genomic               40.795794   \n",
       "GCA_009669545.1_ASM966954v1_genomic               85.463346   \n",
       "GCA_009669555.1_ASM966955v1_genomic               85.463346   \n",
       "GCA_009669565.1_ASM966956v1_genomic               85.463346   \n",
       "...                                                     ...   \n",
       "NODE595                                          -78.833333   \n",
       "NODE596                                          -78.833333   \n",
       "NODE597                                          -78.833333   \n",
       "NODE598                                          -78.833333   \n",
       "NODE599                                          -78.833333   \n",
       "\n",
       "                                     branch_length_sub  branch_length_time  \\\n",
       "sample                                                                       \n",
       "Reference                                    16.935846           162.52800   \n",
       "GCA_009909635.1_ASM990963v1_genomic           8.966111           878.89200   \n",
       "GCA_009669545.1_ASM966954v1_genomic           0.000000             9.23246   \n",
       "GCA_009669555.1_ASM966955v1_genomic           0.996375            24.38400   \n",
       "GCA_009669565.1_ASM966956v1_genomic           1.992328            35.90130   \n",
       "...                                                ...                 ...   \n",
       "NODE595                                       0.933362             0.00000   \n",
       "NODE596                                       0.996375             0.00000   \n",
       "NODE597                                       1.235320            33.20350   \n",
       "NODE598                                       0.211878            41.19420   \n",
       "NODE599                                       0.950701            15.99410   \n",
       "\n",
       "                                         rate_sub  rate_sub_year  \n",
       "sample                                                            \n",
       "Reference                            2.463945e-08       0.104203  \n",
       "GCA_009909635.1_ASM990963v1_genomic  2.412242e-09       0.010202  \n",
       "GCA_009669545.1_ASM966954v1_genomic  0.000000e+00       0.000000  \n",
       "GCA_009669555.1_ASM966955v1_genomic  9.662073e-09       0.040862  \n",
       "GCA_009669565.1_ASM966956v1_genomic  1.312209e-08       0.055495  \n",
       "...                                           ...            ...  \n",
       "NODE595                              0.000000e+00       0.000000  \n",
       "NODE596                              0.000000e+00       0.000000  \n",
       "NODE597                              8.797265e-09       0.037204  \n",
       "NODE598                              1.216191e-09       0.005143  \n",
       "NODE599                              1.405518e-08       0.059441  \n",
       "\n",
       "[1201 rows x 62 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# 1. Get Branch Lengths in Num Substitutions\n",
    "\n",
    "for c in divtree.find_clades():\n",
    "    branch_length_sub = c.branch_length * SEQ_LEN\n",
    "    branch_length_time = timetree.common_ancestor(c.name).branch_length\n",
    "    walk_path = [divtree.root] + divtree.get_path(c)\n",
    "    # Root\n",
    "    if len(walk_path) == 1:\n",
    "        parent = walk_path[0]\n",
    "    else:\n",
    "        parent = walk_path[-2]\n",
    "\n",
    "\n",
    "    # 1. Rate as subs/site/year\n",
    "    rate_sub = c.branch_length / branch_length_time  if branch_length_time != 0 else 0\n",
    "    # 2. Rate as subs/year\n",
    "    rate_sub_year = branch_length_sub / branch_length_time  if branch_length_time != 0 else 0\n",
    "    \n",
    "    metadata_df.at[c.name,\"branch_length_sub\"] = branch_length_sub\n",
    "    metadata_df.at[c.name,\"branch_length_time\"] = branch_length_time\n",
    "    metadata_df.at[c.name,\"rate_sub\"] = rate_sub\n",
    "    metadata_df.at[c.name,\"rate_sub_year\"] = rate_sub_year\n",
    "\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "religious-community",
   "metadata": {},
   "source": [
    "## Colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "passing-christopher",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'country': {'russia': '#8000ff', 'lithuania': '#7215ff', 'estonia': '#652afe', 'germany': '#573ffd', 'china': '#4a53fc', 'azerbaijan': '#3c66fa', 'armenia': '#2f79f7', 'georgia': '#228bf4', 'mongolia': '#149df1', 'tajikistan': '#07adee', 'kyrgyzstan': '#07bcea', 'england': '#14c9e5', 'spain': '#22d5e0', 'france': '#2fe0db', 'nepal': '#3cead5', 'india': '#4af1d0', 'kazakhstan': '#57f7c9', 'turkmenistan': '#65fcc3', 'uzbekistan': '#72febc', 'iran': '#80ffb4', 'italy': '#8dfead', 'poland': '#9afca5', 'switzerland': '#a8f79d', 'norway': '#b5f194', 'the netherlands': '#c3ea8b', 'kenya': '#d0e083', 'uganda': '#ddd579', 'democratic republic of the congo': '#ebc970', 'vietnam': '#f8bc66', 'myanmar': '#ffad5d', 'united states of america': '#ff9d53', 'peru': '#ff8b49', 'canada': '#ff793f', 'indonesia': '#ff6634', 'madagascar': '#ff532a', 'zimbabwe': '#ff3f20', 'algeria': '#ff2a15', 'brazil': '#ff150b', 'bolivia': '#ff0000', '?': '#969696'}, 'province': {'krasnoyarsk krai': '#8000ff', 'panevezys county': '#7b07ff', 'pärnu maakond': '#760eff', 'irkutsk oblast': '#7216ff', 'bavaria': '#6d1dff', 'altai krai': '#6924fe', 'qinghai': '#642bfe', 'republic of dagestan': '#5f32fe', 'goranboy district': '#5b39fd', 'syunik province': '#5640fd', 'shirak province': '#5247fc', 'gegharkunik province': '#4d4efc', 'samtskhe-javakheti': '#4855fb', 'shahbuz rayon': '#445cfb', 'na': '#3f62fa', 'samara oblast': '#3b69f9', 'sečuán': '#3670f8', 'bayankhongor': '#3176f8', 'inner mongolia': '#2d7cf7', 'sughd province': '#2883f6', 'talas region': '#2489f5', 'altai republic': '#1f8ff4', 'govi-altai': '#1a95f3', 'övörkhangai': '#169bf2', 'ömnögovi': '#11a0f0', 'bayan-ölgii': '#0da6ef', 'khovd': '#08abee', 'xinjiang': '#03b0ed', 'issyk-kul region': '#01b6eb', 'east of england': '#06bbea', 'valencia community': '#0abfe8', 'centre-loire valley': '#0fc4e7', 'osh region': '#14c9e5', 'naryn region': '#18cde4', 'gansu': '#1dd1e2', 'hovsgel': '#21d5e0', 'zavkhan province': '#26d9df', 'dornogovi': '#2adddd', 'tuva republic': '#2fe0db', 'jilin': '#34e4d9', 'st. petersberg': '#38e7d7', 'zabaykalsky krai': '#3dead5', 'heilongjiang': '#41edd3', 'tibet': '#46efd1', 'gandaki pradesh': '#4bf2cf', 'maharashtra': '#4ff4cd', 'himachal pradesh': '#54f6cb', 'kabardino-balkaria': '#58f8c9', 'karachay-cherkessia': '#5df9c6', 'ningxia': '#62fbc4', 'shaanxi': '#66fcc2', 'hebei': '#6bfdbf', 'west kazakhstan region': '#6ffebd', 'nagorno-karabakh republic': '#74febb', 'kyzylorda region': '#79ffb8', 'rostov oblast': '#7dffb6', 'astrakhan oblast': '#82ffb3', 'balkan': '#86ffb0', 'shymkent': '#8bfeae', 'navoiy region': '#90feab', 'almaty region': '#94fda8', 'atyrau region': '#99fca6', 'fizuli district': '#9dfba3', 'baku': '#a2f9a0', 'kurdistan': '#a7f89d', 'shamkir district': '#abf69b', 'stavropol krai': '#b0f498', 'ingushetia': '#b4f295', 'mangystau region': '#b9ef92', 'chechnya': '#beed8f', 'aktobe region': '#c2ea8c', 'atyrau': '#c7e789', 'qazakh district': '#cbe486', 'republic of kalmykia': '#d0e083', 'tatarstan': '#d4dd80', 'lazio': '#d9d97c', 'greater london': '#ded579', \"provence-alpes-côte d'azur\": '#e2d176', 'baden-württemberg': '#e7cd73', 'vilnius county': '#ebc970', 'pomeranian voivodeship': '#f0c46c', 'brandenburg': '#f5bf69', 'nidwalden': '#f9bb66', 'oslo': '#feb662', 'catalonia': '#ffb05f', 'occitanie': '#ffab5c', 'north brabant': '#ffa658', 'nairobi': '#ffa055', 'yunnan': '#ff9b52', 'khánh hoà province': '#ff954e', 'lâm đồng province': '#ff8f4b', 'ho chi minh city': '#ff8947', 'guangxi zhuang': '#ff8344', 'california': '#ff7c40', 'la libertad': '#ff763d', 'tamil nadu': '#ff7039', 'saskatchewan': '#ff6936', 'arizona': '#ff6232', 'new mexico': '#ff5c2f', 'colorado': '#ff552b', 'texas': '#ff4e28', 'nevada': '#ff4724', 'central java': '#ff4020', 'province de fianarantsoa': '#ff391d', 'kirov oblast': '#ff3219', 'gujarat': '#ff2b16', 'fujian': '#ff2412', 'midlands': '#ff1d0e', 'msila': '#ff160b', 'state of mato grosso': '#ff0e07', 'la paz': '#ff0704', 'cajamarca': '#ff0000', '?': '#969696'}, 'continent': {'europe': '#8000ff', 'asia': '#00b4ec', 'africa': '#80ffb4', 'north america': '#ffb462', 'south america': '#ff0000', '?': '#969696'}, 'host_order': {'human': '#d62728', 'rodentia': '#9A9A9A', 'siphonaptera': '#2ca02c', 'na': '#9A9A9A', 'lagomorpha': '#1f77b4', 'phthiraptera': '#2ca02c', 'ixodida': '#2ca02c', 'carnivora': '#1f77b4', 'artiodactyla': '#1f77b4', 'lepidoptera': '#2ca02c', 'marsupialia': '#1f77b4', '?': '#969696'}, 'population': {'0.pre': '#8000ff', '0.pe': '#5148fc', '0.ant': '#238af5', '0.ant4': '#0cc1e8', '3.ant': '#3ae8d7', '4.ant': '#68fcc1', '2.ant': '#97fca7', '2.med': '#c5e88a', '1.pre': '#f3c16a', '1.ant': '#ff8a48', '1.in': '#ff4824', '1.ori': '#ff0000', '?': '#969696'}}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEPRECATED: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "out_path_colors = os.path.join(augur_dir, \"colors.tsv\")\n",
    "\n",
    "colors_dict = {}\n",
    "neutral_color = \"#9A9A9A\"\n",
    "\n",
    "host_manual_color = {\n",
    "    # Rodent is baseline\n",
    "    'rodentia': neutral_color, \n",
    "    'human': '#d62728', \n",
    "    # Other mammals\n",
    "    'marsupialia': \"#1f77b4\",\n",
    "    'artiodactyla': \"#1f77b4\", \n",
    "    'carnivora': \"#1f77b4\", \n",
    "    'lagomorpha': \"#1f77b4\", \n",
    "    # Insects\n",
    "    'siphonaptera': \"#2ca02c\", \n",
    "    'phthiraptera': \"#2ca02c\", \n",
    "    'ixodida': \"#2ca02c\", \n",
    "    'lepidoptera': \"#2ca02c\", \n",
    "    # Other\n",
    "    'na': neutral_color,     \n",
    "}\n",
    "\n",
    "for attr in mug_dict:\n",
    "    # Create the color map\n",
    "    colors_dict[attr] = {}\n",
    "    for t in divtree.get_terminals():\n",
    "        attr_val = metadata_df[attr][t.name].lower()\n",
    "        \n",
    "        # Initialize stats and values\n",
    "        if attr_val not in colors_dict[attr] and attr_val != NO_DATA_CHAR:\n",
    "            colors_dict[attr][attr_val] = \"\"\n",
    "\n",
    "    if attr != \"host_order\":\n",
    "        # Create the custom color map (pyplot)\n",
    "        cmap = plt.get_cmap(\"rainbow\", len(colors_dict[attr]))\n",
    "        # Convert the color map to a list of RGB values\n",
    "        cmaplist = [cmap(i) for i in range(cmap.N)]\n",
    "        # Convert RGB values to hex colors\n",
    "        attr_hex = [colors.to_hex(col) for col in cmaplist]\n",
    "    \n",
    "    else:\n",
    "        attr_hex = []\n",
    "        for attr_val in colors_dict[attr]:\n",
    "            attr_col = host_manual_color[attr_val]\n",
    "            attr_hex.append(attr_col)\n",
    "    \n",
    "    # Assign colors to value\n",
    "    for attr_val, attr_col in zip(colors_dict[attr], attr_hex):\n",
    "        colors_dict[attr][attr_val] = attr_col   \n",
    "        \n",
    "    # Add unknown\n",
    "    colors_dict[attr][UNKNOWN_CHAR] = \"#969696\"\n",
    "    \n",
    "print(colors_dict)\n",
    "\n",
    "with open(out_path_colors, \"w\") as outfile:\n",
    "    for attr_key in colors_dict:\n",
    "        for attr_val in colors_dict[attr_key]:\n",
    "            outfile.write(str(attr_key) + \"\\t\" + str(attr_val) + \"\\t\" + str(colors_dict[attr_key][attr_val]) + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "imperial-flooring",
   "metadata": {},
   "source": [
    "## Parse Tree Into Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "regulated-comment",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "      <th>sequencing_technology</th>\n",
       "      <th>assembly_method</th>\n",
       "      <th>host_raw</th>\n",
       "      <th>host_order</th>\n",
       "      <th>population_color</th>\n",
       "      <th>population</th>\n",
       "      <th>geometry</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>population_rtt_dist</th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>timetree_date</th>\n",
       "      <th>timetree_CI_height</th>\n",
       "      <th>timetree_CI_date</th>\n",
       "      <th>mugration_country</th>\n",
       "      <th>mugration_province</th>\n",
       "      <th>mugration_continent</th>\n",
       "      <th>mugration_host_order</th>\n",
       "      <th>mugration_population</th>\n",
       "      <th>mugration_country_confidence</th>\n",
       "      <th>mugration_country_lat_confidence</th>\n",
       "      <th>mugration_country_lon_confidence</th>\n",
       "      <th>mugration_province_confidence</th>\n",
       "      <th>mugration_province_lat_confidence</th>\n",
       "      <th>mugration_province_lon_confidence</th>\n",
       "      <th>mugration_continent_confidence</th>\n",
       "      <th>mugration_host_order_confidence</th>\n",
       "      <th>mugration_population_confidence</th>\n",
       "      <th>mugration_country_lat</th>\n",
       "      <th>mugration_country_lon</th>\n",
       "      <th>mugration_province_lat</th>\n",
       "      <th>mugration_province_lon</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>branch_length_time</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>date_lower</th>\n",
       "      <th>date_upper</th>\n",
       "      <th>divtree_coord_x</th>\n",
       "      <th>divtree_coord_y</th>\n",
       "      <th>rtt_dist</th>\n",
       "      <th>timetree_coord_x</th>\n",
       "      <th>timetree_coord_y</th>\n",
       "      <th>timetree_date_calendar</th>\n",
       "      <th>timetree_CI_lower_err</th>\n",
       "      <th>timetree_CI_upper_err</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Human</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Human</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>POINT (-105.607716 38.7251776)</td>\n",
       "      <td>7.31686e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.3815e-06</td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.004600e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-29</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>North America</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>16.935846</td>\n",
       "      <td>162.52800</td>\n",
       "      <td>2.463945e-08</td>\n",
       "      <td>0.104203</td>\n",
       "      <td>1992</td>\n",
       "      <td>1992</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>509.00</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>5249.006080</td>\n",
       "      <td>509.00</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Human</td>\n",
       "      <td>IonTorrent</td>\n",
       "      <td>Newbler v. 2.6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#c5e88a</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>POINT (40.7957942 47.6222451)</td>\n",
       "      <td>7.30501e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.6582e-06</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.120100e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-98</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Human</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>8.966111</td>\n",
       "      <td>878.89200</td>\n",
       "      <td>2.412242e-09</td>\n",
       "      <td>0.010202</td>\n",
       "      <td>1923</td>\n",
       "      <td>1923</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>339.00</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>5180.006106</td>\n",
       "      <td>339.00</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>5.41847e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.15566e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-15</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.23246</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2006</td>\n",
       "      <td>2006</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>152.00</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>5263.005722</td>\n",
       "      <td>152.00</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Frontopsylla elatoides</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>5.47035e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.20754e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.356000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>0.996375</td>\n",
       "      <td>24.38400</td>\n",
       "      <td>9.662073e-09</td>\n",
       "      <td>0.040862</td>\n",
       "      <td>2005</td>\n",
       "      <td>2005</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>171.00</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>5262.005752</td>\n",
       "      <td>171.00</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>5.4939e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.23109e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.711000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>1.992328</td>\n",
       "      <td>35.90130</td>\n",
       "      <td>1.312209e-08</td>\n",
       "      <td>0.055495</td>\n",
       "      <td>2005</td>\n",
       "      <td>2005</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>173.00</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>5262.005752</td>\n",
       "      <td>173.00</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE595</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.207000e-07</td>\n",
       "      <td>13.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-111.576</td>\n",
       "      <td>{89.1935,137.589}</td>\n",
       "      <td>{-137.589,-89.1935}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.999873</td>\n",
       "      <td>0.999873</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.55</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>0.933362</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>598.25</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>5166.429640</td>\n",
       "      <td>598.25</td>\n",
       "      <td>1910.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE596</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.356000e-07</td>\n",
       "      <td>13.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-111.576</td>\n",
       "      <td>{89.1935,137.589}</td>\n",
       "      <td>{-137.589,-89.1935}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.61</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>0.996375</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>596.75</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>5166.429640</td>\n",
       "      <td>596.75</td>\n",
       "      <td>1910.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE597</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.921000e-07</td>\n",
       "      <td>46.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-78.3728</td>\n",
       "      <td>{50.8548,116.815}</td>\n",
       "      <td>{-116.815,-50.8548}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>1.235320</td>\n",
       "      <td>33.20350</td>\n",
       "      <td>8.797265e-09</td>\n",
       "      <td>0.037204</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>597.50</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>5199.633140</td>\n",
       "      <td>597.50</td>\n",
       "      <td>1943.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE598</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>5.010000e-08</td>\n",
       "      <td>34.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-70.382</td>\n",
       "      <td>{40.845,106.6}</td>\n",
       "      <td>{-106.6,-40.845}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.999874</td>\n",
       "      <td>0.999874</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.60</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>0.211878</td>\n",
       "      <td>41.19420</td>\n",
       "      <td>1.216191e-09</td>\n",
       "      <td>0.005143</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>599.75</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>5207.623840</td>\n",
       "      <td>599.75</td>\n",
       "      <td>1951.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE599</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.248000e-07</td>\n",
       "      <td>58.0</td>\n",
       "      <td></td>\n",
       "      <td>LOW</td>\n",
       "      <td>-54.3879</td>\n",
       "      <td>{27.1416,88.513}</td>\n",
       "      <td>{-88.513,-27.1416}</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>South America</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.991635</td>\n",
       "      <td>0.991635</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.82</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>0.950701</td>\n",
       "      <td>15.99410</td>\n",
       "      <td>1.405518e-08</td>\n",
       "      <td>0.059441</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>600.50</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>5223.617940</td>\n",
       "      <td>600.50</td>\n",
       "      <td>1967.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1201 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    strain    date date_bp  \\\n",
       "sample                                                       \n",
       "Reference                             CO92    1992     -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic   9_10  1923.0     -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic  42126  2006.0     -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic  42123  2005.0     -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic  42118  2005.0     -16   \n",
       "...                                    ...     ...     ...   \n",
       "NODE595                                 NA      NA      NA   \n",
       "NODE596                                 NA      NA      NA   \n",
       "NODE597                                 NA      NA      NA   \n",
       "NODE598                                 NA      NA      NA   \n",
       "NODE599                                 NA      NA      NA   \n",
       "\n",
       "                                                      country       province  \\\n",
       "sample                                                                         \n",
       "Reference                            United States of America       Colorado   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia  Rostov Oblast   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China       Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China       Xinjiang   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China       Xinjiang   \n",
       "...                                                       ...            ...   \n",
       "NODE595                                                    NA             NA   \n",
       "NODE596                                                    NA             NA   \n",
       "NODE597                                                    NA             NA   \n",
       "NODE598                                                    NA             NA   \n",
       "NODE599                                                    NA             NA   \n",
       "\n",
       "                                    country_lat country_lon province_lat  \\\n",
       "sample                                                                     \n",
       "Reference                               39.7837    -100.446      38.7252   \n",
       "GCA_009909635.1_ASM990963v1_genomic     64.6863     97.7453      47.6222   \n",
       "GCA_009669545.1_ASM966954v1_genomic     35.0001         105      42.4805   \n",
       "GCA_009669555.1_ASM966955v1_genomic     35.0001         105      42.4805   \n",
       "GCA_009669565.1_ASM966956v1_genomic     35.0001         105      42.4805   \n",
       "...                                         ...         ...          ...   \n",
       "NODE595                                      NA          NA           NA   \n",
       "NODE596                                      NA          NA           NA   \n",
       "NODE597                                      NA          NA           NA   \n",
       "NODE598                                      NA          NA           NA   \n",
       "NODE599                                      NA          NA           NA   \n",
       "\n",
       "                                    province_lon      biovar branch_major  \\\n",
       "sample                                                                      \n",
       "Reference                               -105.608  Orientalis        1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic      40.7958  Medievalis        2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic      85.4633     Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic      85.4633     Antiqua        0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic      85.4633     Antiqua        0.ANT   \n",
       "...                                          ...         ...          ...   \n",
       "NODE595                                       NA          NA           NA   \n",
       "NODE596                                       NA          NA           NA   \n",
       "NODE597                                       NA          NA           NA   \n",
       "NODE598                                       NA          NA           NA   \n",
       "NODE599                                       NA          NA           NA   \n",
       "\n",
       "                                    branch_minor biosample_accession  \\\n",
       "sample                                                                 \n",
       "Reference                                 1.ORI1        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.ANT1        SAMN07722923   \n",
       "...                                          ...                 ...   \n",
       "NODE595                                       NA                  NA   \n",
       "NODE596                                       NA                  NA   \n",
       "NODE597                                       NA                  NA   \n",
       "NODE598                                       NA                  NA   \n",
       "NODE599                                       NA                  NA   \n",
       "\n",
       "                                                   biosample_comment  \\\n",
       "sample                                                                 \n",
       "Reference                            KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic            KEEP: Assembly Modern   \n",
       "...                                                              ...   \n",
       "NODE595                                                           NA   \n",
       "NODE596                                                           NA   \n",
       "NODE597                                                           NA   \n",
       "NODE598                                                           NA   \n",
       "NODE599                                                           NA   \n",
       "\n",
       "                                    branch_number      continent date_mean  \\\n",
       "sample                                                                       \n",
       "Reference                                       1  North America      1992   \n",
       "GCA_009909635.1_ASM990963v1_genomic             2         Europe      1923   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0           Asia      2006   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0           Asia      2005   \n",
       "GCA_009669565.1_ASM966956v1_genomic             0           Asia      2005   \n",
       "...                                           ...            ...       ...   \n",
       "NODE595                                        NA             NA        NA   \n",
       "NODE596                                        NA             NA        NA   \n",
       "NODE597                                        NA             NA        NA   \n",
       "NODE598                                        NA             NA        NA   \n",
       "NODE599                                        NA             NA        NA   \n",
       "\n",
       "                                    date_bp_mean date_err      lat      lon  \\\n",
       "sample                                                                        \n",
       "Reference                                     29        0  38.7252 -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic           98        0  47.6222  40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic           15        0  42.4805  85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic           16        0  42.4805  85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic           16        0  42.4805  85.4633   \n",
       "...                                          ...      ...      ...      ...   \n",
       "NODE595                                       NA       NA       NA       NA   \n",
       "NODE596                                       NA       NA       NA       NA   \n",
       "NODE597                                       NA       NA       NA       NA   \n",
       "NODE598                                       NA       NA       NA       NA   \n",
       "NODE599                                       NA       NA       NA       NA   \n",
       "\n",
       "                                    host_human sequencing_technology  \\\n",
       "sample                                                                 \n",
       "Reference                                Human                    NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Human            IonTorrent   \n",
       "GCA_009669545.1_ASM966954v1_genomic  Non-Human   Illumina Hiseq 2000   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Non-Human   Illumina Hiseq 2000   \n",
       "GCA_009669565.1_ASM966956v1_genomic  Non-Human   Illumina Hiseq 2000   \n",
       "...                                        ...                   ...   \n",
       "NODE595                                     NA                    NA   \n",
       "NODE596                                     NA                    NA   \n",
       "NODE597                                     NA                    NA   \n",
       "NODE598                                     NA                    NA   \n",
       "NODE599                                     NA                    NA   \n",
       "\n",
       "                                        assembly_method  \\\n",
       "sample                                                    \n",
       "Reference                                            NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Newbler v. 2.6   \n",
       "GCA_009669545.1_ASM966954v1_genomic  SOAPdenovo v. 2.04   \n",
       "GCA_009669555.1_ASM966955v1_genomic  SOAPdenovo v. 2.04   \n",
       "GCA_009669565.1_ASM966956v1_genomic  SOAPdenovo v. 2.04   \n",
       "...                                                 ...   \n",
       "NODE595                                              NA   \n",
       "NODE596                                              NA   \n",
       "NODE597                                              NA   \n",
       "NODE598                                              NA   \n",
       "NODE599                                              NA   \n",
       "\n",
       "                                                   host_raw    host_order  \\\n",
       "sample                                                                      \n",
       "Reference                                             Human         Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic            Homo sapiens         Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic      Citellus undulatus      Rodentia   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Frontopsylla elatoides  Siphonaptera   \n",
       "GCA_009669565.1_ASM966956v1_genomic      Citellus undulatus      Rodentia   \n",
       "...                                                     ...           ...   \n",
       "NODE595                                                  NA            NA   \n",
       "NODE596                                                  NA            NA   \n",
       "NODE597                                                  NA            NA   \n",
       "NODE598                                                  NA            NA   \n",
       "NODE599                                                  NA            NA   \n",
       "\n",
       "                                    population_color population  \\\n",
       "sample                                                            \n",
       "Reference                                    #ff0000      1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic          #c5e88a      2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          #238af5      0.ANT   \n",
       "...                                              ...        ...   \n",
       "NODE595                                           NA         NA   \n",
       "NODE596                                           NA         NA   \n",
       "NODE597                                           NA         NA   \n",
       "NODE598                                           NA         NA   \n",
       "NODE599                                           NA         NA   \n",
       "\n",
       "                                                                 geometry  \\\n",
       "sample                                                                      \n",
       "Reference                                  POINT (-105.607716 38.7251776)   \n",
       "GCA_009909635.1_ASM990963v1_genomic         POINT (40.7957942 47.6222451)   \n",
       "GCA_009669545.1_ASM966954v1_genomic  POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669555.1_ASM966955v1_genomic  POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669565.1_ASM966956v1_genomic  POINT (85.46334640000001 42.4804953)   \n",
       "...                                                                   ...   \n",
       "NODE595                                                                NA   \n",
       "NODE596                                                                NA   \n",
       "NODE597                                                                NA   \n",
       "NODE598                                                                NA   \n",
       "NODE599                                                                NA   \n",
       "\n",
       "                                    root_rtt_dist clade_rtt_dist  \\\n",
       "sample                                                             \n",
       "Reference                             7.31686e-05             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic   7.30501e-05             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic   5.41847e-05             NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic   5.47035e-05             NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic    5.4939e-05             NA   \n",
       "...                                           ...            ...   \n",
       "NODE595                                        NA             NA   \n",
       "NODE596                                        NA             NA   \n",
       "NODE597                                        NA             NA   \n",
       "NODE598                                        NA             NA   \n",
       "NODE599                                        NA             NA   \n",
       "\n",
       "                                    population_rtt_dist node_type  \\\n",
       "sample                                                              \n",
       "Reference                                    6.3815e-06  terminal   \n",
       "GCA_009909635.1_ASM990963v1_genomic          9.6582e-06  terminal   \n",
       "GCA_009669545.1_ASM966954v1_genomic         1.15566e-05  terminal   \n",
       "GCA_009669555.1_ASM966955v1_genomic         1.20754e-05  terminal   \n",
       "GCA_009669565.1_ASM966956v1_genomic         1.23109e-05  terminal   \n",
       "...                                                 ...       ...   \n",
       "NODE595                                              NA  internal   \n",
       "NODE596                                              NA  internal   \n",
       "NODE597                                              NA  internal   \n",
       "NODE598                                              NA  internal   \n",
       "NODE599                                              NA  internal   \n",
       "\n",
       "                                     branch_length  branch_support  \\\n",
       "sample                                                               \n",
       "Reference                             4.004600e-06             0.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic   2.120100e-06             0.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic   0.000000e+00             0.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic   2.356000e-07             0.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic   4.711000e-07             0.0   \n",
       "...                                            ...             ...   \n",
       "NODE595                               2.207000e-07            13.0   \n",
       "NODE596                               2.356000e-07            13.0   \n",
       "NODE597                               2.921000e-07            46.0   \n",
       "NODE598                               5.010000e-08            34.0   \n",
       "NODE599                               2.248000e-07            58.0   \n",
       "\n",
       "                                    branch_support_conf_char  \\\n",
       "sample                                                         \n",
       "Reference                                                      \n",
       "GCA_009909635.1_ASM990963v1_genomic                            \n",
       "GCA_009669545.1_ASM966954v1_genomic                            \n",
       "GCA_009669555.1_ASM966955v1_genomic                            \n",
       "GCA_009669565.1_ASM966956v1_genomic                            \n",
       "...                                                      ...   \n",
       "NODE595                                                        \n",
       "NODE596                                                        \n",
       "NODE597                                                        \n",
       "NODE598                                                        \n",
       "NODE599                                                        \n",
       "\n",
       "                                    branch_support_conf_category  \\\n",
       "sample                                                             \n",
       "Reference                                                    LOW   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          LOW   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          LOW   \n",
       "GCA_009669555.1_ASM966955v1_genomic                          LOW   \n",
       "GCA_009669565.1_ASM966956v1_genomic                          LOW   \n",
       "...                                                          ...   \n",
       "NODE595                                                      LOW   \n",
       "NODE596                                                      LOW   \n",
       "NODE597                                                      LOW   \n",
       "NODE598                                                      LOW   \n",
       "NODE599                                                      LOW   \n",
       "\n",
       "                                    timetree_date timetree_CI_height  \\\n",
       "sample                                                                 \n",
       "Reference                                     -29                 NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic           -98                 NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic           -15                 NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic           -16                 NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic           -16                 NA   \n",
       "...                                           ...                ...   \n",
       "NODE595                                  -111.576  {89.1935,137.589}   \n",
       "NODE596                                  -111.576  {89.1935,137.589}   \n",
       "NODE597                                  -78.3728  {50.8548,116.815}   \n",
       "NODE598                                   -70.382     {40.845,106.6}   \n",
       "NODE599                                  -54.3879   {27.1416,88.513}   \n",
       "\n",
       "                                        timetree_CI_date  \\\n",
       "sample                                                     \n",
       "Reference                                             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic                   NA   \n",
       "...                                                  ...   \n",
       "NODE595                              {-137.589,-89.1935}   \n",
       "NODE596                              {-137.589,-89.1935}   \n",
       "NODE597                              {-116.815,-50.8548}   \n",
       "NODE598                                 {-106.6,-40.845}   \n",
       "NODE599                               {-88.513,-27.1416}   \n",
       "\n",
       "                                            mugration_country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "NODE595                                                  Peru   \n",
       "NODE596                                                  Peru   \n",
       "NODE597                                                  Peru   \n",
       "NODE598                                                  Peru   \n",
       "NODE599                                                  Peru   \n",
       "\n",
       "                                    mugration_province mugration_continent  \\\n",
       "sample                                                                       \n",
       "Reference                                     Colorado       North America   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Rostov Oblast              Europe   \n",
       "GCA_009669545.1_ASM966954v1_genomic           Xinjiang                Asia   \n",
       "GCA_009669555.1_ASM966955v1_genomic           Xinjiang                Asia   \n",
       "GCA_009669565.1_ASM966956v1_genomic           Xinjiang                Asia   \n",
       "...                                                ...                 ...   \n",
       "NODE595                                      Cajamarca       South America   \n",
       "NODE596                                      Cajamarca       South America   \n",
       "NODE597                                      Cajamarca       South America   \n",
       "NODE598                                      Cajamarca       South America   \n",
       "NODE599                                      Cajamarca       South America   \n",
       "\n",
       "                                    mugration_host_order mugration_population  \\\n",
       "sample                                                                          \n",
       "Reference                                          Human                1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic                Human                2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic             Rodentia                0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic         Siphonaptera                0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic             Rodentia                0.ANT   \n",
       "...                                                  ...                  ...   \n",
       "NODE595                                            Human                1.ORI   \n",
       "NODE596                                         Rodentia                1.ORI   \n",
       "NODE597                                         Rodentia                1.ORI   \n",
       "NODE598                                            Human                1.ORI   \n",
       "NODE599                                         Rodentia                1.ORI   \n",
       "\n",
       "                                     mugration_country_confidence  \\\n",
       "sample                                                              \n",
       "Reference                                                     1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.0   \n",
       "...                                                           ...   \n",
       "NODE595                                                       1.0   \n",
       "NODE596                                                       1.0   \n",
       "NODE597                                                       1.0   \n",
       "NODE598                                                       1.0   \n",
       "NODE599                                                       1.0   \n",
       "\n",
       "                                     mugration_country_lat_confidence  \\\n",
       "sample                                                                  \n",
       "Reference                                                    1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          1.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic                          1.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic                          1.000000   \n",
       "...                                                               ...   \n",
       "NODE595                                                      0.999999   \n",
       "NODE596                                                      0.999999   \n",
       "NODE597                                                      0.999999   \n",
       "NODE598                                                      0.999999   \n",
       "NODE599                                                      0.999999   \n",
       "\n",
       "                                     mugration_country_lon_confidence  \\\n",
       "sample                                                                  \n",
       "Reference                                                    1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          1.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic                          1.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic                          1.000000   \n",
       "...                                                               ...   \n",
       "NODE595                                                      0.999999   \n",
       "NODE596                                                      0.999999   \n",
       "NODE597                                                      0.999999   \n",
       "NODE598                                                      0.999999   \n",
       "NODE599                                                      0.999999   \n",
       "\n",
       "                                     mugration_province_confidence  \\\n",
       "sample                                                               \n",
       "Reference                                                     1.00   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.00   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.00   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.00   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.00   \n",
       "...                                                            ...   \n",
       "NODE595                                                       1.00   \n",
       "NODE596                                                       1.00   \n",
       "NODE597                                                       1.00   \n",
       "NODE598                                                       1.00   \n",
       "NODE599                                                       0.99   \n",
       "\n",
       "                                     mugration_province_lat_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                     1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.000000   \n",
       "...                                                                ...   \n",
       "NODE595                                                       0.999873   \n",
       "NODE596                                                       0.999998   \n",
       "NODE597                                                       0.999999   \n",
       "NODE598                                                       0.999874   \n",
       "NODE599                                                       0.991635   \n",
       "\n",
       "                                     mugration_province_lon_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                     1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.000000   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.000000   \n",
       "...                                                                ...   \n",
       "NODE595                                                       0.999873   \n",
       "NODE596                                                       0.999998   \n",
       "NODE597                                                       0.999999   \n",
       "NODE598                                                       0.999874   \n",
       "NODE599                                                       0.991635   \n",
       "\n",
       "                                     mugration_continent_confidence  \\\n",
       "sample                                                                \n",
       "Reference                                                       1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                             1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                             1.0   \n",
       "...                                                             ...   \n",
       "NODE595                                                         1.0   \n",
       "NODE596                                                         1.0   \n",
       "NODE597                                                         1.0   \n",
       "NODE598                                                         1.0   \n",
       "NODE599                                                         1.0   \n",
       "\n",
       "                                     mugration_host_order_confidence  \\\n",
       "sample                                                                 \n",
       "Reference                                                       1.00   \n",
       "GCA_009909635.1_ASM990963v1_genomic                             1.00   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             1.00   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             1.00   \n",
       "GCA_009669565.1_ASM966956v1_genomic                             1.00   \n",
       "...                                                              ...   \n",
       "NODE595                                                         0.55   \n",
       "NODE596                                                         0.61   \n",
       "NODE597                                                         0.56   \n",
       "NODE598                                                         0.60   \n",
       "NODE599                                                         0.82   \n",
       "\n",
       "                                     mugration_population_confidence  \\\n",
       "sample                                                                 \n",
       "Reference                                                        1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                              1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                              1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                              1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                              1.0   \n",
       "...                                                              ...   \n",
       "NODE595                                                          1.0   \n",
       "NODE596                                                          1.0   \n",
       "NODE597                                                          1.0   \n",
       "NODE598                                                          1.0   \n",
       "NODE599                                                          1.0   \n",
       "\n",
       "                                     mugration_country_lat  \\\n",
       "sample                                                       \n",
       "Reference                                        39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic              64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic              35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic              35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic              35.000074   \n",
       "...                                                    ...   \n",
       "NODE595                                          -6.869970   \n",
       "NODE596                                          -6.869970   \n",
       "NODE597                                          -6.869970   \n",
       "NODE598                                          -6.869970   \n",
       "NODE599                                          -6.869970   \n",
       "\n",
       "                                     mugration_country_lon  \\\n",
       "sample                                                       \n",
       "Reference                                      -100.445882   \n",
       "GCA_009909635.1_ASM990963v1_genomic              97.745306   \n",
       "GCA_009669545.1_ASM966954v1_genomic             104.999927   \n",
       "GCA_009669555.1_ASM966955v1_genomic             104.999927   \n",
       "GCA_009669565.1_ASM966956v1_genomic             104.999927   \n",
       "...                                                    ...   \n",
       "NODE595                                         -75.045851   \n",
       "NODE596                                         -75.045851   \n",
       "NODE597                                         -75.045851   \n",
       "NODE598                                         -75.045851   \n",
       "NODE599                                         -75.045851   \n",
       "\n",
       "                                     mugration_province_lat  \\\n",
       "sample                                                        \n",
       "Reference                                         38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic               47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic               42.480495   \n",
       "GCA_009669555.1_ASM966955v1_genomic               42.480495   \n",
       "GCA_009669565.1_ASM966956v1_genomic               42.480495   \n",
       "...                                                     ...   \n",
       "NODE595                                           -6.250000   \n",
       "NODE596                                           -6.250000   \n",
       "NODE597                                           -6.250000   \n",
       "NODE598                                           -6.250000   \n",
       "NODE599                                           -6.250000   \n",
       "\n",
       "                                     mugration_province_lon  \\\n",
       "sample                                                        \n",
       "Reference                                       -105.607716   \n",
       "GCA_009909635.1_ASM990963v1_genomic               40.795794   \n",
       "GCA_009669545.1_ASM966954v1_genomic               85.463346   \n",
       "GCA_009669555.1_ASM966955v1_genomic               85.463346   \n",
       "GCA_009669565.1_ASM966956v1_genomic               85.463346   \n",
       "...                                                     ...   \n",
       "NODE595                                          -78.833333   \n",
       "NODE596                                          -78.833333   \n",
       "NODE597                                          -78.833333   \n",
       "NODE598                                          -78.833333   \n",
       "NODE599                                          -78.833333   \n",
       "\n",
       "                                     branch_length_sub  branch_length_time  \\\n",
       "sample                                                                       \n",
       "Reference                                    16.935846           162.52800   \n",
       "GCA_009909635.1_ASM990963v1_genomic           8.966111           878.89200   \n",
       "GCA_009669545.1_ASM966954v1_genomic           0.000000             9.23246   \n",
       "GCA_009669555.1_ASM966955v1_genomic           0.996375            24.38400   \n",
       "GCA_009669565.1_ASM966956v1_genomic           1.992328            35.90130   \n",
       "...                                                ...                 ...   \n",
       "NODE595                                       0.933362             0.00000   \n",
       "NODE596                                       0.996375             0.00000   \n",
       "NODE597                                       1.235320            33.20350   \n",
       "NODE598                                       0.211878            41.19420   \n",
       "NODE599                                       0.950701            15.99410   \n",
       "\n",
       "                                         rate_sub  rate_sub_year date_lower  \\\n",
       "sample                                                                        \n",
       "Reference                            2.463945e-08       0.104203       1992   \n",
       "GCA_009909635.1_ASM990963v1_genomic  2.412242e-09       0.010202       1923   \n",
       "GCA_009669545.1_ASM966954v1_genomic  0.000000e+00       0.000000       2006   \n",
       "GCA_009669555.1_ASM966955v1_genomic  9.662073e-09       0.040862       2005   \n",
       "GCA_009669565.1_ASM966956v1_genomic  1.312209e-08       0.055495       2005   \n",
       "...                                           ...            ...        ...   \n",
       "NODE595                              0.000000e+00       0.000000         NA   \n",
       "NODE596                              0.000000e+00       0.000000         NA   \n",
       "NODE597                              8.797265e-09       0.037204         NA   \n",
       "NODE598                              1.216191e-09       0.005143         NA   \n",
       "NODE599                              1.405518e-08       0.059441         NA   \n",
       "\n",
       "                                    date_upper  divtree_coord_x  \\\n",
       "sample                                                            \n",
       "Reference                                 1992         0.000073   \n",
       "GCA_009909635.1_ASM990963v1_genomic       1923         0.000073   \n",
       "GCA_009669545.1_ASM966954v1_genomic       2006         0.000054   \n",
       "GCA_009669555.1_ASM966955v1_genomic       2005         0.000055   \n",
       "GCA_009669565.1_ASM966956v1_genomic       2005         0.000055   \n",
       "...                                        ...              ...   \n",
       "NODE595                                     NA         0.000072   \n",
       "NODE596                                     NA         0.000072   \n",
       "NODE597                                     NA         0.000073   \n",
       "NODE598                                     NA         0.000072   \n",
       "NODE599                                     NA         0.000072   \n",
       "\n",
       "                                     divtree_coord_y  rtt_dist  \\\n",
       "sample                                                           \n",
       "Reference                                     509.00  0.000073   \n",
       "GCA_009909635.1_ASM990963v1_genomic           339.00  0.000073   \n",
       "GCA_009669545.1_ASM966954v1_genomic           152.00  0.000054   \n",
       "GCA_009669555.1_ASM966955v1_genomic           171.00  0.000055   \n",
       "GCA_009669565.1_ASM966956v1_genomic           173.00  0.000055   \n",
       "...                                              ...       ...   \n",
       "NODE595                                       598.25  0.000072   \n",
       "NODE596                                       596.75  0.000072   \n",
       "NODE597                                       597.50  0.000073   \n",
       "NODE598                                       599.75  0.000072   \n",
       "NODE599                                       600.50  0.000072   \n",
       "\n",
       "                                     timetree_coord_x  timetree_coord_y  \\\n",
       "sample                                                                    \n",
       "Reference                                 5249.006080            509.00   \n",
       "GCA_009909635.1_ASM990963v1_genomic       5180.006106            339.00   \n",
       "GCA_009669545.1_ASM966954v1_genomic       5263.005722            152.00   \n",
       "GCA_009669555.1_ASM966955v1_genomic       5262.005752            171.00   \n",
       "GCA_009669565.1_ASM966956v1_genomic       5262.005752            173.00   \n",
       "...                                               ...               ...   \n",
       "NODE595                                   5166.429640            598.25   \n",
       "NODE596                                   5166.429640            596.75   \n",
       "NODE597                                   5199.633140            597.50   \n",
       "NODE598                                   5207.623840            599.75   \n",
       "NODE599                                   5223.617940            600.50   \n",
       "\n",
       "                                     timetree_date_calendar  \\\n",
       "sample                                                        \n",
       "Reference                                            1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                  1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                  2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                  2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                  2005.0   \n",
       "...                                                     ...   \n",
       "NODE595                                              1910.0   \n",
       "NODE596                                              1910.0   \n",
       "NODE597                                              1943.0   \n",
       "NODE598                                              1951.0   \n",
       "NODE599                                              1967.0   \n",
       "\n",
       "                                     timetree_CI_lower_err  \\\n",
       "sample                                                       \n",
       "Reference                                              0.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    0.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                    0.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                    0.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                    0.0   \n",
       "...                                                    ...   \n",
       "NODE595                                               26.0   \n",
       "NODE596                                               26.0   \n",
       "NODE597                                               38.0   \n",
       "NODE598                                               36.0   \n",
       "NODE599                                               34.0   \n",
       "\n",
       "                                     timetree_CI_upper_err  \n",
       "sample                                                      \n",
       "Reference                                              0.0  \n",
       "GCA_009909635.1_ASM990963v1_genomic                    0.0  \n",
       "GCA_009669545.1_ASM966954v1_genomic                    0.0  \n",
       "GCA_009669555.1_ASM966955v1_genomic                    0.0  \n",
       "GCA_009669565.1_ASM966956v1_genomic                    0.0  \n",
       "...                                                    ...  \n",
       "NODE595                                               22.0  \n",
       "NODE596                                               22.0  \n",
       "NODE597                                               28.0  \n",
       "NODE598                                               30.0  \n",
       "NODE599                                               27.0  \n",
       "\n",
       "[1201 rows x 72 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "x_posns = get_x_positions(divtree)\n",
    "y_posns = get_y_positions(divtree)\n",
    "\n",
    "# Initialize \n",
    "# Date lower and upper error bars for tips\n",
    "metadata_df[\"date_lower\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"date_upper\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "    \n",
    "for c in divtree.find_clades():\n",
    "    # ------------------------------------------------------------\n",
    "    # Coordinates for the divergence tree\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    metadata_df.at[c.name, 'divtree_coord_x'] = coord_x\n",
    "    metadata_df.at[c.name, 'divtree_coord_y'] = coord_y\n",
    "    \n",
    "    # ------------------------------------------------------------\n",
    "    # Root to tip distance (same as div x)\n",
    "    metadata_df.at[c.name, \"rtt_dist\"] = coord_x \n",
    "\n",
    "    # ------------------------------------------------------------\n",
    "    # Date lower and upper error bars for tips\n",
    "    date_mean = metadata_df[\"date_mean\"][c.name] \n",
    "    if date_mean != NO_DATA_CHAR:\n",
    "        metadata_df.at[c.name, \"date_lower\"] = date_mean - metadata_df[\"date_err\"][c.name]\n",
    "        metadata_df.at[c.name, \"date_upper\"] = date_mean + metadata_df[\"date_err\"][c.name]\n",
    "\n",
    "x_posns = get_x_positions(timetree)\n",
    "y_posns = get_y_positions(timetree)\n",
    "\n",
    "\n",
    "for c in timetree.find_clades():\n",
    "\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    metadata_df.at[c.name, 'timetree_coord_x'] = coord_x\n",
    "    metadata_df.at[c.name, 'timetree_coord_y'] = coord_y\n",
    "    \n",
    "    timetree_date = int(float(metadata_df[\"timetree_date\"][c.name]))\n",
    "    timetree_date_calendar = CURRENT_YEAR + timetree_date\n",
    "    metadata_df.at[c.name, 'timetree_date_calendar'] = timetree_date_calendar\n",
    "    \n",
    "    timetree_CI_date = metadata_df[\"timetree_CI_date\"][c.name]\n",
    "    if timetree_CI_date == NO_DATA_CHAR:\n",
    "        timetree_CI_lower_err = 0\n",
    "        timetree_CI_upper_err = 0\n",
    "    else:\n",
    "        timetree_CI_split = [int(float(d)) for d in timetree_CI_date.strip(\"{}\").split(\",\")]\n",
    "        timetree_CI_calendar = [CURRENT_YEAR + d for d in timetree_CI_split]\n",
    "        timetree_CI_lower_err = timetree_date_calendar - timetree_CI_calendar[0]\n",
    "        timetree_CI_upper_err = timetree_CI_calendar[1] - timetree_date_calendar\n",
    "    metadata_df.at[c.name,\"timetree_CI_lower_err\"] = timetree_CI_lower_err\n",
    "    metadata_df.at[c.name,\"timetree_CI_upper_err\"] = timetree_CI_upper_err   \n",
    "\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bibliographic-learning",
   "metadata": {},
   "source": [
    "## Create Custom Columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "former-bouquet",
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in divtree.get_terminals():\n",
    "    strain = metadata_df[\"strain\"][c.name]\n",
    "    country = metadata_df[\"country\"][c.name]\n",
    "    province = metadata_df[\"country\"][c.name]\n",
    "    date = metadata_df[\"date\"][c.name]\n",
    "    if type(date) == str:\n",
    "        date = date.strip(\"[]\")\n",
    "    metadata_df.at[c.name,\"country_date_strain\"] = \"{} ({}) {}\".format(country, date, strain)\n",
    "    metadata_df.at[c.name,\"province_date_strain\"] = \"{} ({}) {}\".format(province, date, strain)\n",
    "    \n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brutal-oracle",
   "metadata": {},
   "source": [
    "---\n",
    "# 4. EXPORT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "otherwise-screw",
   "metadata": {},
   "source": [
    "## Time Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "caring-affiliate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_timetree = copy.deepcopy(timetree)\n",
    "\n",
    "for c in out_timetree.find_clades():\n",
    "    if not c.confidence:\n",
    "        c.confidence = 0\n",
    "\n",
    "metadata_to_comment(out_timetree, metadata_df)    \n",
    "out_timetree_nex_path = os.path.join(augur_dir, \"all.timetree.nex\")\n",
    "Phylo.write(out_timetree, out_timetree_nex_path, \"nexus\")\n",
    "\n",
    "for c in out_timetree.find_clades():\n",
    "    c.comment = None\n",
    "\n",
    "out_timetree_nwk_path = os.path.join(augur_dir, \"all.timetree.nwk\")\n",
    "Phylo.write(out_timetree, out_timetree_nwk_path, \"newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "running-swimming",
   "metadata": {},
   "source": [
    "# Divergence Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "traditional-sheep",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_divtree = copy.deepcopy(divtree)\n",
    "\n",
    "for c in out_divtree.find_clades():\n",
    "    if not c.confidence:\n",
    "        c.confidence = 0\n",
    "\n",
    "metadata_to_comment(out_divtree, metadata_df)    \n",
    "out_divtree_nex_path = os.path.join(augur_dir, \"all.divtree.nex\")\n",
    "Phylo.write(out_divtree, out_divtree_nex_path, \"nexus\", format_branch_length=\"%1.10f\")\n",
    "\n",
    "for c in out_divtree.find_clades():\n",
    "    c.comment = None\n",
    "\n",
    "out_divtree_nwk_path = os.path.join(augur_dir, \"all.divtree.nwk\")\n",
    "Phylo.write(out_divtree, out_divtree_nwk_path, \"newick\", format_branch_length=\"%1.10f\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "assumed-missouri",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "guilty-trading",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_metadata = out_divtree_nwk_path = os.path.join(augur_dir, \"metadata.tsv\")\n",
    "metadata_df.to_csv(out_path_metadata, sep=\"\\t\", index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "given-wallpaper",
   "metadata": {},
   "source": [
    "## Augur"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "convenient-calculator",
   "metadata": {},
   "source": [
    "### Reduced Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "acceptable-robin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>branch_length_time</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>strain</th>\n",
       "      <th>country_date_strain</th>\n",
       "      <th>province_date_strain</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>population_rtt_dist</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>mugration_country</th>\n",
       "      <th>mugration_country_confidence</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>mugration_province</th>\n",
       "      <th>mugration_province_confidence</th>\n",
       "      <th>mugration_continent</th>\n",
       "      <th>mugration_continent_confidence</th>\n",
       "      <th>mugration_host_order</th>\n",
       "      <th>mugration_host_order_confidence</th>\n",
       "      <th>mugration_population</th>\n",
       "      <th>mugration_population_confidence</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>node_name</th>\n",
       "      <th>blank</th>\n",
       "      <th>country_conf_category</th>\n",
       "      <th>province_conf_category</th>\n",
       "      <th>continent_conf_category</th>\n",
       "      <th>host_order_conf_category</th>\n",
       "      <th>population_conf_category</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>4.004600e-06</td>\n",
       "      <td>17.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>1992</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>CO92</td>\n",
       "      <td>United States of America (1992) CO92</td>\n",
       "      <td>United States of America (1992) CO92</td>\n",
       "      <td>7.31686e-05</td>\n",
       "      <td>6.3815e-06</td>\n",
       "      <td>2.463945e-08</td>\n",
       "      <td>0.104203</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>1.0</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>1.00</td>\n",
       "      <td>North America</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[1992.0, 1992.0]</td>\n",
       "      <td>Reference</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>2.120100e-06</td>\n",
       "      <td>9.0</td>\n",
       "      <td>879.0</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>1923</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>2</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>9_10</td>\n",
       "      <td>Russia (1923.0) 9_10</td>\n",
       "      <td>Russia (1923.0) 9_10</td>\n",
       "      <td>7.30501e-05</td>\n",
       "      <td>9.6582e-06</td>\n",
       "      <td>2.412242e-09</td>\n",
       "      <td>0.010202</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>Russia</td>\n",
       "      <td>1.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>1.00</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[1923.0, 1923.0]</td>\n",
       "      <td>GCA_009909635.1_ASM990963v1_genomic</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>2006</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>0</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>42126</td>\n",
       "      <td>China (2006.0) 42126</td>\n",
       "      <td>China (2006.0) 42126</td>\n",
       "      <td>5.41847e-05</td>\n",
       "      <td>1.15566e-05</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>China</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[2006.0, 2006.0]</td>\n",
       "      <td>GCA_009669545.1_ASM966954v1_genomic</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>2.356000e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>2005</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>0</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>42123</td>\n",
       "      <td>China (2005.0) 42123</td>\n",
       "      <td>China (2005.0) 42123</td>\n",
       "      <td>5.47035e-05</td>\n",
       "      <td>1.20754e-05</td>\n",
       "      <td>9.662073e-09</td>\n",
       "      <td>0.040862</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>China</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[2005.0, 2005.0]</td>\n",
       "      <td>GCA_009669555.1_ASM966955v1_genomic</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>4.711000e-07</td>\n",
       "      <td>2.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>2005</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>0</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>42118</td>\n",
       "      <td>China (2005.0) 42118</td>\n",
       "      <td>China (2005.0) 42118</td>\n",
       "      <td>5.4939e-05</td>\n",
       "      <td>1.23109e-05</td>\n",
       "      <td>1.312209e-08</td>\n",
       "      <td>0.055495</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>China</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[2005.0, 2005.0]</td>\n",
       "      <td>GCA_009669565.1_ASM966956v1_genomic</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE595</th>\n",
       "      <td>2.207000e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1910.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>13.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>internal</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>Peru</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>1.00</td>\n",
       "      <td>South America</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>0.55</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[1884.0, 1932.0]</td>\n",
       "      <td>NODE595</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE596</th>\n",
       "      <td>2.356000e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1910.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>13.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>internal</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>Peru</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>1.00</td>\n",
       "      <td>South America</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>0.61</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[1884.0, 1932.0]</td>\n",
       "      <td>NODE596</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE597</th>\n",
       "      <td>2.921000e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>1943.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>46.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>internal</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.797265e-09</td>\n",
       "      <td>0.037204</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>Peru</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>1.00</td>\n",
       "      <td>South America</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[1905.0, 1971.0]</td>\n",
       "      <td>NODE597</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE598</th>\n",
       "      <td>5.010000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>1951.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>34.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>internal</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.216191e-09</td>\n",
       "      <td>0.005143</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>Peru</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>1.00</td>\n",
       "      <td>South America</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Human</td>\n",
       "      <td>0.60</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[1915.0, 1981.0]</td>\n",
       "      <td>NODE598</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE599</th>\n",
       "      <td>2.248000e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1967.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>58.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>internal</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.405518e-08</td>\n",
       "      <td>0.059441</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>Peru</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>0.99</td>\n",
       "      <td>South America</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>0.82</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[1933.0, 1994.0]</td>\n",
       "      <td>NODE599</td>\n",
       "      <td></td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1201 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     branch_length  branch_length_sub  \\\n",
       "sample                                                                  \n",
       "Reference                             4.004600e-06               17.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic   2.120100e-06                9.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic   0.000000e+00                0.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic   2.356000e-07                1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic   4.711000e-07                2.0   \n",
       "...                                            ...                ...   \n",
       "NODE595                               2.207000e-07                1.0   \n",
       "NODE596                               2.356000e-07                1.0   \n",
       "NODE597                               2.921000e-07                1.0   \n",
       "NODE598                               5.010000e-08                0.0   \n",
       "NODE599                               2.248000e-07                1.0   \n",
       "\n",
       "                                     branch_length_time  timetree_num_date  \\\n",
       "sample                                                                       \n",
       "Reference                                         163.0             1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic               879.0             1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                 9.0             2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                24.0             2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                36.0             2005.0   \n",
       "...                                                 ...                ...   \n",
       "NODE595                                             0.0             1910.0   \n",
       "NODE596                                             0.0             1910.0   \n",
       "NODE597                                            33.0             1943.0   \n",
       "NODE598                                            41.0             1951.0   \n",
       "NODE599                                            16.0             1967.0   \n",
       "\n",
       "                                    date_mean date_err date_bp_mean  \\\n",
       "sample                                                                \n",
       "Reference                                1992        0           29   \n",
       "GCA_009909635.1_ASM990963v1_genomic      1923        0           98   \n",
       "GCA_009669545.1_ASM966954v1_genomic      2006        0           15   \n",
       "GCA_009669555.1_ASM966955v1_genomic      2005        0           16   \n",
       "GCA_009669565.1_ASM966956v1_genomic      2005        0           16   \n",
       "...                                       ...      ...          ...   \n",
       "NODE595                                    NA       NA           NA   \n",
       "NODE596                                    NA       NA           NA   \n",
       "NODE597                                    NA       NA           NA   \n",
       "NODE598                                    NA       NA           NA   \n",
       "NODE599                                    NA       NA           NA   \n",
       "\n",
       "                                     branch_support  \\\n",
       "sample                                                \n",
       "Reference                                       0.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic             0.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic             0.0   \n",
       "...                                             ...   \n",
       "NODE595                                        13.0   \n",
       "NODE596                                        13.0   \n",
       "NODE597                                        46.0   \n",
       "NODE598                                        34.0   \n",
       "NODE599                                        58.0   \n",
       "\n",
       "                                    branch_support_conf_category  \\\n",
       "sample                                                             \n",
       "Reference                                                    LOW   \n",
       "GCA_009909635.1_ASM990963v1_genomic                          LOW   \n",
       "GCA_009669545.1_ASM966954v1_genomic                          LOW   \n",
       "GCA_009669555.1_ASM966955v1_genomic                          LOW   \n",
       "GCA_009669565.1_ASM966956v1_genomic                          LOW   \n",
       "...                                                          ...   \n",
       "NODE595                                                      LOW   \n",
       "NODE596                                                      LOW   \n",
       "NODE597                                                      LOW   \n",
       "NODE598                                                      LOW   \n",
       "NODE599                                                      LOW   \n",
       "\n",
       "                                    branch_support_conf_char node_type  \\\n",
       "sample                                                                   \n",
       "Reference                                                     terminal   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           terminal   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           terminal   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           terminal   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           terminal   \n",
       "...                                                      ...       ...   \n",
       "NODE595                                                       internal   \n",
       "NODE596                                                       internal   \n",
       "NODE597                                                       internal   \n",
       "NODE598                                                       internal   \n",
       "NODE599                                                       internal   \n",
       "\n",
       "                                    branch_number biosample_accession strain  \\\n",
       "sample                                                                         \n",
       "Reference                                       1        SAMEA1705942   CO92   \n",
       "GCA_009909635.1_ASM990963v1_genomic             2        SAMN13632815   9_10   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0        SAMN07722925  42126   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0        SAMN07722924  42123   \n",
       "GCA_009669565.1_ASM966956v1_genomic             0        SAMN07722923  42118   \n",
       "...                                           ...                 ...    ...   \n",
       "NODE595                                        NA                  NA     NA   \n",
       "NODE596                                        NA                  NA     NA   \n",
       "NODE597                                        NA                  NA     NA   \n",
       "NODE598                                        NA                  NA     NA   \n",
       "NODE599                                        NA                  NA     NA   \n",
       "\n",
       "                                                      country_date_strain  \\\n",
       "sample                                                                      \n",
       "Reference                            United States of America (1992) CO92   \n",
       "GCA_009909635.1_ASM990963v1_genomic                  Russia (1923.0) 9_10   \n",
       "GCA_009669545.1_ASM966954v1_genomic                  China (2006.0) 42126   \n",
       "GCA_009669555.1_ASM966955v1_genomic                  China (2005.0) 42123   \n",
       "GCA_009669565.1_ASM966956v1_genomic                  China (2005.0) 42118   \n",
       "...                                                                   ...   \n",
       "NODE595                                                                NA   \n",
       "NODE596                                                                NA   \n",
       "NODE597                                                                NA   \n",
       "NODE598                                                                NA   \n",
       "NODE599                                                                NA   \n",
       "\n",
       "                                                     province_date_strain  \\\n",
       "sample                                                                      \n",
       "Reference                            United States of America (1992) CO92   \n",
       "GCA_009909635.1_ASM990963v1_genomic                  Russia (1923.0) 9_10   \n",
       "GCA_009669545.1_ASM966954v1_genomic                  China (2006.0) 42126   \n",
       "GCA_009669555.1_ASM966955v1_genomic                  China (2005.0) 42123   \n",
       "GCA_009669565.1_ASM966956v1_genomic                  China (2005.0) 42118   \n",
       "...                                                                   ...   \n",
       "NODE595                                                                NA   \n",
       "NODE596                                                                NA   \n",
       "NODE597                                                                NA   \n",
       "NODE598                                                                NA   \n",
       "NODE599                                                                NA   \n",
       "\n",
       "                                    root_rtt_dist population_rtt_dist  \\\n",
       "sample                                                                  \n",
       "Reference                             7.31686e-05          6.3815e-06   \n",
       "GCA_009909635.1_ASM990963v1_genomic   7.30501e-05          9.6582e-06   \n",
       "GCA_009669545.1_ASM966954v1_genomic   5.41847e-05         1.15566e-05   \n",
       "GCA_009669555.1_ASM966955v1_genomic   5.47035e-05         1.20754e-05   \n",
       "GCA_009669565.1_ASM966956v1_genomic    5.4939e-05         1.23109e-05   \n",
       "...                                           ...                 ...   \n",
       "NODE595                                        NA                  NA   \n",
       "NODE596                                        NA                  NA   \n",
       "NODE597                                        NA                  NA   \n",
       "NODE598                                        NA                  NA   \n",
       "NODE599                                        NA                  NA   \n",
       "\n",
       "                                         rate_sub  rate_sub_year  country_lat  \\\n",
       "sample                                                                          \n",
       "Reference                            2.463945e-08       0.104203    39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic  2.412242e-09       0.010202    64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic  0.000000e+00       0.000000    35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic  9.662073e-09       0.040862    35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic  1.312209e-08       0.055495    35.000074   \n",
       "...                                           ...            ...          ...   \n",
       "NODE595                              0.000000e+00       0.000000    -6.869970   \n",
       "NODE596                              0.000000e+00       0.000000    -6.869970   \n",
       "NODE597                              8.797265e-09       0.037204    -6.869970   \n",
       "NODE598                              1.216191e-09       0.005143    -6.869970   \n",
       "NODE599                              1.405518e-08       0.059441    -6.869970   \n",
       "\n",
       "                                     country_lon         mugration_country  \\\n",
       "sample                                                                       \n",
       "Reference                            -100.445882  United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic    97.745306                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic   104.999927                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic   104.999927                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic   104.999927                     China   \n",
       "...                                          ...                       ...   \n",
       "NODE595                               -75.045851                      Peru   \n",
       "NODE596                               -75.045851                      Peru   \n",
       "NODE597                               -75.045851                      Peru   \n",
       "NODE598                               -75.045851                      Peru   \n",
       "NODE599                               -75.045851                      Peru   \n",
       "\n",
       "                                     mugration_country_confidence  \\\n",
       "sample                                                              \n",
       "Reference                                                     1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.0   \n",
       "...                                                           ...   \n",
       "NODE595                                                       1.0   \n",
       "NODE596                                                       1.0   \n",
       "NODE597                                                       1.0   \n",
       "NODE598                                                       1.0   \n",
       "NODE599                                                       1.0   \n",
       "\n",
       "                                     province_lat  province_lon  \\\n",
       "sample                                                            \n",
       "Reference                               38.725178   -105.607716   \n",
       "GCA_009909635.1_ASM990963v1_genomic     47.622245     40.795794   \n",
       "GCA_009669545.1_ASM966954v1_genomic     42.480495     85.463346   \n",
       "GCA_009669555.1_ASM966955v1_genomic     42.480495     85.463346   \n",
       "GCA_009669565.1_ASM966956v1_genomic     42.480495     85.463346   \n",
       "...                                           ...           ...   \n",
       "NODE595                                 -6.250000    -78.833333   \n",
       "NODE596                                 -6.250000    -78.833333   \n",
       "NODE597                                 -6.250000    -78.833333   \n",
       "NODE598                                 -6.250000    -78.833333   \n",
       "NODE599                                 -6.250000    -78.833333   \n",
       "\n",
       "                                    mugration_province  \\\n",
       "sample                                                   \n",
       "Reference                                     Colorado   \n",
       "GCA_009909635.1_ASM990963v1_genomic      Rostov Oblast   \n",
       "GCA_009669545.1_ASM966954v1_genomic           Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic           Xinjiang   \n",
       "GCA_009669565.1_ASM966956v1_genomic           Xinjiang   \n",
       "...                                                ...   \n",
       "NODE595                                      Cajamarca   \n",
       "NODE596                                      Cajamarca   \n",
       "NODE597                                      Cajamarca   \n",
       "NODE598                                      Cajamarca   \n",
       "NODE599                                      Cajamarca   \n",
       "\n",
       "                                     mugration_province_confidence  \\\n",
       "sample                                                               \n",
       "Reference                                                     1.00   \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.00   \n",
       "GCA_009669545.1_ASM966954v1_genomic                           1.00   \n",
       "GCA_009669555.1_ASM966955v1_genomic                           1.00   \n",
       "GCA_009669565.1_ASM966956v1_genomic                           1.00   \n",
       "...                                                            ...   \n",
       "NODE595                                                       1.00   \n",
       "NODE596                                                       1.00   \n",
       "NODE597                                                       1.00   \n",
       "NODE598                                                       1.00   \n",
       "NODE599                                                       0.99   \n",
       "\n",
       "                                    mugration_continent  \\\n",
       "sample                                                    \n",
       "Reference                                 North America   \n",
       "GCA_009909635.1_ASM990963v1_genomic              Europe   \n",
       "GCA_009669545.1_ASM966954v1_genomic                Asia   \n",
       "GCA_009669555.1_ASM966955v1_genomic                Asia   \n",
       "GCA_009669565.1_ASM966956v1_genomic                Asia   \n",
       "...                                                 ...   \n",
       "NODE595                                   South America   \n",
       "NODE596                                   South America   \n",
       "NODE597                                   South America   \n",
       "NODE598                                   South America   \n",
       "NODE599                                   South America   \n",
       "\n",
       "                                     mugration_continent_confidence  \\\n",
       "sample                                                                \n",
       "Reference                                                       1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                             1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                             1.0   \n",
       "...                                                             ...   \n",
       "NODE595                                                         1.0   \n",
       "NODE596                                                         1.0   \n",
       "NODE597                                                         1.0   \n",
       "NODE598                                                         1.0   \n",
       "NODE599                                                         1.0   \n",
       "\n",
       "                                    mugration_host_order  \\\n",
       "sample                                                     \n",
       "Reference                                          Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic                Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic             Rodentia   \n",
       "GCA_009669555.1_ASM966955v1_genomic         Siphonaptera   \n",
       "GCA_009669565.1_ASM966956v1_genomic             Rodentia   \n",
       "...                                                  ...   \n",
       "NODE595                                            Human   \n",
       "NODE596                                         Rodentia   \n",
       "NODE597                                         Rodentia   \n",
       "NODE598                                            Human   \n",
       "NODE599                                         Rodentia   \n",
       "\n",
       "                                     mugration_host_order_confidence  \\\n",
       "sample                                                                 \n",
       "Reference                                                       1.00   \n",
       "GCA_009909635.1_ASM990963v1_genomic                             1.00   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             1.00   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             1.00   \n",
       "GCA_009669565.1_ASM966956v1_genomic                             1.00   \n",
       "...                                                              ...   \n",
       "NODE595                                                         0.55   \n",
       "NODE596                                                         0.61   \n",
       "NODE597                                                         0.56   \n",
       "NODE598                                                         0.60   \n",
       "NODE599                                                         0.82   \n",
       "\n",
       "                                    mugration_population  \\\n",
       "sample                                                     \n",
       "Reference                                          1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic                2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic                0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic                0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic                0.ANT   \n",
       "...                                                  ...   \n",
       "NODE595                                            1.ORI   \n",
       "NODE596                                            1.ORI   \n",
       "NODE597                                            1.ORI   \n",
       "NODE598                                            1.ORI   \n",
       "NODE599                                            1.ORI   \n",
       "\n",
       "                                     mugration_population_confidence  \\\n",
       "sample                                                                 \n",
       "Reference                                                        1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic                              1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                              1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                              1.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic                              1.0   \n",
       "...                                                              ...   \n",
       "NODE595                                                          1.0   \n",
       "NODE596                                                          1.0   \n",
       "NODE597                                                          1.0   \n",
       "NODE598                                                          1.0   \n",
       "NODE599                                                          1.0   \n",
       "\n",
       "                                    timetree_num_date_confidence  \\\n",
       "sample                                                             \n",
       "Reference                                       [1992.0, 1992.0]   \n",
       "GCA_009909635.1_ASM990963v1_genomic             [1923.0, 1923.0]   \n",
       "GCA_009669545.1_ASM966954v1_genomic             [2006.0, 2006.0]   \n",
       "GCA_009669555.1_ASM966955v1_genomic             [2005.0, 2005.0]   \n",
       "GCA_009669565.1_ASM966956v1_genomic             [2005.0, 2005.0]   \n",
       "...                                                          ...   \n",
       "NODE595                                         [1884.0, 1932.0]   \n",
       "NODE596                                         [1884.0, 1932.0]   \n",
       "NODE597                                         [1905.0, 1971.0]   \n",
       "NODE598                                         [1915.0, 1981.0]   \n",
       "NODE599                                         [1933.0, 1994.0]   \n",
       "\n",
       "                                                               node_name  \\\n",
       "sample                                                                     \n",
       "Reference                                                      Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic  GCA_009909635.1_ASM990963v1_genomic   \n",
       "GCA_009669545.1_ASM966954v1_genomic  GCA_009669545.1_ASM966954v1_genomic   \n",
       "GCA_009669555.1_ASM966955v1_genomic  GCA_009669555.1_ASM966955v1_genomic   \n",
       "GCA_009669565.1_ASM966956v1_genomic  GCA_009669565.1_ASM966956v1_genomic   \n",
       "...                                                                  ...   \n",
       "NODE595                                                          NODE595   \n",
       "NODE596                                                          NODE596   \n",
       "NODE597                                                          NODE597   \n",
       "NODE598                                                          NODE598   \n",
       "NODE599                                                          NODE599   \n",
       "\n",
       "                                    blank country_conf_category  \\\n",
       "sample                                                            \n",
       "Reference                                                  HIGH   \n",
       "GCA_009909635.1_ASM990963v1_genomic                        HIGH   \n",
       "GCA_009669545.1_ASM966954v1_genomic                        HIGH   \n",
       "GCA_009669555.1_ASM966955v1_genomic                        HIGH   \n",
       "GCA_009669565.1_ASM966956v1_genomic                        HIGH   \n",
       "...                                   ...                   ...   \n",
       "NODE595                                                    HIGH   \n",
       "NODE596                                                    HIGH   \n",
       "NODE597                                                    HIGH   \n",
       "NODE598                                                    HIGH   \n",
       "NODE599                                                    HIGH   \n",
       "\n",
       "                                    province_conf_category  \\\n",
       "sample                                                       \n",
       "Reference                                             HIGH   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   HIGH   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   HIGH   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   HIGH   \n",
       "GCA_009669565.1_ASM966956v1_genomic                   HIGH   \n",
       "...                                                    ...   \n",
       "NODE595                                               HIGH   \n",
       "NODE596                                               HIGH   \n",
       "NODE597                                               HIGH   \n",
       "NODE598                                               HIGH   \n",
       "NODE599                                               HIGH   \n",
       "\n",
       "                                    continent_conf_category  \\\n",
       "sample                                                        \n",
       "Reference                                              HIGH   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    HIGH   \n",
       "GCA_009669545.1_ASM966954v1_genomic                    HIGH   \n",
       "GCA_009669555.1_ASM966955v1_genomic                    HIGH   \n",
       "GCA_009669565.1_ASM966956v1_genomic                    HIGH   \n",
       "...                                                     ...   \n",
       "NODE595                                                HIGH   \n",
       "NODE596                                                HIGH   \n",
       "NODE597                                                HIGH   \n",
       "NODE598                                                HIGH   \n",
       "NODE599                                                HIGH   \n",
       "\n",
       "                                    host_order_conf_category  \\\n",
       "sample                                                         \n",
       "Reference                                               HIGH   \n",
       "GCA_009909635.1_ASM990963v1_genomic                     HIGH   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     HIGH   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     HIGH   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     HIGH   \n",
       "...                                                      ...   \n",
       "NODE595                                                  LOW   \n",
       "NODE596                                                  LOW   \n",
       "NODE597                                                  LOW   \n",
       "NODE598                                                  LOW   \n",
       "NODE599                                                  LOW   \n",
       "\n",
       "                                    population_conf_category  \n",
       "sample                                                        \n",
       "Reference                                               HIGH  \n",
       "GCA_009909635.1_ASM990963v1_genomic                     HIGH  \n",
       "GCA_009669545.1_ASM966954v1_genomic                     HIGH  \n",
       "GCA_009669555.1_ASM966955v1_genomic                     HIGH  \n",
       "GCA_009669565.1_ASM966956v1_genomic                     HIGH  \n",
       "...                                                      ...  \n",
       "NODE595                                                 HIGH  \n",
       "NODE596                                                 HIGH  \n",
       "NODE597                                                 HIGH  \n",
       "NODE598                                                 HIGH  \n",
       "NODE599                                                 HIGH  \n",
       "\n",
       "[1201 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remember, order mug attrs when dealing with confidence!\n",
    "\n",
    "columns = [\n",
    "    # Draw Divergence Tree\n",
    "    \"branch_length\",\n",
    "    \"branch_length_sub\",\n",
    "    # Draw Time Tree\n",
    "    \"branch_length_time\",\n",
    "    \"timetree_date_calendar\",\n",
    "    \"date_mean\",\n",
    "    \"date_err\",\n",
    "    \"date_bp_mean\",    \n",
    "    # Branch Support\n",
    "    \"branch_support\",\n",
    "    \"branch_support_conf_category\",  \n",
    "    \"branch_support_conf_char\",  \n",
    "    # Filters\n",
    "    \"node_type\",\n",
    "    \"branch_number\",    \n",
    "    # Text Description\n",
    "    \"biosample_accession\",\n",
    "    \"strain\",\n",
    "    \"country_date_strain\",\n",
    "    \"province_date_strain\",\n",
    "    # Stats\n",
    "    \"root_rtt_dist\",\n",
    "    \"population_rtt_dist\",\n",
    "    \"rate_sub\",\n",
    "    \"rate_sub_year\", \n",
    "]\n",
    "\n",
    "for attr in mug_dict:\n",
    "    mug_colname = \"mugration_\" + attr\n",
    "    conf_colname = \"mugration_\" + attr + \"_confidence\"\n",
    "    \n",
    "    lat_colname = attr + \"_lat\"\n",
    "    lon_colname = attr + \"_lon\"\n",
    "    if lat_colname in metadata_df.columns:\n",
    "        lat_colname = \"mugration_\" + lat_colname\n",
    "        lon_colname = \"mugration_\" + lon_colname\n",
    "        columns.append(lat_colname)\n",
    "        columns.append(lon_colname)\n",
    "        \n",
    "    columns.append(mug_colname)\n",
    "    columns.append(conf_colname)\n",
    "    \n",
    "\n",
    "# Edit df\n",
    "auspice_df = copy.copy(metadata_df[columns])\n",
    "auspice_df[\"timetree_num_date_confidence\"] = [NO_DATA_CHAR] * len(auspice_df)\n",
    "auspice_df[\"node_name\"] = list(auspice_df.index)\n",
    "auspice_df[\"blank\"] = [\" \"] * len(auspice_df)\n",
    "\n",
    "# Add mugration confidence categories\n",
    "for attr in mug_dict:\n",
    "    colname = attr + \"_conf_category\"\n",
    "    for rec in auspice_df.iterrows():\n",
    "        sample = rec[0]\n",
    "        conf = auspice_df[\"mugration_\" + attr + \"_confidence\"][sample]\n",
    "        category = \"LOW\"\n",
    "        if conf >= 0.95:\n",
    "            category = \"HIGH\"\n",
    "        auspice_df.at[sample, colname] = category\n",
    "\n",
    "# Date formating\n",
    "for rec in metadata_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    date = rec[1][\"timetree_date_calendar\"]\n",
    "    date_lower = float(date) - rec[1][\"timetree_CI_lower_err\"]\n",
    "    date_upper = float(date) + rec[1][\"timetree_CI_upper_err\"]\n",
    "    date_confidence = [date_lower, date_upper]\n",
    "    auspice_df.at[sample,\"timetree_num_date_confidence\"] = date_confidence\n",
    "    \n",
    "    # Rounding\n",
    "    auspice_df.at[sample, \"branch_length_sub\"] = round(auspice_df[\"branch_length_sub\"][sample])\n",
    "    auspice_df.at[sample, \"branch_length_time\"] = round(auspice_df[\"branch_length_time\"][sample])\n",
    "\n",
    "        \n",
    "# Rename columns\n",
    "auspice_df.rename(columns={\"timetree_date_calendar\": \"timetree_num_date\"}, inplace=True)\n",
    "auspice_df.rename(columns={\"mugration_country_lat\": \"country_lat\"}, inplace=True)\n",
    "auspice_df.rename(columns={\"mugration_country_lon\": \"country_lon\"}, inplace=True)\n",
    "auspice_df.rename(columns={\"mugration_province_lat\": \"province_lat\"}, inplace=True)\n",
    "auspice_df.rename(columns={\"mugration_province_lon\": \"province_lon\"}, inplace=True)\n",
    "display(auspice_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "silent-tuesday",
   "metadata": {},
   "source": [
    "### JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "novel-enough",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'branch_length': 4.0046e-06, 'branch_length_sub': 17.0, 'branch_length_time': 163.0, 'num_date': 1992.0, 'date_mean': 1992.0, 'date_err': 0.0, 'date_bp_mean': 29.0, 'branch_support': 0.0, 'branch_support_conf_category': 'LOW', 'branch_support_conf_char': '', 'node_type': 'terminal', 'branch_number': '1.0', 'biosample_accession': 'SAMEA1705942', 'strain': 'CO92', 'country_date_strain': 'United States of America (1992) CO92', 'province_date_strain': 'United States of America (1992) CO92', 'root_rtt_dist': 7.316859999999999e-05, 'population_rtt_dist': 6.3815e-06, 'rate_sub': 2.4639446741484545e-08, 'rate_sub_year': 0.1042026349355188, 'country_lat': 39.7837304, 'country_lon': -100.4458825, 'country': 'United States of America', 'country_confidence': {'United States of America': 1.0}, 'province_lat': 38.7251776, 'province_lon': -105.607716, 'province': 'Colorado', 'province_confidence': {'Colorado': 1.0}, 'continent': 'North America', 'continent_confidence': {'North America': 1.0}, 'host_order': 'Human', 'host_order_confidence': {'Human': 1.0}, 'population': '1.ORI', 'population_confidence': {'1.ORI': 1.0}, 'num_date_confidence': [1992.0, 1992.0], 'node_name': 'Reference', 'blank': ' ', 'country_conf_category': 'HIGH', 'province_conf_category': 'HIGH', 'continent_conf_category': 'HIGH', 'host_order_conf_category': 'HIGH', 'population_conf_category': 'HIGH'}\n"
     ]
    }
   ],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=None, \n",
    "    aln_path=None,  \n",
    "    tree=divtree, \n",
    "    tree_df=auspice_df, \n",
    "    color_keyword_exclude=[\"geometry\"],\n",
    "    type_convert = {\n",
    "        \"branch_number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])\n",
    "\n",
    "out_path_augur_json = os.path.join(augur_dir, \"all.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hybrid-artwork",
   "metadata": {},
   "source": [
    "## Auspice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "parental-subdivision",
   "metadata": {},
   "source": [
    "### JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "buried-business",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: turkmenistan, myanmar, kyrgyzstan, democratic republic of the congo, madagascar, kazakhstan, georgia, vietnam, uganda. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/all.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/all.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n"
     ]
    }
   ],
   "source": [
    "#import sys, importlib\n",
    "#importlib.reload(sys.modules['functions'])\n",
    "#from functions import auspice_export, branch_attributes\n",
    "\n",
    "auspice_dict = auspice_export(\n",
    "    tree=divtree,\n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=out_path_colors,\n",
    "    auspice_latlons_path=out_path_latlon, \n",
    "    )\n",
    "\n",
    "\n",
    "label_col = list(auspice_df.columns)\n",
    "\n",
    "# Recursively add branch attrs\n",
    "branch_attributes(\n",
    "    tree_dict=auspice_dict[\"tree\"], \n",
    "    sub_dict=auspice_dict[\"tree\"], \n",
    "    df=auspice_df,\n",
    "    label_col=label_col,\n",
    "    )\n",
    "\n",
    "# Last manual changes\n",
    "auspice_dict_copy = copy.deepcopy(auspice_dict)\n",
    "for i in range(0, len(auspice_dict_copy[\"meta\"][\"colorings\"])):\n",
    "    coloring = auspice_dict_copy[\"meta\"][\"colorings\"][i]\n",
    "    for key in coloring:\n",
    "        # Node type as internal or terminal\n",
    "        if coloring[key] == \"node_type\":\n",
    "            auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['internal', '#FFFFFF'], ['terminal', '#000000']]\n",
    "            #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "        # Confidence category\n",
    "        if \"conf_category\" in coloring[key]:\n",
    "            auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['LOW', '#FFFFFF'], ['HIGH', '#000000']]\n",
    "            #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "        # Host Human binary\n",
    "        #if \"host_human\" in coloring[key]:\n",
    "        #    auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['Human', '#CBB742'], ['Non-Human', \"#60B6F2\"], ['NA', \"#D6D6D6\"]]\n",
    "\n",
    "        \n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(auspice_dir, \"all.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\\n\")\n",
    "\n",
    "#out_path_auspice_remote_json = os.path.join(auspice_dir, \"{}_all.json\".format(AUSPICE_PREFIX))\n",
    "#utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "#export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "#print(\"Validation successful for local JSON.\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "amended-danish",
   "metadata": {},
   "source": [
    "## Subtrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "neutral-clause",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.PRE #8000ff\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEPRECATED: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/0.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/0.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_order_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_order_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.PE #5148fc\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: georgia. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/0.PE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/0.PE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "0.ANT #238af5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: kyrgyzstan. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/0.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/0.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "0.ANT4 #0cc1e8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/0.ANT4.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/0.ANT4.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_order_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_order_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.ANT #3ae8d7\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/3.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/3.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.ANT #68fcc1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/4.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/4.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.ANT #97fca7\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/2.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/2.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "2.MED #c5e88a\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: turkmenistan, kazakhstan. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/2.MED.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/2.MED.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "1.PRE #f3c16a\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/1.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/1.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "1.ANT #ff8a48\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_order_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_order_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/1.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/1.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "1.IN #ff4824\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: uganda, democratic republic of the congo. Auspice will create colors for them.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/1.IN.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/1.IN.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "1.ORI #ff0000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"country_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"continent_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait province were not specified in your provided color scale: myanmar, vietnam, madagascar. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/1.ORI.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/auspice/all/chromosome/full/filter5/ml/1.ORI.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"population_conf_category\" specifies a hex code for \"LOW\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n"
     ]
    }
   ],
   "source": [
    "#import sys, importlib\n",
    "#importlib.reload(sys.modules['functions'])\n",
    "#from functions import extract_subtree\n",
    "\n",
    "for population in population_list:\n",
    "    timetree_copy = copy.deepcopy(timetree)\n",
    "    \n",
    "    # Create the subtree df\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    color = df[\"population_color\"][0]\n",
    "    \n",
    "    #if len(subtree_df) < 2: continue\n",
    "    print(population, color)\n",
    "\n",
    "    df.sort_values(\"timetree_coord_y\", inplace=True)\n",
    "    tips = df[df[\"node_type\"] == \"terminal\"]\n",
    "    internals = df[df[\"node_type\"] == \"internal\"]\n",
    "    \n",
    "    # ----------------------------------------------\n",
    "    # Subtree Timetree\n",
    "    subtree = extract_subtree(\n",
    "        tree=timetree_copy, \n",
    "        tips=list(tips.index), \n",
    "        df=metadata_df, \n",
    "        color_branches=False)\n",
    "    \n",
    "    metadata_to_comment(subtree, metadata_df)   \n",
    "\n",
    "\n",
    "    out_subtree_nex_path = os.path.join(augur_dir, \"{}.timetree.nex\".format(population))\n",
    "    Phylo.write(subtree, out_subtree_nex_path, \"nexus\")\n",
    "    out_subtree_nwk_path = os.path.join(augur_dir, \"{}.timetree.nwk\".format(population))\n",
    "    Phylo.write(subtree, out_subtree_nwk_path, \"newick\")\n",
    "    \n",
    "    # ----------------------------------------------\n",
    "    # Subtree divtree\n",
    "    divtree_copy = copy.deepcopy(divtree)\n",
    "    subtree_div = extract_subtree(\n",
    "        tree=divtree_copy, \n",
    "        tips=list(tips.index), \n",
    "        df=metadata_df, \n",
    "        color_branches=False)\n",
    "    \n",
    "    metadata_to_comment(subtree_div, metadata_df)   \n",
    "    out_subtree_path = os.path.join(augur_dir, \"{}.divtree.nex\".format(population))\n",
    "    Phylo.write(subtree, out_subtree_path, \"nexus\", format_branch_length=\"%1.10f\")\n",
    "\n",
    "    #fig, ax = plt.subplots(1)\n",
    "    #Phylo.draw(subtree, axes=ax, show_confidence=False, label_func = lambda x: '', do_show=False)\n",
    "    \n",
    "    augur_dict = augur_export(\n",
    "        tree_path=None, \n",
    "        aln_path=None,  \n",
    "        tree=subtree, \n",
    "        tree_df=auspice_df, \n",
    "        color_keyword_exclude=[\"geometry\"],\n",
    "        type_convert = {\n",
    "            \"branch_number\" : (lambda x : str(x))\n",
    "        },\n",
    "    ) \n",
    "        \n",
    "    out_path_augur_json = os.path.join(augur_dir, \"{}.json\".format(population) )\n",
    "    utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "\n",
    "    auspice_dict = auspice_export(\n",
    "        tree=subtree,\n",
    "        augur_json_paths=[out_path_augur_json], \n",
    "        auspice_config_path=auspice_config_path, \n",
    "        auspice_colors_path=out_path_colors,\n",
    "        auspice_latlons_path=out_path_latlon, \n",
    "        )    \n",
    "    \n",
    "    label_col = list(auspice_df.columns)\n",
    "\n",
    "    # Recursively add branch attrs\n",
    "    branch_attributes(\n",
    "        tree_dict=auspice_dict[\"tree\"], \n",
    "        sub_dict=auspice_dict[\"tree\"], \n",
    "        df=auspice_df,\n",
    "        label_col=label_col,\n",
    "        )    \n",
    "    \n",
    "    # Last manual changes\n",
    "    auspice_dict_copy = copy.deepcopy(auspice_dict)\n",
    "    for i in range(0, len(auspice_dict_copy[\"meta\"][\"colorings\"])):\n",
    "        coloring = auspice_dict_copy[\"meta\"][\"colorings\"][i]\n",
    "        for key in coloring:\n",
    "            # Node type as internal or terminal\n",
    "            if coloring[key] == \"node_type\":\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['internal', '#FFFFFF'], ['terminal', color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            # Confidence category\n",
    "            if \"conf_category\" in coloring[key]:\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['LOW', '#FFFFFF'], ['HIGH', color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i]) \n",
    "    \n",
    "    # Write outputs - For Local Rendering\n",
    "    out_path_auspice_local_json = os.path.join(auspice_dir, \"{}.json\".format(population) )\n",
    "    utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "    export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "    print(\"Validation successful for local JSON.\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acting-karma",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
