{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "proprietary-spice",
   "metadata": {},
   "source": [
    "# BEAST Analysis Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "gross-cutting",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 0. SETUP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "expanded-saudi",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from Bio import Phylo, AlignIO\n",
    "from functions import *\n",
    "import subprocess\n",
    "from matplotlib import colors\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "connected-livestock",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "backed-hybrid",
   "metadata": {},
   "outputs": [],
   "source": [
    "WILDCARDS = [\"all\", \"chromosome\", \"clade\", \"5\"]\n",
    "project_dir = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/\"\n",
    "results_dir = project_dir\n",
    "\n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "PRUNE = WILDCARDS[2]\n",
    "MISSING_DATA = WILDCARDS[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "removable-vulnerability",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_dir            = project_dir + \"/beast/all/chromosome/{}/phylogeography/\".format(PRUNE)\n",
    "metadata_path       = project_dir + \"/iqtree/all/chromosome/full/filter{}/filter-taxa/metadata.tsv\".format(MISSING_DATA)\n",
    "auspice_config_path = project_dir + \"/config/auspice_config.json\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "constant_sites_path = project_dir + \"/snippy_multi/all/chromosome/full/snippy-multi.constant_sites.txt\".format(PRUNE)\n",
    "aln_path            = project_dir + \"iqtree/all/chromosome/full/filter{}/filter-clades/1.ORI/1.ORI.fasta\".format(MISSING_DATA)\n",
    "\n",
    "# ------------------------------------------\n",
    "# Output\n",
    "auspice_dir = os.path.join(results_dir, \"auspice/all/chromosome/full/filter{}/beast/\".format(MISSING_DATA))\n",
    "if not os.path.exists(auspice_dir):\n",
    "    subprocess.run([\"mkdir\", \"-p\", auspice_dir]) \n",
    "    \n",
    "augur_dir = os.path.join(results_dir, \"augur/all/chromosome/full/filter{}/beast/\".format(MISSING_DATA))\n",
    "if not os.path.exists(augur_dir):\n",
    "    subprocess.run([\"mkdir\", \"-p\", augur_dir]) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ancient-syria",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "linear-burden",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", None)\n",
    "\n",
    "# ------------------------------------------\n",
    "BRANCH_LIST = {\n",
    "    \"1.ORI\" : [\"1.ORI1\", \"1.ORI2\", \"1.ORI3\"],\n",
    "    \"1.IN\": [\"1.IN1\",\"1.IN2\",\"1.IN3\"],  \n",
    "    \"1.ANT\": [\"1.ANT1\"], \n",
    "    \"1.PRE\" : [\"1.PRE0\",\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],\n",
    "    \"2.MED\": [\"2.MED0\", \"2.MED1\",\"2.MED2\",\"2.MED3\" ],      \n",
    "    \"2.ANT\": [\"2.ANT1\",\"2.ANT2\",\"2.ANT3\" ],    \n",
    "    \"4.ANT\": [\"4.ANT1\" ],       \n",
    "    \"3.ANT\": [\"3.ANT1\", \"3.ANT2\" ],  \n",
    "    \"0.ANT\": [\"0.ANT1\", \"0.ANT2\",\"0.ANT3\",\"0.ANT5\"],         \n",
    "    \"0.ANT4\" : [\"0.ANT4\"], \n",
    "    \"0.PE\": [\"0.PE2\", \"0.PE4m\", \"0.PE4m\", \"0.PE4t\", \"0.PE4a\", \"0.PE5\", \"0.PE7\", \"0.PE8\", \"0.PE10\"],   \n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"],        \n",
    "}\n",
    "\n",
    "NUM_STATES = 10\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "JSON_INDENT=2\n",
    "CURRENT_YEAR = 2021\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "with open(constant_sites_path) as infile:\n",
    "    data = infile.read().strip().split(\",\")\n",
    "    constant_sites = sum([int(count) for count in data])\n",
    "\n",
    "aln = AlignIO.read(aln_path, \"fasta\")\n",
    "variant_sites = len(aln[0].seq)\n",
    "SEQ_LEN = constant_sites + variant_sites\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "UNKNOWN_CHAR = \"?\"\n",
    "\n",
    "POSTERIOR_THRESH = 95"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mechanical-monte",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 1. IMPORT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "convertible-freeze",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "beneficial-parcel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "      <th>branch_major_color</th>\n",
       "      <th>geometry_size</th>\n",
       "      <th>geometry</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>Human</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POINT (-105.607716 38.7251776)</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>#b3f396</td>\n",
       "      <td>4.0</td>\n",
       "      <td>POINT (40.7957942 47.6222451)</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>#1996f3</td>\n",
       "      <td>105.0</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>#1996f3</td>\n",
       "      <td>105.0</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>#1996f3</td>\n",
       "      <td>105.0</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313243_45</th>\n",
       "      <td>Azov38</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313243_45</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>4.0</td>\n",
       "      <td>POINT (40.7957942 47.6222451)</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.000022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313246_49</th>\n",
       "      <td>Gdansk8</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Poland</td>\n",
       "      <td>Pomeranian Voivodeship</td>\n",
       "      <td>52.215933</td>\n",
       "      <td>19.134422</td>\n",
       "      <td>54.2456</td>\n",
       "      <td>18.1099</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313246_49</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>54.245560</td>\n",
       "      <td>18.109900</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POINT (18.1099 54.24556)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6651390</th>\n",
       "      <td>AGU010</td>\n",
       "      <td>[1435:1477]</td>\n",
       "      <td>[-586:-544]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6651390</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1456.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>3.0</td>\n",
       "      <td>POINT (25.24953400167952 54.82269205)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637004</th>\n",
       "      <td>AGU025</td>\n",
       "      <td>[1441:1612]</td>\n",
       "      <td>[-580:-409]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637004</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1526.5</td>\n",
       "      <td>494.5</td>\n",
       "      <td>85.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>3.0</td>\n",
       "      <td>POINT (25.24953400167952 54.82269205)</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637002</th>\n",
       "      <td>AGU007B</td>\n",
       "      <td>[1463:1632]</td>\n",
       "      <td>[-558:-389]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637002</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1547.5</td>\n",
       "      <td>473.5</td>\n",
       "      <td>84.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>3.0</td>\n",
       "      <td>POINT (25.24953400167952 54.82269205)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      strain         date      date_bp  \\\n",
       "sample                                                                   \n",
       "Reference                               CO92         1992          -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic     9_10       1923.0          -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic    42126       2006.0          -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic    42123       2005.0          -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic    42118       2005.0          -16   \n",
       "...                                      ...          ...          ...   \n",
       "SAMEA7313243_45                       Azov38  [1400:1700]  [-621:-321]   \n",
       "SAMEA7313246_49                      Gdansk8  [1400:1700]  [-621:-321]   \n",
       "SAMEA6651390                          AGU010  [1435:1477]  [-586:-544]   \n",
       "SAMEA6637004                          AGU025  [1441:1612]  [-580:-409]   \n",
       "SAMEA6637002                         AGU007B  [1463:1632]  [-558:-389]   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "SAMEA7313243_45                                        Russia   \n",
       "SAMEA7313246_49                                        Poland   \n",
       "SAMEA6651390                                        Lithuania   \n",
       "SAMEA6637004                                        Lithuania   \n",
       "SAMEA6637002                                        Lithuania   \n",
       "\n",
       "                                                   province  country_lat  \\\n",
       "sample                                                                     \n",
       "Reference                                          Colorado    39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic           Rostov Oblast    64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic                Xinjiang    35.000074   \n",
       "...                                                     ...          ...   \n",
       "SAMEA7313243_45                               Rostov Oblast    64.686314   \n",
       "SAMEA7313246_49                      Pomeranian Voivodeship    52.215933   \n",
       "SAMEA6651390                                 Vilnius County    55.350000   \n",
       "SAMEA6637004                                 Vilnius County    55.350000   \n",
       "SAMEA6637002                                 Vilnius County    55.350000   \n",
       "\n",
       "                                     country_lon province_lat province_lon  \\\n",
       "sample                                                                       \n",
       "Reference                            -100.445882      38.7252     -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic    97.745306      47.6222      40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic   104.999927      42.4805      85.4633   \n",
       "...                                          ...          ...          ...   \n",
       "SAMEA7313243_45                        97.745306      47.6222      40.7958   \n",
       "SAMEA7313246_49                        19.134422      54.2456      18.1099   \n",
       "SAMEA6651390                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637004                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637002                           23.750000      54.8227      25.2495   \n",
       "\n",
       "                                              biovar branch_major  \\\n",
       "sample                                                              \n",
       "Reference                                 Orientalis        1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic       Medievalis        2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          Antiqua        0.ANT   \n",
       "...                                              ...          ...   \n",
       "SAMEA7313243_45                      Second Pandemic        1.PRE   \n",
       "SAMEA7313246_49                      Second Pandemic        1.PRE   \n",
       "SAMEA6651390                         Second Pandemic        1.PRE   \n",
       "SAMEA6637004                         Second Pandemic        1.PRE   \n",
       "SAMEA6637002                         Second Pandemic        1.PRE   \n",
       "\n",
       "                                    branch_minor biosample_accession  \\\n",
       "sample                                                                 \n",
       "Reference                                 1.ORI1        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.ANT1        SAMN07722923   \n",
       "...                                          ...                 ...   \n",
       "SAMEA7313243_45                           1.PRE1     SAMEA7313243_45   \n",
       "SAMEA7313246_49                           1.PRE1     SAMEA7313246_49   \n",
       "SAMEA6651390                              1.PRE1        SAMEA6651390   \n",
       "SAMEA6637004                              1.PRE1        SAMEA6637004   \n",
       "SAMEA6637002                              1.PRE1        SAMEA6637002   \n",
       "\n",
       "                                                     biosample_comment  \\\n",
       "sample                                                                   \n",
       "Reference                              KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic              KEEP: Assembly Modern   \n",
       "...                                                                ...   \n",
       "SAMEA7313243_45                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA7313246_49                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA6651390                                         KEEP: SRA Ancient   \n",
       "SAMEA6637004                                         KEEP: SRA Ancient   \n",
       "SAMEA6637002                                         KEEP: SRA Ancient   \n",
       "\n",
       "                                     branch_number      continent  date_mean  \\\n",
       "sample                                                                         \n",
       "Reference                                        1  North America     1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic              2         Europe     1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic              0           Asia     2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic              0           Asia     2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic              0           Asia     2005.0   \n",
       "...                                            ...            ...        ...   \n",
       "SAMEA7313243_45                                  1         Europe     1550.0   \n",
       "SAMEA7313246_49                                  1         Europe     1550.0   \n",
       "SAMEA6651390                                     1         Europe     1456.0   \n",
       "SAMEA6637004                                     1         Europe     1526.5   \n",
       "SAMEA6637002                                     1         Europe     1547.5   \n",
       "\n",
       "                                     date_bp_mean  date_err        lat  \\\n",
       "sample                                                                   \n",
       "Reference                                    29.0       0.0  38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic          98.0       0.0  47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic          15.0       0.0  42.480495   \n",
       "GCA_009669555.1_ASM966955v1_genomic          16.0       0.0  42.480495   \n",
       "GCA_009669565.1_ASM966956v1_genomic          16.0       0.0  42.480495   \n",
       "...                                           ...       ...        ...   \n",
       "SAMEA7313243_45                             471.0     150.0  47.622245   \n",
       "SAMEA7313246_49                             471.0     150.0  54.245560   \n",
       "SAMEA6651390                                565.0      21.0  54.822692   \n",
       "SAMEA6637004                                494.5      85.5  54.822692   \n",
       "SAMEA6637002                                473.5      84.5  54.822692   \n",
       "\n",
       "                                            lon host_human branch_major_color  \\\n",
       "sample                                                                          \n",
       "Reference                           -105.607716      Human            #ff0000   \n",
       "GCA_009909635.1_ASM990963v1_genomic   40.795794      Human            #b3f396   \n",
       "GCA_009669545.1_ASM966954v1_genomic   85.463346  Non-Human            #1996f3   \n",
       "GCA_009669555.1_ASM966955v1_genomic   85.463346  Non-Human            #1996f3   \n",
       "GCA_009669565.1_ASM966956v1_genomic   85.463346  Non-Human            #1996f3   \n",
       "...                                         ...        ...                ...   \n",
       "SAMEA7313243_45                       40.795794      Human            #e6ce74   \n",
       "SAMEA7313246_49                       18.109900      Human            #e6ce74   \n",
       "SAMEA6651390                          25.249534      Human            #e6ce74   \n",
       "SAMEA6637004                          25.249534      Human            #e6ce74   \n",
       "SAMEA6637002                          25.249534      Human            #e6ce74   \n",
       "\n",
       "                                     geometry_size  \\\n",
       "sample                                               \n",
       "Reference                                      1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic            4.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic          105.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic          105.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic          105.0   \n",
       "...                                            ...   \n",
       "SAMEA7313243_45                                4.0   \n",
       "SAMEA7313246_49                                1.0   \n",
       "SAMEA6651390                                   3.0   \n",
       "SAMEA6637004                                   3.0   \n",
       "SAMEA6637002                                   3.0   \n",
       "\n",
       "                                                                  geometry  \\\n",
       "sample                                                                       \n",
       "Reference                                   POINT (-105.607716 38.7251776)   \n",
       "GCA_009909635.1_ASM990963v1_genomic          POINT (40.7957942 47.6222451)   \n",
       "GCA_009669545.1_ASM966954v1_genomic   POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669555.1_ASM966955v1_genomic   POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669565.1_ASM966956v1_genomic   POINT (85.46334640000001 42.4804953)   \n",
       "...                                                                    ...   \n",
       "SAMEA7313243_45                              POINT (40.7957942 47.6222451)   \n",
       "SAMEA7313246_49                                   POINT (18.1099 54.24556)   \n",
       "SAMEA6651390                         POINT (25.24953400167952 54.82269205)   \n",
       "SAMEA6637004                         POINT (25.24953400167952 54.82269205)   \n",
       "SAMEA6637002                         POINT (25.24953400167952 54.82269205)   \n",
       "\n",
       "                                     root_rtt_dist  clade_rtt_dist  \n",
       "sample                                                              \n",
       "Reference                                 0.000073        0.000006  \n",
       "GCA_009909635.1_ASM990963v1_genomic       0.000073        0.000010  \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.000054        0.000012  \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.000055        0.000012  \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.000055        0.000012  \n",
       "...                                            ...             ...  \n",
       "SAMEA7313243_45                           0.000075        0.000022  \n",
       "SAMEA7313246_49                           0.000060        0.000007  \n",
       "SAMEA6651390                              0.000060        0.000006  \n",
       "SAMEA6637004                              0.000061        0.000007  \n",
       "SAMEA6637002                              0.000060        0.000006  \n",
       "\n",
       "[601 rows x 27 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.set_index(metadata_df.columns[0], inplace=True)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "passing-capitol",
   "metadata": {},
   "source": [
    "## Most Recent Sampling Date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "tribal-manchester",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1.ORI': 2016.0, '1.IN': 2008.0, '1.ANT': 2004.0, '1.PRE': 1767.5, '2.MED': 2018.0, '2.ANT': 2008.0, '4.ANT': 2015.0, '3.ANT': 2017.0, '0.ANT': 2019.0, '0.ANT4': 765.0, '0.PE': 2014.0, '0.PRE': -1686.0}\n"
     ]
    }
   ],
   "source": [
    "out_path = os.path.join(augur_dir, \"most_recent_sampling_dates.tsv\")\n",
    "mrsd_dict = {}\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    for branch in BRANCH_LIST:\n",
    "        branch_df = metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])]\n",
    "        if len(branch_df) == 0: continue\n",
    "        max_date = max(branch_df[\"date_mean\"])\n",
    "        outfile.write(\"{}\\t{}\\n\".format(branch, max_date))\n",
    "        mrsd_dict[branch] = max_date\n",
    "        \n",
    "print(mrsd_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "banned-boutique",
   "metadata": {},
   "source": [
    "## Colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "outside-logistics",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'country': {'Kyrgyzstan': '#8000ff', 'Uzbekistan': '#7215ff', 'Myanmar': '#652afe', 'France': '#573ffd', 'England': '#4a53fc', 'Tajikistan': '#3c66fa', 'Algeria': '#2f79f7', 'Armenia': '#228bf4', 'Canada': '#149df1', 'Lithuania': '#07adee', 'Azerbaijan': '#07bcea', 'Zimbabwe': '#14c9e5', 'Brazil': '#22d5e0', 'Germany': '#2fe0db', 'Iran': '#3cead5', 'Bolivia': '#4af1d0', 'Norway': '#57f7c9', 'Spain': '#65fcc3', 'Poland': '#72febc', 'Kazakhstan': '#80ffb4', 'Mongolia': '#8dfead', 'Georgia': '#9afca5', 'Turkmenistan': '#a8f79d', 'Vietnam': '#b5f194', 'Kenya': '#c3ea8b', 'Switzerland': '#d0e083', 'China': '#ddd579', 'Nepal': '#ebc970', 'Indonesia': '#f8bc66', 'India': '#ffad5d', 'Russia': '#ff9d53', 'Uganda': '#ff8b49', 'Italy': '#ff793f', 'Democratic Republic of the Congo': '#ff6634', 'The Netherlands': '#ff532a', 'Estonia': '#ff3f20', 'United States of America': '#ff2a15', 'Peru': '#ff150b', 'Madagascar': '#ff0000', 'NA': '#969696'}, 'province': {'Shaanxi': '#8000ff', 'Hovsgel': '#7b07ff', 'Tibet': '#760eff', 'Almaty Region': '#7215ff', 'Xinjiang': '#6d1cff', 'Krasnoyarsk Krai': '#6923fe', 'Shymkent': '#642afe', 'Talas Region': '#6031fe', 'Cajamarca': '#5b38fd', 'Greater London': '#573ffd', 'Kurdistan': '#5246fd', 'Brandenburg': '#4e4dfc', 'Khovd': '#4954fb', 'Texas': '#455afb', 'Lazio': '#4061fa', 'Midlands': '#3c67f9', 'Bavaria': '#376ef9', 'Baku': '#3374f8', 'Gegharkunik Province': '#2e7af7', 'Zabaykalsky Krai': '#2a81f6', 'Shahbuz Rayon': '#2587f5', 'Nidwalden': '#218df4', 'Tamil Nadu': '#1c92f3', 'Govi-Altai': '#1898f2', 'Qazakh District': '#139ef1', 'Kyzylorda Region': '#0fa3f0', 'Mangystau Province': '#0aa9ef', 'Astrakhan Oblast': '#06aeed', 'Rostov Oblast': '#01b3ec', 'Central Java': '#03b8eb', 'Syunik Province': '#08bde9', 'Centre-Loire Valley': '#0cc2e8', 'Osh Region': '#11c6e6', 'Lâm Đồng Province': '#15cae5', 'West Kazakhstan region': '#1acfe3', 'Issyk-Kul Region': '#1ed3e1', 'Kabardino-Balkaria': '#23d7e0', 'Irkutsk Oblast': '#27dade', 'Yunnan': '#2cdedc', 'Fujian': '#31e1da', 'Saskatchewan': '#35e5d9', 'Catalonia': '#3ae8d7', 'Nevada': '#3eebd5', 'California': '#43edd3', 'Panevezys County': '#47f0d1', 'NA': '#969696', 'State of Mato Grosso': '#50f4cd', 'Ömnögovi': '#55f6ca', 'Heilongjiang': '#59f8c8', 'Province de Fianarantsoa': '#5ef9c6', 'Pomeranian Voivodeship': '#62fbc4', 'East of England': '#67fcc2', \"Provence-Alpes-Côte d'Azur\": '#6bfdbf', 'New Mexico': '#70febd', 'La Libertad': '#74feba', 'Kyzylorda Province': '#79ffb8', 'Baden-Württemberg': '#7dffb6', 'Sughd Province': '#82ffb3', 'North Brabant': '#86ffb1', 'Tatarstan': '#8bfeae', 'Tuva Republic': '#8ffeab', 'Övörkhangai': '#94fda9', 'Altai Republic': '#98fca6', 'Vilnius County': '#9dfba3', 'Colorado': '#a1f9a1', 'Republic of Dagestan': '#a6f89e', 'Inner Mongolia': '#aaf69b', 'Samtskhe-Javakheti': '#aff498', 'Sečuán': '#b3f295', 'Shirak Province': '#b8f092', 'Himachal Pradesh': '#bced90', 'Hebei': '#c1eb8d', 'MSila': '#c5e88a', 'St. Petersberg': '#cae587', 'Goranboy District': '#cee184', 'Occitanie': '#d3de81', 'Karachay-Cherkessia': '#d8da7d', 'Gujarat': '#dcd77a', 'Naryn Region': '#e1d377', 'Khánh Hoà Province': '#e5cf74', 'Shamkir District': '#eaca71', 'Ho Chi Minh City': '#eec66e', 'Altai Krai': '#f3c26b', 'Ingushetia': '#f7bd67', 'Valencia Community': '#fcb864', 'Oslo': '#ffb361', 'Maharashtra': '#ffae5d', 'Gansu': '#ffa95a', 'Stavropol Krai': '#ffa357', 'Jilin': '#ff9e54', 'Atyrau Region': '#ff9850', 'Guangxi Zhuang': '#ff924d', 'Republic of Kalmykia': '#ff8d49', 'Nagorno-Karabakh Republic': '#ff8746', 'Qinghai': '#ff8143', 'Zavkhan Province': '#ff7a3f', 'Aktobe Region': '#ff743c', 'Gandaki Pradesh': '#ff6e38', 'Navoiy Region': '#ff6735', 'Fizuli District': '#ff6131', 'Nairobi': '#ff5a2e', 'Samara Oblast': '#ff542a', 'Balkan': '#ff4d27', 'Bayan-Ölgii': '#ff4623', 'Chechnya': '#ff3f20', 'Atyrau': '#ff381c', 'Kirov Oblast': '#ff3119', 'Ningxia': '#ff2a15', 'Arizona': '#ff2312', 'La Paz': '#ff1c0e', 'West Kazakhstan Region': '#ff150b', 'Dornogovi': '#ff0e07', 'Bayankhongor': '#ff0704', 'Pärnu maakond': '#ff0000'}, 'branch_major': {'0.PRE': '#8000ff', '0.PE': '#5148fc', '0.ANT4': '#238af5', '0.ANT': '#0cc1e8', '3.ANT': '#3ae8d7', '4.ANT': '#68fcc1', '2.ANT': '#97fca7', '2.MED': '#c5e88a', '1.PRE': '#f3c16a', '1.ANT': '#ff8a48', '1.IN': '#ff4824', '1.ORI': '#ff0000', 'NA': '#969696'}, 'branch_minor': {'0.PE4m': '#8000ff', '0.PE4a': '#7116ff', '0.PE2': '#632cfe', '1.PRE0': '#5542fd', '0.PE7': '#4757fb', '0.PE4t': '#396cf9', '3.ANT2': '#2b7ff6', '1.ORI3': '#1c92f3', '0.ANT5': '#0ea4f0', '1.PRE2': '#00b4ec', '1.PRE1': '#0ec3e7', '0.PE4h': '#1cd1e2', '1.ORI2': '#2adddd', '0.PE8': '#39e7d7', '0.PRE1': '#47f0d1', '2.ANT1': '#55f6ca', '3.ANT1': '#63fbc3', '1.IN2': '#71febc', '0.ANT2': '#80ffb4', '2.MED2': '#8efeac', '1.ANT1': '#9cfba4', '2.MED1': '#aaf69b', '0.PE10': '#b8f092', '0.ANT1': '#c6e789', '2.ANT3': '#d4dd80', '4.ANT1': '#e3d176', '2.MED0': '#f1c36c', '1.IN1': '#ffb462', '1.PRE3': '#ffa457', '0.PRE2': '#ff924d', '1.ORI1': '#ff8042', '2.ANT2': '#ff6c37', '0.ANT3': '#ff572c', '0.ANT4': '#ff4221', '0.PE5': '#ff2c16', '1.IN3': '#ff160b', '2.MED3': '#ff0000', 'NA': '#969696'}, 'continent': {'North America': '#8000ff', 'Europe': '#00b4ec', 'South America': '#80ffb4', 'Asia': '#ffb462', 'Africa': '#ff0000', 'NA': '#969696'}}\n"
     ]
    }
   ],
   "source": [
    "out_path_colors = os.path.join(augur_dir, \"colors.tsv\")\n",
    "attributes = [\"country\", \"province\", \"branch_major\", \"branch_minor\",\"continent\"]\n",
    "colors_dict = {}\n",
    "\n",
    "for attr in attributes:\n",
    "    # Create the color map\n",
    "    colors_dict[attr] = {}\n",
    "    \n",
    "    # Set branch_major based on list\n",
    "    if attr == \"branch_major\":\n",
    "        states = list(BRANCH_LIST.keys())\n",
    "        states.reverse()\n",
    "    else:\n",
    "        states = set(metadata_df[attr])\n",
    "        \n",
    "    # Create the custom color map (pyplot)\n",
    "    cmap = plt.get_cmap(\"rainbow\", len(states))\n",
    "    # Convert the color map to a list of RGB values\n",
    "    cmaplist = [cmap(i) for i in range(cmap.N)]\n",
    "    # Convert RGB values to hex colors\n",
    "    attr_hex = [colors.to_hex(col) for col in cmaplist]\n",
    "\n",
    "    # Assign colors to value\n",
    "    for state, color in zip(states, attr_hex):\n",
    "        colors_dict[attr][state] = color\n",
    "        \n",
    "    # Add unknown\n",
    "    colors_dict[attr][NO_DATA_CHAR] = \"#969696\"\n",
    "    \n",
    "print(colors_dict)\n",
    "\n",
    "with open(out_path_colors, \"w\") as outfile:\n",
    "    for attr_key in colors_dict:\n",
    "        for attr_val in colors_dict[attr_key]:\n",
    "            outfile.write(str(attr_key) + \"\\t\" + str(attr_val) + \"\\t\" + str(colors_dict[attr_key][attr_val]) + \"\\n\")        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "american-swing",
   "metadata": {},
   "source": [
    "## Latitude and Longitude"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bizarre-challenge",
   "metadata": {},
   "outputs": [],
   "source": [
    "latlon_country_df = pd.DataFrame()\n",
    "latlon_province_df = pd.DataFrame()\n",
    "\n",
    "df_list = [latlon_country_df, latlon_province_df]\n",
    "attr_list = [\"country\", \"province\"]\n",
    "\n",
    "# Create a mapping of geo name to lat,lon\n",
    "for df,attr in zip(df_list, attr_list):\n",
    "    # Get data\n",
    "    for rec in metadata_df.iterrows():\n",
    "        node_name = rec[0]\n",
    "        name = rec[1][attr]\n",
    "        country = rec[1][\"country\"]\n",
    "        \n",
    "        #node_type = rec[1][\"node_type\"]\n",
    "        #if node_type == \"internal\":\n",
    "        #    continue\n",
    "\n",
    "        #if attr == \"province\" and name == NO_DATA_CHAR and node_type == \"terminal\" and country != \"Russia\":\n",
    "        if attr == \"province\" and name == NO_DATA_CHAR and country != \"Russia\":\n",
    "            # Use country instead\n",
    "            name = rec[1][\"country\"]\n",
    "            lat = rec[1][\"country_lat\"]\n",
    "            lon = rec[1][\"country_lon\"]\n",
    "        else:\n",
    "            lat = rec[1][attr + \"_lat\"]\n",
    "            lon = rec[1][attr + \"_lon\"]             \n",
    "\n",
    "\n",
    "        if name not in df.index:\n",
    "            df.at[name, \"lat\"] = lat\n",
    "            df.at[name, \"lon\"] = lon\n",
    "            df.at[name, \"size\"] = 1\n",
    "        else:\n",
    "            df[\"size\"][name] += 1\n",
    "        \n",
    "\n",
    "#display(latlon_country_df)\n",
    "#display(latlon_province_df)\n",
    "\n",
    "# Mapping file for auspice\n",
    "out_path_latlon = os.path.join(augur_dir, \"latlon.tsv\")\n",
    "\n",
    "# Countries\n",
    "with open(out_path_latlon, \"w\") as outfile:\n",
    "    for country in latlon_country_df.index:\n",
    "        lat = str(latlon_country_df[\"lat\"][country])\n",
    "        lon = str(latlon_country_df[\"lon\"][country])\n",
    "        outfile.write(\"country\" + \"\\t\" + country + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "# Provinces\n",
    "with open(out_path_latlon, \"a\") as outfile:\n",
    "    for province in latlon_province_df.index:\n",
    "        lat = str(latlon_province_df[\"lat\"][province])\n",
    "        lon = str(latlon_province_df[\"lon\"][province])\n",
    "        outfile.write(\"province\" + \"\\t\" + province + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "        \n",
    "#display(metadata_df[metadata_df[\"continent\"] == \"Europe\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "super-permit",
   "metadata": {},
   "source": [
    "## Continent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "framed-extension",
   "metadata": {},
   "outputs": [],
   "source": [
    "continent_dict = {}\n",
    "\n",
    "for rec in metadata_df.iterrows():\n",
    "    continent = rec[1][\"continent\"]\n",
    "    country = rec[1][\"country\"]\n",
    "    continent_dict[country] = continent"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "economic-raising",
   "metadata": {},
   "source": [
    "## Tree Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "small-message",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Branch: 1.ORI\n",
      "Rename: GCA_000324805.2_EV76_CN_genomic GCA_000324805.2_EV76-CN_genomic\n",
      "Rename: GCA_000986995.1_YPES001_SEQ_2_ASM_1_genomic GCA_000986995.1_YPES001-SEQ-2-ASM-1_genomic\n",
      "Branch: 1.PRE\n",
      "Branch: 2.MED\n",
      "Rename: GCA_001617735.1_Yersinia_pestis_M_1763_genomic GCA_001617735.1_Yersinia_pestis_M-1763_genomic\n",
      "Rename: GCA_002412305.1_Y.pestis_A_1809_genomic GCA_002412305.1_Y.pestis_A-1809_genomic\n",
      "Rename: GCA_001617785.1_Yersinia_pestis_M_549_genomic GCA_001617785.1_Yersinia_pestis_M-549_genomic\n",
      "Rename: GCA_001617725.1_Yersinia_pestis_M_1484_genomic GCA_001617725.1_Yersinia_pestis_M-1484_genomic\n",
      "Rename: GCA_001617715.1_Yersinia_pestis_M_519_genomic GCA_001617715.1_Yersinia_pestis_M-519_genomic\n",
      "Rename: GCA_001617815.1_Yersinia_pestis_M_1453_genomic GCA_001617815.1_Yersinia_pestis_M-1453_genomic\n",
      "Rename: GCA_001617705.1_Yersinia_pestis_C_791_genomic GCA_001617705.1_Yersinia_pestis_C-791_genomic\n",
      "Branch: 0.ANT4\n",
      "Branch: 0.PRE\n"
     ]
    }
   ],
   "source": [
    "# Construct a dictionary to hold the trees\n",
    "tree_dict = {}\n",
    "\n",
    "for branch in BRANCH_LIST:\n",
    "    for filename in os.listdir(tree_dir):\n",
    "        if not filename.endswith(\".tree\"): continue  \n",
    "        filepath = os.path.join(tree_dir, filename)\n",
    "        \n",
    "        # Check file matches branch\n",
    "        if filename.strip(\".tree\") == branch:  \n",
    "        #if filename.split(\"_\")[0] == branch:  \n",
    "        \n",
    "            print(\"Branch:\", branch)\n",
    "            tree_dict[branch] = {}  \n",
    "            \n",
    "            # Add tree files to dict\n",
    "            tree_dict[branch][\"tree_file_raw\"] = filepath \n",
    "            tree_dict[branch][\"tree_file_edit\"] = os.path.join(tree_dir, branch + \".nex\")\n",
    "            tree_dict[branch][\"sample_rename\"] = {}\n",
    "            \n",
    "            # Read in raw tree to deal with dashes\n",
    "            with open(tree_dict[branch][\"tree_file_raw\"],  \"r\") as infile:                    \n",
    "                with open(tree_dict[branch][\"tree_file_edit\"], \"w\") as outfile:           \n",
    "                    raw_tree = infile.read()\n",
    "                    # Remove quotations if they exist\n",
    "                    raw_tree = raw_tree.replace(\"'\",\"\")\n",
    "                    \n",
    "                    # Split into lines to iterate over\n",
    "                    raw_tree_lines = raw_tree.split(\"\\n\")\n",
    "                    \n",
    "                    # By default don't parse a line for dashes\n",
    "                    taxa_line = False\n",
    "                    \n",
    "                    for line in raw_tree_lines: \n",
    "                                \n",
    "                        if \"TREE\" not in line and \"-\" in line:\n",
    "                            \n",
    "                            if len(line.split(\" \")) == 1:\n",
    "                                name_dashes = line.strip()\n",
    "                                name_no_dashes = name_dashes.replace(\"-\",\"_\")\n",
    "                                tree_dict[branch][\"sample_rename\"][name_no_dashes] = name_dashes\n",
    "                            line = line.replace(\"-\",\"_\")\n",
    "                                \n",
    "                        outfile.write(line + \"\\n\")\n",
    "            \n",
    "            # Read in edited tree\n",
    "            trees = Phylo.parse(tree_dict[branch][\"tree_file_edit\"], \"nexus\")\n",
    "            # There should be only 1 tree\n",
    "            for t in trees:\n",
    "                tree_dict[branch][\"tree\"] = t\n",
    "                tree_dict[branch][\"tree\"].ladderize(reverse=False)\n",
    "                break\n",
    "\n",
    "            # Rename sample names back to with dashes\n",
    "            for c in tree_dict[branch][\"tree\"].find_clades():\n",
    "                if c.name in tree_dict[branch][\"sample_rename\"]:\n",
    "                    orig_name = c.name\n",
    "                    c.name = tree_dict[branch][\"sample_rename\"][c.name]\n",
    "                    print(\"Rename:\", orig_name, c.name)\n",
    "                    \n",
    "                # Strip the date suffix\n",
    "                #if c.name:\n",
    "                #    c.name = \"_\".join(c.name.split(\"_\")[0:-1])           \n",
    "                    \n",
    "            # Rename internal nodes\n",
    "            node_i = 0\n",
    "            for c in tree_dict[branch][\"tree\"].find_clades():\n",
    "                if not c.name:\n",
    "                    c.name = \"NODE{}\".format(node_i)\n",
    "                    node_i += 1\n",
    "            \n",
    "            #Phylo.draw(tree_dict[branch][\"tree\"])\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "italic-huntington",
   "metadata": {},
   "source": [
    "## Add Tree Data to Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "identical-circumstances",
   "metadata": {},
   "source": [
    "### Get comments from the first root that is not root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "formal-questionnaire",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n",
      "1.PRE\n",
      "2.MED\n",
      "0.ANT4\n",
      "0.PRE\n"
     ]
    }
   ],
   "source": [
    "TREE_PARAMETERS = None\n",
    "\n",
    "for branch in tree_dict:\n",
    "    \n",
    "    print(branch)\n",
    "    \n",
    "    tree = tree_dict[branch][\"tree\"]\n",
    "    df = copy.deepcopy(metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])])\n",
    "    \n",
    "    # Check if a sample was missed :(\n",
    "    df_samples = df.index\n",
    "    tree_samples = [c.name for c in tree.find_clades()]\n",
    "    \n",
    "    for sample in df_samples:\n",
    "        if sample not in tree_samples:\n",
    "            df.drop(sample, inplace=True)\n",
    "            print(\"Dropping {} from the dataframe.\".format(sample))\n",
    "    \n",
    "    root_comment_dict = parse_comment(tree.root.comment)\n",
    "    \n",
    "    for c in tree.find_clades():\n",
    "        if c.is_terminal():\n",
    "            continue\n",
    "            \n",
    "        if c != tree.root:\n",
    "            comment_dict = parse_comment(c.comment)\n",
    "            for parameter,value in comment_dict.items():\n",
    "                # Intialize parameter values\n",
    "                if \"range\" in parameter or \"95%\" in parameter:\n",
    "                    df[parameter] = [[NO_DATA_CHAR,NO_DATA_CHAR]] * len(df)\n",
    "                    if parameter not in root_comment_dict:\n",
    "                        root_comment_dict[parameter] = '{0,0}'\n",
    "                else:\n",
    "                    df[parameter] = [NO_DATA_CHAR] * len(df)\n",
    "                    if parameter not in root_comment_dict:\n",
    "                        root_comment_dict[parameter] = '0'\n",
    "                    \n",
    "            break\n",
    "\n",
    "    # Update the roots comment to include missing values\n",
    "    new_comment = \"[&\"\n",
    "    new_comment_list = []\n",
    "    for parameter,value in root_comment_dict.items():\n",
    "        param_str = \"{}={}\".format(parameter, value)\n",
    "        new_comment_list.append(param_str)\n",
    "    new_comment += \",\".join(new_comment_list) + \"]\"\n",
    "    tree.root.comment = new_comment\n",
    "    #print(root_comment_dict)\n",
    "    #print(new_comment_list)\n",
    "    \n",
    "    if not TREE_PARAMETERS:\n",
    "        TREE_PARAMETERS = [p for p in root_comment_dict]\n",
    "    \n",
    "    tree_dict[branch][\"df\"] = df \n",
    "    #display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "labeled-apollo",
   "metadata": {},
   "source": [
    "### Parse Tree Comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "optimum-identifier",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n",
      "1.PRE\n",
      "2.MED\n",
      "0.ANT4\n",
      "0.PRE\n"
     ]
    }
   ],
   "source": [
    "parameters = [\n",
    "    \"branch_length\",\n",
    "    \"branch_length_sub\",\n",
    "    \"node_type\",    \n",
    "    \"branch_support\",\n",
    "    \"branch_support_conf_category\",\n",
    "    \"branch_support_conf_char\",\n",
    "    \"country_date_strain\",\n",
    "    \"province_date_strain\",\n",
    "    \"rate\",\n",
    "    \"rate_hpd\",\n",
    "    \"rate_sub\",\n",
    "    \"rate_sub_year\",\n",
    "    \"state\",\n",
    "    \"state_prob\",\n",
    "    \"state_conf_category\",\n",
    "    \"state_rate\",\n",
    "    \"state_rate_hpd\",\n",
    "    \"state_continent\",\n",
    "    \"state_lat\",\n",
    "    \"state_lon\",\n",
    "    \"height\",\n",
    "    \"height_hpd\",\n",
    "    \"timetree_num_date\",\n",
    "    \"timetree_num_date_confidence\",\n",
    "]\n",
    "\n",
    "for branch in tree_dict:\n",
    "    print(branch)\n",
    "    \n",
    "    tree = tree_dict[branch][\"tree\"]\n",
    "    df = tree_dict[branch][\"df\"]\n",
    "    \n",
    "    # Add to dataframe\n",
    "    for param in parameters:\n",
    "        df[param] = [NO_DATA_CHAR] * len(df)\n",
    "        \n",
    "    for c in tree.find_clades():\n",
    "        # Defaults\n",
    "        node_type = \"internal\"\n",
    "        branch_support = 0\n",
    "        branch_support_conf_category = \"LOW\"\n",
    "        branch_support_conf_char = \"\"\n",
    "        branch_length = 0\n",
    "        branch_length_sub = 0\n",
    "        country_date_strain = NO_DATA_CHAR\n",
    "        province_date_strain = NO_DATA_CHAR\n",
    "        rate_hpd = [0,0]\n",
    "        rate = 0\n",
    "        rate_sub_year = 0\n",
    "        state = NO_DATA_CHAR\n",
    "        state_conf_category = \"LOW\"\n",
    "        state_prob = 0\n",
    "        state_rate = 0\n",
    "        state_continent = NO_DATA_CHAR\n",
    "        state_lat = NO_DATA_CHAR\n",
    "        state_lon = NO_DATA_CHAR\n",
    "        \n",
    "        \n",
    "        timetree_num_date = 0\n",
    "        timetree_num_date_confidence = [0,0]\n",
    "        height = 0\n",
    "        height_hpd = [0,0]\n",
    "        \n",
    "        \n",
    "        comment_dict = parse_comment(c.comment)\n",
    "        #print(comment_dict.keys())\n",
    "\n",
    "        # Branch Length\n",
    "        if c.branch_length:\n",
    "            branch_length = c.branch_length\n",
    "\n",
    "        # Branch support\n",
    "        if \"posterior\" in comment_dict:\n",
    "            branch_support = float(round(float(comment_dict[\"posterior\"]) * 100))\n",
    "            if branch_support >= 95:\n",
    "                branch_support_conf_category = \"HIGH\"\n",
    "                branch_support_conf_char = \"*\"\n",
    "\n",
    "        # Rates\n",
    "        if \"default.rate\" in comment_dict:\n",
    "            rate = float(comment_dict[\"default.rate\"])\n",
    "            rate_sub = rate\n",
    "\n",
    "            if c.branch_length:\n",
    "                rate_sub_year = rate * SEQ_LEN\n",
    "                branch_length_sub = rate_sub_year * c.branch_length\n",
    "\n",
    "        if \"default.rate_95%_HPD\" in comment_dict:\n",
    "            rate_hpd_split = comment_dict[\"default.rate_95%_HPD\"].strip(\"{}\").split(\",\")\n",
    "            rate_hpd = [float(r) for r in rate_hpd_split]\n",
    "            \n",
    "        if \"state.rate\" in comment_dict:\n",
    "            state_rate = float(comment_dict[\"state.rate\"])   \n",
    "\n",
    "        if \"state.rate_95%_HPD\" in comment_dict:\n",
    "            state_rate_hpd_split = comment_dict[\"state.rate_95%_HPD\"].strip(\"{}\").split(\",\")\n",
    "            state_rate_hpd = [float(r) for r in state_rate_hpd_split]            \n",
    "    \n",
    "        # States\n",
    "        if \"state\" in comment_dict:\n",
    "            state = comment_dict[\"state\"]\n",
    "            state_continent = continent_dict[state]\n",
    "            state_latlon_dict = dict(latlon_country_df.loc[state])\n",
    "            state_lat = float(state_latlon_dict[\"lat\"])\n",
    "            state_lon = float(state_latlon_dict[\"lon\"])\n",
    "            \n",
    "            \n",
    "        if \"state.prob\" in comment_dict:\n",
    "            state_prob = float(round(float(comment_dict[\"state.prob\"]) * 100))\n",
    "            if state_prob >= 95:\n",
    "                state_conf_category = \"HIGH\"\n",
    "\n",
    "\n",
    "        # Dates\n",
    "        \n",
    "        if \"height\" in comment_dict:\n",
    "            height = float(comment_dict[\"height\"])\n",
    "\n",
    "        if \"height_95%_HPD\" in comment_dict:\n",
    "            height_hpd_split = comment_dict[\"height_95%_HPD\"].strip(\"{}\").split(\",\")\n",
    "            height_hpd = [float(h) for h in height_hpd_split]\n",
    "        \n",
    "        height_hpd_reverse = copy.copy(height_hpd)\n",
    "        height_hpd_reverse.reverse()\n",
    "        timetree_num_date = float(round(mrsd_dict[branch] - height))\n",
    "        timetree_num_date_confidence = [round(mrsd_dict[branch] - h) for h in height_hpd_reverse]  \n",
    "\n",
    "        if c.is_terminal():\n",
    "            node_type = \"terminal\"\n",
    "            country = metadata_df[\"country\"][c.name]\n",
    "            province = metadata_df[\"province\"][c.name]\n",
    "            date = metadata_df[\"date\"][c.name]\n",
    "            strain = metadata_df[\"strain\"][c.name]\n",
    "\n",
    "            country_date_strain = \"{} {} {}\".format(country, date, strain)\n",
    "            province_date_strain = \"{} {} {}\".format(province, date, strain)\n",
    "            \n",
    "        \n",
    "        df.at[c.name, \"branch_length\"] = branch_length\n",
    "        df.at[c.name, \"branch_length_sub\"] = branch_length_sub\n",
    "        df.at[c.name, \"node_type\"] = node_type\n",
    "        df.at[c.name, \"branch_support\"] = branch_support\n",
    "        df.at[c.name, \"branch_support_conf_category\"] = branch_support_conf_category\n",
    "        df.at[c.name, \"branch_support_conf_char\"] = branch_support_conf_char\n",
    "        df.at[c.name, \"country_date_strain\"] = country_date_strain\n",
    "        df.at[c.name, \"province_date_strain\"] = province_date_strain\n",
    "\n",
    "        \n",
    "        df.at[c.name, \"rate\"] = rate\n",
    "        df.at[c.name, \"rate_hpd\"] = rate_hpd\n",
    "        df.at[c.name, \"rate_sub\"] = rate_sub\n",
    "        df.at[c.name, \"rate_sub_year\"] = rate_sub_year\n",
    "        \n",
    "        df.at[c.name, \"state\"] = state\n",
    "        df.at[c.name, \"state_prob\"] = state_prob\n",
    "        df.at[c.name, \"state_conf_category\"] = state_conf_category\n",
    "        df.at[c.name, \"state_continent\"] = state_continent\n",
    "        df.at[c.name, \"state_lat\"] = state_lat\n",
    "        df.at[c.name, \"state_lon\"] = state_lon\n",
    "        \n",
    "        df.at[c.name, \"state_rate\"] = state_rate\n",
    "        df.at[c.name, \"state_rate_hpd\"] = state_rate_hpd\n",
    "        \n",
    "        df.at[c.name, \"height\"] = height\n",
    "        df.at[c.name, \"height_hpd\"] = height_hpd   \n",
    "        \n",
    "        df.at[c.name, \"timetree_num_date\"] = timetree_num_date\n",
    "        df.at[c.name, \"timetree_num_date_confidence\"] = timetree_num_date_confidence        \n",
    "\n",
    "    # Update internal nodes NA\n",
    "    df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "    #display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "surface-confidentiality",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "branch\thigh\tinternal\t% high\n",
      "1.ORI\t102\t116\t88.0\n",
      "1.PRE\t17\t39\t44.0\n",
      "2.MED\t80\t115\t70.0\n",
      "0.ANT4\t8\t11\t73.0\n",
      "0.PRE\t3\t7\t43.0\n"
     ]
    }
   ],
   "source": [
    "## State Confidence Summary\n",
    "print(\"{}\\t{}\\t{}\\t{}\".format(\"branch\", \"high\", \"internal\", \"% high\" ))\n",
    "for branch in tree_dict:\n",
    "    \n",
    "    tree = tree_dict[branch][\"tree\"]\n",
    "    df = tree_dict[branch][\"df\"]\n",
    "    internal_nodes = 0\n",
    "    high_state_prob = 0\n",
    "    \n",
    "    for rec in df.iterrows():\n",
    "        sample = rec[0]\n",
    "        # Focus on internal branches\n",
    "        if rec[1][\"node_type\"] == \"terminal\": continue\n",
    "        state_prob = rec[1][\"state_prob\"]\n",
    "        internal_nodes += 1\n",
    "        if state_prob >= 95:\n",
    "            high_state_prob += 1\n",
    "        perc = round((high_state_prob / internal_nodes) * 100, 0)\n",
    "    print(\"{}\\t{}\\t{}\\t{}\".format(branch, high_state_prob, internal_nodes, perc))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "robust-tracker",
   "metadata": {},
   "source": [
    "## Reduced Dataframe for Auspice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "brutal-somalia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['strain', 'date', 'date_bp', 'country', 'province', 'country_lat',\n",
      "       'country_lon', 'province_lat', 'province_lon', 'biovar', 'branch_major',\n",
      "       'branch_minor', 'biosample_accession', 'biosample_comment',\n",
      "       'branch_number', 'continent', 'date_mean', 'date_bp_mean', 'date_err',\n",
      "       'lat', 'lon', 'host_human', 'branch_major_color', 'geometry_size',\n",
      "       'geometry', 'root_rtt_dist', 'clade_rtt_dist', 'length_range',\n",
      "       'state.rate_range', 'default.rate_95%_HPD', 'length_95%_HPD',\n",
      "       'state.rate_95%_HPD', 'default.rate_range', 'state.rate',\n",
      "       'default.rate', 'length', 'posterior', 'state.prob', 'height_median',\n",
      "       'height_range', 'height_95%_HPD', 'state.rate_median',\n",
      "       'default.rate_median', 'length_median', 'state', 'state.set.prob',\n",
      "       'state.set', 'height', 'branch_length', 'branch_length_sub',\n",
      "       'node_type', 'branch_support', 'branch_support_conf_category',\n",
      "       'branch_support_conf_char', 'country_date_strain',\n",
      "       'province_date_strain', 'rate', 'rate_hpd', 'rate_sub', 'rate_sub_year',\n",
      "       'state_prob', 'state_conf_category', 'state_rate', 'state_rate_hpd',\n",
      "       'state_continent', 'state_lat', 'state_lon', 'height_hpd',\n",
      "       'timetree_num_date', 'timetree_num_date_confidence'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# Options\n",
    "print(tree_dict[branch][\"df\"].columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aquatic-bidding",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>node_type</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>province</th>\n",
       "      <th>country</th>\n",
       "      <th>state_prob</th>\n",
       "      <th>state_conf_category</th>\n",
       "      <th>state_rate</th>\n",
       "      <th>continent</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>strain</th>\n",
       "      <th>country_date_strain</th>\n",
       "      <th>province_date_strain</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>node_name</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>13.502915</td>\n",
       "      <td>22.937615</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>[1992, 1992]</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005133</td>\n",
       "      <td>North America</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>CO92</td>\n",
       "      <td>United States of America 1992 CO92</td>\n",
       "      <td>Colorado 1992 CO92</td>\n",
       "      <td>Human</td>\n",
       "      <td>1992</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>7.31686e-05</td>\n",
       "      <td>6.3815e-06</td>\n",
       "      <td>4.016733e-07</td>\n",
       "      <td>1.698716</td>\n",
       "      <td>Reference</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000834775.1_ASM83477v1_genomic</th>\n",
       "      <td>6.249831</td>\n",
       "      <td>8.900725</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1967.0</td>\n",
       "      <td>[1967, 1967]</td>\n",
       "      <td>Arizona</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005133</td>\n",
       "      <td>North America</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>34.3953</td>\n",
       "      <td>-111.763</td>\n",
       "      <td>SAMN03161124</td>\n",
       "      <td>Dodson</td>\n",
       "      <td>United States of America 1967.0 Dodson</td>\n",
       "      <td>Arizona 1967.0 Dodson</td>\n",
       "      <td>Human</td>\n",
       "      <td>1967</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>7.05773e-05</td>\n",
       "      <td>3.7902e-06</td>\n",
       "      <td>3.367514e-07</td>\n",
       "      <td>1.424154</td>\n",
       "      <td>GCA_000834775.1_ASM83477v1_genomic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000834335.1_ASM83433v1_genomic</th>\n",
       "      <td>4.664658</td>\n",
       "      <td>4.498545</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1954.0</td>\n",
       "      <td>[1954, 1954]</td>\n",
       "      <td>California</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005133</td>\n",
       "      <td>North America</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>36.7015</td>\n",
       "      <td>-118.756</td>\n",
       "      <td>SAMN03120838</td>\n",
       "      <td>Shasta</td>\n",
       "      <td>United States of America 1954.0 Shasta</td>\n",
       "      <td>California 1954.0 Shasta</td>\n",
       "      <td>Human</td>\n",
       "      <td>1954</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>6.94491e-05</td>\n",
       "      <td>2.662e-06</td>\n",
       "      <td>2.280366e-07</td>\n",
       "      <td>0.964389</td>\n",
       "      <td>GCA_000834335.1_ASM83433v1_genomic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000169635.1_ASM16963v1_genomic</th>\n",
       "      <td>23.222619</td>\n",
       "      <td>13.089660</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>[2005, 2005]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005133</td>\n",
       "      <td>Africa</td>\n",
       "      <td>-18.924960</td>\n",
       "      <td>46.441642</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>SAMN02404403</td>\n",
       "      <td>MG05-1020</td>\n",
       "      <td>Madagascar 2005.0 MG05-1020</td>\n",
       "      <td>NA 2005.0 MG05-1020</td>\n",
       "      <td>Human</td>\n",
       "      <td>2005</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>7.15927e-05</td>\n",
       "      <td>4.8056e-06</td>\n",
       "      <td>1.332814e-07</td>\n",
       "      <td>0.563660</td>\n",
       "      <td>GCA_000169635.1_ASM16963v1_genomic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000170275.1_ASM17027v1_genomic</th>\n",
       "      <td>13.009251</td>\n",
       "      <td>4.460811</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1991.0</td>\n",
       "      <td>[1991, 1991]</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>China</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005133</td>\n",
       "      <td>Asia</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>SAMN02404399</td>\n",
       "      <td>F1991016</td>\n",
       "      <td>China 1991.0 F1991016</td>\n",
       "      <td>Yunnan 1991.0 F1991016</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1991</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>7.25252e-05</td>\n",
       "      <td>5.7381e-06</td>\n",
       "      <td>8.108000e-08</td>\n",
       "      <td>0.342895</td>\n",
       "      <td>GCA_000170275.1_ASM17027v1_genomic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE111</th>\n",
       "      <td>0.418109</td>\n",
       "      <td>0.308126</td>\n",
       "      <td>internal</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>[2001, 2008]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Peru</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005103</td>\n",
       "      <td>South America</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>31.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.742575e-07</td>\n",
       "      <td>0.736952</td>\n",
       "      <td>NODE111</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE112</th>\n",
       "      <td>3.334069</td>\n",
       "      <td>1.793268</td>\n",
       "      <td>internal</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>[2006, 2010]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Peru</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005132</td>\n",
       "      <td>South America</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>96.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.271812e-07</td>\n",
       "      <td>0.537862</td>\n",
       "      <td>NODE112</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE113</th>\n",
       "      <td>1.351110</td>\n",
       "      <td>1.145470</td>\n",
       "      <td>internal</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>[2003, 2009]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Peru</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005107</td>\n",
       "      <td>South America</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>52.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.004681e-07</td>\n",
       "      <td>0.847799</td>\n",
       "      <td>NODE113</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE114</th>\n",
       "      <td>1.112510</td>\n",
       "      <td>0.576862</td>\n",
       "      <td>internal</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>[2005, 2009]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Peru</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005129</td>\n",
       "      <td>South America</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>23.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.226085e-07</td>\n",
       "      <td>0.518523</td>\n",
       "      <td>NODE114</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE115</th>\n",
       "      <td>1.645869</td>\n",
       "      <td>1.393134</td>\n",
       "      <td>internal</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>[2005, 2010]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Peru</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.005120</td>\n",
       "      <td>South America</td>\n",
       "      <td>-6.869970</td>\n",
       "      <td>-75.045851</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>64.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.001473e-07</td>\n",
       "      <td>0.846442</td>\n",
       "      <td>NODE115</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>233 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    branch_length  branch_length_sub  \\\n",
       "sample                                                                 \n",
       "Reference                               13.502915          22.937615   \n",
       "GCA_000834775.1_ASM83477v1_genomic       6.249831           8.900725   \n",
       "GCA_000834335.1_ASM83433v1_genomic       4.664658           4.498545   \n",
       "GCA_000169635.1_ASM16963v1_genomic      23.222619          13.089660   \n",
       "GCA_000170275.1_ASM17027v1_genomic      13.009251           4.460811   \n",
       "...                                           ...                ...   \n",
       "NODE111                                  0.418109           0.308126   \n",
       "NODE112                                  3.334069           1.793268   \n",
       "NODE113                                  1.351110           1.145470   \n",
       "NODE114                                  1.112510           0.576862   \n",
       "NODE115                                  1.645869           1.393134   \n",
       "\n",
       "                                   node_type  timetree_num_date  \\\n",
       "sample                                                            \n",
       "Reference                           terminal             1992.0   \n",
       "GCA_000834775.1_ASM83477v1_genomic  terminal             1967.0   \n",
       "GCA_000834335.1_ASM83433v1_genomic  terminal             1954.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic  terminal             2005.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic  terminal             1991.0   \n",
       "...                                      ...                ...   \n",
       "NODE111                             internal             2005.0   \n",
       "NODE112                             internal             2008.0   \n",
       "NODE113                             internal             2006.0   \n",
       "NODE114                             internal             2007.0   \n",
       "NODE115                             internal             2008.0   \n",
       "\n",
       "                                   timetree_num_date_confidence    province  \\\n",
       "sample                                                                        \n",
       "Reference                                          [1992, 1992]    Colorado   \n",
       "GCA_000834775.1_ASM83477v1_genomic                 [1967, 1967]     Arizona   \n",
       "GCA_000834335.1_ASM83433v1_genomic                 [1954, 1954]  California   \n",
       "GCA_000169635.1_ASM16963v1_genomic                 [2005, 2005]          NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic                 [1991, 1991]      Yunnan   \n",
       "...                                                         ...         ...   \n",
       "NODE111                                            [2001, 2008]          NA   \n",
       "NODE112                                            [2006, 2010]          NA   \n",
       "NODE113                                            [2003, 2009]          NA   \n",
       "NODE114                                            [2005, 2009]          NA   \n",
       "NODE115                                            [2005, 2010]          NA   \n",
       "\n",
       "                                                     country  state_prob  \\\n",
       "sample                                                                     \n",
       "Reference                           United States of America       100.0   \n",
       "GCA_000834775.1_ASM83477v1_genomic  United States of America       100.0   \n",
       "GCA_000834335.1_ASM83433v1_genomic  United States of America       100.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic                Madagascar       100.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic                     China       100.0   \n",
       "...                                                      ...         ...   \n",
       "NODE111                                                 Peru       100.0   \n",
       "NODE112                                                 Peru       100.0   \n",
       "NODE113                                                 Peru       100.0   \n",
       "NODE114                                                 Peru       100.0   \n",
       "NODE115                                                 Peru       100.0   \n",
       "\n",
       "                                   state_conf_category  state_rate  \\\n",
       "sample                                                               \n",
       "Reference                                         HIGH    0.005133   \n",
       "GCA_000834775.1_ASM83477v1_genomic                HIGH    0.005133   \n",
       "GCA_000834335.1_ASM83433v1_genomic                HIGH    0.005133   \n",
       "GCA_000169635.1_ASM16963v1_genomic                HIGH    0.005133   \n",
       "GCA_000170275.1_ASM17027v1_genomic                HIGH    0.005133   \n",
       "...                                                ...         ...   \n",
       "NODE111                                           HIGH    0.005103   \n",
       "NODE112                                           HIGH    0.005132   \n",
       "NODE113                                           HIGH    0.005107   \n",
       "NODE114                                           HIGH    0.005129   \n",
       "NODE115                                           HIGH    0.005120   \n",
       "\n",
       "                                        continent  country_lat  country_lon  \\\n",
       "sample                                                                        \n",
       "Reference                           North America    39.783730  -100.445882   \n",
       "GCA_000834775.1_ASM83477v1_genomic  North America    39.783730  -100.445882   \n",
       "GCA_000834335.1_ASM83433v1_genomic  North America    39.783730  -100.445882   \n",
       "GCA_000169635.1_ASM16963v1_genomic         Africa   -18.924960    46.441642   \n",
       "GCA_000170275.1_ASM17027v1_genomic           Asia    35.000074   104.999927   \n",
       "...                                           ...          ...          ...   \n",
       "NODE111                             South America    -6.869970   -75.045851   \n",
       "NODE112                             South America    -6.869970   -75.045851   \n",
       "NODE113                             South America    -6.869970   -75.045851   \n",
       "NODE114                             South America    -6.869970   -75.045851   \n",
       "NODE115                             South America    -6.869970   -75.045851   \n",
       "\n",
       "                                   branch_major branch_minor  branch_support  \\\n",
       "sample                                                                         \n",
       "Reference                                 1.ORI       1.ORI1             0.0   \n",
       "GCA_000834775.1_ASM83477v1_genomic        1.ORI       1.ORI1             0.0   \n",
       "GCA_000834335.1_ASM83433v1_genomic        1.ORI       1.ORI1             0.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic        1.ORI       1.ORI3             0.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic        1.ORI       1.ORI2             0.0   \n",
       "...                                         ...          ...             ...   \n",
       "NODE111                                      NA           NA            31.0   \n",
       "NODE112                                      NA           NA            96.0   \n",
       "NODE113                                      NA           NA            52.0   \n",
       "NODE114                                      NA           NA            23.0   \n",
       "NODE115                                      NA           NA            64.0   \n",
       "\n",
       "                                   branch_support_conf_category  \\\n",
       "sample                                                            \n",
       "Reference                                                   LOW   \n",
       "GCA_000834775.1_ASM83477v1_genomic                          LOW   \n",
       "GCA_000834335.1_ASM83433v1_genomic                          LOW   \n",
       "GCA_000169635.1_ASM16963v1_genomic                          LOW   \n",
       "GCA_000170275.1_ASM17027v1_genomic                          LOW   \n",
       "...                                                         ...   \n",
       "NODE111                                                     LOW   \n",
       "NODE112                                                    HIGH   \n",
       "NODE113                                                     LOW   \n",
       "NODE114                                                     LOW   \n",
       "NODE115                                                     LOW   \n",
       "\n",
       "                                   branch_support_conf_char province_lat  \\\n",
       "sample                                                                     \n",
       "Reference                                                        38.7252   \n",
       "GCA_000834775.1_ASM83477v1_genomic                               34.3953   \n",
       "GCA_000834335.1_ASM83433v1_genomic                               36.7015   \n",
       "GCA_000169635.1_ASM16963v1_genomic                                    NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic                                    25   \n",
       "...                                                     ...          ...   \n",
       "NODE111                                                               NA   \n",
       "NODE112                                                   *           NA   \n",
       "NODE113                                                               NA   \n",
       "NODE114                                                               NA   \n",
       "NODE115                                                               NA   \n",
       "\n",
       "                                   province_lon biosample_accession  \\\n",
       "sample                                                                \n",
       "Reference                              -105.608        SAMEA1705942   \n",
       "GCA_000834775.1_ASM83477v1_genomic     -111.763        SAMN03161124   \n",
       "GCA_000834335.1_ASM83433v1_genomic     -118.756        SAMN03120838   \n",
       "GCA_000169635.1_ASM16963v1_genomic           NA        SAMN02404403   \n",
       "GCA_000170275.1_ASM17027v1_genomic          102        SAMN02404399   \n",
       "...                                         ...                 ...   \n",
       "NODE111                                      NA                  NA   \n",
       "NODE112                                      NA                  NA   \n",
       "NODE113                                      NA                  NA   \n",
       "NODE114                                      NA                  NA   \n",
       "NODE115                                      NA                  NA   \n",
       "\n",
       "                                       strain  \\\n",
       "sample                                          \n",
       "Reference                                CO92   \n",
       "GCA_000834775.1_ASM83477v1_genomic     Dodson   \n",
       "GCA_000834335.1_ASM83433v1_genomic     Shasta   \n",
       "GCA_000169635.1_ASM16963v1_genomic  MG05-1020   \n",
       "GCA_000170275.1_ASM17027v1_genomic   F1991016   \n",
       "...                                       ...   \n",
       "NODE111                                    NA   \n",
       "NODE112                                    NA   \n",
       "NODE113                                    NA   \n",
       "NODE114                                    NA   \n",
       "NODE115                                    NA   \n",
       "\n",
       "                                                       country_date_strain  \\\n",
       "sample                                                                       \n",
       "Reference                               United States of America 1992 CO92   \n",
       "GCA_000834775.1_ASM83477v1_genomic  United States of America 1967.0 Dodson   \n",
       "GCA_000834335.1_ASM83433v1_genomic  United States of America 1954.0 Shasta   \n",
       "GCA_000169635.1_ASM16963v1_genomic             Madagascar 2005.0 MG05-1020   \n",
       "GCA_000170275.1_ASM17027v1_genomic                   China 1991.0 F1991016   \n",
       "...                                                                    ...   \n",
       "NODE111                                                                 NA   \n",
       "NODE112                                                                 NA   \n",
       "NODE113                                                                 NA   \n",
       "NODE114                                                                 NA   \n",
       "NODE115                                                                 NA   \n",
       "\n",
       "                                        province_date_strain host_human  \\\n",
       "sample                                                                    \n",
       "Reference                                 Colorado 1992 CO92      Human   \n",
       "GCA_000834775.1_ASM83477v1_genomic     Arizona 1967.0 Dodson      Human   \n",
       "GCA_000834335.1_ASM83433v1_genomic  California 1954.0 Shasta      Human   \n",
       "GCA_000169635.1_ASM16963v1_genomic       NA 2005.0 MG05-1020      Human   \n",
       "GCA_000170275.1_ASM17027v1_genomic    Yunnan 1991.0 F1991016  Non-Human   \n",
       "...                                                      ...        ...   \n",
       "NODE111                                                   NA         NA   \n",
       "NODE112                                                   NA         NA   \n",
       "NODE113                                                   NA         NA   \n",
       "NODE114                                                   NA         NA   \n",
       "NODE115                                                   NA         NA   \n",
       "\n",
       "                                   date_mean date_err date_bp_mean  \\\n",
       "sample                                                               \n",
       "Reference                               1992        0           29   \n",
       "GCA_000834775.1_ASM83477v1_genomic      1967        0           54   \n",
       "GCA_000834335.1_ASM83433v1_genomic      1954        0           67   \n",
       "GCA_000169635.1_ASM16963v1_genomic      2005        0           16   \n",
       "GCA_000170275.1_ASM17027v1_genomic      1991        0           30   \n",
       "...                                      ...      ...          ...   \n",
       "NODE111                                   NA       NA           NA   \n",
       "NODE112                                   NA       NA           NA   \n",
       "NODE113                                   NA       NA           NA   \n",
       "NODE114                                   NA       NA           NA   \n",
       "NODE115                                   NA       NA           NA   \n",
       "\n",
       "                                   root_rtt_dist clade_rtt_dist      rate_sub  \\\n",
       "sample                                                                          \n",
       "Reference                            7.31686e-05     6.3815e-06  4.016733e-07   \n",
       "GCA_000834775.1_ASM83477v1_genomic   7.05773e-05     3.7902e-06  3.367514e-07   \n",
       "GCA_000834335.1_ASM83433v1_genomic   6.94491e-05      2.662e-06  2.280366e-07   \n",
       "GCA_000169635.1_ASM16963v1_genomic   7.15927e-05     4.8056e-06  1.332814e-07   \n",
       "GCA_000170275.1_ASM17027v1_genomic   7.25252e-05     5.7381e-06  8.108000e-08   \n",
       "...                                          ...            ...           ...   \n",
       "NODE111                                       NA             NA  1.742575e-07   \n",
       "NODE112                                       NA             NA  1.271812e-07   \n",
       "NODE113                                       NA             NA  2.004681e-07   \n",
       "NODE114                                       NA             NA  1.226085e-07   \n",
       "NODE115                                       NA             NA  2.001473e-07   \n",
       "\n",
       "                                    rate_sub_year  \\\n",
       "sample                                              \n",
       "Reference                                1.698716   \n",
       "GCA_000834775.1_ASM83477v1_genomic       1.424154   \n",
       "GCA_000834335.1_ASM83433v1_genomic       0.964389   \n",
       "GCA_000169635.1_ASM16963v1_genomic       0.563660   \n",
       "GCA_000170275.1_ASM17027v1_genomic       0.342895   \n",
       "...                                           ...   \n",
       "NODE111                                  0.736952   \n",
       "NODE112                                  0.537862   \n",
       "NODE113                                  0.847799   \n",
       "NODE114                                  0.518523   \n",
       "NODE115                                  0.846442   \n",
       "\n",
       "                                                             node_name blank  \n",
       "sample                                                                        \n",
       "Reference                                                    Reference        \n",
       "GCA_000834775.1_ASM83477v1_genomic  GCA_000834775.1_ASM83477v1_genomic        \n",
       "GCA_000834335.1_ASM83433v1_genomic  GCA_000834335.1_ASM83433v1_genomic        \n",
       "GCA_000169635.1_ASM16963v1_genomic  GCA_000169635.1_ASM16963v1_genomic        \n",
       "GCA_000170275.1_ASM17027v1_genomic  GCA_000170275.1_ASM17027v1_genomic        \n",
       "...                                                                ...   ...  \n",
       "NODE111                                                        NODE111        \n",
       "NODE112                                                        NODE112        \n",
       "NODE113                                                        NODE113        \n",
       "NODE114                                                        NODE114        \n",
       "NODE115                                                        NODE115        \n",
       "\n",
       "[233 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.PRE\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>node_type</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>province</th>\n",
       "      <th>country</th>\n",
       "      <th>state_prob</th>\n",
       "      <th>state_conf_category</th>\n",
       "      <th>state_rate</th>\n",
       "      <th>continent</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>strain</th>\n",
       "      <th>country_date_strain</th>\n",
       "      <th>province_date_strain</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>node_name</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA5818830</th>\n",
       "      <td>1.704386</td>\n",
       "      <td>0.257695</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1560.0</td>\n",
       "      <td>[1560, 1560]</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007231</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>SAMEA5818830</td>\n",
       "      <td>STN021</td>\n",
       "      <td>Switzerland [1485:1635] STN021</td>\n",
       "      <td>Nidwalden [1485:1635] STN021</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>75</td>\n",
       "      <td>461</td>\n",
       "      <td>6.30679e-05</td>\n",
       "      <td>9.8052e-06</td>\n",
       "      <td>3.575122e-08</td>\n",
       "      <td>0.151195</td>\n",
       "      <td>SAMEA5818830</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818829</th>\n",
       "      <td>1.067785</td>\n",
       "      <td>0.152720</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1560.0</td>\n",
       "      <td>[1560, 1560]</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007231</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>SAMEA5818829</td>\n",
       "      <td>STN020</td>\n",
       "      <td>Switzerland [1485:1635] STN020</td>\n",
       "      <td>Nidwalden [1485:1635] STN020</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>75</td>\n",
       "      <td>461</td>\n",
       "      <td>6.27851e-05</td>\n",
       "      <td>9.5224e-06</td>\n",
       "      <td>3.381921e-08</td>\n",
       "      <td>0.143025</td>\n",
       "      <td>SAMEA5818829</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818828</th>\n",
       "      <td>1.046214</td>\n",
       "      <td>0.156027</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1560.0</td>\n",
       "      <td>[1560, 1560]</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007231</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>SAMEA5818828</td>\n",
       "      <td>STN019</td>\n",
       "      <td>Switzerland [1485:1635] STN019</td>\n",
       "      <td>Nidwalden [1485:1635] STN019</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>75</td>\n",
       "      <td>461</td>\n",
       "      <td>6.28323e-05</td>\n",
       "      <td>9.5696e-06</td>\n",
       "      <td>3.526404e-08</td>\n",
       "      <td>0.149135</td>\n",
       "      <td>SAMEA5818828</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818826</th>\n",
       "      <td>1.067785</td>\n",
       "      <td>0.161386</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1560.0</td>\n",
       "      <td>[1560, 1560]</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007231</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>SAMEA5818826</td>\n",
       "      <td>STN014</td>\n",
       "      <td>Switzerland [1485:1635] STN014</td>\n",
       "      <td>Nidwalden [1485:1635] STN014</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>75</td>\n",
       "      <td>461</td>\n",
       "      <td>6.2735e-05</td>\n",
       "      <td>9.4723e-06</td>\n",
       "      <td>3.573825e-08</td>\n",
       "      <td>0.151141</td>\n",
       "      <td>SAMEA5818826</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818825</th>\n",
       "      <td>1.704386</td>\n",
       "      <td>0.969636</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1560.0</td>\n",
       "      <td>[1560, 1560]</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007231</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>SAMEA5818825</td>\n",
       "      <td>STN013</td>\n",
       "      <td>Switzerland [1485:1635] STN013</td>\n",
       "      <td>Nidwalden [1485:1635] STN013</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>75</td>\n",
       "      <td>461</td>\n",
       "      <td>6.32799e-05</td>\n",
       "      <td>1.00172e-05</td>\n",
       "      <td>1.345219e-07</td>\n",
       "      <td>0.568906</td>\n",
       "      <td>SAMEA5818825</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE34</th>\n",
       "      <td>2.579463</td>\n",
       "      <td>0.544664</td>\n",
       "      <td>internal</td>\n",
       "      <td>1555.0</td>\n",
       "      <td>[1550, 1559]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007265</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>10.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.992887e-08</td>\n",
       "      <td>0.211154</td>\n",
       "      <td>NODE34</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE35</th>\n",
       "      <td>2.622074</td>\n",
       "      <td>1.066974</td>\n",
       "      <td>internal</td>\n",
       "      <td>1558.0</td>\n",
       "      <td>[1554, 1560]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007231</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.621906e-08</td>\n",
       "      <td>0.406920</td>\n",
       "      <td>NODE35</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE36</th>\n",
       "      <td>3.470437</td>\n",
       "      <td>0.488058</td>\n",
       "      <td>internal</td>\n",
       "      <td>1556.0</td>\n",
       "      <td>[1552, 1560]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007239</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.325365e-08</td>\n",
       "      <td>0.140633</td>\n",
       "      <td>NODE36</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE37</th>\n",
       "      <td>2.367701</td>\n",
       "      <td>0.346754</td>\n",
       "      <td>internal</td>\n",
       "      <td>1558.0</td>\n",
       "      <td>[1555, 1560]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007250</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>17.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.462956e-08</td>\n",
       "      <td>0.146452</td>\n",
       "      <td>NODE37</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE38</th>\n",
       "      <td>2.389273</td>\n",
       "      <td>0.334175</td>\n",
       "      <td>internal</td>\n",
       "      <td>1558.0</td>\n",
       "      <td>[1555, 1560]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.813331</td>\n",
       "      <td>8.444947</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>15.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.307203e-08</td>\n",
       "      <td>0.139865</td>\n",
       "      <td>NODE38</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              branch_length  branch_length_sub node_type  timetree_num_date  \\\n",
       "sample                                                                        \n",
       "SAMEA5818830       1.704386           0.257695  terminal             1560.0   \n",
       "SAMEA5818829       1.067785           0.152720  terminal             1560.0   \n",
       "SAMEA5818828       1.046214           0.156027  terminal             1560.0   \n",
       "SAMEA5818826       1.067785           0.161386  terminal             1560.0   \n",
       "SAMEA5818825       1.704386           0.969636  terminal             1560.0   \n",
       "...                     ...                ...       ...                ...   \n",
       "NODE34             2.579463           0.544664  internal             1555.0   \n",
       "NODE35             2.622074           1.066974  internal             1558.0   \n",
       "NODE36             3.470437           0.488058  internal             1556.0   \n",
       "NODE37             2.367701           0.346754  internal             1558.0   \n",
       "NODE38             2.389273           0.334175  internal             1558.0   \n",
       "\n",
       "             timetree_num_date_confidence   province      country  state_prob  \\\n",
       "sample                                                                          \n",
       "SAMEA5818830                 [1560, 1560]  Nidwalden  Switzerland       100.0   \n",
       "SAMEA5818829                 [1560, 1560]  Nidwalden  Switzerland       100.0   \n",
       "SAMEA5818828                 [1560, 1560]  Nidwalden  Switzerland       100.0   \n",
       "SAMEA5818826                 [1560, 1560]  Nidwalden  Switzerland       100.0   \n",
       "SAMEA5818825                 [1560, 1560]  Nidwalden  Switzerland       100.0   \n",
       "...                                   ...        ...          ...         ...   \n",
       "NODE34                       [1550, 1559]         NA  Switzerland       100.0   \n",
       "NODE35                       [1554, 1560]         NA  Switzerland       100.0   \n",
       "NODE36                       [1552, 1560]         NA  Switzerland       100.0   \n",
       "NODE37                       [1555, 1560]         NA  Switzerland       100.0   \n",
       "NODE38                       [1555, 1560]         NA  Switzerland       100.0   \n",
       "\n",
       "             state_conf_category  state_rate continent  country_lat  \\\n",
       "sample                                                                \n",
       "SAMEA5818830                HIGH    0.007231    Europe    46.813331   \n",
       "SAMEA5818829                HIGH    0.007231    Europe    46.813331   \n",
       "SAMEA5818828                HIGH    0.007231    Europe    46.813331   \n",
       "SAMEA5818826                HIGH    0.007231    Europe    46.813331   \n",
       "SAMEA5818825                HIGH    0.007231    Europe    46.813331   \n",
       "...                          ...         ...       ...          ...   \n",
       "NODE34                      HIGH    0.007265    Europe    46.813331   \n",
       "NODE35                      HIGH    0.007231    Europe    46.813331   \n",
       "NODE36                      HIGH    0.007239    Europe    46.813331   \n",
       "NODE37                      HIGH    0.007250    Europe    46.813331   \n",
       "NODE38                      HIGH    0.007225    Europe    46.813331   \n",
       "\n",
       "              country_lon branch_major branch_minor  branch_support  \\\n",
       "sample                                                                \n",
       "SAMEA5818830     8.444947        1.PRE       1.PRE1             0.0   \n",
       "SAMEA5818829     8.444947        1.PRE       1.PRE1             0.0   \n",
       "SAMEA5818828     8.444947        1.PRE       1.PRE1             0.0   \n",
       "SAMEA5818826     8.444947        1.PRE       1.PRE1             0.0   \n",
       "SAMEA5818825     8.444947        1.PRE       1.PRE1             0.0   \n",
       "...                   ...          ...          ...             ...   \n",
       "NODE34           8.444947           NA           NA            10.0   \n",
       "NODE35           8.444947           NA           NA           100.0   \n",
       "NODE36           8.444947           NA           NA             8.0   \n",
       "NODE37           8.444947           NA           NA            17.0   \n",
       "NODE38           8.444947           NA           NA            15.0   \n",
       "\n",
       "             branch_support_conf_category branch_support_conf_char  \\\n",
       "sample                                                               \n",
       "SAMEA5818830                          LOW                            \n",
       "SAMEA5818829                          LOW                            \n",
       "SAMEA5818828                          LOW                            \n",
       "SAMEA5818826                          LOW                            \n",
       "SAMEA5818825                          LOW                            \n",
       "...                                   ...                      ...   \n",
       "NODE34                                LOW                            \n",
       "NODE35                               HIGH                        *   \n",
       "NODE36                                LOW                            \n",
       "NODE37                                LOW                            \n",
       "NODE38                                LOW                            \n",
       "\n",
       "             province_lat province_lon biosample_accession  strain  \\\n",
       "sample                                                               \n",
       "SAMEA5818830      46.9428      8.41198        SAMEA5818830  STN021   \n",
       "SAMEA5818829      46.9428      8.41198        SAMEA5818829  STN020   \n",
       "SAMEA5818828      46.9428      8.41198        SAMEA5818828  STN019   \n",
       "SAMEA5818826      46.9428      8.41198        SAMEA5818826  STN014   \n",
       "SAMEA5818825      46.9428      8.41198        SAMEA5818825  STN013   \n",
       "...                   ...          ...                 ...     ...   \n",
       "NODE34                 NA           NA                  NA      NA   \n",
       "NODE35                 NA           NA                  NA      NA   \n",
       "NODE36                 NA           NA                  NA      NA   \n",
       "NODE37                 NA           NA                  NA      NA   \n",
       "NODE38                 NA           NA                  NA      NA   \n",
       "\n",
       "                         country_date_strain          province_date_strain  \\\n",
       "sample                                                                       \n",
       "SAMEA5818830  Switzerland [1485:1635] STN021  Nidwalden [1485:1635] STN021   \n",
       "SAMEA5818829  Switzerland [1485:1635] STN020  Nidwalden [1485:1635] STN020   \n",
       "SAMEA5818828  Switzerland [1485:1635] STN019  Nidwalden [1485:1635] STN019   \n",
       "SAMEA5818826  Switzerland [1485:1635] STN014  Nidwalden [1485:1635] STN014   \n",
       "SAMEA5818825  Switzerland [1485:1635] STN013  Nidwalden [1485:1635] STN013   \n",
       "...                                      ...                           ...   \n",
       "NODE34                                    NA                            NA   \n",
       "NODE35                                    NA                            NA   \n",
       "NODE36                                    NA                            NA   \n",
       "NODE37                                    NA                            NA   \n",
       "NODE38                                    NA                            NA   \n",
       "\n",
       "             host_human date_mean date_err date_bp_mean root_rtt_dist  \\\n",
       "sample                                                                  \n",
       "SAMEA5818830      Human      1560       75          461   6.30679e-05   \n",
       "SAMEA5818829      Human      1560       75          461   6.27851e-05   \n",
       "SAMEA5818828      Human      1560       75          461   6.28323e-05   \n",
       "SAMEA5818826      Human      1560       75          461    6.2735e-05   \n",
       "SAMEA5818825      Human      1560       75          461   6.32799e-05   \n",
       "...                 ...       ...      ...          ...           ...   \n",
       "NODE34               NA        NA       NA           NA            NA   \n",
       "NODE35               NA        NA       NA           NA            NA   \n",
       "NODE36               NA        NA       NA           NA            NA   \n",
       "NODE37               NA        NA       NA           NA            NA   \n",
       "NODE38               NA        NA       NA           NA            NA   \n",
       "\n",
       "             clade_rtt_dist      rate_sub  rate_sub_year     node_name blank  \n",
       "sample                                                                        \n",
       "SAMEA5818830     9.8052e-06  3.575122e-08       0.151195  SAMEA5818830        \n",
       "SAMEA5818829     9.5224e-06  3.381921e-08       0.143025  SAMEA5818829        \n",
       "SAMEA5818828     9.5696e-06  3.526404e-08       0.149135  SAMEA5818828        \n",
       "SAMEA5818826     9.4723e-06  3.573825e-08       0.151141  SAMEA5818826        \n",
       "SAMEA5818825    1.00172e-05  1.345219e-07       0.568906  SAMEA5818825        \n",
       "...                     ...           ...            ...           ...   ...  \n",
       "NODE34                   NA  4.992887e-08       0.211154        NODE34        \n",
       "NODE35                   NA  9.621906e-08       0.406920        NODE35        \n",
       "NODE36                   NA  3.325365e-08       0.140633        NODE36        \n",
       "NODE37                   NA  3.462956e-08       0.146452        NODE37        \n",
       "NODE38                   NA  3.307203e-08       0.139865        NODE38        \n",
       "\n",
       "[79 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.MED\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>node_type</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>province</th>\n",
       "      <th>country</th>\n",
       "      <th>state_prob</th>\n",
       "      <th>state_conf_category</th>\n",
       "      <th>state_rate</th>\n",
       "      <th>continent</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>strain</th>\n",
       "      <th>country_date_strain</th>\n",
       "      <th>province_date_strain</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>node_name</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>20.502046</td>\n",
       "      <td>11.345819</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>[1923, 1923]</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006089</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>9_10</td>\n",
       "      <td>Russia 1923.0 9_10</td>\n",
       "      <td>Rostov Oblast 1923.0 9_10</td>\n",
       "      <td>Human</td>\n",
       "      <td>1923</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>7.30501e-05</td>\n",
       "      <td>9.6582e-06</td>\n",
       "      <td>1.308552e-07</td>\n",
       "      <td>0.553399</td>\n",
       "      <td>GCA_009909635.1_ASM990963v1_genomic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009296005.1_ASM929600v1_genomic</th>\n",
       "      <td>7.976810</td>\n",
       "      <td>2.350051</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1953.0</td>\n",
       "      <td>[1953, 1953]</td>\n",
       "      <td>Chechnya</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006089</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>43.3976</td>\n",
       "      <td>45.6985</td>\n",
       "      <td>SAMN12991209</td>\n",
       "      <td>C-25</td>\n",
       "      <td>Russia 1953.0 C-25</td>\n",
       "      <td>Chechnya 1953.0 C-25</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1953</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>7.1472e-05</td>\n",
       "      <td>8.0801e-06</td>\n",
       "      <td>6.966271e-08</td>\n",
       "      <td>0.294610</td>\n",
       "      <td>GCA_009296005.1_ASM929600v1_genomic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630485.1_ASM863048v1_genomic</th>\n",
       "      <td>2.961652</td>\n",
       "      <td>9.647097</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1997.0</td>\n",
       "      <td>[1997, 1997]</td>\n",
       "      <td>Kabardino-Balkaria</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006089</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>43.4428</td>\n",
       "      <td>43.4205</td>\n",
       "      <td>SAMN12721152</td>\n",
       "      <td>C-742</td>\n",
       "      <td>Russia 1997.0 C-742</td>\n",
       "      <td>Kabardino-Balkaria 1997.0 C-742</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1997</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>7.53973e-05</td>\n",
       "      <td>1.20054e-05</td>\n",
       "      <td>7.702200e-07</td>\n",
       "      <td>3.257336</td>\n",
       "      <td>GCA_008630485.1_ASM863048v1_genomic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630435.1_ASM863043v1_genomic</th>\n",
       "      <td>19.492362</td>\n",
       "      <td>62.164798</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>[1996, 1996]</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006089</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td>SAMN12721146</td>\n",
       "      <td>C-719</td>\n",
       "      <td>Russia 1996.0 C-719</td>\n",
       "      <td>Karachay-Cherkessia 1996.0 C-719</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1996</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>8.77894e-05</td>\n",
       "      <td>2.43975e-05</td>\n",
       "      <td>7.541059e-07</td>\n",
       "      <td>3.189188</td>\n",
       "      <td>GCA_008630435.1_ASM863043v1_genomic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630395.1_ASM863039v1_genomic</th>\n",
       "      <td>7.962416</td>\n",
       "      <td>11.179134</td>\n",
       "      <td>terminal</td>\n",
       "      <td>1984.0</td>\n",
       "      <td>[1984, 1984]</td>\n",
       "      <td>Republic of Dagestan</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006089</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>43.0883</td>\n",
       "      <td>47.1499</td>\n",
       "      <td>SAMN12715009</td>\n",
       "      <td>C-528</td>\n",
       "      <td>Russia 1984.0 C-528</td>\n",
       "      <td>Republic of Dagestan 1984.0 C-528</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1984</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>7.4455e-05</td>\n",
       "      <td>1.10631e-05</td>\n",
       "      <td>3.319827e-07</td>\n",
       "      <td>1.403988</td>\n",
       "      <td>GCA_008630395.1_ASM863039v1_genomic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE110</th>\n",
       "      <td>8.165731</td>\n",
       "      <td>3.840846</td>\n",
       "      <td>internal</td>\n",
       "      <td>1989.0</td>\n",
       "      <td>[1982, 1996]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006089</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.112203e-07</td>\n",
       "      <td>0.470362</td>\n",
       "      <td>NODE110</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE111</th>\n",
       "      <td>4.595163</td>\n",
       "      <td>2.896447</td>\n",
       "      <td>internal</td>\n",
       "      <td>1993.0</td>\n",
       "      <td>[1989, 1997]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006090</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.490448e-07</td>\n",
       "      <td>0.630325</td>\n",
       "      <td>NODE111</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE112</th>\n",
       "      <td>9.151472</td>\n",
       "      <td>1.017458</td>\n",
       "      <td>internal</td>\n",
       "      <td>1998.0</td>\n",
       "      <td>[1989, 2008]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006108</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>73.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.628923e-08</td>\n",
       "      <td>0.111180</td>\n",
       "      <td>NODE112</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE113</th>\n",
       "      <td>7.560997</td>\n",
       "      <td>2.498844</td>\n",
       "      <td>internal</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>[1997, 2009]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006089</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.814700e-08</td>\n",
       "      <td>0.330491</td>\n",
       "      <td>NODE113</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE114</th>\n",
       "      <td>11.638865</td>\n",
       "      <td>3.247230</td>\n",
       "      <td>internal</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>[2000, 2014]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.006089</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.597123e-08</td>\n",
       "      <td>0.278999</td>\n",
       "      <td>NODE114</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>231 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     branch_length  branch_length_sub  \\\n",
       "sample                                                                  \n",
       "GCA_009909635.1_ASM990963v1_genomic      20.502046          11.345819   \n",
       "GCA_009296005.1_ASM929600v1_genomic       7.976810           2.350051   \n",
       "GCA_008630485.1_ASM863048v1_genomic       2.961652           9.647097   \n",
       "GCA_008630435.1_ASM863043v1_genomic      19.492362          62.164798   \n",
       "GCA_008630395.1_ASM863039v1_genomic       7.962416          11.179134   \n",
       "...                                            ...                ...   \n",
       "NODE110                                   8.165731           3.840846   \n",
       "NODE111                                   4.595163           2.896447   \n",
       "NODE112                                   9.151472           1.017458   \n",
       "NODE113                                   7.560997           2.498844   \n",
       "NODE114                                  11.638865           3.247230   \n",
       "\n",
       "                                    node_type  timetree_num_date  \\\n",
       "sample                                                             \n",
       "GCA_009909635.1_ASM990963v1_genomic  terminal             1923.0   \n",
       "GCA_009296005.1_ASM929600v1_genomic  terminal             1953.0   \n",
       "GCA_008630485.1_ASM863048v1_genomic  terminal             1997.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic  terminal             1996.0   \n",
       "GCA_008630395.1_ASM863039v1_genomic  terminal             1984.0   \n",
       "...                                       ...                ...   \n",
       "NODE110                              internal             1989.0   \n",
       "NODE111                              internal             1993.0   \n",
       "NODE112                              internal             1998.0   \n",
       "NODE113                              internal             2005.0   \n",
       "NODE114                              internal             2009.0   \n",
       "\n",
       "                                    timetree_num_date_confidence  \\\n",
       "sample                                                             \n",
       "GCA_009909635.1_ASM990963v1_genomic                 [1923, 1923]   \n",
       "GCA_009296005.1_ASM929600v1_genomic                 [1953, 1953]   \n",
       "GCA_008630485.1_ASM863048v1_genomic                 [1997, 1997]   \n",
       "GCA_008630435.1_ASM863043v1_genomic                 [1996, 1996]   \n",
       "GCA_008630395.1_ASM863039v1_genomic                 [1984, 1984]   \n",
       "...                                                          ...   \n",
       "NODE110                                             [1982, 1996]   \n",
       "NODE111                                             [1989, 1997]   \n",
       "NODE112                                             [1989, 2008]   \n",
       "NODE113                                             [1997, 2009]   \n",
       "NODE114                                             [2000, 2014]   \n",
       "\n",
       "                                                 province country  state_prob  \\\n",
       "sample                                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic         Rostov Oblast  Russia       100.0   \n",
       "GCA_009296005.1_ASM929600v1_genomic              Chechnya  Russia       100.0   \n",
       "GCA_008630485.1_ASM863048v1_genomic    Kabardino-Balkaria  Russia       100.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic   Karachay-Cherkessia  Russia       100.0   \n",
       "GCA_008630395.1_ASM863039v1_genomic  Republic of Dagestan  Russia       100.0   \n",
       "...                                                   ...     ...         ...   \n",
       "NODE110                                                NA  Russia       100.0   \n",
       "NODE111                                                NA  Russia       100.0   \n",
       "NODE112                                                NA  Russia       100.0   \n",
       "NODE113                                                NA  Russia       100.0   \n",
       "NODE114                                                NA  Russia       100.0   \n",
       "\n",
       "                                    state_conf_category  state_rate continent  \\\n",
       "sample                                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic                HIGH    0.006089    Europe   \n",
       "GCA_009296005.1_ASM929600v1_genomic                HIGH    0.006089    Europe   \n",
       "GCA_008630485.1_ASM863048v1_genomic                HIGH    0.006089    Europe   \n",
       "GCA_008630435.1_ASM863043v1_genomic                HIGH    0.006089    Europe   \n",
       "GCA_008630395.1_ASM863039v1_genomic                HIGH    0.006089    Europe   \n",
       "...                                                 ...         ...       ...   \n",
       "NODE110                                            HIGH    0.006089    Europe   \n",
       "NODE111                                            HIGH    0.006090    Europe   \n",
       "NODE112                                            HIGH    0.006108    Europe   \n",
       "NODE113                                            HIGH    0.006089    Europe   \n",
       "NODE114                                            HIGH    0.006089    Europe   \n",
       "\n",
       "                                     country_lat  country_lon branch_major  \\\n",
       "sample                                                                       \n",
       "GCA_009909635.1_ASM990963v1_genomic    64.686314    97.745306        2.MED   \n",
       "GCA_009296005.1_ASM929600v1_genomic    64.686314    97.745306        2.MED   \n",
       "GCA_008630485.1_ASM863048v1_genomic    64.686314    97.745306        2.MED   \n",
       "GCA_008630435.1_ASM863043v1_genomic    64.686314    97.745306        2.MED   \n",
       "GCA_008630395.1_ASM863039v1_genomic    64.686314    97.745306        2.MED   \n",
       "...                                          ...          ...          ...   \n",
       "NODE110                                64.686314    97.745306           NA   \n",
       "NODE111                                64.686314    97.745306           NA   \n",
       "NODE112                                64.686314    97.745306           NA   \n",
       "NODE113                                64.686314    97.745306           NA   \n",
       "NODE114                                64.686314    97.745306           NA   \n",
       "\n",
       "                                    branch_minor  branch_support  \\\n",
       "sample                                                             \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1             0.0   \n",
       "GCA_009296005.1_ASM929600v1_genomic       2.MED1             0.0   \n",
       "GCA_008630485.1_ASM863048v1_genomic       2.MED1             0.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic       2.MED0             0.0   \n",
       "GCA_008630395.1_ASM863039v1_genomic       2.MED1             0.0   \n",
       "...                                          ...             ...   \n",
       "NODE110                                       NA           100.0   \n",
       "NODE111                                       NA           100.0   \n",
       "NODE112                                       NA            73.0   \n",
       "NODE113                                       NA           100.0   \n",
       "NODE114                                       NA           100.0   \n",
       "\n",
       "                                    branch_support_conf_category  \\\n",
       "sample                                                             \n",
       "GCA_009909635.1_ASM990963v1_genomic                          LOW   \n",
       "GCA_009296005.1_ASM929600v1_genomic                          LOW   \n",
       "GCA_008630485.1_ASM863048v1_genomic                          LOW   \n",
       "GCA_008630435.1_ASM863043v1_genomic                          LOW   \n",
       "GCA_008630395.1_ASM863039v1_genomic                          LOW   \n",
       "...                                                          ...   \n",
       "NODE110                                                     HIGH   \n",
       "NODE111                                                     HIGH   \n",
       "NODE112                                                      LOW   \n",
       "NODE113                                                     HIGH   \n",
       "NODE114                                                     HIGH   \n",
       "\n",
       "                                    branch_support_conf_char province_lat  \\\n",
       "sample                                                                      \n",
       "GCA_009909635.1_ASM990963v1_genomic                               47.6222   \n",
       "GCA_009296005.1_ASM929600v1_genomic                               43.3976   \n",
       "GCA_008630485.1_ASM863048v1_genomic                               43.4428   \n",
       "GCA_008630435.1_ASM863043v1_genomic                               43.7368   \n",
       "GCA_008630395.1_ASM863039v1_genomic                               43.0883   \n",
       "...                                                      ...          ...   \n",
       "NODE110                                                    *           NA   \n",
       "NODE111                                                    *           NA   \n",
       "NODE112                                                                NA   \n",
       "NODE113                                                    *           NA   \n",
       "NODE114                                                    *           NA   \n",
       "\n",
       "                                    province_lon biosample_accession strain  \\\n",
       "sample                                                                        \n",
       "GCA_009909635.1_ASM990963v1_genomic      40.7958        SAMN13632815   9_10   \n",
       "GCA_009296005.1_ASM929600v1_genomic      45.6985        SAMN12991209   C-25   \n",
       "GCA_008630485.1_ASM863048v1_genomic      43.4205        SAMN12721152  C-742   \n",
       "GCA_008630435.1_ASM863043v1_genomic      41.7268        SAMN12721146  C-719   \n",
       "GCA_008630395.1_ASM863039v1_genomic      47.1499        SAMN12715009  C-528   \n",
       "...                                          ...                 ...    ...   \n",
       "NODE110                                       NA                  NA     NA   \n",
       "NODE111                                       NA                  NA     NA   \n",
       "NODE112                                       NA                  NA     NA   \n",
       "NODE113                                       NA                  NA     NA   \n",
       "NODE114                                       NA                  NA     NA   \n",
       "\n",
       "                                     country_date_strain  \\\n",
       "sample                                                     \n",
       "GCA_009909635.1_ASM990963v1_genomic   Russia 1923.0 9_10   \n",
       "GCA_009296005.1_ASM929600v1_genomic   Russia 1953.0 C-25   \n",
       "GCA_008630485.1_ASM863048v1_genomic  Russia 1997.0 C-742   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Russia 1996.0 C-719   \n",
       "GCA_008630395.1_ASM863039v1_genomic  Russia 1984.0 C-528   \n",
       "...                                                  ...   \n",
       "NODE110                                               NA   \n",
       "NODE111                                               NA   \n",
       "NODE112                                               NA   \n",
       "NODE113                                               NA   \n",
       "NODE114                                               NA   \n",
       "\n",
       "                                                  province_date_strain  \\\n",
       "sample                                                                   \n",
       "GCA_009909635.1_ASM990963v1_genomic          Rostov Oblast 1923.0 9_10   \n",
       "GCA_009296005.1_ASM929600v1_genomic               Chechnya 1953.0 C-25   \n",
       "GCA_008630485.1_ASM863048v1_genomic    Kabardino-Balkaria 1997.0 C-742   \n",
       "GCA_008630435.1_ASM863043v1_genomic   Karachay-Cherkessia 1996.0 C-719   \n",
       "GCA_008630395.1_ASM863039v1_genomic  Republic of Dagestan 1984.0 C-528   \n",
       "...                                                                ...   \n",
       "NODE110                                                             NA   \n",
       "NODE111                                                             NA   \n",
       "NODE112                                                             NA   \n",
       "NODE113                                                             NA   \n",
       "NODE114                                                             NA   \n",
       "\n",
       "                                    host_human date_mean date_err  \\\n",
       "sample                                                              \n",
       "GCA_009909635.1_ASM990963v1_genomic      Human      1923        0   \n",
       "GCA_009296005.1_ASM929600v1_genomic  Non-Human      1953        0   \n",
       "GCA_008630485.1_ASM863048v1_genomic  Non-Human      1997        0   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Non-Human      1996        0   \n",
       "GCA_008630395.1_ASM863039v1_genomic  Non-Human      1984        0   \n",
       "...                                        ...       ...      ...   \n",
       "NODE110                                     NA        NA       NA   \n",
       "NODE111                                     NA        NA       NA   \n",
       "NODE112                                     NA        NA       NA   \n",
       "NODE113                                     NA        NA       NA   \n",
       "NODE114                                     NA        NA       NA   \n",
       "\n",
       "                                    date_bp_mean root_rtt_dist clade_rtt_dist  \\\n",
       "sample                                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic           98   7.30501e-05     9.6582e-06   \n",
       "GCA_009296005.1_ASM929600v1_genomic           68    7.1472e-05     8.0801e-06   \n",
       "GCA_008630485.1_ASM863048v1_genomic           24   7.53973e-05    1.20054e-05   \n",
       "GCA_008630435.1_ASM863043v1_genomic           25   8.77894e-05    2.43975e-05   \n",
       "GCA_008630395.1_ASM863039v1_genomic           37    7.4455e-05    1.10631e-05   \n",
       "...                                          ...           ...            ...   \n",
       "NODE110                                       NA            NA             NA   \n",
       "NODE111                                       NA            NA             NA   \n",
       "NODE112                                       NA            NA             NA   \n",
       "NODE113                                       NA            NA             NA   \n",
       "NODE114                                       NA            NA             NA   \n",
       "\n",
       "                                         rate_sub  rate_sub_year  \\\n",
       "sample                                                             \n",
       "GCA_009909635.1_ASM990963v1_genomic  1.308552e-07       0.553399   \n",
       "GCA_009296005.1_ASM929600v1_genomic  6.966271e-08       0.294610   \n",
       "GCA_008630485.1_ASM863048v1_genomic  7.702200e-07       3.257336   \n",
       "GCA_008630435.1_ASM863043v1_genomic  7.541059e-07       3.189188   \n",
       "GCA_008630395.1_ASM863039v1_genomic  3.319827e-07       1.403988   \n",
       "...                                           ...            ...   \n",
       "NODE110                              1.112203e-07       0.470362   \n",
       "NODE111                              1.490448e-07       0.630325   \n",
       "NODE112                              2.628923e-08       0.111180   \n",
       "NODE113                              7.814700e-08       0.330491   \n",
       "NODE114                              6.597123e-08       0.278999   \n",
       "\n",
       "                                                               node_name blank  \n",
       "sample                                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic  GCA_009909635.1_ASM990963v1_genomic        \n",
       "GCA_009296005.1_ASM929600v1_genomic  GCA_009296005.1_ASM929600v1_genomic        \n",
       "GCA_008630485.1_ASM863048v1_genomic  GCA_008630485.1_ASM863048v1_genomic        \n",
       "GCA_008630435.1_ASM863043v1_genomic  GCA_008630435.1_ASM863043v1_genomic        \n",
       "GCA_008630395.1_ASM863039v1_genomic  GCA_008630395.1_ASM863039v1_genomic        \n",
       "...                                                                  ...   ...  \n",
       "NODE110                                                          NODE110        \n",
       "NODE111                                                          NODE111        \n",
       "NODE112                                                          NODE112        \n",
       "NODE113                                                          NODE113        \n",
       "NODE114                                                          NODE114        \n",
       "\n",
       "[231 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.ANT4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>node_type</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>province</th>\n",
       "      <th>country</th>\n",
       "      <th>state_prob</th>\n",
       "      <th>state_conf_category</th>\n",
       "      <th>state_rate</th>\n",
       "      <th>continent</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>strain</th>\n",
       "      <th>country_date_strain</th>\n",
       "      <th>province_date_strain</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>node_name</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA5661390</th>\n",
       "      <td>4.813590</td>\n",
       "      <td>69.534111</td>\n",
       "      <td>terminal</td>\n",
       "      <td>478.0</td>\n",
       "      <td>[478, 478]</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>Germany</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002454</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>SAMEA5661390</td>\n",
       "      <td>UNT004</td>\n",
       "      <td>Germany [419:537] UNT004</td>\n",
       "      <td>Bavaria [419:537] UNT004</td>\n",
       "      <td>Human</td>\n",
       "      <td>478</td>\n",
       "      <td>59</td>\n",
       "      <td>1543</td>\n",
       "      <td>6.24896e-05</td>\n",
       "      <td>1.98615e-05</td>\n",
       "      <td>3.415710e-06</td>\n",
       "      <td>14.445374</td>\n",
       "      <td>SAMEA5661390</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661389</th>\n",
       "      <td>11.613209</td>\n",
       "      <td>0.178004</td>\n",
       "      <td>terminal</td>\n",
       "      <td>516.0</td>\n",
       "      <td>[516, 516]</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>Germany</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002454</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>SAMEA5661389</td>\n",
       "      <td>UNT003</td>\n",
       "      <td>Germany [433:599] UNT003</td>\n",
       "      <td>Bavaria [433:599] UNT003</td>\n",
       "      <td>Human</td>\n",
       "      <td>516</td>\n",
       "      <td>83</td>\n",
       "      <td>1505</td>\n",
       "      <td>5.80821e-05</td>\n",
       "      <td>1.5454e-05</td>\n",
       "      <td>3.624344e-09</td>\n",
       "      <td>0.015328</td>\n",
       "      <td>SAMEA5661389</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661385</th>\n",
       "      <td>12.406262</td>\n",
       "      <td>34.354881</td>\n",
       "      <td>terminal</td>\n",
       "      <td>517.0</td>\n",
       "      <td>[517, 517]</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>Germany</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002454</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>SAMEA5661385</td>\n",
       "      <td>PET004</td>\n",
       "      <td>Germany [434:600] PET004</td>\n",
       "      <td>Bavaria [434:600] PET004</td>\n",
       "      <td>Human</td>\n",
       "      <td>517</td>\n",
       "      <td>83</td>\n",
       "      <td>1504</td>\n",
       "      <td>6.23395e-05</td>\n",
       "      <td>1.97114e-05</td>\n",
       "      <td>6.547866e-07</td>\n",
       "      <td>2.769157</td>\n",
       "      <td>SAMEA5661385</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661384</th>\n",
       "      <td>79.874756</td>\n",
       "      <td>6.056320</td>\n",
       "      <td>terminal</td>\n",
       "      <td>521.0</td>\n",
       "      <td>[521, 521]</td>\n",
       "      <td>Valencia Community</td>\n",
       "      <td>Spain</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002454</td>\n",
       "      <td>Europe</td>\n",
       "      <td>39.326068</td>\n",
       "      <td>-4.837979</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>39.682</td>\n",
       "      <td>-0.765441</td>\n",
       "      <td>SAMEA5661384</td>\n",
       "      <td>VAL001</td>\n",
       "      <td>Spain [432:610] VAL001</td>\n",
       "      <td>Valencia Community [432:610] VAL001</td>\n",
       "      <td>Human</td>\n",
       "      <td>521</td>\n",
       "      <td>89</td>\n",
       "      <td>1500</td>\n",
       "      <td>5.88052e-05</td>\n",
       "      <td>1.61771e-05</td>\n",
       "      <td>1.792881e-08</td>\n",
       "      <td>0.075823</td>\n",
       "      <td>SAMEA5661384</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661372</th>\n",
       "      <td>35.988328</td>\n",
       "      <td>19.306661</td>\n",
       "      <td>terminal</td>\n",
       "      <td>765.0</td>\n",
       "      <td>[765, 765]</td>\n",
       "      <td>Centre-Loire Valley</td>\n",
       "      <td>France</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002454</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.603354</td>\n",
       "      <td>1.888334</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>47.549</td>\n",
       "      <td>1.73241</td>\n",
       "      <td>SAMEA5661372</td>\n",
       "      <td>LSD020</td>\n",
       "      <td>France [650:880] LSD020</td>\n",
       "      <td>Centre-Loire Valley [650:880] LSD020</td>\n",
       "      <td>Human</td>\n",
       "      <td>765</td>\n",
       "      <td>115</td>\n",
       "      <td>1256</td>\n",
       "      <td>6.41122e-05</td>\n",
       "      <td>2.14841e-05</td>\n",
       "      <td>1.268521e-07</td>\n",
       "      <td>0.536470</td>\n",
       "      <td>SAMEA5661372</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE6</th>\n",
       "      <td>40.658299</td>\n",
       "      <td>0.412774</td>\n",
       "      <td>internal</td>\n",
       "      <td>501.0</td>\n",
       "      <td>[474, 517]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Germany</td>\n",
       "      <td>99.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002469</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>93.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.400574e-09</td>\n",
       "      <td>0.010152</td>\n",
       "      <td>NODE6</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE7</th>\n",
       "      <td>224.417933</td>\n",
       "      <td>19.154782</td>\n",
       "      <td>internal</td>\n",
       "      <td>718.0</td>\n",
       "      <td>[637, 765]</td>\n",
       "      <td>NA</td>\n",
       "      <td>France</td>\n",
       "      <td>98.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002454</td>\n",
       "      <td>Europe</td>\n",
       "      <td>46.603354</td>\n",
       "      <td>1.888334</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.018236e-08</td>\n",
       "      <td>0.085353</td>\n",
       "      <td>NODE7</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE8</th>\n",
       "      <td>9.250971</td>\n",
       "      <td>4.780357</td>\n",
       "      <td>internal</td>\n",
       "      <td>471.0</td>\n",
       "      <td>[458, 478]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Germany</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002454</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.221871e-07</td>\n",
       "      <td>0.516741</td>\n",
       "      <td>NODE8</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE9</th>\n",
       "      <td>17.270125</td>\n",
       "      <td>0.390279</td>\n",
       "      <td>internal</td>\n",
       "      <td>489.0</td>\n",
       "      <td>[475, 498]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Germany</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002485</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>85.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.343581e-09</td>\n",
       "      <td>0.022599</td>\n",
       "      <td>NODE9</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE10</th>\n",
       "      <td>13.930256</td>\n",
       "      <td>0.415738</td>\n",
       "      <td>internal</td>\n",
       "      <td>503.0</td>\n",
       "      <td>[490, 511]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Germany</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.002493</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>65.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.056875e-09</td>\n",
       "      <td>0.029844</td>\n",
       "      <td>NODE10</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              branch_length  branch_length_sub node_type  timetree_num_date  \\\n",
       "sample                                                                        \n",
       "SAMEA5661390       4.813590          69.534111  terminal              478.0   \n",
       "SAMEA5661389      11.613209           0.178004  terminal              516.0   \n",
       "SAMEA5661385      12.406262          34.354881  terminal              517.0   \n",
       "SAMEA5661384      79.874756           6.056320  terminal              521.0   \n",
       "SAMEA5661372      35.988328          19.306661  terminal              765.0   \n",
       "...                     ...                ...       ...                ...   \n",
       "NODE6             40.658299           0.412774  internal              501.0   \n",
       "NODE7            224.417933          19.154782  internal              718.0   \n",
       "NODE8              9.250971           4.780357  internal              471.0   \n",
       "NODE9             17.270125           0.390279  internal              489.0   \n",
       "NODE10            13.930256           0.415738  internal              503.0   \n",
       "\n",
       "             timetree_num_date_confidence             province  country  \\\n",
       "sample                                                                    \n",
       "SAMEA5661390                   [478, 478]              Bavaria  Germany   \n",
       "SAMEA5661389                   [516, 516]              Bavaria  Germany   \n",
       "SAMEA5661385                   [517, 517]              Bavaria  Germany   \n",
       "SAMEA5661384                   [521, 521]   Valencia Community    Spain   \n",
       "SAMEA5661372                   [765, 765]  Centre-Loire Valley   France   \n",
       "...                                   ...                  ...      ...   \n",
       "NODE6                          [474, 517]                   NA  Germany   \n",
       "NODE7                          [637, 765]                   NA   France   \n",
       "NODE8                          [458, 478]                   NA  Germany   \n",
       "NODE9                          [475, 498]                   NA  Germany   \n",
       "NODE10                         [490, 511]                   NA  Germany   \n",
       "\n",
       "              state_prob state_conf_category  state_rate continent  \\\n",
       "sample                                                               \n",
       "SAMEA5661390       100.0                HIGH    0.002454    Europe   \n",
       "SAMEA5661389       100.0                HIGH    0.002454    Europe   \n",
       "SAMEA5661385       100.0                HIGH    0.002454    Europe   \n",
       "SAMEA5661384       100.0                HIGH    0.002454    Europe   \n",
       "SAMEA5661372       100.0                HIGH    0.002454    Europe   \n",
       "...                  ...                 ...         ...       ...   \n",
       "NODE6               99.0                HIGH    0.002469    Europe   \n",
       "NODE7               98.0                HIGH    0.002454    Europe   \n",
       "NODE8              100.0                HIGH    0.002454    Europe   \n",
       "NODE9              100.0                HIGH    0.002485    Europe   \n",
       "NODE10             100.0                HIGH    0.002493    Europe   \n",
       "\n",
       "              country_lat  country_lon branch_major branch_minor  \\\n",
       "sample                                                             \n",
       "SAMEA5661390    51.083420    10.423447        0.ANT       0.ANT4   \n",
       "SAMEA5661389    51.083420    10.423447        0.ANT       0.ANT4   \n",
       "SAMEA5661385    51.083420    10.423447        0.ANT       0.ANT4   \n",
       "SAMEA5661384    39.326068    -4.837979        0.ANT       0.ANT4   \n",
       "SAMEA5661372    46.603354     1.888334        0.ANT       0.ANT4   \n",
       "...                   ...          ...          ...          ...   \n",
       "NODE6           51.083420    10.423447           NA           NA   \n",
       "NODE7           46.603354     1.888334           NA           NA   \n",
       "NODE8           51.083420    10.423447           NA           NA   \n",
       "NODE9           51.083420    10.423447           NA           NA   \n",
       "NODE10          51.083420    10.423447           NA           NA   \n",
       "\n",
       "              branch_support branch_support_conf_category  \\\n",
       "sample                                                      \n",
       "SAMEA5661390             0.0                          LOW   \n",
       "SAMEA5661389             0.0                          LOW   \n",
       "SAMEA5661385             0.0                          LOW   \n",
       "SAMEA5661384             0.0                          LOW   \n",
       "SAMEA5661372             0.0                          LOW   \n",
       "...                      ...                          ...   \n",
       "NODE6                   93.0                          LOW   \n",
       "NODE7                  100.0                         HIGH   \n",
       "NODE8                  100.0                         HIGH   \n",
       "NODE9                   85.0                          LOW   \n",
       "NODE10                  65.0                          LOW   \n",
       "\n",
       "             branch_support_conf_char province_lat province_lon  \\\n",
       "sample                                                            \n",
       "SAMEA5661390                               48.9468      11.4039   \n",
       "SAMEA5661389                               48.9468      11.4039   \n",
       "SAMEA5661385                               48.9468      11.4039   \n",
       "SAMEA5661384                                39.682    -0.765441   \n",
       "SAMEA5661372                                47.549      1.73241   \n",
       "...                               ...          ...          ...   \n",
       "NODE6                                           NA           NA   \n",
       "NODE7                               *           NA           NA   \n",
       "NODE8                               *           NA           NA   \n",
       "NODE9                                           NA           NA   \n",
       "NODE10                                          NA           NA   \n",
       "\n",
       "             biosample_accession  strain       country_date_strain  \\\n",
       "sample                                                               \n",
       "SAMEA5661390        SAMEA5661390  UNT004  Germany [419:537] UNT004   \n",
       "SAMEA5661389        SAMEA5661389  UNT003  Germany [433:599] UNT003   \n",
       "SAMEA5661385        SAMEA5661385  PET004  Germany [434:600] PET004   \n",
       "SAMEA5661384        SAMEA5661384  VAL001    Spain [432:610] VAL001   \n",
       "SAMEA5661372        SAMEA5661372  LSD020   France [650:880] LSD020   \n",
       "...                          ...     ...                       ...   \n",
       "NODE6                         NA      NA                        NA   \n",
       "NODE7                         NA      NA                        NA   \n",
       "NODE8                         NA      NA                        NA   \n",
       "NODE9                         NA      NA                        NA   \n",
       "NODE10                        NA      NA                        NA   \n",
       "\n",
       "                              province_date_strain host_human date_mean  \\\n",
       "sample                                                                    \n",
       "SAMEA5661390              Bavaria [419:537] UNT004      Human       478   \n",
       "SAMEA5661389              Bavaria [433:599] UNT003      Human       516   \n",
       "SAMEA5661385              Bavaria [434:600] PET004      Human       517   \n",
       "SAMEA5661384   Valencia Community [432:610] VAL001      Human       521   \n",
       "SAMEA5661372  Centre-Loire Valley [650:880] LSD020      Human       765   \n",
       "...                                            ...        ...       ...   \n",
       "NODE6                                           NA         NA        NA   \n",
       "NODE7                                           NA         NA        NA   \n",
       "NODE8                                           NA         NA        NA   \n",
       "NODE9                                           NA         NA        NA   \n",
       "NODE10                                          NA         NA        NA   \n",
       "\n",
       "             date_err date_bp_mean root_rtt_dist clade_rtt_dist      rate_sub  \\\n",
       "sample                                                                          \n",
       "SAMEA5661390       59         1543   6.24896e-05    1.98615e-05  3.415710e-06   \n",
       "SAMEA5661389       83         1505   5.80821e-05     1.5454e-05  3.624344e-09   \n",
       "SAMEA5661385       83         1504   6.23395e-05    1.97114e-05  6.547866e-07   \n",
       "SAMEA5661384       89         1500   5.88052e-05    1.61771e-05  1.792881e-08   \n",
       "SAMEA5661372      115         1256   6.41122e-05    2.14841e-05  1.268521e-07   \n",
       "...               ...          ...           ...            ...           ...   \n",
       "NODE6              NA           NA            NA             NA  2.400574e-09   \n",
       "NODE7              NA           NA            NA             NA  2.018236e-08   \n",
       "NODE8              NA           NA            NA             NA  1.221871e-07   \n",
       "NODE9              NA           NA            NA             NA  5.343581e-09   \n",
       "NODE10             NA           NA            NA             NA  7.056875e-09   \n",
       "\n",
       "              rate_sub_year     node_name blank  \n",
       "sample                                           \n",
       "SAMEA5661390      14.445374  SAMEA5661390        \n",
       "SAMEA5661389       0.015328  SAMEA5661389        \n",
       "SAMEA5661385       2.769157  SAMEA5661385        \n",
       "SAMEA5661384       0.075823  SAMEA5661384        \n",
       "SAMEA5661372       0.536470  SAMEA5661372        \n",
       "...                     ...           ...   ...  \n",
       "NODE6              0.010152         NODE6        \n",
       "NODE7              0.085353         NODE7        \n",
       "NODE8              0.516741         NODE8        \n",
       "NODE9              0.022599         NODE9        \n",
       "NODE10             0.029844        NODE10        \n",
       "\n",
       "[23 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.PRE\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>node_type</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>province</th>\n",
       "      <th>country</th>\n",
       "      <th>state_prob</th>\n",
       "      <th>state_conf_category</th>\n",
       "      <th>state_rate</th>\n",
       "      <th>continent</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>strain</th>\n",
       "      <th>country_date_strain</th>\n",
       "      <th>province_date_strain</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>node_name</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA104233048</th>\n",
       "      <td>214.525846</td>\n",
       "      <td>38.594663</td>\n",
       "      <td>terminal</td>\n",
       "      <td>-1944.0</td>\n",
       "      <td>[-1686, -1686]</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>Germany</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>SAMEA104233048</td>\n",
       "      <td>Post6</td>\n",
       "      <td>Germany [-2007:-1882] Post6</td>\n",
       "      <td>Bavaria [-2007:-1882] Post6</td>\n",
       "      <td>Human</td>\n",
       "      <td>-1944.5</td>\n",
       "      <td>62.5</td>\n",
       "      <td>3965.5</td>\n",
       "      <td>6.19918e-05</td>\n",
       "      <td>5.85177e-05</td>\n",
       "      <td>4.254023e-08</td>\n",
       "      <td>0.179907</td>\n",
       "      <td>SAMEA104233048</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233049</th>\n",
       "      <td>64.193490</td>\n",
       "      <td>53.214854</td>\n",
       "      <td>terminal</td>\n",
       "      <td>-2272.0</td>\n",
       "      <td>[-2272, -2272]</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>Germany</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>SAMEA104233049</td>\n",
       "      <td>1343UnTal85</td>\n",
       "      <td>Germany [-2396:-2148] 1343UnTal85</td>\n",
       "      <td>Bavaria [-2396:-2148] 1343UnTal85</td>\n",
       "      <td>Human</td>\n",
       "      <td>-2272</td>\n",
       "      <td>124</td>\n",
       "      <td>4293</td>\n",
       "      <td>4.64152e-05</td>\n",
       "      <td>4.29411e-05</td>\n",
       "      <td>1.960172e-07</td>\n",
       "      <td>0.828976</td>\n",
       "      <td>SAMEA104233049</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233047</th>\n",
       "      <td>257.758638</td>\n",
       "      <td>11.121895</td>\n",
       "      <td>terminal</td>\n",
       "      <td>-2457.0</td>\n",
       "      <td>[-2457, -2457]</td>\n",
       "      <td>Pärnu maakond</td>\n",
       "      <td>Estonia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>Europe</td>\n",
       "      <td>58.752378</td>\n",
       "      <td>25.331908</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>58.3195</td>\n",
       "      <td>24.3026</td>\n",
       "      <td>SAMEA104233047</td>\n",
       "      <td>KunilaII</td>\n",
       "      <td>Estonia [-2574:-2340] KunilaII</td>\n",
       "      <td>Pärnu maakond [-2574:-2340] KunilaII</td>\n",
       "      <td>Human</td>\n",
       "      <td>-2457</td>\n",
       "      <td>117</td>\n",
       "      <td>4478</td>\n",
       "      <td>3.29161e-05</td>\n",
       "      <td>2.9442e-05</td>\n",
       "      <td>1.020276e-08</td>\n",
       "      <td>0.043148</td>\n",
       "      <td>SAMEA104233047</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>219.837448</td>\n",
       "      <td>10.901882</td>\n",
       "      <td>terminal</td>\n",
       "      <td>-2546.0</td>\n",
       "      <td>[-2546, -2546]</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>Europe</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "      <td>SAMEA104233046</td>\n",
       "      <td>Gyvakarai1</td>\n",
       "      <td>Lithuania [-2621:-2472] Gyvakarai1</td>\n",
       "      <td>Panevezys County [-2621:-2472] Gyvakarai1</td>\n",
       "      <td>Human</td>\n",
       "      <td>-2546.5</td>\n",
       "      <td>74.5</td>\n",
       "      <td>4567.5</td>\n",
       "      <td>2.08194e-05</td>\n",
       "      <td>1.73453e-05</td>\n",
       "      <td>1.172606e-08</td>\n",
       "      <td>0.049591</td>\n",
       "      <td>SAMEA104233046</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>64.247120</td>\n",
       "      <td>267.079819</td>\n",
       "      <td>terminal</td>\n",
       "      <td>-2776.0</td>\n",
       "      <td>[-2776, -2776]</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "      <td>SAMEA3541827</td>\n",
       "      <td>RISE509</td>\n",
       "      <td>Russia [-2876:-2677] RISE509</td>\n",
       "      <td>Krasnoyarsk Krai [-2876:-2677] RISE509</td>\n",
       "      <td>Human</td>\n",
       "      <td>-2776.5</td>\n",
       "      <td>99.5</td>\n",
       "      <td>4797.5</td>\n",
       "      <td>8.9963e-06</td>\n",
       "      <td>5.5222e-06</td>\n",
       "      <td>9.829686e-07</td>\n",
       "      <td>4.157071</td>\n",
       "      <td>SAMEA3541827</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>51.578810</td>\n",
       "      <td>116.556103</td>\n",
       "      <td>internal</td>\n",
       "      <td>-2723.0</td>\n",
       "      <td>[-2791, -2670]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>97.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.001180</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>98.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.343379e-07</td>\n",
       "      <td>2.259767</td>\n",
       "      <td>NODE2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE3</th>\n",
       "      <td>29.815516</td>\n",
       "      <td>24.849557</td>\n",
       "      <td>internal</td>\n",
       "      <td>-2690.0</td>\n",
       "      <td>[-2727, -2667]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>99.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.970737e-07</td>\n",
       "      <td>0.833444</td>\n",
       "      <td>NODE3</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>13.108865</td>\n",
       "      <td>27.524550</td>\n",
       "      <td>internal</td>\n",
       "      <td>-2675.0</td>\n",
       "      <td>[-2692, -2667]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>Europe</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.964864e-07</td>\n",
       "      <td>2.099690</td>\n",
       "      <td>NODE4</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>348.749632</td>\n",
       "      <td>53.019582</td>\n",
       "      <td>internal</td>\n",
       "      <td>-2348.0</td>\n",
       "      <td>[-2462, -2273]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Germany</td>\n",
       "      <td>89.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.594801e-08</td>\n",
       "      <td>0.152028</td>\n",
       "      <td>NODE5</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE6</th>\n",
       "      <td>177.167644</td>\n",
       "      <td>469.123092</td>\n",
       "      <td>internal</td>\n",
       "      <td>-2173.0</td>\n",
       "      <td>[-2388, -1948]</td>\n",
       "      <td>NA</td>\n",
       "      <td>Germany</td>\n",
       "      <td>88.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>Europe</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.261158e-07</td>\n",
       "      <td>2.647905</td>\n",
       "      <td>NODE6</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                branch_length  branch_length_sub node_type  timetree_num_date  \\\n",
       "sample                                                                          \n",
       "SAMEA104233048     214.525846          38.594663  terminal            -1944.0   \n",
       "SAMEA104233049      64.193490          53.214854  terminal            -2272.0   \n",
       "SAMEA104233047     257.758638          11.121895  terminal            -2457.0   \n",
       "SAMEA104233046     219.837448          10.901882  terminal            -2546.0   \n",
       "SAMEA3541827        64.247120         267.079819  terminal            -2776.0   \n",
       "...                       ...                ...       ...                ...   \n",
       "NODE2               51.578810         116.556103  internal            -2723.0   \n",
       "NODE3               29.815516          24.849557  internal            -2690.0   \n",
       "NODE4               13.108865          27.524550  internal            -2675.0   \n",
       "NODE5              348.749632          53.019582  internal            -2348.0   \n",
       "NODE6              177.167644         469.123092  internal            -2173.0   \n",
       "\n",
       "               timetree_num_date_confidence          province    country  \\\n",
       "sample                                                                     \n",
       "SAMEA104233048               [-1686, -1686]           Bavaria    Germany   \n",
       "SAMEA104233049               [-2272, -2272]           Bavaria    Germany   \n",
       "SAMEA104233047               [-2457, -2457]     Pärnu maakond    Estonia   \n",
       "SAMEA104233046               [-2546, -2546]  Panevezys County  Lithuania   \n",
       "SAMEA3541827                 [-2776, -2776]  Krasnoyarsk Krai     Russia   \n",
       "...                                     ...               ...        ...   \n",
       "NODE2                        [-2791, -2670]                NA     Russia   \n",
       "NODE3                        [-2727, -2667]                NA     Russia   \n",
       "NODE4                        [-2692, -2667]                NA     Russia   \n",
       "NODE5                        [-2462, -2273]                NA    Germany   \n",
       "NODE6                        [-2388, -1948]                NA    Germany   \n",
       "\n",
       "                state_prob state_conf_category  state_rate continent  \\\n",
       "sample                                                                 \n",
       "SAMEA104233048       100.0                HIGH    0.001178    Europe   \n",
       "SAMEA104233049       100.0                HIGH    0.001178    Europe   \n",
       "SAMEA104233047       100.0                HIGH    0.001178    Europe   \n",
       "SAMEA104233046       100.0                HIGH    0.001178    Europe   \n",
       "SAMEA3541827         100.0                HIGH    0.001178    Europe   \n",
       "...                    ...                 ...         ...       ...   \n",
       "NODE2                 97.0                HIGH    0.001180    Europe   \n",
       "NODE3                 99.0                HIGH    0.001178    Europe   \n",
       "NODE4                100.0                HIGH    0.001178    Europe   \n",
       "NODE5                 89.0                 LOW    0.001178    Europe   \n",
       "NODE6                 88.0                 LOW    0.001178    Europe   \n",
       "\n",
       "                country_lat  country_lon branch_major branch_minor  \\\n",
       "sample                                                               \n",
       "SAMEA104233048    51.083420    10.423447        0.PRE       0.PRE1   \n",
       "SAMEA104233049    51.083420    10.423447        0.PRE       0.PRE1   \n",
       "SAMEA104233047    58.752378    25.331908        0.PRE       0.PRE1   \n",
       "SAMEA104233046    55.350000    23.750000        0.PRE       0.PRE1   \n",
       "SAMEA3541827      64.686314    97.745306        0.PRE       0.PRE2   \n",
       "...                     ...          ...          ...          ...   \n",
       "NODE2             64.686314    97.745306           NA           NA   \n",
       "NODE3             64.686314    97.745306           NA           NA   \n",
       "NODE4             64.686314    97.745306           NA           NA   \n",
       "NODE5             51.083420    10.423447           NA           NA   \n",
       "NODE6             51.083420    10.423447           NA           NA   \n",
       "\n",
       "                branch_support branch_support_conf_category  \\\n",
       "sample                                                        \n",
       "SAMEA104233048             0.0                          LOW   \n",
       "SAMEA104233049             0.0                          LOW   \n",
       "SAMEA104233047             0.0                          LOW   \n",
       "SAMEA104233046             0.0                          LOW   \n",
       "SAMEA3541827               0.0                          LOW   \n",
       "...                        ...                          ...   \n",
       "NODE2                     98.0                         HIGH   \n",
       "NODE3                    100.0                         HIGH   \n",
       "NODE4                    100.0                         HIGH   \n",
       "NODE5                    100.0                         HIGH   \n",
       "NODE6                    100.0                         HIGH   \n",
       "\n",
       "               branch_support_conf_char province_lat province_lon  \\\n",
       "sample                                                              \n",
       "SAMEA104233048                               48.9468      11.4039   \n",
       "SAMEA104233049                               48.9468      11.4039   \n",
       "SAMEA104233047                               58.3195      24.3026   \n",
       "SAMEA104233046                               55.9156      25.0312   \n",
       "SAMEA3541827                                 63.3234       97.098   \n",
       "...                                 ...          ...          ...   \n",
       "NODE2                                 *           NA           NA   \n",
       "NODE3                                 *           NA           NA   \n",
       "NODE4                                 *           NA           NA   \n",
       "NODE5                                 *           NA           NA   \n",
       "NODE6                                 *           NA           NA   \n",
       "\n",
       "               biosample_accession       strain  \\\n",
       "sample                                            \n",
       "SAMEA104233048      SAMEA104233048        Post6   \n",
       "SAMEA104233049      SAMEA104233049  1343UnTal85   \n",
       "SAMEA104233047      SAMEA104233047     KunilaII   \n",
       "SAMEA104233046      SAMEA104233046   Gyvakarai1   \n",
       "SAMEA3541827          SAMEA3541827      RISE509   \n",
       "...                            ...          ...   \n",
       "NODE2                           NA           NA   \n",
       "NODE3                           NA           NA   \n",
       "NODE4                           NA           NA   \n",
       "NODE5                           NA           NA   \n",
       "NODE6                           NA           NA   \n",
       "\n",
       "                               country_date_strain  \\\n",
       "sample                                               \n",
       "SAMEA104233048         Germany [-2007:-1882] Post6   \n",
       "SAMEA104233049   Germany [-2396:-2148] 1343UnTal85   \n",
       "SAMEA104233047      Estonia [-2574:-2340] KunilaII   \n",
       "SAMEA104233046  Lithuania [-2621:-2472] Gyvakarai1   \n",
       "SAMEA3541827          Russia [-2876:-2677] RISE509   \n",
       "...                                            ...   \n",
       "NODE2                                           NA   \n",
       "NODE3                                           NA   \n",
       "NODE4                                           NA   \n",
       "NODE5                                           NA   \n",
       "NODE6                                           NA   \n",
       "\n",
       "                                     province_date_strain host_human  \\\n",
       "sample                                                                 \n",
       "SAMEA104233048                Bavaria [-2007:-1882] Post6      Human   \n",
       "SAMEA104233049          Bavaria [-2396:-2148] 1343UnTal85      Human   \n",
       "SAMEA104233047       Pärnu maakond [-2574:-2340] KunilaII      Human   \n",
       "SAMEA104233046  Panevezys County [-2621:-2472] Gyvakarai1      Human   \n",
       "SAMEA3541827       Krasnoyarsk Krai [-2876:-2677] RISE509      Human   \n",
       "...                                                   ...        ...   \n",
       "NODE2                                                  NA         NA   \n",
       "NODE3                                                  NA         NA   \n",
       "NODE4                                                  NA         NA   \n",
       "NODE5                                                  NA         NA   \n",
       "NODE6                                                  NA         NA   \n",
       "\n",
       "               date_mean date_err date_bp_mean root_rtt_dist clade_rtt_dist  \\\n",
       "sample                                                                        \n",
       "SAMEA104233048   -1944.5     62.5       3965.5   6.19918e-05    5.85177e-05   \n",
       "SAMEA104233049     -2272      124         4293   4.64152e-05    4.29411e-05   \n",
       "SAMEA104233047     -2457      117         4478   3.29161e-05     2.9442e-05   \n",
       "SAMEA104233046   -2546.5     74.5       4567.5   2.08194e-05    1.73453e-05   \n",
       "SAMEA3541827     -2776.5     99.5       4797.5    8.9963e-06     5.5222e-06   \n",
       "...                  ...      ...          ...           ...            ...   \n",
       "NODE2                 NA       NA           NA            NA             NA   \n",
       "NODE3                 NA       NA           NA            NA             NA   \n",
       "NODE4                 NA       NA           NA            NA             NA   \n",
       "NODE5                 NA       NA           NA            NA             NA   \n",
       "NODE6                 NA       NA           NA            NA             NA   \n",
       "\n",
       "                    rate_sub  rate_sub_year       node_name blank  \n",
       "sample                                                             \n",
       "SAMEA104233048  4.254023e-08       0.179907  SAMEA104233048        \n",
       "SAMEA104233049  1.960172e-07       0.828976  SAMEA104233049        \n",
       "SAMEA104233047  1.020276e-08       0.043148  SAMEA104233047        \n",
       "SAMEA104233046  1.172606e-08       0.049591  SAMEA104233046        \n",
       "SAMEA3541827    9.829686e-07       4.157071    SAMEA3541827        \n",
       "...                      ...            ...             ...   ...  \n",
       "NODE2           5.343379e-07       2.259767           NODE2        \n",
       "NODE3           1.970737e-07       0.833444           NODE3        \n",
       "NODE4           4.964864e-07       2.099690           NODE4        \n",
       "NODE5           3.594801e-08       0.152028           NODE5        \n",
       "NODE6           6.261158e-07       2.647905           NODE6        \n",
       "\n",
       "[15 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    print(branch)\n",
    "    \n",
    "    columns = [\n",
    "        # Required\n",
    "        \"branch_length\",\n",
    "        \"branch_length_sub\",\n",
    "        \"node_type\",\n",
    "        # Time Tree\n",
    "        \"timetree_num_date\",\n",
    "        \"timetree_num_date_confidence\",\n",
    "        # Geo\n",
    "        #\"country\",\n",
    "        \"province\",\n",
    "        \"state\",\n",
    "        \"state_prob\",\n",
    "        \"state_conf_category\",\n",
    "        \"state_rate\",\n",
    "        \"state_continent\",\n",
    "        \"state_lat\",\n",
    "        \"state_lon\",\n",
    "        #\"continuous_geo\",\n",
    "        # Colors and Filters \n",
    "        \"branch_major\",\n",
    "        \"branch_minor\",\n",
    "        \"branch_support\",\n",
    "        \"branch_support_conf_category\",  \n",
    "        \"branch_support_conf_char\",  \n",
    "        #\"continent\",\n",
    "        \"province_lat\",\n",
    "        \"province_lon\",\n",
    "        #\"country_lat\",\n",
    "        #\"country_lon\",\n",
    "        # Text Description\n",
    "        \"biosample_accession\",\n",
    "        \"strain\",\n",
    "        \"country_date_strain\",\n",
    "        \"province_date_strain\",\n",
    "        \"host_human\",\n",
    "        # Tip Dates\n",
    "        \"date_mean\",\n",
    "        \"date_err\",\n",
    "        \"date_bp_mean\",\n",
    "        # Stats\n",
    "        \"root_rtt_dist\",\n",
    "        \"clade_rtt_dist\",\n",
    "        \"rate_sub\",\n",
    "        \"rate_sub_year\",\n",
    "    ]\n",
    "    \n",
    "    auspice_df = copy.copy(tree_dict[branch][\"df\"][columns])\n",
    "\n",
    "    # Edit df\n",
    "    auspice_df[\"node_name\"] = list(auspice_df.index)\n",
    "    auspice_df[\"blank\"] = [\" \"] * len(auspice_df)       \n",
    "    \n",
    "    # Rename\n",
    "    auspice_df.rename(columns={\"state\": \"country\"}, inplace=True)\n",
    "    auspice_df.rename(columns={\"state_continent\": \"continent\"}, inplace=True)\n",
    "    auspice_df.rename(columns={\"state_lat\": \"country_lat\"}, inplace=True)\n",
    "    auspice_df.rename(columns={\"state_lon\": \"country_lon\"}, inplace=True)\n",
    "    \n",
    "    # Make states grey if low confidence\n",
    "    \"\"\"for rec in auspice_df.iterrows():\n",
    "        state_country = rec[1][\"country\"]\n",
    "        state_continent = rec[1][\"continent\"]\n",
    "        state_prob = rec[1][\"state_prob\"]\n",
    "        if state_prob < 95:\n",
    "            auspice_df.at[rec[0], \"continent\"] = NO_DATA_CHAR\n",
    "            auspice_df.at[rec[0], \"country\"] = NO_DATA_CHAR\"\"\"\n",
    "    \n",
    "    display(auspice_df)\n",
    "    tree_dict[branch][\"auspice_df\"] = auspice_df\n",
    "    #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "integrated-development",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "municipal-dubai",
   "metadata": {},
   "source": [
    "## Time Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "detected-marketing",
   "metadata": {},
   "outputs": [],
   "source": [
    "for branch in tree_dict:\n",
    "    out_timetree = copy.deepcopy(tree_dict[branch][\"tree\"])\n",
    "\n",
    "    metadata_to_comment(out_timetree, tree_dict[branch][\"df\"])    \n",
    "    out_timetree_nex_path = os.path.join(augur_dir, branch + \".timetree.nex\")\n",
    "    Phylo.write(out_timetree, out_timetree_nex_path, \"nexus\")\n",
    "\n",
    "    for c in out_timetree.find_clades():\n",
    "        c.comment = None\n",
    "\n",
    "    out_timetree_nwk_path = os.path.join(augur_dir, branch + \".timetree.nwk\")\n",
    "    Phylo.write(out_timetree, out_timetree_nwk_path, \"newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aware-extraction",
   "metadata": {},
   "source": [
    "## Augur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "hundred-difference",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'branch_length': 0.0, 'branch_length_sub': 0.0, 'node_type': 'internal', 'num_date': 1866.0, 'num_date_confidence': [1810, 1907], 'province': 'NA', 'country': 'China', 'state_prob': 59.0, 'state_conf_category': 'LOW', 'state_rate': 0.0, 'continent': 'Asia', 'country_lat': 35.000074, 'country_lon': 104.999927, 'branch_major': 'NA', 'branch_minor': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'province_lat': 'NA', 'province_lon': 'NA', 'biosample_accession': 'NA', 'strain': 'NA', 'country_date_strain': 'NA', 'province_date_strain': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_err': 'NA', 'date_bp_mean': 'NA', 'root_rtt_dist': 'NA', 'clade_rtt_dist': 'NA', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'node_name': 'NODE0', 'blank': ' '}\n",
      "{'branch_length': 0.0, 'branch_length_sub': 0.0, 'node_type': 'internal', 'num_date': 1274.0, 'num_date_confidence': [1211, 1318], 'province': 'NA', 'country': 'Germany', 'state_prob': 36.0, 'state_conf_category': 'LOW', 'state_rate': 0.0, 'continent': 'Europe', 'country_lat': 51.0834196, 'country_lon': 10.4234469, 'branch_major': 'NA', 'branch_minor': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'province_lat': 'NA', 'province_lon': 'NA', 'biosample_accession': 'NA', 'strain': 'NA', 'country_date_strain': 'NA', 'province_date_strain': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_err': 'NA', 'date_bp_mean': 'NA', 'root_rtt_dist': 'NA', 'clade_rtt_dist': 'NA', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'node_name': 'NODE0', 'blank': ' '}\n",
      "{'branch_length': 0.0, 'branch_length_sub': 0.0, 'node_type': 'internal', 'num_date': 1811.0, 'num_date_confidence': [1716, 1883], 'province': 'NA', 'country': 'Kazakhstan', 'state_prob': 59.0, 'state_conf_category': 'LOW', 'state_rate': 0.0, 'continent': 'Asia', 'country_lat': 47.2286086, 'country_lon': 65.20931970000001, 'branch_major': 'NA', 'branch_minor': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'province_lat': 'NA', 'province_lon': 'NA', 'biosample_accession': 'NA', 'strain': 'NA', 'country_date_strain': 'NA', 'province_date_strain': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_err': 'NA', 'date_bp_mean': 'NA', 'root_rtt_dist': 'NA', 'clade_rtt_dist': 'NA', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'node_name': 'NODE0', 'blank': ' '}\n",
      "{'branch_length': 0.0, 'branch_length_sub': 0.0, 'node_type': 'internal', 'num_date': 173.0, 'num_date_confidence': [49, 237], 'province': 'NA', 'country': 'Kyrgyzstan', 'state_prob': 79.0, 'state_conf_category': 'LOW', 'state_rate': 0.0, 'continent': 'Asia', 'country_lat': 41.5089324, 'country_lon': 74.724091, 'branch_major': 'NA', 'branch_minor': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'province_lat': 'NA', 'province_lon': 'NA', 'biosample_accession': 'NA', 'strain': 'NA', 'country_date_strain': 'NA', 'province_date_strain': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_err': 'NA', 'date_bp_mean': 'NA', 'root_rtt_dist': 'NA', 'clade_rtt_dist': 'NA', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'node_name': 'NODE0', 'blank': ' '}\n",
      "{'branch_length': 0.0, 'branch_length_sub': 0.0, 'node_type': 'internal', 'num_date': -2866.0, 'num_date_confidence': [-3040, -2777], 'province': 'NA', 'country': 'Russia', 'state_prob': 92.0, 'state_conf_category': 'LOW', 'state_rate': 0.0, 'continent': 'Europe', 'country_lat': 64.6863136, 'country_lon': 97.7453061, 'branch_major': 'NA', 'branch_minor': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'province_lat': 'NA', 'province_lon': 'NA', 'biosample_accession': 'NA', 'strain': 'NA', 'country_date_strain': 'NA', 'province_date_strain': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_err': 'NA', 'date_bp_mean': 'NA', 'root_rtt_dist': 'NA', 'clade_rtt_dist': 'NA', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'node_name': 'NODE0', 'blank': ' '}\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    augur_dict = augur_export(\n",
    "        tree_path=None, \n",
    "        aln_path=None,  \n",
    "        tree=tree_dict[branch][\"tree\"], \n",
    "        tree_df=tree_dict[branch][\"auspice_df\"], \n",
    "        color_keyword_exclude=[\"geometry\"],\n",
    "        type_convert = {\n",
    "            \"branch_number\" : (lambda x : str(x))\n",
    "        },\n",
    "    )\n",
    "    \n",
    "    tree_dict[branch][\"augur_dict\"] = augur_dict\n",
    "    \n",
    "    first_taxa = list(augur_dict[\"nodes\"].keys())[0]\n",
    "    print(augur_dict[\"nodes\"][first_taxa])\n",
    "\n",
    "    out_path_augur_json = os.path.join(augur_dir, branch + \"_augur.json\" )\n",
    "    utils.write_json(data=tree_dict[branch][\"augur_dict\"], file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "    tree_dict[branch][\"augur_json_path\"] = out_path_augur_json\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "peaceful-concern",
   "metadata": {},
   "source": [
    "## Auspice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aquatic-entry",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main//config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'state_prob' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_rate' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_conf_category' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/1.ORI.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/1.ORI.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "1.PRE\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main//config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'state_prob' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_rate' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_conf_category' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/1.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/1.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "2.MED\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main//config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'state_prob' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_rate' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_conf_category' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/2.MED.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/2.MED.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "0.ANT4\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main//config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/0.ANT4.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/0.ANT4.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "0.PRE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'state_prob' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_rate' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_conf_category' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main//config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/0.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/auspice/all/chromosome/full/filter5/beast/0.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'state_prob' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_rate' is missing type information. We've guessed 'continuous'.\n",
      "\n",
      "WARNING: [config file] Trait 'state_conf_category' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_minor\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"branch_major\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"province\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    print(branch)\n",
    "    # Store the color\n",
    "    if branch == \"0.ANT4\":\n",
    "        branch_major_color = colors_dict[\"branch_major\"][\"0.ANT\"]\n",
    "    else:\n",
    "        branch_major_color = colors_dict[\"branch_major\"][branch]\n",
    "\n",
    "    auspice_dict = auspice_export(\n",
    "        tree=tree_dict[branch][\"tree\"],\n",
    "        augur_json_paths=tree_dict[branch][\"augur_json_path\"], \n",
    "        auspice_config_path=auspice_config_path, \n",
    "        auspice_colors_path=out_path_colors,\n",
    "        auspice_latlons_path=out_path_latlon, \n",
    "        auspice_geo_res=[\"country\",\"province\"],\n",
    "        )\n",
    "\n",
    "\n",
    "    label_col = list(tree_dict[branch][\"auspice_df\"])\n",
    "    #print(label_col)\n",
    "\n",
    "    # Recursively add branch attrs\n",
    "    branch_attributes(\n",
    "        tree_dict=auspice_dict[\"tree\"], \n",
    "        sub_dict=auspice_dict[\"tree\"], \n",
    "        df=tree_dict[branch][\"auspice_df\"],\n",
    "        label_col=label_col,\n",
    "        )\n",
    "    \n",
    "    \n",
    "    # Last manual changes\n",
    "    auspice_dict_copy = copy.deepcopy(auspice_dict)\n",
    "    for i in range(0, len(auspice_dict_copy[\"meta\"][\"colorings\"])):\n",
    "        coloring = auspice_dict_copy[\"meta\"][\"colorings\"][i]\n",
    "        for key in coloring:\n",
    "            # Node type as internal or terminal\n",
    "            if coloring[key] == \"node_type\":\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['internal', '#FFFFFF'], ['terminal', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            # Confidence category\n",
    "            if \"conf_category\" in coloring[key]:\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['LOW', '#FFFFFF'], ['HIGH', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            # Host Human binary\n",
    "            if \"host_human\" in coloring[key]:\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['Human', '#CBB742'], ['Non-Human', \"#60B6F2\"], ['NA', \"#D6D6D6\"]]\n",
    "\n",
    "    # Write outputs - For Local Rendering\n",
    "    out_path_auspice_local_json = os.path.join(auspice_dir, branch + \".json\" )\n",
    "    utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "    export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "    print(\"Validation successful for local JSON.\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "limiting-thriller",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEPRECATED: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "      <th>branch_major_color</th>\n",
       "      <th>geometry_size</th>\n",
       "      <th>geometry</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>length_range</th>\n",
       "      <th>state.rate_range</th>\n",
       "      <th>default.rate_95%_HPD</th>\n",
       "      <th>length_95%_HPD</th>\n",
       "      <th>state.rate_95%_HPD</th>\n",
       "      <th>default.rate_range</th>\n",
       "      <th>state.rate</th>\n",
       "      <th>default.rate</th>\n",
       "      <th>length</th>\n",
       "      <th>posterior</th>\n",
       "      <th>state.prob</th>\n",
       "      <th>height_median</th>\n",
       "      <th>height_range</th>\n",
       "      <th>height_95%_HPD</th>\n",
       "      <th>state.rate_median</th>\n",
       "      <th>default.rate_median</th>\n",
       "      <th>length_median</th>\n",
       "      <th>state</th>\n",
       "      <th>state.set.prob</th>\n",
       "      <th>state.set</th>\n",
       "      <th>height</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>country_date_strain</th>\n",
       "      <th>province_date_strain</th>\n",
       "      <th>rate</th>\n",
       "      <th>rate_hpd</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>state_prob</th>\n",
       "      <th>state_conf_category</th>\n",
       "      <th>state_rate</th>\n",
       "      <th>state_rate_hpd</th>\n",
       "      <th>state_continent</th>\n",
       "      <th>state_lat</th>\n",
       "      <th>state_lon</th>\n",
       "      <th>height_hpd</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>149.507456</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>internal</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0.0, 0.0]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0.0, 0.0]</td>\n",
       "      <td>Asia</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>[108.807767550407, 206.2509600693948]</td>\n",
       "      <td>1866.0</td>\n",
       "      <td>[1810, 1907]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       strain date date_bp country province country_lat country_lon  \\\n",
       "sample                                                                \n",
       "NODE0      NA   NA      NA      NA       NA          NA          NA   \n",
       "\n",
       "       province_lat province_lon biovar branch_major branch_minor  \\\n",
       "sample                                                              \n",
       "NODE0            NA           NA     NA           NA           NA   \n",
       "\n",
       "       biosample_accession biosample_comment branch_number continent  \\\n",
       "sample                                                                 \n",
       "NODE0                   NA                NA            NA        NA   \n",
       "\n",
       "       date_mean date_bp_mean date_err lat lon host_human branch_major_color  \\\n",
       "sample                                                                         \n",
       "NODE0         NA           NA       NA  NA  NA         NA                 NA   \n",
       "\n",
       "       geometry_size geometry root_rtt_dist clade_rtt_dist length_range  \\\n",
       "sample                                                                    \n",
       "NODE0             NA       NA            NA             NA           NA   \n",
       "\n",
       "       state.rate_range default.rate_95%_HPD length_95%_HPD  \\\n",
       "sample                                                        \n",
       "NODE0                NA                   NA             NA   \n",
       "\n",
       "       state.rate_95%_HPD default.rate_range state.rate default.rate length  \\\n",
       "sample                                                                        \n",
       "NODE0                  NA                 NA         NA           NA     NA   \n",
       "\n",
       "       posterior state.prob height_median height_range height_95%_HPD  \\\n",
       "sample                                                                  \n",
       "NODE0         NA         NA            NA           NA             NA   \n",
       "\n",
       "       state.rate_median default.rate_median length_median  state  \\\n",
       "sample                                                              \n",
       "NODE0                 NA                  NA            NA  China   \n",
       "\n",
       "       state.set.prob state.set      height  branch_length  branch_length_sub  \\\n",
       "sample                                                                          \n",
       "NODE0              NA        NA  149.507456            0.0                0.0   \n",
       "\n",
       "       node_type  branch_support branch_support_conf_category  \\\n",
       "sample                                                          \n",
       "NODE0   internal           100.0                         HIGH   \n",
       "\n",
       "       branch_support_conf_char country_date_strain province_date_strain  \\\n",
       "sample                                                                     \n",
       "NODE0                         *                  NA                   NA   \n",
       "\n",
       "        rate    rate_hpd  rate_sub  rate_sub_year  state_prob  \\\n",
       "sample                                                          \n",
       "NODE0    0.0  [0.0, 0.0]       0.0            0.0        59.0   \n",
       "\n",
       "       state_conf_category  state_rate state_rate_hpd state_continent  \\\n",
       "sample                                                                  \n",
       "NODE0                  LOW         0.0     [0.0, 0.0]            Asia   \n",
       "\n",
       "        state_lat   state_lon                             height_hpd  \\\n",
       "sample                                                                 \n",
       "NODE0   35.000074  104.999927  [108.807767550407, 206.2509600693948]   \n",
       "\n",
       "        timetree_num_date timetree_num_date_confidence  \n",
       "sample                                                  \n",
       "NODE0              1866.0                 [1810, 1907]  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df = tree_dict[\"1.ORI\"][\"df\"]\n",
    "df.loc[df.index.str.endswith('NODE0')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "north-assembly",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
