{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "proprietary-spice",
   "metadata": {},
   "source": [
    "# BEAST Analysis Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "gross-cutting",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 0. SETUP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "expanded-saudi",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from Bio import Phylo, AlignIO\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "connected-livestock",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "removable-vulnerability",
   "metadata": {},
   "outputs": [],
   "source": [
    "#project_dir         = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "#project_dir         = os.path.join(project_dir, \"results\")\n",
    "project_dir          = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main\"\n",
    "\n",
    "#tree_dir            = project_dir + \"/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock\"\n",
    "tree_dir            = project_dir + \"/beast/all/chromosome/clade/phylogeography/\"\n",
    "metadata_path       = project_dir + \"/iqtree/all/chromosome/full/filter5/filter-taxa/metadata.tsv\"\n",
    "beast_dir           = project_dir + \"/beast/all/chromosome/clade\"\n",
    "\n",
    "auspice_config_path = project_dir + \"/config/auspice_config.json\"\n",
    "out_path_colors     = project_dir + \"/augur/all/chromosome/full/filter5/colors.tsv\"\n",
    "out_path_latlon     = project_dir + \"/augur/all/chromosome/full/filter5/latlon.tsv\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "constant_sites_path = project_dir + \"/snippy_multi/all/chromosome/full/snippy-multi.constant_sites.txt\"\n",
    "aln_path            = project_dir + \"/iqtree/all/chromosome/full/filter5/filter-sites/snippy-multi.snps.aln\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ancient-syria",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "linear-burden",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", None)\n",
    "\n",
    "# ------------------------------------------\n",
    "BRANCH_LIST = {\n",
    "    \"1.ORI\" : [\"1.ORI1\", \"1.ORI2\", \"1.ORI3\"],\n",
    "    \"1.IN\": [\"1.IN1\",\"1.IN2\",\"1.IN3\"],  \n",
    "    \"1.ANT\": [\"1.ANT1\"], \n",
    "    \"1.PRE\" : [\"1.PRE0\",\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],\n",
    "    \"2.MED\": [\"2.MED0\", \"2.MED1\",\"2.MED2\",\"2.MED3\" ],      \n",
    "    \"2.ANT\": [\"2.ANT1\",\"2.ANT2\",\"2.ANT3\" ],    \n",
    "    \"4.ANT\": [\"4.ANT1\" ],       \n",
    "    \"3.ANT\": [\"3.ANT1\", \"3.ANT2\" ],  \n",
    "    \"0.ANT\": [\"0.ANT1\", \"0.ANT2\",\"0.ANT3\",\"0.ANT5\"],         \n",
    "    \"0.ANT4\" : [\"0.ANT4\"], \n",
    "    \"0.PE\": [\"0.PE2\", \"0.PE4m\", \"0.PE4m\", \"0.PE4t\", \"0.PE4a\", \"0.PE5\", \"0.PE7\", \"0.PE8\", \"0.PE10\"],   \n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"],        \n",
    "}\n",
    "\n",
    "NUM_STATES = 10\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "JSON_INDENT=2\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "with open(constant_sites_path) as infile:\n",
    "    data = infile.read().strip().split(\",\")\n",
    "    constant_sites = sum([int(count) for count in data])\n",
    "\n",
    "aln = AlignIO.read(aln_path, \"fasta\")\n",
    "variant_sites = len(aln[0].seq)\n",
    "SEQ_LEN = constant_sites + variant_sites\n",
    "\n",
    "POSTERIOR_THRESH = 95"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mechanical-monte",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 1. IMPORT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "convertible-freeze",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "beneficial-parcel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "      <th>branch_major_color</th>\n",
       "      <th>geometry_size</th>\n",
       "      <th>geometry</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>Human</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POINT (-105.607716 38.7251776)</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>#b3f396</td>\n",
       "      <td>4.0</td>\n",
       "      <td>POINT (40.7957942 47.6222451)</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>#1996f3</td>\n",
       "      <td>105.0</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>#1996f3</td>\n",
       "      <td>105.0</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>#1996f3</td>\n",
       "      <td>105.0</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313243_45</th>\n",
       "      <td>Azov38</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313243_45</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>4.0</td>\n",
       "      <td>POINT (40.7957942 47.6222451)</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.000022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313246_49</th>\n",
       "      <td>Gdansk8</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Poland</td>\n",
       "      <td>Pomeranian Voivodeship</td>\n",
       "      <td>52.215933</td>\n",
       "      <td>19.134422</td>\n",
       "      <td>54.2456</td>\n",
       "      <td>18.1099</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313246_49</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>54.245560</td>\n",
       "      <td>18.109900</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POINT (18.1099 54.24556)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6651390</th>\n",
       "      <td>AGU010</td>\n",
       "      <td>[1435:1477]</td>\n",
       "      <td>[-586:-544]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6651390</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1456.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>3.0</td>\n",
       "      <td>POINT (25.24953400167952 54.82269205)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637004</th>\n",
       "      <td>AGU025</td>\n",
       "      <td>[1441:1612]</td>\n",
       "      <td>[-580:-409]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637004</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1526.5</td>\n",
       "      <td>494.5</td>\n",
       "      <td>85.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>3.0</td>\n",
       "      <td>POINT (25.24953400167952 54.82269205)</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637002</th>\n",
       "      <td>AGU007B</td>\n",
       "      <td>[1463:1632]</td>\n",
       "      <td>[-558:-389]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637002</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1547.5</td>\n",
       "      <td>473.5</td>\n",
       "      <td>84.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>#e6ce74</td>\n",
       "      <td>3.0</td>\n",
       "      <td>POINT (25.24953400167952 54.82269205)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      strain         date      date_bp  \\\n",
       "sample                                                                   \n",
       "Reference                               CO92         1992          -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic     9_10       1923.0          -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic    42126       2006.0          -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic    42123       2005.0          -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic    42118       2005.0          -16   \n",
       "...                                      ...          ...          ...   \n",
       "SAMEA7313243_45                       Azov38  [1400:1700]  [-621:-321]   \n",
       "SAMEA7313246_49                      Gdansk8  [1400:1700]  [-621:-321]   \n",
       "SAMEA6651390                          AGU010  [1435:1477]  [-586:-544]   \n",
       "SAMEA6637004                          AGU025  [1441:1612]  [-580:-409]   \n",
       "SAMEA6637002                         AGU007B  [1463:1632]  [-558:-389]   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "SAMEA7313243_45                                        Russia   \n",
       "SAMEA7313246_49                                        Poland   \n",
       "SAMEA6651390                                        Lithuania   \n",
       "SAMEA6637004                                        Lithuania   \n",
       "SAMEA6637002                                        Lithuania   \n",
       "\n",
       "                                                   province  country_lat  \\\n",
       "sample                                                                     \n",
       "Reference                                          Colorado    39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic           Rostov Oblast    64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic                Xinjiang    35.000074   \n",
       "...                                                     ...          ...   \n",
       "SAMEA7313243_45                               Rostov Oblast    64.686314   \n",
       "SAMEA7313246_49                      Pomeranian Voivodeship    52.215933   \n",
       "SAMEA6651390                                 Vilnius County    55.350000   \n",
       "SAMEA6637004                                 Vilnius County    55.350000   \n",
       "SAMEA6637002                                 Vilnius County    55.350000   \n",
       "\n",
       "                                     country_lon province_lat province_lon  \\\n",
       "sample                                                                       \n",
       "Reference                            -100.445882      38.7252     -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic    97.745306      47.6222      40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic   104.999927      42.4805      85.4633   \n",
       "...                                          ...          ...          ...   \n",
       "SAMEA7313243_45                        97.745306      47.6222      40.7958   \n",
       "SAMEA7313246_49                        19.134422      54.2456      18.1099   \n",
       "SAMEA6651390                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637004                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637002                           23.750000      54.8227      25.2495   \n",
       "\n",
       "                                              biovar branch_major  \\\n",
       "sample                                                              \n",
       "Reference                                 Orientalis        1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic       Medievalis        2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          Antiqua        0.ANT   \n",
       "...                                              ...          ...   \n",
       "SAMEA7313243_45                      Second Pandemic        1.PRE   \n",
       "SAMEA7313246_49                      Second Pandemic        1.PRE   \n",
       "SAMEA6651390                         Second Pandemic        1.PRE   \n",
       "SAMEA6637004                         Second Pandemic        1.PRE   \n",
       "SAMEA6637002                         Second Pandemic        1.PRE   \n",
       "\n",
       "                                    branch_minor biosample_accession  \\\n",
       "sample                                                                 \n",
       "Reference                                 1.ORI1        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.ANT1        SAMN07722923   \n",
       "...                                          ...                 ...   \n",
       "SAMEA7313243_45                           1.PRE1     SAMEA7313243_45   \n",
       "SAMEA7313246_49                           1.PRE1     SAMEA7313246_49   \n",
       "SAMEA6651390                              1.PRE1        SAMEA6651390   \n",
       "SAMEA6637004                              1.PRE1        SAMEA6637004   \n",
       "SAMEA6637002                              1.PRE1        SAMEA6637002   \n",
       "\n",
       "                                                     biosample_comment  \\\n",
       "sample                                                                   \n",
       "Reference                              KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic              KEEP: Assembly Modern   \n",
       "...                                                                ...   \n",
       "SAMEA7313243_45                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA7313246_49                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA6651390                                         KEEP: SRA Ancient   \n",
       "SAMEA6637004                                         KEEP: SRA Ancient   \n",
       "SAMEA6637002                                         KEEP: SRA Ancient   \n",
       "\n",
       "                                     branch_number      continent  date_mean  \\\n",
       "sample                                                                         \n",
       "Reference                                        1  North America     1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic              2         Europe     1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic              0           Asia     2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic              0           Asia     2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic              0           Asia     2005.0   \n",
       "...                                            ...            ...        ...   \n",
       "SAMEA7313243_45                                  1         Europe     1550.0   \n",
       "SAMEA7313246_49                                  1         Europe     1550.0   \n",
       "SAMEA6651390                                     1         Europe     1456.0   \n",
       "SAMEA6637004                                     1         Europe     1526.5   \n",
       "SAMEA6637002                                     1         Europe     1547.5   \n",
       "\n",
       "                                     date_bp_mean  date_err        lat  \\\n",
       "sample                                                                   \n",
       "Reference                                    29.0       0.0  38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic          98.0       0.0  47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic          15.0       0.0  42.480495   \n",
       "GCA_009669555.1_ASM966955v1_genomic          16.0       0.0  42.480495   \n",
       "GCA_009669565.1_ASM966956v1_genomic          16.0       0.0  42.480495   \n",
       "...                                           ...       ...        ...   \n",
       "SAMEA7313243_45                             471.0     150.0  47.622245   \n",
       "SAMEA7313246_49                             471.0     150.0  54.245560   \n",
       "SAMEA6651390                                565.0      21.0  54.822692   \n",
       "SAMEA6637004                                494.5      85.5  54.822692   \n",
       "SAMEA6637002                                473.5      84.5  54.822692   \n",
       "\n",
       "                                            lon host_human branch_major_color  \\\n",
       "sample                                                                          \n",
       "Reference                           -105.607716      Human            #ff0000   \n",
       "GCA_009909635.1_ASM990963v1_genomic   40.795794      Human            #b3f396   \n",
       "GCA_009669545.1_ASM966954v1_genomic   85.463346  Non-Human            #1996f3   \n",
       "GCA_009669555.1_ASM966955v1_genomic   85.463346  Non-Human            #1996f3   \n",
       "GCA_009669565.1_ASM966956v1_genomic   85.463346  Non-Human            #1996f3   \n",
       "...                                         ...        ...                ...   \n",
       "SAMEA7313243_45                       40.795794      Human            #e6ce74   \n",
       "SAMEA7313246_49                       18.109900      Human            #e6ce74   \n",
       "SAMEA6651390                          25.249534      Human            #e6ce74   \n",
       "SAMEA6637004                          25.249534      Human            #e6ce74   \n",
       "SAMEA6637002                          25.249534      Human            #e6ce74   \n",
       "\n",
       "                                     geometry_size  \\\n",
       "sample                                               \n",
       "Reference                                      1.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic            4.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic          105.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic          105.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic          105.0   \n",
       "...                                            ...   \n",
       "SAMEA7313243_45                                4.0   \n",
       "SAMEA7313246_49                                1.0   \n",
       "SAMEA6651390                                   3.0   \n",
       "SAMEA6637004                                   3.0   \n",
       "SAMEA6637002                                   3.0   \n",
       "\n",
       "                                                                  geometry  \\\n",
       "sample                                                                       \n",
       "Reference                                   POINT (-105.607716 38.7251776)   \n",
       "GCA_009909635.1_ASM990963v1_genomic          POINT (40.7957942 47.6222451)   \n",
       "GCA_009669545.1_ASM966954v1_genomic   POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669555.1_ASM966955v1_genomic   POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669565.1_ASM966956v1_genomic   POINT (85.46334640000001 42.4804953)   \n",
       "...                                                                    ...   \n",
       "SAMEA7313243_45                              POINT (40.7957942 47.6222451)   \n",
       "SAMEA7313246_49                                   POINT (18.1099 54.24556)   \n",
       "SAMEA6651390                         POINT (25.24953400167952 54.82269205)   \n",
       "SAMEA6637004                         POINT (25.24953400167952 54.82269205)   \n",
       "SAMEA6637002                         POINT (25.24953400167952 54.82269205)   \n",
       "\n",
       "                                     root_rtt_dist  clade_rtt_dist  \n",
       "sample                                                              \n",
       "Reference                                 0.000073        0.000006  \n",
       "GCA_009909635.1_ASM990963v1_genomic       0.000073        0.000010  \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.000054        0.000012  \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.000055        0.000012  \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.000055        0.000012  \n",
       "...                                            ...             ...  \n",
       "SAMEA7313243_45                           0.000075        0.000022  \n",
       "SAMEA7313246_49                           0.000060        0.000007  \n",
       "SAMEA6651390                              0.000060        0.000006  \n",
       "SAMEA6637004                              0.000061        0.000007  \n",
       "SAMEA6637002                              0.000060        0.000006  \n",
       "\n",
       "[601 rows x 27 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.set_index(metadata_df.columns[0], inplace=True)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "passing-capitol",
   "metadata": {},
   "source": [
    "## Most Recent Sampling Date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "tribal-manchester",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path = os.path.join(beast_dir, \"most_recent_sampling_dates.tsv\")\n",
    "mrsd_dict = {}\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    for branch in BRANCH_LIST:\n",
    "        branch_df = metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])]\n",
    "        max_date = max(branch_df[\"date_mean\"])\n",
    "        outfile.write(\"{}\\t{}\\n\".format(branch, max_date))\n",
    "        mrsd_dict[branch] = max_date"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unlikely-beverage",
   "metadata": {},
   "source": [
    "## Colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cleared-dutch",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'branch_major': {'0.PRE': '#8000ff', '0.PE': '#4c4ffc', '0.ANT': '#1996f3', '3.ANT': '#1acee3', '4.ANT': '#4df3ce', '2.ANT': '#80ffb4', '2.MED': '#b3f396', '1.PRE': '#e6ce74', '1.ANT': '#ff964f', '1.IN': '#ff4f28', '1.ORI': '#ff0000', '?': '#969696'}, 'branch_minor': {'0.PRE2': '#8000ff', '0.PRE1': '#7018ff', '0.PE7': '#6130fe', '0.PE2': '#5148fc', '0.PE8': '#425ffa', '0.PE10': '#3275f8', '0.PE4': '#238af5', '0.PE5': '#139ef1', '0.ANT1': '#04b0ed', '0.ANT4': '#0cc1e8', '0.ANT5': '#1bd0e3', '0.ANT2': '#2bdddd', '0.ANT3': '#3ae8d7', '3.ANT1': '#49f1d0', '3.ANT2': '#59f8c8', '4.ANT1': '#68fcc1', '2.ANT3': '#78ffb9', '2.ANT2': '#87ffb0', '2.ANT1': '#97fca7', '2.MED0': '#a6f89e', '2.MED3': '#b6f194', '2.MED2': '#c5e88a', '2.MED1': '#d5dd7f', '1.PRE0': '#e4d075', '1.PRE1': '#f3c16a', '1.PRE2': '#ffb05f', '1.PRE3': '#ff9e53', '1.ANT1': '#ff8a48', '1.IN1': '#ff753c', '1.IN2': '#ff5f30', '1.IN3': '#ff4824', '1.ORI2': '#ff3018', '1.ORI1': '#ff180c', '1.ORI3': '#ff0000', '?': '#969696'}, 'province': {'Krasnoyarsk Krai': '#8000ff', 'Panevezys County': '#7b07ff', 'Pärnu maakond': '#760eff', 'Irkutsk Oblast': '#7215ff', 'Bavaria': '#6d1dff', 'Altai Krai': '#6924fe', 'Qinghai': '#642bfe', 'Republic of Dagestan': '#6032fe', 'Goranboy District': '#5b39fd', 'Syunik Province': '#5740fd', 'Shirak Province': '#5247fc', 'Gegharkunik Province': '#4d4dfc', 'Samtskhe-Javakheti': '#4954fb', 'Shahbuz Rayon': '#445bfb', 'Samara Oblast': '#4062fa', 'Sečuán': '#3b68f9', 'Bayankhongor': '#376ff9', 'Inner Mongolia': '#3275f8', 'Sughd Province': '#2e7bf7', 'Talas Region': '#2982f6', 'Altai Republic': '#2488f5', 'Govi-Altai': '#208ef4', 'Övörkhangai': '#1b94f3', 'Ömnögovi': '#1799f2', 'Bayan-Ölgii': '#129ff1', 'Khovd': '#0ea5ef', 'Xinjiang': '#09aaee', 'Issyk-Kul Region': '#05afed', 'East of England': '#00b4ec', 'Valencia Community': '#05b9ea', 'Centre-Loire Valley': '#09bee9', 'Osh Region': '#0ec3e7', 'Naryn Region': '#12c7e6', 'Gansu': '#17cce4', 'Hovsgel': '#1bd0e3', 'Zavkhan Province': '#20d4e1', 'Dornogovi': '#24d8df', 'Tuva Republic': '#29dcdd', 'Jilin': '#2edfdc', 'St. Petersberg': '#32e3da', 'Zabaykalsky Krai': '#37e6d8', 'Heilongjiang': '#3be9d6', 'Tibet': '#40ecd4', 'Gandaki Pradesh': '#44eed2', 'Maharashtra': '#49f1d0', 'Himachal Pradesh': '#4df3ce', 'Kabardino-Balkaria': '#52f5cc', 'Karachay-Cherkessia': '#57f7ca', 'Ningxia': '#5bf9c7', 'Shaanxi': '#60fac5', 'Hebei': '#64fbc3', 'West Kazakhstan Region': '#69fcc1', 'Nagorno-Karabakh Republic': '#6dfdbe', 'Kyzylorda Region': '#72febc', 'Rostov Oblast': '#76ffb9', 'Astrakhan Oblast': '#7bffb7', 'Balkan': '#80ffb4', 'Kyzylorda Province': '#84ffb2', 'Shymkent': '#89ffaf', 'Navoiy Region': '#8dfead', 'Almaty Region': '#92fdaa', 'West Kazakhstan region': '#96fca7', 'Atyrau Region': '#9bfba5', 'Fizuli District': '#9ffaa2', 'Baku': '#a4f99f', 'Kurdistan': '#a8f79c', 'Shamkir District': '#adf599', 'Stavropol Krai': '#b2f396', 'Ingushetia': '#b6f194', 'Mangystau Province': '#bbee91', 'Chechnya': '#bfec8e', 'Aktobe Region': '#c4e98b', 'Atyrau': '#c8e688', 'Qazakh District': '#cde385', 'Republic of Kalmykia': '#d1df82', 'Tatarstan': '#d6dc7e', 'Lazio': '#dbd87b', 'Greater London': '#dfd478', \"Provence-Alpes-Côte d'Azur\": '#e4d075', 'Baden-Württemberg': '#e8cc72', 'Vilnius County': '#edc76f', 'Pomeranian Voivodeship': '#f1c36b', 'Brandenburg': '#f6be68', 'Nidwalden': '#fab965', 'Oslo': '#ffb462', 'Catalonia': '#ffaf5e', 'Occitanie': '#ffaa5b', 'North Brabant': '#ffa558', 'Nairobi': '#ff9f54', 'Yunnan': '#ff9951', 'Khánh Hoà Province': '#ff944d', 'Lâm Đồng Province': '#ff8e4a', 'Ho Chi Minh City': '#ff8847', 'Guangxi Zhuang': '#ff8243', 'California': '#ff7b40', 'La Libertad': '#ff753c', 'Tamil Nadu': '#ff6f39', 'Saskatchewan': '#ff6835', 'Arizona': '#ff6232', 'New Mexico': '#ff5b2e', 'Colorado': '#ff542b', 'Texas': '#ff4d27', 'Nevada': '#ff4724', 'Central Java': '#ff4020', 'Province de Fianarantsoa': '#ff391d', 'Kirov Oblast': '#ff3219', 'Gujarat': '#ff2b15', 'Fujian': '#ff2412', 'Midlands': '#ff1d0e', 'MSila': '#ff150b', 'State of Mato Grosso': '#ff0e07', 'La Paz': '#ff0704', 'Cajamarca': '#ff0000', '?': '#969696'}, 'country': {'Russia': '#8000ff', 'Lithuania': '#7215ff', 'Estonia': '#652afe', 'Germany': '#573ffd', 'China': '#4a53fc', 'Azerbaijan': '#3c66fa', 'Armenia': '#2f79f7', 'Georgia': '#228bf4', 'Mongolia': '#149df1', 'Tajikistan': '#07adee', 'Kyrgyzstan': '#07bcea', 'England': '#14c9e5', 'Spain': '#22d5e0', 'France': '#2fe0db', 'Nepal': '#3cead5', 'India': '#4af1d0', 'Kazakhstan': '#57f7c9', 'Turkmenistan': '#65fcc3', 'Uzbekistan': '#72febc', 'Iran': '#80ffb4', 'Italy': '#8dfead', 'Poland': '#9afca5', 'Switzerland': '#a8f79d', 'Norway': '#b5f194', 'The Netherlands': '#c3ea8b', 'Kenya': '#d0e083', 'Uganda': '#ddd579', 'Democratic Republic of the Congo': '#ebc970', 'Vietnam': '#f8bc66', 'Myanmar': '#ffad5d', 'United States of America': '#ff9d53', 'Peru': '#ff8b49', 'Canada': '#ff793f', 'Indonesia': '#ff6634', 'Madagascar': '#ff532a', 'Zimbabwe': '#ff3f20', 'Algeria': '#ff2a15', 'Brazil': '#ff150b', 'Bolivia': '#ff0000', '?': '#969696'}}\n"
     ]
    }
   ],
   "source": [
    "colors_dict = {}\n",
    "\n",
    "colors_df = pd.read_csv(out_path_colors, sep='\\t', header=None)\n",
    "colors_df.columns = [\"state\", \"value\", \"color\"]\n",
    "\n",
    "for state in set(colors_df[\"state\"]):\n",
    "    state_df = colors_df[colors_df[\"state\"] == state]    \n",
    "    colors_dict[state] = {}\n",
    "    \n",
    "    for value,color in zip(state_df[\"value\"], state_df[\"color\"]):\n",
    "        colors_dict[state][value] = color\n",
    "\n",
    "print(colors_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "economic-raising",
   "metadata": {},
   "source": [
    "## Tree Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "small-message",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Branch: 1.ORI\n",
      "Rename: GCA_000324805.2_EV76_CN_genomic GCA_000324805.2_EV76-CN_genomic\n",
      "Rename: GCA_000986995.1_YPES001_SEQ_2_ASM_1_genomic GCA_000986995.1_YPES001-SEQ-2-ASM-1_genomic\n",
      "Branch: 1.PRE\n",
      "Branch: 2.MED\n",
      "Rename: GCA_001617735.1_Yersinia_pestis_M_1763_genomic GCA_001617735.1_Yersinia_pestis_M-1763_genomic\n",
      "Rename: GCA_002412305.1_Y.pestis_A_1809_genomic GCA_002412305.1_Y.pestis_A-1809_genomic\n",
      "Rename: GCA_001617785.1_Yersinia_pestis_M_549_genomic GCA_001617785.1_Yersinia_pestis_M-549_genomic\n",
      "Rename: GCA_001617725.1_Yersinia_pestis_M_1484_genomic GCA_001617725.1_Yersinia_pestis_M-1484_genomic\n",
      "Rename: GCA_001617715.1_Yersinia_pestis_M_519_genomic GCA_001617715.1_Yersinia_pestis_M-519_genomic\n",
      "Rename: GCA_001617815.1_Yersinia_pestis_M_1453_genomic GCA_001617815.1_Yersinia_pestis_M-1453_genomic\n",
      "Rename: GCA_001617705.1_Yersinia_pestis_C_791_genomic GCA_001617705.1_Yersinia_pestis_C-791_genomic\n",
      "Branch: 0.ANT4\n",
      "Branch: 0.PRE\n"
     ]
    }
   ],
   "source": [
    "# Construct a dictionary to hold the trees\n",
    "tree_dict = {}\n",
    "\n",
    "for branch in BRANCH_LIST:\n",
    "    for filename in os.listdir(tree_dir):\n",
    "        if not filename.endswith(\".tre\") and not filename.endswith(\".tree\"): continue         \n",
    "        filepath = os.path.join(tree_dir, filename)\n",
    "        \n",
    "        # Option 1 when all trees are present\n",
    "        #if branch in filename: \n",
    "        \n",
    "        # Option 2 when not all trees are present\n",
    "        if filename.split(\"_\")[0] == branch:        \n",
    "            print(\"Branch:\", branch)\n",
    "            tree_dict[branch] = {}    \n",
    "            # Add tree files to dict\n",
    "            tree_dict[branch][\"tree_file_raw\"] = filepath \n",
    "            tree_dict[branch][\"tree_file_edit\"] = os.path.join(tree_dir, branch + \".nex\")\n",
    "            tree_dict[branch][\"sample_rename\"] = {}\n",
    "            \n",
    "            # Read in raw tree to deal with dashes\n",
    "            with open(tree_dict[branch][\"tree_file_raw\"],  \"r\") as infile:                    \n",
    "                with open(tree_dict[branch][\"tree_file_edit\"], \"w\") as outfile:           \n",
    "                    raw_tree = infile.read()\n",
    "                    # Remove quotations if they exist\n",
    "                    raw_tree = raw_tree.replace(\"'\",\"\")\n",
    "                    \n",
    "                    # Split into lines to iterate over\n",
    "                    raw_tree_lines = raw_tree.split(\"\\n\")\n",
    "                    \n",
    "                    # By default don't parse a line for dashes\n",
    "                    taxa_line = False\n",
    "                    \n",
    "                    for line in raw_tree_lines: \n",
    "                                \n",
    "                        if \"TREE\" not in line and \"-\" in line:\n",
    "                            \n",
    "                            if len(line.split(\" \")) == 1:\n",
    "                                name_dashes = line.strip()\n",
    "                                name_no_dashes = name_dashes.replace(\"-\",\"_\")\n",
    "                                tree_dict[branch][\"sample_rename\"][name_no_dashes] = name_dashes\n",
    "                            line = line.replace(\"-\",\"_\")\n",
    "                                \n",
    "                        outfile.write(line + \"\\n\")\n",
    "            \n",
    "            # Read in edited tree\n",
    "            trees = Phylo.parse(tree_dict[branch][\"tree_file_edit\"], \"nexus\")\n",
    "            # There should be only 1 tree\n",
    "            for t in trees:\n",
    "                tree_dict[branch][\"tree\"] = t\n",
    "                tree_dict[branch][\"tree\"].ladderize(reverse=False)\n",
    "                break\n",
    "\n",
    "            # Rename sample names back to with dashes\n",
    "            for c in tree_dict[branch][\"tree\"].find_clades():\n",
    "                if c.name in tree_dict[branch][\"sample_rename\"]:\n",
    "                    orig_name = c.name\n",
    "                    c.name = tree_dict[branch][\"sample_rename\"][c.name]\n",
    "                    print(\"Rename:\", orig_name, c.name)\n",
    "                    \n",
    "                # Strip the date suffix\n",
    "                #if c.name:\n",
    "                #    c.name = \"_\".join(c.name.split(\"_\")[0:-1])           \n",
    "                    \n",
    "            # Rename internal nodes\n",
    "            node_i = 0\n",
    "            for c in tree_dict[branch][\"tree\"].find_clades():\n",
    "                if not c.name:\n",
    "                    c.name = \"NODE{}\".format(node_i)\n",
    "                    node_i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "italic-huntington",
   "metadata": {},
   "source": [
    "## Add Tree Data to Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "identical-circumstances",
   "metadata": {},
   "source": [
    "### Get comments from the first root that is not root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "formal-questionnaire",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n",
      "1.PRE\n",
      "2.MED\n",
      "0.ANT4\n",
      "0.PRE\n"
     ]
    }
   ],
   "source": [
    "TREE_PARAMETERS = None\n",
    "\n",
    "for branch in tree_dict:\n",
    "    \n",
    "    print(branch)\n",
    "    \n",
    "    tree = tree_dict[branch][\"tree\"]\n",
    "    df = copy.deepcopy(metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])])\n",
    "    \n",
    "    # Check if a sample was missed :(\n",
    "    df_samples = df.index\n",
    "    tree_samples = [c.name for c in tree.find_clades()]\n",
    "    \n",
    "    for sample in df_samples:\n",
    "        if sample not in tree_samples:\n",
    "            df.drop(sample, inplace=True)\n",
    "            print(\"Dropping {} from the dataframe.\".format(sample))\n",
    "    \n",
    "    root_comment_dict = parse_comment(tree.root.comment)\n",
    "    \n",
    "    for c in tree.find_clades():\n",
    "        if c.is_terminal():\n",
    "            continue\n",
    "            \n",
    "        if c != tree.root:\n",
    "            comment_dict = parse_comment(c.comment)\n",
    "            for parameter,value in comment_dict.items():\n",
    "                # Intialize parameter values\n",
    "                if \"range\" in parameter or \"95%\" in parameter:\n",
    "                    df[parameter] = [[NO_DATA_CHAR,NO_DATA_CHAR]] * len(df)\n",
    "                    if parameter not in root_comment_dict:\n",
    "                        root_comment_dict[parameter] = '{0,0}'\n",
    "                else:\n",
    "                    df[parameter] = [NO_DATA_CHAR] * len(df)\n",
    "                    if parameter not in root_comment_dict:\n",
    "                        root_comment_dict[parameter] = '0'\n",
    "                    \n",
    "            break\n",
    "\n",
    "    # Update the roots comment to include missing values\n",
    "    new_comment = \"[&\"\n",
    "    new_comment_list = []\n",
    "    for parameter,value in root_comment_dict.items():\n",
    "        param_str = \"{}={}\".format(parameter, value)\n",
    "        new_comment_list.append(param_str)\n",
    "    new_comment += \",\".join(new_comment_list) + \"]\"\n",
    "    tree.root.comment = new_comment\n",
    "    #print(root_comment_dict)\n",
    "    #print(new_comment_list)\n",
    "    \n",
    "    if not TREE_PARAMETERS:\n",
    "        TREE_PARAMETERS = [p for p in root_comment_dict]\n",
    "    \n",
    "    tree_dict[branch][\"df\"] = df \n",
    "    #display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "labeled-apollo",
   "metadata": {},
   "source": [
    "### Parse Tree Comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "annual-mitchell",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n",
      "1.PRE\n",
      "2.MED\n",
      "0.ANT4\n",
      "0.PRE\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    print(branch)\n",
    "    \n",
    "    tree = tree_dict[branch][\"tree\"]\n",
    "    df = tree_dict[branch][\"df\"]\n",
    "    \n",
    "    # Initialize new columns\n",
    "    df[\"timetree_num_date\"] = [NO_DATA_CHAR] * len(df) # Calendar date of height\n",
    "    df[\"timetree_num_date_confidence\"] = [[NO_DATA_CHAR, NO_DATA_CHAR ]] * len(df)\n",
    "    df[\"rate_sub_year\"] = [NO_DATA_CHAR] * len(df)\n",
    "    df[\"branch_length_sub\"] = [NO_DATA_CHAR] * len(df)\n",
    "    df[\"branch_support_conf_category\"] = [NO_DATA_CHAR] * len(df)\n",
    "    df[\"branch_support_conf_char\"] = [NO_DATA_CHAR] * len(df)\n",
    "    \n",
    "    for c in tree.find_clades():\n",
    "        \n",
    "        # --------------------------------------\n",
    "        # Get node type\n",
    "        node_type = \"internal\"\n",
    "        if c.is_terminal():\n",
    "            node_type = \"terminal\"\n",
    "        df.at[c.name, \"node_type\"] = node_type\n",
    "        \n",
    "        # --------------------------------------\n",
    "        # Parse comments\n",
    "        comment_dict = parse_comment(c.comment)\n",
    "        for parameter,val in comment_dict.items():\n",
    "            \n",
    "            # Ranges should be parsed into list     \n",
    "            if \"range\" in parameter or \"95%\" in parameter:\n",
    "                val = [float(v) for v in val.strip(\"{}\").split(\",\")]\n",
    "            # Multiply the posterior by 100 to be comparable to UFboot\n",
    "            elif parameter == \"posterior\":\n",
    "                val = float(val) * 100\n",
    "            # Make sure states are strings\n",
    "            elif parameter == \"state\":\n",
    "                parameter = \"country\"\n",
    "                val = str(val) \n",
    "            elif parameter == \"state.set\":\n",
    "                val = str(val)                  \n",
    "            elif parameter == \"state.set.prob\":\n",
    "                val = str(val)\n",
    "            elif parameter == \"default.rate\":\n",
    "                parameter = \"rate\"\n",
    "                val = float(val)\n",
    "            else:\n",
    "                val = float(val)\n",
    "            df.at[c.name, parameter] = val\n",
    "            \n",
    "            \n",
    "        # Tips don't have posteriors, set to 0\n",
    "        if c.is_terminal():\n",
    "            posterior = 0\n",
    "            df.at[c.name, \"posterior\"] = posterior\n",
    "            comment_dict[\"posterior\"] = posterior\n",
    "            \n",
    "        # Set confidence category for posterior/branch support\n",
    "        posterior = df[\"posterior\"][c.name]\n",
    "        branch_support_conf_category = \"LOW\"\n",
    "        branch_support_conf_char = \"\"\n",
    "        if posterior >= POSTERIOR_THRESH:\n",
    "            branch_support_conf_category = \"HIGH\"\n",
    "            branch_support_conf_char = \"*\"\n",
    "        df.at[c.name,\"branch_support_conf_category\"] = branch_support_conf_category\n",
    "        df.at[c.name,\"branch_support_conf_char\"] = branch_support_conf_char\n",
    "\n",
    "        # Check for missing parameters\n",
    "        # For example, the most recent sample will be missing height uncertainty\n",
    "        for parameter in TREE_PARAMETERS:\n",
    "            if parameter not in comment_dict:\n",
    "                if \"range\" in parameter or \"95%\" in parameter:\n",
    "                    val = [0,0]\n",
    "                else:\n",
    "                    val = 0\n",
    "                df.at[c.name, parameter] = val\n",
    "                comment_dict[parameter] = val\n",
    "                    \n",
    "                #print(c.name, parameter, df[\"date\"][c.name], mrsd_dict[branch])\n",
    "            \n",
    "        # Convert height to calendar dates\n",
    "        height = df[\"height\"][c.name]\n",
    "        height_95_hpd = df[\"height_95%_HPD\"][c.name]\n",
    "        \n",
    "        # The sample with the most recent date does not have a height 95% HPD\n",
    "        if height_95_hpd[0] == NO_DATA_CHAR:\n",
    "            height_95_hpd = [0,0]\n",
    "            df.at[c.name, \"height_95%_HPD\"] = height_95_hpd\n",
    "            \n",
    "        height_calendar = mrsd_dict[branch] - height\n",
    "        height_95_hpd_calendar = [mrsd_dict[branch] -  height_95_hpd[1], mrsd_dict[branch] -  height_95_hpd[0]]\n",
    "    \n",
    "        df.at[c.name, \"timetree_num_date\"] = height_calendar\n",
    "        df.at[c.name, \"timetree_num_date_confidence\"] = height_95_hpd_calendar\n",
    "        \n",
    "        # Add another measure of substitution rate\n",
    "        rate_sub_year = df[\"rate\"][c.name] * SEQ_LEN\n",
    "        df.at[c.name, \"rate_sub_year\"] = rate_sub_year\n",
    "        \n",
    "        # Measure branch length in substitutions\n",
    "        branch_length_sub = rate_sub_year * df[\"length\"][c.name]\n",
    "        df.at[c.name, \"branch_length_sub\"] = branch_length_sub\n",
    "    \n",
    "    df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "robust-tracker",
   "metadata": {},
   "source": [
    "## Reduced Dataframe for Auspice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "brutal-somalia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['strain', 'date', 'date_bp', 'country', 'province', 'country_lat',\n",
      "       'country_lon', 'province_lat', 'province_lon', 'biovar', 'branch_major',\n",
      "       'branch_minor', 'biosample_accession', 'biosample_comment',\n",
      "       'branch_number', 'continent', 'date_mean', 'date_bp_mean', 'date_err',\n",
      "       'lat', 'lon', 'host_human', 'branch_major_color', 'geometry_size',\n",
      "       'geometry', 'root_rtt_dist', 'clade_rtt_dist', 'length_range',\n",
      "       'state.rate_range', 'default.rate_95%_HPD', 'length_95%_HPD',\n",
      "       'state.rate_95%_HPD', 'default.rate_range', 'state.rate',\n",
      "       'default.rate', 'length', 'posterior', 'height_median', 'height_range',\n",
      "       'height_95%_HPD', 'state.rate_median', 'default.rate_median',\n",
      "       'length_median', 'height', 'timetree_num_date',\n",
      "       'timetree_num_date_confidence', 'rate_sub_year', 'branch_length_sub',\n",
      "       'branch_support_conf_category', 'branch_support_conf_char', 'node_type',\n",
      "       'rate', 'state.prob', 'state', 'state.set.prob', 'state.set'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# OPtions\n",
    "print(tree_dict[branch][\"df\"].columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "reasonable-faculty",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>branch_length_time</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>terminal</td>\n",
       "      <td>14.290093</td>\n",
       "      <td>1992.000000</td>\n",
       "      <td>[1991.999999999996, 1992.000000000004]</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>4.231324e-07</td>\n",
       "      <td>2.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>CO92</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>1992</td>\n",
       "      <td>29</td>\n",
       "      <td>14.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000834775.1_ASM83477v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>6.975785</td>\n",
       "      <td>1967.000000</td>\n",
       "      <td>[1966.999999999993, 1967.00000000001]</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Arizona</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>34.3953</td>\n",
       "      <td>-111.763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.163987e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>Dodson</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>1967</td>\n",
       "      <td>54</td>\n",
       "      <td>7.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000834335.1_ASM83433v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>4.681411</td>\n",
       "      <td>1954.000000</td>\n",
       "      <td>[1953.9999999999918, 1954.0000000000082]</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>California</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>36.7015</td>\n",
       "      <td>-118.756</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>2.270539e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Shasta</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>1954</td>\n",
       "      <td>67</td>\n",
       "      <td>5.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000169635.1_ASM16963v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>27.879200</td>\n",
       "      <td>2005.000000</td>\n",
       "      <td>[2004.9999999999675, 2005.00000000003]</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>NA</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.330186e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>MG05-1020</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>28.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000170275.1_ASM17027v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>14.388214</td>\n",
       "      <td>1991.000000</td>\n",
       "      <td>[1990.999999999984, 1991.000000000016]</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>8.240116e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>F1991016</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1991</td>\n",
       "      <td>30</td>\n",
       "      <td>14.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE111</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.850724</td>\n",
       "      <td>2004.741212</td>\n",
       "      <td>[2001.2031002225738, 2007.743490769209]</td>\n",
       "      <td>Peru</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>24.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.684506e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE112</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.986264</td>\n",
       "      <td>2008.107821</td>\n",
       "      <td>[2005.6455957831777, 2009.904784379414]</td>\n",
       "      <td>Peru</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>96.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>1.382195e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE113</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.381419</td>\n",
       "      <td>2006.159421</td>\n",
       "      <td>[2003.379419273439, 2008.5135808224536]</td>\n",
       "      <td>Peru</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>41.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>2.024896e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE114</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.282061</td>\n",
       "      <td>2006.906261</td>\n",
       "      <td>[2004.18074830017, 2009.2828337803026]</td>\n",
       "      <td>Peru</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>39.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.853258e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE115</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.952827</td>\n",
       "      <td>2007.845012</td>\n",
       "      <td>[2005.6845492036232, 2009.6835662363621]</td>\n",
       "      <td>Peru</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>13.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.277932e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>233 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   node_type  branch_length  \\\n",
       "sample                                                        \n",
       "Reference                           terminal      14.290093   \n",
       "GCA_000834775.1_ASM83477v1_genomic  terminal       6.975785   \n",
       "GCA_000834335.1_ASM83433v1_genomic  terminal       4.681411   \n",
       "GCA_000169635.1_ASM16963v1_genomic  terminal      27.879200   \n",
       "GCA_000170275.1_ASM17027v1_genomic  terminal      14.388214   \n",
       "...                                      ...            ...   \n",
       "NODE111                             internal       1.850724   \n",
       "NODE112                             internal       2.986264   \n",
       "NODE113                             internal       1.381419   \n",
       "NODE114                             internal       1.282061   \n",
       "NODE115                             internal       0.952827   \n",
       "\n",
       "                                    timetree_num_date  \\\n",
       "sample                                                  \n",
       "Reference                                 1992.000000   \n",
       "GCA_000834775.1_ASM83477v1_genomic        1967.000000   \n",
       "GCA_000834335.1_ASM83433v1_genomic        1954.000000   \n",
       "GCA_000169635.1_ASM16963v1_genomic        2005.000000   \n",
       "GCA_000170275.1_ASM17027v1_genomic        1991.000000   \n",
       "...                                               ...   \n",
       "NODE111                                   2004.741212   \n",
       "NODE112                                   2008.107821   \n",
       "NODE113                                   2006.159421   \n",
       "NODE114                                   2006.906261   \n",
       "NODE115                                   2007.845012   \n",
       "\n",
       "                                                timetree_num_date_confidence  \\\n",
       "sample                                                                         \n",
       "Reference                             [1991.999999999996, 1992.000000000004]   \n",
       "GCA_000834775.1_ASM83477v1_genomic     [1966.999999999993, 1967.00000000001]   \n",
       "GCA_000834335.1_ASM83433v1_genomic  [1953.9999999999918, 1954.0000000000082]   \n",
       "GCA_000169635.1_ASM16963v1_genomic    [2004.9999999999675, 2005.00000000003]   \n",
       "GCA_000170275.1_ASM17027v1_genomic    [1990.999999999984, 1991.000000000016]   \n",
       "...                                                                      ...   \n",
       "NODE111                              [2001.2031002225738, 2007.743490769209]   \n",
       "NODE112                              [2005.6455957831777, 2009.904784379414]   \n",
       "NODE113                              [2003.379419273439, 2008.5135808224536]   \n",
       "NODE114                               [2004.18074830017, 2009.2828337803026]   \n",
       "NODE115                             [2005.6845492036232, 2009.6835662363621]   \n",
       "\n",
       "                                                     country    province  \\\n",
       "sample                                                                     \n",
       "Reference                           United States of America    Colorado   \n",
       "GCA_000834775.1_ASM83477v1_genomic  United States of America     Arizona   \n",
       "GCA_000834335.1_ASM83433v1_genomic  United States of America  California   \n",
       "GCA_000169635.1_ASM16963v1_genomic                Madagascar          NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic                     China      Yunnan   \n",
       "...                                                      ...         ...   \n",
       "NODE111                                                 Peru          NA   \n",
       "NODE112                                                 Peru          NA   \n",
       "NODE113                                                 Peru          NA   \n",
       "NODE114                                                 Peru          NA   \n",
       "NODE115                                                 Peru          NA   \n",
       "\n",
       "                                   country_lat country_lon province_lat  \\\n",
       "sample                                                                    \n",
       "Reference                              39.7837    -100.446      38.7252   \n",
       "GCA_000834775.1_ASM83477v1_genomic     39.7837    -100.446      34.3953   \n",
       "GCA_000834335.1_ASM83433v1_genomic     39.7837    -100.446      36.7015   \n",
       "GCA_000169635.1_ASM16963v1_genomic     -18.925     46.4416           NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic     35.0001         105           25   \n",
       "...                                        ...         ...          ...   \n",
       "NODE111                                     NA          NA           NA   \n",
       "NODE112                                     NA          NA           NA   \n",
       "NODE113                                     NA          NA           NA   \n",
       "NODE114                                     NA          NA           NA   \n",
       "NODE115                                     NA          NA           NA   \n",
       "\n",
       "                                   province_lon  branch_support  \\\n",
       "sample                                                            \n",
       "Reference                              -105.608             0.0   \n",
       "GCA_000834775.1_ASM83477v1_genomic     -111.763             0.0   \n",
       "GCA_000834335.1_ASM83433v1_genomic     -118.756             0.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic           NA             0.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic          102             0.0   \n",
       "...                                         ...             ...   \n",
       "NODE111                                      NA            24.0   \n",
       "NODE112                                      NA            96.0   \n",
       "NODE113                                      NA            41.0   \n",
       "NODE114                                      NA            39.0   \n",
       "NODE115                                      NA            13.0   \n",
       "\n",
       "                                   branch_support_conf_category  \\\n",
       "sample                                                            \n",
       "Reference                                                   LOW   \n",
       "GCA_000834775.1_ASM83477v1_genomic                          LOW   \n",
       "GCA_000834335.1_ASM83433v1_genomic                          LOW   \n",
       "GCA_000169635.1_ASM16963v1_genomic                          LOW   \n",
       "GCA_000170275.1_ASM17027v1_genomic                          LOW   \n",
       "...                                                         ...   \n",
       "NODE111                                                     LOW   \n",
       "NODE112                                                    HIGH   \n",
       "NODE113                                                     LOW   \n",
       "NODE114                                                     LOW   \n",
       "NODE115                                                     LOW   \n",
       "\n",
       "                                   branch_support_conf_char      rate_sub  \\\n",
       "sample                                                                      \n",
       "Reference                                                    4.231324e-07   \n",
       "GCA_000834775.1_ASM83477v1_genomic                           3.163987e-07   \n",
       "GCA_000834335.1_ASM83433v1_genomic                           2.270539e-07   \n",
       "GCA_000169635.1_ASM16963v1_genomic                           1.330186e-07   \n",
       "GCA_000170275.1_ASM17027v1_genomic                           8.240116e-08   \n",
       "...                                                     ...           ...   \n",
       "NODE111                                                      1.684506e-07   \n",
       "NODE112                                                   *  1.382195e-07   \n",
       "NODE113                                                      2.024896e-07   \n",
       "NODE114                                                      1.853258e-07   \n",
       "NODE115                                                      1.277932e-07   \n",
       "\n",
       "                                    rate_sub_year  branch_length_sub  \\\n",
       "sample                                                                 \n",
       "Reference                                     2.0               26.0   \n",
       "GCA_000834775.1_ASM83477v1_genomic            1.0                9.0   \n",
       "GCA_000834335.1_ASM83433v1_genomic            1.0                4.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic            1.0               16.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic            0.0                5.0   \n",
       "...                                           ...                ...   \n",
       "NODE111                                       1.0                1.0   \n",
       "NODE112                                       1.0                2.0   \n",
       "NODE113                                       1.0                1.0   \n",
       "NODE114                                       1.0                1.0   \n",
       "NODE115                                       1.0                1.0   \n",
       "\n",
       "                                       strain branch_major      biovar  \\\n",
       "sample                                                                   \n",
       "Reference                                CO92        1.ORI  Orientalis   \n",
       "GCA_000834775.1_ASM83477v1_genomic     Dodson        1.ORI  Orientalis   \n",
       "GCA_000834335.1_ASM83433v1_genomic     Shasta        1.ORI  Orientalis   \n",
       "GCA_000169635.1_ASM16963v1_genomic  MG05-1020        1.ORI  Orientalis   \n",
       "GCA_000170275.1_ASM17027v1_genomic   F1991016        1.ORI  Orientalis   \n",
       "...                                       ...          ...         ...   \n",
       "NODE111                                    NA           NA          NA   \n",
       "NODE112                                    NA           NA          NA   \n",
       "NODE113                                    NA           NA          NA   \n",
       "NODE114                                    NA           NA          NA   \n",
       "NODE115                                    NA           NA          NA   \n",
       "\n",
       "                                   host_human date_mean date_bp_mean  \\\n",
       "sample                                                                 \n",
       "Reference                               Human      1992           29   \n",
       "GCA_000834775.1_ASM83477v1_genomic      Human      1967           54   \n",
       "GCA_000834335.1_ASM83433v1_genomic      Human      1954           67   \n",
       "GCA_000169635.1_ASM16963v1_genomic      Human      2005           16   \n",
       "GCA_000170275.1_ASM17027v1_genomic  Non-Human      1991           30   \n",
       "...                                       ...       ...          ...   \n",
       "NODE111                                    NA        NA           NA   \n",
       "NODE112                                    NA        NA           NA   \n",
       "NODE113                                    NA        NA           NA   \n",
       "NODE114                                    NA        NA           NA   \n",
       "NODE115                                    NA        NA           NA   \n",
       "\n",
       "                                    branch_length_time blank  \n",
       "sample                                                        \n",
       "Reference                                         14.0        \n",
       "GCA_000834775.1_ASM83477v1_genomic                 7.0        \n",
       "GCA_000834335.1_ASM83433v1_genomic                 5.0        \n",
       "GCA_000169635.1_ASM16963v1_genomic                28.0        \n",
       "GCA_000170275.1_ASM17027v1_genomic                14.0        \n",
       "...                                                ...   ...  \n",
       "NODE111                                            2.0        \n",
       "NODE112                                            3.0        \n",
       "NODE113                                            1.0        \n",
       "NODE114                                            1.0        \n",
       "NODE115                                            1.0        \n",
       "\n",
       "[233 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.PRE\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>branch_length_time</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA5818830</th>\n",
       "      <td>terminal</td>\n",
       "      <td>2.258639</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1559.9999999999498, 1560.0000000000514]</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.439042e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>STN021</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>461</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818829</th>\n",
       "      <td>terminal</td>\n",
       "      <td>2.972765</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1559.99999999995, 1560.0000000000514]</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.368854e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>STN020</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>461</td>\n",
       "      <td>3.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818828</th>\n",
       "      <td>terminal</td>\n",
       "      <td>2.951818</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1559.99999999995, 1560.0000000000514]</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.335237e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>STN019</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>461</td>\n",
       "      <td>3.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818826</th>\n",
       "      <td>terminal</td>\n",
       "      <td>3.024750</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1559.9999999999498, 1560.0000000000514]</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.443847e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>STN014</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>461</td>\n",
       "      <td>3.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818825</th>\n",
       "      <td>terminal</td>\n",
       "      <td>2.258639</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1559.9999999999498, 1560.0000000000514]</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.172794e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>STN013</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>1560</td>\n",
       "      <td>461</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE34</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.526181</td>\n",
       "      <td>1552.856734</td>\n",
       "      <td>[1547.343805771212, 1557.7926399457879]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>13.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.700155e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE35</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.560544</td>\n",
       "      <td>1554.070167</td>\n",
       "      <td>[1548.899359250475, 1558.5969363982726]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.733455e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE36</th>\n",
       "      <td>internal</td>\n",
       "      <td>3.015103</td>\n",
       "      <td>1556.055502</td>\n",
       "      <td>[1551.3016283684117, 1559.5395201477922]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.328841e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE37</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.895886</td>\n",
       "      <td>1558.412227</td>\n",
       "      <td>[1555.0372533147577, 1559.9999012447547]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>16.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.326123e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE38</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.989116</td>\n",
       "      <td>1558.371627</td>\n",
       "      <td>[1554.9207043676165, 1559.9999586494305]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>16.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.481693e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             node_type  branch_length  timetree_num_date  \\\n",
       "sample                                                     \n",
       "SAMEA5818830  terminal       2.258639        1560.000000   \n",
       "SAMEA5818829  terminal       2.972765        1560.000000   \n",
       "SAMEA5818828  terminal       2.951818        1560.000000   \n",
       "SAMEA5818826  terminal       3.024750        1560.000000   \n",
       "SAMEA5818825  terminal       2.258639        1560.000000   \n",
       "...                ...            ...                ...   \n",
       "NODE34        internal       2.526181        1552.856734   \n",
       "NODE35        internal       2.560544        1554.070167   \n",
       "NODE36        internal       3.015103        1556.055502   \n",
       "NODE37        internal       2.895886        1558.412227   \n",
       "NODE38        internal       2.989116        1558.371627   \n",
       "\n",
       "                          timetree_num_date_confidence      country  \\\n",
       "sample                                                                \n",
       "SAMEA5818830  [1559.9999999999498, 1560.0000000000514]  Switzerland   \n",
       "SAMEA5818829    [1559.99999999995, 1560.0000000000514]  Switzerland   \n",
       "SAMEA5818828    [1559.99999999995, 1560.0000000000514]  Switzerland   \n",
       "SAMEA5818826  [1559.9999999999498, 1560.0000000000514]  Switzerland   \n",
       "SAMEA5818825  [1559.9999999999498, 1560.0000000000514]  Switzerland   \n",
       "...                                                ...          ...   \n",
       "NODE34         [1547.343805771212, 1557.7926399457879]           NA   \n",
       "NODE35         [1548.899359250475, 1558.5969363982726]           NA   \n",
       "NODE36        [1551.3016283684117, 1559.5395201477922]           NA   \n",
       "NODE37        [1555.0372533147577, 1559.9999012447547]           NA   \n",
       "NODE38        [1554.9207043676165, 1559.9999586494305]           NA   \n",
       "\n",
       "               province country_lat country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "SAMEA5818830  Nidwalden     46.8133     8.44495      46.9428      8.41198   \n",
       "SAMEA5818829  Nidwalden     46.8133     8.44495      46.9428      8.41198   \n",
       "SAMEA5818828  Nidwalden     46.8133     8.44495      46.9428      8.41198   \n",
       "SAMEA5818826  Nidwalden     46.8133     8.44495      46.9428      8.41198   \n",
       "SAMEA5818825  Nidwalden     46.8133     8.44495      46.9428      8.41198   \n",
       "...                 ...         ...         ...          ...          ...   \n",
       "NODE34               NA          NA          NA           NA           NA   \n",
       "NODE35               NA          NA          NA           NA           NA   \n",
       "NODE36               NA          NA          NA           NA           NA   \n",
       "NODE37               NA          NA          NA           NA           NA   \n",
       "NODE38               NA          NA          NA           NA           NA   \n",
       "\n",
       "              branch_support branch_support_conf_category  \\\n",
       "sample                                                      \n",
       "SAMEA5818830             0.0                          LOW   \n",
       "SAMEA5818829             0.0                          LOW   \n",
       "SAMEA5818828             0.0                          LOW   \n",
       "SAMEA5818826             0.0                          LOW   \n",
       "SAMEA5818825             0.0                          LOW   \n",
       "...                      ...                          ...   \n",
       "NODE34                  13.0                          LOW   \n",
       "NODE35                   6.0                          LOW   \n",
       "NODE36                   9.0                          LOW   \n",
       "NODE37                  16.0                          LOW   \n",
       "NODE38                  16.0                          LOW   \n",
       "\n",
       "             branch_support_conf_char      rate_sub  rate_sub_year  \\\n",
       "sample                                                               \n",
       "SAMEA5818830                           3.439042e-08            0.0   \n",
       "SAMEA5818829                           3.368854e-08            0.0   \n",
       "SAMEA5818828                           3.335237e-08            0.0   \n",
       "SAMEA5818826                           3.443847e-08            0.0   \n",
       "SAMEA5818825                           1.172794e-07            0.0   \n",
       "...                               ...           ...            ...   \n",
       "NODE34                                 3.700155e-08            0.0   \n",
       "NODE35                                 3.733455e-08            0.0   \n",
       "NODE36                                 3.328841e-08            0.0   \n",
       "NODE37                                 3.326123e-08            0.0   \n",
       "NODE38                                 3.481693e-08            0.0   \n",
       "\n",
       "              branch_length_sub  strain branch_major           biovar  \\\n",
       "sample                                                                  \n",
       "SAMEA5818830                0.0  STN021        1.PRE  Second Pandemic   \n",
       "SAMEA5818829                0.0  STN020        1.PRE  Second Pandemic   \n",
       "SAMEA5818828                0.0  STN019        1.PRE  Second Pandemic   \n",
       "SAMEA5818826                0.0  STN014        1.PRE  Second Pandemic   \n",
       "SAMEA5818825                1.0  STN013        1.PRE  Second Pandemic   \n",
       "...                         ...     ...          ...              ...   \n",
       "NODE34                      0.0      NA           NA               NA   \n",
       "NODE35                      0.0      NA           NA               NA   \n",
       "NODE36                      0.0      NA           NA               NA   \n",
       "NODE37                      0.0      NA           NA               NA   \n",
       "NODE38                      0.0      NA           NA               NA   \n",
       "\n",
       "             host_human date_mean date_bp_mean  branch_length_time blank  \n",
       "sample                                                                    \n",
       "SAMEA5818830      Human      1560          461                 2.0        \n",
       "SAMEA5818829      Human      1560          461                 3.0        \n",
       "SAMEA5818828      Human      1560          461                 3.0        \n",
       "SAMEA5818826      Human      1560          461                 3.0        \n",
       "SAMEA5818825      Human      1560          461                 2.0        \n",
       "...                 ...       ...          ...                 ...   ...  \n",
       "NODE34               NA        NA           NA                 3.0        \n",
       "NODE35               NA        NA           NA                 3.0        \n",
       "NODE36               NA        NA           NA                 3.0        \n",
       "NODE37               NA        NA           NA                 3.0        \n",
       "NODE38               NA        NA           NA                 3.0        \n",
       "\n",
       "[79 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.MED\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>branch_length_time</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>19.016925</td>\n",
       "      <td>1923.000000</td>\n",
       "      <td>[1922.9999999999575, 1923.000000000043]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.699646e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9_10</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>1923</td>\n",
       "      <td>98</td>\n",
       "      <td>19.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009296005.1_ASM929600v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>12.849306</td>\n",
       "      <td>1953.000000</td>\n",
       "      <td>[1952.9999999999568, 1953.0000000000425]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Chechnya</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.3976</td>\n",
       "      <td>45.6985</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>7.077325e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C-25</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1953</td>\n",
       "      <td>68</td>\n",
       "      <td>13.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630485.1_ASM863048v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>3.617000</td>\n",
       "      <td>1997.000000</td>\n",
       "      <td>[1996.9999999999566, 1997.0000000000427]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Kabardino-Balkaria</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.4428</td>\n",
       "      <td>43.4205</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>6.573191e-07</td>\n",
       "      <td>3.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>C-742</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1997</td>\n",
       "      <td>24</td>\n",
       "      <td>4.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630435.1_ASM863043v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>24.303254</td>\n",
       "      <td>1996.000000</td>\n",
       "      <td>[1995.999999999986, 1996.0000000000118]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>6.572837e-07</td>\n",
       "      <td>3.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>C-719</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1996</td>\n",
       "      <td>25</td>\n",
       "      <td>24.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630395.1_ASM863039v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>8.228695</td>\n",
       "      <td>1984.000000</td>\n",
       "      <td>[1983.999999999955, 1984.0000000000412]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Republic of Dagestan</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.0883</td>\n",
       "      <td>47.1499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.738636e-07</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>C-528</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1984</td>\n",
       "      <td>37</td>\n",
       "      <td>8.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE110</th>\n",
       "      <td>internal</td>\n",
       "      <td>8.987725</td>\n",
       "      <td>1989.038486</td>\n",
       "      <td>[1982.178482146981, 1995.4930958935863]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>1.284091e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE111</th>\n",
       "      <td>internal</td>\n",
       "      <td>3.602935</td>\n",
       "      <td>1993.383000</td>\n",
       "      <td>[1988.5689248739607, 1996.8665954433511]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>1.520278e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE112</th>\n",
       "      <td>internal</td>\n",
       "      <td>9.115426</td>\n",
       "      <td>1998.339317</td>\n",
       "      <td>[1988.7757888771998, 2007.4732478904593]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>73.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>2.652168e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE113</th>\n",
       "      <td>internal</td>\n",
       "      <td>12.984397</td>\n",
       "      <td>2009.038383</td>\n",
       "      <td>[2000.3328895240538, 2013.9971872384394]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>6.157763e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>13.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE114</th>\n",
       "      <td>internal</td>\n",
       "      <td>8.846033</td>\n",
       "      <td>2004.919598</td>\n",
       "      <td>[1996.802446804943, 2008.9999946606586]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>8.195099e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>231 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    node_type  branch_length  \\\n",
       "sample                                                         \n",
       "GCA_009909635.1_ASM990963v1_genomic  terminal      19.016925   \n",
       "GCA_009296005.1_ASM929600v1_genomic  terminal      12.849306   \n",
       "GCA_008630485.1_ASM863048v1_genomic  terminal       3.617000   \n",
       "GCA_008630435.1_ASM863043v1_genomic  terminal      24.303254   \n",
       "GCA_008630395.1_ASM863039v1_genomic  terminal       8.228695   \n",
       "...                                       ...            ...   \n",
       "NODE110                              internal       8.987725   \n",
       "NODE111                              internal       3.602935   \n",
       "NODE112                              internal       9.115426   \n",
       "NODE113                              internal      12.984397   \n",
       "NODE114                              internal       8.846033   \n",
       "\n",
       "                                     timetree_num_date  \\\n",
       "sample                                                   \n",
       "GCA_009909635.1_ASM990963v1_genomic        1923.000000   \n",
       "GCA_009296005.1_ASM929600v1_genomic        1953.000000   \n",
       "GCA_008630485.1_ASM863048v1_genomic        1997.000000   \n",
       "GCA_008630435.1_ASM863043v1_genomic        1996.000000   \n",
       "GCA_008630395.1_ASM863039v1_genomic        1984.000000   \n",
       "...                                                ...   \n",
       "NODE110                                    1989.038486   \n",
       "NODE111                                    1993.383000   \n",
       "NODE112                                    1998.339317   \n",
       "NODE113                                    2009.038383   \n",
       "NODE114                                    2004.919598   \n",
       "\n",
       "                                                 timetree_num_date_confidence  \\\n",
       "sample                                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic   [1922.9999999999575, 1923.000000000043]   \n",
       "GCA_009296005.1_ASM929600v1_genomic  [1952.9999999999568, 1953.0000000000425]   \n",
       "GCA_008630485.1_ASM863048v1_genomic  [1996.9999999999566, 1997.0000000000427]   \n",
       "GCA_008630435.1_ASM863043v1_genomic   [1995.999999999986, 1996.0000000000118]   \n",
       "GCA_008630395.1_ASM863039v1_genomic   [1983.999999999955, 1984.0000000000412]   \n",
       "...                                                                       ...   \n",
       "NODE110                               [1982.178482146981, 1995.4930958935863]   \n",
       "NODE111                              [1988.5689248739607, 1996.8665954433511]   \n",
       "NODE112                              [1988.7757888771998, 2007.4732478904593]   \n",
       "NODE113                              [2000.3328895240538, 2013.9971872384394]   \n",
       "NODE114                               [1996.802446804943, 2008.9999946606586]   \n",
       "\n",
       "                                    country              province country_lat  \\\n",
       "sample                                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic  Russia         Rostov Oblast     64.6863   \n",
       "GCA_009296005.1_ASM929600v1_genomic  Russia              Chechnya     64.6863   \n",
       "GCA_008630485.1_ASM863048v1_genomic  Russia    Kabardino-Balkaria     64.6863   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Russia   Karachay-Cherkessia     64.6863   \n",
       "GCA_008630395.1_ASM863039v1_genomic  Russia  Republic of Dagestan     64.6863   \n",
       "...                                     ...                   ...         ...   \n",
       "NODE110                                  NA                    NA          NA   \n",
       "NODE111                                  NA                    NA          NA   \n",
       "NODE112                                  NA                    NA          NA   \n",
       "NODE113                                  NA                    NA          NA   \n",
       "NODE114                                  NA                    NA          NA   \n",
       "\n",
       "                                    country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "GCA_009909635.1_ASM990963v1_genomic     97.7453      47.6222      40.7958   \n",
       "GCA_009296005.1_ASM929600v1_genomic     97.7453      43.3976      45.6985   \n",
       "GCA_008630485.1_ASM863048v1_genomic     97.7453      43.4428      43.4205   \n",
       "GCA_008630435.1_ASM863043v1_genomic     97.7453      43.7368      41.7268   \n",
       "GCA_008630395.1_ASM863039v1_genomic     97.7453      43.0883      47.1499   \n",
       "...                                         ...          ...          ...   \n",
       "NODE110                                      NA           NA           NA   \n",
       "NODE111                                      NA           NA           NA   \n",
       "NODE112                                      NA           NA           NA   \n",
       "NODE113                                      NA           NA           NA   \n",
       "NODE114                                      NA           NA           NA   \n",
       "\n",
       "                                     branch_support  \\\n",
       "sample                                                \n",
       "GCA_009909635.1_ASM990963v1_genomic             0.0   \n",
       "GCA_009296005.1_ASM929600v1_genomic             0.0   \n",
       "GCA_008630485.1_ASM863048v1_genomic             0.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic             0.0   \n",
       "GCA_008630395.1_ASM863039v1_genomic             0.0   \n",
       "...                                             ...   \n",
       "NODE110                                       100.0   \n",
       "NODE111                                       100.0   \n",
       "NODE112                                        73.0   \n",
       "NODE113                                       100.0   \n",
       "NODE114                                       100.0   \n",
       "\n",
       "                                    branch_support_conf_category  \\\n",
       "sample                                                             \n",
       "GCA_009909635.1_ASM990963v1_genomic                          LOW   \n",
       "GCA_009296005.1_ASM929600v1_genomic                          LOW   \n",
       "GCA_008630485.1_ASM863048v1_genomic                          LOW   \n",
       "GCA_008630435.1_ASM863043v1_genomic                          LOW   \n",
       "GCA_008630395.1_ASM863039v1_genomic                          LOW   \n",
       "...                                                          ...   \n",
       "NODE110                                                     HIGH   \n",
       "NODE111                                                     HIGH   \n",
       "NODE112                                                      LOW   \n",
       "NODE113                                                     HIGH   \n",
       "NODE114                                                     HIGH   \n",
       "\n",
       "                                    branch_support_conf_char      rate_sub  \\\n",
       "sample                                                                       \n",
       "GCA_009909635.1_ASM990963v1_genomic                           1.699646e-07   \n",
       "GCA_009296005.1_ASM929600v1_genomic                           7.077325e-08   \n",
       "GCA_008630485.1_ASM863048v1_genomic                           6.573191e-07   \n",
       "GCA_008630435.1_ASM863043v1_genomic                           6.572837e-07   \n",
       "GCA_008630395.1_ASM863039v1_genomic                           3.738636e-07   \n",
       "...                                                      ...           ...   \n",
       "NODE110                                                    *  1.284091e-07   \n",
       "NODE111                                                    *  1.520278e-07   \n",
       "NODE112                                                       2.652168e-08   \n",
       "NODE113                                                    *  6.157763e-08   \n",
       "NODE114                                                    *  8.195099e-08   \n",
       "\n",
       "                                     rate_sub_year  branch_length_sub strain  \\\n",
       "sample                                                                         \n",
       "GCA_009909635.1_ASM990963v1_genomic            1.0               14.0   9_10   \n",
       "GCA_009296005.1_ASM929600v1_genomic            0.0                4.0   C-25   \n",
       "GCA_008630485.1_ASM863048v1_genomic            3.0               10.0  C-742   \n",
       "GCA_008630435.1_ASM863043v1_genomic            3.0               68.0  C-719   \n",
       "GCA_008630395.1_ASM863039v1_genomic            2.0               13.0  C-528   \n",
       "...                                            ...                ...    ...   \n",
       "NODE110                                        1.0                5.0     NA   \n",
       "NODE111                                        1.0                2.0     NA   \n",
       "NODE112                                        0.0                1.0     NA   \n",
       "NODE113                                        0.0                3.0     NA   \n",
       "NODE114                                        0.0                3.0     NA   \n",
       "\n",
       "                                    branch_major      biovar host_human  \\\n",
       "sample                                                                    \n",
       "GCA_009909635.1_ASM990963v1_genomic        2.MED  Medievalis      Human   \n",
       "GCA_009296005.1_ASM929600v1_genomic        2.MED  Medievalis  Non-Human   \n",
       "GCA_008630485.1_ASM863048v1_genomic        2.MED  Medievalis  Non-Human   \n",
       "GCA_008630435.1_ASM863043v1_genomic        2.MED  Medievalis  Non-Human   \n",
       "GCA_008630395.1_ASM863039v1_genomic        2.MED  Medievalis  Non-Human   \n",
       "...                                          ...         ...        ...   \n",
       "NODE110                                       NA          NA         NA   \n",
       "NODE111                                       NA          NA         NA   \n",
       "NODE112                                       NA          NA         NA   \n",
       "NODE113                                       NA          NA         NA   \n",
       "NODE114                                       NA          NA         NA   \n",
       "\n",
       "                                    date_mean date_bp_mean  \\\n",
       "sample                                                       \n",
       "GCA_009909635.1_ASM990963v1_genomic      1923           98   \n",
       "GCA_009296005.1_ASM929600v1_genomic      1953           68   \n",
       "GCA_008630485.1_ASM863048v1_genomic      1997           24   \n",
       "GCA_008630435.1_ASM863043v1_genomic      1996           25   \n",
       "GCA_008630395.1_ASM863039v1_genomic      1984           37   \n",
       "...                                       ...          ...   \n",
       "NODE110                                    NA           NA   \n",
       "NODE111                                    NA           NA   \n",
       "NODE112                                    NA           NA   \n",
       "NODE113                                    NA           NA   \n",
       "NODE114                                    NA           NA   \n",
       "\n",
       "                                     branch_length_time blank  \n",
       "sample                                                         \n",
       "GCA_009909635.1_ASM990963v1_genomic                19.0        \n",
       "GCA_009296005.1_ASM929600v1_genomic                13.0        \n",
       "GCA_008630485.1_ASM863048v1_genomic                 4.0        \n",
       "GCA_008630435.1_ASM863043v1_genomic                24.0        \n",
       "GCA_008630395.1_ASM863039v1_genomic                 8.0        \n",
       "...                                                 ...   ...  \n",
       "NODE110                                             9.0        \n",
       "NODE111                                             4.0        \n",
       "NODE112                                             9.0        \n",
       "NODE113                                            13.0        \n",
       "NODE114                                             9.0        \n",
       "\n",
       "[231 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.ANT4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>branch_length_time</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA5661390</th>\n",
       "      <td>terminal</td>\n",
       "      <td>5.934884</td>\n",
       "      <td>478.000000</td>\n",
       "      <td>[477.9999999999499, 478.0000000000501]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.415710e-06</td>\n",
       "      <td>14.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>UNT004</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>478</td>\n",
       "      <td>1543</td>\n",
       "      <td>6.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661389</th>\n",
       "      <td>terminal</td>\n",
       "      <td>18.106782</td>\n",
       "      <td>516.000000</td>\n",
       "      <td>[515.9999999999488, 516.0000000000514]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>3.624344e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>UNT003</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>516</td>\n",
       "      <td>1505</td>\n",
       "      <td>18.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661385</th>\n",
       "      <td>terminal</td>\n",
       "      <td>18.990485</td>\n",
       "      <td>517.000000</td>\n",
       "      <td>[516.99999999995, 517.00000000005]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>6.547866e-07</td>\n",
       "      <td>3.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>PET004</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>517</td>\n",
       "      <td>1504</td>\n",
       "      <td>19.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661384</th>\n",
       "      <td>terminal</td>\n",
       "      <td>86.679955</td>\n",
       "      <td>521.000000</td>\n",
       "      <td>[520.99999999994, 521.00000000006]</td>\n",
       "      <td>Spain</td>\n",
       "      <td>Valencia Community</td>\n",
       "      <td>39.3261</td>\n",
       "      <td>-4.83798</td>\n",
       "      <td>39.682</td>\n",
       "      <td>-0.765441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.792881e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>VAL001</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>521</td>\n",
       "      <td>1500</td>\n",
       "      <td>87.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661372</th>\n",
       "      <td>terminal</td>\n",
       "      <td>47.054375</td>\n",
       "      <td>765.000000</td>\n",
       "      <td>[765.0, 765.0]</td>\n",
       "      <td>France</td>\n",
       "      <td>Centre-Loire Valley</td>\n",
       "      <td>46.6034</td>\n",
       "      <td>1.88833</td>\n",
       "      <td>47.549</td>\n",
       "      <td>1.73241</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.268521e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>LSD020</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>765</td>\n",
       "      <td>1256</td>\n",
       "      <td>47.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE6</th>\n",
       "      <td>internal</td>\n",
       "      <td>39.631380</td>\n",
       "      <td>500.919910</td>\n",
       "      <td>[474.467398194792, 516.7926070135841]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>93.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>2.400574e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>40.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE7</th>\n",
       "      <td>internal</td>\n",
       "      <td>220.065731</td>\n",
       "      <td>717.945625</td>\n",
       "      <td>[637.2782153708, 764.6461654102749]</td>\n",
       "      <td>France</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>2.018236e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>220.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE8</th>\n",
       "      <td>internal</td>\n",
       "      <td>9.737297</td>\n",
       "      <td>471.254535</td>\n",
       "      <td>[458.37663796541597, 477.97856227659]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>1.221871e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>10.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE9</th>\n",
       "      <td>internal</td>\n",
       "      <td>17.042735</td>\n",
       "      <td>488.863073</td>\n",
       "      <td>[475.22424383178804, 498.499935638548]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>85.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>5.343581e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>17.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE10</th>\n",
       "      <td>internal</td>\n",
       "      <td>15.094803</td>\n",
       "      <td>502.900385</td>\n",
       "      <td>[490.07212059615586, 511.49146655505706]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>65.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>7.056875e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>15.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             node_type  branch_length  timetree_num_date  \\\n",
       "sample                                                     \n",
       "SAMEA5661390  terminal       5.934884         478.000000   \n",
       "SAMEA5661389  terminal      18.106782         516.000000   \n",
       "SAMEA5661385  terminal      18.990485         517.000000   \n",
       "SAMEA5661384  terminal      86.679955         521.000000   \n",
       "SAMEA5661372  terminal      47.054375         765.000000   \n",
       "...                ...            ...                ...   \n",
       "NODE6         internal      39.631380         500.919910   \n",
       "NODE7         internal     220.065731         717.945625   \n",
       "NODE8         internal       9.737297         471.254535   \n",
       "NODE9         internal      17.042735         488.863073   \n",
       "NODE10        internal      15.094803         502.900385   \n",
       "\n",
       "                          timetree_num_date_confidence  country  \\\n",
       "sample                                                            \n",
       "SAMEA5661390    [477.9999999999499, 478.0000000000501]  Germany   \n",
       "SAMEA5661389    [515.9999999999488, 516.0000000000514]  Germany   \n",
       "SAMEA5661385        [516.99999999995, 517.00000000005]  Germany   \n",
       "SAMEA5661384        [520.99999999994, 521.00000000006]    Spain   \n",
       "SAMEA5661372                            [765.0, 765.0]   France   \n",
       "...                                                ...      ...   \n",
       "NODE6            [474.467398194792, 516.7926070135841]  Germany   \n",
       "NODE7              [637.2782153708, 764.6461654102749]   France   \n",
       "NODE8            [458.37663796541597, 477.97856227659]  Germany   \n",
       "NODE9           [475.22424383178804, 498.499935638548]  Germany   \n",
       "NODE10        [490.07212059615586, 511.49146655505706]  Germany   \n",
       "\n",
       "                         province country_lat country_lon province_lat  \\\n",
       "sample                                                                   \n",
       "SAMEA5661390              Bavaria     51.0834     10.4234      48.9468   \n",
       "SAMEA5661389              Bavaria     51.0834     10.4234      48.9468   \n",
       "SAMEA5661385              Bavaria     51.0834     10.4234      48.9468   \n",
       "SAMEA5661384   Valencia Community     39.3261    -4.83798       39.682   \n",
       "SAMEA5661372  Centre-Loire Valley     46.6034     1.88833       47.549   \n",
       "...                           ...         ...         ...          ...   \n",
       "NODE6                          NA          NA          NA           NA   \n",
       "NODE7                          NA          NA          NA           NA   \n",
       "NODE8                          NA          NA          NA           NA   \n",
       "NODE9                          NA          NA          NA           NA   \n",
       "NODE10                         NA          NA          NA           NA   \n",
       "\n",
       "             province_lon  branch_support branch_support_conf_category  \\\n",
       "sample                                                                   \n",
       "SAMEA5661390      11.4039             0.0                          LOW   \n",
       "SAMEA5661389      11.4039             0.0                          LOW   \n",
       "SAMEA5661385      11.4039             0.0                          LOW   \n",
       "SAMEA5661384    -0.765441             0.0                          LOW   \n",
       "SAMEA5661372      1.73241             0.0                          LOW   \n",
       "...                   ...             ...                          ...   \n",
       "NODE6                  NA            93.0                          LOW   \n",
       "NODE7                  NA           100.0                         HIGH   \n",
       "NODE8                  NA           100.0                         HIGH   \n",
       "NODE9                  NA            85.0                          LOW   \n",
       "NODE10                 NA            65.0                          LOW   \n",
       "\n",
       "             branch_support_conf_char      rate_sub  rate_sub_year  \\\n",
       "sample                                                               \n",
       "SAMEA5661390                           3.415710e-06           14.0   \n",
       "SAMEA5661389                           3.624344e-09            0.0   \n",
       "SAMEA5661385                           6.547866e-07            3.0   \n",
       "SAMEA5661384                           1.792881e-08            0.0   \n",
       "SAMEA5661372                           1.268521e-07            1.0   \n",
       "...                               ...           ...            ...   \n",
       "NODE6                                  2.400574e-09            0.0   \n",
       "NODE7                               *  2.018236e-08            0.0   \n",
       "NODE8                               *  1.221871e-07            1.0   \n",
       "NODE9                                  5.343581e-09            0.0   \n",
       "NODE10                                 7.056875e-09            0.0   \n",
       "\n",
       "              branch_length_sub  strain branch_major   biovar host_human  \\\n",
       "sample                                                                     \n",
       "SAMEA5661390               86.0  UNT004        0.ANT  Antiqua      Human   \n",
       "SAMEA5661389                0.0  UNT003        0.ANT  Antiqua      Human   \n",
       "SAMEA5661385               53.0  PET004        0.ANT  Antiqua      Human   \n",
       "SAMEA5661384                7.0  VAL001        0.ANT  Antiqua      Human   \n",
       "SAMEA5661372               25.0  LSD020        0.ANT  Antiqua      Human   \n",
       "...                         ...     ...          ...      ...        ...   \n",
       "NODE6                       0.0      NA           NA       NA         NA   \n",
       "NODE7                      19.0      NA           NA       NA         NA   \n",
       "NODE8                       5.0      NA           NA       NA         NA   \n",
       "NODE9                       0.0      NA           NA       NA         NA   \n",
       "NODE10                      0.0      NA           NA       NA         NA   \n",
       "\n",
       "             date_mean date_bp_mean  branch_length_time blank  \n",
       "sample                                                         \n",
       "SAMEA5661390       478         1543                 6.0        \n",
       "SAMEA5661389       516         1505                18.0        \n",
       "SAMEA5661385       517         1504                19.0        \n",
       "SAMEA5661384       521         1500                87.0        \n",
       "SAMEA5661372       765         1256                47.0        \n",
       "...                ...          ...                 ...   ...  \n",
       "NODE6               NA           NA                40.0        \n",
       "NODE7               NA           NA               220.0        \n",
       "NODE8               NA           NA                10.0        \n",
       "NODE9               NA           NA                17.0        \n",
       "NODE10              NA           NA                15.0        \n",
       "\n",
       "[23 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.PRE\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>branch_support_conf_category</th>\n",
       "      <th>branch_support_conf_char</th>\n",
       "      <th>rate_sub</th>\n",
       "      <th>rate_sub_year</th>\n",
       "      <th>branch_length_sub</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>branch_length_time</th>\n",
       "      <th>blank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA104233048</th>\n",
       "      <td>terminal</td>\n",
       "      <td>140.358180</td>\n",
       "      <td>-1944.500000</td>\n",
       "      <td>[-1944.5, -1944.5]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.668537e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Post6</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>-1944.5</td>\n",
       "      <td>3965.5</td>\n",
       "      <td>140.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233049</th>\n",
       "      <td>terminal</td>\n",
       "      <td>11.894365</td>\n",
       "      <td>-2272.000000</td>\n",
       "      <td>[-2272.00000000008, -2271.999999999927]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.706801e-06</td>\n",
       "      <td>7.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>1343UnTal85</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>-2272</td>\n",
       "      <td>4293</td>\n",
       "      <td>12.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233047</th>\n",
       "      <td>terminal</td>\n",
       "      <td>244.138160</td>\n",
       "      <td>-2457.000000</td>\n",
       "      <td>[-2457.00000000011, -2456.99999999989]</td>\n",
       "      <td>Estonia</td>\n",
       "      <td>Pärnu maakond</td>\n",
       "      <td>58.7524</td>\n",
       "      <td>25.3319</td>\n",
       "      <td>58.3195</td>\n",
       "      <td>24.3026</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.668537e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>KunilaII</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>-2457</td>\n",
       "      <td>4478</td>\n",
       "      <td>244.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>terminal</td>\n",
       "      <td>198.228985</td>\n",
       "      <td>-2546.500000</td>\n",
       "      <td>[-2546.500000000111, -2546.499999999885]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>55.35</td>\n",
       "      <td>23.75</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.668537e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Gyvakarai1</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>-2546.5</td>\n",
       "      <td>4567.5</td>\n",
       "      <td>198.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>terminal</td>\n",
       "      <td>33.455487</td>\n",
       "      <td>-2776.500000</td>\n",
       "      <td>[-2776.5000000001, -2776.499999999893]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td></td>\n",
       "      <td>1.518149e-06</td>\n",
       "      <td>6.0</td>\n",
       "      <td>215.0</td>\n",
       "      <td>RISE509</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>-2776.5</td>\n",
       "      <td>4797.5</td>\n",
       "      <td>33.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>internal</td>\n",
       "      <td>47.574086</td>\n",
       "      <td>-2700.108352</td>\n",
       "      <td>[-2748.16251438543, -2668.2079570299275]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>99.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>1.337582e-06</td>\n",
       "      <td>6.0</td>\n",
       "      <td>269.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>48.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE3</th>\n",
       "      <td>internal</td>\n",
       "      <td>13.587994</td>\n",
       "      <td>-2686.600603</td>\n",
       "      <td>[-2729.2763547802, -2666.6124803596003]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>99.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>6.735803e-07</td>\n",
       "      <td>3.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>14.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>internal</td>\n",
       "      <td>16.606139</td>\n",
       "      <td>-2671.275464</td>\n",
       "      <td>[-2682.524693830586, -2666.500584737746]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>7.313060e-07</td>\n",
       "      <td>3.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>17.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>internal</td>\n",
       "      <td>404.708874</td>\n",
       "      <td>-2283.894365</td>\n",
       "      <td>[-2313.4359420323, -2272.014964033]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>1.668537e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>405.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE6</th>\n",
       "      <td>internal</td>\n",
       "      <td>199.036185</td>\n",
       "      <td>-2084.858180</td>\n",
       "      <td>[-2127.595523122, -2043.0449166393]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>*</td>\n",
       "      <td>9.335778e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>199.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               node_type  branch_length  timetree_num_date  \\\n",
       "sample                                                       \n",
       "SAMEA104233048  terminal     140.358180       -1944.500000   \n",
       "SAMEA104233049  terminal      11.894365       -2272.000000   \n",
       "SAMEA104233047  terminal     244.138160       -2457.000000   \n",
       "SAMEA104233046  terminal     198.228985       -2546.500000   \n",
       "SAMEA3541827    terminal      33.455487       -2776.500000   \n",
       "...                  ...            ...                ...   \n",
       "NODE2           internal      47.574086       -2700.108352   \n",
       "NODE3           internal      13.587994       -2686.600603   \n",
       "NODE4           internal      16.606139       -2671.275464   \n",
       "NODE5           internal     404.708874       -2283.894365   \n",
       "NODE6           internal     199.036185       -2084.858180   \n",
       "\n",
       "                            timetree_num_date_confidence    country  \\\n",
       "sample                                                                \n",
       "SAMEA104233048                        [-1944.5, -1944.5]    Germany   \n",
       "SAMEA104233049   [-2272.00000000008, -2271.999999999927]    Germany   \n",
       "SAMEA104233047    [-2457.00000000011, -2456.99999999989]    Estonia   \n",
       "SAMEA104233046  [-2546.500000000111, -2546.499999999885]  Lithuania   \n",
       "SAMEA3541827      [-2776.5000000001, -2776.499999999893]     Russia   \n",
       "...                                                  ...        ...   \n",
       "NODE2           [-2748.16251438543, -2668.2079570299275]         NA   \n",
       "NODE3            [-2729.2763547802, -2666.6124803596003]         NA   \n",
       "NODE4           [-2682.524693830586, -2666.500584737746]         NA   \n",
       "NODE5                [-2313.4359420323, -2272.014964033]         NA   \n",
       "NODE6                [-2127.595523122, -2043.0449166393]         NA   \n",
       "\n",
       "                        province country_lat country_lon province_lat  \\\n",
       "sample                                                                  \n",
       "SAMEA104233048           Bavaria     51.0834     10.4234      48.9468   \n",
       "SAMEA104233049           Bavaria     51.0834     10.4234      48.9468   \n",
       "SAMEA104233047     Pärnu maakond     58.7524     25.3319      58.3195   \n",
       "SAMEA104233046  Panevezys County       55.35       23.75      55.9156   \n",
       "SAMEA3541827    Krasnoyarsk Krai     64.6863     97.7453      63.3234   \n",
       "...                          ...         ...         ...          ...   \n",
       "NODE2                         NA          NA          NA           NA   \n",
       "NODE3                         NA          NA          NA           NA   \n",
       "NODE4                         NA          NA          NA           NA   \n",
       "NODE5                         NA          NA          NA           NA   \n",
       "NODE6                         NA          NA          NA           NA   \n",
       "\n",
       "               province_lon  branch_support branch_support_conf_category  \\\n",
       "sample                                                                     \n",
       "SAMEA104233048      11.4039             0.0                          LOW   \n",
       "SAMEA104233049      11.4039             0.0                          LOW   \n",
       "SAMEA104233047      24.3026             0.0                          LOW   \n",
       "SAMEA104233046      25.0312             0.0                          LOW   \n",
       "SAMEA3541827         97.098             0.0                          LOW   \n",
       "...                     ...             ...                          ...   \n",
       "NODE2                    NA            99.0                         HIGH   \n",
       "NODE3                    NA            99.0                         HIGH   \n",
       "NODE4                    NA           100.0                         HIGH   \n",
       "NODE5                    NA           100.0                         HIGH   \n",
       "NODE6                    NA           100.0                         HIGH   \n",
       "\n",
       "               branch_support_conf_char      rate_sub  rate_sub_year  \\\n",
       "sample                                                                 \n",
       "SAMEA104233048                           1.668537e-08            0.0   \n",
       "SAMEA104233049                           1.706801e-06            7.0   \n",
       "SAMEA104233047                           1.668537e-08            0.0   \n",
       "SAMEA104233046                           1.668537e-08            0.0   \n",
       "SAMEA3541827                             1.518149e-06            6.0   \n",
       "...                                 ...           ...            ...   \n",
       "NODE2                                 *  1.337582e-06            6.0   \n",
       "NODE3                                 *  6.735803e-07            3.0   \n",
       "NODE4                                 *  7.313060e-07            3.0   \n",
       "NODE5                                 *  1.668537e-08            0.0   \n",
       "NODE6                                 *  9.335778e-08            0.0   \n",
       "\n",
       "                branch_length_sub       strain branch_major      biovar  \\\n",
       "sample                                                                    \n",
       "SAMEA104233048               10.0        Post6        0.PRE  Bronze Age   \n",
       "SAMEA104233049               86.0  1343UnTal85        0.PRE  Bronze Age   \n",
       "SAMEA104233047               17.0     KunilaII        0.PRE  Bronze Age   \n",
       "SAMEA104233046               14.0   Gyvakarai1        0.PRE  Bronze Age   \n",
       "SAMEA3541827                215.0      RISE509        0.PRE  Bronze Age   \n",
       "...                           ...          ...          ...         ...   \n",
       "NODE2                       269.0           NA           NA          NA   \n",
       "NODE3                        39.0           NA           NA          NA   \n",
       "NODE4                        51.0           NA           NA          NA   \n",
       "NODE5                        29.0           NA           NA          NA   \n",
       "NODE6                        79.0           NA           NA          NA   \n",
       "\n",
       "               host_human date_mean date_bp_mean  branch_length_time blank  \n",
       "sample                                                                      \n",
       "SAMEA104233048      Human   -1944.5       3965.5               140.0        \n",
       "SAMEA104233049      Human     -2272         4293                12.0        \n",
       "SAMEA104233047      Human     -2457         4478               244.0        \n",
       "SAMEA104233046      Human   -2546.5       4567.5               198.0        \n",
       "SAMEA3541827        Human   -2776.5       4797.5                33.0        \n",
       "...                   ...       ...          ...                 ...   ...  \n",
       "NODE2                  NA        NA           NA                48.0        \n",
       "NODE3                  NA        NA           NA                14.0        \n",
       "NODE4                  NA        NA           NA                17.0        \n",
       "NODE5                  NA        NA           NA               405.0        \n",
       "NODE6                  NA        NA           NA               199.0        \n",
       "\n",
       "[15 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    print(branch)\n",
    "    \n",
    "    columns = [\n",
    "        # Node type is mandatorya\n",
    "        \"node_type\",        \n",
    "        # Draw Divergence Tree\n",
    "        \"length\",\n",
    "        # Draw Time Tree\n",
    "        \"timetree_num_date\",\n",
    "        \"timetree_num_date_confidence\",  \n",
    "        # Geo\n",
    "        \"country\",\n",
    "        \"province\",\n",
    "        \"country_lat\",\n",
    "        \"country_lon\",\n",
    "        \"province_lat\",\n",
    "        \"province_lon\",           \n",
    "        # Stats\n",
    "        \"posterior\",\n",
    "        \"branch_support_conf_category\",\n",
    "        \"branch_support_conf_char\",\n",
    "        \"rate\",\n",
    "        \"rate_sub_year\",\n",
    "        \"branch_length_sub\",\n",
    "        # Text descriptions\n",
    "        \"strain\",\n",
    "        \"branch_major\",\n",
    "        \"biovar\",\n",
    "        \"host_human\",\n",
    "        \"date_mean\",\n",
    "        \"date_bp_mean\",        \n",
    "    ]\n",
    "    \n",
    "    auspice_df = copy.copy(tree_dict[branch][\"df\"][columns])\n",
    "    \n",
    "    # Round For Pretty Numbers\n",
    "    posteriors = [float(round(p)) for p in list(auspice_df[\"posterior\"])]\n",
    "    auspice_df[\"posterior\"] = posteriors  \n",
    "    \n",
    "    branch_length_times = [float(round(l)) for l in list(auspice_df[\"length\"])] \n",
    "    auspice_df[\"branch_length_time\"] = branch_length_times\n",
    "    \n",
    "    branch_length_subs = [float(round(l)) for l in list(auspice_df[\"branch_length_sub\"])] \n",
    "    auspice_df[\"branch_length_sub\"] = branch_length_subs \n",
    "\n",
    "    rate_sub_years = [float(round(s)) for s in list(auspice_df[\"rate_sub_year\"])] \n",
    "    auspice_df[\"rate_sub_year\"] = rate_sub_years\n",
    "    \n",
    "    auspice_df.rename(columns={\"length\": \"branch_length\"}, inplace=True)\n",
    "    auspice_df.rename(columns={\"posterior\": \"branch_support\"}, inplace=True)\n",
    "    auspice_df.rename(columns={\"rate\": \"rate_sub\"}, inplace=True)\n",
    "    \n",
    "    # Add blank column\n",
    "    auspice_df[\"blank\"] = [\" \"] * len(auspice_df)\n",
    "\n",
    "    \n",
    "    tree_dict[branch][\"auspice_df\"] = auspice_df\n",
    "    display(tree_dict[branch][\"auspice_df\"])\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "integrated-development",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "municipal-dubai",
   "metadata": {},
   "source": [
    "## Time Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "detected-marketing",
   "metadata": {},
   "outputs": [],
   "source": [
    "for branch in tree_dict:\n",
    "    out_timetree = copy.deepcopy(tree_dict[branch][\"tree\"])\n",
    "\n",
    "    metadata_to_comment(out_timetree, tree_dict[branch][\"df\"])    \n",
    "    out_timetree_nex_path = os.path.join(tree_dir, branch + \".timetree.nex\")\n",
    "    Phylo.write(out_timetree, out_timetree_nex_path, \"nexus\")\n",
    "\n",
    "    for c in out_timetree.find_clades():\n",
    "        c.comment = None\n",
    "\n",
    "    out_timetree_nwk_path = os.path.join(tree_dir, branch + \".timetree.nwk\")\n",
    "    Phylo.write(out_timetree, out_timetree_nwk_path, \"newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aware-extraction",
   "metadata": {},
   "source": [
    "## Augur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "hundred-difference",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1867.0942100091504, 'num_date_confidence': [1811.2040243167871, 1907.3711522232752], 'country': 'China', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1274.5194429677265, 'num_date_confidence': [1215.4213839601462, 1318.87619038616], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1809.6838095526111, 'num_date_confidence': [1717.3801696916983, 1879.783064842695], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 173.4771079789549, 'num_date_confidence': [49.12360336217796, 237.49995201676995], 'country': 'Kyrgyzstan', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': -2814.6085616949904, 'num_date_confidence': [-2908.780417379524, -2776.5003057288377], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    \n",
    "    augur_dict = augur_export(\n",
    "        tree_path=None, \n",
    "        aln_path=None,  \n",
    "        tree=tree_dict[branch][\"tree\"], \n",
    "        tree_df=tree_dict[branch][\"auspice_df\"], \n",
    "        color_keyword_exclude=[\"geometry\"],\n",
    "        type_convert = {\n",
    "            \"branch_number\" : (lambda x : str(x))\n",
    "        },\n",
    "    )\n",
    "    \n",
    "    tree_dict[branch][\"augur_dict\"] = augur_dict\n",
    "    \n",
    "    first_taxa = list(augur_dict[\"nodes\"].keys())[0]\n",
    "    print(augur_dict[\"nodes\"][first_taxa])\n",
    "\n",
    "    out_path_augur_json = os.path.join(tree_dir, branch + \"_augur.json\" )\n",
    "    utils.write_json(data=tree_dict[branch][\"augur_dict\"], file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "    tree_dict[branch][\"augur_json_path\"] = out_path_augur_json\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "peaceful-concern",
   "metadata": {},
   "source": [
    "## Auspice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aquatic-entry",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/1.ORI.json'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/1.ORI.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/1.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/1.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/2.MED.json'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/2.MED.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/0.ANT4.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/0.ANT4.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/0.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/phylogeography/0.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    \n",
    "    # Store the color\n",
    "    if branch == \"0.ANT4\":\n",
    "        branch_major_color = colors_dict[\"branch_major\"][\"0.ANT\"]\n",
    "    else:\n",
    "        branch_major_color = colors_dict[\"branch_major\"][branch]\n",
    "\n",
    "    auspice_dict = auspice_export(\n",
    "        tree=tree_dict[branch][\"tree\"],\n",
    "        augur_json_paths=tree_dict[branch][\"augur_json_path\"], \n",
    "        auspice_config_path=auspice_config_path, \n",
    "        auspice_colors_path=out_path_colors,\n",
    "        auspice_latlons_path=out_path_latlon, \n",
    "        )\n",
    "\n",
    "\n",
    "    label_col = list(tree_dict[branch][\"auspice_df\"])\n",
    "    print(label_col)\n",
    "\n",
    "    # Recursively add branch attrs\n",
    "    branch_attributes(\n",
    "        tree_dict=auspice_dict[\"tree\"], \n",
    "        sub_dict=auspice_dict[\"tree\"], \n",
    "        df=tree_dict[branch][\"auspice_df\"],\n",
    "        label_col=label_col,\n",
    "        )\n",
    "    \n",
    "    \n",
    "    # Last manual changes\n",
    "    auspice_dict_copy = copy.deepcopy(auspice_dict)\n",
    "    for i in range(0, len(auspice_dict_copy[\"meta\"][\"colorings\"])):\n",
    "        coloring = auspice_dict_copy[\"meta\"][\"colorings\"][i]\n",
    "        for key in coloring:\n",
    "            # Node type as internal or terminal\n",
    "            if coloring[key] == \"node_type\":\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['internal', '#FFFFFF'], ['terminal', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            # Confidence category\n",
    "            if \"conf_category\" in coloring[key]:\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['LOW', '#FFFFFF'], ['HIGH', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            # Host Human binary\n",
    "            if \"host_human\" in coloring[key]:\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['Human', '#CBB742'], ['Non-Human', \"#60B6F2\"], ['NA', \"#D6D6D6\"]]\n",
    "\n",
    "    # Write outputs - For Local Rendering\n",
    "    out_path_auspice_local_json = os.path.join(tree_dir, branch + \".json\" )\n",
    "    utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "    export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "    print(\"Validation successful for local JSON.\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "regional-inquiry",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
