{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "proprietary-spice",
   "metadata": {},
   "source": [
    "# BEAST Analysis Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "gross-cutting",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 0. SETUP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "expanded-saudi",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from Bio import Phylo, AlignIO\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "connected-livestock",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "removable-vulnerability",
   "metadata": {},
   "outputs": [],
   "source": [
    "log_dir  = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade\"\n",
    "tree_dir = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock\"\n",
    "metadata_path = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/iqtree/all/chromosome/full/filter5/filter-taxa/metadata.tsv\"\n",
    "beast_dir = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade\"\n",
    "\n",
    "auspice_config_path = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json\"\n",
    "out_path_colors = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/augur/all/chromosome/full/filter5/colors.tsv\"\n",
    "out_path_latlon = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/augur/all/chromosome/full/filter5/latlon.tsv\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "constant_sites_path   = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/snippy_multi/all/chromosome/full/snippy-multi.constant_sites.txt\"\n",
    "aln_path              = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/iqtree/all/chromosome/full/filter5/filter-sites/snippy-multi.snps.aln\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ancient-syria",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "linear-burden",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------------------------------------------\n",
    "BRANCH_LIST = {\n",
    "    \"1.ORI\" : [\"1.ORI1\", \"1.ORI2\", \"1.ORI3\"],\n",
    "    \"1.IN\": [\"1.IN1\",\"1.IN2\",\"1.IN3\"],  \n",
    "    \"1.ANT\": [\"1.ANT1\"], \n",
    "    \"1.PRE\" : [\"1.PRE0\",\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],\n",
    "    \"2.MED\": [\"2.MED0\", \"2.MED1\",\"2.MED2\",\"2.MED3\" ],      \n",
    "    \"2.ANT\": [\"2.ANT1\",\"2.ANT2\",\"2.ANT3\" ],    \n",
    "    \"4.ANT\": [\"4.ANT1\" ],       \n",
    "    \"3.ANT\": [\"3.ANT1\", \"3.ANT2\" ],  \n",
    "    \"0.ANT\": [\"0.ANT1\", \"0.ANT2\",\"0.ANT3\",\"0.ANT5\"],         \n",
    "    \"0.ANT4\" : [\"0.ANT4\"], \n",
    "    \"0.PE\": [\"0.PE2\", \"0.PE4m\", \"0.PE4m\", \"0.PE4t\", \"0.PE4a\", \"0.PE5\", \"0.PE7\", \"0.PE8\", \"0.PE10\"],   \n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"],        \n",
    "}\n",
    "\n",
    "NUM_STATES = 10\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "JSON_INDENT=2\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "with open(constant_sites_path) as infile:\n",
    "    data = infile.read().strip().split(\",\")\n",
    "    constant_sites = sum([int(count) for count in data])\n",
    "\n",
    "aln = AlignIO.read(aln_path, \"fasta\")\n",
    "variant_sites = len(aln[0].seq)\n",
    "SEQ_LEN = constant_sites + variant_sites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mechanical-monte",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 1. IMPORT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "convertible-freeze",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "beneficial-parcel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>...</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313243_45</th>\n",
       "      <td>Azov38</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMEA7313243_45</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313246_49</th>\n",
       "      <td>Gdansk8</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Poland</td>\n",
       "      <td>Pomeranian Voivodeship</td>\n",
       "      <td>52.215933</td>\n",
       "      <td>19.134422</td>\n",
       "      <td>54.2456</td>\n",
       "      <td>18.1099</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMEA7313246_49</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>54.245560</td>\n",
       "      <td>18.109900</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6651390</th>\n",
       "      <td>AGU010</td>\n",
       "      <td>[1435:1477]</td>\n",
       "      <td>[-586:-544]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMEA6651390</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1456.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637004</th>\n",
       "      <td>AGU025</td>\n",
       "      <td>[1441:1612]</td>\n",
       "      <td>[-580:-409]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMEA6637004</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1526.5</td>\n",
       "      <td>494.5</td>\n",
       "      <td>85.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637002</th>\n",
       "      <td>AGU007B</td>\n",
       "      <td>[1463:1632]</td>\n",
       "      <td>[-558:-389]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMEA6637002</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1547.5</td>\n",
       "      <td>473.5</td>\n",
       "      <td>84.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      strain         date      date_bp  \\\n",
       "sample                                                                   \n",
       "Reference                               CO92         1992          -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic     9_10       1923.0          -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic    42126       2006.0          -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic    42123       2005.0          -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic    42118       2005.0          -16   \n",
       "...                                      ...          ...          ...   \n",
       "SAMEA7313243_45                       Azov38  [1400:1700]  [-621:-321]   \n",
       "SAMEA7313246_49                      Gdansk8  [1400:1700]  [-621:-321]   \n",
       "SAMEA6651390                          AGU010  [1435:1477]  [-586:-544]   \n",
       "SAMEA6637004                          AGU025  [1441:1612]  [-580:-409]   \n",
       "SAMEA6637002                         AGU007B  [1463:1632]  [-558:-389]   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "SAMEA7313243_45                                        Russia   \n",
       "SAMEA7313246_49                                        Poland   \n",
       "SAMEA6651390                                        Lithuania   \n",
       "SAMEA6637004                                        Lithuania   \n",
       "SAMEA6637002                                        Lithuania   \n",
       "\n",
       "                                                   province  country_lat  \\\n",
       "sample                                                                     \n",
       "Reference                                          Colorado    39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic           Rostov Oblast    64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic                Xinjiang    35.000074   \n",
       "...                                                     ...          ...   \n",
       "SAMEA7313243_45                               Rostov Oblast    64.686314   \n",
       "SAMEA7313246_49                      Pomeranian Voivodeship    52.215933   \n",
       "SAMEA6651390                                 Vilnius County    55.350000   \n",
       "SAMEA6637004                                 Vilnius County    55.350000   \n",
       "SAMEA6637002                                 Vilnius County    55.350000   \n",
       "\n",
       "                                     country_lon province_lat province_lon  \\\n",
       "sample                                                                       \n",
       "Reference                            -100.445882      38.7252     -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic    97.745306      47.6222      40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic   104.999927      42.4805      85.4633   \n",
       "...                                          ...          ...          ...   \n",
       "SAMEA7313243_45                        97.745306      47.6222      40.7958   \n",
       "SAMEA7313246_49                        19.134422      54.2456      18.1099   \n",
       "SAMEA6651390                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637004                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637002                           23.750000      54.8227      25.2495   \n",
       "\n",
       "                                              biovar  ... biosample_accession  \\\n",
       "sample                                                ...                       \n",
       "Reference                                 Orientalis  ...        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       Medievalis  ...        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic          Antiqua  ...        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic          Antiqua  ...        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic          Antiqua  ...        SAMN07722923   \n",
       "...                                              ...  ...                 ...   \n",
       "SAMEA7313243_45                      Second Pandemic  ...     SAMEA7313243_45   \n",
       "SAMEA7313246_49                      Second Pandemic  ...     SAMEA7313246_49   \n",
       "SAMEA6651390                         Second Pandemic  ...        SAMEA6651390   \n",
       "SAMEA6637004                         Second Pandemic  ...        SAMEA6637004   \n",
       "SAMEA6637002                         Second Pandemic  ...        SAMEA6637002   \n",
       "\n",
       "                                                     biosample_comment  \\\n",
       "sample                                                                   \n",
       "Reference                              KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic              KEEP: Assembly Modern   \n",
       "...                                                                ...   \n",
       "SAMEA7313243_45                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA7313246_49                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA6651390                                         KEEP: SRA Ancient   \n",
       "SAMEA6637004                                         KEEP: SRA Ancient   \n",
       "SAMEA6637002                                         KEEP: SRA Ancient   \n",
       "\n",
       "                                    branch_number      continent  date_mean  \\\n",
       "sample                                                                        \n",
       "Reference                                       1  North America     1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic             2         Europe     1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0           Asia     2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0           Asia     2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic             0           Asia     2005.0   \n",
       "...                                           ...            ...        ...   \n",
       "SAMEA7313243_45                                 1         Europe     1550.0   \n",
       "SAMEA7313246_49                                 1         Europe     1550.0   \n",
       "SAMEA6651390                                    1         Europe     1456.0   \n",
       "SAMEA6637004                                    1         Europe     1526.5   \n",
       "SAMEA6637002                                    1         Europe     1547.5   \n",
       "\n",
       "                                    date_bp_mean  date_err        lat  \\\n",
       "sample                                                                  \n",
       "Reference                                   29.0       0.0  38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic         98.0       0.0  47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic         15.0       0.0  42.480495   \n",
       "GCA_009669555.1_ASM966955v1_genomic         16.0       0.0  42.480495   \n",
       "GCA_009669565.1_ASM966956v1_genomic         16.0       0.0  42.480495   \n",
       "...                                          ...       ...        ...   \n",
       "SAMEA7313243_45                            471.0     150.0  47.622245   \n",
       "SAMEA7313246_49                            471.0     150.0  54.245560   \n",
       "SAMEA6651390                               565.0      21.0  54.822692   \n",
       "SAMEA6637004                               494.5      85.5  54.822692   \n",
       "SAMEA6637002                               473.5      84.5  54.822692   \n",
       "\n",
       "                                            lon  host_human  \n",
       "sample                                                       \n",
       "Reference                           -105.607716       Human  \n",
       "GCA_009909635.1_ASM990963v1_genomic   40.795794       Human  \n",
       "GCA_009669545.1_ASM966954v1_genomic   85.463346   Non-Human  \n",
       "GCA_009669555.1_ASM966955v1_genomic   85.463346   Non-Human  \n",
       "GCA_009669565.1_ASM966956v1_genomic   85.463346   Non-Human  \n",
       "...                                         ...         ...  \n",
       "SAMEA7313243_45                       40.795794       Human  \n",
       "SAMEA7313246_49                       18.109900       Human  \n",
       "SAMEA6651390                          25.249534       Human  \n",
       "SAMEA6637004                          25.249534       Human  \n",
       "SAMEA6637002                          25.249534       Human  \n",
       "\n",
       "[601 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.set_index(metadata_df.columns[0], inplace=True)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "passing-capitol",
   "metadata": {},
   "source": [
    "## Most Recent Sampling Date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "tribal-manchester",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path = os.path.join(beast_dir, \"most_recent_sampling_dates.tsv\")\n",
    "mrsd_dict = {}\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    for branch in BRANCH_LIST:\n",
    "        branch_df = metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])]\n",
    "        max_date = max(branch_df[\"date_mean\"])\n",
    "        outfile.write(\"{}\\t{}\\n\".format(branch, max_date))\n",
    "        mrsd_dict[branch] = max_date"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unlikely-beverage",
   "metadata": {},
   "source": [
    "## Colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cleared-dutch",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'province': {'Krasnoyarsk Krai': '#8000ff', 'Panevezys County': '#7b07ff', 'Pärnu maakond': '#760eff', 'Irkutsk Oblast': '#7215ff', 'Bavaria': '#6d1dff', 'Altai Krai': '#6924fe', 'Qinghai': '#642bfe', 'Republic of Dagestan': '#6032fe', 'Goranboy District': '#5b39fd', 'Syunik Province': '#5740fd', 'Shirak Province': '#5247fc', 'Gegharkunik Province': '#4d4dfc', 'Samtskhe-Javakheti': '#4954fb', 'Shahbuz Rayon': '#445bfb', 'Samara Oblast': '#4062fa', 'Sečuán': '#3b68f9', 'Bayankhongor': '#376ff9', 'Inner Mongolia': '#3275f8', 'Sughd Province': '#2e7bf7', 'Talas Region': '#2982f6', 'Altai Republic': '#2488f5', 'Govi-Altai': '#208ef4', 'Övörkhangai': '#1b94f3', 'Ömnögovi': '#1799f2', 'Bayan-Ölgii': '#129ff1', 'Khovd': '#0ea5ef', 'Xinjiang': '#09aaee', 'Issyk-Kul Region': '#05afed', 'East of England': '#00b4ec', 'Valencia Community': '#05b9ea', 'Centre-Loire Valley': '#09bee9', 'Osh Region': '#0ec3e7', 'Naryn Region': '#12c7e6', 'Gansu': '#17cce4', 'Hovsgel': '#1bd0e3', 'Zavkhan Province': '#20d4e1', 'Dornogovi': '#24d8df', 'Tuva Republic': '#29dcdd', 'Jilin': '#2edfdc', 'St. Petersberg': '#32e3da', 'Zabaykalsky Krai': '#37e6d8', 'Heilongjiang': '#3be9d6', 'Tibet': '#40ecd4', 'Gandaki Pradesh': '#44eed2', 'Maharashtra': '#49f1d0', 'Himachal Pradesh': '#4df3ce', 'Kabardino-Balkaria': '#52f5cc', 'Karachay-Cherkessia': '#57f7ca', 'Ningxia': '#5bf9c7', 'Shaanxi': '#60fac5', 'Hebei': '#64fbc3', 'West Kazakhstan Region': '#69fcc1', 'Nagorno-Karabakh Republic': '#6dfdbe', 'Kyzylorda Region': '#72febc', 'Rostov Oblast': '#76ffb9', 'Astrakhan Oblast': '#7bffb7', 'Balkan': '#80ffb4', 'Kyzylorda Province': '#84ffb2', 'Shymkent': '#89ffaf', 'Navoiy Region': '#8dfead', 'Almaty Region': '#92fdaa', 'West Kazakhstan region': '#96fca7', 'Atyrau Region': '#9bfba5', 'Fizuli District': '#9ffaa2', 'Baku': '#a4f99f', 'Kurdistan': '#a8f79c', 'Shamkir District': '#adf599', 'Stavropol Krai': '#b2f396', 'Ingushetia': '#b6f194', 'Mangystau Province': '#bbee91', 'Chechnya': '#bfec8e', 'Aktobe Region': '#c4e98b', 'Atyrau': '#c8e688', 'Qazakh District': '#cde385', 'Republic of Kalmykia': '#d1df82', 'Tatarstan': '#d6dc7e', 'Lazio': '#dbd87b', 'Greater London': '#dfd478', \"Provence-Alpes-Côte d'Azur\": '#e4d075', 'Baden-Württemberg': '#e8cc72', 'Vilnius County': '#edc76f', 'Pomeranian Voivodeship': '#f1c36b', 'Brandenburg': '#f6be68', 'Nidwalden': '#fab965', 'Oslo': '#ffb462', 'Catalonia': '#ffaf5e', 'Occitanie': '#ffaa5b', 'North Brabant': '#ffa558', 'Nairobi': '#ff9f54', 'Yunnan': '#ff9951', 'Khánh Hoà Province': '#ff944d', 'Lâm Đồng Province': '#ff8e4a', 'Ho Chi Minh City': '#ff8847', 'Guangxi Zhuang': '#ff8243', 'California': '#ff7b40', 'La Libertad': '#ff753c', 'Tamil Nadu': '#ff6f39', 'Saskatchewan': '#ff6835', 'Arizona': '#ff6232', 'New Mexico': '#ff5b2e', 'Colorado': '#ff542b', 'Texas': '#ff4d27', 'Nevada': '#ff4724', 'Central Java': '#ff4020', 'Province de Fianarantsoa': '#ff391d', 'Kirov Oblast': '#ff3219', 'Gujarat': '#ff2b15', 'Fujian': '#ff2412', 'Midlands': '#ff1d0e', 'MSila': '#ff150b', 'State of Mato Grosso': '#ff0e07', 'La Paz': '#ff0704', 'Cajamarca': '#ff0000', '?': '#969696'}, 'country': {'Russia': '#8000ff', 'Lithuania': '#7215ff', 'Estonia': '#652afe', 'Germany': '#573ffd', 'China': '#4a53fc', 'Azerbaijan': '#3c66fa', 'Armenia': '#2f79f7', 'Georgia': '#228bf4', 'Mongolia': '#149df1', 'Tajikistan': '#07adee', 'Kyrgyzstan': '#07bcea', 'England': '#14c9e5', 'Spain': '#22d5e0', 'France': '#2fe0db', 'Nepal': '#3cead5', 'India': '#4af1d0', 'Kazakhstan': '#57f7c9', 'Turkmenistan': '#65fcc3', 'Uzbekistan': '#72febc', 'Iran': '#80ffb4', 'Italy': '#8dfead', 'Poland': '#9afca5', 'Switzerland': '#a8f79d', 'Norway': '#b5f194', 'The Netherlands': '#c3ea8b', 'Kenya': '#d0e083', 'Uganda': '#ddd579', 'Democratic Republic of the Congo': '#ebc970', 'Vietnam': '#f8bc66', 'Myanmar': '#ffad5d', 'United States of America': '#ff9d53', 'Peru': '#ff8b49', 'Canada': '#ff793f', 'Indonesia': '#ff6634', 'Madagascar': '#ff532a', 'Zimbabwe': '#ff3f20', 'Algeria': '#ff2a15', 'Brazil': '#ff150b', 'Bolivia': '#ff0000', '?': '#969696'}, 'branch_minor': {'0.PRE2': '#8000ff', '0.PRE1': '#7018ff', '0.PE7': '#6130fe', '0.PE2': '#5148fc', '0.PE8': '#425ffa', '0.PE10': '#3275f8', '0.PE4': '#238af5', '0.PE5': '#139ef1', '0.ANT1': '#04b0ed', '0.ANT4': '#0cc1e8', '0.ANT5': '#1bd0e3', '0.ANT2': '#2bdddd', '0.ANT3': '#3ae8d7', '3.ANT1': '#49f1d0', '3.ANT2': '#59f8c8', '4.ANT1': '#68fcc1', '2.ANT3': '#78ffb9', '2.ANT2': '#87ffb0', '2.ANT1': '#97fca7', '2.MED0': '#a6f89e', '2.MED3': '#b6f194', '2.MED2': '#c5e88a', '2.MED1': '#d5dd7f', '1.PRE0': '#e4d075', '1.PRE1': '#f3c16a', '1.PRE2': '#ffb05f', '1.PRE3': '#ff9e53', '1.ANT1': '#ff8a48', '1.IN1': '#ff753c', '1.IN2': '#ff5f30', '1.IN3': '#ff4824', '1.ORI2': '#ff3018', '1.ORI1': '#ff180c', '1.ORI3': '#ff0000', '?': '#969696'}, 'branch_major': {'0.PRE': '#8000ff', '0.PE': '#4c4ffc', '0.ANT': '#1996f3', '3.ANT': '#1acee3', '4.ANT': '#4df3ce', '2.ANT': '#80ffb4', '2.MED': '#b3f396', '1.PRE': '#e6ce74', '1.ANT': '#ff964f', '1.IN': '#ff4f28', '1.ORI': '#ff0000', '?': '#969696'}}\n"
     ]
    }
   ],
   "source": [
    "colors_dict = {}\n",
    "\n",
    "colors_df = pd.read_csv(out_path_colors, sep='\\t', header=None)\n",
    "colors_df.columns = [\"state\", \"value\", \"color\"]\n",
    "\n",
    "for state in set(colors_df[\"state\"]):\n",
    "    state_df = colors_df[colors_df[\"state\"] == state]    \n",
    "    colors_dict[state] = {}\n",
    "    \n",
    "    for value,color in zip(state_df[\"value\"], state_df[\"color\"]):\n",
    "        colors_dict[state][value] = color\n",
    "\n",
    "print(colors_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "economic-raising",
   "metadata": {},
   "source": [
    "## Tree Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "small-message",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Branch: 1.ORI\n",
      "Branch: 1.IN\n",
      "Branch: 1.ANT\n",
      "Branch: 1.PRE\n",
      "Branch: 2.MED\n",
      "Branch: 2.ANT\n",
      "Branch: 4.ANT\n",
      "Branch: 3.ANT\n",
      "Branch: 0.ANT\n",
      "Branch: 0.ANT4\n",
      "Branch: 0.PE\n",
      "Branch: 0.PRE\n"
     ]
    }
   ],
   "source": [
    "# Construct a dictionary to hold the trees\n",
    "tree_dict = {}\n",
    "\n",
    "for branch in BRANCH_LIST:\n",
    "    print(\"Branch:\", branch)\n",
    "    tree_dict[branch] = {} \n",
    "    for filename in os.listdir(tree_dir):\n",
    "        if not filename.endswith(\".tre\"): continue\n",
    "        filepath = os.path.join(tree_dir, filename)\n",
    "        if branch in filename:            \n",
    "            # Add tree files to dict\n",
    "            tree_dict[branch][\"tree_file_raw\"] = filepath \n",
    "            tree_dict[branch][\"tree_file_edit\"] = os.path.join(tree_dir, branch + \".nex\")\n",
    "            tree_dict[branch][\"sample_rename\"] = {}\n",
    "            \n",
    "            # Read in raw tree to deal with dashes\n",
    "            with open(tree_dict[branch][\"tree_file_raw\"],  \"r\") as infile:                    \n",
    "                with open(tree_dict[branch][\"tree_file_edit\"], \"w\") as outfile:           \n",
    "                    raw_tree = infile.read()\n",
    "                    # Remove quotations if they exist\n",
    "                    raw_tree = raw_tree.replace(\"'\",\"\")\n",
    "                    \n",
    "                    # Split into lines to iterate over\n",
    "                    raw_tree_lines = raw_tree.split(\"\\n\")\n",
    "                    \n",
    "                    # By default don't parse a line for dashes\n",
    "                    taxa_line = False\n",
    "                    \n",
    "                    for line in raw_tree_lines: \n",
    "                                \n",
    "                        if \"TREE\" not in line and \"-\" in line:\n",
    "                            \n",
    "                            if len(line.split(\" \")) == 1:\n",
    "                                name_dashes = line.strip()\n",
    "                                name_no_dashes = name_dashes.replace(\"-\",\"_\")\n",
    "                                tree_dict[branch][\"sample_rename\"][name_no_dashes] = name_dashes\n",
    "                            line = line.replace(\"-\",\"_\")\n",
    "                                \n",
    "                        outfile.write(line + \"\\n\")\n",
    "            \n",
    "            # Read in edited tree\n",
    "            trees = Phylo.parse(tree_dict[branch][\"tree_file_edit\"], \"nexus\")\n",
    "            # There should be only 1 tree\n",
    "            for t in trees:\n",
    "                tree_dict[branch][\"tree\"] = t\n",
    "                tree_dict[branch][\"tree\"].ladderize(reverse=False)\n",
    "                break\n",
    "\n",
    "            # Rename sample names back to with dashes\n",
    "            for c in tree_dict[branch][\"tree\"].find_clades():\n",
    "                if c.name in tree_dict[branch][\"sample_rename\"]:\n",
    "                    orig_name = c.name\n",
    "                    c.name = tree_dict[branch][\"sample_rename\"][c.name]\n",
    "                    #print(\"Rename:\", orig_name, c.name)\n",
    "                    \n",
    "                # Strip the date suffix\n",
    "                if c.name:\n",
    "                    c.name = \"_\".join(c.name.split(\"_\")[0:-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "italic-huntington",
   "metadata": {},
   "source": [
    "## Add Tree Data to Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "arranged-birth",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize new columns\n",
    "\n",
    "metadata_df[\"node_type\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"branch_length\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"rate\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"rate_subs\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"timetree_num_date_confidence\"] = [[0,0]] * len(metadata_df)\n",
    "\n",
    "init_data = {col: [NO_DATA_CHAR] for col in metadata_df.columns}\n",
    "\n",
    "for branch in BRANCH_LIST:\n",
    "    tree = tree_dict[branch][\"tree\"]\n",
    "\n",
    "    df = copy.deepcopy(metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])])\n",
    "    \n",
    "    node_i = 0\n",
    "    \n",
    "    # Get comment headers for df\n",
    "    root_comment_dict = parse_comment(tree.root.comment)\n",
    "    for parameter in root_comment_dict:\n",
    "        df[parameter] = [NO_DATA_CHAR] * len(df)\n",
    "    \n",
    "    for c in tree.find_clades():\n",
    "\n",
    "        # Rename internal nodes\n",
    "        if not c.name:\n",
    "            c.name = \"NODE{}\".format(node_i)\n",
    "            node_i += 1\n",
    "\n",
    "        # Initialize metadata for internal nodes\n",
    "        if \"NODE\" in c.name:\n",
    "            init_data[\"sample\"] = c.name\n",
    "            data_row = pd.DataFrame(init_data)\n",
    "            data_row.set_index(\"sample\", inplace=True)\n",
    "            df = df.append(data_row)\n",
    "            \n",
    "        # Set node type\n",
    "        if \"NODE\" in c.name:\n",
    "            node_type = \"internal\"                    \n",
    "        else:\n",
    "            node_type = \"terminal\"\n",
    "           \n",
    "        # Parse comments\n",
    "        comment_dict = parse_comment(c.comment)\n",
    "        for parameter,val in comment_dict.items():\n",
    "            df.at[c.name, parameter] = val\n",
    "            \n",
    "        # Get calendar date for node\n",
    "        calendar_date = mrsd_dict[branch] - float(comment_dict[\"height\"])\n",
    "        df.at[c.name, \"timetree_num_date\"] = calendar_date\n",
    "        \n",
    "        # Rate\n",
    "        if df[\"rate\"][c.name] == NO_DATA_CHAR:\n",
    "            df.at[c.name, \"rate\"] = 0\n",
    "        else:\n",
    "            df.at[c.name, \"rate\"] = float(df[\"rate\"][c.name])\n",
    "            \n",
    "        rate = df[\"rate\"][c.name]\n",
    "        rate_subs = rate * SEQ_LEN\n",
    "        df.at[c.name, \"rate_subs\"] = rate_subs\n",
    "        \n",
    "        \n",
    "        # Convert\n",
    "        height = comment_dict[\"height\"]\n",
    "        df.at[c.name, \"height\"] = float(height)\n",
    "        if \"height_95%_HPD\" in comment_dict:\n",
    "            height_95_hpd = comment_dict[\"height_95%_HPD\"]\n",
    "            height_95_hpd_list = [float(h) for h in height_95_hpd.strip(\"{}\").split(\",\")]\n",
    "            calendar_95_hpd = [mrsd_dict[branch] - h for h in height_95_hpd_list]\n",
    "            df.at[c.name, \"timetree_num_date_confidence\"] = [calendar_95_hpd[1], calendar_95_hpd[0]]\n",
    "        \n",
    "        # Add data\n",
    "        df.at[c.name, \"node_type\"] = node_type\n",
    "        df.at[c.name, \"branch_length\"] = c.branch_length\n",
    "\n",
    "        \n",
    "        \n",
    "    # Replace NAs for inconsistent parameters\n",
    "    df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "    tree_dict[branch][\"df\"] = copy.deepcopy(df)\n",
    "\n",
    "    \n",
    "    #display(tree_dict[branch][\"df\"])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "robust-tracker",
   "metadata": {},
   "source": [
    "## Reduced Dataframe for Auspice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "brutal-somalia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['strain', 'date', 'date_bp', 'country', 'province', 'country_lat',\n",
      "       'country_lon', 'province_lat', 'province_lon', 'biovar', 'branch_major',\n",
      "       'branch_minor', 'biosample_accession', 'biosample_comment',\n",
      "       'branch_number', 'continent', 'date_mean', 'date_bp_mean', 'date_err',\n",
      "       'lat', 'lon', 'host_human', 'node_type', 'branch_length', 'rate',\n",
      "       'rate_subs', 'timetree_num_date_confidence', 'height_95%_HPD', 'length',\n",
      "       'posterior', 'height_median', 'height_range', 'height',\n",
      "       'timetree_num_date', 'length_range', 'rate_95%_HPD', 'length_95%_HPD',\n",
      "       'rate_median', 'length_median', 'rate_range'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# OPtions\n",
    "print(tree_dict[branch][\"df\"].columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "reasonable-faculty",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>terminal</td>\n",
       "      <td>17.516688</td>\n",
       "      <td>1992.000000</td>\n",
       "      <td>[1992.0, 1992.0]</td>\n",
       "      <td>3.553717e-07</td>\n",
       "      <td>1.502902</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>CO92</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000834775.1_ASM83477v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>9.482611</td>\n",
       "      <td>1967.000000</td>\n",
       "      <td>[1967.0, 1967.0]</td>\n",
       "      <td>3.123540e-07</td>\n",
       "      <td>1.320976</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Arizona</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>34.3953</td>\n",
       "      <td>-111.763</td>\n",
       "      <td>Dodson</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>1967.0</td>\n",
       "      <td>-54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000834335.1_ASM83433v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>4.917912</td>\n",
       "      <td>1954.000000</td>\n",
       "      <td>[1954.0, 1954.0]</td>\n",
       "      <td>2.122180e-07</td>\n",
       "      <td>0.897491</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>California</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>36.7015</td>\n",
       "      <td>-118.756</td>\n",
       "      <td>Shasta</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>1954.0</td>\n",
       "      <td>-67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000169635.1_ASM16963v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>32.332497</td>\n",
       "      <td>2005.000000</td>\n",
       "      <td>[2005.0, 2005.0]</td>\n",
       "      <td>1.026717e-07</td>\n",
       "      <td>0.434209</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>NA</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>MG05-1020</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000170275.1_ASM17027v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>18.461812</td>\n",
       "      <td>1991.000000</td>\n",
       "      <td>[1991.0, 1991.0]</td>\n",
       "      <td>8.033665e-08</td>\n",
       "      <td>0.339752</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>F1991016</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1991.0</td>\n",
       "      <td>-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE111</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.194478</td>\n",
       "      <td>2004.241372</td>\n",
       "      <td>[2000.475519086657, 2007.5041318052265]</td>\n",
       "      <td>1.016505e-07</td>\n",
       "      <td>0.429890</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE112</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.929602</td>\n",
       "      <td>2006.999750</td>\n",
       "      <td>[2003.8227971142646, 2009.6153588063687]</td>\n",
       "      <td>1.311277e-07</td>\n",
       "      <td>0.554552</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE113</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.351704</td>\n",
       "      <td>2005.846571</td>\n",
       "      <td>[2002.70281589384, 2008.677776298874]</td>\n",
       "      <td>1.653129e-07</td>\n",
       "      <td>0.699124</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE114</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.391180</td>\n",
       "      <td>2006.938556</td>\n",
       "      <td>[2004.3050370690294, 2009.1217091576173]</td>\n",
       "      <td>1.153553e-07</td>\n",
       "      <td>0.487849</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE115</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.485981</td>\n",
       "      <td>2007.915086</td>\n",
       "      <td>[2005.7202344890497, 2009.6665137272012]</td>\n",
       "      <td>2.037950e-07</td>\n",
       "      <td>0.861869</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>233 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   node_type  branch_length  \\\n",
       "sample                                                        \n",
       "Reference                           terminal      17.516688   \n",
       "GCA_000834775.1_ASM83477v1_genomic  terminal       9.482611   \n",
       "GCA_000834335.1_ASM83433v1_genomic  terminal       4.917912   \n",
       "GCA_000169635.1_ASM16963v1_genomic  terminal      32.332497   \n",
       "GCA_000170275.1_ASM17027v1_genomic  terminal      18.461812   \n",
       "...                                      ...            ...   \n",
       "NODE111                             internal       0.194478   \n",
       "NODE112                             internal       0.929602   \n",
       "NODE113                             internal       0.351704   \n",
       "NODE114                             internal       2.391180   \n",
       "NODE115                             internal       0.485981   \n",
       "\n",
       "                                    timetree_num_date  \\\n",
       "sample                                                  \n",
       "Reference                                 1992.000000   \n",
       "GCA_000834775.1_ASM83477v1_genomic        1967.000000   \n",
       "GCA_000834335.1_ASM83433v1_genomic        1954.000000   \n",
       "GCA_000169635.1_ASM16963v1_genomic        2005.000000   \n",
       "GCA_000170275.1_ASM17027v1_genomic        1991.000000   \n",
       "...                                               ...   \n",
       "NODE111                                   2004.241372   \n",
       "NODE112                                   2006.999750   \n",
       "NODE113                                   2005.846571   \n",
       "NODE114                                   2006.938556   \n",
       "NODE115                                   2007.915086   \n",
       "\n",
       "                                                timetree_num_date_confidence  \\\n",
       "sample                                                                         \n",
       "Reference                                                   [1992.0, 1992.0]   \n",
       "GCA_000834775.1_ASM83477v1_genomic                          [1967.0, 1967.0]   \n",
       "GCA_000834335.1_ASM83433v1_genomic                          [1954.0, 1954.0]   \n",
       "GCA_000169635.1_ASM16963v1_genomic                          [2005.0, 2005.0]   \n",
       "GCA_000170275.1_ASM17027v1_genomic                          [1991.0, 1991.0]   \n",
       "...                                                                      ...   \n",
       "NODE111                              [2000.475519086657, 2007.5041318052265]   \n",
       "NODE112                             [2003.8227971142646, 2009.6153588063687]   \n",
       "NODE113                                [2002.70281589384, 2008.677776298874]   \n",
       "NODE114                             [2004.3050370690294, 2009.1217091576173]   \n",
       "NODE115                             [2005.7202344890497, 2009.6665137272012]   \n",
       "\n",
       "                                      rate_beast  rate_subs_beast  \\\n",
       "sample                                                              \n",
       "Reference                           3.553717e-07         1.502902   \n",
       "GCA_000834775.1_ASM83477v1_genomic  3.123540e-07         1.320976   \n",
       "GCA_000834335.1_ASM83433v1_genomic  2.122180e-07         0.897491   \n",
       "GCA_000169635.1_ASM16963v1_genomic  1.026717e-07         0.434209   \n",
       "GCA_000170275.1_ASM17027v1_genomic  8.033665e-08         0.339752   \n",
       "...                                          ...              ...   \n",
       "NODE111                             1.016505e-07         0.429890   \n",
       "NODE112                             1.311277e-07         0.554552   \n",
       "NODE113                             1.653129e-07         0.699124   \n",
       "NODE114                             1.153553e-07         0.487849   \n",
       "NODE115                             2.037950e-07         0.861869   \n",
       "\n",
       "                                                     country    province  \\\n",
       "sample                                                                     \n",
       "Reference                           United States of America    Colorado   \n",
       "GCA_000834775.1_ASM83477v1_genomic  United States of America     Arizona   \n",
       "GCA_000834335.1_ASM83433v1_genomic  United States of America  California   \n",
       "GCA_000169635.1_ASM16963v1_genomic                Madagascar          NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic                     China      Yunnan   \n",
       "...                                                      ...         ...   \n",
       "NODE111                                                   NA          NA   \n",
       "NODE112                                                   NA          NA   \n",
       "NODE113                                                   NA          NA   \n",
       "NODE114                                                   NA          NA   \n",
       "NODE115                                                   NA          NA   \n",
       "\n",
       "                                   country_lat country_lon province_lat  \\\n",
       "sample                                                                    \n",
       "Reference                              39.7837    -100.446      38.7252   \n",
       "GCA_000834775.1_ASM83477v1_genomic     39.7837    -100.446      34.3953   \n",
       "GCA_000834335.1_ASM83433v1_genomic     39.7837    -100.446      36.7015   \n",
       "GCA_000169635.1_ASM16963v1_genomic     -18.925     46.4416           NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic     35.0001         105           25   \n",
       "...                                        ...         ...          ...   \n",
       "NODE111                                     NA          NA           NA   \n",
       "NODE112                                     NA          NA           NA   \n",
       "NODE113                                     NA          NA           NA   \n",
       "NODE114                                     NA          NA           NA   \n",
       "NODE115                                     NA          NA           NA   \n",
       "\n",
       "                                   province_lon     strain branch_major  \\\n",
       "sample                                                                    \n",
       "Reference                              -105.608       CO92        1.ORI   \n",
       "GCA_000834775.1_ASM83477v1_genomic     -111.763     Dodson        1.ORI   \n",
       "GCA_000834335.1_ASM83433v1_genomic     -118.756     Shasta        1.ORI   \n",
       "GCA_000169635.1_ASM16963v1_genomic           NA  MG05-1020        1.ORI   \n",
       "GCA_000170275.1_ASM17027v1_genomic          102   F1991016        1.ORI   \n",
       "...                                         ...        ...          ...   \n",
       "NODE111                                      NA         NA           NA   \n",
       "NODE112                                      NA         NA           NA   \n",
       "NODE113                                      NA         NA           NA   \n",
       "NODE114                                      NA         NA           NA   \n",
       "NODE115                                      NA         NA           NA   \n",
       "\n",
       "                                        biovar host_human    date date_bp  \n",
       "sample                                                                     \n",
       "Reference                           Orientalis      Human    1992     -29  \n",
       "GCA_000834775.1_ASM83477v1_genomic  Orientalis      Human  1967.0     -54  \n",
       "GCA_000834335.1_ASM83433v1_genomic  Orientalis      Human  1954.0     -67  \n",
       "GCA_000169635.1_ASM16963v1_genomic  Orientalis      Human  2005.0     -16  \n",
       "GCA_000170275.1_ASM17027v1_genomic  Orientalis  Non-Human  1991.0     -30  \n",
       "...                                        ...        ...     ...     ...  \n",
       "NODE111                                     NA         NA      NA      NA  \n",
       "NODE112                                     NA         NA      NA      NA  \n",
       "NODE113                                     NA         NA      NA      NA  \n",
       "NODE114                                     NA         NA      NA      NA  \n",
       "NODE115                                     NA         NA      NA      NA  \n",
       "\n",
       "[233 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000324785.1_ASM32478v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>7.304944</td>\n",
       "      <td>1979.000000</td>\n",
       "      <td>[1979.0, 1979.0]</td>\n",
       "      <td>2.189105e-07</td>\n",
       "      <td>0.925794</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>E1979001</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1979.0</td>\n",
       "      <td>-42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000022845.1_ASM2284v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>3.860503</td>\n",
       "      <td>1976.000000</td>\n",
       "      <td>[1976.0, 1976.0]</td>\n",
       "      <td>3.444121e-07</td>\n",
       "      <td>1.456553</td>\n",
       "      <td>China</td>\n",
       "      <td>Tibet</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>29.8556</td>\n",
       "      <td>90.875</td>\n",
       "      <td>Z176003</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>-45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000022825.1_ASM2282v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>31.757595</td>\n",
       "      <td>1982.000000</td>\n",
       "      <td>[1982.0, 1982.0]</td>\n",
       "      <td>2.079253e-06</td>\n",
       "      <td>8.793364</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>D182038</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1982.0</td>\n",
       "      <td>-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000022805.1_ASM2280v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>70.350981</td>\n",
       "      <td>2006.000000</td>\n",
       "      <td>[2006.0, 2006.0]</td>\n",
       "      <td>1.477245e-07</td>\n",
       "      <td>0.624742</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>D106004</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000325465.1_ASM32546v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>9.419908</td>\n",
       "      <td>1957.000000</td>\n",
       "      <td>[1957.0, 1957.0]</td>\n",
       "      <td>2.958604e-07</td>\n",
       "      <td>1.251223</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>YN472</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1957.0</td>\n",
       "      <td>-64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE33</th>\n",
       "      <td>internal</td>\n",
       "      <td>8.520717</td>\n",
       "      <td>1972.797250</td>\n",
       "      <td>[1957.7256089593875, 1986.4898330318686]</td>\n",
       "      <td>5.422988e-08</td>\n",
       "      <td>0.229343</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE34</th>\n",
       "      <td>internal</td>\n",
       "      <td>3.206524</td>\n",
       "      <td>1981.521412</td>\n",
       "      <td>[1973.0440875505783, 1988.6716159399791]</td>\n",
       "      <td>2.454978e-07</td>\n",
       "      <td>1.038234</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE35</th>\n",
       "      <td>internal</td>\n",
       "      <td>9.000596</td>\n",
       "      <td>1987.215157</td>\n",
       "      <td>[1980.308465360717, 1993.659574537487]</td>\n",
       "      <td>1.360054e-07</td>\n",
       "      <td>0.575180</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE36</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.179186</td>\n",
       "      <td>1990.707595</td>\n",
       "      <td>[1985.2879987060362, 1995.205487889126]</td>\n",
       "      <td>1.092851e-07</td>\n",
       "      <td>0.462177</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE37</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.921740</td>\n",
       "      <td>1994.610962</td>\n",
       "      <td>[1991.8070085748427, 1995.999762013854]</td>\n",
       "      <td>3.756432e-07</td>\n",
       "      <td>1.588632</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   node_type  branch_length  \\\n",
       "sample                                                        \n",
       "GCA_000324785.1_ASM32478v1_genomic  terminal       7.304944   \n",
       "GCA_000022845.1_ASM2284v1_genomic   terminal       3.860503   \n",
       "GCA_000022825.1_ASM2282v1_genomic   terminal      31.757595   \n",
       "GCA_000022805.1_ASM2280v1_genomic   terminal      70.350981   \n",
       "GCA_000325465.1_ASM32546v1_genomic  terminal       9.419908   \n",
       "...                                      ...            ...   \n",
       "NODE33                              internal       8.520717   \n",
       "NODE34                              internal       3.206524   \n",
       "NODE35                              internal       9.000596   \n",
       "NODE36                              internal       1.179186   \n",
       "NODE37                              internal       1.921740   \n",
       "\n",
       "                                    timetree_num_date  \\\n",
       "sample                                                  \n",
       "GCA_000324785.1_ASM32478v1_genomic        1979.000000   \n",
       "GCA_000022845.1_ASM2284v1_genomic         1976.000000   \n",
       "GCA_000022825.1_ASM2282v1_genomic         1982.000000   \n",
       "GCA_000022805.1_ASM2280v1_genomic         2006.000000   \n",
       "GCA_000325465.1_ASM32546v1_genomic        1957.000000   \n",
       "...                                               ...   \n",
       "NODE33                                    1972.797250   \n",
       "NODE34                                    1981.521412   \n",
       "NODE35                                    1987.215157   \n",
       "NODE36                                    1990.707595   \n",
       "NODE37                                    1994.610962   \n",
       "\n",
       "                                                timetree_num_date_confidence  \\\n",
       "sample                                                                         \n",
       "GCA_000324785.1_ASM32478v1_genomic                          [1979.0, 1979.0]   \n",
       "GCA_000022845.1_ASM2284v1_genomic                           [1976.0, 1976.0]   \n",
       "GCA_000022825.1_ASM2282v1_genomic                           [1982.0, 1982.0]   \n",
       "GCA_000022805.1_ASM2280v1_genomic                           [2006.0, 2006.0]   \n",
       "GCA_000325465.1_ASM32546v1_genomic                          [1957.0, 1957.0]   \n",
       "...                                                                      ...   \n",
       "NODE33                              [1957.7256089593875, 1986.4898330318686]   \n",
       "NODE34                              [1973.0440875505783, 1988.6716159399791]   \n",
       "NODE35                                [1980.308465360717, 1993.659574537487]   \n",
       "NODE36                               [1985.2879987060362, 1995.205487889126]   \n",
       "NODE37                               [1991.8070085748427, 1995.999762013854]   \n",
       "\n",
       "                                      rate_beast  rate_subs_beast country  \\\n",
       "sample                                                                      \n",
       "GCA_000324785.1_ASM32478v1_genomic  2.189105e-07         0.925794   China   \n",
       "GCA_000022845.1_ASM2284v1_genomic   3.444121e-07         1.456553   China   \n",
       "GCA_000022825.1_ASM2282v1_genomic   2.079253e-06         8.793364   China   \n",
       "GCA_000022805.1_ASM2280v1_genomic   1.477245e-07         0.624742   China   \n",
       "GCA_000325465.1_ASM32546v1_genomic  2.958604e-07         1.251223   China   \n",
       "...                                          ...              ...     ...   \n",
       "NODE33                              5.422988e-08         0.229343      NA   \n",
       "NODE34                              2.454978e-07         1.038234      NA   \n",
       "NODE35                              1.360054e-07         0.575180      NA   \n",
       "NODE36                              1.092851e-07         0.462177      NA   \n",
       "NODE37                              3.756432e-07         1.588632      NA   \n",
       "\n",
       "                                   province country_lat country_lon  \\\n",
       "sample                                                                \n",
       "GCA_000324785.1_ASM32478v1_genomic   Yunnan     35.0001         105   \n",
       "GCA_000022845.1_ASM2284v1_genomic     Tibet     35.0001         105   \n",
       "GCA_000022825.1_ASM2282v1_genomic    Yunnan     35.0001         105   \n",
       "GCA_000022805.1_ASM2280v1_genomic    Yunnan     35.0001         105   \n",
       "GCA_000325465.1_ASM32546v1_genomic   Yunnan     35.0001         105   \n",
       "...                                     ...         ...         ...   \n",
       "NODE33                                   NA          NA          NA   \n",
       "NODE34                                   NA          NA          NA   \n",
       "NODE35                                   NA          NA          NA   \n",
       "NODE36                                   NA          NA          NA   \n",
       "NODE37                                   NA          NA          NA   \n",
       "\n",
       "                                   province_lat province_lon    strain  \\\n",
       "sample                                                                   \n",
       "GCA_000324785.1_ASM32478v1_genomic           25          102  E1979001   \n",
       "GCA_000022845.1_ASM2284v1_genomic       29.8556       90.875   Z176003   \n",
       "GCA_000022825.1_ASM2282v1_genomic            25          102   D182038   \n",
       "GCA_000022805.1_ASM2280v1_genomic            25          102   D106004   \n",
       "GCA_000325465.1_ASM32546v1_genomic           25          102     YN472   \n",
       "...                                         ...          ...       ...   \n",
       "NODE33                                       NA           NA        NA   \n",
       "NODE34                                       NA           NA        NA   \n",
       "NODE35                                       NA           NA        NA   \n",
       "NODE36                                       NA           NA        NA   \n",
       "NODE37                                       NA           NA        NA   \n",
       "\n",
       "                                   branch_major   biovar host_human    date  \\\n",
       "sample                                                                        \n",
       "GCA_000324785.1_ASM32478v1_genomic         1.IN  Antiqua  Non-Human  1979.0   \n",
       "GCA_000022845.1_ASM2284v1_genomic          1.IN  Antiqua  Non-Human  1976.0   \n",
       "GCA_000022825.1_ASM2282v1_genomic          1.IN  Antiqua  Non-Human  1982.0   \n",
       "GCA_000022805.1_ASM2280v1_genomic          1.IN  Antiqua  Non-Human  2006.0   \n",
       "GCA_000325465.1_ASM32546v1_genomic         1.IN  Antiqua  Non-Human  1957.0   \n",
       "...                                         ...      ...        ...     ...   \n",
       "NODE33                                       NA       NA         NA      NA   \n",
       "NODE34                                       NA       NA         NA      NA   \n",
       "NODE35                                       NA       NA         NA      NA   \n",
       "NODE36                                       NA       NA         NA      NA   \n",
       "NODE37                                       NA       NA         NA      NA   \n",
       "\n",
       "                                   date_bp  \n",
       "sample                                      \n",
       "GCA_000324785.1_ASM32478v1_genomic     -42  \n",
       "GCA_000022845.1_ASM2284v1_genomic      -45  \n",
       "GCA_000022825.1_ASM2282v1_genomic      -39  \n",
       "GCA_000022805.1_ASM2280v1_genomic      -15  \n",
       "GCA_000325465.1_ASM32546v1_genomic     -64  \n",
       "...                                    ...  \n",
       "NODE33                                  NA  \n",
       "NODE34                                  NA  \n",
       "NODE35                                  NA  \n",
       "NODE36                                  NA  \n",
       "NODE37                                  NA  \n",
       "\n",
       "[77 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000169615.1_ASM16961v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>51.265788</td>\n",
       "      <td>2004.000000</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>6.444323e-08</td>\n",
       "      <td>0.272537</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.53336</td>\n",
       "      <td>32.2167</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>UG05-0454</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_003798225.1_ASM379822v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>2.045508</td>\n",
       "      <td>1954.000000</td>\n",
       "      <td>[1954.0, 1954.0]</td>\n",
       "      <td>4.756261e-07</td>\n",
       "      <td>2.011469</td>\n",
       "      <td>Democratic Republic of the Congo</td>\n",
       "      <td>NA</td>\n",
       "      <td>-2.98143</td>\n",
       "      <td>23.8223</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>FDAARGOS_601</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>NA</td>\n",
       "      <td>1954.0</td>\n",
       "      <td>-67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000835005.1_ASM83500v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>236.991771</td>\n",
       "      <td>1985.000000</td>\n",
       "      <td>[1985.0, 1985.0]</td>\n",
       "      <td>6.444323e-08</td>\n",
       "      <td>0.272537</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>1.44197</td>\n",
       "      <td>38.4314</td>\n",
       "      <td>-1.30317</td>\n",
       "      <td>36.8261</td>\n",
       "      <td>Nairobi</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>NA</td>\n",
       "      <td>1985.0</td>\n",
       "      <td>-36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000013825.1_ASM1382v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>12.265788</td>\n",
       "      <td>1965.000000</td>\n",
       "      <td>[1965.0, 1965.0]</td>\n",
       "      <td>1.159070e-07</td>\n",
       "      <td>0.490182</td>\n",
       "      <td>Democratic Republic of the Congo</td>\n",
       "      <td>NA</td>\n",
       "      <td>-2.98143</td>\n",
       "      <td>23.8223</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>1965.0</td>\n",
       "      <td>-56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1758.910228</td>\n",
       "      <td>[1644.8345812291716, 1855.0439441412525]</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>internal</td>\n",
       "      <td>203.946262</td>\n",
       "      <td>1950.327725</td>\n",
       "      <td>[1943.1927174315974, 1953.999874244536]</td>\n",
       "      <td>6.444323e-08</td>\n",
       "      <td>0.272537</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.779720</td>\n",
       "      <td>1954.969051</td>\n",
       "      <td>[1948.8644061878467, 1964.7555744433698]</td>\n",
       "      <td>2.517850e-07</td>\n",
       "      <td>1.064824</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    node_type  branch_length  \\\n",
       "sample                                                         \n",
       "GCA_000169615.1_ASM16961v1_genomic   terminal      51.265788   \n",
       "GCA_003798225.1_ASM379822v1_genomic  terminal       2.045508   \n",
       "GCA_000835005.1_ASM83500v1_genomic   terminal     236.991771   \n",
       "GCA_000013825.1_ASM1382v1_genomic    terminal      12.265788   \n",
       "NODE0                                internal       0.000000   \n",
       "NODE1                                internal     203.946262   \n",
       "NODE2                                internal       0.779720   \n",
       "\n",
       "                                     timetree_num_date  \\\n",
       "sample                                                   \n",
       "GCA_000169615.1_ASM16961v1_genomic         2004.000000   \n",
       "GCA_003798225.1_ASM379822v1_genomic        1954.000000   \n",
       "GCA_000835005.1_ASM83500v1_genomic         1985.000000   \n",
       "GCA_000013825.1_ASM1382v1_genomic          1965.000000   \n",
       "NODE0                                      1758.910228   \n",
       "NODE1                                      1950.327725   \n",
       "NODE2                                      1954.969051   \n",
       "\n",
       "                                                 timetree_num_date_confidence  \\\n",
       "sample                                                                          \n",
       "GCA_000169615.1_ASM16961v1_genomic                                     [0, 0]   \n",
       "GCA_003798225.1_ASM379822v1_genomic                          [1954.0, 1954.0]   \n",
       "GCA_000835005.1_ASM83500v1_genomic                           [1985.0, 1985.0]   \n",
       "GCA_000013825.1_ASM1382v1_genomic                            [1965.0, 1965.0]   \n",
       "NODE0                                [1644.8345812291716, 1855.0439441412525]   \n",
       "NODE1                                 [1943.1927174315974, 1953.999874244536]   \n",
       "NODE2                                [1948.8644061878467, 1964.7555744433698]   \n",
       "\n",
       "                                       rate_beast  rate_subs_beast  \\\n",
       "sample                                                               \n",
       "GCA_000169615.1_ASM16961v1_genomic   6.444323e-08         0.272537   \n",
       "GCA_003798225.1_ASM379822v1_genomic  4.756261e-07         2.011469   \n",
       "GCA_000835005.1_ASM83500v1_genomic   6.444323e-08         0.272537   \n",
       "GCA_000013825.1_ASM1382v1_genomic    1.159070e-07         0.490182   \n",
       "NODE0                                0.000000e+00         0.000000   \n",
       "NODE1                                6.444323e-08         0.272537   \n",
       "NODE2                                2.517850e-07         1.064824   \n",
       "\n",
       "                                                              country  \\\n",
       "sample                                                                  \n",
       "GCA_000169615.1_ASM16961v1_genomic                             Uganda   \n",
       "GCA_003798225.1_ASM379822v1_genomic  Democratic Republic of the Congo   \n",
       "GCA_000835005.1_ASM83500v1_genomic                              Kenya   \n",
       "GCA_000013825.1_ASM1382v1_genomic    Democratic Republic of the Congo   \n",
       "NODE0                                                              NA   \n",
       "NODE1                                                              NA   \n",
       "NODE2                                                              NA   \n",
       "\n",
       "                                    province country_lat country_lon  \\\n",
       "sample                                                                 \n",
       "GCA_000169615.1_ASM16961v1_genomic        NA     1.53336     32.2167   \n",
       "GCA_003798225.1_ASM379822v1_genomic       NA    -2.98143     23.8223   \n",
       "GCA_000835005.1_ASM83500v1_genomic   Nairobi     1.44197     38.4314   \n",
       "GCA_000013825.1_ASM1382v1_genomic         NA    -2.98143     23.8223   \n",
       "NODE0                                     NA          NA          NA   \n",
       "NODE1                                     NA          NA          NA   \n",
       "NODE2                                     NA          NA          NA   \n",
       "\n",
       "                                    province_lat province_lon        strain  \\\n",
       "sample                                                                        \n",
       "GCA_000169615.1_ASM16961v1_genomic            NA           NA     UG05-0454   \n",
       "GCA_003798225.1_ASM379822v1_genomic           NA           NA  FDAARGOS_601   \n",
       "GCA_000835005.1_ASM83500v1_genomic      -1.30317      36.8261       Nairobi   \n",
       "GCA_000013825.1_ASM1382v1_genomic             NA           NA       Antiqua   \n",
       "NODE0                                         NA           NA            NA   \n",
       "NODE1                                         NA           NA            NA   \n",
       "NODE2                                         NA           NA            NA   \n",
       "\n",
       "                                    branch_major   biovar host_human    date  \\\n",
       "sample                                                                         \n",
       "GCA_000169615.1_ASM16961v1_genomic         1.ANT  Antiqua      Human  2004.0   \n",
       "GCA_003798225.1_ASM379822v1_genomic        1.ANT  Antiqua         NA  1954.0   \n",
       "GCA_000835005.1_ASM83500v1_genomic         1.ANT  Antiqua         NA  1985.0   \n",
       "GCA_000013825.1_ASM1382v1_genomic          1.ANT  Antiqua      Human  1965.0   \n",
       "NODE0                                         NA       NA         NA      NA   \n",
       "NODE1                                         NA       NA         NA      NA   \n",
       "NODE2                                         NA       NA         NA      NA   \n",
       "\n",
       "                                    date_bp  \n",
       "sample                                       \n",
       "GCA_000169615.1_ASM16961v1_genomic      -17  \n",
       "GCA_003798225.1_ASM379822v1_genomic     -67  \n",
       "GCA_000835005.1_ASM83500v1_genomic      -36  \n",
       "GCA_000013825.1_ASM1382v1_genomic       -56  \n",
       "NODE0                                    NA  \n",
       "NODE1                                    NA  \n",
       "NODE2                                    NA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA5818830</th>\n",
       "      <td>terminal</td>\n",
       "      <td>0.080411</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1560.0, 1560.0]</td>\n",
       "      <td>3.588157e-08</td>\n",
       "      <td>0.151747</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>STN021</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>[1485:1635]</td>\n",
       "      <td>[-536:-386]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818829</th>\n",
       "      <td>terminal</td>\n",
       "      <td>1.728035</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1560.0, 1560.0]</td>\n",
       "      <td>3.372235e-08</td>\n",
       "      <td>0.142615</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>STN020</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>[1485:1635]</td>\n",
       "      <td>[-536:-386]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818828</th>\n",
       "      <td>terminal</td>\n",
       "      <td>1.728035</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1560.0, 1560.0]</td>\n",
       "      <td>3.454374e-08</td>\n",
       "      <td>0.146089</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>STN019</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>[1485:1635]</td>\n",
       "      <td>[-536:-386]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818826</th>\n",
       "      <td>terminal</td>\n",
       "      <td>0.023473</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1560.0, 1560.0]</td>\n",
       "      <td>3.387151e-08</td>\n",
       "      <td>0.143246</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>STN014</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>[1485:1635]</td>\n",
       "      <td>[-536:-386]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5818825</th>\n",
       "      <td>terminal</td>\n",
       "      <td>0.080411</td>\n",
       "      <td>1560.000000</td>\n",
       "      <td>[1560.0, 1560.0]</td>\n",
       "      <td>1.428955e-07</td>\n",
       "      <td>0.604319</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Nidwalden</td>\n",
       "      <td>46.8133</td>\n",
       "      <td>8.44495</td>\n",
       "      <td>46.9428</td>\n",
       "      <td>8.41198</td>\n",
       "      <td>STN013</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>Human</td>\n",
       "      <td>[1485:1635]</td>\n",
       "      <td>[-536:-386]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE34</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.944394</td>\n",
       "      <td>1631.712419</td>\n",
       "      <td>[1614.2625699793778, 1647.1824752526363]</td>\n",
       "      <td>5.042076e-08</td>\n",
       "      <td>0.213234</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE35</th>\n",
       "      <td>internal</td>\n",
       "      <td>80.544459</td>\n",
       "      <td>1690.388727</td>\n",
       "      <td>[1655.2814046941703, 1716.0895381907403]</td>\n",
       "      <td>1.671410e-08</td>\n",
       "      <td>0.070686</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE36</th>\n",
       "      <td>internal</td>\n",
       "      <td>17.041200</td>\n",
       "      <td>1708.538003</td>\n",
       "      <td>[1693.4645716880063, 1719.1929733173438]</td>\n",
       "      <td>5.247133e-08</td>\n",
       "      <td>0.221906</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE37</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.765094</td>\n",
       "      <td>1716.336286</td>\n",
       "      <td>[1708.8120020077208, 1720.7597334210177]</td>\n",
       "      <td>3.043283e-08</td>\n",
       "      <td>0.128703</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE38</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.435811</td>\n",
       "      <td>1716.542780</td>\n",
       "      <td>[1708.6363056992125, 1720.8964271898692]</td>\n",
       "      <td>3.067928e-08</td>\n",
       "      <td>0.129746</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             node_type  branch_length  timetree_num_date  \\\n",
       "sample                                                     \n",
       "SAMEA5818830  terminal       0.080411        1560.000000   \n",
       "SAMEA5818829  terminal       1.728035        1560.000000   \n",
       "SAMEA5818828  terminal       1.728035        1560.000000   \n",
       "SAMEA5818826  terminal       0.023473        1560.000000   \n",
       "SAMEA5818825  terminal       0.080411        1560.000000   \n",
       "...                ...            ...                ...   \n",
       "NODE34        internal       2.944394        1631.712419   \n",
       "NODE35        internal      80.544459        1690.388727   \n",
       "NODE36        internal      17.041200        1708.538003   \n",
       "NODE37        internal       1.765094        1716.336286   \n",
       "NODE38        internal       0.435811        1716.542780   \n",
       "\n",
       "                          timetree_num_date_confidence    rate_beast  \\\n",
       "sample                                                                 \n",
       "SAMEA5818830                          [1560.0, 1560.0]  3.588157e-08   \n",
       "SAMEA5818829                          [1560.0, 1560.0]  3.372235e-08   \n",
       "SAMEA5818828                          [1560.0, 1560.0]  3.454374e-08   \n",
       "SAMEA5818826                          [1560.0, 1560.0]  3.387151e-08   \n",
       "SAMEA5818825                          [1560.0, 1560.0]  1.428955e-07   \n",
       "...                                                ...           ...   \n",
       "NODE34        [1614.2625699793778, 1647.1824752526363]  5.042076e-08   \n",
       "NODE35        [1655.2814046941703, 1716.0895381907403]  1.671410e-08   \n",
       "NODE36        [1693.4645716880063, 1719.1929733173438]  5.247133e-08   \n",
       "NODE37        [1708.8120020077208, 1720.7597334210177]  3.043283e-08   \n",
       "NODE38        [1708.6363056992125, 1720.8964271898692]  3.067928e-08   \n",
       "\n",
       "              rate_subs_beast      country   province country_lat country_lon  \\\n",
       "sample                                                                          \n",
       "SAMEA5818830         0.151747  Switzerland  Nidwalden     46.8133     8.44495   \n",
       "SAMEA5818829         0.142615  Switzerland  Nidwalden     46.8133     8.44495   \n",
       "SAMEA5818828         0.146089  Switzerland  Nidwalden     46.8133     8.44495   \n",
       "SAMEA5818826         0.143246  Switzerland  Nidwalden     46.8133     8.44495   \n",
       "SAMEA5818825         0.604319  Switzerland  Nidwalden     46.8133     8.44495   \n",
       "...                       ...          ...        ...         ...         ...   \n",
       "NODE34               0.213234           NA         NA          NA          NA   \n",
       "NODE35               0.070686           NA         NA          NA          NA   \n",
       "NODE36               0.221906           NA         NA          NA          NA   \n",
       "NODE37               0.128703           NA         NA          NA          NA   \n",
       "NODE38               0.129746           NA         NA          NA          NA   \n",
       "\n",
       "             province_lat province_lon  strain branch_major           biovar  \\\n",
       "sample                                                                         \n",
       "SAMEA5818830      46.9428      8.41198  STN021        1.PRE  Second Pandemic   \n",
       "SAMEA5818829      46.9428      8.41198  STN020        1.PRE  Second Pandemic   \n",
       "SAMEA5818828      46.9428      8.41198  STN019        1.PRE  Second Pandemic   \n",
       "SAMEA5818826      46.9428      8.41198  STN014        1.PRE  Second Pandemic   \n",
       "SAMEA5818825      46.9428      8.41198  STN013        1.PRE  Second Pandemic   \n",
       "...                   ...          ...     ...          ...              ...   \n",
       "NODE34                 NA           NA      NA           NA               NA   \n",
       "NODE35                 NA           NA      NA           NA               NA   \n",
       "NODE36                 NA           NA      NA           NA               NA   \n",
       "NODE37                 NA           NA      NA           NA               NA   \n",
       "NODE38                 NA           NA      NA           NA               NA   \n",
       "\n",
       "             host_human         date      date_bp  \n",
       "sample                                             \n",
       "SAMEA5818830      Human  [1485:1635]  [-536:-386]  \n",
       "SAMEA5818829      Human  [1485:1635]  [-536:-386]  \n",
       "SAMEA5818828      Human  [1485:1635]  [-536:-386]  \n",
       "SAMEA5818826      Human  [1485:1635]  [-536:-386]  \n",
       "SAMEA5818825      Human  [1485:1635]  [-536:-386]  \n",
       "...                 ...          ...          ...  \n",
       "NODE34               NA           NA           NA  \n",
       "NODE35               NA           NA           NA  \n",
       "NODE36               NA           NA           NA  \n",
       "NODE37               NA           NA           NA  \n",
       "NODE38               NA           NA           NA  \n",
       "\n",
       "[79 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>13.009036</td>\n",
       "      <td>1923.000000</td>\n",
       "      <td>[1923.0, 1923.0]</td>\n",
       "      <td>1.678153e-07</td>\n",
       "      <td>0.709707</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>9_10</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Human</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009296005.1_ASM929600v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>0.469889</td>\n",
       "      <td>1953.000000</td>\n",
       "      <td>[1953.0, 1953.0]</td>\n",
       "      <td>8.264115e-08</td>\n",
       "      <td>0.349498</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Chechnya</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.3976</td>\n",
       "      <td>45.6985</td>\n",
       "      <td>C-25</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1953.0</td>\n",
       "      <td>-68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630485.1_ASM863048v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>3.250880</td>\n",
       "      <td>1997.000000</td>\n",
       "      <td>[1996.9999999999998, 1997.0]</td>\n",
       "      <td>7.262572e-07</td>\n",
       "      <td>3.071413</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Kabardino-Balkaria</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.4428</td>\n",
       "      <td>43.4205</td>\n",
       "      <td>C-742</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1997.0</td>\n",
       "      <td>-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630435.1_ASM863043v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>26.020483</td>\n",
       "      <td>1996.000000</td>\n",
       "      <td>[1996.0, 1996.0]</td>\n",
       "      <td>6.860444e-07</td>\n",
       "      <td>2.901349</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td>C-719</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630395.1_ASM863039v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>4.468999</td>\n",
       "      <td>1984.000000</td>\n",
       "      <td>[1983.9999999999998, 1984.0]</td>\n",
       "      <td>5.818262e-07</td>\n",
       "      <td>2.460600</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Republic of Dagestan</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.0883</td>\n",
       "      <td>47.1499</td>\n",
       "      <td>C-528</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1984.0</td>\n",
       "      <td>-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE110</th>\n",
       "      <td>internal</td>\n",
       "      <td>6.738293</td>\n",
       "      <td>1988.877114</td>\n",
       "      <td>[1981.978551608946, 1995.1945509657712]</td>\n",
       "      <td>1.440738e-07</td>\n",
       "      <td>0.609302</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE111</th>\n",
       "      <td>internal</td>\n",
       "      <td>9.988554</td>\n",
       "      <td>2004.492834</td>\n",
       "      <td>[1995.7386247694087, 2008.9988057840776]</td>\n",
       "      <td>7.567338e-08</td>\n",
       "      <td>0.320030</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE112</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.784600</td>\n",
       "      <td>1991.215981</td>\n",
       "      <td>[1985.5990761703022, 1996.0199169572434]</td>\n",
       "      <td>5.944240e-08</td>\n",
       "      <td>0.251388</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE113</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.788837</td>\n",
       "      <td>1993.385854</td>\n",
       "      <td>[1988.484805318999, 1996.8918607345145]</td>\n",
       "      <td>1.672413e-07</td>\n",
       "      <td>0.707280</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE114</th>\n",
       "      <td>internal</td>\n",
       "      <td>6.480830</td>\n",
       "      <td>2008.366315</td>\n",
       "      <td>[1998.5679978421094, 2013.9650310570123]</td>\n",
       "      <td>7.510759e-08</td>\n",
       "      <td>0.317637</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>231 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    node_type  branch_length  \\\n",
       "sample                                                         \n",
       "GCA_009909635.1_ASM990963v1_genomic  terminal      13.009036   \n",
       "GCA_009296005.1_ASM929600v1_genomic  terminal       0.469889   \n",
       "GCA_008630485.1_ASM863048v1_genomic  terminal       3.250880   \n",
       "GCA_008630435.1_ASM863043v1_genomic  terminal      26.020483   \n",
       "GCA_008630395.1_ASM863039v1_genomic  terminal       4.468999   \n",
       "...                                       ...            ...   \n",
       "NODE110                              internal       6.738293   \n",
       "NODE111                              internal       9.988554   \n",
       "NODE112                              internal       2.784600   \n",
       "NODE113                              internal       0.788837   \n",
       "NODE114                              internal       6.480830   \n",
       "\n",
       "                                     timetree_num_date  \\\n",
       "sample                                                   \n",
       "GCA_009909635.1_ASM990963v1_genomic        1923.000000   \n",
       "GCA_009296005.1_ASM929600v1_genomic        1953.000000   \n",
       "GCA_008630485.1_ASM863048v1_genomic        1997.000000   \n",
       "GCA_008630435.1_ASM863043v1_genomic        1996.000000   \n",
       "GCA_008630395.1_ASM863039v1_genomic        1984.000000   \n",
       "...                                                ...   \n",
       "NODE110                                    1988.877114   \n",
       "NODE111                                    2004.492834   \n",
       "NODE112                                    1991.215981   \n",
       "NODE113                                    1993.385854   \n",
       "NODE114                                    2008.366315   \n",
       "\n",
       "                                                 timetree_num_date_confidence  \\\n",
       "sample                                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic                          [1923.0, 1923.0]   \n",
       "GCA_009296005.1_ASM929600v1_genomic                          [1953.0, 1953.0]   \n",
       "GCA_008630485.1_ASM863048v1_genomic              [1996.9999999999998, 1997.0]   \n",
       "GCA_008630435.1_ASM863043v1_genomic                          [1996.0, 1996.0]   \n",
       "GCA_008630395.1_ASM863039v1_genomic              [1983.9999999999998, 1984.0]   \n",
       "...                                                                       ...   \n",
       "NODE110                               [1981.978551608946, 1995.1945509657712]   \n",
       "NODE111                              [1995.7386247694087, 2008.9988057840776]   \n",
       "NODE112                              [1985.5990761703022, 1996.0199169572434]   \n",
       "NODE113                               [1988.484805318999, 1996.8918607345145]   \n",
       "NODE114                              [1998.5679978421094, 2013.9650310570123]   \n",
       "\n",
       "                                       rate_beast  rate_subs_beast country  \\\n",
       "sample                                                                       \n",
       "GCA_009909635.1_ASM990963v1_genomic  1.678153e-07         0.709707  Russia   \n",
       "GCA_009296005.1_ASM929600v1_genomic  8.264115e-08         0.349498  Russia   \n",
       "GCA_008630485.1_ASM863048v1_genomic  7.262572e-07         3.071413  Russia   \n",
       "GCA_008630435.1_ASM863043v1_genomic  6.860444e-07         2.901349  Russia   \n",
       "GCA_008630395.1_ASM863039v1_genomic  5.818262e-07         2.460600  Russia   \n",
       "...                                           ...              ...     ...   \n",
       "NODE110                              1.440738e-07         0.609302      NA   \n",
       "NODE111                              7.567338e-08         0.320030      NA   \n",
       "NODE112                              5.944240e-08         0.251388      NA   \n",
       "NODE113                              1.672413e-07         0.707280      NA   \n",
       "NODE114                              7.510759e-08         0.317637      NA   \n",
       "\n",
       "                                                 province country_lat  \\\n",
       "sample                                                                  \n",
       "GCA_009909635.1_ASM990963v1_genomic         Rostov Oblast     64.6863   \n",
       "GCA_009296005.1_ASM929600v1_genomic              Chechnya     64.6863   \n",
       "GCA_008630485.1_ASM863048v1_genomic    Kabardino-Balkaria     64.6863   \n",
       "GCA_008630435.1_ASM863043v1_genomic   Karachay-Cherkessia     64.6863   \n",
       "GCA_008630395.1_ASM863039v1_genomic  Republic of Dagestan     64.6863   \n",
       "...                                                   ...         ...   \n",
       "NODE110                                                NA          NA   \n",
       "NODE111                                                NA          NA   \n",
       "NODE112                                                NA          NA   \n",
       "NODE113                                                NA          NA   \n",
       "NODE114                                                NA          NA   \n",
       "\n",
       "                                    country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "GCA_009909635.1_ASM990963v1_genomic     97.7453      47.6222      40.7958   \n",
       "GCA_009296005.1_ASM929600v1_genomic     97.7453      43.3976      45.6985   \n",
       "GCA_008630485.1_ASM863048v1_genomic     97.7453      43.4428      43.4205   \n",
       "GCA_008630435.1_ASM863043v1_genomic     97.7453      43.7368      41.7268   \n",
       "GCA_008630395.1_ASM863039v1_genomic     97.7453      43.0883      47.1499   \n",
       "...                                         ...          ...          ...   \n",
       "NODE110                                      NA           NA           NA   \n",
       "NODE111                                      NA           NA           NA   \n",
       "NODE112                                      NA           NA           NA   \n",
       "NODE113                                      NA           NA           NA   \n",
       "NODE114                                      NA           NA           NA   \n",
       "\n",
       "                                    strain branch_major      biovar  \\\n",
       "sample                                                                \n",
       "GCA_009909635.1_ASM990963v1_genomic   9_10        2.MED  Medievalis   \n",
       "GCA_009296005.1_ASM929600v1_genomic   C-25        2.MED  Medievalis   \n",
       "GCA_008630485.1_ASM863048v1_genomic  C-742        2.MED  Medievalis   \n",
       "GCA_008630435.1_ASM863043v1_genomic  C-719        2.MED  Medievalis   \n",
       "GCA_008630395.1_ASM863039v1_genomic  C-528        2.MED  Medievalis   \n",
       "...                                    ...          ...         ...   \n",
       "NODE110                                 NA           NA          NA   \n",
       "NODE111                                 NA           NA          NA   \n",
       "NODE112                                 NA           NA          NA   \n",
       "NODE113                                 NA           NA          NA   \n",
       "NODE114                                 NA           NA          NA   \n",
       "\n",
       "                                    host_human    date date_bp  \n",
       "sample                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic      Human  1923.0     -98  \n",
       "GCA_009296005.1_ASM929600v1_genomic  Non-Human  1953.0     -68  \n",
       "GCA_008630485.1_ASM863048v1_genomic  Non-Human  1997.0     -24  \n",
       "GCA_008630435.1_ASM863043v1_genomic  Non-Human  1996.0     -25  \n",
       "GCA_008630395.1_ASM863039v1_genomic  Non-Human  1984.0     -37  \n",
       "...                                        ...     ...     ...  \n",
       "NODE110                                     NA      NA      NA  \n",
       "NODE111                                     NA      NA      NA  \n",
       "NODE112                                     NA      NA      NA  \n",
       "NODE113                                     NA      NA      NA  \n",
       "NODE114                                     NA      NA      NA  \n",
       "\n",
       "[231 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000834275.1_ASM83427v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>8.145852</td>\n",
       "      <td>1940.000000</td>\n",
       "      <td>[1940.0, 1940.0]</td>\n",
       "      <td>3.959404e-06</td>\n",
       "      <td>16.744707</td>\n",
       "      <td>China</td>\n",
       "      <td>Heilongjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>48</td>\n",
       "      <td>128</td>\n",
       "      <td>Harbin35</td>\n",
       "      <td>2.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>1940.0</td>\n",
       "      <td>-81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_002005285.1_ASM200528v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>0.144253</td>\n",
       "      <td>1924.000000</td>\n",
       "      <td>[1924.0, 1924.0]</td>\n",
       "      <td>3.881345e-07</td>\n",
       "      <td>1.641459</td>\n",
       "      <td>India</td>\n",
       "      <td>Maharashtra</td>\n",
       "      <td>22.3511</td>\n",
       "      <td>78.6677</td>\n",
       "      <td>18.9068</td>\n",
       "      <td>75.6742</td>\n",
       "      <td>195/P</td>\n",
       "      <td>2.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[1898:1950]</td>\n",
       "      <td>[-123:-71]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001601285.1_ASM160128v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>46.915533</td>\n",
       "      <td>1970.000000</td>\n",
       "      <td>[1970.0, 1970.0]</td>\n",
       "      <td>3.295522e-08</td>\n",
       "      <td>0.139371</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Zabaykalsky Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>52.2485</td>\n",
       "      <td>115.956</td>\n",
       "      <td>I-1996</td>\n",
       "      <td>2.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1970.0</td>\n",
       "      <td>-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000475135.2_YPS03_v2_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>12.756693</td>\n",
       "      <td>2002.000000</td>\n",
       "      <td>[2002.0, 2002.0]</td>\n",
       "      <td>1.322908e-06</td>\n",
       "      <td>5.594709</td>\n",
       "      <td>India</td>\n",
       "      <td>Himachal Pradesh</td>\n",
       "      <td>22.3511</td>\n",
       "      <td>78.6677</td>\n",
       "      <td>31.8168</td>\n",
       "      <td>77.3493</td>\n",
       "      <td>S3</td>\n",
       "      <td>2.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>2002.0</td>\n",
       "      <td>-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000834885.1_ASM83488v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>76.145852</td>\n",
       "      <td>2008.000000</td>\n",
       "      <td>[2008.0, 2008.0]</td>\n",
       "      <td>2.034777e-07</td>\n",
       "      <td>0.860527</td>\n",
       "      <td>China</td>\n",
       "      <td>Heilongjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>48</td>\n",
       "      <td>128</td>\n",
       "      <td>Nicholisk 41</td>\n",
       "      <td>2.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>NA</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE48</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.014033</td>\n",
       "      <td>1992.403938</td>\n",
       "      <td>[1987.0633803733706, 1995.9881444470643]</td>\n",
       "      <td>4.731970e-08</td>\n",
       "      <td>0.200120</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE49</th>\n",
       "      <td>internal</td>\n",
       "      <td>7.487647</td>\n",
       "      <td>1996.398445</td>\n",
       "      <td>[1989.6421990442122, 2001.937263084358]</td>\n",
       "      <td>2.958198e-08</td>\n",
       "      <td>0.125105</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE50</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.437066</td>\n",
       "      <td>1996.889808</td>\n",
       "      <td>[1989.3174213737238, 2001.9925075131018]</td>\n",
       "      <td>2.068752e-08</td>\n",
       "      <td>0.087490</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE51</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.531915</td>\n",
       "      <td>2000.047771</td>\n",
       "      <td>[1992.943027659938, 2004.9950608365348]</td>\n",
       "      <td>2.862714e-08</td>\n",
       "      <td>0.121067</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE52</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.982837</td>\n",
       "      <td>2002.203465</td>\n",
       "      <td>[1996.468652166283, 2004.998762849756]</td>\n",
       "      <td>3.033474e-08</td>\n",
       "      <td>0.128289</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    node_type  branch_length  \\\n",
       "sample                                                         \n",
       "GCA_000834275.1_ASM83427v1_genomic   terminal       8.145852   \n",
       "GCA_002005285.1_ASM200528v1_genomic  terminal       0.144253   \n",
       "GCA_001601285.1_ASM160128v1_genomic  terminal      46.915533   \n",
       "GCA_000475135.2_YPS03_v2_genomic     terminal      12.756693   \n",
       "GCA_000834885.1_ASM83488v1_genomic   terminal      76.145852   \n",
       "...                                       ...            ...   \n",
       "NODE48                               internal       1.014033   \n",
       "NODE49                               internal       7.487647   \n",
       "NODE50                               internal       1.437066   \n",
       "NODE51                               internal       1.531915   \n",
       "NODE52                               internal       2.982837   \n",
       "\n",
       "                                     timetree_num_date  \\\n",
       "sample                                                   \n",
       "GCA_000834275.1_ASM83427v1_genomic         1940.000000   \n",
       "GCA_002005285.1_ASM200528v1_genomic        1924.000000   \n",
       "GCA_001601285.1_ASM160128v1_genomic        1970.000000   \n",
       "GCA_000475135.2_YPS03_v2_genomic           2002.000000   \n",
       "GCA_000834885.1_ASM83488v1_genomic         2008.000000   \n",
       "...                                                ...   \n",
       "NODE48                                     1992.403938   \n",
       "NODE49                                     1996.398445   \n",
       "NODE50                                     1996.889808   \n",
       "NODE51                                     2000.047771   \n",
       "NODE52                                     2002.203465   \n",
       "\n",
       "                                                 timetree_num_date_confidence  \\\n",
       "sample                                                                          \n",
       "GCA_000834275.1_ASM83427v1_genomic                           [1940.0, 1940.0]   \n",
       "GCA_002005285.1_ASM200528v1_genomic                          [1924.0, 1924.0]   \n",
       "GCA_001601285.1_ASM160128v1_genomic                          [1970.0, 1970.0]   \n",
       "GCA_000475135.2_YPS03_v2_genomic                             [2002.0, 2002.0]   \n",
       "GCA_000834885.1_ASM83488v1_genomic                           [2008.0, 2008.0]   \n",
       "...                                                                       ...   \n",
       "NODE48                               [1987.0633803733706, 1995.9881444470643]   \n",
       "NODE49                                [1989.6421990442122, 2001.937263084358]   \n",
       "NODE50                               [1989.3174213737238, 2001.9925075131018]   \n",
       "NODE51                                [1992.943027659938, 2004.9950608365348]   \n",
       "NODE52                                 [1996.468652166283, 2004.998762849756]   \n",
       "\n",
       "                                       rate_beast  rate_subs_beast country  \\\n",
       "sample                                                                       \n",
       "GCA_000834275.1_ASM83427v1_genomic   3.959404e-06        16.744707   China   \n",
       "GCA_002005285.1_ASM200528v1_genomic  3.881345e-07         1.641459   India   \n",
       "GCA_001601285.1_ASM160128v1_genomic  3.295522e-08         0.139371  Russia   \n",
       "GCA_000475135.2_YPS03_v2_genomic     1.322908e-06         5.594709   India   \n",
       "GCA_000834885.1_ASM83488v1_genomic   2.034777e-07         0.860527   China   \n",
       "...                                           ...              ...     ...   \n",
       "NODE48                               4.731970e-08         0.200120      NA   \n",
       "NODE49                               2.958198e-08         0.125105      NA   \n",
       "NODE50                               2.068752e-08         0.087490      NA   \n",
       "NODE51                               2.862714e-08         0.121067      NA   \n",
       "NODE52                               3.033474e-08         0.128289      NA   \n",
       "\n",
       "                                             province country_lat country_lon  \\\n",
       "sample                                                                          \n",
       "GCA_000834275.1_ASM83427v1_genomic       Heilongjiang     35.0001         105   \n",
       "GCA_002005285.1_ASM200528v1_genomic       Maharashtra     22.3511     78.6677   \n",
       "GCA_001601285.1_ASM160128v1_genomic  Zabaykalsky Krai     64.6863     97.7453   \n",
       "GCA_000475135.2_YPS03_v2_genomic     Himachal Pradesh     22.3511     78.6677   \n",
       "GCA_000834885.1_ASM83488v1_genomic       Heilongjiang     35.0001         105   \n",
       "...                                               ...         ...         ...   \n",
       "NODE48                                             NA          NA          NA   \n",
       "NODE49                                             NA          NA          NA   \n",
       "NODE50                                             NA          NA          NA   \n",
       "NODE51                                             NA          NA          NA   \n",
       "NODE52                                             NA          NA          NA   \n",
       "\n",
       "                                    province_lat province_lon        strain  \\\n",
       "sample                                                                        \n",
       "GCA_000834275.1_ASM83427v1_genomic            48          128      Harbin35   \n",
       "GCA_002005285.1_ASM200528v1_genomic      18.9068      75.6742         195/P   \n",
       "GCA_001601285.1_ASM160128v1_genomic      52.2485      115.956        I-1996   \n",
       "GCA_000475135.2_YPS03_v2_genomic         31.8168      77.3493            S3   \n",
       "GCA_000834885.1_ASM83488v1_genomic            48          128  Nicholisk 41   \n",
       "...                                          ...          ...           ...   \n",
       "NODE48                                        NA           NA            NA   \n",
       "NODE49                                        NA           NA            NA   \n",
       "NODE50                                        NA           NA            NA   \n",
       "NODE51                                        NA           NA            NA   \n",
       "NODE52                                        NA           NA            NA   \n",
       "\n",
       "                                    branch_major   biovar host_human  \\\n",
       "sample                                                                 \n",
       "GCA_000834275.1_ASM83427v1_genomic         2.ANT  Antiqua      Human   \n",
       "GCA_002005285.1_ASM200528v1_genomic        2.ANT  Antiqua      Human   \n",
       "GCA_001601285.1_ASM160128v1_genomic        2.ANT  Antiqua  Non-Human   \n",
       "GCA_000475135.2_YPS03_v2_genomic           2.ANT  Antiqua      Human   \n",
       "GCA_000834885.1_ASM83488v1_genomic         2.ANT  Antiqua         NA   \n",
       "...                                          ...      ...        ...   \n",
       "NODE48                                        NA       NA         NA   \n",
       "NODE49                                        NA       NA         NA   \n",
       "NODE50                                        NA       NA         NA   \n",
       "NODE51                                        NA       NA         NA   \n",
       "NODE52                                        NA       NA         NA   \n",
       "\n",
       "                                            date     date_bp  \n",
       "sample                                                        \n",
       "GCA_000834275.1_ASM83427v1_genomic        1940.0         -81  \n",
       "GCA_002005285.1_ASM200528v1_genomic  [1898:1950]  [-123:-71]  \n",
       "GCA_001601285.1_ASM160128v1_genomic       1970.0         -51  \n",
       "GCA_000475135.2_YPS03_v2_genomic          2002.0         -19  \n",
       "GCA_000834885.1_ASM83488v1_genomic        2008.0         -13  \n",
       "...                                          ...         ...  \n",
       "NODE48                                        NA          NA  \n",
       "NODE49                                        NA          NA  \n",
       "NODE50                                        NA          NA  \n",
       "NODE51                                        NA          NA  \n",
       "NODE52                                        NA          NA  \n",
       "\n",
       "[107 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009363195.1_ASM936319v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>1.242061</td>\n",
       "      <td>1977.000000</td>\n",
       "      <td>[1977.0, 1977.0]</td>\n",
       "      <td>3.121084e-07</td>\n",
       "      <td>1.319937</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Tuva Republic</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>51.4017</td>\n",
       "      <td>93.8583</td>\n",
       "      <td>I-2638</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1977.0</td>\n",
       "      <td>-44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009295945.1_ASM929594v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>0.138360</td>\n",
       "      <td>1984.000000</td>\n",
       "      <td>[1984.0, 1984.0]</td>\n",
       "      <td>7.191196e-07</td>\n",
       "      <td>3.041227</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Tuva Republic</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>51.4017</td>\n",
       "      <td>93.8583</td>\n",
       "      <td>I-3113</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1984.0</td>\n",
       "      <td>-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_003074255.1_ASM307425v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>24.687046</td>\n",
       "      <td>2012.000000</td>\n",
       "      <td>[2012.0, 2012.0]</td>\n",
       "      <td>5.791297e-08</td>\n",
       "      <td>0.244920</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Altai Republic</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>50.7114</td>\n",
       "      <td>86.8572</td>\n",
       "      <td>1454</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_003074315.1_ASM307431v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>3.138360</td>\n",
       "      <td>1987.000000</td>\n",
       "      <td>[1987.0, 1987.0]</td>\n",
       "      <td>7.964326e-08</td>\n",
       "      <td>0.336819</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Tuva Republic</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>51.4017</td>\n",
       "      <td>93.8583</td>\n",
       "      <td>I-3223</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1987.0</td>\n",
       "      <td>-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_003074235.1_ASM307423v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>3.891382</td>\n",
       "      <td>1988.000000</td>\n",
       "      <td>[1988.0, 1988.0]</td>\n",
       "      <td>1.957728e-06</td>\n",
       "      <td>8.279425</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Bayan-Ölgii</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>48.547</td>\n",
       "      <td>89.8549</td>\n",
       "      <td>I-3244</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1988.0</td>\n",
       "      <td>-33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_003074215.1_ASM307421v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>0.718675</td>\n",
       "      <td>2015.000000</td>\n",
       "      <td>[2015.0, 2015.0]</td>\n",
       "      <td>3.394977e-08</td>\n",
       "      <td>0.143577</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Altai Republic</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>50.7114</td>\n",
       "      <td>86.8572</td>\n",
       "      <td>349</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_003074195.1_ASM307419v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>0.718675</td>\n",
       "      <td>2015.000000</td>\n",
       "      <td>[2015.0, 2015.0]</td>\n",
       "      <td>6.296806e-08</td>\n",
       "      <td>0.266298</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Altai Republic</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>50.7114</td>\n",
       "      <td>86.8572</td>\n",
       "      <td>337</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>NA</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_003074155.1_ASM307415v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>1.670902</td>\n",
       "      <td>2015.000000</td>\n",
       "      <td>[2015.0, 2015.0]</td>\n",
       "      <td>3.472378e-08</td>\n",
       "      <td>0.146850</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Altai Republic</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>50.7114</td>\n",
       "      <td>86.8572</td>\n",
       "      <td>338</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_003086015.1_ASM308601v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>36.242061</td>\n",
       "      <td>2012.000000</td>\n",
       "      <td>[2012.0, 2012.0]</td>\n",
       "      <td>2.691015e-08</td>\n",
       "      <td>0.113806</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Tuva Republic</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>51.4017</td>\n",
       "      <td>93.8583</td>\n",
       "      <td>M-1944</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_003085915.1_ASM308591v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>8.253140</td>\n",
       "      <td>2014.000000</td>\n",
       "      <td>[2014.0, 2014.0]</td>\n",
       "      <td>2.957504e-07</td>\n",
       "      <td>1.250757</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Altai Republic</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>50.7114</td>\n",
       "      <td>86.8572</td>\n",
       "      <td>517</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000325225.1_ASM32522v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>23.180183</td>\n",
       "      <td>2002.000000</td>\n",
       "      <td>[2002.0, 2002.0]</td>\n",
       "      <td>1.341812e-07</td>\n",
       "      <td>0.567465</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Bayan-Ölgii</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>48.547</td>\n",
       "      <td>89.8549</td>\n",
       "      <td>MGJZ12</td>\n",
       "      <td>4.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2002.0</td>\n",
       "      <td>-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1927.113141</td>\n",
       "      <td>[1846.7036288560123, 1975.4208770267599]</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>internal</td>\n",
       "      <td>82.471293</td>\n",
       "      <td>1970.048066</td>\n",
       "      <td>[1954.6388498502156, 1976.9987947109212]</td>\n",
       "      <td>4.483042e-07</td>\n",
       "      <td>1.895922</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>internal</td>\n",
       "      <td>31.724589</td>\n",
       "      <td>1965.119300</td>\n",
       "      <td>[1937.6892560020449, 1981.9090257469804]</td>\n",
       "      <td>2.820828e-07</td>\n",
       "      <td>1.192956</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE3</th>\n",
       "      <td>internal</td>\n",
       "      <td>53.808582</td>\n",
       "      <td>1972.083923</td>\n",
       "      <td>[1950.4284467280177, 1983.3743367680318]</td>\n",
       "      <td>3.599239e-07</td>\n",
       "      <td>1.522153</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>internal</td>\n",
       "      <td>5.041823</td>\n",
       "      <td>1979.423317</td>\n",
       "      <td>[1969.552870941631, 1983.9987635807338]</td>\n",
       "      <td>1.629733e-06</td>\n",
       "      <td>6.892302</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>internal</td>\n",
       "      <td>59.097383</td>\n",
       "      <td>1979.923638</td>\n",
       "      <td>[1965.7358539316247, 1987.8579514384953]</td>\n",
       "      <td>1.119827e-07</td>\n",
       "      <td>0.473586</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE6</th>\n",
       "      <td>internal</td>\n",
       "      <td>3.204336</td>\n",
       "      <td>2000.482370</td>\n",
       "      <td>[1986.6115699081618, 2011.583907038533]</td>\n",
       "      <td>3.382658e-08</td>\n",
       "      <td>0.143056</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE7</th>\n",
       "      <td>internal</td>\n",
       "      <td>18.433906</td>\n",
       "      <td>2007.613454</td>\n",
       "      <td>[1998.7240613851786, 2013.6942117954536]</td>\n",
       "      <td>2.743306e-08</td>\n",
       "      <td>0.116017</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE8</th>\n",
       "      <td>internal</td>\n",
       "      <td>7.582238</td>\n",
       "      <td>2010.351388</td>\n",
       "      <td>[2002.5060200738583, 2014.9368329317301]</td>\n",
       "      <td>2.848645e-08</td>\n",
       "      <td>0.120472</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE9</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.952227</td>\n",
       "      <td>2012.529911</td>\n",
       "      <td>[2006.810410065075, 2014.9999324884454]</td>\n",
       "      <td>3.817546e-08</td>\n",
       "      <td>0.161448</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    node_type  branch_length  \\\n",
       "sample                                                         \n",
       "GCA_009363195.1_ASM936319v1_genomic  terminal       1.242061   \n",
       "GCA_009295945.1_ASM929594v1_genomic  terminal       0.138360   \n",
       "GCA_003074255.1_ASM307425v1_genomic  terminal      24.687046   \n",
       "GCA_003074315.1_ASM307431v1_genomic  terminal       3.138360   \n",
       "GCA_003074235.1_ASM307423v1_genomic  terminal       3.891382   \n",
       "GCA_003074215.1_ASM307421v1_genomic  terminal       0.718675   \n",
       "GCA_003074195.1_ASM307419v1_genomic  terminal       0.718675   \n",
       "GCA_003074155.1_ASM307415v1_genomic  terminal       1.670902   \n",
       "GCA_003086015.1_ASM308601v1_genomic  terminal      36.242061   \n",
       "GCA_003085915.1_ASM308591v1_genomic  terminal       8.253140   \n",
       "GCA_000325225.1_ASM32522v1_genomic   terminal      23.180183   \n",
       "NODE0                                internal       0.000000   \n",
       "NODE1                                internal      82.471293   \n",
       "NODE2                                internal      31.724589   \n",
       "NODE3                                internal      53.808582   \n",
       "NODE4                                internal       5.041823   \n",
       "NODE5                                internal      59.097383   \n",
       "NODE6                                internal       3.204336   \n",
       "NODE7                                internal      18.433906   \n",
       "NODE8                                internal       7.582238   \n",
       "NODE9                                internal       0.952227   \n",
       "\n",
       "                                     timetree_num_date  \\\n",
       "sample                                                   \n",
       "GCA_009363195.1_ASM936319v1_genomic        1977.000000   \n",
       "GCA_009295945.1_ASM929594v1_genomic        1984.000000   \n",
       "GCA_003074255.1_ASM307425v1_genomic        2012.000000   \n",
       "GCA_003074315.1_ASM307431v1_genomic        1987.000000   \n",
       "GCA_003074235.1_ASM307423v1_genomic        1988.000000   \n",
       "GCA_003074215.1_ASM307421v1_genomic        2015.000000   \n",
       "GCA_003074195.1_ASM307419v1_genomic        2015.000000   \n",
       "GCA_003074155.1_ASM307415v1_genomic        2015.000000   \n",
       "GCA_003086015.1_ASM308601v1_genomic        2012.000000   \n",
       "GCA_003085915.1_ASM308591v1_genomic        2014.000000   \n",
       "GCA_000325225.1_ASM32522v1_genomic         2002.000000   \n",
       "NODE0                                      1927.113141   \n",
       "NODE1                                      1970.048066   \n",
       "NODE2                                      1965.119300   \n",
       "NODE3                                      1972.083923   \n",
       "NODE4                                      1979.423317   \n",
       "NODE5                                      1979.923638   \n",
       "NODE6                                      2000.482370   \n",
       "NODE7                                      2007.613454   \n",
       "NODE8                                      2010.351388   \n",
       "NODE9                                      2012.529911   \n",
       "\n",
       "                                                 timetree_num_date_confidence  \\\n",
       "sample                                                                          \n",
       "GCA_009363195.1_ASM936319v1_genomic                          [1977.0, 1977.0]   \n",
       "GCA_009295945.1_ASM929594v1_genomic                          [1984.0, 1984.0]   \n",
       "GCA_003074255.1_ASM307425v1_genomic                          [2012.0, 2012.0]   \n",
       "GCA_003074315.1_ASM307431v1_genomic                          [1987.0, 1987.0]   \n",
       "GCA_003074235.1_ASM307423v1_genomic                          [1988.0, 1988.0]   \n",
       "GCA_003074215.1_ASM307421v1_genomic                          [2015.0, 2015.0]   \n",
       "GCA_003074195.1_ASM307419v1_genomic                          [2015.0, 2015.0]   \n",
       "GCA_003074155.1_ASM307415v1_genomic                          [2015.0, 2015.0]   \n",
       "GCA_003086015.1_ASM308601v1_genomic                          [2012.0, 2012.0]   \n",
       "GCA_003085915.1_ASM308591v1_genomic                          [2014.0, 2014.0]   \n",
       "GCA_000325225.1_ASM32522v1_genomic                           [2002.0, 2002.0]   \n",
       "NODE0                                [1846.7036288560123, 1975.4208770267599]   \n",
       "NODE1                                [1954.6388498502156, 1976.9987947109212]   \n",
       "NODE2                                [1937.6892560020449, 1981.9090257469804]   \n",
       "NODE3                                [1950.4284467280177, 1983.3743367680318]   \n",
       "NODE4                                 [1969.552870941631, 1983.9987635807338]   \n",
       "NODE5                                [1965.7358539316247, 1987.8579514384953]   \n",
       "NODE6                                 [1986.6115699081618, 2011.583907038533]   \n",
       "NODE7                                [1998.7240613851786, 2013.6942117954536]   \n",
       "NODE8                                [2002.5060200738583, 2014.9368329317301]   \n",
       "NODE9                                 [2006.810410065075, 2014.9999324884454]   \n",
       "\n",
       "                                       rate_beast  rate_subs_beast   country  \\\n",
       "sample                                                                         \n",
       "GCA_009363195.1_ASM936319v1_genomic  3.121084e-07         1.319937    Russia   \n",
       "GCA_009295945.1_ASM929594v1_genomic  7.191196e-07         3.041227    Russia   \n",
       "GCA_003074255.1_ASM307425v1_genomic  5.791297e-08         0.244920    Russia   \n",
       "GCA_003074315.1_ASM307431v1_genomic  7.964326e-08         0.336819    Russia   \n",
       "GCA_003074235.1_ASM307423v1_genomic  1.957728e-06         8.279425  Mongolia   \n",
       "GCA_003074215.1_ASM307421v1_genomic  3.394977e-08         0.143577    Russia   \n",
       "GCA_003074195.1_ASM307419v1_genomic  6.296806e-08         0.266298    Russia   \n",
       "GCA_003074155.1_ASM307415v1_genomic  3.472378e-08         0.146850    Russia   \n",
       "GCA_003086015.1_ASM308601v1_genomic  2.691015e-08         0.113806    Russia   \n",
       "GCA_003085915.1_ASM308591v1_genomic  2.957504e-07         1.250757    Russia   \n",
       "GCA_000325225.1_ASM32522v1_genomic   1.341812e-07         0.567465  Mongolia   \n",
       "NODE0                                0.000000e+00         0.000000        NA   \n",
       "NODE1                                4.483042e-07         1.895922        NA   \n",
       "NODE2                                2.820828e-07         1.192956        NA   \n",
       "NODE3                                3.599239e-07         1.522153        NA   \n",
       "NODE4                                1.629733e-06         6.892302        NA   \n",
       "NODE5                                1.119827e-07         0.473586        NA   \n",
       "NODE6                                3.382658e-08         0.143056        NA   \n",
       "NODE7                                2.743306e-08         0.116017        NA   \n",
       "NODE8                                2.848645e-08         0.120472        NA   \n",
       "NODE9                                3.817546e-08         0.161448        NA   \n",
       "\n",
       "                                           province country_lat country_lon  \\\n",
       "sample                                                                        \n",
       "GCA_009363195.1_ASM936319v1_genomic   Tuva Republic     64.6863     97.7453   \n",
       "GCA_009295945.1_ASM929594v1_genomic   Tuva Republic     64.6863     97.7453   \n",
       "GCA_003074255.1_ASM307425v1_genomic  Altai Republic     64.6863     97.7453   \n",
       "GCA_003074315.1_ASM307431v1_genomic   Tuva Republic     64.6863     97.7453   \n",
       "GCA_003074235.1_ASM307423v1_genomic     Bayan-Ölgii      46.825      103.85   \n",
       "GCA_003074215.1_ASM307421v1_genomic  Altai Republic     64.6863     97.7453   \n",
       "GCA_003074195.1_ASM307419v1_genomic  Altai Republic     64.6863     97.7453   \n",
       "GCA_003074155.1_ASM307415v1_genomic  Altai Republic     64.6863     97.7453   \n",
       "GCA_003086015.1_ASM308601v1_genomic   Tuva Republic     64.6863     97.7453   \n",
       "GCA_003085915.1_ASM308591v1_genomic  Altai Republic     64.6863     97.7453   \n",
       "GCA_000325225.1_ASM32522v1_genomic      Bayan-Ölgii      46.825      103.85   \n",
       "NODE0                                            NA          NA          NA   \n",
       "NODE1                                            NA          NA          NA   \n",
       "NODE2                                            NA          NA          NA   \n",
       "NODE3                                            NA          NA          NA   \n",
       "NODE4                                            NA          NA          NA   \n",
       "NODE5                                            NA          NA          NA   \n",
       "NODE6                                            NA          NA          NA   \n",
       "NODE7                                            NA          NA          NA   \n",
       "NODE8                                            NA          NA          NA   \n",
       "NODE9                                            NA          NA          NA   \n",
       "\n",
       "                                    province_lat province_lon  strain  \\\n",
       "sample                                                                  \n",
       "GCA_009363195.1_ASM936319v1_genomic      51.4017      93.8583  I-2638   \n",
       "GCA_009295945.1_ASM929594v1_genomic      51.4017      93.8583  I-3113   \n",
       "GCA_003074255.1_ASM307425v1_genomic      50.7114      86.8572    1454   \n",
       "GCA_003074315.1_ASM307431v1_genomic      51.4017      93.8583  I-3223   \n",
       "GCA_003074235.1_ASM307423v1_genomic       48.547      89.8549  I-3244   \n",
       "GCA_003074215.1_ASM307421v1_genomic      50.7114      86.8572     349   \n",
       "GCA_003074195.1_ASM307419v1_genomic      50.7114      86.8572     337   \n",
       "GCA_003074155.1_ASM307415v1_genomic      50.7114      86.8572     338   \n",
       "GCA_003086015.1_ASM308601v1_genomic      51.4017      93.8583  M-1944   \n",
       "GCA_003085915.1_ASM308591v1_genomic      50.7114      86.8572     517   \n",
       "GCA_000325225.1_ASM32522v1_genomic        48.547      89.8549  MGJZ12   \n",
       "NODE0                                         NA           NA      NA   \n",
       "NODE1                                         NA           NA      NA   \n",
       "NODE2                                         NA           NA      NA   \n",
       "NODE3                                         NA           NA      NA   \n",
       "NODE4                                         NA           NA      NA   \n",
       "NODE5                                         NA           NA      NA   \n",
       "NODE6                                         NA           NA      NA   \n",
       "NODE7                                         NA           NA      NA   \n",
       "NODE8                                         NA           NA      NA   \n",
       "NODE9                                         NA           NA      NA   \n",
       "\n",
       "                                    branch_major   biovar host_human    date  \\\n",
       "sample                                                                         \n",
       "GCA_009363195.1_ASM936319v1_genomic        4.ANT  Antiqua  Non-Human  1977.0   \n",
       "GCA_009295945.1_ASM929594v1_genomic        4.ANT  Antiqua  Non-Human  1984.0   \n",
       "GCA_003074255.1_ASM307425v1_genomic        4.ANT  Antiqua  Non-Human  2012.0   \n",
       "GCA_003074315.1_ASM307431v1_genomic        4.ANT  Antiqua  Non-Human  1987.0   \n",
       "GCA_003074235.1_ASM307423v1_genomic        4.ANT  Antiqua  Non-Human  1988.0   \n",
       "GCA_003074215.1_ASM307421v1_genomic        4.ANT  Antiqua  Non-Human  2015.0   \n",
       "GCA_003074195.1_ASM307419v1_genomic        4.ANT  Antiqua         NA  2015.0   \n",
       "GCA_003074155.1_ASM307415v1_genomic        4.ANT  Antiqua  Non-Human  2015.0   \n",
       "GCA_003086015.1_ASM308601v1_genomic        4.ANT  Antiqua  Non-Human  2012.0   \n",
       "GCA_003085915.1_ASM308591v1_genomic        4.ANT  Antiqua      Human  2014.0   \n",
       "GCA_000325225.1_ASM32522v1_genomic         4.ANT  Antiqua  Non-Human  2002.0   \n",
       "NODE0                                         NA       NA         NA      NA   \n",
       "NODE1                                         NA       NA         NA      NA   \n",
       "NODE2                                         NA       NA         NA      NA   \n",
       "NODE3                                         NA       NA         NA      NA   \n",
       "NODE4                                         NA       NA         NA      NA   \n",
       "NODE5                                         NA       NA         NA      NA   \n",
       "NODE6                                         NA       NA         NA      NA   \n",
       "NODE7                                         NA       NA         NA      NA   \n",
       "NODE8                                         NA       NA         NA      NA   \n",
       "NODE9                                         NA       NA         NA      NA   \n",
       "\n",
       "                                    date_bp  \n",
       "sample                                       \n",
       "GCA_009363195.1_ASM936319v1_genomic     -44  \n",
       "GCA_009295945.1_ASM929594v1_genomic     -37  \n",
       "GCA_003074255.1_ASM307425v1_genomic      -9  \n",
       "GCA_003074315.1_ASM307431v1_genomic     -34  \n",
       "GCA_003074235.1_ASM307423v1_genomic     -33  \n",
       "GCA_003074215.1_ASM307421v1_genomic      -6  \n",
       "GCA_003074195.1_ASM307419v1_genomic      -6  \n",
       "GCA_003074155.1_ASM307415v1_genomic      -6  \n",
       "GCA_003086015.1_ASM308601v1_genomic      -9  \n",
       "GCA_003085915.1_ASM308591v1_genomic      -7  \n",
       "GCA_000325225.1_ASM32522v1_genomic      -19  \n",
       "NODE0                                    NA  \n",
       "NODE1                                    NA  \n",
       "NODE2                                    NA  \n",
       "NODE3                                    NA  \n",
       "NODE4                                    NA  \n",
       "NODE5                                    NA  \n",
       "NODE6                                    NA  \n",
       "NODE7                                    NA  \n",
       "NODE8                                    NA  \n",
       "NODE9                                    NA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000325305.1_ASM32530v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>4.884258</td>\n",
       "      <td>1998.000000</td>\n",
       "      <td>[1998.0, 1998.0]</td>\n",
       "      <td>5.196203e-08</td>\n",
       "      <td>0.219753</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Govi-Altai</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>45.3463</td>\n",
       "      <td>95.9326</td>\n",
       "      <td>MGJZ9</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1998.0</td>\n",
       "      <td>-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000325285.1_ASM32528v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>4.504057</td>\n",
       "      <td>1997.000000</td>\n",
       "      <td>[1997.0, 1997.0]</td>\n",
       "      <td>4.693787e-07</td>\n",
       "      <td>1.985049</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Dornogovi</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>44.418</td>\n",
       "      <td>109.778</td>\n",
       "      <td>MGJZ7</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1997.0</td>\n",
       "      <td>-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000325265.1_ASM32526v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>4.504057</td>\n",
       "      <td>1997.000000</td>\n",
       "      <td>[1997.0, 1997.0]</td>\n",
       "      <td>1.792326e-07</td>\n",
       "      <td>0.757992</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Dornogovi</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>44.418</td>\n",
       "      <td>109.778</td>\n",
       "      <td>MGJZ6</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1997.0</td>\n",
       "      <td>-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000325245.1_ASM32524v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>3.242909</td>\n",
       "      <td>1980.000000</td>\n",
       "      <td>[1980.0, 1980.0]</td>\n",
       "      <td>4.604132e-06</td>\n",
       "      <td>19.471327</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Govi-Altai</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>45.3463</td>\n",
       "      <td>95.9326</td>\n",
       "      <td>MGJZ3</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1980.0</td>\n",
       "      <td>-41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000325205.1_ASM32520v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>6.884258</td>\n",
       "      <td>2000.000000</td>\n",
       "      <td>[2000.0, 2000.0]</td>\n",
       "      <td>1.895825e-07</td>\n",
       "      <td>0.801763</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Bayan-Ölgii</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>48.547</td>\n",
       "      <td>89.8549</td>\n",
       "      <td>MGJZ11</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000324325.1_ASM32432v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>12.036414</td>\n",
       "      <td>1961.000000</td>\n",
       "      <td>[1961.0, 1961.0]</td>\n",
       "      <td>3.970867e-07</td>\n",
       "      <td>1.679319</td>\n",
       "      <td>China</td>\n",
       "      <td>Gansu</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>38</td>\n",
       "      <td>102</td>\n",
       "      <td>CMCC71001</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1961.0</td>\n",
       "      <td>-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000323765.1_ASM32376v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>9.721393</td>\n",
       "      <td>1976.000000</td>\n",
       "      <td>[1976.0, 1976.0]</td>\n",
       "      <td>3.723379e-07</td>\n",
       "      <td>1.574654</td>\n",
       "      <td>China</td>\n",
       "      <td>Gansu</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>38</td>\n",
       "      <td>102</td>\n",
       "      <td>C1976001</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>-45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000323525.1_ASM32352v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>22.721393</td>\n",
       "      <td>1989.000000</td>\n",
       "      <td>[1989.0, 1989.0]</td>\n",
       "      <td>4.556191e-08</td>\n",
       "      <td>0.192686</td>\n",
       "      <td>China</td>\n",
       "      <td>Gansu</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>38</td>\n",
       "      <td>102</td>\n",
       "      <td>71021</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1989.0</td>\n",
       "      <td>-32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000323505.1_ASM32350v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>47.714694</td>\n",
       "      <td>2004.000000</td>\n",
       "      <td>[2004.0, 2004.0]</td>\n",
       "      <td>4.412225e-08</td>\n",
       "      <td>0.186597</td>\n",
       "      <td>China</td>\n",
       "      <td>Qinghai</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>35.4071</td>\n",
       "      <td>95.9521</td>\n",
       "      <td>7</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_013391945.1_ASM1339194v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>51.326956</td>\n",
       "      <td>2017.000000</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>4.333042e-08</td>\n",
       "      <td>0.183249</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Zavkhan Province</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>48.3444</td>\n",
       "      <td>96.6954</td>\n",
       "      <td>3256</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_013391965.1_ASM1339196v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>58.342383</td>\n",
       "      <td>2016.000000</td>\n",
       "      <td>[2016.0, 2016.0]</td>\n",
       "      <td>4.488485e-08</td>\n",
       "      <td>0.189822</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Hovsgel</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>50.2206</td>\n",
       "      <td>99.917</td>\n",
       "      <td>3243</td>\n",
       "      <td>3.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>-5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1885.477451</td>\n",
       "      <td>[1769.253672069358, 1960.4467172107932]</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>internal</td>\n",
       "      <td>3.661834</td>\n",
       "      <td>1948.552787</td>\n",
       "      <td>[1921.8204321382502, 1960.9946451441876]</td>\n",
       "      <td>3.560645e-07</td>\n",
       "      <td>1.505832</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>internal</td>\n",
       "      <td>17.315021</td>\n",
       "      <td>1966.115093</td>\n",
       "      <td>[1950.1492549977556, 1975.9902674435257]</td>\n",
       "      <td>4.613641e-08</td>\n",
       "      <td>0.195115</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE3</th>\n",
       "      <td>internal</td>\n",
       "      <td>10.983553</td>\n",
       "      <td>1947.401435</td>\n",
       "      <td>[1904.8097281908897, 1974.9921310749212]</td>\n",
       "      <td>2.751065e-07</td>\n",
       "      <td>1.163452</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.372311</td>\n",
       "      <td>1957.440073</td>\n",
       "      <td>[1924.5761340474273, 1977.6323139807305]</td>\n",
       "      <td>6.440023e-07</td>\n",
       "      <td>2.723549</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>internal</td>\n",
       "      <td>8.015427</td>\n",
       "      <td>1966.555457</td>\n",
       "      <td>[1945.8905076992496, 1978.8105366050793]</td>\n",
       "      <td>3.348716e-07</td>\n",
       "      <td>1.416205</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE6</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.722931</td>\n",
       "      <td>1972.769391</td>\n",
       "      <td>[1961.21927817965, 1979.79871822773]</td>\n",
       "      <td>5.523391e-07</td>\n",
       "      <td>2.335896</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE7</th>\n",
       "      <td>internal</td>\n",
       "      <td>26.719767</td>\n",
       "      <td>1992.520661</td>\n",
       "      <td>[1981.636252065874, 1997.999883756888]</td>\n",
       "      <td>1.870816e-07</td>\n",
       "      <td>0.791186</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE8</th>\n",
       "      <td>internal</td>\n",
       "      <td>10.361115</td>\n",
       "      <td>1976.646935</td>\n",
       "      <td>[1970.0836661018714, 1979.9902564898262]</td>\n",
       "      <td>2.959227e-07</td>\n",
       "      <td>1.251486</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE9</th>\n",
       "      <td>internal</td>\n",
       "      <td>15.738853</td>\n",
       "      <td>1989.957736</td>\n",
       "      <td>[1980.0773130468538, 1996.9715155592116]</td>\n",
       "      <td>6.713474e-08</td>\n",
       "      <td>0.283919</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     node_type  branch_length  \\\n",
       "sample                                                          \n",
       "GCA_000325305.1_ASM32530v1_genomic    terminal       4.884258   \n",
       "GCA_000325285.1_ASM32528v1_genomic    terminal       4.504057   \n",
       "GCA_000325265.1_ASM32526v1_genomic    terminal       4.504057   \n",
       "GCA_000325245.1_ASM32524v1_genomic    terminal       3.242909   \n",
       "GCA_000325205.1_ASM32520v1_genomic    terminal       6.884258   \n",
       "GCA_000324325.1_ASM32432v1_genomic    terminal      12.036414   \n",
       "GCA_000323765.1_ASM32376v1_genomic    terminal       9.721393   \n",
       "GCA_000323525.1_ASM32352v1_genomic    terminal      22.721393   \n",
       "GCA_000323505.1_ASM32350v1_genomic    terminal      47.714694   \n",
       "GCA_013391945.1_ASM1339194v1_genomic  terminal      51.326956   \n",
       "GCA_013391965.1_ASM1339196v1_genomic  terminal      58.342383   \n",
       "NODE0                                 internal       0.000000   \n",
       "NODE1                                 internal       3.661834   \n",
       "NODE2                                 internal      17.315021   \n",
       "NODE3                                 internal      10.983553   \n",
       "NODE4                                 internal       1.372311   \n",
       "NODE5                                 internal       8.015427   \n",
       "NODE6                                 internal       0.722931   \n",
       "NODE7                                 internal      26.719767   \n",
       "NODE8                                 internal      10.361115   \n",
       "NODE9                                 internal      15.738853   \n",
       "\n",
       "                                      timetree_num_date  \\\n",
       "sample                                                    \n",
       "GCA_000325305.1_ASM32530v1_genomic          1998.000000   \n",
       "GCA_000325285.1_ASM32528v1_genomic          1997.000000   \n",
       "GCA_000325265.1_ASM32526v1_genomic          1997.000000   \n",
       "GCA_000325245.1_ASM32524v1_genomic          1980.000000   \n",
       "GCA_000325205.1_ASM32520v1_genomic          2000.000000   \n",
       "GCA_000324325.1_ASM32432v1_genomic          1961.000000   \n",
       "GCA_000323765.1_ASM32376v1_genomic          1976.000000   \n",
       "GCA_000323525.1_ASM32352v1_genomic          1989.000000   \n",
       "GCA_000323505.1_ASM32350v1_genomic          2004.000000   \n",
       "GCA_013391945.1_ASM1339194v1_genomic        2017.000000   \n",
       "GCA_013391965.1_ASM1339196v1_genomic        2016.000000   \n",
       "NODE0                                       1885.477451   \n",
       "NODE1                                       1948.552787   \n",
       "NODE2                                       1966.115093   \n",
       "NODE3                                       1947.401435   \n",
       "NODE4                                       1957.440073   \n",
       "NODE5                                       1966.555457   \n",
       "NODE6                                       1972.769391   \n",
       "NODE7                                       1992.520661   \n",
       "NODE8                                       1976.646935   \n",
       "NODE9                                       1989.957736   \n",
       "\n",
       "                                                  timetree_num_date_confidence  \\\n",
       "sample                                                                           \n",
       "GCA_000325305.1_ASM32530v1_genomic                            [1998.0, 1998.0]   \n",
       "GCA_000325285.1_ASM32528v1_genomic                            [1997.0, 1997.0]   \n",
       "GCA_000325265.1_ASM32526v1_genomic                            [1997.0, 1997.0]   \n",
       "GCA_000325245.1_ASM32524v1_genomic                            [1980.0, 1980.0]   \n",
       "GCA_000325205.1_ASM32520v1_genomic                            [2000.0, 2000.0]   \n",
       "GCA_000324325.1_ASM32432v1_genomic                            [1961.0, 1961.0]   \n",
       "GCA_000323765.1_ASM32376v1_genomic                            [1976.0, 1976.0]   \n",
       "GCA_000323525.1_ASM32352v1_genomic                            [1989.0, 1989.0]   \n",
       "GCA_000323505.1_ASM32350v1_genomic                            [2004.0, 2004.0]   \n",
       "GCA_013391945.1_ASM1339194v1_genomic                                    [0, 0]   \n",
       "GCA_013391965.1_ASM1339196v1_genomic                          [2016.0, 2016.0]   \n",
       "NODE0                                  [1769.253672069358, 1960.4467172107932]   \n",
       "NODE1                                 [1921.8204321382502, 1960.9946451441876]   \n",
       "NODE2                                 [1950.1492549977556, 1975.9902674435257]   \n",
       "NODE3                                 [1904.8097281908897, 1974.9921310749212]   \n",
       "NODE4                                 [1924.5761340474273, 1977.6323139807305]   \n",
       "NODE5                                 [1945.8905076992496, 1978.8105366050793]   \n",
       "NODE6                                     [1961.21927817965, 1979.79871822773]   \n",
       "NODE7                                   [1981.636252065874, 1997.999883756888]   \n",
       "NODE8                                 [1970.0836661018714, 1979.9902564898262]   \n",
       "NODE9                                 [1980.0773130468538, 1996.9715155592116]   \n",
       "\n",
       "                                        rate_beast  rate_subs_beast   country  \\\n",
       "sample                                                                          \n",
       "GCA_000325305.1_ASM32530v1_genomic    5.196203e-08         0.219753  Mongolia   \n",
       "GCA_000325285.1_ASM32528v1_genomic    4.693787e-07         1.985049  Mongolia   \n",
       "GCA_000325265.1_ASM32526v1_genomic    1.792326e-07         0.757992  Mongolia   \n",
       "GCA_000325245.1_ASM32524v1_genomic    4.604132e-06        19.471327  Mongolia   \n",
       "GCA_000325205.1_ASM32520v1_genomic    1.895825e-07         0.801763  Mongolia   \n",
       "GCA_000324325.1_ASM32432v1_genomic    3.970867e-07         1.679319     China   \n",
       "GCA_000323765.1_ASM32376v1_genomic    3.723379e-07         1.574654     China   \n",
       "GCA_000323525.1_ASM32352v1_genomic    4.556191e-08         0.192686     China   \n",
       "GCA_000323505.1_ASM32350v1_genomic    4.412225e-08         0.186597     China   \n",
       "GCA_013391945.1_ASM1339194v1_genomic  4.333042e-08         0.183249  Mongolia   \n",
       "GCA_013391965.1_ASM1339196v1_genomic  4.488485e-08         0.189822  Mongolia   \n",
       "NODE0                                 0.000000e+00         0.000000        NA   \n",
       "NODE1                                 3.560645e-07         1.505832        NA   \n",
       "NODE2                                 4.613641e-08         0.195115        NA   \n",
       "NODE3                                 2.751065e-07         1.163452        NA   \n",
       "NODE4                                 6.440023e-07         2.723549        NA   \n",
       "NODE5                                 3.348716e-07         1.416205        NA   \n",
       "NODE6                                 5.523391e-07         2.335896        NA   \n",
       "NODE7                                 1.870816e-07         0.791186        NA   \n",
       "NODE8                                 2.959227e-07         1.251486        NA   \n",
       "NODE9                                 6.713474e-08         0.283919        NA   \n",
       "\n",
       "                                              province country_lat  \\\n",
       "sample                                                               \n",
       "GCA_000325305.1_ASM32530v1_genomic          Govi-Altai      46.825   \n",
       "GCA_000325285.1_ASM32528v1_genomic           Dornogovi      46.825   \n",
       "GCA_000325265.1_ASM32526v1_genomic           Dornogovi      46.825   \n",
       "GCA_000325245.1_ASM32524v1_genomic          Govi-Altai      46.825   \n",
       "GCA_000325205.1_ASM32520v1_genomic         Bayan-Ölgii      46.825   \n",
       "GCA_000324325.1_ASM32432v1_genomic               Gansu     35.0001   \n",
       "GCA_000323765.1_ASM32376v1_genomic               Gansu     35.0001   \n",
       "GCA_000323525.1_ASM32352v1_genomic               Gansu     35.0001   \n",
       "GCA_000323505.1_ASM32350v1_genomic             Qinghai     35.0001   \n",
       "GCA_013391945.1_ASM1339194v1_genomic  Zavkhan Province      46.825   \n",
       "GCA_013391965.1_ASM1339196v1_genomic           Hovsgel      46.825   \n",
       "NODE0                                               NA          NA   \n",
       "NODE1                                               NA          NA   \n",
       "NODE2                                               NA          NA   \n",
       "NODE3                                               NA          NA   \n",
       "NODE4                                               NA          NA   \n",
       "NODE5                                               NA          NA   \n",
       "NODE6                                               NA          NA   \n",
       "NODE7                                               NA          NA   \n",
       "NODE8                                               NA          NA   \n",
       "NODE9                                               NA          NA   \n",
       "\n",
       "                                     country_lon province_lat province_lon  \\\n",
       "sample                                                                       \n",
       "GCA_000325305.1_ASM32530v1_genomic        103.85      45.3463      95.9326   \n",
       "GCA_000325285.1_ASM32528v1_genomic        103.85       44.418      109.778   \n",
       "GCA_000325265.1_ASM32526v1_genomic        103.85       44.418      109.778   \n",
       "GCA_000325245.1_ASM32524v1_genomic        103.85      45.3463      95.9326   \n",
       "GCA_000325205.1_ASM32520v1_genomic        103.85       48.547      89.8549   \n",
       "GCA_000324325.1_ASM32432v1_genomic           105           38          102   \n",
       "GCA_000323765.1_ASM32376v1_genomic           105           38          102   \n",
       "GCA_000323525.1_ASM32352v1_genomic           105           38          102   \n",
       "GCA_000323505.1_ASM32350v1_genomic           105      35.4071      95.9521   \n",
       "GCA_013391945.1_ASM1339194v1_genomic      103.85      48.3444      96.6954   \n",
       "GCA_013391965.1_ASM1339196v1_genomic      103.85      50.2206       99.917   \n",
       "NODE0                                         NA           NA           NA   \n",
       "NODE1                                         NA           NA           NA   \n",
       "NODE2                                         NA           NA           NA   \n",
       "NODE3                                         NA           NA           NA   \n",
       "NODE4                                         NA           NA           NA   \n",
       "NODE5                                         NA           NA           NA   \n",
       "NODE6                                         NA           NA           NA   \n",
       "NODE7                                         NA           NA           NA   \n",
       "NODE8                                         NA           NA           NA   \n",
       "NODE9                                         NA           NA           NA   \n",
       "\n",
       "                                         strain branch_major   biovar  \\\n",
       "sample                                                                  \n",
       "GCA_000325305.1_ASM32530v1_genomic        MGJZ9        3.ANT  Antiqua   \n",
       "GCA_000325285.1_ASM32528v1_genomic        MGJZ7        3.ANT  Antiqua   \n",
       "GCA_000325265.1_ASM32526v1_genomic        MGJZ6        3.ANT  Antiqua   \n",
       "GCA_000325245.1_ASM32524v1_genomic        MGJZ3        3.ANT  Antiqua   \n",
       "GCA_000325205.1_ASM32520v1_genomic       MGJZ11        3.ANT  Antiqua   \n",
       "GCA_000324325.1_ASM32432v1_genomic    CMCC71001        3.ANT  Antiqua   \n",
       "GCA_000323765.1_ASM32376v1_genomic     C1976001        3.ANT  Antiqua   \n",
       "GCA_000323525.1_ASM32352v1_genomic        71021        3.ANT  Antiqua   \n",
       "GCA_000323505.1_ASM32350v1_genomic            7        3.ANT  Antiqua   \n",
       "GCA_013391945.1_ASM1339194v1_genomic       3256        3.ANT  Antiqua   \n",
       "GCA_013391965.1_ASM1339196v1_genomic       3243        3.ANT  Antiqua   \n",
       "NODE0                                        NA           NA       NA   \n",
       "NODE1                                        NA           NA       NA   \n",
       "NODE2                                        NA           NA       NA   \n",
       "NODE3                                        NA           NA       NA   \n",
       "NODE4                                        NA           NA       NA   \n",
       "NODE5                                        NA           NA       NA   \n",
       "NODE6                                        NA           NA       NA   \n",
       "NODE7                                        NA           NA       NA   \n",
       "NODE8                                        NA           NA       NA   \n",
       "NODE9                                        NA           NA       NA   \n",
       "\n",
       "                                     host_human    date date_bp  \n",
       "sample                                                           \n",
       "GCA_000325305.1_ASM32530v1_genomic    Non-Human  1998.0     -23  \n",
       "GCA_000325285.1_ASM32528v1_genomic    Non-Human  1997.0     -24  \n",
       "GCA_000325265.1_ASM32526v1_genomic    Non-Human  1997.0     -24  \n",
       "GCA_000325245.1_ASM32524v1_genomic    Non-Human  1980.0     -41  \n",
       "GCA_000325205.1_ASM32520v1_genomic    Non-Human  2000.0     -21  \n",
       "GCA_000324325.1_ASM32432v1_genomic    Non-Human  1961.0     -60  \n",
       "GCA_000323765.1_ASM32376v1_genomic    Non-Human  1976.0     -45  \n",
       "GCA_000323525.1_ASM32352v1_genomic    Non-Human  1989.0     -32  \n",
       "GCA_000323505.1_ASM32350v1_genomic        Human  2004.0     -17  \n",
       "GCA_013391945.1_ASM1339194v1_genomic  Non-Human  2017.0      -4  \n",
       "GCA_013391965.1_ASM1339196v1_genomic  Non-Human  2016.0      -5  \n",
       "NODE0                                        NA      NA      NA  \n",
       "NODE1                                        NA      NA      NA  \n",
       "NODE2                                        NA      NA      NA  \n",
       "NODE3                                        NA      NA      NA  \n",
       "NODE4                                        NA      NA      NA  \n",
       "NODE5                                        NA      NA      NA  \n",
       "NODE6                                        NA      NA      NA  \n",
       "NODE7                                        NA      NA      NA  \n",
       "NODE8                                        NA      NA      NA  \n",
       "NODE9                                        NA      NA      NA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>5.410463</td>\n",
       "      <td>2006.000000</td>\n",
       "      <td>[2006.0, 2006.0000000000002]</td>\n",
       "      <td>1.570504e-08</td>\n",
       "      <td>0.066418</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>42126</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>10.677114</td>\n",
       "      <td>2005.000000</td>\n",
       "      <td>[2004.9999999999998, 2005.0000000000002]</td>\n",
       "      <td>3.697239e-08</td>\n",
       "      <td>0.156360</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>42123</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>2.117711</td>\n",
       "      <td>2005.000000</td>\n",
       "      <td>[2005.0, 2005.0000000000002]</td>\n",
       "      <td>5.413330e-08</td>\n",
       "      <td>0.228935</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>42118</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669605.1_ASM966960v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>2.117711</td>\n",
       "      <td>2005.000000</td>\n",
       "      <td>[2004.9999999999998, 2005.0]</td>\n",
       "      <td>3.525254e-08</td>\n",
       "      <td>0.149086</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>42117</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669625.1_ASM966962v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>13.851436</td>\n",
       "      <td>2005.000000</td>\n",
       "      <td>[2005.0, 2005.0]</td>\n",
       "      <td>1.210787e-08</td>\n",
       "      <td>0.051205</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>42116</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE97</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.563553</td>\n",
       "      <td>1995.240889</td>\n",
       "      <td>[1987.4546828191246, 1998.99577465624]</td>\n",
       "      <td>2.856819e-08</td>\n",
       "      <td>0.120818</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE98</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.375887</td>\n",
       "      <td>1998.403159</td>\n",
       "      <td>[1991.859927842856, 2000.997431030353]</td>\n",
       "      <td>2.572350e-08</td>\n",
       "      <td>0.108787</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE99</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.208979</td>\n",
       "      <td>1993.725629</td>\n",
       "      <td>[1984.0716057453021, 2000.136072806417]</td>\n",
       "      <td>1.762941e-08</td>\n",
       "      <td>0.074557</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE100</th>\n",
       "      <td>internal</td>\n",
       "      <td>3.754150</td>\n",
       "      <td>1996.486777</td>\n",
       "      <td>[1989.4158198927937, 2000.9262507455383]</td>\n",
       "      <td>2.582290e-08</td>\n",
       "      <td>0.109208</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE101</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.436299</td>\n",
       "      <td>1998.423605</td>\n",
       "      <td>[1989.6801838227625, 2004.3187217460081]</td>\n",
       "      <td>2.599009e-08</td>\n",
       "      <td>0.109915</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    node_type  branch_length  \\\n",
       "sample                                                         \n",
       "GCA_009669545.1_ASM966954v1_genomic  terminal       5.410463   \n",
       "GCA_009669555.1_ASM966955v1_genomic  terminal      10.677114   \n",
       "GCA_009669565.1_ASM966956v1_genomic  terminal       2.117711   \n",
       "GCA_009669605.1_ASM966960v1_genomic  terminal       2.117711   \n",
       "GCA_009669625.1_ASM966962v1_genomic  terminal      13.851436   \n",
       "...                                       ...            ...   \n",
       "NODE97                               internal       2.563553   \n",
       "NODE98                               internal       2.375887   \n",
       "NODE99                               internal       1.208979   \n",
       "NODE100                              internal       3.754150   \n",
       "NODE101                              internal       2.436299   \n",
       "\n",
       "                                     timetree_num_date  \\\n",
       "sample                                                   \n",
       "GCA_009669545.1_ASM966954v1_genomic        2006.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic        2005.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic        2005.000000   \n",
       "GCA_009669605.1_ASM966960v1_genomic        2005.000000   \n",
       "GCA_009669625.1_ASM966962v1_genomic        2005.000000   \n",
       "...                                                ...   \n",
       "NODE97                                     1995.240889   \n",
       "NODE98                                     1998.403159   \n",
       "NODE99                                     1993.725629   \n",
       "NODE100                                    1996.486777   \n",
       "NODE101                                    1998.423605   \n",
       "\n",
       "                                                 timetree_num_date_confidence  \\\n",
       "sample                                                                          \n",
       "GCA_009669545.1_ASM966954v1_genomic              [2006.0, 2006.0000000000002]   \n",
       "GCA_009669555.1_ASM966955v1_genomic  [2004.9999999999998, 2005.0000000000002]   \n",
       "GCA_009669565.1_ASM966956v1_genomic              [2005.0, 2005.0000000000002]   \n",
       "GCA_009669605.1_ASM966960v1_genomic              [2004.9999999999998, 2005.0]   \n",
       "GCA_009669625.1_ASM966962v1_genomic                          [2005.0, 2005.0]   \n",
       "...                                                                       ...   \n",
       "NODE97                                 [1987.4546828191246, 1998.99577465624]   \n",
       "NODE98                                 [1991.859927842856, 2000.997431030353]   \n",
       "NODE99                                [1984.0716057453021, 2000.136072806417]   \n",
       "NODE100                              [1989.4158198927937, 2000.9262507455383]   \n",
       "NODE101                              [1989.6801838227625, 2004.3187217460081]   \n",
       "\n",
       "                                       rate_beast  rate_subs_beast country  \\\n",
       "sample                                                                       \n",
       "GCA_009669545.1_ASM966954v1_genomic  1.570504e-08         0.066418   China   \n",
       "GCA_009669555.1_ASM966955v1_genomic  3.697239e-08         0.156360   China   \n",
       "GCA_009669565.1_ASM966956v1_genomic  5.413330e-08         0.228935   China   \n",
       "GCA_009669605.1_ASM966960v1_genomic  3.525254e-08         0.149086   China   \n",
       "GCA_009669625.1_ASM966962v1_genomic  1.210787e-08         0.051205   China   \n",
       "...                                           ...              ...     ...   \n",
       "NODE97                               2.856819e-08         0.120818      NA   \n",
       "NODE98                               2.572350e-08         0.108787      NA   \n",
       "NODE99                               1.762941e-08         0.074557      NA   \n",
       "NODE100                              2.582290e-08         0.109208      NA   \n",
       "NODE101                              2.599009e-08         0.109915      NA   \n",
       "\n",
       "                                     province country_lat country_lon  \\\n",
       "sample                                                                  \n",
       "GCA_009669545.1_ASM966954v1_genomic  Xinjiang     35.0001         105   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Xinjiang     35.0001         105   \n",
       "GCA_009669565.1_ASM966956v1_genomic  Xinjiang     35.0001         105   \n",
       "GCA_009669605.1_ASM966960v1_genomic  Xinjiang     35.0001         105   \n",
       "GCA_009669625.1_ASM966962v1_genomic  Xinjiang     35.0001         105   \n",
       "...                                       ...         ...         ...   \n",
       "NODE97                                     NA          NA          NA   \n",
       "NODE98                                     NA          NA          NA   \n",
       "NODE99                                     NA          NA          NA   \n",
       "NODE100                                    NA          NA          NA   \n",
       "NODE101                                    NA          NA          NA   \n",
       "\n",
       "                                    province_lat province_lon strain  \\\n",
       "sample                                                                 \n",
       "GCA_009669545.1_ASM966954v1_genomic      42.4805      85.4633  42126   \n",
       "GCA_009669555.1_ASM966955v1_genomic      42.4805      85.4633  42123   \n",
       "GCA_009669565.1_ASM966956v1_genomic      42.4805      85.4633  42118   \n",
       "GCA_009669605.1_ASM966960v1_genomic      42.4805      85.4633  42117   \n",
       "GCA_009669625.1_ASM966962v1_genomic      42.4805      85.4633  42116   \n",
       "...                                          ...          ...    ...   \n",
       "NODE97                                        NA           NA     NA   \n",
       "NODE98                                        NA           NA     NA   \n",
       "NODE99                                        NA           NA     NA   \n",
       "NODE100                                       NA           NA     NA   \n",
       "NODE101                                       NA           NA     NA   \n",
       "\n",
       "                                    branch_major   biovar host_human    date  \\\n",
       "sample                                                                         \n",
       "GCA_009669545.1_ASM966954v1_genomic        0.ANT  Antiqua  Non-Human  2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic        0.ANT  Antiqua  Non-Human  2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic        0.ANT  Antiqua  Non-Human  2005.0   \n",
       "GCA_009669605.1_ASM966960v1_genomic        0.ANT  Antiqua  Non-Human  2005.0   \n",
       "GCA_009669625.1_ASM966962v1_genomic        0.ANT  Antiqua  Non-Human  2005.0   \n",
       "...                                          ...      ...        ...     ...   \n",
       "NODE97                                        NA       NA         NA      NA   \n",
       "NODE98                                        NA       NA         NA      NA   \n",
       "NODE99                                        NA       NA         NA      NA   \n",
       "NODE100                                       NA       NA         NA      NA   \n",
       "NODE101                                       NA       NA         NA      NA   \n",
       "\n",
       "                                    date_bp  \n",
       "sample                                       \n",
       "GCA_009669545.1_ASM966954v1_genomic     -15  \n",
       "GCA_009669555.1_ASM966955v1_genomic     -16  \n",
       "GCA_009669565.1_ASM966956v1_genomic     -16  \n",
       "GCA_009669605.1_ASM966960v1_genomic     -16  \n",
       "GCA_009669625.1_ASM966962v1_genomic     -16  \n",
       "...                                     ...  \n",
       "NODE97                                   NA  \n",
       "NODE98                                   NA  \n",
       "NODE99                                   NA  \n",
       "NODE100                                  NA  \n",
       "NODE101                                  NA  \n",
       "\n",
       "[205 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA5661390</th>\n",
       "      <td>terminal</td>\n",
       "      <td>2.897253</td>\n",
       "      <td>478.000000</td>\n",
       "      <td>[477.9999999999999, 478.0000000000001]</td>\n",
       "      <td>1.459943e-06</td>\n",
       "      <td>6.174240</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>UNT004</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[419:537]</td>\n",
       "      <td>[-1602:-1484]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661389</th>\n",
       "      <td>terminal</td>\n",
       "      <td>16.676053</td>\n",
       "      <td>516.000000</td>\n",
       "      <td>[515.9999999999999, 516.0000000000001]</td>\n",
       "      <td>3.461309e-09</td>\n",
       "      <td>0.014638</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>UNT003</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[433:599]</td>\n",
       "      <td>[-1588:-1422]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661385</th>\n",
       "      <td>terminal</td>\n",
       "      <td>24.099582</td>\n",
       "      <td>517.000000</td>\n",
       "      <td>[516.9999999999999, 517.0000000000001]</td>\n",
       "      <td>7.160345e-07</td>\n",
       "      <td>3.028180</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>PET004</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[434:600]</td>\n",
       "      <td>[-1587:-1421]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661384</th>\n",
       "      <td>terminal</td>\n",
       "      <td>96.812017</td>\n",
       "      <td>521.000000</td>\n",
       "      <td>[520.9999999999999, 521.0000000000001]</td>\n",
       "      <td>1.858864e-08</td>\n",
       "      <td>0.078613</td>\n",
       "      <td>Spain</td>\n",
       "      <td>Valencia Community</td>\n",
       "      <td>39.3261</td>\n",
       "      <td>-4.83798</td>\n",
       "      <td>39.682</td>\n",
       "      <td>-0.765441</td>\n",
       "      <td>VAL001</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[432:610]</td>\n",
       "      <td>[-1589:-1411]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661372</th>\n",
       "      <td>terminal</td>\n",
       "      <td>6.632183</td>\n",
       "      <td>765.000000</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>9.442931e-08</td>\n",
       "      <td>0.399351</td>\n",
       "      <td>France</td>\n",
       "      <td>Centre-Loire Valley</td>\n",
       "      <td>46.6034</td>\n",
       "      <td>1.88833</td>\n",
       "      <td>47.549</td>\n",
       "      <td>1.73241</td>\n",
       "      <td>LSD020</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[650:880]</td>\n",
       "      <td>[-1371:-1141]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661367</th>\n",
       "      <td>terminal</td>\n",
       "      <td>6.632183</td>\n",
       "      <td>765.000000</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>1.283400e-07</td>\n",
       "      <td>0.542762</td>\n",
       "      <td>France</td>\n",
       "      <td>Centre-Loire Valley</td>\n",
       "      <td>46.6034</td>\n",
       "      <td>1.88833</td>\n",
       "      <td>47.549</td>\n",
       "      <td>1.73241</td>\n",
       "      <td>LSD001</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[650:880]</td>\n",
       "      <td>[-1371:-1141]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661365</th>\n",
       "      <td>terminal</td>\n",
       "      <td>13.841901</td>\n",
       "      <td>575.000000</td>\n",
       "      <td>[574.9999999999999, 575.0000000000001]</td>\n",
       "      <td>3.400581e-08</td>\n",
       "      <td>0.143814</td>\n",
       "      <td>England</td>\n",
       "      <td>East of England</td>\n",
       "      <td>52.531</td>\n",
       "      <td>-1.26491</td>\n",
       "      <td>52.22</td>\n",
       "      <td>0.487578</td>\n",
       "      <td>EDI004</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[500:650]</td>\n",
       "      <td>[-1521:-1371]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661364</th>\n",
       "      <td>terminal</td>\n",
       "      <td>13.841901</td>\n",
       "      <td>575.000000</td>\n",
       "      <td>[574.9999999999999, 575.0000000000001]</td>\n",
       "      <td>1.977523e-07</td>\n",
       "      <td>0.836314</td>\n",
       "      <td>England</td>\n",
       "      <td>East of England</td>\n",
       "      <td>52.531</td>\n",
       "      <td>-1.26491</td>\n",
       "      <td>52.22</td>\n",
       "      <td>0.487578</td>\n",
       "      <td>EDI003</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[500:650]</td>\n",
       "      <td>[-1521:-1371]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661363</th>\n",
       "      <td>terminal</td>\n",
       "      <td>2.289918</td>\n",
       "      <td>555.500000</td>\n",
       "      <td>[555.4999999999999, 555.5000000000001]</td>\n",
       "      <td>2.059089e-08</td>\n",
       "      <td>0.087081</td>\n",
       "      <td>England</td>\n",
       "      <td>East of England</td>\n",
       "      <td>52.531</td>\n",
       "      <td>-1.26491</td>\n",
       "      <td>52.22</td>\n",
       "      <td>0.487578</td>\n",
       "      <td>EDI001</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[474:637]</td>\n",
       "      <td>[-1547:-1384]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA5661360</th>\n",
       "      <td>terminal</td>\n",
       "      <td>12.176053</td>\n",
       "      <td>511.500000</td>\n",
       "      <td>[511.4999999999999, 511.5000000000001]</td>\n",
       "      <td>1.884688e-07</td>\n",
       "      <td>0.797053</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>DIT003</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[428:595]</td>\n",
       "      <td>[-1593:-1426]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA4354665</th>\n",
       "      <td>terminal</td>\n",
       "      <td>19.506773</td>\n",
       "      <td>498.500000</td>\n",
       "      <td>[498.4999999999999, 498.5000000000001]</td>\n",
       "      <td>1.986100e-08</td>\n",
       "      <td>0.083994</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>AE1175</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[426:571]</td>\n",
       "      <td>[-1595:-1450]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA1061800</th>\n",
       "      <td>terminal</td>\n",
       "      <td>128.645841</td>\n",
       "      <td>237.500000</td>\n",
       "      <td>[237.4999999999999, 237.5000000000001]</td>\n",
       "      <td>3.558060e-07</td>\n",
       "      <td>1.504738</td>\n",
       "      <td>Kyrgyzstan</td>\n",
       "      <td>Issyk-Kul Region</td>\n",
       "      <td>41.5089</td>\n",
       "      <td>74.7241</td>\n",
       "      <td>42.061</td>\n",
       "      <td>78.1707</td>\n",
       "      <td>DA101</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>Human</td>\n",
       "      <td>[214:261]</td>\n",
       "      <td>[-1807:-1760]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>172.516705</td>\n",
       "      <td>[38.78639801892041, 237.49042879623642]</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>internal</td>\n",
       "      <td>187.229118</td>\n",
       "      <td>386.240283</td>\n",
       "      <td>[272.30631903094593, 465.1990236272046]</td>\n",
       "      <td>5.455400e-08</td>\n",
       "      <td>0.230714</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>internal</td>\n",
       "      <td>257.126805</td>\n",
       "      <td>532.424545</td>\n",
       "      <td>[501.066165126286, 555.4956198309944]</td>\n",
       "      <td>3.353718e-09</td>\n",
       "      <td>0.014183</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE3</th>\n",
       "      <td>internal</td>\n",
       "      <td>7.948018</td>\n",
       "      <td>558.440263</td>\n",
       "      <td>[537.1388205148279, 574.7139964562191]</td>\n",
       "      <td>1.167448e-08</td>\n",
       "      <td>0.049373</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>internal</td>\n",
       "      <td>128.104707</td>\n",
       "      <td>438.422017</td>\n",
       "      <td>[385.67369228913856, 474.1183372423146]</td>\n",
       "      <td>6.758988e-08</td>\n",
       "      <td>0.285844</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>internal</td>\n",
       "      <td>48.797822</td>\n",
       "      <td>459.324879</td>\n",
       "      <td>[433.44257515217515, 477.2744945199666]</td>\n",
       "      <td>8.806584e-08</td>\n",
       "      <td>0.372439</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE6</th>\n",
       "      <td>internal</td>\n",
       "      <td>19.914613</td>\n",
       "      <td>501.403098</td>\n",
       "      <td>[474.5394556512549, 516.8060533921721]</td>\n",
       "      <td>2.046727e-09</td>\n",
       "      <td>0.008656</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE7</th>\n",
       "      <td>internal</td>\n",
       "      <td>265.467399</td>\n",
       "      <td>701.985064</td>\n",
       "      <td>[598.9859323869073, 764.1220984520477]</td>\n",
       "      <td>2.302237e-08</td>\n",
       "      <td>0.097364</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE8</th>\n",
       "      <td>internal</td>\n",
       "      <td>2.116941</td>\n",
       "      <td>469.995513</td>\n",
       "      <td>[456.1146121324109, 477.8742477227939]</td>\n",
       "      <td>1.098138e-07</td>\n",
       "      <td>0.464414</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE9</th>\n",
       "      <td>internal</td>\n",
       "      <td>3.890480</td>\n",
       "      <td>488.154908</td>\n",
       "      <td>[473.78646507489646, 498.4991489255475]</td>\n",
       "      <td>5.551144e-09</td>\n",
       "      <td>0.023476</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE10</th>\n",
       "      <td>internal</td>\n",
       "      <td>20.330720</td>\n",
       "      <td>502.721482</td>\n",
       "      <td>[489.569888234374, 511.49264299540516]</td>\n",
       "      <td>6.899829e-09</td>\n",
       "      <td>0.029180</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             node_type  branch_length  timetree_num_date  \\\n",
       "sample                                                     \n",
       "SAMEA5661390  terminal       2.897253         478.000000   \n",
       "SAMEA5661389  terminal      16.676053         516.000000   \n",
       "SAMEA5661385  terminal      24.099582         517.000000   \n",
       "SAMEA5661384  terminal      96.812017         521.000000   \n",
       "SAMEA5661372  terminal       6.632183         765.000000   \n",
       "SAMEA5661367  terminal       6.632183         765.000000   \n",
       "SAMEA5661365  terminal      13.841901         575.000000   \n",
       "SAMEA5661364  terminal      13.841901         575.000000   \n",
       "SAMEA5661363  terminal       2.289918         555.500000   \n",
       "SAMEA5661360  terminal      12.176053         511.500000   \n",
       "SAMEA4354665  terminal      19.506773         498.500000   \n",
       "SAMEA1061800  terminal     128.645841         237.500000   \n",
       "NODE0         internal       0.000000         172.516705   \n",
       "NODE1         internal     187.229118         386.240283   \n",
       "NODE2         internal     257.126805         532.424545   \n",
       "NODE3         internal       7.948018         558.440263   \n",
       "NODE4         internal     128.104707         438.422017   \n",
       "NODE5         internal      48.797822         459.324879   \n",
       "NODE6         internal      19.914613         501.403098   \n",
       "NODE7         internal     265.467399         701.985064   \n",
       "NODE8         internal       2.116941         469.995513   \n",
       "NODE9         internal       3.890480         488.154908   \n",
       "NODE10        internal      20.330720         502.721482   \n",
       "\n",
       "                         timetree_num_date_confidence    rate_beast  \\\n",
       "sample                                                                \n",
       "SAMEA5661390   [477.9999999999999, 478.0000000000001]  1.459943e-06   \n",
       "SAMEA5661389   [515.9999999999999, 516.0000000000001]  3.461309e-09   \n",
       "SAMEA5661385   [516.9999999999999, 517.0000000000001]  7.160345e-07   \n",
       "SAMEA5661384   [520.9999999999999, 521.0000000000001]  1.858864e-08   \n",
       "SAMEA5661372                                   [0, 0]  9.442931e-08   \n",
       "SAMEA5661367                                   [0, 0]  1.283400e-07   \n",
       "SAMEA5661365   [574.9999999999999, 575.0000000000001]  3.400581e-08   \n",
       "SAMEA5661364   [574.9999999999999, 575.0000000000001]  1.977523e-07   \n",
       "SAMEA5661363   [555.4999999999999, 555.5000000000001]  2.059089e-08   \n",
       "SAMEA5661360   [511.4999999999999, 511.5000000000001]  1.884688e-07   \n",
       "SAMEA4354665   [498.4999999999999, 498.5000000000001]  1.986100e-08   \n",
       "SAMEA1061800   [237.4999999999999, 237.5000000000001]  3.558060e-07   \n",
       "NODE0         [38.78639801892041, 237.49042879623642]  0.000000e+00   \n",
       "NODE1         [272.30631903094593, 465.1990236272046]  5.455400e-08   \n",
       "NODE2           [501.066165126286, 555.4956198309944]  3.353718e-09   \n",
       "NODE3          [537.1388205148279, 574.7139964562191]  1.167448e-08   \n",
       "NODE4         [385.67369228913856, 474.1183372423146]  6.758988e-08   \n",
       "NODE5         [433.44257515217515, 477.2744945199666]  8.806584e-08   \n",
       "NODE6          [474.5394556512549, 516.8060533921721]  2.046727e-09   \n",
       "NODE7          [598.9859323869073, 764.1220984520477]  2.302237e-08   \n",
       "NODE8          [456.1146121324109, 477.8742477227939]  1.098138e-07   \n",
       "NODE9         [473.78646507489646, 498.4991489255475]  5.551144e-09   \n",
       "NODE10         [489.569888234374, 511.49264299540516]  6.899829e-09   \n",
       "\n",
       "              rate_subs_beast     country             province country_lat  \\\n",
       "sample                                                                       \n",
       "SAMEA5661390         6.174240     Germany              Bavaria     51.0834   \n",
       "SAMEA5661389         0.014638     Germany              Bavaria     51.0834   \n",
       "SAMEA5661385         3.028180     Germany              Bavaria     51.0834   \n",
       "SAMEA5661384         0.078613       Spain   Valencia Community     39.3261   \n",
       "SAMEA5661372         0.399351      France  Centre-Loire Valley     46.6034   \n",
       "SAMEA5661367         0.542762      France  Centre-Loire Valley     46.6034   \n",
       "SAMEA5661365         0.143814     England      East of England      52.531   \n",
       "SAMEA5661364         0.836314     England      East of England      52.531   \n",
       "SAMEA5661363         0.087081     England      East of England      52.531   \n",
       "SAMEA5661360         0.797053     Germany              Bavaria     51.0834   \n",
       "SAMEA4354665         0.083994     Germany              Bavaria     51.0834   \n",
       "SAMEA1061800         1.504738  Kyrgyzstan     Issyk-Kul Region     41.5089   \n",
       "NODE0                0.000000          NA                   NA          NA   \n",
       "NODE1                0.230714          NA                   NA          NA   \n",
       "NODE2                0.014183          NA                   NA          NA   \n",
       "NODE3                0.049373          NA                   NA          NA   \n",
       "NODE4                0.285844          NA                   NA          NA   \n",
       "NODE5                0.372439          NA                   NA          NA   \n",
       "NODE6                0.008656          NA                   NA          NA   \n",
       "NODE7                0.097364          NA                   NA          NA   \n",
       "NODE8                0.464414          NA                   NA          NA   \n",
       "NODE9                0.023476          NA                   NA          NA   \n",
       "NODE10               0.029180          NA                   NA          NA   \n",
       "\n",
       "             country_lon province_lat province_lon  strain branch_major  \\\n",
       "sample                                                                    \n",
       "SAMEA5661390     10.4234      48.9468      11.4039  UNT004        0.ANT   \n",
       "SAMEA5661389     10.4234      48.9468      11.4039  UNT003        0.ANT   \n",
       "SAMEA5661385     10.4234      48.9468      11.4039  PET004        0.ANT   \n",
       "SAMEA5661384    -4.83798       39.682    -0.765441  VAL001        0.ANT   \n",
       "SAMEA5661372     1.88833       47.549      1.73241  LSD020        0.ANT   \n",
       "SAMEA5661367     1.88833       47.549      1.73241  LSD001        0.ANT   \n",
       "SAMEA5661365    -1.26491        52.22     0.487578  EDI004        0.ANT   \n",
       "SAMEA5661364    -1.26491        52.22     0.487578  EDI003        0.ANT   \n",
       "SAMEA5661363    -1.26491        52.22     0.487578  EDI001        0.ANT   \n",
       "SAMEA5661360     10.4234      48.9468      11.4039  DIT003        0.ANT   \n",
       "SAMEA4354665     10.4234      48.9468      11.4039  AE1175        0.ANT   \n",
       "SAMEA1061800     74.7241       42.061      78.1707   DA101        0.ANT   \n",
       "NODE0                 NA           NA           NA      NA           NA   \n",
       "NODE1                 NA           NA           NA      NA           NA   \n",
       "NODE2                 NA           NA           NA      NA           NA   \n",
       "NODE3                 NA           NA           NA      NA           NA   \n",
       "NODE4                 NA           NA           NA      NA           NA   \n",
       "NODE5                 NA           NA           NA      NA           NA   \n",
       "NODE6                 NA           NA           NA      NA           NA   \n",
       "NODE7                 NA           NA           NA      NA           NA   \n",
       "NODE8                 NA           NA           NA      NA           NA   \n",
       "NODE9                 NA           NA           NA      NA           NA   \n",
       "NODE10                NA           NA           NA      NA           NA   \n",
       "\n",
       "               biovar host_human       date        date_bp  \n",
       "sample                                                      \n",
       "SAMEA5661390  Antiqua      Human  [419:537]  [-1602:-1484]  \n",
       "SAMEA5661389  Antiqua      Human  [433:599]  [-1588:-1422]  \n",
       "SAMEA5661385  Antiqua      Human  [434:600]  [-1587:-1421]  \n",
       "SAMEA5661384  Antiqua      Human  [432:610]  [-1589:-1411]  \n",
       "SAMEA5661372  Antiqua      Human  [650:880]  [-1371:-1141]  \n",
       "SAMEA5661367  Antiqua      Human  [650:880]  [-1371:-1141]  \n",
       "SAMEA5661365  Antiqua      Human  [500:650]  [-1521:-1371]  \n",
       "SAMEA5661364  Antiqua      Human  [500:650]  [-1521:-1371]  \n",
       "SAMEA5661363  Antiqua      Human  [474:637]  [-1547:-1384]  \n",
       "SAMEA5661360  Antiqua      Human  [428:595]  [-1593:-1426]  \n",
       "SAMEA4354665  Antiqua      Human  [426:571]  [-1595:-1450]  \n",
       "SAMEA1061800  Antiqua      Human  [214:261]  [-1807:-1760]  \n",
       "NODE0              NA         NA         NA             NA  \n",
       "NODE1              NA         NA         NA             NA  \n",
       "NODE2              NA         NA         NA             NA  \n",
       "NODE3              NA         NA         NA             NA  \n",
       "NODE4              NA         NA         NA             NA  \n",
       "NODE5              NA         NA         NA             NA  \n",
       "NODE6              NA         NA         NA             NA  \n",
       "NODE7              NA         NA         NA             NA  \n",
       "NODE8              NA         NA         NA             NA  \n",
       "NODE9              NA         NA         NA             NA  \n",
       "NODE10             NA         NA         NA             NA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_008630455.1_ASM863045v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>29.2825</td>\n",
       "      <td>2011</td>\n",
       "      <td>[2011.0, 2011.0]</td>\n",
       "      <td>3.0165e-08</td>\n",
       "      <td>0.127571</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Republic of Dagestan</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.0883</td>\n",
       "      <td>47.1499</td>\n",
       "      <td>C-826</td>\n",
       "      <td>0.PE</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630415.1_ASM863041v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>6.28249</td>\n",
       "      <td>1988</td>\n",
       "      <td>[1988.0, 1988.0]</td>\n",
       "      <td>5.68933e-07</td>\n",
       "      <td>2.40607</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Republic of Dagestan</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.0883</td>\n",
       "      <td>47.1499</td>\n",
       "      <td>C-672</td>\n",
       "      <td>0.PE</td>\n",
       "      <td>Caucasica</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1988.0</td>\n",
       "      <td>-33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_006546745.1_ASM654674v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>23.105</td>\n",
       "      <td>1990</td>\n",
       "      <td>[1990.0, 1990.0]</td>\n",
       "      <td>1.54868e-07</td>\n",
       "      <td>0.654951</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Bayan-Ölgii</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>48.547</td>\n",
       "      <td>89.8549</td>\n",
       "      <td>I-3269</td>\n",
       "      <td>0.PE</td>\n",
       "      <td>Ulegeica</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1990.0</td>\n",
       "      <td>-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_006546725.1_ASM654672v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>23.8653</td>\n",
       "      <td>1986</td>\n",
       "      <td>[1986.0, 1986.0]</td>\n",
       "      <td>1.12831e-07</td>\n",
       "      <td>0.477172</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Govi-Altai</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>45.3463</td>\n",
       "      <td>95.9326</td>\n",
       "      <td>I-3196</td>\n",
       "      <td>0.PE</td>\n",
       "      <td>Ulegeica</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1986.0</td>\n",
       "      <td>-35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_006546665.1_ASM654666v1_genomic</th>\n",
       "      <td>terminal</td>\n",
       "      <td>20.4987</td>\n",
       "      <td>1987</td>\n",
       "      <td>[1987.0, 1987.0]</td>\n",
       "      <td>1.64816e-07</td>\n",
       "      <td>0.697021</td>\n",
       "      <td>Mongolia</td>\n",
       "      <td>Khovd</td>\n",
       "      <td>46.825</td>\n",
       "      <td>103.85</td>\n",
       "      <td>48.0552</td>\n",
       "      <td>91.2673</td>\n",
       "      <td>I-3193</td>\n",
       "      <td>0.PE</td>\n",
       "      <td>Ulegeica</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>1987.0</td>\n",
       "      <td>-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE79</th>\n",
       "      <td>internal</td>\n",
       "      <td>9.1079</td>\n",
       "      <td>1984.11</td>\n",
       "      <td>[1977.0886858553683, 1992.827346687587]</td>\n",
       "      <td>7.64583e-08</td>\n",
       "      <td>0.32335</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE80</th>\n",
       "      <td>internal</td>\n",
       "      <td>4.37041</td>\n",
       "      <td>1991.53</td>\n",
       "      <td>[1981.778606751268, 1999.4364370064268]</td>\n",
       "      <td>3.8192e-07</td>\n",
       "      <td>1.61518</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE81</th>\n",
       "      <td>internal</td>\n",
       "      <td>3.51422</td>\n",
       "      <td>1986.39</td>\n",
       "      <td>[1978.653294424603, 1994.975657130774]</td>\n",
       "      <td>9.264e-08</td>\n",
       "      <td>0.391784</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE82</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.377691</td>\n",
       "      <td>1991.58</td>\n",
       "      <td>[1983.1396651849832, 1999.9104904895198]</td>\n",
       "      <td>6.44544e-07</td>\n",
       "      <td>2.72584</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE83</th>\n",
       "      <td>internal</td>\n",
       "      <td>10.6037</td>\n",
       "      <td>1994.37</td>\n",
       "      <td>[1982.7778888689616, 2003.8458223130806]</td>\n",
       "      <td>2.19841e-07</td>\n",
       "      <td>0.92973</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    node_type branch_length timetree_num_date  \\\n",
       "sample                                                                          \n",
       "GCA_008630455.1_ASM863045v1_genomic  terminal       29.2825              2011   \n",
       "GCA_008630415.1_ASM863041v1_genomic  terminal       6.28249              1988   \n",
       "GCA_006546745.1_ASM654674v1_genomic  terminal        23.105              1990   \n",
       "GCA_006546725.1_ASM654672v1_genomic  terminal       23.8653              1986   \n",
       "GCA_006546665.1_ASM654666v1_genomic  terminal       20.4987              1987   \n",
       "...                                       ...           ...               ...   \n",
       "NODE79                               internal        9.1079           1984.11   \n",
       "NODE80                               internal       4.37041           1991.53   \n",
       "NODE81                               internal       3.51422           1986.39   \n",
       "NODE82                               internal      0.377691           1991.58   \n",
       "NODE83                               internal       10.6037           1994.37   \n",
       "\n",
       "                                                 timetree_num_date_confidence  \\\n",
       "sample                                                                          \n",
       "GCA_008630455.1_ASM863045v1_genomic                          [2011.0, 2011.0]   \n",
       "GCA_008630415.1_ASM863041v1_genomic                          [1988.0, 1988.0]   \n",
       "GCA_006546745.1_ASM654674v1_genomic                          [1990.0, 1990.0]   \n",
       "GCA_006546725.1_ASM654672v1_genomic                          [1986.0, 1986.0]   \n",
       "GCA_006546665.1_ASM654666v1_genomic                          [1987.0, 1987.0]   \n",
       "...                                                                       ...   \n",
       "NODE79                                [1977.0886858553683, 1992.827346687587]   \n",
       "NODE80                                [1981.778606751268, 1999.4364370064268]   \n",
       "NODE81                                 [1978.653294424603, 1994.975657130774]   \n",
       "NODE82                               [1983.1396651849832, 1999.9104904895198]   \n",
       "NODE83                               [1982.7778888689616, 2003.8458223130806]   \n",
       "\n",
       "                                      rate_beast rate_subs_beast   country  \\\n",
       "sample                                                                       \n",
       "GCA_008630455.1_ASM863045v1_genomic   3.0165e-08        0.127571    Russia   \n",
       "GCA_008630415.1_ASM863041v1_genomic  5.68933e-07         2.40607    Russia   \n",
       "GCA_006546745.1_ASM654674v1_genomic  1.54868e-07        0.654951  Mongolia   \n",
       "GCA_006546725.1_ASM654672v1_genomic  1.12831e-07        0.477172  Mongolia   \n",
       "GCA_006546665.1_ASM654666v1_genomic  1.64816e-07        0.697021  Mongolia   \n",
       "...                                          ...             ...       ...   \n",
       "NODE79                               7.64583e-08         0.32335        NA   \n",
       "NODE80                                3.8192e-07         1.61518        NA   \n",
       "NODE81                                 9.264e-08        0.391784        NA   \n",
       "NODE82                               6.44544e-07         2.72584        NA   \n",
       "NODE83                               2.19841e-07         0.92973        NA   \n",
       "\n",
       "                                                 province country_lat  \\\n",
       "sample                                                                  \n",
       "GCA_008630455.1_ASM863045v1_genomic  Republic of Dagestan     64.6863   \n",
       "GCA_008630415.1_ASM863041v1_genomic  Republic of Dagestan     64.6863   \n",
       "GCA_006546745.1_ASM654674v1_genomic           Bayan-Ölgii      46.825   \n",
       "GCA_006546725.1_ASM654672v1_genomic            Govi-Altai      46.825   \n",
       "GCA_006546665.1_ASM654666v1_genomic                 Khovd      46.825   \n",
       "...                                                   ...         ...   \n",
       "NODE79                                                 NA          NA   \n",
       "NODE80                                                 NA          NA   \n",
       "NODE81                                                 NA          NA   \n",
       "NODE82                                                 NA          NA   \n",
       "NODE83                                                 NA          NA   \n",
       "\n",
       "                                    country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "GCA_008630455.1_ASM863045v1_genomic     97.7453      43.0883      47.1499   \n",
       "GCA_008630415.1_ASM863041v1_genomic     97.7453      43.0883      47.1499   \n",
       "GCA_006546745.1_ASM654674v1_genomic      103.85       48.547      89.8549   \n",
       "GCA_006546725.1_ASM654672v1_genomic      103.85      45.3463      95.9326   \n",
       "GCA_006546665.1_ASM654666v1_genomic      103.85      48.0552      91.2673   \n",
       "...                                         ...          ...          ...   \n",
       "NODE79                                       NA           NA           NA   \n",
       "NODE80                                       NA           NA           NA   \n",
       "NODE81                                       NA           NA           NA   \n",
       "NODE82                                       NA           NA           NA   \n",
       "NODE83                                       NA           NA           NA   \n",
       "\n",
       "                                     strain branch_major     biovar  \\\n",
       "sample                                                                \n",
       "GCA_008630455.1_ASM863045v1_genomic   C-826         0.PE  Caucasica   \n",
       "GCA_008630415.1_ASM863041v1_genomic   C-672         0.PE  Caucasica   \n",
       "GCA_006546745.1_ASM654674v1_genomic  I-3269         0.PE   Ulegeica   \n",
       "GCA_006546725.1_ASM654672v1_genomic  I-3196         0.PE   Ulegeica   \n",
       "GCA_006546665.1_ASM654666v1_genomic  I-3193         0.PE   Ulegeica   \n",
       "...                                     ...          ...        ...   \n",
       "NODE79                                   NA           NA         NA   \n",
       "NODE80                                   NA           NA         NA   \n",
       "NODE81                                   NA           NA         NA   \n",
       "NODE82                                   NA           NA         NA   \n",
       "NODE83                                   NA           NA         NA   \n",
       "\n",
       "                                    host_human    date date_bp  \n",
       "sample                                                          \n",
       "GCA_008630455.1_ASM863045v1_genomic  Non-Human  2011.0     -10  \n",
       "GCA_008630415.1_ASM863041v1_genomic  Non-Human  1988.0     -33  \n",
       "GCA_006546745.1_ASM654674v1_genomic  Non-Human  1990.0     -31  \n",
       "GCA_006546725.1_ASM654672v1_genomic  Non-Human  1986.0     -35  \n",
       "GCA_006546665.1_ASM654666v1_genomic  Non-Human  1987.0     -34  \n",
       "...                                        ...     ...     ...  \n",
       "NODE79                                      NA      NA      NA  \n",
       "NODE80                                      NA      NA      NA  \n",
       "NODE81                                      NA      NA      NA  \n",
       "NODE82                                      NA      NA      NA  \n",
       "NODE83                                      NA      NA      NA  \n",
       "\n",
       "[170 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>rate_beast</th>\n",
       "      <th>rate_subs_beast</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>strain</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>biovar</th>\n",
       "      <th>host_human</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA104233048</th>\n",
       "      <td>terminal</td>\n",
       "      <td>21.790965</td>\n",
       "      <td>-1944.500000</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>2.660027e-08</td>\n",
       "      <td>0.112495</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>Post6</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>[-2007:-1882]</td>\n",
       "      <td>[-4028:-3903]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233049</th>\n",
       "      <td>terminal</td>\n",
       "      <td>50.342757</td>\n",
       "      <td>-2272.000000</td>\n",
       "      <td>[-2272.0, -2272.0]</td>\n",
       "      <td>2.153232e-07</td>\n",
       "      <td>0.910623</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>1343UnTal85</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>[-2396:-2148]</td>\n",
       "      <td>[-4417:-4169]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233047</th>\n",
       "      <td>terminal</td>\n",
       "      <td>236.800707</td>\n",
       "      <td>-2457.000000</td>\n",
       "      <td>[-2457.0, -2457.0]</td>\n",
       "      <td>1.069577e-08</td>\n",
       "      <td>0.045233</td>\n",
       "      <td>Estonia</td>\n",
       "      <td>Pärnu maakond</td>\n",
       "      <td>58.7524</td>\n",
       "      <td>25.3319</td>\n",
       "      <td>58.3195</td>\n",
       "      <td>24.3026</td>\n",
       "      <td>KunilaII</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>[-2574:-2340]</td>\n",
       "      <td>[-4595:-4361]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>terminal</td>\n",
       "      <td>221.689649</td>\n",
       "      <td>-2546.500000</td>\n",
       "      <td>[-2546.5, -2546.5]</td>\n",
       "      <td>1.120606e-08</td>\n",
       "      <td>0.047392</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>55.35</td>\n",
       "      <td>23.75</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "      <td>Gyvakarai1</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>[-2621:-2472]</td>\n",
       "      <td>[-4642:-4493]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>terminal</td>\n",
       "      <td>9.325354</td>\n",
       "      <td>-2776.500000</td>\n",
       "      <td>[-2776.5, -2776.5]</td>\n",
       "      <td>1.131250e-06</td>\n",
       "      <td>4.784166</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "      <td>RISE509</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>[-2876:-2677]</td>\n",
       "      <td>[-4897:-4698]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541826</th>\n",
       "      <td>terminal</td>\n",
       "      <td>280.290965</td>\n",
       "      <td>-1686.000000</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>7.449197e-08</td>\n",
       "      <td>0.315034</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Altai Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>52.6932</td>\n",
       "      <td>82.6931</td>\n",
       "      <td>RISE505</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>[-1746:-1626]</td>\n",
       "      <td>[-3767:-3647]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6812877</th>\n",
       "      <td>terminal</td>\n",
       "      <td>150.119559</td>\n",
       "      <td>-2518.000000</td>\n",
       "      <td>[-2518.0, -2518.0]</td>\n",
       "      <td>8.381519e-09</td>\n",
       "      <td>0.035446</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Irkutsk Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>56.637</td>\n",
       "      <td>104.719</td>\n",
       "      <td>GLZ002</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>[-2569:-2467]</td>\n",
       "      <td>[-4590:-4488]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6812876</th>\n",
       "      <td>terminal</td>\n",
       "      <td>1.619559</td>\n",
       "      <td>-2666.500000</td>\n",
       "      <td>[-2666.5, -2666.5]</td>\n",
       "      <td>5.942839e-07</td>\n",
       "      <td>2.513285</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Irkutsk Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>56.637</td>\n",
       "      <td>104.719</td>\n",
       "      <td>GLZ001</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>Human</td>\n",
       "      <td>[-2838:-2495]</td>\n",
       "      <td>[-4859:-4516]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>internal</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2852.091245</td>\n",
       "      <td>[-3000.1689872318607, -2776.5095208400344]</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>internal</td>\n",
       "      <td>17.635705</td>\n",
       "      <td>-2765.279235</td>\n",
       "      <td>[-2843.5010841256944, -2683.4217622470696]</td>\n",
       "      <td>5.861306e-07</td>\n",
       "      <td>2.478804</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>internal</td>\n",
       "      <td>74.388943</td>\n",
       "      <td>-2712.824062</td>\n",
       "      <td>[-2777.475232637481, -2669.317429730957]</td>\n",
       "      <td>5.194296e-07</td>\n",
       "      <td>2.196719</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE3</th>\n",
       "      <td>internal</td>\n",
       "      <td>24.271203</td>\n",
       "      <td>-2685.107704</td>\n",
       "      <td>[-2715.2869814642354, -2666.8696733698102]</td>\n",
       "      <td>3.304993e-07</td>\n",
       "      <td>1.397714</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>internal</td>\n",
       "      <td>1.409945</td>\n",
       "      <td>-2672.730652</td>\n",
       "      <td>[-2686.647954927963, -2666.5040578937273]</td>\n",
       "      <td>7.242616e-07</td>\n",
       "      <td>3.062973</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>internal</td>\n",
       "      <td>347.186747</td>\n",
       "      <td>-2439.513700</td>\n",
       "      <td>[-2558.907143703158, -2272.0365347912402]</td>\n",
       "      <td>5.980684e-08</td>\n",
       "      <td>0.252929</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE6</th>\n",
       "      <td>internal</td>\n",
       "      <td>356.051792</td>\n",
       "      <td>-2199.032049</td>\n",
       "      <td>[-2378.9729969836762, -1954.8683498967587]</td>\n",
       "      <td>2.402789e-07</td>\n",
       "      <td>1.016163</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               node_type  branch_length  timetree_num_date  \\\n",
       "sample                                                       \n",
       "SAMEA104233048  terminal      21.790965       -1944.500000   \n",
       "SAMEA104233049  terminal      50.342757       -2272.000000   \n",
       "SAMEA104233047  terminal     236.800707       -2457.000000   \n",
       "SAMEA104233046  terminal     221.689649       -2546.500000   \n",
       "SAMEA3541827    terminal       9.325354       -2776.500000   \n",
       "SAMEA3541826    terminal     280.290965       -1686.000000   \n",
       "SAMEA6812877    terminal     150.119559       -2518.000000   \n",
       "SAMEA6812876    terminal       1.619559       -2666.500000   \n",
       "NODE0           internal       0.000000       -2852.091245   \n",
       "NODE1           internal      17.635705       -2765.279235   \n",
       "NODE2           internal      74.388943       -2712.824062   \n",
       "NODE3           internal      24.271203       -2685.107704   \n",
       "NODE4           internal       1.409945       -2672.730652   \n",
       "NODE5           internal     347.186747       -2439.513700   \n",
       "NODE6           internal     356.051792       -2199.032049   \n",
       "\n",
       "                              timetree_num_date_confidence    rate_beast  \\\n",
       "sample                                                                     \n",
       "SAMEA104233048                                      [0, 0]  2.660027e-08   \n",
       "SAMEA104233049                          [-2272.0, -2272.0]  2.153232e-07   \n",
       "SAMEA104233047                          [-2457.0, -2457.0]  1.069577e-08   \n",
       "SAMEA104233046                          [-2546.5, -2546.5]  1.120606e-08   \n",
       "SAMEA3541827                            [-2776.5, -2776.5]  1.131250e-06   \n",
       "SAMEA3541826                                        [0, 0]  7.449197e-08   \n",
       "SAMEA6812877                            [-2518.0, -2518.0]  8.381519e-09   \n",
       "SAMEA6812876                            [-2666.5, -2666.5]  5.942839e-07   \n",
       "NODE0           [-3000.1689872318607, -2776.5095208400344]  0.000000e+00   \n",
       "NODE1           [-2843.5010841256944, -2683.4217622470696]  5.861306e-07   \n",
       "NODE2             [-2777.475232637481, -2669.317429730957]  5.194296e-07   \n",
       "NODE3           [-2715.2869814642354, -2666.8696733698102]  3.304993e-07   \n",
       "NODE4            [-2686.647954927963, -2666.5040578937273]  7.242616e-07   \n",
       "NODE5            [-2558.907143703158, -2272.0365347912402]  5.980684e-08   \n",
       "NODE6           [-2378.9729969836762, -1954.8683498967587]  2.402789e-07   \n",
       "\n",
       "                rate_subs_beast    country          province country_lat  \\\n",
       "sample                                                                     \n",
       "SAMEA104233048         0.112495    Germany           Bavaria     51.0834   \n",
       "SAMEA104233049         0.910623    Germany           Bavaria     51.0834   \n",
       "SAMEA104233047         0.045233    Estonia     Pärnu maakond     58.7524   \n",
       "SAMEA104233046         0.047392  Lithuania  Panevezys County       55.35   \n",
       "SAMEA3541827           4.784166     Russia  Krasnoyarsk Krai     64.6863   \n",
       "SAMEA3541826           0.315034     Russia        Altai Krai     64.6863   \n",
       "SAMEA6812877           0.035446     Russia    Irkutsk Oblast     64.6863   \n",
       "SAMEA6812876           2.513285     Russia    Irkutsk Oblast     64.6863   \n",
       "NODE0                  0.000000         NA                NA          NA   \n",
       "NODE1                  2.478804         NA                NA          NA   \n",
       "NODE2                  2.196719         NA                NA          NA   \n",
       "NODE3                  1.397714         NA                NA          NA   \n",
       "NODE4                  3.062973         NA                NA          NA   \n",
       "NODE5                  0.252929         NA                NA          NA   \n",
       "NODE6                  1.016163         NA                NA          NA   \n",
       "\n",
       "               country_lon province_lat province_lon       strain  \\\n",
       "sample                                                              \n",
       "SAMEA104233048     10.4234      48.9468      11.4039        Post6   \n",
       "SAMEA104233049     10.4234      48.9468      11.4039  1343UnTal85   \n",
       "SAMEA104233047     25.3319      58.3195      24.3026     KunilaII   \n",
       "SAMEA104233046       23.75      55.9156      25.0312   Gyvakarai1   \n",
       "SAMEA3541827       97.7453      63.3234       97.098      RISE509   \n",
       "SAMEA3541826       97.7453      52.6932      82.6931      RISE505   \n",
       "SAMEA6812877       97.7453       56.637      104.719       GLZ002   \n",
       "SAMEA6812876       97.7453       56.637      104.719       GLZ001   \n",
       "NODE0                   NA           NA           NA           NA   \n",
       "NODE1                   NA           NA           NA           NA   \n",
       "NODE2                   NA           NA           NA           NA   \n",
       "NODE3                   NA           NA           NA           NA   \n",
       "NODE4                   NA           NA           NA           NA   \n",
       "NODE5                   NA           NA           NA           NA   \n",
       "NODE6                   NA           NA           NA           NA   \n",
       "\n",
       "               branch_major      biovar host_human           date  \\\n",
       "sample                                                              \n",
       "SAMEA104233048        0.PRE  Bronze Age      Human  [-2007:-1882]   \n",
       "SAMEA104233049        0.PRE  Bronze Age      Human  [-2396:-2148]   \n",
       "SAMEA104233047        0.PRE  Bronze Age      Human  [-2574:-2340]   \n",
       "SAMEA104233046        0.PRE  Bronze Age      Human  [-2621:-2472]   \n",
       "SAMEA3541827          0.PRE  Bronze Age      Human  [-2876:-2677]   \n",
       "SAMEA3541826          0.PRE  Bronze Age      Human  [-1746:-1626]   \n",
       "SAMEA6812877          0.PRE  Bronze Age      Human  [-2569:-2467]   \n",
       "SAMEA6812876          0.PRE  Bronze Age      Human  [-2838:-2495]   \n",
       "NODE0                    NA          NA         NA             NA   \n",
       "NODE1                    NA          NA         NA             NA   \n",
       "NODE2                    NA          NA         NA             NA   \n",
       "NODE3                    NA          NA         NA             NA   \n",
       "NODE4                    NA          NA         NA             NA   \n",
       "NODE5                    NA          NA         NA             NA   \n",
       "NODE6                    NA          NA         NA             NA   \n",
       "\n",
       "                      date_bp  \n",
       "sample                         \n",
       "SAMEA104233048  [-4028:-3903]  \n",
       "SAMEA104233049  [-4417:-4169]  \n",
       "SAMEA104233047  [-4595:-4361]  \n",
       "SAMEA104233046  [-4642:-4493]  \n",
       "SAMEA3541827    [-4897:-4698]  \n",
       "SAMEA3541826    [-3767:-3647]  \n",
       "SAMEA6812877    [-4590:-4488]  \n",
       "SAMEA6812876    [-4859:-4516]  \n",
       "NODE0                      NA  \n",
       "NODE1                      NA  \n",
       "NODE2                      NA  \n",
       "NODE3                      NA  \n",
       "NODE4                      NA  \n",
       "NODE5                      NA  \n",
       "NODE6                      NA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    \n",
    "    columns = [\n",
    "        # Node type is mandatory\n",
    "        \"node_type\",        \n",
    "        # Draw Divergence Tree\n",
    "        \"branch_length\",\n",
    "        # Draw Time Tree\n",
    "        \"timetree_num_date\",\n",
    "        \"timetree_num_date_confidence\",    \n",
    "        \"rate\",\n",
    "        \"rate_subs\",\n",
    "        # Geo\n",
    "        \"country\",\n",
    "        \"province\",\n",
    "        \"country_lat\",\n",
    "        \"country_lon\",\n",
    "        \"province_lat\",\n",
    "        \"province_lon\",        \n",
    "        # Text descriptions\n",
    "        \"strain\",\n",
    "        \"branch_major\",\n",
    "        \"biovar\",\n",
    "        \"host_human\",\n",
    "        \"date\",\n",
    "        \"date_bp\",        \n",
    "    ]\n",
    "    \n",
    "    auspice_df = copy.copy(tree_dict[branch][\"df\"][columns])\n",
    "    auspice_df.rename(columns={\"rate\": \"rate_beast\"}, inplace=True)\n",
    "    auspice_df.rename(columns={\"rate_subs\": \"rate_subs_beast\"}, inplace=True)\n",
    "    tree_dict[branch][\"auspice_df\"] = auspice_df\n",
    "    display(tree_dict[branch][\"auspice_df\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lonely-execution",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Plot Rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "environmental-length",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rates = [float(rate) for rate in list(tree_dict[\"1.ORI\"][\"df\"][\"rate\"]) if rate != NO_DATA_CHAR]\\nprint(sum(rates) / len(rates))\\nsns.histplot(rates)'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"rates = [float(rate) for rate in list(tree_dict[\"1.ORI\"][\"df\"][\"rate\"]) if rate != NO_DATA_CHAR]\n",
    "print(sum(rates) / len(rates))\n",
    "sns.histplot(rates)\"\"\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "integrated-development",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "municipal-dubai",
   "metadata": {},
   "source": [
    "## Time Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "detected-marketing",
   "metadata": {},
   "outputs": [],
   "source": [
    "for branch in tree_dict:\n",
    "    out_timetree = copy.deepcopy(tree_dict[branch][\"tree\"])\n",
    "\n",
    "    metadata_to_comment(out_timetree, tree_dict[branch][\"df\"])    \n",
    "    out_timetree_nex_path = os.path.join(tree_dir, branch + \".timetree.nex\")\n",
    "    Phylo.write(out_timetree, out_timetree_nex_path, \"nexus\")\n",
    "\n",
    "    for c in out_timetree.find_clades():\n",
    "        c.comment = None\n",
    "\n",
    "    out_timetree_nwk_path = os.path.join(tree_dir, branch + \".timetree.nwk\")\n",
    "    Phylo.write(out_timetree, out_timetree_nwk_path, \"newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aware-extraction",
   "metadata": {},
   "source": [
    "## Augur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "hundred-difference",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1867.1884218560415, 'num_date_confidence': [1801.597968774552, 1910.50330129055], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1824.653130788105, 'num_date_confidence': [1648.352772485698, 1931.4884503335472], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1758.9102280437203, 'num_date_confidence': [1644.8345812291716, 1855.0439441412525], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1278.6526335602698, 'num_date_confidence': [1207.8355605170877, 1322.0491191760786], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1713.1820400313547, 'num_date_confidence': [1443.2176609282997, 1873.6537028813916], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1723.2471631226174, 'num_date_confidence': [1474.5912102054986, 1883.2728041297587], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1927.1131414440563, 'num_date_confidence': [1846.7036288560123, 1975.4208770267599], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1885.4774508225157, 'num_date_confidence': [1769.253672069358, 1960.4467172107932], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1629.1570284697618, 'num_date_confidence': [1348.5131881667462, 1851.0084294367762], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 172.5167052685615, 'num_date_confidence': [38.78639801892041, 237.49042879623642], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1766.6113153814408, 'num_date_confidence': [1561.4370816922892, 1905.538698521477], 'rate_beast': 0, 'rate_subs_beast': 0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': -2852.0912448477893, 'num_date_confidence': [-3000.1689872318607, -2776.5095208400344], 'rate_beast': 0.0, 'rate_subs_beast': 0.0, 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date': 'NA', 'date_bp': 'NA'}\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    \n",
    "    augur_dict = augur_export(\n",
    "        tree_path=None, \n",
    "        aln_path=None,  \n",
    "        tree=tree_dict[branch][\"tree\"], \n",
    "        tree_df=tree_dict[branch][\"auspice_df\"], \n",
    "        color_keyword_exclude=[\"geometry\"],\n",
    "        type_convert = {\n",
    "            \"branch_number\" : (lambda x : str(x))\n",
    "        },\n",
    "    )\n",
    "    \n",
    "    tree_dict[branch][\"augur_dict\"] = augur_dict\n",
    "    \n",
    "    first_taxa = list(augur_dict[\"nodes\"].keys())[0]\n",
    "    print(augur_dict[\"nodes\"][first_taxa])\n",
    "\n",
    "    out_path_augur_json = os.path.join(tree_dir, branch + \"_augur.json\" )\n",
    "    utils.write_json(data=tree_dict[branch][\"augur_dict\"], file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "    tree_dict[branch][\"augur_json_path\"] = out_path_augur_json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "peaceful-concern",
   "metadata": {},
   "source": [
    "## Auspice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aquatic-entry",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.ORI.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.ORI.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.IN.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.IN.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/2.MED.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/2.MED.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/2.ANT.json'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/2.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/4.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/4.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/3.ANT.json'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/3.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.ANT4.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.ANT4.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.PE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.PE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'rate_beast', 'rate_subs_beast', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'strain', 'branch_major', 'biovar', 'host_human', 'date', 'date_bp']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'blank', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_length_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_support_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_bp_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_mean', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_sub_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_time', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    \n",
    "    # Store the color\n",
    "    if branch == \"0.ANT4\":\n",
    "        branch_major_color = colors_dict[\"branch_major\"][\"0.ANT\"]\n",
    "    else:\n",
    "        branch_major_color = colors_dict[\"branch_major\"][branch]\n",
    "\n",
    "    auspice_dict = auspice_export(\n",
    "        tree=tree_dict[branch][\"tree\"],\n",
    "        augur_json_paths=tree_dict[branch][\"augur_json_path\"], \n",
    "        auspice_config_path=auspice_config_path, \n",
    "        auspice_colors_path=out_path_colors,\n",
    "        auspice_latlons_path=out_path_latlon, \n",
    "        )\n",
    "\n",
    "\n",
    "    label_col = list(tree_dict[branch][\"auspice_df\"])\n",
    "    print(label_col)\n",
    "\n",
    "    # Recursively add branch attrs\n",
    "    branch_attributes(\n",
    "        tree_dict=auspice_dict[\"tree\"], \n",
    "        sub_dict=auspice_dict[\"tree\"], \n",
    "        df=tree_dict[branch][\"auspice_df\"],\n",
    "        label_col=label_col,\n",
    "        )\n",
    "    \n",
    "    \n",
    "    # Last manual changes\n",
    "    auspice_dict_copy = copy.deepcopy(auspice_dict)\n",
    "    for i in range(0, len(auspice_dict_copy[\"meta\"][\"colorings\"])):\n",
    "        coloring = auspice_dict_copy[\"meta\"][\"colorings\"][i]\n",
    "        for key in coloring:\n",
    "            # Node type as internal or terminal\n",
    "            if coloring[key] == \"node_type\":\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['internal', '#FFFFFF'], ['terminal', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            # Confidence category\n",
    "            if \"conf_category\" in coloring[key]:\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['LOW', '#FFFFFF'], ['HIGH', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "\n",
    "    # Write outputs - For Local Rendering\n",
    "    out_path_auspice_local_json = os.path.join(tree_dir, branch + \".json\" )\n",
    "    utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "    export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "    print(\"Validation successful for local JSON.\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "regional-inquiry",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
