{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "proprietary-spice",
   "metadata": {},
   "source": [
    "# BEAST Analysis Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "gross-cutting",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 0. SETUP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "expanded-saudi",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from Bio import Phylo, AlignIO\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "connected-livestock",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "removable-vulnerability",
   "metadata": {},
   "outputs": [],
   "source": [
    "#project_dir         = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "#project_dir         = os.path.join(project_dir, \"results\")\n",
    "project_dir          = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main\"\n",
    "\n",
    "tree_dir            = project_dir + \"/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock\"\n",
    "metadata_path       = project_dir + \"/iqtree/all/chromosome/full/filter5/filter-taxa/metadata.tsv\"\n",
    "beast_dir           = project_dir + \"/beast/all/chromosome/clade\"\n",
    "\n",
    "auspice_config_path = project_dir + \"/config/auspice_config.json\"\n",
    "out_path_colors     = project_dir + \"/augur/all/chromosome/full/filter5/colors.tsv\"\n",
    "out_path_latlon     = project_dir + \"/augur/all/chromosome/full/filter5/latlon.tsv\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "constant_sites_path = project_dir + \"/snippy_multi/all/chromosome/full/snippy-multi.constant_sites.txt\"\n",
    "aln_path            = project_dir + \"/iqtree/all/chromosome/full/filter5/filter-sites/snippy-multi.snps.aln\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ancient-syria",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "linear-burden",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", None)\n",
    "\n",
    "# ------------------------------------------\n",
    "BRANCH_LIST = {\n",
    "    \"1.ORI\" : [\"1.ORI1\", \"1.ORI2\", \"1.ORI3\"],\n",
    "    \"1.IN\": [\"1.IN1\",\"1.IN2\",\"1.IN3\"],  \n",
    "    \"1.ANT\": [\"1.ANT1\"], \n",
    "    \"1.PRE\" : [\"1.PRE0\",\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],\n",
    "    \"2.MED\": [\"2.MED0\", \"2.MED1\",\"2.MED2\",\"2.MED3\" ],      \n",
    "    \"2.ANT\": [\"2.ANT1\",\"2.ANT2\",\"2.ANT3\" ],    \n",
    "    \"4.ANT\": [\"4.ANT1\" ],       \n",
    "    \"3.ANT\": [\"3.ANT1\", \"3.ANT2\" ],  \n",
    "    \"0.ANT\": [\"0.ANT1\", \"0.ANT2\",\"0.ANT3\",\"0.ANT5\"],         \n",
    "    \"0.ANT4\" : [\"0.ANT4\"], \n",
    "    \"0.PE\": [\"0.PE2\", \"0.PE4m\", \"0.PE4m\", \"0.PE4t\", \"0.PE4a\", \"0.PE5\", \"0.PE7\", \"0.PE8\", \"0.PE10\"],   \n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"],        \n",
    "}\n",
    "\n",
    "NUM_STATES = 10\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "JSON_INDENT=2\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "with open(constant_sites_path) as infile:\n",
    "    data = infile.read().strip().split(\",\")\n",
    "    constant_sites = sum([int(count) for count in data])\n",
    "\n",
    "aln = AlignIO.read(aln_path, \"fasta\")\n",
    "variant_sites = len(aln[0].seq)\n",
    "SEQ_LEN = constant_sites + variant_sites\n",
    "\n",
    "POSTERIOR_THRESH = 95"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mechanical-monte",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 1. IMPORT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "convertible-freeze",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "beneficial-parcel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313243_45</th>\n",
       "      <td>Azov38</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313243_45</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313246_49</th>\n",
       "      <td>Gdansk8</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Poland</td>\n",
       "      <td>Pomeranian Voivodeship</td>\n",
       "      <td>52.215933</td>\n",
       "      <td>19.134422</td>\n",
       "      <td>54.2456</td>\n",
       "      <td>18.1099</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313246_49</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>54.245560</td>\n",
       "      <td>18.109900</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6651390</th>\n",
       "      <td>AGU010</td>\n",
       "      <td>[1435:1477]</td>\n",
       "      <td>[-586:-544]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6651390</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1456.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637004</th>\n",
       "      <td>AGU025</td>\n",
       "      <td>[1441:1612]</td>\n",
       "      <td>[-580:-409]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637004</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1526.5</td>\n",
       "      <td>494.5</td>\n",
       "      <td>85.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637002</th>\n",
       "      <td>AGU007B</td>\n",
       "      <td>[1463:1632]</td>\n",
       "      <td>[-558:-389]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637002</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1547.5</td>\n",
       "      <td>473.5</td>\n",
       "      <td>84.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      strain         date      date_bp  \\\n",
       "sample                                                                   \n",
       "Reference                               CO92         1992          -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic     9_10       1923.0          -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic    42126       2006.0          -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic    42123       2005.0          -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic    42118       2005.0          -16   \n",
       "...                                      ...          ...          ...   \n",
       "SAMEA7313243_45                       Azov38  [1400:1700]  [-621:-321]   \n",
       "SAMEA7313246_49                      Gdansk8  [1400:1700]  [-621:-321]   \n",
       "SAMEA6651390                          AGU010  [1435:1477]  [-586:-544]   \n",
       "SAMEA6637004                          AGU025  [1441:1612]  [-580:-409]   \n",
       "SAMEA6637002                         AGU007B  [1463:1632]  [-558:-389]   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "SAMEA7313243_45                                        Russia   \n",
       "SAMEA7313246_49                                        Poland   \n",
       "SAMEA6651390                                        Lithuania   \n",
       "SAMEA6637004                                        Lithuania   \n",
       "SAMEA6637002                                        Lithuania   \n",
       "\n",
       "                                                   province  country_lat  \\\n",
       "sample                                                                     \n",
       "Reference                                          Colorado    39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic           Rostov Oblast    64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic                Xinjiang    35.000074   \n",
       "...                                                     ...          ...   \n",
       "SAMEA7313243_45                               Rostov Oblast    64.686314   \n",
       "SAMEA7313246_49                      Pomeranian Voivodeship    52.215933   \n",
       "SAMEA6651390                                 Vilnius County    55.350000   \n",
       "SAMEA6637004                                 Vilnius County    55.350000   \n",
       "SAMEA6637002                                 Vilnius County    55.350000   \n",
       "\n",
       "                                     country_lon province_lat province_lon  \\\n",
       "sample                                                                       \n",
       "Reference                            -100.445882      38.7252     -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic    97.745306      47.6222      40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic   104.999927      42.4805      85.4633   \n",
       "...                                          ...          ...          ...   \n",
       "SAMEA7313243_45                        97.745306      47.6222      40.7958   \n",
       "SAMEA7313246_49                        19.134422      54.2456      18.1099   \n",
       "SAMEA6651390                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637004                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637002                           23.750000      54.8227      25.2495   \n",
       "\n",
       "                                              biovar branch_major  \\\n",
       "sample                                                              \n",
       "Reference                                 Orientalis        1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic       Medievalis        2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          Antiqua        0.ANT   \n",
       "...                                              ...          ...   \n",
       "SAMEA7313243_45                      Second Pandemic        1.PRE   \n",
       "SAMEA7313246_49                      Second Pandemic        1.PRE   \n",
       "SAMEA6651390                         Second Pandemic        1.PRE   \n",
       "SAMEA6637004                         Second Pandemic        1.PRE   \n",
       "SAMEA6637002                         Second Pandemic        1.PRE   \n",
       "\n",
       "                                    branch_minor biosample_accession  \\\n",
       "sample                                                                 \n",
       "Reference                                 1.ORI1        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.ANT1        SAMN07722923   \n",
       "...                                          ...                 ...   \n",
       "SAMEA7313243_45                           1.PRE1     SAMEA7313243_45   \n",
       "SAMEA7313246_49                           1.PRE1     SAMEA7313246_49   \n",
       "SAMEA6651390                              1.PRE1        SAMEA6651390   \n",
       "SAMEA6637004                              1.PRE1        SAMEA6637004   \n",
       "SAMEA6637002                              1.PRE1        SAMEA6637002   \n",
       "\n",
       "                                                     biosample_comment  \\\n",
       "sample                                                                   \n",
       "Reference                              KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic              KEEP: Assembly Modern   \n",
       "...                                                                ...   \n",
       "SAMEA7313243_45                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA7313246_49                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA6651390                                         KEEP: SRA Ancient   \n",
       "SAMEA6637004                                         KEEP: SRA Ancient   \n",
       "SAMEA6637002                                         KEEP: SRA Ancient   \n",
       "\n",
       "                                     branch_number      continent  date_mean  \\\n",
       "sample                                                                         \n",
       "Reference                                        1  North America     1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic              2         Europe     1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic              0           Asia     2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic              0           Asia     2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic              0           Asia     2005.0   \n",
       "...                                            ...            ...        ...   \n",
       "SAMEA7313243_45                                  1         Europe     1550.0   \n",
       "SAMEA7313246_49                                  1         Europe     1550.0   \n",
       "SAMEA6651390                                     1         Europe     1456.0   \n",
       "SAMEA6637004                                     1         Europe     1526.5   \n",
       "SAMEA6637002                                     1         Europe     1547.5   \n",
       "\n",
       "                                     date_bp_mean  date_err        lat  \\\n",
       "sample                                                                   \n",
       "Reference                                    29.0       0.0  38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic          98.0       0.0  47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic          15.0       0.0  42.480495   \n",
       "GCA_009669555.1_ASM966955v1_genomic          16.0       0.0  42.480495   \n",
       "GCA_009669565.1_ASM966956v1_genomic          16.0       0.0  42.480495   \n",
       "...                                           ...       ...        ...   \n",
       "SAMEA7313243_45                             471.0     150.0  47.622245   \n",
       "SAMEA7313246_49                             471.0     150.0  54.245560   \n",
       "SAMEA6651390                                565.0      21.0  54.822692   \n",
       "SAMEA6637004                                494.5      85.5  54.822692   \n",
       "SAMEA6637002                                473.5      84.5  54.822692   \n",
       "\n",
       "                                            lon host_human  \n",
       "sample                                                      \n",
       "Reference                           -105.607716      Human  \n",
       "GCA_009909635.1_ASM990963v1_genomic   40.795794      Human  \n",
       "GCA_009669545.1_ASM966954v1_genomic   85.463346  Non-Human  \n",
       "GCA_009669555.1_ASM966955v1_genomic   85.463346  Non-Human  \n",
       "GCA_009669565.1_ASM966956v1_genomic   85.463346  Non-Human  \n",
       "...                                         ...        ...  \n",
       "SAMEA7313243_45                       40.795794      Human  \n",
       "SAMEA7313246_49                       18.109900      Human  \n",
       "SAMEA6651390                          25.249534      Human  \n",
       "SAMEA6637004                          25.249534      Human  \n",
       "SAMEA6637002                          25.249534      Human  \n",
       "\n",
       "[601 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.set_index(metadata_df.columns[0], inplace=True)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "passing-capitol",
   "metadata": {},
   "source": [
    "## Most Recent Sampling Date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "tribal-manchester",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path = os.path.join(beast_dir, \"most_recent_sampling_dates.tsv\")\n",
    "mrsd_dict = {}\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    for branch in BRANCH_LIST:\n",
    "        branch_df = metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])]\n",
    "        max_date = max(branch_df[\"date_mean\"])\n",
    "        outfile.write(\"{}\\t{}\\n\".format(branch, max_date))\n",
    "        mrsd_dict[branch] = max_date"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unlikely-beverage",
   "metadata": {},
   "source": [
    "## Colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cleared-dutch",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'country': {'Russia': '#8000ff', 'Lithuania': '#7215ff', 'Estonia': '#652afe', 'Germany': '#573ffd', 'China': '#4a53fc', 'Azerbaijan': '#3c66fa', 'Armenia': '#2f79f7', 'Georgia': '#228bf4', 'Mongolia': '#149df1', 'Tajikistan': '#07adee', 'Kyrgyzstan': '#07bcea', 'England': '#14c9e5', 'Spain': '#22d5e0', 'France': '#2fe0db', 'Nepal': '#3cead5', 'India': '#4af1d0', 'Kazakhstan': '#57f7c9', 'Turkmenistan': '#65fcc3', 'Uzbekistan': '#72febc', 'Iran': '#80ffb4', 'Italy': '#8dfead', 'Poland': '#9afca5', 'Switzerland': '#a8f79d', 'Norway': '#b5f194', 'The Netherlands': '#c3ea8b', 'Kenya': '#d0e083', 'Uganda': '#ddd579', 'Democratic Republic of the Congo': '#ebc970', 'Vietnam': '#f8bc66', 'Myanmar': '#ffad5d', 'United States of America': '#ff9d53', 'Peru': '#ff8b49', 'Canada': '#ff793f', 'Indonesia': '#ff6634', 'Madagascar': '#ff532a', 'Zimbabwe': '#ff3f20', 'Algeria': '#ff2a15', 'Brazil': '#ff150b', 'Bolivia': '#ff0000', '?': '#969696'}, 'branch_major': {'0.PRE': '#8000ff', '0.PE': '#4c4ffc', '0.ANT': '#1996f3', '3.ANT': '#1acee3', '4.ANT': '#4df3ce', '2.ANT': '#80ffb4', '2.MED': '#b3f396', '1.PRE': '#e6ce74', '1.ANT': '#ff964f', '1.IN': '#ff4f28', '1.ORI': '#ff0000', '?': '#969696'}, 'province': {'Krasnoyarsk Krai': '#8000ff', 'Panevezys County': '#7b07ff', 'Pärnu maakond': '#760eff', 'Irkutsk Oblast': '#7215ff', 'Bavaria': '#6d1dff', 'Altai Krai': '#6924fe', 'Qinghai': '#642bfe', 'Republic of Dagestan': '#6032fe', 'Goranboy District': '#5b39fd', 'Syunik Province': '#5740fd', 'Shirak Province': '#5247fc', 'Gegharkunik Province': '#4d4dfc', 'Samtskhe-Javakheti': '#4954fb', 'Shahbuz Rayon': '#445bfb', 'Samara Oblast': '#4062fa', 'Sečuán': '#3b68f9', 'Bayankhongor': '#376ff9', 'Inner Mongolia': '#3275f8', 'Sughd Province': '#2e7bf7', 'Talas Region': '#2982f6', 'Altai Republic': '#2488f5', 'Govi-Altai': '#208ef4', 'Övörkhangai': '#1b94f3', 'Ömnögovi': '#1799f2', 'Bayan-Ölgii': '#129ff1', 'Khovd': '#0ea5ef', 'Xinjiang': '#09aaee', 'Issyk-Kul Region': '#05afed', 'East of England': '#00b4ec', 'Valencia Community': '#05b9ea', 'Centre-Loire Valley': '#09bee9', 'Osh Region': '#0ec3e7', 'Naryn Region': '#12c7e6', 'Gansu': '#17cce4', 'Hovsgel': '#1bd0e3', 'Zavkhan Province': '#20d4e1', 'Dornogovi': '#24d8df', 'Tuva Republic': '#29dcdd', 'Jilin': '#2edfdc', 'St. Petersberg': '#32e3da', 'Zabaykalsky Krai': '#37e6d8', 'Heilongjiang': '#3be9d6', 'Tibet': '#40ecd4', 'Gandaki Pradesh': '#44eed2', 'Maharashtra': '#49f1d0', 'Himachal Pradesh': '#4df3ce', 'Kabardino-Balkaria': '#52f5cc', 'Karachay-Cherkessia': '#57f7ca', 'Ningxia': '#5bf9c7', 'Shaanxi': '#60fac5', 'Hebei': '#64fbc3', 'West Kazakhstan Region': '#69fcc1', 'Nagorno-Karabakh Republic': '#6dfdbe', 'Kyzylorda Region': '#72febc', 'Rostov Oblast': '#76ffb9', 'Astrakhan Oblast': '#7bffb7', 'Balkan': '#80ffb4', 'Kyzylorda Province': '#84ffb2', 'Shymkent': '#89ffaf', 'Navoiy Region': '#8dfead', 'Almaty Region': '#92fdaa', 'West Kazakhstan region': '#96fca7', 'Atyrau Region': '#9bfba5', 'Fizuli District': '#9ffaa2', 'Baku': '#a4f99f', 'Kurdistan': '#a8f79c', 'Shamkir District': '#adf599', 'Stavropol Krai': '#b2f396', 'Ingushetia': '#b6f194', 'Mangystau Province': '#bbee91', 'Chechnya': '#bfec8e', 'Aktobe Region': '#c4e98b', 'Atyrau': '#c8e688', 'Qazakh District': '#cde385', 'Republic of Kalmykia': '#d1df82', 'Tatarstan': '#d6dc7e', 'Lazio': '#dbd87b', 'Greater London': '#dfd478', \"Provence-Alpes-Côte d'Azur\": '#e4d075', 'Baden-Württemberg': '#e8cc72', 'Vilnius County': '#edc76f', 'Pomeranian Voivodeship': '#f1c36b', 'Brandenburg': '#f6be68', 'Nidwalden': '#fab965', 'Oslo': '#ffb462', 'Catalonia': '#ffaf5e', 'Occitanie': '#ffaa5b', 'North Brabant': '#ffa558', 'Nairobi': '#ff9f54', 'Yunnan': '#ff9951', 'Khánh Hoà Province': '#ff944d', 'Lâm Đồng Province': '#ff8e4a', 'Ho Chi Minh City': '#ff8847', 'Guangxi Zhuang': '#ff8243', 'California': '#ff7b40', 'La Libertad': '#ff753c', 'Tamil Nadu': '#ff6f39', 'Saskatchewan': '#ff6835', 'Arizona': '#ff6232', 'New Mexico': '#ff5b2e', 'Colorado': '#ff542b', 'Texas': '#ff4d27', 'Nevada': '#ff4724', 'Central Java': '#ff4020', 'Province de Fianarantsoa': '#ff391d', 'Kirov Oblast': '#ff3219', 'Gujarat': '#ff2b15', 'Fujian': '#ff2412', 'Midlands': '#ff1d0e', 'MSila': '#ff150b', 'State of Mato Grosso': '#ff0e07', 'La Paz': '#ff0704', 'Cajamarca': '#ff0000', '?': '#969696'}, 'branch_minor': {'0.PRE2': '#8000ff', '0.PRE1': '#7018ff', '0.PE7': '#6130fe', '0.PE2': '#5148fc', '0.PE8': '#425ffa', '0.PE10': '#3275f8', '0.PE4': '#238af5', '0.PE5': '#139ef1', '0.ANT1': '#04b0ed', '0.ANT4': '#0cc1e8', '0.ANT5': '#1bd0e3', '0.ANT2': '#2bdddd', '0.ANT3': '#3ae8d7', '3.ANT1': '#49f1d0', '3.ANT2': '#59f8c8', '4.ANT1': '#68fcc1', '2.ANT3': '#78ffb9', '2.ANT2': '#87ffb0', '2.ANT1': '#97fca7', '2.MED0': '#a6f89e', '2.MED3': '#b6f194', '2.MED2': '#c5e88a', '2.MED1': '#d5dd7f', '1.PRE0': '#e4d075', '1.PRE1': '#f3c16a', '1.PRE2': '#ffb05f', '1.PRE3': '#ff9e53', '1.ANT1': '#ff8a48', '1.IN1': '#ff753c', '1.IN2': '#ff5f30', '1.IN3': '#ff4824', '1.ORI2': '#ff3018', '1.ORI1': '#ff180c', '1.ORI3': '#ff0000', '?': '#969696'}}\n"
     ]
    }
   ],
   "source": [
    "colors_dict = {}\n",
    "\n",
    "colors_df = pd.read_csv(out_path_colors, sep='\\t', header=None)\n",
    "colors_df.columns = [\"state\", \"value\", \"color\"]\n",
    "\n",
    "for state in set(colors_df[\"state\"]):\n",
    "    state_df = colors_df[colors_df[\"state\"] == state]    \n",
    "    colors_dict[state] = {}\n",
    "    \n",
    "    for value,color in zip(state_df[\"value\"], state_df[\"color\"]):\n",
    "        colors_dict[state][value] = color\n",
    "\n",
    "print(colors_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "economic-raising",
   "metadata": {},
   "source": [
    "## Tree Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "small-message",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Branch: 1.ORI\n",
      "Branch: 1.IN\n",
      "Branch: 1.ANT\n",
      "Branch: 1.PRE\n",
      "Branch: 2.MED\n",
      "Branch: 2.ANT\n",
      "Branch: 4.ANT\n",
      "Branch: 3.ANT\n",
      "Branch: 0.ANT\n",
      "Branch: 0.ANT4\n",
      "Branch: 0.PE\n",
      "Branch: 0.PRE\n"
     ]
    }
   ],
   "source": [
    "# Construct a dictionary to hold the trees\n",
    "tree_dict = {}\n",
    "\n",
    "for branch in BRANCH_LIST:\n",
    "    print(\"Branch:\", branch)\n",
    "    tree_dict[branch] = {} \n",
    "    for filename in os.listdir(tree_dir):\n",
    "        if not filename.endswith(\".tre\"): continue\n",
    "        filepath = os.path.join(tree_dir, filename)\n",
    "        if branch in filename:            \n",
    "            # Add tree files to dict\n",
    "            tree_dict[branch][\"tree_file_raw\"] = filepath \n",
    "            tree_dict[branch][\"tree_file_edit\"] = os.path.join(tree_dir, branch + \".nex\")\n",
    "            tree_dict[branch][\"sample_rename\"] = {}\n",
    "            \n",
    "            # Read in raw tree to deal with dashes\n",
    "            with open(tree_dict[branch][\"tree_file_raw\"],  \"r\") as infile:                    \n",
    "                with open(tree_dict[branch][\"tree_file_edit\"], \"w\") as outfile:           \n",
    "                    raw_tree = infile.read()\n",
    "                    # Remove quotations if they exist\n",
    "                    raw_tree = raw_tree.replace(\"'\",\"\")\n",
    "                    \n",
    "                    # Split into lines to iterate over\n",
    "                    raw_tree_lines = raw_tree.split(\"\\n\")\n",
    "                    \n",
    "                    # By default don't parse a line for dashes\n",
    "                    taxa_line = False\n",
    "                    \n",
    "                    for line in raw_tree_lines: \n",
    "                                \n",
    "                        if \"TREE\" not in line and \"-\" in line:\n",
    "                            \n",
    "                            if len(line.split(\" \")) == 1:\n",
    "                                name_dashes = line.strip()\n",
    "                                name_no_dashes = name_dashes.replace(\"-\",\"_\")\n",
    "                                tree_dict[branch][\"sample_rename\"][name_no_dashes] = name_dashes\n",
    "                            line = line.replace(\"-\",\"_\")\n",
    "                                \n",
    "                        outfile.write(line + \"\\n\")\n",
    "            \n",
    "            # Read in edited tree\n",
    "            trees = Phylo.parse(tree_dict[branch][\"tree_file_edit\"], \"nexus\")\n",
    "            # There should be only 1 tree\n",
    "            for t in trees:\n",
    "                tree_dict[branch][\"tree\"] = t\n",
    "                tree_dict[branch][\"tree\"].ladderize(reverse=False)\n",
    "                break\n",
    "\n",
    "            # Rename sample names back to with dashes\n",
    "            for c in tree_dict[branch][\"tree\"].find_clades():\n",
    "                if c.name in tree_dict[branch][\"sample_rename\"]:\n",
    "                    orig_name = c.name\n",
    "                    c.name = tree_dict[branch][\"sample_rename\"][c.name]\n",
    "                    #print(\"Rename:\", orig_name, c.name)\n",
    "                    \n",
    "                # Strip the date suffix\n",
    "                if c.name:\n",
    "                    c.name = \"_\".join(c.name.split(\"_\")[0:-1])\n",
    "                    \n",
    "            # Rename internal nodes\n",
    "            node_i = 0\n",
    "            for c in tree_dict[branch][\"tree\"].find_clades():\n",
    "                if not c.name:\n",
    "                    c.name = \"NODE{}\".format(node_i)\n",
    "                    node_i += 1\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "italic-huntington",
   "metadata": {},
   "source": [
    "## Add Tree Data to Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "identical-circumstances",
   "metadata": {},
   "source": [
    "### Get comments from the first root that is not root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "formal-questionnaire",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n",
      "{'height_95%_HPD': '{105.49669870945002,214.40203122544787}', 'length': '0.0', 'posterior': '1.0', 'height_median': '140.51015101320417', 'height_range': '{96.56887471108806,409.75872259127294}', 'height': '148.81157814395854', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={105.49669870945002,214.40203122544787}', 'length=0.0', 'posterior=1.0', 'height_median=140.51015101320417', 'height_range={96.56887471108806,409.75872259127294}', 'height=148.81157814395854', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "1.IN\n",
      "{'height_95%_HPD': '{76.51154966645281,359.64722751430185}', 'length': '0.0', 'posterior': '1.0', 'height_median': '158.79244103045914', 'height_range': '{64.03240410762909,1346.3719576228896}', 'height': '183.34686921189493', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={76.51154966645281,359.64722751430185}', 'length=0.0', 'posterior=1.0', 'height_median=158.79244103045914', 'height_range={64.03240410762909,1346.3719576228896}', 'height=183.34686921189493', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "1.ANT\n",
      "{'height_95%_HPD': '{148.95605585874753,359.16541877082847}', 'length': '0.0', 'posterior': '1.0', 'height_median': '236.60128389929963', 'height_range': '{114.633666853396,680.730948495061}', 'height': '245.08977195627978', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={148.95605585874753,359.16541877082847}', 'length=0.0', 'posterior=1.0', 'height_median=236.60128389929963', 'height_range={114.633666853396,680.730948495061}', 'height=245.08977195627978', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "1.PRE\n",
      "{'height_95%_HPD': '{445.45088082392135,559.6644394829121}', 'length': '0.0', 'posterior': '1.0', 'height_median': '480.0332740689476', 'height_range': '{438.27940329995164,865.9244965967442}', 'height': '488.84736643973014', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={445.45088082392135,559.6644394829121}', 'length=0.0', 'posterior=1.0', 'height_median=480.0332740689476', 'height_range={438.27940329995164,865.9244965967442}', 'height=488.84736643973014', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "2.MED\n",
      "{'height_95%_HPD': '{144.34629711860848,574.7823390717003}', 'length': '0.0', 'posterior': '1.0', 'height_median': '267.8180939238116', 'height_range': '{113.36353250524456,1711.013674441649}', 'height': '304.81795996864525', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={144.34629711860848,574.7823390717003}', 'length=0.0', 'posterior=1.0', 'height_median=267.8180939238116', 'height_range={113.36353250524456,1711.013674441649}', 'height=304.81795996864525', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "2.ANT\n",
      "{'height_95%_HPD': '{124.72719587024137,533.4087897945013}', 'length': '0.0', 'posterior': '1.0', 'height_median': '252.1900890021193', 'height_range': '{85.92435375902556,1691.0112663075213}', 'height': '284.75283687738266', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={124.72719587024137,533.4087897945013}', 'length=0.0', 'posterior=1.0', 'height_median=252.1900890021193', 'height_range={85.92435375902556,1691.0112663075213}', 'height=284.75283687738266', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "4.ANT\n",
      "{'height_95%_HPD': '{39.57912297324021,168.29637114398767}', 'length': '0.0', 'posterior': '1.0', 'height_median': '76.58218150305166', 'height_range': '{38.10353914048194,498.6843006606764}', 'height': '87.88685855594382', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={39.57912297324021,168.29637114398767}', 'length=0.0', 'posterior=1.0', 'height_median=76.58218150305166', 'height_range={38.10353914048194,498.6843006606764}', 'height=87.88685855594382', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "3.ANT\n",
      "{'height_95%_HPD': '{56.55328278920684,247.74632793064202}', 'length': '0.0', 'posterior': '1.0', 'height_median': '115.37926776007328', 'height_range': '{56.031375094789944,872.407313998455}', 'height': '131.5225491774843', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={56.55328278920684,247.74632793064202}', 'length=0.0', 'posterior=1.0', 'height_median=115.37926776007328', 'height_range={56.031375094789944,872.407313998455}', 'height=131.5225491774843', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "0.ANT\n",
      "{'height_95%_HPD': '{167.99157056322392,670.4868118332539}', 'length': '0.0', 'posterior': '1.0', 'height_median': '360.52232835499626', 'height_range': '{77.82138914894533,1555.0661946681244}', 'height': '389.8429715302382', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={167.99157056322392,670.4868118332539}', 'length=0.0', 'posterior=1.0', 'height_median=360.52232835499626', 'height_range={77.82138914894533,1555.0661946681244}', 'height=389.8429715302382', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "0.ANT4\n",
      "{'height_95%_HPD': '{527.5095712037636,726.2136019810796}', 'length': '0.0', 'posterior': '1.0', 'height_median': '570.4734674344344', 'height_range': '{527.5095712037636,1138.0768560507051}', 'height': '592.4832947314385', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={527.5095712037636,726.2136019810796}', 'length=0.0', 'posterior=1.0', 'height_median=570.4734674344344', 'height_range={527.5095712037636,1138.0768560507051}', 'height=592.4832947314385', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "0.PE\n",
      "Dropping SAMEA104488961 from the dataframe.\n",
      "{'height_95%_HPD': '{108.46130147852293,452.5629183077107}', 'length': '0.0', 'posterior': '1.0', 'height_median': '224.12023865618067', 'height_range': '{62.664086907698625,1482.9287161187194}', 'height': '247.38868461855913', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={108.46130147852293,452.5629183077107}', 'length=0.0', 'posterior=1.0', 'height_median=224.12023865618067', 'height_range={62.664086907698625,1482.9287161187194}', 'height=247.38868461855913', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n",
      "0.PRE\n",
      "{'height_95%_HPD': '{1090.5095208400346,1314.1689872318607}', 'length': '0.0', 'posterior': '1.0', 'height_median': '1143.3570853433978', 'height_range': '{1090.5095208400346,1721.9029061404672}', 'height': '1166.091244847789', 'length_range': '{0,0}', 'rate_95%_HPD': '{0,0}', 'length_95%_HPD': '{0,0}', 'rate_range': '{0,0}', 'rate': '0', 'rate_median': '0', 'length_median': '0'}\n",
      "['height_95%_HPD={1090.5095208400346,1314.1689872318607}', 'length=0.0', 'posterior=1.0', 'height_median=1143.3570853433978', 'height_range={1090.5095208400346,1721.9029061404672}', 'height=1166.091244847789', 'length_range={0,0}', 'rate_95%_HPD={0,0}', 'length_95%_HPD={0,0}', 'rate_range={0,0}', 'rate=0', 'rate_median=0', 'length_median=0']\n"
     ]
    }
   ],
   "source": [
    "TREE_PARAMETERS = None\n",
    "\n",
    "for branch in tree_dict:\n",
    "    print(branch)\n",
    "    \n",
    "    tree = tree_dict[branch][\"tree\"]\n",
    "    df = copy.deepcopy(metadata_df[metadata_df[\"branch_minor\"].isin(BRANCH_LIST[branch])])\n",
    "    \n",
    "    # Check if a sample was missed :(\n",
    "    df_samples = df.index\n",
    "    tree_samples = [c.name for c in tree.find_clades()]\n",
    "    for sample in df_samples:\n",
    "        if sample not in tree_samples:\n",
    "            df.drop(sample, inplace=True)\n",
    "            print(\"Dropping {} from the dataframe.\".format(sample))\n",
    "    \n",
    "    root_comment_dict = parse_comment(tree.root.comment)\n",
    "    \n",
    "    for c in tree.find_clades():\n",
    "        if c.is_terminal():\n",
    "            continue\n",
    "            \n",
    "        if c != tree.root:\n",
    "            comment_dict = parse_comment(c.comment)\n",
    "            for parameter,value in comment_dict.items():\n",
    "                # Intialize parameter values\n",
    "                if \"range\" in parameter or \"95%\" in parameter:\n",
    "                    df[parameter] = [[NO_DATA_CHAR,NO_DATA_CHAR]] * len(df)\n",
    "                    if parameter not in root_comment_dict:\n",
    "                        root_comment_dict[parameter] = '{0,0}'\n",
    "                else:\n",
    "                    df[parameter] = [NO_DATA_CHAR] * len(df)\n",
    "                    if parameter not in root_comment_dict:\n",
    "                        root_comment_dict[parameter] = '0'\n",
    "                    \n",
    "            break\n",
    "\n",
    "    # Update the roots comment to include missing values\n",
    "    new_comment = \"[&\"\n",
    "    new_comment_list = []\n",
    "    for parameter,value in root_comment_dict.items():\n",
    "        param_str = \"{}={}\".format(parameter, value)\n",
    "        new_comment_list.append(param_str)\n",
    "    new_comment += \",\".join(new_comment_list) + \"]\"\n",
    "    tree.root.comment = new_comment\n",
    "    print(root_comment_dict)\n",
    "    print(new_comment_list)\n",
    "    \n",
    "    if not TREE_PARAMETERS:\n",
    "        TREE_PARAMETERS = [p for p in root_comment_dict]\n",
    "    \n",
    "    tree_dict[branch][\"df\"] = df \n",
    "    #display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "labeled-apollo",
   "metadata": {},
   "source": [
    "### Parse Tree Comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "annual-mitchell",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n",
      "GCA_001693595.1_ASM169359v1_genomic height_95%_HPD 2016.0 2016.0\n",
      "GCA_001693595.1_ASM169359v1_genomic height_median 2016.0 2016.0\n",
      "GCA_001693595.1_ASM169359v1_genomic height_range 2016.0 2016.0\n",
      "1.IN\n",
      "GCA_015158475.1_ASM1515847v1_genomic height_95%_HPD 2008.0 2008.0\n",
      "GCA_015158475.1_ASM1515847v1_genomic height_median 2008.0 2008.0\n",
      "GCA_015158475.1_ASM1515847v1_genomic height_range 2008.0 2008.0\n",
      "1.ANT\n",
      "GCA_000169615.1_ASM16961v1_genomic height_95%_HPD 2004.0 2004.0\n",
      "GCA_000169615.1_ASM16961v1_genomic height_median 2004.0 2004.0\n",
      "GCA_000169615.1_ASM16961v1_genomic height_range 2004.0 2004.0\n",
      "1.PRE\n",
      "SAMEA7313236_38 height_95%_HPD [1762:1773] 1767.5\n",
      "SAMEA7313236_38 height_median [1762:1773] 1767.5\n",
      "SAMEA7313236_38 height_range [1762:1773] 1767.5\n",
      "2.MED\n",
      "2.ANT\n",
      "4.ANT\n",
      "3.ANT\n",
      "GCA_013391945.1_ASM1339194v1_genomic height_95%_HPD 2017.0 2017.0\n",
      "GCA_013391945.1_ASM1339194v1_genomic height_median 2017.0 2017.0\n",
      "GCA_013391945.1_ASM1339194v1_genomic height_range 2017.0 2017.0\n",
      "0.ANT\n",
      "NODE75 height_95%_HPD nan 2019.0\n",
      "NODE75 height_median nan 2019.0\n",
      "NODE75 height_range nan 2019.0\n",
      "NODE75 length_range nan 2019.0\n",
      "NODE75 rate_95%_HPD nan 2019.0\n",
      "NODE75 length_95%_HPD nan 2019.0\n",
      "NODE75 rate_range nan 2019.0\n",
      "NODE75 rate_median nan 2019.0\n",
      "NODE75 length_median nan 2019.0\n",
      "0.ANT4\n",
      "SAMEA5661367 height_95%_HPD [650:880] 765.0\n",
      "SAMEA5661367 height_median [650:880] 765.0\n",
      "SAMEA5661367 height_range [650:880] 765.0\n",
      "SAMEA5661372 height_95%_HPD [650:880] 765.0\n",
      "SAMEA5661372 height_median [650:880] 765.0\n",
      "SAMEA5661372 height_range [650:880] 765.0\n",
      "0.PE\n",
      "GCA_003085955.1_ASM308595v1_genomic height_95%_HPD 2014.0 2014.0\n",
      "GCA_003085955.1_ASM308595v1_genomic height_median 2014.0 2014.0\n",
      "GCA_003085955.1_ASM308595v1_genomic height_range 2014.0 2014.0\n",
      "0.PRE\n",
      "SAMEA104233048 height_95%_HPD [-2007:-1882] -1686.0\n",
      "SAMEA104233048 height_median [-2007:-1882] -1686.0\n",
      "SAMEA104233048 height_range [-2007:-1882] -1686.0\n",
      "SAMEA3541826 height_95%_HPD [-1746:-1626] -1686.0\n",
      "SAMEA3541826 height_median [-1746:-1626] -1686.0\n",
      "SAMEA3541826 height_range [-1746:-1626] -1686.0\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    print(branch)\n",
    "    \n",
    "    tree = tree_dict[branch][\"tree\"]\n",
    "    df = tree_dict[branch][\"df\"]\n",
    "    \n",
    "    # Initialize new columns\n",
    "    df[\"timetree_num_date\"] = [NO_DATA_CHAR] * len(df) # Calendar date of height\n",
    "    df[\"timetree_num_date_confidence\"] = [[NO_DATA_CHAR, NO_DATA_CHAR ]] * len(df)\n",
    "    df[\"rate_sub_year\"] = [NO_DATA_CHAR] * len(df)\n",
    "    df[\"branch_length_sub\"] = [NO_DATA_CHAR] * len(df)\n",
    "    df[\"branch_support_conf_category\"] = [NO_DATA_CHAR] * len(df)\n",
    "    df[\"branch_support_conf_char\"] = [NO_DATA_CHAR] * len(df)\n",
    "    \n",
    "    for c in tree.find_clades():\n",
    "        \n",
    "        # --------------------------------------\n",
    "        # Get node type\n",
    "        node_type = \"internal\"\n",
    "        if c.is_terminal():\n",
    "            node_type = \"terminal\"\n",
    "        df.at[c.name, \"node_type\"] = node_type\n",
    "        \n",
    "        # --------------------------------------\n",
    "        # Parse comments\n",
    "        comment_dict = parse_comment(c.comment)\n",
    "        for parameter,val in comment_dict.items():\n",
    "\n",
    "            # Ranges should be parsed into list\n",
    "            if \"range\" in parameter or \"95%\" in parameter:\n",
    "                val = [float(v) for v in val.strip(\"{}\").split(\",\")]\n",
    "            # Multiply the posterior by 100 to be comparable to UFboot\n",
    "            elif parameter == \"posterior\":\n",
    "                val = float(val) * 100\n",
    "            else:\n",
    "                val = float(val)\n",
    "            df.at[c.name, parameter] = val\n",
    "            \n",
    "            \n",
    "        # Tips don't have posteriors, set to 0\n",
    "        if c.is_terminal():\n",
    "            posterior = 0\n",
    "            df.at[c.name, \"posterior\"] = posterior\n",
    "            comment_dict[\"posterior\"] = posterior\n",
    "            \n",
    "        # Set confidence category for posterior/branch support\n",
    "        posterior = df[\"posterior\"][c.name]\n",
    "        branch_support_conf_category = \"LOW\"\n",
    "        branch_support_conf_char = \"\"\n",
    "        if posterior >= POSTERIOR_THRESH:\n",
    "            branch_support_conf_category = \"HIGH\"\n",
    "            branch_support_conf_char = \"*\"\n",
    "        df.at[c.name,\"branch_support_conf_category\"] = branch_support_conf_category\n",
    "        df.at[c.name,\"branch_support_conf_char\"] = branch_support_conf_char\n",
    "\n",
    "        # Check for missing parameters\n",
    "        # For example, the most recent sample will be missing height uncertainty\n",
    "        for parameter in TREE_PARAMETERS:\n",
    "            if parameter not in comment_dict:\n",
    "                if \"range\" in parameter or \"95%\" in parameter:\n",
    "                    val = [0,0]\n",
    "                else:\n",
    "                    val = 0\n",
    "                df.at[c.name, parameter] = val\n",
    "                comment_dict[parameter] = val\n",
    "                    \n",
    "                print(c.name, parameter, df[\"date\"][c.name], mrsd_dict[branch])\n",
    "            \n",
    "        # Convert height to calendar dates\n",
    "        height = df[\"height\"][c.name]\n",
    "        height_95_hpd = df[\"height_95%_HPD\"][c.name]\n",
    "        \n",
    "        # The sample with the most recent date does not have a height 95% HPD\n",
    "        if height_95_hpd[0] == NO_DATA_CHAR:\n",
    "            height_95_hpd = [0,0]\n",
    "            df.at[c.name, \"height_95%_HPD\"] = height_95_hpd\n",
    "            \n",
    "        height_calendar = mrsd_dict[branch] - height\n",
    "        height_95_hpd_calendar = [mrsd_dict[branch] -  height_95_hpd[1], mrsd_dict[branch] -  height_95_hpd[0]]\n",
    "    \n",
    "        df.at[c.name, \"timetree_num_date\"] = height_calendar\n",
    "        df.at[c.name, \"timetree_num_date_confidence\"] = height_95_hpd_calendar\n",
    "        \n",
    "        # Add another measure of substitution rate\n",
    "        rate_sub_year = df[\"rate\"][c.name] * SEQ_LEN\n",
    "        df.at[c.name, \"rate_sub_year\"] = rate_sub_year\n",
    "        \n",
    "        # Measure branch length in substitutions\n",
    "        branch_length_sub = rate_sub_year * df[\"length\"][c.name]\n",
    "        df.at[c.name, \"branch_length_sub\"] = branch_length_sub\n",
    "    \n",
    "    df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "robust-tracker",
   "metadata": {},
   "source": [
    "## Reduced Dataframe for Auspice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "brutal-somalia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['strain', 'date', 'date_bp', 'country', 'province', 'country_lat',\n",
      "       'country_lon', 'province_lat', 'province_lon', 'biovar', 'branch_major',\n",
      "       'branch_minor', 'biosample_accession', 'biosample_comment',\n",
      "       'branch_number', 'continent', 'date_mean', 'date_bp_mean', 'date_err',\n",
      "       'lat', 'lon', 'host_human', 'length_range', 'rate_95%_HPD',\n",
      "       'length_95%_HPD', 'length', 'posterior', 'height_median', 'rate_range',\n",
      "       'height_range', 'height_95%_HPD', 'rate', 'rate_median',\n",
      "       'length_median', 'height', 'timetree_num_date',\n",
      "       'timetree_num_date_confidence', 'rate_sub_year', 'branch_length_sub',\n",
      "       'branch_support_conf_category', 'branch_support_conf_char',\n",
      "       'node_type'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# OPtions\n",
    "print(tree_dict[branch][\"df\"].columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "reasonable-faculty",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n",
      "1.IN\n",
      "1.ANT\n",
      "1.PRE\n",
      "2.MED\n",
      "2.ANT\n",
      "4.ANT\n",
      "3.ANT\n",
      "0.ANT\n",
      "0.ANT4\n",
      "0.PE\n",
      "0.PRE\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    print(branch)\n",
    "    \n",
    "    columns = [\n",
    "        # Node type is mandatorya\n",
    "        \"node_type\",        \n",
    "        # Draw Divergence Tree\n",
    "        \"length\",\n",
    "        # Draw Time Tree\n",
    "        \"timetree_num_date\",\n",
    "        \"timetree_num_date_confidence\",  \n",
    "        # Geo\n",
    "        \"country\",\n",
    "        \"province\",\n",
    "        \"country_lat\",\n",
    "        \"country_lon\",\n",
    "        \"province_lat\",\n",
    "        \"province_lon\",           \n",
    "        # Stats\n",
    "        \"posterior\",\n",
    "        \"branch_support_conf_category\",\n",
    "        \"branch_support_conf_char\",\n",
    "        \"rate\",\n",
    "        \"rate_sub_year\",\n",
    "        \"branch_length_sub\",\n",
    "        # Text descriptions\n",
    "        \"strain\",\n",
    "        \"branch_major\",\n",
    "        \"biovar\",\n",
    "        \"host_human\",\n",
    "        \"date_mean\",\n",
    "        \"date_bp_mean\",        \n",
    "    ]\n",
    "    \n",
    "    auspice_df = copy.copy(tree_dict[branch][\"df\"][columns])\n",
    "    \n",
    "    # Round For Pretty Numbers\n",
    "    posteriors = [float(round(p)) for p in list(auspice_df[\"posterior\"])]\n",
    "    auspice_df[\"posterior\"] = posteriors  \n",
    "    \n",
    "    branch_length_times = [float(round(l)) for l in list(auspice_df[\"length\"])] \n",
    "    auspice_df[\"branch_length_time\"] = branch_length_times\n",
    "    \n",
    "    branch_length_subs = [float(round(l)) for l in list(auspice_df[\"branch_length_sub\"])] \n",
    "    auspice_df[\"branch_length_sub\"] = branch_length_subs \n",
    "\n",
    "    rate_sub_years = [float(round(s)) for s in list(auspice_df[\"rate_sub_year\"])] \n",
    "    auspice_df[\"rate_sub_year\"] = rate_sub_years\n",
    "    \n",
    "    auspice_df.rename(columns={\"length\": \"branch_length\"}, inplace=True)\n",
    "    auspice_df.rename(columns={\"posterior\": \"branch_support\"}, inplace=True)\n",
    "    auspice_df.rename(columns={\"rate\": \"rate_sub\"}, inplace=True)\n",
    "    \n",
    "    # Add blank column\n",
    "    auspice_df[\"blank\"] = [\" \"] * len(auspice_df)\n",
    "\n",
    "    \n",
    "    tree_dict[branch][\"auspice_df\"] = auspice_df\n",
    "    #display(tree_dict[branch][\"auspice_df\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "integrated-development",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "municipal-dubai",
   "metadata": {},
   "source": [
    "## Time Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "detected-marketing",
   "metadata": {},
   "outputs": [],
   "source": [
    "for branch in tree_dict:\n",
    "    out_timetree = copy.deepcopy(tree_dict[branch][\"tree\"])\n",
    "\n",
    "    metadata_to_comment(out_timetree, tree_dict[branch][\"df\"])    \n",
    "    out_timetree_nex_path = os.path.join(tree_dir, branch + \".timetree.nex\")\n",
    "    Phylo.write(out_timetree, out_timetree_nex_path, \"nexus\")\n",
    "\n",
    "    for c in out_timetree.find_clades():\n",
    "        c.comment = None\n",
    "\n",
    "    out_timetree_nwk_path = os.path.join(tree_dir, branch + \".timetree.nwk\")\n",
    "    Phylo.write(out_timetree, out_timetree_nwk_path, \"newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aware-extraction",
   "metadata": {},
   "source": [
    "## Augur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "hundred-difference",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1867.1884218560415, 'num_date_confidence': [1801.597968774552, 1910.50330129055], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1824.653130788105, 'num_date_confidence': [1648.352772485698, 1931.4884503335472], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1758.9102280437203, 'num_date_confidence': [1644.8345812291716, 1855.0439441412525], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1278.6526335602698, 'num_date_confidence': [1207.8355605170877, 1322.0491191760786], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1713.1820400313547, 'num_date_confidence': [1443.2176609282997, 1873.6537028813916], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1723.2471631226174, 'num_date_confidence': [1474.5912102054986, 1883.2728041297587], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1927.1131414440563, 'num_date_confidence': [1846.7036288560123, 1975.4208770267599], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1885.4774508225157, 'num_date_confidence': [1769.253672069358, 1960.4467172107932], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1629.1570284697618, 'num_date_confidence': [1348.5131881667462, 1851.0084294367762], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 172.5167052685615, 'num_date_confidence': [38.78639801892041, 237.49042879623642], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': 1766.6113153814408, 'num_date_confidence': [1561.4370816922892, 1905.538698521477], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n",
      "{'node_type': 'internal', 'branch_length': 0.0, 'num_date': -2852.0912448477893, 'num_date_confidence': [-3000.1689872318607, -2776.5095208400344], 'country': 'NA', 'province': 'NA', 'country_lat': 'NA', 'country_lon': 'NA', 'province_lat': 'NA', 'province_lon': 'NA', 'branch_support': 100.0, 'branch_support_conf_category': 'HIGH', 'branch_support_conf_char': '*', 'rate_sub': 0.0, 'rate_sub_year': 0.0, 'branch_length_sub': 0.0, 'strain': 'NA', 'branch_major': 'NA', 'biovar': 'NA', 'host_human': 'NA', 'date_mean': 'NA', 'date_bp_mean': 'NA', 'branch_length_time': 0.0, 'blank': ' '}\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    \n",
    "    augur_dict = augur_export(\n",
    "        tree_path=None, \n",
    "        aln_path=None,  \n",
    "        tree=tree_dict[branch][\"tree\"], \n",
    "        tree_df=tree_dict[branch][\"auspice_df\"], \n",
    "        color_keyword_exclude=[\"geometry\"],\n",
    "        type_convert = {\n",
    "            \"branch_number\" : (lambda x : str(x))\n",
    "        },\n",
    "    )\n",
    "    \n",
    "    tree_dict[branch][\"augur_dict\"] = augur_dict\n",
    "    \n",
    "    first_taxa = list(augur_dict[\"nodes\"].keys())[0]\n",
    "    print(augur_dict[\"nodes\"][first_taxa])\n",
    "\n",
    "    out_path_augur_json = os.path.join(tree_dir, branch + \"_augur.json\" )\n",
    "    utils.write_json(data=tree_dict[branch][\"augur_dict\"], file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "    tree_dict[branch][\"augur_json_path\"] = out_path_augur_json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "peaceful-concern",
   "metadata": {},
   "source": [
    "## Auspice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aquatic-entry",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.ORI.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.ORI.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.IN.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.IN.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.ANT.json'...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/1.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/2.MED.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/2.MED.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/2.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/2.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/4.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/4.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/3.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/3.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.ANT.json'...\n",
      "Validating that the JSON is internally consistent...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.ANT.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.ANT4.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.ANT4.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.PE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.PE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/config/auspice_config.json'...\n",
      "Validation success.\n",
      "['node_type', 'branch_length', 'timetree_num_date', 'timetree_num_date_confidence', 'country', 'province', 'country_lat', 'country_lon', 'province_lat', 'province_lon', 'branch_support', 'branch_support_conf_category', 'branch_support_conf_char', 'rate_sub', 'rate_sub_year', 'branch_length_sub', 'strain', 'branch_major', 'biovar', 'host_human', 'date_mean', 'date_bp_mean', 'branch_length_time', 'blank']\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.PRE.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/beast/all/chromosome/clade/summary_trees_noHyperPrior/relaxed_clock/0.PRE.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n",
      "WARNING: You asked for a color-by for trait 'biosample_accession', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_major_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_minor_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'branch_number', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'clade_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'continent', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'country_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'date_err', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'node_name', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_date_strain', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'province_joint_conf_category', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'root_rtt_dist', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_sub', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_geo_year', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'biovar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"Non-Human\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  Color option \"host_human\" specifies a hex code for \"NA\" but this isn't ever seen on the tree nodes.\n",
      "\tWARNING:  The coloring \"blank\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The filter \"continent\" does not appear as a property on any tree nodes.\n",
      "\tWARNING:  The filter \"branch_minor\" does not appear as a property on any tree nodes.\n"
     ]
    }
   ],
   "source": [
    "for branch in tree_dict:\n",
    "    \n",
    "    # Store the color\n",
    "    if branch == \"0.ANT4\":\n",
    "        branch_major_color = colors_dict[\"branch_major\"][\"0.ANT\"]\n",
    "    else:\n",
    "        branch_major_color = colors_dict[\"branch_major\"][branch]\n",
    "\n",
    "    auspice_dict = auspice_export(\n",
    "        tree=tree_dict[branch][\"tree\"],\n",
    "        augur_json_paths=tree_dict[branch][\"augur_json_path\"], \n",
    "        auspice_config_path=auspice_config_path, \n",
    "        auspice_colors_path=out_path_colors,\n",
    "        auspice_latlons_path=out_path_latlon, \n",
    "        )\n",
    "\n",
    "\n",
    "    label_col = list(tree_dict[branch][\"auspice_df\"])\n",
    "    print(label_col)\n",
    "\n",
    "    # Recursively add branch attrs\n",
    "    branch_attributes(\n",
    "        tree_dict=auspice_dict[\"tree\"], \n",
    "        sub_dict=auspice_dict[\"tree\"], \n",
    "        df=tree_dict[branch][\"auspice_df\"],\n",
    "        label_col=label_col,\n",
    "        )\n",
    "    \n",
    "    \n",
    "    # Last manual changes\n",
    "    auspice_dict_copy = copy.deepcopy(auspice_dict)\n",
    "    for i in range(0, len(auspice_dict_copy[\"meta\"][\"colorings\"])):\n",
    "        coloring = auspice_dict_copy[\"meta\"][\"colorings\"][i]\n",
    "        for key in coloring:\n",
    "            # Node type as internal or terminal\n",
    "            if coloring[key] == \"node_type\":\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['internal', '#FFFFFF'], ['terminal', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            # Confidence category\n",
    "            if \"conf_category\" in coloring[key]:\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['LOW', '#FFFFFF'], ['HIGH', branch_major_color]]\n",
    "                #print(auspice_dict[\"meta\"][\"colorings\"][i])\n",
    "            # Host Human binary\n",
    "            if \"host_human\" in coloring[key]:\n",
    "                auspice_dict[\"meta\"][\"colorings\"][i]['scale'] = [['Human', '#CBB742'], ['Non-Human', \"#60B6F2\"], ['NA', \"#D6D6D6\"]]\n",
    "\n",
    "    # Write outputs - For Local Rendering\n",
    "    out_path_auspice_local_json = os.path.join(tree_dir, branch + \".json\" )\n",
    "    utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "    export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "    print(\"Validation successful for local JSON.\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "regional-inquiry",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
