{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "1. Estimate Clock Model\n",
    "1. Filter Outliers\n",
    "    - Plot Prune Compare: Divtree\n",
    "    - Plot Prune Compare: Timetree\n",
    "1. Add Clock Model to Dataframe\n",
    "\n",
    "NOTE: Filter outliers after clock model estimated?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bio\n",
    "import treetime\n",
    "from Bio import Phylo,AlignIO, Align,SeqIO\n",
    "\n",
    "# Plotting\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "import seaborn as sns\n",
    "sns.set_style('whitegrid')\n",
    "\n",
    "# Stats\n",
    "import pandas as pd\n",
    "\n",
    "# System IO\n",
    "import dill\n",
    "import copy\n",
    "import os\n",
    "import io\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "\n",
    "# Custom script variables\n",
    "SCRIPT_NAME = \"clock\"\n",
    "PREV_DIR_NAME = \"parse_tree\"\n",
    "PREV_SCRIPT_NAME = \"parse_tree\"\n",
    "\n",
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"5\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "MISSING_DATA = WILDCARDS[2]\n",
    "\n",
    "NAME_COL = \"Name\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tree_dill:\t /2/scratch/keaton/plague-phylogeography/results/parse_tree/all/chromosome_filter5/parse_tree.phylo.obj\n",
      "tree_df_dill:\t /2/scratch/keaton/plague-phylogeography/results/parse_tree/all/chromosome_filter5/parse_tree.df.obj\n",
      "aln path:\t /2/scratch/keaton/plague-phylogeography/results/snippy_multi/all/snippy-core_chromosome.snps.filter5.aln\n",
      "auspice_latlon_path: /2/scratch/keaton/plague-phylogeography/results/parse_tree/all/chromosome_filter5/parse_tree_latlon.tsv\n",
      "auspice_colors_path: /2/scratch/keaton/plague-phylogeography/results/parse_tree/all/chromosome_filter5/parse_tree_colors.tsv\n",
      "auspice_config_path: /2/scratch/keaton/plague-phylogeography/results/config/auspice_config.json\n",
      "auspice_remote_dir_path: /2/scratch/keaton/plague-phylogeography/auspice/\n",
      "outdir: /2/scratch/keaton/plague-phylogeography/results/clock/all/chromosome_filter5\n"
     ]
    }
   ],
   "source": [
    "results_dir = os.path.join(project_dir, \"results\")\n",
    "config_dir = os.path.join(results_dir, \"config\")\n",
    "\n",
    "\n",
    "outdir       = os.path.join(results_dir, SCRIPT_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA),)\n",
    "tree_dill    = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA),PREV_SCRIPT_NAME + \".phylo.obj\")\n",
    "tree_df_dill = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), PREV_SCRIPT_NAME + \".df.obj\")\n",
    "tree_df_path = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), PREV_SCRIPT_NAME + \".tsv\")\n",
    "aln_path     = os.path.join(results_dir,\"snippy_multi\",READS_ORIGIN,\"snippy-core_{}.snps.filter{}.aln\".format(LOCUS_NAME, MISSING_DATA))\n",
    "\n",
    "\n",
    "# Auspice\n",
    "auspice_latlon_path = os.path.join(results_dir, \"parse_tree\", READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), \"parse_tree\" + \"_latlon.tsv\")\n",
    "auspice_colors_path = os.path.join(results_dir, \"parse_tree\", READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), \"parse_tree\" + \"_colors.tsv\")\n",
    "auspice_config_path = os.path.join(config_dir, \"auspice_config.json\")\n",
    "auspice_remote_dir_path = os.path.join(project_dir, \"auspice/\")\n",
    "\n",
    "constant_sites_path = os.path.join(results_dir,\"snippy_multi\",READS_ORIGIN,\"snippy-core_chromosome.full.constant_sites.txt\")\n",
    "\n",
    "\n",
    "print(\"tree_dill:\\t\", tree_dill)\n",
    "print(\"tree_df_dill:\\t\", tree_df_dill)\n",
    "print(\"aln path:\\t\", aln_path)\n",
    "print(\"auspice_latlon_path:\", auspice_latlon_path)\n",
    "print(\"auspice_colors_path:\", auspice_colors_path)\n",
    "print(\"auspice_config_path:\", auspice_config_path)\n",
    "print(\"auspice_remote_dir_path:\", auspice_remote_dir_path)\n",
    "print(\"outdir:\", outdir)\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "while not os.path.exists(outdir):\n",
    "    os.makedirs(outdir)   \n",
    "    \n",
    "SCRIPT_NAME = \"clock_model\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(tree_dill, \"rb\") as infile:\n",
    "    tree = dill.load(infile)\n",
    "tree.ladderize(reverse=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "      <th>Continent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.664821</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.724800e-06</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>2.078125</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.840800e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>RISE509</td>\n",
       "      <td>[-2876:-2677]</td>\n",
       "      <td>[-4897:-4698]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE2</td>\n",
       "      <td>SAMEA3541827</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.342240e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>3.156250</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.233000e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>Gyvakarai1</td>\n",
       "      <td>[-2621:-2472]</td>\n",
       "      <td>[-4642:-4493]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>55.35</td>\n",
       "      <td>23.75</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>SAMEA104233046</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE612</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.345000e-07</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>613.750000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000268845.1_ASM26884v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.340000e-08</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>613.000000</td>\n",
       "      <td>PY-56</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141479</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "      <td>South America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE613</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.345000e-07</td>\n",
       "      <td>black</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>614.500000</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000268905.1_ASM26890v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.034000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>614.000000</td>\n",
       "      <td>PY-60</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141482</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "      <td>South America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269445.1_ASM26944v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.034000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>615.000000</td>\n",
       "      <td>PY-95</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-6.25</td>\n",
       "      <td>-78.8333</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141498</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "      <td>South America</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1229 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot sCF  Branch_Length  \\\n",
       "Name                                                           \n",
       "NODE0                                  NA  NA   0.000000e+00   \n",
       "NODE1                                  NA  NA   2.724800e-06   \n",
       "SAMEA3541827                           NA  NA   3.840800e-06   \n",
       "NODE2                                  NA  NA   1.342240e-05   \n",
       "SAMEA104233046                         NA  NA   2.233000e-06   \n",
       "...                                   ...  ..            ...   \n",
       "NODE612                                NA  NA   2.345000e-07   \n",
       "GCA_000268845.1_ASM26884v1_genomic     NA  NA   2.340000e-08   \n",
       "NODE613                                NA  NA   2.345000e-07   \n",
       "GCA_000268905.1_ASM26890v1_genomic     NA  NA   7.034000e-07   \n",
       "GCA_000269445.1_ASM26944v1_genomic     NA  NA   7.034000e-07   \n",
       "\n",
       "                                   Branch_Support_Color   coord_x     coord_y  \\\n",
       "Name                                                                            \n",
       "NODE0                                             black  0.000000   17.664821   \n",
       "NODE1                                             black  0.000003    2.078125   \n",
       "SAMEA3541827                                       grey  0.000007    1.000000   \n",
       "NODE2                                             black  0.000016    3.156250   \n",
       "SAMEA104233046                                     grey  0.000018    2.000000   \n",
       "...                                                 ...       ...         ...   \n",
       "NODE612                                           black  0.000074  613.750000   \n",
       "GCA_000268845.1_ASM26884v1_genomic                 grey  0.000074  613.000000   \n",
       "NODE613                                           black  0.000074  614.500000   \n",
       "GCA_000268905.1_ASM26890v1_genomic                 grey  0.000075  614.000000   \n",
       "GCA_000269445.1_ASM26944v1_genomic                 grey  0.000075  615.000000   \n",
       "\n",
       "                                        Strain           Date         DateBP  \\\n",
       "Name                                                                           \n",
       "NODE0                                       NA             NA             NA   \n",
       "NODE1                                       NA             NA             NA   \n",
       "SAMEA3541827                           RISE509  [-2876:-2677]  [-4897:-4698]   \n",
       "NODE2                                       NA             NA             NA   \n",
       "SAMEA104233046                      Gyvakarai1  [-2621:-2472]  [-4642:-4493]   \n",
       "...                                        ...            ...            ...   \n",
       "NODE612                                     NA             NA             NA   \n",
       "GCA_000268845.1_ASM26884v1_genomic       PY-56           2010            -11   \n",
       "NODE613                                     NA             NA             NA   \n",
       "GCA_000268905.1_ASM26890v1_genomic       PY-60           2010            -11   \n",
       "GCA_000269445.1_ASM26944v1_genomic       PY-95           2010            -11   \n",
       "\n",
       "                                      Country          Province CountryLat  \\\n",
       "Name                                                                         \n",
       "NODE0                                      NA                NA         NA   \n",
       "NODE1                                      NA                NA         NA   \n",
       "SAMEA3541827                           Russia  Krasnoyarsk Krai    64.6863   \n",
       "NODE2                                      NA                NA         NA   \n",
       "SAMEA104233046                      Lithuania  Panevezys County      55.35   \n",
       "...                                       ...               ...        ...   \n",
       "NODE612                                    NA                NA         NA   \n",
       "GCA_000268845.1_ASM26884v1_genomic       Peru       La Libertad   -6.86997   \n",
       "NODE613                                    NA                NA         NA   \n",
       "GCA_000268905.1_ASM26890v1_genomic       Peru       La Libertad   -6.86997   \n",
       "GCA_000269445.1_ASM26944v1_genomic       Peru         Cajamarca   -6.86997   \n",
       "\n",
       "                                   CountryLon ProvinceLat ProvinceLon  \\\n",
       "Name                                                                    \n",
       "NODE0                                      NA          NA          NA   \n",
       "NODE1                                      NA          NA          NA   \n",
       "SAMEA3541827                          97.7453     63.3234      97.098   \n",
       "NODE2                                      NA          NA          NA   \n",
       "SAMEA104233046                          23.75     55.9156     25.0312   \n",
       "...                                       ...         ...         ...   \n",
       "NODE612                                    NA          NA          NA   \n",
       "GCA_000268845.1_ASM26884v1_genomic   -75.0459          -8       -78.5   \n",
       "NODE613                                    NA          NA          NA   \n",
       "GCA_000268905.1_ASM26890v1_genomic   -75.0459          -8       -78.5   \n",
       "GCA_000269445.1_ASM26944v1_genomic   -75.0459       -6.25    -78.8333   \n",
       "\n",
       "                                        Biovar Branch_Major Branch_Minor  \\\n",
       "Name                                                                       \n",
       "NODE0                                       NA           NA           NA   \n",
       "NODE1                                       NA           NA           NA   \n",
       "SAMEA3541827                        Bronze Age        0.PRE       0.PRE2   \n",
       "NODE2                                       NA           NA           NA   \n",
       "SAMEA104233046                      Bronze Age        0.PRE       0.PRE1   \n",
       "...                                        ...          ...          ...   \n",
       "NODE612                                     NA           NA           NA   \n",
       "GCA_000268845.1_ASM26884v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "NODE613                                     NA           NA           NA   \n",
       "GCA_000268905.1_ASM26890v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "GCA_000269445.1_ASM26944v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "\n",
       "                                         BioSample       BioSampleComment  \\\n",
       "Name                                                                        \n",
       "NODE0                                           NA                     NA   \n",
       "NODE1                                           NA                     NA   \n",
       "SAMEA3541827                          SAMEA3541827      KEEP: SRA Ancient   \n",
       "NODE2                                           NA                     NA   \n",
       "SAMEA104233046                      SAMEA104233046      KEEP: SRA Ancient   \n",
       "...                                            ...                    ...   \n",
       "NODE612                                         NA                     NA   \n",
       "GCA_000268845.1_ASM26884v1_genomic    SAMN02141479  KEEP: Assembly Modern   \n",
       "NODE613                                         NA                     NA   \n",
       "GCA_000268905.1_ASM26890v1_genomic    SAMN02141482  KEEP: Assembly Modern   \n",
       "GCA_000269445.1_ASM26944v1_genomic    SAMN02141498  KEEP: Assembly Modern   \n",
       "\n",
       "                                   Branch_Number      Continent  \n",
       "Name                                                             \n",
       "NODE0                                         NA             NA  \n",
       "NODE1                                         NA             NA  \n",
       "SAMEA3541827                                   0         Europe  \n",
       "NODE2                                         NA             NA  \n",
       "SAMEA104233046                                 0         Europe  \n",
       "...                                          ...            ...  \n",
       "NODE612                                       NA             NA  \n",
       "GCA_000268845.1_ASM26884v1_genomic             1  South America  \n",
       "NODE613                                       NA             NA  \n",
       "GCA_000268905.1_ASM26890v1_genomic             1  South America  \n",
       "GCA_000269445.1_ASM26944v1_genomic             1  South America  \n",
       "\n",
       "[1229 rows x 22 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(tree_df_dill, \"rb\") as infile:\n",
    "    tree_df = dill.load(infile)\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Constant Sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(constant_sites_path, \"r\") as infile:\n",
    "    constant_sites_str = infile.read().strip().split(\",\")\n",
    "    constant_sites_list = [int(c) for c in constant_sites_str]\n",
    "    constant_sites = sum(constant_sites_list)\n",
    "    constant_sites_freq = [c/constant_sites for c in constant_sites_list]\n",
    "    fixed_pi = np.array(constant_sites_freq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Filter Outlier Tips"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse Tip Dates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Attempting to parse dates...\n",
      "\tUsing column 'Name' as name. This needs match the taxon names in the tree!!\n",
      "\tUsing column 'Date' as date.\n"
     ]
    }
   ],
   "source": [
    "# Use the utils function to parse the metadata dates\n",
    "dates_raw = treetime.utils.parse_dates(tree_df_path, \n",
    "                                   date_col=DATE_COL, \n",
    "                                   name_col = NAME_COL)\n",
    "\n",
    "# Remove nan elements (internal nodes)\n",
    "dates = {}\n",
    "for k,v in dates_raw.items():\n",
    "    if type(v) == list:\n",
    "        dates[k] = v\n",
    "    elif not pd.isnull(v):\n",
    "        dates[k] = v\n",
    "        \n",
    "# Add Reference\n",
    "dates[\"Reference\"] = REF_DATE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Estimate Clock Model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize treetime object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "0.00\t-TreeAnc: set-up\n",
      "\n",
      "1.39\t-SequenceData: loaded alignment.\n",
      "1.40\t--SequenceData.aln: specified sequence length doesn't match alignment\n",
      "    \t  length. Treating difference as constant sites.\n",
      "\n",
      "1.40\t-SeqData: making compressed alignment...\n",
      "\n",
      "5.04\t-SequenceData: constructed compressed alignment...\n",
      "\n",
      "5.47\t-ClockTree._assign_dates: assigned date contraints to 615 out of 615 tips.\n"
     ]
    }
   ],
   "source": [
    "tt = treetime.TreeTime(\n",
    "    dates=dates,\n",
    "    aln=aln_path,\n",
    "    tree=tree,\n",
    "    verbose=2,\n",
    "    fill_overhangs=False,\n",
    "    seq_len=REF_LEN,\n",
    "    alphabet=\"nuc_nogap\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estimate the clock model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "5.48\t-TreeAnc.optimize_tree: sequences...\n",
      "\n",
      "5.48\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, joint\n",
      "\n",
      "5.48\tWARNING: Previous versions of TreeTime (<0.7.0) RECONSTRUCTED sequences of\n",
      "    \ttips at positions with AMBIGUOUS bases. This resulted in unexpected\n",
      "    \tbehavior is some cases and is no longer done by default. If you want to\n",
      "    \treplace those ambiguous sites with their most likely state, rerun with\n",
      "    \t`reconstruct_tip_states=True` or `--reconstruct-tip-states`.\n",
      "\n",
      "14.79\t-GTR: with alphabet: ['A', 'C', 'G', 'T']\n",
      "\n",
      "14.79\t-GTR: model inference\n",
      "\n",
      "22.46\t-TreeAnc.optimize_branch_length: running branch length optimization using\n",
      "     \t jointML ancestral sequences\n",
      "\n",
      "27.85\t-TreeAnc.prune_short_branches: pruning short branches (max prob at\n",
      "     \t zero)...\n",
      "\n",
      "28.05\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, joint\n",
      "\n",
      "35.35\t-TreeAnc.optimize_branch_length: running branch length optimization using\n",
      "     \t jointML ancestral sequences\n",
      "\n",
      "40.31\tTreeTime.reroot: with method or node: NODE0\n",
      "\n",
      "41.51\tTreeTime.reroot: with method or node: NODE0\n",
      "\n",
      "42.53\t-TreeAnc.optimize_tree: sequences...\n",
      "\n",
      "42.53\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, joint\n",
      "\n",
      "49.76\t-TreeAnc.optimize_branch_length: running branch length optimization using\n",
      "     \t jointML ancestral sequences\n",
      "\n",
      "54.69\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, joint\n",
      "\n",
      "62.00\t-TreeAnc.optimize_branch_length: running branch length optimization using\n",
      "     \t jointML ancestral sequences\n",
      "\n",
      "66.96\t###TreeTime.run: INITIAL ROUND\n",
      "\n",
      "66.96\t-ClockTree: Maximum likelihood tree optimization with temporal constraints\n",
      "\n",
      "168.98\tTreeTime.reroot: with method or node: NODE0\n",
      "\n",
      "169.99\t###TreeTime.run: rerunning timetree after rerooting\n",
      "\n",
      "170.00\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "      \t constraints\n",
      "\n",
      "170.00\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, joint\n",
      "\n",
      "280.86\t###TreeTime.run: ITERATION 1 out of 10 iterations\n",
      "\n",
      "280.86\t-TreeTime.run: adding coalescent prior with Tc=const\n",
      "relaxed_clock {'slack': 0.1, 'coupling': 0}\n",
      "\n",
      "290.72\t-TreeTime.resolve_polytomies: resolving multiple mergers...\n",
      "\n",
      "294.16\t-TreeAnc.optimize_tree: sequences...\n",
      "\n",
      "294.16\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, joint\n",
      "\n",
      "301.72\t-TreeAnc.optimize_branch_length: running branch length optimization using\n",
      "      \t jointML ancestral sequences\n",
      "\n",
      "306.96\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "      \t constraints\n",
      "\n",
      "416.85\t-TreeAnc.infer_ancestral_sequences with method: ml, joint\n",
      "\n",
      "424.65\t##TreeTime.run: ndiff=4, n_resolved=83, Tc=skyline\n",
      "\n",
      "424.65\t###TreeTime.run: ITERATION 2 out of 10 iterations\n",
      "\n",
      "424.65\t-TreeTime.run: adding coalescent prior with Tc=const\n",
      "relaxed_clock {'slack': 0.1, 'coupling': 0}\n",
      "\n",
      "435.37\t-TreeTime.resolve_polytomies: resolving multiple mergers...\n",
      "\n",
      "436.09\t-TreeAnc.optimize_tree: sequences...\n",
      "\n",
      "436.09\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, joint\n",
      "\n",
      "443.63\t-TreeAnc.optimize_branch_length: running branch length optimization using\n",
      "      \t jointML ancestral sequences\n",
      "\n",
      "448.96\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "      \t constraints\n",
      "\n",
      "565.15\t-TreeAnc.infer_ancestral_sequences with method: ml, joint\n",
      "\n",
      "572.93\t##TreeTime.run: ndiff=2, n_resolved=12, Tc=skyline\n",
      "\n",
      "572.94\t###TreeTime.run: ITERATION 3 out of 10 iterations\n",
      "\n",
      "572.94\t-TreeTime.run: adding coalescent prior with Tc=const\n",
      "relaxed_clock {'slack': 0.1, 'coupling': 0}\n",
      "\n",
      "583.81\t-TreeTime.resolve_polytomies: resolving multiple mergers...\n",
      "\n",
      "584.62\t-TreeAnc.optimize_tree: sequences...\n",
      "\n",
      "584.62\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, joint\n",
      "\n",
      "592.26\t-TreeAnc.optimize_branch_length: running branch length optimization using\n",
      "      \t jointML ancestral sequences\n",
      "\n",
      "597.61\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "      \t constraints\n",
      "\n",
      "716.07\t-TreeAnc.infer_ancestral_sequences with method: ml, joint\n",
      "\n",
      "724.17\t##TreeTime.run: ndiff=18, n_resolved=15, Tc=skyline\n",
      "\n",
      "724.17\t###TreeTime.run: ITERATION 4 out of 10 iterations\n",
      "\n",
      "724.17\t-TreeTime.run: adding coalescent prior with Tc=const\n",
      "relaxed_clock {'slack': 0.1, 'coupling': 0}\n",
      "\n",
      "735.17\t-TreeTime.resolve_polytomies: resolving multiple mergers...\n",
      "\n",
      "735.50\t-TreeAnc.optimize_tree: sequences...\n",
      "\n",
      "735.50\t-TreeAnc.infer_ancestral_sequences with method: probabilistic, joint\n",
      "\n",
      "743.10\t-TreeAnc.optimize_branch_length: running branch length optimization using\n",
      "      \t jointML ancestral sequences\n",
      "\n",
      "748.48\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "      \t constraints\n",
      "\n",
      "867.74\t-TreeAnc.infer_ancestral_sequences with method: ml, joint\n",
      "\n",
      "875.54\t##TreeTime.run: ndiff=41, n_resolved=1, Tc=skyline\n",
      "\n",
      "875.54\t###TreeTime.run: ITERATION 5 out of 10 iterations\n",
      "\n",
      "875.54\t-TreeTime.run: adding coalescent prior with Tc=const\n",
      "relaxed_clock {'slack': 0.1, 'coupling': 0}\n",
      "\n",
      "886.55\t-TreeTime.resolve_polytomies: resolving multiple mergers...\n",
      "\n",
      "886.65\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "      \t constraints\n",
      "\n",
      "1007.49\t-TreeAnc.infer_ancestral_sequences with method: ml, joint\n",
      "\n",
      "1015.39\t##TreeTime.run: ndiff=9, n_resolved=0, Tc=skyline\n",
      "\n",
      "1015.39\t###TreeTime.run: ITERATION 6 out of 10 iterations\n",
      "\n",
      "1015.39\t-TreeTime.run: adding coalescent prior with Tc=const\n",
      "relaxed_clock {'slack': 0.1, 'coupling': 0}\n",
      "\n",
      "1026.45\t-TreeTime.resolve_polytomies: resolving multiple mergers...\n",
      "\n",
      "1026.55\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "       \t constraints\n",
      "\n",
      "1147.99\t-TreeAnc.infer_ancestral_sequences with method: ml, joint\n",
      "\n",
      "1155.84\t##TreeTime.run: ndiff=0, n_resolved=0, Tc=skyline\n",
      "\n",
      "1155.84\t###TreeTime.run: NEAR CONVERGED with skyline\n",
      "\n",
      "1155.85\t-###ClockTree.calc_rate_susceptibility: run with upper bound of rate\n",
      "       \t estimate\n",
      "\n",
      "1155.85\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "       \t constraints\n",
      "\n",
      "1276.38\t-###ClockTree.calc_rate_susceptibility: run with lower bound of rate\n",
      "       \t estimate\n",
      "\n",
      "1276.38\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "       \t constraints\n",
      "\n",
      "1395.84\t-###ClockTree.calc_rate_susceptibility: run with central rate estimate\n",
      "\n",
      "1395.84\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "       \t constraints\n",
      "\n",
      "1515.38\t###TreeTime.run: FINAL ROUND - confidence estimation via marginal\n",
      "       \treconstruction\n",
      "\n",
      "1515.38\t-ClockTree: Maximum likelihood tree optimization with temporal\n",
      "       \t constraints\n",
      "ERROR in distribution multiplication: Distributions do not overlap\n",
      "ERROR in distribution multiplication: Distributions do not overlap\n",
      "ERROR in distribution multiplication: Distributions do not overlap\n",
      "ERROR in distribution multiplication: Distributions do not overlap\n",
      "THIS SHOULDN'T BE CALLED ON A DELTA FUNCTION\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'function' object has no attribute 'x'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-edc2efc1eeae>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0mbranch_length_mode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"joint\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;31m# joint or marginal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0mtime_marginal\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m   \u001b[0;31m# True, False or \"assign\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m     \u001b[0mTc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"skyline\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m )\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/treetime/treetime.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, root, infer_gtr, relaxed_clock, n_iqd, resolve_polytomies, max_iter, Tc, fixed_clock_rate, time_marginal, sequence_marginal, branch_length_mode, vary_rate, use_covariation, **kwargs)\u001b[0m\n\u001b[1;32m    259\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlogger\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"###TreeTime.run: FINAL ROUND - confidence estimation via marginal reconstruction\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    260\u001b[0m             \u001b[0mtt_kwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'time_marginal'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtime_marginal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 261\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmake_time_tree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mtt_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    262\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    263\u001b[0m         \u001b[0;31m# explicitly print out which branches are bad and whose dates don't correspond to the input dates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/treetime/clock_tree.py\u001b[0m in \u001b[0;36mmake_time_tree\u001b[0;34m(self, time_marginal, clock_rate, **kwargs)\u001b[0m\n\u001b[1;32m    341\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    342\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtime_marginal\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 343\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ml_t_marginal\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0massign_dates\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime_marginal\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"assign\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    344\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    345\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ml_t_joint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/treetime/clock_tree.py\u001b[0m in \u001b[0;36m_ml_t_marginal\u001b[0;34m(self, assign_dates)\u001b[0m\n\u001b[1;32m    592\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcomplementary_msgs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    593\u001b[0m                     \u001b[0mmsg_parent_to_node\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mNodeInterpolator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmultiply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcomplementary_msgs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 594\u001b[0;31m                     \u001b[0mmsg_parent_to_node\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_adjust_grid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrel_tol\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrel_tol_prune\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    595\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    596\u001b[0m                     \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnumdate\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnumeric_date\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/treetime/distribution.py\u001b[0m in \u001b[0;36m_adjust_grid\u001b[0;34m(self, rel_tol, yc)\u001b[0m\n\u001b[1;32m    272\u001b[0m                 \u001b[0mn_iter\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    273\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 274\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_peak_idx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_func\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    275\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_peak_pos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_func\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_peak_idx\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    276\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_peak_val\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpeak_pos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'function' object has no attribute 'x'"
     ]
    }
   ],
   "source": [
    "#tt.run(max_iter=1)\n",
    "tt.run(\n",
    "    max_iter=10,\n",
    "    n_iqd=3,\n",
    "    verbose=4,\n",
    "    infer_gtr=True,\n",
    "    use_covariation=False,\n",
    "    root=tt.tree.root,\n",
    "    resolve_polytomies=True,\n",
    "    vary_rate=True,\n",
    "    relaxed_clock={\"slack\" : 0.1, \"coupling\": 0},\n",
    "    branch_length_mode = \"joint\", # joint or marginal\n",
    "    time_marginal=True,   # True, False or \"assign\"\n",
    "    Tc=\"skyline\",\n",
    ")\n",
    "\n",
    "# Order nodes\n",
    "tt.tree.ladderize(reverse=False)\n",
    "\n",
    "# Add extra length to root\n",
    "tt_orig = copy.deepcopy(tt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tt = copy.deepcopy(tt_orig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quick Checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common Ancestor\n",
    "tt.tree.common_ancestor(\"NODE0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RTT Regression\n",
    "tt.plot_root_to_tip(add_internal=True, label=True)\n",
    "\n",
    "# Save\n",
    "out_path = os.path.join(outdir, SCRIPT_NAME + \"_rtt-pre.\" + FMT) \n",
    "plt.savefig(out_path, \n",
    "            dpi=dpi, \n",
    "            bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tt.print_lh()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(tt.gtr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Filter Outliers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prune 'Bad' Branches from Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tt_pre = copy.deepcopy(tt)\n",
    "tt_prune = copy.deepcopy(tt)\n",
    "\n",
    "# Color branches and set to divtree\n",
    "for n in tt_pre.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "    if n.bad_branch:\n",
    "        n.color = \"red\"\n",
    "for n in tt_prune.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "\n",
    "bad_nodes = [c for c in tt_prune.tree.find_clades() if c.bad_branch]\n",
    "\n",
    "while len(bad_nodes) > 0:\n",
    "    for node in bad_nodes:\n",
    "        if node.is_terminal():\n",
    "            print(\"PRUNING:\", node.name)\n",
    "            tt_prune.tree.prune(node)\n",
    "    bad_nodes = [c for c in tt_prune.tree.find_clades() if c.bad_branch]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot Prune Comparison: Divtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")\n",
    "fig, (ax1,ax2) = plt.subplots(2, sharex=True, dpi=dpi)\n",
    "\n",
    "Phylo.draw(tt_pre.tree,   show_confidence=False, label_func = lambda x: '', do_show=False, axes=ax1,)  \n",
    "Phylo.draw(tt_prune.tree, show_confidence=False, label_func = lambda x: '', do_show=False, axes=ax2,)\n",
    "\n",
    "ax1.set_title(\"Divergence Tree Pre-Pruning\")\n",
    "ax1.set_xlabel(\"\")\n",
    "ax1.set_ylabel(\"\")\n",
    "ax1.set_yticks([])\n",
    "ax1.spines[\"right\"].set_visible(False)\n",
    "ax1.spines[\"top\"].set_visible(False)\n",
    "ax1.spines[\"left\"].set_visible(False)\n",
    "\n",
    "ax2.set_title(\"Divergence Tree Post-Pruning\")\n",
    "ax2.set_xlabel(\"Branch Length\")\n",
    "ax2.set_ylabel(\"\")\n",
    "ax2.set_ylabel(\"\")\n",
    "ax2.set_yticks([])\n",
    "ax2.spines[\"right\"].set_visible(False)\n",
    "ax2.spines[\"top\"].set_visible(False)\n",
    "ax2.spines[\"left\"].set_visible(False)\n",
    "\n",
    "# Save\n",
    "out_path = os.path.join(outdir, SCRIPT_NAME + \"_divtree-prune.\" + FMT) \n",
    "plt.savefig(out_path, dpi=dpi, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot Prune Comparison: Timetree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, (ax1,ax2) = plt.subplots(2, sharex=True, dpi=dpi)\n",
    "\n",
    "tt_pre.branch_length_to_years()\n",
    "tt_prune.branch_length_to_years()\n",
    "\n",
    "Phylo.draw(tt_pre.tree,   show_confidence=False, label_func = lambda x: '', do_show=False, axes=ax1)  \n",
    "Phylo.draw(tt_prune.tree, show_confidence=False, label_func = lambda x: '', do_show=False, axes=ax2)\n",
    "\n",
    "ax1.set_title(\"Time Tree Pre-Pruning\")\n",
    "ax2.set_title(\"Time Tree Post-Pruning\")\n",
    "\n",
    "ax1.set_title(\"Time Tree Pre-Pruning\")\n",
    "ax1.set_xlabel(\"\")\n",
    "ax1.set_ylabel(\"\")\n",
    "ax1.set_yticks([])\n",
    "ax1.spines[\"right\"].set_visible(False)\n",
    "ax1.spines[\"top\"].set_visible(False)\n",
    "ax1.spines[\"left\"].set_visible(False)\n",
    "\n",
    "ax2.set_title(\"Time Tree Post-Pruning\")\n",
    "ax2.set_xlabel(\"Clock Length\")\n",
    "ax2.set_ylabel(\"\")\n",
    "ax2.set_ylabel(\"\")\n",
    "ax2.set_yticks([])\n",
    "ax2.spines[\"right\"].set_visible(False)\n",
    "ax2.spines[\"top\"].set_visible(False)\n",
    "ax2.spines[\"left\"].set_visible(False)\n",
    "\n",
    "fig.subplots_adjust(hspace=0.5)\n",
    "\n",
    "# Save\n",
    "out_path = os.path.join(outdir, SCRIPT_NAME + \"_timetree-prune.\" + FMT) \n",
    "plt.savefig(out_path, dpi=dpi, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Update tt variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tt = tt_prune"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove Pruned Nodes from Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tt_nodes = [c.name for c in tt.tree.find_clades()]\n",
    "for rec in tree_df.iterrows():\n",
    "    node = rec[0]\n",
    "    if node not in tt_nodes:\n",
    "        tree_df.drop(node, inplace=True)\n",
    "        \n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove Pruned Nodes from Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aln = AlignIO.read(aln_path, \"fasta\")\n",
    "filtered_seq = [rec for rec in aln if rec.id in tt_nodes]\n",
    "new_aln = Align.MultipleSeqAlignment(filtered_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Add Clock Model to Dataframe\n",
    "\n",
    "- Rates\n",
    "- Dates\n",
    "- RTT Regression\n",
    "- Skyline\n",
    "- coord_x and coord_y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_df[\"timetree_rate\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_rate_fold_change\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_mutation_length\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "\n",
    "# The mean rate is the slope\n",
    "mean_rate = tt.clock_model[\"slope\"]\n",
    "\n",
    "for c in tt.tree.find_clades():\n",
    "    tree_df.at[c.name, \"timetree_mutation_length\"] = c.mutation_length\n",
    "    \n",
    "    # Relaxed Clock\n",
    "    if hasattr(c, \"branch_length_interpolator\") and c.branch_length_interpolator:\n",
    "        g = c.branch_length_interpolator.gamma\n",
    "        tree_df.at[c.name, \"timetree_rate_fold_change\"] = g\n",
    "        tree_df.at[c.name, \"timetree_rate\"] = mean_rate * g\n",
    "        \n",
    "    # Strict Clock\n",
    "    else:\n",
    "        tree_df.at[c.name, \"timetree_rate_fold_change\"] = 1\n",
    "        tree_df.at[c.name, \"timetree_rate\"] = mean_rate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create new columns\n",
    "tree_df[\"timetree_raw_date\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_num_date\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "\n",
    "if hasattr(tt.tree.root, \"marginal_inverse_cdf\"):\n",
    "    tree_df[\"timetree_num_date_confidence\"] = [[NO_DATA_CHAR,NO_DATA_CHAR] for row in range(0,len(tree_df))]\n",
    "    tree_df[\"timetree_num_date_bar\"] = [[NO_DATA_CHAR,NO_DATA_CHAR] for row in range(0,len(tree_df))]\n",
    "    \n",
    "# clock_length is the same as branch_length until running branch_length_to_years()\n",
    "tree_df[\"timetree_clock_length\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "\n",
    "# Make a copy to change branch_length\n",
    "tt_copy = copy.deepcopy(tt)\n",
    "tt_copy.branch_length_to_years()\n",
    "\n",
    "CONF_BAD_RANGE = 10000\n",
    "CONF_EXTREME = 100000\n",
    "bad_conf = [[\"node\", \"lower_bad\", \"upper_bad\", \"lower_corrected\", \"upper_corrected\"]]\n",
    "        \n",
    "for c in tt_copy.tree.find_clades():\n",
    "    # Marginal Probability\n",
    "    if hasattr(c, \"marginal_inverse_cdf\"):   \n",
    "        # Retrieve the region containing the confidence interval\n",
    "        conf = tt.get_max_posterior_region(c, fraction=CONFIDENCE)\n",
    "        conf_range = abs(conf[1] - conf[0])\n",
    "        if conf_range > CONF_BAD_RANGE or conf[0] < -CONF_EXTREME or conf[1] > CONF_EXTREME:\n",
    "            bad_conf.append([c.name, conf[0], conf[1], c.numdate, c.numdate])\n",
    "            conf[0] = c.numdate\n",
    "            conf[1] = c.numdate \n",
    "        \n",
    "        # Set as lower and upper bounds on date\n",
    "        tree_df.at[c.name, \"timetree_num_date_confidence\"][0] = conf[0]\n",
    "        tree_df.at[c.name, \"timetree_num_date_confidence\"][1] = conf[1]\n",
    "        \n",
    "        # Convert to YBP present for drawing bars\n",
    "        tree_df.at[c.name, \"timetree_num_date_bar\"][0] = CURRENT_YEAR - conf[0]\n",
    "        tree_df.at[c.name, \"timetree_num_date_bar\"][1] = CURRENT_YEAR - conf[1]\n",
    "        \n",
    "    tree_df.at[c.name, \"timetree_raw_date\"] = c.date  \n",
    "    tree_df.at[c.name, \"timetree_num_date\"] = c.numdate\n",
    "    tree_df.at[c.name, \"timetree_clock_length\"] = c.branch_length\n",
    "    \n",
    "out_path_bad_conf = os.path.join(outdir, SCRIPT_NAME + \"_bad-confidence.tsv\")\n",
    "with open(out_path_bad_conf, \"w\") as outfile:\n",
    "    for node in bad_conf:\n",
    "        outfile.write(\"\\t\".join([str(n) for n in node]) + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RTT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a copy of the tree\n",
    "tt_copy = copy.deepcopy(tt)\n",
    "tt_copy.branch_length_to_years()\n",
    "\n",
    "# Plotting the tree\n",
    "tree_df[\"timetree_coord_x\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_coord_y\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))] \n",
    "\n",
    "# Plotting the regression\n",
    "tree_df[\"timetree_reg_x\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_reg_y\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "tree_df[\"timetree_reg_bad\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "\n",
    "x_posns = get_x_positions(tt_copy.tree)\n",
    "y_posns = get_y_positions(tt_copy.tree)\n",
    "tt_reg = tt_copy.setup_TreeRegression()\n",
    "\n",
    "# Add x and y coordinates\n",
    "for c in tt_copy.tree.find_clades():     \n",
    "        \n",
    "    # Tree Node Coordinates\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    tree_df.at[c.name, 'timetree_coord_x'] = coord_x\n",
    "    tree_df.at[c.name, 'timetree_coord_y'] = coord_y\n",
    "    \n",
    "    # Regression Node Coordinates\n",
    "    reg_y = c._v\n",
    "    if c.is_terminal():\n",
    "        reg_x = tt_reg.tip_value(c)\n",
    "    else:\n",
    "        reg_x = c.numdate\n",
    "    reg_bad = c.bad_branch  if hasattr(c, 'bad_branch') else False\n",
    "    tree_df.at[c.name, 'timetree_reg_x'] = reg_x\n",
    "    tree_df.at[c.name, 'timetree_reg_y'] = reg_y    \n",
    "    tree_df.at[c.name, 'timetree_reg_bad'] = reg_bad\n",
    "\n",
    "# Fix up new values that could be none\n",
    "tree_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Divergence coord_x and coord_y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tt_copy = copy.deepcopy(tt)\n",
    "for n in tt_copy.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "    \n",
    "x_posns = get_x_positions(tt_copy.tree)\n",
    "y_posns = get_y_positions(tt_copy.tree)\n",
    "\n",
    "# Add x and y coordinates as other attributes\n",
    "for c in tt_copy.tree.find_clades():\n",
    "    # x coordinates will be of branch length units\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    # y coordinates will be arbitrary, based on number of tips\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    # Update coordinates in dataframe\n",
    "    tree_df.at[c.name, 'coord_x'] = coord_x\n",
    "    tree_df.at[c.name, 'coord_y'] = coord_y    \n",
    "\n",
    "# Visualize dataframe\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add Metadata as Comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_to_comment(tt.tree, tree_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GTR Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_gtr = os.path.join(outdir, SCRIPT_NAME + \"_gtr.txt\" )\n",
    "gtr_model_str = str(tt.gtr)\n",
    "with open(out_path_gtr,\"w\") as outfile:\n",
    "    outfile.write(gtr_model_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_aln = os.path.join(outdir, SCRIPT_NAME + \".fasta\" )\n",
    "with open(out_path_aln, \"w\") as outfile:\n",
    "    count = SeqIO.write(new_aln, outfile, \"fasta\")\n",
    "    print(count, \"alignments written.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframe\n",
    "out_path_df = os.path.join(outdir, SCRIPT_NAME + \".tsv\" )\n",
    "out_path_pickle_df = os.path.join(outdir,  SCRIPT_NAME + \".df.obj\" )\n",
    "\n",
    "tree_df.to_csv(out_path_df, sep=\"\\t\")\n",
    "with open(out_path_pickle_df,\"wb\") as outfile:\n",
    "    dill.dump(tree_df, outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Timetrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tt_copy = copy.deepcopy(tt)\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \"_timetree.nexus\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_timetree.treetime.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tt_copy, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tt_copy.tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \"_timetree.nwk\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Divtrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tt_copy = copy.deepcopy(tt)\n",
    "# Convert to divtree\n",
    "for n in tt_copy.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \"_divtree.nexus\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_divtree.treetime.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tt_copy, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tt_copy.tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \"_divtree.nwk\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Augur JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=out_path_nwk, \n",
    "    aln_path=aln_path, \n",
    "    tree=tt.tree,     \n",
    "    #tree=tree_div, \n",
    "    tree_df=tree_df, \n",
    "    color_keyword_exclude=[\"color\", \"coord\", \"reg\", \"lat\", \"lon\"],\n",
    "    type_convert = {\n",
    "        \"Branch_Number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])\n",
    "\n",
    "if hasattr(tt, \"merger_model\"):\n",
    "    skyline, conf = tt.merger_model.skyline_inferred(gen=50, confidence=2)\n",
    "    augur_dict['skyline'] = [[float(x) for x in skyline.x], [float(y) for y in conf[0]],\n",
    "                           [float(y) for y in skyline.y], [float(y) for y in conf[1]]]\n",
    "\n",
    "out_path_augur_json = os.path.join(outdir, SCRIPT_NAME + \"_augur.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Auspice JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "auspice_dict = auspice_export(\n",
    "    tree=tt.tree, \n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=auspice_colors_path,\n",
    "    auspice_latlons_path=auspice_latlon_path, \n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(outdir, SCRIPT_NAME + \"_auspice.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "#out_path_auspice_remote_json = os.path.join(auspice_remote_dir_path, AUSPICE_PREFIX + SCRIPT_NAME.replace(\"_\",\"-\") + \".json\" )\n",
    "#utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "#export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "#print(\"Validation successful for remote JSON.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
