{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "collectible-letter",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sqlite3\n",
    "import random\n",
    "import pandas as pd\n",
    "import copy\n",
    "\n",
    "random.seed(1234)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "understood-sleep",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "database = os.path.join(project_dir, \"results\", \"sqlite_db\", \"yersinia_pestis_db.sqlite\")\n",
    "conn = sqlite3.connect(database)\n",
    "cur = conn.cursor()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "damaged-being",
   "metadata": {},
   "source": [
    "## 1 Sample Per Branch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "preliminary-tobago",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.MED1 ('SAMN15063847', 'modern')\n",
      "0.ANT1 ('SAMN07722911', 'modern')\n",
      "4.ANT1 ('SAMN12991207', 'modern')\n",
      "0.PE2 ('SAMN05521424', 'modern')\n",
      "2.MED0 ('SAMN07176224', 'modern')\n",
      "1.PRE1 ('SAMEA5818828', 'ancient')\n",
      "1.PRE0 ('SAMEA5818806', 'ancient')\n",
      "0.PE5 ('SAMN12138635', 'modern')\n",
      "0.ANT4 ('SAMEA5661367', 'ancient')\n",
      "1.ORI1 ('SAMN02403946', 'modern')\n",
      "0.ANT3 ('SAMN02777961', 'modern')\n",
      "2.ANT3 ('SAMN02769655', 'modern')\n",
      "1.ANT1 ('SAMN02404404', 'modern')\n",
      "1.ORI3 ('SAMN02470617', 'modern')\n",
      "2.MED2 ('SAMN02403012', 'modern')\n",
      "0.ANT2 ('SAMN02402994', 'modern')\n",
      "1.ORI2 ('SAMN02403056', 'modern')\n",
      "1.IN3 ('SAMN02403073', 'modern')\n",
      "1.PRE2 ('SAMEA5054090', 'ancient')\n",
      "0.PRE1 ('SAMEA104233048', 'ancient')\n",
      "0.PE8 ('SAMEA104488961', 'ancient')\n",
      "1.PRE3 ('SAMEA3937654', 'ancient')\n",
      "3.ANT2 ('SAMN05521438', 'modern')\n",
      "0.PE4 ('SAMN02403037', 'modern')\n",
      "0.ANT5 ('SAMN05150065', 'modern')\n",
      "2.ANT1 ('SAMN02403941', 'modern')\n",
      "1.IN2 ('SAMN02403071', 'modern')\n",
      "2.MED3 ('SAMN13907443', 'modern')\n",
      "1.IN1 ('SAMN02403014', 'modern')\n",
      "2.ANT2 ('SAMN13021964', 'modern')\n",
      "3.ANT1 ('SAMN02403051', 'modern')\n",
      "0.PE7 ('SAMN02403027', 'modern')\n",
      "0.PE10 ('SAMN02403015', 'modern')\n",
      "0.PRE2 ('SAMEA3541827', 'ancient')\n"
     ]
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "    SELECT\n",
    "      BioSampleAccession,\n",
    "      BioSampleBranch,\n",
    "      BioSampleComment\n",
    "    FROM\n",
    "      BioSample\n",
    "    LEFT Join\n",
    "      Assembly ON BioSampleAccession==AssemblyBioSampleAccession\n",
    "    WHERE\n",
    "      BioSampleComment LIKE '%KEEP%' AND BioSampleComment NOT LIKE \"%Local%\"\n",
    "    \"\"\"\n",
    "    \n",
    "\n",
    "result = cur.execute(query).fetchall()\n",
    "\n",
    "# Create a mapping of branches to samples\n",
    "branch_dict = {}\n",
    "for rec in result:\n",
    "    accession = rec[0]\n",
    "    branch = rec[1]\n",
    "    comment = rec[2]\n",
    "    origin = \"ancient\" if \"Ancient\" in comment else \"modern\"\n",
    "    # Strip sub branches that are letter designations\n",
    "    while branch[-1].isalpha():\n",
    "        branch = branch[:-1]\n",
    "    if branch not in branch_dict:\n",
    "        branch_dict[branch] = []        \n",
    "    branch_dict[branch].append((accession, origin))        \n",
    "\n",
    "random_samples = {}\n",
    "# Randomly sample\n",
    "for branch in branch_dict:\n",
    "    sample = random.choice(branch_dict[branch])\n",
    "    random_samples[branch] = sample\n",
    "\n",
    "# print out\n",
    "for branch in random_samples:\n",
    "    print(branch, random_samples[branch])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "binary-rubber",
   "metadata": {},
   "source": [
    "## Explore Diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "southern-currency",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669605.1_ASM966960v1_genomic</th>\n",
       "      <td>42117</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722922</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313243_45</th>\n",
       "      <td>Azov38</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313243_45</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313246_49</th>\n",
       "      <td>Gdansk8</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Poland</td>\n",
       "      <td>Pomeranian Voivodeship</td>\n",
       "      <td>52.215933</td>\n",
       "      <td>19.134422</td>\n",
       "      <td>54.2456</td>\n",
       "      <td>18.1099</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313246_49</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6651390</th>\n",
       "      <td>AGU010</td>\n",
       "      <td>[1435:1477]</td>\n",
       "      <td>[-586:-544]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6651390</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637004</th>\n",
       "      <td>AGU025</td>\n",
       "      <td>[1441:1612]</td>\n",
       "      <td>[-580:-409]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637004</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637002</th>\n",
       "      <td>AGU007B</td>\n",
       "      <td>[1463:1632]</td>\n",
       "      <td>[-558:-389]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637002</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>612 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      strain         date      date_bp  \\\n",
       "sample                                                                   \n",
       "GCA_009909635.1_ASM990963v1_genomic     9_10         1923          -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic    42126         2006          -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic    42123         2005          -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic    42118         2005          -16   \n",
       "GCA_009669605.1_ASM966960v1_genomic    42117         2005          -16   \n",
       "...                                      ...          ...          ...   \n",
       "SAMEA7313243_45                       Azov38  [1400:1700]  [-621:-321]   \n",
       "SAMEA7313246_49                      Gdansk8  [1400:1700]  [-621:-321]   \n",
       "SAMEA6651390                          AGU010  [1435:1477]  [-586:-544]   \n",
       "SAMEA6637004                          AGU025  [1441:1612]  [-580:-409]   \n",
       "SAMEA6637002                         AGU007B  [1463:1632]  [-558:-389]   \n",
       "\n",
       "                                       country                province  \\\n",
       "sample                                                                   \n",
       "GCA_009909635.1_ASM990963v1_genomic     Russia           Rostov Oblast   \n",
       "GCA_009669545.1_ASM966954v1_genomic      China                Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic      China                Xinjiang   \n",
       "GCA_009669565.1_ASM966956v1_genomic      China                Xinjiang   \n",
       "GCA_009669605.1_ASM966960v1_genomic      China                Xinjiang   \n",
       "...                                        ...                     ...   \n",
       "SAMEA7313243_45                         Russia           Rostov Oblast   \n",
       "SAMEA7313246_49                         Poland  Pomeranian Voivodeship   \n",
       "SAMEA6651390                         Lithuania          Vilnius County   \n",
       "SAMEA6637004                         Lithuania          Vilnius County   \n",
       "SAMEA6637002                         Lithuania          Vilnius County   \n",
       "\n",
       "                                     country_lat  country_lon province_lat  \\\n",
       "sample                                                                       \n",
       "GCA_009909635.1_ASM990963v1_genomic    64.686314    97.745306      47.6222   \n",
       "GCA_009669545.1_ASM966954v1_genomic    35.000074   104.999927      42.4805   \n",
       "GCA_009669555.1_ASM966955v1_genomic    35.000074   104.999927      42.4805   \n",
       "GCA_009669565.1_ASM966956v1_genomic    35.000074   104.999927      42.4805   \n",
       "GCA_009669605.1_ASM966960v1_genomic    35.000074   104.999927      42.4805   \n",
       "...                                          ...          ...          ...   \n",
       "SAMEA7313243_45                        64.686314    97.745306      47.6222   \n",
       "SAMEA7313246_49                        52.215933    19.134422      54.2456   \n",
       "SAMEA6651390                           55.350000    23.750000      54.8227   \n",
       "SAMEA6637004                           55.350000    23.750000      54.8227   \n",
       "SAMEA6637002                           55.350000    23.750000      54.8227   \n",
       "\n",
       "                                    province_lon           biovar  \\\n",
       "sample                                                              \n",
       "GCA_009909635.1_ASM990963v1_genomic      40.7958       Medievalis   \n",
       "GCA_009669545.1_ASM966954v1_genomic      85.4633          Antiqua   \n",
       "GCA_009669555.1_ASM966955v1_genomic      85.4633          Antiqua   \n",
       "GCA_009669565.1_ASM966956v1_genomic      85.4633          Antiqua   \n",
       "GCA_009669605.1_ASM966960v1_genomic      85.4633          Antiqua   \n",
       "...                                          ...              ...   \n",
       "SAMEA7313243_45                          40.7958  Second Pandemic   \n",
       "SAMEA7313246_49                          18.1099  Second Pandemic   \n",
       "SAMEA6651390                             25.2495  Second Pandemic   \n",
       "SAMEA6637004                             25.2495  Second Pandemic   \n",
       "SAMEA6637002                             25.2495  Second Pandemic   \n",
       "\n",
       "                                    branch_major branch_minor  \\\n",
       "sample                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic        2.MED       2.MED1   \n",
       "GCA_009669545.1_ASM966954v1_genomic        0.ANT       0.ANT1   \n",
       "GCA_009669555.1_ASM966955v1_genomic        0.ANT       0.ANT1   \n",
       "GCA_009669565.1_ASM966956v1_genomic        0.ANT       0.ANT1   \n",
       "GCA_009669605.1_ASM966960v1_genomic        0.ANT       0.ANT1   \n",
       "...                                          ...          ...   \n",
       "SAMEA7313243_45                            1.PRE       1.PRE1   \n",
       "SAMEA7313246_49                            1.PRE       1.PRE1   \n",
       "SAMEA6651390                               1.PRE       1.PRE1   \n",
       "SAMEA6637004                               1.PRE       1.PRE1   \n",
       "SAMEA6637002                               1.PRE       1.PRE1   \n",
       "\n",
       "                                    biosample_accession  \\\n",
       "sample                                                    \n",
       "GCA_009909635.1_ASM990963v1_genomic        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic        SAMN07722923   \n",
       "GCA_009669605.1_ASM966960v1_genomic        SAMN07722922   \n",
       "...                                                 ...   \n",
       "SAMEA7313243_45                         SAMEA7313243_45   \n",
       "SAMEA7313246_49                         SAMEA7313246_49   \n",
       "SAMEA6651390                               SAMEA6651390   \n",
       "SAMEA6637004                               SAMEA6637004   \n",
       "SAMEA6637002                               SAMEA6637002   \n",
       "\n",
       "                                                     biosample_comment  \\\n",
       "sample                                                                   \n",
       "GCA_009909635.1_ASM990963v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669605.1_ASM966960v1_genomic              KEEP: Assembly Modern   \n",
       "...                                                                ...   \n",
       "SAMEA7313243_45                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA7313246_49                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA6651390                                         KEEP: SRA Ancient   \n",
       "SAMEA6637004                                         KEEP: SRA Ancient   \n",
       "SAMEA6637002                                         KEEP: SRA Ancient   \n",
       "\n",
       "                                     branch_number continent  \n",
       "sample                                                        \n",
       "GCA_009909635.1_ASM990963v1_genomic              2    Europe  \n",
       "GCA_009669545.1_ASM966954v1_genomic              0      Asia  \n",
       "GCA_009669555.1_ASM966955v1_genomic              0      Asia  \n",
       "GCA_009669565.1_ASM966956v1_genomic              0      Asia  \n",
       "GCA_009669605.1_ASM966960v1_genomic              0      Asia  \n",
       "...                                            ...       ...  \n",
       "SAMEA7313243_45                                  1    Europe  \n",
       "SAMEA7313246_49                                  1    Europe  \n",
       "SAMEA6651390                                     1    Europe  \n",
       "SAMEA6637004                                     1    Europe  \n",
       "SAMEA6637002                                     1    Europe  \n",
       "\n",
       "[612 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse the metadata\n",
    "metadata_path = os.path.join(project_dir, \"results\", \"metadata\", \"all\",\"metadata.tsv\")\n",
    "\n",
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.fillna(\"NA\", inplace=True)\n",
    "metadata_df.set_index(\"sample\", inplace=True)\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "round-wedding",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the distance matrix\n",
    "#dist_mat_path = os.path.join(project_dir, \"results\", \"snippy_multi\", \"all\",\"chromosome\", \"filter100\", \"snippy-multi.snps.dist\")#\n",
    "#dist_mat_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "#dist_mat_df.set_index(dist_mat_df.columns[0], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "senior-communication",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2.MED1\n",
      "\t Rostov Oblast\n",
      "\t\t 1923 GCA_009909635.1_ASM990963v1_genomic\n",
      "\t Chechnya\n",
      "\t\t 1953 GCA_009296005.1_ASM929600v1_genomic\n",
      "\t Kabardino-Balkaria\n",
      "\t\t 1997 GCA_008630485.1_ASM863048v1_genomic\n",
      "\t Republic of Dagestan\n",
      "\t\t 1984 GCA_008630395.1_ASM863039v1_genomic\n",
      "\t Qazakh District\n",
      "\t\t 1984 GCA_008630575.1_ASM863057v1_genomic\n",
      "\t Ingushetia\n",
      "\t\t 1970 GCA_006376535.1_ASM637653v1_genomic\n",
      "\t Goranboy District\n",
      "\t\t 1968 GCA_006376495.1_ASM637649v1_genomic\n",
      "\t Baku\n",
      "\t\t 1968 GCA_006376475.1_ASM637647v1_genomic\n",
      "\t Shamkir District\n",
      "\t\t 1959 GCA_006376435.1_ASM637643v1_genomic\n",
      "\t Stavropol Krai\n",
      "\t\t 1950 GCA_006376415.1_ASM637641v1_genomic\n",
      "\t West Kazakhstan Region\n",
      "\t\t 1990 GCA_003086155.1_ASM308615v1_genomic\n",
      "\t\t 1950 GCA_016102755.1_ASM1610275v1_genomic\n",
      "\t\t 1917 GCA_016102575.1_ASM1610257v1_genomic\n",
      "\t Astrakhan Oblast\n",
      "\t\t 1992 GCA_001617725.1_Yersinia_pestis_M-1484_genomic\n",
      "\t\t 1962 GCA_016102745.1_ASM1610274v1_genomic\n",
      "\t\t 1930 GCA_016102665.1_ASM1610266v1_genomic\n",
      "\t Xinjiang\n",
      "\t\t 2005 GCA_004023595.1_ASM402359v1_genomic\n",
      "\t Fizuli District\n",
      "\t\t 1967 GCA_003074355.1_ASM307435v1_genomic\n",
      "\t Kyzylorda Province\n",
      "\t\t 1967 GCA_003074295.1_ASM307429v1_genomic\n",
      "\t Republic of Kalmykia\n",
      "\t\t 2014 GCA_003085995.1_ASM308599v1_genomic\n",
      "\t\t 1986 GCA_016102905.1_ASM1610290v1_genomic\n",
      "\t Altai Republic\n",
      "\t\t 2014 GCA_003085975.1_ASM308597v1_genomic\n",
      "\t Talas Region\n",
      "\t\t 1980 GCA_002412305.1_Y.pestis_A-1809_genomic\n",
      "\t Shymkent\n",
      "\t\t 1993 GCA_003086095.1_ASM308609v1_genomic\n",
      "\t Almaty Region\n",
      "\t\t 1988 GCA_003086115.1_ASM308611v1_genomic\n",
      "\t Navoiy Region\n",
      "\t\t 1983 GCA_003086075.1_ASM308607v1_genomic\n",
      "\t Atyrau\n",
      "\t\t 1991 GCA_001617815.1_Yersinia_pestis_M-1453_genomic\n",
      "\t Mangystau Province\n",
      "\t\t 1978 GCA_001617805.1_Yersinia_pestis_173_genomic\n",
      "\t Balkan\n",
      "\t\t 1977 GCA_001617785.1_Yersinia_pestis_M-549_genomic\n",
      "\t Kurdistan\n",
      "\t\t 2007 GCA_000970105.1_ASM97010v1_genomic\n",
      "\t\t 1968 GCA_000006645.1_ASM664v1_genomic\n",
      "\t Atyrau Region\n",
      "\t\t 1990 GCA_016102965.1_ASM1610296v1_genomic\n",
      "\t\t 1959 GCA_016102875.1_ASM1610287v1_genomic\n",
      "\t Kazakhstan\n",
      "\t\t 1955 GCA_016102945.1_ASM1610294v1_genomic\n",
      "\t Aktobe Region\n",
      "\t\t 1955 GCA_016102765.1_ASM1610276v1_genomic\n",
      "\t Kyzylorda Region\n",
      "\t\t 1945 GCA_016102655.1_ASM1610265v1_genomic\n",
      "\t Nagorno-Karabakh Republic\n",
      "\t\t 1931 GCA_016102545.1_ASM1610254v1_genomic\n",
      "\t  West Kazakhstan region\n",
      "\t\t 1912 GCA_016102555.1_ASM1610255v1_genomic\n",
      "\t Russia\n",
      "\t\t 2015 GCA_015336695.1_ASM1533669v1_genomic\n",
      "\t\t 1989 GCA_015337645.1_ASM1533764v1_genomic\n",
      "\t Turkmenistan\n",
      "\t\t 1949 GCA_015350455.1_ASM1535045v1_genomic\n",
      "\n",
      "0.ANT1\n",
      "\t Xinjiang\n",
      "\t\t 2006 GCA_009669545.1_ASM966954v1_genomic\n",
      "\t\t 1980 GCA_009669805.1_ASM966980v1_genomic\n",
      "\n",
      "4.ANT1\n",
      "\t Tuva Republic\n",
      "\t\t 1977 GCA_009363195.1_ASM936319v1_genomic\n",
      "\t\t 2012 GCA_003086015.1_ASM308601v1_genomic\n",
      "\t Altai Republic\n",
      "\t\t 2012 GCA_003074255.1_ASM307425v1_genomic\n",
      "\t Bayan-Ölgii\n",
      "\t\t 1988 GCA_003074235.1_ASM307423v1_genomic\n",
      "\t Russia\n",
      "\t\t 2015 GCA_015338045.1_ASM1533804v1_genomic\n",
      "\n",
      "0.PE2\n",
      "\t Republic of Dagestan\n",
      "\t\t 2011 GCA_008630455.1_ASM863045v1_genomic\n",
      "\t\t 1978 GCA_006376585.1_ASM637658v1_genomic\n",
      "\t Shirak Province\n",
      "\t\t 1969 GCA_006376545.1_ASM637654v1_genomic\n",
      "\t Shahbuz Rayon\n",
      "\t\t 1968 GCA_006376505.1_ASM637650v1_genomic\n",
      "\t Goranboy District\n",
      "\t\t 1989 GCA_001295055.1_ASM129505v1_genomic\n",
      "\t Samtskhe-Javakheti\n",
      "\t\t 1979 GCA_002993145.1_ASM299314v1_genomic\n",
      "\t Syunik Province\n",
      "\t\t 1984 GCA_003074435.1_ASM307443v1_genomic\n",
      "\t Gegharkunik Province\n",
      "\t\t 1979 GCA_003074415.1_ASM307441v1_genomic\n",
      "\t Georgia\n",
      "\t\t 1975 GCA_000382505.1_Yersinia_pestis_G1670E_assembly_genomic\n",
      "\t Russia\n",
      "\t\t 2012 GCA_015338205.1_ASM1533820v1_genomic\n",
      "\n",
      "2.MED0\n",
      "\t Karachay-Cherkessia\n",
      "\t\t 1996 GCA_008630435.1_ASM863043v1_genomic\n",
      "\t\t 1971 GCA_006376575.1_ASM637657v1_genomic\n",
      "\t Kabardino-Balkaria\n",
      "\t\t 1986 GCA_003074275.1_ASM307427v1_genomic\n",
      "\t Russia\n",
      "\t\t 2001 GCA_015337285.1_ASM1533728v1_genomic\n",
      "\n",
      "0.PE5\n",
      "\t Bayan-Ölgii\n",
      "\t\t 1990 GCA_006546745.1_ASM654674v1_genomic\n",
      "\t Govi-Altai\n",
      "\t\t 1986 GCA_006546725.1_ASM654672v1_genomic\n",
      "\t Khovd\n",
      "\t\t 1987 GCA_006546665.1_ASM654666v1_genomic\n",
      "\t Övörkhangai\n",
      "\t\t 1982 GCA_006517425.1_ASM651742v1_genomic\n",
      "\t Omnogovi\n",
      "\t\t 1986 GCA_000950945.1_ASM95094v1_genomic\n",
      "\t Ömnögovi\n",
      "\t\t 1973 GCA_002981995.1_ASM298199v1_genomic\n",
      "\n",
      "1.ORI1\n",
      "\t Arizona\n",
      "\t\t 1967 GCA_000834775.1_ASM83477v1_genomic\n",
      "\t\t 2011 GCA_001509855.1_ASM150985v1_genomic\n",
      "\t California\n",
      "\t\t 1954 GCA_000834335.1_ASM83433v1_genomic\n",
      "\t\t 1988 GCA_000181455.2_ASM18145v2_genomic\n",
      "\t State of Mato Grosso\n",
      "\t\t 1966 GCA_000182525.1_ASM18252v1_genomic\n",
      "\t Maharashtra\n",
      "\t\t 1924 GCA_000182505.1_ASM18250v1_genomic\n",
      "\t New Mexico\n",
      "\t\t 2002 GCA_000834495.1_ASM83449v1_genomic\n",
      "\t Midlands\n",
      "\t\t 1994 GCA_000986995.1_YPES001-SEQ-2-ASM-1_genomic\n",
      "\t La Paz\n",
      "\t\t 1969 GCA_003797205.1_ASM379720v1_genomic\n",
      "\t Central Java\n",
      "\t\t 1957 GCA_003798085.1_ASM379808v1_genomic\n",
      "\t MSila\n",
      "\t\t 2008 GCA_001487065.1_Yersinia_pestis_genomic\n",
      "\t Texas\n",
      "\t\t 2009 GCA_001509865.1_ASM150986v1_genomic\n",
      "\t Tamil Nadu\n",
      "\t\t 1998 GCA_000475155.2_YP24H_v2_genomic\n",
      "\t Saskatchewan\n",
      "\t\t 2010 GCA_000705315.1_EDB10_058v1_genomic\n",
      "\t La Libertad\n",
      "\t\t 2010 GCA_000269245.1_ASM26924v1_genomic\n",
      "\t Cajamarca\n",
      "\t\t 2008 GCA_000268465.1_ASM26846v1_genomic\n",
      "\t Nevada\n",
      "\t\t 2009 GCA_000255755.1_ASM25575v1_genomic\n",
      "\t Fujian\n",
      "\t\t 1946 GCA_000324825.1_ASM32482v1_genomic\n",
      "\n",
      "0.ANT3\n",
      "\t Naryn Region\n",
      "\t\t 1947 GCA_000734805.1_ASM73480v1_genomic\n",
      "\t Osh Region\n",
      "\t\t 1966 GCA_002412225.1_Y.pestis_A-1486_genomic\n",
      "\t Kyrgyzstan\n",
      "\t\t 1968 GCA_001188675.1_ASM118867v1_genomic\n",
      "\t Xinjiang\n",
      "\t\t 1979 GCA_000324165.1_ASM32416v1_genomic\n",
      "\n",
      "2.ANT3\n",
      "\t Heilongjiang\n",
      "\t\t 1940 GCA_000834275.1_ASM83427v1_genomic\n",
      "\t\t 2008 GCA_000834885.1_ASM83488v1_genomic\n",
      "\t Zabaykalsky Krai\n",
      "\t\t 1970 GCA_001601285.1_ASM160128v1_genomic\n",
      "\t Jilin\n",
      "\t\t 1959 GCA_000324965.1_ASM32496v1_genomic\n",
      "\t Inner Mongolia\n",
      "\t\t 1989 GCA_000324585.1_ASM32458v1_genomic\n",
      "\t\t 1957 GCA_000324565.1_ASM32456v1_genomic\n",
      "\t St. Petersberg\n",
      "\t\t 1965 GCA_000323585.1_ASM32358v1_genomic\n",
      "\t Russia\n",
      "\t\t 1966 GCA_015336465.1_ASM1533646v1_genomic\n",
      "\n",
      "1.ANT1\n",
      "\t Uganda\n",
      "\t\t 2004 GCA_000169615.1_ASM16961v1_genomic\n",
      "\t Democratic Republic of the Congo\n",
      "\t\t 1954 GCA_003798225.1_ASM379822v1_genomic\n",
      "\t Nairobi\n",
      "\t\t 1985 GCA_000835005.1_ASM83500v1_genomic\n",
      "\n",
      "1.ORI3\n",
      "\t Madagascar\n",
      "\t\t 2005 GCA_000169635.1_ASM16963v1_genomic\n",
      "\t\t 1970 GCA_001261985.1_YPW_3_IHQD_genomic\n",
      "\t Province de Fianarantsoa\n",
      "\t\t 1995 GCA_000168235.2_ASM16823v2_genomic\n",
      "\t Gujarat\n",
      "\t\t 1994 GCA_000475115.2_YP113_v2_genomic\n",
      "\t Lanzou\n",
      "\t\t 1926 GCA_000324805.2_EV76-CN_genomic\n",
      "\t Kirov Oblast\n",
      "\t\t 1941 GCA_000590535.1_EV_NIIEG_genomic\n",
      "\n",
      "2.MED2\n",
      "\t Xinjiang\n",
      "\t\t 1973 GCA_000169655.1_ASM16965v1_genomic\n",
      "\n",
      "0.ANT2\n",
      "\t Xinjiang\n",
      "\t\t 2003 GCA_000169675.1_ASM16967v1_genomic\n",
      "\n",
      "1.ORI2\n",
      "\t Yunnan\n",
      "\t\t 1991 GCA_000170275.1_ASM17027v1_genomic\n",
      "\t Guangxi Zhuang\n",
      "\t\t 2000 GCA_000325445.1_ASM32544v1_genomic\n",
      "\t Myanmar\n",
      "\t\t 1995 GCA_000325405.1_ASM32540v1_genomic\n",
      "\t  Ho Chi Minh City\n",
      "\t\t 1988 GCA_013421535.1_ASM1342153v1_genomic\n",
      "\t Vietnam\n",
      "\t\t 1986 GCA_013421545.1_ASM1342154v1_genomic\n",
      "\t Khánh Hoà Province\n",
      "\t\t 1962 GCA_013421465.1_ASM1342146v1_genomic\n",
      "\t Lâm Đồng Province\n",
      "\t\t 1967 GCA_013421455.1_ASM1342145v1_genomic\n",
      "\n",
      "1.IN3\n",
      "\t Yunnan\n",
      "\t\t 1979 GCA_000324785.1_ASM32478v1_genomic\n",
      "\t\t 2006 GCA_000022805.1_ASM2280v1_genomic\n",
      "\t\t 1954 GCA_000325365.1_ASM32536v1_genomic\n",
      "\n",
      "0.PE4\n",
      "\t Altai Republic\n",
      "\t\t 2007 GCA_002188775.1_ASM218877v1_genomic\n",
      "\t\t 1961 GCA_001295115.1_ASM129511v1_genomic\n",
      "\t Talas Region\n",
      "\t\t 1980 GCA_001294975.1_ASM129497v1_genomic\n",
      "\t Sughd Province\n",
      "\t\t 1979 GCA_001294885.1_ASM129488v1_genomic\n",
      "\t Bayankhongor\n",
      "\t\t 1984 GCA_001294805.1_ASM129480v1_genomic\n",
      "\t Inner Mongolia\n",
      "\t\t 1970 GCA_000007885.1_ASM788v1_genomic\n",
      "\t Qinghai\n",
      "\t\t 2001 GCA_000325185.1_ASM32518v1_genomic\n",
      "\t Sečuán\n",
      "\t\t 2000 GCA_000324665.1_ASM32466v1_genomic\n",
      "\t Russia\n",
      "\t\t 2012 GCA_015336865.1_ASM1533686v1_genomic\n",
      "\n",
      "0.ANT5\n",
      "\t Issyk-Kul Region\n",
      "\t\t 1952 GCA_003074535.1_ASM307453v1_genomic\n",
      "\t\t 1983 GCA_002412255.1_Y.pestis_A-1836_genomic\n",
      "\t Kyrgyzstan\n",
      "\t\t 2019 GCA_015274745.1_ASM1527474v1_genomic\n",
      "\n",
      "2.ANT1\n",
      "\t Maharashtra\n",
      "\t\t 1924 GCA_002005285.1_ASM200528v1_genomic\n",
      "\t Himachal Pradesh\n",
      "\t\t 2002 GCA_000475135.2_YPS03_v2_genomic\n",
      "\t Gandaki Pradesh\n",
      "\t\t 1967 GCA_000182485.1_ASM18248v1_genomic\n",
      "\t Tibet\n",
      "\t\t 1990 GCA_000323325.1_ASM32332v1_genomic\n",
      "\n",
      "1.IN2\n",
      "\t Tibet\n",
      "\t\t 1976 GCA_000022845.1_ASM2284v1_genomic\n",
      "\t\t 2007 GCA_015158485.1_ASM1515848v1_genomic\n",
      "\t Qinghai\n",
      "\t\t 1991 GCA_000324745.1_ASM32474v1_genomic\n",
      "\t\t 1964 GCA_000324705.1_ASM32470v1_genomic\n",
      "\t Gansu\n",
      "\t\t 1982 GCA_000324725.1_ASM32472v1_genomic\n",
      "\n",
      "2.MED3\n",
      "\t Shaanxi\n",
      "\t\t 2006 GCA_000325345.1_ASM32534v1_genomic\n",
      "\t Ningxia\n",
      "\t\t 1978 GCA_000325125.1_ASM32512v1_genomic\n",
      "\t Gansu\n",
      "\t\t 1963 GCA_000325105.1_ASM32510v1_genomic\n",
      "\t Hebei\n",
      "\t\t 1994 GCA_000325065.1_ASM32506v1_genomic\n",
      "\t Inner Mongolia\n",
      "\t\t 1991 GCA_000325045.1_ASM32504v1_genomic\n",
      "\t\t 1957 GCA_000324485.1_ASM32448v1_genomic\n",
      "\t\t 2018 GCA_014734395.1_ASM1473439v1_genomic\n",
      "\t Jilin\n",
      "\t\t 1958 GCA_000324945.1_ASM32494v1_genomic\n",
      "\t Qinghai\n",
      "\t\t 1961 GCA_000324005.1_ASM32400v1_genomic\n",
      "\t  Dornogovi\n",
      "\t\t 2018 GCA_013392085.1_ASM1339208v1_genomic\n",
      "\n",
      "3.ANT2\n",
      "\t Govi-Altai\n",
      "\t\t 1998 GCA_000325305.1_ASM32530v1_genomic\n",
      "\t Dornogovi\n",
      "\t\t 1997 GCA_000325285.1_ASM32528v1_genomic\n",
      "\t Bayan-Ölgii\n",
      "\t\t 2000 GCA_000325205.1_ASM32520v1_genomic\n",
      "\t Mongolia\n",
      "\t\t 2017 GCA_013391945.1_ASM1339194v1_genomic\n",
      "\n",
      "1.IN1\n",
      "\t Xinjiang\n",
      "\t\t 1985 GCA_000325165.1_ASM32516v1_genomic\n",
      "\t Qinghai\n",
      "\t\t 1954 GCA_000323965.1_ASM32396v1_genomic\n",
      "\n",
      "2.ANT2\n",
      "\t Tibet\n",
      "\t\t 1996 GCA_000324905.1_ASM32490v1_genomic\n",
      "\t Qinghai\n",
      "\t\t 2004 GCA_000320445.1_ASM32044v1_genomic\n",
      "\n",
      "3.ANT1\n",
      "\t Gansu\n",
      "\t\t 1961 GCA_000324325.1_ASM32432v1_genomic\n",
      "\t\t 1989 GCA_000323525.1_ASM32352v1_genomic\n",
      "\t Qinghai\n",
      "\t\t 2004 GCA_000323505.1_ASM32350v1_genomic\n",
      "\n",
      "0.PE7\n",
      "\t Qinghai\n",
      "\t\t 1961 GCA_000323845.1_ASM32384v1_genomic\n",
      "\n",
      "0.PE10\n",
      "\t Qinghai\n",
      "\t\t 2004 GCA_000323625.1_ASM32362v1_genomic\n",
      "\n",
      "1.PRE1\n",
      "\t Nidwalden\n",
      "\t\t 1560 SAMEA5818818\n",
      "\t Bavaria\n",
      "\t\t 1525 SAMEA5818817\n",
      "\t\t 1342 SAMEA5818811\n",
      "\t\t 1336 SAMEA5818809\n",
      "\t\t 1543 SAMEA5818808\n",
      "\t East of England\n",
      "\t\t 1505 SAMEA5818815\n",
      "\t Baden-Württemberg\n",
      "\t\t 1556 SAMEA5818805\n",
      "\t Brandenburg\n",
      "\t\t 1633 SAMEA5818803\n",
      "\t Greater London\n",
      "\t\t 1597 SAMEA5818798\n",
      "\t\t 1349 SAMN00715800\n",
      "\t Occitanie\n",
      "\t\t 1334 SAMEA5054093\n",
      "\t Oslo\n",
      "\t\t 1330 SAMEA5054092\n",
      "\t Catalonia\n",
      "\t\t 1360 SAMEA3937653\n",
      "\t Provence-Alpes-Côte d'Azur\n",
      "\t\t 1721 SAMEA3713711\n",
      "\t Lazio\n",
      "\t\t 1350 SAMEA7293136\n",
      "\t Chechnya\n",
      "\t\t 1650 SAMEA7293135\n",
      "\t Rostov Oblast\n",
      "\t\t 1767 SAMEA7313236_38\n",
      "\t\t 1550 SAMEA7313243_45\n",
      "\t Pomeranian Voivodeship\n",
      "\t\t 1550 SAMEA7313246_49\n",
      "\t Vilnius County\n",
      "\t\t 1456 SAMEA6651390\n",
      "\t\t 1526 SAMEA6637004\n",
      "\t\t 1547 SAMEA6637002\n",
      "\n",
      "1.PRE0\n",
      "\t Tatarstan\n",
      "\t\t 1350 SAMEA5818806\n",
      "\n",
      "0.ANT4\n",
      "\t Bavaria\n",
      "\t\t 478 SAMEA5661390\n",
      "\t\t 516 SAMEA5661389\n",
      "\t\t 517 SAMEA5661385\n",
      "\t\t 511 SAMEA5661360\n",
      "\t\t 498 SAMEA4354665\n",
      "\t Valencia Community\n",
      "\t\t 521 SAMEA5661384\n",
      "\t Centre-Loire Valley\n",
      "\t\t 765 SAMEA5661367\n",
      "\t East of England\n",
      "\t\t 575 SAMEA5661364\n",
      "\t\t 555 SAMEA5661363\n",
      "\t Issyk-Kul Region\n",
      "\t\t 237 SAMEA1061800\n",
      "\n",
      "1.PRE2\n",
      "\t North Brabant\n",
      "\t\t 1350 SAMEA5054089\n",
      "\n",
      "0.PRE1\n",
      "\t Bavaria\n",
      "\t\t -1944 SAMEA104233048\n",
      "\t\t -2272 SAMEA104233049\n",
      "\t Pärnu maakond\n",
      "\t\t -2457 SAMEA104233047\n",
      "\t Panevezys County\n",
      "\t\t -2546 SAMEA104233046\n",
      "\t Altai Krai\n",
      "\t\t -1686 SAMEA3541826\n",
      "\t Irkutsk Oblast\n",
      "\t\t -2518 SAMEA6812877\n",
      "\t\t -2666 SAMEA6812876\n",
      "\n",
      "0.PE8\n",
      "\t Samara Oblast\n",
      "\t\t -1836 SAMEA104488961\n",
      "\n",
      "1.PRE3\n",
      "\t Tatarstan\n",
      "\t\t 1381 SAMEA3937654\n",
      "\n",
      "0.PRE2\n",
      "\t Krasnoyarsk Krai\n",
      "\t\t -2776 SAMEA3541827\n",
      "Number of samples: 207\n"
     ]
    }
   ],
   "source": [
    "branch_dict = {}\n",
    "samples_df = copy.deepcopy(metadata_df)\n",
    "TIME_WINDOW = 25 # exclude if dates are within this range\n",
    "ANCIENT_DATE_THRESHOLD = 1900 # samples older than this are considered ancient\n",
    "GEO = \"province\"\n",
    "GEO_ALT = \"country\"\n",
    "\n",
    "for rec in metadata_df.iterrows():      \n",
    "    sample = rec[0]\n",
    "    branch = rec[1][\"branch_minor\"]\n",
    "    \n",
    "    # Remove the subclade letter\n",
    "    while branch[-1].isalpha():\n",
    "        branch = branch[:-1]\n",
    "        \n",
    "    if branch not in branch_dict:\n",
    "        branch_dict[branch] = {GEO : {},}\n",
    "                \n",
    "    geo_val = metadata_df[GEO][sample]\n",
    "    if geo_val == \"NA\":\n",
    "        geo_val = metadata_df[GEO_ALT][sample]\n",
    "\n",
    "    date = samples_df[\"date\"][sample].lstrip(\"[\").rstrip(\"]\")\n",
    "    # If it's a range, take the mean\n",
    "    date_split = [int(d) for d in date.split(\":\")]\n",
    "    if len(date_split) > 1:\n",
    "        date = sum(date_split) / len(date_split)\n",
    "    date = int(date)\n",
    "        \n",
    "    # Add the country if it hasn't been observed\n",
    "    if geo_val not in branch_dict[branch][GEO]:    \n",
    "        branch_dict[branch][GEO][geo_val] = {\"dates\" : {date : sample}}\n",
    "        continue\n",
    "    \n",
    "    # Keep all ancient samples\n",
    "    if date < ANCIENT_DATE_THRESHOLD:\n",
    "        branch_dict[branch][GEO][geo_val][\"dates\"][date] = sample\n",
    "    \n",
    "    else:\n",
    "        # Compare dates\n",
    "        i_date = 0\n",
    "        # By default, assume we're adding the sample\n",
    "        add_sample = True\n",
    "        for c_date in branch_dict[branch][GEO][geo_val][\"dates\"]:\n",
    "            date_diff = abs(date - c_date)\n",
    "            # If the date difference is too small, exclude\n",
    "            if date_diff < TIME_WINDOW:\n",
    "                # How to resolve ties? Want to minimize terminal branch length\n",
    "                add_sample = False\n",
    "            i_date += 1\n",
    "            \n",
    "        if add_sample:           \n",
    "            branch_dict[branch][GEO][geo_val][\"dates\"][date] = sample\n",
    "\n",
    "sample_counter = 0\n",
    "for branch in branch_dict:\n",
    "    print()\n",
    "    print(branch)\n",
    "    for geo_val in branch_dict[branch][GEO]:\n",
    "        print(\"\\t\", geo_val)\n",
    "        for date in branch_dict[branch][GEO][geo_val][\"dates\"]:\n",
    "            sample = branch_dict[branch][GEO][geo_val][\"dates\"][date]\n",
    "            sample_counter += 1\n",
    "            print(\"\\t\\t\", date, sample)\n",
    "            \n",
    "print(\"Number of samples:\", sample_counter)\n",
    "\n",
    "#display(samples_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "legitimate-advisory",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
