{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "coupled-nation",
   "metadata": {},
   "source": [
    "---\n",
    "# 0. SETUP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "racial-reconstruction",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "laughing-atlanta",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import geopandas\n",
    "import shapely\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors,lines,ticker\n",
    "import cartopy.crs as ccrs\n",
    "import scipy\n",
    "import datetime\n",
    "from Bio import AlignIO, Phylo\n",
    "import seaborn as sns\n",
    "import subprocess\n",
    "import geopy\n",
    "from geopy import distance\n",
    "import statsmodels.api as sma\n",
    "import statsmodels.formula.api as smfa\n",
    "import statsmodels.stats.multitest as smmt\n",
    "import math\n",
    "from skbio.stats import distance as skbio_dist\n",
    "from functions import *\n",
    "from mpl_toolkits.axes_grid1.inset_locator import inset_axes\n",
    "import sklearn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dimensional-worst",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "described-shipping",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"full\", \"5\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "      \n",
    "results_dir = os.path.join(project_dir, \"results/\")\n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "PRUNE = WILDCARDS[2]\n",
    "MISSING_DATA = WILDCARDS[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "foster-survivor",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/\n"
     ]
    }
   ],
   "source": [
    "# Manual Override\n",
    "results_dir = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/\"\n",
    "print(results_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cloudy-catering",
   "metadata": {},
   "outputs": [],
   "source": [
    "iqtree_dir    = os.path.join(results_dir, \"iqtree/{}/{}/{}/filter{}/\".format(READS_ORIGIN, LOCUS_NAME, PRUNE, MISSING_DATA))\n",
    "variant_qc_dir    = os.path.join(results_dir, \"variant_qc/{}/{}/\".format(READS_ORIGIN, LOCUS_NAME))\n",
    "\n",
    "metadata_path = iqtree_dir    + \"filter-taxa/metadata.tsv\"\n",
    "variant_qc_path = variant_qc_dir    + \"variant_qc.txt\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Trees\n",
    "divtree_path  = iqtree_dir    + \"filter-taxa/iqtree.treefile\"\n",
    "\n",
    "if PRUNE == \"prune\":\n",
    "    full_divtree_path = divtree_path.replace(\"prune\",\"full\")\n",
    "    full_metadata_path = metadata_path.replace(\"prune\",\"full\")\n",
    "else:\n",
    "    full_divtree_path = divtree_path\n",
    "    full_metadata_path = metadata_path\n",
    "\n",
    "# ------------------------------------------\n",
    "# Output\n",
    "out_dir = iqtree_dir.replace(\"iqtree\", \"iqtree_stats\")\n",
    "if not os.path.exists(out_dir):\n",
    "    subprocess.run([\"mkdir\", \"-p\", out_dir]) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "color-batch",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "addressed-liver",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", None)\n",
    "\n",
    "np.random.seed(1235423134)\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "ALPHA = 0.05\n",
    "\n",
    "# ------------------------------------------\n",
    "# Time\n",
    "CURRENT_YEAR = datetime.datetime.utcnow().year\n",
    "\n",
    "# ------------------------------------------\n",
    "# Geo\n",
    "CRS = \"epsg:3857\"\n",
    "CRS_EPSG = ccrs.epsg('3857')\n",
    "world_polygons = geopandas.read_file(geopandas.datasets.get_path(\"naturalearth_lowres\"))\n",
    "\n",
    "# ------------------------------------------\n",
    "# Plotting\n",
    "plt.rcParams['axes.facecolor']    ='white'\n",
    "plt.rcParams['savefig.facecolor'] ='white'\n",
    "plt.rcParams['savefig.dpi']       = 400\n",
    "plt.rcParams['svg.fonttype']      = 'none'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pleasant-objective",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. IMPORT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "assumed-elements",
   "metadata": {},
   "source": [
    "## Variant QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "every-portrait",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tstv</th>\n",
       "      <th>cds_sites</th>\n",
       "      <th>ns_sites</th>\n",
       "      <th>ss_sites</th>\n",
       "      <th>ns_ss_ratio</th>\n",
       "      <th>other_var</th>\n",
       "      <th>other_var_ratio</th>\n",
       "      <th>homo_het_sites</th>\n",
       "      <th>homo_sites</th>\n",
       "      <th>het_sites</th>\n",
       "      <th>het_ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>1.854</td>\n",
       "      <td>216</td>\n",
       "      <td>121</td>\n",
       "      <td>51</td>\n",
       "      <td>2.37255</td>\n",
       "      <td>1</td>\n",
       "      <td>0.019608</td>\n",
       "      <td>242</td>\n",
       "      <td>216</td>\n",
       "      <td>26</td>\n",
       "      <td>0.107438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>1.935</td>\n",
       "      <td>246</td>\n",
       "      <td>134</td>\n",
       "      <td>60</td>\n",
       "      <td>2.23333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>271</td>\n",
       "      <td>246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.092251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>2.119</td>\n",
       "      <td>241</td>\n",
       "      <td>133</td>\n",
       "      <td>58</td>\n",
       "      <td>2.29310</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>258</td>\n",
       "      <td>241</td>\n",
       "      <td>17</td>\n",
       "      <td>0.065891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>2.000</td>\n",
       "      <td>247</td>\n",
       "      <td>135</td>\n",
       "      <td>60</td>\n",
       "      <td>2.25000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>265</td>\n",
       "      <td>247</td>\n",
       "      <td>18</td>\n",
       "      <td>0.067924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669605.1_ASM966960v1_genomic</th>\n",
       "      <td>2.012</td>\n",
       "      <td>239</td>\n",
       "      <td>132</td>\n",
       "      <td>58</td>\n",
       "      <td>2.27586</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>260</td>\n",
       "      <td>239</td>\n",
       "      <td>21</td>\n",
       "      <td>0.080769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313243_45</th>\n",
       "      <td>1.917</td>\n",
       "      <td>96</td>\n",
       "      <td>48</td>\n",
       "      <td>28</td>\n",
       "      <td>1.71429</td>\n",
       "      <td>3</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>120</td>\n",
       "      <td>96</td>\n",
       "      <td>24</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313246_49</th>\n",
       "      <td>1.813</td>\n",
       "      <td>171</td>\n",
       "      <td>74</td>\n",
       "      <td>45</td>\n",
       "      <td>1.64444</td>\n",
       "      <td>17</td>\n",
       "      <td>0.377778</td>\n",
       "      <td>260</td>\n",
       "      <td>171</td>\n",
       "      <td>89</td>\n",
       "      <td>0.342308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6651390</th>\n",
       "      <td>2.378</td>\n",
       "      <td>138</td>\n",
       "      <td>72</td>\n",
       "      <td>40</td>\n",
       "      <td>1.80000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>218</td>\n",
       "      <td>138</td>\n",
       "      <td>80</td>\n",
       "      <td>0.366972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637004</th>\n",
       "      <td>2.143</td>\n",
       "      <td>140</td>\n",
       "      <td>75</td>\n",
       "      <td>41</td>\n",
       "      <td>1.82927</td>\n",
       "      <td>1</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>193</td>\n",
       "      <td>140</td>\n",
       "      <td>53</td>\n",
       "      <td>0.274611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637002</th>\n",
       "      <td>1.838</td>\n",
       "      <td>154</td>\n",
       "      <td>75</td>\n",
       "      <td>43</td>\n",
       "      <td>1.74419</td>\n",
       "      <td>8</td>\n",
       "      <td>0.186047</td>\n",
       "      <td>231</td>\n",
       "      <td>154</td>\n",
       "      <td>77</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>602 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      tstv  cds_sites  ns_sites  ss_sites  \\\n",
       "sample                                                                      \n",
       "GCA_009909635.1_ASM990963v1_genomic  1.854        216       121        51   \n",
       "GCA_009669545.1_ASM966954v1_genomic  1.935        246       134        60   \n",
       "GCA_009669555.1_ASM966955v1_genomic  2.119        241       133        58   \n",
       "GCA_009669565.1_ASM966956v1_genomic  2.000        247       135        60   \n",
       "GCA_009669605.1_ASM966960v1_genomic  2.012        239       132        58   \n",
       "...                                    ...        ...       ...       ...   \n",
       "SAMEA7313243_45                      1.917         96        48        28   \n",
       "SAMEA7313246_49                      1.813        171        74        45   \n",
       "SAMEA6651390                         2.378        138        72        40   \n",
       "SAMEA6637004                         2.143        140        75        41   \n",
       "SAMEA6637002                         1.838        154        75        43   \n",
       "\n",
       "                                     ns_ss_ratio  other_var  other_var_ratio  \\\n",
       "sample                                                                         \n",
       "GCA_009909635.1_ASM990963v1_genomic      2.37255          1         0.019608   \n",
       "GCA_009669545.1_ASM966954v1_genomic      2.23333          0         0.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic      2.29310          0         0.000000   \n",
       "GCA_009669565.1_ASM966956v1_genomic      2.25000          0         0.000000   \n",
       "GCA_009669605.1_ASM966960v1_genomic      2.27586          0         0.000000   \n",
       "...                                          ...        ...              ...   \n",
       "SAMEA7313243_45                          1.71429          3         0.107143   \n",
       "SAMEA7313246_49                          1.64444         17         0.377778   \n",
       "SAMEA6651390                             1.80000          1         0.025000   \n",
       "SAMEA6637004                             1.82927          1         0.024390   \n",
       "SAMEA6637002                             1.74419          8         0.186047   \n",
       "\n",
       "                                     homo_het_sites  homo_sites  het_sites  \\\n",
       "sample                                                                       \n",
       "GCA_009909635.1_ASM990963v1_genomic             242         216         26   \n",
       "GCA_009669545.1_ASM966954v1_genomic             271         246         25   \n",
       "GCA_009669555.1_ASM966955v1_genomic             258         241         17   \n",
       "GCA_009669565.1_ASM966956v1_genomic             265         247         18   \n",
       "GCA_009669605.1_ASM966960v1_genomic             260         239         21   \n",
       "...                                             ...         ...        ...   \n",
       "SAMEA7313243_45                                 120          96         24   \n",
       "SAMEA7313246_49                                 260         171         89   \n",
       "SAMEA6651390                                    218         138         80   \n",
       "SAMEA6637004                                    193         140         53   \n",
       "SAMEA6637002                                    231         154         77   \n",
       "\n",
       "                                     het_ratio  \n",
       "sample                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic   0.107438  \n",
       "GCA_009669545.1_ASM966954v1_genomic   0.092251  \n",
       "GCA_009669555.1_ASM966955v1_genomic   0.065891  \n",
       "GCA_009669565.1_ASM966956v1_genomic   0.067924  \n",
       "GCA_009669605.1_ASM966960v1_genomic   0.080769  \n",
       "...                                        ...  \n",
       "SAMEA7313243_45                       0.200000  \n",
       "SAMEA7313246_49                       0.342308  \n",
       "SAMEA6651390                          0.366972  \n",
       "SAMEA6637004                          0.274611  \n",
       "SAMEA6637002                          0.333333  \n",
       "\n",
       "[602 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "variant_qc_df = pd.read_csv(variant_qc_path, sep=' ', index_col = 0)\n",
    "variant_qc_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "display(variant_qc_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "composite-render",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "lined-timothy",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t', index_col = 0)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "full_metadata_df = pd.read_csv(full_metadata_path, sep='\\t', index_col = 0)\n",
    "full_metadata_df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "revised-production",
   "metadata": {},
   "source": [
    "## Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "satellite-lover",
   "metadata": {},
   "outputs": [],
   "source": [
    "for rec in variant_qc_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    if sample in metadata_df.index:\n",
    "        for column in variant_qc_df.columns:\n",
    "            metadata_df.at[sample, column] = variant_qc_df[column][sample]                \n",
    "                \n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "peaceful-conversation",
   "metadata": {},
   "source": [
    "## Trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "nuclear-hepatitis",
   "metadata": {},
   "outputs": [],
   "source": [
    "divtree = Phylo.read(divtree_path, \"newick\")\n",
    "divtree.ladderize(reverse=False)\n",
    "\n",
    "# Rename internal nodes      \n",
    "node_i = 0\n",
    "for c in divtree.find_clades():\n",
    "    if not c.name or \"NODE\" in c.name:\n",
    "        c.name = \"NODE{}\".format(node_i)\n",
    "        node_i += 1\n",
    "\n",
    "# Add external branch length\n",
    "for t in divtree.get_terminals():\n",
    "    sample = t.name\n",
    "    metadata_df.at[sample, \"external_branch_length\"] = t.branch_length\n",
    "    \n",
    "\n",
    "    \n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "#Phylo.draw(divtree, label_func = lambda x: metadata_df[\"external_branch_length\"][x.name] if x.is_terminal() else '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "finite-james",
   "metadata": {},
   "source": [
    "## Colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "paperback-ability",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "      <th>sequencing_technology</th>\n",
       "      <th>assembly_method</th>\n",
       "      <th>host_raw</th>\n",
       "      <th>host_order</th>\n",
       "      <th>population_color</th>\n",
       "      <th>population</th>\n",
       "      <th>geometry</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>population_rtt_dist</th>\n",
       "      <th>tstv</th>\n",
       "      <th>cds_sites</th>\n",
       "      <th>ns_sites</th>\n",
       "      <th>ss_sites</th>\n",
       "      <th>ns_ss_ratio</th>\n",
       "      <th>other_var</th>\n",
       "      <th>other_var_ratio</th>\n",
       "      <th>homo_het_sites</th>\n",
       "      <th>homo_sites</th>\n",
       "      <th>het_sites</th>\n",
       "      <th>het_ratio</th>\n",
       "      <th>external_branch_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>Human</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Human</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>POINT (-105.607716 38.7251776)</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.004600e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>IonTorrent</td>\n",
       "      <td>Newbler v. 2.6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#c5e88a</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>POINT (40.7957942 47.6222451)</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>1.854</td>\n",
       "      <td>216</td>\n",
       "      <td>121</td>\n",
       "      <td>51</td>\n",
       "      <td>2.37255</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0196078</td>\n",
       "      <td>242</td>\n",
       "      <td>216</td>\n",
       "      <td>26</td>\n",
       "      <td>0.107438</td>\n",
       "      <td>2.120100e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>1.935</td>\n",
       "      <td>246</td>\n",
       "      <td>134</td>\n",
       "      <td>60</td>\n",
       "      <td>2.23333</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>271</td>\n",
       "      <td>246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.0922509</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Frontopsylla elatoides</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>2.119</td>\n",
       "      <td>241</td>\n",
       "      <td>133</td>\n",
       "      <td>58</td>\n",
       "      <td>2.2931</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>258</td>\n",
       "      <td>241</td>\n",
       "      <td>17</td>\n",
       "      <td>0.0658915</td>\n",
       "      <td>2.356000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46334640000001 42.4804953)</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>2</td>\n",
       "      <td>247</td>\n",
       "      <td>135</td>\n",
       "      <td>60</td>\n",
       "      <td>2.25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>265</td>\n",
       "      <td>247</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0679245</td>\n",
       "      <td>4.711000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313243_45</th>\n",
       "      <td>Azov38</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313243_45</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (40.7957942 47.6222451)</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>1.917</td>\n",
       "      <td>96</td>\n",
       "      <td>48</td>\n",
       "      <td>28</td>\n",
       "      <td>1.71429</td>\n",
       "      <td>3</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>120</td>\n",
       "      <td>96</td>\n",
       "      <td>24</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1.202300e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313246_49</th>\n",
       "      <td>Gdansk8</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Poland</td>\n",
       "      <td>Pomeranian Voivodeship</td>\n",
       "      <td>52.215933</td>\n",
       "      <td>19.134422</td>\n",
       "      <td>54.2456</td>\n",
       "      <td>18.1099</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313246_49</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>54.245560</td>\n",
       "      <td>18.109900</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (18.1099 54.24556)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>1.813</td>\n",
       "      <td>171</td>\n",
       "      <td>74</td>\n",
       "      <td>45</td>\n",
       "      <td>1.64444</td>\n",
       "      <td>17</td>\n",
       "      <td>0.377778</td>\n",
       "      <td>260</td>\n",
       "      <td>171</td>\n",
       "      <td>89</td>\n",
       "      <td>0.342308</td>\n",
       "      <td>5.131000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6651390</th>\n",
       "      <td>AGU010</td>\n",
       "      <td>[1435:1477]</td>\n",
       "      <td>[-586:-544]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6651390</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1456.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (25.24953400167952 54.82269205)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>2.378</td>\n",
       "      <td>138</td>\n",
       "      <td>72</td>\n",
       "      <td>40</td>\n",
       "      <td>1.8</td>\n",
       "      <td>1</td>\n",
       "      <td>0.025</td>\n",
       "      <td>218</td>\n",
       "      <td>138</td>\n",
       "      <td>80</td>\n",
       "      <td>0.366972</td>\n",
       "      <td>2.360000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637004</th>\n",
       "      <td>AGU025</td>\n",
       "      <td>[1441:1612]</td>\n",
       "      <td>[-580:-409]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637004</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1526.5</td>\n",
       "      <td>494.5</td>\n",
       "      <td>85.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (25.24953400167952 54.82269205)</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>2.143</td>\n",
       "      <td>140</td>\n",
       "      <td>75</td>\n",
       "      <td>41</td>\n",
       "      <td>1.82927</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0243902</td>\n",
       "      <td>193</td>\n",
       "      <td>140</td>\n",
       "      <td>53</td>\n",
       "      <td>0.274611</td>\n",
       "      <td>9.423000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637002</th>\n",
       "      <td>AGU007B</td>\n",
       "      <td>[1463:1632]</td>\n",
       "      <td>[-558:-389]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637002</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1547.5</td>\n",
       "      <td>473.5</td>\n",
       "      <td>84.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (25.24953400167952 54.82269205)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>1.838</td>\n",
       "      <td>154</td>\n",
       "      <td>75</td>\n",
       "      <td>43</td>\n",
       "      <td>1.74419</td>\n",
       "      <td>8</td>\n",
       "      <td>0.186047</td>\n",
       "      <td>231</td>\n",
       "      <td>154</td>\n",
       "      <td>77</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>2.360000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      strain         date      date_bp  \\\n",
       "sample                                                                   \n",
       "Reference                               CO92         1992          -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic     9_10       1923.0          -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic    42126       2006.0          -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic    42123       2005.0          -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic    42118       2005.0          -16   \n",
       "...                                      ...          ...          ...   \n",
       "SAMEA7313243_45                       Azov38  [1400:1700]  [-621:-321]   \n",
       "SAMEA7313246_49                      Gdansk8  [1400:1700]  [-621:-321]   \n",
       "SAMEA6651390                          AGU010  [1435:1477]  [-586:-544]   \n",
       "SAMEA6637004                          AGU025  [1441:1612]  [-580:-409]   \n",
       "SAMEA6637002                         AGU007B  [1463:1632]  [-558:-389]   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "SAMEA7313243_45                                        Russia   \n",
       "SAMEA7313246_49                                        Poland   \n",
       "SAMEA6651390                                        Lithuania   \n",
       "SAMEA6637004                                        Lithuania   \n",
       "SAMEA6637002                                        Lithuania   \n",
       "\n",
       "                                                   province  country_lat  \\\n",
       "sample                                                                     \n",
       "Reference                                          Colorado    39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic           Rostov Oblast    64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic                Xinjiang    35.000074   \n",
       "...                                                     ...          ...   \n",
       "SAMEA7313243_45                               Rostov Oblast    64.686314   \n",
       "SAMEA7313246_49                      Pomeranian Voivodeship    52.215933   \n",
       "SAMEA6651390                                 Vilnius County    55.350000   \n",
       "SAMEA6637004                                 Vilnius County    55.350000   \n",
       "SAMEA6637002                                 Vilnius County    55.350000   \n",
       "\n",
       "                                     country_lon province_lat province_lon  \\\n",
       "sample                                                                       \n",
       "Reference                            -100.445882      38.7252     -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic    97.745306      47.6222      40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic   104.999927      42.4805      85.4633   \n",
       "...                                          ...          ...          ...   \n",
       "SAMEA7313243_45                        97.745306      47.6222      40.7958   \n",
       "SAMEA7313246_49                        19.134422      54.2456      18.1099   \n",
       "SAMEA6651390                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637004                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637002                           23.750000      54.8227      25.2495   \n",
       "\n",
       "                                              biovar branch_major  \\\n",
       "sample                                                              \n",
       "Reference                                 Orientalis        1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic       Medievalis        2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          Antiqua        0.ANT   \n",
       "...                                              ...          ...   \n",
       "SAMEA7313243_45                      Second Pandemic        1.PRE   \n",
       "SAMEA7313246_49                      Second Pandemic        1.PRE   \n",
       "SAMEA6651390                         Second Pandemic        1.PRE   \n",
       "SAMEA6637004                         Second Pandemic        1.PRE   \n",
       "SAMEA6637002                         Second Pandemic        1.PRE   \n",
       "\n",
       "                                    branch_minor biosample_accession  \\\n",
       "sample                                                                 \n",
       "Reference                                 1.ORI1        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.ANT1        SAMN07722923   \n",
       "...                                          ...                 ...   \n",
       "SAMEA7313243_45                           1.PRE1     SAMEA7313243_45   \n",
       "SAMEA7313246_49                           1.PRE1     SAMEA7313246_49   \n",
       "SAMEA6651390                              1.PRE1        SAMEA6651390   \n",
       "SAMEA6637004                              1.PRE1        SAMEA6637004   \n",
       "SAMEA6637002                              1.PRE1        SAMEA6637002   \n",
       "\n",
       "                                                     biosample_comment  \\\n",
       "sample                                                                   \n",
       "Reference                              KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic              KEEP: Assembly Modern   \n",
       "...                                                                ...   \n",
       "SAMEA7313243_45                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA7313246_49                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA6651390                                         KEEP: SRA Ancient   \n",
       "SAMEA6637004                                         KEEP: SRA Ancient   \n",
       "SAMEA6637002                                         KEEP: SRA Ancient   \n",
       "\n",
       "                                     branch_number      continent  date_mean  \\\n",
       "sample                                                                         \n",
       "Reference                                        1  North America     1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic              2         Europe     1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic              0           Asia     2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic              0           Asia     2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic              0           Asia     2005.0   \n",
       "...                                            ...            ...        ...   \n",
       "SAMEA7313243_45                                  1         Europe     1550.0   \n",
       "SAMEA7313246_49                                  1         Europe     1550.0   \n",
       "SAMEA6651390                                     1         Europe     1456.0   \n",
       "SAMEA6637004                                     1         Europe     1526.5   \n",
       "SAMEA6637002                                     1         Europe     1547.5   \n",
       "\n",
       "                                     date_bp_mean  date_err        lat  \\\n",
       "sample                                                                   \n",
       "Reference                                    29.0       0.0  38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic          98.0       0.0  47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic          15.0       0.0  42.480495   \n",
       "GCA_009669555.1_ASM966955v1_genomic          16.0       0.0  42.480495   \n",
       "GCA_009669565.1_ASM966956v1_genomic          16.0       0.0  42.480495   \n",
       "...                                           ...       ...        ...   \n",
       "SAMEA7313243_45                             471.0     150.0  47.622245   \n",
       "SAMEA7313246_49                             471.0     150.0  54.245560   \n",
       "SAMEA6651390                                565.0      21.0  54.822692   \n",
       "SAMEA6637004                                494.5      85.5  54.822692   \n",
       "SAMEA6637002                                473.5      84.5  54.822692   \n",
       "\n",
       "                                            lon host_human  \\\n",
       "sample                                                       \n",
       "Reference                           -105.607716      Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic   40.795794      Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic   85.463346  Non-Human   \n",
       "GCA_009669555.1_ASM966955v1_genomic   85.463346  Non-Human   \n",
       "GCA_009669565.1_ASM966956v1_genomic   85.463346  Non-Human   \n",
       "...                                         ...        ...   \n",
       "SAMEA7313243_45                       40.795794      Human   \n",
       "SAMEA7313246_49                       18.109900      Human   \n",
       "SAMEA6651390                          25.249534      Human   \n",
       "SAMEA6637004                          25.249534      Human   \n",
       "SAMEA6637002                          25.249534      Human   \n",
       "\n",
       "                                    sequencing_technology     assembly_method  \\\n",
       "sample                                                                          \n",
       "Reference                                              NA                  NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic            IonTorrent      Newbler v. 2.6   \n",
       "GCA_009669545.1_ASM966954v1_genomic   Illumina Hiseq 2000  SOAPdenovo v. 2.04   \n",
       "GCA_009669555.1_ASM966955v1_genomic   Illumina Hiseq 2000  SOAPdenovo v. 2.04   \n",
       "GCA_009669565.1_ASM966956v1_genomic   Illumina Hiseq 2000  SOAPdenovo v. 2.04   \n",
       "...                                                   ...                 ...   \n",
       "SAMEA7313243_45                               NextSeq 500                  NA   \n",
       "SAMEA7313246_49                               NextSeq 500                  NA   \n",
       "SAMEA6651390                                  NextSeq 500                  NA   \n",
       "SAMEA6637004                                  NextSeq 500                  NA   \n",
       "SAMEA6637002                          Illumina HiSeq 4000                  NA   \n",
       "\n",
       "                                                   host_raw    host_order  \\\n",
       "sample                                                                      \n",
       "Reference                                             Human         Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic            Homo sapiens         Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic      Citellus undulatus      Rodentia   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Frontopsylla elatoides  Siphonaptera   \n",
       "GCA_009669565.1_ASM966956v1_genomic      Citellus undulatus      Rodentia   \n",
       "...                                                     ...           ...   \n",
       "SAMEA7313243_45                                Homo sapiens         Human   \n",
       "SAMEA7313246_49                                Homo sapiens         Human   \n",
       "SAMEA6651390                                   Homo sapiens         Human   \n",
       "SAMEA6637004                                   Homo sapiens         Human   \n",
       "SAMEA6637002                                   Homo sapiens         Human   \n",
       "\n",
       "                                    population_color population  \\\n",
       "sample                                                            \n",
       "Reference                                    #ff0000      1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic          #c5e88a      2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          #238af5      0.ANT   \n",
       "...                                              ...        ...   \n",
       "SAMEA7313243_45                              #f3c16a      1.PRE   \n",
       "SAMEA7313246_49                              #f3c16a      1.PRE   \n",
       "SAMEA6651390                                 #f3c16a      1.PRE   \n",
       "SAMEA6637004                                 #f3c16a      1.PRE   \n",
       "SAMEA6637002                                 #f3c16a      1.PRE   \n",
       "\n",
       "                                                                  geometry  \\\n",
       "sample                                                                       \n",
       "Reference                                   POINT (-105.607716 38.7251776)   \n",
       "GCA_009909635.1_ASM990963v1_genomic          POINT (40.7957942 47.6222451)   \n",
       "GCA_009669545.1_ASM966954v1_genomic   POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669555.1_ASM966955v1_genomic   POINT (85.46334640000001 42.4804953)   \n",
       "GCA_009669565.1_ASM966956v1_genomic   POINT (85.46334640000001 42.4804953)   \n",
       "...                                                                    ...   \n",
       "SAMEA7313243_45                              POINT (40.7957942 47.6222451)   \n",
       "SAMEA7313246_49                                   POINT (18.1099 54.24556)   \n",
       "SAMEA6651390                         POINT (25.24953400167952 54.82269205)   \n",
       "SAMEA6637004                         POINT (25.24953400167952 54.82269205)   \n",
       "SAMEA6637002                         POINT (25.24953400167952 54.82269205)   \n",
       "\n",
       "                                     root_rtt_dist clade_rtt_dist  \\\n",
       "sample                                                              \n",
       "Reference                                 0.000073             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic       0.000073             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.000054             NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.000055             NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.000055             NA   \n",
       "...                                            ...            ...   \n",
       "SAMEA7313243_45                           0.000075             NA   \n",
       "SAMEA7313246_49                           0.000060             NA   \n",
       "SAMEA6651390                              0.000060             NA   \n",
       "SAMEA6637004                              0.000061             NA   \n",
       "SAMEA6637002                              0.000060             NA   \n",
       "\n",
       "                                     population_rtt_dist   tstv cds_sites  \\\n",
       "sample                                                                      \n",
       "Reference                                       0.000006     NA        NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic             0.000010  1.854       216   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0.000012  1.935       246   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0.000012  2.119       241   \n",
       "GCA_009669565.1_ASM966956v1_genomic             0.000012      2       247   \n",
       "...                                                  ...    ...       ...   \n",
       "SAMEA7313243_45                                 0.000022  1.917        96   \n",
       "SAMEA7313246_49                                 0.000007  1.813       171   \n",
       "SAMEA6651390                                    0.000006  2.378       138   \n",
       "SAMEA6637004                                    0.000007  2.143       140   \n",
       "SAMEA6637002                                    0.000006  1.838       154   \n",
       "\n",
       "                                    ns_sites ss_sites ns_ss_ratio other_var  \\\n",
       "sample                                                                        \n",
       "Reference                                 NA       NA          NA        NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      121       51     2.37255         1   \n",
       "GCA_009669545.1_ASM966954v1_genomic      134       60     2.23333         0   \n",
       "GCA_009669555.1_ASM966955v1_genomic      133       58      2.2931         0   \n",
       "GCA_009669565.1_ASM966956v1_genomic      135       60        2.25         0   \n",
       "...                                      ...      ...         ...       ...   \n",
       "SAMEA7313243_45                           48       28     1.71429         3   \n",
       "SAMEA7313246_49                           74       45     1.64444        17   \n",
       "SAMEA6651390                              72       40         1.8         1   \n",
       "SAMEA6637004                              75       41     1.82927         1   \n",
       "SAMEA6637002                              75       43     1.74419         8   \n",
       "\n",
       "                                    other_var_ratio homo_het_sites homo_sites  \\\n",
       "sample                                                                          \n",
       "Reference                                        NA             NA         NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic       0.0196078            242        216   \n",
       "GCA_009669545.1_ASM966954v1_genomic               0            271        246   \n",
       "GCA_009669555.1_ASM966955v1_genomic               0            258        241   \n",
       "GCA_009669565.1_ASM966956v1_genomic               0            265        247   \n",
       "...                                             ...            ...        ...   \n",
       "SAMEA7313243_45                            0.107143            120         96   \n",
       "SAMEA7313246_49                            0.377778            260        171   \n",
       "SAMEA6651390                                  0.025            218        138   \n",
       "SAMEA6637004                              0.0243902            193        140   \n",
       "SAMEA6637002                               0.186047            231        154   \n",
       "\n",
       "                                    het_sites  het_ratio  \\\n",
       "sample                                                     \n",
       "Reference                                  NA         NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic        26   0.107438   \n",
       "GCA_009669545.1_ASM966954v1_genomic        25  0.0922509   \n",
       "GCA_009669555.1_ASM966955v1_genomic        17  0.0658915   \n",
       "GCA_009669565.1_ASM966956v1_genomic        18  0.0679245   \n",
       "...                                       ...        ...   \n",
       "SAMEA7313243_45                            24        0.2   \n",
       "SAMEA7313246_49                            89   0.342308   \n",
       "SAMEA6651390                               80   0.366972   \n",
       "SAMEA6637004                               53   0.274611   \n",
       "SAMEA6637002                               77   0.333333   \n",
       "\n",
       "                                     external_branch_length  \n",
       "sample                                                       \n",
       "Reference                                      4.004600e-06  \n",
       "GCA_009909635.1_ASM990963v1_genomic            2.120100e-06  \n",
       "GCA_009669545.1_ASM966954v1_genomic            0.000000e+00  \n",
       "GCA_009669555.1_ASM966955v1_genomic            2.356000e-07  \n",
       "GCA_009669565.1_ASM966956v1_genomic            4.711000e-07  \n",
       "...                                                     ...  \n",
       "SAMEA7313243_45                                1.202300e-06  \n",
       "SAMEA7313246_49                                5.131000e-07  \n",
       "SAMEA6651390                                   2.360000e-08  \n",
       "SAMEA6637004                                   9.423000e-07  \n",
       "SAMEA6637002                                   2.360000e-08  \n",
       "\n",
       "[601 rows x 44 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Manual\n",
    "colors_dict = {'0.PRE': '#8000ff', '0.PE': '#5148fc', '0.ANT': '#238af5', '0.ANT4': '#0cc1e8', '3.ANT': '#3ae8d7', '4.ANT': '#68fcc1', '2.ANT': '#97fca7', '2.MED': '#c5e88a', '1.PRE': '#f3c16a', '1.ANT': '#ff8a48', '1.IN': '#ff4824', '1.ORI': '#ff0000', 'NA': '#c4c4c4'}\n",
    "\n",
    "population_list = list(colors_dict.keys())\n",
    "population_list.remove(NO_DATA_CHAR)\n",
    "population_list.reverse()\n",
    "\n",
    "# Initialize columns\n",
    "metadata_df[\"population_color\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"population\"]       = [NO_DATA_CHAR] * len(metadata_df)\n",
    "\n",
    "for t in divtree.get_terminals():       \n",
    "    # Assign population\n",
    "    branch_minor = full_metadata_df[\"branch_minor\"][t.name]\n",
    "    branch_major = full_metadata_df[\"branch_major\"][t.name]\n",
    "    population = branch_major\n",
    "    if branch_minor == \"0.ANT4\":\n",
    "        population = branch_minor\n",
    "    metadata_df.at[t.name, \"population\"] = population        \n",
    "    \n",
    "    # Assign color\n",
    "    population_color = colors_dict[population]\n",
    "    metadata_df.at[t.name, \"population_color\"] = population_color\n",
    "    \n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "touched-victoria",
   "metadata": {},
   "source": [
    "## Sampling Age Range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "powered-register",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/iqtree_stats/all/chromosome/full/filter5/population_time_frame.tsv\n",
      "population\tsamples\tmin_date\tmax_date\trange_date\n",
      "1.ORI\t117\t1924\t2016\t92\n",
      "1.IN\t39\t1954\t2008\t54\n",
      "1.ANT\t4\t1954\t2004\t50\n",
      "1.PRE\t40\t1270\t1800\t530\n",
      "2.MED\t116\t1912\t2018\t106\n",
      "2.ANT\t54\t1924\t2008\t84\n",
      "4.ANT\t11\t1977\t2015\t38\n",
      "3.ANT\t11\t1961\t2017\t56\n",
      "0.ANT4\t12\t214\t880\t666\n",
      "0.ANT\t103\t1947\t2019\t72\n",
      "0.PE\t85\t1958\t2014\t56\n",
      "0.PRE\t8\t-2876\t-1626\t1250\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ktmeaton/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n"
     ]
    }
   ],
   "source": [
    "out_path = os.path.join(out_dir, \"population_time_frame.tsv\")\n",
    "print(out_path)\n",
    "time_frame_dict = {}\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    header = \"{}\\t{}\\t{}\\t{}\\t{}\".format(\"population\", \"samples\", \"min_date\", \"max_date\", \"range_date\")\n",
    "    print(header)\n",
    "    outfile.write(header + \"\\n\")\n",
    "    \n",
    "    for population in population_list:\n",
    "        df = metadata_df[metadata_df[\"population\"] == population]\n",
    "        \n",
    "        drop_sample = \"SAMEA104488961\"\n",
    "        if drop_sample in df.index:\n",
    "            df.drop(drop_sample, inplace=True)    \n",
    "        \n",
    "        comment = df[\"biosample_comment\"][0]\n",
    "        samples = len(df)\n",
    "        min_date = 2021\n",
    "        max_date = -999999\n",
    "        for sample in df.index:\n",
    "            date_mean = df[\"date_mean\"][sample]\n",
    "            if \"Modern\" in comment:\n",
    "                sample_min = date_mean\n",
    "                sample_max = date_mean\n",
    "            if \"Ancient\" in comment:\n",
    "                date_err = df[\"date_err\"][sample]\n",
    "                sample_min = date_mean - date_err\n",
    "                sample_max = date_mean + date_err\n",
    "            if sample_min <= min_date:\n",
    "                min_date = int(sample_min)\n",
    "            if sample_max >= max_date:\n",
    "                max_date = int(sample_max)\n",
    "        range_date = max_date - min_date\n",
    "        time_frame_dict[population] = range_date\n",
    "        line = \"{}\\t{}\\t{}\\t{}\\t{}\".format(population, samples, min_date, max_date, range_date)\n",
    "        print(line) \n",
    "        outfile.write(line + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "oriental-relation",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "capable-demographic",
   "metadata": {},
   "source": [
    "## Convert Dataframe to Geopandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "tutorial-found",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "      <th>sequencing_technology</th>\n",
       "      <th>assembly_method</th>\n",
       "      <th>host_raw</th>\n",
       "      <th>host_order</th>\n",
       "      <th>population_color</th>\n",
       "      <th>population</th>\n",
       "      <th>geometry</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>population_rtt_dist</th>\n",
       "      <th>tstv</th>\n",
       "      <th>cds_sites</th>\n",
       "      <th>ns_sites</th>\n",
       "      <th>ss_sites</th>\n",
       "      <th>ns_ss_ratio</th>\n",
       "      <th>other_var</th>\n",
       "      <th>other_var_ratio</th>\n",
       "      <th>homo_het_sites</th>\n",
       "      <th>homo_sites</th>\n",
       "      <th>het_sites</th>\n",
       "      <th>het_ratio</th>\n",
       "      <th>external_branch_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>Human</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Human</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>POINT (-105.60772 38.72518)</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.004600e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>IonTorrent</td>\n",
       "      <td>Newbler v. 2.6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#c5e88a</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>POINT (40.79579 47.62225)</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>1.854</td>\n",
       "      <td>216</td>\n",
       "      <td>121</td>\n",
       "      <td>51</td>\n",
       "      <td>2.37255</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0196078</td>\n",
       "      <td>242</td>\n",
       "      <td>216</td>\n",
       "      <td>26</td>\n",
       "      <td>0.107438</td>\n",
       "      <td>2.120100e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46335 42.48050)</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>1.935</td>\n",
       "      <td>246</td>\n",
       "      <td>134</td>\n",
       "      <td>60</td>\n",
       "      <td>2.23333</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>271</td>\n",
       "      <td>246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.0922509</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Frontopsylla elatoides</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46335 42.48050)</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>2.119</td>\n",
       "      <td>241</td>\n",
       "      <td>133</td>\n",
       "      <td>58</td>\n",
       "      <td>2.2931</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>258</td>\n",
       "      <td>241</td>\n",
       "      <td>17</td>\n",
       "      <td>0.0658915</td>\n",
       "      <td>2.356000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46335 42.48050)</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>2</td>\n",
       "      <td>247</td>\n",
       "      <td>135</td>\n",
       "      <td>60</td>\n",
       "      <td>2.25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>265</td>\n",
       "      <td>247</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0679245</td>\n",
       "      <td>4.711000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313243_45</th>\n",
       "      <td>Azov38</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313243_45</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (40.79579 47.62225)</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>1.917</td>\n",
       "      <td>96</td>\n",
       "      <td>48</td>\n",
       "      <td>28</td>\n",
       "      <td>1.71429</td>\n",
       "      <td>3</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>120</td>\n",
       "      <td>96</td>\n",
       "      <td>24</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1.202300e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313246_49</th>\n",
       "      <td>Gdansk8</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Poland</td>\n",
       "      <td>Pomeranian Voivodeship</td>\n",
       "      <td>52.215933</td>\n",
       "      <td>19.134422</td>\n",
       "      <td>54.2456</td>\n",
       "      <td>18.1099</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313246_49</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>54.245560</td>\n",
       "      <td>18.109900</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (18.10990 54.24556)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>1.813</td>\n",
       "      <td>171</td>\n",
       "      <td>74</td>\n",
       "      <td>45</td>\n",
       "      <td>1.64444</td>\n",
       "      <td>17</td>\n",
       "      <td>0.377778</td>\n",
       "      <td>260</td>\n",
       "      <td>171</td>\n",
       "      <td>89</td>\n",
       "      <td>0.342308</td>\n",
       "      <td>5.131000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6651390</th>\n",
       "      <td>AGU010</td>\n",
       "      <td>[1435:1477]</td>\n",
       "      <td>[-586:-544]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6651390</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1456.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (25.24953 54.82269)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>2.378</td>\n",
       "      <td>138</td>\n",
       "      <td>72</td>\n",
       "      <td>40</td>\n",
       "      <td>1.8</td>\n",
       "      <td>1</td>\n",
       "      <td>0.025</td>\n",
       "      <td>218</td>\n",
       "      <td>138</td>\n",
       "      <td>80</td>\n",
       "      <td>0.366972</td>\n",
       "      <td>2.360000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637004</th>\n",
       "      <td>AGU025</td>\n",
       "      <td>[1441:1612]</td>\n",
       "      <td>[-580:-409]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637004</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1526.5</td>\n",
       "      <td>494.5</td>\n",
       "      <td>85.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (25.24953 54.82269)</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>2.143</td>\n",
       "      <td>140</td>\n",
       "      <td>75</td>\n",
       "      <td>41</td>\n",
       "      <td>1.82927</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0243902</td>\n",
       "      <td>193</td>\n",
       "      <td>140</td>\n",
       "      <td>53</td>\n",
       "      <td>0.274611</td>\n",
       "      <td>9.423000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637002</th>\n",
       "      <td>AGU007B</td>\n",
       "      <td>[1463:1632]</td>\n",
       "      <td>[-558:-389]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637002</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1547.5</td>\n",
       "      <td>473.5</td>\n",
       "      <td>84.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (25.24953 54.82269)</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>1.838</td>\n",
       "      <td>154</td>\n",
       "      <td>75</td>\n",
       "      <td>43</td>\n",
       "      <td>1.74419</td>\n",
       "      <td>8</td>\n",
       "      <td>0.186047</td>\n",
       "      <td>231</td>\n",
       "      <td>154</td>\n",
       "      <td>77</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>2.360000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      strain         date      date_bp  \\\n",
       "sample                                                                   \n",
       "Reference                               CO92         1992          -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic     9_10       1923.0          -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic    42126       2006.0          -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic    42123       2005.0          -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic    42118       2005.0          -16   \n",
       "...                                      ...          ...          ...   \n",
       "SAMEA7313243_45                       Azov38  [1400:1700]  [-621:-321]   \n",
       "SAMEA7313246_49                      Gdansk8  [1400:1700]  [-621:-321]   \n",
       "SAMEA6651390                          AGU010  [1435:1477]  [-586:-544]   \n",
       "SAMEA6637004                          AGU025  [1441:1612]  [-580:-409]   \n",
       "SAMEA6637002                         AGU007B  [1463:1632]  [-558:-389]   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "SAMEA7313243_45                                        Russia   \n",
       "SAMEA7313246_49                                        Poland   \n",
       "SAMEA6651390                                        Lithuania   \n",
       "SAMEA6637004                                        Lithuania   \n",
       "SAMEA6637002                                        Lithuania   \n",
       "\n",
       "                                                   province  country_lat  \\\n",
       "sample                                                                     \n",
       "Reference                                          Colorado    39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic           Rostov Oblast    64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic                Xinjiang    35.000074   \n",
       "...                                                     ...          ...   \n",
       "SAMEA7313243_45                               Rostov Oblast    64.686314   \n",
       "SAMEA7313246_49                      Pomeranian Voivodeship    52.215933   \n",
       "SAMEA6651390                                 Vilnius County    55.350000   \n",
       "SAMEA6637004                                 Vilnius County    55.350000   \n",
       "SAMEA6637002                                 Vilnius County    55.350000   \n",
       "\n",
       "                                     country_lon province_lat province_lon  \\\n",
       "sample                                                                       \n",
       "Reference                            -100.445882      38.7252     -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic    97.745306      47.6222      40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic   104.999927      42.4805      85.4633   \n",
       "...                                          ...          ...          ...   \n",
       "SAMEA7313243_45                        97.745306      47.6222      40.7958   \n",
       "SAMEA7313246_49                        19.134422      54.2456      18.1099   \n",
       "SAMEA6651390                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637004                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637002                           23.750000      54.8227      25.2495   \n",
       "\n",
       "                                              biovar branch_major  \\\n",
       "sample                                                              \n",
       "Reference                                 Orientalis        1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic       Medievalis        2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          Antiqua        0.ANT   \n",
       "...                                              ...          ...   \n",
       "SAMEA7313243_45                      Second Pandemic        1.PRE   \n",
       "SAMEA7313246_49                      Second Pandemic        1.PRE   \n",
       "SAMEA6651390                         Second Pandemic        1.PRE   \n",
       "SAMEA6637004                         Second Pandemic        1.PRE   \n",
       "SAMEA6637002                         Second Pandemic        1.PRE   \n",
       "\n",
       "                                    branch_minor biosample_accession  \\\n",
       "sample                                                                 \n",
       "Reference                                 1.ORI1        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.ANT1        SAMN07722923   \n",
       "...                                          ...                 ...   \n",
       "SAMEA7313243_45                           1.PRE1     SAMEA7313243_45   \n",
       "SAMEA7313246_49                           1.PRE1     SAMEA7313246_49   \n",
       "SAMEA6651390                              1.PRE1        SAMEA6651390   \n",
       "SAMEA6637004                              1.PRE1        SAMEA6637004   \n",
       "SAMEA6637002                              1.PRE1        SAMEA6637002   \n",
       "\n",
       "                                                     biosample_comment  \\\n",
       "sample                                                                   \n",
       "Reference                              KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic              KEEP: Assembly Modern   \n",
       "...                                                                ...   \n",
       "SAMEA7313243_45                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA7313246_49                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA6651390                                         KEEP: SRA Ancient   \n",
       "SAMEA6637004                                         KEEP: SRA Ancient   \n",
       "SAMEA6637002                                         KEEP: SRA Ancient   \n",
       "\n",
       "                                     branch_number      continent  date_mean  \\\n",
       "sample                                                                         \n",
       "Reference                                        1  North America     1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic              2         Europe     1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic              0           Asia     2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic              0           Asia     2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic              0           Asia     2005.0   \n",
       "...                                            ...            ...        ...   \n",
       "SAMEA7313243_45                                  1         Europe     1550.0   \n",
       "SAMEA7313246_49                                  1         Europe     1550.0   \n",
       "SAMEA6651390                                     1         Europe     1456.0   \n",
       "SAMEA6637004                                     1         Europe     1526.5   \n",
       "SAMEA6637002                                     1         Europe     1547.5   \n",
       "\n",
       "                                     date_bp_mean  date_err        lat  \\\n",
       "sample                                                                   \n",
       "Reference                                    29.0       0.0  38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic          98.0       0.0  47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic          15.0       0.0  42.480495   \n",
       "GCA_009669555.1_ASM966955v1_genomic          16.0       0.0  42.480495   \n",
       "GCA_009669565.1_ASM966956v1_genomic          16.0       0.0  42.480495   \n",
       "...                                           ...       ...        ...   \n",
       "SAMEA7313243_45                             471.0     150.0  47.622245   \n",
       "SAMEA7313246_49                             471.0     150.0  54.245560   \n",
       "SAMEA6651390                                565.0      21.0  54.822692   \n",
       "SAMEA6637004                                494.5      85.5  54.822692   \n",
       "SAMEA6637002                                473.5      84.5  54.822692   \n",
       "\n",
       "                                            lon host_human  \\\n",
       "sample                                                       \n",
       "Reference                           -105.607716      Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic   40.795794      Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic   85.463346  Non-Human   \n",
       "GCA_009669555.1_ASM966955v1_genomic   85.463346  Non-Human   \n",
       "GCA_009669565.1_ASM966956v1_genomic   85.463346  Non-Human   \n",
       "...                                         ...        ...   \n",
       "SAMEA7313243_45                       40.795794      Human   \n",
       "SAMEA7313246_49                       18.109900      Human   \n",
       "SAMEA6651390                          25.249534      Human   \n",
       "SAMEA6637004                          25.249534      Human   \n",
       "SAMEA6637002                          25.249534      Human   \n",
       "\n",
       "                                    sequencing_technology     assembly_method  \\\n",
       "sample                                                                          \n",
       "Reference                                              NA                  NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic            IonTorrent      Newbler v. 2.6   \n",
       "GCA_009669545.1_ASM966954v1_genomic   Illumina Hiseq 2000  SOAPdenovo v. 2.04   \n",
       "GCA_009669555.1_ASM966955v1_genomic   Illumina Hiseq 2000  SOAPdenovo v. 2.04   \n",
       "GCA_009669565.1_ASM966956v1_genomic   Illumina Hiseq 2000  SOAPdenovo v. 2.04   \n",
       "...                                                   ...                 ...   \n",
       "SAMEA7313243_45                               NextSeq 500                  NA   \n",
       "SAMEA7313246_49                               NextSeq 500                  NA   \n",
       "SAMEA6651390                                  NextSeq 500                  NA   \n",
       "SAMEA6637004                                  NextSeq 500                  NA   \n",
       "SAMEA6637002                          Illumina HiSeq 4000                  NA   \n",
       "\n",
       "                                                   host_raw    host_order  \\\n",
       "sample                                                                      \n",
       "Reference                                             Human         Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic            Homo sapiens         Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic      Citellus undulatus      Rodentia   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Frontopsylla elatoides  Siphonaptera   \n",
       "GCA_009669565.1_ASM966956v1_genomic      Citellus undulatus      Rodentia   \n",
       "...                                                     ...           ...   \n",
       "SAMEA7313243_45                                Homo sapiens         Human   \n",
       "SAMEA7313246_49                                Homo sapiens         Human   \n",
       "SAMEA6651390                                   Homo sapiens         Human   \n",
       "SAMEA6637004                                   Homo sapiens         Human   \n",
       "SAMEA6637002                                   Homo sapiens         Human   \n",
       "\n",
       "                                    population_color population  \\\n",
       "sample                                                            \n",
       "Reference                                    #ff0000      1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic          #c5e88a      2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          #238af5      0.ANT   \n",
       "...                                              ...        ...   \n",
       "SAMEA7313243_45                              #f3c16a      1.PRE   \n",
       "SAMEA7313246_49                              #f3c16a      1.PRE   \n",
       "SAMEA6651390                                 #f3c16a      1.PRE   \n",
       "SAMEA6637004                                 #f3c16a      1.PRE   \n",
       "SAMEA6637002                                 #f3c16a      1.PRE   \n",
       "\n",
       "                                                        geometry  \\\n",
       "sample                                                             \n",
       "Reference                            POINT (-105.60772 38.72518)   \n",
       "GCA_009909635.1_ASM990963v1_genomic    POINT (40.79579 47.62225)   \n",
       "GCA_009669545.1_ASM966954v1_genomic    POINT (85.46335 42.48050)   \n",
       "GCA_009669555.1_ASM966955v1_genomic    POINT (85.46335 42.48050)   \n",
       "GCA_009669565.1_ASM966956v1_genomic    POINT (85.46335 42.48050)   \n",
       "...                                                          ...   \n",
       "SAMEA7313243_45                        POINT (40.79579 47.62225)   \n",
       "SAMEA7313246_49                        POINT (18.10990 54.24556)   \n",
       "SAMEA6651390                           POINT (25.24953 54.82269)   \n",
       "SAMEA6637004                           POINT (25.24953 54.82269)   \n",
       "SAMEA6637002                           POINT (25.24953 54.82269)   \n",
       "\n",
       "                                     root_rtt_dist clade_rtt_dist  \\\n",
       "sample                                                              \n",
       "Reference                                 0.000073             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic       0.000073             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.000054             NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.000055             NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.000055             NA   \n",
       "...                                            ...            ...   \n",
       "SAMEA7313243_45                           0.000075             NA   \n",
       "SAMEA7313246_49                           0.000060             NA   \n",
       "SAMEA6651390                              0.000060             NA   \n",
       "SAMEA6637004                              0.000061             NA   \n",
       "SAMEA6637002                              0.000060             NA   \n",
       "\n",
       "                                     population_rtt_dist   tstv cds_sites  \\\n",
       "sample                                                                      \n",
       "Reference                                       0.000006     NA        NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic             0.000010  1.854       216   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0.000012  1.935       246   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0.000012  2.119       241   \n",
       "GCA_009669565.1_ASM966956v1_genomic             0.000012      2       247   \n",
       "...                                                  ...    ...       ...   \n",
       "SAMEA7313243_45                                 0.000022  1.917        96   \n",
       "SAMEA7313246_49                                 0.000007  1.813       171   \n",
       "SAMEA6651390                                    0.000006  2.378       138   \n",
       "SAMEA6637004                                    0.000007  2.143       140   \n",
       "SAMEA6637002                                    0.000006  1.838       154   \n",
       "\n",
       "                                    ns_sites ss_sites ns_ss_ratio other_var  \\\n",
       "sample                                                                        \n",
       "Reference                                 NA       NA          NA        NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      121       51     2.37255         1   \n",
       "GCA_009669545.1_ASM966954v1_genomic      134       60     2.23333         0   \n",
       "GCA_009669555.1_ASM966955v1_genomic      133       58      2.2931         0   \n",
       "GCA_009669565.1_ASM966956v1_genomic      135       60        2.25         0   \n",
       "...                                      ...      ...         ...       ...   \n",
       "SAMEA7313243_45                           48       28     1.71429         3   \n",
       "SAMEA7313246_49                           74       45     1.64444        17   \n",
       "SAMEA6651390                              72       40         1.8         1   \n",
       "SAMEA6637004                              75       41     1.82927         1   \n",
       "SAMEA6637002                              75       43     1.74419         8   \n",
       "\n",
       "                                    other_var_ratio homo_het_sites homo_sites  \\\n",
       "sample                                                                          \n",
       "Reference                                        NA             NA         NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic       0.0196078            242        216   \n",
       "GCA_009669545.1_ASM966954v1_genomic               0            271        246   \n",
       "GCA_009669555.1_ASM966955v1_genomic               0            258        241   \n",
       "GCA_009669565.1_ASM966956v1_genomic               0            265        247   \n",
       "...                                             ...            ...        ...   \n",
       "SAMEA7313243_45                            0.107143            120         96   \n",
       "SAMEA7313246_49                            0.377778            260        171   \n",
       "SAMEA6651390                                  0.025            218        138   \n",
       "SAMEA6637004                              0.0243902            193        140   \n",
       "SAMEA6637002                               0.186047            231        154   \n",
       "\n",
       "                                    het_sites  het_ratio  \\\n",
       "sample                                                     \n",
       "Reference                                  NA         NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic        26   0.107438   \n",
       "GCA_009669545.1_ASM966954v1_genomic        25  0.0922509   \n",
       "GCA_009669555.1_ASM966955v1_genomic        17  0.0658915   \n",
       "GCA_009669565.1_ASM966956v1_genomic        18  0.0679245   \n",
       "...                                       ...        ...   \n",
       "SAMEA7313243_45                            24        0.2   \n",
       "SAMEA7313246_49                            89   0.342308   \n",
       "SAMEA6651390                               80   0.366972   \n",
       "SAMEA6637004                               53   0.274611   \n",
       "SAMEA6637002                               77   0.333333   \n",
       "\n",
       "                                     external_branch_length  \n",
       "sample                                                       \n",
       "Reference                                      4.004600e-06  \n",
       "GCA_009909635.1_ASM990963v1_genomic            2.120100e-06  \n",
       "GCA_009669545.1_ASM966954v1_genomic            0.000000e+00  \n",
       "GCA_009669555.1_ASM966955v1_genomic            2.356000e-07  \n",
       "GCA_009669565.1_ASM966956v1_genomic            4.711000e-07  \n",
       "...                                                     ...  \n",
       "SAMEA7313243_45                                1.202300e-06  \n",
       "SAMEA7313246_49                                5.131000e-07  \n",
       "SAMEA6651390                                   2.360000e-08  \n",
       "SAMEA6637004                                   9.423000e-07  \n",
       "SAMEA6637002                                   2.360000e-08  \n",
       "\n",
       "[601 rows x 44 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<GeoAxesSubplot:>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjwAAAEbCAYAAADeTl6JAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAClQ0lEQVR4nOy9d5hcaXnmfZ9T4VQOXbFzjpJampEmADZr7NldL9gwgDEwxLHBZv2xmDUMjMGwa+LADDb441vby8KMB89ijMEyeMB4wdheoxkJZalb6hyqu6u6u7pyDud8f5TeV5W7qrs6v7/r0iWpwqlTVafOe58n3A8nSRIYDAaDwWAwDjP8Xu8Ag8FgMBgMxk7DBA+DwWAwGIxDDxM8DAaDwWAwDj1M8DAYDAaDwTj0MMHDYDAYDAbj0MMED4PBYDAYjEOPvNqdv/zLvyx5vd7d2hcGg8FgMBiMLXPp0qUfSpL0y+Xuqyp4vF4vLl68uDN7xWAwGAwGg9FAOI6zVrqPpbQYDAaDwWAcepjgYTAYDAaDcehhgofBYDAYDMahhwkeBoPBYDAYhx4meBgMBoPBYBx6mOBhMBgMBoNx6GGCh8FgMBgMxqGHCR4Gg8FgMBiHHiZ4GAwGg8FgHHqqOi0zGIy9RZIkSJIEURTpH47jKv4BQP9mMBgMxl2Y4GEcCDKZDBKJBFKpFDKZDLLZLBUD5I8gCJAkCZlMhj6G4zjwPE//FP+f/JHL5ZDJZFAoFJDLd/5nkU6nEYlEEI1GkUqlSoRLLBaDz+dDKpWqe9v525HL5VCr1VCpVCV/BEFg4ojBYBwZmOBh7DmiKCKdTiOVSiEUCiEYDEIURcjlchiNRgSDQXg8nl3bH5VKBZvNhp6enhJBIEkSFVv5yOVy+th0Oo1wOIxUKgWNRgONRkNF1OTkJLxe75aETCVkMhlkMhkVc8XiSRRFxONxJBIJAKBCyGg0wmq1QqVSNWxfGAwGY7/CBA9j11laWoLb7UYymaRpmkrspNDRaDTQarUQBAFarRYKhYL+3+Vy4datW0gmkwVRpUr7qlAoYDAYEIvFEI/Hy96vUCgKIkqZTKZh70Umk0EURWQyGfqZymQyqFQqKJVKqFQqGjmKxWKwWq1obW2FKIqIRqNIp9Nl/0iSBLlcTiNfJpMJer2+YfvNYDAYuwUTPIwdI5lMIhQKYW1tDZlMBpIkQa/Xo6WlBTabDWtra1hZWSkrEHYSmUwGnU4Ho9EIs9kMQRAQi8UQiUTg8XiQSqUQi8XqEiTpdBobGxtV70+n043Y/RKy2WzZzzCTySASiSASiRTczvM8QqEQXnjhhU0jTRzHQavV0j+NFGlHHVEUabRQLpcjm80iFovRNCTPH/6eEkmSEIlEEAwGkUwm6UWQwWCA2WyGTqdjaVdGw2CCh9FwJEnCzMwMlpeXS1I/gUAALpcLCoUCKpUKOp0OZrMZcrkcOp0OgUAAMpmsIBISDoe3XM9CkMlkaGlpQUtLC1QqFRKJBObn53Hr1q2GppcOAqIoIhQKFdzG8zyNQhkMBgiCAKvVCo1Gc6AWHFLgDQCrq6s0tZhOp8HzPDQaDURRRDAYhEwmg1KphFKphFqthk6ng06ngyAIO7qPoigilUohkUhgbm4O4XAYZrMZer0eCoUCSqUSPM9DrVbv6H7sJaIowuv1YmlpqeRYBACv1wsgl15uaWlBc3MzFArFbu8m45DBFS9I+Zw5c0a6ePHiLu4O47AQCoWwvLyMQCBAi4o3i3A4HA5wHIdEIoFgMFgilqqRL5JkMhnUajVdQBQKBcxmc8kJUxRFJBIJbGxsYGZmZkvv8zChUqnQ398Pi8Wy17tSEUmSsLCwgHA4jHg8jmQyCYfDAYVCgWQySVOgpEZJqVRCp9MhmUzSqF1+QTupf5LJZGhqaoLdbt/1hVWSpAMlKhuJKIqYnJzE6upq1d87z/M4deoUDAYDPZ8chQgYo344jrskSdKZcvexCM8RhoSTE4kEVCoVNBoNZDJZyeOCwSCWl5cRiUQgl8uhUqlo509TUxO9It7Y2EAmk4FCoYBer8fw8DCAXGrr0qVLVffFbDZDqVTSepFQKFSX4AGAoaGhksWadG0lEgn4/X4kEgkkEglaxJtIJOp+ncOIyWTCsWPH9v1VNMdxsFqtWF9fRywWAwCsrKyUPC6TydAUI8/z0Ol0cDqdsNvtu9KFVw9HVewAue9maGgI/f39CIfDiEajtG4umUzSdB85L9y+fZuKI0EQYDAYaGpaq9Xu9dth7HP21y+fsWtIkgSv10sjKRqNBoIglBU8BoMBcrkcsVgMkiQhlUohEolgY2MDLpcLfX19aGpqQjabxezsLFKpFNRqNQYHB2E0GiEIAu677z74fD6k02laWJt/ZU3qfeLxONbW1qoWMufD8zy0Wi30en3BFZ8kSRgbG4Pf70c2m23Y53ZYsVgs+17sEHQ6Hc6cOYNAIIDl5eWq37HNZsPIyAg4jkM6ncb8/Dytk2lvb68o8hm7i0wmg8lkgslkKnt/LBbD5OQk0uk0TCZTQSF9recKBoMJnkMGidqQVBIApFIpWCwWmM1m+rhoNIpoNIpMJgOVSlXVfya/cLUadrsdNpsNABCJRBAKhRAKhaBUKmG32+FwOMo+TxRF3Lp1C5FIBGq1GgaDAaFQqOqJjBQdd3V1lV2oOY7D8ePHkU6nEY1GEY/HkclkEI1GEYvFaBs8E0M5NjY2YLVaD0zdCMdxMJvNMJvNyGQyVNwSVCoVhoaGYDAYaARFkiSoVCoaNUgkEjRieZSjLAcBjUaDkydPAsgV6YdCIXoOi8VitAZNqVTu8Z4y9jNM8BwSYrEYVlZWsL6+jmQySW8nHUktLS0FtQKkQJNAigiz2WxBmmoz0uk0YrEY9XzR6/Xw+XyYnJxEJpMBz/MwGAxIJpP0tTUaDZqamgAA165do7UY5H2QVEUlOI6jhZ2pVKpqZIKk18LhMFwu15ErUK4Gx3Gw2+20SJmkNg/a4i+Xy3HixAmEQiH4/X6IogiLxVISLVAqlWhra9ubndxDMpkMQqEQIpEItVgQBAGdnZ0Hrg7G7XZjZmampFswHA7D7XZDoVCgr68P8Xic1mkBuWOd2FActOOb0ThY0fIhwO/3w+VyQRRF6rliNBphMBigUCgQj8cxMzMDn89HvVlIHY7FYoHBYEAqlcLExAR8Ph90Oh0GBgaQSCTAcRwsFkvBiTEQCNCTTnE7dF9fH9ra2mjtTDabRSaTgc/ng8vlooXLPT09NKJAOmsymQx1HZbJZJDL5eB5HtlslnbakHRZcQqrFkh3ElkU892X8//kj3KIxWIIhULUtO+wwfM8zGYzenp6WA3EAUSSJCQSCYTDYercnV+YTf4mv3fyx2AwwGAw7PXuQ5Ik+Hw+xGKxAv8orVZL29Xzo73bTV/JZDLo9Xo4HA44nc49FT/ZbBZerxfxeLxgZAzP8/T7YinX+mFFy4ccEtovRyQSwdjYGOLxOHXdTSQSSCaTCIfDWFpagkKhgE6no3nxSCSCy5cvAwAtDMyP+Gi1WvT19UEmk+HKlSsFaSGdTgdJkrC6ugqfz4dEIgGr1Yr29nZIkoS5uTkAwOzsLGZnZ6HRaKghHxE4pJiYnKwJPM/TNNZWrkx5nq9aJ1CNVCqFpaUlLC4u1v3c/QxxYQ6Hw0zwHABIOicQCCAYDCISiRREO9RqNYxGI3X41mg0UKlUexbJkSSJeuuEw2H4/X6Ew2Gk02lks9ldTylns1maCtvrSM/y8jJmZ2c3fZwgCBgZGYHRaMTy8jLi8TiUSiUUCgUEQYAgCFCpVEwY1QCL8BxgSCSCnDhIN0O5H3J+OiuVSlH/DwC0PidffJCCYvL/SpAOKCJYlEolRFHExsYGwuEwTWt1dXWB4zh4vV4qaBKJBDXkI1c4AApmW5GiRJK60mg0cDgcMBqNO/CJVkeSJESjUaysrJTtDDqoaDQaWmDO2L9IkoS1tTXcunWr4HYyIsRisUCj0ezR3pUSiUQwMTFBzzN7Ban3MxgM0Ov1UKvVey52CIlEAuvr6zRynX9hSdZmhUJBf5tEQBJ/qUwmQ0sEmODJUS3CwwTPASUYDOL27dslKSUSpSGhUdL94HA4tnwyrOYTQlx81Wo1DaGTE0ooFMLq6iqNLnEcRw0HbTZbwf6QFFgymYRGoykQWel0ms5/SiaTyGQyVJBVEng7zfLyMqampnb9dRuN2WxGd3c31Gr1genSOmpks1msrKxgeXm5ILVqNBoxODi4ZyJHFEXqjpzJZArSwiQ9vbCwsKd1c0qlEiaTqSQ9ZLVa6YVUfscomUlXDjLzb7OLwO1A7DIkSYLRaGQiZguwlNYhhIiLYkj+nqBUKulsKGLC19zcDLlcXiJkyP/j8Tj8fj9tXff7/ejp6UFHR0fBa7ndbkxMTBTcZrVaMTAwAAC4cuVKiceNUqlET08Prd1ZWVmB2+2mOfy2tjb09vZiaWkJXq8XsVgMqVQKcrkcPT09yGQyWFhYoKFwuVwOq9UKp9O5pVTVVmltbYUgCBgfHz9wbbFmsxlNTU2Qy+UQBAF6vX7fXPEyCslms1hdXYXb7aZiRy6Xo7u7Gy0tLdv+3kizgFKprHtbPp8PN2/e3Nbr7zSpVApra2slt+t0upKRKwS9Xo/u7u4SM9NEIoGrV68CuGt0CtwdNuxwOLbl0h2NRuH1emkqf2NjA06nk82uayAswnMA8Xq9GBsbq2iYp1QqaVeWz+dDJBKhjyVFe6QzSq/Xw2azIRgMIhAIwOl0oqOjAz/72c+ocDKbzRgdHaUnxGQyidnZWbqN4q4ws9mMwcFBjI2NIRAIgOM4dHV1wel0FpxY5+bmsLCwULDvJCJVaS5VZ2cnWltbEQ6HkUgk4PV6aZ2R1Wrd9bZUv9+Pa9eu7eprbhWdTgeTyYRwOAydTof+/v693iXGJly4cKGga5FcUDTiOJ+amsLy8jL9v1KphNPpRE9PD4BcFDOTyaCzsxMACgbo8jwPURRx8eJFZu1wB51Oh9OnT29JhPp8PiwvLyMajRZ4DOX/Xe42UvfIuMu+jvCQ7hyShmFUh1T2kzbiYqxWKwYHB7GwsIClpaWyzw8EAvT/4XC4IMdOToBKpZIKHr/fj5WVFbS2tgLIFdENDw/TsQwej4dugxQFyuVytLa2UmPAclc/wWCw7P5VG8K5srKCpaWlkpOsx+PB5OQkjEYj2traYLPZdtyCPh6Pl9RT7GcikQji8TiamprgdDr3encYVYjFYrh69WpJOkiv129b7GQyGbhcrgKxA+TOxU1NTUin03Swr9FohMvlKjvkV6PRwGAwFPgfHWW0Wm1J1HxjY4NaYwCg6T+1Wg2ZTIZMJoN0Og25XI5AILAl8dje3o7Ozs595yBezH4YobLnEZ6VlRUsLCxAEAS0trZWNKdrJNlslrZHe71ebGxs0FbI1tZWWkBLakT2K8lkEvPz80ilUtBqtfQKnhQO+/1+rK6uUnOu7YxQEAQBJ0+eLFsvIEkSLVDWarUQBAHz8/OYn58veFxLSwv6+vqoACGT1D0eT0OncJOhkIlEAjKZDN3d3bDZbDsS/SH7nUwm4fV6sbq6uqmP0F7gcDgKxi0AwD333HPkC5WJCMwv6E8mk1AqlTAajXvm20KKU2OxGNbX1+F2uwHcHa3R1tZWYKpYKy6XCwsLC2V/byqVCseOHUM2m8XExATi8Tjt3sy/SGJURq1Wo7W1FSaTCdFolF4QkRKCfEGjUqlot9p24TgOJpMJTU1NtICZNLOQP8RwU6FQUMG1W5AmmdHR0R1/rSNftEwiEfF4HKFQCOFwuOLiT0K1QC7KQWY8iaIISZKojwVpAySdTKRQj+R8twsJHSsUCvh8PsTjcTidTnqQ1qKW0+k0fD4f/aHlD++MRqMIBAKbDvQshnj3mM1mGI3GAkGY7yVx/fp1+Hy+kucbjUaMjo7S9xEKhahZYiqVKjAhrAbxy6kH0uVlMpnQ2traMAF05coVBINBcBwHQRBK6qj2Gr1ej46ODmg0GkSjUUQiEdq5stOTwfcT5Pgiv5ulpaWS+g4SESTHllqtRldXF2w22562dpNFi3RlSpIErVZb9z6VSyPnY7Vacfz4cQQCAVqvwjh8yOVyHD9+fMt1j/VGa3YzurOvU1o7zfLyMqanp2teHPMLUFOpFFZXVzd9DjGvy2azNIVD2ryVSiXUajW0Wm1JtIg4oBJhE4/HsbGxgVAohGQySethyKyg+fl5dHV1AchdqZ06dQqJRIJe/ZH2RJ1ORzui7HY75ufnq57k6iEej2N1dZX6tgwMDCCTycDr9WJ+fh6iKMJsNqOlpQUWiwUul6sg9aZUKuH3+6HVaisaoPl8PszNzdFOMzL8kbSq54eIA4EAdVnd7DtOp9MIBoMIBoNwu904efJkQ7xniOCVJAkymQx2ux2RSKSs4NsLwuEwxsbGIJPJYLFYYLPZjkQbqyRJBSnXakKadAFmMhlIkgSlUolkMkk9ilKpFEKhEJqammCz2XY18kv2rdpriqKImZkZAHcbFQRBgE6nK+i+s1qt1KS0HCQScZBStYz6yWQyWF9fp+fXeDyOdDpNhQk5hsi6JooijEYjeJ5HJpPBhQsXqNEtGelBfN5EUYQgCHQNJFH//cCeR3ii0Sh4nm/IDJ9MJkO7i8gXupf5ZUEQ6EmW4zg61VcQBPh8Pqyvr+/IpG6e50uMx/IdVxOJBBYXF0tM/chJlRykJAIik8kQi8XojyKTyeDEiRP0IJYkCYFAAH6/nz42kUhgZGQECoUCV65cKfHiaG5uht1uh8vlouKIXAGU6/4Ccikx0gGWjyRJWFhYgMvlqik8LAgCXbjIiIvtQl6XdHNcvnx5X4yxUKlUsFgsdNTCUShwFEURq6urcLlcNL1ITt6krotAorJkuChBpVLR+sJicUCccMmJnrQ+74SIJN5PkiSVdOtIkoRYLEbd0wOBAK5fv16yvxqNBhaLBVarlabBVlZWMDc3VxLhtdlsGBoawvr6OrxeL7xeb0PfD+PgolKp0NnZSeu6yIV2LajVajQ3N5d0+u4E+zrC4/f7MT09Da1WC6vVCrvdvuWr7kAggJWVFQSDwX3XKkxEwW7kwkVRRCQSKdt2aTAYMDAwAKfTSSMSpMWSLARTU1MFESGFQoH29nb09PQgm82W5H/JIMdIJAK32w2NRoPm5mZwHEct74tZW1ujP5hgMIiXvvSlkMlk1LujGLlcjubm5orvV6VSoa+vj87NWl9frxhhIamNSCRC01zbXahkMhmWlpbg8XgqtrvuFGTMBhmEqVKp6NX9fjJZ22nS6TT1q0mn01AqlbSLhZhbAqC3EUGaP3OJQC6ayp1HyMiRfJaXl+nvgAhMlUpV8JhEIoGZmRnwPA9BEMo6pOcP/yVuykBuFEsikcDa2hpNacViMSSTSfA8D7vdDp/PV3Z/yXw6l8sFpVIJq9UKs9kMtVpdIniICd6pU6f2TYSSsT9IJBJlL0RrIR6Pw+Px7IrgqcaeR3iWlpYwPT1dcJter4fT6YTdbt+SGVo6naZjAHYiglIrJJqyX+A4jrp5KhSKgpoAEonq6upCLBbD8vJyRdHIcRwMBgM9YZL5V4lEgi4iOp0O2Wy2pLOjHDzPY2BgAGq1uqyZIqGzsxPd3d0lty8sLGBubg48z+PEiRN0ESG314IgCOjo6Niyt4nH48Ht27frft52sNvt6OnpKVlYjxqSJMHtdmN6epqG2WuJruVHYMtBomFbvXiy2+0YGhqi28lkMrh48SISiQQsFguampqoiF9YWEAoFEIoFNrTNm9BEHD8+HFoNBrMz8/D5XLt2b4wDhcajQb333//jr/Ovo7wlIO0SgeDQWg0GkQiEVq/YTKZYLVaqy5KJCJhs9mwsbFRYNq1m+yl2CoHuSqs1kVUi4KXJInWwlSiniiHKIpVxYJSqYTD4ajYwdfS0gKr1QpBEGidQzweryscn0wmMTU1hdXVVRiNRnoM1Sp+djNH3UjjuYOOKIqYmpqiNVwqlaqmovdaUCgUZbdF0kJkUGclAoEAMplMQWq4vb0dU1NTEEURarUaXq8XwWCwpEV8L7DZbBgYGKCDhMsZ9jEYB5k9FzwkWpAfHSCsra1BoVAUhF2Xl5eh0WjQ19dXUH8Rj8fh8/kQCoUQDAbp0EqLxYKWlpYCd17G/kWpVOKee+6hbfQkNbOZwM2PBEajUVy5cqXqYtTe3o6mpibE43HMzs7Sx5KrbDL7q1Z2sgCYpKqISeR2anF8Ph9WVlaQTCah1WrR1tYGtVqNjY0NBINB+lmT+jeO49DS0oLe3t4Gv6vtk0wmMTY2hlAoVLfYIXVulRzLyfaLzz/k9tbWVoyOjiKTydBhlNFoFOl0GhqNBjqdDgaDASsrK/B4PDTdRS42/H7/vvKvIRYcxCyUwTiM7AvBMzIygmAwSA3rEokEHc6Yf7Jpa2ujk2Hz2+kikQimpqZKIg6s6O5goVarcfz4cYyNjaGlpQU6nW5TW3XSQUMGqWYymZIJ0uVIJBJQq9Uwm82wWq1YWFjA2toaZDIZBgcHK06fr4TBYMA999xTNSVXK4IgoK2tjXb7bcX2vxykLkWv10Mul8Pj8cDj8RRYMZRjdXWViq39ElEKBAIYHx+nF0m1ih2dToehoSHqryOKIrxeL8bHx8s+vpywJE7jKpUKdrud1uPkfzbLy8u4cOFCQZR3u8fFTpLNZusqQmUwDiJ7LnicTiftesjHYrEgEAjA6/XSE0VXV1fZ1kydTod77rkHbrcba2triMVitJUzFotBr9dDp9PRgjzG/kOj0eD48eO0mHhycpLeTuodSFskIRKJ4PLlywWLCrFb3wzShWI2m+FwONDX14fe3l5ayL0VjEYjOjs7t1XLI5fLcebMmR0Z5MlxHJqamqj1Aela2qxGZbOhiruJJElYXFwsqM3arBYnn8HBQeh0OgB3h3KSi6tyVPtsyMVYOp3G5cuXacrdbDbva8PS/U4sEoHrb/8Wercb4eZmtL/2tdDc+c4YBwsylFilUu0Ld/c9/1VWCs2TToeenp4CAZNvbldMc3NzSSdP/twXMqySsffwPA+n00n9Qux2O20vz0epVCIcDsPlcqG1tZWOqZDJZNDpdLjvvvuQzWZpjReZnh6JRLC+vo61tbWK9Vtk1pjP50MymWxIB4HZbC6bBqkVvV6/o1PL5XI5HA4HHd4ajUZLxIJcLofRaITRaITJZIJOp9vzdnYyyHZubo6mhXieh0KhqKmoWBAEdHZ2QqvVIpPJYGlpCUtLS5tGAsl5o9z5Znp6GqurqzCbzdBqtfB6vYhEImVHujBqIxaJIPDYY/it27fBAZAAfOPcOeDJJ5noOYAkEgm0tbXRsUR7zZ53adXLysoKbd0kQzLLFYySVIdcLqf+LLvdrl7PledRY2hoqETxp1IpiKKIlZUVpNNpOJ1OakoYj8cRi8XoYlxrtCEUCuHWrVsF6QS5XI6mpiZwHIeNjQ1kMhnI5XK87GUva0gUIxqNYnx8HNFotObncBwHh8OBnp6eXR+AKooiHalAbOf3QzQHAB0QW67dvxZhqdPp0NbWBrvdDp7nEQgEcOvWrZp/lyS6xSLDu8PkX/4l3v3VryL/6JMAfOU3fxMDb33rXu0WYxtwHIdjx47BarXu1usdrC6tarS0tNB/V7ry2myaOGNvUCqV6Ovrg8FgKCtSyUJPpjWLoohbt24hEomgo6ODCiQiZHmeRyQSQSwWoyaLxAk0k8lApVLBYDDg/vvvp87Hfr+fzvAym804deoUrW1p1CKv1Wpx+vRpzMzM1NR9w3Ec7rvvvrJzynYDklLeL+3tJBKbP5S2HJtFZ4aHh2G32+n3Kooibt68WdfYD1LULAgCEz27gG5lBcW/Qg6AlkXmdwWTyQSZTFZ1gHOtaLVaWls3NTWFpqamPY8UHzjBk0+5BWpxcRGzs7N7sDeMaphMJhw7dqyudA3P8+ju7i6oy5EkCbdv30ZHRwdMJhOmp6cLukoUCgUdD2AymdDR0QGz2Qy9Xg+9Xo/Ozk46mJFEjex2e6PfLnieR19fH1KpFNbX16s+VpIkxOPxPRM8+42JiQl4PJ6S24trO7p+7dcgVHBoN5lMJTYGPM/DarWW3XY5vB4PJp96CvbVVaza7Rh87DFY90EdwmEm3NwMCSiJ8ETzLnQZW4fjODoHsjhiarPZMDIyQkeU1LtdMsCaGGrudqS6Fg604CkHmVnVqOGNPM9TP5/9NBDyIGEwGHD8+PEtFXIWjxzhOK5g4q7BYCgQPPkpDuJW63A4YDabIZPJqP3/bkQ0OI7DwMAA7YiqFHFUqVR1d4UdNcrVdvzVCy/A+PnPl63t0Gg0ZSPAtUbxNlZXEX30Ufx+IpF7vaUlPPvoo8DTTzPRs4O0v/a1+Ma5c3hzfg3P0BDaHn54j/fsYNPe3o7W1taCESuXL19GOBxGU1MT2traYDab4fP56vKEksvlaGlpodve7xy4Gp5KJBIJRCIRCIKAjY2NmgoSN8NgMKC1tRW3b98uWKza29vR1dVFrd79fn/ZCeyCIFT1+Tgq3H///TsWvQiFQrh8+XLNj3/wwQfLCh2S1trJgmGCz+fDzMwMrfEpV890lCnnkL2V2g6n01lSExUMBrG0tFQQdZPJZNDr9XTECMdx+Nab34wPXbhQ8nqfPX0aL33qqe2/SUZFYpEIls6ehXZlBdGWFrQ9/DArWN4iGo0GQ0NDJQOagVxdpCRJUKvVdPZcpbpDpVIJi8UCmUwGt9sNuVyOzs5O2kCynzhUNTyVIIWFmUwGXV1daG1txdraGpRKJdLpNG1zrgXistva2orx8XFqgGe1WqmhmEwmg1arRXd3N7q7u+kYBfLlRyIR6hgdCoUgl8upXwyZxgxgy908BwVSW7NTqFSquiJ68Xi8RPBks1lcv34dmUwG99xzz463FDc1NcFsNiORSNDp74y7lBskvJXaDo/HQ6fDE4xGIwwGAxbm5nDtq1+F2uVCursbPe96Fyw2Gz1fONbWyr6ercZ0WL14PR5MPPkkHKurWHU4jnT6TKPTsQLlGpHJZDAYDEin04jH4wXmuhqNBqdPn64oSNRqNYLBIMbHx2l6y2KxwGAwIJvNwul0IhwOQ61WQ6/X0+hoZ2cnHUR90Dg0gkehUBQUNCsUCtoKl0qlMDk5Sf03kskkFRrl5u4YDAa0tbWB53k6v4kYlVWCtEkT1Go1PdGmUiksLi7C4/FAJpMhm83S1z/snVxkaOpOpJASiQSuXr1aVwTN4/FAqVTSAbWSJGF8fJyaVt64cQMnTpzYcdFD5poxSiEmfvkR03prO1QqFVpaWgrEDiESCmHskUfw6vPn76ZN/vqvYbrT+iwIApabmiDNz5e83rrTif7tv8UCvB4Poo8+io+Q9NnyMkufHXHsdjt0Oh0UCgWUSiWUSiVkMhkymUxB1oCk6AFgZmYGGxsb1JldEATE4/GCdYmQzWYxOztbkL7SarUYGBgoSE2Vu1gtdrY/SByalFY1SLqC+PgQwaFUKnH79u2CinS5XI6BgYEdKWRNpVJYW1tDMBikk5hFUUQikQDHcYe+C6TRqS1RFLG0tIT5+fm67QZaWlpobU05wTs6OsrM47ZBIBDA1NQUZDIZ7HY72tra6no+mYEXi8VyjtMaDcbe/Ga8Ml+kDA1RkZKPTqfDmTNlI9oAgB99+tP4pT/4g7LpscG3vS3n+XNHhLydiBAAz6pU0O6ACDn3wQ/i9y9dYumzLXDYImN6vR79/f1lU1CNgFi63Lp1q2CmoslkwokTJ/ZdemorHImUVjU4jqOKlPhqKJVKRKNR+Hw+cBwHq9UKm80Gnufp5OXOzs6GGiYplUq0tbWVnPxXV1dx69atbRnW7QSNdjydn5/HyMhIw/aP53l0dHRAq9Xixo0bdT23mgFlKBTCjRs3MDo6eihOALtNJpPBjRs3aNg7k8lgZmYGdrt901EhBGI8mo/p+9/H337+8zCsriLsdML0speVPR6NRmPF7SaTScSvXQMHIAzgReROghkA8qUlKBQKpFKp3KL59NP47FNPwebxYN3pxMAHP7gji6nd49nV9Nlh4TBFxuRyOXp6etDc3LwjHliSJOHGjRsIhUJlI+KJRAKzs7Po7u4+1Bd6h/edVSEej8PlcmFlZYUOUCQOzWtra+B5HoODgwXDSXcSu92OZDKJxcXFhnaYbYedcDxdW1tDMpmkzsfEzdd2p3Ziq1gsFpjN5oYOYwwGg3C5XOjq6mrYNo8KPp+P1hIkEgnMz89DJpPVHeUhpFIp3L59G9FoFO2vfz2N3oyNjZWdlVcslAiiKOLq1atY0esRAvB/AfQB+BYAJ4DJ//N/kHnVq3DizBkkk0lYnU5Y70RYGp3GymfV4YC0vAw3gL8DYAAQBOAymXb0dQ86k089RbvogJxIfHsigc8+9RT93g4CTqcTvb29O5YmymazuH37Nnw+X8XHJBIJLC8vH/rz3ZERPNlsltbSuN1u8DyPU6dOFQwhBXLiYyfSWdXgOA4dHR2wWCy4ePEiBEGg7rd75Xi7dPYs3n1H7AC5k8mbb9/GV86e3VZBYf6A17W1NaytrWF6ehrNzc3o6urakvBJpVJ1uRrXSiaTQSqV2pd+EvsZq9UKrVZLvxOO4zA0NLTlz3FjY4OerGUyGQKBAHQ6HVpbW+H3+2kkiYz1aGpqQjgYxItf/jLkMzPI9Pbiwfe+F4o7v6veN70JX/ne9/Ar6+t4AcBHkTu+JVHEM+97Hxa+/nW0d3fvWrR18LHH8KW3vQ3WTAbvIfsC4C+mp+H1eA5ctGK3OOiRMVIzUy0i2QjW19c39QJTq9UwGAyHOroDHHLBQ/KVLpcLgUAAVqsVBoMBJpMJFotl3325Wq0WLS0tcLvdBdX2e8FuOp4SZ92VlRWoVCrYbDbYbLaC+ViVUks+nw+Tk5M7Uv+0tLSE5eVltLe3o7Ozk6W3aoTnebS3t2NtbQ0+nw+SJGFsbIwam9Ur4kkRpUwmwz333INYLIZz585BqVSitbUV0WgUnZ2dtO4hHAziX//jfyyo9/n+976Hn/+Hf8jV8ikUsJ48ib/50Y/wEaBA1L8TwMc+/GE4n3uOtqjLZLIdbSywOp0YGxrC7968WbAv70gmdyxaQTpFJUlCNps9kK70JDK2G4XljYTneXR1ddHGmJ2mXBS0mMHBwZKL/8PI/lrxG0Q2m4Xb7Ybf74dcLkdzczOGh4f3zXygavT396OzsxM3b95EKBTas/3YK8fTRCIBl8sFl8tVcPuZM2cKplwvLi5iY2OjxC200ZDp3C6Xi7ZndnZ2MlfkPCRJwsbGBhYWFhCNRtHa2oquri4kk0lwHAe/3w9RFLG+vk6HwNYjHolzaywWQyQSgdlsxn333Qef14tbX/865DMz8NyJ4uiNRvzfP/5jKnaA3DH8yvPn8fTjj6PnTW+CXC5HsrMTDqCsqG+9Y2RJLjpIx9hOnj9a/f6y+9Lkdjdk+zzP03EspIkjP4K1WSq92vvfq+nmg489hmfLFJYPfPCDO/7aW0Umk6Gnpwd6vX7XxixsFlmtta7uMHAoBQ/P82hpadlyzcBe4PP5EIvFoNPpYDKZ0N7ejsnJyT0rYt5vjqf5CyRxSt5psZOPJEmIxWKIxWJIJBI4derUgRDQO008Hsfk5CS9uFAoFIhGo9TSAcilMa9cuQIAmJ2dxU/+4R8w8fjj6AiHsajX49eefx73vuQlFV+D4zgIggC/349r167l2tYlCbfe+lb8+tgYPT6/861voekLX0CgyDAQyIkH9dIS5HI5OI5D9xvegB89+yykdLpE1C+bTBjOuy2dTkOlUtFuyp2gUrQiEAggFonULCDkcnnBGBZJkiCKItLpdNVOxvzGDp7nkUwmqf8Yub/c+9/L6eaNKCzfbbGWzWYxNTUFIDcz0Ol00plxxA6F4zjE43FEo9GGDNw0mUwVmzQeeOCBI2WPcSTa0g8CpDNoeHiYFkun02m8+OKLe5be2k+Op729vWhvb6f/F0UR586d27MCb5lMhr6+PlrsflSJx+O4cOFCQUqkubkZg4ODCIVC8Hg84DiO+n18/5vfRNOf/RnuB5AF8ACAbwM4ee4c7n3JS7A0P49vv+MdMCwswNvcjIHHHoO9ubkkpVTNeVmSJPzW1762qSvz3MQE8J734J24WzfzDMdB/pWvoL23t+D1SIQknU7vSPrH6/HA/7a34V2ZzN00HICfB/BXdUwKFwRh28KMvFdinUEgpoz573+7080nb9zA1Ec+gu5YDHMaDfo/8xkMnDix5X2vByLWii/q8q0OSEq9mM0E5FZpampCX18fZmdnkc1mcfLkyW1vM5VK4dy5cwW38TwPi8WCY8eObXv7+40j35Z+EDAYDDh16hS8Xi81XlMoFOju7obb7d6RotzN2E+OpwsLC3A4HDQ8y/M8zGbzpsV4O0U2m8XGxsaRFzzEhIzUUJHoKpAr+na73XSBvH7+PLr+7M/wn4GCRf31AD73qlfBefUq/mlkBO+Lx3P3Lyzg2be8BekybcbVaszMv/M7NUUnh0ZHMf2//hc+9vGPozUQwLLJhMFPfKJE7ACgTuk7hdXphPdlL8OP/+VfIENODL4cgB61180VmzVuFdIwUQwp4M+/r9r38Nn3vx9D165hEMAEgNsnT+L3v/hF+rjJGzegeN/78Pk7z5EiETzzvvdh8k/+ZFdEz2aNGST9V65+S5KkHYn6kVlWpO4mm81uu3ZQqVRCp9PRiHhnZye6urqOZISaCZ59hFarhVKpRDabpVcVbW1tcDgc+OlPf7rHe7e3ZDIZ3Lx5E0NDQ9BoNMhkMg1tQ68XQRDKuvgeNVZWVsDzPLRaLWw2G+x2OzQaDbLZLJaWlgoeu/jxj+NJFBYJvxLAjwG0RSL420cfxe/cETvk/kptxtVqzFp1OuDJJ/GVTaKToiiie3AQvd/6FgBgVJLKppCJyFYqlchkMpAkqUDkNQqutxe/9C//sqW6OY7jqCjbyYUslUoVRJEqfQ/fv3IFb/J4CqNn167hs+9/PxU9Mx/9KJ5AadH44x/9KAa++93ctiQJ2UwG89/6FtQuFyItLQ1LO23WmCFJEjWrLYbjOCSTSQiC0NDjgOf5HampsVgsiEQiUCgUcDgcR1LsAEzw7DvKeTHI5XIYjcaClu6jSCgUwsWLF9HS0oJAILDr6Sy9Xg+j0QidTge/3w+PxwO73X5kTx4A0NHRgY6OjpLbScpvdnYWGxsbkCQJPalU2QWGB7Co0aBvaqrmNuNyNWZ/NTJCozi1RCd5ni97hU46s0hRaXFqhzym3Fia7eDS6/E/APwO7oqEbwwPo+eNbyxJJ5FaJOIcL90Ra7sxqiaZTNIxMb1velPZaJr59m0qdoC7Yub6tWt0O53RaNnvuzPPATibyWDtfe/Db+Zt/5svvgh87nPbFj2bNWbwPL9pDWU6nabjghpBT08PHA4HVldX4ff74ff7G1LH09LSglQqhZ6engM7FqIRMMFzAFhcXKRDOFOpFCKRyKGev1UNMk6iUZArY5lMVlDgmY9CoUBvby8cDgcA4Nq1a4jH4zh9+vSRFjvV8Pv9mJ6eRjqdhsVigV6vx/eUSkhFokcC8DMAA5/9LFa++lVILlfJ/as2G8Svfx16txv/sLgI89gYTZP8n9FR/Ie2NoSbm9H28MMwms2QJAk8z28qRirVYJAupmqQ358gCEin0+A4bkuLHokW/cvzz6P3S1/C65GLePHIfS7h//Af0JIXTSJRJrJ/pL2cvP5uNTkQ81C5QgHzU0/hK3/7t9CurCDe1oa2hx/GyKteVSJm3ABUAMZf8QpMANhAzvhRgZzT9YMAdAAWNBoQP/aFv/kb/EZR2umN4+P42ve+h75HHqHdY8RCQBRFZDKZmn6XbQ8/XDH1WatwJFGgRgkenU4HjuNgMpng9/vhcrlgsVi2fZ4RBAGDg4MN2ceDDBM8BwCHwwGv14uVlRXEYjGaNlhaWtoXrswHFZVKhdHRUdpiHolEsLGxQdvPyWJCTqYAsLy8jEAgAI1Gg+npaQwPD9MFlnEX4pmTX/D5ym9/G8/86q8WpDn+FMDGe96DB4aHYXjsMTz7zncWtBn/hSDAHArht772NZwF8AqgME1y/Tr+7cQJvOFONCd/wVcoFGVn1DWy/iJ/Ucz3iyJDHjeD7If3iSeoJ9BDd+57BYDf+9KXgLzaI/Je9nqGVP7nptZqC6JpkiRhAiiInqwA+CcAn7lzWwi5YvV/j7vf5fMAFgEMPfkkVCoVJEmCukgA487jVS4X/d0V+wgRf6Fy50YiMHmeB/R6KL/0JXzt29/OpT6bm9H22tfCYDLVFblbXljA5FNPwe7xYNVux+CHPrTl74JEzogYDwaD8Hq9LH3eIJjgOQCoVCo6/mJ1dRVTU1NQq9Xo6elBOp2G2+3ekY6BRkBy3fsJo9GI9vZ2aDQaeL1eenIjn7Hf78f6+nqJMVg4HMbs7CwA0Bb1cDiMZDKJ5uZm9PfvZ7uz3YWYAJKiT6VSiXte+lJMPPUUHvvIR9CTSuG2XA5lZyeGvvc9/PTCBQw+9hi0eW3G3uZmKPv78Xvf+AY45EZBfAFl0iTPPQfpN3+zoMYGyB17JPoil8tpBI/U3zQ6OpcfeZmbmMD4Bz9Ytfvo+9/8JpJ/9mcYBGAGEEGuSJnAASgtn65vhpRrZga3PvYxdAQCWDSZMPzJT5Ytym4kHMfh9smTeObaNSpOvwvgt3H3u7sAlKS8XgXgs696FUYGB2kEabO0kyiKJd8j+c5J7RW5n0wZzxdCCqUSA299K0RRpMdNPecr8l38ftF3IT39NGx1ih6NRkMFTygUglqtRnt7O21PZ9Hk7cMEzwFCJpOhpaUFNpsNKysr2NjYwPDwMFKpFG373W/kX1GRLpJsNluwGO3m0FSbzQaHw4Hl5eWComelUklncpHUVT75AzHzicfjAO52VzRy2OxBJxwO4/Lly3QhEQQBI6dPY+SHP4TX44H20Ufx9pmZgoVC+/TTdEJ4P4CL/8//Qxe7PpQ3CuwHSo4hktIiKQeyWJBjsXihJMcpSW1up/28WvdRIhbD/Mc+Bn06jSEAj6IwwvHvcFf0SADmAYjnz2P0gQfubr/GGVKumRlk3v1ufFqScq8Rj+OZd78brjJt943m97/4RTzxX/8rrl+9in4AKRR+dzKU/y47iy7ctuoHJklSTVGa/CjRVqj0XTzxhS/A9uST9DXIsVUt8pff8ZnJZNDW1kY7HhmNgcXhDyAKhQJ2ux0bGxu4cOHCjs9i2Q4cx9GZVMlkEqlUil5pcRxXIHwEQdjRcR+tra1QqVS4efNmSYeX1WqF2Wyu+FxiDEameNvtdhgMBnAcB4vFglQqhaWlpQNp0b9TaLVaDA8Po6WlJedunHflPPnUUzR1BdxdKCbzFuwfnT2LzPg4yCc6BaD405Xu3F5w1X4nlSUIAhQKBZ2jl06nkUwmIYoidR2Wy+VQKpUFTsPEdG8rx6IgCJj56EfLFuze+tCHoH/8cXwhnUYz7ood8phXITe9nbyvZwC8GbnutnxqnSE18fGP4513xA7dD0nCRNH2dgJJkvD4H/8xXvOTn2DkJz/BHAq/uwzKf5fFHWkanQ6mJ5/El3791/HHTU14Qq/HmEyGG1//OlY+/3lMfP3riG3TgJREgwRBgEwmgyAINUdTqn0XgiDQY5AInUolCGTUBEGhULAZfjsAi/AcUNRqNZ0QPjMzs9e7s2WIQMj3uyCLUSM7YCwWC5LJJAKBAK21yF+AyZgKhUJB3WWBXHEmGUxZLIgikQjC4TDsdjvOnTtHCw4ZOXiep8N4Ozo6cPPmTeoFUsuiHf3Sl/Ah5Lx6XomcL80zKKrhAZB6y1sA3L1aJ1HEclfSSqWyoHanOBVMOpDIsZE/VqKSmOU4Dk1NTTAYDEgkEuiMxcq+t55Egu67HuUjHJcATANYQs6f6F4A54t+B7XMkLp+/jxiKyv4c+TqYn7tzrY43B2dUSu1uhHnp5CKi/+F97wHz/zZn9H3/wBKv8u/6O7G5Pe+B/F//++C9FssEoH5u9+9W9s1NobnxsbwCwCasT1n5/waL0mSaMcVMZncjErfxZrDgb688wtJp5Y7P3R0dJRMKU+lUrh9+zaUSuW+vqA9aDDBc4Ah7dH7rUZmu5CrINIB04j6pI2NDVrjAaCkyDiZTMLtdtOws81mg1arxdWrVzEyMkJrUvLR6XR0vtfLXvYyJnaqoFKpcPLkSVy9ehXRaHTTRVsmk6EfgAE5ofNjACYA/wzgvwIYQC6yk3rLW/C2976XLrAkukPGIwB3Iz5kcdvseyJmbyS9NTw8DJPJhPn5eTrPK5vNwmg0YmBggNZd+P1+zM3NYU6thhSJlLy3RdwVOSGgbG1KFsB/LrptVqmkXUvA5jOkrp8/n4skoVAYAsA9KB2dUY1aR0eQCF6lz/aVb3wjvg/g9/7sz9CP3Hd3HMCPACwYDFj7+Z+H6fnnqU9TfvrN9ad/WpI2eguAPwfwHhSaBebzo7NnEf7Sl2hHn/53fxcPlUmFkSJ2cq6p53xT6zwvYitQjEajoSNY8iGeVoft3L7XsNESB5ilpSVMT0/v9W7sKI2K9hR7puS3oZP6IlEUIZPJoNPpaPpDrVaju7ubjvtgbI+ZmRm4XC5a7PnaRALnkavp+JlMhpY//3P4vV7Mf+xj6E2nMQ/g1ciNWAByC8rvAXjNT34CQRDQ0dGBcDhMU5T5CwTP8wURn1rHLpB0WDabRW9vL5x3ik/JINnl5WWkUimYTCZ0dHTQ+ozbt28jFArRGp53Ii96AeAccos0h7tdS29BoShZAWi3FrnNdfo0HCdPFkRW1j0eTH/hC7CtrmLd4UD/Bz5AC5b//j/+RzxZpv3/DwD0VRidUYlaRkeQOrxaf6PxaBRLZ89Cs7xMTSHP/fZv45NFRoASgI+1tKCL4/CuMjWKzyH3+QHAc//pP6H1Qx+i9/3o7Fn0fulLJdHAmTzRQ8ZG1FrvUwmvx4PJCvO8SGSw3HBWs9mMY8eOlU2dxmIxNqB4i7DREoeUvRosupuQk0QttvkkF5/fpQOA1nDkk81m4XQ60d/fj0QigcXFRaytrVHRY7PZYDaboVarWeSmgZDvwep0Iv7lL+Ov3vc+vPtOCugV2Sz+28c+hm63u2x04ufu/Fv7u78LIJembG1tRTAYRGdnJ8bHxwGAdmqRFnFSlErSVaSYuRJk6GZHRwcVO0DueGptbYXP50M2m4VWq6V/e71exO4Y5g2cOIEf/e7v4vEvfQldADzIRXTaAfwPAK13blMA+CAAJ4BlALePH8dH/+iP8KFXvxrdiQRmkavhefTSJUiXLtHIitFshs3ppEWxfchFQ0mdSCWDx2agLrEDVHcjJq7wpEavVtRaLfoeeYRG0TiOQ1sgUPZ1WgMBrA4NlY0GLgH4FHKf38S//RvE17+evrdontgh23oncm3+T/zLv2Dw6tWKIy9qoVL3G0kpkohiNpsFz/MlYod0dZazs0gkErh48SKGhoZgt9vr2i9GdZjgOcCQ1s2jwGY5dZKu2iwELAgCrFYrHA4HDAYDJEnCz372M3Ach87OTrS2th5pJ9KdJv+Yjb/wAhU7QG5RSrjd5RcqAN8BYPjAB/DQr/wKgNwk8UQiQWscBgYGcPnyZRptIKmJ/JqsZDIJnudp6jIfYuBHzBLzh9US5HI5jh07RgueCa2trWhubobP58PMzAzSX/sangDwdQDdAN6BXNv5d5E76ZIWbQl3W7a/ffMm/vprX8MHbt3CP37qU3gyL7LC4W7qRlPGQZoUWqtUKkwrFGWnwM8rlfiVO4KguC4nOTAA16c+VdJGT9rCI8gVVMsBpAFMZrPoubOgF18QvP4Vr8BbASoo/hLAt3/yk4LHkIgq+Y4WjUZIRbVPZHL9qTJpo68hl+58D/kcw+GCDrR+lK+R6gUwevVq1ZEXm1FL9xup+yp3odbT04P29vaKF1LBYBCiKGJ8fBwejwdOpxPRaBQdHR3bnqt11GGC54AiSdKRGjVRfHIgCw45mWx2hSmTyTA6Oko7q/K3e+zYMWi1WhZC3iLRaJTWt6jVapoiVKlUdBYRERLkmI1FIgj967/in3HXZVcPoAsVWs85Dr8jiggGg7hy5QpikQgmvv51zLjdUJ08iYcefxx6oxGnTp3C0tISwuEwEokEjSCQiA9ZZMnCQf7OT2uoVKqq9gKVumd4nqfdfpfuLN4TAF6HXCRHi1xq7iEUCrpXI1ej9E4A1//qr9D1jW/AsrFRMbJSCeJ51ffEE3jmAx8oSee0f+ITAErrcogJIGmjvxSJ4H+/731I8zxmZDJ8mOPwcknCq/K25/rHf8Slhx7C8fvuK9iH17/iFXgKhcXIp+7cXix68hn+5CfxzLvfTbvKyOT6wU98IpceyvNnWrXZ4J2cxB8XieV3ShI+9vGPo/2550qMD3Hn/zMA/gilgjp/5MVmTHz84/hkme438tr55KfSlUplgWN7MZlMBhsbG1hdXaW3+Xw++Hw+yOVy2O12aLXamveTUQoTPAeUaDR6ZCI8xfUBPM9DoVDUVdDX2dlZsduBuZhuHUmS6LwsIHdcVnusQqFAIh5H6MMfxgfuzM4iU9NfDtD25eKFak4Q4Ha74XQ6oVGpsPKf//PdQtof/ADf+9GP8Iof/QhGoxFGoxGRSARutxs+nw+JRAIajQaJRIIOeyTRnXKdXN3d3dtaWGQyGZYMBkh+P6IArgB4CrmCa6C8oCO+NCQlorv3Xkhnz5Y13MuPGpBoZH4xdtfAAC61tOAPVlbQhlzqR9XSggdPnQJQOiU83wTwMoDrd/aXE0VIoojPA1TskP39DQCP/cEf4PgPf1jwXt6OUkPBd97ZZjXae3vh+spX8LGPfxxtgQCWTCYc/9zn0NnbC0mS4GxrQ8uf/AkAYFgUcfM//IeKKTBJkmB8//vxzBe/WCL6MqggqDfZv3za/P6Krw2AtqKTwc9ra2vgOA5Wq7WqI/vs7CxWVlbo89PpNDQaDU6cOAGVSsVS6w2A+fAcULRabUGo/jBDiv7ImAdRFOvuXgiHwwiHwzu0h0cXjuOg1+uph1K1dCDxZJr95jfxxvHxgkXxlcilTKx9fXgGdz1ayEI1+qUvIRKJgOM4BH/8Y2pER57/qxcv4h8+9Smsr68DyHXQ9ff345577oHZbEY8HqeFyMXpKODuTCpBEBAKhcruvyiKCAQCcLlcWFhYwMLCQsVj6teef57WHv3mnX3MIJcOKus/c+fvVQCf6u2F7aGH8P0HHqCPnQDwuEKB7HPP4YePPIKV+Xkq3sjQUMLS2bP4g5UVfBq5jq9PA/iDlRUsfuc7kCSppC4n3wTw2ygVLPehvEjoKRNVrZRKqkVQtPf24qHnnsPQ88/joeeeg7OtjXp3kToh8m+X2Vz2c1w2mcBxHH7xNa/BzO/+Ln4Pubqp30OuYJlHZS+nWlk0maq+tt1ux8jICJxOJziOg8PhgN1u33T8DLmo6+/vx0te8hIMDQ1hdHSU1RE2EBbhOaCQzojNFv5aPTT2M6QItZ4fPc/z0Ol0EAQBZrOZmoAxGo9Wq6XHYX4KqRhyLMp++MOyi+Lqvffid/7pn3D1wgV8+NWvRlcyiXlBwK9+97u496UvpdEY+R135uLnp8bGMDk5iVQqBYfDQUWy2WyGz+dDJpOhreuk+6/4uEomk1hdXQXP87BYLFRsRyIRzM/Pl/ze5ubmoNfr4XA40NTUBLVajWQyiWP33ovIP/4jYnmRiAcB/ANyjsr56aHvAkgC+BZyqS/d7Cz+4ud+DqP/9m94/gc/gO/GDWTPnsUT6TS4dBrSygqe+Y3fwPRXvgIAtHjWZTbjxOc+B32FQmOVywVBELBuNBZE0YgJIIdcYXXxc7MoH3UrbpcHUDGVNAWUPHY7DH3iExVTYISHHn6YziEjr/3Zf/7ngpEXRFCPnzyJ19T42tXSbzzPo6enZ0sCpbu7m4obnucLCuYZjYEJngOKKIq0K6QStXpoVHv+fhFL9ZxAlEoldVU2GAxQq9U7uGeM/KhOsaMyIf9Y/BHKL4rO170OeqMRP//v/z1+/s7IDiAnQi5cuACVSoUzZ84g09tb9vmx1lak02lMT0/D7XbDZrOhvb2dXjkTU7n8YuWRkRFotVrE43GMj4+jq6sLJpMJ2WwW169fr8mTpWL0UKHAZN571QP4ZQBfBfA/kavpMQH4ReSmhD+Pu2Ml3iGK+Mzb3ob3nD+PP7nnHvx3FEZd3ilJeO+HPoQzfn9h8ezb347l17624vyplcVFCN/5ToHoyjcBXETpd/MAckNe/zMKRUJHnrgg/CVyNTvvLHrss0DNgqIW8lNgrYEAlk0mDH7iE5t2oP3+F7+Iz77//bh+7Rr1A/qZyYTfaWnBxNe/XnCO43ke7e3tsFqtuHHjxt2Zez09WH32WXzswx8ueO2ugQEMDAxsORqzsrKC5eVl+Hw+NDU1sajODsAEzwGF4zg0NzdXnaFVnKvP7/QoNukqZrtiaS9Rq9VlzbwYOwOp3wFKnYsJ+cfig7jrnkxreB54AC9/73vLPlcQBOrEzXEcHnzve/H9730Przx/vuD5v/qHf4iZuTk0NTVhaGgIfr8fmUwGHR0diEQiCAaDNPI0MDCAVCoFlUpFo3+Dg4O4desW5ufnG/bZcI8+iv/19NN41533qkNuUOg7yjw2v2qIA2BfX4cgCGipVMDs85XWy0gSPnruHL45MkLThvnzp67+9/+O30+lEEGuUFqGXITnak8PPrK4CG0mU+KA/G0ASgAfQK4FflapRMcnPlEw34vw7Z/8BK9/xStwHaCC4lmUdmltlQv//M9wf/KT6BdFTPE8mj/2MQz/wi/UbKQIgHZjxSIRmB57DH90+za4H/yg4BznbGnByMgIVCoVlpaW6LFiMBhoO73zToHyCMehv78fVqt1W+MgiMAh9ZnsQq3xMMFzQOE4Dk6ns6rgqeahsRnbEUt7Aem2cTqdZduJGTtDPB6Hy+UCcLdGpxz5x6Ied92TV/v74XjHO/Dy974X+ioW+vmjPfRGI17+wx/ix1/+MmSzs8j29NDnczIZbt26RVNMhL6+Ply+fJn+3+Vy4d577y24io5GoxX3f6s8/Pa340uTkxj76U/RBiBx508ll+X8/6/ZbNDpdFixWMq2bHegNP3EAWgPhWD4ylfw1e9+F9qVFYSdTrQ9/DA0Oh0d6aFHrluMsJBKgfuN38Bv/M//iSvImRSSgudYczOGX/lKPHRnG5ulpvLFzQgaF9m58M//DNsf/iEeu/M+JVHEM3/4h7gA4P5f+IW6t1fpHPftf/1XjH7mM7QLNJ1OY2BgAM3NzfR4sdvtuHans0uhUMBsNm979hWx3SCCitF4mOA5wOh0OrS1tWFpaangdmKKZV9fxz8CeAkKJzAXD+gru+1tiKW9wGw2w+v1Qq/XU6t/xs4jl8uh0+kQiUSqmvkRP5d80fNLAH78jnfgoY9+dNPXOXHiRMH/9UZjyfMymQzC4TBGRkZKnLG1Wi1GRkZw48YNALk5aEtLS2hra6OL2E6lEH73U5/C//2Hf8DS5z6HDgAnkKvbeTXuRlGeBfDaO4+XAPwFz+MNZ88CAB55/nk8c/JkSc3IosUCyest610zrNOh75FHAABdggBRFJFKpaqO9Dj5q7+Kb/zrv+LNt2/jXtyNDJn2SVTX88lPUrED3O0A+8AnPwlXe3tZI8BqVDrHWTY2aFE7x3Ho7u4ueIwkSQUXmtlsFteuXas4gqYWYrEYVldXqU0GGxy6MzDBc4DhOA69vb3IZrNwu90AyphiIVcb8O+QC6eT0PZmFC9QQO1iaS9QKBTo7++v6HHB2BkUCgWGhoZw8eJFarRWLq3V/trX4hvnztHuqs3SWLUQi8Vw9epV8DxPW3gHBgbofLNiLBYLNBoNYrEYnauV78ljtVrhdrs3rY3bCj//y7+M2M/9HG79l/8C7/w8+gF8CDlTwhkALwPw/ykUMAkCfA4H3nD2LAaOHwcA9I2M4AJp2Q4GsWQ0YvhTn8K9CgWeefvbqxbuAqCpQIVCUXH20+Bjj0Gt1QJPPomvnD0L7coKEm1taH71q/eF2AGAPlEsK1C6RbGsEeDfmM04Ho9XFECVznFrJhOAXNeU2+2mx01+l5XFYoHBYMDq6iqdrXbr1i2cPn267KiIani9XkxMTADIpePZGJudg83SOgQkEglcuHABoijiR295S9mZNH+oVqPlkUdoaHszSA1P/gK1n672itFqtRgdHWWdWHtAOBzG3NwcVCoVJEmCx+MpG+1RCwJC//RPkN9JQz1wJw0lSRL8Gxt44f/9fyGfm4PY24uht74VInLhfYPBgFAoBK/XC5vNBr1eTxeFdDoNr9cLr9cLmUyGkZHqCZeVlRVMTk6io6MD3d3dJVGdUChUkPqqBulGI51eZD5bNpul87XI49LpNDiOQyIex/c/9zmc+PGPSwp7w088gVMveQm6urpgs9nouIhYLIaJiQkkEomSz9U1M4OJPCFUrnA3f+yETCaDb30d45/9LGweD9YcDgx88IOwNTfTVEoqlSow9qw2FLSRPP6Od+CBxUXq0Hy+owNP/MVf0Pv/7pd+CX9UJHrIbLV8M0FyOxkuSoRg8ViNaue45tZWmuIkHVMDAwMF+ytJEtxuNyYnJ+ltPT096Ojo2PS9er1epNNpBAKBAqNBo9GIe+65Z9PnMypTbZYWEzyHhIWFBczNzeH2K1+J9+R1uBD+VKPB8PPPF9zm9Xgw8eSTcKyuYtXhwOBjj9GhdwDwrf/1vyB/7jl6Asq85S14w7vetcPvZOscO3aMmQjuA+bm5uByuWghJxEEcrm8oDBzeXkZfr8f/o0N+D7wgYrimnRXEe65556yJpJkEOxmBINBpFKpiscKmZq9srKCYDCIvr4+KBQKLC8vw+v1QqFQwG63w+l0YmNjA5OTk7T2p3iOmyAIJV1rP3j1q/FEOFyyQD+u1+M/ffe7m+5/OSRJosXd+SNYyOexWbcZETiZTKbkscWDd3eCx9/xDvz24mKJCPzzPNFDaniKH7PI8/hvZd5f/nBRMoj0oSIn5FgkgqU7ES0yyDT/gq6zsxNdXV1lBZ8kSTh//nyBAazT6cTQ0FDV95rNZnHhwoWy3Yz9/f1VXb4Zm8OGhx4BOjo6EI/H8Y8mE6R4vHxeP+82Mq36IyS0vbyMZx99FHj6aVidTpx99lmcee65wpPLc8/hrFKJh9/+9t17Y3Ww0ydlRm1YLBbE43FoNBpIkkQX03Q6DZfLhVAoVNAxtfid71QtkBdFscAvJxgMlhU8tYgdAPS5gUCA1koolUqaiiBt9sW1G+TqPX84qclkKoi61BIJ6YxGy6ZmOreRSiMpOpVKhWw2S/ePRJZqoVIkJ5PJ0MhpvYaftfKSPLED5Dk0Ly7Sx9z/C7+AC8jV7PSJIqZ5Hs6PfQzJ//k/IbndJee8fKMADnedkPPR6HRlmzDUajWGh4er1uRwHAebzUaL9gGUNbQMBoOIxWKIx+OQy+WIx+MVP0c2x29nYYLnkMBxHAYHBzH6+c/XlNeffOop/P4dsQPkTghvTyTw2aeegvWpp8A9/XT5IY5PPw3sM8FDTjzNzc17vStHHkmSMD09jVAohLW1NXo7abGNl4k+blYgXzw4tlGLgk6nQywWw8bGBgRBgN1ur1p/kUgkMDY2Bo1Gg8HBQQC5q/Vy4ykI5e6b02ggRSIlC/SCRrMtcz4S3SGvWY9ZJ5l5Rsif9k3cqUkUKZFINDzFVatD8/2/8AvAnY4s8lm52ttLjACfQ64onFDuoq8a7e3tNRUgk+iO0WiE1WotMAvMZrOYmJgo+B0UI5PJqElqc3Mzm46+wzDBc4jgOA6vef3rcb2zE//tkUfQ4vNhqYIhF2lPLXg+AJvHA2B7FvG7idVqxdDQUN2FgozGI0kSFhcXy45miMfjFTtPqhXI8zxfsBA3UtjK5XIYDIaaO2tUKlVJK/v6Ha+c4it2Mu+tnDjo/8xn8Mz73leSmun99Ke383agVCrriugQ8ofwEsj8OrKt/NljZKRFI9mOQ3OxCeG8Vgv7+jrIUVLpoq8ak5OTiEQiJXU7xXR1daGrqwsajabkuJiamoIkSbDb7VAqlchms1Cr1ZidnYVer8exY8dYR+kuw1aJQ4ZarcYDP//zeMDlQjgcxtTUVNkFqFp7aj9qPwHtlRuzWq3GyZMn2QljHyBJErxeLxYXF6vOK6s0IqRcB1d+N2F+lGRjYwO3b9/G4ODgnjjR5r9mOp3G8vJygaiRyWTUbbpSimjgxAnM/emf4sOPPYaueBwLGg16P/1pDBS13m+FekewkChOseDJ30bxb7z3TW+CvMGpl/MdHXimTA3PCx0dNfn4tPf20knlw8gVc9frwlyM2+1GZ2dn1UaI4iGzyWQSbrcbi4uL6OzshNVqxaVLl2CxWADkanx4nkfznSJxxu7CipYPOdlsFi+++CKtpSAhalLDU9yeqs2r4TmZl9YiJ6Brjz5Ka3j2qpPLZrOht7eXiZ19wvT0dIkXVCXKRUOA6sWjpJg2H61WC71eTwd+WiyWTY+HQCCAQCAAQRBgNBqh0WhqfIflmZubg8fjodEOpVKJRCIBuVxeNc1FUCqVEEWxYWaH9RYXk5lNlfZVLpcjFAiU/Ma/OTICw+c+1/Df+OPveAdesrhIHZpfKOrS2gusViuOHTtWVx3UxYsXkU6nwfM8zpw5A5/Ph7a2th3eUwaBdWkdcaampgqMspRKJTiOg9vlwu3Pfx42jwfrTicGPvjBgi6ts88+C+7pp+kJSMoTOwAw+Zd/iXd/9aslUaCv/OZv7pgbM+ti2F+kUimcO3eu6mMkSaJihLRH10t+ikWpVJbtJrJYLDh+/HjZxWlubg4LCwsF23M6nTAajTAYDHUbvUmShKtXryIWi9H6IkmS6GBSURSRTqermjGS55BU1HYhHkj11O1UE1sKhQJjTz+967/xzdjtqHKtreaERCKBF198saaOLUbjYV1aR5z29na43W6IooihoSHaYmu22fDyL36xbGg1kUjgNW97G1TvfjeSyWTZPPpuuzE7HA4mdvYZxPCyGpUGitYDmXIuk8kqbmtjYwNzc3OwWq3Q6/UFC39xWpd0jLlcLuh0OtoybzQa0dTUBL1eD1EUsbS0hHA4DKVSiba2NtrNtbq6ilgsRsWbQqFAJpOhf4CcAFGpVEgmkxWFD4myNAIiLElbfTkUCgU1atxMeKbT6Yb/xt/yilfg1wBqdfE3AJ6rY87WXsz4m5+fr0vwZLNZaDQaNuJmH8IEzxFApVKhvb0dCwsLiMfjcDqdUKlU8Pl8BSfofMhikUwmqSAqPkE2yo05vzC1uEg1H71eX/Z2xt5RLYJBFtdGdfVUOzYIi4uLWFxchEKhgMViQVtbG3ieL9sdRkilUohEIgAAv9+P+fl5WmSaf8y73W5IkgS5XE5TQWR/SEt4vhiTJAmJRKKgLb8cpIamEZ8RMQ0sh1wuLyhErgbpxqr0G4/fEX/1mBK+5RWvwKdQOJj01J3baxU9ezHjr7fO2h+tVov77ruPTTvfhzDBc0To6OiAx+PBxsYGOjo6oNVqYbFYCiZdVyKbzdIFAAB1lW17+OGqxaYEnuehVqsRj8dLFqwzZ85Ap9Mhm83SE8Ty8jJWVlbgdDrh9XppIWwwGGS58H1GuVRQfuEu0LgZVcWmetVIp9PweDzweDzUEbkS5QR/uVoYInJI8bVaraZiThCEivsmiiKy2WyJIAJAU2BtbW0wm82IRqOYnp7eVNiVg6THKtXxyGSymuqFFAoFFTKVCspbXv1qZDIZKBQKen7wb2xUTTX9OlDea6eO97gXM/68Xi/sdntddghM7OxPmOA5IshkMpw6dQqTk5O4fv06Tpw4gcHBQbzwwgub1hkAuUgPMTXjeR6pVAoGkwmyL3wBX/nOd8oWm3Ich56eHrS0tEAmk9GTeDgchsvlKlgs89Nq7e3tNBzc0tKCK1euIBaLwev1IpVKscF6+wSfz4fp6WkAhSJnM2+arUKiF7Ucr/ls9ngS+ahlO0ajEcFgEBzHQa1W09ESm6XsOI5DMpmk0aH8bcZiMbS2tkKhUMBoNCKZTBbUG9VKcacYed/k+6glrVg8C02j0xXM18r/jRNhBAD+jQ0EP/Qh/NatWxVTTdWsLkhtFol0VdrXvZjx5/f76z7mGPsTJniOEGq1GidOnMCVK1cwPj6O48ePo6+vD1NTU5s+t9xJKJvNQm80ov8tbym5orFarejo6CjwOCEneqPRWNYptxwKhQIdHR24ffs2JEnC3NwcNX1j7A2SJGFqagoreVfVWy1GrofNIjXb2W4tiKJIC/51Oh1SqVTdtUnlIiwqlQpra2u0Pq2jowNra2tV03CVSKVSNNq0lSgRKQ7Pp5IbMSnSTqfTuVTTHbEDlE81VbO6OIZCF+dKhembWRhslWpjdrRaLbvIOiQwwXPE2NjYQDweRzgcxvr6OlpbW6FWqzE3N1fVQ6US6XS6JFTf19fX0NRT/gmYjCRgIePdJ5FIUEO2/O9kt76LWmp4dhqfz4fe3l6o1WrwPI+bN29uW+glk0ksLS3RtAkZgjo/Pw+fz1e3yEsmkxXb/4shhc4kUlVPW3sqlaLuy4YKRqZ6jweCIIDneXwbuZqdd6LQ6uKvAbym6D2mUima7svfp2oRp61SbczO8VOnNjUfZBwcmOA5YqjVauh0OgQCAUxOTkKj0aCpqQkcx+HGjRtbWlCKvUca3UnlcDiwsLCAbDaLWCxGZymRq1hBEJgA2mEkScLt27cRuDOPiHQ1EX+nammIRrEfBE82m0U8HqeCfnBwEOPj49vebjwex6VLl3Dy5Emo1Wro9XqcOHECkUgEY2NjNNqjVqthNpsRDAYRjUYrbi+dTm/adg6gbF1RPZAZXkGHo2z0Jux0UgH2rXPn8GsvfSmuA9Tq4q9R2qVVPLeLRJEIlSJOW6XSmJ0nvvAFPPzCCw3romPsPUzwHDF0Oh1OnjxJXZivXLmCl7zkJTCbzTh9+jSuXr1atycIKWKWJAmiKEIUxYa6iAqCgO7ublovkkgkYDAYsLq6CqPRiI2NDdauvoNkMhlMT09TsVNpKCWZC9QIT5ly7BdR63a70dHRQedvbWxsYHV1ddvbTSQSuHr1Kl7ykpfQKKZOp8P999+PcDiMVCoFi8WCRCIBm82Ga9euld2OQqEAx3E1fQ/ke9xuqrBcqumvRkbQ88Y3QhAE2pTwl//0TwBy3+UIUOKiTDq/KvGFxx9Hz/nztK199oEH8IEnntjWvlcas9Pi87GRNYcM9m0eQTiOg8FgwL333ouf/vSnWF1dRVtbGzQaDfXxqBcSdlYoFIjFYg1vIW9tbUU4HMbq6io0Gg14nqeFzdt1zGVUZ2FhAZ47M9YAlBTekqt80sGkUqlouuMwIooiZmZmMDKSc6fq6elBMBikgyS3QzabRTgcxsTEBM6cyXmnkd8rkOsYCoVC6O7uRkdHBxKJBCKRCGJ3Jq2Ti45yabZyETISId1udK5Sqkl+p7CZRGmqzeEqJ5bzI8dfePxxvO78+cKU2Pnz+MLjj5eInnrMCSuN2Yn39Gzps2DsX5jgOcJwHAeFQoG5uTk64O7ee+/F8vIyjabUiyAI0O2AARjHcejv76emckBOZK2urjKDrx0mGAzSf5NOmkqimERhyFV9I6M9xHxwpyJI9bC2tgan04mmpiYIgoC2trYt/2byyWQyFZsIVlZWMDk5CYfDQTsgCSTdG4vFaMcTibLKZDIolUrwPI9bt25hfX29YLukBX27ArVaqon4F5HaHDKNPb/YPb/rC7hrA0D2qy9P7AB5be3nzxe8Vr3mhIOPPYZni8bsPKfV4tVf/SpisRhmZ2frGi/B2L+w5OQR59SpUzCbzZibm6Ph7ba2NvT19W3pB24ymXbsxCCXy2Gz2ej2eZ6H3+/fkddi5Bap2dlZ6lJc66ymbDaLRCJBLQQaWQMhiiItZt1rxsfHaW2Nw+GA2WxuyHZDoRAkSUIwGITX64XP58P8/DwmJyfB8zxWV1dLRItMJoNer4fD4YDdbofJZIJer4der4dGo6FRuaGhIVp/JQgCBEFAyO/Hv33gA5h+61vx0w98AN68aF6jIENKgZxwJe7TmUymZCI7ofi4qdbWns/S2bM0tUYe8+bbt7F09mzZfbM6ndA+/TQ+e/o0vtrejv/xi7+IXxwfh6O1FdeuXSvYR8bBZu/PGow9RRAEHDt2DPPz8xgbG8PJkyep6NFoNLhx40ZdV371mHNtF7lcjv7+4tMdo1F4PB4sLi4CuOu8W0ttFlmoSKqL53n6/O0uHMRTRiaTlY32kKjAbixQmUwG8XgcarUaCoUCo6OjmJ2dhcvl2va2I5EIrl+/XpKaIlGQsbExaLVaGAwG9Pb21iwAZTIZ7rvvPiwvL2NxcbFqh1L+XL3tUuwArVAoIIoieJ6nrfDF75V8t6Swulpbe/7om62YEzpaW/Hg2bNobW2lx47L5UIymUR3d3ed75axX2ERHgY4jkNXVxdsNlvB7U1NTXUPv2uk4EmlUvTK9urVq7hw4QIuXbqE+fl5enJUq9UNez1GIfl1O8RzJZvNbhrlKRYboigimUw2LNojCAIdzkkiFTKZjKbRFAoFTb3tNB6Pp8CBvLOzsyHRJzJjSxCEgvRUfsonGo3C7XZjfHy8rtZ4QRDQ09OD0dFRTH3hCzSVA9ztUJp86qltv4d8SJoNyJ0jmpqa6L8riVOS9iJpsNkHHsAzyIkc4G5b++yDD9LtKxQKak6YTzlzQp7nYTKZ0NLSgvvvvx9tbW0F+yKXyyGTyUrOi4yDy5GM8IRCoQJDPEbu5FKu08nhcMDj8dScOmrUIkM8X8xmM/r6+hCLxbC6uop4PI75+XmsrKxAFEW0tbWhs7OThZwbjN/vp6ks0pUlk8m29TmTK/atRHuI30uxYzBxOc5vjyepkc2GjRJILdtW8Pv9uH79Oo4fP05TbWq1ekueVsUUR6+IZ07x+/H5fPD5fHUvzFqtFrYKHUq2HUhrkcGrBoMBg4ODWFpawsLCwqZWA+Sc8oEnnsAXHn8c18+fp23t0w88gA989rPUQ0gQhE1H3hiNRhoVq9bw4HA4IJPJGtpxythbjqTgye94MBqN4HmeHdRV6OjoqFnwuFwutLS01LSYpdNp+P1+6icik8noVGJBEHDixAm6HZ1OR4uhU6kUgsEg3G435ufn4ff70d/fD61Wy4RPg3C5XNQPBUBd/jebiV4S7SE1HJtBak6qRZbKTR0ng3GLZ3uV29/tFEIHg0Gsr6+j5U4EQRCEhgieYioVFlutVlit1rq3JwgCIl1dkJaWStJE605nSW3MVlCpVOjp6YFGo4FSqSxwLLbb7TRttBmkHuyjX/wiFbmjZXyGkskkmqxWDJw9i7//+tchTkxg1WjEqXe9C63t7Whqaqo5KszzPOx2e31vmLGvOZKCh2C1WiFJEgKBADQaTcEJnnGXetptE4kEgsEgTCZT1cdFIhFcunSp4ASef3KpJlyUSiVsNhv0ej2Wlpawvr6OixcvwmAwoLm5GXa7nQnYbdLU1ASfz7el59YS5cuP9lQTIvUY4xUbYBJI3Q/P81AqlQ1pH69Gd3c3fD7fjpgkFgsznU6HkZGRLQv9X3/uOfzl8DDeGovRaMizKhUGPvjBquMWakWtVlcUDUqlEnq9vuYZXxzHIR6P0/dK0pkkvQnkzumDg4NQKBTo/9SnEA6HceXKFdx75sy+KHRn7C1HvoaH4ziYzeaGmYcdRur16Lh169amdR7RaLRgYZTL5XXPyFKpVOjq6oLFYoHdbodcLofP54Pf74coitjY2KDOzOFw+ND6wuwEbW1t0Gq1W3puPQt9IpGgowfyIa3U9Rx7xACz2n7lTzeXyWRQqVQNESb5Ikur1eL06dNb/vzKQfY5/xh2Op04c+bMtuqiWjo68Eu3buHPH3oIz/X34/97xSugffppAMgVM1++jHctL+Mjly8j+uijNXdwcRwHp9OJ4eHhqo8bGhra9OKIHAuiKBYIO9LtRW4zmUwYGRmh6UlRFBGPx2Gz2XYk4sY4eDDJe4d4PI5MJgO73c7SIkV0dHQgHo/XLAiTySQWFxfR19dX9v5sNgu3211wGykQrBe5XF521o0oigiHw7hx4wbt5tFqteju7obFYmHfcQ1sdSEl3im1fMZkJAVZ0PKLc7cyo6qW55CUiFwub1i0JxAIFPhBabVanDx5Ei+88MKWhXb+BPFkMlki/hoVsWjp6MB7/s//AZCLvF68eBHnPvjBsuMWPvvUU7BuUtCs0+lw/PhxqFSqTV9bLpfj1KlTOH/+fNlhqSQCSI6p4ohfvllhT09PwTE7OTkJj8dDu04ZDCZ47tDT08MWwQrwPI/h4WG0trbixo0bNdU7LC0tUQt8u92OcDiMaDSKYDAIn89XEAFSKBRbypWTKI5SqYTBYKDf39raGrLZLHQ6HWQyGbLZLORyOdLpNObm5rC+vo7h4WFEo1FEo1FotdqGXo0fFjQazZavjGUyWV2RE3I81DoSoRykYLnW33Ejf+9+vx+RSKTAdJOkXtfW1jZ9PmnPBkCdkvMLrvP3led5nDhxAkajsWH7TyAu5pXGLWxWzKzRaDA6Olr3dPG+vj6cf+GFAnfknl//9cLXL1PDlJ+q9Hq9Bc0oZNRJV1fXlp3fA4EA7frbTcsNxs7ABM8dmNjZHIPBgAcffBA+nw+BQADLy8tVH+/1euH1ejE1NVVQXJg/oZl87l1dXVvap1gsBo/HA7vdDofDASAX2ia+H/feey8WFxcxMDBAC1vJSTOTySCdTiOVSjHBU4btpEr2YtAnEbW10sgUpyiKuHTpEoaHhwvEe29vL4LBIO0kK3eeIUM+iyM4pMU+/7djtVphNpsbZnJYDM/z0Ov1FcctVCtmNhqNtFutXpRyOcKPP47fGhu7O4vrxRdheOIJaPPECkntAYWeTOl0GsvLywWfTX9/P3p7e7cVqbx+/TpEUURPTw9tqGAcXI58DQ+jPogvRX9/P+69917qp1GN4k4KhUJBnXiB3El2KyclnufR2dmJEydOULED5K6sdTodNBoNtFothoeH6dUecFfcGo1GtLa27tjicdAh85m2wl5MmN7LixZSEzQ7O1sg9ARBoGaeCoWCtq4Tl+NqkZD8zjONRgOTyUQjrTuJXq/PjVtQqQo8b0gxczl6enpw6tSpLUdBzn/5y/j1O2IHyEWT3jQ+jtXnny94XCKRoOm9TCZDRaQkSZDJZLh161aB6N3OcRiLxeh3SQbnMg42LMLD2DIGgwGjo6NIJBIIBAJwu90Fc5eA8lf6ZMAkuaJlgmP/QVyE86+ma2khJ2xFfOz2nKxGRXhIjQ2Q+9z8fj8sFgu9X6PRoKWlpSAiWstnyXEcWlpaYDabC1K2O43BYMh1Yz39ND771FOweTxYdzox8MEPlnRpcRwHq9WK9vb2be0fPz1dNoUmLC4WjDQp9xqkbov48Fy/fh0jIyPbNiXNr+9iA4oPB0zwMLaNSqWC0+mEw+FAPB7HxsYGotEonE4njEYj4vE4rZ1RKpU0lQTcndY8PT1N621IZKGpqanA6p2xe0Sj0YIOmGot4o2I5pC2460+VyaT1TTni0C8ebZSGF2MUqksEGoulwsmk6mgCH8rn5HFYtlyqnc7mM1mKBQKWJ1OWqBcLo3V1dWF5ubmhth5ZHp7y46NiNbg6SWKIhVF5Pi8dOkSTp8+vS3RMzc3ByBnl9Hb27vl7TD2D0zwMBoGx3HQaDQlV0MajQYjIyOYmZnB2tpaQcSH4zh6YimGOMgeP358T1IkR5VUKoWJiYmChYbjOCQSCVp7RWonMplM2VqdcsJIkiQ6XoKkIkhaopaRFZWot3292j5uBXJskjlefr+f1vOQYtlaxZwgCGhqaoLNZtuzyKdCoUB/fz/Gx8erPq6jo6Nhv8uX/Jf/gu///d/jlefP0xqev7/vPnS/4Q01uXLnC861lRXc/MxnMOX1ItHfj9d87Wto2UL9jV6vRywWY4amhwgmeBi7Asdx6O3thVKpxNzcHB0FUNyKnn97JpNBNBplYmeXqZS+IqKHCJZ6oyOCIFBRUyxSdnNB2YnUGTHAI8dqJpPBjRs30N3djebmZthsNjqItRI9PT3bTg01CpvNBp1Oh0gkUvExyWSyYbPs9EYjXv7DH+LHX/4yZLOzyPb04Bfe+17cmphAOp0u+30pFIqCurxkMomgz4fg296Gj8TjOeG0uIjnRkbwi+PjdYue/v5+ZLNZeL1eqFQqKBQKmM3mffH9MLYGEzyMXYPjOHR0dKClpQWJRAKrq6vwer0FDrgk5UUWXDa4b/dRq9Voa2srqDkhs4pI8W2t0RhSkC6KYsFzUqkU7U7KJ184pNNpiKJYtruJ3LYV8dJoAU0+m3K3LywsgOM4OByOkloUhUJBvaE0Gs2+cgLmOA5Go7Gq4CGT4huF3mjEQx/9aMFtNpsNS0tLAEBnqWUyGSqYi6OQNz/9abw3HsffAwgCiABwRqP467e/He//53+ua384joNOp4PRaMTGxgb8fj8MBgNOnDixvTfK2DP2zy+McWSQy+V0NlZPTw/thhgbGysxgmPRnb0j37I/P3pRq8DIb/8vB8dxBaJHoVCUmMqRCe2k+yeTyVDvFblcjlQqBZVKRaNP9exXI+B5vqL4I6MQ5ubmqC9UMBiEwWDAwMBAQ8XCTlDsXROLRAp9cj72sZq6NLdDR0cHFhYWIEkSjh8/jpWVFayurpat+ZIkCYblZfwEwKsBmhp7HoBw7hzCwSD0dXgX8TwPv9+PtrY2tLW1IZVKYXJyks71Yhw82GrC2FM4joNWq4Ver0d/f3+JwNnuohAIBDA5OQmgvplgRx1SP0LEgUKhoM6/crmcnvCVSmXFotXNxjak02nasZefniAQwSWKIm1FJmk0MvCTpDLyXZo3Y6uFyjzPl7zXzV6TLMqzs7Po7e3Fz/3cz+HkyZNVj+tUKtWQYurtYrPZ6PuNRSIIPPYYfutrX8MjP/gBfutrX8P1N70J4aKuzEZDjh/Sxj84OIiXvexldFBrPrFIBAvRKBU7uPP3qwD0pdM4/+Uv1/XaHMdhZGSEdsgJgoDjx48jGAziwoUL++I7YtQHEzyMfYPFYsHIyEjBwrcVh9R0Oo1EIoHp6WmMj49Dr9cjm81ibGwMy8vLZS3sGYXo9Xr09fXRKdekbieVSlGzRgC0M6Z44ZfJZJsKTEmSIAgC9WTaauSFPK/WaCBx3q4XUldGEAShpmhXMpmkDsbl6j8kSYLf74fH48HExAQuXLiwLQ+kRiGTydDfn+vPWjp7Fm++fbtQSJw/X7eIqBeFQgGDwQC9Xk9FtkwmQ0dHR8G5IRaJIPThD+PfhcNl29vlAGSzszW/br7QUiqVGB8fp+NwyLFTbMHB2P+wlBZjX2G1WmGxWOD3+yGTyUrs+quRyWQwPj6OYDBIu4gkSUIsFsNPf/pTOl9LLpfjnnvuoXUiOp0OPM8jkUggGo0im81Cq9UiEolApVLtiIX/QcDhcMBisWBsbAx+v7/qY4mlAJBbwItbvslMKHK/TCYrSENtpy2dtMvXkmYgUSqgNi+cfHieRyaToXVJtab2SOptbGwMp06dgiiKCAaD1L+KDLvN/wz3S8rEarXm0s8rK2WFRD0iYiuQKF9xRIfjOAwMDGB8fBzpdBru730PvzE+jh8BZdvbMwC4np6aXlOSJMzMzFCxR6KIsVgMi4uLMJlMcDgcR/a8cJBhgoex71CpVHSO0MzMDPR6fU2jH/x+P3w+H/2/XC4Hx3GwWCy08BHILXQ/+9nP6P/1ej3MZjPW19cRj8fp7LBAILDjrrb7Hblcjt7eXty4cYPW15Cp3fm1K/nTvCVJgiiKtP4GQEn7ev7iTthKhIeIJkEQahJMpNiVtMin0+my74dsmzyGDBoldUekpb4WiEiKRCL46U9/uun7NBgMDfG2aRQOhwOXmpvLColsjSJiO4iiWLZ5Qa/X44EHHkAoFIJraQkcgAcBfBelNTzx06fxy+99b02v5/F4CpzbAeD48eO4dOkSTCYTOjs7AWyezmTsP1hKi7Hv6O3txenTp9HX10eLmaPRKIC7k9avXbuGubk5GiFIJpMFE9glSaIFriTSU4lwOIzFxUWa6iKvubq6ytJfyE2/vvfee6FUKqlgILPKCKTOhqS8SNor37MnH1KDk18LtBVfHKVSCY7jkM1mK3r/kI4ukj4D7tbWkKt30jVGBkUSAUXqg0iHGoAC48xaSSaTVFiVg+M4Kgj3m/O4zWZD+2tfi28MDRWMmvjr48dx72//9o6+djweRyqVQjgcrvjZabVahO8IMj2AX0RO5PwlgKeamyF+9KP45R//eNOCZUmSsLy8DEEQCoaQkvuUSiU1INTpdBgbG9v1eXGM7cFVWwjOnDkjXbx4cRd3h8EoZHJyEhsbG7TLJRaLQalU4tixY5ibm6NRGLfbXVIzUq7tuV4MBgM6OzvR1NR05P03lpaWMD09Tf9PalhIBxcRF+UQBKGkA4u4bJMaoHqLQElNTDabpZGXYiFSrn05/z4iauqh3NTuevaZRL5I0XWxi/UDDzyw7zq4pqamMDUxgcRPfwrT6irCTicUL30punt7MTw8vGOvu7y8jKmpKQDAyZMnK4rBmakp3HrrW/GqCxdoZOf7DzyAl//whzV3ZkWjUWqeWo5kMkmP95s3byKVSuH06dNbnsTO2Bk4jrskSdKZcvexlBZjX9Pf34+Ojg6k02lMTk5CFEWMjo5CqVSis7MTa2trcLlcZYVNJpPZtugJhUKYmJiAWq3GsWPH9k1txV5ATNfIYp9/dUvqVGqFtJYTtnKlTKZkA4XpMCJuqokdAAUmiLXuu0wm29ZVPYmEAbnPjBTAktusVuu+EzsA0NfXB5PJBOurXgWO47C4uIjZ2Vmsrq6is7Nzx2ZNGY1G9PX1YXp6GmtraxUFT29/P+z/+I8FxoUvf+9762pDr5Y2d7vdmJ6ehlwuLxDI0WiUCZ4DBBM8jH0Nx3FQqVRQqVQ4ffo0FhcX6eKm0WjQ3NxckMoqhqSzSCSg3itzUpyaSqVw69YtnDx5clvv5yCj1WoxMjKCyclJ8DxPa2GqpWqAu3U2+cKDRITqhYxvIAKr2HiOFEvnD/TcjHqEMYkmbSUiVW5b+bVBAGh9yH6D47iCOhqVSkX/HQwGd0zwEL+uWCwGr9dLU4PlKGdcWCvkWKzURefz+TAyMgKj0Yjz588jnU5Dq9VueRwKY29gNTyMA0VHRwc94XEch/7+/qrFg8S7hRSO1puWyl+YA4EAxsbGjrT/hs1mw4MPPkingXMct+lJnxT55kM6uQhkmGe170epVNJi6HLuy+R7rldIiaJYts6oEvW2wVeC1BSR92G1Wg9MtMBgMNButVq7KLdDV1cXVCoVLl68SCO9jeTWrVu4fv162fuIH4/FYqG1VqSxYaeNFxmNhUV4GAca0klTS3ExuTonbHaVTtqQCZIkYX19HalUCgaDgba8a7VaNDU1wWQyHYk6H+LPotPpMDMzU2D3X078kIhL/mej0WjgdDrR1NREPWoAIBKJ4Nq1ayW1OCSVUKk7qp60VDlIqzzZ380g7s9bXXjJKBXymfA8f6AmcqtUKtx7770QBIG6YO8kSqUSJpMJ8XgcGxsbiEQiGB0d3fY4DjL+Y21tDQAwPz8Pp9NZEMEC7kZ+ZDIZTp06hYWFBchksn2ZfmRUhgkexoGnnogLuTqvJYUhiiJ1C86fgZRIJApMx6LRKFwuF5RKJUZGRmAymep/EwcMjuPQ0tICk8mEGzduIB6PF9TOmM1mqNVqcByHcDiMZDKJcDgMo9GI7u5uGI3GsuJQp9Ohu7ubumMTthKdIygUioqRH1KTRbad30pfjfzOr1pTZ2QoLvF8yn8/XV1dB27x3InIzsrKCubn56FSqWAwGNDb21twXBEhnEwmMT09jaGhoS2/FrG9IGIHyAmeQCCA48ePVxRTer0ex48f3/LrMvYOJngYBxpST1EvxemI4ggE6aYhoogU61Zrn06lUrh27RqOHTsGq9Va9z4dRDQaDe677z54PB7EYjGYTCY0NTWVTffUmjZqbm5GJBJBPB6H0WiE0WiEx+PB6upqxeeQlGVxxIWkjchYiPz0VTabLRsVIt1bZJ/LCeP8ye9ku8VRKbI/JF1HWvaBwloRi8WC9vb2TT+Xo0BzczP8fj/W19cRCoWQzWYxMDAAjuPQ3Nxc0CW4sbGx5ddJpVK4ceMG9VgiqWulUomBgYF9NciV0TjYt8o48KhUqrr9ctLpdEFKi6TGRFGkdSn5i2d+23WxOMpf6DiOw9jYGE6fPr0rtQ37AZ7ny842KqZWYUpcdPPR6XSQJKngajyf/BZ3AAXihohVch8xRqxEOUPFfIpTo6R4W6VSFfgOkQLbdDpdUSRrtVoMDw8fiVRoLXAch+HhYchkMng8HrjdbmQyGQwNDUEmk8HhcFDhu1XjPzJmJhwO00L3VCoFQRBwzz33MLFziGHfLONAs76+vunYg3KU81LJrw/JX4AUCgUVNWQ6dzlPGfLYZDKJtbW1IyN4dgOFQoG+vr6KggfIDYcltRdEhJRbFOsRF0TgBny+gknhHa97HdR5bczFfj5EFImiWNWbqBF1KIcNnudpqsrj8VAH9FOnTmFgYADr6+sQRRGJRKKuYvNQKIRAIICNjQ0Eg0F6kUNqw4jDO+Pwwn5pjAPL7OwsFhcXt/TcbDZbkJYAQMcMFF+NF6cqkskk5HI59WSRyWS0/Z1sb319Hd3d3ezKvYHE4/GKLcnksyeRHBKZaUTbcDQczk0KvzM8UwLwjXPnoPiTP4G8SsFutfETRqPxyPs6bUZvby+SyST8fj8ikQhu3ryJnp4eHD9+HLOzs4hEIpiYmMDIyEjFbZCOqmw2iytXrhSITxKNI7/RYDAIj8eD5ubmuveVWBsw9jesLZ1xIMlms1Wv9svB8zwdGyAIQkmdiUKhqDjhWxAEyGQy+jxiIEcms5PxBOSEGo/HcevWra29OUZZfD4fkskk/dxJFCeTyVBhk7+ANUpsLn772yWTwt98+zYW/uZv6t4Wx3Ho6enBqVOnmNjZBIVCgdHRUWo2GAgEcPnyZdy8eRPt7e1obW2F1+utKCpFUcSNGzdw9epVXL16lUb8FAoFFcTFBerE1T1/G8Vks9mCqDKpBwqHw41424wdhElSxoFjY2MDMzMzNO9e72iA/FoOkgIhC2ilRZI8p56OsN1o1z1KSJJEFyvyXRGRSQwHN1us6kUQBKjvDKbMhwOgcrnq2pZarcbx48drGoTLyMFxHNrb2wsEhiiKmJiYoPP2Kv1mo9Eo/H4/LUIn5wpik1Au+sfzPHw+H2w2G4LBIG7evAmNRgOtVkv9nwKBQME4DdK16fV6D4yP0lGFCR7GgUKSJCwtLSEWiwHInWzyu2rS6XTFmglSX0PYyhylWpHJZGhtbaUhdcb2CQaDZbuc8rvp8tnqvKt8RFGkgymLJ4VHW1oQi0QKanvaX/taaMrUbun1eoyOjjIRvAXMZjO0Wi0dIAzkvpcrV67g+PHjFW0giOAtJ2zKiSSNRkMLyEVRxOrqKkZGRqBQKKgnVzweR0tLS4FPD/Hhstvt23ynjJ2GCR7GgSISiRRc7XEcVzBPibQAl5tmnU6ny7Yu7wTZbBYXLlwAkHMn7uzsRCaTqeg/w6hOLBYr8D7Kh0xdL+6WI8M5t0Mmk8lNCj93jqa1JADfGBpC00MPla3twZNPFogetVqNEydOMLGzRTiOw6lTp3D16tUC0UPMLiuh1WrrGgGS7/QdCASgUChgNptpXVgmk4HD4Sh5nkwmg8ViKRivIYoiYrEYPB4POI6D2Wyms+gYewcTPIwDxezsbMX7SBEx8dMovrITRXHbjrxbIRgM4vLlyxBFEXK5HHa7HQ6HA8Y6BhsedapF4jiOK4ii5c9Pq4TX48HEk0/CsbqKVYcDg489BqvTWfAYUoxutlgg/+M/xle/8x1ol5cRaW5G9xvfiLm//mu8q0xtz1fOnsXAW99KjSuHhoZYvc42USgU6O/vx9WrVwtuv3r1KoaHh8sOFZXL5ejr68Pk5OSm0T6dTleQplpfX6cz+ohoOXHiRMXnt7S0FBxvS0tLWFxcpJFHl8uFe+65h/3m9xgmeBgHikqLGFn0iODJZDK09Ty/Bb0RaY56yRdYmUwGKysrWFlZwdDQEJxFiyyjPGazGRaLpcBsjuO4ktbv/EGxlfB6PIg++ig+kkjkIjPLy3j20UeBp58uED1knIUgCFCqVOh75JGC7WiXl8vW9mhXVgDk6n8GBgbYItcgTCZTyTGQSqWwsrJScYo6MTLcrMHB6XRSp2tJkhAKhQAACwsLaGpqwrFjx6q2rBeflzo6OtDe3o7x8XGEw2Fks1lEIhF6LMRiMczPzyMcDkOj0aC1tZVFgHYBVlzAOFB0dXUVzBwixYgymaygjoOkrciCyHFcxVqPvcLtduPKlSu4cuUK5ufnEQwGEY1GEQ6H90SY7XcGBwdpW7pSqaRmkESYkNs3+44nn3oKb78jdoCcSHl7IoHJp54qeFx+MXQ5SG1PwXOQq+1RKBQ4efIkHbLKaAzlXJDX19dx7ty5is7LtUxyz4/A+f1+RKNRyGQymEwm2hZfjXQ6DZ/PB5/PV+Dl1dzcDK1WC7lcjkAgAEmSEIlEcPHiRaytrdHZYNevX8eFCxcwPj6O1dXVutPuJH3GzhvVYREexoFCqVRi5c4VNADqaFsO8uMn0Z9sNkvFz34gvyYlGAxifn6e/l+j0aC3t5ctmHkolUqcPHmSpjWIECHRPKC2VnS7x1M2MmPzeApvu+OjIwhCWRFVqban7eGHYbPZWDfWDkDMGq9evVogClKpFNbX18v+Xrq6utDR0YF0Oo3l5WU0NTXB5/PB7/dTt+X8wmetVgu9Xg+e52G32xEIBOB2u0sidalUCmNjY+jq6gKQs03geR4rKytob2+nY1FmZmaQyWSwvr6Oixcv0vNQMfF4nJqWTk9Pw263w2q1wmg0btr4oNFoIJPJMDs7i46ODlYvVgEmeBgHhkQigStXrtDi1Hpa0uuZqr4fiMViuHHjBtra2goGKB51NBoNTp06hcuXL5fcV2uB6qrDAakoHSUBWHc60V/0WDKLq2xXj04HPPkkvnL2LLQrK4i1tqLrDW+AQqlER0dHfW+MUTMGgwEDAwOYmJgoiGhUczYns9R6enoA5NJjJCpjNpsLIjyCIKCrqwszMzNwOp3w+XxYX19HR0dHQbSICHCSWs1PqxFBI5PJ0NPTA51Oh1AohOnp6ZrOWUScLS8vQy6Xo7e3d1NDRI7jYDAYMDs7i8HBwU1f4yjCVQuBnTlzRrp48eIu7g6DUZmVlRVMTk7iC48/jp7z5zEIYALA7AMP4ANPPFH2OSqVCul0uiAKcNCwWq3o6uqCRqNhLe538Hq9mJqaot1YpDOvllQAqeEhaS0JwLMqFbRFNTz5YwdqRZIkDAwMoLW1dQvvilEP4XAYExMTNN00MjLSsNbwtbU1zM7O4sEHH4TP50MkEkEwGKxauFwLoijSaNTy8nKJ0Wm1LtKuri4aTcqHjNfweDxYXl7G8PAwNBoNHYty1M4ZHMddkiTpTNn7mOBhHBQymQxe93M/h9eeP493AnSxegbAd8qInq2YEu53uru70dnZude7sS+4fv06fD7flp7r9Xgw+dRTsHk8WHc6MfDBD5Z0aeXPUKsVnufx0pe+lI0Z2CWy2Sxu3LiBQCCAY8eOwWaz0fsmb97EX7/mNXCur8Njs+HX/+7vMHD8eE3b9Xq98Pv96O/PxfwCgQCSyWTZtvRaCYVC0Ov1VEATT7GZmRn6mGqCRy6XY2BgAFarFTzPw+12U9dnQRAQiUTQ09ND03OSJGFhYQGdnZ0H9mJvKzDBwzgUpFIpfFgQ8EcoNYH7rwAe/slPCh6/W547u4lWq8V9992317uxL1hcXMTCwkJJGqtR1gMkalTPYqFWq3H//fcfqQVmryGRjFu3biEYDEKtViMRCmHiF38R7xBFemH0FzyPl167VrPoaTQ3b97EyMhIQcRFkiRcvHiR+gvVcs6Sy+UYHh6G0WjE6uoqgsEgzGYzHA5HSTRnYWEBQK5r7Kgck9UEz9GKdTEONEqlEv1A2YLTAQBvfsUr6G2CIBw6sQPk7PIPW9Rqq7S2tpaNpJAFkEA6+eoN7WcymQJH3VpgM9R2H1JcrtfrkU6nEQwG8fwjj1CxA+TOEe8QRTz3qlft2Xmhv7+/5BjkOA5NTU11baelpQVGoxFyuRytra0YGRlBc3Nz2eO7o6MDarUaqVQKa2trR76LiwkexoFiAijbCjwN4E3k/3d8eA4rh/m91YNMJitIYRDI1HtSv0BGihDjx3pIJpN1C6V6RRJj+yjvFIqfOXMGAwMDaA8Gy14YOb1e3L59G8vLy/D7/bsqAIilQj61eAQ1NTXB6XTSY3d1dZWO1tkMjuNgt9shCAKampogSRLW19fh9XqPpPhhgodxoIj8yq/gGdwVPaSG5+cB2mHDcdyhraFQKBQ1+YocFSpdHadSKWpAWGz8WEton3T1CIJQt0tyNYM6xs6i1WrR0tKCdYej7IXRqtUKlUqFQCCAGzduYHJykhpXrq+v4/bt21hYWIAkSXC73Zibm4PP56MmgYR0Oo319XUsLS1Rk8KtEI1GYbPZcOrUKfT29paI65MnT2J0dBRDQ0N46UtfCoPBgGQyiStXrtQseghyuRw8z8Nms0GpVCIcDiMYDNY1EPmgczhXBcah5avf+x4e5jjcANCHXGTn5wE8DOD3AIzceVwikahqGndQsdlsRyYXXwv5c9XKUfxZ5btuV6PcoFmlUolsNrtphG11dfVI1UzsR3797/4Of3HyZEkNzyPPP09b0/MH+6bTacTjcdhsNuh0OkiSBL1eTwcTt7a2QqFQQBRFeDwerK6uQq1Wo6WlZVsT0tva2ui/TSYT9Ho9PB4PUqkURFGk7s9AToQPDg7iypUryGQyuH37Nrq7uyu6TFeDCCdRFLG+vn5kHN+Z4GEcOK4AeA1Q0qn1V3duB+4udMRdmbQY50NmbvE8fyDqYgwGQ4HLNCP3mdRDpTlrtTyPtKjzPF91fEUsFoPP52OmkXvIwPHjwLVr+MzDD8O+vo41mw1vOHu2oGA5P5qiUCiodxLpbvJ6vTh27FiB6JidncXa2hruvffeHUldmkymitPfgVwE65577sHNmzcRCoVw69Yt9PT0wOFw1C2w19bWIJPJYDAYNp09d1hgXVqMA0knx+F1yKWxppATO98o6tLKRy6XF1yZk8nI6XT6QPzYjUYjTpw4cWhTdVtFkiRcvny5IN2wGbXYFVR7DLnqr7YNjUaD0dFRVs9zgEkmk5DJZAW/uVAoBI1Gs+O/Q1JYXal+TBRFJBIJOprCaDRuuWU+FothY2MDbW1t+/48WAvVurTY2ZNxIFm4I9QDgQCuXr1KIzuVyGazUCqV1I03/wo/f/DofkSlUmF0dJTVhpSB4zhYrda6BE8mk6HpCblcDkmSSiI+lS4EyWNJPUT+MUOEkCiKiEajSKVSTPAcYMoVGdcTUZQkCWtraxAEoWrUphyLi4tQqVQVU008z0Oj0UCj0Ww7kki2cxSKmJngYRxoyk1QLgdZqCoJm+Kajf1Eb28vEztVqPd7y2azyGaz4DiOjo1QKBT0arqau3L+zC5yHJFasfznbeVKOZ1OsxlIhwRijWA2m3Hu3DnY7XY0NTVBp9MhlUohk8lUrMcjBdSCIOxqbc1hiO5sBhM8jANPb28v/H5/TREaURTLOugmk0m66NVSmLpbtLS0wGq17vVu7FuCwSDcbnddzyGzj/IpPh7IkNniYbNEMOenGvKnY+eTSCTqighcunQJo6OjrAvvgCOKIm7evFnQ9r62tlbQfi6Xy2GxWMpeyGQyGUSjUUSjUYRCobrr1BiVYW3pjAOPRqOpa1hjpWLTdDqNZDJJi5xJQfNe0dzcjP7+/iNx5bUV0uk0bt68WXcongiZYjGTTzabhSAIJfeTCAyJEFWjng7BQCCARCIBl8tV83MY+5O1tTX4fL6Kx6VcLsfx48crRm3z29w3i1wz6oNFeBiHAq1WW/Nja4kEkZoO0sacv7jFIhG4/vZvoXe7EW5uRvtrX5ubnN1gzGYzEztVWF1drXvWVa2QolCCJEm00J2MnJDJZAWprXLbqIVAIIDr168DwLY8XRi7SzKZxOLiIkRRRE9PDxQKBWKxGCYmJqo+z2g0Vq3pyRdKsVgMqVQKCoWCnQsaABM8jEOB1WqtqZYnnUxi5pvfhMHjqUmskDZmsrDGIhEEHnsMv3X7Nm2J/8a5c8CTTzZc9ByFIsLtsJkHTyXq+VyJU3M6naa1P0CuoDWVSlXt5orH45tuP5VKYXZ2loqjeDxOp18z9i8ejwfT09M0nRkKhdDd3Y1oNLrp8UV8gCqR3wG2vr4On88Hu92OgYGBXRc9kiQdKosFltJiHAo4jsPx48erFvjGIhFs/N7v4beffhqP/OAH+K2vfQ2Bxx5DLBKpuu38ep6ls2fx5jtiB8j5AL359m0snT3bmDeSx2EzTWwkmUxmy4KnVorHUuRTy8JTrU1YkiREo1FcvHixIKqjUqnq6jhj7D5erxcTExMF54VoNIqbN29ibm5u0+ePjY0hEAhUFEbFHYPZbBarq6vb2+ktwnHcoUqr7bngqeUqiMGoBY7j0N7ejvvvv7/sjKWls2fxxvHxusUKqedRKpXQLi+XndGjXVlpwDsopFxbLCNHMBjcso1ALdET8phEIgGZTAalUklb2IG7Rc2VRGlra2tZB95UKoX5+Xn87Gc/w89+9rOSxa29vR1Go7Het8TYBSRJgt/vh8vl2lb0NRaL4erVq7h69WrZ9W99fb3kNlEU92yt7OzsRCAQ2NY2yCiOvWbPU1rMp4LRaARBwLFjxxAKhTA/Pw+fzwcA0K2sbEms5Kcsws3NkFA4sV0CEG1pacSuF6Dbgbqgw0Imk6GmcJlMpq5oGBkqWkkw5TsxE1FDtq9SqZBMJmkNT3E3n0ajQWtrK1oqHA/r6+uYn5+H1+PBxJNPwrG6ilWHA4OPPQar0wmVSsVqNfYhyWQSt2/fhlarRTAYbMg2g8EgLl++jJaWFrS1tdHUaaWIisvlwuDgYENeux4EQdiWXUI2m8Xi4iLsdnsD92pr7LngYT9uxk5hMBgwOjoKl8uFmZmZhoiV3je9Cd944QWa1pIAfGNoCG0PP9zQfZfL5XUPrTxKCIJQIHTy6x7yr76JR05+VxbP85DJZPQ2YkaZzWahUCiQSqUqFqbnn6/yRZZGo0FfX9+mheZutxtejwfRRx/FRxKJ3DG0vIxnH30Uiuee29JcJMbO4/F44Pf7G55GTafTWFhYwMbGBpqbm7GxsVFRiLvdblocvdtstVt1cnIS2WwWQ0ND+2Kt33PBw2DsNG1tbdBoNMi+6U34qxdfxJvupLW2IlYUSiWavvAFfOU734F2ZQXRlha0PfxwwwuWD4vN+06h0+mgVqtpJ1Ul36R8UULckSVJQiKRgFKpRCaToYZ/oigimUwiHo1WLEwvl2bU6XQYHR2tSaAqFApMPvUUfv+O2AFyAvztiQSe/Pzngde9rs5PgrHTiKK47ZTOZkQiEUxNTdW0LwcJvV5Puxn3A0zwMA49HMfBYrFAbzQi8cQT+Or3vgf10tKWxIpMJoOgVmPgrW/dwT3GtiYwHwXkcjlOnDiBixcv1lxPwfN8Qc1MftqKwHEcls6exbvLFKZ/5exZGN75zoJttrS01OWEbbFYYPd4yqZWHXnGdIz9w8zMzI4XyNfK7Owsent7D0z0t7m5ea93oQAmeBhHBrVaDa1ej2OPPrrpOAJJkqBSqQpcl8mspJ3unpLL5Uzw1IBWq4XZbKY1WptR7Sozk8nQFvNqtV75of2Ojg50d3fXdfVqMBiw6nBAKip+lwBEN2lXZuwOkiQhmUwikUhgbW0NKzvQkLBVVldXsb6+jq6uLrS3t++byMlBgQkexpGhngnHpDg1H+K7stMMDw8fmCu4vaatra1mwbNZJIh8v5GWloq1XvmpM4fDUfeCs7q6io7f+R088+53452SRFNmf8FxeOiLX6xrW4zGkc1mkU6n4fP5sLa2tuMprO0giiJmZ2cRCAQwOjra8O273W5Eo1H09PTsqdP8TnC43g2DUQWSdshms/vW1K+pqenQmHztBmazueZZQ7WIk1QqhZ43vhHfGBoCOUJIrVfPG99YUEOxuLhY175GIhG43W7EX3gBr5ck/BjATwD8GMDrJAmLf/d3dW2PsX0kScLi4iLOnTuHF198EZOTk/ta7OTj8/l2xJ+HDCy9evXqvh2ovFVYhIdxZCBX52RWFilWzW89JsWr1UYWFI+aaCRNTU07st3DCsdxGBoawuXLl6sOfBUEoWaRK1coYPviF/G1b38b6qUlhJub0fPGN0Je1B2ztraG/v7+miOHU1NTEEURupUVGAA8VHS/bHa2pu0wtk86nUYkEsHMzAwimxiP7mcWFhZgt9sbej7iOA59fX3weDy4du0aBgcHD403FBM8jCNDvnEXcdHNZDKQJInW7CSTSUiSREcHFJNMJmmnT7WxAlvh2LFjZQ0TGdXRaDSw2+1Vay02E7H5cBwHvdGIvre8hR4b+eNFCJIkIRgM1h2Rq2SPkGU1PLuG2+2Gy+XasVlsu0UsFsO5c+fQ09PT8AJhp9MJq9Va9ULioMEED+PIkC94kskklEolvernOI6KF/JvlUqFRCJRcvVEnrOdkyXHcXA4HDQlw/M8q9vZBo3yJpHJZLRtncBxXMUr6Hpqumw2G4LBINpf+1p849y5Ai+n7z/wAF7+3vduc+8ZtSJJ0oEXO4R0Oo3l5WU4nc6GR57lcnldtY/7ncPzThiMKqTT6YIrFRKlqQQRPYIgFIihfLLZLFQqFSRJKhsNKgeJRtjtdmg0mvrfCKMsjRy2Wc7rpNL3W48vCkkLaHQ64Mkn8ZWzZ6FdWUHTAw/g5e9/P/SHJG1wENhPnVeNIBKJYGJiAgMDA4eu0LiRMMHDOBLEYjH6bzKOoJbFiix0MpkMMpmsYOEjQqjWyExnZyc6OzvZCWkH2KyNf7MrX2JKWE24EkdmALBarWhra6u5YBrI2SKQY0+j02HgrW+FxWLBiRMnat4GY/vkp68PEx6PB8lkEsePH2/oBcBhggkexpEgX/BsBVLUTBasfGoJI3d3d6Ozs3Nb+8CojMFgAMdxFaN2xfOziKcSSWtUEzukSD2TycBms6G3txdqtbrufZTL5bj//vuxvLyMeDyOtrY25re0B3Ach+7ubly9enWvd6Xh+P1+TE9P78nMrYMAu9RkHAnyazK2Q7krp3Q6XbWWQ6vVoqOjoyGvzyiPTCbbdA4VqfMhw0GTySSdp1UtskO+W7lcDqfTuSWxQ1Aqleju7sbIyAgVaYzdRZIkeDyeQ1szt7a2tm9tN/YaJngYR4L8BW2rKaVyUQClUrmp8zJpfWfsLCaTqer92+msI51atdZqMfYvJBJ4WL/LgzZvazdhgodxJCCLHM/zWx4Nkc1mc7O0BAGCINAFkHRzVeIwtXXuZxwOR1UxS6I69YrPVCoFmUyGWCzGfJIOCX19fXu9CztKoyLahw0meBiHnkQiAb/fD0mSCgpP64XUcSSTSZoO4Xl+00W0nsJWxtYRBAEjIyMQBKGgTZ146WwHURShUCh2ZbQIY+dRKBSHVrxKknToutAaBRM8jEPP6upqRfO4raJQKJDNZjcNHxsMBnR1dTXkNRmbY7Va4XA4kM1maSROEISao2wk7aVQKArqteRyOYaHh1lq8hDR19d3aL9P1qVVHiZ4GIce8uNvVDt4vkNzNZRKJUZHRw9tceR+pbOzE3K5nEbi6qnVIO3KqVSKLoZyuRz33nsvm3F2yNBoNIdmZEIx7FgtDxM8jENPI2toyCJYi4jp6+s7VC6lBwWZTIaeLY5pIPPUiKgFgJ6eHmYSeUg5jClKuVwOnU6317uxL2GCh3HoCQQCBZ4r20GhUCCdTiORSEAul1ccSul0OtlcrD2EjO2ol0wmU+DK7XA4Gj6jiLF/aGtr2+tdaDisK7Qy7PKTcehRq9UIBoPIZrPbPhGQ55MC5kwmQ2tESD1PR0cHuru72UlnDyETn3/2s59teRstLS3o7+9n3+MhRq/Xw2g0IhgMbms7giCgqakJZrMZKpUKMpkMyWQSLpcLsVhsV12drVbrrr3WQYMJHsahR6FQNKxouVw0h7QtcxyH3t7eQ3nVeBDZjs9Kd3c3Ojo6mNg5Amwl7azVatHe3g6dTgeO46DRaEqOFa1Wi6amJkiSBL/fj4mJiR0XPkqlkp1/qsAED+PQEwgEADSmaLlSSzvP8xAEgaU/9hFbmaDe1NSErq4uZiVwRJAkqW7PmtOnT9c1EoTjODQ1NWFgYABjY2M0EiwIAnQ6HTY2Nup6/UqoVCqcOXOG1Q1WgX0yjENNKpVCKBQCUFms1AqJFFW6795772XtoPsIlUpV4o7McRysVis0Gg2USiXi8TgtUrbZbFsSSYyDy/j4OKLRaMntPM/D4XDQouZEIgGlUgmdTrfl+WcWiwX9/f2YmJiAyWTCwMAAbt++va39z6erq4uJnU1gnw7jUON2u+m/yw3+rAdJktDb2wulUom1tTXE43EkEgmYTCZ0d3ezk80+Qy6X4/Tp01haWkIsFkN7ezv0ej0TpQzK4OAgdDodFhYWIIoiOI5DW1sbHA7HjnQ6mc1mWCwWDA0NIRKJQCaTbfu8BOTqFB0OR4P2sj5IU8hBaPHnqnmJnDlzRrp48eIu7g6D0ThEUcSLL75Ir/CrTdOuh8HBQZa6YjAOEZlMhjY17JZvViAQgEajweXLl7c9CmJgYAAtLS0N2rP6EEURV69ehUql2hfmnBzHXZIk6Uy5+1hbOuPQ4vP5qNiRyWQNmyAcCoXYNGIG4xBBLCZ20yTUZDJBqVTi/vvvx+Dg4Ja3o1ar9/QCjOd5HDt2DKFQCJFIZM/2oxaY4GEcWvLTWY2cIOx2u8vm/RkMBqNeeJ5Hc3PzlmuDzGbznkdVBEHA6OgopqamsLy8vG8vCJngYRxKlpaWCrofGil4nE4nczJlMBgNxel01v0cnuf3Te0MaQSYmpqijSKZTGZfXRyyKkvGoaS4MLURHjw8z8NqtaK/v39b22EwGIxiWltboVQqsb6+jkgkglgsBgDUsX19fR1ALoXV09MDi8XSsPmAjYI0bly5cgUqlQrZbBZ2u33fnDOZ4GEcSkwmE/23UqmsuwvCbDZDq9WC4zg6cdtsNrNOLAaDsWPYbDYqcJLJJMLhMB0E6vP54PP50NXVtS/tEwKBAHw+H/1/IpGAQqHYs+6xcrCzN+NQEo/HC/5fb0pLo9Ggr6+vkbvEYDAYNUMutAgWi2VfT0EPBAIFtTtOpxNDQ0N7uEel7K94GIPRIJaWlui/t5LKOoxTlBkMBqPREEPXrq4unDlzhhZQ76fIDoFFeBiHElIox/N8TdEdi8UCs9kMv98PQRD29ZUUg8E4Wly/fh1KpRImk4nW9wwMDEClUu3pfgWDQYyNjWFkZASiKCKRSECSJHR0dMBsNu/pvpWDCR7GoSOdTtMhfbVOSFepVGhra0Nra+uet3gyGAxGPjabDRMTE/B4PPS2ZDK5p4InnU5jenoaqVQKV69ehc1mw9DQEJ0cvx9hKS3GoSMUCqGpqQnt7e0QBKEmTwi3240bN25gcXFxF/aQwWAwasfpdBZ0nmo0mj21xpAkCTMzMwiHwwByJQB9fX2QyWSwWCz79qKRCR7GoYO0QpJRErX8+ARBQGdnJzo7O3dhDxkMBqN2OI5DV1cXAECn06G3t3fPZsKlUimMj4/TaJNKpcLJkycPRN0jS2kxDh1WqxU8z+Py5ctIpVI1efD09fXBYDDs0h4yGAxGfbS1tcFqtUKhUOyJPUYikcDq6ipcLhe1+TAYDLBYLFCr1bu+P1uBCR7GoWN8fBypVArhcBgcx9VkzsWckxkMxn6G47g9ERaSJGFlZQXT09MF5QEOhwPDw8O7vj/bgQkexqEjFArRoaEAKtbwaDQa9Pb2lvhdMBgMBiPH3NxcSW0jx3EHMv3PBA/jUOHxeArEDlBZ8Bw7dgxarXY3dovBYDAOHJlMBh6Ph9ZDAjmrj56eHmg0mj3eu/phgodxaEilUvD7/VCpVEgkEgByP85UKlW2cNnlcqG/v3/Piv8YDAZjPyOXy/HSl74U2WwWa2tryGQysNlse+7/s1WY4GEcGpRKJVpbWxEMBultCoWiovGgx+PB+vo6urq60N7evlu7yWAwGAcKmUyG5ubmvd6NbcMED+PQkM1mcf369YJBocT2vBJyubxg0CiDwWAwDidM8DAODX6/v0DsKBSKqu3obW1t6OnpqamLi8E4bCwsLECv10On00GpVO717jAYOw470zMODfm1OHK5vKR4OR9BENDd3c3EDuPIsra2hpmZGfh8vr3eFQZjV2ARHsahIb+1vNo4iba2NrS3t7NiZcaR5r777tvrXWAwdhV2ecs4NGg0GtjtdgC52h2FQgGFQlHwGKVSSb13aoWIp2QyWdNcLgaDwWDsP1iEh3GoyK/ZyWQyJcKGODDXMkZCkiQsLS1hZWUFcrkc0WgU7e3t6O7ubvh+MxgMBmNnYYKHcajo6upCOp1GJBIBkJv/wvM8eJ6HRqOBXq9HIBCAXq/fdKgox3EwGAxwOp1IJpNQKpWsuJPBYDAOKCylxThUGAwGdHV1obm5GSqVijqEkknDKysrWF9fL+jmqobRaIRCoWCdLAwGg3HAYYKHcajgOA4WiwVarRYcx8FsNsNsNiMajSIcDgPIpbUmJiY2naDOYDAYjMMDS2kxDh0cx6GtrQ02mw0LCwvweDwlbsterxd+vx8OhwMGgwFms5kNEGUwGIxDDBM8jEOLIAgYGBhAd3c3NjY2EIvFsLa2hkQiAblcjvb2dkiShEgkgqampr3eXQaDwWDsIEzwMA49CoUCTqcTANDT04NMJgOO45gPD4PBYBwhmOBhHDnkcnbYMxgMxlGDFS0zGAwGg8E49DDBw2AwGAwG49DDBA+DwWAwGIxDDxM8DAaDwWAwDj1M8DAYDAaDwTj0MMHDYDAYDAbj0MMED4PBYDAYjEMPEzwMBoPBYDAOPUzwMBh1kM1m93oXGAwGg7EFmOUsg1EFSZKQTqexsbEBl8uFbDaLY8eOwWAw7PWuMRgMBqMOmOBhMIqIx+NIJpOIRqPgeR6CIEAmk2FoaAh6vR4cx+31LjIYDAajTpjgYTDuEA6HMTU1hVAohJMnT6K1tXWvd4nBYBwiMpkMZDIZu2jaI5jgYRx5YrEYVlZW4PF46CR1QRD2ercYRxxRFBGNRhGLxZBMJiFJEhQKBQRBgMlkgkwm2+td3BOy2SwWFxfh9/thNpvR0tJS0+9VFEWk02kkk0nE43GYzWYolcpd2OO7TE1NQRAE9PT07OrrMnIwwcM4siSTSczP///t3WtTGlm3B/B/N90N3Q0IBm/gY5IyUzVJVZKZOB/ifPSZqpyZqSSVm5MbhhjQNJe+0PRtnxc5vR/NqFGjovj/VflCQGgR6cXaa6/1Htvb2wAAXddx+/ZtlEolTlSnqYmiCK9fv4bjOMiy7MDbWJaFX3755cJP2Kfhui4+fPiAu3fv/lCQlmUZtra20Ol0EEURAGA0GmFrawsrKyv4z3/+g1KptO/2k8kEQRDgy5cv6Ha7+zYdqKqK1dVVzM3NAcChQWSapuh2uxiPx9B1Xb4/hGEoMzbFYhG6rsul8PF4jDRNIYSAEAJRFCGKIsRxDEVRYNs2FhYWoKpf9w0JIZBlmfyZ/DJFUaCqKgqFgrwtnZ6SP7kH+e2338Tjx48v8HCILsZkMsHjx48RxzEAYHV1Fbdu3WKgQxcuSRJ0u10Z4IxGo2PtBlQUBQsLC6jX6zLrs/ekKISA53kwTfPcXtdCCIxGIwwGAwghUCgUMJlMZEZKCAHHcSCEQLFYRLVaRbFYlNkWTdNQKBSQJAmyLINlWTBNE2EYyg0DcRzDMAzs7u4iDMMjn4+9y0X5//ZxqaqKcrmMcrkMXdcBfH2f2N3dRZIkp3+SjjheVVWPvfPTtm38+uuvfI/6DkVR/lcI8dtB1/GZo2tJ13X5Scq2bayvr3Ndnc7dZDJBv9+H4zjo9/tQFEWe7E9KCIFer4derwfg62u60WjANE0kSYIvX77A930AQLFYRKlUQrFYhGEYUFUV9Xod9Xr9VI/reR6GwyF6vR5Go9Gxfm4ymWBnZ+fEj3eS4/qRwCQPNo/7+/woIcSJ2lz4vo9nz56h0WigWq2iXC4z63NCDHjoWoiiCIqiyE9uo9FIvjn6vo8nT56g1WqhXq9f29oI+j4hBNrtNnzfx40bN7C0tHTobZMkkcFNEAQIguDEWYeTiONYLs9+K8+67NXpdLCxsQHLso79GMPhEC9fvsR4PP6hY6XTGQwGGAwGAL5mpEzThGmaMpi1LEsuc+7s7MCyLCwsLPA97f8x4KGZ9+rVK2xvb8MwDDx69AhRFOHJkyf7btPv99Hv96FpGu7cuYPl5eUpHS1dRnkBcbvdllmKXq+H9+/fwzRNZFmGlZUVNBoNRFGEN2/eYHd3d8pHfbQ0TREEwZEBT541GY/H8DwPb968YfPNSyJ/TeZZvMNsbm5ifn4elUpFFr1blrWv3um6YMBDM89xHABfszxPnz7F7du3D11CSJIEr169QqVSgW3bF3mY9I0oitDr9eC6LmzbRpIkaLVackdOEATo9XoYj8dQVRXLy8sYDAbwPA9xHMsTc6FQgGEY0HUdlmWh2WweunwphMBkMkEYhrLY1XVdWST7rfF4LLMdg8EAiqLgqLrIy6BQKODmzZtYXFw88KQXBAFevHiBIAiQZdml/33oaGmaYmdnZ99yoqIoaDabqFaraDQa1yYDxICHZlq+OyLn+z4cx0Gr1UKn0znwZ4QQ+Ouvv/DgwQN2VD4h13UhhDh2g8Z8B0sYhgjDEL7vYzQayeaP3/r06ROazSaiKMLOzs6+bMNhyznf+vjxI0zTRBAE+3bSAPjhE/xVCA7yGpxKpbIv4MkLjDc3N48sDqarTwiBTqeDTqcDTdNkMTnw3+LvnKZpsgasUqlc6eCIAQ+dShRFSNMUxWLxUhXOeZ6Hra0t9Pt9uQX025PQaDTCxsYGgiBAv98/8H6SJMH29jYDnkM4joN2u400TWHbNhYXFxFFEV6+fAkAqNfrcqknD17yv8f8/Dwsy4LjOPA870QFu0mSoN1u/9Cx783KXEdZlqFUKqFWq+277MWLF+daVEyXU5IkMgv+PYqiwLIsVKtVrK2twTTNcz66s8WAh05lZ2cH3W4XYRjCMAxomgZd16HrOpIkwdraGsrl8nfvJ+81cVby+5qfn0epVMJ4PJa1ObZtwzRNLCwsyH/cwwIe4OtJPUkSbgPdIwxDdDodfPz4UQaSruvi8+fP+26X10QdJN9VRNMzGAzw7t07aJqGKIrQ7/e/WwtCJISA7/tXIpN5EL6TXyH5i2ya26f/+ecf+L4v6yI0TYOmabAsC/V6HdVq9cDjy7IMURRB13WZEnVdFy9evMDy8jKq1SoMw0Acx7AsS+6m+lZeRHnY9bZt4+7du/suS9MUcRwjiiJMJhM4joM3b97I3Q6HmUwm6Ha713rERL7E5LouhsMhT4oz4iK3X9NsqVQqePDgwaHvwZcZA54rQgiB3d1dBEGAUqkE27ZhWRZUVUW/30e73UahUJDbE2u1GiqVyr/uJ01TDAaDU7emX15ext9//31gv4v379/L5QpFUWSAsbceQ1EUlEolGYQIIfD27dt991MoFDA3NwdVVeX6sWVZ0DQNW1tbGI1GqNVqsjtppVLB8vIyisUi4jiG53kYjUYYDocYj8fn0jTsOsh3txHR9aZpGmq1GkzTxM2bN69s1pudli+YEAJBEMDzPARBAEVRYBiGzIBomoZSqSRrT/J25IVCAbquy4xKfrJXFEVmVNI0RZIkUFUVw+EQrutiMpnAMAyUSiXZ9jxfw7937x50XYfv+5ibmzt2xB7HsayDePv27YHFpVeVoigol8soFotYWVnBjRs3pn1IUxOGIZ4+fcqsDtE1pCgKVldXUS6XMT8/f2UyOuy0fI7yk79pmtB1HWmawvd9uK4rd4AkSYI4juUgwIvuYxFFETzP23dZGIb4888/5feKosA0TRiGIYvSFEVBmqYIwxCTyUS2QjcMQ1btr6+v4/nz5xf6+5yXarWK9fV1OVvnusmD8cFggG63yyUPomuq0Wjg1q1bx6rDvEoY8PyAJEnw+++/I01TFAoFFAqFQ/t1XHb5yS4/4X369Gnah3RuisWiXBIslUowTRNzc3NXNk17GvnuKd/34XkeXNeF67pc/iO6xhRFwcOHD/ft4Jsl1+cd/gwFQSBHExiGISfjsgPpdOS7wwDIycL5kmCWZVBVFaqqYmVlBfPz8zMf2OTLpXlmbjKZIIoiObsnH8hIRLTX0tLSzAY7wDUPeJ4/fw7TNGFZFpIkQZIkiKIIWZah2WzCMAwkSSJrX1zXxWg0YlOuS0bTNDQaDRSLRZTLZVQqlUvVG+i85UMPh8Mhdnd34brutA+JiK4g3/cRRZGcxzVrZj7gOarPi23bePfu3YHXcXfK1RGGIXZ2dpBlGZIkQa1Ww88//3xpiux835dF3lEUyV1ntVrtxIFZEARwXRdBEMj79DzvyvbFIKLLw3VdPHv2DPfu3UOxWJxqC5TzMNMBTxiGcgu1ruuoVCpYXFzEjRs3oCiKnCfjOA6+fPly7G6TdHncuXMHzWbzUmZ0kiTBu3fvDh1hoWka1tfXsbKycqz7cxznX0NPiYjO0mg0wh9//CE3sPz0009HDpi9SmY64CmVStjY2ECn04HjOHAcB6PRCJ7n4datWwAA0zTRarXQarUwHo/l0MF8Z1Ucx8iyDFmWIU1TeJ534nb4dH76/T6KxSIajca5fBo5bSdo13Xx/PnzI0cYJEmCzc1NLC4uHqsnkmEYmJ+fl9/n3a01TZN1SvmwTSKi46jX61hdXYWqqlAUBcVi8dKNDDor7MNzCnl77e3t7UM/vdPFWlpawsLCAmzblt2cjzu8cjwey35GedFz3stoc3MThUIB9+/fP3KJbDAYyCWm0Wj03S7OAGBZFiqVCu7cuXNmy2/j8RivX78+cmQGERHwtWvyo0ePZmrpin14TiGvB9F1fd+LIY7jfQWidDl0u110u135ff5JxbIs2Vconwi89++pKAqSJJGZkTzoyS8HcOSoC+Dra8VxHAwGA5kRNAxDNpXc27eoVCrBsiyYpnniTtd7J4vHcSwL7fNMZN4TKgiCEz57RHTdlMtl3L9/f6aCne9hhgdflxZ838dwOMTnz58xmUzkFnNVVeXE7CAIrmyfHfqqUCjAtm2USiXZtTofX7H36yzSuXnQnP+PCSFQKBSgadq/3mTiOIbjOLIXTv4zWZb9a2s5EdFZMAwDzWYTa2trM7OEdVSGhwHPN4QQCMNQ1mDQ9ZNPUrdtG+VyGbZtH9qYMH+9eJ6HMAzlV94H57D/r3xUSB5g9ft99sYhoqkoFotYWlqSA6BPM2fxsuCS1gnkIxZ6vd60D4WmJK/R8n1fvg40TcPa2hqWlpZk8Xt+m9M0nMwbVbKnExFN22QyQbvdRrvdlvMEq9UqKpUKbNuGaZoz0bD1Sv4GQggIIc4tBZckCQaDAcrlMuI45lICIUkSvH379l+T3YmIZokQQo6a2UvXdTSbTbRaLbkz9KqZesAjhECn00EQBLBtG4ZhYDAYYDKZoF6vo1KpoN/vw3EcRFEkCzWFECiVSqjVapibm0O1WoVpmmfyR9A0DQ8fPpTfp2mKbrcLz/P2FYnmw0H31mjkdRtERESzIo5jfPjwAR8+fADw32X5Wq2Ger0uJxXkCYJ8TmH+dRkawU69hqfT6WBzc/PM7q9QKMidMXufcMMw9hWp5um5NE3lHymOY7ktOf9SVXXf7b+VZRmiKJJFpY7jsEszERHRHpVKBRsbG+f+OJe6huesl4rSNJX9UIbD4ZG3zQdMHoemaXJbsRBC7pxhoSkREdHRLsNw7akHPNN0kmArSRLZZZmIiIiulqtXdURERER0Qgx4iIiIaOYx4CEiIqKZx4CHiIiIZt7Ui5bL5TJarda0D4OIiIjOiWEY0z6E6Qc8tVoNtVpt2odBREREM4xLWkRERDTzGPAQERHRzGPAQ0RERDOPAQ8RERHNPAY8RERENPMY8BAREdHMY8BDREREM48BDxEREc08BjxEREQ08xjwEBER0cxjwENEREQzTxFCHH6lojy+wGMhIiIi+hG7Qoj/OeiKIwMeIiIiolnAJS0iIiKaeQx4iIiIaOYx4CEiIqKZx4CHiIiIZh4DHiIiIpp5/wcN1ncPlIMZ9AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "size_by_pop   = {}\n",
    "size_by_coord  = {}\n",
    "geometry = []\n",
    "\n",
    "# Create geometry size dict\n",
    "for rec in metadata_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    population = rec[1][\"population\"]\n",
    "    \n",
    "    # Store the province, unless it's missing, then use country\n",
    "    geo_name = rec[1][\"province\"]\n",
    "    lat = rec[1][\"province_lat\"]\n",
    "    lon = rec[1][\"province_lon\"]  \n",
    "    \n",
    "    if geo_name == NO_DATA_CHAR:\n",
    "        geo_name = rec[1][\"country\"]\n",
    "        lat = rec[1][\"country_lat\"]\n",
    "        lon = rec[1][\"country_lon\"]   \n",
    "        \n",
    "    metadata_df.at[sample, \"lat\"] = lat\n",
    "    metadata_df.at[sample, \"lon\"] = lon\n",
    "    point = shapely.geometry.Point(lon,lat)  \n",
    "    geometry.append(point)\n",
    "    str_point = str(point.x) + \",\" + str(point.y)\n",
    "        \n",
    "    # Add Population to Dict\n",
    "    if population not in size_by_pop:\n",
    "        size_by_pop[population] = {}         \n",
    "        \n",
    "    # Add counter to the size_by_pop\n",
    "    if geo_name not in size_by_pop[population]:\n",
    "        size_by_pop[population][str_point] = 0\n",
    "    size_by_pop[population][str_point] += 1\n",
    "    \n",
    "    # Add counter to size by coord\n",
    "    if str_point not in size_by_coord:\n",
    "        size_by_coord[str_point] = 0\n",
    "    size_by_coord[str_point] += 1\n",
    "\n",
    "metadata_gdf = geopandas.GeoDataFrame(metadata_df, crs=\"epsg:3857\", geometry=geometry)\n",
    "display(metadata_gdf)\n",
    "\n",
    "#print(size_by_pop[\"0.PRE\"])\n",
    "#print(size_by_coord)\n",
    "\n",
    "# Generate a figure with two axes, one for CartoPy, one for GeoPandas\n",
    "fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG},figsize=(10, 5))\n",
    "world_polygons.plot(ax=ax, color=\"grey\", alpha=0.5)\n",
    "metadata_gdf.plot(ax=ax, c=\"red\", ec=\"black\", lw=0.25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "successful-amsterdam",
   "metadata": {},
   "source": [
    "## Calculate Root Distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "trained-pottery",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>host_human</th>\n",
       "      <th>sequencing_technology</th>\n",
       "      <th>assembly_method</th>\n",
       "      <th>host_raw</th>\n",
       "      <th>host_order</th>\n",
       "      <th>population_color</th>\n",
       "      <th>population</th>\n",
       "      <th>geometry</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "      <th>population_rtt_dist</th>\n",
       "      <th>tstv</th>\n",
       "      <th>cds_sites</th>\n",
       "      <th>ns_sites</th>\n",
       "      <th>ss_sites</th>\n",
       "      <th>ns_ss_ratio</th>\n",
       "      <th>other_var</th>\n",
       "      <th>other_var_ratio</th>\n",
       "      <th>homo_het_sites</th>\n",
       "      <th>homo_sites</th>\n",
       "      <th>het_sites</th>\n",
       "      <th>het_ratio</th>\n",
       "      <th>external_branch_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>Human</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Human</td>\n",
       "      <td>Human</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>POINT (-105.60772 38.72518)</td>\n",
       "      <td>7.31686e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.3815e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.004600e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>IonTorrent</td>\n",
       "      <td>Newbler v. 2.6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#c5e88a</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>POINT (40.79579 47.62225)</td>\n",
       "      <td>7.30501e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.6582e-06</td>\n",
       "      <td>1.854</td>\n",
       "      <td>216</td>\n",
       "      <td>121</td>\n",
       "      <td>51</td>\n",
       "      <td>2.37255</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0196078</td>\n",
       "      <td>242</td>\n",
       "      <td>216</td>\n",
       "      <td>26</td>\n",
       "      <td>0.107438</td>\n",
       "      <td>2.120100e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46335 42.48050)</td>\n",
       "      <td>5.41847e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.15566e-05</td>\n",
       "      <td>1.935</td>\n",
       "      <td>246</td>\n",
       "      <td>134</td>\n",
       "      <td>60</td>\n",
       "      <td>2.23333</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>271</td>\n",
       "      <td>246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.0922509</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Frontopsylla elatoides</td>\n",
       "      <td>Siphonaptera</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46335 42.48050)</td>\n",
       "      <td>5.47035e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.20754e-05</td>\n",
       "      <td>2.119</td>\n",
       "      <td>241</td>\n",
       "      <td>133</td>\n",
       "      <td>58</td>\n",
       "      <td>2.2931</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>258</td>\n",
       "      <td>241</td>\n",
       "      <td>17</td>\n",
       "      <td>0.0658915</td>\n",
       "      <td>2.356000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>Non-Human</td>\n",
       "      <td>Illumina Hiseq 2000</td>\n",
       "      <td>SOAPdenovo v. 2.04</td>\n",
       "      <td>Citellus undulatus</td>\n",
       "      <td>Rodentia</td>\n",
       "      <td>#238af5</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>POINT (85.46335 42.48050)</td>\n",
       "      <td>5.4939e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.23109e-05</td>\n",
       "      <td>2</td>\n",
       "      <td>247</td>\n",
       "      <td>135</td>\n",
       "      <td>60</td>\n",
       "      <td>2.25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>265</td>\n",
       "      <td>247</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0679245</td>\n",
       "      <td>4.711000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313243_45</th>\n",
       "      <td>Azov38</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313243_45</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (40.79579 47.62225)</td>\n",
       "      <td>7.54024e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.21397e-05</td>\n",
       "      <td>1.917</td>\n",
       "      <td>96</td>\n",
       "      <td>48</td>\n",
       "      <td>28</td>\n",
       "      <td>1.71429</td>\n",
       "      <td>3</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>120</td>\n",
       "      <td>96</td>\n",
       "      <td>24</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1.202300e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7313246_49</th>\n",
       "      <td>Gdansk8</td>\n",
       "      <td>[1400:1700]</td>\n",
       "      <td>[-621:-321]</td>\n",
       "      <td>Poland</td>\n",
       "      <td>Pomeranian Voivodeship</td>\n",
       "      <td>52.215933</td>\n",
       "      <td>19.134422</td>\n",
       "      <td>54.2456</td>\n",
       "      <td>18.1099</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA7313246_49</td>\n",
       "      <td>KEEP: SRA Ancient Combined Record</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>54.245560</td>\n",
       "      <td>18.109900</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (18.10990 54.24556)</td>\n",
       "      <td>6.01348e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.8721e-06</td>\n",
       "      <td>1.813</td>\n",
       "      <td>171</td>\n",
       "      <td>74</td>\n",
       "      <td>45</td>\n",
       "      <td>1.64444</td>\n",
       "      <td>17</td>\n",
       "      <td>0.377778</td>\n",
       "      <td>260</td>\n",
       "      <td>171</td>\n",
       "      <td>89</td>\n",
       "      <td>0.342308</td>\n",
       "      <td>5.131000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6651390</th>\n",
       "      <td>AGU010</td>\n",
       "      <td>[1435:1477]</td>\n",
       "      <td>[-586:-544]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6651390</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1456.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (25.24953 54.82269)</td>\n",
       "      <td>5.95981e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.3354e-06</td>\n",
       "      <td>2.378</td>\n",
       "      <td>138</td>\n",
       "      <td>72</td>\n",
       "      <td>40</td>\n",
       "      <td>1.8</td>\n",
       "      <td>1</td>\n",
       "      <td>0.025</td>\n",
       "      <td>218</td>\n",
       "      <td>138</td>\n",
       "      <td>80</td>\n",
       "      <td>0.366972</td>\n",
       "      <td>2.360000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637004</th>\n",
       "      <td>AGU025</td>\n",
       "      <td>[1441:1612]</td>\n",
       "      <td>[-580:-409]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637004</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1526.5</td>\n",
       "      <td>494.5</td>\n",
       "      <td>85.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (25.24953 54.82269)</td>\n",
       "      <td>6.0564e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.3013e-06</td>\n",
       "      <td>2.143</td>\n",
       "      <td>140</td>\n",
       "      <td>75</td>\n",
       "      <td>41</td>\n",
       "      <td>1.82927</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0243902</td>\n",
       "      <td>193</td>\n",
       "      <td>140</td>\n",
       "      <td>53</td>\n",
       "      <td>0.274611</td>\n",
       "      <td>9.423000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA6637002</th>\n",
       "      <td>AGU007B</td>\n",
       "      <td>[1463:1632]</td>\n",
       "      <td>[-558:-389]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Vilnius County</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "      <td>54.8227</td>\n",
       "      <td>25.2495</td>\n",
       "      <td>Second Pandemic</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>1.PRE1</td>\n",
       "      <td>SAMEA6637002</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>1</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1547.5</td>\n",
       "      <td>473.5</td>\n",
       "      <td>84.5</td>\n",
       "      <td>54.822692</td>\n",
       "      <td>25.249534</td>\n",
       "      <td>Human</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>#f3c16a</td>\n",
       "      <td>1.PRE</td>\n",
       "      <td>POINT (25.24953 54.82269)</td>\n",
       "      <td>5.96217e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.359e-06</td>\n",
       "      <td>1.838</td>\n",
       "      <td>154</td>\n",
       "      <td>75</td>\n",
       "      <td>43</td>\n",
       "      <td>1.74419</td>\n",
       "      <td>8</td>\n",
       "      <td>0.186047</td>\n",
       "      <td>231</td>\n",
       "      <td>154</td>\n",
       "      <td>77</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>2.360000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      strain         date      date_bp  \\\n",
       "sample                                                                   \n",
       "Reference                               CO92         1992          -29   \n",
       "GCA_009909635.1_ASM990963v1_genomic     9_10       1923.0          -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic    42126       2006.0          -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic    42123       2005.0          -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic    42118       2005.0          -16   \n",
       "...                                      ...          ...          ...   \n",
       "SAMEA7313243_45                       Azov38  [1400:1700]  [-621:-321]   \n",
       "SAMEA7313246_49                      Gdansk8  [1400:1700]  [-621:-321]   \n",
       "SAMEA6651390                          AGU010  [1435:1477]  [-586:-544]   \n",
       "SAMEA6637004                          AGU025  [1441:1612]  [-580:-409]   \n",
       "SAMEA6637002                         AGU007B  [1463:1632]  [-558:-389]   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                     China   \n",
       "...                                                       ...   \n",
       "SAMEA7313243_45                                        Russia   \n",
       "SAMEA7313246_49                                        Poland   \n",
       "SAMEA6651390                                        Lithuania   \n",
       "SAMEA6637004                                        Lithuania   \n",
       "SAMEA6637002                                        Lithuania   \n",
       "\n",
       "                                                   province  country_lat  \\\n",
       "sample                                                                     \n",
       "Reference                                          Colorado    39.783730   \n",
       "GCA_009909635.1_ASM990963v1_genomic           Rostov Oblast    64.686314   \n",
       "GCA_009669545.1_ASM966954v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669555.1_ASM966955v1_genomic                Xinjiang    35.000074   \n",
       "GCA_009669565.1_ASM966956v1_genomic                Xinjiang    35.000074   \n",
       "...                                                     ...          ...   \n",
       "SAMEA7313243_45                               Rostov Oblast    64.686314   \n",
       "SAMEA7313246_49                      Pomeranian Voivodeship    52.215933   \n",
       "SAMEA6651390                                 Vilnius County    55.350000   \n",
       "SAMEA6637004                                 Vilnius County    55.350000   \n",
       "SAMEA6637002                                 Vilnius County    55.350000   \n",
       "\n",
       "                                     country_lon province_lat province_lon  \\\n",
       "sample                                                                       \n",
       "Reference                            -100.445882      38.7252     -105.608   \n",
       "GCA_009909635.1_ASM990963v1_genomic    97.745306      47.6222      40.7958   \n",
       "GCA_009669545.1_ASM966954v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic   104.999927      42.4805      85.4633   \n",
       "GCA_009669565.1_ASM966956v1_genomic   104.999927      42.4805      85.4633   \n",
       "...                                          ...          ...          ...   \n",
       "SAMEA7313243_45                        97.745306      47.6222      40.7958   \n",
       "SAMEA7313246_49                        19.134422      54.2456      18.1099   \n",
       "SAMEA6651390                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637004                           23.750000      54.8227      25.2495   \n",
       "SAMEA6637002                           23.750000      54.8227      25.2495   \n",
       "\n",
       "                                              biovar branch_major  \\\n",
       "sample                                                              \n",
       "Reference                                 Orientalis        1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic       Medievalis        2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          Antiqua        0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          Antiqua        0.ANT   \n",
       "...                                              ...          ...   \n",
       "SAMEA7313243_45                      Second Pandemic        1.PRE   \n",
       "SAMEA7313246_49                      Second Pandemic        1.PRE   \n",
       "SAMEA6651390                         Second Pandemic        1.PRE   \n",
       "SAMEA6637004                         Second Pandemic        1.PRE   \n",
       "SAMEA6637002                         Second Pandemic        1.PRE   \n",
       "\n",
       "                                    branch_minor biosample_accession  \\\n",
       "sample                                                                 \n",
       "Reference                                 1.ORI1        SAMEA1705942   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1        SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1        SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic       0.ANT1        SAMN07722923   \n",
       "...                                          ...                 ...   \n",
       "SAMEA7313243_45                           1.PRE1     SAMEA7313243_45   \n",
       "SAMEA7313246_49                           1.PRE1     SAMEA7313246_49   \n",
       "SAMEA6651390                              1.PRE1        SAMEA6651390   \n",
       "SAMEA6637004                              1.PRE1        SAMEA6637004   \n",
       "SAMEA6637002                              1.PRE1        SAMEA6637002   \n",
       "\n",
       "                                                     biosample_comment  \\\n",
       "sample                                                                   \n",
       "Reference                              KEEP: Assembly Modern Reference   \n",
       "GCA_009909635.1_ASM990963v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669545.1_ASM966954v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic              KEEP: Assembly Modern   \n",
       "GCA_009669565.1_ASM966956v1_genomic              KEEP: Assembly Modern   \n",
       "...                                                                ...   \n",
       "SAMEA7313243_45                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA7313246_49                      KEEP: SRA Ancient Combined Record   \n",
       "SAMEA6651390                                         KEEP: SRA Ancient   \n",
       "SAMEA6637004                                         KEEP: SRA Ancient   \n",
       "SAMEA6637002                                         KEEP: SRA Ancient   \n",
       "\n",
       "                                     branch_number      continent  date_mean  \\\n",
       "sample                                                                         \n",
       "Reference                                        1  North America     1992.0   \n",
       "GCA_009909635.1_ASM990963v1_genomic              2         Europe     1923.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic              0           Asia     2006.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic              0           Asia     2005.0   \n",
       "GCA_009669565.1_ASM966956v1_genomic              0           Asia     2005.0   \n",
       "...                                            ...            ...        ...   \n",
       "SAMEA7313243_45                                  1         Europe     1550.0   \n",
       "SAMEA7313246_49                                  1         Europe     1550.0   \n",
       "SAMEA6651390                                     1         Europe     1456.0   \n",
       "SAMEA6637004                                     1         Europe     1526.5   \n",
       "SAMEA6637002                                     1         Europe     1547.5   \n",
       "\n",
       "                                     date_bp_mean  date_err        lat  \\\n",
       "sample                                                                   \n",
       "Reference                                    29.0       0.0  38.725178   \n",
       "GCA_009909635.1_ASM990963v1_genomic          98.0       0.0  47.622245   \n",
       "GCA_009669545.1_ASM966954v1_genomic          15.0       0.0  42.480495   \n",
       "GCA_009669555.1_ASM966955v1_genomic          16.0       0.0  42.480495   \n",
       "GCA_009669565.1_ASM966956v1_genomic          16.0       0.0  42.480495   \n",
       "...                                           ...       ...        ...   \n",
       "SAMEA7313243_45                             471.0     150.0  47.622245   \n",
       "SAMEA7313246_49                             471.0     150.0  54.245560   \n",
       "SAMEA6651390                                565.0      21.0  54.822692   \n",
       "SAMEA6637004                                494.5      85.5  54.822692   \n",
       "SAMEA6637002                                473.5      84.5  54.822692   \n",
       "\n",
       "                                            lon host_human  \\\n",
       "sample                                                       \n",
       "Reference                           -105.607716      Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic   40.795794      Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic   85.463346  Non-Human   \n",
       "GCA_009669555.1_ASM966955v1_genomic   85.463346  Non-Human   \n",
       "GCA_009669565.1_ASM966956v1_genomic   85.463346  Non-Human   \n",
       "...                                         ...        ...   \n",
       "SAMEA7313243_45                       40.795794      Human   \n",
       "SAMEA7313246_49                       18.109900      Human   \n",
       "SAMEA6651390                          25.249534      Human   \n",
       "SAMEA6637004                          25.249534      Human   \n",
       "SAMEA6637002                          25.249534      Human   \n",
       "\n",
       "                                    sequencing_technology     assembly_method  \\\n",
       "sample                                                                          \n",
       "Reference                                              NA                  NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic            IonTorrent      Newbler v. 2.6   \n",
       "GCA_009669545.1_ASM966954v1_genomic   Illumina Hiseq 2000  SOAPdenovo v. 2.04   \n",
       "GCA_009669555.1_ASM966955v1_genomic   Illumina Hiseq 2000  SOAPdenovo v. 2.04   \n",
       "GCA_009669565.1_ASM966956v1_genomic   Illumina Hiseq 2000  SOAPdenovo v. 2.04   \n",
       "...                                                   ...                 ...   \n",
       "SAMEA7313243_45                               NextSeq 500                  NA   \n",
       "SAMEA7313246_49                               NextSeq 500                  NA   \n",
       "SAMEA6651390                                  NextSeq 500                  NA   \n",
       "SAMEA6637004                                  NextSeq 500                  NA   \n",
       "SAMEA6637002                          Illumina HiSeq 4000                  NA   \n",
       "\n",
       "                                                   host_raw    host_order  \\\n",
       "sample                                                                      \n",
       "Reference                                             Human         Human   \n",
       "GCA_009909635.1_ASM990963v1_genomic            Homo sapiens         Human   \n",
       "GCA_009669545.1_ASM966954v1_genomic      Citellus undulatus      Rodentia   \n",
       "GCA_009669555.1_ASM966955v1_genomic  Frontopsylla elatoides  Siphonaptera   \n",
       "GCA_009669565.1_ASM966956v1_genomic      Citellus undulatus      Rodentia   \n",
       "...                                                     ...           ...   \n",
       "SAMEA7313243_45                                Homo sapiens         Human   \n",
       "SAMEA7313246_49                                Homo sapiens         Human   \n",
       "SAMEA6651390                                   Homo sapiens         Human   \n",
       "SAMEA6637004                                   Homo sapiens         Human   \n",
       "SAMEA6637002                                   Homo sapiens         Human   \n",
       "\n",
       "                                    population_color population  \\\n",
       "sample                                                            \n",
       "Reference                                    #ff0000      1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic          #c5e88a      2.MED   \n",
       "GCA_009669545.1_ASM966954v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic          #238af5      0.ANT   \n",
       "GCA_009669565.1_ASM966956v1_genomic          #238af5      0.ANT   \n",
       "...                                              ...        ...   \n",
       "SAMEA7313243_45                              #f3c16a      1.PRE   \n",
       "SAMEA7313246_49                              #f3c16a      1.PRE   \n",
       "SAMEA6651390                                 #f3c16a      1.PRE   \n",
       "SAMEA6637004                                 #f3c16a      1.PRE   \n",
       "SAMEA6637002                                 #f3c16a      1.PRE   \n",
       "\n",
       "                                                        geometry  \\\n",
       "sample                                                             \n",
       "Reference                            POINT (-105.60772 38.72518)   \n",
       "GCA_009909635.1_ASM990963v1_genomic    POINT (40.79579 47.62225)   \n",
       "GCA_009669545.1_ASM966954v1_genomic    POINT (85.46335 42.48050)   \n",
       "GCA_009669555.1_ASM966955v1_genomic    POINT (85.46335 42.48050)   \n",
       "GCA_009669565.1_ASM966956v1_genomic    POINT (85.46335 42.48050)   \n",
       "...                                                          ...   \n",
       "SAMEA7313243_45                        POINT (40.79579 47.62225)   \n",
       "SAMEA7313246_49                        POINT (18.10990 54.24556)   \n",
       "SAMEA6651390                           POINT (25.24953 54.82269)   \n",
       "SAMEA6637004                           POINT (25.24953 54.82269)   \n",
       "SAMEA6637002                           POINT (25.24953 54.82269)   \n",
       "\n",
       "                                    root_rtt_dist clade_rtt_dist  \\\n",
       "sample                                                             \n",
       "Reference                             7.31686e-05             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic   7.30501e-05             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic   5.41847e-05             NA   \n",
       "GCA_009669555.1_ASM966955v1_genomic   5.47035e-05             NA   \n",
       "GCA_009669565.1_ASM966956v1_genomic    5.4939e-05             NA   \n",
       "...                                           ...            ...   \n",
       "SAMEA7313243_45                       7.54024e-05             NA   \n",
       "SAMEA7313246_49                       6.01348e-05             NA   \n",
       "SAMEA6651390                          5.95981e-05             NA   \n",
       "SAMEA6637004                           6.0564e-05             NA   \n",
       "SAMEA6637002                          5.96217e-05             NA   \n",
       "\n",
       "                                    population_rtt_dist   tstv cds_sites  \\\n",
       "sample                                                                     \n",
       "Reference                                    6.3815e-06     NA        NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic          9.6582e-06  1.854       216   \n",
       "GCA_009669545.1_ASM966954v1_genomic         1.15566e-05  1.935       246   \n",
       "GCA_009669555.1_ASM966955v1_genomic         1.20754e-05  2.119       241   \n",
       "GCA_009669565.1_ASM966956v1_genomic         1.23109e-05      2       247   \n",
       "...                                                 ...    ...       ...   \n",
       "SAMEA7313243_45                             2.21397e-05  1.917        96   \n",
       "SAMEA7313246_49                              6.8721e-06  1.813       171   \n",
       "SAMEA6651390                                 6.3354e-06  2.378       138   \n",
       "SAMEA6637004                                 7.3013e-06  2.143       140   \n",
       "SAMEA6637002                                  6.359e-06  1.838       154   \n",
       "\n",
       "                                    ns_sites ss_sites ns_ss_ratio other_var  \\\n",
       "sample                                                                        \n",
       "Reference                                 NA       NA          NA        NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      121       51     2.37255         1   \n",
       "GCA_009669545.1_ASM966954v1_genomic      134       60     2.23333         0   \n",
       "GCA_009669555.1_ASM966955v1_genomic      133       58      2.2931         0   \n",
       "GCA_009669565.1_ASM966956v1_genomic      135       60        2.25         0   \n",
       "...                                      ...      ...         ...       ...   \n",
       "SAMEA7313243_45                           48       28     1.71429         3   \n",
       "SAMEA7313246_49                           74       45     1.64444        17   \n",
       "SAMEA6651390                              72       40         1.8         1   \n",
       "SAMEA6637004                              75       41     1.82927         1   \n",
       "SAMEA6637002                              75       43     1.74419         8   \n",
       "\n",
       "                                    other_var_ratio homo_het_sites homo_sites  \\\n",
       "sample                                                                          \n",
       "Reference                                        NA             NA         NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic       0.0196078            242        216   \n",
       "GCA_009669545.1_ASM966954v1_genomic               0            271        246   \n",
       "GCA_009669555.1_ASM966955v1_genomic               0            258        241   \n",
       "GCA_009669565.1_ASM966956v1_genomic               0            265        247   \n",
       "...                                             ...            ...        ...   \n",
       "SAMEA7313243_45                            0.107143            120         96   \n",
       "SAMEA7313246_49                            0.377778            260        171   \n",
       "SAMEA6651390                                  0.025            218        138   \n",
       "SAMEA6637004                              0.0243902            193        140   \n",
       "SAMEA6637002                               0.186047            231        154   \n",
       "\n",
       "                                    het_sites  het_ratio  \\\n",
       "sample                                                     \n",
       "Reference                                  NA         NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic        26   0.107438   \n",
       "GCA_009669545.1_ASM966954v1_genomic        25  0.0922509   \n",
       "GCA_009669555.1_ASM966955v1_genomic        17  0.0658915   \n",
       "GCA_009669565.1_ASM966956v1_genomic        18  0.0679245   \n",
       "...                                       ...        ...   \n",
       "SAMEA7313243_45                            24        0.2   \n",
       "SAMEA7313246_49                            89   0.342308   \n",
       "SAMEA6651390                               80   0.366972   \n",
       "SAMEA6637004                               53   0.274611   \n",
       "SAMEA6637002                               77   0.333333   \n",
       "\n",
       "                                     external_branch_length  \n",
       "sample                                                       \n",
       "Reference                                      4.004600e-06  \n",
       "GCA_009909635.1_ASM990963v1_genomic            2.120100e-06  \n",
       "GCA_009669545.1_ASM966954v1_genomic            0.000000e+00  \n",
       "GCA_009669555.1_ASM966955v1_genomic            2.356000e-07  \n",
       "GCA_009669565.1_ASM966956v1_genomic            4.711000e-07  \n",
       "...                                                     ...  \n",
       "SAMEA7313243_45                                1.202300e-06  \n",
       "SAMEA7313246_49                                5.131000e-07  \n",
       "SAMEA6651390                                   2.360000e-08  \n",
       "SAMEA6637004                                   9.423000e-07  \n",
       "SAMEA6637002                                   2.360000e-08  \n",
       "\n",
       "[601 rows x 44 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata_df[\"root_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"population_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "\n",
    "population_mrca_dict = {}\n",
    "\n",
    "# Get the MRCA for Each Population\n",
    "for population in colors_dict:\n",
    "    if population == NO_DATA_CHAR: continue\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    tips = list(df.index)\n",
    "    mrca = divtree.common_ancestor(tips)\n",
    "    population_mrca_dict[population] = mrca\n",
    "\n",
    "for c in divtree.get_terminals():\n",
    "    sample = c.name\n",
    "    if sample not in metadata_df.index: continue\n",
    "    population = metadata_df[\"population\"][sample]    \n",
    "    # Genetic Distance to Root\n",
    "    root_dist = divtree.distance(divtree.root, c)\n",
    "    # Genetic Distance to Population MRCA \n",
    "    mrca = population_mrca_dict[population]\n",
    "    mrca_dist = divtree.distance(mrca, c)\n",
    "\n",
    "    metadata_df.at[sample, \"root_rtt_dist\"] = float(root_dist)\n",
    "    metadata_df.at[sample, \"population_rtt_dist\"] = float(mrca_dist)\n",
    "\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "exceptional-label",
   "metadata": {},
   "source": [
    "## Time Dependency of Rate Variation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ranging-mounting",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.ORI\n",
      "1.IN\n",
      "1.ANT\n",
      "1.PRE\n",
      "2.MED\n",
      "2.ANT\n",
      "4.ANT\n",
      "3.ANT\n",
      "0.ANT4\n",
      "0.ANT\n",
      "0.PE\n",
      "0.PRE\n",
      "y = 0.002x - 6.66e-07x² + 0.0019\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ktmeaton/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWYAAADCCAYAAAB63GhaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAexAAAHsQEGxWGGAAAn+klEQVR4nO3dd3hUZdr48e89k0YCGgkIkST0IqCUgDQBpRiUjoqK0gIEEHeFWFfWfXVf3X1FRX+LGjpIX1FAhUUQAcEVewMWhAXpvUMSQiZ5fn+cSUxCMgNMMiW5P9c1V2bOec557pOEO4fnPEWMMSillPIfNl8HoJRSKj9NzEop5Wc0MSullJ/RxKyUUn7Ga4lZRGqJyAwRWVRg+x0iMltE5olItLfiUUopf+W1xGyM2W2MGVbIrhHAUOD/gML2K6VUmRLk6wAAMcYYEdkLxObbIZIAJLRt23Zcq1atfBOdUkqVgDfeeGOJMebewvb5Q2I2IiJAHHAg3w5jVgGrkpOTx02cONEnwSmlVEl444039ha1z2uJWUSigJeBFiLyNNDQGDMEmAFMB0KAZ7wVj1JK+SuvJWZjzElgVCHb1wJrvRWHUkr5O+0up5RSfkYTs1JK+RlNzEop5WdKXWJ2OBycO3fO12EopdQ184fucsVm5sQlrH37CEFpFbHX28PLi4ZSNbqKr8NSSqmrUmoS8/ff/MJXL0dQ99SjAGQfyeLFoa/zwruDSUtLIzU1lbS0tNxXRkYGWVlZuS+rKzW5X/PKmbPaZrNht9sJCgoiKCiIsLAwwsPDiYiIuOxrcHCw9y5eKVWqlJrE/M36X6h8ql/uZxt2Tu+0sXXrVsLDwwkPD6dixYrExMQQERFBSEgIQUFB2O12bLYra9HJzs4mKysLh8NBZmYmGRkZpKamkpqaypkzZzh48GDuHwGHw4ExBhHJTex53+eIjIykatWqREdHU7lyZYKCSs2PRCl1jUpNFmgUX4dfIn6iampbAAyGG2rZ6NSpU7HVYbPZsNlsBAcHU65cOQAqV658zeczxnD27FmOHDnC9u3b2bBhA1lZWfn259zBV65cmerVqxMXF0dYWJhnF6KU8mulJjF36NyaTaNnsm3WYULORZPR8Cv+9GYvX4flkogQGRlJZGQkDRo0KLKcMYbjx4+zd+9eVqxYQUZGRu7xxhhsNhuxsbHUq1ePSpUqFdoco5QKHKUmMQM882oih8Yd4uTJk9SpMzr3rjbQiQg33ngjN954Iy1btrxsf3Z2Nvv37+enn37i+PHj+faFhobSoEED6tWrp+3eSgWIUpWYAW666SZuuukmX4fhVTabjerVq1O9evXL9l28eJFt27axZMkSHA5H7t10ZGQkt9xyCzExMXqHrZSfKXWJWeUXFhZGs2bNaNasWb7tp06dYsuWLWzcuDH3gWRYWBhNmjShdu3amqyV8iFNzGVUxYoV6dChQ75t6enp/PLLL3zzzTe5ybpcuXLEx8cTFxenyVopL9HErHKVK1eOVq1akXdRgtTUVH744Qc2btyYu61+/fo0bdpU26yVKiGamJVLERERtG/fnvbt2wNWD5EdO3awZMkSMjMzAYiKiqJdu3Zcd911vgxVqVJDE7O6KiJC/fr1qV+/fu62EydOsG7dOs6fPw9AlSpVaNeuHeHh4b4KU6mApolZeaxSpUr07t079/Phw4f55JNPSEtLAyA2NpbWrVsTGhrqqxCVCiiamFWxi46Opl+/34fH79u3j2XLlnHp0iWCgoJo06YNNWrU8F2ASvk5l4lZRBoD3YFQAGPMX70RlCpd4uLiiIuLAyAzM5Ovvvoq92FilSpV6NChgw4zVyoPd3fMa4DDwAXnZ03MyiPBwcH5HiYeOXKE5cuXk56eTnBwMB07diQ6OtrHUSrlW+4S81agjzHmvDeCUWVP1apVue+++wDIyMhgw4YNrFmzBoCWLVu6nENEqdLKXWJuDhwXkWOAMcZcPub3ColIBDAJcADrjDELndsTgU5AOPCiMebna61DBbbQ0FC6du0KWPN/fP/998ydOxeA2rVr07p16yueolWpQOYuMY8txrr6AYuNMStFZBGw0Lm9PTAcaAa0BTQxK2w2Gy1btsydtOm///0vCxYswBijSVqVeu4S81zgbuf7lR7WFQP86HyfnWf7B8ByrAeMD+U9QEQSgIS+fft6WLUKdHXq1KFOnToA7Ny5kwULFpCdnU39+vW57bbbdLi4KlXcJeZ5QF/gElZiftCDug5gJect5F8EdqgxpouIVAOeIs9dujFmFbAqOTl5nAf1qlKmbt261K1bF4Dt27czb948jDE0atSI5s2ba5JWAc9dYr4DqOh8v8vDupYAk0SkN/CxiMw2xgwB/i0iU4Hrgeke1qHKmAYNGuQ+INyyZQtz5szBZrPRsWPH3C56/mrfvj2sXLYAm81GnweGebQajipd3CXmE8AMIB046UlFxphUIDHPpvnO7RM9Oa9SORo3bkzjxo3Jysri888/Z926dURERJCQkECFChV8HV4+O379DytmP8+whCpkZWeT8n+jGPrkJKKjy9Zc4qpw7hLzAODPQITzvVJ+z2635671eOHCBVavXs358+e56aab6NSpE3a73ccRwqolMxjerQoigs1mZ0yPG5k77x0ee+olX4em/ECRiVlEvgGqAjkNdisA//6/oVIFlC9fPnd4+IEDB3IfGnbs2NGnw8Lttux8beF2uw27ZLk4QpUlru6YXwXqYT20g/wP7JQKODExMQwcOJDs7Gw2bNjA+vXriYqKIiEhgZCQEK/GElu3Jd/v+Bfx9a4HYP1Pp2gU39vNUaqscJWYlwC/AdWBEGA/MMsbQSlVkmw2G3fccQcAJ0+e5P333+fSpUu0bt3aayMNe/YbwOIFaby78VuMgdpN7qXDnXd5pW7l/1wl5kzAYI3UA9hb8uEo5V1RUVEMGDAAYwxfffUVs2fPJjIyknvuuafE76LvHzAca2yVUvm5SsydgFeAp4EsYJtXIlLKB0SENm3a0KZNG06dOsXixYtxOBx07dq1zK26rnyvyMRsjFkPtBKRYKCc1yJSyscqVqzIww8/TFZWFp999hmrV6+mTp06tGvXTgevKK9wNx/zq1gj8WxYzRo6sb4qM+x2O3fdZbX77ty5k3fffZfw8HC6d+9ORESEj6NTpZm7RPsgcBQYBSSVfDhKeUd2djavz1nAjpNnqRlZnmeGPOKyf3POMPDU1FRWrFhBenq6NnOoEuMuMac6v3YB2pVwLEp5zYi/T+T9Bp3JrnUTnDzGlr9OYP4Lz7ptqoiIiKB///5kZ2fz6aefsmrVKpo3b06TJk28FLkqC9wl5uewemUkYU0wpFTAO3fuHGuyK5Bd2Xm3G3Uj6yKqcfjw4Su+A7bZbCQkJADw/fffM2vWLGJjY+ncubO2QyuPuUvMO7DW/PsGa2Y4pQKew+EgOyg437bs4BAcDkcRR7gWHx9PfHw8+/bt49133yUiIoKePXvqOobqml3Nmn8GXfNPlQI33HADLc4f5NML55Dy12HSU4k/vJ2YmP4enTcuLo4hQ4Zw7tw5PvjgAwB69erldxMoKf+na/6pMkdEWPiXp3g2ZQb7M7KIDrIx4YWni21FlOuuu46HH36Yixcv8tFHH5Genk6PHj2IiooqlvOr0u9q1vzDGKOTGKlSISwsjDfHjSnxOvr3709mZibLly/n9OnT3H333boKuHLratb8MyUYh1KlVnBwMH379iU7O5uVK1dy5MgROnXqRM2aNX0dmvJTVzLtJ1hTfxpgjjeCUqo0stlsdO/eHWMMa9euZc2aNXTq1InatWv7OjTlgXPnzjHv9RcJvnQee1Q1Bo0dT1CQZ2PxXB09AWuCfKVUMRIROnfuTKdOnVi3bh1r166lS5cuegcdgDIzM5n0xBCeqHmBsPJ2TlzYzf/702M88epkj85b5NMOY8z7wEGgJdbgEh1golQxEhE6derE8OHD2bVrF9OnT2fvXp3EMZBs+89WOpU7RliwNWq0UkQwFU/v4NKlSx6d19399nvA58BFtI1ZqRIhInTp0gVjDJ9++imffvop3bp1IyZGhw74O7s9iMwCmdGRjcc9fNwl5oXAJmCfR7UopdwSEe666y6MMaxevZpVq1Zxzz33aC8OP3Zzw4a8YmrSOO0QFcND2H0mg8waLUq0jRngNmC0870BfL+KpVKlnIiQkJBAdnY2//rXvzhx4gR9+vQhMjLS16GpAmw2G09MnMGCt1/n0tljRNWtz+jBIz0+r7vE7ACqAKc8rUhEIoBJznOuM8YsdG6/GcjpUPqOMeY/ntalVGlgs9no0aMHmZmZLF26lMzMTPr160e5cjo9uj8JCQlhyLg/Fes53SXm+ljNGMex7pire1BXP2CxMWaliCzCaiYBSAZOY03Gf9SD8ytVKgUHB9O/f39SU1N5//33CQ8Pp3fv3h7/d1n5L3c/2XHFWFcM8KPzfXae7S2BO5z7xwLP5+wQkQQgoW/fvsUYhlKBKSIigoEDB3Ly5EnmzJlDdHQ03bp109nsSiF3jw6rGmPezXl5WNcBfp+hLm+9u7EmSToFlM97gDFmlTEmuUaNGh5WrVTpERUVRWJiIo0bN2b69Ol89dVXvg5JFTN3ibmbiNxQTHUtAfqLSArwsYjMdm5/E5iCNaBlSjHVpVSpFxsby4gRIyhXrhxTpkxh9+7dvg5JFRN3TRlNgIMicgI8m8TIGJMKJObZNN+5fQOw4VrPq1RZ16RJE2699VbWrFnD+vXruffee7n++ut9HZbygDfbmJVSJURE6Nq1K5mZmSxZsgSAe++9Vx8QBih3P7V5QFvgJyCkxKNRSnkkODiYBx54gNOnTzNnzhzi4uJ0uasA5K6NeQGwDKub3OclHo1SqljccMMNJCYmUrNmTaZOncq2bdt8HZK6Cu4ScyvgrDFmC1DJC/EopYpR7dq1GTlyJMeOHWPGjBmcO3fO1yGpK+CuKeMEUEtE5qODP5QKWB07dqRNmza5A1R69epVbEtpqeLn7ifzOLAaK4E/UvLhKKVKSkhICAMGDKBVq1ZMnz6dn376ydchqSIUmZhFxI41bPohYBDwmbeCUkqVnOjoaJKSksjIyGDatGmcPHnS1yGpAlw1ZWRizY/hcH7WqT+VKkVatWpFixYtWLp0KcYY+vXrh92uE0j6A1dNGZ2A75xfOwItvBKRUspr7HY79913H3feeSfTp0/n559/9nVICtdLS63HWk4qG9CfllKlWKVKlRg5ciSpqalMnz6dCxcu+DqkMs3dw7/5aD9mpcqMtm3bMmjQIJYuXcqqVat8HU6Zpf2YlVL5hISEMHDgQOrUqUNKSgr79unjJW9zl5hPAJHaj1mpsqd27dqMGjWKzZs3s2DBAhwOh/uDVLFwN8BkKPAD1lp/DxdVSESaAq2BG4AdwCfO2eSUUgFMROjevTtnzpxh1qxZxMfH07x5c1+HVeq56sf8AL8n5X5A4yLKvYmVlH/Gao9OB/4iIl2LOVallI9ERkYyYsQIMjIymDlzJunp6b4OqVRzdcf8AjAEa4L7vsBfgUUFCxljxhbYtA34V7FEp5TyK23atKFZs2YsXLiQ+vXr07ZtW1+HVCq5amOuYIyZb4xJN8YsAMILFhCRF0VktIgkFnK8UqoUCgsLY+jQoYSHhzN16lSdGKkEuLpjriIieR/H3lhImcPACqx5m2cWZ2BKKf/WtGlTGjVqxD//+U+io6Pp3Lmzr0MqNVwl5uFXcPy7xph0EemZs0FEKhljTngemlLK3wUHB/PII4+wfft2UlJSuP/++6lUSXvWeqrIxHwlq2I7k3IX4B4RmQo8AMQBw4ovRKWUv2vQoAH16tVj8eLFREZGkpCQ4OuQAlpxTMg6CmuF6/eBpcYYTcpKlUE2m40HHniAWrVqkZKSwunTp30dUsBy1V2upoiUv4JzHAVOAeuB7SJS6NqAIhIhIjNFZKqIPFRgX4KIbL+KuJVSfqpu3bokJSWxcuVK1q1b5+twApKrO+aNwFax2ESkqLKVgMlAhPPrZAARCStQrh+w2BiTBPTO2SgiN2DNXvfTNV2BUsrv2O12BgwYQJUqVZgyZYr23LhKrhKzHYjFmo/ZgTU/c2EOYd0xfw58DGwRkRlAywLlYoD9zvfZebY/B7xa2Imdd9IT9+zZ4yJMpZS/atiwIYmJiSxbtowvvvjC1+EEDFeJuS3wLdZ8zHc6v17GGDMO+BvWRPo24FNjzDBjzMYCRQ9gJefcekUkAmgI/B1oISKDC5x7lTEmuUaNGldzTUopPxIcHMygQYMoX74806ZNIy0tzdch+T1XvTJ+E5G2QDfnppUuyp4C1rqpawkwSUR6Ax+LyGxjzBCgO4CILLqSniBKqcDUtGlTbr75ZubPn0+TJk2Ij4/3dUh+y12vjLlYvS0WOF/XzBiTaoxJNMaMdo4oHFJg/4OenF8p5f9CQ0NJTEwkIyODuXPnkpWV5euQ/JK7xHwHUBGoCnQo8WiUUmVC27Zt6dmzJ1OmTEGfIV3uSuZjngG8BehSukqpYhMZGcno0aPZvHkzH3/8McYYX4fkN9wl5gHOMuVxMR+zUkpdCxGhZ8+eNG7cmJSUFM6cOePrkPyCy4nynUtKaduvUqpE1axZk6SkJBYuXEjt2rXL/HSiLu+Ynb0yct7XLPlwlFJlVVBQEAMHDiQkJISZM2dy6dIlX4fkM66GZNuBRc6Rf6HA194LSylVVrVo0YL+/fszY8YMdu/e7etwfMLVHXMmUA1r1F8aoGv4KaW8onz58owePZqtW7eycmWRQyhKLVeJuRPwnfNrR6CFVyJSSimnnj17UqtWLaZMmVKm1hksMjEbY9ZjTZbfGis5j/FSTEoplat+/foMHjyYuXPnsn172ZiE0l13uU+xemV0db6UUsrrwsLCSEpKYu/evSxbtqzU93l2l5i3Ah2MMe2NMe29EZBSShUlISGBpk2bkpKSwvnz530dTolxl5ibA8dFZF+BhVmVUsonatSowYgRI1i8eDE///yzr8MpEe4S81hgJPC886WUUj4XHBxMYmIip06dYsmSJaWuacNdYp4H7MaasnN5yYejlFJX7s477yQ+Pp6UlJRSNc+zu8S8AFgGVMdaoUQppfxK9erVGTZsGHPnzmXnzp2+DqdYuEvMrYCzzjkzKnkhHqWUumqhoaGMHDmSHTt28Mknn/g6HI9dybSfkSIyH2s1bKWU8lvdu3cnLi6OadOmkZlZ1DKl/s9dYh4MrMJamFWn/VRK+b2GDRvy4IMPMmXKFA4ePOjrcK6Jq0mMvsFa5+92oB3wL28FpZRSnqhQoQJjxozhyy+/ZOPGgutC+z9X8zFPACK8FYhSShUnEeH+++/n22+/Zf78+QwYMAAR8XVYV8TVXBnvA3WwFmNdDNT1VlBKKVVcWrZsSefOnXn77bdJTQ2MSTLdtTEnOle3TgMSvRGQUkoVt6pVq5KUlMS8efPYtWuXr8Nxy11iPiIi/xSRhXjYK0NEIkRkpohMFZGH8mx/RUSmi8hiESnvSR1KKVWUkJAQRo4cyebNm1m/fr2vw3HJXWJ+BMgGxPneE/2AxcaYJKB3zkZjzDPGmOHAl8AtHtahlFIu9enThwoVKjBv3jy/HcrtMjEbY7YZYx4yxjxojNnqYV0xwH7n++y8O0SkCtaESd8U2J4gIhP37NnjYdVKKfW7+Ph4unbtyltvvcWFCxd8Hc5lXHaXy5lVrphmlzuAlZzz1Ssi0cCrwGPGmKy8BxhjVhljkmvUqOFh1UoplV+VKlUYNWoUCxYs8Luh3K66y70PpAPniqmuJcAkEekNfCwis40xQ7Dm4vgNeEVEUowxpXMeP6WU3wkODiYpKYmPPvqIgwcPcscdd/g6JMB1Yv4jkAXUwGpj9ogxJpX8PTvmO7e38vTcSinliV69evHdd9+xaNEiHnjgAZ/3d3bVxmwHYrFWyc50vpRSqlRq0aIFt99+OykpKWRkZPg0FleJuR3wI9ZCrJ2dX5VSqtSKiYlh8ODBTJ06lWPHjvksDleJ+XOgMrDB+Qq8AedKKXWVIiIiGDNmDCtXrmTz5s0+icFVYg4CqqFNGUqpMsZmszF48GAOHjzImjVrvF+/i31tge+wmjByXkopVWZ069aNqKgorw9GcTWJ0W9Y7czZwA/AFm8FpZRS/qJZs2a5kyBdvHjRK3W6G5I9H13zTylVxkVHRzNs2DCmT5/O0aMlv5iTq37MkGfNPxEJmDX/HA4H27dv93UYyksaNGhAUJC7X2WlPFOuXDnGjBnDnDlzaNmyJQ0bNiyxukrlmn/bt2/3uyGWqmTs3LlT/wgrrxERBg8ezK5du0p0ZRR3txmDsZowugEdSyyKElC3bl0aN27s6zCUUqVQz5492bRpE8uWLaNPnz7Ffn53d8x/xlpeKsj5XimlFNCmTRsaNmzIzJkzyc7Odn/AVXB3x9wRqOh87//T/qsScfToUZYsWcL58+d5+umnfR2OUl516tQp3l40GYcti5qVqjP43oG5c2l8+uP3rD51hGnPjGNg81Y8+tCAYqnTXWI+AczAmmXuZLHUGGAWLVrE2rVruXTpEu+88w7h4eEAGGMYM2YMNpuN2rVrM27cuMvKLlq0iE2bNnH06FFefvllbrnl6tYB+Oijj1i9ejV2u51XXnmFsLAwALKzs3n++ec5e/Ys8fHxDB06tMiyV3JNa9euZfny5Zw5cwYRYeHChfnKV6lShTZt2rBt27ZCz7dixQq6d+9+2fb169cze/ZsHA4Hr776KtHR0bn7xo8fz/nz5ylfvjx/+9vfLiubnp7Oyy+/TGpqKosWLSItLY0RI0Zw3XXXERsby3PPPXdV30ulrkVaWhrPpPyZpk+0xR4cxIEdR5k4402eGD6Oecs/Ym5kMPY7+mEDpv/7G0IXv8ew+/t7XK+7powBzjLlne8D0rhx4zh06BDp6ekkJl7d0oUfffQRU6dO5b777mPJkiW527/44gtuueUW3nrrLX788UcyMzMvK5uYmMi0adN48cUXWb58OXv27CEpKYlDhw4xePBgl/VmZWWRkpJCUFAQkZGRhIaG5u778MMPOXjwIMYYqlWrVmjZNWvWMHbsWJKSkvj3v//t8pp69OjB5MmTadOmDUOGDLkszt27d5OcnFxkz4fFixcXun3atGnMmjWLZ599lhkzZuRu379/Pw6Hg3/84x9kZmayf//+y8rWqlUr3zHbtm2jUaNGpKSkcPToUQ4cOODy+6eUp86dO8fTjzxAjQdvxh5s/e5XqleFPZkHAdi4Zxf2pr8/xyrX7jbe+3Ijv/76q8d1u5oo/yWsrnIPAk8BD3pcm488+uijTJ06lQULFvDQQw/l2/fCCy8wduzY3Ne6desKPUf16tXZv39/7ucDBw4QGxsLQOXKlTlx4kShZR0OB5MmTeKRRx6hRo0atGvXjj59+jBhwgSXMR8/fpwzZ87w5ptvUrlyZdauXZu779dff6V169a89dZbTJ48udCykyZNIjIykqpVq/L1119f0TV99tlndO3a9bI4a9Wqxdq1a7n//vvzHf/aa68xatQoNm3axKhRo5g1a1a+/cYYRMTl9y4uLo4DBw4UWTZHs2bNSEtLIzk5mX379nHw4EGX3z+lPJGZmcmkHj24c81nl+3LmRA02JGNydO2bLKyuLlmTbZv387PP3s2rbyrpoxBxpg/Axhj9onIYGC8R7X5SN26dTl48CDbtm1j0aJF+fZlZGTkG82TlZVV8HAA9u3bR0xMTO7nmJgYtm61Vts6fvw4UVFRl5XNzMzkj3/8I+PGjctNRPv376dixYqkpaXllnc4HDz55JPExcWRnJwMQMWKFalWrVru+7zL38TExBASEoKIEBQUVGhZYwzPP/88drsdgDlz5vDDDz/w7LPPFnpNX3/9NS1atMBmsxUZZ0FPPvkkAEOGDGHy5MmX7RcRjDGXfe+qVavGhx9+mFtPnz59iiybw2az8dJLLwGQmJhIzZo1i4xLKU9t/ukn7vrySxpnZfHJ47O57r2xBIcGc+w/h6kZGgfAk/f2Z+iMaVwcfB8YKDf7PZ4a/ijVbrrJ4+HbrhKziEhNY8xvIhLw/wo6duyY24aa19///neXx/Xo0YNHH32UtLQ03n77bRYvXkxERAR33303Cxcu5PHHH6dJkyaEhIRcVnb8+PFs3bqVt99+m86dO1O9enXS09NZuHBh7pI2drudoKAg3nzzzXz1hoSE0KFDBx5//HHOnj3L5MmTmThxIl26dKFfv3784Q9/YOPGjbRv377QsmFhYQwfPpzIyEg6duzIoEGDGDRoUKHXBDB79mzGj7f+7n7zzTeFxnm1hg0bxvDhw7l06RKvvPIKmzZt4pdffmHkyJHYbDaSk5MJDQ0lNjb2srInT55k/PjxfPfdd0yYMIGnn36a0aNH43A4aNGiBTfeeONVx6PUlbIHBXHRZqNcVhavrfiBt1qNZ1tsFRKGj+WRoVarbo3YWBaOGMOMD5dyMS2dx0b+gSpVqgB4PtG+MabQF9Y6fGeBTc6vrxVVtqRf48aNM1dj8+bNZvPmzbmfN2zYYPr3728uXLhwVedR/q/gz1qp4uBwOMzL3bqZM2AMmF3BweadMWMuK7f+o6XmuXZNzDu31TavJbQx332+/orrACaaInKeqzvmp4FfgaZYPTNmevYnwHfat29P+/btfR2GUsrP/bpjGx+vspo7O/z5WT5o2hTH4cNUbNKEUWPH5it74sQJnk+eTnDk/yBkU213Co6Xn6XZ7f/ObRK8VkUmZmdGn+7R2X1Ih2SXDTt37qRu3bq+DkOVAjt2bmfe8lfpNqABIsIXKxbRtueD3N628EHPL/zPdK6vMY/goAoAnLz+Drbu68GJEyc8bmorlTO/NGjQwNchKC+pW7eu/rxVsfh41Xu5SRng9u612LDok3yJ+df/7ub12SvIJIijv6US5kzKACHBkeyiCs8sfIusUDs3XgrmpaRkl2MKiuK1xCwiEcAkrBVR1hljFjq33wEMccbylDHmsKd1BQUF6TwZSqmrInL5Qzux/f55/4GDPPLSUk41HIWIcCl9ArfYjhIaYj3wu3hxP+mNKhD2WBcATp9P47kprzPx8avvzOZZQ8jV6QcsNsYkAb3zbB8BDAX+DxjmxXiUUipXl469WLfk9ybQb9ftpXnD359NTV24nFM3J+Um7+CEZPaee4FzaXM5c2EWmaETuPWd3/v6h1QI51iY45pi8WZTRgzWqttgrYqSQ4wxRkT2ArF5DxCRBCChb9++XgpRKVVW3dq4KZmXRrBm0VJsdhvNG/Wk850JufttNhtk/d4/WexB1L79Vha8ZCXjdV/GMP/YccpFReaWsV8qfFyEO968Yz6AlZwL1mvE+hMU5yzz+w5jVhljkmvUqOGdCJVSZVp885Y88/jfeOqxl/IlZYDRj/Si0vZ3ckf7BR/YQN/W1YmIiCAiIoJ7Ot/F9cu3c373IbIuZXLygy+5v/Ht1xSHGC8tMJinjTkD+ALoaowZIiKdgIeBEOAZY8yhgscmJyebiRMneiVOpZQqyr79B5gwYxmZxk5C6/r0uzv/GtXZ2dksXvEhh04c4662HWlUv+gH0yLyhjEmudB93krMntDErJQqbVwlZm82ZSillLoCmpiVUsrPaGJWSik/o4lZKaX8jCZmpZTyM5qYlVLKz2hiVkopP6OJWSml/IwmZqWU8jOamJVSys8ExJBsEfkA2FuCVVQv4fOXNI3fdwI5dtD4fek2Y0yhsxwFxAomxph7S/L8IjKxqDHrgUDj951Ajh00fl8SkSInANKmDMsqXwfgIY3fdwI5dtD4fanI2AOiKUMppcoSvWNWSik/ExBtzJ4SkVrAeCACGAn8P6w/SkeNMU+JSEPgT85tLwPbgLexlsDaZYx5wyeBO+WN3xjzoIiMBeoAduBR4Gb8OP4cIhKHtVjCSeBXrEUTirwOY8x/fBTqZUTEBvwvcD3wvfNrQMSew7lYxQas36UGBFD8ItIL6AlUxvoduoUAir+gohanzmWMKTMvYFGBz/8EBJgCVMD6xzYFaA+MdpaZAwT7Ovac+LFWepnt/PwYcHsAxd8VSHS+X+DuOnwdb4HY+wKzsf7g3RVIsee5hr8CTwO9AjF+Z7w3ALMCNf481zEQuNv5flHB/WW2KUNE2gPbjfWdqWCMOW+MOYv1w40B9juLHgcq+SjMwkQBJ5zvcxawDZT4fwAeEpF/AVtxfx3+pD7wFVYi+BOBFTsichewBTiG9T/HgIo/j/HAdAI3/hx5/41mF9xZJpoyCnIm5V5Ydw8A50WkAtbd83msRWEbOfdVxvqvt784ye+JNg74hcCJfyjwF2PMJhFZCpx1bi/qOvzJAeCSMcaIyGnc/wz8TScgEusPzEXgqHN7oMSPiPwNWAl8C4xwbg6Y+AvIWZx6C4U86ysTvTJEJAqr7bULMBV4AlgGGGAcUBt4CuuHOgHrbu5tIBPYZ4x53ftR/66Q+B1Yv5BhwGisJOy38ecQkcbAX4BTwAXgEC6uwxizxUehXkZEwrHaBNOAHUAwARJ7XiIyBDgCNCSA4heRR7GS8dfAT0A4ARR/QQUXpzbGzM+3vywkZqWUCiRlto1ZKaX8lSZmpZTyM5qYlVLKz2hiVkopP6OJWSml/IwmZuVVIlJXRDaLyBER+c053NzTcw4RkQPO9/eLyLfXeJ67RcSIyEL3pV3G4hCRY87ru+1az6XKLk3Mytt6Yw1saup8HRaRe0XkVxE5ICKvO5PbaRH5WkS+E5F3nJ/vyrNvnYgcEpFWBc4fAUQ7y50RkbUiclREbhWRVs5j1ojIKWef3ryGAKuBPiISCVDYMSLS2BnXXhF5T0SCC5znCFAFq795q4LX5zzvFcV3BXWpUkgTs/K2D7CG0x4AvsBagcKBNTDABowFQrEGEPQC4rGGQm8A+jjPEY6V4D8HnnRRVzjWHBeHge5YAxA2APcB5fMWdCbiXsAzWANIHnDuKuyYP2NNoGOw5v9oW6DeqsBprMEPcwten4iEXkV87upSpZAmZuVtgpWEbsCa96MHkAR8gjXznw0ohzXHR85Q8n1AOtYETjmuZGTUMefcCRewkr3BGrlX2LEPOc+/HqiHdfdMEcfYgJXGmBpAB6yRaHkdwfqDkw0ML+L6rjQ+d3WpUqhMzpWhfKoZ8AbW3eRRrKHxZ4BXgZpXeI404EOsqR77Yc3/cCVeA5Y668ly1ptjMPC/xpgXRKQ6sEdEGhRxzEvAPBE5ivXHo2PBiowxZ51LBz0DvAC8cgXXd011qdJHh2SrgOJsF37JGBNzDcc2x5pDJAr4ERhkjMko7mOulTfrUv5NE7NSSvkZbWNWSik/o4lZKaX8jCZmpZTyM5qYVZkgIm86B2jYRKS+iPzjKo6VAp+biMjQ4o9SKYt2l1OlnnMFmCysteKeAJoDzzqTswCbgXnAc1iLef5gjJklIhuBj4HPReRJrDXalhljNojIH7EWBVWq2OkdsyoLmgJbjDGrgcZYo/EeBVKx+gbfijWow4Y1Yu8h53EXjDETsJbAygbeBzY69zmcywMpVew0Mauy4AZ+X/j1N2A31u/+XGPMC8aYx4B7gM3GmL9gredHzjHGmP3AH7AGx/wlz75Ir0SvyhxtylBlwQ6seSbyegt4WUSOAMew7ob/LiJxgD1vQRG5BWsh0DCsSY7AWn38UEkGrcouHWCiSj3nw7uZxphieWDnXDH7r8YYVxMoKXXNNDGrMkFEOgLfGmPSiuFcdYEsY8xuzyNT6nKamJVSys/owz+llPIzmpiVUsrP/H95ZwUdQ9V+wAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 400x200 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "DPI = 50\n",
    "FIGSIZE = [8,4]\n",
    "\n",
    "fig, axes = plt.subplots(1, figsize=FIGSIZE, dpi=DPI)\n",
    "ax = axes\n",
    "\n",
    "x = []\n",
    "y = []\n",
    "c_list = []\n",
    "\n",
    "for population in population_list:\n",
    "    print(population)\n",
    "    \n",
    "    # Prep pop data frame\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    \n",
    "    drop_sample = \"SAMEA104488961\"\n",
    "    if drop_sample in df.index:\n",
    "        df.drop(drop_sample, inplace=True)    \n",
    "        \n",
    "    # Perform regression\n",
    "    dates = list(df[\"date_mean\"])\n",
    "    dists = list(df[\"population_rtt_dist\"])        \n",
    "    slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(dates, dists)\n",
    "    r_squared = r_value * r_value\n",
    "    \n",
    "    color = df[\"population_color\"][0]\n",
    "    time_frame = time_frame_dict[population]\n",
    "\n",
    "\n",
    "    x.append(time_frame)\n",
    "    y.append(r_squared)\n",
    "    c_list.append(color)\n",
    "\n",
    "# Fit an exponential model\n",
    "def func_exp(x, a, b, c):\n",
    "    return a * x + b * x**2 + c\n",
    "popt, _ = scipy.optimize.curve_fit(func_exp, x, y)\n",
    "\n",
    "# summarize the parameter values\n",
    "a, b, c = popt\n",
    "op_1 = \"+\"\n",
    "if b < 0: op_1 = \"-\"\n",
    "op_2 = \"+\"\n",
    "if c < 0: op_2 = \"-\"\n",
    "formula = \"y = {:.3f}x {} {:.2e}x² {} {:.4f}\".format(a,op_1,abs(b),op_2,abs(c))\n",
    "print(formula)\n",
    "# Create points for the line\n",
    "x_line = np.arange(min(x), max(x), 1)\n",
    "y_line = func_exp(x_line, a, b, c)\n",
    "  \n",
    "ax.plot(x_line, y_line, ls=\"-\", lw=0.5, c=\"black\", zorder=1)\n",
    "    \n",
    "sns.scatterplot(\n",
    "        ax=ax,\n",
    "        x=x,\n",
    "        y=y,  \n",
    "        s=50,\n",
    "        c = c_list, \n",
    "        ec = \"black\",\n",
    "        zorder=2,\n",
    "    )\n",
    "\n",
    "\n",
    "ax.set_xlim(1300,0)\n",
    "ax.set_ylim(-0.1,1)\n",
    "ax.set_xlabel(r\"$\\bf{Sampling\\ Age\\ Range}$\" + \"\\n(Years)\")\n",
    "ax.set_ylabel(r\"$\\bf{Coefficient\\ of\\ Determination}$\" + \"\\n(R²)\")\n",
    "\n",
    "# Use the stats from bootstrapping!\n",
    "ax.annotate(\n",
    "    (formula),\n",
    "    xy=(0.05, 0.15),\n",
    "    xycoords=\"axes fraction\",\n",
    "    #size=5,\n",
    "    ha=\"left\",\n",
    "    va=\"center\",\n",
    "    bbox=dict(fc=\"w\", lw=0.5),\n",
    ")\n",
    "\n",
    "out_path = os.path.join(out_dir, \"rtt_curve_fit\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "certified-aging",
   "metadata": {},
   "source": [
    "---\n",
    "# External Branch Lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acquired-shakespeare",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ----------------\n",
    "# CI of external branch lengths\n",
    "\n",
    "data = list(metadata_df[\"external_branch_length\"])\n",
    "data_arr = np.array(data)\n",
    "confidence = 90\n",
    "tails = 1\n",
    "peak, ci = calc_peak_ci(data=data, confidence=confidence, tails=tails)\n",
    "branch_len_thresh = ci[1]\n",
    "print(branch_len_thresh)\n",
    "\n",
    "# -----------------------------\n",
    "# Long Branch Lengths\n",
    "long_branch_dict = {pop: {\"total_branches\" : 0, \"long_branches\" : 0, \"perc\": 0,} for pop in population_list}\n",
    "\n",
    "\n",
    "for rec in metadata_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    population = rec[1][\"population\"]\n",
    "    length = rec[1][\"external_branch_length\"]\n",
    "    long_branch_dict[population][\"total_branches\"] += 1\n",
    "    if length >= branch_len_thresh:\n",
    "        long_branch_dict[population][\"long_branches\"] += 1\n",
    "\n",
    "for pop in long_branch_dict:\n",
    "    long_branch_dict[pop][\"perc\"] = (long_branch_dict[pop][\"long_branches\"] / long_branch_dict[pop][\"total_branches\"]) * 100\n",
    "\n",
    "df = pd.DataFrame.from_dict(long_branch_dict, orient=\"index\")\n",
    "df[\"population\"] = df.index\n",
    "\n",
    "FONTSIZE=14\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "DPI=50\n",
    "FIGSIZE=[12,6]\n",
    "## All\n",
    "fig, axes = plt.subplots(1,2,figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.5)\n",
    "ax = axes[0]\n",
    "sns.histplot(\n",
    "    ax=ax,\n",
    "    data=metadata_df,\n",
    "    x=\"external_branch_length\",\n",
    "    kde=True,\n",
    "    bins=50,\n",
    "    color=\"grey\",\n",
    ")\n",
    "\n",
    "ax.axvline(branch_len_thresh, color=\"black\", ls=\"--\", label=\"Threshold\\n{:.2e}\".format(branch_len_thresh))\n",
    "ax.axvline(0, color=\"black\", ls=\"--\")\n",
    "ax.set_ylabel(\"Number of External Branches\")\n",
    "ax.set_xlabel(\"External Branch Length\\n(Substitutions/Site)\")\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(1e-5))\n",
    "ax.legend(fontsize=FONTSIZE-2, edgecolor=\"black\")\n",
    "\n",
    "ax = axes[1]\n",
    "\n",
    "\n",
    "sns.barplot(\n",
    "    ax=ax,\n",
    "    data=df,\n",
    "    #x=\"perc\",\n",
    "    x=\"long_branches\",\n",
    "    y=\"population\",\n",
    "    palette=colors_dict,\n",
    "    ec=\"black\",\n",
    ")\n",
    "\n",
    "ax.set_ylabel(\"Population\")\n",
    "ax.set_xlabel(\"Number of External Branches\\nLonger Than Threshold\")\n",
    "\n",
    "out_path = os.path.join(out_dir, \"long_branches\")\n",
    "print(out_path)\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fallen-impact",
   "metadata": {},
   "source": [
    "---\n",
    "# Tree Shapes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adequate-raise",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path = os.path.join(out_dir, \"branch_lengths.tsv\")\n",
    "print(out_path)\n",
    "dpi=200\n",
    "figsize=[6,3]\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    header = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\".format(\"population\",\"total_length\", \"mean_dist_to_root\", \"ratio\", \"internal_prop\", \"external_prop\",)\n",
    "    print(header)\n",
    "    outfile.write(header + \"\\n\")\n",
    "\n",
    "\n",
    "    for population in population_list:\n",
    "        #if population != \"1.ORI\": continue\n",
    "        out_img  =os.path.join(out_dir, \"branch_lengths_{}\".format(population))\n",
    "            \n",
    "        # Extract subtree\n",
    "        df = metadata_df[metadata_df[\"population\"] == population]\n",
    "        #if population == \"0.ANT\":\n",
    "        #    df = metadata_df[(metadata_df[\"population\"] == \"0.ANT\") | (metadata_df[\"population\"] == \"0.ANT4\")]\n",
    "        color = df[\"population_color\"][0]\n",
    "        tips = list(df.index)\n",
    "        divtree_copy = copy.deepcopy(divtree)\n",
    "        subtree = extract_subtree(\n",
    "            tree=divtree_copy, \n",
    "            tips=tips, \n",
    "            df=metadata_df, \n",
    "            color=color)\n",
    "        \n",
    "        \n",
    "        # Drop RT5\n",
    "        for t in subtree.get_terminals():\n",
    "            if t.name == \"SAMEA104488961\":\n",
    "                subtree.prune(\"SAMEA104488961\")\n",
    "                df.drop(\"SAMEA104488961\", inplace=True)\n",
    "        subtree.ladderize(reverse=True)\n",
    "        \n",
    "        # Identify long branches\n",
    "        tip_colors = []\n",
    "        for c in subtree.find_clades():\n",
    "\n",
    "            # Condition 1 : terminal and long branch\n",
    "            if c.is_terminal():\n",
    "                tip_color = color\n",
    "                if c.branch_length >= branch_len_thresh:\n",
    "                    # Condition 2: parent is not the root\n",
    "                    c_path = [subtree.root] + subtree.get_path(c)\n",
    "                    if len(c_path) > 2:\n",
    "                        c.color = \"#c4c4c4\"\n",
    "                        tip_color = \"#c4c4c4\"\n",
    "                tip_colors.append(tip_color)\n",
    "            \n",
    "            # Condition 3: internal and all tips are long\n",
    "            else:\n",
    "                prune = True\n",
    "                for c_term in c.get_terminals():\n",
    "                    if c_term.branch_length < branch_len_thresh:\n",
    "                        prune = False\n",
    "                if prune:\n",
    "                    c.color = \"#c4c4c4\"\n",
    "             \n",
    "            \n",
    "        \n",
    "        # Branch Length Stats\n",
    "        internal_lengths = []\n",
    "        external_lengths = []\n",
    "        dist_to_root = []\n",
    "\n",
    "        for c in subtree.find_clades():\n",
    "            if c == subtree.root:\n",
    "                c.branch_length = 0\n",
    "                continue\n",
    "            if c.is_terminal():\n",
    "                external_lengths.append(c.branch_length)\n",
    "            else:\n",
    "                internal_lengths.append(c.branch_length)\n",
    "            dist_to_root.append(subtree.distance(c))\n",
    "            \n",
    "        mean_dist_to_root = sum(dist_to_root) / len(dist_to_root)\n",
    "\n",
    "        internal_length = sum(internal_lengths)\n",
    "        external_length = sum(external_lengths)\n",
    "        \n",
    "        total_length = internal_length + external_length\n",
    "        internal_prop = round(internal_length / total_length, 2)\n",
    "        external_prop = round(external_length / total_length, 2)\n",
    "        ratio = round(internal_length / external_length, 2)\n",
    "        line = \"{}\\t{:.2e}\\t{:.2e}\\t{}\\t{}\\t{}\".format(population, total_length,  mean_dist_to_root, ratio, internal_prop, external_prop)\n",
    "        print(line)\n",
    "        outfile.write(line + \"\\n\")\n",
    "\n",
    "        \n",
    "        #-----------------------------\n",
    "        # Plotting\n",
    "        \n",
    "        g = sns.JointGrid(       \n",
    "                x=dist_to_root,\n",
    "                space=0,\n",
    "            )\n",
    "        g.fig.suptitle(population, x=0.45, y=1.1, fontweight=\"bold\")\n",
    "        g.fig.set_figheight(figsize[1])\n",
    "        g.fig.set_figwidth(figsize[0])\n",
    "        g.fig.set_dpi(dpi)\n",
    "        \n",
    "        # Histogram\n",
    "        ax=g.ax_marg_x\n",
    "        sns.kdeplot(\n",
    "            ax=ax,\n",
    "            x=dist_to_root,\n",
    "            fill=True,\n",
    "            color=color,\n",
    "            label=population,\n",
    "            #cut=0,\n",
    "            clip=[0,max(dist_to_root)]\n",
    "        )\n",
    "        \n",
    "        \n",
    "        # Tree\n",
    "        ax = g.ax_joint\n",
    "        plt.rcParams['lines.linewidth'] = 1         \n",
    "        Phylo.draw(do_show=False,tree=subtree, label_func = lambda x: \"\", axes=ax, show_confidence=False,)\n",
    "        \n",
    "        ylim = ax.get_ylim()\n",
    "        ylim_range = ylim[0] - ylim[1]\n",
    "        xlim = ax.get_xlim()\n",
    "        xlim_range = xlim[1] - xlim[0]\n",
    "        xbuff = xlim_range * 0.01\n",
    "        \n",
    "        # Scatter     \n",
    "        x = []\n",
    "        y = []\n",
    "        labels = []\n",
    "        y_i = 1\n",
    "        for t in subtree.get_terminals():\n",
    "            x.append(subtree.distance(t))\n",
    "            y.append(y_i)\n",
    "            y_i += 1\n",
    "            labels.append(round(metadata_df[\"date_mean\"][t.name]))\n",
    "        \n",
    "        sns.scatterplot(ax=ax, x=x, y=y, s=10, c=tip_colors, edgecolor=\"black\", zorder=2)\n",
    "        \n",
    "        # Critical Window\n",
    "        #g.ax_joint.axvline(x=0, ls=\"--\", color=\"black\", lw=0.5)\n",
    "        #g.ax_marg_x.axvline(x=0, ls=\"--\", color=\"black\", lw=0.5)\n",
    "        \n",
    "        #g.ax_joint.axvline(x=1e-5, ls=\"--\", color=\"black\", lw=0.5)\n",
    "        #g.ax_marg_x.axvline(x=1e-5, ls=\"--\", color=\"black\", lw=0.5)\n",
    "        \n",
    "        #g.ax_joint.axvspan(xmin=0, xmax=1e-5, edgecolor=\"none\", facecolor=\"#c4c4c4\", alpha=0.25,)\n",
    "        #g.ax_marg_x.axvspan(xmin=0, xmax=1e-5, edgecolor=\"none\", facecolor=\"#c4c4c4\", alpha=0.25)\n",
    "\n",
    "        # General\n",
    "        ax.set_ylabel(\"\")\n",
    "        ax.set_yticks([])\n",
    "        ax.set_xlabel(\"Distance to the MRCA\\n(Substitutions/Site)\")\n",
    "        ax.set_ylim(ylim[0] + (ylim_range * 0.1), ylim[1] - (ylim_range * 0.1))\n",
    "        ax.set_xlim(0 - (xlim_range* 0.1),)\n",
    "        ax.ticklabel_format(axis=\"x\", style=\"sci\", scilimits=(0,0)) \n",
    "\n",
    "        # Special\n",
    "        if xlim_range < 1e-4:\n",
    "            ax.set_xlim(-0.5e-5, 7e-5)\n",
    "        else:\n",
    "            ax.set_xlim(-0.5e-5, 2e-4)      \n",
    "            \n",
    "        #if population == \"0.PRE\" or population == \"1.PRE\" or population == \"2.MED\":\n",
    "        #    ax.set_xlim(-0.5e-5, 2e-4) \n",
    "\n",
    "        plt.savefig(out_img + \".png\", bbox_inches=\"tight\")\n",
    "        plt.savefig(out_img + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "raised-pastor",
   "metadata": {},
   "source": [
    "---\n",
    "# 0. Variant QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "outside-interpretation",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_list = list(colors_dict.keys())\n",
    "population_list.remove(NO_DATA_CHAR)\n",
    "population_list.reverse()\n",
    "\n",
    "population_colors = []\n",
    "\n",
    "for population in population_list:\n",
    "    population_colors.append(colors_dict[population])\n",
    "    \n",
    "ancient_df = df[df[\"biosample_comment\"].str.contains(\"Ancient\")]\n",
    "modern_df = df[df[\"biosample_comment\"].str.contains(\"Modern\")]\n",
    "\n",
    "variables = [\"tstv\", \"ns_ss_ratio\", \"other_var_ratio\"]\n",
    "xlabels = [\"TsTv Ratio\", \"NS / SS Ratio\", \"Other Variants Ratio\"]\n",
    "tails = [2, 2, 1]\n",
    "confidence = 95\n",
    "\n",
    "ylab=\"External Branch Length\\n(Substitutions/Site)\"\n",
    "y=\"external_branch_length\"\n",
    "df = metadata_df[metadata_df[variables[0]] != NO_DATA_CHAR]\n",
    "#df = df[(df[\"population\"] == \"0.ANT\")]\n",
    "#df = df[(df[\"external_branch_length\"] >= 1e-5)]\n",
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "FONTSIZE=14\n",
    "plt.rc('font', size=FONTSIZE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "romance-parish",
   "metadata": {},
   "source": [
    "## Elbow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "immediate-gravity",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", None, \"display.max_columns\", None)\n",
    "\n",
    "display(df[[\"strain\", \"tstv\",  \"ns_ss_ratio\", \"other_var_ratio\", \"external_branch_length\"]].sort_values(\"other_var_ratio\"))\n",
    "\n",
    "#pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", None)\n",
    "\n",
    "\n",
    "\"\"\"vals = range(50, 160, 5)\n",
    "\n",
    "header = \"{}\\t{}\\t{}\".format(\"val\", \"long\", \"short\")\n",
    "long = []\n",
    "short = []\n",
    "print(header)\n",
    "for val in vals:\n",
    "    val = val /100\n",
    "    sub_df = df[df[x] <= val]\n",
    "    long_branches = len(sub_df[sub_df[\"external_branch_length\"] >= 1e-5])\n",
    "    short_branches = len(sub_df) - long_branches\n",
    "    line = \"{}\\t{}\\t{}\".format(val, long_branches, short_branches)\n",
    "    long.append(long_branches)\n",
    "    short.append(short_branches)\n",
    "    print(line)\n",
    "    \n",
    "sns.scatterplot(x=[v / 100 for v in vals], y=short, label=\"short\")\n",
    "sns.scatterplot(x=[v / 100 for v in vals], y=long, label=\"long\")\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "equal-iraqi",
   "metadata": {},
   "source": [
    "## Branch Lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "tropical-welding",
   "metadata": {},
   "outputs": [],
   "source": [
    "long_branch_dict = {pop: {\"total_branches\" : 0, \"long_branches\" : 0, \"perc\": 0,} for pop in population_list}\n",
    "threshold = 1e-5\n",
    "\n",
    "for rec in metadata_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    population = rec[1][\"population\"]\n",
    "    length = rec[1][\"external_branch_length\"]\n",
    "    long_branch_dict[population][\"total_branches\"] += 1\n",
    "    if length >= threshold:\n",
    "        long_branch_dict[population][\"long_branches\"] += 1\n",
    "\n",
    "for pop in long_branch_dict:\n",
    "    long_branch_dict[pop][\"perc\"] = (long_branch_dict[pop][\"long_branches\"] / long_branch_dict[pop][\"total_branches\"]) * 100\n",
    "\n",
    "df = pd.DataFrame.from_dict(long_branch_dict, orient=\"index\")\n",
    "df[\"population\"] = df.index\n",
    "\n",
    "FONTSIZE=14\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "DPI=50\n",
    "FIGSIZE=[12,6]\n",
    "## All\n",
    "fig, axes = plt.subplots(1,2,figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.5)\n",
    "ax = axes[0]\n",
    "sns.histplot(\n",
    "    ax=ax,\n",
    "    data=metadata_df,\n",
    "    x=\"external_branch_length\",\n",
    "    kde=True,\n",
    "    bins=50,\n",
    "    color=\"grey\",\n",
    ")\n",
    "\n",
    "ax.axvline(1e-5, color=\"black\", ls=\"--\", label=\"Threshold\")\n",
    "ax.set_ylabel(\"Number of External Branches\")\n",
    "ax.set_xlabel(\"External Branch Length\\n(Substitutions/Site)\")\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(1e-5))\n",
    "ax.legend(fontsize=FONTSIZE-2, edgecolor=\"black\")\n",
    "\n",
    "ax = axes[1]\n",
    "\n",
    "\n",
    "sns.barplot(\n",
    "    ax=ax,\n",
    "    data=df,\n",
    "    #x=\"perc\",\n",
    "    x=\"long_branches\",\n",
    "    y=\"population\",\n",
    "    palette=colors_dict,\n",
    "    ec=\"black\",\n",
    ")\n",
    "\n",
    "\"\"\"y = 0\n",
    "for rec in df.iterrows():\n",
    "    #x = rec[1][\"perc\"]\n",
    "    x = rec[1][\"perc\"]\n",
    "    t = rec[1][\"long_branches\"]\n",
    "    ax.text(x + 0.25, y + 0.1, str(round(t)), fontsize=FONTSIZE-4)\n",
    "    y += 1\"\"\"\n",
    "\n",
    "\"\"\"for y, x in enumerate(list(df[\"perc\"])):\n",
    "    print(y,x)\n",
    "    ax.text(x + 0.5, y + .25, str(round(x)) + \"%\", fontsize=FONTSIZE-4)\"\"\"\n",
    "\n",
    "ax.set_ylabel(\"Population\")\n",
    "ax.set_xlabel(\"Number of External Branches\\nLonger Than Threshold\")\n",
    "\n",
    "out_path = os.path.join(out_dir, \"long_branches\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accessible-sarah",
   "metadata": {},
   "source": [
    "## TsTv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "amateur-forwarding",
   "metadata": {},
   "outputs": [],
   "source": [
    "FONTSIZE=14\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "DPI=50\n",
    "FIGSIZE=[18,6]\n",
    "## All\n",
    "fig, axes = plt.subplots(1,3,figsize=FIGSIZE, dpi=DPI)\n",
    "    \n",
    "df = metadata_df[metadata_df[variables[0]] != NO_DATA_CHAR]\n",
    "outliers = []\n",
    "ax_i = 0\n",
    "for x, xlab, tail in zip(variables, xlabels, tails):\n",
    "    ax = axes[ax_i]\n",
    "    ax_i += 1\n",
    "    samples = list(df.index)\n",
    "    data_x = list(df[x])\n",
    "    data_y = list(df[y])\n",
    "    \n",
    "    if tail == 2:\n",
    "        adjust = (100 - confidence) / 2\n",
    "        ci = [0 + adjust, 100 - adjust]\n",
    "    elif tail == 1:\n",
    "        ci = [0, confidence]\n",
    "    x_ci = np.array(\n",
    "        np.percentile(\n",
    "            np.array(list(df[x])),\n",
    "            ci,\n",
    "            axis=0,\n",
    "        )\n",
    "    )\n",
    "    print(x, ci, x_ci)\n",
    "\n",
    "\n",
    "    \"\"\"sns.histplot(\n",
    "        ax=ax,\n",
    "        data=df,\n",
    "        x=data_x,\n",
    "        kde=True,\n",
    "        line_kws={\"lw\": 2},\n",
    "        bins=10,\n",
    "        kde_kws={\"bw_adjust\" : 0.7}\n",
    "    )\"\"\"\n",
    "    sns.kdeplot(\n",
    "        ax=ax,\n",
    "        data=df,\n",
    "        x=data_x,\n",
    "        fill=True,\n",
    "        bw_adjust=0.7,\n",
    "    )\n",
    "        \n",
    "    \n",
    "    ax.set_xlabel(xlab)\n",
    "    ax.set_ylabel(\"Number of Samples\")\n",
    "    #ax.axvline(x_ci[0], ls=\"--\", color=\"black\", lw=1)\n",
    "    #ax.axvline(x_ci[1], ls=\"--\", color=\"black\", lw=1)\n",
    "    \n",
    "    for sample in df.index:\n",
    "        val = df[x][sample]\n",
    "        if x == \"other_var_ratio\" and \"Ancient\" in df[\"biosample_comment\"][sample]:\n",
    "            continue\n",
    "        if x_ci[0] <= val <= x_ci[1]:\n",
    "            outlier = False\n",
    "        else:\n",
    "            outlier = True\n",
    "        if outlier and sample not in outliers:\n",
    "            outliers.append(sample)\n",
    "            print(df[\"population\"][sample], df[\"strain\"][sample], x, val)\n",
    "\n",
    "metadata_df[\"variant_qc\"] = [\"pass\"] * len(metadata_df)\n",
    "for outlier in outliers:\n",
    "    metadata_df.at[outlier, \"variant_qc\"] = \"fail\"\n",
    "    \n",
    "print(\"Outliers:\", len(outliers))\n",
    "#display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aboriginal-asthma",
   "metadata": {},
   "source": [
    "### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sunset-ranking",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for x, xlab in zip(variables, xlabels):\n",
    "    confidence = 0.90\n",
    "    samples = list(df.index)\n",
    "    data_x = list(df[x])\n",
    "    data_y = list(df[y])\n",
    "\n",
    "\n",
    "    g = sns.JointGrid(       \n",
    "                x=data_x,\n",
    "                y=data_y,\n",
    "                space=0,\n",
    "            )\n",
    "\n",
    "\n",
    "    # ------------------------------\n",
    "    # Plot Bivariate KDE\n",
    "    sns.kdeplot(\n",
    "        ax=g.ax_joint,    \n",
    "        x=data_x,\n",
    "        y=data_y,    \n",
    "        thresh=1-confidence,\n",
    "        fill=True,\n",
    "        color=\"grey\",\n",
    "    )\n",
    "\n",
    "    # ------------------------------\n",
    "    # Detect outliers\n",
    "    outer_poly = g.ax_joint.collections[0]\n",
    "\n",
    "    i = 0\n",
    "    for px,py,ps in zip(data_x, data_y, samples):\n",
    "        point = [px, py]\n",
    "        outlier=1 # red\n",
    "\n",
    "        for path in outer_poly.get_paths():        \n",
    "            contained = path.contains_points([point])\n",
    "            if contained:\n",
    "                outlier=0 # blue\n",
    "                break\n",
    "\n",
    "        df.at[ps, x + \"_outlier\"] = outlier\n",
    "\n",
    "    # ------------------------------\n",
    "    # Plot Marginals\n",
    "    g.plot_marginals(\n",
    "        sns.kdeplot, \n",
    "        #bw_adjust=0.3,\n",
    "        fill=True,\n",
    "        color=\"grey\",\n",
    "    )\n",
    "\n",
    "    # ------------------------------\n",
    "    # Plot Scatter\n",
    "    labels = [\"Yes\",\"No\"]\n",
    "    vals = [1,0]\n",
    "\n",
    "    \"\"\"for val,label in zip(vals, labels):\n",
    "        sub_df = df[df[x + \"_outlier\"] == val]\n",
    "        sns.scatterplot(\n",
    "            ax=g.ax_joint,\n",
    "            data=sub_df,\n",
    "            x=x,\n",
    "            y=y,\n",
    "            s=30,\n",
    "            ec = \"black\",      \n",
    "            lw=1,\n",
    "            label=label,\n",
    "            alpha=0.9,\n",
    "        )\"\"\"\n",
    "\n",
    "    for population in population_list:\n",
    "        pop_df = df[df[\"population\"] == population]\n",
    "        color = pop_df[\"population_color\"][0]\n",
    "        sns.scatterplot(\n",
    "            ax=g.ax_joint,\n",
    "            x=list(pop_df[x]),\n",
    "            y=list(pop_df[y]),\n",
    "            s=30,\n",
    "            color=color,\n",
    "            ec = \"black\",      \n",
    "            #alpha=1.00,\n",
    "            lw=1,\n",
    "            zorder=2,   \n",
    "            label=population,\n",
    "        )\n",
    "\n",
    "    g.ax_joint.set_xlabel(xlab)\n",
    "    g.ax_joint.set_ylabel(ylab)\n",
    "    g.ax_joint.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))  \n",
    "\n",
    "    #for ax in [g.ax_joint, g.ax_marg_y]:\n",
    "    #    ax.axhline(1e-5, ls=\"--\", c=\"black\", lw=1)\n",
    "\n",
    "    g.ax_joint.get_legend().remove()\n",
    "    #g.fig.legend(title=\"Outlier\", bbox_to_anchor=(1.3,0.8), edgecolor=\"black\")\n",
    "    g.fig.legend(title=\"Population\", bbox_to_anchor=(1.3,0.8), edgecolor=\"black\")\n",
    "\n",
    "\n",
    "    out_path = os.path.join(out_dir, x)\n",
    "    plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "    plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\n",
    "\n",
    "    for sample in df.index:\n",
    "        metadata_df.at[sample, x + \"_outlier\"] = df[x + \"_outlier\"][sample]\n",
    "    # Hmm\n",
    "    metadata_df.at[\"Reference\", x + \"_outlier\"] = 0\n",
    "\n",
    "    out_path = os.path.join(out_dir, x + \".tsv\")\n",
    "    \n",
    "    header = \"{}\\t{}\\t{}\\t{}\".format(\"population\", \"samples\", \"outliers\", \"perc_outliers\")\n",
    "    with open(out_path, \"w\") as outfile:\n",
    "        outfile.write(header + \"\\n\")\n",
    "        for population in population_list:\n",
    "            pop_df = metadata_df[metadata_df[\"population\"] == population]\n",
    "            samples = len(pop_df)\n",
    "            outliers = len(pop_df[pop_df[x + \"_outlier\"] == 1])\n",
    "            perc_outliers = round((outliers/samples) * 100, 1)\n",
    "            line = \"{}\\t{}\\t{}\\t{}\".format(population, samples, outliers, perc_outliers)\n",
    "            outfile.write(line + \"\\n\")\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "leading-photograph",
   "metadata": {},
   "source": [
    "### Long Branch Percentage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "built-russia",
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 1e-5\n",
    "long_branches = 0\n",
    "tstv_outlier = 0\n",
    "\n",
    "tstv_min = 1.36\n",
    "tstv_max = 3.9\n",
    "\n",
    "\n",
    "for sample in metadata_df.index:\n",
    "    branch_length = metadata_df[\"external_branch_length\"][sample]\n",
    "    tstv = metadata_df[\"tstv\"][sample]\n",
    "    \n",
    "    if branch_length >= threshold:\n",
    "        long_branches += 1\n",
    "    \n",
    "    if tstv == NO_DATA_CHAR: continue\n",
    "    if tstv > tstv_max or tstv < tstv_min:\n",
    "        tstv_outlier += 1\n",
    "        \n",
    "        \n",
    "        \n",
    "print(long_branches, tstv_outlier)\n",
    "print(long_branches / len(metadata_df) * 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "distributed-windsor",
   "metadata": {},
   "source": [
    "## TsTv Outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "swedish-transportation",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Final Values\n",
    "\n",
    "tstv_min = 1.36\n",
    "tstv_max = 3.9\n",
    "\n",
    "out_path = os.path.join(out_dir, \"tstv_outliers.tsv\")\n",
    "\n",
    "perc_outliers_dict = {}\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    header = \"{}\\t{}\\t{}\\t{}\".format(\"population\", \"samples\", \"outliers\", \"perc_outliers\")\n",
    "    print(\"|\" + header.replace(\"\\t\", \"|\") + \"|\")\n",
    "    outfile.write(header + \"\\n\")\n",
    "\n",
    "    for population in population_list:\n",
    "        pop_df = df[df[\"population\"] == population]\n",
    "        samples = len(pop_df)\n",
    "        outliers = len(pop_df[(pop_df[\"tstv\"] <= tstv_min) | (pop_df[\"tstv\"] >= tstv_max)])\n",
    "        perc_outliers = round((outliers / samples) * 100)\n",
    "        line = \"{}\\t{}\\t{}\\t{}\".format(population, samples, outliers, perc_outliers)\n",
    "        print(\"|\" + line.replace(\"\\t\", \"|\") + \"|\")\n",
    "        outfile.write(line + \"\\n\")\n",
    "\n",
    "        perc_outliers_dict[population] = perc_outliers\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "extra-shore",
   "metadata": {},
   "source": [
    "## Manual BEAST Rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "distinguished-gambling",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bad placement in this notebook!!!\n",
    "\n",
    "rates = {\n",
    " '1.ORI': 1.27e-07,\n",
    " '1.IN': 8.32e-08,\n",
    " '1.ANT': 5.84e-08,\n",
    " '1.PRE': 4.68e-08,\n",
    " '2.MED': 2.41e-07,\n",
    " '2.ANT': 8.18e-08,\n",
    " '4.ANT': 7.99e-08,\n",
    " '3.ANT': 6.25e-08,\n",
    " '0.ANT4': 3.63e-08,\n",
    " '0.ANT': 5.65e-08,\n",
    " '0.PE': 6.16e-07,\n",
    " '0.PRE': 5.56e-08}\n",
    "\n",
    "x = list(perc_outliers_dict.values())\n",
    "y = list(rates.values())\n",
    "\n",
    "sns.scatterplot(x,y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "green-cycling",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"# Final Values\n",
    "\n",
    "tstv_min = 1.36\n",
    "tstv_max = 3.9\n",
    "\n",
    "header = \"{}\\t{}\\t{}\".format(\"population\", \"samples\", \"outliers\", \"perc_outliers\")\n",
    "print(header)\n",
    "\n",
    "\n",
    "#print(\"tstv: \", tstv_min, tstv_max)\n",
    "#print(\"ns_ss: \", ns_ss_min, ns_ss_max)\n",
    "#print(\"other_var: \", other_var_min, other_var_max)\n",
    "#print(\"external branch length:\", 1e-05)\n",
    "\n",
    "drop_samples = 0\n",
    "\n",
    "out_path = os.path.join(out_dir, \"variant_qc_fail.txt\")\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    header = \"{}\\t{}\".format(\"BioSampleAccession\", \"BioSampleComment\")\n",
    "    outfile.write(header + \"\\n\")\n",
    "    for rec in metadata_df.iterrows():\n",
    "        if rec[1][\"ns_ss_ratio\"] == NO_DATA_CHAR: continue        \n",
    "        if \"Modern\" not in rec[1][\"biosample_comment\"]: continue\n",
    "\n",
    "        sample = rec[0]\n",
    "\n",
    "        ns_ss = rec[1][\"ns_ss_ratio\"]\n",
    "        tstv = rec[1][\"tstv\"]\n",
    "        other_var = rec[1][\"other_var_ratio\"]\n",
    "        external_branch_length = rec[1][\"external_branch_length\"]\n",
    "        if (\n",
    "            sample in metadata_df.index\n",
    "            and (\n",
    "            tstv >= tstv_max\n",
    "            or tstv <= tstv_min\n",
    "            or ns_ss >= ns_ss_max\n",
    "            or ns_ss <= ns_ss_min\n",
    "            or other_var >= other_var_max\n",
    "            or external_branch_length >= 1e-05)\n",
    "        ):\n",
    "            drop_samples += 1\n",
    "            #metadata_df.drop(sample, inplace=True)\n",
    "            acc = metadata_df[\"biosample_accession\"][sample]\n",
    "            comment = metadata_df[\"biosample_comment\"][sample]\n",
    "            comment += \" QC Fail\"\n",
    "            outfile.write(\"{}\\t{}\\n\".format(acc, comment))\n",
    "        \n",
    "print(\"Dropped\", drop_samples, \"out of\", len(metadata_df_bak))\n",
    "#display(metadata_df)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "useful-attachment",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "stuffed-running",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "\n",
    "FONTSIZE=14\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "DPI=400\n",
    "FIGSIZE=[18,6]\n",
    "\n",
    "## All\n",
    "fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "for spine in ax.spines:\n",
    "    ax.spines[spine].set_linewidth(0.25)\n",
    "\n",
    "# Sorry Antarctica\n",
    "ax.set_ylim(-60,90)\n",
    "\n",
    "world_polygons.plot(\n",
    "    ax=ax, \n",
    "    zorder=1, \n",
    "    alpha=0.5, \n",
    "    color=\"grey\",\n",
    "    edgecolor=\"white\", \n",
    "    linewidth=0.1,\n",
    ") \n",
    "\n",
    "# Rescale the sizes for pie chart/inset axes\n",
    "sizes = list(size_by_coord.values())\n",
    "# Try to log values\n",
    "#sizes = [math.log(s) for s in sizes]\n",
    "# Convert to 2D array\n",
    "sizes = np.reshape(np.array(sizes), (-1, 1))\n",
    "rescale_max = 0.15\n",
    "rescale_min = 0.025\n",
    "sizes_rescale = sklearn.preprocessing.minmax_scale(sizes, feature_range=(rescale_min,rescale_max))\n",
    "size_by_coord_rescale = {c:s[0] for c,s in zip(size_by_coord.keys(), sizes_rescale)}\n",
    "\n",
    "\n",
    "i = 0\n",
    "for coord in size_by_coord:\n",
    "    # Get all samples from this coordinate\n",
    "    coord_list = [float(c) for c in coord.split(\",\")]\n",
    "    lon,lat = coord_list[0], coord_list[1]\n",
    "    df = metadata_df[(metadata_df[\"lon\"] == lon) & (metadata_df[\"lat\"] == lat)]\n",
    "    populations = set(df[\"population\"])\n",
    "    coord_colors = []\n",
    "    sizes = []\n",
    "    for pop in populations:\n",
    "        num_samples = len(df[df[\"population\"] == pop])\n",
    "        color = colors_dict[pop]\n",
    "        sizes.append(num_samples)    \n",
    "        coord_colors.append(color)\n",
    "        \n",
    "    total_sizes = sum(sizes)\n",
    "    ax_width = size_by_coord_rescale[coord]\n",
    "    ax_width = size_by_coord_rescale[coord] * 5\n",
    "    \n",
    "   \n",
    "    # Create an axis inset\n",
    "    ax_sub=inset_axes(ax, width=ax_width, height=ax_width, loc=10, \n",
    "                       bbox_to_anchor=(lon, lat),\n",
    "                       bbox_transform=ax.transData,\n",
    "                       borderpad=0) \n",
    "    \n",
    "    # Create pie chart\n",
    "    wedges,texts= ax_sub.pie(\n",
    "        sizes, \n",
    "        colors=coord_colors, \n",
    "        normalize=True,\n",
    "        startangle=90,\n",
    "        wedgeprops = {'linewidth': 0.25, 'edgecolor' : 'black', \"alpha\" : 0.75},\n",
    "    )\n",
    "    \n",
    "    i += 1\n",
    "    #if i > 20: break\n",
    "    \n",
    "\n",
    "# Population legend\n",
    "populations = list(colors_dict.keys())\n",
    "populations.remove(NO_DATA_CHAR)\n",
    "populations.reverse()\n",
    "\n",
    "\n",
    "population_lines = [lines.Line2D([0], [0], color=colors_dict[population], lw=0, markersize=2, markeredgecolor=\"black\", markeredgewidth=0.25, marker=\"o\") \n",
    "                for population in populations]\n",
    "population_labels = populations\n",
    "\n",
    "population_legend = fig.legend(population_lines, \n",
    "           populations,\n",
    "           #ncol=len(populations),\n",
    "           ncol=1,\n",
    "           loc=\"center\", \n",
    "           bbox_to_anchor=(0.2, 0.475),\n",
    "           frameon=True,\n",
    "            borderpad=0.5, \n",
    "            handletextpad=0.25, \n",
    "           labelspacing=0.5,\n",
    "           edgecolor=\"black\",\n",
    "           title=\"Population\",\n",
    "           fontsize=FONTSIZE / 2.25,\n",
    "           title_fontsize=FONTSIZE / 1.5,\n",
    "          )\n",
    "population_legend.get_frame().set_linewidth(0.25)\n",
    "\n",
    "    \n",
    "\n",
    "# Size Legend\n",
    "s = [ax.scatter(0,-9000, s=1, ec=\"black\", c=\"white\", lw=0.25),\n",
    "     ax.scatter(0,-9000, s=10, ec=\"black\", c=\"white\", lw=0.25), \n",
    "    ax.scatter(0,-9000, s=100, ec=\"black\", c=\"white\", lw=0.25),     \n",
    "    ]\n",
    "sl = [\"1\", \"10\", \"100\"]\n",
    "leg_size = fig.legend(s, sl, title=\"Genomes\", loc=\"center\",  borderpad=0.5, bbox_to_anchor=[0.3,0.45], edgecolor=\"black\",\n",
    "                    fontsize=FONTSIZE / 2, title_fontsize=FONTSIZE / 1.5)\n",
    "leg_size.get_frame().set_linewidth(0.25)\n",
    "\n",
    "\n",
    "out_path = os.path.join(out_dir, \"map_all\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nasty-indonesian",
   "metadata": {},
   "source": [
    "## Individual map by population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "approved-potter",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_list = list(colors_dict.keys())\n",
    "population_list.remove(NO_DATA_CHAR)\n",
    "\n",
    "FONTSIZE=12\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "DPI=400\n",
    "FIGSIZE=[6,3]\n",
    "\n",
    "for population in population_list:\n",
    "    print(population)\n",
    "\n",
    "    fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG}, figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "    # Sorry Antarctica\n",
    "    ax.set_ylim(-60,90)\n",
    "    color = colors_dict[population]\n",
    "\n",
    "    world_polygons.plot(\n",
    "        ax=ax, \n",
    "        zorder=1, \n",
    "        alpha=0.5, \n",
    "        color=\"grey\",\n",
    "        edgecolor=\"grey\", \n",
    "        linewidth=0,\n",
    "    ) \n",
    "    \n",
    "    for coord in size_by_pop[population]:\n",
    "        coord_list = [float(c) for c in coord.split(\",\")]\n",
    "        lon,lat = coord_list[0], coord_list[1]\n",
    "        gdf = metadata_gdf[(metadata_gdf[\"population\"] == population) & (metadata_gdf[\"lon\"] == lon) & (metadata_gdf[\"lat\"] == lat)]\n",
    "        num_samples = len(gdf)\n",
    "        marker_scale = 20\n",
    "        gdf.plot(\n",
    "            ax=ax, \n",
    "            c=color, \n",
    "            ec=\"black\", \n",
    "            lw=0.25, \n",
    "            markersize=num_samples * marker_scale,\n",
    "            alpha=0.75,\n",
    "        )\n",
    "        \n",
    "    out_path = os.path.join(out_dir, \"map_{}\".format(population))\n",
    "    plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "    plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rocky-sixth",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. RTT (Root-To-Tip Regression)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deadly-escape",
   "metadata": {},
   "source": [
    "## Calculate Population And Root Distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "innovative-butter",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df[\"root_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"population_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "\n",
    "population_mrca_dict = {}\n",
    "\n",
    "# Get the MRCA for Each Population\n",
    "for population in colors_dict:\n",
    "    if population == NO_DATA_CHAR: continue\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    tips = list(df.index)\n",
    "    mrca = divtree.common_ancestor(tips)\n",
    "    population_mrca_dict[population] = mrca\n",
    "\n",
    "for c in divtree.get_terminals():\n",
    "    sample = c.name\n",
    "    if sample not in metadata_df.index: continue\n",
    "    population = metadata_df[\"population\"][sample]    \n",
    "    # Genetic Distance to Root\n",
    "    root_dist = divtree.distance(divtree.root, c)\n",
    "    # Genetic Distance to Population MRCA \n",
    "    mrca = population_mrca_dict[population]\n",
    "    mrca_dist = divtree.distance(mrca, c)\n",
    "\n",
    "    metadata_df.at[sample, \"root_rtt_dist\"] = float(root_dist)\n",
    "    metadata_df.at[sample, \"population_rtt_dist\"] = float(mrca_dist)\n",
    "\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nuclear-backing",
   "metadata": {},
   "source": [
    "## Report Average RTT by Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "underlying-field",
   "metadata": {},
   "outputs": [],
   "source": [
    "rtt_all_list = list(metadata_df[\"root_rtt_dist\"])\n",
    "rtt_all_mean = sum(rtt_all_list) / len(rtt_all_list)\n",
    "rtt_dist_dict = {\"All\": rtt_all_mean}\n",
    "\n",
    "for population in colors_dict:\n",
    "    if population == NO_DATA_CHAR: continue\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    rtt_list = list(df[\"population_rtt_dist\"])\n",
    "    rtt_mean = sum(rtt_list) / len(rtt_list)\n",
    "    \n",
    "    rtt_dist_dict[population] = rtt_mean\n",
    "    \n",
    "print(rtt_dist_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "defensive-bacon",
   "metadata": {},
   "source": [
    "## B. Root-To-Tip Regression With Bootstrapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "recorded-injection",
   "metadata": {},
   "source": [
    "## All"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dimensional-number",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"import sys, importlib\n",
    "importlib.reload(sys.modules['functions'])\n",
    "from functions import linregress_bootstrap\"\"\"\n",
    "\n",
    "np.random.seed(5435455)\n",
    "NBOOTS = 1000\n",
    "\n",
    "bootstrap_mega_dict = {}\n",
    "\n",
    "import sys, importlib\n",
    "importlib.reload(sys.modules['functions'])\n",
    "from functions import linregress_bootstrap\n",
    "\n",
    "# ALL\n",
    "df = copy.deepcopy(metadata_df[(metadata_df[\"date_mean\"] != NO_DATA_CHAR)])\n",
    "label = \"All\"\n",
    "print(\"Population:\", label)\n",
    "color = \"grey\"\n",
    "x = list(float(n) for n in df[\"date_mean\"])\n",
    "y = list(df[\"root_rtt_dist\"])\n",
    "xerr=list(df[\"date_err\"])\n",
    "\n",
    "bootstraps = linregress_bootstrap(x=x,y=y,xerr=xerr,nboots=NBOOTS, plot=False, color=color, label=label)\n",
    "bootstrap_mega_dict[\"All\"] = bootstraps\n",
    "bootstrap_mega_dict[\"All\"][\"samples\"] = list(df.index)\n",
    "bootstrap_mega_dict[\"All\"][\"color\"] = \"#c4c4c4\"\n",
    "bootstrap_mega_dict[\"All\"][\"xerr\"] = xerr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "reflected-chuck",
   "metadata": {},
   "source": [
    "## Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pediatric-customs",
   "metadata": {},
   "outputs": [],
   "source": [
    "for population in colors_dict:\n",
    "    if population == NO_DATA_CHAR: continue\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    print(\"Population:\", population) \n",
    "    if population == \"0.PE\":\n",
    "        df.drop(\"SAMEA104488961\", inplace=True)\n",
    "    \n",
    "    if len(df) == 0: continue    \n",
    "    color = df[\"population_color\"][0]\n",
    "    x = list(df[\"date_mean\"])\n",
    "    y = list(df[\"population_rtt_dist\"])\n",
    "    #y = list(df[\"root_rtt_dist\"])\n",
    "    xerr=list(df[\"date_err\"])\n",
    "\n",
    "    bootstraps = linregress_bootstrap(x=x,y=y,xerr=xerr,nboots=NBOOTS, plot=False, color=color, label=label)\n",
    "    bootstrap_mega_dict[population] = bootstraps\n",
    "    bootstrap_mega_dict[population][\"samples\"] = list(df.index)\n",
    "    bootstrap_mega_dict[population][\"color\"] = colors_dict[population]\n",
    "    bootstrap_mega_dict[population][\"xerr\"] = xerr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "banner-decimal",
   "metadata": {},
   "source": [
    "## Correct p values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "invisible-earth",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct p_values, \"All\" is the first one\n",
    "raw_p_values = [bootstrap_mega_dict[population][\"p_value\"] for population in bootstrap_mega_dict]\n",
    "\n",
    "reject, corrected_p_values, alphacSidak, alphacBonf = smmt.multipletests(raw_p_values, alpha = 0.05, method = \"holm\")\n",
    "\n",
    "i = 0\n",
    "for population in bootstrap_mega_dict:\n",
    "    pop_p_value = corrected_p_values[i]\n",
    "    pop_reject = reject[i]\n",
    "    bootstrap_mega_dict[population][\"corrected_p_value\"] = pop_p_value\n",
    "    bootstrap_mega_dict[population][\"reject\"] = pop_reject\n",
    "    i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "informed-omaha",
   "metadata": {},
   "source": [
    "## Write Stats to File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "white-hungarian",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save stats\n",
    "out_path = os.path.join(out_dir, \"rtt.tsv\")\n",
    "\n",
    "#print(bootstrap_mega_dict[\"All\"].keys())\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    headers = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\".format(\n",
    "        \"population\",\n",
    "        \"mean_rtt_dist\",\n",
    "        \"coefficient\",\n",
    "        \"raw_p_value\",\n",
    "        \"n\",\n",
    "        \"corrected_p_value\",\n",
    "        \"reject\",\n",
    "        \"rate\",\n",
    "        \"rate_bootstrap_peak\",\n",
    "        \"rate_bootstrap_ci_low\",\n",
    "        \"rate_bootstrap_ci_high\",\n",
    "        \"mrca\",\n",
    "        \"mrca_bootstrap_peak\",\n",
    "        \"mrca_bootstrap_ci_low\",\n",
    "        \"mrca_bootstrap_ci_high\", \n",
    "    )\n",
    "    outfile.write(headers + \"\\n\")\n",
    "    \n",
    "    # Population Values\n",
    "    for population in bootstrap_mega_dict:\n",
    "        data = [\n",
    "            population, \n",
    "            \"{:.2e}\".format(rtt_dist_dict[population]),            \n",
    "            round(bootstrap_mega_dict[population][\"r_squared\"],2),\n",
    "            \"{:.3e}\".format(bootstrap_mega_dict[population][\"p_value\"]),\n",
    "            bootstrap_mega_dict[population][\"n\"],\n",
    "            \"{:.3e}\".format(bootstrap_mega_dict[population][\"corrected_p_value\"]),\n",
    "            bootstrap_mega_dict[population][\"reject\"],\n",
    "            \"{:.2e}\".format(bootstrap_mega_dict[population][\"slope\"]),\n",
    "            \"{:.2e}\".format(bootstrap_mega_dict[population][\"bootstrap_slope_peak\"]),\n",
    "            \"{:.2e}\".format(bootstrap_mega_dict[population][\"bootstrap_slope_ci\"][0]),\n",
    "            \"{:.2e}\".format(bootstrap_mega_dict[population][\"bootstrap_slope_ci\"][1]),\n",
    "            round(bootstrap_mega_dict[population][\"x_intercept\"]),\n",
    "            round(bootstrap_mega_dict[population][\"bootstrap_x_intercept_peak\"]),\n",
    "            round(bootstrap_mega_dict[population][\"bootstrap_x_intercept_ci\"][0]),\n",
    "            round(bootstrap_mega_dict[population][\"bootstrap_x_intercept_ci\"][1]),\n",
    "        ]\n",
    "        data_str = \"\\t\".join([str(d) for d in data])\n",
    "        outfile.write(data_str + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "comparable-belly",
   "metadata": {},
   "source": [
    "## Extract Temporal Constraints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "quality-thumbnail",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_constraints = os.path.join(out_dir, \"temporal_constraints.txt\")\n",
    "with open(out_path_constraints, \"w\") as outfile:\n",
    "    for population in colors_dict:\n",
    "        if population == NO_DATA_CHAR: continue\n",
    "        df = metadata_df[metadata_df[\"population\"] == population]\n",
    "        tips = list(df.index)\n",
    "\n",
    "        population_bootstraps = bootstrap_mega_dict[population]\n",
    "        p_value = population_bootstraps[\"corrected_p_value\"]\n",
    "        slope = population_bootstraps[\"bootstrap_slope_peak\"]\n",
    "        if slope < 0:\n",
    "            continue\n",
    "        if p_value >= ALPHA: \n",
    "            continue\n",
    "        print(\"Population:\", population)\n",
    "        mrca_ci = population_bootstraps[\"bootstrap_x_intercept_ci\"]\n",
    "        mrca_ci_bp = [\"-\" + str(CURRENT_YEAR - d) for d in mrca_ci]\n",
    "\n",
    "        # LSD mrca constraint format mrca(A,B,C) b(1998.21,2000.5)\n",
    "        print(mrca_ci)\n",
    "        outfile.write(\"mrca({}) b({})\\n\".format(\",\".join(tips), \",\".join(mrca_ci_bp)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "expired-physiology",
   "metadata": {},
   "source": [
    "## Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ceramic-relay",
   "metadata": {},
   "source": [
    "### All"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "anticipated-intranet",
   "metadata": {},
   "outputs": [],
   "source": [
    "TARGET_RES = [720, 360]\n",
    "DPI = 200\n",
    "FIGSIZE = [TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE = 5\n",
    "plt.rc(\"font\", size=FONTSIZE)\n",
    "\n",
    "                \n",
    "dates = [float(date) for date in metadata_df[\"date_mean\"]]\n",
    "dists = [float(dist) for dist in metadata_df[\"root_rtt_dist\"]]\n",
    "dates_err = [float(date) for date in metadata_df[\"date_err\"]]\n",
    "color = [colors_dict[population] for population in metadata_df[\"population\"]]\n",
    "\n",
    "\n",
    "p_sig = \"\"\n",
    "if bootstrap_mega_dict[\"All\"][\"reject\"]:\n",
    "    p_sig = \"*\"\n",
    "    \n",
    "corrected_p_value = bootstrap_mega_dict[\"All\"][\"corrected_p_value\"]\n",
    "pretty_p_value = corrected_p_value\n",
    "if pretty_p_value != 1:\n",
    "    pretty_p_value = \"{:.3e}\".format(corrected_p_value)\n",
    "\n",
    "fig, axes = plt.subplots(1, figsize=FIGSIZE, dpi=DPI)\n",
    "ax = axes\n",
    "\n",
    "# plot regression line of populations with strict clock signal\n",
    "for population in bootstrap_mega_dict:\n",
    "    # Distance of population root to phylogeny root\n",
    "    #print(bootstrap_mega_dict[population][\"samples\"])\n",
    "    root_node =  divtree.common_ancestor(bootstrap_mega_dict[population][\"samples\"])\n",
    "    root_dist = divtree.distance(root_node)\n",
    "    \n",
    "    y_adjust = [y + root_dist for y in bootstrap_mega_dict[population][\"y\"]]\n",
    "    \n",
    "    linestyle = \"--\" if population != \"All\" else \"-\"\n",
    "    reject = bootstrap_mega_dict[population][\"reject\"] \n",
    "    slope = bootstrap_mega_dict[population][\"slope\"] \n",
    "    if reject and slope >= 0:\n",
    "        sns.regplot(\n",
    "            ax=ax,\n",
    "            x=bootstrap_mega_dict[population][\"x\"],\n",
    "            y=y_adjust,\n",
    "            ci=None,\n",
    "            scatter_kws={\"s\": 0},\n",
    "            line_kws={\"linewidth\": 0.5, \"zorder\":10, \"linestyle\": linestyle},\n",
    "            color=\"black\",\n",
    "        )\n",
    "\n",
    "ax.scatter(\n",
    "    x=dates,\n",
    "    y=dists,\n",
    "    s=10,\n",
    "    color=color,\n",
    "    ec=\"black\",\n",
    "    lw=0.25,\n",
    "    label=None,\n",
    "    alpha=0.8,\n",
    "    zorder=3,\n",
    ")\n",
    "\n",
    "ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))  \n",
    "ax.set_xlim(-3500, 2500)\n",
    "ax.set_xlabel(\"Date\")\n",
    "ax.set_ylabel(\"Root-To-Tip Distance \\n (substitutions/site)\", fontsize=5)\n",
    "\n",
    "# Use the stats from bootstrapping!\n",
    "ax.annotate(\n",
    "    (\"      R²: {}\".format(round(bootstrap_mega_dict[\"All\"][\"r_squared\"], 2))\n",
    "        + \"\\n       p: {}{}\".format(pretty_p_value, p_sig)\n",
    "        + \"\\n  Rate: {:.2e}\".format(bootstrap_mega_dict[\"All\"][\"slope\"])\n",
    "        + \"\\nMRCA: {:.0f}\".format(bootstrap_mega_dict[\"All\"][\"x_intercept\"])\n",
    "    ),\n",
    "    xy=(0.05, 0.8),\n",
    "    xycoords=\"axes fraction\",\n",
    "    size=5,\n",
    "    ha=\"left\",\n",
    "    va=\"center\",\n",
    "    bbox=dict(fc=\"w\", lw=0.5),\n",
    ")\n",
    "\n",
    "# Population legend\n",
    "populations = list(colors_dict.keys())\n",
    "populations.remove(NO_DATA_CHAR)\n",
    "populations.reverse()\n",
    "\n",
    "population_lines = [lines.Line2D([0], [0], color=colors_dict[population], lw=0, markersize=5, markeredgecolor=\"black\", markeredgewidth=0.25, marker=\"o\") \n",
    "                for population  in populations ]\n",
    "\n",
    "population_legend = fig.legend(population_lines, \n",
    "           populations,\n",
    "           #ncol=len(populations),\n",
    "           ncol=1,\n",
    "           loc=\"center\", \n",
    "           bbox_to_anchor=(1, 0.5),\n",
    "           frameon=True,\n",
    "           borderpad=0.5, \n",
    "           handletextpad=0.25, \n",
    "           labelspacing=0.5,\n",
    "           edgecolor=\"black\",\n",
    "           title=\"Population\",\n",
    "           fontsize=FONTSIZE,\n",
    "           title_fontsize=FONTSIZE,\n",
    "          )\n",
    "population_legend.get_frame().set_linewidth(1)\n",
    "\n",
    "\n",
    "out_path = os.path.join(out_dir, \"rtt_all_population\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accessible-tender",
   "metadata": {},
   "source": [
    "### Multiplot by Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "gothic-custody",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "#TARGET_RES = [1280,  1280]\n",
    "#DPI=200\n",
    "#FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "#FONTSIZE=5\n",
    "#plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "FIGSIZE=[8,6]\n",
    "DPI=50\n",
    "\n",
    "reg_x = \"date_mean\"\n",
    "#reg_y = \"population_rtt_dist\"\n",
    "reg_y = \"root_rtt_dist\"\n",
    "reg_err = \"date_err\"\n",
    "\n",
    "nrow=3\n",
    "ncol=4\n",
    "\n",
    "fig, axes = plt.subplots(nrow,ncol,figsize=FIGSIZE, dpi=DPI)\n",
    "#fig.subplots_adjust(wspace=0.35, hspace=1)\n",
    "fig.subplots_adjust(wspace=0.35, hspace=0.65)\n",
    "\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for population in colors_dict:\n",
    "    if population == NO_DATA_CHAR: continue\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    \n",
    "    if label not in bootstrap_mega_dict: continue\n",
    "    print(\"Population:\", population)\n",
    "    \n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "    \n",
    "    p_sig = \"\"\n",
    "    if bootstrap_mega_dict[population][\"reject\"]:\n",
    "        p_sig = \"*\"    \n",
    "    corrected_p_value = bootstrap_mega_dict[population][\"corrected_p_value\"]\n",
    "    pretty_p_value = corrected_p_value\n",
    "    if pretty_p_value != 1:\n",
    "        pretty_p_value = \"{:.3e}\".format(corrected_p_value)\n",
    "        \n",
    "    # Manual regression with scipy\n",
    "    # Trying this because the negative coefficients of statsmodels confuse me\n",
    "    x = list(df[\"date_mean\"])\n",
    "    y = list(df[\"population_rtt_dist\"])        \n",
    "    slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(x, y)\n",
    "    r_squared = r_value * r_value\n",
    "\n",
    "    sns.regplot(\n",
    "        ax=ax, \n",
    "        x=bootstrap_mega_dict[population][\"x\"],\n",
    "        y=bootstrap_mega_dict[population][\"y\"], \n",
    "        ci=95,\n",
    "        scatter_kws={\"s\": 0},\n",
    "        line_kws={\"linewidth\" : 0.75},\n",
    "        color=\"grey\",\n",
    "        label=(\n",
    "        \"R² = {}\".format(round(r_squared,2))            \n",
    "        #\"R² = {}\".format(round(bootstrap_mega_dict[population][\"r_squared\"],2))\n",
    "        #+ \"\\n        p: {}{}\".format(pretty_p_value, p_sig)\n",
    "        #+ \"\\n   Rate: {:.2}\".format(bootstrap_mega_dict[population][\"bootstrap_slope_peak\"])  \n",
    "        #+ \"\\ntMRCA: {}\".format(round(bootstrap_mega_dict[population][\"bootstrap_x_intercept_peak\"]))                \n",
    "        )\n",
    "    )\n",
    "    \n",
    "    ax.errorbar(\n",
    "        x=bootstrap_mega_dict[population][\"x\"],\n",
    "        y=bootstrap_mega_dict[population][\"y\"], \n",
    "        xerr=bootstrap_mega_dict[population][\"xerr\"],       \n",
    "        yerr=None,\n",
    "        ls='none',\n",
    "        color = bootstrap_mega_dict[population][\"color\"],  \n",
    "        capsize=1,\n",
    "        label=None,\n",
    "        zorder=1,\n",
    "        lw=0.5,\n",
    "        )\n",
    "    \n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        x=bootstrap_mega_dict[population][\"x\"],\n",
    "        y=bootstrap_mega_dict[population][\"y\"],  \n",
    "        s=20,\n",
    "        color = bootstrap_mega_dict[population][\"color\"], \n",
    "        ec = \"black\",\n",
    "        alpha=0.75,\n",
    "        zorder=2,\n",
    "    )\n",
    "\n",
    "    # Set xlimits\n",
    "    xlim = ax.get_xlim()\n",
    "    xrange = abs(xlim[1] - xlim[0])\n",
    "    x_buff = xrange * 0.1\n",
    "    new_xlim = [xlim[0] - x_buff, xlim[1] + x_buff]\n",
    "\n",
    "    ax.set_xlim(new_xlim)\n",
    "\n",
    "    ylim = ax.get_ylim()\n",
    "    \n",
    "    if ylim[0] < 0:\n",
    "        ylim = ylim[0], ylim[1]\n",
    "        \n",
    "    yrange = abs(xlim[1] - xlim[0])\n",
    "    y_buff = yrange * 0.1\n",
    "    new_ylim = [ylim[0] - y_buff, ylim[1] + y_buff]\n",
    "    #ax.set_ylim(-0.000005, 0.000105)\n",
    "    \n",
    "\n",
    "    # Format and position legend\n",
    "    legend = ax.legend(\n",
    "        borderpad=0.8, \n",
    "        handletextpad=-2, \n",
    "        edgecolor=\"black\", \n",
    "        #bbox_to_anchor=(0.5, -0.55), \n",
    "        loc=0,\n",
    "        #loc='center',\n",
    "        #fontsize=FONTSIZE * 0.80,\n",
    "    )\n",
    "    frame = legend.get_frame().set_linewidth(0.5)\n",
    "        \n",
    "    # Format axis\n",
    "    ax.set_title(r\"$\\bf{}$\".format(population))\n",
    "\n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(r\"$\\bf{Distance\\ to\\ MRCA}$\" + \"\\n(Substitutions / Site)\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(r\"$\\bf{Sampling\\ Age}$\")    \n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "    \n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_linewidth(0.5)    \n",
    "            \n",
    "\n",
    "    # Update axis\n",
    "    if i_col == ncol - 1:\n",
    "        i_col = 0\n",
    "        i_row += 1\n",
    "    else:\n",
    "        i_col += 1\n",
    "    \n",
    "    \n",
    "out_path = os.path.join(out_dir, \"rtt_regression_population\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "modular-bottom",
   "metadata": {},
   "source": [
    "## Long Branch Dependency\n",
    "\n",
    "Plot the relationship between rate variation (R2) and the number of long external branches (>= 1e-5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "formed-beach",
   "metadata": {},
   "outputs": [],
   "source": [
    "dpi = 50\n",
    "figsize = [6,6]\n",
    "fontsize = 14\n",
    "plt.rc(\"font\", size=fontsize)\n",
    "\n",
    "fig, axes = plt.subplots(1, figsize=figsize, dpi=dpi)\n",
    "ax = axes\n",
    "\n",
    "for population in population_list:\n",
    "    pop_df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    long_branches = long_branch_dict[population][\"long_branches\"]\n",
    "    r_squared = bootstrap_mega_dict[population][\"r_squared\"]\n",
    "    print(population, long_branches, r_squared)\n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        x=[long_branches],\n",
    "        y=[r_squared],\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "public-yacht",
   "metadata": {},
   "source": [
    "## Time Dependency"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stunning-debate",
   "metadata": {},
   "source": [
    "### Sampling Time Frame By R2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "given-engagement",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"df = metadata_df[metadata_df[\"date_mean\"] >= 1900]\n",
    "\n",
    "confidence = 95\n",
    "ci = np.array(\n",
    "        np.percentile(\n",
    "            np.array(list(df[\"date_mean\"])),\n",
    "            (100 - confidence, confidence),\n",
    "            axis=0,\n",
    "        )\n",
    "    )\n",
    "\n",
    "sns.kdeplot(\n",
    "    data=df,\n",
    "    x=\"date_mean\"\n",
    ")\n",
    "\n",
    "plt.axvline(ci[0])\n",
    "plt.axvline(ci[1])\n",
    "print(ci)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "failing-tennessee",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"TARGET_RES = [720, 360]\n",
    "DPI = 200\n",
    "FIGSIZE = [TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE = 5\n",
    "plt.rc(\"font\", size=FONTSIZE)\n",
    "\n",
    "fig, axes = plt.subplots(1, figsize=FIGSIZE, dpi=DPI)\n",
    "ax = axes\n",
    "\n",
    "x = []\n",
    "y = []\n",
    "c_list = []\n",
    "\n",
    "for population in bootstrap_mega_dict:\n",
    "    if population == \"All\": continue\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    if \"SAMEA104488961\" in df.index:\n",
    "        df.drop(\"SAMEA104488961\", inplace=True)\n",
    "    min_date, max_date = 10000,-10000\n",
    "    for sample in df.index:i\n",
    "        sample_date_mean = df[\"date_mean\"][sample]\n",
    "        sample_date_err = df[\"date_err\"][sample]\n",
    "        sample_min_date = sample_date_mean - sample_date_err\n",
    "        sample_max_date = sample_date_mean + sample_date_err\n",
    "        if sample_min_date < min_date:\n",
    "            min_date = sample_min_date\n",
    "        if sample_max_date > max_date:\n",
    "            max_date = sample_max_date\n",
    "    \n",
    "    diff_date = abs(max_date - min_date)\n",
    "    n = len(df)\n",
    "    print(population, diff_date, n)\n",
    "    r_squared = bootstrap_mega_dict[population][\"r_squared\"]\n",
    "    color = df[\"population_color\"][0]\n",
    "    x.append(diff_date)\n",
    "    y.append(r_squared)\n",
    "    c_list.append(color)\n",
    "\n",
    "\n",
    "# Fit an exponential model\n",
    "def func_exp(x, a, b, c):\n",
    "    return a * x + b * x**2 + c\n",
    "popt, _ = scipy.optimize.curve_fit(func_exp, x, y)\n",
    "\n",
    "# summarize the parameter values\n",
    "a, b, c = popt\n",
    "op_1 = \"+\"\n",
    "if b < 0: op_1 = \"-\"\n",
    "op_2 = \"+\"\n",
    "if c < 0: op_2 = \"-\"\n",
    "formula = \"y = {:.3f}x {} {:.2e}x² {} {:.2f}\".format(a,op_1,abs(b),op_2,abs(c))\n",
    "print(formula)\n",
    "# Create points for the line\n",
    "x_line = np.arange(min(x), max(x), 1)\n",
    "y_line = func_exp(x_line, a, b, c)\n",
    "  \n",
    "ax.plot(x_line, y_line, ls=\"-\", lw=0.5, c=\"black\", zorder=1)\n",
    "    \n",
    "sns.scatterplot(\n",
    "        ax=ax,\n",
    "        x=x,\n",
    "        y=y,  \n",
    "        s=20,\n",
    "        c = c_list, \n",
    "        ec = \"black\",\n",
    "        zorder=2,\n",
    "    )\n",
    "\n",
    "\n",
    "ax.set_xlim(1300,0)\n",
    "ax.set_ylim(-0.2,1)\n",
    "ax.set_xlabel(\"Sampling Time Frame (Years)\")\n",
    "ax.set_ylabel(\"Coefficient of Determination\")\n",
    "\n",
    "# Use the stats from bootstrapping!\n",
    "ax.annotate(\n",
    "    (formula),\n",
    "    xy=(0.05, 0.15),\n",
    "    xycoords=\"axes fraction\",\n",
    "    size=5,\n",
    "    ha=\"left\",\n",
    "    va=\"center\",\n",
    "    bbox=dict(fc=\"w\", lw=0.5),\n",
    ")\n",
    "\n",
    "out_path = os.path.join(out_dir, \"rtt_curve_fit\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "genetic-upgrade",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 3. Geographic Structure"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "marine-anthropology",
   "metadata": {},
   "source": [
    "## All x All Distance Matrix and Mantel Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "extended-crack",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\"\"\"NUM_SAMPLES = len(metadata_gdf.index)\n",
    "#NUM_SAMPLES = 200\n",
    "\n",
    "gene_dist_mat = []\n",
    "geo_dist_mat = []\n",
    "\n",
    "all_dist_dict = {\"gene\":[],\"geo\":[], \"mantel\": {}}\n",
    "\n",
    "i = 0\n",
    "progress_size = 1\n",
    "progress_next = 1\n",
    "\n",
    "analysis_gdf = copy.deepcopy(metadata_gdf)\n",
    "\n",
    "#for sample1 in metadata_gdf.index[:NUM_SAMPLES]:\n",
    "for sample1 in analysis_gdf.index:\n",
    "    progress = 100 * (i / len(analysis_gdf))\n",
    "    if progress >= progress_next:\n",
    "        print(\"{}%\".format(round(progress,2)))\n",
    "        progress_next += progress_size\n",
    "    \n",
    "    # Store distances in list for this sample\n",
    "    geo_dist_list = []\n",
    "    gene_dist_list = []\n",
    "    \n",
    "    geom1 = metadata_gdf[\"geometry\"][sample1]\n",
    "    coord1 = (geom1.y, geom1.x)     \n",
    "    \n",
    "    #for sample2 in metadata_gdf.index[:NUM_SAMPLES]:\n",
    "    for sample2 in analysis_gdf.index:\n",
    "        geom2 = metadata_gdf[\"geometry\"][sample2]\n",
    "        coord2 = (geom2.y, geom2.x)  \n",
    "\n",
    "        # Calculate geographic and genetic distance\n",
    "        geo_dist = round(geopy.distance.great_circle(coord1, coord2).km, 2)\n",
    "        gene_dist = divtree.distance(sample1, sample2)   \n",
    "        \n",
    "        # Add to arrays\n",
    "        geo_dist_list.append(geo_dist)\n",
    "        gene_dist_list.append(gene_dist)   \n",
    "        \n",
    "    # Add to matrix\n",
    "    all_dist_dict[\"geo\"].append(geo_dist_list)\n",
    "    all_dist_dict[\"gene\"].append(gene_dist_list)\n",
    "    \n",
    "    # increment progress counter\n",
    "    i += 1\n",
    "    \n",
    "print(\"100%\")\n",
    "\n",
    "# Check symmetry\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"arr1 = np.array(geo_dist_mat)\n",
    "arr2 = np.array(geo_dist_mat).T\n",
    "\n",
    "print(np.all(np.abs(arr1-arr2) < 1e-10))\n",
    "\n",
    "# Check symmetry\n",
    "arr1 = np.array(gene_dist_mat)\n",
    "arr2 = np.array(gene_dist_mat).T\n",
    "\n",
    "print(np.all(np.abs(arr1-arr2) < 1e-10))\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "coeff, p_value, n = skbio_dist.mantel(all_dist_dict[\"gene\"], all_dist_dict[\"geo\"])\n",
    "all_dist_dict[\"mantel\"][\"coeff\"] = coeff\n",
    "all_dist_dict[\"mantel\"][\"p_value\"] = p_value\n",
    "all_dist_dict[\"mantel\"][\"n\"] = n\n",
    "\n",
    "print(all_dist_dict[\"mantel\"])\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "governing-indianapolis",
   "metadata": {},
   "source": [
    "## Within Clade Distance Matrices and Mantel Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "developing-yugoslavia",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_dist_dict = {}\n",
    "\n",
    "for population in colors_dict:\n",
    "    if population == NO_DATA_CHAR: continue\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]\n",
    "    if len(df) == 0: continue\n",
    "    print(population)\n",
    "    \n",
    "    population_dist_dict[population] = {\"gene\":[],\"geo\":[], \"mantel\": {}}\n",
    "    \n",
    "    for sample1 in df.index:\n",
    "        \n",
    "        # Store distances in list for this sample\n",
    "        geo_dist_list = []\n",
    "        gene_dist_list = []        \n",
    "        \n",
    "        geom1 = metadata_gdf[\"geometry\"][sample1]\n",
    "        coord1 = (geom1.y, geom1.x)  \n",
    "    \n",
    "        for sample2 in df.index:\n",
    "            geom2 = metadata_gdf[\"geometry\"][sample2]\n",
    "            coord2 = (geom2.y, geom2.x)  \n",
    "            \n",
    "            # Calculate geographic and genetic distance\n",
    "            geo_dist = round(geopy.distance.great_circle(coord1, coord2).km, 8)\n",
    "            gene_dist = divtree.distance(sample1, sample2)   \n",
    "\n",
    "            # Add to arrays\n",
    "            geo_dist_list.append(geo_dist)\n",
    "            gene_dist_list.append(gene_dist)   \n",
    "\n",
    "        # Add to matrix\n",
    "        population_dist_dict[population][\"geo\"].append(geo_dist_list)\n",
    "        population_dist_dict[population][\"gene\"].append(gene_dist_list)  \n",
    "        \n",
    "    # Mantel Test\n",
    "    coeff, p_value, n = skbio_dist.mantel(population_dist_dict[population][\"gene\"], population_dist_dict[population][\"geo\"])\n",
    "    population_dist_dict[population][\"mantel\"][\"coeff\"] = coeff\n",
    "    population_dist_dict[population][\"mantel\"][\"p_value\"] = p_value\n",
    "    population_dist_dict[population][\"mantel\"][\"n\"] = n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "muslim-hierarchy",
   "metadata": {},
   "source": [
    "## Correct P Values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "explicit-channel",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct p_values\n",
    "raw_p_values = [population_dist_dict[population][\"mantel\"][\"p_value\"] for population in population_dist_dict]\n",
    "\n",
    "# Add the p value from the all comparison to the end\n",
    "#raw_p_values += [all_dist_dict[\"mantel\"][\"p_value\"]]\n",
    "\n",
    "reject, corrected_p_values, alphacSidak, alphacBonf = smmt.multipletests(raw_p_values, alpha = 0.05, method = \"holm\")\n",
    "\n",
    "i = 0\n",
    "for population in population_dist_dict:\n",
    "    population_p_value = corrected_p_values[i]\n",
    "    population_reject = reject[i]\n",
    "    population_dist_dict[population][\"mantel\"][\"corrected_p_value\"] = population_p_value\n",
    "    population_dist_dict[population][\"mantel\"][\"reject\"] = population_reject\n",
    "    i += 1\n",
    "    \n",
    "# The last corrected p_value is associated with all\n",
    "\"\"\"all_p_value = corrected_p_values[i]\n",
    "all_reject = reject[i]\n",
    "all_dist_dict[\"mantel\"][\"corrected_p_value\"] = all_p_value\n",
    "all_dist_dict[\"mantel\"][\"reject\"] = all_reject\"\"\"\n",
    "    \n",
    "# Save results to a file\n",
    "out_path = os.path.join(out_dir, \"mantel.tsv\")\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    headers = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\".format(\n",
    "        \"population\",\n",
    "        \"coefficient\",\n",
    "        \"raw_p_value\",\n",
    "        \"n\",\n",
    "        \"corrected_p_value\",\n",
    "        \"reject\",\n",
    "    )\n",
    "    outfile.write(headers + \"\\n\")\n",
    "    \n",
    "    # All values\n",
    "    #data = [\"all\"] + list(all_dist_dict[\"mantel\"].values())\n",
    "    #data_str = \"\\t\".join([str(d) for d in data])\n",
    "    #outfile.write(data_str + \"\\n\")\n",
    "    \n",
    "    # Population Values\n",
    "    for population in population_dist_dict:\n",
    "        data = [population] + list(population_dist_dict[population][\"mantel\"].values())\n",
    "        data_str = \"\\t\".join([str(d) for d in data])\n",
    "        outfile.write(data_str + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "center-investigation",
   "metadata": {},
   "source": [
    "## Plot Populations All in 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "vulnerable-stuart",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\"\"\"# PLOT SETUP\n",
    "TARGET_RES = [1280, 480]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "fig, axes = plt.subplots(1,figsize=FIGSIZE, dpi=DPI)\n",
    "\"\"\"\n",
    "FONTSIZE=12\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "DPI=400\n",
    "FIGSIZE=[18,6]\n",
    "\n",
    "fig, ax = plt.subplots(1,1, dpi=400, figsize=FIGSIZE)\n",
    "\n",
    "#for spine in ax.spines:\n",
    "#    ax.spines[spine].set_linewidth(0.5)\n",
    "\n",
    "ax.set_ylabel(\"Genetic Distance\\n(Substitutions/Site)\", fontsize=FONTSIZE+2)\n",
    "ax.set_xlabel(\"Great Circle Distance (km)\", fontsize=FONTSIZE+2)\n",
    "ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0)) \n",
    "    \n",
    "# Plot by Population\n",
    "\n",
    "populations=list(colors_dict.keys())\n",
    "populations.remove(\"1.ORI\"); populations.remove(\"1.PRE\"); populations.remove(\"0.ANT4\");\n",
    "populations.reverse()\n",
    "populations.append(\"1.ORI\"); populations.append(\"1.PRE\"); populations.append(\"0.ANT4\");\n",
    "for population in populations:\n",
    "    \n",
    "    if population not in population_dist_dict: continue\n",
    "    \n",
    "    print(population)        \n",
    "    \n",
    "    # Color\n",
    "    color = colors_dict[population]\n",
    "        \n",
    "    # Extract the upper triangle of the distance matrices\n",
    "    gene_dist_mat = np.array(population_dist_dict[population][\"gene\"])\n",
    "    gene_dist_arr = gene_dist_mat[np.triu_indices(len(gene_dist_mat), k = 1)]\n",
    "\n",
    "    geo_dist_mat = np.array(population_dist_dict[population][\"geo\"])\n",
    "    geo_dist_arr = geo_dist_mat[np.triu_indices(len(geo_dist_mat), k = 1)]\n",
    "\n",
    "    # Add to the all plot\n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        x=geo_dist_arr,\n",
    "        y=gene_dist_arr,\n",
    "        color=color,       \n",
    "        s=50,\n",
    "        ec = \"black\", \n",
    "        lw=0.1,\n",
    "        alpha=0.50,\n",
    "        rasterized=True\n",
    "    )  \n",
    "\n",
    "# Population legend\n",
    "populations = list(colors_dict.keys())\n",
    "populations.remove(NO_DATA_CHAR)\n",
    "populations.reverse()\n",
    "\n",
    "population_lines = [lines.Line2D([0], [0], color=colors_dict[population], lw=0, markersize=10, markeredgecolor=\"black\", markeredgewidth=0.25, marker=\"o\") \n",
    "                for population in populations]\n",
    "\n",
    "population_legend = fig.legend(population_lines, \n",
    "           populations,\n",
    "           #ncol=len(populations),\n",
    "           ncol=1,\n",
    "           loc=\"center\", \n",
    "           bbox_to_anchor=(1.1, 0.5),\n",
    "           frameon=True,\n",
    "           borderpad=0.5, \n",
    "           handletextpad=0.25, \n",
    "           labelspacing=0.5,\n",
    "           edgecolor=\"black\",\n",
    "           title=\"Population\",\n",
    "           fontsize=FONTSIZE,\n",
    "           title_fontsize=FONTSIZE+2,\n",
    "          )\n",
    "#population_legend.get_frame().set_linewidth(0.50)\n",
    "   \n",
    "\n",
    "out_path = os.path.join(out_dir, \"mantel_all\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "under-tuning",
   "metadata": {},
   "source": [
    "## Plot Populations Multi Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "atlantic-purchase",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "nrow=3\n",
    "ncol=4\n",
    "\n",
    "fig, axes = plt.subplots(nrow,ncol,figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.35, hspace=0.80)\n",
    "\n",
    "\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for population in colors_dict:\n",
    "    if population == NO_DATA_CHAR: continue\n",
    "    df = metadata_df[metadata_df[\"population\"] == population]  \n",
    "    print(population)\n",
    "    \n",
    "    if len(df) < 3:\n",
    "        # Update axis\n",
    "        if i_col == ncol - 1:\n",
    "            i_col = 0\n",
    "            i_row += 1\n",
    "        else:\n",
    "            i_col += 1        \n",
    "        continue  \n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "    \n",
    "    frame = legend.get_frame().set_linewidth(0.5)        \n",
    "        \n",
    "    # Format axis\n",
    "    ax.set_title(population)\n",
    "    \n",
    "\n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Genetic Distance \\n (substitutions/site)\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Great Circle Distance (km)\")\n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "    \n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_linewidth(0.5)  \n",
    "        \n",
    "    # Color\n",
    "    color = colors_dict[population]\n",
    "        \n",
    "    # Prepare data\n",
    "    gene_dist_mat = np.array(population_dist_dict[population][\"gene\"])\n",
    "    gene_dist_arr = gene_dist_mat[np.triu_indices(len(gene_dist_mat), k = 1)]\n",
    "\n",
    "    geo_dist_mat = np.array(population_dist_dict[population][\"geo\"])\n",
    "    geo_dist_arr = geo_dist_mat[np.triu_indices(len(geo_dist_mat), k = 1)]\n",
    "\n",
    "    # Add scatter\n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        x=geo_dist_arr,\n",
    "        y=gene_dist_arr,\n",
    "        color=color,       \n",
    "        s=1,\n",
    "        ec = \"black\",      \n",
    "        alpha=1.00,\n",
    "        zorder=2,      \n",
    "    )\n",
    "    \n",
    "    sns.kdeplot(\n",
    "        ax=ax,\n",
    "        x=geo_dist_arr, \n",
    "        y=gene_dist_arr, \n",
    "        fill=True,\n",
    "        zorder=1,\n",
    "        color=color, \n",
    "        alpha=0.5,\n",
    "        thresh=0.05,\n",
    "    )\n",
    "    \n",
    "    # Dummy plot for label\n",
    "    p_sig = \"\"\n",
    "    if population_dist_dict[population][\"mantel\"][\"reject\"]:\n",
    "        p_sig = \"*\"\n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        x=[0],\n",
    "        y=[0],\n",
    "        s=0,\n",
    "        label=(\n",
    "            \"Coefficient: {}\".format(round(population_dist_dict[population][\"mantel\"][\"coeff\"],2))\n",
    "            + \"\\n               p: {:.2e}{}\".format(population_dist_dict[population][\"mantel\"][\"corrected_p_value\"],p_sig)      \n",
    "        + \"\\n               N: {}\".format(population_dist_dict[population][\"mantel\"][\"n\"])              \n",
    "            )          \n",
    "    )\n",
    "    \n",
    "    # Format and position legend\n",
    "    legend = ax.legend(\n",
    "        borderpad=0.8, \n",
    "        handletextpad=-2, \n",
    "        edgecolor=\"black\", \n",
    "        bbox_to_anchor=(0.5, -0.45), \n",
    "        loc='center',\n",
    "        fontsize=FONTSIZE * 0.80,\n",
    "    )\n",
    "    frame = legend.get_frame().set_linewidth(0.5)    \n",
    "            \n",
    "\n",
    "    # Update axis\n",
    "    if i_col == ncol - 1:\n",
    "        i_col = 0\n",
    "        i_row += 1\n",
    "    else:\n",
    "        i_col += 1 \n",
    "        \n",
    "        \n",
    "        \n",
    "out_path = os.path.join(out_dir, \"mantel_populations\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\", dpi=400)\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "demonstrated-provincial",
   "metadata": {},
   "source": [
    "---\n",
    "# 1.ORI Example"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "amber-croatia",
   "metadata": {},
   "source": [
    "## Geographic Distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dying-space",
   "metadata": {},
   "outputs": [],
   "source": [
    "continent_colors_dict = {c:\"\" for c in set(metadata_df[\"continent\"])}\n",
    "\n",
    "# Create the custom color map (pyplot)\n",
    "cmap = plt.get_cmap(\"rainbow\", len(continent_colors_dict))\n",
    "# Convert the color map to a list of RGB values\n",
    "cmaplist = [cmap(i) for i in range(cmap.N)]\n",
    "# Convert RGB values to hex colors\n",
    "attr_hex = [colors.to_hex(col) for col in cmaplist]\n",
    "\n",
    "# Assign colors to value\n",
    "for continent, color in zip(continent_colors_dict, attr_hex):\n",
    "    continent_colors_dict[continent] = color\n",
    "# Add NA\n",
    "continent_colors_dict[\"Intercontinental\"] = \"#c4c4c4\"\n",
    "\n",
    "print(continent_colors_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fundamental-contact",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_list = list(colors_dict.keys())\n",
    "population_list.remove(NO_DATA_CHAR)\n",
    "\n",
    "dist_dict = {}\n",
    "\n",
    "for population in population_list:\n",
    "    print(population)\n",
    "    \n",
    "    dist_dict[population] = {}\n",
    "    \n",
    "    df = metadata_gdf[metadata_gdf[\"population\"] == population]\n",
    "    #human_df = df[df[\"host_order\"] == \"Human\"]\n",
    "    #non_human_df = df[(df[\"host_order\"] != \"Human\") & (df[\"host_order\"] != NO_DATA_CHAR)]\n",
    "    #label_list = [\"human\", \"non_human\"]\n",
    "    #df_list = [human_df, non_human_df]    \n",
    "    #\n",
    "    #if len(human_df) > 0 and len(non_human_df) > 0:\n",
    "    #    all_df = df\n",
    "    #    label_list.append(\"all\")\n",
    "    #    df_list.append(all_df)\n",
    "    label_list = [\"all\"]\n",
    "    df_list = [df]\n",
    "    \n",
    "    for label,df in zip(label_list, df_list):\n",
    "        if len(df) < 3: continue\n",
    "        print(label)\n",
    "        dist_dict[population][label] = {\"gene\":[],\"geo\":[], \"mantel\": {}, \"continent\": []}\n",
    "        \n",
    "        for sample1 in df.index:\n",
    "\n",
    "            # Store distances in list for this sample\n",
    "            geo_dist_list = []\n",
    "            gene_dist_list = []  \n",
    "            continent_list = []\n",
    "\n",
    "            # Get geographic coordinates of sample 1\n",
    "            geom1 = metadata_gdf[\"geometry\"][sample1]\n",
    "            coord1 = (geom1.y, geom1.x)\n",
    "            continent1 = metadata_gdf[\"continent\"][sample1]\n",
    "\n",
    "            for sample2 in df.index:\n",
    "                geom2 = metadata_gdf[\"geometry\"][sample2]\n",
    "                coord2 = (geom2.y, geom2.x) \n",
    "                continent2 = metadata_gdf[\"continent\"][sample2]            \n",
    "\n",
    "                # Calculate geographic and genetic distance\n",
    "                geo_dist = round(geopy.distance.great_circle(coord1, coord2).km, 8)\n",
    "                gene_dist = divtree.distance(sample1, sample2)   \n",
    "\n",
    "                if continent1 != continent2:\n",
    "                    continent = \"Intercontinental\"\n",
    "                else:\n",
    "                    continent = continent1\n",
    "\n",
    "                # Add to arrays\n",
    "                geo_dist_list.append(geo_dist)\n",
    "                gene_dist_list.append(gene_dist) \n",
    "                continent_list.append(continent)\n",
    "\n",
    "\n",
    "            # Add to matrix\n",
    "            dist_dict[population][label][\"geo\"].append(geo_dist_list)\n",
    "            dist_dict[population][label][\"gene\"].append(gene_dist_list)  \n",
    "            dist_dict[population][label][\"continent\"].append(continent_list)  \n",
    "\n",
    "        # Mantel Test\n",
    "        coeff, p_value, n = skbio_dist.mantel(dist_dict[population][label][\"gene\"], dist_dict[population][label][\"geo\"])\n",
    "        dist_dict[population][label][\"mantel\"][\"coeff\"] = coeff\n",
    "        dist_dict[population][label][\"mantel\"][\"p_value\"] = p_value\n",
    "        dist_dict[population][label][\"mantel\"][\"n\"] = n\n",
    "        \n",
    "        print(\"\\t\", dist_dict[population][label][\"mantel\"])\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "functional-franchise",
   "metadata": {},
   "source": [
    "## Correct P Values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "solar-participant",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct p_values\n",
    "raw_p_values = []\n",
    "for population in dist_dict:\n",
    "    for label in dist_dict[population]:\n",
    "        values = dist_dict[population][label][\"mantel\"][\"p_value\"]\n",
    "        raw_p_values.append(values)\n",
    "\n",
    "reject, corrected_p_values, alphacSidak, alphacBonf = smmt.multipletests(raw_p_values, alpha = 0.05, method = \"holm\")\n",
    "\n",
    "i = 0\n",
    "\n",
    "for population in dist_dict:\n",
    "    for label in dist_dict[population]:\n",
    "        label_p_value = corrected_p_values[i]\n",
    "        label_reject = reject[i]\n",
    "        dist_dict[population][label][\"mantel\"][\"corrected_p_value\"] = label_p_value\n",
    "        dist_dict[population][label][\"mantel\"][\"reject\"] = label_reject\n",
    "        i += 1   \n",
    "        \n",
    "        print(population, label, dist_dict[population][label][\"mantel\"])\n",
    "        \n",
    "    \n",
    "# Save results to a file\n",
    "out_path = os.path.join(out_dir, \"mantel_host.tsv\")\n",
    "\n",
    "with open(out_path, \"w\") as outfile:\n",
    "    headers = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\".format(\n",
    "        \"population\",\n",
    "        \"label\",\n",
    "        \"coefficient\",\n",
    "        \"raw_p_value\",\n",
    "        \"n\",\n",
    "        \"corrected_p_value\",\n",
    "        \"reject\",\n",
    "    )\n",
    "    outfile.write(headers + \"\\n\")\n",
    "    \n",
    "    # Population Values\n",
    "    for population in dist_dict:\n",
    "        for label in dist_dict[population]:\n",
    "            data = [population] + [label] + list(dist_dict[population][label][\"mantel\"].values())\n",
    "            data_str = \"\\t\".join([str(d) for d in data])\n",
    "            outfile.write(data_str + \"\\n\")        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "decreased-consistency",
   "metadata": {},
   "source": [
    "## Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "prostate-arthritis",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "FONTSIZE=12\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "\n",
    "for population in dist_dict:\n",
    "    if population != \"1.ORI\": continue\n",
    "    color = colors_dict[population]\n",
    "    for label in dist_dict[population]:\n",
    "        if label != \"all\": continue\n",
    "        print(population, label)\n",
    "        \n",
    "        # Extract the upper triangle of the distance matrices\n",
    "        gene_dist_mat = np.array(dist_dict[population][label][\"gene\"])\n",
    "        gene_dist_arr = gene_dist_mat[np.triu_indices(len(gene_dist_mat), k = 1)]\n",
    "\n",
    "        geo_dist_mat = np.array(dist_dict[population][label][\"geo\"])\n",
    "        geo_dist_arr = geo_dist_mat[np.triu_indices(len(geo_dist_mat), k = 1)]\n",
    "        \n",
    "        continent_mat = np.array(dist_dict[population][label][\"continent\"])\n",
    "        continent_arr = continent_mat[np.triu_indices(len(continent_mat), k = 1)]\n",
    "        continent_colors = [continent_colors_dict[c] for c in continent_arr]\n",
    "\n",
    "        g = sns.JointGrid(       \n",
    "            x=geo_dist_arr,\n",
    "            y=gene_dist_arr,\n",
    "            space=0,\n",
    "        )\n",
    "        \n",
    "        g.plot_joint(\n",
    "            sns.scatterplot,\n",
    "            color=color,     \n",
    "            #c=continent_colors,\n",
    "            s=30,\n",
    "            ec = \"black\",      \n",
    "            alpha=1.00,\n",
    "            zorder=2,   \n",
    "            rasterized=True, \n",
    "        )\n",
    "        \n",
    "        g.plot_joint(\n",
    "            sns.kdeplot,\n",
    "            fill=True,\n",
    "            zorder=1,\n",
    "            color=color,\n",
    "            alpha=0.5,\n",
    "            thresh=0.05,\n",
    "        )    \n",
    "\n",
    "        g.plot_marginals(\n",
    "            sns.kdeplot, \n",
    "            bw_adjust=0.3,\n",
    "            fill=True,\n",
    "            color=color,\n",
    "        )   \n",
    "        \n",
    "        g.fig.suptitle(\"{}\\n{}\".format(population,label), x=0, y=1)\n",
    "        g.ax_joint.set_ylabel(\"Genetic Distance \\n (Substitutions/Site)\", fontsize=FONTSIZE+2)\n",
    "        g.ax_joint.set_xlabel(\"Great Circle Distance\\n(km)\", fontsize=FONTSIZE+2)\n",
    "\n",
    "        #g.ax_joint.set_xlim(-7000,25000)\n",
    "\n",
    "        # Dummy plot for label\n",
    "        p_sig = \"\"\n",
    "        if dist_dict[population][label][\"mantel\"][\"reject\"]:\n",
    "            p_sig = \"*\"\n",
    "        sns.scatterplot(\n",
    "            ax=g.ax_joint,\n",
    "            x=[0],\n",
    "            y=[0],\n",
    "            s=0,\n",
    "            label=(\n",
    "                \"Coefficient: {}\".format(round(dist_dict[population][label][\"mantel\"][\"coeff\"],2))\n",
    "                + \"\\n               p: {:.2e}{}\".format(dist_dict[population][label][\"mantel\"][\"corrected_p_value\"],p_sig)      \n",
    "            + \"\\n               N: {}\".format(dist_dict[population][label][\"mantel\"][\"n\"])              \n",
    "                )          \n",
    "        )\n",
    "\n",
    "        # Format and position legend\n",
    "        legend = g.ax_joint.legend(\n",
    "            borderpad=0.8, \n",
    "            handletextpad=-2, \n",
    "            edgecolor=\"black\", \n",
    "            bbox_to_anchor=(0.5, -0.3), \n",
    "            loc='center',\n",
    "            fontsize=FONTSIZE - 2,\n",
    "        )\n",
    "\n",
    "\n",
    "        out_path = os.path.join(out_dir, \"mantel_{}_{}\".format(population, label))\n",
    "        print(out_path)\n",
    "        #plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "        #plt.savefig(out_path + \".svg\", bbox_inches=\"tight\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hungarian-rolling",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# 4. Timeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fourth-counter",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 240]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "fig, axes = plt.subplots(1,4,figsize=FIGSIZE, dpi=DPI, sharey=False)\n",
    "fig.subplots_adjust(wspace=0.25)\n",
    "\n",
    "# Bronze Age\n",
    "ax = axes[-4]\n",
    "ax.set_title(\"Bronze Age\")\n",
    "ax_range = [-3001, -1]\n",
    "ax.set_xlim(ax_range)\n",
    "ax.set_ylim(0,80)\n",
    "df = metadata_gdf[(metadata_gdf[\"date_mean\"] > ax_range[0]) & (metadata_gdf[\"date_mean\"] <= ax_range[1]) ]\n",
    "sns.histplot(\n",
    "    ax=ax, \n",
    "    x=df[\"date_mean\"], \n",
    "    multiple=\"stack\",    \n",
    "    hue=df[\"population\"], \n",
    "    palette=colors_dict, \n",
    "    binwidth=200,\n",
    "    alpha=1.0,\n",
    "    lw=0.25,\n",
    ")\n",
    "ax.legend().remove()\n",
    "\n",
    "# Roman\n",
    "ax = axes[-3]\n",
    "ax.set_title(\"Late Antiquity\")\n",
    "ax.set_ylim(0,80)\n",
    "\n",
    "ax_range = [0, 999]\n",
    "ax.set_xlim(ax_range)\n",
    "df = metadata_gdf[(metadata_gdf[\"date_mean\"] > ax_range[0]) & (metadata_gdf[\"date_mean\"] <= ax_range[1]) ]\n",
    "sns.histplot(\n",
    "    ax=ax, \n",
    "    x=df[\"date_mean\"], \n",
    "    multiple=\"stack\",    \n",
    "    hue=df[\"population\"], \n",
    "    palette=colors_dict, \n",
    "    binwidth=50,\n",
    "    alpha=1.0,\n",
    "    lw=0.25,\n",
    ")\n",
    "ax.legend().remove()\n",
    "\n",
    "\n",
    "# Medieval\n",
    "ax = axes[-2]\n",
    "ax.set_title(\"Medieval/Early Modern\")\n",
    "ax.set_ylim(0,80)\n",
    "ax_range = [999, 1899]\n",
    "ax.set_xlim(ax_range)\n",
    "df = metadata_gdf[(metadata_gdf[\"date_mean\"] > ax_range[0]) & (metadata_gdf[\"date_mean\"] <= ax_range[1]) ]\n",
    "sns.histplot(\n",
    "    ax=ax, \n",
    "    x=df[\"date_mean\"], \n",
    "    multiple=\"stack\",    \n",
    "    hue=df[\"population\"], \n",
    "    palette=colors_dict, \n",
    "    binwidth=50,\n",
    "    alpha=1.0,\n",
    "    lw=0.25,\n",
    ")\n",
    "ax.legend().remove()\n",
    "\n",
    "# Modern\n",
    "ax = axes[-1]\n",
    "ax.set_title(\"Modern\")\n",
    "ax_range = [1899, 2020]\n",
    "ax.set_xlim(ax_range)\n",
    "ax.set_ylim(0,80)\n",
    "df = metadata_gdf[(metadata_gdf[\"date_mean\"] > ax_range[0]) & (metadata_gdf[\"date_mean\"] <= ax_range[1]) ]\n",
    "sns.histplot(\n",
    "    ax=ax, \n",
    "    x=df[\"date_mean\"], \n",
    "    multiple=\"stack\",    \n",
    "    hue=df[\"population\"], \n",
    "    palette=colors_dict, \n",
    "    binwidth=10,\n",
    "    alpha=1.0,\n",
    "    lw=0.25,\n",
    ")\n",
    "ax.legend().remove()\n",
    "\n",
    "for ax in axes:\n",
    "    ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Collection Date\")\n",
    "axes[0].set_ylabel(\"Number of Genomes\")\n",
    "\n",
    "for ax in axes:\n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_linewidth(0)\n",
    "    ax.spines[\"bottom\"].set_linewidth(0.5)\n",
    "    ax.spines[\"left\"].set_linewidth(0.5)\n",
    "    \n",
    "# Population legend\n",
    "populations = list(colors_dict.keys())\n",
    "populations.remove(NO_DATA_CHAR)\n",
    "populations.reverse()\n",
    "\n",
    "population_lines = [lines.Line2D([0], [0], color=colors_dict[population], lw=0, markersize=5, markeredgecolor=\"black\", markeredgewidth=0.25, marker=\"s\") \n",
    "                for population in populations]\n",
    "\n",
    "population_legend = fig.legend(population_lines, \n",
    "           populations,\n",
    "           #ncol=len(populations),\n",
    "           ncol=1,\n",
    "           loc=\"center\", \n",
    "           bbox_to_anchor=(0.95, 0.50),\n",
    "           frameon=True,\n",
    "           borderpad=0.5, \n",
    "           handletextpad=0.25, \n",
    "           labelspacing=0.5,\n",
    "           edgecolor=\"black\",\n",
    "           title=\"Population\",\n",
    "           fontsize=FONTSIZE,\n",
    "           title_fontsize=FONTSIZE,\n",
    "          )\n",
    "population_legend.get_frame().set_linewidth(0.50)\n",
    "    \n",
    "out_path = os.path.join(out_dir, \"timeline_all\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "composed-house",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "useful-passport",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incident-feeding",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_metadata = os.path.join(out_dir, \"metadata.tsv\")\n",
    "metadata_df.to_csv(out_path_metadata, sep=\"\\t\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sublime-chocolate",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
