{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "juvenile-testimony",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "1. Plot Locus Coverage of Whole Genome.\n",
    "1. Add Locus Coverage to Dataframe.\n",
    "1. Plot Locus Coverage: Distribution.\n",
    "1. Plot Locus Coverage: Phlyo.\n",
    "1. Identify Interesting Samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "excellent-saver",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "serious-cannon",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "effective-train",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import dill\n",
    "from Bio import Phylo\n",
    "import copy\n",
    "import math\n",
    "import pandas as pd\n",
    "from matplotlib import lines, patches, colors\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "departmental-bibliography",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "covered-secretariat",
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "\n",
    "# Custom script variables\n",
    "SCRIPT_NAME = \"locus_coverage_collect\"\n",
    "PREV_DIR_NAME = \"mugration\"\n",
    "PREV_SCRIPT_NAME = \"mugration_model\"\n",
    "\n",
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"50\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "MISSING_DATA = WILDCARDS[2]\n",
    "\n",
    "NAME_COL = \"Sample\"\n",
    "\n",
    "pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deadly-legislature",
   "metadata": {},
   "source": [
    "## File Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "concerned-train",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tree_dill:\t /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/mugration/all/chromosome_filter50/mugration_model_timetree.treetime.obj\n",
      "tree_df_dill:\t /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/mugration/all/chromosome_filter50/mugration_model.df.obj\n",
      "aln path:\t /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/snippy_multi/all/snippy-core_chromosome.snps.filter50.aln\n",
      "auspice_latlon_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome_filter50/parse_tree_latlon.tsv\n",
      "auspice_colors_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome_filter50/parse_tree_colors.tsv\n",
      "auspice_config_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/config/auspice_config.json\n",
      "auspice_remote_dir_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/auspice/\n",
      "outdir: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/locus_coverage_collect/all/chromosome_filter50\n"
     ]
    }
   ],
   "source": [
    "config_dir = os.path.join(project_dir, \"config\")\n",
    "results_dir = os.path.join(project_dir, \"results\")\n",
    "\n",
    "\n",
    "outdir       = os.path.join(results_dir, SCRIPT_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA),)\n",
    "tree_dill    = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA),PREV_SCRIPT_NAME + \"_timetree.treetime.obj\")\n",
    "tree_df_dill = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), PREV_SCRIPT_NAME + \".df.obj\")\n",
    "tree_df_path = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), PREV_SCRIPT_NAME + \".tsv\")\n",
    "aln_path     = os.path.join(results_dir,\"snippy_multi\",READS_ORIGIN,\"snippy-core_{}.snps.filter{}.aln\".format(LOCUS_NAME, MISSING_DATA))\n",
    "\n",
    "# Auspice\n",
    "auspice_latlon_path = os.path.join(results_dir, \"parse_tree\", READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), \"parse_tree\" + \"_latlon.tsv\")\n",
    "auspice_colors_path = os.path.join(results_dir, \"parse_tree\", READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), \"parse_tree\" + \"_colors.tsv\")\n",
    "auspice_config_path = os.path.join(config_dir, \"auspice_config.json\")\n",
    "auspice_remote_dir_path = os.path.join(project_dir, \"auspice/\")\n",
    "\n",
    "# Uniq\n",
    "cov_df_path  = os.path.join(results_dir, SCRIPT_NAME, READS_ORIGIN, \"locus_coverage\" + \".txt\")\n",
    "dep_df_path  = os.path.join(results_dir, SCRIPT_NAME, READS_ORIGIN, \"locus_depth\" + \".txt\")\n",
    "\n",
    "print(\"tree_dill:\\t\", tree_dill)\n",
    "print(\"tree_df_dill:\\t\", tree_df_dill)\n",
    "print(\"aln path:\\t\", aln_path)\n",
    "print(\"auspice_latlon_path:\", auspice_latlon_path)\n",
    "print(\"auspice_colors_path:\", auspice_colors_path)\n",
    "print(\"auspice_config_path:\", auspice_config_path)\n",
    "print(\"auspice_remote_dir_path:\", auspice_remote_dir_path)\n",
    "print(\"outdir:\", outdir)\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "while not os.path.exists(outdir):\n",
    "    os.makedirs(outdir)   \n",
    "    \n",
    "SCRIPT_NAME = \"locus_coverage_collect\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "responsible-joshua",
   "metadata": {},
   "source": [
    "## Import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "quality-emergency",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(tree_dill, \"rb\") as infile:\n",
    "    tt = dill.load(infile)\n",
    "tt.tree.ladderize(reverse=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "played-wesley",
   "metadata": {},
   "source": [
    "## Import Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "another-audit",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "      <th>timetree_rate</th>\n",
       "      <th>timetree_rate_fold_change</th>\n",
       "      <th>timetree_mutation_length</th>\n",
       "      <th>timetree_raw_date</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>timetree_num_date_bar</th>\n",
       "      <th>timetree_clock_length</th>\n",
       "      <th>timetree_coord_x</th>\n",
       "      <th>timetree_coord_y</th>\n",
       "      <th>timetree_reg_x</th>\n",
       "      <th>timetree_reg_y</th>\n",
       "      <th>timetree_reg_bad</th>\n",
       "      <th>Mugration_Branch_Number</th>\n",
       "      <th>Mugration_Branch_Number_Confidence</th>\n",
       "      <th>Mugration_Branch_Number_Entropy</th>\n",
       "      <th>Mugration_Branch_Major</th>\n",
       "      <th>Mugration_Branch_Major_Confidence</th>\n",
       "      <th>Mugration_Branch_Major_Entropy</th>\n",
       "      <th>Mugration_Branch_Minor</th>\n",
       "      <th>Mugration_Branch_Minor_Confidence</th>\n",
       "      <th>Mugration_Branch_Minor_Entropy</th>\n",
       "      <th>Mugration_Country</th>\n",
       "      <th>Mugration_Country_Confidence</th>\n",
       "      <th>Mugration_Country_Entropy</th>\n",
       "      <th>Mugration_Country_Lat</th>\n",
       "      <th>Mugration_Country_Lon</th>\n",
       "      <th>Mugration_Province</th>\n",
       "      <th>Mugration_Province_Confidence</th>\n",
       "      <th>Mugration_Province_Entropy</th>\n",
       "      <th>Mugration_Province_Lat</th>\n",
       "      <th>Mugration_Province_Lon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>black</td>\n",
       "      <td>2.148815e-07</td>\n",
       "      <td>2.25</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.942141e-08</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.148815e-07</td>\n",
       "      <td>0855-09-28</td>\n",
       "      <td>855.740923</td>\n",
       "      <td>[855.7409234792349, 855.7409234792349]</td>\n",
       "      <td>[1165.2590765207651, 1165.2590765207651]</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>2.25</td>\n",
       "      <td>855.740923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>Russia</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.447080e-05</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.278202e-05</td>\n",
       "      <td>1.00</td>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>2.052208e-08</td>\n",
       "      <td>1.056673</td>\n",
       "      <td>2.256714e-05</td>\n",
       "      <td>1938-03-30</td>\n",
       "      <td>1938.242634</td>\n",
       "      <td>[1923.0, 1923.0]</td>\n",
       "      <td>[98.0, 98.0]</td>\n",
       "      <td>1082.501711</td>\n",
       "      <td>1082.601711</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1923.000000</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>Russia</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.307300e-05</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.149205e-05</td>\n",
       "      <td>2.00</td>\n",
       "      <td>NA</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>NA</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>1.821491e-08</td>\n",
       "      <td>0.937878</td>\n",
       "      <td>2.127716e-05</td>\n",
       "      <td>1996-07-12</td>\n",
       "      <td>1996.529295</td>\n",
       "      <td>[1992.0, 1992.0]</td>\n",
       "      <td>[29.0, 29.0]</td>\n",
       "      <td>1140.788371</td>\n",
       "      <td>1140.888371</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1992.000000</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>2.721210e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>2.536025e-05</td>\n",
       "      <td>3.50</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.136642e-08</td>\n",
       "      <td>1.100148</td>\n",
       "      <td>2.514537e-05</td>\n",
       "      <td>2005-09-05</td>\n",
       "      <td>2005.678813</td>\n",
       "      <td>[1985.457689492314, 2005.0]</td>\n",
       "      <td>[35.5423105076859, 16.0]</td>\n",
       "      <td>1149.937889</td>\n",
       "      <td>1150.037889</td>\n",
       "      <td>3.50</td>\n",
       "      <td>2005.678813</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>China</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.536025e-05</td>\n",
       "      <td>3.00</td>\n",
       "      <td>42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>5.812589e-09</td>\n",
       "      <td>0.299288</td>\n",
       "      <td>2.828252e-18</td>\n",
       "      <td>2008-05-10</td>\n",
       "      <td>2008.355574</td>\n",
       "      <td>[2006.0, 2006.0]</td>\n",
       "      <td>[15.0, 15.0]</td>\n",
       "      <td>2.676761</td>\n",
       "      <td>1152.714650</td>\n",
       "      <td>3.00</td>\n",
       "      <td>2006.000000</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>China</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.331000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.557514e-05</td>\n",
       "      <td>4.00</td>\n",
       "      <td>42123</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>3.247561e-08</td>\n",
       "      <td>1.672155</td>\n",
       "      <td>2.148899e-07</td>\n",
       "      <td>2007-07-06</td>\n",
       "      <td>2007.510839</td>\n",
       "      <td>[2005.0, 2005.0]</td>\n",
       "      <td>[16.0, 16.0]</td>\n",
       "      <td>1.832026</td>\n",
       "      <td>1151.869916</td>\n",
       "      <td>4.00</td>\n",
       "      <td>2005.000000</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>China</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    UFboot sCF  Branch_Length  \\\n",
       "Name                                                            \n",
       "NODE0                                   NA  NA   0.000000e+00   \n",
       "GCA_009909635.1_ASM990963v1_genomic     NA  NA   2.447080e-05   \n",
       "Reference                               NA  NA   2.307300e-05   \n",
       "NODE1                                   91   0   2.721210e-05   \n",
       "GCA_009669545.1_ASM966954v1_genomic     NA  NA   2.330000e-08   \n",
       "GCA_009669555.1_ASM966955v1_genomic     NA  NA   2.331000e-07   \n",
       "\n",
       "                                    Branch_Support_Color       coord_x  \\\n",
       "Name                                                                     \n",
       "NODE0                                              black  2.148815e-07   \n",
       "GCA_009909635.1_ASM990963v1_genomic                 grey  2.278202e-05   \n",
       "Reference                                           grey  2.149205e-05   \n",
       "NODE1                                              black  2.536025e-05   \n",
       "GCA_009669545.1_ASM966954v1_genomic                 grey  2.536025e-05   \n",
       "GCA_009669555.1_ASM966955v1_genomic                 grey  2.557514e-05   \n",
       "\n",
       "                                     coord_y Strain  Date DateBP  \\\n",
       "Name                                                               \n",
       "NODE0                                   2.25     NA    NA     NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic     1.00   9_10  1923    -98   \n",
       "Reference                               2.00     NA  1992    -29   \n",
       "NODE1                                   3.50     NA    NA     NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic     3.00  42126  2006    -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic     4.00  42123  2005    -16   \n",
       "\n",
       "                                                      Country       Province  \\\n",
       "Name                                                                           \n",
       "NODE0                                                      NA             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia  Rostov Oblast   \n",
       "Reference                            United States of America       Colorado   \n",
       "NODE1                                                      NA             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China       Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China       Xinjiang   \n",
       "\n",
       "                                    CountryLat CountryLon ProvinceLat  \\\n",
       "Name                                                                    \n",
       "NODE0                                       NA         NA          NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic    64.6863    97.7453     47.6222   \n",
       "Reference                              39.7837   -100.446     38.7252   \n",
       "NODE1                                       NA         NA          NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic    35.0001        105     42.4805   \n",
       "GCA_009669555.1_ASM966955v1_genomic    35.0001        105     42.4805   \n",
       "\n",
       "                                    ProvinceLon      Biovar Branch_Major  \\\n",
       "Name                                                                       \n",
       "NODE0                                        NA          NA           NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic     40.7958  Medievalis        2.MED   \n",
       "Reference                              -105.608  Orientalis        1.ORI   \n",
       "NODE1                                        NA          NA           NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic     85.4633     Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic     85.4633     Antiqua        0.ANT   \n",
       "\n",
       "                                    Branch_Minor     BioSample  \\\n",
       "Name                                                             \n",
       "NODE0                                         NA            NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1  SAMN13632815   \n",
       "Reference                                 1.ORI1            NA   \n",
       "NODE1                                         NA            NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1  SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1  SAMN07722924   \n",
       "\n",
       "                                                    BioSampleComment  \\\n",
       "Name                                                                   \n",
       "NODE0                                                             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic            KEEP: Assembly Modern   \n",
       "Reference                            KEEP: Assembly Modern Reference   \n",
       "NODE1                                                             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic            KEEP: Assembly Modern   \n",
       "\n",
       "                                    Branch_Number  timetree_rate  \\\n",
       "Name                                                               \n",
       "NODE0                                          NA   1.942141e-08   \n",
       "GCA_009909635.1_ASM990963v1_genomic             2   2.052208e-08   \n",
       "Reference                                       1   1.821491e-08   \n",
       "NODE1                                          NA   2.136642e-08   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0   5.812589e-09   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0   3.247561e-08   \n",
       "\n",
       "                                     timetree_rate_fold_change  \\\n",
       "Name                                                             \n",
       "NODE0                                                 1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   1.056673   \n",
       "Reference                                             0.937878   \n",
       "NODE1                                                 1.100148   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   0.299288   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   1.672155   \n",
       "\n",
       "                                     timetree_mutation_length  \\\n",
       "Name                                                            \n",
       "NODE0                                            2.148815e-07   \n",
       "GCA_009909635.1_ASM990963v1_genomic              2.256714e-05   \n",
       "Reference                                        2.127716e-05   \n",
       "NODE1                                            2.514537e-05   \n",
       "GCA_009669545.1_ASM966954v1_genomic              2.828252e-18   \n",
       "GCA_009669555.1_ASM966955v1_genomic              2.148899e-07   \n",
       "\n",
       "                                    timetree_raw_date  timetree_num_date  \\\n",
       "Name                                                                       \n",
       "NODE0                                      0855-09-28         855.740923   \n",
       "GCA_009909635.1_ASM990963v1_genomic        1938-03-30        1938.242634   \n",
       "Reference                                  1996-07-12        1996.529295   \n",
       "NODE1                                      2005-09-05        2005.678813   \n",
       "GCA_009669545.1_ASM966954v1_genomic        2008-05-10        2008.355574   \n",
       "GCA_009669555.1_ASM966955v1_genomic        2007-07-06        2007.510839   \n",
       "\n",
       "                                               timetree_num_date_confidence  \\\n",
       "Name                                                                          \n",
       "NODE0                                [855.7409234792349, 855.7409234792349]   \n",
       "GCA_009909635.1_ASM990963v1_genomic                        [1923.0, 1923.0]   \n",
       "Reference                                                  [1992.0, 1992.0]   \n",
       "NODE1                                           [1985.457689492314, 2005.0]   \n",
       "GCA_009669545.1_ASM966954v1_genomic                        [2006.0, 2006.0]   \n",
       "GCA_009669555.1_ASM966955v1_genomic                        [2005.0, 2005.0]   \n",
       "\n",
       "                                                        timetree_num_date_bar  \\\n",
       "Name                                                                            \n",
       "NODE0                                [1165.2590765207651, 1165.2590765207651]   \n",
       "GCA_009909635.1_ASM990963v1_genomic                              [98.0, 98.0]   \n",
       "Reference                                                        [29.0, 29.0]   \n",
       "NODE1                                                [35.5423105076859, 16.0]   \n",
       "GCA_009669545.1_ASM966954v1_genomic                              [15.0, 15.0]   \n",
       "GCA_009669555.1_ASM966955v1_genomic                              [16.0, 16.0]   \n",
       "\n",
       "                                     timetree_clock_length  timetree_coord_x  \\\n",
       "Name                                                                           \n",
       "NODE0                                             0.100000          0.100000   \n",
       "GCA_009909635.1_ASM990963v1_genomic            1082.501711       1082.601711   \n",
       "Reference                                      1140.788371       1140.888371   \n",
       "NODE1                                          1149.937889       1150.037889   \n",
       "GCA_009669545.1_ASM966954v1_genomic               2.676761       1152.714650   \n",
       "GCA_009669555.1_ASM966955v1_genomic               1.832026       1151.869916   \n",
       "\n",
       "                                     timetree_coord_y  timetree_reg_x  \\\n",
       "Name                                                                    \n",
       "NODE0                                            2.25      855.740923   \n",
       "GCA_009909635.1_ASM990963v1_genomic              1.00     1923.000000   \n",
       "Reference                                        2.00     1992.000000   \n",
       "NODE1                                            3.50     2005.678813   \n",
       "GCA_009669545.1_ASM966954v1_genomic              3.00     2006.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic              4.00     2005.000000   \n",
       "\n",
       "                                     timetree_reg_y  timetree_reg_bad  \\\n",
       "Name                                                                    \n",
       "NODE0                                      0.000000             False   \n",
       "GCA_009909635.1_ASM990963v1_genomic        0.000023             False   \n",
       "Reference                                  0.000021             False   \n",
       "NODE1                                      0.000025             False   \n",
       "GCA_009669545.1_ASM966954v1_genomic        0.000025             False   \n",
       "GCA_009669555.1_ASM966955v1_genomic        0.000025             False   \n",
       "\n",
       "                                    Mugration_Branch_Number  \\\n",
       "Name                                                          \n",
       "NODE0                                                     1   \n",
       "GCA_009909635.1_ASM990963v1_genomic                       2   \n",
       "Reference                                                 1   \n",
       "NODE1                                                     0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                       0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                       0   \n",
       "\n",
       "                                    Mugration_Branch_Number_Confidence  \\\n",
       "Name                                                                     \n",
       "NODE0                                                         0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                                  1   \n",
       "Reference                                                            1   \n",
       "NODE1                                                         0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                                  1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                                  1   \n",
       "\n",
       "                                    Mugration_Branch_Number_Entropy  \\\n",
       "Name                                                                  \n",
       "NODE0                                                       1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    -1.00009e-12   \n",
       "Reference                                              -1.00009e-12   \n",
       "NODE1                                                      0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic                    -9.99867e-13   \n",
       "GCA_009669555.1_ASM966955v1_genomic                    -9.99867e-13   \n",
       "\n",
       "                                    Mugration_Branch_Major  \\\n",
       "Name                                                         \n",
       "NODE0                                                1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic                  2.MED   \n",
       "Reference                                            1.ORI   \n",
       "NODE1                                                0.ANT   \n",
       "GCA_009669545.1_ASM966954v1_genomic                  0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic                  0.ANT   \n",
       "\n",
       "                                    Mugration_Branch_Major_Confidence  \\\n",
       "Name                                                                    \n",
       "NODE0                                                        0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                                 1   \n",
       "Reference                                                           1   \n",
       "NODE1                                                        0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                                 1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                                 1   \n",
       "\n",
       "                                    Mugration_Branch_Major_Entropy  \\\n",
       "Name                                                                 \n",
       "NODE0                                                      1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   -1.00009e-12   \n",
       "Reference                                             -1.00009e-12   \n",
       "NODE1                                                     0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   -9.99867e-13   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   -9.99867e-13   \n",
       "\n",
       "                                    Mugration_Branch_Minor  \\\n",
       "Name                                                         \n",
       "NODE0                                               1.ORI1   \n",
       "GCA_009909635.1_ASM990963v1_genomic                 2.MED1   \n",
       "Reference                                           1.ORI1   \n",
       "NODE1                                               0.ANT1   \n",
       "GCA_009669545.1_ASM966954v1_genomic                 0.ANT1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                 0.ANT1   \n",
       "\n",
       "                                    Mugration_Branch_Minor_Confidence  \\\n",
       "Name                                                                    \n",
       "NODE0                                                        0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                                 1   \n",
       "Reference                                                           1   \n",
       "NODE1                                                        0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                                 1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                                 1   \n",
       "\n",
       "                                    Mugration_Branch_Minor_Entropy  \\\n",
       "Name                                                                 \n",
       "NODE0                                                      1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   -1.00009e-12   \n",
       "Reference                                             -1.00009e-12   \n",
       "NODE1                                                     0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   -9.99867e-13   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   -9.99867e-13   \n",
       "\n",
       "                                            Mugration_Country  \\\n",
       "Name                                                            \n",
       "NODE0                                                  Russia   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "Reference                            United States of America   \n",
       "NODE1                                                   China   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "\n",
       "                                    Mugration_Country_Confidence  \\\n",
       "Name                                                               \n",
       "NODE0                                                   0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                            1   \n",
       "Reference                                                      1   \n",
       "NODE1                                                   0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                            1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                            1   \n",
       "\n",
       "                                    Mugration_Country_Entropy  \\\n",
       "Name                                                            \n",
       "NODE0                                                 1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic              -1.00009e-12   \n",
       "Reference                                        -1.00009e-12   \n",
       "NODE1                                                0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic              -1.00009e-12   \n",
       "GCA_009669555.1_ASM966955v1_genomic              -1.00009e-12   \n",
       "\n",
       "                                    Mugration_Country_Lat  \\\n",
       "Name                                                        \n",
       "NODE0                                             64.6863   \n",
       "GCA_009909635.1_ASM990963v1_genomic               64.6863   \n",
       "Reference                                         39.7837   \n",
       "NODE1                                             35.0001   \n",
       "GCA_009669545.1_ASM966954v1_genomic               35.0001   \n",
       "GCA_009669555.1_ASM966955v1_genomic               35.0001   \n",
       "\n",
       "                                    Mugration_Country_Lon Mugration_Province  \\\n",
       "Name                                                                           \n",
       "NODE0                                             97.7453      Rostov Oblast   \n",
       "GCA_009909635.1_ASM990963v1_genomic               97.7453      Rostov Oblast   \n",
       "Reference                                        -100.446           Colorado   \n",
       "NODE1                                                 105           Xinjiang   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   105           Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   105           Xinjiang   \n",
       "\n",
       "                                    Mugration_Province_Confidence  \\\n",
       "Name                                                                \n",
       "NODE0                                                    0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                             1   \n",
       "Reference                                                       1   \n",
       "NODE1                                                    0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             1   \n",
       "\n",
       "                                    Mugration_Province_Entropy  \\\n",
       "Name                                                             \n",
       "NODE0                                                  1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic               -1.00009e-12   \n",
       "Reference                                         -1.00009e-12   \n",
       "NODE1                                                 0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic               -1.00009e-12   \n",
       "GCA_009669555.1_ASM966955v1_genomic               -1.00009e-12   \n",
       "\n",
       "                                    Mugration_Province_Lat  \\\n",
       "Name                                                         \n",
       "NODE0                                              47.6222   \n",
       "GCA_009909635.1_ASM990963v1_genomic                47.6222   \n",
       "Reference                                          38.7252   \n",
       "NODE1                                              42.4805   \n",
       "GCA_009669545.1_ASM966954v1_genomic                42.4805   \n",
       "GCA_009669555.1_ASM966955v1_genomic                42.4805   \n",
       "\n",
       "                                    Mugration_Province_Lon  \n",
       "Name                                                        \n",
       "NODE0                                              40.7958  \n",
       "GCA_009909635.1_ASM990963v1_genomic                40.7958  \n",
       "Reference                                         -105.608  \n",
       "NODE1                                              85.4633  \n",
       "GCA_009669545.1_ASM966954v1_genomic                85.4633  \n",
       "GCA_009669555.1_ASM966955v1_genomic                85.4633  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(tree_df_dill, \"rb\") as infile:\n",
    "    tree_df = dill.load(infile)\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "communist-marketing",
   "metadata": {},
   "source": [
    "## Import Locus Dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "killing-commissioner",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AL590842.1:1..4653728</th>\n",
       "      <th>gene-YPO0001</th>\n",
       "      <th>cds-CAL18691.1</th>\n",
       "      <th>gene-YPO0002</th>\n",
       "      <th>cds-CAL18692.1</th>\n",
       "      <th>...</th>\n",
       "      <th>gene-YPPCP1.08c</th>\n",
       "      <th>cds-CAB53171.1</th>\n",
       "      <th>gene-YPPCP1.09c</th>\n",
       "      <th>cds-CAB53172.1</th>\n",
       "      <th>id-AL109969.1:8528..8532</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>0.919553</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>0.947885</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>0.945635</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18878 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     AL590842.1:1..4653728  gene-YPO0001  \\\n",
       "Sample                                                                     \n",
       "GCA_009909635.1_ASM990963v1_genomic               0.919553           1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic               0.947885           1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic               0.945635           1.0   \n",
       "\n",
       "                                     cds-CAL18691.1  gene-YPO0002  \\\n",
       "Sample                                                              \n",
       "GCA_009909635.1_ASM990963v1_genomic             1.0           1.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic             1.0           1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic             1.0           1.0   \n",
       "\n",
       "                                     cds-CAL18692.1  ...  gene-YPPCP1.08c  \\\n",
       "Sample                                               ...                    \n",
       "GCA_009909635.1_ASM990963v1_genomic             1.0  ...              0.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic             1.0  ...              1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic             1.0  ...              1.0   \n",
       "\n",
       "                                     cds-CAB53171.1  gene-YPPCP1.09c  \\\n",
       "Sample                                                                 \n",
       "GCA_009909635.1_ASM990963v1_genomic             0.0              0.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic             1.0              1.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic             1.0              1.0   \n",
       "\n",
       "                                     cds-CAB53172.1  id-AL109969.1:8528..8532  \n",
       "Sample                                                                         \n",
       "GCA_009909635.1_ASM990963v1_genomic             0.0                       0.0  \n",
       "GCA_009669545.1_ASM966954v1_genomic             1.0                       1.0  \n",
       "GCA_009669555.1_ASM966955v1_genomic             1.0                       1.0  \n",
       "\n",
       "[3 rows x 18878 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AL590842.1:1..4653728</th>\n",
       "      <th>gene-YPO0001</th>\n",
       "      <th>cds-CAL18691.1</th>\n",
       "      <th>gene-YPO0002</th>\n",
       "      <th>cds-CAL18692.1</th>\n",
       "      <th>...</th>\n",
       "      <th>gene-YPPCP1.08c</th>\n",
       "      <th>cds-CAB53171.1</th>\n",
       "      <th>gene-YPPCP1.09c</th>\n",
       "      <th>cds-CAB53172.1</th>\n",
       "      <th>id-AL109969.1:8528..8532</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>18.131132</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>18.714769</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>18.661659</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>19.200001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18878 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     AL590842.1:1..4653728  gene-YPO0001  \\\n",
       "Sample                                                                     \n",
       "GCA_009909635.1_ASM990963v1_genomic              18.131132          20.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic              18.714769          20.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic              18.661659          20.0   \n",
       "\n",
       "                                     cds-CAL18691.1  gene-YPO0002  \\\n",
       "Sample                                                              \n",
       "GCA_009909635.1_ASM990963v1_genomic            20.0          20.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic            20.0          20.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic            20.0          20.0   \n",
       "\n",
       "                                     cds-CAL18692.1  ...  gene-YPPCP1.08c  \\\n",
       "Sample                                               ...                    \n",
       "GCA_009909635.1_ASM990963v1_genomic            20.0  ...              0.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic            20.0  ...             20.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic            20.0  ...             20.0   \n",
       "\n",
       "                                     cds-CAB53171.1  gene-YPPCP1.09c  \\\n",
       "Sample                                                                 \n",
       "GCA_009909635.1_ASM990963v1_genomic             0.0              0.0   \n",
       "GCA_009669545.1_ASM966954v1_genomic            20.0             20.0   \n",
       "GCA_009669555.1_ASM966955v1_genomic            20.0             20.0   \n",
       "\n",
       "                                     cds-CAB53172.1  id-AL109969.1:8528..8532  \n",
       "Sample                                                                         \n",
       "GCA_009909635.1_ASM990963v1_genomic             0.0                  0.000000  \n",
       "GCA_009669545.1_ASM966954v1_genomic            20.0                 20.000000  \n",
       "GCA_009669555.1_ASM966955v1_genomic            20.0                 19.200001  \n",
       "\n",
       "[3 rows x 18878 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cov_df = pd.read_csv(cov_df_path, sep='\\t')\n",
    "cov_df.set_index(NAME_COL, inplace=True)\n",
    "cov_df_samples = cov_df.index\n",
    "\n",
    "dep_df = pd.read_csv(dep_df_path, sep='\\t')\n",
    "dep_df.set_index(NAME_COL, inplace=True)\n",
    "dep_df_samples = dep_df.index\n",
    "\n",
    "# Remove samples not in tree dataframe\n",
    "for sample in cov_df_samples:\n",
    "    if sample not in tree_df.index:\n",
    "        cov_df.drop(sample, inplace=True)\n",
    "        dep_df.drop(sample, inplace=True)\n",
    "        \n",
    "display(cov_df)        \n",
    "display(dep_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "vertical-checkout",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AL590842.1:1..4653728</th>\n",
       "      <th>AL117189.1:1..70305</th>\n",
       "      <th>AL117211.1:1..96210</th>\n",
       "      <th>AL109969.1:1..9612</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>0.919553</td>\n",
       "      <td>0.919181</td>\n",
       "      <td>0.938385</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>0.947885</td>\n",
       "      <td>0.941384</td>\n",
       "      <td>0.941461</td>\n",
       "      <td>0.800666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>0.945635</td>\n",
       "      <td>0.931143</td>\n",
       "      <td>0.940838</td>\n",
       "      <td>0.800250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     AL590842.1:1..4653728  \\\n",
       "Sample                                                       \n",
       "GCA_009909635.1_ASM990963v1_genomic               0.919553   \n",
       "GCA_009669545.1_ASM966954v1_genomic               0.947885   \n",
       "GCA_009669555.1_ASM966955v1_genomic               0.945635   \n",
       "\n",
       "                                     AL117189.1:1..70305  AL117211.1:1..96210  \\\n",
       "Sample                                                                          \n",
       "GCA_009909635.1_ASM990963v1_genomic             0.919181             0.938385   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0.941384             0.941461   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0.931143             0.940838   \n",
       "\n",
       "                                     AL109969.1:1..9612  \n",
       "Sample                                                   \n",
       "GCA_009909635.1_ASM990963v1_genomic            0.000000  \n",
       "GCA_009669545.1_ASM966954v1_genomic            0.800666  \n",
       "GCA_009669555.1_ASM966955v1_genomic            0.800250  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "replicon_cov_df = copy.deepcopy(cov_df)\n",
    "replicon_loci = []\n",
    "non_replicon_loci = []\n",
    "\n",
    "for col in cov_df.columns:\n",
    "    #if (\":\" not in col) or (\"id\" in col) or (\"rna\" in col) or (\"exon\" in col):\n",
    "    if (\":\" in col) and (\"id\" not in col) and (\"rna\" not in col) and (\"exon\" not in col):\n",
    "        replicon_loci.append(col)\n",
    "    else:\n",
    "        non_replicon_loci.append(col)\n",
    "\n",
    "replicon_cov_df.drop(columns=non_replicon_loci, inplace=True)\n",
    "display(replicon_cov_df)\n",
    "\n",
    "#replicon_dep_df = copy.deepcopy(dep_df)\n",
    "#replicon_dep_df.drop(columns=drop_loci, inplace=True)\n",
    "#display(replicon_dep_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "systematic-disorder",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Plot Locus Coverage of Whole Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "conscious-berry",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>data</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.919553</td>\n",
       "      <td>GCA_009909635.1_ASM990963v1_genomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.947885</td>\n",
       "      <td>GCA_009669545.1_ASM966954v1_genomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.945635</td>\n",
       "      <td>GCA_009669555.1_ASM966955v1_genomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GCA_009909635.1_ASM990963v1_genomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GCA_009669545.1_ASM966954v1_genomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14995</th>\n",
       "      <td>4999</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCA_009669545.1_ASM966954v1_genomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14996</th>\n",
       "      <td>4999</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCA_009669555.1_ASM966955v1_genomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14997</th>\n",
       "      <td>5000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCA_009909635.1_ASM990963v1_genomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14998</th>\n",
       "      <td>5000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCA_009669545.1_ASM966954v1_genomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14999</th>\n",
       "      <td>5000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCA_009669555.1_ASM966955v1_genomic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          x    y      data                               sample\n",
       "0         1  1.0  0.919553  GCA_009909635.1_ASM990963v1_genomic\n",
       "1         1  3.0  0.947885  GCA_009669545.1_ASM966954v1_genomic\n",
       "2         1  4.0  0.945635  GCA_009669555.1_ASM966955v1_genomic\n",
       "3         2  1.0  1.000000  GCA_009909635.1_ASM990963v1_genomic\n",
       "4         2  3.0  1.000000  GCA_009669545.1_ASM966954v1_genomic\n",
       "...     ...  ...       ...                                  ...\n",
       "14995  4999  3.0  0.000000  GCA_009669545.1_ASM966954v1_genomic\n",
       "14996  4999  4.0  0.000000  GCA_009669555.1_ASM966955v1_genomic\n",
       "14997  5000  1.0  0.000000  GCA_009909635.1_ASM990963v1_genomic\n",
       "14998  5000  3.0  0.000000  GCA_009669545.1_ASM966954v1_genomic\n",
       "14999  5000  4.0  0.000000  GCA_009669555.1_ASM966955v1_genomic\n",
       "\n",
       "[15000 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# X is the positional index of the loci\n",
    "# Y is the samples y coord in the tree\n",
    "# D is the data about the loci\n",
    "\n",
    "i = 1\n",
    "s = []\n",
    "x = []\n",
    "y = []\n",
    "d = []\n",
    "r = {}\n",
    "\n",
    "#for loci in replicon_cov_df:\n",
    "#for loci in cov_df.columns[0:5000]:\n",
    "for loci in cov_df.columns[0:5000]:\n",
    "    if loci in replicon_loci:\n",
    "        r[loci] = i\n",
    "    for sample in cov_df.index:\n",
    "        s.append(sample)\n",
    "        x.append(i)\n",
    "        y.append(tree_df[\"coord_y\"][sample])\n",
    "        d.append(cov_df[loci][sample])\n",
    "        #d.append(replicon_cov_df[loci][sample])\n",
    "    i += 1\n",
    "    \n",
    "# Create a dataframe\n",
    "df = pd.DataFrame({\"x\":x,\"y\":y,\"data\":d, \"sample\" : s})\n",
    "display(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "controlled-evidence",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAB+cAAAVKCAYAAAA7dXk0AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAD2EAAA9hAHVrK90AABxAUlEQVR4nOzdf5Tdd13v+9cnP9uS9FfqgVMotQI9gOk6KFCTerRyQFQOqHDABahLpHjrj+u5l+sRWKDXHD0uEX9yjt5jOeCtshQpQgB/IBYpXLQNtB6LjbVWagutLdSmSTtJ2qRJPveP2Ul3JpnMr/3OzJ48HmvNynfv/Z3P55Pv7PnOnnnO3tN67wEAAAAAAAAA6qxY7AUAAAAAAAAAwHInzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAiq1a7AVQq7X2r5O8dMrV/5Rk7yIsBwAAAAAAAGCUzkjyNVOu++Pe+32LsZgTEeeXv5cmeddiLwIAAAAAAADgJPnfkvzPxV7EVF7WHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi/ub88vdPU6+46qqrcskllyzGWpaNt73tbdm1a1fOPvvs/PzP//xiLwcAAAAAAABOSbfcckuuvPLKqVcf00iXAnF++ds79YpLLrkkmzdvXoy1LBsbNmxI7z0bNmxwLAEAAAAAAGBpOaaRLgVe1h4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQbFnE+dbaO1prfejtW0Y07qtbax9vrd3XWnu0tXZXa+29rbVNoxgfAAAAAAAAgFPDqsVewEK11v5tkjeOeMzTknwgyUun3HTh4O21rbUtvfefG+W8AAAAAAAAACxPY/3M+dbaiiT/M5O/ZHD/CId+Tx4P89cl+e4klya5IskdmTxuP9tae8MI5wQAAAAAAABgmRrrOJ/kPyV5fpLbMhnUF6y1dnmS1w4u/lGSb+29f6T3fmPv/beTbErypcHt72itnT2KeQEAAAAAAABYvsY2zrfWLkhy+GXlfyTJ/hEN/abBvweT/Gjv/eDwjb33B5K8eXDxnEw+mx4AAAAAAAAApjW2cT7J/5NkXZLf6b1/ahQDttbWJXnh4OK1vfd7ptn1Q0keHmy/YhRzAwAAAAAAALB8rVrsBcxHa+17Mvk34R9M8pMjHPrSJGsH25+ebqfe+/7W2rYkL05yaWttde/9sRGu44jW2uYFDrFxJAvhGDff949ZseOubH7HG466fvu9dxx1eeP5T5txrC1bJ46+/PL12bJ1Iltevv6EY9+65/zcuv3AMfvN1ZatE3lo++1JkiuuODvJ7NY9W/M5Jktt7LmOu2Jly6GD/Zjrr/nHR3Lr9gN55Tfcf+S6v3v4Sfn7vz905PJCP56zWfcv/MlE9u2f+1wLOSaLNfYo7iMrVq/IoccOHfO5OvxxnO/Y06m8b7/1l/9X3vTt63LPuStHPvZS/3z/b9dN5MFdj18+/Dmw1Nc9TmMf7+vXsBt+7NVZ/7afn/O4szWO56kkeeVv3pwLHjyYK644e8mv++1/OpFH98VjB2OPdOyZzh0zjf2e9+zKr/30c+e+wFmMndQfk6mPHY/3/cF03n/bnvz93x/KlpevP+G6Dz8One9jzapzSTJ5Pqk4lyTJ/33dLZn4y/0znl9PP/f0PPLgI3Mae/u9d+Qj//vL8lMfuiE9Zy10qUeNO2wUx+TwferwuftH3/ik7Hv40WV3Lhnl2IeP2UK+PzvZ657LufT6H3hR/nHtvtz5HX86488eZrvu63/w2/Ln3/mHWbUy+anvPP46luL9ZDbn3KX6s4eqsacekyPXDx2bpbjuuYxr7JnH/tVPTOThKXeFuZwTl8IxueH/eF0+ef/n8rb33Xrkfr36mm/I2953a170TRfmE5/5YpLkI18+mL+5YW+S5Pk/uD7nvOjZefaWD+eec1eelHW//U8ncvBQ8raXnvyvOVM/33/ph56UPQ/sGcnYM8070+PXhVqs89TUY/p1f/GafNdv/PG8x/71T05k10ND48/y8/Caf9ibW289uGzO3avPWJPf+u/356yNFx+5bZQ/R09Gu+4fef/nc9rtB45cHl738XzbDe8f2dwsfWMX5wd/4/2dg4tv7r3/ywiHf9bQ9m0z7HtbJuP8qiTPSHLrCNcx7PqicVmgiX17c/Cxlm13bj/hfjPdniQ7tu08+n2ec052bNuZbc8554Tvt2/757Nj22Mz7jeX+Xfvm/z9lNmse77Gcey5jrti1YocOnDomOv3bd+dHdsey+6hj9mjt3w+O256/C9oLPTjOWy6dT/4uZ159NGFzTWOH8f5jr1yzcoc3H/wmM/V3VOO31Jb93Qm/mVPbnvCmmTf6MeeaqmN/dBf78yOoUcO030OLLV1j9PYM379Orgju/c9Hj+WwpqXwtg//qWJHMjk1+Glvu6dn9uZvUP9ymMHY49i7Nk89j2RA7v3jvUxmfrY8XjfH0zn0Vsezo6bDh53n+F1H34cOorHmuN0Lnn6g7tz4+4DM55f1+1dl91f2T3n8S9+0rp89u5/zqEDdy9kmSc0imMy9T61Z/8jeXTfo2P9eVM99uFjdjK+PxvV2HM6l+7fmL0r92bHtptnfJ9Zr/ux52THtpuzalWy7ZLZrWMp3E/mcs6d69hztRSOR3LsMTkyhu+fTqmxH7ppZx6cclcYu58j9Z15cN9dR86RSbJm3xOz7c7tufCSA9l252QC2P/FA9mxbTKonv3TL0j6Y7ntCfuTfSdn3Ts/tzMHDibbvvbkf82Z+vl+40suzMP3PjzN3nMbe6Z5Z3r8Okon8/439ZjuP/TCec+/7c7teeimndmxY+i6WX4ePrp9Ijs+d2DZnLvXrjmUA7snH7scmWeEj9OS0a777V+eyD/vfvwXsIfXDeP4svbvSPKkTEbr94x47AuGtqd7SfvDhr/zvmDavQAAAAAAAAA45Y1VnG+t/bskb0hyIMkP996Pfc3ohRl+DYyZfmV++LVd1o14HQAAAAAAAAAsI2MT51tra5K8K0lL8mu991sKpjltaHv/DPvuG9o+vWAtAAAAAAAAACwT4/Q359+ayb8J/6Uk/6VojkeHttfMsO/aoe1Hpt1r4S5b4PtvzOQvNTBi69eekRVrVufiizYedf32e+846vLG858241gbNk0cdXnTReuzYdNENl20/qjrp469duPTsiEHjtlvrjZsmshD229PkqxbO/m7JrNZ92zN55gstbHnOu6KlS2HDh774h5rDzySDTmQdWvvP3LdaZdclA2rHv8bowv9eM5m3edeOpF9++c+10KOyWKNPYr7yIrVK3LosUPHfK4OfxznO/Z0Ku/b67/qf+WZe9bknnNXjnzspf75ftZzJ3Jw1+OXD38OLPV1j9PYx/v6dZSVG458rZnLuLM1juepJLnvqTfnggcPZt3a05f8us+5dCKn74vHDsYe6dgznjtmGHvVun3ZNOVx+UKc7GMy9bHj8b4/mM5p+/Zkw6pD2XTR+hOu+/Dj0Pk+1qw6lyST55OKc0mSfOGuW7Jq3f4Zz6+nn3t6Hjljbt/ab7/3jtz+5d35hguenJ6zFrrUo8YdNopjcvg+dfjc/YQ1p2fV2rbsziWjHPvwMVvI92cne91zOpeueVLOWLUvGzY9Z8afPcx63aufmA2bnpNVK6c/bkvxfjKbc+5S/dlD1dhTj8lhw8dmKa57LuMae+axz3reRNqUu8JczolL4pi0c3Lu2q/Opov2Hrlfrx5cfsbZF2bTRZN5ZM3pB7Nh594kya57DuScZ64+8nOTk7Hucy6dyMFDi/M1Z+rn+/Of/KTsWbvnqOuqHpvM9Ph1oRbrPDX1mK5Z8a/m9L3K1LHPet5E+kOP3z7b+8lp+/dmw4qDy+bcvXrVmqxad0bO2njxkdtG+XP0ZLTr/uKTPp/THj5w5PLwuqGN/pXhR6+19swkn89kMP+u3vtHj7PPliQ/M7j4gt77p+Yxz9uTvHlw8fm995tOsO9/TvJLg4vf3nv/+FznOxlaa5uTXD983fXXX5/Nmzcv0oqWh1e96lV54IEHct555+UDH/jAYi8HAAAAAAAATkk33HBDLrvsmOc7X9Z7v2Ex1nMi4/LM+TdmMsz/U5IzWmuvPs4+w7969O9ba08abP9R733PcfY/nnuGtp+SZNo4n+SCoe27Zzk+AAAAAAAAAKegcYnzh19C/muSvG8W+//00PZFSWYb528d2n7mDPsevv1Aki/McnwAAAAAAAAATkErFnsBS8yNSfYPti+fbqfW2pokmw6/T+99/3T7AgAAAAAAAMBYxPne++t67+1Eb0n+y9C7vGDotrvmMM9Ekr8YXHxRa+0p0+z6iiRnDra3zvX/AwAAAAAAAMCpZSzi/Ki01l7XWuuDty3T7PbLg39XJfnN1trKKWOcl+QXBxd3JXl3xVoBAAAAAAAAWD5OqTg/G733Tyb5g8HF70xybWvtO1trz2ut/WCSbUmeOrj9Lb33nYuxTgAAAAAAAADGx6rFXsAS9fpMvmz9S5K8YPA27FCSn+u9X3WyFwYAAAAAAADA+PHM+ePovT/Se/8PSb43ybVJ7k+yP8ndSX4/yb/rvW9ZvBUCAAAAAAAAME6WzTPnB7F8ywz7XJ3k6jmM+fuZjPEAAAAAAAAAMG+eOQ8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXGJs631s5srb26tfYrrbVPt9a+0Fp7qLW2v7V2f2vtU621N7XWNixwni2ttT7Lt28Zzf8OAAAAAAAAgOVs1WIvYA4uTfK+aW77qiSXD95+srX2fb33j5+0lQEAAAAAAADACYxTnE+Su5Ncl+SvB9v3ZfLZ/09J8sokr0hyXpKPttae33v/2wXOd8kMt9+5wPEBAAAAAAAAOAWMU5y/rvf+1BPcfk1r7buTbE2yJsnPJPmPC5mw9759Ie8PAAAAAAAAAMkY/c353vvBWezz4SS3DS5+c+mCAAAAAAAAAGCWxibOz8Gewb+nLeoqAAAAAAAAAGBgnF7WfkattWclec7g4m0n2HVstNY2L3CIjSNZCAAAAAAAAADzNvZxvrV2RpInJ3lZkjclWTm46Z0jGPvaJF+fZH2SXUluTfJnSa7qve9c6PizdP1JmgcAAAAAAACAImP5svattde11nprrWfyZexvT/IrSZ442OWXk/zeCKZ6UZJzk6xO8lVJLk/yC0n+qbX2XSMYHwAAAAAAAIBTwNg/c36Km5P8cO/9swsc55YkH07yuST3ZjLO/5sk35vkxUnOTvLB1trLeu8fW+BcAAAAAAAAACxz4xrnP5zkpsH26UmeluR7krw8ye+11v7P3vsfz3PsX++9bznO9Z9N8ruttSuT/FYmXz7/3a21p/feH5nnXAAAAAAAAACcAsYyzvfed2Xyb8AfdmOSP2itfX+S30nykdbaFb33q+c59oluv6q19rwkb0hyfpJXZDQvoT+dyxb4/huTvGsUCwEAAAAAAABgfsYyzk+n9/7e1tpLM/ks+t9orX2k976zYKqrMhnnk8m/Q18W53vvNyzk/Vtro1oKAAAAAAAAAPO0YrEXUOAjg3+fkOQ7iua4dWj7yUVzAAAAAAAAALBMLMc4/y9D2xcWzeHp6AAAAAAAAADM2nKM88PPZN9dNMezh7bvLZoDAAAAAAAAgGViOcb5Vw1t31I0x5VD258umgMAAAAAAACAZWJs4nxr7XWttdNm2OeNSV4yuHhXkr88zhh98LblOO9/SWvt6TPMcWWSKwYXv5xk6+z+BwAAAAAAAACcqlYt9gLmYEuSX2mtfTCT0f2OTL5s/foklyT53iTfONh3f5If6r0fmOMcz03y7tbadUk+lsln3u/I5HF6ZpLvS/Ktg30PJrmy975nvv8hAAAAAAAAAE4N4xTnk+TcJD80eJvOPUle33v/xDznWJnkRYO36exIckXv/aPznAMAAAAAAACAU8g4xfkXZjKYvyDJs5I8McmGJI8m+UqSm5P8cZJreu975znHn2byJes3J/m6oTlakgeTfD7JnyW5uvf+8Hz/IwAAAAAAAACcWsYmzvfe78jkS9lftYAxrk5y9Qluvz/Jbw/eAAAAAAAAAGAkViz2AgAAAAAAAABguRPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQbmzjfWjuztfbq1tqvtNY+3Vr7Qmvtodba/tba/a21T7XW3tRa2zDCOV/dWvt4a+2+1tqjrbW7Wmvvba1tGtUcAAAAAAAAACx/qxZ7AXNwaZL3TXPbVyW5fPD2k6217+u9f3y+E7XWTkvygSQvnXLThYO317bWtvTef26+cwAAAAAAAABw6hinOJ8kdye5LslfD7bvy+Sz/5+S5JVJXpHkvCQfba09v/f+t/Oc5z15PMxfl+SdSe5NckmStyZ5WpKfba3d13t/9zznAAAAAAAAAOAUMU5x/rre+1NPcPs1rbXvTrI1yZokP5PkP851ktba5UleO7j4R0le3ns/OLh8Y2vto5n85YCnJnlHa+0Pe++75joPAAAAAAAAAKeOsfmb80OB/ET7fDjJbYOL3zzPqd40+Pdgkh+dOm/v/YEkbx5cPCfJFfOcBwAAAAAAAIBTxNjE+TnYM/j3tLm+Y2ttXZIXDi5e23u/Z5pdP5Tk4cH2K+Y6DwAAAAAAAACnlnF6WfsZtdaeleQ5g4u3nWDX6VyaZO1g+9PT7dR7399a25bkxUkuba2t7r0/No/5ZtRa27zAITaOZCEc4+b7/jErdtyVze94w1HXb7/3jqMubzz/aTOOtWXrxNGXX74+W7ZOZMvL159w7DtWfHX+5oa9x+w3V1u2TuSh7bcnSa644uwks1v3bM3nmCy1sUc17kfuP5S/+as9eeU33H/kuj/87L86ap+Ffjxns+5f+vhE9uyd+1wVx6R67IXeR9qKln6oJzn2c3X44zifsU+k8r79tl/7m/zktz4h95y7cuRjL/XP99/4/ybywI7HLx/+HFjq6x6nsY/39WvYVa//hnzjf/39OY87W+N4nkqSV/7mzbngwYO54oqzl/y6D38N8djB2KMce6Zzx0xjv+c9u/JrP/3cuS9wFmMnJ/+YHO/7g+lsvWd/Pn/jvmx5+foTjv2RLx9c0PcOoz6XvOPPJrL3kcnth7bfXnIuSZL/etM/ZMfHds94fj393NPzyIOPzGns7ffekWt+9IX52Q/9bbLizIUu9ahxh43imBy+Tx0+d//YT/zrPLrrkWV3Lhnl2IeP2UK+PzvZ657PuXS2Y8/Wlq0TOW1t8paXHH8dS/F+Mptz7rj97GGhY089JkeuHzo2S3HdcxnX2DOPfbz7wVzOMUvhmFz/ph/JvgOP5AW/enW2bJ3It133Y9n83343SfKyF1yUP7ruziTJnzzUcuMnJ5+L95Rf25hnP/W8XHb1tdl+/86Tsu5f/+REHnss+clvO/lfc6Z+nH/2lWfm0ME+krFnmnemx68LtVjnqWM+d/7g+dny/tnnqqljT/ezrJl8+N7HcvNnH1025+7VZ6zJb/33+3PWxouP3DbKn6Mno133f/rQ57Py7w4cuTy87uP5thveP7K5WfrGPs631s5I8uQkL8vkS9KvHNz0znkM96yh7ZnOlrdlMs6vSvKMJLfOY77ZuL5oXBZoYt/eHHysZdud20+430y3J8mObTuPfp/nnJMd23Zm23POOeH77f/iLdmxbWLG/eYy/+59k7+fMpt1z9c4jj2qcR+752B2bHs4u4c+Zju23Xz0XAv8eB411jTr3nXjzjw8sbC5xvHjOJ+xh+P81M/V3VOO31Ja94k8/OXdue0Jq5N9ox97qqU29sN/vSs7vvz4N5nTfQ4stXWP09gzff3a/diTs3vf4/FjKax5KYz941+ayIFMfh1e6us+/DXkMI8djD2KsWfz2PdEDuzeu6yOyfG+P5jO/i/sy45te4+7z/DY+794YCTfO0wdd752fm5ndu95/HLVueTih/fnM7v3znh+Xbd3XXZ/Zfecx3/6E8/Iti9+aSFLnNEojsnU+9Se/Y/kkX2PLKvPm1GPffiYnYzvz0Y19kLPpScae7Z2bNuZM05Ptj1rdutYCveTuZxz5zr2XC2F45Ece0yOjOH7p1Nq7OPdD8bv50gP51B2HzlHpu88su/XPrdn251/lyQ5sKNlx7YHkyRPf9bmZPdj2Xb3vSdt3Q/dtDP79yfbLj75X3OOOQc+b0MOHTg0krFnmnemx6+jdDLvf8d87jx63rzn33bn9ln/LGuq/Xfuz45te5bNuXvtmkM5sHvvUT9LH/XjnlGu+1e/MpE7dz/+cZvaADi1jWWcb629Lsn/e4JdfjnJ781j6AuGtqd7SfvD7p7yflVxHgAAAAAAAIAxN5Zx/gRuTvLDvffPzvP9h18DY6ZfmR/6Hf+sm+d8AAAAAAAAAJwCViz2Aubpw0kuGbxdmuQ1SbZm8u/N/15r7aXzHPe0oe39M+y7b2j79HnOBwAAAAAAAMApYCyfOd9735Vk19BVNyb5g9ba9yf5nSQfaa1d0Xu/eo5DPzq0vWaGfdcObT8y7V4Ld9kC339jkneNYiEcbf3aM7JizepcfNHGo67ffu8dR13eeP7TZhxrw6aJoy5vumh9NmyayKaL1h91/dSx1zzzq7Nh595j9purDZsm8tD225Mk69ZO/q7JbNY9W/M5Jktt7FGNu/oJh7LhgT1Zt/b+I9dt2HThUfss9OM5m3Wf/fyJrN4797kqjkn12Au9jwz/zfmpn6vDH8f5jH0ilfftM5/0N3nmnjW559yVIx97qX++n/nciezf8fjlw58DS33d4zT28b5+DVu3+ilHvtbMZdzZGsfzVJLc99Sbc8GDB7Nu7elLft2Hv4Z47GDsUY4907ljprFXrduXTVMely/EYh+T431/MJ01q/dnw+592XTR+hOOveb0gwv63mHU55JzLp3I2sF30g9tv73kXJIkt+/4h6xa12c8v55+7ul55Iy5fWu//d478oWv7M2mC5+arDhzoUs9atxhozgmh+9Th8/dT1hzelauXX7nklGOffiYLeT7s5O97vmcS2c79mxt2DSR09ZOf9yW4v1kNufccfvZw0LHnnpMDhs+Nktx3XMZ19gzj328+8FczjFL45icmRVZnU0XbZz8/7Rzjjxm/JozL8qmi1qSZNW5LRs2PZwkefTvJ5KnnpdNF5yf7ffvPCnrPut5E3nsscX5mnPMOfCrz8yhg/2o66oem8z0+HWhFus8dcznzmkXZdNF+zJbU8ee7mdZM1mz9rFseOjRZXPuXr1qTVatOyNnbbz4yG2j/Dl6Mtp13/XEz2fVAweOXB5eN7Te+8x7jZHW2vuTfE8mX3b+gt77zjm879uTvHlw8fm995tOsO9/TvJLg4vf3nv/+DyXXKq1tjnJ9cPXXX/99dm8efMirWh5eNWrXpUHHngg5513Xj7wgQ8s9nIAAAAAAADglHTDDTfkssuOeb7zZb33GxZjPScyri9rfyIfGfz7hCTfMcf3vWdo+ykz7HvB0Pbdc5wHAAAAAAAAgFPIcozz/zK0feG0ex3frUPbz5xh38O3H0jyhTnOAwAAAAAAAMApZDnG+ScPbe+e4/vemGT/YPvy6XZqra1Jsunw+/Te90+3LwAAAAAAAAAsxzj/qqHtW+byjr33iSR/Mbj4otbadC9t/4okZw62t85teQAAAAAAAACcasYmzrfWXtdaO22Gfd6Y5CWDi3cl+cvjjNEHb1umGeaXB/+uSvKbrbWVU8Y4L8kvDi7uSvLu2f4fAAAAAAAAADg1rVrsBczBliS/0lr7YCaj+x2ZfNn69UkuSfK9Sb5xsO/+JD/Uez8w10l6759srf1Bklcn+c4k17bWfj3JvYN53pbkqYPd39J73znf/xAAAAAAAAAAp4ZxivNJcm6SHxq8TeeeJK/vvX9iAfO8PpMvW/+SJC8YvA07lOTneu9XLWAOAAAAAAAAAE4R4xTnX5jkRZkM5c9K8sQkG5I8muQrSW5O8sdJrum9713IRL33R5L8h9baa5O8Lsm/TXL2YJ7PJPmN3vsNC5kDAAAAAAAAgFPH2MT53vsdmXwp+3k/W733fnWSq+ew/+8n+f35zgcAAAAAAAAASbJisRcAAAAAAAAAAMudOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACg2FjF+dba17fW3tpa+1hr7e7W2r7W2u7W2u2ttatba980onm2tNb6LN++ZRRzAgAAAAAAALB8rVrsBcxWa+3TSb75ODetSfKMwdsPtNbem+QNvff9J3N9AAAAAAAAADCdsYnzSZ48+PfeJB9I8pkkX0qyMsnmJD8x2Of7M/n/eu2I5r1khtvvHNE8AAAAAAAAACxT4xTnb0vy1iQf7L0fnHLbtsEz5v8qycVJXtNa+x+9988sdNLe+/aFjgEAAAAAAADAqW1s/uZ87/2lvfdrjhPmD9/+QCafPX/YK0/OygAAAAAAAADgxMYmzs/Sp4a2n7ZYiwAAAAAAAACAYeP0svazsWZo+9CirWKEWmubFzjExpEsBAAAAAAAAIB5W25x/vKh7dtGMWBr7dokX59kfZJdSW5N8mdJruq97xzFHDO4/iTMAQAAAAAAAEChZfOy9q21FUneMnTVNSMa+kVJzk2yOslXZfIXAH4hyT+11r5rRHMAAAAAAAAAsIwtp2fOvzHJpYPtrb33mxY43i1JPpzkc0nuzWSc/zdJvjfJi5OcneSDrbWX9d4/tsC5AAAAAAAAAFjGlkWcb61dnuTtg4v3J/mRBQ756733Lce5/rNJfre1dmWS30qyMsm7W2tP770/ssA5AQAAAAAAAFimxj7Ot9a+NsnWTP5f9iX5nt77VxYyZu991wy3X9Vae16SNyQ5P8krkvzeQuY8gcsW+P4bk7xrFAsBAAAAAAAAYH7GOs631i5K8udJzklyMMlreu+fPknTX5XJOJ9M/h36kjjfe79hIe/fWhvVUgAAAAAAAACYpxWLvYD5aq2dn+QTmXzmek/y+t771pO4hFuHtp98EucFAAAAAAAAYMyMZZxvrZ2X5NokXzO46sd77797spdxkucDAAAAAAAAYEyNXZxvrZ2V5ONJnj246i29999chKU8e2j73kWYHwAAAAAAAIAxMVZxvrV2RpI/SfL1g6t+vvf+i4u0nCuHtk/W37kHAAAAAAAAYAyNTZxvra1JsjXJNw6uemfv/afmMc7rWmt98LblOLdf0lp7+gxjXJnkisHFLw/WBQAAAAAAAADHtWqxFzAH70vy4sH2J5O8p7W28QT77++93z6PeZ6b5N2tteuSfCzJLUl2ZPJYPTPJ9yX51sG+B5Nc2XvfM495AAAAAAAAADhFjFOcf8XQ9r9P8rcz7P/FJF89z7lWJnnR4G06O5Jc0Xv/6DznAAAAAAAAAOAUMU5x/mT500y+ZP3mJF+X5IlJNiRpSR5M8vkkf5bk6t77w4u1SAAAAAAAAADGx9jE+d57G9E4Vye5+gS335/ktwdvAAAAAAAAALBgKxZ7AQAAAAAAAACw3InzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqJ8wAAAAAAAABQTJwHAAAAAAAAgGLiPAAAAAAAAAAUE+cBAAAAAAAAoJg4DwAAAAAAAADFxHkAAAAAAAAAKCbOAwAAAAAAAEAxcR4AAAAAAAAAionzAAAAAAAAAFBMnAcAAAAAAACAYuI8AAAAAAAAABQT5wEAAAAAAACgmDgPAAAAAAAAAMXEeQAAAAAAAAAoJs4DAAAAAAAAQDFxHgAAAAAAAACKifMAAAAAAAAAUEycBwAAAAAAAIBi4jwAAAAAAAAAFBPnAQAAAAAAAKCYOA8AAAAAAAAAxcR5AAAAAAAAACgmzgMAAAAAAABAMXEeAAAAAAAAAIqNVZxvrX19a+2trbWPtdbubq3ta63tbq3d3lq7urX2TQVzvrq19vHW2n2ttUdba3e11t7bWts06rkAAAAAAAAAWJ5WLfYCZqu19ukk33ycm9Ykecbg7Qdaa+9N8obe+/4Fzndakg8keemUmy4cvL22tbal9/5zC5kHAAAAAAAAgOVvnJ45/+TBv/cmeWeSVya5NMnmJP9Xkn8e3P79Sa4ewXzvyeNh/rok3z2Y74okd2Ty2P1sa+0NI5gLAAAAAAAAgGVsbJ45n+S2JG9N8sHe+8Ept20bPGP+r5JcnOQ1rbX/0Xv/zHwmaq1dnuS1g4t/lOTlQ3Pe2Fr7aJK/TvLUJO9orf1h733XfOYCAAAAAAAAYPkbm2fO995f2nu/5jhh/vDtDyT5iaGrXrmA6d40+Pdgkh+dOudgrjcPLp6TyWfTAwAAAAAAAMBxjU2cn6VPDW0/bT4DtNbWJXnh4OK1vfd7ptn1Q0keHmy/Yj5zAQAAAAAAAHBqGKeXtZ+NNUPbh+Y5xqVJ1g62Pz3dTr33/a21bUlenOTS1trq3vtj85xzWq21zQscYuNIFgIAAAAAAADAvC23OH/50PZt8xzjWXMY47ZMxvlVSZ6R5NZ5znki1xeMCQAAAAAAAMBJtGxe1r61tiLJW4auumaeQ10wtD3dS9ofdvc07wcAAAAAAAAARyybOJ/kjZl8Sfok2dp7v2me46wf2t49w757hrbXzXM+AAAAAAAAAJa5ZRHnW2uXJ3n74OL9SX5kAcOdNrS9f4Z99w1tn76AOQEAAAAAAABYxsb+b8631r42ydZM/l/2Jfme3vtXFjDko0Pba2bYd+3Q9iMLmPNELlvg+29M8q5RLAQAAAAAAACA+RnrON9auyjJnyc5J8nBJK/pvX96gcNODG3P9FL1Txjanukl8Oel937DQt6/tTaqpQAAAAAAAAAwT2P7svattfOTfCLJ+Ul6ktf33reOYOh7hrafMsO+Fwxt3z2CuQEAAAAAAABYhsYyzrfWzktybZKvGVz147333x3R8LcObT9zhn0P334g/397dx4l2VXfB/z7k0YatKEV5AiEJCSxRWADBo9YIikG4oCcgAKEGMeWEQbsHGJITCB2EmSw4yU4gdg4rMlgB9sxBgHBJiAbUIylAcFBDgqrQCISwpK1MlqH0dz8Ua81NT1d3V3dfae7pz+fc97pt9x3351afvOqvlWvkqtW6PgAAAAAAAAA7GfWXThfVUcm+ViSxwyrXtdae+sKHuLyJDuG+bPmGcfBSbbM7NNa2zGpLQAAAAAAAAAb27oK56vq0CR/kuQJw6pfaa39+koeo7W2PcmfD4vPqKpJl7Y/L8kDh/mVuJw+AAAAAAAAAPupdRPOD99UvyjJU4dVb2mt/Zsl9HN+VbVhunBCszcNfzcleWtVHTirj+OSzHwo4LYk75p2HAAAAAAAAABsHJtWewBT+IMkzxrmP5Hk3VV1xjztd7TWvraUA7XWPlFVf5jkRUn+QZKLq+rNSa5P8tgkv5jkYUPz17XWbl3KcQAAAAAAAADYGNZTOH/e2PzfTfJ/Fmj/rSQnL+N4L8nosvXPTnLOMI3bleSNrbW3L+MYAAAAAAAAAGwA6+ay9vtaa+3u1tpzkrw4ycVJbkyyI8m1SX4/ydNaaxeu3ggBAAAAAAAAWC/WzTfnW2u1Qv1sTbJ1iva/n1EYDwAAAAAAAABL4pvzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0Nm6Cuer6sFVdW5VvaGqPlpVN1VVG6atK3icC8f6XWg6e6WOCwAAAAAAAMD+adNqD2BKN6z2AAAAAAAAAABgWustnB93bZIvJ3lW5+M8doHtV3c+PgAAAAAAAADr3HoL59+Q5PIkl7fWbqiqk9M5HG+tXdmzfwAAAAAAAAD2f+sqnG+tvX61xwAAAAAAAAAA0zpgtQcAAAAAAAAAAPu7dfXN+Y2oqs5cZhdnrMhAAAAAAAAAAFgy4fwCquriJE9IckSS25J8Kcn/SvL21tqt+2AIl+6DYwAAAAAAAADQkcvaL+wZSY5JclCSByU5K8mvJvlmVf3D1RwYAAAAAAAAAOuDb85P9sUkH0zy2STXZxTOPzLJi5M8K8lRSd5fVT/aWvvoKo0RAAAAAAAAgHVAOD+3N7fWLpxj/WeS/G5VvTzJ25IcmORdVXVaa+3ufTlAAAAAAAAAANYP4fwcWmu3LbD97VX1g0lemuSEJOcleW+n4TxlmfufkeQdKzEQAAAAAAAAAJZGOL90b88onE9Gv0PfJZxvrV22nP2raqWGAgAAAAAAAMASHbDaA1jHvjQ2/5BVGwUAAAAAAAAAa55wful8JR0AAAAAAACARRHOL91jxuavX7VRAAAAAAAAALDmCeeX7uVj85es2igAAAAAAAAAWPM2XDhfVedXVRumC+fY/tiqOm2BPl6e5IJh8a+TXLTyIwUAAAAAAABgf7FptQcwjap6WpLx4Py4sfnTqur88fatta1LOMwTk7yrqj6Z5KNJvpjk5oxuq0cl+fEkzxza3pfk5a21O5dwHAAAAAAAAAA2iHUVzid5aZKfnLDtqcM0busSj3NgkmcM0yQ3J7mgtfbhJR4DAAAAAAAAgA1ivYXz+8KfZnTJ+jOTPD7J8UmOTVJJbknyV0n+V5KtrbXvrtYgAQAAAAAAAFg/1lU431o7P8n5y+xja+b5Rn1r7cYk/3WYAAAAAAAAAGDZDljtAQAAAAAAAADA/k44DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADobF2F81X14Ko6t6reUFUfraqbqqoN09ZOx3xRVX2sqr5TVfdU1TVV9XtVtaXH8QAAAAAAAADY/2xa7QFM6YZ9daCqekCS9yU5d9amk4bpx6rqwtbaG/fVmAAAAAAAAABYn9bVN+dnuTbJxzv2/+7sDuY/meS5SZ6c5IIk38jotntDVb204xgAAAAAAAAA2A+st2/OvyHJ5Ukub63dUFUnJ7l6pQ9SVWcl+bFh8X8meV5r7b5h+fKq+nCSzyd5WJLfqKo/bq3dttLjAAAAAAAAAGD/sK6+Od9ae31r7SOttd6Xt/9Xw9/7kvzsWDA/M46bkrx2WDw6o2/TAwAAAAAAAMCc1lU4vy9U1eFJfnhYvLi1dt2Eph9I8t1h/rzuAwMAAAAAAABg3Vpvl7XfF56cZPMwf8mkRq21HVW1Lcmzkjy5qg5qrX1vpQdTVWcus4szVmQg7OX/3HhVcvM1OfM3XrrH+iuv/8Yey2eccOqCfV140fY9l593RP7b57fnp554xLx9f+DyB2fXrlH7pbr03Cfm4xd8Krdf+bUkyQUXHLXocS/WUm6Ttdb3SvX7F487MZ9845dy3pNuvH/dH3/mwXu0Wc79mSxu3DOPuWf+2uPy1M8s/tdBetwmvfte7mPkgAMru+5rSfZ+rj7/h27cY3ktjXs+r37j5/P65x2R6445cMX7XuvP9zd8aHt27dq9PPN8W+vjXk99X3jR9nnr2B+84uw89t+9e+p+F2s91qkkeeHvXJGH3HxfLrjgqDU/7pnn0cy5w6/92APz9UMOmLPvD17/vTz3hIOmPsZafGzru2/fC9WOhfp+97tvy3/6t0+cfoCL6DvZ97fJXK8PJvnlD2/PzvtGbebre7zdUvSqJcmonsxXS5bj+W+9IifesnB9PeaUY3LL1bdM1feV138jf/CKc/IrH/q/SS3vHH52v+PGx33hRdvzyNOTr3599/bF3Kczj6mZ2v2zr/6+3Pvde/a7WrKSfc/cZst5fbavxz1NLZ1dZ8bNPE5mHHnGI6Ye3w+/5vQ8/aq/3mv97HH/0OMelztvunPq/ueynPPXPZbnuA3X23sPy+170uNj/LZZi+Oepl99L9z3XI+DaWriWrhN/vLZj8+OR31/zvmPW++vkc94+kn5s7/4Vp719JPz8b+4Jkny2/97e57wtnPzpVd9JL/zzM05+nFH5q0f+kp23rNzn/6f88zffEKe+umv77XfcvpeyOz7+Td/9oRs/86e63rclzP3x/5YX1fyuXPGCadOfC9rIXOdy6zn2n3QoQfnbb914x7nJSv5PnqysuP+8XddkQd9Z/dFuRc6n/p7l/2PFTs2a59wfm+PHpv/ygJtv5JROL8pyelJvtRhPJd26JMVsP3eu/K9lmy7+sp52y20PUlu3nbrnvv8wNG584rbsu2YoxbY74rs2jVqv2R3Pjg3b7vi/sU77h19NmUx416q9dj3ivV7wr25edsVueNxu++z8ds/Web9Ocukcd//mLv18Gy7+v+uaN8rYS31fcCmA7Jr5+gMePZz9Y5Z99VaGvd8dt5xV75y2Obk3pXve7a11vfN227d4wXNpOfbWhv3eur75m23zlvH7rnvkbnj3run7ncp1lPfP3ft9uzcNfp/eK2Pe/bz6AsHjOrJXH3vuHpHtt178LKPudZvE30vv++FasdCdt5x1351m8z1+mCSmz9za3bunLvNeN/ztZvWvq4ly/HK/7c9O7NwfT3+0ONzw9XT/4Leaccfkm3XfGsZI1zYHvfjtltz992Vm/+q7d6+iPt09mPqzh13555779mvnjcr3ffMbbYvXp+tVN/T1NLZj4n5214x9XjaDcm2qxd+blz+7S/nu9d/d8F2SzHN+ese+y3iNlzz7z0ss+9Jjw+vnzZW33M9DpZTE1flNrnj2OzKw/aokSc9dme2Xf2VPPxx92Xb1V9Oknz387cl9z4hd3zhr3LjT23JYbvuzW23b99n477/tr7t6BU73lJr4Oee+93ces38/0es1LngQuevK2lfPv5W8rkz89hdzHtZsy10LrPe6tTmg3dl5x137XFespLnacnKjvu1123PXbvf/lrS+RT7L5e139uJY/OTLmk/49oJ+wEAAAAAAADA/YTzexu/DsYdC7Qdv+7W4R3GAgAAAAAAAMB+QDi/tweMze9YoO29Y/OHdBgLAAAAAAAAAPsBvzm/t3vG5hf6gczNY/N3T2y1PE9Z5v5nJHnHSgyEPR2x+dDkwE15xCln7LH+yuu/scfyGSecumBfx27ZvsfyllOOyGE/sD1bTjlij/Wz+z52y0nZtSt7tZvKYSfm2C0/kNuv/FqS5PDNhyx63Iu1lNtkrfW9Yv0ed2KO3bIjh2++8f5Vx245aY8my7o/s7hx3/+YO/rh2XJKrWjfS9Wr7+U+Rg44sLLrvtFves5+ro7fj0vpez49H9ubDv98HnXnwbnumANXvO+1/nw/dsv2PX6na+b5ttbHvZ76PnbL3v9/jXvAgX/r/v9rpul3sdZjnUqS75x4RR5y8305fPMha37cM8+jmXOHx+/anK8fcsCcfR+8+XvZcsJBUx9jLT629d2374Vqx0J9bzr83myZdV6+HKt9m8z1+mCSY39oe3beN2ozX9/j7ZaiVy1JRvVkvlqyHN952BU58ZaF6+sxxx+TW+560FR9X3n9N3LVDXdny8knJbW8c/jZ/Y7b437csj2HnJ4cO/b1gMXcpzOPqZnafdjBh2TT5trvaslK9j1zmy3n9dm+Hvc0tXR2nRk38ziZceQZj5h6fHX86XOOZfa4n/SQR+fOzXfu1W4plnP+Om4x417z7z0ss+9Jj4/x22YtjnuafvW9cN9zPQ6mqYlr4jY5/GE5IEdnyyln3F8jTz/qpGw5ZVNOO/LkbDll9L7IA5+4Pdl8Qg5//PfnwdftysHHbM5RRx6Rnffs3Kf/5+SoU7LllJ0r2vdCZt/PP3jCCdlee67rcV/O3B/7Y31dyefOzP/vc72XtZC5zmXWc+0+aNPB2XT4oXucl6zk++jJyo772w+9Ig/6zn33Ly/lfIr9V7XWVnsMS1ZVJye5elh8T2vt/BXo89eSvHZYfFJr7XPztP35JP9hWPyR1trHlnv8lVZVZya5dHzdpZdemjPPPHOVRrR/eMELXpCbbropxx13XN73vvet9nAAAAAAAABgQ7rsssvylKfs9X3np7TWLluN8czHZe33dt3Y/EMXaHvi2Py1HcYCAAAAAAAAwH5AOL+3L43NP2qBtjPbdya5qs9wAAAAAAAAAFjvhPN7uzzJjmH+rEmNqurgJFtm9mmt7ZjUFgAAAAAAAICNTTg/S2tte5I/HxafUVWTLm1/XpIHDvMXdR8YAAAAAAAAAOvWhgvnq+r8qmrDdOGEZm8a/m5K8taqOnBWH8cl+fVh8bYk7+oxVgAAAAAAAAD2D5tWewDTqKqnJTltbNVxY/OnVdX54+1ba1uXcpzW2ieq6g+TvCjJP0hycVW9Ocn1SR6b5BeTPGxo/rrW2q1LOQ4AAAAAAAAAG8O6CueTvDTJT07Y9tRhGrd1Gcd6SUaXrX92knOGadyuJG9srb19GccAAAAAAAAAYAPYcJe1X6zW2t2tteckeXGSi5PcmGRHkmuT/H6Sp7XWLly9EQIAAAAAAACwXqyrb8631s5Pcv4y+9iaKb5R31r7/YzCeAAAAAAAAABYEt+cBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoLNNqz0Aujt09oovfvGLqzGO/crNN9+c22+/PVWVyy67bLWHAwAAAAAAABvShOxzr4x0LajW2mqPgY6q6qeTvGO1xwEAAAAAAACwj7ystfbO1R7EbC5rDwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB05jfn93NV9beSnDtr9TeT3LUKw9kfnJHkHbPWvSzJlaswFoD1SB0FWD61FGB51FGA5VFHAZZHHV15hyZ5+Kx1H2mtfWc1BjOfTas9APoaHnTvXO1x7C+qaq7VV7bWLtvXYwFYj9RRgOVTSwGWRx0FWB51FGB51NFu/ny1B7AYLmsPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ9VaW+0xAAAAAAAAAMB+zTfnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzsMUquphVfWmqvpyVd1ZVbdU1Wer6uer6tDVHh/ANKrqwVV1blW9oao+WlU3VVUbpq1L6O9HquoDVXVdVd07/P1AVf3IFH0cWlWvGWrrLVV1x1Bz31RVD5uin79dVW+rqquq6u6q+puq+t9V9fKq2jTtvw1gLlX1hKr6haGGXjvUvjuq6mtVtbWqnj5lf+oosKFU1QOr6kVV9ZtVdclQc26vqh1VdWNVfaqq/lVVHbvI/tRRgEFV/cbYa/xWVWcvYh91FNhwZtXK+aZPLaIvdZQFVWtttccA60JVPSfJe5McOaHJV5M8u7X2zX03KoClq6r5TgLe01o7f5H9VJK3JXnZPM3ekeQVbZ4Tj6o6NcmfJHnkhCa3J/mx1tqfLjCeC5K8NcnmCU22JTm3tXbzfP0AzKeqLknydxbR9PeSvLS1tmOevtRRYEOqqmckuXgRTW9K8uOttY9N6EcdBRhTVd+f5HNJxsOXc1prn5rQXh0FNqwF3iMdd0lr7ewJfaijLJpvzsMiDCe0f5RRMH9Hkl9M8pQkP5zknUOzRyb5k6o6fFUGCbA81yb5+BL3/eXsPvH8QpJ/kuTJw98vDOtfluSNkzoYaudHsvvE850Z1dinZFRz78ioBr+vqh43Tz9/L6MT3c1Jbkjyz5P8UJK/n+QDQ7MtST5QVc6DgOV4yPD3+iRvSfL8jGrfmUn+RZJvD9v/aZKtC/SljgIb2bVJfjfJzyU5L6M6+tQk/zjJ+5Lcl+S4JB+ep36powCDoba8M6Ng/sZF7qaOAiT/Jclj55l+ap591VEWr7VmMpkWmJJ8MklL8r0kZ86x/TXD9pbk3632eE0mk2kxU5JfSnJukuOH5ZPHatnWRfZx2lAbW5LLkxwya/uhw/qZGnrqhH4uHDv2a+bYfubYcT4xoY9NSb4+tLl9rmNl9InRmeP8xGrfByaTaf1OGb1gfmGSAydsPy6jKyvN1JynT2injppMpg07Taqhs9o8d6zuvH+O7eqoyWQyjU1JXjXUmC8n+fdjNefsCe3VUZPJtKGnsZpy4RL3V0dNU00+EQELqKonJTl7WHx3a+2yOZr9ZkYnvEnyqqo6aF+MDWA5Wmuvb619pLV2wzK6eXV2Xybvla21u2cd464krxwWN2X0JsEehpr5c8PilzOqqbPHelmSdw+L51TVE+cYy/MyOhlOkl9trX1jjjavSXLr2DzAkrTWzm2t/VFr7b4J229K8i/HVj1/QlfqKLBhTaqhs9p8MMlXhsW5fk5EHQUYVNWJ2f2tzJ9JMvGnlcaoowDLo44yFeE8LOy5Y/P/ba4GrbVdGV2GL0mOzu4wH2C/NfyW0j8cFr/SWts2V7th/VeHxecO+407O8lRw/x7hpo6l61j8+fNsf25E9qOj+WujH6mJEnOqKrTJxwLYCV8amz+1Nkb1VGARbtz+PuA8ZXqKMBefifJ4RnVsk8t1FgdBVgedZSlEM7Dwp4+/L0zyefnaXfJ2PzT+g0HYM04Jbt/c/mS+RqObX9oRpfPH/f0OdrN5XPZ/cbsXHV2pp+vttb+ehFjmdQPwEo5eGx+rhfW6ijAAqrq0Ul+YFj8yqzN6ijAoKpemNFP192SxX8TUh0FWB51lKkJ52Fhjx7+XtVa2zlPu/E3CR49sRXA/mO81s1+o3S2+WrkovoZavDMpZj26KOqDs/oxHa5YwFYSWeNzc9Vm9RRgDlU1aFVdXpV/Yskn0xy4LDpLbOaqqMASarqqOyuka9trf3NIndVRwF2e0FVfbWq7q6q7VX19ap6T1WdM88+6ihTE87DPKrqAUmOGxavm69ta+3W7P7E0ok9xwWwRozXunlrZJJrJ+w3vnxna+22RfbzoKraPLb+oUlmLge1nLEArIiqOiDJ68ZW/dEczdRRgEFVnV9VrapaRq+tv5bRb20ePzR5U5L3ztpNHQUY+Y0k35fk0uz+PeLFUEcBdntMkkdk9FNKh2f02+0/keQTVXVRVR05xz7qKFPbtNoDgDXuiLH5OxbR/s4kh2VUuAH2d9PUyDvH5mfXyJl+Fltnx/u5d4XHArBSXp3kycP8Ra21z83RRh0FWNgVSV7RWvvMHNvUUWDDq6qnJXlpkp0Z1cs2xe7qKEByV5IPJ/nzjL5RfkeSB2V0NbxXJDk2o99y/1BVPbO19r2xfdVRpiach/k9YGx+xyLazxTBQzqMBWCtmaZG3js2P7tGzvQzTZ2d3c9KjQVg2arqrCS/NizemORnJjRVRwF2+2BGv6GZjGrLqUlemOR5Sd5bVa9qrX1k1j7qKLChVdXBSd6R0Tcl/1Nr7YtTdqGOAiQPmfBt9Yur6reSfDTJ4zMK638myX8ea6OOMjWXtYf53TM2f/Ai2s9cQuTuDmMBWGumqZHjl1iaXSNn+pmmzs7uZ6XGArAsVfW3k1yU0Qeh703ywtbaDROaq6MAg9baba21K4fp8tbaH7bWzsvoUqIPz+ibSufP2k0dBTa6X8jot4L/X5JfWsL+6iiw4c13Gfnh9fzzszvsfuWsJuooUxPOw/y2j80v5tIehw1/F3PpEYD1bpoaedjY/OwaOdPPNHV2dj8rNRaAJauqU5J8PMnRSe5L8k9aa5fMs4s6CrCA1trvJXlfRu9h/XZVHT22WR0FNqyqelSSfz0svrK1dud87SdQRwEW0Fr7ZpKLh8XTquqEsc3qKFMTzsM8Wmv3JLlpWHzofG2HNwhmCtq1PccFsEZcNzY/b41McuLY/OwaOdPPYVV11CL7+ZvW2vjll1ZqLABLMrw4/7MkJyRpSV7SWrtogd3UUYDF+dDw97Akf39svToKbGSvzuibkd9McmhVvWj2lOSMsfZ/d2zbzHuY6ijA4nxpbP4hY/PqKFMTzsPCvjz8Pa2qNs3T7lFz7AOwPxs/KX3UxFZ7b59dIxfVz1CDT52rj9baHdl9IrmcsQBMraqOy+hT9A8fVr2ytfa7i9hVHQVYnL8Zmz9pbF4dBTaymUsSPzzJH0yY/tFY+387tv5Bwzp1FGBxasJ6dZSpCedhYZ8e/h6W5InztDtrbP4v+w0HYM24Osn1w/xZ8zVM8neGv99Ocs2sbZ8em5+vnx/M7iuUzFVnZ/p5ZFV93zz9qNfAiqmqI5N8LMljhlWva629dZG7q6MAizP+7aTxy26qowDLo44CLM5jxuavH5tXR5macB4W9sGx+Z+aq0FVHZDkJ4bF25J8su+QAFZfa61l9yVGH1VVW+ZqN6yf+TTmh4b9xn0qye3D/E9W1aRPop4/Nj/XpaI/OKHt+FgOTfLCYfFLrbWvTTgWwIKGmvInSZ4wrPqV1tqvL3Z/dRRg0V4wNv/FmRl1FNjIWmvnt9ZqvinJL43tcs7YtmuGPtRRgAVU1cOTPHNY/GZr7dsz29RRlkI4DwtorX02yV8MixdU1ZlzNPuXSR49zL+ltfa9fTI4gNX35iQ7h/nfqqpDxjcOy781LO4c2u+htbYjyX8eFh+d5Odntxlq7wXD4iWttcvnGMtFSb4xzP/rqjp1jjb/IcnRY/MAS1JVB2dUd546rHpLa+3fLKGrN0cdBTaoqjq/qh6wQJtXJ3n2sHhN9vxWUaKOAizXm6OOAhtUVf3ofD9nXFXHJ/njJAcNq+a6Ut6bo44yhdr7wxnAbFX1+Iwu7XFIRpfQ+/cZfTv+kCQvSvKyoenXkvxga237aowTYBpV9bQkp42tOi67T8j+Msm7xtu31rZO6OdXk7xuWPxCkl/P6CTw1CSvTfL4YduvttZ+YUIfRyT5XJJHDKvekeQPk9yd5Jwkv5Dk8GH5Ka21Kyb08+wk/zOjDyDekOSXk3w2oxPOn87u39v7dJKzW2v3zdUPwEKq6v1JzhsWP5HkVUnme3G1Y9Kn0dVRYKOqqmuSHJHk/RnVlW9k9Jr7iCSPTfLi7P4Q1I4kz2mt/dkc/aijAHOoqguTvH5YPKe19qkJ7dRRYEMazkcPyuh89LKMPgx6d0bvk56d5BVJjh2afzrJM1pr987RjzrKognnYZGq6keT/PckD5zQ5GsZvVFw1b4bFcDSVdXWJD+52PbDJfHm6ueAJO9M8pJ5dn93kpe11nbNM57TkvxpktMnNPlukhe31j4y3zir6qeT/HaSgyc0+WxG9fqm+foBmE9VTftC6luttZMn9KWOAhvS8GboSYtoel2Sl7TWLp7QjzoKMIcpwnl1FNiQpjgffX+Sl7bWbpvQjzrKognnYQpVdVKSn0vynCQPzeiT+1cleV+S326t3bWKwwOYykqF82P9PTujK4k8KaNPl96U5PIkb2+tfXSRYzosyT/L6HdFT8voBPLajE5K39Ja+9Yi+zkjyT9P8sNJTkhyZ5IvJ3lvkne11nbOszvAglYynB/rUx0FNpThMpvPyOibQI9OcnxG30y6J6Nv+VyR5CNJ/mgxr7fVUYA9LTacH2uvjgIbSlWdleSsJGcmeXhGte+BGV3N6doklyZ5T2vtskX2p46yIOE8AAAAAAAAAHR2wGoPAAAAAAAAAAD2d8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoDPhPAAAAAAAAAB0JpwHAAAAAAAAgM6E8wAAAAAAAADQmXAeAAAAAAAAADoTzgMAAAAAAABAZ8J5AAAAAAAAAOhMOA8AAAAAAAAAnQnnAQAAAAAAAKAz4TwAAAAAAAAAdCacBwAAAAAAAIDOhPMAAAAAAAAA0JlwHgAAAAAAAAA6E84DAAAAAAAAQGfCeQAAAAAAAADoTDgPAAAAAAAAAJ0J5wEAAAAAAACgM+E8AAAAAAAAAHQmnAcAAAAAAACAzoTzAAAAAAAAANCZcB4AAAAAAAAAOhPOAwAAAAAAAEBnwnkAAAAAAAAA6Ew4DwAAAAAAAACdCecBAAAAAAAAoLP/D7ysWhZQkbmMAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 2400x1600 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "cmap = plt.cm.RdYlGn\n",
    "norm = colors.Normalize(vmin=min(d), vmax=max(d))\n",
    "\n",
    "fig, ax1 = plt.subplots(1, dpi=dpi)       \n",
    "marker_size = 25 / max(x)\n",
    "ax1.scatter(df.x, \n",
    "            df.y, \n",
    "            color=cmap(norm(df.data.values)),\n",
    "            marker=\"s\",\n",
    "            s=marker_size)\n",
    "\n",
    "# Draw lines for replicons\n",
    "for replicon in r:\n",
    "    ax1.axvline(x=r[replicon] -1, color=\"black\", linestyle=\"-\", alpha=0.75)\n",
    "\n",
    "#ax1.set_ylim(0,10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "confident-feature",
   "metadata": {},
   "source": [
    "## Import Colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cardiovascular-contributor",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors_df = pd.read_csv(auspice_colors_path, sep='\\t', header=None)\n",
    "colors_df.columns = [\"Attribute\", \"State\", \"Color\"]\n",
    "colors_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "colors_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "forty-compatibility",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Add Locus Info to Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "valuable-convert",
   "metadata": {},
   "outputs": [],
   "source": [
    "for locus in locus_df.columns:\n",
    "    tree_df[locus] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "    for i in range(0, len(locus_df[locus])):\n",
    "        sample = locus_df.index[i]\n",
    "        if sample not in tree_df.index: continue\n",
    "        cov = locus_df[locus][i]\n",
    "        tree_df.at[sample, locus] = cov\n",
    "\n",
    "display(tree_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "raising-orbit",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add ratios\n",
    "locus = [\"pPCP1\"]\n",
    "gene = [\"pla\"]\n",
    "\n",
    "for l,g in zip(locus, gene):\n",
    "    ratio_col = l + \"_\" + g + \"_dep_ratio\"\n",
    "    tree_df[ratio_col] = [NO_DATA_CHAR for row in range(0,len(tree_df))]\n",
    "    ratio_df = locus_df[[l + \"_dep\", g + \"_dep\"]]\n",
    "    for rec in ratio_df.iterrows():\n",
    "        sample = rec[0]\n",
    "        l_dep = rec[1][l + \"_dep\"]\n",
    "        g_dep = rec[1][g + \"_dep\"]\n",
    "        if l_dep == 0 or g_dep == 0 or l_dep < 10:\n",
    "            ratio = NO_DATA_CHAR\n",
    "        else:\n",
    "            ratio = l_dep / g_dep\n",
    "        tree_df.at[sample,ratio_col] = ratio\n",
    "        \n",
    "display(tree_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "integral-continent",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. Plot Locus Distributions: Histogram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "angry-pencil",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(2,1, dpi=dpi, figsize=(4,3), sharex=True)\n",
    "covs = axes[0].violinplot(\n",
    "    dataset=[locus_cov_df[col] * 100 for col in locus_cov_df.columns],\n",
    "    #dataset=[cov_data[col] * 100 for col in cov_data.columns],\n",
    "    #showmeans=True,\n",
    "    showmedians=False,\n",
    "    showextrema=False,\n",
    ")  \n",
    "\n",
    "\n",
    "deps = axes[1].violinplot(\n",
    "    dataset=locus_dep_df, \n",
    "    #showmeans=True,\n",
    "    showmedians=False,\n",
    "    showextrema=False,\n",
    ")    \n",
    "\n",
    "labels = [NO_DATA_CHAR] + [locus.split(\"_\")[0] for locus in locus_cov_df.columns if \"cov\" in locus]\n",
    "axes[1].set_xticklabels(labels, fontsize=4)\n",
    "\n",
    "axes[0].set_ylabel(\"Genomic Coverage (%)\", fontsize=4)\n",
    "axes[1].set_ylabel(\"Genomic Depth (X)\", fontsize=4)\n",
    "axes[1].set_ylim(-2,50)\n",
    "\n",
    "for cov,dep in zip(covs[\"bodies\"], deps[\"bodies\"]):\n",
    "    cov.set_linewidth(1)\n",
    "    dep.set_facecolor('#D43F3A')    \n",
    "    cov.set_alpha(0.8); dep.set_alpha(0.8)\n",
    "    cov.set_linewidth(0.5); dep.set_linewidth(0.5)\n",
    "    cov.set_edgecolor('black'); dep.set_edgecolor('black')\n",
    "    \n",
    "out_path = os.path.join(outdir, SCRIPT_NAME + \"_summary.\" + FMT) \n",
    "plt.savefig(out_path, dpi=dpi, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "disabled-public",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Plot Locus Coverage: Phylo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "equal-destiny",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Plot setup\n",
    "fig, ax1 = plt.subplots(1, dpi=dpi)\n",
    "\n",
    "# Colors Setup\n",
    "attr = \"Branch_Major\"\n",
    "attr_lower = attr.lower()\n",
    "attr_colors_df = colors_df[colors_df[\"Attribute\"] == attr_lower][[\"State\",\"Color\"]]\n",
    "attr_colors_dict = {}\n",
    "for rec in attr_colors_df.iterrows():\n",
    "    state = rec[1][\"State\"]\n",
    "    color = rec[1][\"Color\"]\n",
    "    attr_colors_dict[state] = color\n",
    "\n",
    "tt_copy = copy.deepcopy(tt)\n",
    "\n",
    "\n",
    "# --------------------------------------------\n",
    "# Phylogeny Lines\n",
    "Phylo.draw(tt_copy.tree,\n",
    "           axes=ax1,\n",
    "           show_confidence=False, \n",
    "           label_func = lambda x:'', \n",
    "           do_show=False,\n",
    "          )\n",
    "\n",
    "# Node Bars\n",
    "x_min = 0\n",
    "if hasattr(tt_copy.tree.root, \"marginal_inverse_cdf\"):\n",
    "    # Need to invert the node bar units\n",
    "    present_date = abs(tt_copy.tree.root.numdate - CURRENT_YEAR)\n",
    "    for rec in tree_df.iterrows():\n",
    "        conf_bar = rec[1][\"timetree_num_date_bar\"]\n",
    "        bar_lower = present_date - conf_bar[0]\n",
    "        bar_upper = present_date - conf_bar[1]\n",
    "        y = rec[1][\"timetree_coord_y\"]\n",
    "        ax1.plot([bar_lower, bar_upper], [y,y], alpha=0.5, c=\"grey\", linewidth=2)\n",
    "        \n",
    "        # Possibly update axis x-limits\n",
    "        if bar_lower < x_min:\n",
    "            x_min = bar_lower\n",
    "\n",
    "# --------------------------------------------\n",
    "# Phylogeny Scatter\n",
    "tip_data = tree_df[(tree_df[\"Branch_Support_Color\"] == TERM_COL)]\n",
    "#ax1.scatter(data=tip_data, x=\"timetree_coord_x\", y=\"timetree_coord_y\", s=0.5, c=\"grey\")\n",
    "\n",
    "# --------------------------------------------\n",
    "# Heatmap - Attribute\n",
    "\n",
    "# Heatmap x coordinates will stay the same\n",
    "heatmap_x_buffer = max(tree_df[\"timetree_coord_x\"]) * 0.05\n",
    "heatmap_x1 = max(tree_df[\"timetree_coord_x\"]) + heatmap_x_buffer\n",
    "heatmap_x2 = heatmap_x1 + (0.5 * heatmap_x_buffer)\n",
    "\n",
    "for rec in tip_data.iterrows():\n",
    "    y = rec[1][\"timetree_coord_y\"]\n",
    "    state = rec[1][attr]\n",
    "    color = attr_colors_dict[state]\n",
    "    ax1.plot([heatmap_x1, heatmap_x2], [y,y], alpha=1, c=color, linewidth=0.5)\n",
    "\n",
    "# Label\n",
    "ax1.text(x=heatmap_x1, y=-5, s=\"Clade\", ha=\"left\", rotation=45)\n",
    "\n",
    "# --------------------------------------------\n",
    "# Heatmap - Locus Coverage\n",
    "\n",
    "# Labels\n",
    "locus_x1 = heatmap_x1\n",
    "locus_x2 = heatmap_x2        \n",
    "locus_cov_df = tree_df[tree_df[\"pCD1_cov\"] != NO_DATA_CHAR][[\"pCD1_cov\", \"pMT1_cov\", \"pPCP1_cov\"]]\n",
    "\n",
    "\n",
    "for locus in locus_cov_df.columns:\n",
    "    locus_x1 = locus_x2 + (0.25 * heatmap_x_buffer)\n",
    "    locus_x2 = locus_x1 + (0.5 * heatmap_x_buffer)      \n",
    "    locus_name = locus.split(\"_\")[0]\n",
    "    ax1.text(x=locus_x1, y=-10, s=locus_name, ha=\"left\", rotation=45)\n",
    "\n",
    "# Bars\n",
    "\n",
    "# Black Background\n",
    "locus_x1 = heatmap_x1\n",
    "locus_x2 = heatmap_x2 \n",
    "for locus in locus_cov_df.columns:\n",
    "    # Update locus coordinates\n",
    "    locus_x1 = locus_x2 + (0.25 * heatmap_x_buffer)\n",
    "    locus_x2 = locus_x1 + (0.5 * heatmap_x_buffer)\n",
    "    rect_x = locus_x1\n",
    "    rect_y = min(tree_df[\"timetree_coord_y\"])\n",
    "    rect_width = locus_x2 - locus_x1\n",
    "    rect_height = max(tree_df[\"timetree_coord_y\"])\n",
    "    rect = patches.Rectangle(xy=(rect_x, rect_y), width=rect_width, height=rect_height, linewidth=0.5, edgecolor='black', facecolor=\"black\")\n",
    "    ax1.add_patch(rect)  \n",
    "\n",
    "# White lines for absence\n",
    "for rec in locus_cov_df.iterrows():   \n",
    "    locus_x1 = heatmap_x1\n",
    "    locus_x2 = heatmap_x2 \n",
    "    sample = rec[0]\n",
    "        \n",
    "    for locus in locus_cov_df.columns:\n",
    "        locus_x1 = locus_x2 + (0.25 * heatmap_x_buffer)\n",
    "        locus_x2 = locus_x1 + (0.5 * heatmap_x_buffer)\n",
    "        \n",
    "        cov = locus_cov_df[locus][sample]\n",
    "        y = tree_df[\"timetree_coord_y\"][sample]\n",
    "        if cov < 0.1: \n",
    "            ax1.plot([locus_x1, locus_x2], [y,y], alpha=1, c=\"white\", linewidth=1)\n",
    "\"\"\"\n",
    "# --------------------------------------------\n",
    "# Heatmap - pla depth\n",
    "\n",
    "locus_x1 = locus_x2 + (0.25 * heatmap_x_buffer)\n",
    "locus_x2 = locus_x1 + (0.5 * heatmap_x_buffer)\n",
    "\n",
    "# label\n",
    "ax1.text(x=locus_x1, y=-10, s=\"pla\", ha=\"left\", rotation=45)\n",
    "\n",
    "dep_df = tree_df[\n",
    "                   (tree_df[l + \"_\" + g + \"_dep_ratio\"] != NO_DATA_CHAR) &\n",
    "                   (~tree_df[\"BioSampleComment\"].str.contains(\"Assembly\")) &\n",
    "                   (tree_df[\"pPCP1_dep\"] != NO_DATA_CHAR)\n",
    "                   ][[\"pPCP1_dep\", \"pla_dep\", \"pPCP1_pla_dep_ratio\"]]\n",
    "dep_df = dep_df[dep_df[\"pPCP1_dep\"] >= 10]\n",
    "\n",
    "for rec in dep_df.sort_values(by=\"pPCP1_pla_dep_ratio\").iterrows():\n",
    "    sample = rec[0]\n",
    "    val = rec[1][\"pPCP1_pla_dep_ratio\"]\n",
    "    y = tree_df[\"timetree_coord_y\"][sample]\n",
    "    if val < (20/15):\n",
    "        color = \"green\"\n",
    "    else:\n",
    "        color = \"red\"\n",
    "    ax1.plot([locus_x1, locus_x2], [y,y], alpha=1, c=color, linewidth=1)\n",
    "\n",
    "\"\"\"\n",
    "# --------------------------------------------\n",
    "# Ticks\n",
    "ticks_dict = convert_timetree_ticks(tree=tt_copy.tree, step=500)\n",
    "# Set the new tick locations and labels\n",
    "# Leave out last one\n",
    "ax1.set_xticks(ticks_dict[\"tick_locs\"][:-1])\n",
    "ax1.set_xticklabels(ticks_dict[\"tick_labels\"][:-1])\n",
    "\n",
    "# --------------------------------------------\n",
    "# Lines\n",
    "# Vertical lines for dates\n",
    "for loc in ticks_dict[\"tick_locs\"][:-1]:\n",
    "    ax1.axvline(x=loc, color=\"darkgrey\", alpha=0.5)\n",
    "\n",
    "# --------------------------------------------\n",
    "# Limits\n",
    "\n",
    "x_buffer = math.ceil((ticks_dict[\"tick_locs\"][-1] - ticks_dict[\"tick_locs\"][0]) * 0.05)\n",
    "y_buffer = math.ceil(len(tt_copy.tree.get_terminals()) * 0.05)\n",
    "\n",
    "ax1.set_xlim(0 - x_buffer,)    \n",
    "ax1.set_ylim(len(tt_copy.tree.get_terminals()) + y_buffer, 0 - y_buffer)\n",
    "ax1.set_yticks([])    \n",
    "\n",
    "\n",
    "# --------------------------------------------\n",
    "# Labels\n",
    "ax1.set_xlabel(\"Year\")\n",
    "ax1.set_ylabel(\"\")\n",
    "\n",
    "# --------------------------------------------\n",
    "# Frames\n",
    "ax1.spines[\"right\"].set_visible(False)\n",
    "ax1.spines[\"top\"].set_visible(False)\n",
    "ax1.spines[\"left\"].set_visible(False)\n",
    "\n",
    "# -------------------------------\n",
    "# Legend\n",
    "\n",
    "legend_lines = [lines.Line2D([0], [0], color=attr_colors_dict[state], lw=4) \n",
    "                for state in attr_colors_dict]\n",
    "legend_labels = [state for state in attr_colors_dict]\n",
    "\n",
    "legend = fig.legend(legend_lines, \n",
    "           legend_labels,\n",
    "           ncol=len(legend_labels),\n",
    "           loc=\"center\", \n",
    "           bbox_to_anchor=(0.5, -0.05),\n",
    "           #bbox_to_anchor=(0.14,-0.13),\n",
    "           frameon=True,\n",
    "           borderpad=1.0,\n",
    "           edgecolor=\"black\",\n",
    "           title=\"Clade\",\n",
    "           title_fontsize=SM_FONT,\n",
    "           fontsize=4,\n",
    "          )\n",
    "\n",
    "locus_patches = [patches.Patch(facecolor=\"black\", edgecolor=\"black\",),\n",
    "                 patches.Patch(facecolor=\"white\", edgecolor=\"black\",),]\n",
    "\n",
    "locus_labels = [\">=10%\", \"<10%\", ]\n",
    "\n",
    "locus_legend = fig.legend(locus_patches, \n",
    "           locus_labels,\n",
    "           ncol=len(locus_labels),\n",
    "           loc=\"center\", \n",
    "           bbox_to_anchor=(0.5, -0.15),\n",
    "           frameon=True,\n",
    "           borderpad=1.0,\n",
    "           edgecolor=\"black\",\n",
    "           title=\"Plasmid Coverage\",\n",
    "           title_fontsize=SM_FONT\n",
    "           )\n",
    "\n",
    "\"\"\"gene_patches = [patches.Patch(facecolor=\"green\", edgecolor=\"black\",),\n",
    "                patches.Patch(facecolor=\"red\", edgecolor=\"black\",),\n",
    "                patches.Patch(facecolor=\"white\", edgecolor=\"black\",),]\n",
    "gene_labels = [\">= Plasmid Depth\", \"< Plasmid Depth\", \"No Sequence Data\"]\n",
    "\n",
    "gene_legend = fig.legend(gene_patches, \n",
    "           gene_labels,\n",
    "           ncol=len(gene_labels),\n",
    "           loc=\"center\", \n",
    "           bbox_to_anchor=(0.75, -0.15),\n",
    "           frameon=True,\n",
    "           borderpad=1.0,\n",
    "           edgecolor=\"black\",\n",
    "           title=\"Gene Depth\",\n",
    "           title_fontsize=SM_FONT\n",
    "           )\n",
    "\"\"\"\n",
    "# Save\n",
    "out_path = os.path.join(outdir, SCRIPT_NAME + \"_phylo.\" + FMT) \n",
    "plt.savefig(out_path, dpi=dpi, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "green-jackson",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Plot Locus vs. Gene Ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dedicated-airfare",
   "metadata": {},
   "outputs": [],
   "source": [
    "dep_df = tree_df[\n",
    "                   (tree_df[l + \"_\" + g + \"_dep_ratio\"] != NO_DATA_CHAR) &\n",
    "                   (~tree_df[\"BioSampleComment\"].str.contains(\"Assembly\"))\n",
    "                   ][[\"pPCP1_dep\", \"pla_dep\", \"pPCP1_pla_dep_ratio\"]]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "fig, ax1 = plt.subplots(1, 1, dpi=dpi, figsize=(3,2))\n",
    "\n",
    "colors = []\n",
    "for rec in dep_df.iterrows():\n",
    "    ratio = rec[1][\"pPCP1_pla_dep_ratio\"]\n",
    "    cov = rec[1][\"pPCP1_dep\"]\n",
    "    # If coverage is low, ratio doesn't matter\n",
    "    if cov < 10:\n",
    "        color = \"white\"\n",
    "    elif ratio <= (20/15):\n",
    "        color = \"green\"\n",
    "    else:\n",
    "        color = \"red\"\n",
    "    colors.append(color)\n",
    "\n",
    "#colors = [\"red\" if r > (20/15) else \"green\" for r,c in dep_df[[\"pPCP1_pla_dep_ratio\",\"pPCP1_dep\"]]]\n",
    "ax1.scatter(x=dep_df[\"pPCP1_dep\"], \n",
    "            y=dep_df[\"pla_dep\"], \n",
    "            s=10, \n",
    "            alpha=0.8, \n",
    "            edgecolor=\"black\", \n",
    "            linewidth=0.25, \n",
    "            color=colors\n",
    "            )\n",
    "\n",
    "x = 200\n",
    "y_norm = x * (20/15)\n",
    "y_low = x\n",
    "ax1.plot([0,x], [0,x*(20/15)], color=\"green\", linestyle=\"--\")\n",
    "#ax1.plot([0,175], [0,225], color=\"green\", linestyle=\"--\")\n",
    "#ax1.plot([0,x], [0,x/3], color=\"red\", linestyle=\"--\")\n",
    "ax1.plot([0,x], [0,x * (1/3)], color=\"red\", linestyle=\"--\")\n",
    "\n",
    "\n",
    "ax1.set_xlabel(\"Plasmid pPCP1 Depth (X)\",fontsize=4)\n",
    "ax1.set_ylabel(\"Gene pla Depth (X)\",fontsize=4)\n",
    "\n",
    "ax1.set_xlim(-5,x)\n",
    "ax1.set_ylim(-10,x*20/15)\n",
    "\n",
    "#ax1.set_title(\"Comparison of pPCP1 Depth and pla Depth\", fontsize=SM_FONT)\n",
    "\n",
    "legend_labels = [\"Gene >= Plasmid\", \"Gene < Plasmid\", \"Low Coverage\"]\n",
    "legend_lines = [\n",
    "    lines.Line2D([], [], markeredgecolor=\"black\", markerfacecolor='green', markeredgewidth=0.25, marker='.', linestyle='None', markersize=6,),\n",
    "    lines.Line2D([], [], markeredgecolor=\"black\", markerfacecolor='red', markeredgewidth=0.25, marker='.', linestyle='None', markersize=6,),\n",
    "    lines.Line2D([], [], markeredgecolor=\"black\", markerfacecolor='white', markeredgewidth=0.25, marker='.', linestyle='None', markersize=6,),    \n",
    "               ]\n",
    "\"\"\"fig.legend(legend_lines, \n",
    "           legend_labels,\n",
    "           loc='upper left',\n",
    "           frameon=True,\n",
    "           edgecolor=\"black\",\n",
    "          )\"\"\"\n",
    "\n",
    "legend = fig.legend(legend_lines, \n",
    "                    legend_labels, \n",
    "                    loc='center', \n",
    "                    bbox_to_anchor=(0.25, 0.79), \n",
    "                    frameon=True, \n",
    "                    fontsize=4,\n",
    "                    edgecolor=\"black\")\n",
    "\n",
    "legend.get_frame().set_linewidth(0.5)\n",
    "\n",
    "# Save\n",
    "out_path = os.path.join(outdir, SCRIPT_NAME + \"_pPCP1-pla-depth.\" + FMT) \n",
    "plt.savefig(out_path, dpi=dpi, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "excess-paint",
   "metadata": {},
   "source": [
    "# ? Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "interesting-handling",
   "metadata": {},
   "source": [
    "## Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "young-sound",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_df = os.path.join(outdir, SCRIPT_NAME + \".tsv\" )\n",
    "out_path_pickle_df = os.path.join(outdir,  SCRIPT_NAME + \".df.obj\" )\n",
    "\n",
    "tree_df.to_csv(out_path_df, sep=\"\\t\")\n",
    "with open(out_path_pickle_df,\"wb\") as outfile:\n",
    "    dill.dump(tree_df, outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "supported-error",
   "metadata": {},
   "source": [
    "## Timetrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "stupid-backing",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Phyloxml\n",
    "tt_copy = copy.deepcopy(tt)\n",
    "out_path_xml = os.path.join(outdir,  SCRIPT_NAME + \"_timetree.xml\" )\n",
    "Phylo.write(tt_copy.tree, out_path_xml, 'phyloxml')\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \"_timetree.nexus\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_timetree.treetime.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tt_copy, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tt_copy.tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \"_timetree.nwk\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "touched-delivery",
   "metadata": {},
   "source": [
    "## Divtrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "congressional-divorce",
   "metadata": {},
   "outputs": [],
   "source": [
    "tt_copy = copy.deepcopy(tt)\n",
    "# Convert to divtree\n",
    "for n in tt_copy.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "\n",
    "# Phyloxml\n",
    "out_path_xml = os.path.join(outdir,  SCRIPT_NAME + \"_divtree.xml\" )\n",
    "Phylo.write(tt_copy.tree, out_path_xml, 'phyloxml')\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \"_divtree.nexus\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_divtree.treetime.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tt_copy, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tt_copy.tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \"_divtree.nwk\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "secret-secondary",
   "metadata": {},
   "source": [
    "## Augur JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rubber-gilbert",
   "metadata": {},
   "outputs": [],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=out_path_nwk, \n",
    "    aln_path=aln_path, \n",
    "    tree=tt_copy.tree, \n",
    "    tree_df=tree_df, \n",
    "    color_keyword_exclude=[\"color\", \"coord\", \"lat\", \"lon\"],\n",
    "    type_convert = {\n",
    "        \"Branch_Number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"NODE0\"])\n",
    "\n",
    "out_path_augur_json = os.path.join(outdir, SCRIPT_NAME + \"_augur.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "smart-string",
   "metadata": {},
   "source": [
    "## Auspice JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "intense-harvard",
   "metadata": {},
   "outputs": [],
   "source": [
    "auspice_dict = auspice_export(\n",
    "    tree=tt_copy.tree, \n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=auspice_colors_path,\n",
    "    auspice_latlons_path=auspice_latlon_path, \n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(outdir, SCRIPT_NAME + \"_auspice.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "out_path_auspice_remote_json = os.path.join(auspice_remote_dir_path, AUSPICE_PREFIX + SCRIPT_NAME.replace(\"_\",\"-\") + \".json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "print(\"Validation successful for remote JSON.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "complex-ozone",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
