{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "juvenile-testimony",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "1. Plot Locus Coverage of Whole Genome.\n",
    "1. Add Locus Coverage to Dataframe.\n",
    "1. Plot Locus Coverage: Distribution.\n",
    "1. Plot Locus Coverage: Phlyo.\n",
    "1. Identify Interesting Samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "excellent-saver",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "serious-cannon",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "effective-train",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "from Bio import Phylo\n",
    "import copy\n",
    "import math\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import lines, patches, colors, gridspec, ticker\n",
    "import seaborn as sns\n",
    "import scipy.stats\n",
    "from functions import *\n",
    "\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.datasets import make_blobs\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_samples, silhouette_score\n",
    "\n",
    "import matplotlib.cm as cm\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "natural-combination",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "amino-depth",
   "metadata": {},
   "outputs": [],
   "source": [
    "#WILDCARDS = [\"all\", \"chromosome\", \"full\", \"30\"]\n",
    "WILDCARDS = [\"all\", \"chromosome\", \"full\", \"5\"]\n",
    "#project_dir = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/denmark/\"\n",
    "#project_dir = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/pla/\"\n",
    "project_dir = \"/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/\"\n",
    "results_dir = project_dir\n",
    "\n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "PRUNE = WILDCARDS[2]\n",
    "MISSING_DATA = WILDCARDS[3]\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "\n",
    "pd.set_option('display.max_rows', 100)\n",
    "pd.set_option('display.max_columns', 10)\n",
    "\n",
    "plt.rcParams['lines.linewidth'] = 0.5\n",
    "\n",
    "flierprops = dict(marker='o', markerfacecolor='black', markersize=1,\n",
    "                   markeredgecolor='none')\n",
    "\n",
    "D3_COL_PAL = [\"#1f77b4\", \"#ff7f0e\", \"#2ca02c\", \"#d62728\", \"#9467bd\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "removed-sponsorship",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "committed-involvement",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/locus_coverage_collect/all/\n"
     ]
    }
   ],
   "source": [
    "colors_path = os.path.join(\n",
    "    results_dir,\n",
    "    \"augur/{}/{}/{}/filter{}/beast/colors.tsv\"\n",
    "    .format(READS_ORIGIN, LOCUS_NAME, PRUNE, MISSING_DATA)\n",
    ")\n",
    "\n",
    "tree_path = os.path.join(\n",
    "    results_dir, \n",
    "    \"augur/{}/{}/{}/filter{}/beast/all.timetree.nwk\"\n",
    "    .format(READS_ORIGIN, LOCUS_NAME, PRUNE, MISSING_DATA)\n",
    ")\n",
    "\"\"\"metadata_path = os.path.join(\n",
    "    results_dir, \n",
    "    \"augur/{}/{}/{}/filter{}/beast/metadata.tsv\"\n",
    "    .format(READS_ORIGIN, LOCUS_NAME, PRUNE, MISSING_DATA)\n",
    ")\"\"\"\n",
    "metadata_path = os.path.join(results_dir, \"metadata/all/metadata.tsv\"\n",
    ")\n",
    "\n",
    "cov_df_path = os.path.join(\n",
    "    results_dir, \n",
    "    \"locus_coverage_collect/{}/locus_coverage.txt\"\n",
    "    .format(READS_ORIGIN)\n",
    ")\n",
    "\n",
    "dep_df_path = os.path.join(\n",
    "    results_dir, \n",
    "    \"locus_coverage_collect/{}/locus_depth.txt\"\n",
    "    .format(READS_ORIGIN)\n",
    ")\n",
    "ref_gff_path = os.path.join(\n",
    "    results_dir,\n",
    "    \"data/reference/GCA_000009065.1_ASM906v1_genomic/GCA_000009065.1_ASM906v1_genomic.gff\"\n",
    ")\n",
    "\n",
    "# ------------------------------------------\n",
    "# Output\n",
    "out_dir = os.path.join(results_dir, \"locus_coverage_collect/{}/\".format(READS_ORIGIN))\n",
    "print(out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "together-spread",
   "metadata": {},
   "source": [
    "## Import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "prostate-noise",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tree = Phylo.read(tree_path, format=\"newick\")\\ntree.ladderize(reverse=True)'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"tree = Phylo.read(tree_path, format=\"newick\")\n",
    "tree.ladderize(reverse=True)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "constitutional-malta",
   "metadata": {},
   "source": [
    "## Import Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "juvenile-spare",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.set_index(metadata_df.columns[0], inplace=True)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "#display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "communist-marketing",
   "metadata": {},
   "source": [
    "## Import Locus Dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "killing-commissioner",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] File /mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/locus_coverage_collect/all/locus_coverage.txt does not exist: '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/locus_coverage_collect/all/locus_coverage.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-2feb3e24b4a1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcov_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcov_df_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mcov_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Sample\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m#display(cov_df)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    674\u001b[0m         )\n\u001b[1;32m    675\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 676\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    677\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    678\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    446\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    447\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 448\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfp_or_buf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    449\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    450\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    878\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    879\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 880\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    881\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    882\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1112\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"c\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1113\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"c\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1114\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1115\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1116\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1889\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"usecols\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1890\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1891\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1892\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1893\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] File /mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/locus_coverage_collect/all/locus_coverage.txt does not exist: '/mnt/c/Users/ktmea/Projects/plague-phylogeography-projects/main/locus_coverage_collect/all/locus_coverage.txt'"
     ]
    }
   ],
   "source": [
    "cov_df = pd.read_csv(cov_df_path, sep='\\t')\n",
    "cov_df.set_index(\"Sample\", inplace=True)\n",
    "\n",
    "#display(cov_df)  \n",
    "\n",
    "dep_df = pd.read_csv(dep_df_path, sep='\\t')\n",
    "dep_df.set_index(\"Sample\", inplace=True)\n",
    "               \n",
    "#display(dep_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "presidential-tractor",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove samples not in tree dataframe (ex. outgroup)\n",
    "for sample in cov_df.index:\n",
    "    if sample not in metadata_df.index:\n",
    "        print(\"Dropping:\",sample)\n",
    "        cov_df.drop(sample, inplace=True)\n",
    "        dep_df.drop(sample, inplace=True)\n",
    "\n",
    "# exception\n",
    "\n",
    "#sample = \"SAMN01991268\"\n",
    "#if sample in metadata_df.index:\n",
    "#    cov_df.drop(sample, inplace=True)\n",
    "#    dep_df.drop(sample, inplace=True)    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pediatric-familiar",
   "metadata": {},
   "source": [
    "## Import Reference GFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "further-crisis",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gff_columns = [\n",
    "        \"seqname\",\n",
    "        \"source\",\n",
    "        \"feature\",\n",
    "        \"start\", #sequence numbering starting at 1.\n",
    "        \"end\",\n",
    "        \"score\",\n",
    "        \"strand\",\n",
    "        \"frame\",\n",
    "        \"attribute\",\n",
    "        \"id\",\n",
    "]\n",
    "\n",
    "ref_gff_tsv = ref_gff_path + \".tsv\"\n",
    "subprocess.run([\"rm\", \"-f\", ref_gff_tsv]) \n",
    "\n",
    "with open(ref_gff_path) as infile:\n",
    "    with open(ref_gff_tsv, \"a\") as outfile:\n",
    "        ref_gff_read = infile.read().split(\"\\n\")\n",
    "        for line in ref_gff_read:\n",
    "            if not line.startswith(\"#\"):\n",
    "                split_line = line.strip().split(\"\\t\")\n",
    "                attr = split_line[-1].split(\";\")\n",
    "                attr_id = attr[0].split(\"=\")\n",
    "                if len(attr_id) < 2:\n",
    "                    continue\n",
    "                attr_id = attr_id[1]\n",
    "                line += \"\\t{}\".format(attr_id)\n",
    "                outfile.write(line + \"\\n\")\n",
    "                \n",
    "ref_df = pd.read_csv(ref_gff_tsv, sep='\\t', header=None)\n",
    "ref_df.columns = ref_gff_columns\n",
    "ref_df.set_index(\"id\", inplace=True)\n",
    "#display(ref_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "considerable-portugal",
   "metadata": {},
   "source": [
    "## Create Dataframes for replicons and genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "southeast-scope",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_rep_df = ref_df[ref_df[\"feature\"] == \"region\"]\n",
    "#display(ref_rep_df)\n",
    "\n",
    "ref_gene_df = ref_df[ref_df[\"feature\"] == \"gene\"]\n",
    "#display(ref_gene_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "worse-breach",
   "metadata": {},
   "source": [
    "## Separate Dataframe by Type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "raised-color",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov_rep_df = copy.deepcopy(cov_df)\n",
    "cov_gene_df = copy.deepcopy(cov_df)\n",
    "dep_rep_df = copy.deepcopy(dep_df)\n",
    "dep_gene_df = copy.deepcopy(dep_df)\n",
    "non_replicon_loci = []\n",
    "non_gene_loci = []\n",
    "\n",
    "for col in cov_df.columns:\n",
    "    if col not in ref_rep_df.index:\n",
    "        non_replicon_loci.append(col)\n",
    "    if col not in ref_gene_df.index:\n",
    "        non_gene_loci.append(col)\n",
    "        \n",
    "cov_rep_df.drop(columns=non_replicon_loci, inplace=True)\n",
    "display(cov_rep_df)\n",
    "\n",
    "dep_rep_df.drop(columns=non_replicon_loci, inplace=True)\n",
    "#display(dep_rep_df)\n",
    "       \n",
    "cov_gene_df.drop(columns=non_gene_loci, inplace=True)\n",
    "#display(cov_gene_df)\n",
    "\n",
    "dep_gene_df.drop(columns=non_gene_loci, inplace=True)\n",
    "#display(dep_gene_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ideal-bathroom",
   "metadata": {},
   "source": [
    "---\n",
    "# pPCP1 Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "available-miller",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_seqname = \"AL109969\"\n",
    "target_locus = \"AL109969.1:1..9612\" # pPCP1\n",
    "target_gene = \"gene-YPPCP1.07\" # pla\n",
    "\n",
    "# Danish exlude\n",
    "# exclude_samples = [\"P187\",\"P212\",\"P387\",]\n",
    "\n",
    "# Exclude IS100 and the IS100 transposase\n",
    "exclude_genes = [\"gene-YPPCP1.01\", \"gene-YPPCP1.02\" ]\n",
    "gene_labels = {\n",
    "    \"gene-YPPCP1.01\": \n",
    "        {\n",
    "            \"label\": \"IS100\",\n",
    "            \"coord\": [87, 1109],\n",
    "            \"color\" : \"#c4c4c4\",\n",
    "        },\n",
    "    \"gene-YPPCP1.02\": \n",
    "        {\n",
    "            \"label\": \"IS100\",\n",
    "            \"coord\": [1106, 1888],\n",
    "            \"color\" : \"#c4c4c4\",            \n",
    "        },\n",
    "    \"gene-YPPCP1.03\": \n",
    "        {\n",
    "            \"label\": \"rop\",\n",
    "            \"coord\": [2925, 3119],\n",
    "            \"color\": \"#1f77b4\",\n",
    "        },    \n",
    "    \"gene-YPPCP1.04\": \n",
    "        {\n",
    "            \"label\": \"pim\",\n",
    "            \"coord\": [4355, 4780],\n",
    "            \"color\": \"#1f77b4\",            \n",
    "        },  \n",
    "    \"gene-YPPCP1.05c\": \n",
    "        {\n",
    "            \"label\": \"pst\",\n",
    "             \"coord\": [4815, 5888],\n",
    "            \"color\": \"#1f77b4\",            \n",
    "        }, \n",
    "    \"gene-YPPCP1.06\": \n",
    "        {\n",
    "            \"label\": \"Hypothetical protein\",\n",
    "            \"coord\": [6006, 6422],\n",
    "            \"color\": \"#1f77b4\",            \n",
    "        },     \n",
    "    \"gene-YPPCP1.07\": \n",
    "        {\n",
    "            \"label\": \"pla\",\n",
    "            \"coord\": [6665, 7603],\n",
    "            \"color\" : \"#ff7f0e\",\n",
    "        }, \n",
    "    \"gene-YPPCP1.08c\": \n",
    "        {\n",
    "            \"label\": \"Probable transcriptional regulator\",\n",
    "            \"coord\": [7790, 8089],\n",
    "            \"color\" : \"#ff7f0e\",            \n",
    "        }, \n",
    "    \"gene-YPPCP1.09c\": \n",
    "        {\n",
    "            \"label\": \"Hypothetical protein\",\n",
    "            \"coord\": [8089, 8436],\n",
    "            \"color\" : \"#ff7f0e\",            \n",
    "        },    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fossil-alliance",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the plasmid genes from the reference\n",
    "ppcp1_genes = list(ref_gene_df[ref_gene_df[\"seqname\"] == target_seqname].index)\n",
    "\n",
    "# Select the depth\n",
    "ppcp1_genes_df = dep_gene_df[ppcp1_genes]\n",
    "\n",
    "display(ppcp1_genes_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "genuine-start",
   "metadata": {},
   "source": [
    "---\n",
    "## Calculate Depletion Ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "israeli-evening",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene ID\n",
    "depleted_gene = \"gene-YPPCP1.07\" # pla\n",
    "baseline_gene = \"gene-YPPCP1.05c\" #pst\n",
    "\n",
    "# Gene Label\n",
    "depleted_gene_label = gene_labels[depleted_gene][\"label\"]\n",
    "baseline_gene_label = gene_labels[baseline_gene][\"label\"]\n",
    "\n",
    "# Gene Depth\n",
    "baseline_dep = list(ppcp1_genes_df[baseline_gene])\n",
    "depleted_dep = list(ppcp1_genes_df[depleted_gene])\n",
    "\n",
    "# Accessory plotting variables\n",
    "sample = list(ppcp1_genes_df.index)\n",
    "#timetree_num_date = metadata_df[\"timetree_num_date\"][ppcp1_genes_df.index]\n",
    "timetree_num_date = metadata_df[\"date_mean\"][ppcp1_genes_df.index]\n",
    "strain = metadata_df[\"strain\"][ppcp1_genes_df.index]\n",
    "#root_rtt_dist = metadata_df[\"root_rtt_dist\"][ppcp1_genes_df.index]\n",
    "country = metadata_df[\"country\"][ppcp1_genes_df.index]\n",
    "ratio = []\n",
    "population = metadata_df[\"population\"][ppcp1_genes_df.index]\n",
    "\n",
    "for depleted,baseline in zip(depleted_dep, baseline_dep):\n",
    "    if baseline == 0:\n",
    "        ratio.append(0)\n",
    "    else:\n",
    "        ratio.append(depleted/baseline)\n",
    "\n",
    "data = {\n",
    "    \"baseline\": baseline_dep, \n",
    "    \"depleted\": depleted_dep, \n",
    "    \"ratio\": ratio, \n",
    "    \"sample\" : sample,\n",
    "    \"timetree_num_date\": timetree_num_date,\n",
    "    \"strain\" : strain,\n",
    "    \"country\" : country,\n",
    "    \"population\" : population,\n",
    "    #\"root_rtt_dist\" : root_rtt_dist,\n",
    "    }\n",
    "\n",
    "ratio_df = pd.DataFrame(data)\n",
    "ratio_df.set_index(\"sample\", inplace=True)\n",
    "ratio_df.sort_values(by=\"ratio\", inplace=True)\n",
    "#ratio_df.sort_values(by=\"baseline\", inplace=True)\n",
    "\n",
    "display(ratio_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "advanced-state",
   "metadata": {},
   "source": [
    "## Filter Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "endless-coordinate",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only high cov samples\n",
    "ppcp1_genes_orig_df = copy.deepcopy(ppcp1_genes_df)\n",
    "ratio_orig_df = copy.deepcopy(ratio_df)\n",
    "\n",
    "\n",
    "high_cov_samples = list(dep_rep_df[dep_rep_df[target_locus] >= 10].index)\n",
    "ppcp1_genes_df = ppcp1_genes_df.loc[high_cov_samples]\n",
    "\n",
    "ratio_df =ratio_df.loc[high_cov_samples]\n",
    "\n",
    "# Exclude Danish samples\n",
    "#ppcp1_genes_df.drop(index=exclude_samples, inplace=True)\n",
    "#ratio_df.drop(index=exclude_samples, inplace=True)\n",
    "\n",
    "display(ppcp1_genes_df)\n",
    "display(ratio_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "administrative-closer",
   "metadata": {},
   "source": [
    "## Plot Gene coverage across plasmid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "confirmed-richards",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [480, 480]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "LOCUS_DEP = 10\n",
    "RATIO = 0.7\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "fig, axes = plt.subplots(1,1,figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "ax = axes\n",
    "df = ppcp1_genes_df[ppcp1_genes_df[\"gene-YPPCP1.03\"] > 10]\n",
    "data = {\n",
    "    col:list(df[col]) for col in df.columns\n",
    "}\n",
    "df = pd.DataFrame(data)\n",
    "colors = [gene_labels[gene][\"color\"] for gene in gene_labels]\n",
    "sns.boxplot(\n",
    "    ax=ax,\n",
    "    data=df,\n",
    "    flierprops=flierprops,\n",
    "    palette=colors,\n",
    ")\n",
    "xtick_labels = [\"{} {}\".format(gene_labels[gene][\"label\"],gene_labels[gene][\"coord\"])  for gene in gene_labels]\n",
    "ax.set_xticklabels(xtick_labels)\n",
    "plt.xticks(rotation=45, ha=\"right\")\n",
    "ax.set_xlabel(\"Gene\")\n",
    "ax.set_ylabel(\"Average Depth (X)\")\n",
    "plt.title(\"Sequencing Depth of pPCP1 Genes (N={})\".format(len(df)))\n",
    "\n",
    "out_path = os.path.join(out_dir, \"ppcp1_gene_depth\")\n",
    "plt.savefig(out_path + \".png\", dpi=DPI, bbox_inches = \"tight\", facecolor=\"white\")\n",
    "plt.savefig(out_path + \".svg\", dpi=DPI, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "roman-baltimore",
   "metadata": {},
   "source": [
    "---\n",
    "# Detect K Cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "official-restaurant",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = np.array(ratio_df[\"ratio\"]).reshape(-1, 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "polish-samoa",
   "metadata": {},
   "source": [
    "### Elbow Method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "electrical-governor",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate distortion for a range of number of cluster\n",
    "distortions = []\n",
    "\n",
    "iter = 10\n",
    "for i in range(1, iter):\n",
    "    km = KMeans(\n",
    "        n_clusters=i, \n",
    "        init='random',\n",
    "        n_init=10, \n",
    "        max_iter=300,\n",
    "        tol=1e-04, \n",
    "        random_state=0\n",
    "    )\n",
    "    km.fit(X)\n",
    "    distortions.append(km.inertia_)\n",
    "\n",
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [480, 480]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "LOCUS_DEP = 10\n",
    "RATIO = 0.7\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "fig, ax = plt.subplots(1,1,figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "sns.lineplot(\n",
    "    ax=ax,\n",
    "    x=range(1, iter), \n",
    "    y=distortions,\n",
    ")\n",
    "sns.scatterplot(\n",
    "    ax=ax,\n",
    "    x=range(1, iter), \n",
    "    y=distortions,\n",
    "    s=20,\n",
    "    ec=\"black\",\n",
    "    lw=0.5,\n",
    ")\n",
    "\n",
    "ax.set_xlabel('Number of Clusters')\n",
    "ax.set_ylabel('Distortion')\n",
    "fig.suptitle(\"Cluster Detection by Elbow Method\")\n",
    "\n",
    "out_path = os.path.join(out_dir, \"elbow_{}_{}\".format(depleted_gene_label, baseline_gene_label))\n",
    "plt.savefig(out_path + \".png\", bbox_inches = \"tight\", facecolor=\"white\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "signed-librarian",
   "metadata": {},
   "source": [
    "## Silhoutte Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "korean-valley",
   "metadata": {},
   "outputs": [],
   "source": [
    "range_n_clusters = [2, 3, 4, 5, 6]\n",
    "\n",
    "highest_score = 0\n",
    "highest_cluster = 0\n",
    "\n",
    "for n_clusters in range_n_clusters:\n",
    "    \n",
    "    clusterer = KMeans(n_clusters=n_clusters, random_state=10)\n",
    "    cluster_labels = clusterer.fit_predict(X)\n",
    "\n",
    "    silhouette_avg = silhouette_score(X, cluster_labels)\n",
    "    print(\"For n_clusters =\", n_clusters,\n",
    "          \"The average silhouette_score is :\", silhouette_avg)\n",
    "    \n",
    "    if silhouette_avg > highest_score:\n",
    "        highest_cluster = n_clusters\n",
    "        highest_score = silhouette_avg\n",
    "\n",
    "print(highest_cluster)\n",
    "# To Plot see: https://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cathedral-louisiana",
   "metadata": {},
   "source": [
    "---\n",
    "# Cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stuck-diploma",
   "metadata": {},
   "source": [
    "### Manual Override"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "canadian-berlin",
   "metadata": {},
   "outputs": [],
   "source": [
    "highest_cluster = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accessory-clinton",
   "metadata": {},
   "outputs": [],
   "source": [
    "km = KMeans(\n",
    "    n_clusters=highest_cluster, init='random',\n",
    "    n_init=10, max_iter=300, \n",
    "    tol=1e-04, random_state=0\n",
    ")\n",
    "y_km = km.fit_predict(X)\n",
    "\n",
    "for ratio, cluster, sample in zip(X, y_km, ratio_df.index):\n",
    "    ratio_df.at[sample, \"cluster\"] = cluster\n",
    "    \n",
    "#display(ratio_df)\n",
    "cluster_max_ratios = {cluster:0 for cluster in range(0, highest_cluster)}\n",
    "\n",
    "for cluster in range(0, highest_cluster):\n",
    "    ratios = [r[0] for r in X[y_km == cluster]]\n",
    "    cluster_max_ratios[cluster] = max(ratios)\n",
    "\n",
    "cluster_order = list({cluster:ratio for cluster,ratio in sorted(cluster_max_ratios.items(), key=lambda item: item[1])}.keys())\n",
    "\n",
    "for rec in ratio_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    cluster = rec[1][\"cluster\"]\n",
    "    new_cluster = cluster_order.index(cluster)\n",
    "    ratio_df.at[sample, \"cluster\"] = new_cluster\n",
    "    \n",
    "    \n",
    "display(ratio_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "enclosed-courtesy",
   "metadata": {},
   "source": [
    "---\n",
    "## Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ruled-radical",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [2400, 720]\n",
    "DPI=400\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "LOCUS_DEP = 10\n",
    "RATIO = 0.7\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "# 2 x2\n",
    "#fig, axes = plt.subplots(2,2,figsize=FIGSIZE, dpi=DPI)\n",
    "# 1 x 3\n",
    "fig, axes = plt.subplots(1,3,figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "fig.subplots_adjust(wspace=0.4, hspace=0.4)\n",
    "\n",
    "#------------------------------------------------\n",
    "# Regresion\n",
    "#ax = axes[0][0]\n",
    "ax = axes[0]\n",
    "\n",
    "ax.set_title(\"A\", fontsize=FONTSIZE * 1.5, fontweight=\"bold\")\n",
    "\n",
    "df = ratio_df[ratio_df[\"cluster\"] == cluster]\n",
    "#for cluster in range(0, highest_cluster):\n",
    "for country in set(ratio_df[\"country\"]):\n",
    "    df = ratio_df[ratio_df[\"country\"] == country]\n",
    "    sns.regplot(\n",
    "        ax=ax,\n",
    "        data=df,\n",
    "        x=\"baseline\",\n",
    "        y=\"depleted\", \n",
    "        #ci=None,\n",
    "        scatter_kws={\"s\": 5, \"ec\":\"black\", \"lw\": 0.25, \"zorder\":1,},   \n",
    "        line_kws={\"zorder\":0, },\n",
    "    )\n",
    "    \n",
    "ax.set_xlabel(\"$\\it{\" + \"{}\".format(baseline_gene_label) + \"}$ Depth (X)\")\n",
    "ax.set_ylabel(\"$\\it{\" + \"{}\".format(depleted_gene_label) + \"}$ Depth (X)\")\n",
    "ax.set_xlim(0, max(ratio_df[\"baseline\"] + 10))\n",
    "\n",
    "reg_handles = []\n",
    "for cluster in range(highest_cluster):\n",
    "    cluster_df = ratio_df[ratio_df[\"cluster\"] ==cluster]\n",
    "    x = list(cluster_df[\"baseline\"])\n",
    "    y = list(cluster_df[\"depleted\"])\n",
    "    slope, intercept, r_value, p_value, stderr, = scipy.stats.linregress(x,y)\n",
    "    \n",
    "    p_sig = \"\"\n",
    "    if p_value < 0.05:\n",
    "        p_sig = \"*\"\n",
    "        \n",
    "    r_squared = r_value * r_value\n",
    "    r_squared_pretty = str(round(r_squared, 2))\n",
    "    \n",
    "    intercept_pretty = \"+ \" + str(round(intercept, 2)).replace(\"+\",\"\")\n",
    "    if \"-\" in str(intercept):\n",
    "        intercept_pretty = \"- \" + str(round(intercept, 2)).replace(\"-\",\"\")\n",
    "        \n",
    "    reg_handles.append(\n",
    "        lines.Line2D([0], [0], \n",
    "        label = (\n",
    "            \"R$^{2}$ : \" + \"{}\".format(r_squared_pretty)\n",
    "            #\"y = {}x {}\".format(round(slope,2),intercept_pretty)\n",
    "            + \"\\np   : {:0.2e}{}\".format(p_value, p_sig)\n",
    "        ),\n",
    "        color=D3_COL_PAL[cluster])\n",
    "    )\n",
    "\n",
    "legend = ax.legend(handles=reg_handles, loc=2, fontsize=FONTSIZE * 0.75, edgecolor=\"black\")\n",
    "legend.get_frame().set_linewidth(0.25)\n",
    "\n",
    "#------------------------------------------------\n",
    "# Histogram\n",
    "#ax = axes[0][1]\n",
    "ax = axes[1]\n",
    "\n",
    "ax.set_title(\"B\", fontsize=FONTSIZE * 1.5, fontweight=\"bold\")\n",
    "sns.histplot(\n",
    "    ax=ax,\n",
    "    data=ratio_df,\n",
    "    x=\"ratio\",\n",
    "    bins=len(ratio_df),\n",
    "    #hue=\"cluster\",\n",
    "    hue=\"country\",\n",
    "    palette = D3_COL_PAL[0:len(set(ratio_df[\"country\"]))],\n",
    "    #palette = D3_COL_PAL[0:highest_cluster],\n",
    "    #alpha=0.75,\n",
    "    zorder=0,\n",
    ")\n",
    "sns.kdeplot(\n",
    "    ax=ax,\n",
    "    data=ratio_df,\n",
    "    x=\"ratio\",\n",
    "    #hue=\"cluster\",\n",
    "    hue = \"country\",\n",
    "    #palette = D3_COL_PAL[0:highest_cluster],\n",
    "    palette = D3_COL_PAL[0:len(set(ratio_df[\"country\"]))],    \n",
    "    #alpha=1.0,\n",
    "    fill=True,\n",
    "    zorder=1,\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "ax.set_xlim(0)\n",
    "ax.set_ylabel(\"Number of Samples\")\n",
    "ax.set_xlabel(\n",
    "    \"$\\it{\" \n",
    "    + \"{}\".format(depleted_gene_label)\n",
    "    + \"}$ Depth (X)  / $\\it{\"\n",
    "    + \"{}\".format(baseline_gene_label)\n",
    "    + \"}$ Depth (X)\")\n",
    "ax.legend().remove()\n",
    "\n",
    "#------------------------------------------------\n",
    "# Boxplot\n",
    "\n",
    "\"\"\"\n",
    "ax = axes[1][0]\n",
    "ax.set_title(\"C\", fontsize=FONTSIZE * 1.5, fontweight=\"bold\")\n",
    "\n",
    "sns.boxplot(\n",
    "    ax=ax,\n",
    "    data=ratio_df,\n",
    "    x=\"cluster\",\n",
    "    y=\"ratio\",\n",
    "    flierprops=flierprops,\n",
    ")\n",
    "ax.set_xlabel(\"Cluster\")\n",
    "ax.set_xticklabels([\"Depleted\", \"Normal\"])\n",
    "ax.set_ylabel(\n",
    "    \"$\\it{\" \n",
    "    + \"{}\".format(depleted_gene_label)\n",
    "    + \"}$ Depth (X)  / $\\it{\"\n",
    "    + \"{}\".format(baseline_gene_label)\n",
    "    + \"}$ Depth (X)\")\n",
    "\"\"\"\n",
    "\n",
    "# --------------------------------------\n",
    "# TIMELINE\n",
    "#ax = axes[1][1]\n",
    "ax = axes[2]\n",
    "ax.set_title(\"C\", fontsize=FONTSIZE * 1.5, fontweight=\"bold\")\n",
    "sns.scatterplot(\n",
    "    ax=ax, \n",
    "    data=ratio_df, \n",
    "    x=\"timetree_num_date\", \n",
    "    y=\"ratio\",\n",
    "    s=10,\n",
    "    ec=\"black\",\n",
    "    #palette = D3_COL_PAL[0:highest_cluster],\n",
    "    #hue=\"cluster\",\n",
    "    hue=\"country\",\n",
    "    palette = D3_COL_PAL[0:len(set(ratio_df[\"country\"]))],     \n",
    ")\n",
    "ax.legend().remove()\n",
    "\n",
    "#------------------------------------------------\n",
    "# Date Confidences\n",
    "\"\"\"for sample in ratio_df.index:\n",
    "    conf = [float(c) for c in metadata_df[\"timetree_num_date_confidence\"][sample].strip(\"[]\").split(\",\")]\n",
    "    ratio = ratio_df[\"ratio\"][sample]\n",
    "    cluster = int(ratio_df[\"cluster\"][sample])\n",
    "    color = D3_COL_PAL[cluster]\n",
    "    ax.add_patch(\n",
    "        patches.Rectangle(\n",
    "            (conf[0], ratio), conf[1] - conf[0], 0.02, linewidth=0, facecolor=color, alpha=0.20, zorder=0))\"\"\"\n",
    "\n",
    "\n",
    "#ax.set_xlim(1150,1850)\n",
    "ax.set_xlabel(\"Date\")\n",
    "ax.set_ylabel(\n",
    "    \"$\\it{\" \n",
    "    + \"{}\".format(depleted_gene_label)\n",
    "    + \"}$ Depth (X)  / $\\it{\"\n",
    "    + \"{}\".format(baseline_gene_label)\n",
    "    + \"}$ Depth (X)\")\n",
    "\n",
    "#------------------------------------------------\n",
    "# Legend\n",
    "#normal_patch = patches.Patch(color=D3_COL_PAL[1], label='Normal $\\it{pla}$')\n",
    "normal_patch = patches.Patch(color=D3_COL_PAL[1], label='United States of America')\n",
    "#low_patch = patches.Patch(color=D3_COL_PAL[0], label='Depleted $\\it{pla}$')\n",
    "low_patch = patches.Patch(color=D3_COL_PAL[0], label='Madagascar')\n",
    "#legend = fig.legend(handles=[normal_patch, low_patch], bbox_to_anchor=(0.6,0.98), edgecolor=\"black\")\n",
    "legend = fig.legend(handles=[normal_patch, low_patch], bbox_to_anchor=(0.575,1.2), edgecolor=\"black\")\n",
    "legend.get_frame().set_linewidth(0.5)\n",
    "\n",
    "\"\"\"fig.suptitle(\n",
    "    \"Relative depletion of the plasminogen activator ($\\it{pla}$) virulence factor\", \n",
    "    x=0.5, \n",
    "    y=1.3, \n",
    "    fontsize=FONTSIZE * 2,\n",
    ")\"\"\"\n",
    "\n",
    "out_path = os.path.join(out_dir, \"depletion_{}\".format(depleted_gene_label))\n",
    "plt.savefig(out_path + \".png\", bbox_inches = \"tight\", facecolor=\"white\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "gorgeous-harris",
   "metadata": {},
   "source": [
    "## Add Depletion Statistics to Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "timely-accountability",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df[\"baseline_{}\".format(baseline_gene_label)] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"depleted_{}\".format(depleted_gene_label)] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"ratio_{}_{}\".format(depleted_gene_label, baseline_gene_label)] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "metadata_df[\"cluster_{}_{}\".format(depleted_gene_label, baseline_gene_label)] = [NO_DATA_CHAR] * len(metadata_df)\n",
    "\n",
    "# Including filter\n",
    "#for rec in ratio_df.iterrows():\n",
    "# Excluding filtering\n",
    "for sample in ratio_orig_df.index:\n",
    "    metadata_df.at[sample, \"baseline_{}\".format(baseline_gene_label)] = ratio_orig_df[\"baseline\"][sample]\n",
    "    metadata_df.at[sample, \"depleted_{}\".format(depleted_gene_label)] = ratio_orig_df[\"depleted\"][sample]\n",
    "    metadata_df.at[sample, \"ratio_{}_{}\".format(depleted_gene_label, baseline_gene_label)] = ratio_orig_df[\"ratio\"][sample]\n",
    "    if sample in ratio_df.index:\n",
    "        metadata_df.at[sample, \"cluster_{}_{}\".format(depleted_gene_label, baseline_gene_label)] = ratio_df[\"cluster\"][sample]\n",
    "    \n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "martial-victoria",
   "metadata": {},
   "source": [
    "## Draw Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "divine-meter",
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in tree.get_terminals():\n",
    "    sample = c.name\n",
    "    cluster = metadata_df[\"cluster_{}_{}\".format(depleted_gene_label, baseline_gene_label)][sample]\n",
    "    if cluster == NO_DATA_CHAR:\n",
    "        color = \"#c4c4c4\"\n",
    "    else:\n",
    "        color = D3_COL_PAL[int(cluster)]\n",
    "    c.color = color\n",
    "\n",
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [960, 960]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=4\n",
    "LOCUS_DEP = 10\n",
    "RATIO = 0.7\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "fig, ax = plt.subplots(1,figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "Phylo.draw(\n",
    "    tree, \n",
    "    axes=ax, \n",
    "    show_confidence=False, \n",
    "    label_func = lambda x: metadata_df[\"country_date_strain\"][x.name] if x.is_terminal() else \"\", \n",
    "    do_show=False\n",
    ")\n",
    "\n",
    "ax.axis(\"off\")\n",
    "#ax.set_ylabel(\"\")\n",
    "#ax.set_yticklabels([])\n",
    "#ax.set_xlabel(\"Years\")\n",
    "\n",
    "#------------------------------------------------\n",
    "# Legend\n",
    "normal_patch = patches.Patch(color=D3_COL_PAL[1], label='Normal $\\it{pla}$')\n",
    "low_patch = patches.Patch(color=D3_COL_PAL[0], label='Depleted $\\it{pla}$')\n",
    "unknown_patch = patches.Patch(color=\"#c4c4c4\", label='Unknown $\\it{pla}$')\n",
    "legend = fig.legend(handles=[normal_patch, low_patch, unknown_patch], loc=6, edgecolor=\"black\")\n",
    "legend.get_frame().set_linewidth(0.5)\n",
    "\n",
    "out_path = os.path.join(out_dir, \"depletion_tree_{}\".format(depleted_gene_label))\n",
    "plt.savefig(out_path + \".png\", bbox_inches = \"tight\", facecolor=\"white\", dpi=DPI)\n",
    "plt.savefig(out_path + \".svg\", bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "excess-paint",
   "metadata": {},
   "source": [
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "flush-destination",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "complex-ozone",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_metadata = os.path.join(out_dir, \"metadata.tsv\")\n",
    "metadata_df.to_csv(out_path_metadata, sep=\"\\t\", index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "little-therapy",
   "metadata": {},
   "source": [
    "## Timetree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "excited-sport",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_tree = copy.deepcopy(tree)\n",
    "\n",
    "metadata_to_comment(out_tree, metadata_df)    \n",
    "out_path_tree_nex = os.path.join(out_dir, \"all.timetree.nex\")\n",
    "Phylo.write(out_tree, out_path_tree_nex, \"nexus\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "elegant-spider",
   "metadata": {},
   "source": [
    "## Prior Distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "iraqi-oxide",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [480, 480]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=4\n",
    "LOCUS_DEP = 10\n",
    "RATIO = 0.7\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "priors = {\n",
    "    \"Denmark\": {\"mu\": 1330, \"range\" : 230}, \n",
    "    \"Rostov2033\": {\"mu\": 1767.5, \"range\": 2.75 },\n",
    "    \"Azov38\": {\"mu\":2021-471.0, \"range\": 75.0},\n",
    "    \"CHE1\" : {\"mu\": 2021-371.0, \"range\" :75.0}, \n",
    "    \"BED\" : {\"mu\": 2021-423.5, \"range\" : 18.75}\n",
    "}\n",
    "\n",
    "for prior in priors:\n",
    "          \n",
    "    fig, ax = plt.subplots(1,figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "\n",
    "    mu = priors[prior][\"mu\"]\n",
    "    date_range = priors[prior][\"range\"]\n",
    "    sigma = date_range / 2\n",
    "    x = np.linspace(mu - 3*sigma, mu + 3*sigma, 100)\n",
    "    y = scipy.stats.norm.pdf(x, mu, sigma)\n",
    "\n",
    "    ax.fill_between(x, y, alpha=0.25)\n",
    "\n",
    "    sns.lineplot(ax=ax, x=x, y=y)\n",
    "    #ax.plot(x, scipy.stats.norm.pdf(x, mu, sigma))\n",
    "    ax.set_title(\"{} Collection Date Prior\".format(prior), fontsize=FONTSIZE * 1.5, y=1.1)\n",
    "    ax.set_xlabel(\"Date\", fontsize=FONTSIZE * 1.5)\n",
    "    ax.set_ylabel(\"Density\", fontsize=FONTSIZE * 1.5)\n",
    "\n",
    "    ax.axvline(x=mu, color=\"black\", label=\"Mean Site Date\")\n",
    "    ax.axvline(x=mu - date_range, color=\"black\", linestyle=\"--\", label=\"Site Date Range\")\n",
    "    ax.axvline(x=mu + date_range, color=\"black\", linestyle=\"--\",)\n",
    "    ax.legend(bbox_to_anchor=(0.30,1), edgecolor=\"black\").get_frame().set_linewidth(0.5)\n",
    "\n",
    "\n",
    "#out_path = os.path.join(out_dir, \"prior_denmark\")\n",
    "#plt.savefig(out_path + \".png\", bbox_inches = \"tight\", facecolor=\"white\")\n",
    "#plt.savefig(out_path + \".svg\", bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "prompt-slide",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
