{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Objectives"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Convert newick tree to PhyloXML format.\n",
    "1. Fixes internal node names from an IQTREE tree. \n",
    "   - internal nodes by default get named the confidence (ex. \"99/100\")\n",
    "   - this notebook instead gives them the name NODEi where i is an integer starting from 0.\n",
    "   - NODE0 therefore becomes the root.\n",
    "1. Convert tree to data frame.\n",
    "1. Combine metadata with tree data frame."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Module Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Phylo\n",
    "import os\n",
    "import copy\n",
    "import augur\n",
    "import json\n",
    "from matplotlib import colors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input File Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tree_path = \"../../docs/results/latest/iqtree/iqtree-core_chromosome.filter5.treefile\"\n",
    "tree_path = \"../../docs/results/latest/iqtree/iqtree-core_chromosome.filter5_post.cf.tree\"\n",
    "metadata_path = \"../../docs/results/latest/metadata/metadata.tsv\"\n",
    "outdir = \"../../docs/results/latest/parse_tree\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "\n",
    "# Custom script variables\n",
    "SCRIPT_NAME = \"parse_tree\"\n",
    "NAME_COL = \"Sample\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Convert tree to PhyloXML format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_xml = os.path.join(outdir, \"temp.xml\")\n",
    "Phylo.convert(tree_path, \"newick\", tmp_xml, \"phyloxml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. Fix internal node names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in XML tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the new tree\n",
    "tree = Phylo.read(tmp_xml, \"phyloxml\")\n",
    "\n",
    "# Deepest nodes first\n",
    "tree.ladderize(reverse=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fix node names and confidences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Counter for integer node names\n",
    "node_i = 0\n",
    "\n",
    "# Iterate through the nodes in the tree\n",
    "for c in tree.find_clades():\n",
    "    # Check if the name has 'accidentally' become the confidence vals\n",
    "    ufboot_val = NO_DATA_CHAR    \n",
    "    scf_val = NO_DATA_CHAR\n",
    "    # If it's a named node, parse it\n",
    "    if c.name:\n",
    "        name_split = c.name.split(\"/\")\n",
    "        \n",
    "        # Check if the name has 'accidentally' become the confidence vals\n",
    "        if len(name_split) > 1:\n",
    "            # Name the internal node\n",
    "            c.name = \"NODE\" + str(node_i)    \n",
    "\n",
    "            # Assign confidence values     \n",
    "            ufboot_val = float(name_split[0])\n",
    "            scf_val = float(name_split[1])\n",
    "    # If it's not a named node, give it a name\n",
    "    else:\n",
    "        c.name = \"NODE\" + str(node_i) \n",
    "    \n",
    "    ufboot_conf = Phylo.PhyloXML.Confidence(ufboot_val, type=\"ufboot\")\n",
    "    scf_conf = Phylo.PhyloXML.Confidence(scf_val, type=\"scf\")\n",
    "    \n",
    "    c.confidences.append(ufboot_conf)\n",
    "    c.confidences.append(scf_conf)\n",
    "        \n",
    "    # Increment node counter\n",
    "    node_i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Convert tree to dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize the tree dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.850300e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>1.327050e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.248300e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>90.2</td>\n",
       "      <td>100</td>\n",
       "      <td>1.363600e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000268965.1_ASM26896v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.692000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269425.1_ASM26942v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.350000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1263</th>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269185.1_ASM26918v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269305.1_ASM26930v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1266 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot  sCF  Branch_Length  \\\n",
       "Name                                                            \n",
       "NODE0                                 100  100   0.000000e+00   \n",
       "SAMEA3541827                           NA   NA   3.850300e-06   \n",
       "NODE2                                 100  100   1.327050e-05   \n",
       "SAMEA104233046                         NA   NA   2.248300e-06   \n",
       "NODE4                                90.2  100   1.363600e-05   \n",
       "...                                   ...  ...            ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic     NA   NA   4.692000e-07   \n",
       "GCA_000269425.1_ASM26942v1_genomic     NA   NA   2.350000e-08   \n",
       "NODE1263                                0   97   2.346000e-07   \n",
       "GCA_000269185.1_ASM26918v1_genomic     NA   NA   2.346000e-07   \n",
       "GCA_000269305.1_ASM26930v1_genomic     NA   NA   2.346000e-07   \n",
       "\n",
       "                                   Branch_Support_Color coord_x coord_y  \n",
       "Name                                                                     \n",
       "NODE0                                                NA      NA      NA  \n",
       "SAMEA3541827                                         NA      NA      NA  \n",
       "NODE2                                                NA      NA      NA  \n",
       "SAMEA104233046                                       NA      NA      NA  \n",
       "NODE4                                                NA      NA      NA  \n",
       "...                                                 ...     ...     ...  \n",
       "GCA_000268965.1_ASM26896v1_genomic                   NA      NA      NA  \n",
       "GCA_000269425.1_ASM26942v1_genomic                   NA      NA      NA  \n",
       "NODE1263                                             NA      NA      NA  \n",
       "GCA_000269185.1_ASM26918v1_genomic                   NA      NA      NA  \n",
       "GCA_000269305.1_ASM26930v1_genomic                   NA      NA      NA  \n",
       "\n",
       "[1266 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# initialize empty dataframe\n",
    "tree_df = pd.DataFrame(columns = [\n",
    "    \"Name\",\n",
    "    \"UFboot\", \n",
    "    \"sCF\", \n",
    "    \"Branch_Length\", \n",
    "    \"Branch_Support_Color\", \n",
    "    \"coord_x\", \n",
    "    \"coord_y\",]\n",
    ")\n",
    "\n",
    "for c in tree.find_clades():\n",
    "    if not c.branch_length: c.branch_length = 0.0\n",
    "    \n",
    "    node_data = {\n",
    "        \"Name\" : c.name,\n",
    "        \"UFboot\" : [conf.value for conf in c.confidences if conf.type==\"ufboot\"][0],\n",
    "        \"sCF\" :  [conf.value for conf in c.confidences if conf.type==\"scf\"][0],\n",
    "        \"Branch_Support_Color\" :  NO_DATA_CHAR,\n",
    "        \"Branch_Length\" :  c.branch_length,\n",
    "        \"coord_x\" :  NO_DATA_CHAR,\n",
    "        \"coord_y\" :  NO_DATA_CHAR,\n",
    "    }\n",
    "    \n",
    "    tree_df = tree_df.append(node_data, ignore_index=True)\n",
    "\n",
    "# Set the index to the node name\n",
    "tree_df.set_index(\"Name\", inplace=True)\n",
    "\n",
    "# Visualize data frame\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add plotting x and y coordinates of each node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>17.8552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.850300e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.8503e-06</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>1.327050e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.32705e-05</td>\n",
       "      <td>3.15625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.248300e-06</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.55188e-05</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>90.2</td>\n",
       "      <td>100</td>\n",
       "      <td>1.363600e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.69065e-05</td>\n",
       "      <td>4.3125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000268965.1_ASM26896v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.692000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.69589e-05</td>\n",
       "      <td>631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269425.1_ASM26942v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.350000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.65132e-05</td>\n",
       "      <td>632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1263</th>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.67008e-05</td>\n",
       "      <td>633.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269185.1_ASM26918v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.69354e-05</td>\n",
       "      <td>633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269305.1_ASM26930v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.69354e-05</td>\n",
       "      <td>634</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1266 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot  sCF  Branch_Length  \\\n",
       "Name                                                            \n",
       "NODE0                                 100  100   0.000000e+00   \n",
       "SAMEA3541827                           NA   NA   3.850300e-06   \n",
       "NODE2                                 100  100   1.327050e-05   \n",
       "SAMEA104233046                         NA   NA   2.248300e-06   \n",
       "NODE4                                90.2  100   1.363600e-05   \n",
       "...                                   ...  ...            ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic     NA   NA   4.692000e-07   \n",
       "GCA_000269425.1_ASM26942v1_genomic     NA   NA   2.350000e-08   \n",
       "NODE1263                                0   97   2.346000e-07   \n",
       "GCA_000269185.1_ASM26918v1_genomic     NA   NA   2.346000e-07   \n",
       "GCA_000269305.1_ASM26930v1_genomic     NA   NA   2.346000e-07   \n",
       "\n",
       "                                   Branch_Support_Color      coord_x  coord_y  \n",
       "Name                                                                           \n",
       "NODE0                                                NA            0  17.8552  \n",
       "SAMEA3541827                                         NA   3.8503e-06        1  \n",
       "NODE2                                                NA  1.32705e-05  3.15625  \n",
       "SAMEA104233046                                       NA  1.55188e-05        2  \n",
       "NODE4                                                NA  2.69065e-05   4.3125  \n",
       "...                                                 ...          ...      ...  \n",
       "GCA_000268965.1_ASM26896v1_genomic                   NA  7.69589e-05      631  \n",
       "GCA_000269425.1_ASM26942v1_genomic                   NA  7.65132e-05      632  \n",
       "NODE1263                                             NA  7.67008e-05    633.5  \n",
       "GCA_000269185.1_ASM26918v1_genomic                   NA  7.69354e-05      633  \n",
       "GCA_000269305.1_ASM26930v1_genomic                   NA  7.69354e-05      634  \n",
       "\n",
       "[1266 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_posns = get_x_positions(tree)\n",
    "y_posns = get_y_positions(tree)\n",
    "\n",
    "# Add x and y coordinates as other attributes\n",
    "for c in tree.find_clades():\n",
    "    # x coordinates will be of branch length units\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    # y coordinates will be arbitrary, based on number of tips\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    # Add data to tree dataframe\n",
    "    tree_df.at[c.name, 'coord_x'] = coord_x\n",
    "    tree_df.at[c.name, 'coord_y'] = coord_y    \n",
    "\n",
    "# Visualize dataframe\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add branch support color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>red</td>\n",
       "      <td>0</td>\n",
       "      <td>17.8552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.850300e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>3.8503e-06</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>1.327050e-05</td>\n",
       "      <td>red</td>\n",
       "      <td>1.32705e-05</td>\n",
       "      <td>3.15625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.248300e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>1.55188e-05</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>90.2</td>\n",
       "      <td>100</td>\n",
       "      <td>1.363600e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>2.69065e-05</td>\n",
       "      <td>4.3125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000268965.1_ASM26896v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.692000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.69589e-05</td>\n",
       "      <td>631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269425.1_ASM26942v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.350000e-08</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.65132e-05</td>\n",
       "      <td>632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1263</th>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>black</td>\n",
       "      <td>7.67008e-05</td>\n",
       "      <td>633.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269185.1_ASM26918v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.69354e-05</td>\n",
       "      <td>633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269305.1_ASM26930v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.69354e-05</td>\n",
       "      <td>634</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1266 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot  sCF  Branch_Length  \\\n",
       "Name                                                            \n",
       "NODE0                                 100  100   0.000000e+00   \n",
       "SAMEA3541827                           NA   NA   3.850300e-06   \n",
       "NODE2                                 100  100   1.327050e-05   \n",
       "SAMEA104233046                         NA   NA   2.248300e-06   \n",
       "NODE4                                90.2  100   1.363600e-05   \n",
       "...                                   ...  ...            ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic     NA   NA   4.692000e-07   \n",
       "GCA_000269425.1_ASM26942v1_genomic     NA   NA   2.350000e-08   \n",
       "NODE1263                                0   97   2.346000e-07   \n",
       "GCA_000269185.1_ASM26918v1_genomic     NA   NA   2.346000e-07   \n",
       "GCA_000269305.1_ASM26930v1_genomic     NA   NA   2.346000e-07   \n",
       "\n",
       "                                   Branch_Support_Color      coord_x  coord_y  \n",
       "Name                                                                           \n",
       "NODE0                                               red            0  17.8552  \n",
       "SAMEA3541827                                       grey   3.8503e-06        1  \n",
       "NODE2                                               red  1.32705e-05  3.15625  \n",
       "SAMEA104233046                                     grey  1.55188e-05        2  \n",
       "NODE4                                             black  2.69065e-05   4.3125  \n",
       "...                                                 ...          ...      ...  \n",
       "GCA_000268965.1_ASM26896v1_genomic                 grey  7.69589e-05      631  \n",
       "GCA_000269425.1_ASM26942v1_genomic                 grey  7.65132e-05      632  \n",
       "NODE1263                                          black  7.67008e-05    633.5  \n",
       "GCA_000269185.1_ASM26918v1_genomic                 grey  7.69354e-05      633  \n",
       "GCA_000269305.1_ASM26930v1_genomic                 grey  7.69354e-05      634  \n",
       "\n",
       "[1266 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for c in tree.find_clades():   \n",
    "    # Default Color\n",
    "    branch_color = LOW_COL\n",
    "    # Terminal branches will be grey\n",
    "    if c.is_terminal():\n",
    "        branch_color = TERM_COL\n",
    "    # High support branches\n",
    "    ufboot_val = tree_df[\"UFboot\"][c.name]\n",
    "    scf_val = tree_df[\"sCF\"][c.name]\n",
    "    if scf_val != NO_DATA_CHAR and ufboot_val != NO_DATA_CHAR:\n",
    "        if scf_val >= SCF_THRESH and ufboot_val >= UFBOOT_THRESH:\n",
    "            branch_color = HIGH_COL\n",
    "    tree_df.at[c.name, \"Branch_Support_Color\"] = branch_color            \n",
    "\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 4. Combine metadata into tree dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse metadata into df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669565.1_ASM966956v1_genomic</th>\n",
       "      <td>42118</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722923</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669605.1_ASM966960v1_genomic</th>\n",
       "      <td>42117</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722922</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP542</th>\n",
       "      <td>IP542</td>\n",
       "      <td>1952</td>\n",
       "      <td>-69</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.44197</td>\n",
       "      <td>38.4314</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>1.ANT1</td>\n",
       "      <td>IP542</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP543</th>\n",
       "      <td>IP543</td>\n",
       "      <td>1953</td>\n",
       "      <td>-68</td>\n",
       "      <td>Democratic Republic of the Congo</td>\n",
       "      <td>NA</td>\n",
       "      <td>-2.98143</td>\n",
       "      <td>23.8223</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>1.ANT</td>\n",
       "      <td>1.ANT1</td>\n",
       "      <td>IP543</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP557</th>\n",
       "      <td>IP557</td>\n",
       "      <td>1963</td>\n",
       "      <td>-58</td>\n",
       "      <td>Iran</td>\n",
       "      <td>Kurdistan</td>\n",
       "      <td>32.6475</td>\n",
       "      <td>54.5644</td>\n",
       "      <td>35.6728</td>\n",
       "      <td>47.0124</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>IP557</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP562</th>\n",
       "      <td>IP562</td>\n",
       "      <td>1947</td>\n",
       "      <td>-74</td>\n",
       "      <td>Iran</td>\n",
       "      <td>Kurdistan</td>\n",
       "      <td>32.6475</td>\n",
       "      <td>54.5644</td>\n",
       "      <td>35.6728</td>\n",
       "      <td>47.0124</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>IP562</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP579</th>\n",
       "      <td>IP579</td>\n",
       "      <td>1908</td>\n",
       "      <td>-113</td>\n",
       "      <td>India</td>\n",
       "      <td>NA</td>\n",
       "      <td>22.3511</td>\n",
       "      <td>78.6677</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>IP579</td>\n",
       "      <td>KEEP: Local Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>633 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Strain  Date DateBP  \\\n",
       "Sample                                                    \n",
       "GCA_009909635.1_ASM990963v1_genomic   9_10  1923    -98   \n",
       "GCA_009669545.1_ASM966954v1_genomic  42126  2006    -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic  42123  2005    -16   \n",
       "GCA_009669565.1_ASM966956v1_genomic  42118  2005    -16   \n",
       "GCA_009669605.1_ASM966960v1_genomic  42117  2005    -16   \n",
       "...                                    ...   ...    ...   \n",
       "IP542                                IP542  1952    -69   \n",
       "IP543                                IP543  1953    -68   \n",
       "IP557                                IP557  1963    -58   \n",
       "IP562                                IP562  1947    -74   \n",
       "IP579                                IP579  1908   -113   \n",
       "\n",
       "                                                              Country  \\\n",
       "Sample                                                                  \n",
       "GCA_009909635.1_ASM990963v1_genomic                            Russia   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             China   \n",
       "GCA_009669565.1_ASM966956v1_genomic                             China   \n",
       "GCA_009669605.1_ASM966960v1_genomic                             China   \n",
       "...                                                               ...   \n",
       "IP542                                                           Kenya   \n",
       "IP543                                Democratic Republic of the Congo   \n",
       "IP557                                                            Iran   \n",
       "IP562                                                            Iran   \n",
       "IP579                                                           India   \n",
       "\n",
       "                                          Province CountryLat CountryLon  \\\n",
       "Sample                                                                     \n",
       "GCA_009909635.1_ASM990963v1_genomic  Rostov Oblast    64.6863    97.7453   \n",
       "GCA_009669545.1_ASM966954v1_genomic       Xinjiang    35.0001        105   \n",
       "GCA_009669555.1_ASM966955v1_genomic       Xinjiang    35.0001        105   \n",
       "GCA_009669565.1_ASM966956v1_genomic       Xinjiang    35.0001        105   \n",
       "GCA_009669605.1_ASM966960v1_genomic       Xinjiang    35.0001        105   \n",
       "...                                            ...        ...        ...   \n",
       "IP542                                           NA    1.44197    38.4314   \n",
       "IP543                                           NA   -2.98143    23.8223   \n",
       "IP557                                    Kurdistan    32.6475    54.5644   \n",
       "IP562                                    Kurdistan    32.6475    54.5644   \n",
       "IP579                                           NA    22.3511    78.6677   \n",
       "\n",
       "                                    ProvinceLat ProvinceLon      Biovar  \\\n",
       "Sample                                                                    \n",
       "GCA_009909635.1_ASM990963v1_genomic     47.6222     40.7958  Medievalis   \n",
       "GCA_009669545.1_ASM966954v1_genomic     42.4805     85.4633     Antiqua   \n",
       "GCA_009669555.1_ASM966955v1_genomic     42.4805     85.4633     Antiqua   \n",
       "GCA_009669565.1_ASM966956v1_genomic     42.4805     85.4633     Antiqua   \n",
       "GCA_009669605.1_ASM966960v1_genomic     42.4805     85.4633     Antiqua   \n",
       "...                                         ...         ...         ...   \n",
       "IP542                                        NA          NA     Antiqua   \n",
       "IP543                                        NA          NA     Antiqua   \n",
       "IP557                                   35.6728     47.0124  Medievalis   \n",
       "IP562                                   35.6728     47.0124  Medievalis   \n",
       "IP579                                        NA          NA  Orientalis   \n",
       "\n",
       "                                    Branch_Major Branch_Minor     BioSample  \\\n",
       "Sample                                                                        \n",
       "GCA_009909635.1_ASM990963v1_genomic        2.MED       2.MED1  SAMN13632815   \n",
       "GCA_009669545.1_ASM966954v1_genomic        0.ANT       0.ANT1  SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic        0.ANT       0.ANT1  SAMN07722924   \n",
       "GCA_009669565.1_ASM966956v1_genomic        0.ANT       0.ANT1  SAMN07722923   \n",
       "GCA_009669605.1_ASM966960v1_genomic        0.ANT       0.ANT1  SAMN07722922   \n",
       "...                                          ...          ...           ...   \n",
       "IP542                                      1.ANT       1.ANT1         IP542   \n",
       "IP543                                      1.ANT       1.ANT1         IP543   \n",
       "IP557                                      2.MED       2.MED1         IP557   \n",
       "IP562                                      2.MED       2.MED1         IP562   \n",
       "IP579                                      1.ORI       1.ORI1         IP579   \n",
       "\n",
       "                                          BioSampleComment  Branch_Number  \n",
       "Sample                                                                     \n",
       "GCA_009909635.1_ASM990963v1_genomic  KEEP: Assembly Modern              2  \n",
       "GCA_009669545.1_ASM966954v1_genomic  KEEP: Assembly Modern              0  \n",
       "GCA_009669555.1_ASM966955v1_genomic  KEEP: Assembly Modern              0  \n",
       "GCA_009669565.1_ASM966956v1_genomic  KEEP: Assembly Modern              0  \n",
       "GCA_009669605.1_ASM966960v1_genomic  KEEP: Assembly Modern              0  \n",
       "...                                                    ...            ...  \n",
       "IP542                                   KEEP: Local Modern              1  \n",
       "IP543                                   KEEP: Local Modern              1  \n",
       "IP557                                   KEEP: Local Modern              2  \n",
       "IP562                                   KEEP: Local Modern              2  \n",
       "IP579                                   KEEP: Local Modern              1  \n",
       "\n",
       "[633 rows x 15 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse the metadata\n",
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "# set the metadata index\n",
    "metadata_df.set_index(NAME_COL, inplace=True)\n",
    "\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add metadata to tree and tree dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>red</td>\n",
       "      <td>0</td>\n",
       "      <td>17.8552</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.850300e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>3.8503e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>RISE509</td>\n",
       "      <td>[-2876:-2677]</td>\n",
       "      <td>[-4897:-4698]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE2</td>\n",
       "      <td>SAMEA3541827</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>1.327050e-05</td>\n",
       "      <td>red</td>\n",
       "      <td>1.32705e-05</td>\n",
       "      <td>3.15625</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.248300e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>1.55188e-05</td>\n",
       "      <td>2</td>\n",
       "      <td>Gyvakarai1</td>\n",
       "      <td>[-2621:-2472]</td>\n",
       "      <td>[-4642:-4493]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>55.35</td>\n",
       "      <td>23.75</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>SAMEA104233046</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE4</th>\n",
       "      <td>90.2</td>\n",
       "      <td>100</td>\n",
       "      <td>1.363600e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>2.69065e-05</td>\n",
       "      <td>4.3125</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000268965.1_ASM26896v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.692000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.69589e-05</td>\n",
       "      <td>631</td>\n",
       "      <td>PY-65</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141486</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269425.1_ASM26942v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.350000e-08</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.65132e-05</td>\n",
       "      <td>632</td>\n",
       "      <td>PY-94</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-6.25</td>\n",
       "      <td>-78.8333</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141497</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1263</th>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>black</td>\n",
       "      <td>7.67008e-05</td>\n",
       "      <td>633.5</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269185.1_ASM26918v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.69354e-05</td>\n",
       "      <td>633</td>\n",
       "      <td>PY-99</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141501</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000269305.1_ASM26930v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.346000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>7.69354e-05</td>\n",
       "      <td>634</td>\n",
       "      <td>PY-46</td>\n",
       "      <td>2010</td>\n",
       "      <td>-11</td>\n",
       "      <td>Peru</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMN02141472</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1266 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UFboot  sCF  Branch_Length  \\\n",
       "Name                                                            \n",
       "NODE0                                 100  100   0.000000e+00   \n",
       "SAMEA3541827                           NA   NA   3.850300e-06   \n",
       "NODE2                                 100  100   1.327050e-05   \n",
       "SAMEA104233046                         NA   NA   2.248300e-06   \n",
       "NODE4                                90.2  100   1.363600e-05   \n",
       "...                                   ...  ...            ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic     NA   NA   4.692000e-07   \n",
       "GCA_000269425.1_ASM26942v1_genomic     NA   NA   2.350000e-08   \n",
       "NODE1263                                0   97   2.346000e-07   \n",
       "GCA_000269185.1_ASM26918v1_genomic     NA   NA   2.346000e-07   \n",
       "GCA_000269305.1_ASM26930v1_genomic     NA   NA   2.346000e-07   \n",
       "\n",
       "                                   Branch_Support_Color      coord_x  coord_y  \\\n",
       "Name                                                                            \n",
       "NODE0                                               red            0  17.8552   \n",
       "SAMEA3541827                                       grey   3.8503e-06        1   \n",
       "NODE2                                               red  1.32705e-05  3.15625   \n",
       "SAMEA104233046                                     grey  1.55188e-05        2   \n",
       "NODE4                                             black  2.69065e-05   4.3125   \n",
       "...                                                 ...          ...      ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic                 grey  7.69589e-05      631   \n",
       "GCA_000269425.1_ASM26942v1_genomic                 grey  7.65132e-05      632   \n",
       "NODE1263                                          black  7.67008e-05    633.5   \n",
       "GCA_000269185.1_ASM26918v1_genomic                 grey  7.69354e-05      633   \n",
       "GCA_000269305.1_ASM26930v1_genomic                 grey  7.69354e-05      634   \n",
       "\n",
       "                                        Strain           Date         DateBP  \\\n",
       "Name                                                                           \n",
       "NODE0                                       NA             NA             NA   \n",
       "SAMEA3541827                           RISE509  [-2876:-2677]  [-4897:-4698]   \n",
       "NODE2                                       NA             NA             NA   \n",
       "SAMEA104233046                      Gyvakarai1  [-2621:-2472]  [-4642:-4493]   \n",
       "NODE4                                       NA             NA             NA   \n",
       "...                                        ...            ...            ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic       PY-65           2010            -11   \n",
       "GCA_000269425.1_ASM26942v1_genomic       PY-94           2010            -11   \n",
       "NODE1263                                    NA             NA             NA   \n",
       "GCA_000269185.1_ASM26918v1_genomic       PY-99           2010            -11   \n",
       "GCA_000269305.1_ASM26930v1_genomic       PY-46           2010            -11   \n",
       "\n",
       "                                      Country          Province CountryLat  \\\n",
       "Name                                                                         \n",
       "NODE0                                      NA                NA         NA   \n",
       "SAMEA3541827                           Russia  Krasnoyarsk Krai    64.6863   \n",
       "NODE2                                      NA                NA         NA   \n",
       "SAMEA104233046                      Lithuania  Panevezys County      55.35   \n",
       "NODE4                                      NA                NA         NA   \n",
       "...                                       ...               ...        ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic       Peru       La Libertad   -6.86997   \n",
       "GCA_000269425.1_ASM26942v1_genomic       Peru         Cajamarca   -6.86997   \n",
       "NODE1263                                   NA                NA         NA   \n",
       "GCA_000269185.1_ASM26918v1_genomic       Peru       La Libertad   -6.86997   \n",
       "GCA_000269305.1_ASM26930v1_genomic       Peru       La Libertad   -6.86997   \n",
       "\n",
       "                                   CountryLon ProvinceLat ProvinceLon  \\\n",
       "Name                                                                    \n",
       "NODE0                                      NA          NA          NA   \n",
       "SAMEA3541827                          97.7453     63.3234      97.098   \n",
       "NODE2                                      NA          NA          NA   \n",
       "SAMEA104233046                          23.75     55.9156     25.0312   \n",
       "NODE4                                      NA          NA          NA   \n",
       "...                                       ...         ...         ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic   -75.0459          -8       -78.5   \n",
       "GCA_000269425.1_ASM26942v1_genomic   -75.0459       -6.25    -78.8333   \n",
       "NODE1263                                   NA          NA          NA   \n",
       "GCA_000269185.1_ASM26918v1_genomic   -75.0459          -8       -78.5   \n",
       "GCA_000269305.1_ASM26930v1_genomic   -75.0459          -8       -78.5   \n",
       "\n",
       "                                        Biovar Branch_Major Branch_Minor  \\\n",
       "Name                                                                       \n",
       "NODE0                                       NA           NA           NA   \n",
       "SAMEA3541827                        Bronze Age        0.PRE       0.PRE2   \n",
       "NODE2                                       NA           NA           NA   \n",
       "SAMEA104233046                      Bronze Age        0.PRE       0.PRE1   \n",
       "NODE4                                       NA           NA           NA   \n",
       "...                                        ...          ...          ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "GCA_000269425.1_ASM26942v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "NODE1263                                    NA           NA           NA   \n",
       "GCA_000269185.1_ASM26918v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "GCA_000269305.1_ASM26930v1_genomic  Orientalis        1.ORI       1.ORI1   \n",
       "\n",
       "                                         BioSample       BioSampleComment  \\\n",
       "Name                                                                        \n",
       "NODE0                                           NA                     NA   \n",
       "SAMEA3541827                          SAMEA3541827      KEEP: SRA Ancient   \n",
       "NODE2                                           NA                     NA   \n",
       "SAMEA104233046                      SAMEA104233046      KEEP: SRA Ancient   \n",
       "NODE4                                           NA                     NA   \n",
       "...                                            ...                    ...   \n",
       "GCA_000268965.1_ASM26896v1_genomic    SAMN02141486  KEEP: Assembly Modern   \n",
       "GCA_000269425.1_ASM26942v1_genomic    SAMN02141497  KEEP: Assembly Modern   \n",
       "NODE1263                                        NA                     NA   \n",
       "GCA_000269185.1_ASM26918v1_genomic    SAMN02141501  KEEP: Assembly Modern   \n",
       "GCA_000269305.1_ASM26930v1_genomic    SAMN02141472  KEEP: Assembly Modern   \n",
       "\n",
       "                                   Branch_Number  \n",
       "Name                                              \n",
       "NODE0                                         NA  \n",
       "SAMEA3541827                                   0  \n",
       "NODE2                                         NA  \n",
       "SAMEA104233046                                 0  \n",
       "NODE4                                         NA  \n",
       "...                                          ...  \n",
       "GCA_000268965.1_ASM26896v1_genomic             1  \n",
       "GCA_000269425.1_ASM26942v1_genomic             1  \n",
       "NODE1263                                      NA  \n",
       "GCA_000269185.1_ASM26918v1_genomic             1  \n",
       "GCA_000269305.1_ASM26930v1_genomic             1  \n",
       "\n",
       "[1266 rows x 21 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Iterate through the different metadata attributes\n",
    "for attr in metadata_df.columns:\n",
    "    # Initialize an empty column for the attribute\n",
    "    tree_df[attr] = [NO_DATA_CHAR for row in range(0,len(tree_df))]\n",
    "    # Iterate over the nodes in the tree\n",
    "    for c in tree.find_clades():\n",
    "        # Initialize to empty\n",
    "        attr_val = NO_DATA_CHAR\n",
    "        # If the node shows up in the metadata\n",
    "        if c.name in metadata_df.index:\n",
    "            tree_df.at[c.name, attr] = metadata_df[attr][c.name]\n",
    "            attr_val = metadata_df[attr][c.name]         \n",
    "\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Export"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save tree and tree dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_path_df = os.path.join(outdir, SCRIPT_NAME + \".tsv\" )\n",
    "tree_df.to_csv(out_path_df, sep=\"\\t\")\n",
    "\n",
    "out_path_xml = os.path.join(outdir,  SCRIPT_NAME + \".xml\" )\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \".nwk\" )\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \".nexus\" )\n",
    "Phylo.write(tree, out_path_xml, 'phyloxml')\n",
    "Phylo.write(tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "Phylo.write(tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save lat and lon\n",
    "\n",
    "Write an output tsv that is:  \n",
    "resolution   name   lat    lon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "latlon_dict = {\"Country\": {}, \"Province\": {}}\n",
    "out_path_lat_lon = os.path.join(outdir, SCRIPT_NAME + \"_latlon.tsv\")\n",
    "\n",
    "# Columns in data frame will be ex: CountryLat, CountryLon\n",
    "for c in tree.find_clades():\n",
    "    # Iterate through each geographic level\n",
    "    for level in latlon_dict:\n",
    "        # Store the location name\n",
    "        loc = tree_df[level][c.name]\n",
    "        # If the node has location data and is new\n",
    "        if loc != NO_DATA_CHAR and loc not in latlon_dict[level]:\n",
    "            latlon_dict[level][loc] = {}\n",
    "            latlon_dict[level][loc][\"lat\"] = tree_df[level + \"Lat\"][c.name]\n",
    "            latlon_dict[level][loc][\"lon\"] = tree_df[level + \"Lon\"][c.name]\n",
    "\n",
    "with open(out_path_lat_lon, \"w\") as outfile:\n",
    "    for level in latlon_dict:\n",
    "        for loc in latlon_dict[level]:\n",
    "            outfile.write(level.lower() + \"\\t\"\n",
    "                         + loc + \"\\t\"\n",
    "                         + str(latlon_dict[level][loc][\"lat\"]) + \"\\t\"\n",
    "                         + str(latlon_dict[level][loc][\"lon\"]) + \"\\n\"\n",
    "                 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save colors\n",
    "\n",
    "Write an output tsv that is:\n",
    "attribute   attribute_val   hex_color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'branch_number': {0: '#8000ff', 3: '#00b4ec', 4: '#80ffb4', 2: '#ffb462', 1: '#ff0000'}, 'branch_major': {'0.PRE': '#8000ff', '0.PE': '#4c4ffc', '0.ANT': '#1996f3', '3.ANT': '#1acee3', '4.ANT': '#4df3ce', '2.ANT': '#80ffb4', '2.MED': '#b3f396', '1.PRE': '#e6ce74', '1.ANT': '#ff964f', '1.IN': '#ff4f28', '1.ORI': '#ff0000'}, 'branch_minor': {'0.PRE2': '#8000ff', '0.PRE1': '#7116ff', '0.PE7': '#632cfe', '0.PE2': '#5542fd', '0.PE8': '#4757fb', '0.PE10': '#396cf9', '0.PE4m': '#2b7ff6', '0.PE4h': '#1c92f3', '0.PE4t': '#0ea4f0', '0.PE4a': '#00b4ec', '0.PE5': '#0ec3e7', '0.ANT1': '#1cd1e2', '0.ANT4': '#2adddd', '0.ANT2': '#39e7d7', '0.ANT5': '#47f0d1', '0.ANT3': '#55f6ca', '3.ANT1': '#63fbc3', '3.ANT2': '#71febc', '4.ANT1': '#80ffb4', '2.ANT3': '#8efeac', '2.ANT2': '#9cfba4', '2.ANT1': '#aaf69b', '2.MED0': '#b8f092', '2.MED3': '#c6e789', '2.MED2': '#d4dd80', '2.MED1': '#e3d176', '1.PRE0': '#f1c36c', '1.PRE1': '#ffb462', '1.PRE2': '#ffa457', '1.PRE3': '#ff924d', '1.ANT1': '#ff8042', '1.IN1': '#ff6c37', '1.IN2': '#ff572c', '1.IN3': '#ff4221', '1.ORI1': '#ff2c16', '1.ORI3': '#ff160b', '1.ORI2': '#ff0000'}, 'country': {'Russia': '#8000ff', 'Lithuania': '#7215ff', 'Estonia': '#652afe', 'Germany': '#573ffd', 'China': '#4a53fc', 'Azerbaijan': '#3c66fa', 'Armenia': '#2f79f7', 'Georgia': '#228bf4', 'Mongolia': '#149df1', 'Tajikistan': '#07adee', 'Kyrgyzstan': '#07bcea', 'England': '#14c9e5', 'Spain': '#22d5e0', 'France': '#2fe0db', 'Nepal': '#3cead5', 'India': '#4af1d0', 'Kazakhstan': '#57f7c9', 'Turkmenistan': '#65fcc3', 'Uzbekistan': '#72febc', 'Iran': '#80ffb4', 'United States of America': '#8dfead', 'Norway': '#9afca5', 'Italy': '#a8f79d', 'Poland': '#b5f194', 'Switzerland': '#c3ea8b', 'The Netherlands': '#d0e083', 'Uganda': '#ddd579', 'Democratic Republic of the Congo': '#ebc970', 'Kenya': '#f8bc66', 'Indonesia': '#ffad5d', 'Madagascar': '#ff9d53', 'Vietnam': '#ff8b49', 'Myanmar': '#ff793f', 'Peru': '#ff6634', 'Canada': '#ff532a', 'Algeria': '#ff3f20', 'Brazil': '#ff2a15', 'Zimbabwe': '#ff150b', 'Bolivia': '#ff0000'}, 'province': {'Krasnoyarsk Krai': '#8000ff', 'Panevezys County': '#7b07ff', 'Pärnu maakond': '#760eff', 'Irkutsk Oblast': '#7215ff', 'Bavaria': '#6d1dff', 'Altai Krai': '#6924fe', 'Qinghai': '#642bfe', 'Republic of Dagestan': '#6032fe', 'Goranboy District': '#5b39fd', 'Syunik Province': '#5740fd', 'Shirak Province': '#5247fc', 'Gegharkunik Province': '#4d4dfc', 'Samtskhe-Javakheti': '#4954fb', 'Shahbuz Rayon': '#445bfb', 'Samara Oblast': '#4062fa', 'Sečuán': '#3b68f9', 'Inner Mongolia': '#376ff9', 'Bayankhongor': '#3275f8', 'Sughd Province': '#2e7bf7', 'Talas Region': '#2982f6', 'Altai Republic': '#2488f5', 'Govi-Altai': '#208ef4', 'Övörkhangai': '#1b94f3', 'Ömnögovi': '#1799f2', 'Omnogovi': '#129ff1', 'Bayan-Ölgii': '#0ea5ef', 'Khovd': '#09aaee', 'Xinjiang': '#05afed', 'Issyk-Kul Region': '#00b4ec', 'East of England': '#05b9ea', 'Valencia Community': '#09bee9', 'Centre-Loire Valley': '#0ec3e7', 'Osh Region': '#12c7e6', 'Naryn Region': '#17cce4', 'Gansu': '#1bd0e3', 'Dornogovi': '#20d4e1', 'Tuva Republic': '#24d8df', 'Jilin': '#29dcdd', 'St. Petersberg': '#2edfdc', 'Heilongjiang': '#32e3da', 'Zabaykalsky Krai': '#37e6d8', 'Tibet': '#3be9d6', 'Gandaki Pradesh': '#40ecd4', 'Maharashtra': '#44eed2', 'Himachal Pradesh': '#49f1d0', 'Karachay-Cherkessia': '#4df3ce', 'Kabardino-Balkaria': '#52f5cc', 'Ningxia': '#57f7ca', 'Shaanxi': '#5bf9c7', 'Hebei': '#60fac5', ' Dornogovi': '#64fbc3', 'West Kazakhstan Region': '#69fcc1', 'Nagorno-Karabakh Republic': '#6dfdbe', 'Kyzylorda Region': '#72febc', 'Rostov Oblast': '#76ffb9', 'Astrakhan Oblast': '#7bffb7', 'Balkan': '#80ffb4', 'Kyzylorda Province': '#84ffb2', 'Shymkent': '#89ffaf', 'Navoiy Region': '#8dfead', 'Almaty Region': '#92fdaa', ' West Kazakhstan region': '#96fca7', 'Atyrau Region': '#9bfba5', 'Kurdistan': '#9ffaa2', 'Fizuli District': '#a4f99f', 'Baku': '#a8f79c', 'Illinois': '#adf599', 'Shamkir District': '#b2f396', 'Stavropol Krai': '#b6f194', 'Ingushetia': '#bbee91', 'Mangystau Province': '#bfec8e', 'Chechnya': '#c4e98b', 'Aktobe Region': '#c8e688', 'Atyrau': '#cde385', 'Qazakh District': '#d1df82', 'Republic of Kalmykia': '#d6dc7e', 'Tatarstan': '#dbd87b', 'Oslo': '#dfd478', 'Occitanie': '#e4d075', 'Greater London': '#e8cc72', 'Catalonia': '#edc76f', 'Lazio': '#f1c36b', 'Baden-Württemberg': '#f6be68', \"Provence-Alpes-Côte d'Azur\": '#fab965', 'Pomeranian Voivodeship': '#ffb462', 'Brandenburg': '#ffaf5e', 'Nidwalden': '#ffaa5b', 'North Brabant': '#ffa558', 'Nairobi': '#ff9f54', 'Yunnan': '#ff9951', 'Fujian': '#ff944d', 'Central Java': '#ff8e4a', 'Province de Fianarantsoa': '#ff8847', 'Gujarat': '#ff8243', 'Kirov Oblast': '#ff7b40', 'Lanzou': '#ff753c', 'Khánh Hoà Province': '#ff6f39', 'Lâm Đồng Province': '#ff6835', ' Ho Chi Minh City': '#ff6232', 'Guangxi Zhuang': '#ff5b2e', 'California': '#ff542b', 'La Libertad': '#ff4d27', 'Tamil Nadu': '#ff4724', 'Arizona': '#ff4020', 'New Mexico': '#ff391d', 'Saskatchewan': '#ff3219', 'Texas': '#ff2b15', 'Nevada': '#ff2412', 'MSila': '#ff1d0e', 'State of Mato Grosso': '#ff150b', 'Midlands': '#ff0e07', 'La Paz': '#ff0704', 'Cajamarca': '#ff0000'}}\n"
     ]
    }
   ],
   "source": [
    "out_path_colors = os.path.join(outdir, SCRIPT_NAME + \"_colors.tsv\")\n",
    "file_colors = open(out_path_colors, \"w\")\n",
    "\n",
    "tree_df[\"dummy_conf\"] = [0 for row in range(0,len(tree_df))]\n",
    "\n",
    "hex_dict = {}\n",
    "\n",
    "for attr in ATTRIBUTE_LIST:\n",
    "    # Create the color map\n",
    "    attr_key = attr.lower()\n",
    "    hex_dict[attr_key] = {}\n",
    "    for t in tree.get_terminals():\n",
    "        attr_val = tree_df[attr][t.name]\n",
    "        if attr_val not in hex_dict[attr_key] and attr_val != NO_DATA_CHAR:\n",
    "            hex_dict[attr_key][attr_val] = {}\n",
    "            \n",
    "    # Create the custom color map (pyplot)\n",
    "    cmap = plt.get_cmap(CONT_COLOR_PAL, len(hex_dict[attr_key]))\n",
    "    # Convert the color map to a list of RGB values\n",
    "    cmaplist = [cmap(i) for i in range(cmap.N)]\n",
    "    # Convert RGB values to hex colors\n",
    "    attr_hex = [colors.to_hex(col) for col in cmaplist]\n",
    "    \n",
    "    # Assign colors to value\n",
    "    for attr_val, attr_col in zip(hex_dict[attr_key], attr_hex):\n",
    "        hex_dict[attr_key][attr_val] = attr_col   \n",
    "\n",
    "        \n",
    "print(hex_dict)\n",
    "\n",
    "for attr_key in hex_dict:\n",
    "    for attr_val in hex_dict[attr_key]:\n",
    "        file_colors.write(str(attr_key) + \"\\t\" + str(attr_val) + \"\\t\" + str(hex_dict[attr_key][attr_val]) + \"\\n\")\n",
    "\n",
    "file_colors.close()        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### JSON\n",
    "\n",
    "This can then be used for auspice via:\n",
    "\n",
    "```\n",
    "augur export v2 \\\n",
    "  --auspice-config auspice.config \\\n",
    "  --tree parse_tree.nwk \\\n",
    "  --node-data parse_tree.json \\\n",
    "  --output parse_tree_auspice.json \\\n",
    "  --lat-longs parse_tree_latlon.tsv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exclude these internal variables for plotting\n",
    "COLOR_KEYWORD_EXCLUDE = [\"color\", \"coord\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# There is no alignment used\n",
    "node_data = {\"alignment\" : \"\"}\n",
    "node_data[\"input_tree\"] = tree_path\n",
    "\n",
    "node_dict = {}\n",
    "\n",
    "# Iterate through all nodes in the tree\n",
    "for c in tree.find_clades():\n",
    "    # Add the node to the dictionary\n",
    "    node_dict[c.name] = {}\n",
    "    # Iterate through all attributes in the dataframe\n",
    "    for attr in tree_df.columns:\n",
    "        # Check if this attribute should be excluded\n",
    "        exclude = False\n",
    "        for keyword in COLOR_KEYWORD_EXCLUDE:\n",
    "            if keyword in attr.lower():\n",
    "                exclude = True    \n",
    "        if not exclude:            \n",
    "            # Make attribute name in dict lowercase (ex. Branch_Length -> branch_length)\n",
    "            node_dict[c.name][attr.lower()] = tree_df[attr][c.name]\n",
    "\n",
    "        \n",
    "# Add the dataframe information to the node dict for the json\n",
    "node_data[\"nodes\"] = node_dict\n",
    "\n",
    "out_path_json = os.path.join(outdir, SCRIPT_NAME + \".json\" )\n",
    "augur.utils.write_json(data=node_data, file_name=out_path_json, indent=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.remove(tmp_xml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
