{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "resident-export",
   "metadata": {},
   "source": [
    "# Paper Figures\n",
    "\n",
    "1. Setup\n",
    "1. Import\n",
    "  - Metadata\n",
    "  - Trees\n",
    "    - Divergence tree\n",
    "    - Timetree\n",
    "    - Mugration Trees\n",
    "1. Merge Metadata\n",
    "1. Export\n",
    "  - Timetree\n",
    "  - Divtree\n",
    "  - Augur and Auspice JSON"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "careful-ordering",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bronze-wealth",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "proof-ultimate",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import Phylo, AlignIO\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors, gridspec, lines\n",
    "import scipy\n",
    "import math\n",
    "import numpy as np\n",
    "import geopandas\n",
    "import datetime\n",
    "import shapely\n",
    "import copy\n",
    "from geopy import distance\n",
    "import geopy\n",
    "import cartopy.crs as ccrs\n",
    "from augur import utils, export_v2\n",
    "\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "violent-vampire",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "noted-savannah",
   "metadata": {},
   "outputs": [],
   "source": [
    "#AUSPICE_PREFIX=\"plague-phylogeography-projects_main\"\n",
    "\n",
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"full\", \"5\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "PRUNE = WILDCARDS[2]\n",
    "MISSING_DATA = WILDCARDS[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "proud-territory",
   "metadata": {},
   "outputs": [],
   "source": [
    "iqtree_dir    = os.path.join(project_dir, \"iqtree/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "lsd_dir       = os.path.join(project_dir, \"lsd/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "mugration_dir = os.path.join(project_dir, \"mugration/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "snippy_dir    = os.path.join(project_dir, \"snippy_multi/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "\n",
    "# ------------------------------------------\n",
    "# Trees\n",
    "divtree_path              = iqtree_dir    + \"filter-taxa/iqtree.treefile\"\n",
    "timetree_path             = lsd_dir       + \"lsd.timetree.nex\"\n",
    "#timetree_path             = lsd_dir       + \"lsd.nex\"\n",
    "mugtree_country_path      = mugration_dir + \"country.nex\".format(MISSING_DATA)\n",
    "mugtree_province_path     = mugration_dir + \"province.nex\".format(MISSING_DATA)\n",
    "mugtree_branch_major_path = mugration_dir + \"branch_major.nex\".format(MISSING_DATA)\n",
    "mugtree_branch_minor_path = mugration_dir + \"branch_minor.nex\".format(MISSING_DATA)\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "constant_sites_path   = project_dir + \"snippy_multi/all/chromosome/full/snippy-multi.constant_sites.txt\"\n",
    "aln_path              = iqtree_dir + \"filter-sites/snippy-multi.snps.aln\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Metadata\n",
    "auspice_config_path          = project_dir + \"config/auspice_config.json\"\n",
    "metadata_path                = iqtree_dir    + \"filter-taxa/metadata.tsv\"\n",
    "mugconf_country_path         = mugration_dir + \"country_confidence.csv\"\n",
    "mugstates_country_path       = mugration_dir + \"country_states.csv\"\n",
    "mugconf_province_path        = mugration_dir + \"province_confidence.csv\"\n",
    "mugstates_province_path      = mugration_dir + \"province_states.csv\"\n",
    "mugconf_branch_major_path    = mugration_dir + \"branch_major_confidence.csv\"\n",
    "mugstates_branch_major_path  = mugration_dir + \"branch_major_states.csv\"\n",
    "mugconf_branch_minor_path    = mugration_dir + \"branch_minor_confidence.csv\"\n",
    "mugstates_branch_minor_path  = mugration_dir + \"branch_minor_states.csv\"\n",
    "# ------------------------------------------\n",
    "# Other\n",
    "xu_2019_path          = project_dir + \"figures/xu2019.csv\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Output\n",
    "out_dir = os.path.join(project_dir, \"auspice/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "collaborative-rescue",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "exact-blood",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", None)\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "UNKNOWN_CHAR = \"?\"\n",
    "JSON_INDENT=2\n",
    "CURRENT_YEAR = datetime.datetime.utcnow().year\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "with open(constant_sites_path) as infile:\n",
    "    data = infile.read().strip().split(\",\")\n",
    "    constant_sites = sum([int(count) for count in data])\n",
    "\n",
    "aln = AlignIO.read(aln_path, \"fasta\")\n",
    "variant_sites = len(aln[0].seq)\n",
    "SEQ_LEN = constant_sites + variant_sites\n",
    "\n",
    "# ------------------------------------------\n",
    "# Geo\n",
    "CRS = \"epsg:3857\"\n",
    "CRS_EPSG = ccrs.epsg('3857')\n",
    "world_polygons = geopandas.read_file(geopandas.datasets.get_path(\"naturalearth_lowres\"))\n",
    "\n",
    "# ------------------------------------------\n",
    "# Reference\n",
    "REF_STRAIN = \"CO92\"\n",
    "REF_LAT = \"38.7251776\"\n",
    "REF_LON = \"-105.607716\"\n",
    "REF_DATE_BP = 1992 - CURRENT_YEAR\n",
    "REF_BRANCH_MAJOR = \"1.ORI\"\n",
    "REF_BRANCH_MINOR = \"1.ORI1\"\n",
    "REF_BRANCH_NUMBER = \"1\"\n",
    "REF_COUNTRY = \"United States of America\"\n",
    "REF_PROVINCE = \"Colorado\"\n",
    "REF_CONTINENT = \"North America\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Plotting\n",
    "SM_FONT = 4\n",
    "MED_FONT = 6\n",
    "LG_FONT = 8\n",
    "D3_PAL = [\"#1f77b4\", \"#ff7f0e\", \"#2ca02c\", \"#d62728\", \"#9467bd\", \"#8c564b\", \"#e377c2\", \"#7f7f7f\", \"#bcbd22\", \"#17becf\" ]\n",
    "\n",
    "# ------------------------------------------\n",
    "BRANCH_LIST = {\n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"], \n",
    "    \"0.ANT4\" : [\"0.ANT4\"], \n",
    "    \"0.PE\": [\"0.PE4m\", \"0.PE4m\", \"0.PE4t\", \"0.PE4a\", \"0.PE5\"],   \n",
    "    \"0.ANT\": [\"0.ANT1\", \"0.ANT2\",\"0.ANT3\",\"0.ANT5\"],     \n",
    "    \"1.PRE\" : [\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],    \n",
    "    \"1.ANT\": [\"1.ANT1\"], \n",
    "    \"1.IN\": [\"1.IN1\",\"1.IN2\",\"1.IN3\"],   \n",
    "    \"1.ORI\" : [\"1.ORI1\", \"1.ORI2\", \"1.ORI3\"],\n",
    "    \"2.ANT\": [\"2.ANT1\",\"2.ANT2\",\"2.ANT3\" ], \n",
    "    \"2.MED\": [\"2.MED0\", \"2.MED1\",\"2.MED2\",\"2.MED3\" ], \n",
    "    \"3.ANT\": [\"3.ANT1\", \"3.ANT2\" ],     \n",
    "    \"4.ANT\": [\"4.ANT1\" ],           \n",
    "}\n",
    "\n",
    "ANCIENT_BRANCH_LIST = {\n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"], \n",
    "    \"0.ANT4\" : [\"0.ANT4\"],  \n",
    "    \"1.PRE\" : [\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],     \n",
    "}\n",
    "\n",
    "MUG_ATTRIBUTE_LIST = [\n",
    "    \"branch_major\",\n",
    "    \"branch_minor\",\n",
    "    \"country\",\n",
    "    \"province\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nonprofit-cedar",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. Import"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "processed-mitchell",
   "metadata": {},
   "source": [
    "## Mugration Confidence Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "operational-investigation",
   "metadata": {},
   "outputs": [],
   "source": [
    "states_files = [mugstates_country_path, mugstates_province_path, mugstates_branch_major_path, mugstates_branch_minor_path]\n",
    "conf_files = [mugconf_country_path, mugconf_province_path, mugconf_branch_major_path, mugconf_branch_minor_path]\n",
    "\n",
    "# ----------------------------------------\n",
    "# Convert character states to numbers\n",
    "for mugstates,mugconf in zip(states_files, conf_files):\n",
    "    node_i = 0\n",
    "    conf_basename = os.path.splitext(mugconf)[0]\n",
    "    conf_outname = conf_basename + \"_tidy.csv\"\n",
    "    i = 0\n",
    "    state_dict = {}\n",
    "    with open(mugstates, \"r\") as infile:\n",
    "        read_line = infile.readline()\n",
    "        while read_line:\n",
    "            split_line = read_line.strip().split(\",\")\n",
    "            state = split_line[1]\n",
    "            if state == \"nan\":\n",
    "                state = \"NA\"\n",
    "            if state == UNKNOWN_CHAR:\n",
    "                state_dict[-1] = state\n",
    "            else:\n",
    "                state_dict[i] = state\n",
    "                i += 1\n",
    "            read_line = infile.readline()\n",
    "    # Swap first to last place\n",
    "    final_i = len(state_dict) - 1   \n",
    "    state_dict[final_i] = \"?\"\n",
    "    # Remove old first\n",
    "    state_dict.pop(-1)\n",
    "\n",
    "    with open(mugconf, \"r\") as infile:            \n",
    "        i = 0\n",
    "        read_line = infile.readline()\n",
    "        header = read_line.strip().split(\", \")\n",
    "        header[0] = \"name\"\n",
    "\n",
    "        data = \"\"\n",
    "        for i in range(0, len(header) - 1):\n",
    "            char = header[i + 1]\n",
    "            header[i + 1] = state_dict[i]\n",
    "        header = \",\".join(header)\n",
    "        read_line = infile.readline()            \n",
    "        while read_line:\n",
    "            # Rename internal nodes\n",
    "            split_line = read_line.split(\", \")\n",
    "            node_name = split_line[0]\n",
    "            if \"NODE\" in node_name:\n",
    "                node_name = \"NODE{}\".format(node_i)\n",
    "                node_i += 1\n",
    "            new_line = \",\".join([node_name] + split_line[1:])\n",
    "            data += new_line\n",
    "            read_line = infile.readline()\n",
    "    \n",
    "    with open(conf_outname, \"w\") as outfile:\n",
    "        outfile.write(header + \"\\n\")\n",
    "        outfile.write(data)\n",
    "\n",
    "# ----------------------------------------\n",
    "# Import the tidy files as dataframes\n",
    "mugconf_country_tidy_path = os.path.splitext(mugconf_country_path)[0] + \"_tidy.csv\"\n",
    "mugconf_country_df = pd.read_csv(mugconf_country_tidy_path, sep=',', index_col = 0)\n",
    "\n",
    "mugconf_province_tidy_path = os.path.splitext(mugconf_province_path)[0] + \"_tidy.csv\"\n",
    "mugconf_province_df = pd.read_csv(mugconf_province_tidy_path, sep=',', index_col = 0)\n",
    "\n",
    "mugconf_branch_major_tidy_path = os.path.splitext(mugconf_branch_major_path)[0] + \"_tidy.csv\"\n",
    "mugconf_branch_major_df = pd.read_csv(mugconf_branch_major_tidy_path, sep=',', index_col = 0)\n",
    "\n",
    "mugconf_branch_minor_tidy_path = os.path.splitext(mugconf_branch_minor_path)[0] + \"_tidy.csv\"\n",
    "mugconf_branch_minor_df = pd.read_csv(mugconf_branch_minor_tidy_path, sep=',', index_col = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bottom-thanks",
   "metadata": {},
   "source": [
    "## Trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "loving-rochester",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_paths = [divtree_path, timetree_path, mugtree_country_path, mugtree_province_path, mugtree_branch_major_path, mugtree_branch_minor_path  ]\n",
    "\n",
    "for tree_path in tree_paths:\n",
    "    tree_ext = os.path.splitext(tree_path)[1]\n",
    "    if tree_ext == \".nwk\" or tree_ext == \".newick\" or tree_ext == \".treefile\":\n",
    "        tree = Phylo.read(tree_path, \"newick\")\n",
    "    elif tree_ext == \".nex\" or tree_ext == \".nexus\":\n",
    "        parse_trees = Phylo.parse(tree_path, \"nexus\")\n",
    "        for parse_tree in parse_trees:\n",
    "            clades = [c for c in parse_tree.find_clades()]\n",
    "            if len(clades) > 1:\n",
    "                break\n",
    "        tree = parse_tree\n",
    "    tree.ladderize(reverse=False)\n",
    "    \n",
    "    if tree_path == divtree_path:\n",
    "        print(\"divtree:\", os.path.basename(divtree_path))\n",
    "        divtree = tree\n",
    "    elif tree_path == timetree_path:\n",
    "        print(\"timetree:\", os.path.basename(timetree_path))        \n",
    "        timetree = tree    \n",
    "    elif tree_path == mugtree_country_path:\n",
    "        print(\"mugtree_country:\", os.path.basename(mugtree_country_path))       \n",
    "        mugtree_country = tree    \n",
    "    elif tree_path == mugtree_province_path:\n",
    "        print(\"mugtree_province:\",os.path.basename(mugtree_province_path))\n",
    "        mugtree_province = tree  \n",
    "    elif tree_path == mugtree_branch_major_path:\n",
    "        print(\"mugtree_branch_major:\",os.path.basename(mugtree_branch_major_path))        \n",
    "        mugtree_branch_major = tree            \n",
    "    elif tree_path == mugtree_branch_minor_path:\n",
    "        print(\"mugtree_branch_minor:\",os.path.basename(mugtree_branch_minor_path))        \n",
    "        mugtree_branch_minor = tree                \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "falling-character",
   "metadata": {},
   "source": [
    "### Modern Human Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "leading-collector",
   "metadata": {},
   "outputs": [],
   "source": [
    "#xu_2019_df = pd.read_csv(xu_2019_path, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "emerging-waters",
   "metadata": {},
   "source": [
    "### Project Genomic Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "wooden-patio",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.set_index(metadata_df.columns[0], inplace=True)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "behind-spice",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Merge Metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "intense-election",
   "metadata": {},
   "source": [
    "### Tree Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "unknown-recipe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if they're synchronized\n",
    "node_i = 0\n",
    "\n",
    "metadata_df[\"branch_support_char\"] = [\"\"] * len(metadata_df)\n",
    "\n",
    "for cd,ct,cmc,cmp,cmbmaj, cmbmin in zip(\n",
    "        divtree.find_clades(), \n",
    "        timetree.find_clades(), \n",
    "        mugtree_country.find_clades(),\n",
    "        mugtree_province.find_clades(),\n",
    "        mugtree_branch_major.find_clades(),    \n",
    "        mugtree_branch_minor.find_clades(),      \n",
    "    ):\n",
    "    # Rename internal nodes\n",
    "    if not cd.name or \"NODE\" in cmc.name:\n",
    "        cd.name = \"NODE{}\".format(node_i)\n",
    "        ct.name = \"NODE{}\".format(node_i)\n",
    "        cmc.name = \"NODE{}\".format(node_i) \n",
    "        cmp.name = \"NODE{}\".format(node_i) \n",
    "        cmbmaj.name = \"NODE{}\".format(node_i) \n",
    "        cmbmin.name = \"NODE{}\".format(node_i) \n",
    "        node_i += 1\n",
    "        \n",
    "    if cd.name != ct.name or cd.name != cmc.name or cd.name != cmp.name:\n",
    "        print(\"ERROR. Trees are not synchronized.\")\n",
    "        print(cd.name, ct.name, cmc.name, cmp.name, cmbmaj.name, cmbmin.name)\n",
    "        break\n",
    "        \n",
    "    node_name = cd.name\n",
    "    \n",
    "    # node type\n",
    "    node_type = \"internal\"\n",
    "    if cd.is_terminal():\n",
    "        node_type = \"terminal\"\n",
    "    metadata_df.at[node_name, \"node_type\"] = node_type\n",
    "\n",
    "    \n",
    "    # Branch Length\n",
    "    branch_length = cd.branch_length\n",
    "    if not branch_length:\n",
    "        branch_length = 0\n",
    "    metadata_df.at[node_name, \"branch_length\"] = branch_length\n",
    "    # Branch Support\n",
    "    # Give terminal nodes 100% confidence\n",
    "    branch_support = cd.confidence\n",
    "    if cd.is_terminal():\n",
    "        branch_support = 100\n",
    "        \n",
    "    # Internal nodes without confidence can be 0\n",
    "    elif not cd.confidence:\n",
    "        branch_support = 0\n",
    "    metadata_df.at[node_name, \"branch_support\"] = branch_support\n",
    "    \n",
    "    if branch_support >= 95 and not cd.is_terminal():\n",
    "        branch_support_char = \"*\"\n",
    "        metadata_df.at[node_name, \"branch_support_char\"] = branch_support_char\n",
    "    \n",
    "    comment_dict = {}\n",
    "    # Mugration  Country\n",
    "    if cmc.comment:\n",
    "        cmc_comment_dict = parse_comment(cmc.comment)\n",
    "        for attr in cmc_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmc_comment_dict[attr]\n",
    "            \n",
    "    # Mugration  Province\n",
    "    if cmp.comment:\n",
    "        cmp_comment_dict = parse_comment(cmp.comment)\n",
    "        for attr in cmp_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmp_comment_dict[attr]            \n",
    "            \n",
    "    # Mugration Branch Major\n",
    "    if cmbmaj.comment:\n",
    "        cmbmaj_comment_dict = parse_comment(cmbmaj.comment)\n",
    "        for attr in cmbmaj_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmbmaj_comment_dict[attr]            \n",
    "            \n",
    "    # Mugration Branch Minor\n",
    "    if cmbmin.comment:\n",
    "        cmbmin_comment_dict = parse_comment(cmbmin.comment)\n",
    "        for attr in cmbmin_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmbmin_comment_dict[attr]              \n",
    "\n",
    "    # Time Tree         \n",
    "    if ct.comment:\n",
    "        ct_comment_dict = parse_comment(ct.comment)\n",
    "        for attr in ct_comment_dict:\n",
    "            comment_dict[\"timetree_\" + attr] = ct_comment_dict[attr]\n",
    "            \n",
    "    for attr in comment_dict:\n",
    "        metadata_df.at[node_name, attr] = comment_dict[attr]\n",
    "           \n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "common-transcription",
   "metadata": {},
   "source": [
    "### Add Confidence and Handle Root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "built-turning",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list = [mugconf_country_df, mugconf_province_df, mugconf_branch_major_df, mugconf_branch_minor_df]\n",
    "attr_list = [\"country\", \"province\", \"branch_major\", \"branch_minor\"]\n",
    "\n",
    "for df,attr in zip(df_list, attr_list):\n",
    "    for sample in df.index:\n",
    "        max_state = \"\"\n",
    "        max_val = 0        \n",
    "        for state,val in zip(df.columns, df.loc[sample]):\n",
    "            if val >= max_val:\n",
    "                max_val = val\n",
    "                max_state = state\n",
    "        metadata_df.at[sample,\"mugration_\" + attr] = max_state      \n",
    "        metadata_df.at[sample,\"mugration_\" + attr + \"_confidence\"] = round(max_val,2)\n",
    "\n",
    "display(metadata_df.loc[\"NODE0\"])\n",
    "#print(list(metadata_df[\"mugration_branch_minor_confidence\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "controlled-desperate",
   "metadata": {},
   "source": [
    "### Lat Lon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "destroyed-burton",
   "metadata": {},
   "outputs": [],
   "source": [
    "latlon_country_df = pd.DataFrame()\n",
    "latlon_province_df = pd.DataFrame()\n",
    "\n",
    "df_list = [latlon_country_df, latlon_province_df]\n",
    "attr_list = [\"country\", \"province\"]\n",
    "conf_df_list = [mugconf_country_df, mugconf_province_df]\n",
    "\n",
    "# Create a mapping of geo name to lat,lon\n",
    "for df,attr in zip(df_list, attr_list):\n",
    "    # Get data\n",
    "    for rec in metadata_df.iterrows():\n",
    "        node_name = rec[0]\n",
    "        name = rec[1][attr]\n",
    "        lat = rec[1][attr + \"_lat\"]\n",
    "        lon = rec[1][attr + \"_lon\"] \n",
    "        \n",
    "        if name == NO_DATA_CHAR: continue\n",
    "\n",
    "        if name not in df.index:\n",
    "\n",
    "            df.at[name, \"lat\"] = lat\n",
    "            df.at[name, \"lon\"] = lon\n",
    "            df.at[name, \"size\"] = 1\n",
    "        else:\n",
    "            df[\"size\"][name] += 1\n",
    "\n",
    "#display(latlon_country_df)\n",
    "#display(latlon_province_df)\n",
    "\n",
    "\n",
    "# Update lat lon for mugration\n",
    "for latlon_df, conf_df, attr in zip(df_list, conf_df_list, attr_list):\n",
    "    for rec in metadata_df.iterrows():\n",
    "        sample = rec[0]\n",
    "        name = rec[1][\"mugration_\" + attr]  \n",
    "        lat = latlon_df[\"lat\"][name]\n",
    "        lon = latlon_df[\"lon\"][name]\n",
    "                \n",
    "        metadata_df.at[sample, \"mugration_\" + attr +\"_lat\"] = lat\n",
    "        metadata_df.at[sample, \"mugration_\" + attr +\"_lon\"] = lon\n",
    "        \n",
    "        \n",
    "# Mapping file for auspice\n",
    "out_path_latlon = os.path.join(out_dir, \"latlon.tsv\")\n",
    "\n",
    "# Countries\n",
    "with open(out_path_latlon, \"w\") as outfile:\n",
    "    for country in latlon_country_df.index:\n",
    "        lat = str(latlon_country_df[\"lat\"][country])\n",
    "        lon = str(latlon_country_df[\"lon\"][country])\n",
    "        outfile.write(\"country\" + \"\\t\" + country + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "# Provinces\n",
    "with open(out_path_latlon, \"a\") as outfile:\n",
    "    for province in latlon_province_df.index:\n",
    "        lat = str(latlon_province_df[\"lat\"][province])\n",
    "        lon = str(latlon_province_df[\"lon\"][province])\n",
    "        outfile.write(\"province\" + \"\\t\" + province + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "offshore-connectivity",
   "metadata": {},
   "source": [
    "---\n",
    "# Convert Dataframes to Geopandas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "separate-bailey",
   "metadata": {},
   "source": [
    "### Modern Human Cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "tired-manual",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"geometry = []\n",
    "\n",
    "for rec in xu_2019_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    lon = float(rec[1][\"X\"].replace(\"'\",\"\"))\n",
    "    lat = float(rec[1][\"Y\"].replace(\"'\",\"\"))\n",
    "    point = shapely.geometry.Point(lon,lat)\n",
    "    geometry.append(point)\n",
    "\n",
    "xu_2019_gdf = geopandas.GeoDataFrame(xu_2019_df, crs=CRS, geometry=geometry)\n",
    "\n",
    "# Generate a figure with two axes, one for CartoPy, one for GeoPandas\n",
    "fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=(10, 5))\n",
    "world_polygons.plot(ax=ax, color=\"grey\", alpha=0.5)\n",
    "xu_2019_gdf.plot(ax=ax, c=\"red\", markersize=2)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "collectible-superintendent",
   "metadata": {},
   "source": [
    "### Project Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "filled-border",
   "metadata": {},
   "outputs": [],
   "source": [
    "geometry = []\n",
    "\n",
    "for rec in metadata_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    # Terminal nodes use the \"lat/lon\" attribute        \n",
    "    if rec[1][\"node_type\"] == \"terminal\":  \n",
    "        province = rec[1][\"province\"]\n",
    "        country = rec[1][\"country\"]\n",
    "        if province != NO_DATA_CHAR:\n",
    "            attr = \"province\"\n",
    "            name = province\n",
    "        else:\n",
    "            attr = \"country\"\n",
    "            name = country\n",
    "        lat = rec[1][\"lat\"]\n",
    "        lon = rec[1][\"lon\"]\n",
    "    # Internal nodes use mugration province\n",
    "    else:\n",
    "        name = rec[1][\"mugration_province\"]\n",
    "        lat  = rec[1][\"mugration_province_lat\"]\n",
    "        lon  = rec[1][\"mugration_province_lon\"]     \n",
    "    \n",
    "    if attr == \"province\":\n",
    "        size = latlon_province_df[\"size\"][name]\n",
    "    elif attr == \"country\":\n",
    "        size = latlon_country_df[\"size\"][name]\n",
    "    point = shapely.geometry.Point(lon,lat)\n",
    "    geometry.append(point)\n",
    "\n",
    "\n",
    "metadata_gdf = geopandas.GeoDataFrame(metadata_df, crs=\"epsg:3857\", geometry=geometry)\n",
    "\n",
    "# Add geometry size\n",
    "for rec in metadata_gdf.iterrows():\n",
    "    sample = rec[0]\n",
    "    name = rec[1][\"mugration_\" + attr]\n",
    "    size = latlon_df[\"size\"][name]  \n",
    "    metadata_gdf.at[sample, \"geometry_size\"] = size\n",
    "    \n",
    "\n",
    "\n",
    "# Test map\n",
    "\n",
    "# Generate a figure with two axes, one for CartoPy, one for GeoPandas\n",
    "fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=(10, 5))\n",
    "world_polygons.plot(ax=ax, color=\"grey\", alpha=0.5)\n",
    "metadata_gdf.plot(ax=ax, c=\"red\", markersize=\"geometry_size\", ec=\"black\", lw=0.25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "opened-rabbit",
   "metadata": {},
   "source": [
    "# Colors\n",
    "\n",
    "Use the full tree for color order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "crazy-reach",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_colors = os.path.join(out_dir, \"colors.tsv\")\n",
    "\n",
    "colors_dict = {}\n",
    "\n",
    "for attr in MUG_ATTRIBUTE_LIST:\n",
    "    # Create the color map\n",
    "    attr_key = attr.lower()\n",
    "    colors_dict[attr_key] = {}\n",
    "    for t in divtree.get_terminals():\n",
    "        attr_val = metadata_df[attr][t.name]\n",
    "        # Remove the letter suffix from branch_minor\n",
    "        if attr == \"branch_minor\":\n",
    "            while attr_val[-1].isalpha():\n",
    "                attr_val = attr_val[:-1] \n",
    "        if attr_val not in colors_dict[attr_key] and attr_val != NO_DATA_CHAR:\n",
    "            colors_dict[attr_key][attr_val] = {}\n",
    "            \n",
    "    # Create the custom color map (pyplot)\n",
    "    cmap = plt.get_cmap(\"rainbow\", len(colors_dict[attr_key]))\n",
    "    # Convert the color map to a list of RGB values\n",
    "    cmaplist = [cmap(i) for i in range(cmap.N)]\n",
    "    # Convert RGB values to hex colors\n",
    "    attr_hex = [colors.to_hex(col) for col in cmaplist]\n",
    "    \n",
    "    # Assign colors to value\n",
    "    for attr_val, attr_col in zip(colors_dict[attr_key], attr_hex):\n",
    "        colors_dict[attr_key][attr_val] = attr_col   \n",
    "        \n",
    "    # Add unknown\n",
    "    colors_dict[attr_key][UNKNOWN_CHAR] = \"#969696\"\n",
    "\n",
    "with open(out_path_colors, \"w\") as outfile:\n",
    "    for attr_key in colors_dict:\n",
    "        for attr_val in colors_dict[attr_key]:\n",
    "            outfile.write(str(attr_key) + \"\\t\" + str(attr_val) + \"\\t\" + str(colors_dict[attr_key][attr_val]) + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "regulated-bunny",
   "metadata": {},
   "source": [
    "---\n",
    "# Parse Tree Into Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "elegant-second",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_posns = get_x_positions(divtree)\n",
    "y_posns = get_y_positions(divtree)\n",
    "\n",
    "# Initialize \n",
    "# Date lower and upper error bars for tips\n",
    "metadata_gdf[\"date_lower\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"date_upper\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "    \n",
    "for c in divtree.find_clades():\n",
    "    # ------------------------------------------------------------\n",
    "    # Coordinates for the divergence tree\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    metadata_gdf.at[c.name, 'divtree_coord_x'] = coord_x\n",
    "    metadata_gdf.at[c.name, 'divtree_coord_y'] = coord_y\n",
    "    \n",
    "    # ------------------------------------------------------------\n",
    "    # Root to tip distance (same as div x)\n",
    "    metadata_gdf.at[c.name, \"rtt_dist\"] = coord_x\n",
    "    \n",
    "    # ------------------------------------------------------------\n",
    "    # Branch Major and Branch Minor and Colors\n",
    "    branch_minor = metadata_gdf[\"mugration_branch_minor\"][c.name]\n",
    "    if branch_minor != NO_DATA_CHAR:\n",
    "        # Remove letters at end\n",
    "        while branch_minor[-1].isalpha():\n",
    "            branch_minor = branch_minor[:-1]    \n",
    "    metadata_gdf.at[c.name, \"mugration_branch_minor\"] = branch_minor\n",
    "    branch_major = metadata_gdf[\"mugration_branch_major\"][c.name]\n",
    "    branch_major_color = colors_dict[\"branch_major\"][branch_major]\n",
    "    metadata_gdf.at[c.name, \"branch_major_color\"] = branch_major_color    \n",
    "\n",
    "    # ------------------------------------------------------------\n",
    "    # Date lower and upper error bars for tips\n",
    "    date_mean = metadata_gdf[\"date_mean\"][c.name] \n",
    "    if date_mean != NO_DATA_CHAR:\n",
    "        metadata_gdf.at[c.name, \"date_lower\"] = date_mean - metadata_gdf[\"date_err\"][c.name]\n",
    "        metadata_gdf.at[c.name, \"date_upper\"] = date_mean + metadata_gdf[\"date_err\"][c.name]\n",
    "\n",
    "x_posns = get_x_positions(timetree)\n",
    "y_posns = get_y_positions(timetree)\n",
    "\n",
    "\n",
    "for c in timetree.find_clades():\n",
    "\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    metadata_gdf.at[c.name, 'timetree_coord_x'] = coord_x\n",
    "    metadata_gdf.at[c.name, 'timetree_coord_y'] = coord_y\n",
    "    \n",
    "    timetree_date = int(float(metadata_gdf[\"timetree_date\"][c.name]))\n",
    "    timetree_date_calendar = CURRENT_YEAR + timetree_date\n",
    "    metadata_gdf.at[c.name, 'timetree_date_calendar'] = timetree_date_calendar\n",
    "    \n",
    "    timetree_CI_date = metadata_gdf[\"timetree_CI_date\"][c.name]\n",
    "    if timetree_CI_date == NO_DATA_CHAR:\n",
    "        timetree_CI_lower_err = 0\n",
    "        timetree_CI_upper_err = 0\n",
    "    else:\n",
    "        timetree_CI_split = [int(float(d)) for d in timetree_CI_date.strip(\"{}\").split(\",\")]\n",
    "        timetree_CI_calendar = [CURRENT_YEAR + d for d in timetree_CI_split]\n",
    "        timetree_CI_lower_err = timetree_date_calendar - timetree_CI_calendar[0]\n",
    "        timetree_CI_upper_err = timetree_CI_calendar[1] - timetree_date_calendar\n",
    "    metadata_gdf.at[c.name,\"timetree_CI_lower_err\"] = timetree_CI_lower_err\n",
    "    metadata_gdf.at[c.name,\"timetree_CI_upper_err\"] = timetree_CI_upper_err   \n",
    "\n",
    "display(metadata_gdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "korean-magazine",
   "metadata": {},
   "source": [
    "## Calculate Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "original-afternoon",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_gdf[\"root_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"clade_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "\n",
    "for c in divtree.find_clades():\n",
    "    \n",
    "\n",
    "display(metadata_gdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rotary-document",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ancient_gdf = metadata_gdf[metadata_gdf[\"biosample_comment\"].str.contains(\"Ancient\")]\n",
    "#modern_gdf  = metadata_gdf[metadata_gdf[\"biosample_comment\"].str.contains(\"Modern\")]\n",
    "\n",
    "#display(ancient_gdf)\n",
    "#display(modern_gdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "black-classroom",
   "metadata": {},
   "source": [
    "---\n",
    "# 4. Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "featured-digit",
   "metadata": {},
   "source": [
    "## Timetree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "earlier-wisconsin",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_timetree = copy.deepcopy(timetree)\n",
    "\n",
    "for c in out_timetree.find_clades():\n",
    "    if not c.confidence:\n",
    "        c.confidence = 0\n",
    "\n",
    "metadata_to_comment(out_timetree, metadata_gdf)    \n",
    "out_timetree_path = os.path.join(out_dir, \"final.timetree.nex\")\n",
    "Phylo.write(out_timetree, out_timetree_path, \"nexus\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "subjective-calendar",
   "metadata": {},
   "source": [
    "## Divtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "southeast-facing",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_divtree = copy.deepcopy(divtree)\n",
    "\n",
    "for c in out_divtree.find_clades():\n",
    "    if not c.confidence:\n",
    "        c.confidence = 0\n",
    "\n",
    "metadata_to_comment(out_divtree, metadata_gdf)    \n",
    "out_divtree_path = os.path.join(out_dir, \"final.divtree.nex\")\n",
    "Phylo.write(out_divtree, out_divtree_path, \"nexus\", format_branch_length=\"%1.10f\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "limited-modem",
   "metadata": {},
   "source": [
    "## Augur JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "distinct-index",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "manual-scottish",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple_gdf = metadata_gdf[[\"strain\",\"branch_length\",\"timetree_date_calendar\"]]\n",
    "simple_gdf.rename(columns={\"timetree_date_calendar\": \"timetree_numdate\"}, inplace=True)\n",
    "\n",
    "columns = [\n",
    "    # Draw Divergence Tree\n",
    "    \"branch_length\",\n",
    "    # Draw Time Tree\n",
    "    \"timetree_date_calendar\",\n",
    "    # Draw Country Map\n",
    "    \"mugration_country\",\n",
    "    \"mugration_country_confidence\",\n",
    "    # Draw Province Map\n",
    "    \"mugration_province\",\n",
    "    \"mugration_province_confidence\",\n",
    "    # Colors and Filters\n",
    "    \"mugration_branch_major\",\n",
    "    \"mugration_branch_major_confidence\",\n",
    "    \"mugration_branch_minor\",\n",
    "    \"mugration_branch_minor_confidence\",   \n",
    "    \"branch_number\",\n",
    "    \"branch_support\",\n",
    "    \"continent\",\n",
    "    \"node_type\",\n",
    "    \"geometry_size\",\n",
    "    # Text Description\n",
    "    \"biosample_accession\",\n",
    "    \"strain\",\n",
    "    # Tip Dates\n",
    "    \"date_mean\",\n",
    "    \"date_err\",\n",
    "    \"date_bp_mean\",\n",
    "]\n",
    "\n",
    "# Edit df\n",
    "auspice_gdf = metadata_gdf[columns]\n",
    "auspice_gdf[\"timetree_num_date_confidence\"] = [NO_DATA_CHAR] * len(auspice_gdf)\n",
    "\n",
    "for attr in MUG_ATTRIBUTE_LIST:\n",
    "    col = attr + \"_conf_category\"\n",
    "    auspice_gdf[col] = [\"NA\"] * len(auspice_gdf)\n",
    "    for rec in auspice_gdf.iterrows():\n",
    "        sample = rec[0]\n",
    "        conf = auspice_gdf[\"mugration_\" + attr + \"_confidence\"][sample]\n",
    "        if conf >= 0.95:\n",
    "            category = \"HIGH\"\n",
    "        elif conf < 0.95:\n",
    "            category = \"LOW\"\n",
    "        auspice_gdf.at[sample, col] = category\n",
    "\n",
    "for rec in metadata_gdf.iterrows():\n",
    "    sample = rec[0]\n",
    "    date = rec[1][\"timetree_date_calendar\"]\n",
    "    date_lower = float(date) - rec[1][\"timetree_CI_lower_err\"]\n",
    "    date_upper = float(date) + rec[1][\"timetree_CI_upper_err\"]\n",
    "    #date_confidence = \"{}:{}\".format(date_lower, date_upper)\n",
    "    date_confidence = [date_lower, date_upper]\n",
    "    auspice_gdf.at[sample,\"timetree_num_date_confidence\"] = date_confidence\n",
    "    \n",
    "auspice_gdf.rename(columns={\"timetree_date_calendar\": \"timetree_num_date\"}, inplace=True)\n",
    "\n",
    "display(auspice_gdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "legendary-listening",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import sys, importlib\n",
    "#importlib.reload(sys.modules['plot_phylo_functions'])\n",
    "#from plot_phylo_functions import augur_export,auspice_export\n",
    "\n",
    "augur_dict = augur_export(\n",
    "    tree_path=out_divtree_path, \n",
    "    aln_path=aln_path, \n",
    "    #tree=timetree,     \n",
    "    tree=divtree, \n",
    "    tree_df=auspice_gdf, \n",
    "    color_keyword_exclude=[\"geometry\"],\n",
    "    #color_keyword_exclude=[\"color\", \"coord\", \"reg\", \"lat\", \"lon\"],\n",
    "    type_convert = {\n",
    "        \"branch_number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])\n",
    "\n",
    "out_path_augur_json = os.path.join(out_dir, \"augur_All.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "\n",
    "auspice_dict = auspice_export(\n",
    "    #tree=timetree, \n",
    "    tree=divtree,\n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=out_path_colors,\n",
    "    auspice_latlons_path=out_path_latlon, \n",
    "    )\n",
    "\n",
    "label_col = list(auspice_gdf.columns)\n",
    "\n",
    "# Recursively add branch attrs\n",
    "branch_attributes(\n",
    "    tree_dict=auspice_dict[\"tree\"], \n",
    "    sub_dict=auspice_dict[\"tree\"], \n",
    "    df=auspice_gdf,\n",
    "    label_col=label_col,\n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(out_dir, \"auspice_All.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\\n\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "#out_path_auspice_remote_json = os.path.join(auspice_remote_dir, AUSPICE_PREFIX + \"_All.json\" )\n",
    "#utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "#export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "#print(\"Validation successful for remote JSON.\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "boxed-conclusion",
   "metadata": {},
   "source": [
    "## Auspice Subtrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hispanic-bracelet",
   "metadata": {},
   "outputs": [],
   "source": [
    "for branch in BRANCH_LIST:\n",
    "    print(branch)\n",
    "    timetree_copy = copy.deepcopy(timetree)\n",
    "    \n",
    "    # Create the subtree df\n",
    "    subtree_df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[branch])]\n",
    "\n",
    "    subtree_df.sort_values(\"timetree_coord_y\", inplace=True)\n",
    "    subtree_tips = subtree_df[subtree_df[\"node_type\"] == \"terminal\"]\n",
    "    subtree_internals = subtree_df[subtree_df[\"node_type\"] == \"internal\"]\n",
    "\n",
    "    subtree = extract_subtree(\n",
    "        tree=timetree_copy, \n",
    "        tips=list(subtree_tips.index), \n",
    "        df=metadata_gdf, \n",
    "        color_branches=False)\n",
    "    \n",
    "    metadata_to_comment(subtree, metadata_gdf)   \n",
    "\n",
    "    # Export subtree\n",
    "    out_subtree_path = os.path.join(out_dir, \"{}.nex\".format(branch))\n",
    "    Phylo.write(subtree, out_subtree_path, \"nexus\")\n",
    "\n",
    "    #fig, ax = plt.subplots(1)\n",
    "    #Phylo.draw(subtree, axes=ax, show_confidence=False, label_func = lambda x: '', do_show=False)\n",
    "    \n",
    "    augur_dict = augur_export(\n",
    "        tree_path=out_subtree_path, \n",
    "        aln_path=aln_path,  \n",
    "        tree=subtree, \n",
    "        tree_df=auspice_gdf, \n",
    "        color_keyword_exclude=[\"geometry\"],\n",
    "        type_convert = {\n",
    "            \"branch_number\" : (lambda x : str(x))\n",
    "        },\n",
    "    ) \n",
    "        \n",
    "    out_path_augur_json = os.path.join(out_dir, \"augur_{}.json\".format(branch) )\n",
    "    utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "    \n",
    "    auspice_dict = auspice_export(\n",
    "        tree=subtree,\n",
    "        augur_json_paths=[out_path_augur_json], \n",
    "        auspice_config_path=auspice_config_path, \n",
    "        auspice_colors_path=out_path_colors,\n",
    "        auspice_latlons_path=out_path_latlon, \n",
    "        )    \n",
    "    \n",
    "    label_col = list(auspice_gdf.columns)\n",
    "\n",
    "    # Recursively add branch attrs\n",
    "    branch_attributes(\n",
    "        tree_dict=auspice_dict[\"tree\"], \n",
    "        sub_dict=auspice_dict[\"tree\"], \n",
    "        df=auspice_gdf,\n",
    "        label_col=label_col,\n",
    "        )    \n",
    "    \n",
    "    # Write outputs - For Local Rendering\n",
    "    out_path_auspice_local_json = os.path.join(out_dir, \"auspice_{}.json\".format(branch) )\n",
    "    utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "    export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "    print(\"Validation successful for local JSON.\\n\")\n",
    "\n",
    "    # Write outputs - For Remote Rendering\n",
    "    #out_path_auspice_remote_json = os.path.join(auspice_remote_dir, AUSPICE_PREFIX + \"_{}.json\".format(branch) )\n",
    "    #utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "    #export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "    #print(\"Validation successful for remote JSON.\\n\")    \n",
    "\n",
    "\"\"\"\n",
    "# Terminal df\n",
    "subtree_tips = subtree_df[subtree_df[\"node_type\"] == \"terminal\"]\n",
    "\n",
    "# Internal df\n",
    "subtree_internals = subtree_df[subtree_df[\"node_type\"] == \"internal\"]\n",
    "timetree_copy = copy.deepcopy(timetree)\n",
    "subtree = extract_subtree(tree=timetree_copy, tips=list(subtree_tips.index), df=metadata_gdf, color_branches=False)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "respective-million",
   "metadata": {},
   "source": [
    "---\n",
    "# 5. Plot RTT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beneficial-linux",
   "metadata": {},
   "source": [
    "## All"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "moving-stuff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [480, 480]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "geo_dist_all = []\n",
    "genetic_dist_all = []\n",
    "\n",
    "fig, ax = plt.subplots(1,figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "df = metadata_gdf[(metadata_gdf[\"node_type\"] == \"terminal\") & (metadata_gdf[\"date_mean\"] != NO_DATA_CHAR)]\n",
    "reg_x = \"date_mean\"\n",
    "reg_y = \"rtt_dist\"\n",
    "reg_err = \"date_err\"\n",
    "\n",
    "# -----------------------\n",
    "# Regression\n",
    "slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(list(df[reg_x]),list(df[reg_y]))\n",
    "r2 = r_value * r_value    \n",
    "y_intercept = CURRENT_YEAR - (0 - intercept)/slope\n",
    "p_sig = \"\"\n",
    "p_thresh = 0.05    \n",
    "if p_value < p_thresh:\n",
    "    p_sig = \"*\"\n",
    "\n",
    "mut_per_year = SEQ_LEN * slope\n",
    "year_per_mut = 1 / mut_per_year\n",
    "reg_dict = {\n",
    "    \"slope\": slope,\n",
    "    \"x_intercept\" : intercept,\n",
    "    \"y_intercept\" : y_intercept,\n",
    "    \"p_value\" : p_value,\n",
    "    \"p_sig\" : p_sig,\n",
    "    \"year_per_mut\" : year_per_mut\n",
    "}\n",
    "\n",
    "sns.regplot(\n",
    "    ax=ax,\n",
    "    data=df, \n",
    "    x=list(df[reg_x]), \n",
    "    y=list(df[reg_y]), \n",
    "    #robust=True,\n",
    "    ci=95,\n",
    "    scatter_kws={\"s\": 0},\n",
    "    line_kws={\"linewidth\" : 0.5},\n",
    "    color=\"grey\",\n",
    "    #color=df[\"branch_major_color\"][0],\n",
    "    label=(\n",
    "    \"   RÂ²: {}\".format(round(r2,2))\n",
    "    + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)\n",
    "    + \"\\nRate: {:.2} +/ {:.2}\".format(slope, std_err)  \n",
    "    + \"\\n         {} years/substitution\".format(round(year_per_mut))\n",
    "    + \"\\nRoot: {} YBP\".format(round(y_intercept))                \n",
    "    )\n",
    ")\n",
    "\n",
    "ax.errorbar(\n",
    "    data=df, \n",
    "    x=reg_x, \n",
    "    y=reg_y, \n",
    "    xerr=reg_err,\n",
    "    yerr=None,\n",
    "    ls='none',\n",
    "    c = \"black\",  \n",
    "    capsize=1,\n",
    "    label=None,\n",
    "    zorder=1,\n",
    "    lw=0.5,\n",
    "    )  \n",
    "\n",
    "sns.scatterplot(\n",
    "    ax=ax,\n",
    "    data=df, \n",
    "    x=reg_x, \n",
    "    y=reg_y, \n",
    "    s=10,    \n",
    "    ec = \"white\", \n",
    "    color=\"black\",\n",
    "    alpha=0.75,\n",
    "    zorder=2,\n",
    ") \n",
    "\n",
    "    \n",
    "# Set xlimits\n",
    "xlim = ax.get_xlim()\n",
    "x_buff = round((xlim[1] - xlim[0]) * 0.1,1)\n",
    "new_xlim = [xlim[0]-x_buff, xlim[1]+x_buff]\n",
    "if new_xlim[1] > CURRENT_YEAR:\n",
    "    new_xlim = [new_xlim[0], CURRENT_YEAR + 200]\n",
    "ax.set_xlim(new_xlim)\n",
    "    \n",
    "ax.set_ylabel(\"Root to Tip Distance\")\n",
    "ax.set_xlabel(\"Date\")    \n",
    "ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0)) \n",
    "\n",
    "# Format and position legend\n",
    "legend = ax.legend(\n",
    "    borderpad=0.8, \n",
    "    handletextpad=-2, \n",
    "    edgecolor=\"black\", \n",
    "    #bbox_to_anchor=(0.5, -0.30), \n",
    "    #loc='center',\n",
    "    fontsize=FONTSIZE,\n",
    ")\n",
    "frame = legend.get_frame().set_linewidth(0.5)\n",
    "\n",
    "out_path = os.path.join(out_dir, \"rtt_all\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "painted-reply",
   "metadata": {},
   "source": [
    "## Clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "reserved-separate",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "reg_x = \"date_mean\"\n",
    "reg_y = \"rtt_dist\"\n",
    "reg_err = \"date_err\"\n",
    "\n",
    "nrow=3\n",
    "ncol=4\n",
    "\n",
    "fig, axes = plt.subplots(nrow,ncol,figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.35, hspace=1)\n",
    "\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for label in BRANCH_LIST:\n",
    "    print(\"Branch:\", label)\n",
    "    \n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]  \n",
    "    # If no records for this branch, continue\n",
    "    if len(df) < 2: continue\n",
    "    # -----------------------\n",
    "    # Regression\n",
    "    slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(list(df[reg_x]),list(df[reg_y]))\n",
    "    r2 = r_value * r_value    \n",
    "    y_intercept = CURRENT_YEAR - (0 - intercept)/slope\n",
    "    p_sig = \"\"\n",
    "    p_thresh = 0.05    \n",
    "    if p_value < p_thresh:\n",
    "        p_sig = \"*\"\n",
    "\n",
    "    mut_per_year = SEQ_LEN * slope\n",
    "    year_per_mut = 1 / mut_per_year\n",
    "    \n",
    "    reg_dict = {\n",
    "        \"slope\": slope,\n",
    "        \"x_intercept\" : intercept,\n",
    "        \"y_intercept\" : y_intercept,\n",
    "        \"p_value\" : p_value,\n",
    "        \"p_sig\" : p_sig,\n",
    "        \"year_per_mut\" : year_per_mut\n",
    "    }\n",
    "              \n",
    "    sns.regplot(\n",
    "        ax=ax,\n",
    "        data=df, \n",
    "        x=list(df[reg_x]), \n",
    "        y=list(df[reg_y]), \n",
    "        #robust=True,\n",
    "        ci=95,\n",
    "        scatter_kws={\"s\": 0},\n",
    "        line_kws={\"linewidth\" : 0.5},\n",
    "        color=\"grey\",\n",
    "        #color=df[\"branch_major_color\"][0],\n",
    "        label=(\n",
    "        \"   RÂ²: {}\".format(round(r2,2))\n",
    "        + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)\n",
    "        + \"\\nRate: {:.2} +/ {:.2}\".format(slope, std_err)  \n",
    "        + \"\\n         {} years/substitution\".format(round(year_per_mut))\n",
    "        + \"\\nRoot: {} YBP\".format(round(y_intercept))                \n",
    "        )\n",
    "    )\n",
    "    \n",
    "    ax.errorbar(\n",
    "        data=df, \n",
    "        x=\"date_mean\",\n",
    "        y=\"rtt_dist\", \n",
    "        xerr=\"date_err\",\n",
    "        yerr=None,\n",
    "        ls='none',\n",
    "        c = df[\"branch_major_color\"][0],  \n",
    "        capsize=1,\n",
    "        label=None,\n",
    "        zorder=1,\n",
    "        lw=0.5,\n",
    "        )  \n",
    "    \n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        data=df, \n",
    "        x=\"date_mean\",\n",
    "        s=10,\n",
    "        y=\"rtt_dist\", \n",
    "        c = df[\"branch_major_color\"], \n",
    "        ec = \"black\",\n",
    "        alpha=0.75,\n",
    "        zorder=2,\n",
    "    )    \n",
    "    \n",
    "    # Set xlimits\n",
    "    xlim = ax.get_xlim()\n",
    "    x_buff = round((xlim[1] - xlim[0]) * 0.1,1)\n",
    "    new_xlim = [xlim[0]-x_buff, xlim[1]+x_buff]\n",
    "    if new_xlim[1] > CURRENT_YEAR:\n",
    "        new_xlim = [new_xlim[0], CURRENT_YEAR]\n",
    "    ax.set_xlim(new_xlim)\n",
    "\n",
    "    # Format and position legend\n",
    "    legend = ax.legend(\n",
    "        borderpad=0.8, \n",
    "        handletextpad=-2, \n",
    "        edgecolor=\"black\", \n",
    "        bbox_to_anchor=(0.5, -0.55), \n",
    "        loc='center',\n",
    "        fontsize=FONTSIZE * 0.80,\n",
    "    )\n",
    "    frame = legend.get_frame().set_linewidth(0.5)\n",
    "        \n",
    "    # Format axis\n",
    "    ax.set_title(label)\n",
    "\n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Root to Tip Distance\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Date\")    \n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "    \n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_linewidth(0.5)    \n",
    "            \n",
    "\n",
    "    # Update axis\n",
    "    if i_col == ncol - 1:\n",
    "        i_col = 0\n",
    "        i_row += 1\n",
    "    else:\n",
    "        i_col += 1\n",
    "        \n",
    "    \n",
    "out_path = os.path.join(out_dir, \"rtt_clades\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "successful-egyptian",
   "metadata": {},
   "source": [
    "# Tip Dating"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "current-prairie",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "reg_x = \"date_mean\"\n",
    "reg_y = \"rtt_dist\"\n",
    "reg_err = \"date_err\"\n",
    "\n",
    "nrow=3\n",
    "ncol=3\n",
    "\n",
    "fig, axes = plt.subplots(nrow,ncol,figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.35, hspace=0.35)\n",
    "#fig.subplots_adjust(wspace=0.35, hspace=0.85)\n",
    "\n",
    "for row in axes:\n",
    "    for ax in row:\n",
    "        for spine in ax.spines:\n",
    "            ax.spines[spine].set_linewidth(0.5) \n",
    "\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for label in ANCIENT_BRANCH_LIST:\n",
    "    print(\"Branch:\", label)\n",
    "    \n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    if len(df) < 2: continue\n",
    "    \n",
    "    # -----------------------\n",
    "    # Regression\n",
    "    slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(list(df[reg_x]),list(df[reg_y]))\n",
    "    r2 = r_value * r_value    \n",
    "    y_intercept = CURRENT_YEAR - (0 - intercept)/slope\n",
    "    p_sig = \"\"\n",
    "    p_thresh = 0.05    \n",
    "    if p_value < p_thresh:\n",
    "        p_sig = \"*\"\n",
    "        \n",
    "    mut_per_year = SEQ_LEN * slope\n",
    "    year_per_mut = 1 / mut_per_year\n",
    "    \n",
    "    reg_dict = {\n",
    "        \"slope\": slope,\n",
    "        \"x_intercept\" : intercept,\n",
    "        \"y_intercept\" : y_intercept,\n",
    "        \"p_value\" : p_value,\n",
    "        \"p_sig\" : p_sig,\n",
    "        \"year_per_mut\" : year_per_mut\n",
    "    }\n",
    "    \n",
    "              \n",
    "    sns.regplot(\n",
    "        ax=ax,\n",
    "        data=df, \n",
    "        x=list(df[reg_x]),\n",
    "        y=list(df[reg_y]),\n",
    "        #robust=True,\n",
    "        ci=95,\n",
    "        scatter_kws={\"s\": 0},\n",
    "        line_kws={\"linewidth\" : 0.5},\n",
    "        color=\"grey\",\n",
    "        #color=df[\"branch_major_color\"][0],\n",
    "        label=(\n",
    "        \"   RÂ²: {}\".format(round(r2,2))\n",
    "        + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)\n",
    "        + \"\\nRate: {:.2} +/ {:.2}\".format(slope, std_err)  \n",
    "        + \"\\n         {} years/substitution\".format(round(year_per_mut))\n",
    "        + \"\\nRoot: {} YBP\".format(round(y_intercept))                \n",
    "        )\n",
    "    )\n",
    "    \n",
    "    ax.errorbar(\n",
    "        data=df, \n",
    "        x=\"date_mean\",\n",
    "        y=\"rtt_dist\", \n",
    "        xerr=\"date_err\",\n",
    "        yerr=None,\n",
    "        ls='none',\n",
    "        c = df[\"branch_major_color\"][0],  \n",
    "        capsize=1,\n",
    "        label=None,\n",
    "        zorder=1,\n",
    "        lw=0.5,\n",
    "        )  \n",
    "    \n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        data=df, \n",
    "        x=\"date_mean\",\n",
    "        s=10,\n",
    "        y=\"rtt_dist\", \n",
    "        c = df[\"branch_major_color\"], \n",
    "        ec = \"black\",\n",
    "        alpha=0.75,\n",
    "        zorder=2,\n",
    "    )    \n",
    "    \n",
    "        \n",
    "    # Set xlimits\n",
    "    xlim = ax.get_xlim()\n",
    "    x_buff = round((xlim[1] - xlim[0]) * 0.1,1)\n",
    "    new_xlim = [xlim[0]-x_buff, xlim[1]+x_buff]\n",
    "    if new_xlim[1] > CURRENT_YEAR:\n",
    "        new_xlim = [new_xlim[0], CURRENT_YEAR]\n",
    "    ax.set_xlim(new_xlim)\n",
    "\n",
    "    # Format and position legend\n",
    "    legend = ax.legend(\n",
    "        borderpad=0.8, \n",
    "        handletextpad=-2, \n",
    "        edgecolor=\"black\", \n",
    "        bbox_to_anchor=(0.5, -0.55), \n",
    "        loc='center',\n",
    "        fontsize=FONTSIZE * 0.80,\n",
    "    )\n",
    "    frame = legend.get_frame().set_linewidth(0.5)\n",
    "    ax.get_legend().remove()\n",
    "    \n",
    "    # Format axis\n",
    "    ax.set_title(label)\n",
    "\n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Root to Tip Distance\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Date\")    \n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "      \n",
    "            \n",
    "    # Now predicted values\n",
    "    i_row += 1\n",
    "    ax = axes[i_row][i_col]    \n",
    "    \n",
    "    date_diffs = []\n",
    "    for date_mean, rtt_dist in zip(df[\"date_mean\"], df[\"rtt_dist\"]):\n",
    "        pred_date = (rtt_dist - reg_dict[\"x_intercept\"]) / reg_dict[\"slope\"]\n",
    "        date_diff = pred_date - date_mean\n",
    "        date_diffs.append(date_diff)\n",
    "        \n",
    "    sns.kdeplot(\n",
    "       date_diffs,\n",
    "       ax=ax,\n",
    "       fill=True, \n",
    "       common_norm=True,\n",
    "       alpha=0.5, \n",
    "       linewidth=1,\n",
    "       bw_adjust=0.25,\n",
    "       color=df[\"branch_major_color\"][0],\n",
    "    )\n",
    "    \n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Density\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Regression Date - Actual Date\")    \n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))  \n",
    "    ax.set_xlim(-300,300)    \n",
    "\n",
    "    # ---------------------------------------------------------------\n",
    "    # Diff based on LSD2 estimate\n",
    "    i_row += 1\n",
    "    ax = axes[i_row][i_col]  \n",
    "    \n",
    "    date_diffs = []\n",
    "    for date_mean, timetree_date in zip(df[\"date_mean\"], df[\"timetree_date\"]):\n",
    "        date_est = CURRENT_YEAR + float(timetree_date)\n",
    "        date_diff = date_est - date_mean\n",
    "        date_diffs.append(date_diff)  \n",
    "        \n",
    "    sns.kdeplot(\n",
    "       date_diffs,\n",
    "       ax=ax,\n",
    "       fill=True, \n",
    "       common_norm=True,\n",
    "       alpha=0.5, \n",
    "       linewidth=1,\n",
    "       bw_adjust=0.25,\n",
    "       color=df[\"branch_major_color\"][0],\n",
    "    )\n",
    "    \n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Density\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"LSD2 Date - Actual Date\")    \n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))   \n",
    "    ax.set_xlim(-300,300)\n",
    "    \n",
    "    # Update axis\n",
    "    i_row = 0\n",
    "    i_col += 1\n",
    "\n",
    "\n",
    "out_path = os.path.join(out_dir, \"tip-dating\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "stainless-reporter",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"All:\", len(metadata_gdf[metadata_gdf[\"node_type\"] == \"terminal\"]))\n",
    "print(\"Ancient:\",len(ancient_gdf), len(ancient_gdf) / len(metadata_gdf))\n",
    "print(\"Modern:\",len(modern_gdf), len(modern_gdf) / len(metadata_gdf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "terminal-degree",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"TARGET_RES = [1280, 720]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "fig, ax = plt.subplots(1, dpi=DPI, figsize=FIGSIZE)\n",
    "\n",
    "rtt_dist_ln = []\n",
    "for rtt_dist in df[\"rtt_dist\"]:\n",
    "    rtt_dist_ln.append(math.log(rtt_dist))\n",
    "\n",
    "sns.kdeplot(\n",
    "   rtt_dist_ln,\n",
    "    ax=ax,\n",
    ")\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "worldwide-survival",
   "metadata": {},
   "source": [
    "## 1.PRE Subtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "toxic-fellowship",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the subtree df\n",
    "subtree_df = metadata_gdf[metadata_gdf[\"mugration_branch_major\"] == \"1.PRE\"]\n",
    "subtree_df.sort_values(\"timetree_coord_y\", inplace=True)\n",
    "\n",
    "# Terminal df\n",
    "subtree_tips = subtree_df[subtree_df[\"node_type\"] == \"terminal\"]\n",
    "\n",
    "# Internal df\n",
    "subtree_internals = subtree_df[subtree_df[\"node_type\"] == \"internal\"]\n",
    "timetree_copy = copy.deepcopy(timetree)\n",
    "subtree = extract_subtree(tree=timetree_copy, tips=list(subtree_tips.index), df=metadata_gdf, color_branches=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "quiet-split",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1080, 1080]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "#fig, axes = plt.subplots(1,figsize=FIGSIZE, dpi=DPI, sharey=True)\n",
    "#ax_tree=axes\n",
    "\n",
    "fig, axes = plt.subplots(1,2,figsize=FIGSIZE, dpi=DPI, sharey=True)\n",
    "fig.subplots_adjust(wspace=0.5)  \n",
    "        \n",
    "##############################################################\n",
    "# TIPS\n",
    "ax=axes[0]\n",
    "\n",
    "# Give the numdate attribute\n",
    "for c in subtree.find_clades():\n",
    "    c.numdate = metadata_gdf[\"timetree_date_calendar\"][c.name]\n",
    "\n",
    "Phylo.draw(subtree,\n",
    "           axes=ax,\n",
    "             show_confidence=False, \n",
    "             label_func = lambda x: '',\n",
    "             #label_func = lambda x: \"  {}\".format(metadata_gdf[\"country\"][x.name]) if x.is_terminal() else '',\n",
    "             do_show=False,\n",
    "             branch_labels = lambda x: (\n",
    "                 '*' if not x.is_terminal() and metadata_gdf[\"branch_support\"][x.name] >= 95\n",
    "                 else ''),\n",
    "            )\n",
    "\n",
    "\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_yticks([i for i in range(1,len(subtree_tips)+1)])\n",
    "ax.set_yticklabels(list(subtree_tips[\"country\"]), fontsize=FONTSIZE, ha=\"center\",position=(1.2,0))\n",
    "ax.yaxis.tick_right()\n",
    "\n",
    "\n",
    "# --------------------------------------------\n",
    "# Ticks\n",
    "ticks_dict = convert_timetree_ticks(tree=subtree, step=100)\n",
    "# Set the new tick locations and labels\n",
    "# Leave out last one\n",
    "ax.set_xticks(ticks_dict[\"tick_locs\"][:])\n",
    "ax.set_xticklabels(ticks_dict[\"tick_labels\"][:], ROTATION=45, ha=\"right\")\n",
    "date_to_x = ticks_dict[\"tick_locs\"][0] - float(ticks_dict[\"tick_labels\"][0])\n",
    "\n",
    "# --------------------------------------------\n",
    "# Lines\n",
    "# Vertical lines for dates\n",
    "for loc in ticks_dict[\"tick_locs\"][:]:\n",
    "    ax.axvline(x=loc, color=\"darkgrey\", alpha=0.5, lw=0.25)\n",
    "        \n",
    "\n",
    "\n",
    "# --------------------------------------------\n",
    "# Tip Locations\n",
    "timetree_date_err = [[],[]]\n",
    "for lower,upper in zip(subtree_tips[\"timetree_CI_lower_err\"], subtree_tips[\"timetree_CI_upper_err\"] ):\n",
    "    timetree_date_err[0].append(lower)\n",
    "    timetree_date_err[1].append(upper)\n",
    "    \n",
    "ax.errorbar(\n",
    "    data=subtree_tips, \n",
    "    x=[float(d) + date_to_x for d in subtree_tips[\"timetree_date_calendar\"]],\n",
    "    y=[i-0.1 for i in range(1,len(subtree_tips)+1,1)], \n",
    "    yerr=None, \n",
    "    xerr=timetree_date_err,\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=2,\n",
    "    label=\"Estimated Date\",\n",
    "    zorder=3,\n",
    "    lw=1, \n",
    "    color=\"darkgoldenrod\",\n",
    "    marker=\"o\",\n",
    "    markersize=5,\n",
    "    markerfacecolor=subtree_tips[\"branch_major_color\"][0], \n",
    "    \n",
    ")\n",
    "\n",
    "ax.errorbar(\n",
    "    data=subtree_tips, \n",
    "    x=[d + date_to_x for d in subtree_tips[\"date_mean\"]],\n",
    "    y=[i+0.1 for i in range(1,len(subtree_tips)+1,1)], \n",
    "    yerr=None, \n",
    "    xerr=\"date_err\",\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=2,\n",
    "    label=\"Archaeological Range\",\n",
    "    zorder=2,\n",
    "    lw=1, \n",
    "    color=\"blue\", \n",
    "    #c=subtree_tips[\"branch_major_color\"][0],    \n",
    ")\n",
    "\n",
    "ax.set_title(\"Tip Dating\", fontsize=FONTSIZE * 2)\n",
    "ax.set_xlabel(\"Date\")\n",
    "\n",
    "legend = ax.legend(fontsize=FONTSIZE, loc=\"center\", bbox_to_anchor=(1.25,1.05), edgecolor=\"black\")\n",
    "legend.get_frame().set_linewidth(0.5)\n",
    "\n",
    "##############################################################\n",
    "# INTERNAL AXIS\n",
    "ax=axes[1]\n",
    "Phylo.draw(subtree,\n",
    "           axes=ax,\n",
    "             show_confidence=False, \n",
    "             label_func = lambda x: '',\n",
    "             #label_func = lambda x: \"  {}\".format(metadata_gdf[\"country\"][x.name]) if x.is_terminal() else '',\n",
    "             #label_func = lambda x: x.name if not x.is_terminal() else '',\n",
    "             do_show=False,\n",
    "            )\n",
    "# --------------------------------------------\n",
    "# Ticks\n",
    "ticks_dict = convert_timetree_ticks(tree=subtree, step=100)\n",
    "# Set the new tick locations and labels\n",
    "# Leave out last one\n",
    "ax.set_xticks(ticks_dict[\"tick_locs\"][:])\n",
    "ax.set_xticklabels(ticks_dict[\"tick_labels\"][:], ROTATION=45, ha=\"right\")\n",
    "date_to_x = ticks_dict[\"tick_locs\"][0] - float(ticks_dict[\"tick_labels\"][0])\n",
    "\n",
    "# --------------------------------------------\n",
    "# Lines\n",
    "# Vertical lines for dates\n",
    "for loc in ticks_dict[\"tick_locs\"][:]:\n",
    "    ax.axvline(x=loc, color=\"darkgrey\", alpha=0.5, lw=0.25)\n",
    "    \n",
    "ax.set_title(\"Node Dating\", fontsize=FONTSIZE * 2)\n",
    "ax.set_xlabel(\"Date\")\n",
    "\n",
    "# --------------------------------------------\n",
    "# Interanl Locations\n",
    "timetree_date_err = [[],[]]\n",
    "for lower,upper in zip(subtree_internals[\"timetree_CI_lower_err\"], subtree_internals[\"timetree_CI_upper_err\"] ):\n",
    "    timetree_date_err[0].append(lower)\n",
    "    timetree_date_err[1].append(upper)\n",
    "    \n",
    "ax.errorbar(\n",
    "    data=subtree_internals, \n",
    "    x=[d + date_to_x for d in subtree_internals[\"timetree_date_calendar\"]],\n",
    "    y=[i+0.1 for i in range(1,len(subtree_internals)+1,1)], \n",
    "    yerr=None, \n",
    "    xerr=timetree_date_err,\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=2,\n",
    "    zorder=2,\n",
    "    lw=1, \n",
    "    color=\"darkgoldenrod\",\n",
    "    marker=\"o\",\n",
    "    markersize=5,\n",
    "    markerfacecolor=subtree_tips[\"branch_major_color\"][0], \n",
    ")\n",
    "\n",
    "ax.set_ylabel(\"\")\n",
    "\n",
    "out_path = os.path.join(out_dir, \"dating_1.PRE\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "still-carrier",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"       \n",
    "#----------------------------------------------------\n",
    "# Actual Date\n",
    "ax_plot.errorbar(\n",
    "    data=subtree_tips, \n",
    "    x=\"date_mean\",  \n",
    "    y=[float(i) for i in range(1,len(subtree_tips)+1,1)],  \n",
    "    #y=[i for i in range(len(subtree_tips)+1,1,-1)],    \n",
    "    yerr=None, \n",
    "    xerr=\"date_err\",\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=1,\n",
    "    label=\"Archaeological Date\",\n",
    "    zorder=1,\n",
    "    lw=0.5, \n",
    "    color=\"black\", \n",
    "    #c=subtree_tips[\"branch_major_color\"][0],    \n",
    ")\n",
    "\n",
    "\n",
    "sns.scatterplot(\n",
    "    ax=ax,\n",
    "    data=subtree_tips,\n",
    "    x=\"date_mean\", \n",
    "    #y=\"timetree_coord_y\", \n",
    "    y=[i for i in range(0,len(subtree_tips)*data_spacing,data_spacing)],\n",
    "    color=\"black\",\n",
    "    #hue=subtree_tips[\"country\"],\n",
    "    #c=subtree_tips[\"branch_major_color\"],\n",
    "    s=5)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "timetree_date_err = [[],[]]\n",
    "\n",
    "for lower,upper in zip(subtree_tips[\"timetree_CI_lower_err\"], subtree_tips[\"timetree_CI_upper_err\"] ):\n",
    "    timetree_date_err[0].append(lower)\n",
    "    timetree_date_err[1].append(upper)\n",
    "\n",
    "sns.scatterplot(\n",
    "    ax=ax,\n",
    "    data=subtree_tips,\n",
    "    x=\"timetree_date_calendar\",\n",
    "    #y=\"timetree_coord_y\", \n",
    "    y=[i for i in range(3,len(subtree_tips)*data_spacing,data_spacing)],    \n",
    "    color=\"red\",\n",
    "    #hue=subtree_tips[\"country\"],\n",
    "    #c=subtree_tips[\"branch_major_color\"],\n",
    "    s=10) \n",
    "\n",
    "\n",
    "ax.errorbar(\n",
    "    data=subtree_tips, \n",
    "    x=\"timetree_date_calendar\", \n",
    "    #y=[i for i in range(4,len(subtree_tips)*data_spacing,data_spacing)],\n",
    "    y=\"timetree_coord_y\",\n",
    "    yerr=None, \n",
    "    xerr=timetree_date_err,\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=1,\n",
    "    label=\"Estimated Date\",\n",
    "    zorder=1,\n",
    "    lw=0.5, \n",
    "    color=\"red\", \n",
    "    #c=subtree_tips[\"branch_major_color\"][0],    \n",
    ")\n",
    "\n",
    "\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_yticklabels(list(subtree_tips[\"country\"]), fontsize=FONTSIZE)\n",
    "ax.set_xlabel(\"Date\")\n",
    "ax.set_title(\"Leave-One-Out Tip Dating (1.PRE)\")\n",
    "ax.legend(bbox_to_anchor=(1,1), fontsize=FONTSIZE)\n",
    "\n",
    "#display(subtree_tips)\n",
    "\n",
    "out_path = os.path.join(out_dir, \"leave-one-out\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "passive-tissue",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(subtree_tips)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "confidential-essence",
   "metadata": {},
   "source": [
    "---\n",
    "# Geography"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prepared-import",
   "metadata": {},
   "source": [
    "## Isolation By Distance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "growing-plasma",
   "metadata": {},
   "source": [
    "### Prep Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "northern-government",
   "metadata": {},
   "outputs": [],
   "source": [
    "ibd_dict = {}\n",
    "\n",
    "for label in BRANCH_LIST:\n",
    "    print(\"Label:\", label)    \n",
    "    ibd_dict[label] = {\"geo_dist\" : [], \"genetic_dist\" : []}\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    \n",
    "    i_compare = 0\n",
    "    i = 0\n",
    "    for sample1 in df.index:\n",
    "        geom1 = df[\"geometry\"][sample1]\n",
    "        coord1 = (geom1.y, geom1.x)  \n",
    "        for sample2 in df.index[i_compare:]:\n",
    "            if sample1 == sample2: continue\n",
    "            geom2 = df[\"geometry\"][sample2]\n",
    "            coord2 = (geom2.y, geom2.x)\n",
    "            geo_dist = geopy.distance.great_circle(coord1, coord2).km\n",
    "            genetic_dist = divtree.distance(sample1, sample2)\n",
    "\n",
    "            ibd_dict[label][\"geo_dist\"].append(geo_dist)\n",
    "            ibd_dict[label][\"genetic_dist\"].append(genetic_dist)\n",
    "        i_compare += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "handy-harris",
   "metadata": {},
   "source": [
    "### All"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suited-ratio",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [480, 480]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "geo_dist_all = []\n",
    "genetic_dist_all = []\n",
    "\n",
    "fig, ax = plt.subplots(1,figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "for branch in ibd_dict:\n",
    "    geo_dist_all = geo_dist_all + ibd_dict[branch][\"geo_dist\"]\n",
    "    genetic_dist_all = genetic_dist_all + ibd_dict[branch][\"genetic_dist\"]\n",
    "\n",
    "sns.regplot(\n",
    "    ax=ax,\n",
    "    x=geo_dist_all,\n",
    "    y=genetic_dist_all,\n",
    "    ci=95,\n",
    "    scatter_kws={\"s\": 0},\n",
    "    color=\"grey\",\n",
    "    line_kws={\"linewidth\":0.5},\n",
    "    label=(\n",
    "    \"   RÂ²: {}\".format(round(r2,2))\n",
    "    + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)             \n",
    "    )\n",
    ")\n",
    "    \n",
    "sns.scatterplot(\n",
    "    ax=ax,\n",
    "    x=geo_dist_all,\n",
    "    y=genetic_dist_all,\n",
    "    s=1,\n",
    "    ec = \"white\", \n",
    "    color=\"black\",\n",
    "    alpha=0.75,\n",
    "    zorder=2,\n",
    ")\n",
    "\n",
    "# Format and position legend\n",
    "legend = ax.legend(\n",
    "    borderpad=0.8, \n",
    "    handletextpad=-2, \n",
    "    edgecolor=\"black\", \n",
    "    #bbox_to_anchor=(0.5, -0.30), \n",
    "    #loc='center',\n",
    "    fontsize=FONTSIZE,\n",
    ")\n",
    "frame = legend.get_frame().set_linewidth(0.5)\n",
    "\n",
    "ax.set_ylabel(\"Genetic Distance (subs / site)\")\n",
    "ax.set_xlabel(\"Great Circle Distance (km)\")\n",
    "ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "\n",
    "out_path = os.path.join(out_dir, \"ibd_all\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "manual-merchant",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "nrow=3\n",
    "ncol=4\n",
    "\n",
    "fig, axes = plt.subplots(nrow,ncol,figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.35, hspace=0.75)\n",
    "\n",
    "\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for label in BRANCH_LIST:\n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    \n",
    "    for branch_major in colors_dict[\"branch_major\"]:\n",
    "        if branch_major in label:\n",
    "            break  \n",
    "            \n",
    "    # -----------------------\n",
    "    # Regression\n",
    "    slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(ibd_dict[label][\"geo_dist\"],ibd_dict[label][\"genetic_dist\"])\n",
    "    r2 = r_value * r_value    \n",
    "    y_intercept = CURRENT_YEAR - (0 - intercept)/slope\n",
    "    p_sig = \"\"\n",
    "    p_thresh = 0.05    \n",
    "    if p_value < p_thresh:\n",
    "        p_sig = \"*\"\n",
    "        \n",
    "    mut_per_year = SEQ_LEN * slope\n",
    "    year_per_mut = 1 / mut_per_year\n",
    "    \n",
    "    reg_dict = {\n",
    "        \"slope\": slope,\n",
    "        \"x_intercept\" : intercept,\n",
    "        \"y_intercept\" : y_intercept,\n",
    "        \"p_value\" : p_value,\n",
    "        \"p_sig\" : p_sig,\n",
    "        \"year_per_mut\" : year_per_mut\n",
    "    }\n",
    "    \n",
    "    sns.regplot(\n",
    "        ax=ax,\n",
    "        x=ibd_dict[label][\"geo_dist\"],\n",
    "        y=ibd_dict[label][\"genetic_dist\"],\n",
    "        ci=95,\n",
    "        scatter_kws={\"s\": 0},\n",
    "        color=\"grey\",\n",
    "        line_kws={\"linewidth\":0.5},\n",
    "        label=(\n",
    "        \"   RÂ²: {}\".format(round(r2,2))\n",
    "        + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)             \n",
    "        )\n",
    "    )\n",
    "\n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        x=ibd_dict[label][\"geo_dist\"],\n",
    "        y=ibd_dict[label][\"genetic_dist\"],\n",
    "        s=10,\n",
    "        ec = \"black\",\n",
    "        color = colors_dict[\"branch_major\"][branch_major],         \n",
    "        alpha=0.75,\n",
    "        zorder=2,\n",
    "    )\n",
    "     \n",
    "    # Format and position legend\n",
    "    legend = ax.legend(\n",
    "        borderpad=0.8, \n",
    "        handletextpad=-2, \n",
    "        edgecolor=\"black\", \n",
    "        bbox_to_anchor=(0.5, -0.40), \n",
    "        loc='center',\n",
    "        fontsize=FONTSIZE * 0.80,\n",
    "    )\n",
    "    frame = legend.get_frame().set_linewidth(0.5)        \n",
    "    \n",
    "    # Set xlimits\n",
    "    xlim = ax.get_xlim()\n",
    "    if xlim[1] > 20000:\n",
    "        xbuff = 5000   \n",
    "    elif xlim[1] > 2000:\n",
    "        xbuff = 500\n",
    "    elif xlim[1] > 200:\n",
    "        xbuff = 50\n",
    "    ax.set_xlim(0-xbuff, xlim[1] + xbuff)\n",
    "        \n",
    "    # Format axis\n",
    "    ax.set_title(label)\n",
    "\n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Genetic Distance (subs / site)\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Great Circle Distance (km)\")\n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "    \n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_linewidth(0.5)    \n",
    "            \n",
    "\n",
    "    # Update axis\n",
    "    if i_col == ncol - 1:\n",
    "        i_col = 0\n",
    "        i_row += 1\n",
    "    else:\n",
    "        i_col += 1\n",
    "\n",
    "    \n",
    "out_path = os.path.join(out_dir, \"ibd_clades\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prescribed-stranger",
   "metadata": {},
   "source": [
    "## Map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eight-cover",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "nrow=3\n",
    "ncol=4\n",
    "\n",
    "fig, axes = plt.subplots(nrow, ncol, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.1, hspace=-0.75)\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for label in BRANCH_LIST:\n",
    "    \n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    \n",
    "    for branch_major in colors_dict[\"branch_major\"]:\n",
    "        if branch_major in label:\n",
    "            break  \n",
    "    \n",
    "    # --------------------------------------\n",
    "    # World Basemap\n",
    "    world_polygons.plot(\n",
    "        ax=ax, \n",
    "        zorder=1, \n",
    "        alpha=0.5, \n",
    "        color=\"grey\",\n",
    "        edgecolor=\"black\", \n",
    "        linewidth=0.1,\n",
    "    ) \n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    df.plot(ax=ax, ec=\"black\", lw=0.5, color=colors_dict[\"branch_major\"][branch_major], markersize=df[\"geometry_size\"], linewidth=0.5)\n",
    "\n",
    "    # Sorry Antarctica\n",
    "    ax.set_ylim(-60,90)\n",
    "    \n",
    "    ax.set_title(label)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_linewidth(0.5)        \n",
    "    \n",
    "    # Update axis\n",
    "    if i_col == ncol - 1:\n",
    "        i_col = 0\n",
    "        i_row += 1\n",
    "    else:\n",
    "        i_col += 1\n",
    "        \n",
    "out_path = os.path.join(out_dir, \"map_clades\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cultural-template",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=400\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "\n",
    "## All\n",
    "fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=FIGSIZE, dpi=DPI)\n",
    "world_polygons.plot(\n",
    "    ax=ax, \n",
    "    zorder=1, \n",
    "    alpha=0.5, \n",
    "    color=\"grey\",\n",
    "    edgecolor=\"white\", \n",
    "    linewidth=0.1,\n",
    ") \n",
    "\n",
    "reduce_gdf = copy.deepcopy(metadata_gdf)\n",
    "\n",
    "plotted_coord = []\n",
    "\n",
    "for rec in metadata_gdf.iterrows():\n",
    "    sample=rec[0]\n",
    "    x,y = rec[1][\"geometry\"].x, rec[1][\"geometry\"].y\n",
    "    if [x,y] not in plotted_coord:\n",
    "        plotted_coord.append([x,y])\n",
    "    else:\n",
    "        reduce_gdf.drop(sample, inplace=True)\n",
    "\n",
    "reduce_gdf.plot(\n",
    "    ax=ax, \n",
    "    ec=\"black\", \n",
    "    color=reduce_gdf[\"branch_major_color\"], \n",
    "    markersize=reduce_gdf[\"geometry_size\"], \n",
    "    linewidth=0.25,\n",
    "    alpha=0.80,\n",
    ")\n",
    "\n",
    "#ax.axis(\"off\")\n",
    "for spine in ax.spines:\n",
    "    ax.spines[spine].set_linewidth(0.25)\n",
    "\n",
    "# Sorry Antarctica\n",
    "ax.set_ylim(-60,90)\n",
    "\n",
    "colors_dict_known = copy.copy(colors_dict)\n",
    "del colors_dict_known[\"branch_major\"][UNKNOWN_CHAR]\n",
    "\n",
    "clade_lines = [lines.Line2D([0], [0], color=colors_dict_known[\"branch_major\"][state], lw=0, markersize=2, markeredgecolor=\"black\", markeredgewidth=0.25, marker=\"o\") \n",
    "                for state in colors_dict_known[\"branch_major\"]]\n",
    "clade_labels = [state for state in colors_dict_known[\"branch_major\"]]\n",
    "\n",
    "clade_legend = fig.legend(clade_lines, \n",
    "           clade_labels,\n",
    "           #ncol=len(clade_labels),\n",
    "           ncol=1,\n",
    "           loc=\"center\", \n",
    "           bbox_to_anchor=(0.2, 0.475),\n",
    "           frameon=True,\n",
    "            borderpad=0.5, \n",
    "            handletextpad=0.25, \n",
    "           labelspacing=0.5,\n",
    "           edgecolor=\"black\",\n",
    "           title=\"Clade\",\n",
    "           fontsize=FONTSIZE / 2,\n",
    "           title_fontsize=FONTSIZE / 1.5,\n",
    "          )\n",
    "clade_legend.get_frame().set_linewidth(0.25)\n",
    "\n",
    "# Size Legend\n",
    "s = [ax.scatter(0,-9000, s=1, ec=\"black\", c=\"white\", lw=0.25),\n",
    "     ax.scatter(0,-9000, s=10, ec=\"black\", c=\"white\", lw=0.25),  \n",
    "    ]\n",
    "sl = [\"1\", \"10\"]\n",
    "leg_size = plt.legend(s, sl, title=\"Genomes\", loc=\"center\",  borderpad=0.5, bbox_to_anchor=[0.2,0.35], edgecolor=\"black\",\n",
    "                    fontsize=FONTSIZE / 2, title_fontsize=FONTSIZE / 1.5)\n",
    "leg_size.get_frame().set_linewidth(0.25)\n",
    "\n",
    "out_path = os.path.join(out_dir, \"map_all\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "following-gauge",
   "metadata": {},
   "source": [
    "## Modern Human Plague Cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "advisory-franchise",
   "metadata": {},
   "outputs": [],
   "source": [
    "TARGET_RES = [1280, 1280]\n",
    "DPI=400\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "plt.rc(\"legend\", title_fontsize=MED_FONT)  # fontsize of the legend title\n",
    "plt.rc(\"legend\", fontsize=MED_FONT)  # fontsize of the legend title\n",
    "\n",
    "fig, ax = plt.subplots(1, dpi=DPI, figsize=FIGSIZE)\n",
    "# --------------------------------------\n",
    "# World Basemap\n",
    "world_poly = world_polygons.plot(\n",
    "    ax=ax, \n",
    "    zorder=1, \n",
    "    alpha=0.5, \n",
    "    color=\"grey\",\n",
    "    edgecolor=\"white\", \n",
    "    linewidth=0.25,\n",
    ")\n",
    "xu_2019_gdf.plot(ax=ax,  markersize=0.5, alpha=0.75, ec=\"black\", lw=0.1, c=D3_PAL[2])\n",
    "#ancient_gdf.plot(ax=ax_map,  markersize=ancient_gdf[\"geometry_size\"], alpha=0.75, ec=\"black\", lw=0.25, c=D3_PAL[1])\n",
    "#modern_gdf.plot(ax=ax_map,  markersize=modern_gdf[\"geometry_size\"], alpha=0.75, ec=\"black\", lw=0.25, c=D3_PAL[0])\n",
    "# Sorry Antarctica\n",
    "ax.set_ylim(-60,90)\n",
    "#ax_map.axis('off')  \n",
    "for spine in ax.spines:\n",
    "    ax.spines[spine].set_linewidth(0.25)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "\n",
    "out_path = os.path.join(out_dir, \"map_xu-2019\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\n",
    "\"\"\"\n",
    "# Origin Legend\n",
    "o = [ax_map.scatter(0,-9000, s=5, ec=\"black\", marker=\"o\", c=D3_PAL[1], lw=0.25),\n",
    "     ax_map.scatter(0,-9000, s=5, ec=\"black\", marker=\"o\", c=D3_PAL[0], lw=0.25),\n",
    "     ax_map.scatter(0,-9000, s=5, ec=\"black\", marker=\"o\", c=D3_PAL[2], lw=0.25),     \n",
    "    ]\n",
    "ol = [\"Ancient Genome\", \"Modern Genome\", \"Modern Human Cases\"]\n",
    "\n",
    "leg_origin = plt.legend(\n",
    "    o, ol, title=\"Type\", loc=\"center\", bbox_to_anchor=[0.35,-0.15], edgecolor=\"black\",\n",
    "    fontsize=FONTSIZE / 2, title_fontsize=FONTSIZE - 2)\n",
    "leg_origin.get_frame().set_linewidth(0.25)\n",
    "\n",
    "# Size Legend\n",
    "s = [ax_map.scatter(0,-9000, s=1, ec=\"black\", c=\"black\", lw=0.5),\n",
    "     ax_map.scatter(0,-9000, s=10, ec=\"black\", c=\"black\", lw=0.5),  \n",
    "    ]\n",
    "sl = [\"1\", \"10\"]\n",
    "leg_size = plt.legend(s, sl, title=\"Genomes\", loc=\"center\", bbox_to_anchor=[0.55,-0.15], edgecolor=\"black\",\n",
    "                    fontsize=FONTSIZE / 2, title_fontsize=FONTSIZE - 2)\n",
    "leg_size.get_frame().set_linewidth(0.25)\n",
    "\n",
    "ax_map.add_artist(leg_origin)\n",
    "\n",
    "out_path = os.path.join(out_dir, \"map_comparison\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
