{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "resident-export",
   "metadata": {},
   "source": [
    "# Paper Figures\n",
    "\n",
    "1. Setup\n",
    "1. Import\n",
    "  - Metadata\n",
    "  - Trees\n",
    "    - Divergence tree\n",
    "    - Timetree\n",
    "    - Mugration Trees\n",
    "1. Merge Metadata\n",
    "1. Export\n",
    "  - Timetree\n",
    "  - Divtree\n",
    "  - Augur and Auspice JSON"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "careful-ordering",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bronze-wealth",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "proof-ultimate",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import Phylo, AlignIO\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors, gridspec, lines\n",
    "import scipy\n",
    "import math\n",
    "import numpy as np\n",
    "import geopandas\n",
    "import datetime\n",
    "import shapely\n",
    "import copy\n",
    "from geopy import distance\n",
    "import geopy\n",
    "import cartopy.crs as ccrs\n",
    "from augur import utils, export_v2\n",
    "\n",
    "from plot_phylo_functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "violent-vampire",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "noted-savannah",
   "metadata": {},
   "outputs": [],
   "source": [
    "#AUSPICE_PREFIX=\"plague-phylogeography-projects_main\"\n",
    "\n",
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"full\", \"5\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "PRUNE = WILDCARDS[2]\n",
    "MISSING_DATA = WILDCARDS[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "proud-territory",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "divtree: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/iqtree/all/chromosome/full/filter5/filter-taxa/iqtree.treefile\n",
      "timetree: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/lsd/all/chromosome/full/filter5/lsd.timetree.nex\n"
     ]
    }
   ],
   "source": [
    "iqtree_dir    = os.path.join(project_dir, \"iqtree/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "lsd_dir       = os.path.join(project_dir, \"lsd/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "mugration_dir = os.path.join(project_dir, \"mugration/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "snippy_dir    = os.path.join(project_dir, \"snippy_multi/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "\n",
    "# ------------------------------------------\n",
    "# Trees\n",
    "divtree_path              = iqtree_dir    + \"filter-taxa/iqtree.treefile\"\n",
    "timetree_path             = lsd_dir       + \"lsd.timetree.nex\"\n",
    "#timetree_path             = lsd_dir       + \"lsd.nex\"\n",
    "mugtree_country_path      = mugration_dir + \"country.nex\".format(MISSING_DATA)\n",
    "mugtree_province_path     = mugration_dir + \"province.nex\".format(MISSING_DATA)\n",
    "mugtree_branch_major_path = mugration_dir + \"branch_major.nex\".format(MISSING_DATA)\n",
    "mugtree_branch_minor_path = mugration_dir + \"branch_minor.nex\".format(MISSING_DATA)\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "constant_sites_path   = project_dir + \"snippy_multi/all/chromosome/full/snippy-multi.constant_sites.txt\"\n",
    "aln_path              = iqtree_dir + \"filter-sites/snippy-multi.snps.aln\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Metadata\n",
    "auspice_config_path          = project_dir + \"config/auspice_config.json\"\n",
    "metadata_path                = iqtree_dir    + \"filter-taxa/metadata.tsv\"\n",
    "mugconf_country_path         = mugration_dir + \"country_confidence.csv\"\n",
    "mugstates_country_path       = mugration_dir + \"country_states.csv\"\n",
    "mugconf_province_path        = mugration_dir + \"province_confidence.csv\"\n",
    "mugstates_province_path      = mugration_dir + \"province_states.csv\"\n",
    "mugconf_branch_major_path    = mugration_dir + \"branch_major_confidence.csv\"\n",
    "mugstates_branch_major_path  = mugration_dir + \"branch_major_states.csv\"\n",
    "mugconf_branch_minor_path    = mugration_dir + \"branch_minor_confidence.csv\"\n",
    "mugstates_branch_minor_path  = mugration_dir + \"branch_minor_states.csv\"\n",
    "# ------------------------------------------\n",
    "# Other\n",
    "xu_2019_path          = project_dir + \"figures/xu2019.csv\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Output\n",
    "out_dir = os.path.join(project_dir, \"auspice/all/chromosome/{}/filter{}/\".format(PRUNE, MISSING_DATA))\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "collaborative-rescue",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "exact-blood",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 10, \"display.max_columns\", None)\n",
    "\n",
    "NO_DATA_CHAR = \"NA\"\n",
    "UNKNOWN_CHAR = \"?\"\n",
    "JSON_INDENT=2\n",
    "CURRENT_YEAR = datetime.datetime.utcnow().year\n",
    "\n",
    "# ------------------------------------------\n",
    "# Alignment\n",
    "with open(constant_sites_path) as infile:\n",
    "    data = infile.read().strip().split(\",\")\n",
    "    constant_sites = sum([int(count) for count in data])\n",
    "\n",
    "aln = AlignIO.read(aln_path, \"fasta\")\n",
    "variant_sites = len(aln[0].seq)\n",
    "SEQ_LEN = constant_sites + variant_sites\n",
    "\n",
    "# ------------------------------------------\n",
    "# Geo\n",
    "CRS = \"epsg:3857\"\n",
    "CRS_EPSG = ccrs.epsg('3857')\n",
    "world_polygons = geopandas.read_file(geopandas.datasets.get_path(\"naturalearth_lowres\"))\n",
    "\n",
    "# ------------------------------------------\n",
    "# Reference\n",
    "REF_STRAIN = \"CO92\"\n",
    "REF_LAT = \"38.7251776\"\n",
    "REF_LON = \"-105.607716\"\n",
    "REF_DATE_BP = 1992 - CURRENT_YEAR\n",
    "REF_BRANCH_MAJOR = \"1.ORI\"\n",
    "REF_BRANCH_MINOR = \"1.ORI1\"\n",
    "REF_BRANCH_NUMBER = \"1\"\n",
    "REF_COUNTRY = \"United States of America\"\n",
    "REF_PROVINCE = \"Colorado\"\n",
    "REF_CONTINENT = \"North America\"\n",
    "\n",
    "# ------------------------------------------\n",
    "# Plotting\n",
    "SM_FONT = 4\n",
    "MED_FONT = 6\n",
    "LG_FONT = 8\n",
    "D3_PAL = [\"#1f77b4\", \"#ff7f0e\", \"#2ca02c\", \"#d62728\", \"#9467bd\", \"#8c564b\", \"#e377c2\", \"#7f7f7f\", \"#bcbd22\", \"#17becf\" ]\n",
    "\n",
    "# ------------------------------------------\n",
    "BRANCH_LIST = {\n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"], \n",
    "    \"0.ANT4\" : [\"0.ANT4\"], \n",
    "    \"0.PE\": [\"0.PE4m\", \"0.PE4m\", \"0.PE4t\", \"0.PE4a\", \"0.PE5\"],   \n",
    "    \"0.ANT\": [\"0.ANT1\", \"0.ANT2\",\"0.ANT3\",\"0.ANT5\"],     \n",
    "    \"1.PRE\" : [\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],    \n",
    "    \"1.ANT\": [\"1.ANT1\"], \n",
    "    \"1.IN\": [\"1.IN1\",\"1.IN2\",\"1.IN3\"],   \n",
    "    \"1.ORI\" : [\"1.ORI1\", \"1.ORI2\", \"1.ORI3\"],\n",
    "    \"2.ANT\": [\"2.ANT1\",\"2.ANT2\",\"2.ANT3\" ], \n",
    "    \"2.MED\": [\"2.MED0\", \"2.MED1\",\"2.MED2\",\"2.MED3\" ], \n",
    "    \"3.ANT\": [\"3.ANT1\", \"3.ANT2\" ],     \n",
    "    \"4.ANT\": [\"4.ANT1\" ],           \n",
    "}\n",
    "\n",
    "ANCIENT_BRANCH_LIST = {\n",
    "    \"0.PRE\": [\"0.PRE1\", \"0.PRE2\"], \n",
    "    \"0.ANT4\" : [\"0.ANT4\"],  \n",
    "    \"1.PRE\" : [\"1.PRE1\", \"1.PRE2\", \"1.PRE3\"],     \n",
    "}\n",
    "\n",
    "MUG_ATTRIBUTE_LIST = [\n",
    "    \"branch_major\",\n",
    "    \"branch_minor\",\n",
    "    \"country\",\n",
    "    \"province\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nonprofit-cedar",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. Import"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "processed-mitchell",
   "metadata": {},
   "source": [
    "## Mugration Confidence Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "operational-investigation",
   "metadata": {},
   "outputs": [],
   "source": [
    "states_files = [mugstates_country_path, mugstates_province_path, mugstates_branch_major_path, mugstates_branch_minor_path]\n",
    "conf_files = [mugconf_country_path, mugconf_province_path, mugconf_branch_major_path, mugconf_branch_minor_path]\n",
    "\n",
    "# ----------------------------------------\n",
    "# Convert character states to numbers\n",
    "for mugstates,mugconf in zip(states_files, conf_files):\n",
    "    node_i = 0\n",
    "    conf_basename = os.path.splitext(mugconf)[0]\n",
    "    conf_outname = conf_basename + \"_tidy.csv\"\n",
    "    i = 0\n",
    "    state_dict = {}\n",
    "    with open(mugstates, \"r\") as infile:\n",
    "        read_line = infile.readline()\n",
    "        while read_line:\n",
    "            split_line = read_line.strip().split(\",\")\n",
    "            state = split_line[1]\n",
    "            if state == \"nan\":\n",
    "                state = \"NA\"\n",
    "            if state == UNKNOWN_CHAR:\n",
    "                state_dict[-1] = state\n",
    "            else:\n",
    "                state_dict[i] = state\n",
    "                i += 1\n",
    "            read_line = infile.readline()\n",
    "    # Swap first to last place\n",
    "    final_i = len(state_dict) - 1   \n",
    "    state_dict[final_i] = \"?\"\n",
    "    # Remove old first\n",
    "    state_dict.pop(-1)\n",
    "\n",
    "    with open(mugconf, \"r\") as infile:            \n",
    "        i = 0\n",
    "        read_line = infile.readline()\n",
    "        header = read_line.strip().split(\", \")\n",
    "        header[0] = \"name\"\n",
    "\n",
    "        data = \"\"\n",
    "        for i in range(0, len(header) - 1):\n",
    "            char = header[i + 1]\n",
    "            header[i + 1] = state_dict[i]\n",
    "        header = \",\".join(header)\n",
    "        read_line = infile.readline()            \n",
    "        while read_line:\n",
    "            # Rename internal nodes\n",
    "            split_line = read_line.split(\", \")\n",
    "            node_name = split_line[0]\n",
    "            if \"NODE\" in node_name:\n",
    "                node_name = \"NODE{}\".format(node_i)\n",
    "                node_i += 1\n",
    "            new_line = \",\".join([node_name] + split_line[1:])\n",
    "            data += new_line\n",
    "            read_line = infile.readline()\n",
    "    \n",
    "    with open(conf_outname, \"w\") as outfile:\n",
    "        outfile.write(header + \"\\n\")\n",
    "        outfile.write(data)\n",
    "\n",
    "# ----------------------------------------\n",
    "# Import the tidy files as dataframes\n",
    "mugconf_country_tidy_path = os.path.splitext(mugconf_country_path)[0] + \"_tidy.csv\"\n",
    "mugconf_country_df = pd.read_csv(mugconf_country_tidy_path, sep=',', index_col = 0)\n",
    "\n",
    "mugconf_province_tidy_path = os.path.splitext(mugconf_province_path)[0] + \"_tidy.csv\"\n",
    "mugconf_province_df = pd.read_csv(mugconf_province_tidy_path, sep=',', index_col = 0)\n",
    "\n",
    "mugconf_branch_major_tidy_path = os.path.splitext(mugconf_branch_major_path)[0] + \"_tidy.csv\"\n",
    "mugconf_branch_major_df = pd.read_csv(mugconf_branch_major_tidy_path, sep=',', index_col = 0)\n",
    "\n",
    "mugconf_branch_minor_tidy_path = os.path.splitext(mugconf_branch_minor_path)[0] + \"_tidy.csv\"\n",
    "mugconf_branch_minor_df = pd.read_csv(mugconf_branch_minor_tidy_path, sep=',', index_col = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bottom-thanks",
   "metadata": {},
   "source": [
    "## Trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "loving-rochester",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "divtree: iqtree.treefile\n",
      "timetree: lsd.timetree.nex\n",
      "mugtree_country: country.nex\n",
      "mugtree_province: province.nex\n",
      "mugtree_branch_major: branch_major.nex\n",
      "mugtree_branch_minor: branch_minor.nex\n"
     ]
    }
   ],
   "source": [
    "tree_paths = [divtree_path, timetree_path, mugtree_country_path, mugtree_province_path, mugtree_branch_major_path, mugtree_branch_minor_path  ]\n",
    "\n",
    "for tree_path in tree_paths:\n",
    "    tree_ext = os.path.splitext(tree_path)[1]\n",
    "    if tree_ext == \".nwk\" or tree_ext == \".newick\" or tree_ext == \".treefile\":\n",
    "        tree = Phylo.read(tree_path, \"newick\")\n",
    "    elif tree_ext == \".nex\" or tree_ext == \".nexus\":\n",
    "        parse_trees = Phylo.parse(tree_path, \"nexus\")\n",
    "        for parse_tree in parse_trees:\n",
    "            clades = [c for c in parse_tree.find_clades()]\n",
    "            if len(clades) > 1:\n",
    "                break\n",
    "        tree = parse_tree\n",
    "    tree.ladderize(reverse=False)\n",
    "    \n",
    "    if tree_path == divtree_path:\n",
    "        print(\"divtree:\", os.path.basename(divtree_path))\n",
    "        divtree = tree\n",
    "    elif tree_path == timetree_path:\n",
    "        print(\"timetree:\", os.path.basename(timetree_path))        \n",
    "        timetree = tree    \n",
    "    elif tree_path == mugtree_country_path:\n",
    "        print(\"mugtree_country:\", os.path.basename(mugtree_country_path))       \n",
    "        mugtree_country = tree    \n",
    "    elif tree_path == mugtree_province_path:\n",
    "        print(\"mugtree_province:\",os.path.basename(mugtree_province_path))\n",
    "        mugtree_province = tree  \n",
    "    elif tree_path == mugtree_branch_major_path:\n",
    "        print(\"mugtree_branch_major:\",os.path.basename(mugtree_branch_major_path))        \n",
    "        mugtree_branch_major = tree            \n",
    "    elif tree_path == mugtree_branch_minor_path:\n",
    "        print(\"mugtree_branch_minor:\",os.path.basename(mugtree_branch_minor_path))        \n",
    "        mugtree_branch_minor = tree                \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "falling-character",
   "metadata": {},
   "source": [
    "### Modern Human Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "leading-collector",
   "metadata": {},
   "outputs": [],
   "source": [
    "#xu_2019_df = pd.read_csv(xu_2019_path, sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "emerging-waters",
   "metadata": {},
   "source": [
    "### Project Genomic Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "wooden-patio",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "metadata_df.set_index(metadata_df.columns[0], inplace=True)\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "behind-spice",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Merge Metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "intense-election",
   "metadata": {},
   "source": [
    "### Tree Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "unknown-recipe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>branch_support_char</th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>timetree_date</th>\n",
       "      <th>timetree_CI_height</th>\n",
       "      <th>timetree_CI_date</th>\n",
       "      <th>mugration_country</th>\n",
       "      <th>mugration_province</th>\n",
       "      <th>mugration_branch_major</th>\n",
       "      <th>mugration_branch_minor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>6.334000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-29</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009670865.1_ASM967086v1_genomic</th>\n",
       "      <td>42012</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>-50</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722853</td>\n",
       "      <td>KEEP: Assembly Modern Placement</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1971</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>1.078700e-05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-50</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630435.1_ASM863043v1_genomic</th>\n",
       "      <td>C-719</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>-25</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED0</td>\n",
       "      <td>SAMN12721146</td>\n",
       "      <td>KEEP: Assembly Modern Placement</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1996</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.417700e-05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-25</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000169635.1_ASM16963v1_genomic</th>\n",
       "      <td>MG05-1020</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>NA</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>SAMN02404403</td>\n",
       "      <td>KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...</td>\n",
       "      <td>1</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.927000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000170275.1_ASM17027v1_genomic</th>\n",
       "      <td>F1991016</td>\n",
       "      <td>1991.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>SAMN02404399</td>\n",
       "      <td>KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...</td>\n",
       "      <td>1</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1991</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>5.396000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE22</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>7.030000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-697.117</td>\n",
       "      <td>{138.712,50502.4}</td>\n",
       "      <td>{-50502.4,-138.712}</td>\n",
       "      <td>China</td>\n",
       "      <td>Gansu</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>1.IN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE23</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>7.040000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-697.117</td>\n",
       "      <td>{138.712,50502.4}</td>\n",
       "      <td>{-50502.4,-138.712}</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>1.IN3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE24</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>5.630000e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE25</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.300000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE26</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.300000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        strain    date date_bp  \\\n",
       "sample                                                           \n",
       "Reference                                 CO92    1992     -29   \n",
       "GCA_009670865.1_ASM967086v1_genomic      42012  1971.0     -50   \n",
       "GCA_008630435.1_ASM863043v1_genomic      C-719  1996.0     -25   \n",
       "GCA_000169635.1_ASM16963v1_genomic   MG05-1020  2005.0     -16   \n",
       "GCA_000170275.1_ASM17027v1_genomic    F1991016  1991.0     -30   \n",
       "...                                        ...     ...     ...   \n",
       "NODE22                                      NA      NA      NA   \n",
       "NODE23                                      NA      NA      NA   \n",
       "NODE24                                      NA      NA      NA   \n",
       "NODE25                                      NA      NA      NA   \n",
       "NODE26                                      NA      NA      NA   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009670865.1_ASM967086v1_genomic                     China   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    Russia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                 Madagascar   \n",
       "GCA_000170275.1_ASM17027v1_genomic                      China   \n",
       "...                                                       ...   \n",
       "NODE22                                                     NA   \n",
       "NODE23                                                     NA   \n",
       "NODE24                                                     NA   \n",
       "NODE25                                                     NA   \n",
       "NODE26                                                     NA   \n",
       "\n",
       "                                                province country_lat  \\\n",
       "sample                                                                 \n",
       "Reference                                       Colorado     39.7837   \n",
       "GCA_009670865.1_ASM967086v1_genomic             Xinjiang     35.0001   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Karachay-Cherkessia     64.6863   \n",
       "GCA_000169635.1_ASM16963v1_genomic                    NA     -18.925   \n",
       "GCA_000170275.1_ASM17027v1_genomic                Yunnan     35.0001   \n",
       "...                                                  ...         ...   \n",
       "NODE22                                                NA          NA   \n",
       "NODE23                                                NA          NA   \n",
       "NODE24                                                NA          NA   \n",
       "NODE25                                                NA          NA   \n",
       "NODE26                                                NA          NA   \n",
       "\n",
       "                                    country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "Reference                              -100.446      38.7252     -105.608   \n",
       "GCA_009670865.1_ASM967086v1_genomic         105      42.4805      85.4633   \n",
       "GCA_008630435.1_ASM863043v1_genomic     97.7453      43.7368      41.7268   \n",
       "GCA_000169635.1_ASM16963v1_genomic      46.4416           NA           NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic          105           25          102   \n",
       "...                                         ...          ...          ...   \n",
       "NODE22                                       NA           NA           NA   \n",
       "NODE23                                       NA           NA           NA   \n",
       "NODE24                                       NA           NA           NA   \n",
       "NODE25                                       NA           NA           NA   \n",
       "NODE26                                       NA           NA           NA   \n",
       "\n",
       "                                         biovar branch_major branch_minor  \\\n",
       "sample                                                                      \n",
       "Reference                            Orientalis        1.ORI       1.ORI1   \n",
       "GCA_009670865.1_ASM967086v1_genomic     Antiqua        0.ANT       0.ANT1   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Medievalis        2.MED       2.MED0   \n",
       "GCA_000169635.1_ASM16963v1_genomic   Orientalis        1.ORI       1.ORI3   \n",
       "GCA_000170275.1_ASM17027v1_genomic   Orientalis        1.ORI       1.ORI2   \n",
       "...                                         ...          ...          ...   \n",
       "NODE22                                       NA           NA           NA   \n",
       "NODE23                                       NA           NA           NA   \n",
       "NODE24                                       NA           NA           NA   \n",
       "NODE25                                       NA           NA           NA   \n",
       "NODE26                                       NA           NA           NA   \n",
       "\n",
       "                                    biosample_accession  \\\n",
       "sample                                                    \n",
       "Reference                                  SAMEA1705942   \n",
       "GCA_009670865.1_ASM967086v1_genomic        SAMN07722853   \n",
       "GCA_008630435.1_ASM863043v1_genomic        SAMN12721146   \n",
       "GCA_000169635.1_ASM16963v1_genomic         SAMN02404403   \n",
       "GCA_000170275.1_ASM17027v1_genomic         SAMN02404399   \n",
       "...                                                 ...   \n",
       "NODE22                                               NA   \n",
       "NODE23                                               NA   \n",
       "NODE24                                               NA   \n",
       "NODE25                                               NA   \n",
       "NODE26                                               NA   \n",
       "\n",
       "                                                                     biosample_comment  \\\n",
       "sample                                                                                   \n",
       "Reference                                              KEEP: Assembly Modern Reference   \n",
       "GCA_009670865.1_ASM967086v1_genomic                    KEEP: Assembly Modern Placement   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    KEEP: Assembly Modern Placement   \n",
       "GCA_000169635.1_ASM16963v1_genomic   KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...   \n",
       "GCA_000170275.1_ASM17027v1_genomic   KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...   \n",
       "...                                                                                ...   \n",
       "NODE22                                                                              NA   \n",
       "NODE23                                                                              NA   \n",
       "NODE24                                                                              NA   \n",
       "NODE25                                                                              NA   \n",
       "NODE26                                                                              NA   \n",
       "\n",
       "                                    branch_number      continent date_mean  \\\n",
       "sample                                                                       \n",
       "Reference                                       1  North America      1992   \n",
       "GCA_009670865.1_ASM967086v1_genomic             0           Asia      1971   \n",
       "GCA_008630435.1_ASM863043v1_genomic             2         Europe      1996   \n",
       "GCA_000169635.1_ASM16963v1_genomic              1         Africa      2005   \n",
       "GCA_000170275.1_ASM17027v1_genomic              1           Asia      1991   \n",
       "...                                           ...            ...       ...   \n",
       "NODE22                                         NA             NA        NA   \n",
       "NODE23                                         NA             NA        NA   \n",
       "NODE24                                         NA             NA        NA   \n",
       "NODE25                                         NA             NA        NA   \n",
       "NODE26                                         NA             NA        NA   \n",
       "\n",
       "                                    date_bp_mean date_err      lat      lon  \\\n",
       "sample                                                                        \n",
       "Reference                                     29        0  38.7252 -105.608   \n",
       "GCA_009670865.1_ASM967086v1_genomic           50        0  42.4805  85.4633   \n",
       "GCA_008630435.1_ASM863043v1_genomic           25        0  43.7368  41.7268   \n",
       "GCA_000169635.1_ASM16963v1_genomic            16        0  -18.925  46.4416   \n",
       "GCA_000170275.1_ASM17027v1_genomic            30        0       25      102   \n",
       "...                                          ...      ...      ...      ...   \n",
       "NODE22                                        NA       NA       NA       NA   \n",
       "NODE23                                        NA       NA       NA       NA   \n",
       "NODE24                                        NA       NA       NA       NA   \n",
       "NODE25                                        NA       NA       NA       NA   \n",
       "NODE26                                        NA       NA       NA       NA   \n",
       "\n",
       "                                    branch_support_char node_type  \\\n",
       "sample                                                              \n",
       "Reference                                                terminal   \n",
       "GCA_009670865.1_ASM967086v1_genomic                      terminal   \n",
       "GCA_008630435.1_ASM863043v1_genomic                      terminal   \n",
       "GCA_000169635.1_ASM16963v1_genomic                       terminal   \n",
       "GCA_000170275.1_ASM17027v1_genomic                       terminal   \n",
       "...                                                 ...       ...   \n",
       "NODE22                                               NA  internal   \n",
       "NODE23                                               NA  internal   \n",
       "NODE24                                               NA  internal   \n",
       "NODE25                                               NA  internal   \n",
       "NODE26                                               NA  internal   \n",
       "\n",
       "                                     branch_length  branch_support  \\\n",
       "sample                                                               \n",
       "Reference                             6.334000e-06           100.0   \n",
       "GCA_009670865.1_ASM967086v1_genomic   1.078700e-05           100.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic   2.417700e-05           100.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic    4.927000e-06           100.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic    5.396000e-06           100.0   \n",
       "...                                            ...             ...   \n",
       "NODE22                                7.030000e-07             0.0   \n",
       "NODE23                                7.040000e-07             0.0   \n",
       "NODE24                                5.630000e-06             0.0   \n",
       "NODE25                                2.300000e-08             0.0   \n",
       "NODE26                                2.300000e-08             0.0   \n",
       "\n",
       "                                    timetree_date timetree_CI_height  \\\n",
       "sample                                                                 \n",
       "Reference                                     -29                 NA   \n",
       "GCA_009670865.1_ASM967086v1_genomic           -50                 NA   \n",
       "GCA_008630435.1_ASM863043v1_genomic           -25                 NA   \n",
       "GCA_000169635.1_ASM16963v1_genomic            -16                 NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic            -30                 NA   \n",
       "...                                           ...                ...   \n",
       "NODE22                                   -697.117  {138.712,50502.4}   \n",
       "NODE23                                   -697.117  {138.712,50502.4}   \n",
       "NODE24                                     -689.7  {129.229,45096.4}   \n",
       "NODE25                                     -689.7  {129.229,45096.4}   \n",
       "NODE26                                     -689.7  {129.229,45096.4}   \n",
       "\n",
       "                                        timetree_CI_date  \\\n",
       "sample                                                     \n",
       "Reference                                             NA   \n",
       "GCA_009670865.1_ASM967086v1_genomic                   NA   \n",
       "GCA_008630435.1_ASM863043v1_genomic                   NA   \n",
       "GCA_000169635.1_ASM16963v1_genomic                    NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic                    NA   \n",
       "...                                                  ...   \n",
       "NODE22                               {-50502.4,-138.712}   \n",
       "NODE23                               {-50502.4,-138.712}   \n",
       "NODE24                               {-45096.4,-129.229}   \n",
       "NODE25                               {-45096.4,-129.229}   \n",
       "NODE26                               {-45096.4,-129.229}   \n",
       "\n",
       "                                            mugration_country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009670865.1_ASM967086v1_genomic                     China   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    Russia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                 Madagascar   \n",
       "GCA_000170275.1_ASM17027v1_genomic                      China   \n",
       "...                                                       ...   \n",
       "NODE22                                                  China   \n",
       "NODE23                                                  China   \n",
       "NODE24                                                  China   \n",
       "NODE25                                             Madagascar   \n",
       "NODE26                               United States of America   \n",
       "\n",
       "                                      mugration_province  \\\n",
       "sample                                                     \n",
       "Reference                                       Colorado   \n",
       "GCA_009670865.1_ASM967086v1_genomic             Xinjiang   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Karachay-Cherkessia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                Yunnan   \n",
       "GCA_000170275.1_ASM17027v1_genomic                Yunnan   \n",
       "...                                                  ...   \n",
       "NODE22                                             Gansu   \n",
       "NODE23                                            Yunnan   \n",
       "NODE24                                            Yunnan   \n",
       "NODE25                                            Yunnan   \n",
       "NODE26                                       La Libertad   \n",
       "\n",
       "                                    mugration_branch_major  \\\n",
       "sample                                                       \n",
       "Reference                                            1.ORI   \n",
       "GCA_009670865.1_ASM967086v1_genomic                  0.ANT   \n",
       "GCA_008630435.1_ASM863043v1_genomic                  2.MED   \n",
       "GCA_000169635.1_ASM16963v1_genomic                   1.ORI   \n",
       "GCA_000170275.1_ASM17027v1_genomic                   1.ORI   \n",
       "...                                                    ...   \n",
       "NODE22                                                1.IN   \n",
       "NODE23                                                1.IN   \n",
       "NODE24                                               1.ORI   \n",
       "NODE25                                               1.ORI   \n",
       "NODE26                                               1.ORI   \n",
       "\n",
       "                                    mugration_branch_minor  \n",
       "sample                                                      \n",
       "Reference                                           1.ORI1  \n",
       "GCA_009670865.1_ASM967086v1_genomic                 0.ANT1  \n",
       "GCA_008630435.1_ASM863043v1_genomic                 2.MED0  \n",
       "GCA_000169635.1_ASM16963v1_genomic                  1.ORI3  \n",
       "GCA_000170275.1_ASM17027v1_genomic                  1.ORI2  \n",
       "...                                                    ...  \n",
       "NODE22                                               1.IN2  \n",
       "NODE23                                               1.IN3  \n",
       "NODE24                                              1.ORI2  \n",
       "NODE25                                              1.ORI3  \n",
       "NODE26                                              1.ORI1  \n",
       "\n",
       "[55 rows x 32 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check if they're synchronized\n",
    "node_i = 0\n",
    "\n",
    "metadata_df[\"branch_support_char\"] = [\"\"] * len(metadata_df)\n",
    "\n",
    "for cd,ct,cmc,cmp,cmbmaj, cmbmin in zip(\n",
    "        divtree.find_clades(), \n",
    "        timetree.find_clades(), \n",
    "        mugtree_country.find_clades(),\n",
    "        mugtree_province.find_clades(),\n",
    "        mugtree_branch_major.find_clades(),    \n",
    "        mugtree_branch_minor.find_clades(),      \n",
    "    ):\n",
    "    # Rename internal nodes\n",
    "    if not cd.name or \"NODE\" in cmc.name:\n",
    "        cd.name = \"NODE{}\".format(node_i)\n",
    "        ct.name = \"NODE{}\".format(node_i)\n",
    "        cmc.name = \"NODE{}\".format(node_i) \n",
    "        cmp.name = \"NODE{}\".format(node_i) \n",
    "        cmbmaj.name = \"NODE{}\".format(node_i) \n",
    "        cmbmin.name = \"NODE{}\".format(node_i) \n",
    "        node_i += 1\n",
    "        \n",
    "    if cd.name != ct.name or cd.name != cmc.name or cd.name != cmp.name:\n",
    "        print(\"ERROR. Trees are not synchronized.\")\n",
    "        print(cd.name, ct.name, cmc.name, cmp.name, cmbmaj.name, cmbmin.name)\n",
    "        break\n",
    "        \n",
    "    node_name = cd.name\n",
    "    \n",
    "    # node type\n",
    "    node_type = \"internal\"\n",
    "    if cd.is_terminal():\n",
    "        node_type = \"terminal\"\n",
    "    metadata_df.at[node_name, \"node_type\"] = node_type\n",
    "\n",
    "    \n",
    "    # Branch Length\n",
    "    branch_length = cd.branch_length\n",
    "    if not branch_length:\n",
    "        branch_length = 0\n",
    "    metadata_df.at[node_name, \"branch_length\"] = branch_length\n",
    "    # Branch Support\n",
    "    # Give terminal nodes 100% confidence\n",
    "    branch_support = cd.confidence\n",
    "    if cd.is_terminal():\n",
    "        branch_support = 100\n",
    "        \n",
    "    # Internal nodes without confidence can be 0\n",
    "    elif not cd.confidence:\n",
    "        branch_support = 0\n",
    "    metadata_df.at[node_name, \"branch_support\"] = branch_support\n",
    "    \n",
    "    if branch_support >= 95 and not cd.is_terminal():\n",
    "        branch_support_char = \"*\"\n",
    "        metadata_df.at[node_name, \"branch_support_char\"] = branch_support_char\n",
    "    \n",
    "    comment_dict = {}\n",
    "    # Mugration  Country\n",
    "    if cmc.comment:\n",
    "        cmc_comment_dict = parse_comment(cmc.comment)\n",
    "        for attr in cmc_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmc_comment_dict[attr]\n",
    "            \n",
    "    # Mugration  Province\n",
    "    if cmp.comment:\n",
    "        cmp_comment_dict = parse_comment(cmp.comment)\n",
    "        for attr in cmp_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmp_comment_dict[attr]            \n",
    "            \n",
    "    # Mugration Branch Major\n",
    "    if cmbmaj.comment:\n",
    "        cmbmaj_comment_dict = parse_comment(cmbmaj.comment)\n",
    "        for attr in cmbmaj_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmbmaj_comment_dict[attr]            \n",
    "            \n",
    "    # Mugration Branch Minor\n",
    "    if cmbmin.comment:\n",
    "        cmbmin_comment_dict = parse_comment(cmbmin.comment)\n",
    "        for attr in cmbmin_comment_dict:\n",
    "            comment_dict[\"mugration_\" + attr] = cmbmin_comment_dict[attr]              \n",
    "\n",
    "    # Time Tree         \n",
    "    if ct.comment:\n",
    "        ct_comment_dict = parse_comment(ct.comment)\n",
    "        for attr in ct_comment_dict:\n",
    "            comment_dict[\"timetree_\" + attr] = ct_comment_dict[attr]\n",
    "            \n",
    "    for attr in comment_dict:\n",
    "        metadata_df.at[node_name, attr] = comment_dict[attr]\n",
    "           \n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "common-transcription",
   "metadata": {},
   "source": [
    "### Add Confidence and Handle Root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "built-turning",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "strain                                  NA\n",
       "date                                    NA\n",
       "date_bp                                 NA\n",
       "country                                 NA\n",
       "province                                NA\n",
       "                                     ...  \n",
       "mugration_branch_minor               0.PE7\n",
       "mugration_country_confidence          0.64\n",
       "mugration_province_confidence         0.48\n",
       "mugration_branch_major_confidence        1\n",
       "mugration_branch_minor_confidence      0.5\n",
       "Name: NODE0, Length: 36, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_list = [mugconf_country_df, mugconf_province_df, mugconf_branch_major_df, mugconf_branch_minor_df]\n",
    "attr_list = [\"country\", \"province\", \"branch_major\", \"branch_minor\"]\n",
    "\n",
    "for df,attr in zip(df_list, attr_list):\n",
    "    for sample in df.index:\n",
    "        max_state = \"\"\n",
    "        max_val = 0        \n",
    "        for state,val in zip(df.columns, df.loc[sample]):\n",
    "            if val >= max_val:\n",
    "                max_val = val\n",
    "                max_state = state\n",
    "        metadata_df.at[sample,\"mugration_\" + attr] = max_state      \n",
    "        metadata_df.at[sample,\"mugration_\" + attr + \"_confidence\"] = round(max_val,2)\n",
    "\n",
    "display(metadata_df.loc[\"NODE0\"])\n",
    "#print(list(metadata_df[\"mugration_branch_minor_confidence\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "controlled-desperate",
   "metadata": {},
   "source": [
    "### Lat Lon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "destroyed-burton",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>branch_support_char</th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>timetree_date</th>\n",
       "      <th>timetree_CI_height</th>\n",
       "      <th>timetree_CI_date</th>\n",
       "      <th>mugration_country</th>\n",
       "      <th>mugration_province</th>\n",
       "      <th>mugration_branch_major</th>\n",
       "      <th>mugration_branch_minor</th>\n",
       "      <th>mugration_country_confidence</th>\n",
       "      <th>mugration_province_confidence</th>\n",
       "      <th>mugration_branch_major_confidence</th>\n",
       "      <th>mugration_branch_minor_confidence</th>\n",
       "      <th>mugration_country_lat</th>\n",
       "      <th>mugration_country_lon</th>\n",
       "      <th>mugration_province_lat</th>\n",
       "      <th>mugration_province_lon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>6.334000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-29</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009670865.1_ASM967086v1_genomic</th>\n",
       "      <td>42012</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>-50</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722853</td>\n",
       "      <td>KEEP: Assembly Modern Placement</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1971</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>1.078700e-05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-50</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630435.1_ASM863043v1_genomic</th>\n",
       "      <td>C-719</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>-25</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED0</td>\n",
       "      <td>SAMN12721146</td>\n",
       "      <td>KEEP: Assembly Modern Placement</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1996</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.417700e-05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-25</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>43.736833</td>\n",
       "      <td>41.726799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000169635.1_ASM16963v1_genomic</th>\n",
       "      <td>MG05-1020</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>NA</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>SAMN02404403</td>\n",
       "      <td>KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...</td>\n",
       "      <td>1</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.927000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.21</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-18.924960</td>\n",
       "      <td>46.441642</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000170275.1_ASM17027v1_genomic</th>\n",
       "      <td>F1991016</td>\n",
       "      <td>1991.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>SAMN02404399</td>\n",
       "      <td>KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...</td>\n",
       "      <td>1</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1991</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>5.396000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE22</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>7.030000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-697.117</td>\n",
       "      <td>{138.712,50502.4}</td>\n",
       "      <td>{-50502.4,-138.712}</td>\n",
       "      <td>China</td>\n",
       "      <td>Gansu</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>1.IN2</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.42</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>102.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE23</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>7.040000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-697.117</td>\n",
       "      <td>{138.712,50502.4}</td>\n",
       "      <td>{-50502.4,-138.712}</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>1.IN3</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.42</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE24</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>5.630000e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.41</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE25</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.300000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.36</td>\n",
       "      <td>-18.924960</td>\n",
       "      <td>46.441642</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE26</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.300000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>0.37</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.98</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>-8.000000</td>\n",
       "      <td>-78.500000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        strain    date date_bp  \\\n",
       "sample                                                           \n",
       "Reference                                 CO92    1992     -29   \n",
       "GCA_009670865.1_ASM967086v1_genomic      42012  1971.0     -50   \n",
       "GCA_008630435.1_ASM863043v1_genomic      C-719  1996.0     -25   \n",
       "GCA_000169635.1_ASM16963v1_genomic   MG05-1020  2005.0     -16   \n",
       "GCA_000170275.1_ASM17027v1_genomic    F1991016  1991.0     -30   \n",
       "...                                        ...     ...     ...   \n",
       "NODE22                                      NA      NA      NA   \n",
       "NODE23                                      NA      NA      NA   \n",
       "NODE24                                      NA      NA      NA   \n",
       "NODE25                                      NA      NA      NA   \n",
       "NODE26                                      NA      NA      NA   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009670865.1_ASM967086v1_genomic                     China   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    Russia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                 Madagascar   \n",
       "GCA_000170275.1_ASM17027v1_genomic                      China   \n",
       "...                                                       ...   \n",
       "NODE22                                                     NA   \n",
       "NODE23                                                     NA   \n",
       "NODE24                                                     NA   \n",
       "NODE25                                                     NA   \n",
       "NODE26                                                     NA   \n",
       "\n",
       "                                                province country_lat  \\\n",
       "sample                                                                 \n",
       "Reference                                       Colorado     39.7837   \n",
       "GCA_009670865.1_ASM967086v1_genomic             Xinjiang     35.0001   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Karachay-Cherkessia     64.6863   \n",
       "GCA_000169635.1_ASM16963v1_genomic                    NA     -18.925   \n",
       "GCA_000170275.1_ASM17027v1_genomic                Yunnan     35.0001   \n",
       "...                                                  ...         ...   \n",
       "NODE22                                                NA          NA   \n",
       "NODE23                                                NA          NA   \n",
       "NODE24                                                NA          NA   \n",
       "NODE25                                                NA          NA   \n",
       "NODE26                                                NA          NA   \n",
       "\n",
       "                                    country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "Reference                              -100.446      38.7252     -105.608   \n",
       "GCA_009670865.1_ASM967086v1_genomic         105      42.4805      85.4633   \n",
       "GCA_008630435.1_ASM863043v1_genomic     97.7453      43.7368      41.7268   \n",
       "GCA_000169635.1_ASM16963v1_genomic      46.4416           NA           NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic          105           25          102   \n",
       "...                                         ...          ...          ...   \n",
       "NODE22                                       NA           NA           NA   \n",
       "NODE23                                       NA           NA           NA   \n",
       "NODE24                                       NA           NA           NA   \n",
       "NODE25                                       NA           NA           NA   \n",
       "NODE26                                       NA           NA           NA   \n",
       "\n",
       "                                         biovar branch_major branch_minor  \\\n",
       "sample                                                                      \n",
       "Reference                            Orientalis        1.ORI       1.ORI1   \n",
       "GCA_009670865.1_ASM967086v1_genomic     Antiqua        0.ANT       0.ANT1   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Medievalis        2.MED       2.MED0   \n",
       "GCA_000169635.1_ASM16963v1_genomic   Orientalis        1.ORI       1.ORI3   \n",
       "GCA_000170275.1_ASM17027v1_genomic   Orientalis        1.ORI       1.ORI2   \n",
       "...                                         ...          ...          ...   \n",
       "NODE22                                       NA           NA           NA   \n",
       "NODE23                                       NA           NA           NA   \n",
       "NODE24                                       NA           NA           NA   \n",
       "NODE25                                       NA           NA           NA   \n",
       "NODE26                                       NA           NA           NA   \n",
       "\n",
       "                                    biosample_accession  \\\n",
       "sample                                                    \n",
       "Reference                                  SAMEA1705942   \n",
       "GCA_009670865.1_ASM967086v1_genomic        SAMN07722853   \n",
       "GCA_008630435.1_ASM863043v1_genomic        SAMN12721146   \n",
       "GCA_000169635.1_ASM16963v1_genomic         SAMN02404403   \n",
       "GCA_000170275.1_ASM17027v1_genomic         SAMN02404399   \n",
       "...                                                 ...   \n",
       "NODE22                                               NA   \n",
       "NODE23                                               NA   \n",
       "NODE24                                               NA   \n",
       "NODE25                                               NA   \n",
       "NODE26                                               NA   \n",
       "\n",
       "                                                                     biosample_comment  \\\n",
       "sample                                                                                   \n",
       "Reference                                              KEEP: Assembly Modern Reference   \n",
       "GCA_009670865.1_ASM967086v1_genomic                    KEEP: Assembly Modern Placement   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    KEEP: Assembly Modern Placement   \n",
       "GCA_000169635.1_ASM16963v1_genomic   KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...   \n",
       "GCA_000170275.1_ASM17027v1_genomic   KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...   \n",
       "...                                                                                ...   \n",
       "NODE22                                                                              NA   \n",
       "NODE23                                                                              NA   \n",
       "NODE24                                                                              NA   \n",
       "NODE25                                                                              NA   \n",
       "NODE26                                                                              NA   \n",
       "\n",
       "                                    branch_number      continent date_mean  \\\n",
       "sample                                                                       \n",
       "Reference                                       1  North America      1992   \n",
       "GCA_009670865.1_ASM967086v1_genomic             0           Asia      1971   \n",
       "GCA_008630435.1_ASM863043v1_genomic             2         Europe      1996   \n",
       "GCA_000169635.1_ASM16963v1_genomic              1         Africa      2005   \n",
       "GCA_000170275.1_ASM17027v1_genomic              1           Asia      1991   \n",
       "...                                           ...            ...       ...   \n",
       "NODE22                                         NA             NA        NA   \n",
       "NODE23                                         NA             NA        NA   \n",
       "NODE24                                         NA             NA        NA   \n",
       "NODE25                                         NA             NA        NA   \n",
       "NODE26                                         NA             NA        NA   \n",
       "\n",
       "                                    date_bp_mean date_err      lat      lon  \\\n",
       "sample                                                                        \n",
       "Reference                                     29        0  38.7252 -105.608   \n",
       "GCA_009670865.1_ASM967086v1_genomic           50        0  42.4805  85.4633   \n",
       "GCA_008630435.1_ASM863043v1_genomic           25        0  43.7368  41.7268   \n",
       "GCA_000169635.1_ASM16963v1_genomic            16        0  -18.925  46.4416   \n",
       "GCA_000170275.1_ASM17027v1_genomic            30        0       25      102   \n",
       "...                                          ...      ...      ...      ...   \n",
       "NODE22                                        NA       NA       NA       NA   \n",
       "NODE23                                        NA       NA       NA       NA   \n",
       "NODE24                                        NA       NA       NA       NA   \n",
       "NODE25                                        NA       NA       NA       NA   \n",
       "NODE26                                        NA       NA       NA       NA   \n",
       "\n",
       "                                    branch_support_char node_type  \\\n",
       "sample                                                              \n",
       "Reference                                                terminal   \n",
       "GCA_009670865.1_ASM967086v1_genomic                      terminal   \n",
       "GCA_008630435.1_ASM863043v1_genomic                      terminal   \n",
       "GCA_000169635.1_ASM16963v1_genomic                       terminal   \n",
       "GCA_000170275.1_ASM17027v1_genomic                       terminal   \n",
       "...                                                 ...       ...   \n",
       "NODE22                                               NA  internal   \n",
       "NODE23                                               NA  internal   \n",
       "NODE24                                               NA  internal   \n",
       "NODE25                                               NA  internal   \n",
       "NODE26                                               NA  internal   \n",
       "\n",
       "                                     branch_length  branch_support  \\\n",
       "sample                                                               \n",
       "Reference                             6.334000e-06           100.0   \n",
       "GCA_009670865.1_ASM967086v1_genomic   1.078700e-05           100.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic   2.417700e-05           100.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic    4.927000e-06           100.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic    5.396000e-06           100.0   \n",
       "...                                            ...             ...   \n",
       "NODE22                                7.030000e-07             0.0   \n",
       "NODE23                                7.040000e-07             0.0   \n",
       "NODE24                                5.630000e-06             0.0   \n",
       "NODE25                                2.300000e-08             0.0   \n",
       "NODE26                                2.300000e-08             0.0   \n",
       "\n",
       "                                    timetree_date timetree_CI_height  \\\n",
       "sample                                                                 \n",
       "Reference                                     -29                 NA   \n",
       "GCA_009670865.1_ASM967086v1_genomic           -50                 NA   \n",
       "GCA_008630435.1_ASM863043v1_genomic           -25                 NA   \n",
       "GCA_000169635.1_ASM16963v1_genomic            -16                 NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic            -30                 NA   \n",
       "...                                           ...                ...   \n",
       "NODE22                                   -697.117  {138.712,50502.4}   \n",
       "NODE23                                   -697.117  {138.712,50502.4}   \n",
       "NODE24                                     -689.7  {129.229,45096.4}   \n",
       "NODE25                                     -689.7  {129.229,45096.4}   \n",
       "NODE26                                     -689.7  {129.229,45096.4}   \n",
       "\n",
       "                                        timetree_CI_date  \\\n",
       "sample                                                     \n",
       "Reference                                             NA   \n",
       "GCA_009670865.1_ASM967086v1_genomic                   NA   \n",
       "GCA_008630435.1_ASM863043v1_genomic                   NA   \n",
       "GCA_000169635.1_ASM16963v1_genomic                    NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic                    NA   \n",
       "...                                                  ...   \n",
       "NODE22                               {-50502.4,-138.712}   \n",
       "NODE23                               {-50502.4,-138.712}   \n",
       "NODE24                               {-45096.4,-129.229}   \n",
       "NODE25                               {-45096.4,-129.229}   \n",
       "NODE26                               {-45096.4,-129.229}   \n",
       "\n",
       "                                            mugration_country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009670865.1_ASM967086v1_genomic                     China   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    Russia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                 Madagascar   \n",
       "GCA_000170275.1_ASM17027v1_genomic                      China   \n",
       "...                                                       ...   \n",
       "NODE22                                                  China   \n",
       "NODE23                                                  China   \n",
       "NODE24                                                  China   \n",
       "NODE25                                             Madagascar   \n",
       "NODE26                               United States of America   \n",
       "\n",
       "                                      mugration_province  \\\n",
       "sample                                                     \n",
       "Reference                                       Colorado   \n",
       "GCA_009670865.1_ASM967086v1_genomic             Xinjiang   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Karachay-Cherkessia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                Yunnan   \n",
       "GCA_000170275.1_ASM17027v1_genomic                Yunnan   \n",
       "...                                                  ...   \n",
       "NODE22                                             Gansu   \n",
       "NODE23                                            Yunnan   \n",
       "NODE24                                            Yunnan   \n",
       "NODE25                                            Yunnan   \n",
       "NODE26                                       La Libertad   \n",
       "\n",
       "                                    mugration_branch_major  \\\n",
       "sample                                                       \n",
       "Reference                                            1.ORI   \n",
       "GCA_009670865.1_ASM967086v1_genomic                  0.ANT   \n",
       "GCA_008630435.1_ASM863043v1_genomic                  2.MED   \n",
       "GCA_000169635.1_ASM16963v1_genomic                   1.ORI   \n",
       "GCA_000170275.1_ASM17027v1_genomic                   1.ORI   \n",
       "...                                                    ...   \n",
       "NODE22                                                1.IN   \n",
       "NODE23                                                1.IN   \n",
       "NODE24                                               1.ORI   \n",
       "NODE25                                               1.ORI   \n",
       "NODE26                                               1.ORI   \n",
       "\n",
       "                                    mugration_branch_minor  \\\n",
       "sample                                                       \n",
       "Reference                                           1.ORI1   \n",
       "GCA_009670865.1_ASM967086v1_genomic                 0.ANT1   \n",
       "GCA_008630435.1_ASM863043v1_genomic                 2.MED0   \n",
       "GCA_000169635.1_ASM16963v1_genomic                  1.ORI3   \n",
       "GCA_000170275.1_ASM17027v1_genomic                  1.ORI2   \n",
       "...                                                    ...   \n",
       "NODE22                                               1.IN2   \n",
       "NODE23                                               1.IN3   \n",
       "NODE24                                              1.ORI2   \n",
       "NODE25                                              1.ORI3   \n",
       "NODE26                                              1.ORI1   \n",
       "\n",
       "                                     mugration_country_confidence  \\\n",
       "sample                                                              \n",
       "Reference                                                    1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                          1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                          1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                           1.00   \n",
       "GCA_000170275.1_ASM17027v1_genomic                           1.00   \n",
       "...                                                           ...   \n",
       "NODE22                                                       0.96   \n",
       "NODE23                                                       0.96   \n",
       "NODE24                                                       0.87   \n",
       "NODE25                                                       0.39   \n",
       "NODE26                                                       0.37   \n",
       "\n",
       "                                     mugration_province_confidence  \\\n",
       "sample                                                               \n",
       "Reference                                                     1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                           1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                           1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                            0.21   \n",
       "GCA_000170275.1_ASM17027v1_genomic                            1.00   \n",
       "...                                                            ...   \n",
       "NODE22                                                        0.36   \n",
       "NODE23                                                        0.83   \n",
       "NODE24                                                        0.84   \n",
       "NODE25                                                        0.36   \n",
       "NODE26                                                        0.36   \n",
       "\n",
       "                                     mugration_branch_major_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                         1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                               1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                               1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                                1.00   \n",
       "GCA_000170275.1_ASM17027v1_genomic                                1.00   \n",
       "...                                                                ...   \n",
       "NODE22                                                            1.00   \n",
       "NODE23                                                            0.97   \n",
       "NODE24                                                            0.97   \n",
       "NODE25                                                            1.00   \n",
       "NODE26                                                            1.00   \n",
       "\n",
       "                                     mugration_branch_minor_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                         1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                               1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                               1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                                1.00   \n",
       "GCA_000170275.1_ASM17027v1_genomic                                1.00   \n",
       "...                                                                ...   \n",
       "NODE22                                                            0.42   \n",
       "NODE23                                                            0.42   \n",
       "NODE24                                                            0.41   \n",
       "NODE25                                                            0.36   \n",
       "NODE26                                                            0.98   \n",
       "\n",
       "                                     mugration_country_lat  \\\n",
       "sample                                                       \n",
       "Reference                                        39.783730   \n",
       "GCA_009670865.1_ASM967086v1_genomic              35.000074   \n",
       "GCA_008630435.1_ASM863043v1_genomic              64.686314   \n",
       "GCA_000169635.1_ASM16963v1_genomic              -18.924960   \n",
       "GCA_000170275.1_ASM17027v1_genomic               35.000074   \n",
       "...                                                    ...   \n",
       "NODE22                                           35.000074   \n",
       "NODE23                                           35.000074   \n",
       "NODE24                                           35.000074   \n",
       "NODE25                                          -18.924960   \n",
       "NODE26                                           39.783730   \n",
       "\n",
       "                                     mugration_country_lon  \\\n",
       "sample                                                       \n",
       "Reference                                      -100.445882   \n",
       "GCA_009670865.1_ASM967086v1_genomic             104.999927   \n",
       "GCA_008630435.1_ASM863043v1_genomic              97.745306   \n",
       "GCA_000169635.1_ASM16963v1_genomic               46.441642   \n",
       "GCA_000170275.1_ASM17027v1_genomic              104.999927   \n",
       "...                                                    ...   \n",
       "NODE22                                          104.999927   \n",
       "NODE23                                          104.999927   \n",
       "NODE24                                          104.999927   \n",
       "NODE25                                           46.441642   \n",
       "NODE26                                         -100.445882   \n",
       "\n",
       "                                     mugration_province_lat  \\\n",
       "sample                                                        \n",
       "Reference                                         38.725178   \n",
       "GCA_009670865.1_ASM967086v1_genomic               42.480495   \n",
       "GCA_008630435.1_ASM863043v1_genomic               43.736833   \n",
       "GCA_000169635.1_ASM16963v1_genomic                25.000000   \n",
       "GCA_000170275.1_ASM17027v1_genomic                25.000000   \n",
       "...                                                     ...   \n",
       "NODE22                                            38.000000   \n",
       "NODE23                                            25.000000   \n",
       "NODE24                                            25.000000   \n",
       "NODE25                                            25.000000   \n",
       "NODE26                                            -8.000000   \n",
       "\n",
       "                                     mugration_province_lon  \n",
       "sample                                                       \n",
       "Reference                                       -105.607716  \n",
       "GCA_009670865.1_ASM967086v1_genomic               85.463346  \n",
       "GCA_008630435.1_ASM863043v1_genomic               41.726799  \n",
       "GCA_000169635.1_ASM16963v1_genomic               102.000000  \n",
       "GCA_000170275.1_ASM17027v1_genomic               102.000000  \n",
       "...                                                     ...  \n",
       "NODE22                                           102.000000  \n",
       "NODE23                                           102.000000  \n",
       "NODE24                                           102.000000  \n",
       "NODE25                                           102.000000  \n",
       "NODE26                                           -78.500000  \n",
       "\n",
       "[55 rows x 40 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "latlon_country_df = pd.DataFrame()\n",
    "latlon_province_df = pd.DataFrame()\n",
    "\n",
    "df_list = [latlon_country_df, latlon_province_df]\n",
    "attr_list = [\"country\", \"province\"]\n",
    "conf_df_list = [mugconf_country_df, mugconf_province_df]\n",
    "\n",
    "# Create a mapping of geo name to lat,lon\n",
    "for df,attr in zip(df_list, attr_list):\n",
    "    # Get data\n",
    "    for rec in metadata_df.iterrows():\n",
    "        node_name = rec[0]\n",
    "        name = rec[1][attr]\n",
    "        lat = rec[1][attr + \"_lat\"]\n",
    "        lon = rec[1][attr + \"_lon\"] \n",
    "        \n",
    "        if name == NO_DATA_CHAR: continue\n",
    "\n",
    "        if name not in df.index:\n",
    "\n",
    "            df.at[name, \"lat\"] = lat\n",
    "            df.at[name, \"lon\"] = lon\n",
    "            df.at[name, \"size\"] = 1\n",
    "        else:\n",
    "            df[\"size\"][name] += 1\n",
    "\n",
    "#display(latlon_country_df)\n",
    "#display(latlon_province_df)\n",
    "\n",
    "\n",
    "# Update lat lon for mugration\n",
    "for latlon_df, conf_df, attr in zip(df_list, conf_df_list, attr_list):\n",
    "    for rec in metadata_df.iterrows():\n",
    "        sample = rec[0]\n",
    "        name = rec[1][\"mugration_\" + attr]  \n",
    "        lat = latlon_df[\"lat\"][name]\n",
    "        lon = latlon_df[\"lon\"][name]\n",
    "                \n",
    "        metadata_df.at[sample, \"mugration_\" + attr +\"_lat\"] = lat\n",
    "        metadata_df.at[sample, \"mugration_\" + attr +\"_lon\"] = lon\n",
    "        \n",
    "        \n",
    "# Mapping file for auspice\n",
    "out_path_latlon = os.path.join(out_dir, \"latlon.tsv\")\n",
    "\n",
    "# Countries\n",
    "with open(out_path_latlon, \"w\") as outfile:\n",
    "    for country in latlon_country_df.index:\n",
    "        lat = str(latlon_country_df[\"lat\"][country])\n",
    "        lon = str(latlon_country_df[\"lon\"][country])\n",
    "        outfile.write(\"country\" + \"\\t\" + country + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "# Provinces\n",
    "with open(out_path_latlon, \"a\") as outfile:\n",
    "    for province in latlon_province_df.index:\n",
    "        lat = str(latlon_province_df[\"lat\"][province])\n",
    "        lon = str(latlon_province_df[\"lon\"][province])\n",
    "        outfile.write(\"province\" + \"\\t\" + province + \"\\t\" + lat + \"\\t\" + lon + \"\\n\")\n",
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "offshore-connectivity",
   "metadata": {},
   "source": [
    "---\n",
    "# Convert Dataframes to Geopandas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "separate-bailey",
   "metadata": {},
   "source": [
    "### Modern Human Cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "tired-manual",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'xu_2019_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-886aedfcdd3e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mgeometry\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mrec\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mxu_2019_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterrows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0msample\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrec\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mlon\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrec\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"X\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"'\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'xu_2019_df' is not defined"
     ]
    }
   ],
   "source": [
    "\"\"\"geometry = []\n",
    "\n",
    "for rec in xu_2019_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    lon = float(rec[1][\"X\"].replace(\"'\",\"\"))\n",
    "    lat = float(rec[1][\"Y\"].replace(\"'\",\"\"))\n",
    "    point = shapely.geometry.Point(lon,lat)\n",
    "    geometry.append(point)\n",
    "\n",
    "xu_2019_gdf = geopandas.GeoDataFrame(xu_2019_df, crs=CRS, geometry=geometry)\n",
    "\n",
    "# Generate a figure with two axes, one for CartoPy, one for GeoPandas\n",
    "fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=(10, 5))\n",
    "world_polygons.plot(ax=ax, color=\"grey\", alpha=0.5)\n",
    "xu_2019_gdf.plot(ax=ax, c=\"red\", markersize=2)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "collectible-superintendent",
   "metadata": {},
   "source": [
    "### Project Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "filled-border",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<GeoAxesSubplot:>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjwAAAEbCAYAAADeTl6JAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAACO6ElEQVR4nO39V4hs277mB37Th/cuvTfL273P7XtuFbQMfYXoFqKE9CD0JNQI6kV6UlFCAtEUKj2IggYVemlJBQXSi6CppgsKNSpQ1dnnnH2W28vkWmlX+oyMDO/NNP2QZ4w9IzIiMiIzIsPk+MFi78wMMyJixhzf/JvvzxmGAQaDwWAwGIxxhh/0AhgMBoPBYDD6DRM8DAaDwWAwxh4meBgMBoPBYIw9TPAwGAwGg8EYe5jgYTAYDAaDMfYwwcNgMBgMBmPsEdv98a//+q+NeDx+V2thMBgMBoPBuDFv3779Z4Zh/HWzv7UVPPF4HG/evOnPqhgMBoPBYDB6CMdxgVZ/YyktBoPBYDAYYw8TPAwGg8FgMMYeJngYDAaDwWCMPUzwMBgMBoPBGHuY4GEwGAwGgzH2MMHDYDAYDAZj7GGCh8FgMBgMxtjDBA+DwWAwGIyxhwkeBoPBYDAYY09bp2UGgzFYDMOAYRjQdZ3+4ziu5T8A9L8MBoPB+BUmeBgjgaqqKJfLqFarUFUVmqZRMUD+KYoCwzCgqiq9Dcdx4Hme/mv8mfwTRRGCIECSJIhi/78WtVoN+XwehUIB1Wr1inApFotIJpOoVqtdP7b5cURRhNVqhcViufJPURQmjhgMxr2BCR7GwNF1HbVaDdVqFdlsFplMBrquQxRFuN1uZDIZRKPRO1uPxWJBMBjE4uLiFUFgGAYVW2ZEUaS3rdVqyOVyqFarsNlssNlsVERtbW0hHo/fSMi0QhAECIJAxVyjeNJ1HaVSCeVyGQCoEHK73QgEArBYLD1bC4PBYAwrTPAw7pzj42OcnZ2hUqnQNE0r+il0bDYb7HY7FEWB3W6HJEn056OjI3z9+hWVSqUuqtRqrZIkweVyoVgsolQqNf27JEl1ESVVVXv2WgRBgK7rUFWVvqeCIMBisUCWZVgsFho5KhaLCAQCmJqagq7rKBQKqNVqTf8ZhgFRFGnky+PxwOl09mzdDAaDcVcwwcPoG5VKBdlsFrFYDKqqwjAMOJ1OTE5OIhgMIhaL4fT0tKlA6CeCIMDhcMDtdsPr9UJRFBSLReTzeUSjUVSrVRSLxa4ESa1WQyKRaPv3Wq3Wi+VfQdO0pu+hqqrI5/PI5/N1v+d5HtlsFr///e+vjTRxHAe73U7/9VKk3Xd0XafRQlEUoWkaisUiTUPy/Pj3lBiGgXw+j0wmg0qlQi+CXC4XvF4vHA4HS7syegYTPIyeYxgGdnd3cXJyciX1k06ncXR0BEmSYLFY4HA44PV6IYoiHA4H0uk0BEGoi4Tkcrkb17MQBEHA5OQkJicnYbFYUC6Xsb+/j69fv/Y0vTQK6LqObDZb9zue52kUyuVyQVEUBAIB2Gy2kdpwSIE3AJyfn9PUYq1WA8/zsNls0HUdmUwGgiBAlmXIsgyr1QqHwwGHwwFFUfq6Rl3XUa1WUS6X8f37d+RyOXi9XjidTkiSBFmWwfM8rFZrX9cxSHRdRzwex/Hx8ZVjEQDi8TiAy/Ty5OQkJiYmIEnSXS+TMWZwjRuSmdevXxtv3ry5w+UwxoVsNouTkxOk02laVHxdhCMcDoPjOJTLZWQymStiqR1mkSQIAqxWK91AJEmC1+u9csLUdR3lchmJRAK7u7s3ep3jhMViwcrKCvx+/6CX0hLDMHBwcIBcLodSqYRKpYJwOAxJklCpVGgKlNQoybIMh8OBSqVCo3bmgnZS/yQIAnw+H0Kh0J1vrIZhjJSo7CW6rmNrawvn5+dtv+88z+P58+dwuVz0fHIfImCM7uE47q1hGK+b/Y1FeO4xJJxcLpdhsVhgs9kgCMKV22UyGZycnCCfz0MURVgsFtr54/P56BVxIpGAqqqQJAlOpxMPHjwAcJnaevv2bdu1eL1eyLJM60Wy2WxXggcA1tfXr2zWpGurXC4jlUqhXC6jXC7TIt5yudz184wjHo8Hjx49GvqraI7jEAgEcHFxgWKxCAA4PT29cjtVVWmKked5OBwORCIRhEKhO+nC64b7KnaAy89mfX0dKysryOVyKBQKtG6uUqnQdB85L3z79o2KI0VR4HK5aGrabrcP+uUwhpzh+uYz7gzDMBCPx2kkxWazQVGUpoLH5XJBFEUUi0UYhoFqtYp8Po9EIoGjoyMsLy/D5/NB0zTs7e2hWq3CarVibW0NbrcbiqLghx9+QDKZRK1Wo4W15itrUu9TKpUQi8XaFjKb4XkedrsdTqez7orPMAx8+fIFqVQKmqb17H0bV/x+/9CLHYLD4cDr16+RTqdxcnLS9jMOBoN4+PAhOI5DrVbD/v4+rZOZmZlpKfIZd4sgCPB4PPB4PE3/XiwWsbW1hVqtBo/HU1dI3+m5gsFggmfMIFEbkkoCgGq1Cr/fD6/XS29XKBRQKBSgqiosFktb/xlz4Wo7QqEQgsEgACCfzyObzSKbzUKWZYRCIYTD4ab303UdX79+RT6fh9VqhcvlQjabbXsiI0XH8/PzTTdqjuPw+PFj1Go1FAoFlEolqKqKQqGAYrFI2+CZGLokkUggEAiMTN0Ix3Hwer3wer1QVZWKW4LFYsH6+jpcLheNoBiGAYvFQqMG5XKZRizvc5RlFLDZbHj27BmAyyL9bDZLz2HFYpHWoMmyPOCVMoYZJnjGhGKxiNPTU1xcXKBSqdDfk46kycnJuloBUqBJIEWEmqbVpamuo1aroVgsUs8Xp9OJZDKJra0tqKoKnufhcrlQqVToc9tsNvh8PgDAL7/8QmsxyOsgqYpWcBxHCzur1WrbyARJr+VyORwdHd27AuV2cByHUChEi5RJanPUNn9RFPHkyRNks1mkUinoug6/338lWiDLMqanpwezyAGiqiqy2Szy+Ty1WFAUBXNzcyNXB3N2dobd3d0r3YK5XA5nZ2eQJAnLy8solUq0Tgu4PNaJDcWoHd+M3sGKlseAVCqFo6Mj6LpOPVfcbjdcLhckSUKpVMLu7i6SyST1ZiF1OH6/Hy6XC9VqFZubm0gmk3A4HFhdXUW5XAbHcfD7/XUnxnQ6TU86je3Qy8vLmJ6eprUzmqZBVVUkk0kcHR3RwuXFxUUaUSCdNaqqUtdhQRAgiiJ4noemabTThqTLGlNYnUC6k8imaHZfNv8zj3IoFovIZrPUtG/c4HkeXq8Xi4uLrAZiBDEMA+VyGblcjjp3mwuzyX/J9538c7lccLlcg14+DMNAMplEsVis84+y2+20Xd0c7b1t+koQBDidToTDYUQikYGKH03TEI/HUSqV6kbG8DxPPy+Wcu0eVrQ85pDQfjPy+Ty+fPmCUqlEXXfL5TIqlQpyuRyOj48hSRIcDgfNi+fzebx79w4AaGGgOeJjt9uxvLwMQRDw/v37urSQw+GAYRg4Pz9HMplEuVxGIBDAzMwMDMPA9+/fAQB7e3vY29uDzWajhnxE4JBiYnKyJvA8T9NYN7ky5Xm+bZ1AO6rVKo6Pj3F4eNj1fYcZ4sKcy+WY4BkBSDonnU4jk8kgn8/XRTusVivcbjd1+LbZbLBYLAOL5BiGQb11crkcUqkUcrkcarUaNE2785Sypmk0FTboSM/JyQn29vauvZ2iKHj48CHcbjdOTk5QKpUgyzIkSYKiKFAUBRaLhQmjDmARnhGGRCLIiYN0MzT7IpvTWdVqlfp/AKD1OWbxQQqKyc+tIB1QRLDIsgxd15FIJJDL5Whaa35+HhzHIR6PU0FTLpepIR+5wgFQN9uKFCWS1JXNZkM4HIbb7e7DO9oewzBQKBRwenratDNoVLHZbLTAnDG8GIaBWCyGr1+/1v2ejAjx+/2w2WwDWt1V8vk8Njc36XlmUJB6P5fLBafTCavVOnCxQyiXy7i4uKCRa/OFJdmbJUmi300iIIm/lKqqtESACZ5L2kV4mOAZUTKZDL59+3YlpUSiNCQ0SrofwuHwjU+G7XxCiIuv1WqlIXRyQslmszg/P6fRJY7jqOFgMBisWw9JgVUqFdhstjqRVavV6PynSqUCVVWpIGsl8PrNyckJtre37/x5e43X68XCwgKsVuvIdGndNzRNw+npKU5OTupSq263G2trawMTObquU3dkVVXr0sIkPX1wcDDQujlZluHxeK6khwKBAL2QMneMkpl0zSAz/667CLwNxC7DMAy43W4mYm4AS2mNIURcNELy9wRZlulsKGLCNzExAVEUrwgZ8nOpVEIqlaKt66lUCouLi5idna17rrOzM2xubtb9LhAIYHV1FQDw/v37Kx43sixjcXGR1u6cnp7i7OyM5vCnp6extLSE4+NjxONxFItFVKtViKKIxcVFqKqKg4MDGgoXRRGBQACRSORGqaqbMjU1BUVRsLGxMXJtsV6vFz6fD6IoQlEUOJ3OobniZdSjaRrOz89xdnZGxY4oilhYWMDk5OStPzfSLCDLctePlUwm8fnz51s9f7+pVquIxWJXfu9wOK6MXCE4nU4sLCxcMTMtl8v48OEDgF+NToFfhw2Hw+FbuXQXCgXE43Gayk8kEohEImx2XQ9hEZ4RJB6P48uXLy0N82RZpl1ZyWQS+Xye3pYU7ZHOKKfTiWAwiEwmg3Q6jUgkgtnZWfzpT3+iwsnr9eLp06f0hFipVLC3t0cfo7ErzOv1Ym1tDV++fEE6nQbHcZifn0ckEqk7sX7//h0HBwd1aycRqVZzqebm5jA1NYVcLodyuYx4PE7rjAKBwJ23paZSKfzyyy93+pw3xeFwwOPxIJfLweFwYGVlZdBLYlzDzz//XNe1SC4oenGcb29v4+TkhP4syzIikQgWFxcBXEYxVVXF3NwcANQN0OV5Hrqu482bN8za4c84HA68evXqRiI0mUzi5OQEhUKhzmPI/N9mvyN1j4xfGeoID+nOIWkYRntIZT9pI24kEAhgbW0NBwcHOD4+bnr/dDpNf87lcnU5dnIClGWZCp5UKoXT01NMTU0BuCyie/DgAR3LEI1G6WOQokBRFDE1NUWNAZtd/WQymabrazeE8/T0FMfHx1dOstFoFFtbW3C73ZienkYwGOy7BX2pVLpSTzHM5PN5lEol+Hw+RCKRQS+H0YZisYgPHz5cSQc5nc5bix1VVXF0dFQndoDLc7HP50OtVqODfd1uN46OjpoO+bXZbHC5XHX+R/cZu91+JWqeSCSoNQYAmv6zWq0QBAGqqqJWq0EURaTT6RuJx5mZGczNzQ2dg3gjwzBCZeARntPTUxwcHEBRFExNTbU0p+slmqbR9uh4PI5EIkFbIaempmgBLakRGVYqlQr29/dRrVZht9vpFTwpHE6lUjg/P6fmXLcZoaAoCp49e9a0XsAwDFqgbLfboSgK9vf3sb+/X3e7yclJLC8vUwFCJqlHo9GeTuEmQyHL5TIEQcDCwgKCwWBfoj9k3ZVKBfF4HOfn59f6CA2CcDhcN24BAF68eHHvC5WJCDQX9FcqFciyDLfbPTDfFlKcWiwWcXFxgbOzMwC/jtaYnp6uM1XslKOjIxwcHDT9vlksFjx69AiapmFzcxOlUol2b5ovkhitsVqtmJqagsfjQaFQoBdEpITALGgsFgvtVrstHMfB4/HA5/PRAmbSzEL+EcNNSZKo4LorSJPM06dP+/5c975omUQiSqUSstkscrlcy82fhGqByygHmfGk6zoMw6A+FqQNkHQykUI9kvO9LSR0LEkSkskkSqUSIpEIPUg7Ucu1Wg3JZJJ+0czDOwuFAtLp9LUDPRsh3j1erxdut7tOEJq9JD5+/IhkMnnl/m63G0+fPqWvI5vNUrPEarVaZ0LYDuKX0w2ky8vj8WBqaqpnAuj9+/fIZDLgOA6Kolypoxo0TqcTs7OzsNlsKBQKyOfztHOl35PBhwlyfJHvzfHx8ZX6DhIRJMeW1WrF/Pw8gsHgQFu7yaZFujINw4Ddbu96Tc3SyGYCgQAeP36MdDpN61UY44coinj8+PGN6x67jdbcZXRnqFNa/ebk5AQ7Ozsdb47mAtRqtYrz8/Nr70PM6zRNoykc0uYtyzKsVivsdvuVaBFxQCXCplQqIZFIIJvNolKp0HoYMitof38f8/PzAC6v1J4/f45yuUyv/kh7osPhoB1RoVAI+/v7bU9y3VAqlXB+fk59W1ZXV6GqKuLxOPb396HrOrxeLyYnJ+H3+3F0dFSXepNlGalUCna7vaUBWjKZxPfv32mnGRn+SFrVzSHidDpNXVav+4xrtRoymQwymQzOzs7w7NmznnjPEMFrGAYEQUAoFEI+n28q+AZBLpfDly9fIAgC/H4/gsHgvWhjNQyjLuXaTkiTLkBVVWEYBmRZRqVSoR5F1WoV2WwWPp8PwWDwTiO/ZG3tnlPXdezu7gL4tVFBURQ4HI667rtAIEBNSptBIhGjlKpldI+qqri4uKDn11KphFqtRoUJOYbIvqbrOtxuN3ieh6qq+Pnnn6nRLRnpQXzedF2Hoih0DyRR/2Fg4BGeQqEAnud7MsNHVVXaXUQ+0EHmlxVFoSdZjuPoVF9FUZBMJnFxcdGXSd08z18xHjM7rpbLZRweHl4x9SMnVXKQkgiIIAgoFov0S6GqKp48eUIPYsMwkE6nkUql6G3L5TIePnwISZLw/v37K14cExMTCIVCODo6ouKIXAE06/4CLlNipAPMjGEYODg4wNHRUUfhYUVR6MZFRlzcFvK8pJvj3bt3QzHGwmKxwO/301EL96HAUdd1nJ+f4+joiKYXycmb1HURSFSWDBclWCwWWl/YKA6IEy450ZPW536ISOL9ZBjGlW4dwzBQLBape3o6ncbHjx+vrNdms8Hv9yMQCNA02OnpKb5//34lwhsMBrG+vo6LiwvE43HE4/Gevh7G6GKxWDA3N0frusiFdidYrVZMTExc6fTtB0Md4UmlUtjZ2YHdbkcgEEAoFLrxVXc6ncbp6SkymczQtQoTUXAXuXBd15HP55u2XbpcLqyuriISidCIBGmxJBvB9vZ2XURIkiTMzMxgcXERmqZdyf+SQY75fB5nZ2ew2WyYmJgAx3HU8r6RWCxGvzCZTAZ/+Zd/CUEQqHdHI6IoYmJiouXrtVgsWF5epnOzLi4uWkZYSGojn8/TNNdtNypBEHB8fIxoNNqy3bVfkDEbZBCmxWKhV/fDZLLWb2q1GvWrqdVqkGWZdrEQc0sA9HdEkJpnLhHIRVOz8wgZOWLm5OSEfg+IwLRYLHW3KZfL2N3dBc/zUBSlqUO6efgvcVMGLkexlMtlxGIxmtIqFouoVCrgeR6hUAjJZLLpesl8uqOjI8iyjEAgAK/XC6vVekXwEBO858+fD02EkjEclMvlpheinVAqlRCNRu9E8LRj4BGe4+Nj7Ozs1P3O6XQiEokgFArdyAytVqvRMQD9iKB0CommDAscx1E3T0mS6moCSCRqfn4exWIRJycnLUUjx3FwuVz0hEnmX5XLZbqJOBwOaJp2pbOjGTzPY3V1FVartamZImFubg4LCwtXfn9wcIDv37+D53k8efKEbiLk952gKApmZ2dv7G0SjUbx7du3ru93G0KhEBYXF69srPcNwzBwdnaGnZ0dGmbvJLpmjsA2g0TDbnrxFAqFsL6+Th9HVVW8efMG5XIZfr8fPp+PiviDgwNks1lks9mBtnkrioLHjx/DZrNhf38fR0dHA1sLY7yw2Wz48ccf+/48Qx3haQZplc5kMrDZbMjn87R+w+PxIBAItN2USEQiGAwikUjUmXbdJYMUW80gV4Xtuog6UfCGYdBamFZ0E+XQdb2tWJBlGeFwuGUH3+TkJAKBABRFoXUOpVKpq3B8pVLB9vY2zs/P4Xa76THUqfi5yxx1L43nRh1d17G9vU1ruCwWS0dF750gSVLTxyJpITKosxXpdBqqqtalhmdmZrC9vQ1d12G1WhGPx5HJZK60iA+CYDCI1dVVOki4mWEfgzHKDFzwkGiBOTpAiMVikCSpLux6cnICm82G5eXluvqLUqmEZDKJbDaLTCZDh1b6/X5MTk7WufMyhhdZlvHixQvaRk9SM9cJXHMksFAo4P379203o5mZGfh8PpRKJezt7dHbkqtsMvurU/pZAExSVcQk8ja1OMlkEqenp6hUKrDb7ZienobVakUikUAmk6HvNal/4zgOk5OTWFpa6vGruj2VSgVfvnxBNpvtWuyQOrdWjuXk8RvPP+T3U1NTePr0KVRVpcMoC4UCarUabDYbHA4HXC4XTk9PEY1GabqLXGykUqmh8q8hFhzELJTBGEeGQvA8fPgQmUyGGtaVy2U6nNF8spmenqaTYc3tdPl8Htvb21ciDqzobrSwWq14/Pgxvnz5gsnJSTgcjmtt1UkHDRmkqqrqlQnSzSiXy7BarfB6vQgEAjg4OEAsFoMgCFhbW2s5fb4VLpcLL168aJuS6xRFUTA9PU27/W5i+98MUpfidDohiiKi0Sii0WidFUMzzs/PqdgalohSOp3GxsYGvUjqVOw4HA6sr69Tfx1d1xGPx7GxsdH09s2EJXEat1gsCIVCtB7H/N6cnJzg559/rovy3va46CeapnVVhMpgjCIDFzyRSIR2PZjx+/1Ip9OIx+P0RDE/P9+0NdPhcODFixc4OztDLBZDsVikrZzFYhFOpxMOh4MW5DGGD5vNhsePH9Ni4q2tLfp7Uu9A2iIJ+Xwe7969q9tUiN36dZAuFK/Xi3A4jOXlZSwtLdFC7pvgdrsxNzd3q1oeURTx+vXrvgzy5DgOPp+PWh+QrqXralSuG6p4lxiGgcPDw7rarOtqccysra3B4XAA+HUoJ7m4aka794ZcjNVqNbx7946m3L1e71AbljIYdwUZSmyxWIbC3X3g38pWoXnS6bC4uFgnYMzmdo1MTExc6eQxz30hwyoZg4fneUQiEeoXEgqFaHu5GVmWkcvlcHR0hKmpKTqmQhAEOBwO/PDDD9A0jdZ4kenp+XweFxcXiMViLeu3yKyxZDKJSqXSkw4Cr9fbNA3SKU6ns69Ty0VRRDgcpsNbC4XCFbEgiiLcbjfcbjc8Hg8cDsfA29nJINvv37/TtBDP85AkqaOiYkVRMDc3B7vdDlVVcXx8jOPj42sjgeS80ex8s7Ozg/Pzc3i9XtjtdsTjceTz+aYjXRiM+0i5XMb09DQdSzRoBt6l1S2np6e0dZMMyWxWMEpSHaIoUn+Wu25X7+bK876xvr5+RfFXq1Xouo7T01PUajVEIhFqSlgqlVAsFulm3Gm0IZvN4uvXr3XpBFEU4fP5wHEcEokEVFWFKIr47W9/25MoRqFQwMbGBgqFQsf34TgO4XAYi4uLdz4AVdd1OlKB2M4PQzQHAB0Q26zdvxNh6XA4MD09jVAoBJ7nkU6n8fXr146/lyS6xSLDDMbN4DgOjx49QiAQuKvnG8/REq2uvK6bJn5XMMFTjyzLWF5ehsvlokZw7SDdW/l8HrOzs1QgESHL8zzy+TyKxSI1WSROoKqq0jQpEcjJZBKpVIq6W3u9XprGUlW167qd69a+u7vbUfcNx3H44Ycfms4pu4+QSKx5KG0zrhsv8uDBA4RCIXqc6bqOn376qeuxH6R4nokexrjj8XggCELbAc6dYrfbaW1drVbDb37zmzuJFI9cW3qnNNswDw8Psbe3N4DVMNrh8Xjw6NGjrtI1PM9jYWGhri7HMAx8+/YNs7Oz8Hg82NnZqesqkSSJjgfweDyYnZ2F1+uF0+mE0+nE3NwcHcxIokahUKjXLxc8z2N5eRnVahUXFxdtb2sYBkqlEhM8f2ZzcxPRaPTa2/E837Lz0uPxXLEx4HkegUCgo8cmkBQpEzuMcYDjODoHsjFiGgwG8fDhQzqipNvHJQOsiaHmXUeqO2GkBU8zyMyqXg1v5Hme+vkM00DIUcLlcuHx48c3KuRsHDnCcVzdxF2Xy1UneMwpDuJWGw6H4fV6IQgCtf8njsT9hOM4rK6u0o6oVtEIi8XS0+gS47LYvVkEuNNUHZknxCK0jHFgZmYGU1NTdSNW3r17h1wuB5/Ph+npaXi9XiSTya48oURRxOTkJH3sYWdsBE+5XEY+n6ctvZ0UJF6Hy+XC1NQUvn37VrdZzczMYH5+nlq9p1KplhPYeym+RpX19fW+da0EAgEcHh62vc35+TkdAvsXf/EXTbuwSFqr1wXDkiRhbW0Na2tr9HfJZBK7u7u0xmd+fn7gRcHDRKdz9QRBaBnhOT09ha7rV2qiIpEInbNnfhyn00lHjHAch2QyeecjQhiMXmOz2bC+vn5lQDOxgzEMA1arFefn53jz5k3LukNZluH3+yEIAs7OziCKIubm5mgDyagwNoKHFBaqqor5+XlMTU0hFotBlmXUajXa5twJxGV3amoKGxsbNIcfCASooZggCLDb7VhYWMDCwgIdo0A+/Hw+Tx2js9ksRFGkfjFkGjOAG3fzjAqktqZfWCyWrkRlqVS6EtnRNA0fP36Eqqp48eJF31uKfT4fvF4vyuUynf7O+JVOBE87sUOIRqN0OjzB7XbD5XIhkUjQmVZk1hYZBFqr1cDzPB3ayWAMK4IgwOVyoVaroVQq1X0nbDYbXr161VKQWK1WZDIZbGxsUHHv9/vhcrmgaRoikQhyuRysViucTieNjs7NzdFB1KPG2AgeSZIwOTlZ9zNphatWq9ja2qL+G6SICkDTuTsulwvT09PgeZ7ObyJGZa0gbdIEq9VKT7TVahWHh4eIRqP0RE2ef9wLm8nQ1H6kkMrlMj58+NBVBC0ajUKWZTqg1jAMbGxsUNPKT58+4cmTJ30XPWSuGeMqxMSvUWz4/X4sLS3h6OgI5+fnbTsuLRYLJicn68QOgeM4BAIBbGxsIJFIXBFO4/6dZAw/oVAIDocDkiRRIS4IAlRVrXMHJyl6ANjd3UUikaDO7IqioFQq1e1LBE3TsLe3V5e+stvtWF1drUtNNbtYbXS2HyVGukurU0i6gvj4EMEhyzK+fftWV5EuiiJWV1f7UsharVYRi8WQyWToJGZd11Eul+9FYeSPP/7Y02iPrus4Pj7G/v5+13YDk5OTtLammeB9+vQpM4+7Bel0Gtvb2xAEAaFQCNPT013dn8zAKxaLsNvtiEQi8Pv9AC4/959//rmlv5LD4cDr102bNOqIx+M4Ojqqc2i/rvOLwegnTqcTKysrV1JQvYJ0rH79+rVupqLH48GTJ09GKj3VirFtS78thUIB5PUFAgEEg0HwPI+zszPkcjnMzc3diWHS+fk5vn79eivDulEgFArh4cOHPX/cRCKBT58+9fQx3W43nj59OhYngLtGVVX8/ve/p2HvSCQCTdMQCoWuHRXSDk3TEIvFaBPB169fm95uamoKKysrTf9WqVRoqsvhcNDBo2ZxO+4XHozhQxRFLC4uYmJioi8eWIZh4NOnT8hms00j4haLBX6/HwsLCyN/oTe2bek3pVQq4ejoCKenp3SAInFojsVi4Hkea2trdcNJ+0koFEKlUsHh4eFYFznHYjFUKhV6ZU7cfIPB4K1aGP1+P7xeb0+HMWYyGRwdHWF+fr5nj3lfSCaTNE1ULpexv78PQRC6jvIQqtUqvn37RmftRSIRhEIhOh6kERIJakTXdXz48IF+/8ljEONFAktpMe6SSCSCpaWlvqWJNE3Dt2/fkEwmW96mXC7j5ORk7M9390bwaJpGa2nOzs7A8zyeP39eN4QUuBQf/UhntYPjOMzOzsLv9+PNmzdQFIWehIfF8bZXmNMHsVgMsVgMOzs7mJiYwPz8/I2ET7Va7crVuFNUVUW1Wh1KP4lhJhAIwG6308+E4zisr6/f+H1MJBL0ZC0IAtLpNBwOB6amppBKpWgkiYz1aHWhQpy8SY1CIpEAx3FQFIXWRhDGPdrKGDykZsbtdvf1eS4uLq71ArNarXC5XCMf3bmOsX51JF95dHSEdDqNQCAAl8sFj8cDv98/dB+u3W7H5OQkzs7Oru1AGSeIs+7p6SksFguCwSCCwWDdfKxWqaVkMomtra2+pCGOj49xcnKCmZkZzM3NsfRWh/A8j5mZGcRiMSSTSRiGgS9fvlBjs25FPBEogiDgxYsXKBaL+OmnnyDLMqamplAoFDA3N3dt3UMzYUMwDAM2mw2VSoVGd0iLuiAIYxfxIZ2ihmFA0zRWt3SH8DyP+fl52hjTb5pFQRtZW1u7cvE/jgzXjt8jNE3D2dkZUqkURFHExMQEHjx4MBLRkpWVFczNzeHz58/IZrODXs6dUy6XcXR0hKOjo7rfv379um7K9eHhIRKJRN+9Ush07qOjI9qeOTc3x1yRTRiGgUQigYODAxQKBUxNTWF+fh6VSgUcxyGVSkHXdVxcXNAhsN2IR+LcWiwWkc/n4fV68cMPP6BWqzW9Oi6VSkilUlBVFX6/HzabDdvb2zTVRpoXzIXuFosF5XK5buMnwogUMo/C+aMVPM/TcSykicMcwboulT7qr39YEAQBi4uLcDqdd+a9dV1k9TZ1daPGWAoenucxOTl545qBQZBMJlEsFuFwOODxeDAzM4OtrS0WVv8z5g2SOCXfpTGcYRgoFosoFosol8t4/vw52wBwKS62trboxYUkSSgUCtTSAbhMY75//x4AsLe3h7OzM/z4449duR4rioJUKoVffvmFjgqJx+PU6dVutyMWiyGRSNR1n3z//p0+D9n0SdSGHFM8z1OPrGbUajUqiIb9MxdFsW4Mi2EY0HUdtVqtbScjx3G0hoTneVQqFeo/Rv4+Cq9/2NE0Ddvb2wCAxcVFRCIROjOO2KFwHIdSqYRCodCTgZsejwenp6dN//ab3/zmXtljjKXg4Thu5L6Yoiji4OAADx48AHA518Tj8eAPf/jDvUpvtSIej2NmZob+HA6Hsbu7O5AC70wmg3/5L/8llpeXabH7fYaM9iBDW8kk+mw2i2g0euW7WCqVcHp62rQDUtd1pNNpOsollUqB47i6lFIqlaorUM9msy1rbsimTx7bTDfFybVaDYqioFarDXX6RxAElEqlrkdqmN87nuchyzKtYSOMwuvvNSSl3sh1ArIT9vb26uY+XlxcIJVKYXl5GXt7e9A0rWeCpxGe5+H3+++V2AHGVPCMIi6XC8+fP0c8HqfGa5IkYWFhAWdnZ30pyh0lDg4OEA6HaXiW53l4vd5ri/H6haZpSCQS917wEBMysjGS6CpwKYDOzs6abpC7u7u4uLjA1NRUnTngwcEBDg4Oul5Hv9MDxCl9mOmVh1Bj1xqBFPAP+/vQK0j6r5koNgyjL1E/MsuK1N1omnbr2kFZluFwOGhEfG5uDvPz8yMXFOgFbIDPEEGKls0RnenpaTx//nxwixoSVFXF58+fabqCXP0PCkVRmrr43jdOT0/B8zzsdjvm5+fx+vVrOJ1OaJqG4+PjlvfTdR2FQqHOfZu0iwOgYx865SZRUF3XIYoidbJt1RZMRLYsy+B5ng4W7SfmlFsncBxHRVk/N7JqtQpFUe4sykPSapIk3XlkyTCMlp8BiTr2emAmz/N9qakhVg2SJCEcDt9LsQOwCM/Q0eykK4oi3G53XUv3fSSbzeLNmzeYnJxEOp2+83SW0+mE2+2Gw+FAKpVCNBpFKBS6tycPAJidncXs7OyV3wuCQEPziUSi6WZF5oglk0mk02kkk0laY9PqyroZJG3VbSs5z/NNr9BJjQ+JGhFX9Mbb9CPaQdJG5oJp0k1lTieR94k4x5O/34WHUKVSqROq5q62XqffLRYLfXye56kQvIuuOTJjrR21Wq2nr3txcRHhcBjn5+c0dduLtNbk5CSq1SoWFxdHdixEL2CCZwQ4PDykQzir1Sry+fzYtcl2Chkn0SvIlbEgCHUFnmYkScLS0hLC4TAA4JdffkGpVMKrV6/utdhpRyqVws7ODmq1Gvx+Px0+aLfb6QZut9vpPCtJkujmDaCj6AYRA6TVXNM0GgkghcjtaFWDQbqY2kG+f0SgcBx3o02PCDUi2hq/142/I6+XrI+0l5Pnv6smB2IeSuolSXt7sw64RkjUxtz9ZhjGtZ8X+VxUVYWiKLTQnAhUIgBVVe3oe9mu86xT4WgYBkRR7JngcTgc4DgOHo8HqVQKR0dH8Pv9tz7PKIqCtbW1nqxxlGGCZwQIh8OIx+M4PT1FsVhEKBSCzWbD8fHx2Loy3wUWiwVPnz6lLeb5fB6JRIK2n5PNhJxMAeDk5ATpdBo2mw07Ozt48OAB3WAZv0I8c8wFn+VyGbIsQ9d12sZO5tgR0UA2b+D6NJUoilc2JfOGL0lS0xl1vay/MD+/2S+qld9PI2QdkiR1dPvG18Jx3EA6Oc3vmzl6p6oqfX/J38iayedbq9WaihtzF911UT5z1KfRR4gI6mbnRiIwiSAmx5s5WtZtvValUoEsy3So522OKfK+EcGYyWQQj8dZ+rxHMMEzAlgsFjr+4vz8HNvb27BarVhcXEStVsPZ2dmtOwb6RWOHzTDgdrsxMzMDm82GeDxOT27kPU6lUri4uLhiDJbL5WhXBWlRz+VyqFQqmJiYaDm/6T5CTADJpiXLMorFIt6+fQtN05oer+bN0RwFME+HNkMGABNxQf6fbH5EDJAoHongkSLrXkfnzJEXRVE6EjC6roPjuLG6cGn8zpP3QlVV/K//1r8F65Mn+Ff/i//iyv1aFUu3g7x/ZszimawHAF2D+b2uVqsQBAG6rtPj5ibRGrJuInqJB1U32Gw2Kniy2SysVitmZmZoezqLJt8eJnhGCEEQMDk5iWAwiNPTUyQSCTx48ADVahUnJyeDXl5TzFdU5tC3eTO6Sxv/YDCIcDiMk5OTuqJnWZbpTC6SujKjqio+ffp05WRYKpUA/NpdcRfDZkeFXC6Hd+/e0Y2EpH86wRyZuS56RkRMY2s1SXmQqAEAeiw2bpTkOCWpzdu2X3dyX7LGcd3IiLM1Ke4tl8t4dHGBw//j/+j7c3eSIiO3I1GiXkBSq+baI/Ox1S7yZ+74VFUV09PTtOOR0RuY4BlBJElCKBTC9+/f8fPPP2N5eXloBU+7q1cSSjZfkbW6mu8FU1NT4Hkenz9/vvK3QCAAr9fb8r7EGIxsZIIgoFwuI5fLwefzIZ1O4/j4GJOTk2O7gXWL3W7HgwcPkE6n6eDYbiEdQWaxTMSz+X02HzOkHsgcWTBvMoqi0IgPz/N0w1NVte7q/CaDfMl6OxF2uq7Tzq92xoejSGM6qlKpQJIknP3n/zkev3gxwJVdxZxKVVWVprRu83mQOiMAdee0VikvMmqCIEkSm+HXB5jgGVGsViudEL67uzvo5dwYclIxnyCJBX4vO2D8fj8qlQrS6XRd2JlAxlRIklTXaloul+lgykZBlM/nkcvlEAqF8NNPP9GCQ8YlPM/TYbyzs7P4/PlzV+7YsizXfUakQ6dVoSsRMEQYNbuSlmW5rnanMbVGOpDI8zYW1jaD4zj4fD64XC6Uy2U6P6wVjSk3si7yWlwuF9xuN3K5HAzDGJnuTHMKqVnxv2EY+Mt/7V8DcP0oi7vCHEkkbeikjuc2UefG44+kU5udH2ZnZ69MKa9Wq/j27RtkWe77cNH7BBM8Iwxpjx62GpnbYq6D6IWjKXApaEiNB3A1TVKpVHB2dkbDzsFgEHa7HR8+fMDDhw+bDqZ0OBx0vtdvf/tbJnbaYLFY8OzZM3z48KEjE01zlxb5rEhUpnEjslgsdIMlhcpkPALwa/Ey2dyu+5yI2RsRVw8ePIDH48H+/j6d56VpGtxuN1ZXV2ndRSqVwvfv3699bZ//h/8B+sUFnv2dv0N/p6oqVlZW4Pf761q+AeD3v//90H/HSQF5p91Rw1RzSIqsybmmH2sjxdqN2Gw2OoLFTDAYRCgUGvrPfdRggmeE6bXp1bBRqVTobKDbRnsaPVNIakTXdZoiOT09hSAIcDgc2NnZQaVSgdVq7ehKlHVpXY8kSfD5fB0JHlLYTFp+m3XsKIqC2dlZ5HI5pFKpphEh87BMc3rsuucm6bCVlRXqg7K0tITDw0MqeHRdR7FYpF4snYgdAFj4d/4dVP9c+wUACwsL8Hq9daKaPHarSIO5LqRXFwU3hQjMTgV/u+/K+f4+zv7H/xHP/6v/qlfLawkZG9GN51O3kMhgs4if1+vFo0ePmlowBINBWK3Wezf6od8wwTPC3IfBokRsdGKbbzZpIz8DlxtYo2DSNA2RSAQrKysol8s4PDxELBaDrusQBAHBYBBerxdWq5VFbnpIN8KVGL+1ev/9fj+mpqaQyWQwNzeHjY0NAL+2RpuHg2qaRtNV19VnkA1qdnYWkUiE/p7jOExNTdGp63a7nf43Ho/XDS1th8PtBv6cplhfX697DrL+w8NDHB4eNr0/KdBu7ArrtBW+V5hr7rr5XEmEh0TRyOdbqVTgDASQePSoL+ttFLyVSqVv7xeJKJILq8aLJtLV2Uz8lctlvHnzBuvr6wiFQn1Z332FCZ4Rhph/3Qeuy6mTdNV1V2qKoiAQCCAcDtMi5D/96U/gOA5zc3OYmpq6106k/aabY1ZVVVgsFgiCgIWFBWSz2bri53Q6jXK5TGscVldX8e7dO9p2TqIe5khopVKhjr2NmxApXCVmieZhtQRRFPHo0SPqGUOYmprCxMQEkskkdnd3affex//pfwInCHjyH/wHTV9jLpe7YvV/fHzcUuwAzS90yHTzu5zqTjb0mzwXqfMxR6YkSYLF74ft3/13e7nMOszCjKRA+wERVs0u1BYXFzEzM9PyfctkMtB1HRsbG4hGo4hEIigUCpidnb31XK37DhM8I8ooFTP2gsaTA9lwyMnkuitMQRDw9OlTuFyuusfiOA6PHj2C3W6nBoSM7igUCrS+xWq10vSGxWKhzrtESLQ7Zm02GxwOB1RVpYJhcXERfr8fPM8jGAwiEAjg/fv3AC69kH7++Wd6Jex0OvH8+XMcHx8jl8uhXC7TCAKJ+JBNlmwc5L/mAmiLxdLWXqBV9wzP87Tbb39/H7FYDNr+PmAqfG7k5OQEkUikbn7S/Pw8crkcNWXslMb5TuZUCukq7FScNHrTNKPVBUixWMTf+1t/C3/9d/8u/sbf+Bsdr59EqEiquZeQ45Dw//k3/g2slstI//2/j9/85jc9fS4z5lS6LMt1ju2NqKqKRCKB8/Nz+rtkMklHroRCIdjt9r6t9T7ABM+IUigU7k2ExzyNG7jcWCRJ6irvPjc317LbgbmY3hzDMOi8LABt63MavW5IEbGu61hYWLjSqWImnU6jVCohEokgGAzi4uICwGWty/b2Nux2O+x2O9xuN9xuN/L5PM7OzpBMJlEul2Gz2VAul+lIAhLdaZbSWFhYuNXGIggClpaWMDMzA/c//IeIxWJX0q1mn59MJnNlYOSjR4+QzWaxvb195T01Rw1IhMJcjN1M/Kuq2rJwFrjcjLf/0T+CZXkZM7/9LS1AbhzamU6n4ff7aQt/s+f6//6b/yb+3wD+X//lfwn883/ewTv2K0SYmiMvJAVG0j+kQL2VeCPeSwCo4L1yrlhYwOm3b3iytNTV+jqBDDudnp5GOBymn38gEGhbv7S3t4fT01N6/1qtBpvNhidPnsBisbDUeg9ggmdEsdvtdzIocBggJzBSi0HqMbohl8shl8v1ZRLxfYbjODidTlrI227UAfE5aUwlmV1xW3F6ekq7rR4+fIjj42Ps7e3Rrq23b99ibm4ONpsNwWAQDocDKysrtL03mUzCYrFAVVXqvWNep9nZOZvNNr0K13Ud2WwWuVyORgt8Pl/LY0qWZSwsLCAWi2HvT39CZnMTz//9fx9A/diLnZ0diKJYV8vD8zw8Hg9sNhsCgQCi0WhdFIoU3JKoDRFy7SAF3OaNsy5S6nDA+PPjkNuY36Pd9+9x+A/+Af7Vf/yP25qFVnkev9N1ZNuupjXNjpFGzMKZRO2I/5J5+GqrGp3/6z/8hzdcXXNIPVKtVkMoFMLExASNGLeK6DRCPr+VlRX4fD7EYjF4PJ4rXXuMm8O1C1m+fv3aePPmzR0uh9ENb9686crXZJTppLvGDM/zcDgcUBQFXq8XiqLA6XQyM68+cHFxgS9fvgDoLBVCEEURdrsd4XAYExMTbT/fQqEATdPqOpmKxSKOjo6QyWRowbAkSZifn0c4HKZX+UdHR9jd3aV1YJIk0c2x8bginYETExPw+/1UbOfzeezv7zcV2k6nE+FwGD6fD1arlUZHSDF8JpPB7//e34O+uYm/8Y//cdM5VKqq4sGDB1eijUSkn5+fY3Nzk0Y+Gj2ISDSGCM5WKSEyy6xWq1G/IXIxQQRpM5FABnSWy+W62qVWFIvFO0sRd3tu6CWkcJsIaEEQ8Fd/9Vc3Wk+hUEA0GoXH44Hf7+/Dau8HHMe9NQzjdbO/sQjPiELaVu8L3ZxAZFnG1NQULBYLXC4Xa+3sM+b0Q7OBnq3u8/Lly44+m0qlgnfv3sFiseD169f0WLDZbFhbW4OmaYhGo9jd3UWtVsPOzg7Ozs4QDAYxMzNTFxVpdPJ++PAh7HY7SqUSNjY2MD8/D4/HA03T8PHjx45qSUj0sB0//Gf/GU3NkmhSY0Tm27dv0DStroi5Wq1SMUnei2aYBVCjBYMZ8nvSdk9GbZinvxPBYy7MJlPIOxE7AO60Hq5fYofneczMzCAQCODTp091BoXEt4f8jng3ra6u3ng9p6enODk5QTKZhM/nYymsPsAEz4jCcRwmJiaGdqTEILFarU3NvBj9wVxc22mx6ePHjzsWoqSduJWxnSAImJqagsfjwYcPH+Dz+bC+vo5UKgVVVTE7O4t8Po9MJkMjKqurq6hWq7BYLFAUBYqiYG1tDV+/fsX+/n5H6+oWMuy0cbq3OULR6OatKAoKhUJXFzeNbtGNEHFExA7wqzeOWfhUKpWeup2PEi6XCw8fPoTFYsHx8TE9VlwuFwRBwPn5OT3WOY6jfk23iSCTY4DUZ7ILtd7DBM+IwnEcIpFI3wVPqVjE93/2z/Dw3/63+/o8t4V020QikabtxIz+UCqVcHR0BKD93DQzoihSh+pOaTbaoxG73Y7l5WV8/fqVppgIy8vLePfuHf356OgIL1++rBNQhUKh7+MOSKuyOQVF5isFg8ErZqIOhwM+nw+lUqnj2U6kSJdEYxpfE/EiIjU4jcXkxKjxPhIOh+FwODAxMVHn7L26ulqXdg2FQvjll18AXEYrvV7vrdPl5PMg9UCM3sPsYUcYh8NRN3CuH+QuLlD++eehH2zo9XqhaRqcTicr8rtDzOKllaNsI8QkrxuePHly7bGuqipyuRwePnxYJ3aASzH08OFD+nM+n8fx8XHdeu8ihUDEBOkUI11qoihiYWGh6X2WlpZutDZS/Eu6fshrrVQqdZszmSZPIjq1Wg08zw9tvdun//l/xub/9r/1/HF5nsf8/DxmZmaoWOQ4DgsLC3VDgQ3DqLvQ1DQNv/zyC7LZm5ZpX9Y8nZ+fU5uMYX3vRx0meEYYjuOwtLSEiYmJvj1HaG4OL//r/3ro88mSJGFlZaXjjghGb5AkCevr6wB+HTHQDlmWsby83JOxKMViET/99BP+8Ic/4F/8i3+BX375BeFwuKXNgN/vp7Ul5pZ4QiAQ6HvtiVkUmsVWpVJBNBptmhIkXUhEhJAiY/PU93ZUKhWoqnql1ZtEdoiBoPlxhnl6e+l3v0Pmd7/r+ePquo63b9/SYczVahUHBwfI5/NXPhe/34/FxUXY7XaoqopyuYyvX7/eKDIWj8ept5TVar0i1hm9g3VpjQHlchk///zzUA3ku2vsdjuePn069vPFhpFcLofv37/DYrHAMAzaQt1IIBDAo0eProhnEvUgHUKyLCOTySCbzUIQBLhcLmSzWcTjcQSDQTidTrop1Go1xONxxONxCIJQF8VpxunpKba2tjA7O4uFhYUra8lms3Wpr3aY5ySRLiYyqZ2kiMjtiG8M+V2r+hoSZSDpLVVVUSwWsbm5iXK53PR9JemxVvU25rET5nEO5onmZG3EW8fcrt7pUNBxwu120xQnz/OIRCJYXV2tu41hGDg7O8PW1hb93eLiImZnZ699/Hg8jlqthnQ6XWc06Ha78eLFi969kHsI69IacywWC+bm5joeXjiOFAoFZLNZZiI4AJxOJ54+fUp/lmUZR0dHtJCTCAKSOgEuP6+TkxOkUilUKpW2Yp10VwGXqSjzhiBJEiYmJjAxMdGR4J+cnITdbm85Nd3lcuG3v/0tdF3H6ekpMpkMlpeXIUkSTk5OEI/HIUkSQqEQIpEIEokEtra2qKAwt5mTdJLZIJQY9rVC13Xs7e1hb2/v2tdCIKKEFHeb/XFIuzR5b4iPVSMkOkcMEc2368Tjp1f8P//D/xCuvT34/uP/GP+3f+/fu5PnbAZxBJ+bm8P8/HxLwdc4AqST4nJN07C9vd1U9LLZWf2FCZ4xYXZ2FqVSCdFodNBLGRj3taNk2PD7/SiVSrDZbDAMg0YLarUajo6OkM1m6zqmrqPRLyeTyTR1ze50Yj25bzqdprUSJFUE/Npm31hTQ67ezcNJPR5P13VAmqbRUQe9isqSFJ3FYqHCixjhdRqdaRXJIcKN3Kaf/F/29vB/B/D/+O//e2CAgsdqteLBgwd1vk+NkEJzUrQP4ErbPhkBVCwWUSqVIIoiSqVSy/eRzfHrL0zwjAkcx2FtbY1eMd8nyImnn7VMjM4wDAM7Ozt00CeBtNiSGVnd0Dg4tlebgsPhQLFYRCKRgKIoCIVCbX1myuUyvnz5Qv1/gNYRE8JdTi8n0R3ynN0Y8jXOmjJP+ybGesQ/qJ/DSc8A/Awg3ZdH75yZmZm2YodAondutxuBQKDOLVvTNGxubtZ9DxoRBIGapE5MTLAIT59hNTxjRqlUoo6dqVTqTk+4gyAQCGB9fb1jQzRG/zAMA4eHhy1Tq+0M8VpBHIDJcRwMBvHo0aNbr/WmNIqIw8NDnJycXLnIIPPeWomDm7wX19Fu3EM7SKrRfN926+vH2gn5fB7/6B/9I/ztv/23+/L43TA5OXmlbqcRMufMZrPVfc4XFxfY3t6GYRi0ZV3TNFitVuzt7cHpdOLRo0eso7QPsBqee4TVaoXVakUgEEAul8P29vat2iWHFavVimfPnrETxhBgGAbi8TgODw/bOg6TVFC30QGzaE8kEvj27RvW1tYGUkhrfs5arYaTk5M6UUNmXFUqlWuLffsxEqHbx2zWoUV+3+4+/cLhcAyF2AGAs7MzzM3NtW2EaBwyW6lUcHZ2hsPDQ8zNzSEQCODt27d0VEQkEgHP85iYmGBeOwOACZ4xxul04tmzZ/jDH/5AaymuC8GPAsFgEEtLS0zsDAm7u7s4Pj6+9na6rrd1AG51H1EUabuvruuIRqN0xpQsy1AUBX6//9rjIZ1OI51OQ1EUuN3uW7egk9dMaodkWa4rUG5HtVqFoih0ZEMvIGmobm5PzgmNEOPCZmsb1nb1XmMYBra3t5t2Frbj5OQEuq7j4OAAwWAQi4uLdR5S/fZOY7SGCZ4xRxAEhEKhOqMsWZavHTI4rKysrGBqamrQy2D8mWq1eq3YIbOHgKuuv51APGRIZ5UsyzR1S9je3obf78fjx4+bbk7fv3/HwcEB/VmSJEQiEbjdbrhcrq6N3gzDQDqdrhuUWi6XIUkSRFGkBcPtxAHxupFl+UapqEbI83W6OQuC0PbzaPY4xKDwvrSpx+NxHB0dddRqDlx2tL169Qp/+MMfEAqFYLPZ7nSuGKM9TPDcA2ZmZnB2dgZd17G+vk5bbBOJRJ1fiBlypdrtFXk/CYfDTOwMGWdnZ9feptOBou0gU84FQWj5WIlEAt+/f0cgEIDT6azblBvTuqRj7OjoCA6Hg9axuN1u+Hw+OJ1O6LqO4+Nj5HI5yLKM6elp2s11fn6OYrFIoyOSJNFWdCIiOI6j359WwqcTs8ZOMQ+1bHUhQ+ZnmYubW9FMhJHnGJZzwl2wv7/fseABLlOwNpuNjbgZQpjguQdYLBbMzMzg4OAApVIJkUgEFosFyWSy7gRthmwWlUqFCqJ+pcJ4nqcnaPP/N+J0Ovvy/Iyb0y6CQTbXXnX1tDs2CIeHhzg8PIQkSfD7/ZiengbP8227w6rVKvL5PAAglUphf3+fFpmaj/mzszMYhkGdjjVNo+shLeFmIWAYBsrlcl1bfjNIDU0v3iNiGtgMURRb+g810q4bi6S7yJT1cY/2LC0tdXV7u92OH374Yezfl1GECZ57wuzsLKLRKBKJBGZnZ2G32+H3++smXbeCuN+Skz9xZe3U44PneVitVpRKpSsb1uvXr+FwOKBpGn2sk5MTnJ6eIhKJIB6P00LYTCbD8t9DRrNUkLlwF+hdkWujqV47arUaotEootEodURuRTPB36wLiYgcUnxttVqpKFAUpeXadF2HpmlNIyMkBTY9PQ2v14tCoYCdnZ0bpZpJeqxVB9V1KSzzmq4TMsSBWJIken7oNOqz88//OYr/7J/h6d//+x3dftDE43GEQqGu7BCY2BlOmOC5JwiCgOfPn2NrawsfP37EkydPsLa2ht///vcdFSFWKhVqasbzPKrVKrWeb3Vy5jgOi4uLmJychCAI9Ha5XA5HR0d1m6U5rTYzM0PDwZOTk3j//j2KxSLi8Tiq1SobrDckJJNJ7OzsAKgXOf0qjCfRi26LZq+7PYl8dPI4brcbmUwGHMfBarXS0RLXbfbmCebmCIxhGCgWi5iamoIkSXC73ahUKnX1Rp3S2ClGXjf5PDoRJGTkRCcQYQT8GgnuJH059Zvf4MLrvfK8sizTSNcwpcxSqdS9KdQed9jw0HuE1WrFkydPoOs6NjY2IIoilpeXO7ovOQmRCc8A6GyeVnOTXrx4gZmZGSpmeJ4Hz/Nwu914/PgxVldX6aTtVkiSRPPnhmHc6/EZw4JhGFQ4k81RVdW+b1LXRWpu87idoOs6Lfh3Op2oVqt0wninkO8P+Ver1WCxWOrM6WZnZ6lRY7eQC5FarUafoxvxSSI2nUCGjxKIqJIkiY65aIbVZsPs8+f0Z7OLM3lPZVnuWW3TbbHb7ewia0xgEZ57RiKRQKlUQi6Xw8XFBaampmC1WvH9+/e2HiqtICds82a3vLzc09STeUMhIwlYyPjuKZfL2NraQj6fr/tM7uqz6KSGp98kk0ksLS3BarWC53l8/vz51tGsSqWC4+NjmjYhQ1D39/eRTCa7FnlkrlYnApQUIZNIVTfijXSZmQWL2cCQDCMlFzok0qTrOv0bgKa1QtVqlab7BjkyptnQUMboMhwSmnFnWK1WGlXZ2tpCLpeDz+fD4uLija+oyuVyXUqq151U4XCYPn6xWKSip1qttpwgzegthmHg27dvSCaTNIpANiRyRd9vhuGKX9M0lEol+P1+eL1eOmLitpRKJbx9+5YWVzudTjx58gSvXr2qi/ZYrVY6ALUdtVqtI/dxcrFCok3dQmZ4kWOC+CIBoNPYyXBYYnJI/kuigo1iR1EUWhNFhM8gUBQFy8vLQ3HcMXoD+yTvGQ6HA8+ePcPLly9hs9nw/v171Go1eL1evHr16kYnF1LETE4Mvb4KVxSlbpAjETnn5+eoVqs4PT3t6fMx6lFVFZubm0in0wBAC9ZrtRpN0ZAajn5uTsMS1Ts7O6PRk1AohHA43JPHLZfL+PDhA4Bf644cDgd+/PFHvHz5Eo8fP8aPP/6ImZmZtqloIkQ7iTx1M1z0OjRNo8cCx3FUCCuKQiM7ZBI7+W8jpPh5GGp4VlZW2MiaMYMJnnsIx3FwuVx4+fIleJ7H+fk5gMt5MDd1L65Wq9B1HZIkoVgs9nK5AC6jRmRjsdls4HmeDvjr1YbDaM7BwQGi0Sj9mVzJS5IESZJofQPZ8BRFGRpx0g90Xcfu7i79eXFxsWeu35qmIZfL4e3bt/R35PsaCASQSCRwdnYGj8eD2dlZam5HIE0EJN1kplmkgtQl9RrSUWcesVGr1doeG8QfyEwz4fa//p2/g3/6D/5Bz9dsJhgMIhAI9PU5GHcPk6/3GHIV9v37d4RCIciyjJcvX+Lk5IR233SLoijXFiLfBI7jsLKyQk3lgEuRdX5+zgy++kwmk6H/TzppWqU/yGZGrup74SBMIOaDvXzMmxKLxRCJRODz+aAoCqanp2/8nTGjqiq2t7eb/u309BRbW1sIh8O0A5KgaRqKxSKKxSLteCJpYEEQaBHw169fcXFxUfe4RJD0MzVM/ItIioqMwVBVtc680RzZIXU/jev68Y9/RA4A/tP/tC9rtVqtNFVZLBaxt7fX9XgJxnDCIjz3nOfPn8Pr9eL79+80vD09PY3l5eUbfcE9Hk/fTgyiKCIYDNLH53keqVSqL8/FuNyk9vb2qEtxp1OyNU1DuVymFgK9rIEgUcRhSDVsbGzQmptwOAxvQ6v1TclmszAMA5lMBvF4HMlkEvv7+9ja2qIR2UbRIggCnE4nwuEwQqEQPB4PnE4nnE4nbDYbbYdfX1+nrtIk3XQXETnzxHsS+TEMA6qq0udu9Ahqddz88fFj7P4r/0pf1ul2u/H06VMaWfrll1/q1sgYbbh2qv7169fGmzdv7nA5jEFgGAb29/eRyWTw7Nkz+uVOJpP49OlTV1d+CwsLmJub69dSr1AqlW7cwstoz9nZGTY3NwH86rxLvFbaCZ9GYUQM6nrluAyA1ow1RntIVKCb5+m0o6kZT58+hc/nA/CrQDw6OrrRY5khEZHGdI55rXa7HS6XC0tLS10JwEqlgpOTExweHt56nZ1ChheTcwlZr6ZpdE5as/sAdzPeRhAELCwsYGpqih47R0dH2N3dxfr6OiKRSF+fn9E7OI57axjG62Z/G/xlEmPgcByH+fn5K8W/Pp8P6+vr+Pr1a8eP1cuiVVKQXK1WUSwWUa1WIQgC/H4/9fdhYqd/mOt2iOdKJwNnG8WGruvUW8U8juGmKIpC/WVI+kZVVYiiSAd43tXYg2g0CrfbDUEQwHEc5ubmcHZ2dusJ6GTGFjEQbGYeWCgUUCgUUKlU8OjRo6Yz8ZqhKAoWFxfh8Xjw+fPnO2n1J58T8enxer1IJpM0vdXqPsCvLepmccvzPH3PSaeYJElduXHzPA+XywWbzYa5ubkrnYZkCGwwGLzhq2YMG/dS8GSzWbhcrkEvY6jgOK5pO3k4HEY0Gu04ddSrOgDi+eL1erG8vIxisYjz83OUSiXs7+/j9PQUuq5jenoac3NzLOTcY1KpFE1lka4sssHcFLIRtZvT1Ari99K46ROXYxJBIHUhAK4dNkogtWw3IZVK4ePHj3j8+DFNtVmt1ht5WjXSuHG3GtyZTCaRTCa73pjtdvudehuRwasulwtra2s4Pj7GwcHBtc9vPqeQ9wDAlfeBFMxf59PldrtpVKzdJHNih9GpkGQMP/dS8BCxE4/H4Xa76dUCozmzs7MdC56joyNMTk52tJnVajWkUikUCgUAlxsrcVVWFAVPnjyhj+NwOGgxdLVaRSaTwdnZGfb395FKpbCysgK73c6ET484Ojqqu+LtZlO8TvSSaA+p4bgOUnPSLo3WbOo4GYx73ciDbqICzchkMri4uMDk5CSAy2O3F4KnkVaFxYFA4EYdRSTSs7W1RR+fiNFeYbFYsLi4CJvNBlmW6zrCQqEQjo6OOkpXkXow4uxOxnQ0Hj9kBM78/DysViuSySQODw8RDAbh8Xjg8/k6jgrzPI9QKNTdC2YMNfdS8BACgQAMw0A6nYbNZrsT87RRpJsTYLlcRiaTgcfjaXu7fD6Pt2/f1p3AzSeXdsJFlmUEg0E4nU4cHx/j4uICb968gcvlwsTEBEKhEBOwt8Tn8yGZTN7ovp1E+czRnnZCpJv6jXK53LTuhaSDeJ6HLMs93dCbsbCwgGQy2ZfISaMwczgcePjw4Y2F/uTkJGw2G/L5PILBIPL5PD59+tSLpQK47HhqJRpkWYbT6ex4xhfHcSiVSvS1EoFGDA6By3P62toajdgJgoDj42Osra0NRaE7Y7Dc+y4tjuPg9XqRSCSoHw2jnm4LBr9+/XptN0+hUKjbGEVR7Nq1llzJ+f1+hEIhiKKIZDKJVCoFXdeRSCSoM3Mul2OOzF0wPT19rZtvK7rZ6MvlMhRFuRKdIa3U3Rx7xACz3brM080FQYDFYumJMDGLLLvdjlevXt34/WsGWbP5GI5EInj9+vWtu+A8Hg+mp6dpx1Yv4DgOkUgEDx48aHu79fX1ay+OyLFA3JoJpNuL/M7j8eDhw4dU7Oi6jlKphGAw2JeIG2P0YJL3z5RKJaiqilAoxNIiDczOzqJUKnUsCCuVCg4PD1u6wWqahrOzs7rfkQLBbhFFsemsG13Xkcvl8OnTJ1rwaLfbsbCwAL/fzz7jDrjpRkq8Uzp5j8lQWrKhkWOATF2/yXNfhzkl0qtoTzqdrvODstvtePbsGX7/+9/fWGibJ4g3cx++i4jF2//mv4Hz5Uus/uv/esf3cTgcePz4cUdmjKIo4vnz5/jjH/9IW/zNkAggOaYaI37mTr3G8ThbW1uIRqPUaoPBYILnzywuLrJNsAU8z+PBgweYmprCp0+fOqp3OD4+RrlcRjAYRCgUQi6XQ6FQQCaTofOYCJIk3ShXTqI4sizD5XLRzy8Wi0HTNDgcDgiCAE3T6ATp79+/4+LiAg8ePKBdLna7vadX4+OCzWa78ZWxIAhdRU7I8WAePtktpGC50+9xL7/vqVQK+Xy+znSTpF7Nk9BbQVr3gcvjWtO0uoJr81p5nseTJ0/gdrt7tn4CcTEnn512eIii39/V/Z8+fdq1e/Py8vKVVFqjEWGzGiZzqjIej9c1oxDrgvn5eWpW2i3pdJq6iQ9qphejdzAfHkZXaJqGZDKJdDqNk5OTju7TWFxontBMTuS/+c1vuo4o6LqOo6MjZLPZuplGZJgh8f04PDzE6uoqLWwlz5vJZJDP52Gz2XpmGjdObG5uXonEdUqzgtJOuI2Tcrf3vekaW8FxHB48eFAn3iuVCt69e0c7yZqJrFbrIFYA5r8FAgF4vd6eD+g18/79+zp37U5xu920W+0mvH37tk5gk4sU83tmTrmR2ixyUcPzPB4/flz3XdZ1/VaRyt/97nfQdR2Li4u0oYIx3DAfHkbPIL4UwWAQ4XAY+/v71xa3Np7MzSZ0pH7jJiclnuebmhw2doM01hGQE6jb7e7LVfK4cJuZaIOYMD3ICC1pmd/b20MgEKCvX1EUPHv2DH/6059ot5lhGDR1R2ZfNcPceUa6nB48eND3gnyn09m14FlcXMTMzMytPgOLxVIneFRVhdVqrUs7NrMzML+nX79+xQ8//EBF122Ow2KxSD+bdDrNBM8YwAQP48a4XC48ffoU5XIZ6XQaZ2dnV06UzXw+VFWty8Wz6MrwoaoqSqUSvaLWNK2raMhNNr67npPVqyJ2UmMDXL5vqVQKflMayGazYXJysi4i2sl7yXEcJicn4fV661K2/aYbjzKO4xAIBG4tdsgoDeAysrOysoJsNovj42NqNEmerxFSt0V8eD5+/IiHDx/e2pTULLTa+fUwRgcmeBi3xmKxIBKJIBwOo1QqIZFIoFAoIBKJwO12o1Qq0doZ4opLNjZd16EoCnZ2dmhomkQWfD5fndU74+4gDr7kvW/XIt6LaA5pO77pfQVB6GjOF4F489ykMLoRWZbrhNrR0RE8Hk9dJOYm75Hf78f8/Pyt19ctXq+3I/E5Pz+PiYmJnnV2LS4uolgsYmpqCoqiwGq14uTk5Nrjgkx8r1ar9Ph8+/YtXr16dSvR8/37dwCXdhlLS0s3fhzG8MAED6NncBwHm8125WrIZrPh4cOH2N3dRSwWq4v4cBxHTyyNEAfZx48fDyRFcl+pVqvY3Nys22g4jkO5XKa1V6R2QlXVpimZZsLIMAw6TJTUs6iqSh+vG8Fiptv29XZrvAnk2CRzvFKpFN6+fYsHDx7QYtlOxZyiKPD5fAgGgwOLfEqShJWVFWxsbLS93ezsbM++l6SN3YzVau14BptZnOm6DlVVsbu7i8ePH994TU6nE8VikRmajhFM8DDuBI7jsLS0BFmW8f37dzoKoLEewfx7VVVRKBSY2LljWqWviOghgqXb6Ig5NdEoUu5yQ+lH6owY4JFjVVVVfPr0CQsLC5iYmEAwGLx2WGcv6mB6RTAYhMPhQD6fb3mbSqXS11l2kiTRWpxmnxeZmQbU2xuQVvV0Ot2xPUIzVlZWoGka4vE4LBYLnQE2DJ8P42YwwcO4MziOw+zsLCYnJ1Eul3F+fo54PF7ngEtSXmTDZYP77h6r1Yrp6em6mhPDMFCtVuncqU6jMaQgXdf1uvtUq9Wm3Ulm4UAGlTbbtMjvbiJeei2gyXvT7PcHBwfgOA7hcLhuijxZO/GGstlsQ+UEzHEc3G53W8FTKpX6Prw3GAzi+PgYwK+F4aqqUsHcGIUkw0kB0Asms1VAN3AcB4fDAbfbjUQigVQqBZfLhSdPntz+hTEGwvB8wxj3BlEU6WwskrfXdR1fvny5YgTHojuDw2zZb45edCowSFFwqwJdjuPqRE+j7wrxptE0rW4TIy3IoiiiWq3CYrHQ6FM36+oFPM+3FH9kFML379+pL1Qmk4HL5cLq6mrfxcJtuc675i4iHbOzszg4OIBhGHj8+DFOT09xfn7etOar2ed6GxdtnueRSqUwPT2N6elpVKtVbG1t0blejNGD7SaMgcJxHOx2O5xOJ1ZWVq4InNtuCul0mg5H7PcMpXGC1I+QTUSSJOr8K4oiPeHLstyyaPW6sQ21Wo127JnTEwQiuHRdp07DJI1GBn6Sq3qzS/N13LRQmef5K6/1uuckm/Le3h6WlpbwV3/1V3j27Fnb47parfakmPq2BIPBlp+t1Wq9diRELyDHDxl7sba2ht/+9rd0UGsjPM/XCaHbmAVyHIeHDx/SDjlFUfD48WNkMhn8/PPPQ/EZMbqDCR7G0OD3+/Hw4cO6je8mDqm1Wg3lchk7OzvY2NiA0+mEpmn48uULTk5OmlrYM+pxOp1YXl6m/i+kbqdaraJWq9GoBumMadz4BUG4VmAahkGnc1er1RtHXsj9Oo0GEuftbiF1ZQRFUTqKdlUqFepg3CwqYhgGUqkUotEoNjc38fPPP9/KA6lXCIKAlZWVpn8jUbV+I0kSXC4XnE4nFdmCIGB2dvbKuYGkQ8mx6fF4btxBZhZasixjY2ODmnCSY+cm5oyMwcJSWoyhIhAIwO/3I5VKQRCEK3b97VBVFRsbG8hkMrSLyDAMFItF6piay+UgiiJevHhB60QcDgd4nke5XEahUICmabDb7cjn87BYLPfWnDAcDsPv9+PLly9IpVJtb0ssBYDLDbyx5ZvMhCJ/FwShLg11m7Z00i7fSZqBRKmAzrxwzPA8D1VVaV1Sp6k9knr78uULnj9/Dl3XkclkqH8VGXZrfg+HJWUSCASuLV7uJyTK1xjR4TgOq6ur2NjYQK1WqxtlIggC5ubmblwAbhgGdnd3qdgjUcRisYjDw0N4PB6Ew+F7e14YZZjgYQwdFouFzhHa3d2F0+nsaNZVKpWqc30mzrZ+v58WPgKXG92f/vQn+rPT6YTX68XFxQVKpRKdHZZOp/tq4T8KiKKIpaUlfPr0idbXkKnd5toV8zRv4h5M6m8AXGlfN2/uhJtEeIhoUhSlo82NFLuSFvlardb09ZDHJrchg0ZJ3RFpqe8EIpLy+Tx+97vfXfs6XS5Xz7xtekE4HK4TPHa7Haurq7fqgOoGXdebNi84nU785je/QTabxcePH6GqKmw2G549e3ar9y8ajdIxNYTHjx/j7du38Hg81N29347XjN7DUlqMoWNpaQmvXr3C8vIyLWYuFAoAfp20/ssvv+D79+80QlCpVOrmPhmGQQtcSaSnFblcDoeHhzTVRZ7z/Pycpb9wOf365cuXdGRHpVKhs8oIpM6GpLxI2svs2WOG1OCYa4Fu4osjyzI4joOmaS29f8jGTNJnwK+1NeTqnXSNkUGRRECR+iDSoQagzjizUyqVChVWzeA4jgrCYXMebxQbwWAQkiThw4cPfXfGLpVKqFaryOVyLd87u90OXdfhcrluJXYMw8DJyQkURbniNk0EMjEgdDgc+PLly62Kohl3DxM8jKFElmVMT08jHA5D0zS8f/8eHz58wB//+EecnJxgdXUVpVIJ7969w8HBAd6/f18X3SEpklqthrdv395oDWQ4aSKR6GlnzyiiKApmZ2evTLknokfTtJYzjswzoQgkOkLu0+z+10FSTABoRKnZuoFfxz80tjGbU0dEyJC6pHabWbfHAxFPpPBZFMW6FJ855ddowDdoLBYLjXSur69jbm4O6XQamUwGOzs7fX1u4qXz4cMHpNPpprcRBAFra2t4/vz5rSI7xWIRXq8XPp/vyt9kWcajR49QqVSQzWbxxz/+ERcXF/RCjDEasJQWY6hZWVnB7OwsarUatra2oOs6nj59ClmWMTc3h1gshqOjo6bpBVVVbz0RO5vNYnNzE1arFY8ePRqa2opBQEzXyMZsFgSkTqVTSGs54SZXyoIg0AiDWYAQUdPMq8WM2QSx07Wba0VuAomEAZfvGSmAJb8LBAJD2a6+vLwMj8eDQCAAjuNoUfX5+Tnm5ub6NmvK7XZjeXkZOzs7iMViLaNfjSmom9AubX52doadnR2IolgXSSwUCjdqrGAMBhbhYQw1HMfBYrHA6XTi1atXmJ2dpZubzWbDxMREW0EjCAKdm3STegNSnJrJZPD169cbv45xwG634+HDh5AkiXYoaZpWF51oRjOPHNL11S3kfs1mZ5H6GqB+oOd1EGHcCaTuqBf1G+SxzO8NqQ8ZNjiOQzAYpN8hi8VC/9bPbiWHw4Hp6WlMTk4ikUj0bBxIIyT12epvyWQSDx8+rJvEbrfbbzwOhTEYmOBhjBSzs7M0bM1xHFZWVtpuPsS7hRSO3iRtQk6E6XQaX758udf+G8FgEH/xF39Bp4FzHHftSb+Z2CSdXIRORKksyzR11axglnzO3QopXdeb1hm1ots2+FaQmiLyOgKBwMhEC1wuF+1Wu6mTcTfMz8/DYrHgzZs3NNLbS75+/YqPHz82/Rvx4/H7/bTWijQ2NEt/MYYXltJijDSkDqOT4uLGOpFmnUJmzDUiwOVGd3FxgWq1CpfLRVve7XY7fD4fPB7PvZizQ/xZHA4Hdnd36+z+m4kfTdOuCBSbzYZIJAKfz0c9agAgn8/jl19+uVIMS1IJrbqjuklLNYNEjch6r4O4P9904yWjVMh7wvP8SE3ktlgsePnyJRRFuZW5X6fIsgyPx4NSqYREIoF8Po+nT5/eehwHGf8Ri8UAAPv7+4hEInURLODXNKkgCHj+/DkODg4gCMJQph8ZrWGChzHydBNxIVfnndT26LpO3YLNM5DK5XJdGL9QKODo6AiyLOPhw4d34kA7aDiOw+TkJDweDz59+oRSqVRXO+P1emG1WsFxHHK5HCqVCnK5HNxuNxYWFuB2u5uKQ4fDgYWFBeqOTbhJdI4gSVLLyA+pySKPbW6lb4e586vTNAsZiks8n8yvZ35+fuQ2z35Edk5PT7G/vw+LxQKXy4WlpaW644oI4Uqlgp2dHayvr9/4uYjtBRE7wKXgSafTePz4cUsx5XQ6bzWFnTE4mOBhjDQcx92onqIxHdEYgSBznMzdRKSbptUGV61W8csvv+DRo0cIBAJdr2kUsdls+OGHHxCNRlEsFuHxeODz+ZqmezpNG01MTCCfz6NUKsHtdsPtdiMajeL8/LzlfUjKsjHiQtJGpDvKnL7SNK1pVIi0upM1NxPG5snv5HEbo1JkPSRdR1r2gfo5VH6/HzMzM9e+L/eBiYkJpFIpXFxcIJvNQtM0rK6uguM4TExM1HWFJRKJGz9PtVrFp0+fqMcSSV3LsozV1dWhGuTK6B3sU2WMPBaLpWu/HOLOSq7mSWpM13Val2LePMkJ0ezrYn4sAsdx+PLlC169enUntQ3DAM/zLWcbmelUmBIXXTMOhwOGYdRdjZvRdb0u2mIWN0Sskr8RY8RWNDNUNNOYGiVGhhaLpc53iHjv1Gq1liLZbrfjwYMH9yIV2gkcx+HBgwcQBAHRaBRnZ2dQVRXr6+sQBAHhcJgK35sWjpMxM7lcjha6V6tVKIqCFy9eMLEzxrBPljHSXFxcXDv2oBnm9mrz7xr9YYDLNAcRNWQ6d+NUb7KBkonfsVjs3gieu0CSJCwvL7cUPMBltxOpvSAipNmm2I24IAK3MRLUmPYiPjvmv5MC11ZF1Iqi9KQOZdzgeZ6mqqLRKHVAf/78OVZXV3FxcQFd11Eul7sqNs9ms0in00gkEshkMvQih9SGEYd3xvjCurQYI8ve3h6+fPlyo/s2GyBJxgw00piqqFQqEEWRFmyS/zcXzl5cXNx7s8JeUyqV6NTsxn8kDWUYBm3zJlGe29LscySDT9vRbvyE2+3Gq1evhmqExLCxtLREfXfy+Tw+f/6MQqGAx48f04uJzc3Nto9BPn9iXrq3t0fr70g0jgjgTCaDaDR6o7XexuuLcXewSwvGSKJpWtur/WbwPE9bnw3DuFJnIknSlWJSgqIo1K+lVqtB13V6kmtmr18qlfD161c8fPiwqzUyWpNMJmkUhUTVZFmGpmk04mLewIYxTcRxHBYWFm482PI+IUkSnj59io8fPyKVSiGdTuPdu3fgeR5ra2vIZrM05dUsSqbrOj59+kSFDYn4ke7LZp19iUQCoVCIRnpIC7oZTdOQzWapGCODYZeXl0fGVuC+wgQPY+RIJBLY3d2lefduzcjMtRwkBULqLFptQuQ+3XSE3UW77n3CMAxIkkTTR6TWyjyd3fz59CK60+r4amwr7wSr1YrHjx93NAiXcQnHcZiZmalLW+u6js3NTTpvr9VnUCgUkEqlaPSPfJZE/DQrWOd5HslkEsFgEJlMBp8/f4bNZoPdbqf+T+l0Gg8ePKD3IV2b8XicCZ4hhwkexkhhGAaOj4+ptX25XK7rqqnVai1TSaS+htBYd9FLBEHA1NRU0ytExs3IZDJNu5zM3XRmepFSbCWaVFXtSuw4nU48ffqUieAb4PV6Ybfb6+ZW6bqO9+/f4/Hjxy1tIMhn10zYNPvsbDYbLSDXdR3n5+fUWZxEhUqlEiYnJ+t8eogPVygUuuUrZfQbJngYI0U+n6+72iMDQoFfzeNI2qmRWq3WtHW5H2iahp9//hnApTvx3NwcVFVt6T/DaE+xWGw5woBMXW/slqvVard+r1vVZpiPIZLebDVnzWq14smTJ0zs3BCO4/D8+XN8+PChTvS0e8+BSyFynbmoGbPTdzqdhiRJ8Hq9tMNPVdWmM7sEQYDf76+bJ6brOorFIqLRKDiOg9frpbPoGIODCR7GSLG3t9fyb2SwI/HTaLyyIzUfdz3/JpPJ4N27d9B1HaIoIhQKIRwOw+123+k6Rpl2kbjGaeyCIDQdPdEtZA4bx3F00yQt7URg6bqO//0/+U8gxGL4P/8v/0vd/Ylx5fr6+r0eOtsLJEnCysoKPnz4UPf7Dx8+4MGDB02HioqiiOXlZWxtbV0b7XM4HHVpqouLC5ydnQEAFS1Pnjxpef/Jycm64+34+BiHh4dUMB8dHeHFixfsOz9gmOBhjBStNjGy6RHBQ1IOZOMjJ7xBdE6ZBZaqqjg9PcXp6SnW19cRiUTufD2jiNfrhd/vrzOb4zjuSus3+bx7Me+MjLMgBetmSA2Roih4/nf/LgpNrBEURcHq6irb5HqEx+O5cgxUq1Wcnp62nKJOjAyva3CIRCLU6dowDGSzWQDAwcEBfD4fHj161LZlvfG8NDs7i5mZGWxsbCCXy0HTNOTzeXosFItF7O/vI5fLwWazYWpqikWA7gAmeBgjxfz8PLxeL3Z3dwH8ahjY6IhLDATNw/46GSdxl5ydndGrSBLyFkURuq7D4XCwk18Da2trePv2LSqVCv3MSdSMCJDG1NZtMBdDt/p7pVJBaHIS4uxsXdG7JEl49uwZK1DuMaurq/jTn/5U9z2+uLjATz/9hLW1NTrU1ow51dQKcwQulUqhUChAEAQ4nU6kUqk6sdKMWq2GXC4H4HKwKkmPTUxM0PRWOp3G5OQkCoUCjfgCoPPBrFYrnE4n/H4/gsFgV7V/xWIR2WwW4XCYnTfawAQPY6SQZRmnp6f0Z+Jo2wxy1U+iP2SDHJYTgrkmJZPJYH9/n/5ss9mwtLTU9AR+X5FlGc+ePaNpDSJEzAXEvfxsiY9OswiPGTJrzWxIGQwGmdjpA8Ss8cOHD3V1VNVqFRcXF02/L/Pz85idnUWtVsPJyQl8Ph+SySRSqRR1WzYXPtvtdjidTvA8j1AohHQ6jbOzsyuCh7Sjz8/PA7i0TeB5Hqenp5iZmaFjUXZ3d6GqKi4uLvDmzRt6HmqkVCpR09KdnR2EQiEEAgG43e5rxY/NZoMgCNjb28Ps7CyrF2sB1y7E//r1a+PNmzd3uBwGozXlchnv37+ntRPdtKSLoghJkroeQTFopqen6wYoMi6vZt+9e3dFhHQz7bwTiClgp+3nJIpYrVbxF3/xF1cmbjN6RzQaxebmZl2Kenl5GdPT0x0/Rq1WQzKZhNfrvVJjRawvXr9+jY2NDaRSKbx69epKtMgcSW78PREpiUQCDocD2WwWOzs7bc9ZzZoqRFHE0tISJiYm2r6earWKTCaDZDKJtbW1a1//uMJx3FvDMF43+xuL8DBGBmI8R/xzOkldkNvVarWhSmd1yvHxMcrlMubn52Gz2ViLOy6vZtfX17G9vU27scyGkL2CpKw6FZu6rqNSqWB1dZWJnT4TiURgt9uxubmJfD4PAF0XhkuS1LTrCgCNwpA5cS6XC7u7u1cKl1t9H82/J1GnYDAIv99Po1EnJyfUFbwdqqpic3MTlUqFRpMa1yoIApLJJE5OTmjxNRmLws4Zv8IED2NkCIVC2N3dhaZpHUV2yKRsYDhddzslHo8jHo8DABYWFjA3NzfgFQ2eQCCA09NTJJNJAL2L6pi5SacXGXDJ6D9OpxMvXrzAp0+fkE6ne/od53meChWfzwee5289BiSbzcLpdMJisWBmZgbT09M4Pj6m9YjXcXx8DJvNhkAgAJ7ncXZ2Bk3TkEqloCgK8vk8lpaW6qJQh4eHmJubG+nzXy9h0o8xMpA0Vqf0qnh1mOh2nMY44/F4mnbO9KoFnFwhd4OiKGwA5R0iCAKePXuGv/k3/yZisRh++uknvH//HrFYrC7a160gDgQCWFlZoT97PJ5bC9nDw8O644njOExPT3dc66WqKjY2NvDTTz8hkUggGAyC4zhaXP38+fO6WiSSamt83vsMi/AwRgZZlvH69Wtsbm7i/Py87W1vMnJiFCgUCrRV+r4zNTWFk5OTK5sZESrmQmazb06nqKpaV4jcCWyG2t1D0ppOpxPxeJzWsrjdbjoBPRqNwmq1wuv1YnFxcSBpnpWVlSvPy3EcfD5fnaHidUxOTsLtdkMURUxNTWFqaqrlbWdnZ3FxcUHfEyKS7isswsMYKcjgQJ7n8b//R/8Rzg4OrtyG+PCMK+P82rpBEAQEg8ErvydT70n9AhkpQlrYu6FSqXS9ObL6nbtHlmXMzs7i9evXWF1dRSQSgdPpRK1WQzQaha7rKBQKOD4+xrdv33BycoJUKnWnkY9mFymdeAT5fD5EIhF67J6fn9PROtfBcRxCoRAURYHP54NhGLi4uEA8Hr+XUR8W4WGMHDzPY319Hf+/ahV6k1A1x3Ft/VNGGUmSOvIVuS/4fD4cHx9f+X21WqUGhOaojtmQsh1EKAGXx1MnxaUEltIaHHa7HXa7HZOTk/R3qqoik8mgWCzSNHc6ncbu7i7C4TBWVlbAcRzi8Tj1w5mdnUU0GkW5XIbb7UY2m4Xf76fDQWu1GtLpNCqVClwuF1wu143WWygUEAwGEQgEkMvlcHBwUHe8Pnv2jJoq6rqODx8+IJvN4v379/jhhx+6OhcQwRQMBpHNZpHL5WAYBhwOx705ZpngYYwkoVAIf+uf/JO6uVpmyuXyWIqe+x6SbqTV509ofK86ETtA80GzzQwum3F+fo7Z2Vn2OQ0JoijC7/df8egxt47XajWUSiUEg0E4HA4YhgGn00kHE09NTUGSJOi6jmg0ivPzc1itVkxOTt5qQrq5jd7j8cDpdCIajaJarULXder+DPwa3X7//j1UVcW3b9+wsLDQ0mW6HS6Xi0Y9Ly4u7o3jOxM8jJFldna25YZndrwlQwYbazHIzC2e50ei3sflcmFpaWnQyxgqur2ybjVnrZP7kRZ1nufbjq8oFotIJpPMNHLIMacqJUnC7OwsgMvP+uDgAPF4HI8ePaoTHXt7e4jFYnj58mVfUpcej6fl9HfgMoL14sULfP78GdlsFl+/fsXi4uKNHJZjsRgEQYDL5erJ7LlRgAkexsji8Xjgcrno3JtGNE2jm1Lj5kSs32u12kh82d1uN548eXJvQs+dEggE4HQ6qa1/J9zmszYMA5IkXTuva3d3F3a7ndXzjCAcx2F+fh4TExNXvm+hUAjz8/Nd14J1C0lrNasfs9vt+OGHH1Aul+lYCo7juu4im5mZAXAp0I+PjzE9PT3058HbwgQPY2ThOA6Li4tXJig3Q9M0yLIMQRCgaVrdFb558OgwYrFY8PTpUyZ2msBxHK1/6BRVVWl6QhRFGIZxJeLTKu1FbiuK4pVjhqQ/SIEsGTfBGE2aFRl3E1E0DAOxWAyKorSN2jTj8PAQFoulZaqJ53nYbDbYbLZbRxLJ49yHImYmeBgjTbMJys0gG1UrYdNYszFMLC0tMbHThm4/NxL5I8XIHMdBkiR6Nd3OXdnc6k6OI1IrZr7fTa6Ua7Uam4E0JhBrBK/Xi59++gmhUAg+nw8OhwPVahWqqrasxyOdVIqi3GltzbhHdwAmeBhjwNLSElKpVEcRGl3XIUnSFVPCSqVCN71OClPvisnJSQQCgUEvY2jJZDJ04nynNJt91Hg8kBlJjcNmiWA2pxrIsdL4mOVyuauIwNu3b/H06VPWhTfi6LqOz58/17W9x2KxuvZzUkjd7EJGVVUUCgUUCgVks9kbd4AxrsJ8eBgjj81mowWHndCq9qJWq6FSqdAiZ1LQPCgmJiZoyyzjKrVaDZ8/f+46FE+ETKOYMaNpGhRFufJ3EoEhEaJ2dNMhmE6nUS6XcXR01PF9GMNJLBZDMplseVyKoojHjx+3jNqaaxKvi1wzuoNFeBhjQaf27AA6igSRmg7SxjwI0eH1epnYacP5+Xnfxofoul7nvWMYBi10J4NKBUGoS201e4xOSKfT+PjxIwC0LMBnDB+VSgWHh4fQdR2Li4uQJAnFYhGbm5tt7+d2u9vW9JiFUrFYRLVapYXyjNvBBA9jLAgEAh3V8kiSRE3pOoG0MQ9iLtd9KCK8Ddd58LSim/eVGBDWarW6rj9FUVCtVtuOMCmVStc+frVaxd7eHhVHpVKJTr9mDC/RaBQ7Ozs0nZnNZrGwsIBCoXDt8bW4uNj27+YOsIuLCySTSYRCIayurt656DEMY6wsFlhKizEWcByHx48fX1vgSzxUumFQ9TzjZprYS1RVvbHg6ZTGsRRmOjmG2rUJG4aBQqGAN2/e1EV1LBZLVx1njLsnHo9jc3Oz7rxQKBTw+fNnfP/+/dr7f/nyBel0uqUwauwY1DTt2tmB/YLjuLFKqw08wlMqleqMnRiMm8JxHGZmZhAKhbCzs4OLi4srt9F1vesUlSzL9OTUroOn17ABoa3JZDI3thEQBOFaESsIAgzDoI7dgiBA13XUarW6YuZWonRqaqqpA2+1WsXp6SlisVjTeUgzMzNwu903el2M/mIYBtLpNI6Ojm4VfS0Wi/jw4QPcbjfW19ev7H+tzlulUmkgBe1zc3NIp9Ndt9abIaM4ms2+u0sGHuFhPhWMXqMoCh49eoSXL1/C5/PV/e0mYqVSqaBarXbtzntbHA7HnT7fKKGqKgRBgKIoXad/yFDRVpAREmZRQ1qJyfmqVquB5/krwslms2FlZQXLy8tNH/vi4gL7+/sthz9aLBZWqzGEVCoVfPz4EYlEAplMpiePmclk8O7dO3z//p2mzGu1WsuIyqAK2hVFuVWnmKZp1Fdo0Aw8wsO+3Ix+4XK58PTpUxwdHWF3d5f+/jbHHKnZ6PdxK4oiZFnu63OMMoqiQFVVGmEx1z2Yr76JR465K4vneRqx0XWdmlFqmkZrvFph/tzN0R2bzYbl5eVrC83btdDLsnyjuUiM/hONRpFKpXqeRq3Vajg4OEAikcDExAQSiUTLyOXZ2Rktjr5rbtqturW1BU3TsL6+PhR7/cAFD4PRb6anp2Gz2bC9vX3ruphqtUo3y35yH2zeb4PD4YDVaqWdVK1SVObPm7gjk1SVLMtQVZUa/um6fq2YbZbKcDgcePr0aUcClWxWuq5j65/8E0z+zb8J15+jkCyiN5zouo50Ot3X58jn89je3u5oLaOE0+mk3YzDwMBTWgxGv+E4Dn6/HzzP07EAN4VEA/rNbSYw3wdEUcSTJ0+6OpGSz5+kD8zWA4RuvXUmJyfx4sWLjqNxpNsleniI/9N/99/h5J/+U/q3UCjU0WMw7pbd3d2+F8h3yt7e3p2n1m/DxMTEUB3XLMLDuDdYrVYUi8WOilYNw4DFYqlzXSazkvoteERRZIKnA+x2O7xeL5LJZEe3bydmVFXtKF1pDu3Pzs5iYWGhK9FFaiEm5+fx/r/9b7H84AF9XOaoPRwYhoFKpYJyuYxYLIbT09NBL4lyfn6Oi4sLzM/PY2ZmZmgiJ6MCEzyMe0M3kR2LxXLFX4X4rvSbBw8esPqdDpmenu5Y8FzXWUM+X1VVW6YOzEI5HA53veGY24vnnz6l/+/3+/s+gZvRGk3TUKvVkEwmEYvF+p7Cug26rmNvbw/pdBpPTcdQrzg7O0OhUMDi4uJAneb7wXi9GgajDaSbR9O0oTX18/l8Y2PydRd4vd6OO0g6ESfE1bYZiqLUCaHDw8POFvln8vl8y6LlYQr73ycMw8Dh4SF++ukn/OEPf8DW1tZQix0zyWSyL/48ZGDphw8fhnag8k1hgodxbyBX52RWFnBZRCrLMhVDkiRBEIS2zsr9FEuNbfSM9nAch/X19WujI4qidPy5kUGyxM6fuG03nvxjsVhXppTb29tNI0c8z7PP/Y6p1WpIpVJ4+/Yt9vb2Rtbk8+DgoOfnI47jsLy8jMnJSfzyyy89a8MfBpjgYdwbzFb/xEVXVVXqsaIoCmq1GlRVbbmBVioVGubtdXrr0aNHmJ6e7ulj3gdsNtu1ERJd1zsu9uQ4jtZ5kc2k2aZiGEZPNgOfz8dGSdwxZ2dn2NjYQD6fH/RSbkWxWMRPP/3U1u7gpkQiEbx8+XKsDFBZ0phxbzALnkqlUuegzHEcvYIn/2+xWFAul6+kQsh9bjNfi+M4hMNhmpLheZ7V7dyCXnmTmB2WCRzHtUyHdbMZBIPBpgKJRXfuHsMwBjIfrx/UajWcnJwgEon0vIhZFMWxqi0bn1fCYLSBRG4IxI+lFUT0KIpSJ4bMaJoGi8UCwzA6jh6QaEQoFBqITfy40ssISbO0U6vPtxtflFYjI5jguXuGqfOqF+TzeWxubmJ1dXXsCo17CRM8jHuB2cpfFMW2nThmyEYnCAIEQajb+IgQ6jQyMzc3h7m5OXZC6gPXtfFfd+VLTAnbCVezB1MgEMD09HRXlvtWq5UeewS/3z8Ulvv3CWI5MW4FudFoFJVKBY8fP2Yp0hYwwcO4F7SaXdQpZPRA44YFdNb9s7CwgLm5uVutgdEal8tFC4ybQeZnEZFLPJVIWqOd2CHDZlVVRTAYxNLS0o0GHouiiB9//BEnJycolUqYnp5mfksDgOM4LCws4MOHD4NeSs9JpVLY2dnB2traoJcylLBLTca9wFyTcRuaXTnVarW2tRx2ux2zs7M9eX5GcwRBuHYOFanzkWUZ1WoVlUqFztNqF9khn60oiohEIjcSOwRZlrGwsICHDx9Skca4WwzDQDQaHduauVgsNrS2G4OGCR7GvcC8od00pdQsCkAma7eDtDcz+ovH42n7d8MwqJtyt5DW9FGy9Wc0h0QCx/WzHLV5W3cJEzyMewHZ5Hiev7HnhqZpEAQBiqJAURS6AZJurlZ049XCuDnhcLitmCVRnW7FJxkYWywWWYHxmLC8vDzoJfSVXkW0xw0meBhjT7lcRiqVgmEYtxr+Seo4KpUKTYfwPH/tJtpNYSvj5iiKgocPH0JRlLo2dcMwbh3i13UdkiSNlSfJfUaSpLEVr4ZhjF0XWq9ggocx9pyfn9OURK+8NyRJgqZp14aPXS4X5ufne/KcjOsJBAIIh8PQNI1G4sh8rE4gaS/iuE0QRREPHjxgqckxYnl5eWw/T9al1RwmeBhjD/ny96odnDg0Xxc1kGUZT58+HdviyGFlbm4OoijSSFw3tRqkXblardLNUBRFvHz5ks04GzNsNltLb6RRhx2rzWGChzH29LKGhmyCnYiY5eXlsXIpHRUEQcDi4uKN7qtpGp2jRY6bxcVFZhI5poxjilIURTgcjkEvYyhhgocx9qTT6TrPldsgSRJqtRrK5TJEUWw5lDISiSAYDN76+Rg3g4zt6BZVVetcucPhMCYmJnq9PMaQMI6z61hXaGvY5Sdj7LFarchkMtA07dYnAnJ/UsBMho6anZtnZ2exsLDATjoDhEx8/tOf/nTjx5icnMTKygr7HMcYp9MJt9t96yGwiqLA5/PB6/XCYrFAEARUKhUcHR2hWCzeqatzIBC4s+caNZjgYYw9kiT1rGi5WTSHtC1zHIelpaWxvGocRW7js7KwsIDZ2Vkmdu4BN0k72+12zMzMwOFwgOM42Gy2K8eK3W6Hz+eDYRhIpVLY3Nzsu/CRZZmdf9rABA9j7Emn0wB6U7TcqqWd53koisLSH0PETSao+3w+zM/PMyuBe4JhGF171rx69aqrkSAcx8Hn82F1dRVfvnyhkWBFUeBwOJBIJLp6/lZYLBa8fv2a1Q22gb0zjLGmWq0im80CaC1WOoVEilr97eXLl6wddIiwWCxX3JE5jkMgEIDNZoMsyyiVSrRIORgM3kgkMUaXjY0NFAqFK7/neR7hcJgWNZfLZciyDIfDceP5Z36/HysrK9jc3ITH48Hq6iq+fft2q/WbmZ+fZ2LnGti7wxhrzs7O6P83G/zZDYZhYGlpCbIsIxaLoVQqoVwuw+PxYGFhgZ1shgxRFPHq1SscHx+jWCxiZmYGTqeTiVIGZW1tDQ6HAwcHB9B1HRzHYXp6GuFwuC+dTl6vF36/H+vr68jn8xAE4dbnJeCyTjEcDvdold1BmkJGocWfa+cl8vr1a+PNmzd3uBwGo3fouo4//OEP9Aq/3TTtblhbW2OpKwZjjFBVlTY13JVvVjqdhs1mw7t37249CmJ1dRWTk5M9Wll36LqODx8+wGKxDIU5J8dxbw3DeN3sb6wtnTG2JJNJKnYEQejZBOFsNsumETMYYwSxmLhLk1CPxwNZlvHjjz9ibW3txo9jtVoHegHG8zwePXqEbDaLfD4/sHV0AhM8jLHFnM7q5QThs7Ozpnl/BoPB6Bae5zExMXHj2iCv1zvwqIqiKHj69Cm2t7dxcnIytBeETPAwxpLj4+O67odeCp5IJMKcTBkMRk+JRCJd34fn+aGpnSGNANvb27RRRFXVobo4ZFWWjLGksTC1Fx48PM8jEAhgZWXlVo/DYDAYjUxNTUGWZVxcXCCfz6NYLAIAdWy/uLgAcJnCWlxchN/v79l8wF5BGjfev38Pi8UCTdMQCoWG5pzJBA9jLPF4PPT/ZVnuugvC6/XCbreD4zg6cdvr9bJOLAaD0TeCwSAVOJVKBblcjg4CTSaTSCaTmJ+fH0r7hHQ6jWQySX8ul8uQJGlg3WPNYGdvxlhSKpXqfu42pWWz2bC8vNzLJTEYDEbHkAstgt/vH+op6Ol0uq52JxKJYH19fYAruspwxcMYjB5xfHxM//8mqaxxnKLMYDAYvYYYus7Pz+P169e0gHqYIjsEFuFhjCWkUI7n+Y6iO36/H16vF6lUCoqiDPWVFIPBuF98/PgRsizD4/HQ+p7V1VVYLJaBriuTyeDLly94+PAhdF1HuVyGYRiYnZ2F1+sd6NqawQQPY+yo1Wp0SF+nE9ItFgump6cxNTU18BZPBoPBMBMMBrG5uYloNEp/V6lUBip4arUadnZ2UK1W8eHDBwSDQayvr9PJ8cMIS2kxxo5sNgufz4eZmRkoitKRJ8TZ2Rk+ffqEw8PDO1ghg8FgdE4kEqnrPLXZbAO1xjAMA7u7u8jlcgAuSwCWl5chCAL8fv/QXjQywcMYO0grJBkl0cmXT1EUzM3NYW5u7g5WyGAwGJ3DcRzm5+cBAA6HA0tLSwObCVetVrGxsUGjTRaLBc+ePRuJukeW0mKMHYFAADzP4927d6hWqx158CwvL8Plct3RChkMBqM7pqenEQgEIEnSQOwxyuUyzs/PcXR0RG0+XC4X/H4/rFbrna/nJjDBwxg7NjY2UK1WkcvlwHFcR+ZczDmZwWAMMxzHDURYGIaB09NT7Ozs1JUHhMNhPHjw4M7XcxuY4GGMHdlslg4NBdCyhsdms2FpaemK3wWDwWAwLvn+/fuV2kaO40Yy/c8ED2OsiEajdWIHaC14Hj16BLvdfhfLYjAYjJFDVVVEo1FaDwlcWn0sLi7CZrMNeHXdwwQPY2yoVqtIpVKwWCwol8sALr+c1Wq1aeHy0dERVlZWBlb8x2AwGMOMKIr4y7/8S2iahlgsBlVVEQwGB+7/c1OY4GGMDbIsY2pqCplMhv5OkqSWxoPRaBQXFxeYn5/HzMzMXS2TwWAwRgpBEDAxMTHoZdwaJngYY4Omafj48WPdoFBie94KURTrBo0yGAwGYzxhgocxNqRSqTqxI0lS23b06elpLC4udtTFxWCMGwcHB3A6nXA4HJBledDLYTD6DjvTM8YGcy2OKIpXipfNKIqChYUFJnYY95ZYLIbd3V0kk8lBL4XBuBNYhIcxNphby9uNk5iensbMzAwrVmbca3744YdBL4HBuFPY5S1jbLDZbAiFQgAua3ckSYIkSXW3kWWZeu90ChFPlUqlo7lcDAaDwRg+WISHMVaYa3ZUVb0ibIgDcydjJAzDwPHxMU5PTyGKIgqFAmZmZrCwsNDzdTMYDAajvzDBwxgr5ufnUavVkM/nAVzOf+F5HjzPw2azwel0Ip1Ow+l0XjtUlOM4uFwuRCIRVCoVyLLMijsZDAZjRGEpLcZY4XK5MD8/j4mJCVgsFuoQSiYNn56e4uLioq6bqx1utxuSJLFOFgaDwRhxmOBhjBUcx8Hv98Nut4PjOHi9Xni9XhQKBeRyOQCXaa3Nzc1rJ6gzGAwGY3xgKS3G2MFxHKanpxEMBnFwcIBoNHrFbTkejyOVSiEcDsPlcsHr9bIBogwGgzHGMMHDGFsURcHq6ioWFhaQSCRQLBYRi8VQLpchiiJmZmZgGAby+Tx8Pt+gl8tgMBiMPsIED2PskSQJkUgEALC4uAhVVcFxHPPhYTAYjHsEEzyMe4cossOewWAw7husaJnBYDAYDMbYwwQPg8FgMBiMsYcJHgaDwWAwGGMPEzwMBoPBYDDGHiZ4GAwGg8FgjD1M8DAYDAaDwRh7mOBhMBgMBoMx9jDBw2AwGAwGY+xhgofB6AJN0wa9BAaDwWDcAGY5y2C0wTAM1Go1JBIJHB0dQdM0PHr0CC6Xa9BLYzAYDEYXMMHDYDRQKpVQqVRQKBTA8zwURYEgCFhfX4fT6QTHcYNeIoPBYDC6hAkeBuPP5HI5bG9vI5vN4tmzZ5iamhr0khgMxhihqioEQWAXTQOCCR7GvadYLOL09BTRaJROUlcUZdDLYtxzdF1HoVBAsVhEpVKBYRiQJAmKosDj8UAQhEEvcSBomobDw0OkUil4vV5MTk529H3VdR21Wg2VSgWlUglerxeyLN/Bin9le3sbiqJgcXHxTp+XcQkTPIx7S6VSwf7+Ps7OzgAAkiRhYWEBFouFTVRnDIxqtYqtrS0kk0nout70NjabDc+fP7/zDfsm5HI5HBwc4MGDB7cSabqu4+joCCcnJ6hWqwCAbDaLo6MjTExMYGZmBhaLpe72lUoFxWIRiUQC5+fndU0HPM9jenoabrcbAFqKSE3TcH5+jlKpBEmS6PmhXC7TiI2iKJAkiabCS6USNE2DYRgwDAPVahXVahW1Wg0cx8FutyMYDILnL/uGDMOAruv0PuR3HMeB53kIgkBvy7g5HHlzm/H69WvjzZs3d7gcBuNuqFQqePPmDWq1GgBgenoa8/PzTOgw7hxVVXF+fk4FTjab7agbkOM4BINBeL1eGvUxb4qGYSCfz8NqtfbtuDYMA9lsFul0GoZhQBAEVCoVGpEyDAPJZBKGYUBRFLhcLiiKQqMtoihCEASoqgpd12Gz2WC1WlEul2nDQK1WgyzLiMfjKJfLbd8Pc7qIfLc7hed5OBwOOBwOSJIE4PI8EY/Hoarqzd+kNuvleb7jzk+73Y4XL16wc9Q1cBz31jCM183+xt45xr1EkiR6JWW327G0tMTy6oy+U6lUkEqlkEwmkUqlwHEc3ey7xTAMxGIxxGIxAJfHdCAQgNVqhaqqSCQSKBQKAABFUWCxWKAoCmRZBs/z8Hq98Hq9N3refD6PTCaDWCyGbDbb0f0qlQouLi66fr5u1nUbYULEZqev57YYhtGVzUWhUMDnz58RCATgcrngcDhY1KdLmOBh3Auq1So4jqNXbtlslp4cC4UCPn78iKmpKXi93ntbG8G4HsMwcHh4iEKhAL/fj3A43PK2qqpScVMsFlEsFruOOnRDrVaj6dlGSNTFzMnJCV69egWbzdbxc2QyGXz79g2lUulWa2XcjHQ6jXQ6DeAyImW1WmG1WqmYtdlsNM15cXEBm82GYDDIzml/hgkextizubmJs7MzyLKMly9folqt4uPHj3W3SaVSSKVSEEURy8vLiEQiA1otYxghBcSHh4c0ShGLxbC/vw+r1Qpd1zExMYFAIIBqtYrd3V3E4/EBr7o9mqahWCy2FTwkalIqlZDP57G7u8vMN4cEckySKF4rtre34fP54HQ6adG7zWarq3e6LzDBwxh7kskkgMsoz6dPn7CwsNAyhaCqKjY3N+F0OmG32+9ymYwGqtUqYrEYcrkc7HY7VFXF1NQU7cgpFouIxWIolUrgeR6RSATpdBr5fB61Wo1uzIIgQJZlSJIEm82GycnJlulLwzBQqVRQLpdpsWsul6NFso2USiUa7Uin0+A4Du3qIocBQRAwNzeHUCjUdNMrFov4+vUrisUidF0f+tfDaI+mabi4uKhLJ3Ich8nJSbhcLgQCgXsTAWKChzHWkO4IQqFQQDKZxNTUFE5OTprexzAMvH//Hk+fPmWOyl2Sy+VgGEbHBo2kg6VcLqNcLqNQKCCbzVLzx0ZOT08xOTmJarWKi4uLumhDq3ROI8fHx7BarSgWi3WdNABuvcGPgjggNThOp7NO8JAC4+3t7bbFwYzRxzAMnJyc4OTkBKIo0mJy4Nfib4IoirQGzOl0jrQ4YoKHcSOq1So0TYOiKENVOJfP53F0dIRUKkVbQBs3oWw2i1evXqFYLCKVSjV9HFVVcXZ2xgRPC5LJJA4PD6FpGux2O0KhEKrVKr59+wYA8Hq9NNVDxAv5PHw+H2w2G5LJJPL5fFcFu6qq4vDw8FZrN0dl7iO6rsNiscDj8dT97uvXr30tKmYMJ6qq0ij4dXAcB5vNBpfLhdnZWVit1j6vrrcwwcO4ERcXFzg/P0e5XIYsyxBFEZIkQZIkqKqK2dlZOByOax+HeE30CvJYPp8PFosFpVKJ1ubY7XZYrVYEg0H6xW0leIDLTV1VVdYGaqJcLuPk5ATHx8dUSOZyOUSj0brbkZqoZpCuIsbgSKfT+P79O0RRRLVaRSqVurYWhMEwDAOFQmEkIpnNYGfyEYIcZINsn97Z2UGhUKB1EaIoQhRF2Gw2eL1euFyupuvTdR3VahWSJNGQaC6Xw9evXxGJROByuSDLMmq1Gmw2G+2maoQUUbb6u91ux4MHD+p+p2kaarUaqtUqKpUKkskkdnd3abdDKyqVCs7Pz+/1iAmSYsrlcshkMmxTHBPusv2aMV44nU48ffq05Tl4mGGCZ0QwDAPxeBzFYhEWiwV2ux02mw08zyOVSuHw8BCCIND2RI/HA6fTeeVxNE1DOp2+sTV9JBLBhw8fmvpd7O/v03QFx3FUYJjrMTiOg8VioSLEMAzs7e3VPY4gCHC73eB5nuaPbTYbRFHE0dERstksPB4PdSd1Op2IRCJQFAW1Wg35fB7ZbBaZTAalUqkvpmH3AdLdxmAw7jeiKMLj8cBqtWJubm5ko97MafmOMQwDxWIR+XwexWIRHMdBlmUaARFFERaLhdaeEDtyQRAgSRKNqJDNnuM4GlHRNA2qqoLneWQyGeRyOVQqFciyDIvFQm3PSQ7/4cOHkCQJhUIBbre7Y8Veq9VoHcTe3l7T4tJRheM4OBwOKIqCiYkJ+P3+QS9pYJTLZXz69IlFdRiMewjHcZienobD4YDP5xuZiA5zWu4jZPO3Wq2QJAmapqFQKCCXy9EOEFVVUavV6CDAu/axqFaryOfzdb8rl8t49+4d/ZnjOFitVsiyTIvSOI6Dpmkol8uoVCrUCl2WZVq1v7S0hI2NjTt9Pf3C5XJhaWmJzta5bxAxnk6ncX5+zlIeDMY9JRAIYH5+vqM6zFGCCZ5boKoq/vCHP0DTNAiCAEEQWvp1DDtksyMb3unp6aCX1DcURaEpQYvFAqvVCrfbPbJh2ptAuqcKhQLy+TxyuRxyuRxL/zEY9xiO4/Ds2bO6Dr5x4v6c4XtIsVikowlkWaaTcZkD6WAg3WEA6GRhkhLUdR08z4PneUxMTMDn8429sCHpUhKZq1QqqFardHYPGcjIYDAYZsLh8NiKHeCeC56NjQ1YrVbYbDaoqgpVVVGtVqHrOiYnJyHLMlRVpbUvuVwO2WyWmXINGaIoIhAIQFEUOBwOOJ3OofIG6jdk6GEmk0E8Hkculxv0khgMxghSKBRQrVbpPK5xY+wFTzufF7vdju/fvzf9G+tOGR3K5TIuLi6g6zpUVYXH48H6+vrQFNkVCgVa5F2tVmnXmcfj6VqYFYtF5HI5FItF+pj5fH5kfTEYDMbwkMvl8PnzZzx8+BCKogzUAqUfjLXgKZfLtIVakiQ4nU6EQiH4/X5wHEfnySSTSSQSiY7dJhnDw/LyMiYnJ4cyoqOqKr5//95yhIUoilhaWsLExERHj5dMJq8MPWUwGIxeks1m8cc//pE2sKysrLQdMDtKjLXgsVgsePXqFU5OTpBMJpFMJpHNZpHP5zE/Pw8AsFqtmJqawtTUFEqlEh06SDqrarUadF2HruvQNA35fL5rO3xG/0ilUlAUBYFAoC9XIzd1gs7lctjY2Gg7wkBVVWxvbyMUCnXkiSTLMnw+H/2ZuFuLokjrlMiwTQaDwegEr9eL6elp8DwPjuOgKMrQjQzqFcyH5wYQe+2zs7OWV++MuyUcDiMYDMJut1M3506HV5ZKJepnRIqeiZfR9vY2BEHAkydP2qbI0uk0TTFls9lrXZwBwGazwel0Ynl5uWfpt1KphK2trbYjMxgMBgO4dE1++fLlWKWumA/PDSD1IJIk1R0MtVqtrkCUMRycn5/j/Pyc/kyuVGw2G/UVIhOBzZ8nx3FQVZVGRojoIb8H0HbUBXB5rCSTSaTTaRoRlGWZmkqafYssFgtsNhusVmvXTtfmyeK1Wo0W2pNIJPGEKhaLXb57DAbjvuFwOPDkyZOxEjvXwSI8uEwtFAoFZDIZRKNRVCoV2mLO8zydmF0sFkfWZ4dxiSAIsNvtsFgs1LWajK8w/+tFOJeIZvIdMwwDgiBAFMUrJ5larYZkMkm9cMh9dF2/0lrOYDAYvUCWZUxOTmJ2dnZsUljtIjxM8DRgGAbK5TKtwWDcP8gkdbvdDofDAbvd3tKYkBwv+Xwe5XKZ/iM+OK2+X2RUCBFYqVSKeeMwGIyBoCgKwuEwHQB9kzmLwwJLaXUBGbEQi8UGvRTGgCA1WoVCgR4HoihidnYW4XCYFr+T29zEcJIYVTJPJwaDMWgqlQoODw9xeHhI5wm6XC44nU7Y7XZYrdaxMGwdyVdgGAYMw+hbCE5VVaTTaTgcDtRqNZZKYEBVVezt7V2Z7M5gMBjjhGEYdNSMGUmSMDk5iampKdoZOmoMXPAYhoGTkxMUi0XY7XbIsox0Oo1KpQKv1wun04lUKoVkMolqtUoLNQ3DgMVigcfjgdvthsvlgtVq7cmHIIoinj17Rn/WNA3n5+fI5/N1RaJkOKi5RoPUbTAYDAaDMS7UajUcHBzg4OAAwK9peY/HA6/XSycVkAABmVNI/g2DEezAa3hOTk6wvb3ds8cTBIF2xpjfcFmW64pUSXhO0zT6IdVqNdqWTP7xPF93+0Z0XUe1WqVFpclkkrk0MxgMBoNhwul04tWrV31/nqGu4el1qkjTNOqHkslk2t6WDJjsBFEUaVuxYRi0c4YVmjIYDAaD0Z5hGK49cMEzSLoRW6qqUpdlBoPBYDAYo8XoVR0xGAwGg8FgdAkTPAwGg8FgMMYeJngYDAaDwWCMPUzwMBgMBoPBGHsGXrTscDgwNTU16GUwGAwGg8HoE7IsD3oJgxc8Ho8HHo9n0MtgMBgMBoMxxrCUFoPBYDAYjLGHCR4Gg8FgMBhjDxM8DAaDwWAwxh4meBgMBoPBYIw9TPAwGAwGg8EYe5jgYTAYDAaDMfYwwcNgMBgMBmPsYYKHwWAwGAzG2MMED4PBYDAYjLGHCR4Gg8FgMBhjDxM8DAaDwWAwxh7OMIzWf+S4N3e4FgaDwWAwGIzbEDcM46+b/aGt4GEwGAwGg8EYB1hKi8FgMBgMxtjDBA+DwWAwGIyxhwkeBoPBYDAYYw8TPAwGg8FgMMYeJngYDAaDwWCMPf9/BdMEzrCU+PcAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "geometry = []\n",
    "\n",
    "for rec in metadata_df.iterrows():\n",
    "    sample = rec[0]\n",
    "    # Terminal nodes use the \"lat/lon\" attribute        \n",
    "    if rec[1][\"node_type\"] == \"terminal\":  \n",
    "        province = rec[1][\"province\"]\n",
    "        country = rec[1][\"country\"]\n",
    "        if province != NO_DATA_CHAR:\n",
    "            attr = \"province\"\n",
    "            name = province\n",
    "        else:\n",
    "            attr = \"country\"\n",
    "            name = country\n",
    "        lat = rec[1][\"lat\"]\n",
    "        lon = rec[1][\"lon\"]\n",
    "    # Internal nodes use mugration province\n",
    "    else:\n",
    "        name = rec[1][\"mugration_province\"]\n",
    "        lat  = rec[1][\"mugration_province_lat\"]\n",
    "        lon  = rec[1][\"mugration_province_lon\"]     \n",
    "    \n",
    "    if attr == \"province\":\n",
    "        size = latlon_province_df[\"size\"][name]\n",
    "    elif attr == \"country\":\n",
    "        size = latlon_country_df[\"size\"][name]\n",
    "    point = shapely.geometry.Point(lon,lat)\n",
    "    geometry.append(point)\n",
    "\n",
    "\n",
    "metadata_gdf = geopandas.GeoDataFrame(metadata_df, crs=\"epsg:3857\", geometry=geometry)\n",
    "\n",
    "# Add geometry size\n",
    "for rec in metadata_gdf.iterrows():\n",
    "    sample = rec[0]\n",
    "    name = rec[1][\"mugration_\" + attr]\n",
    "    size = latlon_df[\"size\"][name]  \n",
    "    metadata_gdf.at[sample, \"geometry_size\"] = size\n",
    "    \n",
    "\n",
    "\n",
    "# Test map\n",
    "\n",
    "# Generate a figure with two axes, one for CartoPy, one for GeoPandas\n",
    "fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=(10, 5))\n",
    "world_polygons.plot(ax=ax, color=\"grey\", alpha=0.5)\n",
    "metadata_gdf.plot(ax=ax, c=\"red\", markersize=\"geometry_size\", ec=\"black\", lw=0.25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "opened-rabbit",
   "metadata": {},
   "source": [
    "# Colors\n",
    "\n",
    "Use the full tree for color order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "crazy-reach",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_colors = os.path.join(out_dir, \"colors.tsv\")\n",
    "\n",
    "colors_dict = {}\n",
    "\n",
    "for attr in MUG_ATTRIBUTE_LIST:\n",
    "    # Create the color map\n",
    "    attr_key = attr.lower()\n",
    "    colors_dict[attr_key] = {}\n",
    "    for t in divtree.get_terminals():\n",
    "        attr_val = metadata_df[attr][t.name]\n",
    "        # Remove the letter suffix from branch_minor\n",
    "        if attr == \"branch_minor\":\n",
    "            while attr_val[-1].isalpha():\n",
    "                attr_val = attr_val[:-1] \n",
    "        if attr_val not in colors_dict[attr_key] and attr_val != NO_DATA_CHAR:\n",
    "            colors_dict[attr_key][attr_val] = {}\n",
    "            \n",
    "    # Create the custom color map (pyplot)\n",
    "    cmap = plt.get_cmap(\"rainbow\", len(colors_dict[attr_key]))\n",
    "    # Convert the color map to a list of RGB values\n",
    "    cmaplist = [cmap(i) for i in range(cmap.N)]\n",
    "    # Convert RGB values to hex colors\n",
    "    attr_hex = [colors.to_hex(col) for col in cmaplist]\n",
    "    \n",
    "    # Assign colors to value\n",
    "    for attr_val, attr_col in zip(colors_dict[attr_key], attr_hex):\n",
    "        colors_dict[attr_key][attr_val] = attr_col   \n",
    "        \n",
    "    # Add unknown\n",
    "    colors_dict[attr_key][UNKNOWN_CHAR] = \"#969696\"\n",
    "\n",
    "with open(out_path_colors, \"w\") as outfile:\n",
    "    for attr_key in colors_dict:\n",
    "        for attr_val in colors_dict[attr_key]:\n",
    "            outfile.write(str(attr_key) + \"\\t\" + str(attr_val) + \"\\t\" + str(colors_dict[attr_key][attr_val]) + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "regulated-bunny",
   "metadata": {},
   "source": [
    "---\n",
    "# Parse Tree Into Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "elegant-second",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>branch_support_char</th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>timetree_date</th>\n",
       "      <th>timetree_CI_height</th>\n",
       "      <th>timetree_CI_date</th>\n",
       "      <th>mugration_country</th>\n",
       "      <th>mugration_province</th>\n",
       "      <th>mugration_branch_major</th>\n",
       "      <th>mugration_branch_minor</th>\n",
       "      <th>mugration_country_confidence</th>\n",
       "      <th>mugration_province_confidence</th>\n",
       "      <th>mugration_branch_major_confidence</th>\n",
       "      <th>mugration_branch_minor_confidence</th>\n",
       "      <th>mugration_country_lat</th>\n",
       "      <th>mugration_country_lon</th>\n",
       "      <th>mugration_province_lat</th>\n",
       "      <th>mugration_province_lon</th>\n",
       "      <th>geometry</th>\n",
       "      <th>geometry_size</th>\n",
       "      <th>date_lower</th>\n",
       "      <th>date_upper</th>\n",
       "      <th>divtree_coord_x</th>\n",
       "      <th>divtree_coord_y</th>\n",
       "      <th>rtt_dist</th>\n",
       "      <th>branch_major_color</th>\n",
       "      <th>timetree_coord_x</th>\n",
       "      <th>timetree_coord_y</th>\n",
       "      <th>timetree_date_calendar</th>\n",
       "      <th>timetree_CI_lower_err</th>\n",
       "      <th>timetree_CI_upper_err</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>6.334000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-29</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>POINT (-105.60772 38.72518)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1992</td>\n",
       "      <td>1992</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>28.00000</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>8352.42496</td>\n",
       "      <td>28.00000</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009670865.1_ASM967086v1_genomic</th>\n",
       "      <td>42012</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>-50</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722853</td>\n",
       "      <td>KEEP: Assembly Modern Placement</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1971</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>1.078700e-05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-50</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>POINT (85.46335 42.48050)</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1971</td>\n",
       "      <td>1971</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>7.00000</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>#4062fa</td>\n",
       "      <td>8331.42500</td>\n",
       "      <td>7.00000</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630435.1_ASM863043v1_genomic</th>\n",
       "      <td>C-719</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>-25</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED0</td>\n",
       "      <td>SAMN12721146</td>\n",
       "      <td>KEEP: Assembly Modern Placement</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1996</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.417700e-05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-25</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>43.736833</td>\n",
       "      <td>41.726799</td>\n",
       "      <td>POINT (41.72680 43.73683)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1996</td>\n",
       "      <td>1996</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>15.00000</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>#80ffb4</td>\n",
       "      <td>8356.42610</td>\n",
       "      <td>15.00000</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000169635.1_ASM16963v1_genomic</th>\n",
       "      <td>MG05-1020</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>NA</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>SAMN02404403</td>\n",
       "      <td>KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...</td>\n",
       "      <td>1</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.927000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.21</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-18.924960</td>\n",
       "      <td>46.441642</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (46.44164 -18.92496)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2005</td>\n",
       "      <td>2005</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>26.00000</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>8365.42496</td>\n",
       "      <td>26.00000</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000170275.1_ASM17027v1_genomic</th>\n",
       "      <td>F1991016</td>\n",
       "      <td>1991.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>SAMN02404399</td>\n",
       "      <td>KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...</td>\n",
       "      <td>1</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1991</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>5.396000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (102.00000 25.00000)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1991</td>\n",
       "      <td>1991</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>25.00000</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>8351.42496</td>\n",
       "      <td>25.00000</td>\n",
       "      <td>1991.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE22</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>7.030000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-697.117</td>\n",
       "      <td>{138.712,50502.4}</td>\n",
       "      <td>{-50502.4,-138.712}</td>\n",
       "      <td>China</td>\n",
       "      <td>Gansu</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>1.IN2</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.42</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (102.00000 38.00000)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>23.96875</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>#ff6232</td>\n",
       "      <td>7684.30810</td>\n",
       "      <td>23.96875</td>\n",
       "      <td>1324.0</td>\n",
       "      <td>49805.0</td>\n",
       "      <td>559.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE23</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>7.040000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-697.117</td>\n",
       "      <td>{138.712,50502.4}</td>\n",
       "      <td>{-50502.4,-138.712}</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>1.IN3</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.42</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (102.00000 25.00000)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>24.93750</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>#ff6232</td>\n",
       "      <td>7684.30810</td>\n",
       "      <td>24.93750</td>\n",
       "      <td>1324.0</td>\n",
       "      <td>49805.0</td>\n",
       "      <td>559.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE24</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>5.630000e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.41</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (102.00000 25.00000)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>25.87500</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>7691.72496</td>\n",
       "      <td>25.87500</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>44407.0</td>\n",
       "      <td>560.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE25</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.300000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.36</td>\n",
       "      <td>-18.924960</td>\n",
       "      <td>46.441642</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (102.00000 25.00000)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>26.75000</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>7691.72496</td>\n",
       "      <td>26.75000</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>44407.0</td>\n",
       "      <td>560.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE26</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.300000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>0.37</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.98</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>-8.000000</td>\n",
       "      <td>-78.500000</td>\n",
       "      <td>POINT (-78.50000 -8.00000)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>27.50000</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>7691.72496</td>\n",
       "      <td>27.50000</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>44407.0</td>\n",
       "      <td>560.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        strain    date date_bp  \\\n",
       "sample                                                           \n",
       "Reference                                 CO92    1992     -29   \n",
       "GCA_009670865.1_ASM967086v1_genomic      42012  1971.0     -50   \n",
       "GCA_008630435.1_ASM863043v1_genomic      C-719  1996.0     -25   \n",
       "GCA_000169635.1_ASM16963v1_genomic   MG05-1020  2005.0     -16   \n",
       "GCA_000170275.1_ASM17027v1_genomic    F1991016  1991.0     -30   \n",
       "...                                        ...     ...     ...   \n",
       "NODE22                                      NA      NA      NA   \n",
       "NODE23                                      NA      NA      NA   \n",
       "NODE24                                      NA      NA      NA   \n",
       "NODE25                                      NA      NA      NA   \n",
       "NODE26                                      NA      NA      NA   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009670865.1_ASM967086v1_genomic                     China   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    Russia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                 Madagascar   \n",
       "GCA_000170275.1_ASM17027v1_genomic                      China   \n",
       "...                                                       ...   \n",
       "NODE22                                                     NA   \n",
       "NODE23                                                     NA   \n",
       "NODE24                                                     NA   \n",
       "NODE25                                                     NA   \n",
       "NODE26                                                     NA   \n",
       "\n",
       "                                                province country_lat  \\\n",
       "sample                                                                 \n",
       "Reference                                       Colorado     39.7837   \n",
       "GCA_009670865.1_ASM967086v1_genomic             Xinjiang     35.0001   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Karachay-Cherkessia     64.6863   \n",
       "GCA_000169635.1_ASM16963v1_genomic                    NA     -18.925   \n",
       "GCA_000170275.1_ASM17027v1_genomic                Yunnan     35.0001   \n",
       "...                                                  ...         ...   \n",
       "NODE22                                                NA          NA   \n",
       "NODE23                                                NA          NA   \n",
       "NODE24                                                NA          NA   \n",
       "NODE25                                                NA          NA   \n",
       "NODE26                                                NA          NA   \n",
       "\n",
       "                                    country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "Reference                              -100.446      38.7252     -105.608   \n",
       "GCA_009670865.1_ASM967086v1_genomic         105      42.4805      85.4633   \n",
       "GCA_008630435.1_ASM863043v1_genomic     97.7453      43.7368      41.7268   \n",
       "GCA_000169635.1_ASM16963v1_genomic      46.4416           NA           NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic          105           25          102   \n",
       "...                                         ...          ...          ...   \n",
       "NODE22                                       NA           NA           NA   \n",
       "NODE23                                       NA           NA           NA   \n",
       "NODE24                                       NA           NA           NA   \n",
       "NODE25                                       NA           NA           NA   \n",
       "NODE26                                       NA           NA           NA   \n",
       "\n",
       "                                         biovar branch_major branch_minor  \\\n",
       "sample                                                                      \n",
       "Reference                            Orientalis        1.ORI       1.ORI1   \n",
       "GCA_009670865.1_ASM967086v1_genomic     Antiqua        0.ANT       0.ANT1   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Medievalis        2.MED       2.MED0   \n",
       "GCA_000169635.1_ASM16963v1_genomic   Orientalis        1.ORI       1.ORI3   \n",
       "GCA_000170275.1_ASM17027v1_genomic   Orientalis        1.ORI       1.ORI2   \n",
       "...                                         ...          ...          ...   \n",
       "NODE22                                       NA           NA           NA   \n",
       "NODE23                                       NA           NA           NA   \n",
       "NODE24                                       NA           NA           NA   \n",
       "NODE25                                       NA           NA           NA   \n",
       "NODE26                                       NA           NA           NA   \n",
       "\n",
       "                                    biosample_accession  \\\n",
       "sample                                                    \n",
       "Reference                                  SAMEA1705942   \n",
       "GCA_009670865.1_ASM967086v1_genomic        SAMN07722853   \n",
       "GCA_008630435.1_ASM863043v1_genomic        SAMN12721146   \n",
       "GCA_000169635.1_ASM16963v1_genomic         SAMN02404403   \n",
       "GCA_000170275.1_ASM17027v1_genomic         SAMN02404399   \n",
       "...                                                 ...   \n",
       "NODE22                                               NA   \n",
       "NODE23                                               NA   \n",
       "NODE24                                               NA   \n",
       "NODE25                                               NA   \n",
       "NODE26                                               NA   \n",
       "\n",
       "                                                                     biosample_comment  \\\n",
       "sample                                                                                   \n",
       "Reference                                              KEEP: Assembly Modern Reference   \n",
       "GCA_009670865.1_ASM967086v1_genomic                    KEEP: Assembly Modern Placement   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    KEEP: Assembly Modern Placement   \n",
       "GCA_000169635.1_ASM16963v1_genomic   KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...   \n",
       "GCA_000170275.1_ASM17027v1_genomic   KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...   \n",
       "...                                                                                ...   \n",
       "NODE22                                                                              NA   \n",
       "NODE23                                                                              NA   \n",
       "NODE24                                                                              NA   \n",
       "NODE25                                                                              NA   \n",
       "NODE26                                                                              NA   \n",
       "\n",
       "                                    branch_number      continent date_mean  \\\n",
       "sample                                                                       \n",
       "Reference                                       1  North America      1992   \n",
       "GCA_009670865.1_ASM967086v1_genomic             0           Asia      1971   \n",
       "GCA_008630435.1_ASM863043v1_genomic             2         Europe      1996   \n",
       "GCA_000169635.1_ASM16963v1_genomic              1         Africa      2005   \n",
       "GCA_000170275.1_ASM17027v1_genomic              1           Asia      1991   \n",
       "...                                           ...            ...       ...   \n",
       "NODE22                                         NA             NA        NA   \n",
       "NODE23                                         NA             NA        NA   \n",
       "NODE24                                         NA             NA        NA   \n",
       "NODE25                                         NA             NA        NA   \n",
       "NODE26                                         NA             NA        NA   \n",
       "\n",
       "                                    date_bp_mean date_err      lat      lon  \\\n",
       "sample                                                                        \n",
       "Reference                                     29        0  38.7252 -105.608   \n",
       "GCA_009670865.1_ASM967086v1_genomic           50        0  42.4805  85.4633   \n",
       "GCA_008630435.1_ASM863043v1_genomic           25        0  43.7368  41.7268   \n",
       "GCA_000169635.1_ASM16963v1_genomic            16        0  -18.925  46.4416   \n",
       "GCA_000170275.1_ASM17027v1_genomic            30        0       25      102   \n",
       "...                                          ...      ...      ...      ...   \n",
       "NODE22                                        NA       NA       NA       NA   \n",
       "NODE23                                        NA       NA       NA       NA   \n",
       "NODE24                                        NA       NA       NA       NA   \n",
       "NODE25                                        NA       NA       NA       NA   \n",
       "NODE26                                        NA       NA       NA       NA   \n",
       "\n",
       "                                    branch_support_char node_type  \\\n",
       "sample                                                              \n",
       "Reference                                                terminal   \n",
       "GCA_009670865.1_ASM967086v1_genomic                      terminal   \n",
       "GCA_008630435.1_ASM863043v1_genomic                      terminal   \n",
       "GCA_000169635.1_ASM16963v1_genomic                       terminal   \n",
       "GCA_000170275.1_ASM17027v1_genomic                       terminal   \n",
       "...                                                 ...       ...   \n",
       "NODE22                                               NA  internal   \n",
       "NODE23                                               NA  internal   \n",
       "NODE24                                               NA  internal   \n",
       "NODE25                                               NA  internal   \n",
       "NODE26                                               NA  internal   \n",
       "\n",
       "                                     branch_length  branch_support  \\\n",
       "sample                                                               \n",
       "Reference                             6.334000e-06           100.0   \n",
       "GCA_009670865.1_ASM967086v1_genomic   1.078700e-05           100.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic   2.417700e-05           100.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic    4.927000e-06           100.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic    5.396000e-06           100.0   \n",
       "...                                            ...             ...   \n",
       "NODE22                                7.030000e-07             0.0   \n",
       "NODE23                                7.040000e-07             0.0   \n",
       "NODE24                                5.630000e-06             0.0   \n",
       "NODE25                                2.300000e-08             0.0   \n",
       "NODE26                                2.300000e-08             0.0   \n",
       "\n",
       "                                    timetree_date timetree_CI_height  \\\n",
       "sample                                                                 \n",
       "Reference                                     -29                 NA   \n",
       "GCA_009670865.1_ASM967086v1_genomic           -50                 NA   \n",
       "GCA_008630435.1_ASM863043v1_genomic           -25                 NA   \n",
       "GCA_000169635.1_ASM16963v1_genomic            -16                 NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic            -30                 NA   \n",
       "...                                           ...                ...   \n",
       "NODE22                                   -697.117  {138.712,50502.4}   \n",
       "NODE23                                   -697.117  {138.712,50502.4}   \n",
       "NODE24                                     -689.7  {129.229,45096.4}   \n",
       "NODE25                                     -689.7  {129.229,45096.4}   \n",
       "NODE26                                     -689.7  {129.229,45096.4}   \n",
       "\n",
       "                                        timetree_CI_date  \\\n",
       "sample                                                     \n",
       "Reference                                             NA   \n",
       "GCA_009670865.1_ASM967086v1_genomic                   NA   \n",
       "GCA_008630435.1_ASM863043v1_genomic                   NA   \n",
       "GCA_000169635.1_ASM16963v1_genomic                    NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic                    NA   \n",
       "...                                                  ...   \n",
       "NODE22                               {-50502.4,-138.712}   \n",
       "NODE23                               {-50502.4,-138.712}   \n",
       "NODE24                               {-45096.4,-129.229}   \n",
       "NODE25                               {-45096.4,-129.229}   \n",
       "NODE26                               {-45096.4,-129.229}   \n",
       "\n",
       "                                            mugration_country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009670865.1_ASM967086v1_genomic                     China   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    Russia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                 Madagascar   \n",
       "GCA_000170275.1_ASM17027v1_genomic                      China   \n",
       "...                                                       ...   \n",
       "NODE22                                                  China   \n",
       "NODE23                                                  China   \n",
       "NODE24                                                  China   \n",
       "NODE25                                             Madagascar   \n",
       "NODE26                               United States of America   \n",
       "\n",
       "                                      mugration_province  \\\n",
       "sample                                                     \n",
       "Reference                                       Colorado   \n",
       "GCA_009670865.1_ASM967086v1_genomic             Xinjiang   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Karachay-Cherkessia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                Yunnan   \n",
       "GCA_000170275.1_ASM17027v1_genomic                Yunnan   \n",
       "...                                                  ...   \n",
       "NODE22                                             Gansu   \n",
       "NODE23                                            Yunnan   \n",
       "NODE24                                            Yunnan   \n",
       "NODE25                                            Yunnan   \n",
       "NODE26                                       La Libertad   \n",
       "\n",
       "                                    mugration_branch_major  \\\n",
       "sample                                                       \n",
       "Reference                                            1.ORI   \n",
       "GCA_009670865.1_ASM967086v1_genomic                  0.ANT   \n",
       "GCA_008630435.1_ASM863043v1_genomic                  2.MED   \n",
       "GCA_000169635.1_ASM16963v1_genomic                   1.ORI   \n",
       "GCA_000170275.1_ASM17027v1_genomic                   1.ORI   \n",
       "...                                                    ...   \n",
       "NODE22                                                1.IN   \n",
       "NODE23                                                1.IN   \n",
       "NODE24                                               1.ORI   \n",
       "NODE25                                               1.ORI   \n",
       "NODE26                                               1.ORI   \n",
       "\n",
       "                                    mugration_branch_minor  \\\n",
       "sample                                                       \n",
       "Reference                                           1.ORI1   \n",
       "GCA_009670865.1_ASM967086v1_genomic                 0.ANT1   \n",
       "GCA_008630435.1_ASM863043v1_genomic                 2.MED0   \n",
       "GCA_000169635.1_ASM16963v1_genomic                  1.ORI3   \n",
       "GCA_000170275.1_ASM17027v1_genomic                  1.ORI2   \n",
       "...                                                    ...   \n",
       "NODE22                                               1.IN2   \n",
       "NODE23                                               1.IN3   \n",
       "NODE24                                              1.ORI2   \n",
       "NODE25                                              1.ORI3   \n",
       "NODE26                                              1.ORI1   \n",
       "\n",
       "                                     mugration_country_confidence  \\\n",
       "sample                                                              \n",
       "Reference                                                    1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                          1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                          1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                           1.00   \n",
       "GCA_000170275.1_ASM17027v1_genomic                           1.00   \n",
       "...                                                           ...   \n",
       "NODE22                                                       0.96   \n",
       "NODE23                                                       0.96   \n",
       "NODE24                                                       0.87   \n",
       "NODE25                                                       0.39   \n",
       "NODE26                                                       0.37   \n",
       "\n",
       "                                     mugration_province_confidence  \\\n",
       "sample                                                               \n",
       "Reference                                                     1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                           1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                           1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                            0.21   \n",
       "GCA_000170275.1_ASM17027v1_genomic                            1.00   \n",
       "...                                                            ...   \n",
       "NODE22                                                        0.36   \n",
       "NODE23                                                        0.83   \n",
       "NODE24                                                        0.84   \n",
       "NODE25                                                        0.36   \n",
       "NODE26                                                        0.36   \n",
       "\n",
       "                                     mugration_branch_major_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                         1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                               1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                               1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                                1.00   \n",
       "GCA_000170275.1_ASM17027v1_genomic                                1.00   \n",
       "...                                                                ...   \n",
       "NODE22                                                            1.00   \n",
       "NODE23                                                            0.97   \n",
       "NODE24                                                            0.97   \n",
       "NODE25                                                            1.00   \n",
       "NODE26                                                            1.00   \n",
       "\n",
       "                                     mugration_branch_minor_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                         1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                               1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                               1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                                1.00   \n",
       "GCA_000170275.1_ASM17027v1_genomic                                1.00   \n",
       "...                                                                ...   \n",
       "NODE22                                                            0.42   \n",
       "NODE23                                                            0.42   \n",
       "NODE24                                                            0.41   \n",
       "NODE25                                                            0.36   \n",
       "NODE26                                                            0.98   \n",
       "\n",
       "                                     mugration_country_lat  \\\n",
       "sample                                                       \n",
       "Reference                                        39.783730   \n",
       "GCA_009670865.1_ASM967086v1_genomic              35.000074   \n",
       "GCA_008630435.1_ASM863043v1_genomic              64.686314   \n",
       "GCA_000169635.1_ASM16963v1_genomic              -18.924960   \n",
       "GCA_000170275.1_ASM17027v1_genomic               35.000074   \n",
       "...                                                    ...   \n",
       "NODE22                                           35.000074   \n",
       "NODE23                                           35.000074   \n",
       "NODE24                                           35.000074   \n",
       "NODE25                                          -18.924960   \n",
       "NODE26                                           39.783730   \n",
       "\n",
       "                                     mugration_country_lon  \\\n",
       "sample                                                       \n",
       "Reference                                      -100.445882   \n",
       "GCA_009670865.1_ASM967086v1_genomic             104.999927   \n",
       "GCA_008630435.1_ASM863043v1_genomic              97.745306   \n",
       "GCA_000169635.1_ASM16963v1_genomic               46.441642   \n",
       "GCA_000170275.1_ASM17027v1_genomic              104.999927   \n",
       "...                                                    ...   \n",
       "NODE22                                          104.999927   \n",
       "NODE23                                          104.999927   \n",
       "NODE24                                          104.999927   \n",
       "NODE25                                           46.441642   \n",
       "NODE26                                         -100.445882   \n",
       "\n",
       "                                     mugration_province_lat  \\\n",
       "sample                                                        \n",
       "Reference                                         38.725178   \n",
       "GCA_009670865.1_ASM967086v1_genomic               42.480495   \n",
       "GCA_008630435.1_ASM863043v1_genomic               43.736833   \n",
       "GCA_000169635.1_ASM16963v1_genomic                25.000000   \n",
       "GCA_000170275.1_ASM17027v1_genomic                25.000000   \n",
       "...                                                     ...   \n",
       "NODE22                                            38.000000   \n",
       "NODE23                                            25.000000   \n",
       "NODE24                                            25.000000   \n",
       "NODE25                                            25.000000   \n",
       "NODE26                                            -8.000000   \n",
       "\n",
       "                                     mugration_province_lon  \\\n",
       "sample                                                        \n",
       "Reference                                       -105.607716   \n",
       "GCA_009670865.1_ASM967086v1_genomic               85.463346   \n",
       "GCA_008630435.1_ASM863043v1_genomic               41.726799   \n",
       "GCA_000169635.1_ASM16963v1_genomic               102.000000   \n",
       "GCA_000170275.1_ASM17027v1_genomic               102.000000   \n",
       "...                                                     ...   \n",
       "NODE22                                           102.000000   \n",
       "NODE23                                           102.000000   \n",
       "NODE24                                           102.000000   \n",
       "NODE25                                           102.000000   \n",
       "NODE26                                           -78.500000   \n",
       "\n",
       "                                                        geometry  \\\n",
       "sample                                                             \n",
       "Reference                            POINT (-105.60772 38.72518)   \n",
       "GCA_009670865.1_ASM967086v1_genomic    POINT (85.46335 42.48050)   \n",
       "GCA_008630435.1_ASM863043v1_genomic    POINT (41.72680 43.73683)   \n",
       "GCA_000169635.1_ASM16963v1_genomic    POINT (46.44164 -18.92496)   \n",
       "GCA_000170275.1_ASM17027v1_genomic    POINT (102.00000 25.00000)   \n",
       "...                                                          ...   \n",
       "NODE22                                POINT (102.00000 38.00000)   \n",
       "NODE23                                POINT (102.00000 25.00000)   \n",
       "NODE24                                POINT (102.00000 25.00000)   \n",
       "NODE25                                POINT (102.00000 25.00000)   \n",
       "NODE26                                POINT (-78.50000 -8.00000)   \n",
       "\n",
       "                                     geometry_size date_lower date_upper  \\\n",
       "sample                                                                     \n",
       "Reference                                      1.0       1992       1992   \n",
       "GCA_009670865.1_ASM967086v1_genomic            4.0       1971       1971   \n",
       "GCA_008630435.1_ASM863043v1_genomic            1.0       1996       1996   \n",
       "GCA_000169635.1_ASM16963v1_genomic             2.0       2005       2005   \n",
       "GCA_000170275.1_ASM17027v1_genomic             2.0       1991       1991   \n",
       "...                                            ...        ...        ...   \n",
       "NODE22                                         1.0         NA         NA   \n",
       "NODE23                                         2.0         NA         NA   \n",
       "NODE24                                         2.0         NA         NA   \n",
       "NODE25                                         2.0         NA         NA   \n",
       "NODE26                                         1.0         NA         NA   \n",
       "\n",
       "                                     divtree_coord_x  divtree_coord_y  \\\n",
       "sample                                                                  \n",
       "Reference                                   0.000072         28.00000   \n",
       "GCA_009670865.1_ASM967086v1_genomic         0.000053          7.00000   \n",
       "GCA_008630435.1_ASM863043v1_genomic         0.000086         15.00000   \n",
       "GCA_000169635.1_ASM16963v1_genomic          0.000071         26.00000   \n",
       "GCA_000170275.1_ASM17027v1_genomic          0.000071         25.00000   \n",
       "...                                              ...              ...   \n",
       "NODE22                                      0.000059         23.96875   \n",
       "NODE23                                      0.000060         24.93750   \n",
       "NODE24                                      0.000066         25.87500   \n",
       "NODE25                                      0.000066         26.75000   \n",
       "NODE26                                      0.000066         27.50000   \n",
       "\n",
       "                                     rtt_dist branch_major_color  \\\n",
       "sample                                                             \n",
       "Reference                            0.000072            #ff0000   \n",
       "GCA_009670865.1_ASM967086v1_genomic  0.000053            #4062fa   \n",
       "GCA_008630435.1_ASM863043v1_genomic  0.000086            #80ffb4   \n",
       "GCA_000169635.1_ASM16963v1_genomic   0.000071            #ff0000   \n",
       "GCA_000170275.1_ASM17027v1_genomic   0.000071            #ff0000   \n",
       "...                                       ...                ...   \n",
       "NODE22                               0.000059            #ff6232   \n",
       "NODE23                               0.000060            #ff6232   \n",
       "NODE24                               0.000066            #ff0000   \n",
       "NODE25                               0.000066            #ff0000   \n",
       "NODE26                               0.000066            #ff0000   \n",
       "\n",
       "                                     timetree_coord_x  timetree_coord_y  \\\n",
       "sample                                                                    \n",
       "Reference                                  8352.42496          28.00000   \n",
       "GCA_009670865.1_ASM967086v1_genomic        8331.42500           7.00000   \n",
       "GCA_008630435.1_ASM863043v1_genomic        8356.42610          15.00000   \n",
       "GCA_000169635.1_ASM16963v1_genomic         8365.42496          26.00000   \n",
       "GCA_000170275.1_ASM17027v1_genomic         8351.42496          25.00000   \n",
       "...                                               ...               ...   \n",
       "NODE22                                     7684.30810          23.96875   \n",
       "NODE23                                     7684.30810          24.93750   \n",
       "NODE24                                     7691.72496          25.87500   \n",
       "NODE25                                     7691.72496          26.75000   \n",
       "NODE26                                     7691.72496          27.50000   \n",
       "\n",
       "                                     timetree_date_calendar  \\\n",
       "sample                                                        \n",
       "Reference                                            1992.0   \n",
       "GCA_009670865.1_ASM967086v1_genomic                  1971.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic                  1996.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic                   2005.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic                   1991.0   \n",
       "...                                                     ...   \n",
       "NODE22                                               1324.0   \n",
       "NODE23                                               1324.0   \n",
       "NODE24                                               1332.0   \n",
       "NODE25                                               1332.0   \n",
       "NODE26                                               1332.0   \n",
       "\n",
       "                                     timetree_CI_lower_err  \\\n",
       "sample                                                       \n",
       "Reference                                              0.0   \n",
       "GCA_009670865.1_ASM967086v1_genomic                    0.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    0.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic                     0.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic                     0.0   \n",
       "...                                                    ...   \n",
       "NODE22                                             49805.0   \n",
       "NODE23                                             49805.0   \n",
       "NODE24                                             44407.0   \n",
       "NODE25                                             44407.0   \n",
       "NODE26                                             44407.0   \n",
       "\n",
       "                                     timetree_CI_upper_err  \n",
       "sample                                                      \n",
       "Reference                                              0.0  \n",
       "GCA_009670865.1_ASM967086v1_genomic                    0.0  \n",
       "GCA_008630435.1_ASM863043v1_genomic                    0.0  \n",
       "GCA_000169635.1_ASM16963v1_genomic                     0.0  \n",
       "GCA_000170275.1_ASM17027v1_genomic                     0.0  \n",
       "...                                                    ...  \n",
       "NODE22                                               559.0  \n",
       "NODE23                                               559.0  \n",
       "NODE24                                               560.0  \n",
       "NODE25                                               560.0  \n",
       "NODE26                                               560.0  \n",
       "\n",
       "[55 rows x 53 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "x_posns = get_x_positions(divtree)\n",
    "y_posns = get_y_positions(divtree)\n",
    "\n",
    "# Initialize \n",
    "# Date lower and upper error bars for tips\n",
    "metadata_gdf[\"date_lower\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"date_upper\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "    \n",
    "for c in divtree.find_clades():\n",
    "    # ------------------------------------------------------------\n",
    "    # Coordinates for the divergence tree\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    metadata_gdf.at[c.name, 'divtree_coord_x'] = coord_x\n",
    "    metadata_gdf.at[c.name, 'divtree_coord_y'] = coord_y\n",
    "    \n",
    "    # ------------------------------------------------------------\n",
    "    # Root to tip distance (same as div x)\n",
    "    metadata_gdf.at[c.name, \"rtt_dist\"] = coord_x\n",
    "    \n",
    "    # ------------------------------------------------------------\n",
    "    # Branch Major and Branch Minor and Colors\n",
    "    branch_minor = metadata_gdf[\"mugration_branch_minor\"][c.name]\n",
    "    if branch_minor != NO_DATA_CHAR:\n",
    "        # Remove letters at end\n",
    "        while branch_minor[-1].isalpha():\n",
    "            branch_minor = branch_minor[:-1]    \n",
    "    metadata_gdf.at[c.name, \"mugration_branch_minor\"] = branch_minor\n",
    "    branch_major = metadata_gdf[\"mugration_branch_major\"][c.name]\n",
    "    branch_major_color = colors_dict[\"branch_major\"][branch_major]\n",
    "    metadata_gdf.at[c.name, \"branch_major_color\"] = branch_major_color    \n",
    "\n",
    "    # ------------------------------------------------------------\n",
    "    # Date lower and upper error bars for tips\n",
    "    date_mean = metadata_gdf[\"date_mean\"][c.name] \n",
    "    if date_mean != NO_DATA_CHAR:\n",
    "        metadata_gdf.at[c.name, \"date_lower\"] = date_mean - metadata_gdf[\"date_err\"][c.name]\n",
    "        metadata_gdf.at[c.name, \"date_upper\"] = date_mean + metadata_gdf[\"date_err\"][c.name]\n",
    "\n",
    "x_posns = get_x_positions(timetree)\n",
    "y_posns = get_y_positions(timetree)\n",
    "\n",
    "\n",
    "for c in timetree.find_clades():\n",
    "\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    metadata_gdf.at[c.name, 'timetree_coord_x'] = coord_x\n",
    "    metadata_gdf.at[c.name, 'timetree_coord_y'] = coord_y\n",
    "    \n",
    "    timetree_date = int(float(metadata_gdf[\"timetree_date\"][c.name]))\n",
    "    timetree_date_calendar = CURRENT_YEAR + timetree_date\n",
    "    metadata_gdf.at[c.name, 'timetree_date_calendar'] = timetree_date_calendar\n",
    "    \n",
    "    timetree_CI_date = metadata_gdf[\"timetree_CI_date\"][c.name]\n",
    "    if timetree_CI_date == NO_DATA_CHAR:\n",
    "        timetree_CI_lower_err = 0\n",
    "        timetree_CI_upper_err = 0\n",
    "    else:\n",
    "        timetree_CI_split = [int(float(d)) for d in timetree_CI_date.strip(\"{}\").split(\",\")]\n",
    "        timetree_CI_calendar = [CURRENT_YEAR + d for d in timetree_CI_split]\n",
    "        timetree_CI_lower_err = timetree_date_calendar - timetree_CI_calendar[0]\n",
    "        timetree_CI_upper_err = timetree_CI_calendar[1] - timetree_date_calendar\n",
    "    metadata_gdf.at[c.name,\"timetree_CI_lower_err\"] = timetree_CI_lower_err\n",
    "    metadata_gdf.at[c.name,\"timetree_CI_upper_err\"] = timetree_CI_upper_err   \n",
    "\n",
    "display(metadata_gdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "korean-magazine",
   "metadata": {},
   "source": [
    "## Calculate Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "original-afternoon",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "      <th>date_mean</th>\n",
       "      <th>date_bp_mean</th>\n",
       "      <th>date_err</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>branch_support_char</th>\n",
       "      <th>node_type</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>branch_support</th>\n",
       "      <th>timetree_date</th>\n",
       "      <th>timetree_CI_height</th>\n",
       "      <th>timetree_CI_date</th>\n",
       "      <th>mugration_country</th>\n",
       "      <th>mugration_province</th>\n",
       "      <th>mugration_branch_major</th>\n",
       "      <th>mugration_branch_minor</th>\n",
       "      <th>mugration_country_confidence</th>\n",
       "      <th>mugration_province_confidence</th>\n",
       "      <th>mugration_branch_major_confidence</th>\n",
       "      <th>mugration_branch_minor_confidence</th>\n",
       "      <th>mugration_country_lat</th>\n",
       "      <th>mugration_country_lon</th>\n",
       "      <th>mugration_province_lat</th>\n",
       "      <th>mugration_province_lon</th>\n",
       "      <th>geometry</th>\n",
       "      <th>geometry_size</th>\n",
       "      <th>date_lower</th>\n",
       "      <th>date_upper</th>\n",
       "      <th>divtree_coord_x</th>\n",
       "      <th>divtree_coord_y</th>\n",
       "      <th>rtt_dist</th>\n",
       "      <th>branch_major_color</th>\n",
       "      <th>timetree_coord_x</th>\n",
       "      <th>timetree_coord_y</th>\n",
       "      <th>timetree_date_calendar</th>\n",
       "      <th>timetree_CI_lower_err</th>\n",
       "      <th>timetree_CI_upper_err</th>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <th>clade_rtt_dist</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>CO92</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>SAMEA1705942</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>North America</td>\n",
       "      <td>1992</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>6.334000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-29</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "      <td>POINT (-105.60772 38.72518)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1992</td>\n",
       "      <td>1992</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>28.0000</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>8352.42496</td>\n",
       "      <td>28.0000</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009670865.1_ASM967086v1_genomic</th>\n",
       "      <td>42012</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>-50</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722853</td>\n",
       "      <td>KEEP: Assembly Modern Placement</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1971</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>1.078700e-05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-50</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "      <td>POINT (85.46335 42.48050)</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1971</td>\n",
       "      <td>1971</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>#4062fa</td>\n",
       "      <td>8331.42500</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_008630435.1_ASM863043v1_genomic</th>\n",
       "      <td>C-719</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>-25</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED0</td>\n",
       "      <td>SAMN12721146</td>\n",
       "      <td>KEEP: Assembly Modern Placement</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "      <td>1996</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>43.7368</td>\n",
       "      <td>41.7268</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.417700e-05</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-25</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Karachay-Cherkessia</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "      <td>43.736833</td>\n",
       "      <td>41.726799</td>\n",
       "      <td>POINT (41.72680 43.73683)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1996</td>\n",
       "      <td>1996</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>15.0000</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>#80ffb4</td>\n",
       "      <td>8356.42610</td>\n",
       "      <td>15.0000</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000169635.1_ASM16963v1_genomic</th>\n",
       "      <td>MG05-1020</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>NA</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>SAMN02404403</td>\n",
       "      <td>KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...</td>\n",
       "      <td>1</td>\n",
       "      <td>Africa</td>\n",
       "      <td>2005</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>-18.925</td>\n",
       "      <td>46.4416</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>4.927000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-16</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.21</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-18.924960</td>\n",
       "      <td>46.441642</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (46.44164 -18.92496)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2005</td>\n",
       "      <td>2005</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>26.0000</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>8365.42496</td>\n",
       "      <td>26.0000</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000170275.1_ASM17027v1_genomic</th>\n",
       "      <td>F1991016</td>\n",
       "      <td>1991.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>SAMN02404399</td>\n",
       "      <td>KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...</td>\n",
       "      <td>1</td>\n",
       "      <td>Asia</td>\n",
       "      <td>1991</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>102</td>\n",
       "      <td></td>\n",
       "      <td>terminal</td>\n",
       "      <td>5.396000e-06</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (102.00000 25.00000)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1991</td>\n",
       "      <td>1991</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>25.0000</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>8351.42496</td>\n",
       "      <td>25.0000</td>\n",
       "      <td>1991.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE23</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>7.040000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-697.117</td>\n",
       "      <td>{138.712,50502.4}</td>\n",
       "      <td>{-50502.4,-138.712}</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.IN</td>\n",
       "      <td>1.IN3</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.42</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (102.00000 25.00000)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>24.9375</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>#ff6232</td>\n",
       "      <td>7684.30810</td>\n",
       "      <td>24.9375</td>\n",
       "      <td>1324.0</td>\n",
       "      <td>49805.0</td>\n",
       "      <td>559.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE24</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>5.630000e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>China</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI2</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.41</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (102.00000 25.00000)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>25.8750</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>7691.72496</td>\n",
       "      <td>25.8750</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>44407.0</td>\n",
       "      <td>560.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE25</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.300000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Yunnan</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI3</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.36</td>\n",
       "      <td>-18.924960</td>\n",
       "      <td>46.441642</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>POINT (102.00000 25.00000)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>26.7500</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>7691.72496</td>\n",
       "      <td>26.7500</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>44407.0</td>\n",
       "      <td>560.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE26</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.300000e-08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-689.7</td>\n",
       "      <td>{129.229,45096.4}</td>\n",
       "      <td>{-45096.4,-129.229}</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>0.37</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.98</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "      <td>-8.000000</td>\n",
       "      <td>-78.500000</td>\n",
       "      <td>POINT (-78.50000 -8.00000)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>27.5000</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>#ff0000</td>\n",
       "      <td>7691.72496</td>\n",
       "      <td>27.5000</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>44407.0</td>\n",
       "      <td>560.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>root_rtt_dist</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        strain    date date_bp  \\\n",
       "sample                                                           \n",
       "Reference                                 CO92    1992     -29   \n",
       "GCA_009670865.1_ASM967086v1_genomic      42012  1971.0     -50   \n",
       "GCA_008630435.1_ASM863043v1_genomic      C-719  1996.0     -25   \n",
       "GCA_000169635.1_ASM16963v1_genomic   MG05-1020  2005.0     -16   \n",
       "GCA_000170275.1_ASM17027v1_genomic    F1991016  1991.0     -30   \n",
       "...                                        ...     ...     ...   \n",
       "NODE23                                      NA      NA      NA   \n",
       "NODE24                                      NA      NA      NA   \n",
       "NODE25                                      NA      NA      NA   \n",
       "NODE26                                      NA      NA      NA   \n",
       "root_rtt_dist                              NaN     NaN     NaN   \n",
       "\n",
       "                                                      country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009670865.1_ASM967086v1_genomic                     China   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    Russia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                 Madagascar   \n",
       "GCA_000170275.1_ASM17027v1_genomic                      China   \n",
       "...                                                       ...   \n",
       "NODE23                                                     NA   \n",
       "NODE24                                                     NA   \n",
       "NODE25                                                     NA   \n",
       "NODE26                                                     NA   \n",
       "root_rtt_dist                                             NaN   \n",
       "\n",
       "                                                province country_lat  \\\n",
       "sample                                                                 \n",
       "Reference                                       Colorado     39.7837   \n",
       "GCA_009670865.1_ASM967086v1_genomic             Xinjiang     35.0001   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Karachay-Cherkessia     64.6863   \n",
       "GCA_000169635.1_ASM16963v1_genomic                    NA     -18.925   \n",
       "GCA_000170275.1_ASM17027v1_genomic                Yunnan     35.0001   \n",
       "...                                                  ...         ...   \n",
       "NODE23                                                NA          NA   \n",
       "NODE24                                                NA          NA   \n",
       "NODE25                                                NA          NA   \n",
       "NODE26                                                NA          NA   \n",
       "root_rtt_dist                                        NaN         NaN   \n",
       "\n",
       "                                    country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "Reference                              -100.446      38.7252     -105.608   \n",
       "GCA_009670865.1_ASM967086v1_genomic         105      42.4805      85.4633   \n",
       "GCA_008630435.1_ASM863043v1_genomic     97.7453      43.7368      41.7268   \n",
       "GCA_000169635.1_ASM16963v1_genomic      46.4416           NA           NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic          105           25          102   \n",
       "...                                         ...          ...          ...   \n",
       "NODE23                                       NA           NA           NA   \n",
       "NODE24                                       NA           NA           NA   \n",
       "NODE25                                       NA           NA           NA   \n",
       "NODE26                                       NA           NA           NA   \n",
       "root_rtt_dist                               NaN          NaN          NaN   \n",
       "\n",
       "                                         biovar branch_major branch_minor  \\\n",
       "sample                                                                      \n",
       "Reference                            Orientalis        1.ORI       1.ORI1   \n",
       "GCA_009670865.1_ASM967086v1_genomic     Antiqua        0.ANT       0.ANT1   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Medievalis        2.MED       2.MED0   \n",
       "GCA_000169635.1_ASM16963v1_genomic   Orientalis        1.ORI       1.ORI3   \n",
       "GCA_000170275.1_ASM17027v1_genomic   Orientalis        1.ORI       1.ORI2   \n",
       "...                                         ...          ...          ...   \n",
       "NODE23                                       NA           NA           NA   \n",
       "NODE24                                       NA           NA           NA   \n",
       "NODE25                                       NA           NA           NA   \n",
       "NODE26                                       NA           NA           NA   \n",
       "root_rtt_dist                               NaN          NaN          NaN   \n",
       "\n",
       "                                    biosample_accession  \\\n",
       "sample                                                    \n",
       "Reference                                  SAMEA1705942   \n",
       "GCA_009670865.1_ASM967086v1_genomic        SAMN07722853   \n",
       "GCA_008630435.1_ASM863043v1_genomic        SAMN12721146   \n",
       "GCA_000169635.1_ASM16963v1_genomic         SAMN02404403   \n",
       "GCA_000170275.1_ASM17027v1_genomic         SAMN02404399   \n",
       "...                                                 ...   \n",
       "NODE23                                               NA   \n",
       "NODE24                                               NA   \n",
       "NODE25                                               NA   \n",
       "NODE26                                               NA   \n",
       "root_rtt_dist                                       NaN   \n",
       "\n",
       "                                                                     biosample_comment  \\\n",
       "sample                                                                                   \n",
       "Reference                                              KEEP: Assembly Modern Reference   \n",
       "GCA_009670865.1_ASM967086v1_genomic                    KEEP: Assembly Modern Placement   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    KEEP: Assembly Modern Placement   \n",
       "GCA_000169635.1_ASM16963v1_genomic   KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...   \n",
       "GCA_000170275.1_ASM17027v1_genomic   KEEP: Assembly Modern Morelli 2010 Cui 2013 Ke...   \n",
       "...                                                                                ...   \n",
       "NODE23                                                                              NA   \n",
       "NODE24                                                                              NA   \n",
       "NODE25                                                                              NA   \n",
       "NODE26                                                                              NA   \n",
       "root_rtt_dist                                                                      NaN   \n",
       "\n",
       "                                    branch_number      continent date_mean  \\\n",
       "sample                                                                       \n",
       "Reference                                       1  North America      1992   \n",
       "GCA_009670865.1_ASM967086v1_genomic             0           Asia      1971   \n",
       "GCA_008630435.1_ASM863043v1_genomic             2         Europe      1996   \n",
       "GCA_000169635.1_ASM16963v1_genomic              1         Africa      2005   \n",
       "GCA_000170275.1_ASM17027v1_genomic              1           Asia      1991   \n",
       "...                                           ...            ...       ...   \n",
       "NODE23                                         NA             NA        NA   \n",
       "NODE24                                         NA             NA        NA   \n",
       "NODE25                                         NA             NA        NA   \n",
       "NODE26                                         NA             NA        NA   \n",
       "root_rtt_dist                                 NaN            NaN       NaN   \n",
       "\n",
       "                                    date_bp_mean date_err      lat      lon  \\\n",
       "sample                                                                        \n",
       "Reference                                     29        0  38.7252 -105.608   \n",
       "GCA_009670865.1_ASM967086v1_genomic           50        0  42.4805  85.4633   \n",
       "GCA_008630435.1_ASM863043v1_genomic           25        0  43.7368  41.7268   \n",
       "GCA_000169635.1_ASM16963v1_genomic            16        0  -18.925  46.4416   \n",
       "GCA_000170275.1_ASM17027v1_genomic            30        0       25      102   \n",
       "...                                          ...      ...      ...      ...   \n",
       "NODE23                                        NA       NA       NA       NA   \n",
       "NODE24                                        NA       NA       NA       NA   \n",
       "NODE25                                        NA       NA       NA       NA   \n",
       "NODE26                                        NA       NA       NA       NA   \n",
       "root_rtt_dist                                NaN      NaN      NaN      NaN   \n",
       "\n",
       "                                    branch_support_char node_type  \\\n",
       "sample                                                              \n",
       "Reference                                                terminal   \n",
       "GCA_009670865.1_ASM967086v1_genomic                      terminal   \n",
       "GCA_008630435.1_ASM863043v1_genomic                      terminal   \n",
       "GCA_000169635.1_ASM16963v1_genomic                       terminal   \n",
       "GCA_000170275.1_ASM17027v1_genomic                       terminal   \n",
       "...                                                 ...       ...   \n",
       "NODE23                                               NA  internal   \n",
       "NODE24                                               NA  internal   \n",
       "NODE25                                               NA  internal   \n",
       "NODE26                                               NA  internal   \n",
       "root_rtt_dist                                       NaN       NaN   \n",
       "\n",
       "                                     branch_length  branch_support  \\\n",
       "sample                                                               \n",
       "Reference                             6.334000e-06           100.0   \n",
       "GCA_009670865.1_ASM967086v1_genomic   1.078700e-05           100.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic   2.417700e-05           100.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic    4.927000e-06           100.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic    5.396000e-06           100.0   \n",
       "...                                            ...             ...   \n",
       "NODE23                                7.040000e-07             0.0   \n",
       "NODE24                                5.630000e-06             0.0   \n",
       "NODE25                                2.300000e-08             0.0   \n",
       "NODE26                                2.300000e-08             0.0   \n",
       "root_rtt_dist                                  NaN             NaN   \n",
       "\n",
       "                                    timetree_date timetree_CI_height  \\\n",
       "sample                                                                 \n",
       "Reference                                     -29                 NA   \n",
       "GCA_009670865.1_ASM967086v1_genomic           -50                 NA   \n",
       "GCA_008630435.1_ASM863043v1_genomic           -25                 NA   \n",
       "GCA_000169635.1_ASM16963v1_genomic            -16                 NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic            -30                 NA   \n",
       "...                                           ...                ...   \n",
       "NODE23                                   -697.117  {138.712,50502.4}   \n",
       "NODE24                                     -689.7  {129.229,45096.4}   \n",
       "NODE25                                     -689.7  {129.229,45096.4}   \n",
       "NODE26                                     -689.7  {129.229,45096.4}   \n",
       "root_rtt_dist                                 NaN                NaN   \n",
       "\n",
       "                                        timetree_CI_date  \\\n",
       "sample                                                     \n",
       "Reference                                             NA   \n",
       "GCA_009670865.1_ASM967086v1_genomic                   NA   \n",
       "GCA_008630435.1_ASM863043v1_genomic                   NA   \n",
       "GCA_000169635.1_ASM16963v1_genomic                    NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic                    NA   \n",
       "...                                                  ...   \n",
       "NODE23                               {-50502.4,-138.712}   \n",
       "NODE24                               {-45096.4,-129.229}   \n",
       "NODE25                               {-45096.4,-129.229}   \n",
       "NODE26                               {-45096.4,-129.229}   \n",
       "root_rtt_dist                                        NaN   \n",
       "\n",
       "                                            mugration_country  \\\n",
       "sample                                                          \n",
       "Reference                            United States of America   \n",
       "GCA_009670865.1_ASM967086v1_genomic                     China   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    Russia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                 Madagascar   \n",
       "GCA_000170275.1_ASM17027v1_genomic                      China   \n",
       "...                                                       ...   \n",
       "NODE23                                                  China   \n",
       "NODE24                                                  China   \n",
       "NODE25                                             Madagascar   \n",
       "NODE26                               United States of America   \n",
       "root_rtt_dist                                             NaN   \n",
       "\n",
       "                                      mugration_province  \\\n",
       "sample                                                     \n",
       "Reference                                       Colorado   \n",
       "GCA_009670865.1_ASM967086v1_genomic             Xinjiang   \n",
       "GCA_008630435.1_ASM863043v1_genomic  Karachay-Cherkessia   \n",
       "GCA_000169635.1_ASM16963v1_genomic                Yunnan   \n",
       "GCA_000170275.1_ASM17027v1_genomic                Yunnan   \n",
       "...                                                  ...   \n",
       "NODE23                                            Yunnan   \n",
       "NODE24                                            Yunnan   \n",
       "NODE25                                            Yunnan   \n",
       "NODE26                                       La Libertad   \n",
       "root_rtt_dist                                        NaN   \n",
       "\n",
       "                                    mugration_branch_major  \\\n",
       "sample                                                       \n",
       "Reference                                            1.ORI   \n",
       "GCA_009670865.1_ASM967086v1_genomic                  0.ANT   \n",
       "GCA_008630435.1_ASM863043v1_genomic                  2.MED   \n",
       "GCA_000169635.1_ASM16963v1_genomic                   1.ORI   \n",
       "GCA_000170275.1_ASM17027v1_genomic                   1.ORI   \n",
       "...                                                    ...   \n",
       "NODE23                                                1.IN   \n",
       "NODE24                                               1.ORI   \n",
       "NODE25                                               1.ORI   \n",
       "NODE26                                               1.ORI   \n",
       "root_rtt_dist                                          NaN   \n",
       "\n",
       "                                    mugration_branch_minor  \\\n",
       "sample                                                       \n",
       "Reference                                           1.ORI1   \n",
       "GCA_009670865.1_ASM967086v1_genomic                 0.ANT1   \n",
       "GCA_008630435.1_ASM863043v1_genomic                 2.MED0   \n",
       "GCA_000169635.1_ASM16963v1_genomic                  1.ORI3   \n",
       "GCA_000170275.1_ASM17027v1_genomic                  1.ORI2   \n",
       "...                                                    ...   \n",
       "NODE23                                               1.IN3   \n",
       "NODE24                                              1.ORI2   \n",
       "NODE25                                              1.ORI3   \n",
       "NODE26                                              1.ORI1   \n",
       "root_rtt_dist                                          NaN   \n",
       "\n",
       "                                     mugration_country_confidence  \\\n",
       "sample                                                              \n",
       "Reference                                                    1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                          1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                          1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                           1.00   \n",
       "GCA_000170275.1_ASM17027v1_genomic                           1.00   \n",
       "...                                                           ...   \n",
       "NODE23                                                       0.96   \n",
       "NODE24                                                       0.87   \n",
       "NODE25                                                       0.39   \n",
       "NODE26                                                       0.37   \n",
       "root_rtt_dist                                                 NaN   \n",
       "\n",
       "                                     mugration_province_confidence  \\\n",
       "sample                                                               \n",
       "Reference                                                     1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                           1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                           1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                            0.21   \n",
       "GCA_000170275.1_ASM17027v1_genomic                            1.00   \n",
       "...                                                            ...   \n",
       "NODE23                                                        0.83   \n",
       "NODE24                                                        0.84   \n",
       "NODE25                                                        0.36   \n",
       "NODE26                                                        0.36   \n",
       "root_rtt_dist                                                  NaN   \n",
       "\n",
       "                                     mugration_branch_major_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                         1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                               1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                               1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                                1.00   \n",
       "GCA_000170275.1_ASM17027v1_genomic                                1.00   \n",
       "...                                                                ...   \n",
       "NODE23                                                            0.97   \n",
       "NODE24                                                            0.97   \n",
       "NODE25                                                            1.00   \n",
       "NODE26                                                            1.00   \n",
       "root_rtt_dist                                                      NaN   \n",
       "\n",
       "                                     mugration_branch_minor_confidence  \\\n",
       "sample                                                                   \n",
       "Reference                                                         1.00   \n",
       "GCA_009670865.1_ASM967086v1_genomic                               1.00   \n",
       "GCA_008630435.1_ASM863043v1_genomic                               1.00   \n",
       "GCA_000169635.1_ASM16963v1_genomic                                1.00   \n",
       "GCA_000170275.1_ASM17027v1_genomic                                1.00   \n",
       "...                                                                ...   \n",
       "NODE23                                                            0.42   \n",
       "NODE24                                                            0.41   \n",
       "NODE25                                                            0.36   \n",
       "NODE26                                                            0.98   \n",
       "root_rtt_dist                                                      NaN   \n",
       "\n",
       "                                     mugration_country_lat  \\\n",
       "sample                                                       \n",
       "Reference                                        39.783730   \n",
       "GCA_009670865.1_ASM967086v1_genomic              35.000074   \n",
       "GCA_008630435.1_ASM863043v1_genomic              64.686314   \n",
       "GCA_000169635.1_ASM16963v1_genomic              -18.924960   \n",
       "GCA_000170275.1_ASM17027v1_genomic               35.000074   \n",
       "...                                                    ...   \n",
       "NODE23                                           35.000074   \n",
       "NODE24                                           35.000074   \n",
       "NODE25                                          -18.924960   \n",
       "NODE26                                           39.783730   \n",
       "root_rtt_dist                                          NaN   \n",
       "\n",
       "                                     mugration_country_lon  \\\n",
       "sample                                                       \n",
       "Reference                                      -100.445882   \n",
       "GCA_009670865.1_ASM967086v1_genomic             104.999927   \n",
       "GCA_008630435.1_ASM863043v1_genomic              97.745306   \n",
       "GCA_000169635.1_ASM16963v1_genomic               46.441642   \n",
       "GCA_000170275.1_ASM17027v1_genomic              104.999927   \n",
       "...                                                    ...   \n",
       "NODE23                                          104.999927   \n",
       "NODE24                                          104.999927   \n",
       "NODE25                                           46.441642   \n",
       "NODE26                                         -100.445882   \n",
       "root_rtt_dist                                          NaN   \n",
       "\n",
       "                                     mugration_province_lat  \\\n",
       "sample                                                        \n",
       "Reference                                         38.725178   \n",
       "GCA_009670865.1_ASM967086v1_genomic               42.480495   \n",
       "GCA_008630435.1_ASM863043v1_genomic               43.736833   \n",
       "GCA_000169635.1_ASM16963v1_genomic                25.000000   \n",
       "GCA_000170275.1_ASM17027v1_genomic                25.000000   \n",
       "...                                                     ...   \n",
       "NODE23                                            25.000000   \n",
       "NODE24                                            25.000000   \n",
       "NODE25                                            25.000000   \n",
       "NODE26                                            -8.000000   \n",
       "root_rtt_dist                                           NaN   \n",
       "\n",
       "                                     mugration_province_lon  \\\n",
       "sample                                                        \n",
       "Reference                                       -105.607716   \n",
       "GCA_009670865.1_ASM967086v1_genomic               85.463346   \n",
       "GCA_008630435.1_ASM863043v1_genomic               41.726799   \n",
       "GCA_000169635.1_ASM16963v1_genomic               102.000000   \n",
       "GCA_000170275.1_ASM17027v1_genomic               102.000000   \n",
       "...                                                     ...   \n",
       "NODE23                                           102.000000   \n",
       "NODE24                                           102.000000   \n",
       "NODE25                                           102.000000   \n",
       "NODE26                                           -78.500000   \n",
       "root_rtt_dist                                           NaN   \n",
       "\n",
       "                                                        geometry  \\\n",
       "sample                                                             \n",
       "Reference                            POINT (-105.60772 38.72518)   \n",
       "GCA_009670865.1_ASM967086v1_genomic    POINT (85.46335 42.48050)   \n",
       "GCA_008630435.1_ASM863043v1_genomic    POINT (41.72680 43.73683)   \n",
       "GCA_000169635.1_ASM16963v1_genomic    POINT (46.44164 -18.92496)   \n",
       "GCA_000170275.1_ASM17027v1_genomic    POINT (102.00000 25.00000)   \n",
       "...                                                          ...   \n",
       "NODE23                                POINT (102.00000 25.00000)   \n",
       "NODE24                                POINT (102.00000 25.00000)   \n",
       "NODE25                                POINT (102.00000 25.00000)   \n",
       "NODE26                                POINT (-78.50000 -8.00000)   \n",
       "root_rtt_dist                                               None   \n",
       "\n",
       "                                     geometry_size date_lower date_upper  \\\n",
       "sample                                                                     \n",
       "Reference                                      1.0       1992       1992   \n",
       "GCA_009670865.1_ASM967086v1_genomic            4.0       1971       1971   \n",
       "GCA_008630435.1_ASM863043v1_genomic            1.0       1996       1996   \n",
       "GCA_000169635.1_ASM16963v1_genomic             2.0       2005       2005   \n",
       "GCA_000170275.1_ASM17027v1_genomic             2.0       1991       1991   \n",
       "...                                            ...        ...        ...   \n",
       "NODE23                                         2.0         NA         NA   \n",
       "NODE24                                         2.0         NA         NA   \n",
       "NODE25                                         2.0         NA         NA   \n",
       "NODE26                                         1.0         NA         NA   \n",
       "root_rtt_dist                                  NaN        NaN        NaN   \n",
       "\n",
       "                                     divtree_coord_x  divtree_coord_y  \\\n",
       "sample                                                                  \n",
       "Reference                                   0.000072          28.0000   \n",
       "GCA_009670865.1_ASM967086v1_genomic         0.000053           7.0000   \n",
       "GCA_008630435.1_ASM863043v1_genomic         0.000086          15.0000   \n",
       "GCA_000169635.1_ASM16963v1_genomic          0.000071          26.0000   \n",
       "GCA_000170275.1_ASM17027v1_genomic          0.000071          25.0000   \n",
       "...                                              ...              ...   \n",
       "NODE23                                      0.000060          24.9375   \n",
       "NODE24                                      0.000066          25.8750   \n",
       "NODE25                                      0.000066          26.7500   \n",
       "NODE26                                      0.000066          27.5000   \n",
       "root_rtt_dist                                    NaN              NaN   \n",
       "\n",
       "                                     rtt_dist branch_major_color  \\\n",
       "sample                                                             \n",
       "Reference                            0.000072            #ff0000   \n",
       "GCA_009670865.1_ASM967086v1_genomic  0.000053            #4062fa   \n",
       "GCA_008630435.1_ASM863043v1_genomic  0.000086            #80ffb4   \n",
       "GCA_000169635.1_ASM16963v1_genomic   0.000071            #ff0000   \n",
       "GCA_000170275.1_ASM17027v1_genomic   0.000071            #ff0000   \n",
       "...                                       ...                ...   \n",
       "NODE23                               0.000060            #ff6232   \n",
       "NODE24                               0.000066            #ff0000   \n",
       "NODE25                               0.000066            #ff0000   \n",
       "NODE26                               0.000066            #ff0000   \n",
       "root_rtt_dist                             NaN                NaN   \n",
       "\n",
       "                                     timetree_coord_x  timetree_coord_y  \\\n",
       "sample                                                                    \n",
       "Reference                                  8352.42496           28.0000   \n",
       "GCA_009670865.1_ASM967086v1_genomic        8331.42500            7.0000   \n",
       "GCA_008630435.1_ASM863043v1_genomic        8356.42610           15.0000   \n",
       "GCA_000169635.1_ASM16963v1_genomic         8365.42496           26.0000   \n",
       "GCA_000170275.1_ASM17027v1_genomic         8351.42496           25.0000   \n",
       "...                                               ...               ...   \n",
       "NODE23                                     7684.30810           24.9375   \n",
       "NODE24                                     7691.72496           25.8750   \n",
       "NODE25                                     7691.72496           26.7500   \n",
       "NODE26                                     7691.72496           27.5000   \n",
       "root_rtt_dist                                     NaN               NaN   \n",
       "\n",
       "                                     timetree_date_calendar  \\\n",
       "sample                                                        \n",
       "Reference                                            1992.0   \n",
       "GCA_009670865.1_ASM967086v1_genomic                  1971.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic                  1996.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic                   2005.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic                   1991.0   \n",
       "...                                                     ...   \n",
       "NODE23                                               1324.0   \n",
       "NODE24                                               1332.0   \n",
       "NODE25                                               1332.0   \n",
       "NODE26                                               1332.0   \n",
       "root_rtt_dist                                           NaN   \n",
       "\n",
       "                                     timetree_CI_lower_err  \\\n",
       "sample                                                       \n",
       "Reference                                              0.0   \n",
       "GCA_009670865.1_ASM967086v1_genomic                    0.0   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    0.0   \n",
       "GCA_000169635.1_ASM16963v1_genomic                     0.0   \n",
       "GCA_000170275.1_ASM17027v1_genomic                     0.0   \n",
       "...                                                    ...   \n",
       "NODE23                                             49805.0   \n",
       "NODE24                                             44407.0   \n",
       "NODE25                                             44407.0   \n",
       "NODE26                                             44407.0   \n",
       "root_rtt_dist                                          NaN   \n",
       "\n",
       "                                     timetree_CI_upper_err root_rtt_dist  \\\n",
       "sample                                                                     \n",
       "Reference                                              0.0            NA   \n",
       "GCA_009670865.1_ASM967086v1_genomic                    0.0            NA   \n",
       "GCA_008630435.1_ASM863043v1_genomic                    0.0            NA   \n",
       "GCA_000169635.1_ASM16963v1_genomic                     0.0            NA   \n",
       "GCA_000170275.1_ASM17027v1_genomic                     0.0            NA   \n",
       "...                                                    ...           ...   \n",
       "NODE23                                               559.0            NA   \n",
       "NODE24                                               560.0            NA   \n",
       "NODE25                                               560.0            NA   \n",
       "NODE26                                               560.0            NA   \n",
       "root_rtt_dist                                          NaN            NA   \n",
       "\n",
       "                                    clade_rtt_dist  \n",
       "sample                                              \n",
       "Reference                                       NA  \n",
       "GCA_009670865.1_ASM967086v1_genomic             NA  \n",
       "GCA_008630435.1_ASM863043v1_genomic             NA  \n",
       "GCA_000169635.1_ASM16963v1_genomic              NA  \n",
       "GCA_000170275.1_ASM17027v1_genomic              NA  \n",
       "...                                            ...  \n",
       "NODE23                                          NA  \n",
       "NODE24                                          NA  \n",
       "NODE25                                          NA  \n",
       "NODE26                                          NA  \n",
       "root_rtt_dist                                   NA  \n",
       "\n",
       "[56 rows x 55 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata_gdf[\"root_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "metadata_gdf[\"clade_rtt_dist\"] = [NO_DATA_CHAR] * len(metadata_gdf)\n",
    "\n",
    "for c in divtree.find_clades():\n",
    "    \n",
    "\n",
    "display(metadata_gdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rotary-document",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ancient_gdf = metadata_gdf[metadata_gdf[\"biosample_comment\"].str.contains(\"Ancient\")]\n",
    "#modern_gdf  = metadata_gdf[metadata_gdf[\"biosample_comment\"].str.contains(\"Modern\")]\n",
    "\n",
    "#display(ancient_gdf)\n",
    "#display(modern_gdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "black-classroom",
   "metadata": {},
   "source": [
    "---\n",
    "# 4. Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "featured-digit",
   "metadata": {},
   "source": [
    "## Timetree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "earlier-wisconsin",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_timetree = copy.deepcopy(timetree)\n",
    "\n",
    "for c in out_timetree.find_clades():\n",
    "    if not c.confidence:\n",
    "        c.confidence = 0\n",
    "\n",
    "metadata_to_comment(out_timetree, metadata_gdf)    \n",
    "out_timetree_path = os.path.join(out_dir, \"final.timetree.nex\")\n",
    "Phylo.write(out_timetree, out_timetree_path, \"nexus\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "subjective-calendar",
   "metadata": {},
   "source": [
    "## Divtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "southeast-facing",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_divtree = copy.deepcopy(divtree)\n",
    "\n",
    "for c in out_divtree.find_clades():\n",
    "    if not c.confidence:\n",
    "        c.confidence = 0\n",
    "\n",
    "metadata_to_comment(out_divtree, metadata_gdf)    \n",
    "out_divtree_path = os.path.join(out_dir, \"final.divtree.nex\")\n",
    "Phylo.write(out_divtree, out_divtree_path, \"nexus\", format_branch_length=\"%1.10f\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "limited-modem",
   "metadata": {},
   "source": [
    "## Augur JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "distinct-index",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(metadata_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "manual-scottish",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple_gdf = metadata_gdf[[\"strain\",\"branch_length\",\"timetree_date_calendar\"]]\n",
    "simple_gdf.rename(columns={\"timetree_date_calendar\": \"timetree_numdate\"}, inplace=True)\n",
    "\n",
    "columns = [\n",
    "    # Draw Divergence Tree\n",
    "    \"branch_length\",\n",
    "    # Draw Time Tree\n",
    "    \"timetree_date_calendar\",\n",
    "    # Draw Country Map\n",
    "    \"mugration_country\",\n",
    "    \"mugration_country_confidence\",\n",
    "    # Draw Province Map\n",
    "    \"mugration_province\",\n",
    "    \"mugration_province_confidence\",\n",
    "    # Colors and Filters\n",
    "    \"mugration_branch_major\",\n",
    "    \"mugration_branch_major_confidence\",\n",
    "    \"mugration_branch_minor\",\n",
    "    \"mugration_branch_minor_confidence\",   \n",
    "    \"branch_number\",\n",
    "    \"branch_support\",\n",
    "    \"continent\",\n",
    "    \"node_type\",\n",
    "    \"geometry_size\",\n",
    "    # Text Description\n",
    "    \"biosample_accession\",\n",
    "    \"strain\",\n",
    "    # Tip Dates\n",
    "    \"date_mean\",\n",
    "    \"date_err\",\n",
    "    \"date_bp_mean\",\n",
    "]\n",
    "\n",
    "# Edit df\n",
    "auspice_gdf = metadata_gdf[columns]\n",
    "auspice_gdf[\"timetree_num_date_confidence\"] = [NO_DATA_CHAR] * len(auspice_gdf)\n",
    "\n",
    "for attr in MUG_ATTRIBUTE_LIST:\n",
    "    col = attr + \"_conf_category\"\n",
    "    auspice_gdf[col] = [\"NA\"] * len(auspice_gdf)\n",
    "    for rec in auspice_gdf.iterrows():\n",
    "        sample = rec[0]\n",
    "        conf = auspice_gdf[\"mugration_\" + attr + \"_confidence\"][sample]\n",
    "        if conf >= 0.95:\n",
    "            category = \"HIGH\"\n",
    "        elif conf < 0.95:\n",
    "            category = \"LOW\"\n",
    "        auspice_gdf.at[sample, col] = category\n",
    "\n",
    "for rec in metadata_gdf.iterrows():\n",
    "    sample = rec[0]\n",
    "    date = rec[1][\"timetree_date_calendar\"]\n",
    "    date_lower = float(date) - rec[1][\"timetree_CI_lower_err\"]\n",
    "    date_upper = float(date) + rec[1][\"timetree_CI_upper_err\"]\n",
    "    #date_confidence = \"{}:{}\".format(date_lower, date_upper)\n",
    "    date_confidence = [date_lower, date_upper]\n",
    "    auspice_gdf.at[sample,\"timetree_num_date_confidence\"] = date_confidence\n",
    "    \n",
    "auspice_gdf.rename(columns={\"timetree_date_calendar\": \"timetree_num_date\"}, inplace=True)\n",
    "\n",
    "display(auspice_gdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "legendary-listening",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import sys, importlib\n",
    "#importlib.reload(sys.modules['plot_phylo_functions'])\n",
    "#from plot_phylo_functions import augur_export,auspice_export\n",
    "\n",
    "augur_dict = augur_export(\n",
    "    tree_path=out_divtree_path, \n",
    "    aln_path=aln_path, \n",
    "    #tree=timetree,     \n",
    "    tree=divtree, \n",
    "    tree_df=auspice_gdf, \n",
    "    color_keyword_exclude=[\"geometry\"],\n",
    "    #color_keyword_exclude=[\"color\", \"coord\", \"reg\", \"lat\", \"lon\"],\n",
    "    type_convert = {\n",
    "        \"branch_number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])\n",
    "\n",
    "out_path_augur_json = os.path.join(out_dir, \"augur_All.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "\n",
    "auspice_dict = auspice_export(\n",
    "    #tree=timetree, \n",
    "    tree=divtree,\n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=out_path_colors,\n",
    "    auspice_latlons_path=out_path_latlon, \n",
    "    )\n",
    "\n",
    "label_col = list(auspice_gdf.columns)\n",
    "\n",
    "# Recursively add branch attrs\n",
    "branch_attributes(\n",
    "    tree_dict=auspice_dict[\"tree\"], \n",
    "    sub_dict=auspice_dict[\"tree\"], \n",
    "    df=auspice_gdf,\n",
    "    label_col=label_col,\n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(out_dir, \"auspice_All.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\\n\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "#out_path_auspice_remote_json = os.path.join(auspice_remote_dir, AUSPICE_PREFIX + \"_All.json\" )\n",
    "#utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "#export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "#print(\"Validation successful for remote JSON.\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "boxed-conclusion",
   "metadata": {},
   "source": [
    "## Auspice Subtrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hispanic-bracelet",
   "metadata": {},
   "outputs": [],
   "source": [
    "for branch in BRANCH_LIST:\n",
    "    print(branch)\n",
    "    timetree_copy = copy.deepcopy(timetree)\n",
    "    \n",
    "    # Create the subtree df\n",
    "    subtree_df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[branch])]\n",
    "\n",
    "    subtree_df.sort_values(\"timetree_coord_y\", inplace=True)\n",
    "    subtree_tips = subtree_df[subtree_df[\"node_type\"] == \"terminal\"]\n",
    "    subtree_internals = subtree_df[subtree_df[\"node_type\"] == \"internal\"]\n",
    "\n",
    "    subtree = extract_subtree(\n",
    "        tree=timetree_copy, \n",
    "        tips=list(subtree_tips.index), \n",
    "        df=metadata_gdf, \n",
    "        color_branches=False)\n",
    "    \n",
    "    metadata_to_comment(subtree, metadata_gdf)   \n",
    "\n",
    "    # Export subtree\n",
    "    out_subtree_path = os.path.join(out_dir, \"{}.nex\".format(branch))\n",
    "    Phylo.write(subtree, out_subtree_path, \"nexus\")\n",
    "\n",
    "    #fig, ax = plt.subplots(1)\n",
    "    #Phylo.draw(subtree, axes=ax, show_confidence=False, label_func = lambda x: '', do_show=False)\n",
    "    \n",
    "    augur_dict = augur_export(\n",
    "        tree_path=out_subtree_path, \n",
    "        aln_path=aln_path,  \n",
    "        tree=subtree, \n",
    "        tree_df=auspice_gdf, \n",
    "        color_keyword_exclude=[\"geometry\"],\n",
    "        type_convert = {\n",
    "            \"branch_number\" : (lambda x : str(x))\n",
    "        },\n",
    "    ) \n",
    "        \n",
    "    out_path_augur_json = os.path.join(out_dir, \"augur_{}.json\".format(branch) )\n",
    "    utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "    \n",
    "    auspice_dict = auspice_export(\n",
    "        tree=subtree,\n",
    "        augur_json_paths=[out_path_augur_json], \n",
    "        auspice_config_path=auspice_config_path, \n",
    "        auspice_colors_path=out_path_colors,\n",
    "        auspice_latlons_path=out_path_latlon, \n",
    "        )    \n",
    "    \n",
    "    label_col = list(auspice_gdf.columns)\n",
    "\n",
    "    # Recursively add branch attrs\n",
    "    branch_attributes(\n",
    "        tree_dict=auspice_dict[\"tree\"], \n",
    "        sub_dict=auspice_dict[\"tree\"], \n",
    "        df=auspice_gdf,\n",
    "        label_col=label_col,\n",
    "        )    \n",
    "    \n",
    "    # Write outputs - For Local Rendering\n",
    "    out_path_auspice_local_json = os.path.join(out_dir, \"auspice_{}.json\".format(branch) )\n",
    "    utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "    export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "    print(\"Validation successful for local JSON.\\n\")\n",
    "\n",
    "    # Write outputs - For Remote Rendering\n",
    "    #out_path_auspice_remote_json = os.path.join(auspice_remote_dir, AUSPICE_PREFIX + \"_{}.json\".format(branch) )\n",
    "    #utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "    #export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "    #print(\"Validation successful for remote JSON.\\n\")    \n",
    "\n",
    "\"\"\"\n",
    "# Terminal df\n",
    "subtree_tips = subtree_df[subtree_df[\"node_type\"] == \"terminal\"]\n",
    "\n",
    "# Internal df\n",
    "subtree_internals = subtree_df[subtree_df[\"node_type\"] == \"internal\"]\n",
    "timetree_copy = copy.deepcopy(timetree)\n",
    "subtree = extract_subtree(tree=timetree_copy, tips=list(subtree_tips.index), df=metadata_gdf, color_branches=False)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "respective-million",
   "metadata": {},
   "source": [
    "---\n",
    "# 5. Plot RTT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beneficial-linux",
   "metadata": {},
   "source": [
    "## All"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "moving-stuff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [480, 480]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "geo_dist_all = []\n",
    "genetic_dist_all = []\n",
    "\n",
    "fig, ax = plt.subplots(1,figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "df = metadata_gdf[(metadata_gdf[\"node_type\"] == \"terminal\") & (metadata_gdf[\"date_mean\"] != NO_DATA_CHAR)]\n",
    "reg_x = \"date_mean\"\n",
    "reg_y = \"rtt_dist\"\n",
    "reg_err = \"date_err\"\n",
    "\n",
    "# -----------------------\n",
    "# Regression\n",
    "slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(list(df[reg_x]),list(df[reg_y]))\n",
    "r2 = r_value * r_value    \n",
    "y_intercept = CURRENT_YEAR - (0 - intercept)/slope\n",
    "p_sig = \"\"\n",
    "p_thresh = 0.05    \n",
    "if p_value < p_thresh:\n",
    "    p_sig = \"*\"\n",
    "\n",
    "mut_per_year = SEQ_LEN * slope\n",
    "year_per_mut = 1 / mut_per_year\n",
    "reg_dict = {\n",
    "    \"slope\": slope,\n",
    "    \"x_intercept\" : intercept,\n",
    "    \"y_intercept\" : y_intercept,\n",
    "    \"p_value\" : p_value,\n",
    "    \"p_sig\" : p_sig,\n",
    "    \"year_per_mut\" : year_per_mut\n",
    "}\n",
    "\n",
    "sns.regplot(\n",
    "    ax=ax,\n",
    "    data=df, \n",
    "    x=list(df[reg_x]), \n",
    "    y=list(df[reg_y]), \n",
    "    #robust=True,\n",
    "    ci=95,\n",
    "    scatter_kws={\"s\": 0},\n",
    "    line_kws={\"linewidth\" : 0.5},\n",
    "    color=\"grey\",\n",
    "    #color=df[\"branch_major_color\"][0],\n",
    "    label=(\n",
    "    \"   R²: {}\".format(round(r2,2))\n",
    "    + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)\n",
    "    + \"\\nRate: {:.2} +/ {:.2}\".format(slope, std_err)  \n",
    "    + \"\\n         {} years/substitution\".format(round(year_per_mut))\n",
    "    + \"\\nRoot: {} YBP\".format(round(y_intercept))                \n",
    "    )\n",
    ")\n",
    "\n",
    "ax.errorbar(\n",
    "    data=df, \n",
    "    x=reg_x, \n",
    "    y=reg_y, \n",
    "    xerr=reg_err,\n",
    "    yerr=None,\n",
    "    ls='none',\n",
    "    c = \"black\",  \n",
    "    capsize=1,\n",
    "    label=None,\n",
    "    zorder=1,\n",
    "    lw=0.5,\n",
    "    )  \n",
    "\n",
    "sns.scatterplot(\n",
    "    ax=ax,\n",
    "    data=df, \n",
    "    x=reg_x, \n",
    "    y=reg_y, \n",
    "    s=10,    \n",
    "    ec = \"white\", \n",
    "    color=\"black\",\n",
    "    alpha=0.75,\n",
    "    zorder=2,\n",
    ") \n",
    "\n",
    "    \n",
    "# Set xlimits\n",
    "xlim = ax.get_xlim()\n",
    "x_buff = round((xlim[1] - xlim[0]) * 0.1,1)\n",
    "new_xlim = [xlim[0]-x_buff, xlim[1]+x_buff]\n",
    "if new_xlim[1] > CURRENT_YEAR:\n",
    "    new_xlim = [new_xlim[0], CURRENT_YEAR + 200]\n",
    "ax.set_xlim(new_xlim)\n",
    "    \n",
    "ax.set_ylabel(\"Root to Tip Distance\")\n",
    "ax.set_xlabel(\"Date\")    \n",
    "ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0)) \n",
    "\n",
    "# Format and position legend\n",
    "legend = ax.legend(\n",
    "    borderpad=0.8, \n",
    "    handletextpad=-2, \n",
    "    edgecolor=\"black\", \n",
    "    #bbox_to_anchor=(0.5, -0.30), \n",
    "    #loc='center',\n",
    "    fontsize=FONTSIZE,\n",
    ")\n",
    "frame = legend.get_frame().set_linewidth(0.5)\n",
    "\n",
    "out_path = os.path.join(out_dir, \"rtt_all\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "painted-reply",
   "metadata": {},
   "source": [
    "## Clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "reserved-separate",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "reg_x = \"date_mean\"\n",
    "reg_y = \"rtt_dist\"\n",
    "reg_err = \"date_err\"\n",
    "\n",
    "nrow=3\n",
    "ncol=4\n",
    "\n",
    "fig, axes = plt.subplots(nrow,ncol,figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.35, hspace=1)\n",
    "\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for label in BRANCH_LIST:\n",
    "    print(\"Branch:\", label)\n",
    "    \n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]  \n",
    "    # If no records for this branch, continue\n",
    "    if len(df) < 2: continue\n",
    "    # -----------------------\n",
    "    # Regression\n",
    "    slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(list(df[reg_x]),list(df[reg_y]))\n",
    "    r2 = r_value * r_value    \n",
    "    y_intercept = CURRENT_YEAR - (0 - intercept)/slope\n",
    "    p_sig = \"\"\n",
    "    p_thresh = 0.05    \n",
    "    if p_value < p_thresh:\n",
    "        p_sig = \"*\"\n",
    "\n",
    "    mut_per_year = SEQ_LEN * slope\n",
    "    year_per_mut = 1 / mut_per_year\n",
    "    \n",
    "    reg_dict = {\n",
    "        \"slope\": slope,\n",
    "        \"x_intercept\" : intercept,\n",
    "        \"y_intercept\" : y_intercept,\n",
    "        \"p_value\" : p_value,\n",
    "        \"p_sig\" : p_sig,\n",
    "        \"year_per_mut\" : year_per_mut\n",
    "    }\n",
    "              \n",
    "    sns.regplot(\n",
    "        ax=ax,\n",
    "        data=df, \n",
    "        x=list(df[reg_x]), \n",
    "        y=list(df[reg_y]), \n",
    "        #robust=True,\n",
    "        ci=95,\n",
    "        scatter_kws={\"s\": 0},\n",
    "        line_kws={\"linewidth\" : 0.5},\n",
    "        color=\"grey\",\n",
    "        #color=df[\"branch_major_color\"][0],\n",
    "        label=(\n",
    "        \"   R²: {}\".format(round(r2,2))\n",
    "        + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)\n",
    "        + \"\\nRate: {:.2} +/ {:.2}\".format(slope, std_err)  \n",
    "        + \"\\n         {} years/substitution\".format(round(year_per_mut))\n",
    "        + \"\\nRoot: {} YBP\".format(round(y_intercept))                \n",
    "        )\n",
    "    )\n",
    "    \n",
    "    ax.errorbar(\n",
    "        data=df, \n",
    "        x=\"date_mean\",\n",
    "        y=\"rtt_dist\", \n",
    "        xerr=\"date_err\",\n",
    "        yerr=None,\n",
    "        ls='none',\n",
    "        c = df[\"branch_major_color\"][0],  \n",
    "        capsize=1,\n",
    "        label=None,\n",
    "        zorder=1,\n",
    "        lw=0.5,\n",
    "        )  \n",
    "    \n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        data=df, \n",
    "        x=\"date_mean\",\n",
    "        s=10,\n",
    "        y=\"rtt_dist\", \n",
    "        c = df[\"branch_major_color\"], \n",
    "        ec = \"black\",\n",
    "        alpha=0.75,\n",
    "        zorder=2,\n",
    "    )    \n",
    "    \n",
    "    # Set xlimits\n",
    "    xlim = ax.get_xlim()\n",
    "    x_buff = round((xlim[1] - xlim[0]) * 0.1,1)\n",
    "    new_xlim = [xlim[0]-x_buff, xlim[1]+x_buff]\n",
    "    if new_xlim[1] > CURRENT_YEAR:\n",
    "        new_xlim = [new_xlim[0], CURRENT_YEAR]\n",
    "    ax.set_xlim(new_xlim)\n",
    "\n",
    "    # Format and position legend\n",
    "    legend = ax.legend(\n",
    "        borderpad=0.8, \n",
    "        handletextpad=-2, \n",
    "        edgecolor=\"black\", \n",
    "        bbox_to_anchor=(0.5, -0.55), \n",
    "        loc='center',\n",
    "        fontsize=FONTSIZE * 0.80,\n",
    "    )\n",
    "    frame = legend.get_frame().set_linewidth(0.5)\n",
    "        \n",
    "    # Format axis\n",
    "    ax.set_title(label)\n",
    "\n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Root to Tip Distance\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Date\")    \n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "    \n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_linewidth(0.5)    \n",
    "            \n",
    "\n",
    "    # Update axis\n",
    "    if i_col == ncol - 1:\n",
    "        i_col = 0\n",
    "        i_row += 1\n",
    "    else:\n",
    "        i_col += 1\n",
    "        \n",
    "    \n",
    "out_path = os.path.join(out_dir, \"rtt_clades\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "successful-egyptian",
   "metadata": {},
   "source": [
    "# Tip Dating"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "current-prairie",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "reg_x = \"date_mean\"\n",
    "reg_y = \"rtt_dist\"\n",
    "reg_err = \"date_err\"\n",
    "\n",
    "nrow=3\n",
    "ncol=3\n",
    "\n",
    "fig, axes = plt.subplots(nrow,ncol,figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.35, hspace=0.35)\n",
    "#fig.subplots_adjust(wspace=0.35, hspace=0.85)\n",
    "\n",
    "for row in axes:\n",
    "    for ax in row:\n",
    "        for spine in ax.spines:\n",
    "            ax.spines[spine].set_linewidth(0.5) \n",
    "\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for label in ANCIENT_BRANCH_LIST:\n",
    "    print(\"Branch:\", label)\n",
    "    \n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    if len(df) < 2: continue\n",
    "    \n",
    "    # -----------------------\n",
    "    # Regression\n",
    "    slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(list(df[reg_x]),list(df[reg_y]))\n",
    "    r2 = r_value * r_value    \n",
    "    y_intercept = CURRENT_YEAR - (0 - intercept)/slope\n",
    "    p_sig = \"\"\n",
    "    p_thresh = 0.05    \n",
    "    if p_value < p_thresh:\n",
    "        p_sig = \"*\"\n",
    "        \n",
    "    mut_per_year = SEQ_LEN * slope\n",
    "    year_per_mut = 1 / mut_per_year\n",
    "    \n",
    "    reg_dict = {\n",
    "        \"slope\": slope,\n",
    "        \"x_intercept\" : intercept,\n",
    "        \"y_intercept\" : y_intercept,\n",
    "        \"p_value\" : p_value,\n",
    "        \"p_sig\" : p_sig,\n",
    "        \"year_per_mut\" : year_per_mut\n",
    "    }\n",
    "    \n",
    "              \n",
    "    sns.regplot(\n",
    "        ax=ax,\n",
    "        data=df, \n",
    "        x=list(df[reg_x]),\n",
    "        y=list(df[reg_y]),\n",
    "        #robust=True,\n",
    "        ci=95,\n",
    "        scatter_kws={\"s\": 0},\n",
    "        line_kws={\"linewidth\" : 0.5},\n",
    "        color=\"grey\",\n",
    "        #color=df[\"branch_major_color\"][0],\n",
    "        label=(\n",
    "        \"   R²: {}\".format(round(r2,2))\n",
    "        + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)\n",
    "        + \"\\nRate: {:.2} +/ {:.2}\".format(slope, std_err)  \n",
    "        + \"\\n         {} years/substitution\".format(round(year_per_mut))\n",
    "        + \"\\nRoot: {} YBP\".format(round(y_intercept))                \n",
    "        )\n",
    "    )\n",
    "    \n",
    "    ax.errorbar(\n",
    "        data=df, \n",
    "        x=\"date_mean\",\n",
    "        y=\"rtt_dist\", \n",
    "        xerr=\"date_err\",\n",
    "        yerr=None,\n",
    "        ls='none',\n",
    "        c = df[\"branch_major_color\"][0],  \n",
    "        capsize=1,\n",
    "        label=None,\n",
    "        zorder=1,\n",
    "        lw=0.5,\n",
    "        )  \n",
    "    \n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        data=df, \n",
    "        x=\"date_mean\",\n",
    "        s=10,\n",
    "        y=\"rtt_dist\", \n",
    "        c = df[\"branch_major_color\"], \n",
    "        ec = \"black\",\n",
    "        alpha=0.75,\n",
    "        zorder=2,\n",
    "    )    \n",
    "    \n",
    "        \n",
    "    # Set xlimits\n",
    "    xlim = ax.get_xlim()\n",
    "    x_buff = round((xlim[1] - xlim[0]) * 0.1,1)\n",
    "    new_xlim = [xlim[0]-x_buff, xlim[1]+x_buff]\n",
    "    if new_xlim[1] > CURRENT_YEAR:\n",
    "        new_xlim = [new_xlim[0], CURRENT_YEAR]\n",
    "    ax.set_xlim(new_xlim)\n",
    "\n",
    "    # Format and position legend\n",
    "    legend = ax.legend(\n",
    "        borderpad=0.8, \n",
    "        handletextpad=-2, \n",
    "        edgecolor=\"black\", \n",
    "        bbox_to_anchor=(0.5, -0.55), \n",
    "        loc='center',\n",
    "        fontsize=FONTSIZE * 0.80,\n",
    "    )\n",
    "    frame = legend.get_frame().set_linewidth(0.5)\n",
    "    ax.get_legend().remove()\n",
    "    \n",
    "    # Format axis\n",
    "    ax.set_title(label)\n",
    "\n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Root to Tip Distance\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Date\")    \n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "      \n",
    "            \n",
    "    # Now predicted values\n",
    "    i_row += 1\n",
    "    ax = axes[i_row][i_col]    \n",
    "    \n",
    "    date_diffs = []\n",
    "    for date_mean, rtt_dist in zip(df[\"date_mean\"], df[\"rtt_dist\"]):\n",
    "        pred_date = (rtt_dist - reg_dict[\"x_intercept\"]) / reg_dict[\"slope\"]\n",
    "        date_diff = pred_date - date_mean\n",
    "        date_diffs.append(date_diff)\n",
    "        \n",
    "    sns.kdeplot(\n",
    "       date_diffs,\n",
    "       ax=ax,\n",
    "       fill=True, \n",
    "       common_norm=True,\n",
    "       alpha=0.5, \n",
    "       linewidth=1,\n",
    "       bw_adjust=0.25,\n",
    "       color=df[\"branch_major_color\"][0],\n",
    "    )\n",
    "    \n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Density\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Regression Date - Actual Date\")    \n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))  \n",
    "    ax.set_xlim(-300,300)    \n",
    "\n",
    "    # ---------------------------------------------------------------\n",
    "    # Diff based on LSD2 estimate\n",
    "    i_row += 1\n",
    "    ax = axes[i_row][i_col]  \n",
    "    \n",
    "    date_diffs = []\n",
    "    for date_mean, timetree_date in zip(df[\"date_mean\"], df[\"timetree_date\"]):\n",
    "        date_est = CURRENT_YEAR + float(timetree_date)\n",
    "        date_diff = date_est - date_mean\n",
    "        date_diffs.append(date_diff)  \n",
    "        \n",
    "    sns.kdeplot(\n",
    "       date_diffs,\n",
    "       ax=ax,\n",
    "       fill=True, \n",
    "       common_norm=True,\n",
    "       alpha=0.5, \n",
    "       linewidth=1,\n",
    "       bw_adjust=0.25,\n",
    "       color=df[\"branch_major_color\"][0],\n",
    "    )\n",
    "    \n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Density\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"LSD2 Date - Actual Date\")    \n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))   \n",
    "    ax.set_xlim(-300,300)\n",
    "    \n",
    "    # Update axis\n",
    "    i_row = 0\n",
    "    i_col += 1\n",
    "\n",
    "\n",
    "out_path = os.path.join(out_dir, \"tip-dating\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "stainless-reporter",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"All:\", len(metadata_gdf[metadata_gdf[\"node_type\"] == \"terminal\"]))\n",
    "print(\"Ancient:\",len(ancient_gdf), len(ancient_gdf) / len(metadata_gdf))\n",
    "print(\"Modern:\",len(modern_gdf), len(modern_gdf) / len(metadata_gdf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "terminal-degree",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"TARGET_RES = [1280, 720]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "fig, ax = plt.subplots(1, dpi=DPI, figsize=FIGSIZE)\n",
    "\n",
    "rtt_dist_ln = []\n",
    "for rtt_dist in df[\"rtt_dist\"]:\n",
    "    rtt_dist_ln.append(math.log(rtt_dist))\n",
    "\n",
    "sns.kdeplot(\n",
    "   rtt_dist_ln,\n",
    "    ax=ax,\n",
    ")\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "worldwide-survival",
   "metadata": {},
   "source": [
    "## 1.PRE Subtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "toxic-fellowship",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the subtree df\n",
    "subtree_df = metadata_gdf[metadata_gdf[\"mugration_branch_major\"] == \"1.PRE\"]\n",
    "subtree_df.sort_values(\"timetree_coord_y\", inplace=True)\n",
    "\n",
    "# Terminal df\n",
    "subtree_tips = subtree_df[subtree_df[\"node_type\"] == \"terminal\"]\n",
    "\n",
    "# Internal df\n",
    "subtree_internals = subtree_df[subtree_df[\"node_type\"] == \"internal\"]\n",
    "timetree_copy = copy.deepcopy(timetree)\n",
    "subtree = extract_subtree(tree=timetree_copy, tips=list(subtree_tips.index), df=metadata_gdf, color_branches=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "quiet-split",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1080, 1080]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "#fig, axes = plt.subplots(1,figsize=FIGSIZE, dpi=DPI, sharey=True)\n",
    "#ax_tree=axes\n",
    "\n",
    "fig, axes = plt.subplots(1,2,figsize=FIGSIZE, dpi=DPI, sharey=True)\n",
    "fig.subplots_adjust(wspace=0.5)  \n",
    "        \n",
    "##############################################################\n",
    "# TIPS\n",
    "ax=axes[0]\n",
    "\n",
    "# Give the numdate attribute\n",
    "for c in subtree.find_clades():\n",
    "    c.numdate = metadata_gdf[\"timetree_date_calendar\"][c.name]\n",
    "\n",
    "Phylo.draw(subtree,\n",
    "           axes=ax,\n",
    "             show_confidence=False, \n",
    "             label_func = lambda x: '',\n",
    "             #label_func = lambda x: \"  {}\".format(metadata_gdf[\"country\"][x.name]) if x.is_terminal() else '',\n",
    "             do_show=False,\n",
    "             branch_labels = lambda x: (\n",
    "                 '*' if not x.is_terminal() and metadata_gdf[\"branch_support\"][x.name] >= 95\n",
    "                 else ''),\n",
    "            )\n",
    "\n",
    "\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_yticks([i for i in range(1,len(subtree_tips)+1)])\n",
    "ax.set_yticklabels(list(subtree_tips[\"country\"]), fontsize=FONTSIZE, ha=\"center\",position=(1.2,0))\n",
    "ax.yaxis.tick_right()\n",
    "\n",
    "\n",
    "# --------------------------------------------\n",
    "# Ticks\n",
    "ticks_dict = convert_timetree_ticks(tree=subtree, step=100)\n",
    "# Set the new tick locations and labels\n",
    "# Leave out last one\n",
    "ax.set_xticks(ticks_dict[\"tick_locs\"][:])\n",
    "ax.set_xticklabels(ticks_dict[\"tick_labels\"][:], ROTATION=45, ha=\"right\")\n",
    "date_to_x = ticks_dict[\"tick_locs\"][0] - float(ticks_dict[\"tick_labels\"][0])\n",
    "\n",
    "# --------------------------------------------\n",
    "# Lines\n",
    "# Vertical lines for dates\n",
    "for loc in ticks_dict[\"tick_locs\"][:]:\n",
    "    ax.axvline(x=loc, color=\"darkgrey\", alpha=0.5, lw=0.25)\n",
    "        \n",
    "\n",
    "\n",
    "# --------------------------------------------\n",
    "# Tip Locations\n",
    "timetree_date_err = [[],[]]\n",
    "for lower,upper in zip(subtree_tips[\"timetree_CI_lower_err\"], subtree_tips[\"timetree_CI_upper_err\"] ):\n",
    "    timetree_date_err[0].append(lower)\n",
    "    timetree_date_err[1].append(upper)\n",
    "    \n",
    "ax.errorbar(\n",
    "    data=subtree_tips, \n",
    "    x=[float(d) + date_to_x for d in subtree_tips[\"timetree_date_calendar\"]],\n",
    "    y=[i-0.1 for i in range(1,len(subtree_tips)+1,1)], \n",
    "    yerr=None, \n",
    "    xerr=timetree_date_err,\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=2,\n",
    "    label=\"Estimated Date\",\n",
    "    zorder=3,\n",
    "    lw=1, \n",
    "    color=\"darkgoldenrod\",\n",
    "    marker=\"o\",\n",
    "    markersize=5,\n",
    "    markerfacecolor=subtree_tips[\"branch_major_color\"][0], \n",
    "    \n",
    ")\n",
    "\n",
    "ax.errorbar(\n",
    "    data=subtree_tips, \n",
    "    x=[d + date_to_x for d in subtree_tips[\"date_mean\"]],\n",
    "    y=[i+0.1 for i in range(1,len(subtree_tips)+1,1)], \n",
    "    yerr=None, \n",
    "    xerr=\"date_err\",\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=2,\n",
    "    label=\"Archaeological Range\",\n",
    "    zorder=2,\n",
    "    lw=1, \n",
    "    color=\"blue\", \n",
    "    #c=subtree_tips[\"branch_major_color\"][0],    \n",
    ")\n",
    "\n",
    "ax.set_title(\"Tip Dating\", fontsize=FONTSIZE * 2)\n",
    "ax.set_xlabel(\"Date\")\n",
    "\n",
    "legend = ax.legend(fontsize=FONTSIZE, loc=\"center\", bbox_to_anchor=(1.25,1.05), edgecolor=\"black\")\n",
    "legend.get_frame().set_linewidth(0.5)\n",
    "\n",
    "##############################################################\n",
    "# INTERNAL AXIS\n",
    "ax=axes[1]\n",
    "Phylo.draw(subtree,\n",
    "           axes=ax,\n",
    "             show_confidence=False, \n",
    "             label_func = lambda x: '',\n",
    "             #label_func = lambda x: \"  {}\".format(metadata_gdf[\"country\"][x.name]) if x.is_terminal() else '',\n",
    "             #label_func = lambda x: x.name if not x.is_terminal() else '',\n",
    "             do_show=False,\n",
    "            )\n",
    "# --------------------------------------------\n",
    "# Ticks\n",
    "ticks_dict = convert_timetree_ticks(tree=subtree, step=100)\n",
    "# Set the new tick locations and labels\n",
    "# Leave out last one\n",
    "ax.set_xticks(ticks_dict[\"tick_locs\"][:])\n",
    "ax.set_xticklabels(ticks_dict[\"tick_labels\"][:], ROTATION=45, ha=\"right\")\n",
    "date_to_x = ticks_dict[\"tick_locs\"][0] - float(ticks_dict[\"tick_labels\"][0])\n",
    "\n",
    "# --------------------------------------------\n",
    "# Lines\n",
    "# Vertical lines for dates\n",
    "for loc in ticks_dict[\"tick_locs\"][:]:\n",
    "    ax.axvline(x=loc, color=\"darkgrey\", alpha=0.5, lw=0.25)\n",
    "    \n",
    "ax.set_title(\"Node Dating\", fontsize=FONTSIZE * 2)\n",
    "ax.set_xlabel(\"Date\")\n",
    "\n",
    "# --------------------------------------------\n",
    "# Interanl Locations\n",
    "timetree_date_err = [[],[]]\n",
    "for lower,upper in zip(subtree_internals[\"timetree_CI_lower_err\"], subtree_internals[\"timetree_CI_upper_err\"] ):\n",
    "    timetree_date_err[0].append(lower)\n",
    "    timetree_date_err[1].append(upper)\n",
    "    \n",
    "ax.errorbar(\n",
    "    data=subtree_internals, \n",
    "    x=[d + date_to_x for d in subtree_internals[\"timetree_date_calendar\"]],\n",
    "    y=[i+0.1 for i in range(1,len(subtree_internals)+1,1)], \n",
    "    yerr=None, \n",
    "    xerr=timetree_date_err,\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=2,\n",
    "    zorder=2,\n",
    "    lw=1, \n",
    "    color=\"darkgoldenrod\",\n",
    "    marker=\"o\",\n",
    "    markersize=5,\n",
    "    markerfacecolor=subtree_tips[\"branch_major_color\"][0], \n",
    ")\n",
    "\n",
    "ax.set_ylabel(\"\")\n",
    "\n",
    "out_path = os.path.join(out_dir, \"dating_1.PRE\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "still-carrier",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"       \n",
    "#----------------------------------------------------\n",
    "# Actual Date\n",
    "ax_plot.errorbar(\n",
    "    data=subtree_tips, \n",
    "    x=\"date_mean\",  \n",
    "    y=[float(i) for i in range(1,len(subtree_tips)+1,1)],  \n",
    "    #y=[i for i in range(len(subtree_tips)+1,1,-1)],    \n",
    "    yerr=None, \n",
    "    xerr=\"date_err\",\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=1,\n",
    "    label=\"Archaeological Date\",\n",
    "    zorder=1,\n",
    "    lw=0.5, \n",
    "    color=\"black\", \n",
    "    #c=subtree_tips[\"branch_major_color\"][0],    \n",
    ")\n",
    "\n",
    "\n",
    "sns.scatterplot(\n",
    "    ax=ax,\n",
    "    data=subtree_tips,\n",
    "    x=\"date_mean\", \n",
    "    #y=\"timetree_coord_y\", \n",
    "    y=[i for i in range(0,len(subtree_tips)*data_spacing,data_spacing)],\n",
    "    color=\"black\",\n",
    "    #hue=subtree_tips[\"country\"],\n",
    "    #c=subtree_tips[\"branch_major_color\"],\n",
    "    s=5)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "timetree_date_err = [[],[]]\n",
    "\n",
    "for lower,upper in zip(subtree_tips[\"timetree_CI_lower_err\"], subtree_tips[\"timetree_CI_upper_err\"] ):\n",
    "    timetree_date_err[0].append(lower)\n",
    "    timetree_date_err[1].append(upper)\n",
    "\n",
    "sns.scatterplot(\n",
    "    ax=ax,\n",
    "    data=subtree_tips,\n",
    "    x=\"timetree_date_calendar\",\n",
    "    #y=\"timetree_coord_y\", \n",
    "    y=[i for i in range(3,len(subtree_tips)*data_spacing,data_spacing)],    \n",
    "    color=\"red\",\n",
    "    #hue=subtree_tips[\"country\"],\n",
    "    #c=subtree_tips[\"branch_major_color\"],\n",
    "    s=10) \n",
    "\n",
    "\n",
    "ax.errorbar(\n",
    "    data=subtree_tips, \n",
    "    x=\"timetree_date_calendar\", \n",
    "    #y=[i for i in range(4,len(subtree_tips)*data_spacing,data_spacing)],\n",
    "    y=\"timetree_coord_y\",\n",
    "    yerr=None, \n",
    "    xerr=timetree_date_err,\n",
    "    ls='none',  \n",
    "    #c = \"black\",  \n",
    "    capsize=1,\n",
    "    label=\"Estimated Date\",\n",
    "    zorder=1,\n",
    "    lw=0.5, \n",
    "    color=\"red\", \n",
    "    #c=subtree_tips[\"branch_major_color\"][0],    \n",
    ")\n",
    "\n",
    "\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_yticklabels(list(subtree_tips[\"country\"]), fontsize=FONTSIZE)\n",
    "ax.set_xlabel(\"Date\")\n",
    "ax.set_title(\"Leave-One-Out Tip Dating (1.PRE)\")\n",
    "ax.legend(bbox_to_anchor=(1,1), fontsize=FONTSIZE)\n",
    "\n",
    "#display(subtree_tips)\n",
    "\n",
    "out_path = os.path.join(out_dir, \"leave-one-out\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "passive-tissue",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(subtree_tips)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "confidential-essence",
   "metadata": {},
   "source": [
    "---\n",
    "# Geography"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prepared-import",
   "metadata": {},
   "source": [
    "## Isolation By Distance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "growing-plasma",
   "metadata": {},
   "source": [
    "### Prep Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "northern-government",
   "metadata": {},
   "outputs": [],
   "source": [
    "ibd_dict = {}\n",
    "\n",
    "for label in BRANCH_LIST:\n",
    "    print(\"Label:\", label)    \n",
    "    ibd_dict[label] = {\"geo_dist\" : [], \"genetic_dist\" : []}\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    \n",
    "    i_compare = 0\n",
    "    i = 0\n",
    "    for sample1 in df.index:\n",
    "        geom1 = df[\"geometry\"][sample1]\n",
    "        coord1 = (geom1.y, geom1.x)  \n",
    "        for sample2 in df.index[i_compare:]:\n",
    "            if sample1 == sample2: continue\n",
    "            geom2 = df[\"geometry\"][sample2]\n",
    "            coord2 = (geom2.y, geom2.x)\n",
    "            geo_dist = geopy.distance.great_circle(coord1, coord2).km\n",
    "            genetic_dist = divtree.distance(sample1, sample2)\n",
    "\n",
    "            ibd_dict[label][\"geo_dist\"].append(geo_dist)\n",
    "            ibd_dict[label][\"genetic_dist\"].append(genetic_dist)\n",
    "        i_compare += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "handy-harris",
   "metadata": {},
   "source": [
    "### All"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suited-ratio",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [480, 480]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "geo_dist_all = []\n",
    "genetic_dist_all = []\n",
    "\n",
    "fig, ax = plt.subplots(1,figsize=FIGSIZE, dpi=DPI)\n",
    "\n",
    "for branch in ibd_dict:\n",
    "    geo_dist_all = geo_dist_all + ibd_dict[branch][\"geo_dist\"]\n",
    "    genetic_dist_all = genetic_dist_all + ibd_dict[branch][\"genetic_dist\"]\n",
    "\n",
    "sns.regplot(\n",
    "    ax=ax,\n",
    "    x=geo_dist_all,\n",
    "    y=genetic_dist_all,\n",
    "    ci=95,\n",
    "    scatter_kws={\"s\": 0},\n",
    "    color=\"grey\",\n",
    "    line_kws={\"linewidth\":0.5},\n",
    "    label=(\n",
    "    \"   R²: {}\".format(round(r2,2))\n",
    "    + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)             \n",
    "    )\n",
    ")\n",
    "    \n",
    "sns.scatterplot(\n",
    "    ax=ax,\n",
    "    x=geo_dist_all,\n",
    "    y=genetic_dist_all,\n",
    "    s=1,\n",
    "    ec = \"white\", \n",
    "    color=\"black\",\n",
    "    alpha=0.75,\n",
    "    zorder=2,\n",
    ")\n",
    "\n",
    "# Format and position legend\n",
    "legend = ax.legend(\n",
    "    borderpad=0.8, \n",
    "    handletextpad=-2, \n",
    "    edgecolor=\"black\", \n",
    "    #bbox_to_anchor=(0.5, -0.30), \n",
    "    #loc='center',\n",
    "    fontsize=FONTSIZE,\n",
    ")\n",
    "frame = legend.get_frame().set_linewidth(0.5)\n",
    "\n",
    "ax.set_ylabel(\"Genetic Distance (subs / site)\")\n",
    "ax.set_xlabel(\"Great Circle Distance (km)\")\n",
    "ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "\n",
    "out_path = os.path.join(out_dir, \"ibd_all\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "manual-merchant",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "nrow=3\n",
    "ncol=4\n",
    "\n",
    "fig, axes = plt.subplots(nrow,ncol,figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.35, hspace=0.75)\n",
    "\n",
    "\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for label in BRANCH_LIST:\n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    \n",
    "    for branch_major in colors_dict[\"branch_major\"]:\n",
    "        if branch_major in label:\n",
    "            break  \n",
    "            \n",
    "    # -----------------------\n",
    "    # Regression\n",
    "    slope, intercept, r_value, p_value, std_err = scipy.stats.linregress(ibd_dict[label][\"geo_dist\"],ibd_dict[label][\"genetic_dist\"])\n",
    "    r2 = r_value * r_value    \n",
    "    y_intercept = CURRENT_YEAR - (0 - intercept)/slope\n",
    "    p_sig = \"\"\n",
    "    p_thresh = 0.05    \n",
    "    if p_value < p_thresh:\n",
    "        p_sig = \"*\"\n",
    "        \n",
    "    mut_per_year = SEQ_LEN * slope\n",
    "    year_per_mut = 1 / mut_per_year\n",
    "    \n",
    "    reg_dict = {\n",
    "        \"slope\": slope,\n",
    "        \"x_intercept\" : intercept,\n",
    "        \"y_intercept\" : y_intercept,\n",
    "        \"p_value\" : p_value,\n",
    "        \"p_sig\" : p_sig,\n",
    "        \"year_per_mut\" : year_per_mut\n",
    "    }\n",
    "    \n",
    "    sns.regplot(\n",
    "        ax=ax,\n",
    "        x=ibd_dict[label][\"geo_dist\"],\n",
    "        y=ibd_dict[label][\"genetic_dist\"],\n",
    "        ci=95,\n",
    "        scatter_kws={\"s\": 0},\n",
    "        color=\"grey\",\n",
    "        line_kws={\"linewidth\":0.5},\n",
    "        label=(\n",
    "        \"   R²: {}\".format(round(r2,2))\n",
    "        + \"\\n     p: {:.2e}{}\".format(p_value, p_sig)             \n",
    "        )\n",
    "    )\n",
    "\n",
    "    sns.scatterplot(\n",
    "        ax=ax,\n",
    "        x=ibd_dict[label][\"geo_dist\"],\n",
    "        y=ibd_dict[label][\"genetic_dist\"],\n",
    "        s=10,\n",
    "        ec = \"black\",\n",
    "        color = colors_dict[\"branch_major\"][branch_major],         \n",
    "        alpha=0.75,\n",
    "        zorder=2,\n",
    "    )\n",
    "     \n",
    "    # Format and position legend\n",
    "    legend = ax.legend(\n",
    "        borderpad=0.8, \n",
    "        handletextpad=-2, \n",
    "        edgecolor=\"black\", \n",
    "        bbox_to_anchor=(0.5, -0.40), \n",
    "        loc='center',\n",
    "        fontsize=FONTSIZE * 0.80,\n",
    "    )\n",
    "    frame = legend.get_frame().set_linewidth(0.5)        \n",
    "    \n",
    "    # Set xlimits\n",
    "    xlim = ax.get_xlim()\n",
    "    if xlim[1] > 20000:\n",
    "        xbuff = 5000   \n",
    "    elif xlim[1] > 2000:\n",
    "        xbuff = 500\n",
    "    elif xlim[1] > 200:\n",
    "        xbuff = 50\n",
    "    ax.set_xlim(0-xbuff, xlim[1] + xbuff)\n",
    "        \n",
    "    # Format axis\n",
    "    ax.set_title(label)\n",
    "\n",
    "    if i_col == 0:\n",
    "        ax.set_ylabel(\"Genetic Distance (subs / site)\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "    ax.set_xlabel(\"Great Circle Distance (km)\")\n",
    "    ax.ticklabel_format(axis=\"y\", style=\"sci\", scilimits=(0,0))    \n",
    "    \n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_linewidth(0.5)    \n",
    "            \n",
    "\n",
    "    # Update axis\n",
    "    if i_col == ncol - 1:\n",
    "        i_col = 0\n",
    "        i_row += 1\n",
    "    else:\n",
    "        i_col += 1\n",
    "\n",
    "    \n",
    "out_path = os.path.join(out_dir, \"ibd_clades\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prescribed-stranger",
   "metadata": {},
   "source": [
    "## Map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eight-cover",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=200\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "plt.rc('font', size=FONTSIZE)\n",
    "\n",
    "nrow=3\n",
    "ncol=4\n",
    "\n",
    "fig, axes = plt.subplots(nrow, ncol, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=FIGSIZE, dpi=DPI)\n",
    "fig.subplots_adjust(wspace=0.1, hspace=-0.75)\n",
    "i_row = 0\n",
    "i_col = 0\n",
    "\n",
    "for label in BRANCH_LIST:\n",
    "    \n",
    "    # ----------------------------------------\n",
    "    # Setup axis\n",
    "    ax = axes[i_row][i_col]\n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    \n",
    "    for branch_major in colors_dict[\"branch_major\"]:\n",
    "        if branch_major in label:\n",
    "            break  \n",
    "    \n",
    "    # --------------------------------------\n",
    "    # World Basemap\n",
    "    world_polygons.plot(\n",
    "        ax=ax, \n",
    "        zorder=1, \n",
    "        alpha=0.5, \n",
    "        color=\"grey\",\n",
    "        edgecolor=\"black\", \n",
    "        linewidth=0.1,\n",
    "    ) \n",
    "\n",
    "    df = metadata_gdf[metadata_gdf[\"branch_minor\"].isin(BRANCH_LIST[label])]   \n",
    "    df.plot(ax=ax, ec=\"black\", lw=0.5, color=colors_dict[\"branch_major\"][branch_major], markersize=df[\"geometry_size\"], linewidth=0.5)\n",
    "\n",
    "    # Sorry Antarctica\n",
    "    ax.set_ylim(-60,90)\n",
    "    \n",
    "    ax.set_title(label)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_linewidth(0.5)        \n",
    "    \n",
    "    # Update axis\n",
    "    if i_col == ncol - 1:\n",
    "        i_col = 0\n",
    "        i_row += 1\n",
    "    else:\n",
    "        i_col += 1\n",
    "        \n",
    "out_path = os.path.join(out_dir, \"map_clades\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cultural-template",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------\n",
    "# PLOT SETUP\n",
    "TARGET_RES = [1280, 1280]\n",
    "DPI=400\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "FONTSIZE=5\n",
    "\n",
    "## All\n",
    "fig, ax = plt.subplots(1, subplot_kw={'projection': CRS_EPSG},\n",
    "                        figsize=FIGSIZE, dpi=DPI)\n",
    "world_polygons.plot(\n",
    "    ax=ax, \n",
    "    zorder=1, \n",
    "    alpha=0.5, \n",
    "    color=\"grey\",\n",
    "    edgecolor=\"white\", \n",
    "    linewidth=0.1,\n",
    ") \n",
    "\n",
    "reduce_gdf = copy.deepcopy(metadata_gdf)\n",
    "\n",
    "plotted_coord = []\n",
    "\n",
    "for rec in metadata_gdf.iterrows():\n",
    "    sample=rec[0]\n",
    "    x,y = rec[1][\"geometry\"].x, rec[1][\"geometry\"].y\n",
    "    if [x,y] not in plotted_coord:\n",
    "        plotted_coord.append([x,y])\n",
    "    else:\n",
    "        reduce_gdf.drop(sample, inplace=True)\n",
    "\n",
    "reduce_gdf.plot(\n",
    "    ax=ax, \n",
    "    ec=\"black\", \n",
    "    color=reduce_gdf[\"branch_major_color\"], \n",
    "    markersize=reduce_gdf[\"geometry_size\"], \n",
    "    linewidth=0.25,\n",
    "    alpha=0.80,\n",
    ")\n",
    "\n",
    "#ax.axis(\"off\")\n",
    "for spine in ax.spines:\n",
    "    ax.spines[spine].set_linewidth(0.25)\n",
    "\n",
    "# Sorry Antarctica\n",
    "ax.set_ylim(-60,90)\n",
    "\n",
    "colors_dict_known = copy.copy(colors_dict)\n",
    "del colors_dict_known[\"branch_major\"][UNKNOWN_CHAR]\n",
    "\n",
    "clade_lines = [lines.Line2D([0], [0], color=colors_dict_known[\"branch_major\"][state], lw=0, markersize=2, markeredgecolor=\"black\", markeredgewidth=0.25, marker=\"o\") \n",
    "                for state in colors_dict_known[\"branch_major\"]]\n",
    "clade_labels = [state for state in colors_dict_known[\"branch_major\"]]\n",
    "\n",
    "clade_legend = fig.legend(clade_lines, \n",
    "           clade_labels,\n",
    "           #ncol=len(clade_labels),\n",
    "           ncol=1,\n",
    "           loc=\"center\", \n",
    "           bbox_to_anchor=(0.2, 0.475),\n",
    "           frameon=True,\n",
    "            borderpad=0.5, \n",
    "            handletextpad=0.25, \n",
    "           labelspacing=0.5,\n",
    "           edgecolor=\"black\",\n",
    "           title=\"Clade\",\n",
    "           fontsize=FONTSIZE / 2,\n",
    "           title_fontsize=FONTSIZE / 1.5,\n",
    "          )\n",
    "clade_legend.get_frame().set_linewidth(0.25)\n",
    "\n",
    "# Size Legend\n",
    "s = [ax.scatter(0,-9000, s=1, ec=\"black\", c=\"white\", lw=0.25),\n",
    "     ax.scatter(0,-9000, s=10, ec=\"black\", c=\"white\", lw=0.25),  \n",
    "    ]\n",
    "sl = [\"1\", \"10\"]\n",
    "leg_size = plt.legend(s, sl, title=\"Genomes\", loc=\"center\",  borderpad=0.5, bbox_to_anchor=[0.2,0.35], edgecolor=\"black\",\n",
    "                    fontsize=FONTSIZE / 2, title_fontsize=FONTSIZE / 1.5)\n",
    "leg_size.get_frame().set_linewidth(0.25)\n",
    "\n",
    "out_path = os.path.join(out_dir, \"map_all\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "following-gauge",
   "metadata": {},
   "source": [
    "## Modern Human Plague Cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "advisory-franchise",
   "metadata": {},
   "outputs": [],
   "source": [
    "TARGET_RES = [1280, 1280]\n",
    "DPI=400\n",
    "FIGSIZE=[TARGET_RES[0] / DPI, TARGET_RES[1] / DPI]\n",
    "plt.rc(\"legend\", title_fontsize=MED_FONT)  # fontsize of the legend title\n",
    "plt.rc(\"legend\", fontsize=MED_FONT)  # fontsize of the legend title\n",
    "\n",
    "fig, ax = plt.subplots(1, dpi=DPI, figsize=FIGSIZE)\n",
    "# --------------------------------------\n",
    "# World Basemap\n",
    "world_poly = world_polygons.plot(\n",
    "    ax=ax, \n",
    "    zorder=1, \n",
    "    alpha=0.5, \n",
    "    color=\"grey\",\n",
    "    edgecolor=\"white\", \n",
    "    linewidth=0.25,\n",
    ")\n",
    "xu_2019_gdf.plot(ax=ax,  markersize=0.5, alpha=0.75, ec=\"black\", lw=0.1, c=D3_PAL[2])\n",
    "#ancient_gdf.plot(ax=ax_map,  markersize=ancient_gdf[\"geometry_size\"], alpha=0.75, ec=\"black\", lw=0.25, c=D3_PAL[1])\n",
    "#modern_gdf.plot(ax=ax_map,  markersize=modern_gdf[\"geometry_size\"], alpha=0.75, ec=\"black\", lw=0.25, c=D3_PAL[0])\n",
    "# Sorry Antarctica\n",
    "ax.set_ylim(-60,90)\n",
    "#ax_map.axis('off')  \n",
    "for spine in ax.spines:\n",
    "    ax.spines[spine].set_linewidth(0.25)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "\n",
    "out_path = os.path.join(out_dir, \"map_xu-2019\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\n",
    "\"\"\"\n",
    "# Origin Legend\n",
    "o = [ax_map.scatter(0,-9000, s=5, ec=\"black\", marker=\"o\", c=D3_PAL[1], lw=0.25),\n",
    "     ax_map.scatter(0,-9000, s=5, ec=\"black\", marker=\"o\", c=D3_PAL[0], lw=0.25),\n",
    "     ax_map.scatter(0,-9000, s=5, ec=\"black\", marker=\"o\", c=D3_PAL[2], lw=0.25),     \n",
    "    ]\n",
    "ol = [\"Ancient Genome\", \"Modern Genome\", \"Modern Human Cases\"]\n",
    "\n",
    "leg_origin = plt.legend(\n",
    "    o, ol, title=\"Type\", loc=\"center\", bbox_to_anchor=[0.35,-0.15], edgecolor=\"black\",\n",
    "    fontsize=FONTSIZE / 2, title_fontsize=FONTSIZE - 2)\n",
    "leg_origin.get_frame().set_linewidth(0.25)\n",
    "\n",
    "# Size Legend\n",
    "s = [ax_map.scatter(0,-9000, s=1, ec=\"black\", c=\"black\", lw=0.5),\n",
    "     ax_map.scatter(0,-9000, s=10, ec=\"black\", c=\"black\", lw=0.5),  \n",
    "    ]\n",
    "sl = [\"1\", \"10\"]\n",
    "leg_size = plt.legend(s, sl, title=\"Genomes\", loc=\"center\", bbox_to_anchor=[0.55,-0.15], edgecolor=\"black\",\n",
    "                    fontsize=FONTSIZE / 2, title_fontsize=FONTSIZE - 2)\n",
    "leg_size.get_frame().set_linewidth(0.25)\n",
    "\n",
    "ax_map.add_artist(leg_origin)\n",
    "\n",
    "out_path = os.path.join(out_dir, \"map_comparison\")\n",
    "plt.savefig(out_path + \".png\", bbox_inches=\"tight\")\n",
    "plt.savefig(out_path + \".svg\", bbox_inches=\"tight\")\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
