{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "resistant-strand",
   "metadata": {
    "tags": [
     "snakemake-job-properties"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/home/ktmeaton/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages', '/mnt/c/Users/ktmea/Projects/plague-phylogeography/workflow/notebooks']); import pickle; snakemake = pickle.loads(b'\\x80\\x03csnakemake.script\\nSnakemake\\nq\\x00)\\x81q\\x01}q\\x02(X\\x05\\x00\\x00\\x00inputq\\x03csnakemake.io\\nInputFiles\\nq\\x04)\\x81q\\x05(Xp\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/clock/all/chromosome_filter50/clock_model_timetree.nwkq\\x06Xg\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/clock/all/chromosome_filter50/clock_model.tsvq\\x07e}q\\x08(X\\x06\\x00\\x00\\x00_namesq\\t}q\\n(X\\x04\\x00\\x00\\x00treeq\\x0bK\\x00N\\x86q\\x0cX\\x03\\x00\\x00\\x00tsvq\\rK\\x01N\\x86q\\x0euX\\x12\\x00\\x00\\x00_allowed_overridesq\\x0f]q\\x10(X\\x05\\x00\\x00\\x00indexq\\x11X\\x04\\x00\\x00\\x00sortq\\x12eh\\x11cfunctools\\npartial\\nq\\x13cbuiltins\\ngetattr\\nq\\x14csnakemake.io\\nNamedlist\\nq\\x15X\\x0f\\x00\\x00\\x00_used_attributeq\\x16\\x86q\\x17Rq\\x18\\x85q\\x19Rq\\x1a(h\\x18)}q\\x1bX\\x05\\x00\\x00\\x00_nameq\\x1ch\\x11sNtq\\x1dbh\\x12h\\x13h\\x18\\x85q\\x1eRq\\x1f(h\\x18)}q h\\x1ch\\x12sNtq!bh\\x0bh\\x06h\\rh\\x07ubX\\x06\\x00\\x00\\x00outputq\"csnakemake.io\\nOutputFiles\\nq#)\\x81q$(Xx\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/mugration/all/chromosome_filter50/mugration_model_timetree.nwkq%Xo\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/mugration/all/chromosome_filter50/mugration_model.tsvq&e}q\\'(h\\t}q((h\\x0bK\\x00N\\x86q)h\\rK\\x01N\\x86q*uh\\x0f]q+(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85q,Rq-(h\\x18)}q.h\\x1ch\\x11sNtq/bh\\x12h\\x13h\\x18\\x85q0Rq1(h\\x18)}q2h\\x1ch\\x12sNtq3bh\\x0bh%h\\rh&ubX\\x06\\x00\\x00\\x00paramsq4csnakemake.io\\nParams\\nq5)\\x81q6}q7(h\\t}q8h\\x0f]q9(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85q:Rq;(h\\x18)}q<h\\x1ch\\x11sNtq=bh\\x12h\\x13h\\x18\\x85q>Rq?(h\\x18)}q@h\\x1ch\\x12sNtqAbubX\\t\\x00\\x00\\x00wildcardsqBcsnakemake.io\\nWildcards\\nqC)\\x81qD(X\\x03\\x00\\x00\\x00allqEX\\n\\x00\\x00\\x00chromosomeqFX\\x02\\x00\\x00\\x0050qGe}qH(h\\t}qI(X\\x0c\\x00\\x00\\x00reads_originqJK\\x00N\\x86qKX\\n\\x00\\x00\\x00locus_nameqLK\\x01N\\x86qMX\\x0c\\x00\\x00\\x00missing_dataqNK\\x02N\\x86qOuh\\x0f]qP(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85qQRqR(h\\x18)}qSh\\x1ch\\x11sNtqTbh\\x12h\\x13h\\x18\\x85qURqV(h\\x18)}qWh\\x1ch\\x12sNtqXbX\\x0c\\x00\\x00\\x00reads_originqYhEX\\n\\x00\\x00\\x00locus_nameqZhFX\\x0c\\x00\\x00\\x00missing_dataq[hGubX\\x07\\x00\\x00\\x00threadsq\\\\K\\x01X\\t\\x00\\x00\\x00resourcesq]csnakemake.io\\nResources\\nq^)\\x81q_(K\\x01K\\x01K\\x01M\\xb8\\x0bK<e}q`(h\\t}qa(X\\x06\\x00\\x00\\x00_coresqbK\\x00N\\x86qcX\\x06\\x00\\x00\\x00_nodesqdK\\x01N\\x86qeX\\x04\\x00\\x00\\x00cpusqfK\\x02N\\x86qgX\\x06\\x00\\x00\\x00mem_mbqhK\\x03N\\x86qiX\\x08\\x00\\x00\\x00time_minqjK\\x04N\\x86qkuh\\x0f]ql(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85qmRqn(h\\x18)}qoh\\x1ch\\x11sNtqpbh\\x12h\\x13h\\x18\\x85qqRqr(h\\x18)}qsh\\x1ch\\x12sNtqtbhbK\\x01hdK\\x01hfK\\x01X\\x06\\x00\\x00\\x00mem_mbquM\\xb8\\x0bhjK<ubX\\x03\\x00\\x00\\x00logqvcsnakemake.io\\nLog\\nqw)\\x81qxX~\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/mugration/all/chromosome_filter50/mugration_model_processed.py.ipynbqya}qz(h\\t}q{X\\x08\\x00\\x00\\x00notebookq|K\\x00N\\x86q}sh\\x0f]q~(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85q\\x7fRq\\x80(h\\x18)}q\\x81h\\x1ch\\x11sNtq\\x82bh\\x12h\\x13h\\x18\\x85q\\x83Rq\\x84(h\\x18)}q\\x85h\\x1ch\\x12sNtq\\x86bh|hyubX\\x06\\x00\\x00\\x00configq\\x87}q\\x88(X\\t\\x00\\x00\\x00sqlite_dbq\\x89X\\x19\\x00\\x00\\x00yersinia_pestis_db.sqliteq\\x8aX\\x19\\x00\\x00\\x00sqlite_select_command_asmq\\x8bX\\xc8\\x00\\x00\\x00SELECT AssemblyFTPGenbank FROM BioSample LEFT JOIN Assembly ON AssemblyBioSampleAccession = BioSampleAccession WHERE (BioSampleComment LIKE \\'%KEEP%Assembly%Modern%\\' AND length(AssemblyFTPGenbank) > 0)q\\x8cX\\x19\\x00\\x00\\x00sqlite_select_command_sraq\\x8dX\\xbe\\x00\\x00\\x00SELECT BioSampleAccession, SRARunAccession FROM BioSample LEFT JOIN SRA ON SRABioSampleAccession = BioSampleAccession WHERE (SRARunAccession = \\'SRR1048902\\' OR SRARunAccession = \\'SRR1048905\\')q\\x8eX\\x1b\\x00\\x00\\x00sqlite_select_command_localq\\x8fXU\\x00\\x00\\x00SELECT BioSampleAccession FROM BioSample WHERE (BioSampleComment LIKE \\'%Local%Test%\\')q\\x90X\\x19\\x00\\x00\\x00sqlite_select_command_refq\\x91X\\xaa\\x00\\x00\\x00SELECT AssemblyFTPGenbank FROM BioSample LEFT JOIN Assembly ON AssemblyBioSampleAccession = BioSampleAccession WHERE (BioSampleComment LIKE \\'%Assembly%Modern%Reference%\\')q\\x92X\\x15\\x00\\x00\\x00max_datasets_assemblyq\\x93K\\x03X\\x10\\x00\\x00\\x00max_datasets_sraq\\x94K\\x02X\\x0c\\x00\\x00\\x00reads_originq\\x95]q\\x96(X\\x08\\x00\\x00\\x00assemblyq\\x97X\\x03\\x00\\x00\\x00sraq\\x98X\\x05\\x00\\x00\\x00localq\\x99eX\\x18\\x00\\x00\\x00detect_repeats_thresholdq\\x9aKZX\\x15\\x00\\x00\\x00detect_repeats_lengthq\\x9bK2X\\x0f\\x00\\x00\\x00reference_locusq\\x9cX\\x08\\x00\\x00\\x00AL590842q\\x9dX\\x14\\x00\\x00\\x00reference_locus_nameq\\x9eX\\n\\x00\\x00\\x00chromosomeq\\x9fX\\x15\\x00\\x00\\x00reference_locus_startq\\xa0X\\x01\\x00\\x00\\x000q\\xa1X\\x13\\x00\\x00\\x00reference_locus_endq\\xa2X\\x07\\x00\\x00\\x004653728q\\xa3X\\t\\x00\\x00\\x00eager_revq\\xa4X\\x05\\x00\\x00\\x002.2.1q\\xa5X\\x15\\x00\\x00\\x00eager_clip_readlengthq\\xa6K#X\\r\\x00\\x00\\x00eager_bwaalnnq\\xa7G?\\x84z\\xe1G\\xae\\x14{X\\r\\x00\\x00\\x00eager_bwaalnlq\\xa8K\\x10X\\x0b\\x00\\x00\\x00eager_otherq\\xa9X\\x0c\\x00\\x00\\x00--mergedonlyq\\xaaX\\x08\\x00\\x00\\x00organismq\\xabX\\x0f\\x00\\x00\\x00Yersinia pestisq\\xacX\\x15\\x00\\x00\\x00eager_forward_adapterq\\xadX@\\x00\\x00\\x00AGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNATCTCGTATGCCGTCTTCTGCTTGq\\xaeX\\x15\\x00\\x00\\x00eager_reverse_adapterq\\xafX:\\x00\\x00\\x00AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATTq\\xb0X\\x10\\x00\\x00\\x00snippy_ctg_depthq\\xb1K\\nX\\x10\\x00\\x00\\x00snippy_bam_depthq\\xb2K\\x03X\\x10\\x00\\x00\\x00snippy_base_qualq\\xb3K\\x14X\\x0f\\x00\\x00\\x00snippy_map_qualq\\xb4K\\x1eX\\x0f\\x00\\x00\\x00snippy_min_fracq\\xb5G?\\xec\\xcc\\xcc\\xcc\\xcc\\xcc\\xcdX\\x10\\x00\\x00\\x00snippy_mask_charq\\xb6X\\x01\\x00\\x00\\x00Xq\\xb7X\\x13\\x00\\x00\\x00snippy_missing_dataq\\xb8K2X\\x12\\x00\\x00\\x00snippy_snp_densityq\\xb9K\\nX\\x15\\x00\\x00\\x00snippy_keep_singletonq\\xbaX\\x10\\x00\\x00\\x00--keep-singletonq\\xbbX\\x1e\\x00\\x00\\x00snippy_multi_plot_missing_dataq\\xbc]q\\xbd(K\\x00K\\x19K2KKKdeX\\x0c\\x00\\x00\\x00iqtree_modelq\\xbeX\\x06\\x00\\x00\\x00-m MFPq\\xbfX\\x0b\\x00\\x00\\x00iqtree_seedq\\xc0X\\x08\\x00\\x00\\x0047321424q\\xc1X\\x0f\\x00\\x00\\x00iqtree_outgroupq\\xc2X\\t\\x00\\x00\\x00Referenceq\\xc3X\\x0c\\x00\\x00\\x00iqtree_otherq\\xc4X\\r\\x00\\x00\\x00--ufboot 1000q\\xc5X\\x0b\\x00\\x00\\x00iqtree_runsq\\xc6K\\x01uX\\x04\\x00\\x00\\x00ruleq\\xc7X\\x0f\\x00\\x00\\x00mugration_modelq\\xc8X\\x0f\\x00\\x00\\x00bench_iterationq\\xc9NX\\t\\x00\\x00\\x00scriptdirq\\xcaXD\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/workflow/notebooksq\\xcbub.'); from snakemake.logging import logger; logger.printshellcmds = True; import os; os.chdir('/mnt/c/Users/ktmea/Projects/plague-phylogeography');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "1. Run mugration analysis on select traits.\n",
    "1. Add mugration results to dataframe.\n",
    "1. Plot confidence boxplots.\n",
    "1. Plot colored trees.\n",
    "1. Export:\n",
    "   - Tree Dataframe (tsv)\n",
    "   - Augur JSON\n",
    "   - Auspice JSON"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Module Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "from treetime.utils import parse_dates\n",
    "from treetime import wrappers\n",
    "\n",
    "import copy\n",
    "import os\n",
    "import sys\n",
    "import IPython\n",
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors, lines, patches, gridspec\n",
    "import seaborn as sns\n",
    "\n",
    "import dill"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "\n",
    "# Custom script variables\n",
    "SCRIPT_NAME = \"mugration\"\n",
    "PREV_DIR_NAME = \"clock\"\n",
    "PREV_SCRIPT_NAME = \"clock_model\"\n",
    "\n",
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"50\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "    \n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "MISSING_DATA = WILDCARDS[2]\n",
    "\n",
    "NAME_COL = \"Name\"\n",
    "\n",
    "plt.rc('legend', frameon=False) # legend frame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input File Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tree_dill:\t /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/clock/all/chromosome_filter50/clock_model_timetree.treetime.obj\n",
      "tree_df_dill:\t /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/clock/all/chromosome_filter50/clock_model.df.obj\n",
      "aln path:\t /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/snippy_multi/all/snippy-core_chromosome.snps.filter50.aln\n",
      "auspice_latlon_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome_filter50/parse_tree_latlon.tsv\n",
      "auspice_colors_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome_filter50/parse_tree_colors.tsv\n",
      "auspice_config_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/config/auspice_config.json\n",
      "auspice_remote_dir_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/auspice/\n",
      "outdir: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/mugration/all/chromosome_filter50\n"
     ]
    }
   ],
   "source": [
    "config_dir = os.path.join(project_dir, \"config\")\n",
    "results_dir = os.path.join(project_dir, \"results\")\n",
    "\n",
    "\n",
    "outdir       = os.path.join(results_dir, SCRIPT_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA),)\n",
    "tree_dill    = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA),PREV_SCRIPT_NAME + \"_timetree.treetime.obj\")\n",
    "tree_df_dill = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), PREV_SCRIPT_NAME + \".df.obj\")\n",
    "tree_df_path = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), PREV_SCRIPT_NAME + \".tsv\")\n",
    "aln_path     = os.path.join(results_dir,\"snippy_multi\",READS_ORIGIN,\"snippy-core_{}.snps.filter{}.aln\".format(LOCUS_NAME, MISSING_DATA))\n",
    "\n",
    "# Auspice\n",
    "auspice_latlon_path = os.path.join(results_dir, \"parse_tree\", READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), \"parse_tree\" + \"_latlon.tsv\")\n",
    "auspice_colors_path = os.path.join(results_dir, \"parse_tree\", READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), \"parse_tree\" + \"_colors.tsv\")\n",
    "auspice_config_path = os.path.join(config_dir, \"auspice_config.json\")\n",
    "auspice_remote_dir_path = os.path.join(project_dir, \"auspice/\")\n",
    "\n",
    "print(\"tree_dill:\\t\", tree_dill)\n",
    "print(\"tree_df_dill:\\t\", tree_df_dill)\n",
    "print(\"aln path:\\t\", aln_path)\n",
    "print(\"auspice_latlon_path:\", auspice_latlon_path)\n",
    "print(\"auspice_colors_path:\", auspice_colors_path)\n",
    "print(\"auspice_config_path:\", auspice_config_path)\n",
    "print(\"auspice_remote_dir_path:\", auspice_remote_dir_path)\n",
    "print(\"outdir:\", outdir)\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "while not os.path.exists(outdir):\n",
    "    os.makedirs(outdir)   \n",
    "    \n",
    "SCRIPT_NAME = \"mugration_model\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tree_div = Phylo.read(tree_path, \"newick\")\n",
    "#tree_div.ladderize(reverse=False)\n",
    "\n",
    "with open(tree_dill, \"rb\") as infile:\n",
    "    tt = dill.load(infile)\n",
    "tt.tree.ladderize(reverse=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "      <th>timetree_rate</th>\n",
       "      <th>timetree_rate_fold_change</th>\n",
       "      <th>timetree_mutation_length</th>\n",
       "      <th>timetree_raw_date</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>timetree_num_date_bar</th>\n",
       "      <th>timetree_clock_length</th>\n",
       "      <th>timetree_coord_x</th>\n",
       "      <th>timetree_coord_y</th>\n",
       "      <th>timetree_reg_x</th>\n",
       "      <th>timetree_reg_y</th>\n",
       "      <th>timetree_reg_bad</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>black</td>\n",
       "      <td>2.148815e-07</td>\n",
       "      <td>2.25</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.942141e-08</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.148815e-07</td>\n",
       "      <td>0855-09-28</td>\n",
       "      <td>855.740923</td>\n",
       "      <td>[855.7409234792349, 855.7409234792349]</td>\n",
       "      <td>[1165.2590765207651, 1165.2590765207651]</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>2.25</td>\n",
       "      <td>855.740923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.447080e-05</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.278202e-05</td>\n",
       "      <td>1.00</td>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>2.052208e-08</td>\n",
       "      <td>1.056673</td>\n",
       "      <td>2.256714e-05</td>\n",
       "      <td>1938-03-30</td>\n",
       "      <td>1938.242634</td>\n",
       "      <td>[1923.0, 1923.0]</td>\n",
       "      <td>[98.0, 98.0]</td>\n",
       "      <td>1082.501711</td>\n",
       "      <td>1082.601711</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1923.000000</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.307300e-05</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.149205e-05</td>\n",
       "      <td>2.00</td>\n",
       "      <td>NA</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>NA</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>1.821491e-08</td>\n",
       "      <td>0.937878</td>\n",
       "      <td>2.127716e-05</td>\n",
       "      <td>1996-07-12</td>\n",
       "      <td>1996.529295</td>\n",
       "      <td>[1992.0, 1992.0]</td>\n",
       "      <td>[29.0, 29.0]</td>\n",
       "      <td>1140.788371</td>\n",
       "      <td>1140.888371</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1992.000000</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>2.721210e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>2.536025e-05</td>\n",
       "      <td>3.50</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.136642e-08</td>\n",
       "      <td>1.100148</td>\n",
       "      <td>2.514537e-05</td>\n",
       "      <td>2005-09-05</td>\n",
       "      <td>2005.678813</td>\n",
       "      <td>[1985.457689492314, 2005.0]</td>\n",
       "      <td>[35.5423105076859, 16.0]</td>\n",
       "      <td>1149.937889</td>\n",
       "      <td>1150.037889</td>\n",
       "      <td>3.50</td>\n",
       "      <td>2005.678813</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.536025e-05</td>\n",
       "      <td>3.00</td>\n",
       "      <td>42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>5.812589e-09</td>\n",
       "      <td>0.299288</td>\n",
       "      <td>2.828252e-18</td>\n",
       "      <td>2008-05-10</td>\n",
       "      <td>2008.355574</td>\n",
       "      <td>[2006.0, 2006.0]</td>\n",
       "      <td>[15.0, 15.0]</td>\n",
       "      <td>2.676761</td>\n",
       "      <td>1152.714650</td>\n",
       "      <td>3.00</td>\n",
       "      <td>2006.000000</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.331000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.557514e-05</td>\n",
       "      <td>4.00</td>\n",
       "      <td>42123</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>3.247561e-08</td>\n",
       "      <td>1.672155</td>\n",
       "      <td>2.148899e-07</td>\n",
       "      <td>2007-07-06</td>\n",
       "      <td>2007.510839</td>\n",
       "      <td>[2005.0, 2005.0]</td>\n",
       "      <td>[16.0, 16.0]</td>\n",
       "      <td>1.832026</td>\n",
       "      <td>1151.869916</td>\n",
       "      <td>4.00</td>\n",
       "      <td>2005.000000</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    UFboot sCF  Branch_Length  \\\n",
       "Name                                                            \n",
       "NODE0                                   NA  NA   0.000000e+00   \n",
       "GCA_009909635.1_ASM990963v1_genomic     NA  NA   2.447080e-05   \n",
       "Reference                               NA  NA   2.307300e-05   \n",
       "NODE1                                   91   0   2.721210e-05   \n",
       "GCA_009669545.1_ASM966954v1_genomic     NA  NA   2.330000e-08   \n",
       "GCA_009669555.1_ASM966955v1_genomic     NA  NA   2.331000e-07   \n",
       "\n",
       "                                    Branch_Support_Color       coord_x  \\\n",
       "Name                                                                     \n",
       "NODE0                                              black  2.148815e-07   \n",
       "GCA_009909635.1_ASM990963v1_genomic                 grey  2.278202e-05   \n",
       "Reference                                           grey  2.149205e-05   \n",
       "NODE1                                              black  2.536025e-05   \n",
       "GCA_009669545.1_ASM966954v1_genomic                 grey  2.536025e-05   \n",
       "GCA_009669555.1_ASM966955v1_genomic                 grey  2.557514e-05   \n",
       "\n",
       "                                     coord_y Strain  Date DateBP  \\\n",
       "Name                                                               \n",
       "NODE0                                   2.25     NA    NA     NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic     1.00   9_10  1923    -98   \n",
       "Reference                               2.00     NA  1992    -29   \n",
       "NODE1                                   3.50     NA    NA     NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic     3.00  42126  2006    -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic     4.00  42123  2005    -16   \n",
       "\n",
       "                                                      Country       Province  \\\n",
       "Name                                                                           \n",
       "NODE0                                                      NA             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia  Rostov Oblast   \n",
       "Reference                            United States of America       Colorado   \n",
       "NODE1                                                      NA             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China       Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China       Xinjiang   \n",
       "\n",
       "                                    CountryLat CountryLon ProvinceLat  \\\n",
       "Name                                                                    \n",
       "NODE0                                       NA         NA          NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic    64.6863    97.7453     47.6222   \n",
       "Reference                              39.7837   -100.446     38.7252   \n",
       "NODE1                                       NA         NA          NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic    35.0001        105     42.4805   \n",
       "GCA_009669555.1_ASM966955v1_genomic    35.0001        105     42.4805   \n",
       "\n",
       "                                    ProvinceLon      Biovar Branch_Major  \\\n",
       "Name                                                                       \n",
       "NODE0                                        NA          NA           NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic     40.7958  Medievalis        2.MED   \n",
       "Reference                              -105.608  Orientalis        1.ORI   \n",
       "NODE1                                        NA          NA           NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic     85.4633     Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic     85.4633     Antiqua        0.ANT   \n",
       "\n",
       "                                    Branch_Minor     BioSample  \\\n",
       "Name                                                             \n",
       "NODE0                                         NA            NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1  SAMN13632815   \n",
       "Reference                                 1.ORI1            NA   \n",
       "NODE1                                         NA            NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1  SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1  SAMN07722924   \n",
       "\n",
       "                                                    BioSampleComment  \\\n",
       "Name                                                                   \n",
       "NODE0                                                             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic            KEEP: Assembly Modern   \n",
       "Reference                            KEEP: Assembly Modern Reference   \n",
       "NODE1                                                             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic            KEEP: Assembly Modern   \n",
       "\n",
       "                                    Branch_Number  timetree_rate  \\\n",
       "Name                                                               \n",
       "NODE0                                          NA   1.942141e-08   \n",
       "GCA_009909635.1_ASM990963v1_genomic             2   2.052208e-08   \n",
       "Reference                                       1   1.821491e-08   \n",
       "NODE1                                          NA   2.136642e-08   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0   5.812589e-09   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0   3.247561e-08   \n",
       "\n",
       "                                     timetree_rate_fold_change  \\\n",
       "Name                                                             \n",
       "NODE0                                                 1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   1.056673   \n",
       "Reference                                             0.937878   \n",
       "NODE1                                                 1.100148   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   0.299288   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   1.672155   \n",
       "\n",
       "                                     timetree_mutation_length  \\\n",
       "Name                                                            \n",
       "NODE0                                            2.148815e-07   \n",
       "GCA_009909635.1_ASM990963v1_genomic              2.256714e-05   \n",
       "Reference                                        2.127716e-05   \n",
       "NODE1                                            2.514537e-05   \n",
       "GCA_009669545.1_ASM966954v1_genomic              2.828252e-18   \n",
       "GCA_009669555.1_ASM966955v1_genomic              2.148899e-07   \n",
       "\n",
       "                                    timetree_raw_date  timetree_num_date  \\\n",
       "Name                                                                       \n",
       "NODE0                                      0855-09-28         855.740923   \n",
       "GCA_009909635.1_ASM990963v1_genomic        1938-03-30        1938.242634   \n",
       "Reference                                  1996-07-12        1996.529295   \n",
       "NODE1                                      2005-09-05        2005.678813   \n",
       "GCA_009669545.1_ASM966954v1_genomic        2008-05-10        2008.355574   \n",
       "GCA_009669555.1_ASM966955v1_genomic        2007-07-06        2007.510839   \n",
       "\n",
       "                                               timetree_num_date_confidence  \\\n",
       "Name                                                                          \n",
       "NODE0                                [855.7409234792349, 855.7409234792349]   \n",
       "GCA_009909635.1_ASM990963v1_genomic                        [1923.0, 1923.0]   \n",
       "Reference                                                  [1992.0, 1992.0]   \n",
       "NODE1                                           [1985.457689492314, 2005.0]   \n",
       "GCA_009669545.1_ASM966954v1_genomic                        [2006.0, 2006.0]   \n",
       "GCA_009669555.1_ASM966955v1_genomic                        [2005.0, 2005.0]   \n",
       "\n",
       "                                                        timetree_num_date_bar  \\\n",
       "Name                                                                            \n",
       "NODE0                                [1165.2590765207651, 1165.2590765207651]   \n",
       "GCA_009909635.1_ASM990963v1_genomic                              [98.0, 98.0]   \n",
       "Reference                                                        [29.0, 29.0]   \n",
       "NODE1                                                [35.5423105076859, 16.0]   \n",
       "GCA_009669545.1_ASM966954v1_genomic                              [15.0, 15.0]   \n",
       "GCA_009669555.1_ASM966955v1_genomic                              [16.0, 16.0]   \n",
       "\n",
       "                                     timetree_clock_length  timetree_coord_x  \\\n",
       "Name                                                                           \n",
       "NODE0                                             0.100000          0.100000   \n",
       "GCA_009909635.1_ASM990963v1_genomic            1082.501711       1082.601711   \n",
       "Reference                                      1140.788371       1140.888371   \n",
       "NODE1                                          1149.937889       1150.037889   \n",
       "GCA_009669545.1_ASM966954v1_genomic               2.676761       1152.714650   \n",
       "GCA_009669555.1_ASM966955v1_genomic               1.832026       1151.869916   \n",
       "\n",
       "                                     timetree_coord_y  timetree_reg_x  \\\n",
       "Name                                                                    \n",
       "NODE0                                            2.25      855.740923   \n",
       "GCA_009909635.1_ASM990963v1_genomic              1.00     1923.000000   \n",
       "Reference                                        2.00     1992.000000   \n",
       "NODE1                                            3.50     2005.678813   \n",
       "GCA_009669545.1_ASM966954v1_genomic              3.00     2006.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic              4.00     2005.000000   \n",
       "\n",
       "                                     timetree_reg_y  timetree_reg_bad  \n",
       "Name                                                                   \n",
       "NODE0                                      0.000000             False  \n",
       "GCA_009909635.1_ASM990963v1_genomic        0.000023             False  \n",
       "Reference                                  0.000021             False  \n",
       "NODE1                                      0.000025             False  \n",
       "GCA_009669545.1_ASM966954v1_genomic        0.000025             False  \n",
       "GCA_009669555.1_ASM966955v1_genomic        0.000025             False  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(tree_df_dill, \"rb\") as infile:\n",
    "    tree_df = dill.load(infile)\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import lat lon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geo</th>\n",
       "      <th>Name</th>\n",
       "      <th>Lat</th>\n",
       "      <th>Lon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>country</td>\n",
       "      <td>Russia</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>country</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>39.783730</td>\n",
       "      <td>-100.445882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>country</td>\n",
       "      <td>China</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>province</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>47.622245</td>\n",
       "      <td>40.795794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>province</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>38.725178</td>\n",
       "      <td>-105.607716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>province</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>42.480495</td>\n",
       "      <td>85.463346</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Geo                      Name        Lat         Lon\n",
       "0   country                    Russia  64.686314   97.745306\n",
       "1   country  United States of America  39.783730 -100.445882\n",
       "2   country                     China  35.000074  104.999927\n",
       "3  province             Rostov Oblast  47.622245   40.795794\n",
       "4  province                  Colorado  38.725178 -105.607716\n",
       "5  province                  Xinjiang  42.480495   85.463346"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "latlon_df = pd.read_csv(auspice_latlon_path, sep='\\t', header=None)\n",
    "latlon_df.columns = [\"Geo\",\"Name\",\"Lat\",\"Lon\"]\n",
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "latlon_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "latlon_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Run mugration analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup Mugration Dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "mug_dict = {}\n",
    "\n",
    "\n",
    "tt_copy = copy.deepcopy(tt)\n",
    "# Set branch length to mutations\n",
    "for n in tt_copy.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "tree_div = tt_copy.tree\n",
    "\n",
    "for attr in ATTRIBUTE_LIST:\n",
    "    mug_dict[attr] = {}\n",
    "    # Make a copy of the input tree\n",
    "    mug_dict[attr][\"tree_div\"] = copy.deepcopy(tree_div)\n",
    "    # Map the taxon name to the attribute for mugration analysis\n",
    "    #mug_dict[attr][\"leaf_to_attr\"] = {sample:str(data[attr]) for sample,data in tree_df.iterrows()\n",
    "    #                if data[attr]!=NO_DATA_CHAR and data[attr]}    \n",
    "    mug_dict[attr][\"leaf_to_attr\"] = {sample:str(data[attr]) for sample,data in tree_df.iterrows()\n",
    "                    if data[attr]!=NO_DATA_CHAR}     \n",
    "    # Make blank dicts for the mugration output\n",
    "    mug_dict[attr][\"tree_mug\"] = {}\n",
    "    mug_dict[attr][\"letter_to_state\"] = {}\n",
    "    mug_dict[attr][\"reverse_alphabet\"] = {}\n",
    "    mug_dict[attr][\"unique_states\"] = {}\n",
    "\n",
    "    \n",
    "# Perform necessary type conversions\n",
    "\n",
    "if \"Branch_Number\" in mug_dict.keys():\n",
    "    for sample,attr_val in mug_dict[\"Branch_Number\"][\"leaf_to_attr\"].items():\n",
    "        float_to_str = str(math.ceil(float(attr_val)))\n",
    "        mug_dict[\"Branch_Number\"][\"leaf_to_attr\"][sample] = float_to_str"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run mugration to capture log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running mugration for attribute: Branch_Number\n",
      "Standard output restored, logging to file disabled.\n",
      "Running mugration for attribute: Branch_Major\n",
      "Standard output restored, logging to file disabled.\n",
      "Running mugration for attribute: Branch_Minor\n",
      "Standard output restored, logging to file disabled.\n",
      "Running mugration for attribute: Country\n",
      "Standard output restored, logging to file disabled.\n",
      "Running mugration for attribute: Province\n",
      "Standard output restored, logging to file disabled.\n"
     ]
    }
   ],
   "source": [
    "for attr in ATTRIBUTE_LIST:\n",
    "    # Initialize stdout capture\n",
    "    print(\"Running mugration for attribute: {}\".format(attr))    \n",
    "    old_stdout = sys.stdout\n",
    "    new_stdout = io.StringIO()\n",
    "    sys.stdout = new_stdout\n",
    "    \n",
    "    # Run mugration\n",
    "    mug, letter_to_state, reverse_alphabet = wrappers.reconstruct_discrete_traits(mug_dict[attr][\"tree_div\"], \n",
    "                                                                     traits=mug_dict[attr][\"leaf_to_attr\"], \n",
    "                                                                     missing_data=NO_DATA_CHAR,\n",
    "                                                                     #missing_data=\"nan\",\n",
    "                                                                     #pc=pc, \n",
    "                                                                     #sampling_bias_correction=sampling_bias_correction, \n",
    "                                                                     verbose=4, \n",
    "                                                                     #weights=params.weights\n",
    "                                                                    )\n",
    "    mug_dict[attr][\"tree_mug\"] = mug\n",
    "    mug_dict[attr][\"letter_to_state\"] = letter_to_state\n",
    "    mug_dict[attr][\"reverse_alphabet\"] = reverse_alphabet\n",
    "    mug_dict[attr][\"unique_states\"] = sorted(letter_to_state.values())\n",
    "    \n",
    "    # Save stdout to file\n",
    "    output = new_stdout.getvalue()\n",
    "    out_path = os.path.join(outdir, SCRIPT_NAME + \"_{}.log\".format(attr.lower())) \n",
    "    with open(out_path, \"w\") as file:\n",
    "        file.write(output)\n",
    "    # Restore stdout\n",
    "    sys.stdout = old_stdout\n",
    "    print(\"Standard output restored, logging to file disabled.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. Add mugration to dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "      <th>timetree_rate</th>\n",
       "      <th>timetree_rate_fold_change</th>\n",
       "      <th>timetree_mutation_length</th>\n",
       "      <th>timetree_raw_date</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>timetree_num_date_bar</th>\n",
       "      <th>timetree_clock_length</th>\n",
       "      <th>timetree_coord_x</th>\n",
       "      <th>timetree_coord_y</th>\n",
       "      <th>timetree_reg_x</th>\n",
       "      <th>timetree_reg_y</th>\n",
       "      <th>timetree_reg_bad</th>\n",
       "      <th>Mugration_Branch_Number</th>\n",
       "      <th>Mugration_Branch_Number_Confidence</th>\n",
       "      <th>Mugration_Branch_Number_Entropy</th>\n",
       "      <th>Mugration_Branch_Major</th>\n",
       "      <th>Mugration_Branch_Major_Confidence</th>\n",
       "      <th>Mugration_Branch_Major_Entropy</th>\n",
       "      <th>Mugration_Branch_Minor</th>\n",
       "      <th>Mugration_Branch_Minor_Confidence</th>\n",
       "      <th>Mugration_Branch_Minor_Entropy</th>\n",
       "      <th>Mugration_Country</th>\n",
       "      <th>Mugration_Country_Confidence</th>\n",
       "      <th>Mugration_Country_Entropy</th>\n",
       "      <th>Mugration_Country_Lat</th>\n",
       "      <th>Mugration_Country_Lon</th>\n",
       "      <th>Mugration_Province</th>\n",
       "      <th>Mugration_Province_Confidence</th>\n",
       "      <th>Mugration_Province_Entropy</th>\n",
       "      <th>Mugration_Province_Lat</th>\n",
       "      <th>Mugration_Province_Lon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>black</td>\n",
       "      <td>2.148815e-07</td>\n",
       "      <td>2.25</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.942141e-08</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.148815e-07</td>\n",
       "      <td>0855-09-28</td>\n",
       "      <td>855.740923</td>\n",
       "      <td>[855.7409234792349, 855.7409234792349]</td>\n",
       "      <td>[1165.2590765207651, 1165.2590765207651]</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>2.25</td>\n",
       "      <td>855.740923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>Russia</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>0.379404</td>\n",
       "      <td>1.08856</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.447080e-05</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.278202e-05</td>\n",
       "      <td>1.00</td>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>2.052208e-08</td>\n",
       "      <td>1.056673</td>\n",
       "      <td>2.256714e-05</td>\n",
       "      <td>1938-03-30</td>\n",
       "      <td>1938.242634</td>\n",
       "      <td>[1923.0, 1923.0]</td>\n",
       "      <td>[98.0, 98.0]</td>\n",
       "      <td>1082.501711</td>\n",
       "      <td>1082.601711</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1923.000000</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>Russia</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.307300e-05</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.149205e-05</td>\n",
       "      <td>2.00</td>\n",
       "      <td>NA</td>\n",
       "      <td>1992</td>\n",
       "      <td>-29</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>NA</td>\n",
       "      <td>KEEP: Assembly Modern Reference</td>\n",
       "      <td>1</td>\n",
       "      <td>1.821491e-08</td>\n",
       "      <td>0.937878</td>\n",
       "      <td>2.127716e-05</td>\n",
       "      <td>1996-07-12</td>\n",
       "      <td>1996.529295</td>\n",
       "      <td>[1992.0, 1992.0]</td>\n",
       "      <td>[29.0, 29.0]</td>\n",
       "      <td>1140.788371</td>\n",
       "      <td>1140.888371</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1992.000000</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>39.7837</td>\n",
       "      <td>-100.446</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>38.7252</td>\n",
       "      <td>-105.608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>2.721210e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>2.536025e-05</td>\n",
       "      <td>3.50</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.136642e-08</td>\n",
       "      <td>1.100148</td>\n",
       "      <td>2.514537e-05</td>\n",
       "      <td>2005-09-05</td>\n",
       "      <td>2005.678813</td>\n",
       "      <td>[1985.457689492314, 2005.0]</td>\n",
       "      <td>[35.5423105076859, 16.0]</td>\n",
       "      <td>1149.937889</td>\n",
       "      <td>1150.037889</td>\n",
       "      <td>3.50</td>\n",
       "      <td>2005.678813</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>China</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>0.460385</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.536025e-05</td>\n",
       "      <td>3.00</td>\n",
       "      <td>42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>5.812589e-09</td>\n",
       "      <td>0.299288</td>\n",
       "      <td>2.828252e-18</td>\n",
       "      <td>2008-05-10</td>\n",
       "      <td>2008.355574</td>\n",
       "      <td>[2006.0, 2006.0]</td>\n",
       "      <td>[15.0, 15.0]</td>\n",
       "      <td>2.676761</td>\n",
       "      <td>1152.714650</td>\n",
       "      <td>3.00</td>\n",
       "      <td>2006.000000</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>China</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.331000e-07</td>\n",
       "      <td>grey</td>\n",
       "      <td>2.557514e-05</td>\n",
       "      <td>4.00</td>\n",
       "      <td>42123</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>3.247561e-08</td>\n",
       "      <td>1.672155</td>\n",
       "      <td>2.148899e-07</td>\n",
       "      <td>2007-07-06</td>\n",
       "      <td>2007.510839</td>\n",
       "      <td>[2005.0, 2005.0]</td>\n",
       "      <td>[16.0, 16.0]</td>\n",
       "      <td>1.832026</td>\n",
       "      <td>1151.869916</td>\n",
       "      <td>4.00</td>\n",
       "      <td>2005.000000</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>China</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    UFboot sCF  Branch_Length  \\\n",
       "Name                                                            \n",
       "NODE0                                   NA  NA   0.000000e+00   \n",
       "GCA_009909635.1_ASM990963v1_genomic     NA  NA   2.447080e-05   \n",
       "Reference                               NA  NA   2.307300e-05   \n",
       "NODE1                                   91   0   2.721210e-05   \n",
       "GCA_009669545.1_ASM966954v1_genomic     NA  NA   2.330000e-08   \n",
       "GCA_009669555.1_ASM966955v1_genomic     NA  NA   2.331000e-07   \n",
       "\n",
       "                                    Branch_Support_Color       coord_x  \\\n",
       "Name                                                                     \n",
       "NODE0                                              black  2.148815e-07   \n",
       "GCA_009909635.1_ASM990963v1_genomic                 grey  2.278202e-05   \n",
       "Reference                                           grey  2.149205e-05   \n",
       "NODE1                                              black  2.536025e-05   \n",
       "GCA_009669545.1_ASM966954v1_genomic                 grey  2.536025e-05   \n",
       "GCA_009669555.1_ASM966955v1_genomic                 grey  2.557514e-05   \n",
       "\n",
       "                                     coord_y Strain  Date DateBP  \\\n",
       "Name                                                               \n",
       "NODE0                                   2.25     NA    NA     NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic     1.00   9_10  1923    -98   \n",
       "Reference                               2.00     NA  1992    -29   \n",
       "NODE1                                   3.50     NA    NA     NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic     3.00  42126  2006    -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic     4.00  42123  2005    -16   \n",
       "\n",
       "                                                      Country       Province  \\\n",
       "Name                                                                           \n",
       "NODE0                                                      NA             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia  Rostov Oblast   \n",
       "Reference                            United States of America       Colorado   \n",
       "NODE1                                                      NA             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China       Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China       Xinjiang   \n",
       "\n",
       "                                    CountryLat CountryLon ProvinceLat  \\\n",
       "Name                                                                    \n",
       "NODE0                                       NA         NA          NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic    64.6863    97.7453     47.6222   \n",
       "Reference                              39.7837   -100.446     38.7252   \n",
       "NODE1                                       NA         NA          NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic    35.0001        105     42.4805   \n",
       "GCA_009669555.1_ASM966955v1_genomic    35.0001        105     42.4805   \n",
       "\n",
       "                                    ProvinceLon      Biovar Branch_Major  \\\n",
       "Name                                                                       \n",
       "NODE0                                        NA          NA           NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic     40.7958  Medievalis        2.MED   \n",
       "Reference                              -105.608  Orientalis        1.ORI   \n",
       "NODE1                                        NA          NA           NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic     85.4633     Antiqua        0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic     85.4633     Antiqua        0.ANT   \n",
       "\n",
       "                                    Branch_Minor     BioSample  \\\n",
       "Name                                                             \n",
       "NODE0                                         NA            NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic       2.MED1  SAMN13632815   \n",
       "Reference                                 1.ORI1            NA   \n",
       "NODE1                                         NA            NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic       0.ANT1  SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic       0.ANT1  SAMN07722924   \n",
       "\n",
       "                                                    BioSampleComment  \\\n",
       "Name                                                                   \n",
       "NODE0                                                             NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic            KEEP: Assembly Modern   \n",
       "Reference                            KEEP: Assembly Modern Reference   \n",
       "NODE1                                                             NA   \n",
       "GCA_009669545.1_ASM966954v1_genomic            KEEP: Assembly Modern   \n",
       "GCA_009669555.1_ASM966955v1_genomic            KEEP: Assembly Modern   \n",
       "\n",
       "                                    Branch_Number  timetree_rate  \\\n",
       "Name                                                               \n",
       "NODE0                                          NA   1.942141e-08   \n",
       "GCA_009909635.1_ASM990963v1_genomic             2   2.052208e-08   \n",
       "Reference                                       1   1.821491e-08   \n",
       "NODE1                                          NA   2.136642e-08   \n",
       "GCA_009669545.1_ASM966954v1_genomic             0   5.812589e-09   \n",
       "GCA_009669555.1_ASM966955v1_genomic             0   3.247561e-08   \n",
       "\n",
       "                                     timetree_rate_fold_change  \\\n",
       "Name                                                             \n",
       "NODE0                                                 1.000000   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   1.056673   \n",
       "Reference                                             0.937878   \n",
       "NODE1                                                 1.100148   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   0.299288   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   1.672155   \n",
       "\n",
       "                                     timetree_mutation_length  \\\n",
       "Name                                                            \n",
       "NODE0                                            2.148815e-07   \n",
       "GCA_009909635.1_ASM990963v1_genomic              2.256714e-05   \n",
       "Reference                                        2.127716e-05   \n",
       "NODE1                                            2.514537e-05   \n",
       "GCA_009669545.1_ASM966954v1_genomic              2.828252e-18   \n",
       "GCA_009669555.1_ASM966955v1_genomic              2.148899e-07   \n",
       "\n",
       "                                    timetree_raw_date  timetree_num_date  \\\n",
       "Name                                                                       \n",
       "NODE0                                      0855-09-28         855.740923   \n",
       "GCA_009909635.1_ASM990963v1_genomic        1938-03-30        1938.242634   \n",
       "Reference                                  1996-07-12        1996.529295   \n",
       "NODE1                                      2005-09-05        2005.678813   \n",
       "GCA_009669545.1_ASM966954v1_genomic        2008-05-10        2008.355574   \n",
       "GCA_009669555.1_ASM966955v1_genomic        2007-07-06        2007.510839   \n",
       "\n",
       "                                               timetree_num_date_confidence  \\\n",
       "Name                                                                          \n",
       "NODE0                                [855.7409234792349, 855.7409234792349]   \n",
       "GCA_009909635.1_ASM990963v1_genomic                        [1923.0, 1923.0]   \n",
       "Reference                                                  [1992.0, 1992.0]   \n",
       "NODE1                                           [1985.457689492314, 2005.0]   \n",
       "GCA_009669545.1_ASM966954v1_genomic                        [2006.0, 2006.0]   \n",
       "GCA_009669555.1_ASM966955v1_genomic                        [2005.0, 2005.0]   \n",
       "\n",
       "                                                        timetree_num_date_bar  \\\n",
       "Name                                                                            \n",
       "NODE0                                [1165.2590765207651, 1165.2590765207651]   \n",
       "GCA_009909635.1_ASM990963v1_genomic                              [98.0, 98.0]   \n",
       "Reference                                                        [29.0, 29.0]   \n",
       "NODE1                                                [35.5423105076859, 16.0]   \n",
       "GCA_009669545.1_ASM966954v1_genomic                              [15.0, 15.0]   \n",
       "GCA_009669555.1_ASM966955v1_genomic                              [16.0, 16.0]   \n",
       "\n",
       "                                     timetree_clock_length  timetree_coord_x  \\\n",
       "Name                                                                           \n",
       "NODE0                                             0.100000          0.100000   \n",
       "GCA_009909635.1_ASM990963v1_genomic            1082.501711       1082.601711   \n",
       "Reference                                      1140.788371       1140.888371   \n",
       "NODE1                                          1149.937889       1150.037889   \n",
       "GCA_009669545.1_ASM966954v1_genomic               2.676761       1152.714650   \n",
       "GCA_009669555.1_ASM966955v1_genomic               1.832026       1151.869916   \n",
       "\n",
       "                                     timetree_coord_y  timetree_reg_x  \\\n",
       "Name                                                                    \n",
       "NODE0                                            2.25      855.740923   \n",
       "GCA_009909635.1_ASM990963v1_genomic              1.00     1923.000000   \n",
       "Reference                                        2.00     1992.000000   \n",
       "NODE1                                            3.50     2005.678813   \n",
       "GCA_009669545.1_ASM966954v1_genomic              3.00     2006.000000   \n",
       "GCA_009669555.1_ASM966955v1_genomic              4.00     2005.000000   \n",
       "\n",
       "                                     timetree_reg_y  timetree_reg_bad  \\\n",
       "Name                                                                    \n",
       "NODE0                                      0.000000             False   \n",
       "GCA_009909635.1_ASM990963v1_genomic        0.000023             False   \n",
       "Reference                                  0.000021             False   \n",
       "NODE1                                      0.000025             False   \n",
       "GCA_009669545.1_ASM966954v1_genomic        0.000025             False   \n",
       "GCA_009669555.1_ASM966955v1_genomic        0.000025             False   \n",
       "\n",
       "                                    Mugration_Branch_Number  \\\n",
       "Name                                                          \n",
       "NODE0                                                     1   \n",
       "GCA_009909635.1_ASM990963v1_genomic                       2   \n",
       "Reference                                                 1   \n",
       "NODE1                                                     0   \n",
       "GCA_009669545.1_ASM966954v1_genomic                       0   \n",
       "GCA_009669555.1_ASM966955v1_genomic                       0   \n",
       "\n",
       "                                    Mugration_Branch_Number_Confidence  \\\n",
       "Name                                                                     \n",
       "NODE0                                                         0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                                  1   \n",
       "Reference                                                            1   \n",
       "NODE1                                                         0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                                  1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                                  1   \n",
       "\n",
       "                                    Mugration_Branch_Number_Entropy  \\\n",
       "Name                                                                  \n",
       "NODE0                                                       1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    -1.00009e-12   \n",
       "Reference                                              -1.00009e-12   \n",
       "NODE1                                                      0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic                    -9.99867e-13   \n",
       "GCA_009669555.1_ASM966955v1_genomic                    -9.99867e-13   \n",
       "\n",
       "                                    Mugration_Branch_Major  \\\n",
       "Name                                                         \n",
       "NODE0                                                1.ORI   \n",
       "GCA_009909635.1_ASM990963v1_genomic                  2.MED   \n",
       "Reference                                            1.ORI   \n",
       "NODE1                                                0.ANT   \n",
       "GCA_009669545.1_ASM966954v1_genomic                  0.ANT   \n",
       "GCA_009669555.1_ASM966955v1_genomic                  0.ANT   \n",
       "\n",
       "                                    Mugration_Branch_Major_Confidence  \\\n",
       "Name                                                                    \n",
       "NODE0                                                        0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                                 1   \n",
       "Reference                                                           1   \n",
       "NODE1                                                        0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                                 1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                                 1   \n",
       "\n",
       "                                    Mugration_Branch_Major_Entropy  \\\n",
       "Name                                                                 \n",
       "NODE0                                                      1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   -1.00009e-12   \n",
       "Reference                                             -1.00009e-12   \n",
       "NODE1                                                     0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   -9.99867e-13   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   -9.99867e-13   \n",
       "\n",
       "                                    Mugration_Branch_Minor  \\\n",
       "Name                                                         \n",
       "NODE0                                               1.ORI1   \n",
       "GCA_009909635.1_ASM990963v1_genomic                 2.MED1   \n",
       "Reference                                           1.ORI1   \n",
       "NODE1                                               0.ANT1   \n",
       "GCA_009669545.1_ASM966954v1_genomic                 0.ANT1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                 0.ANT1   \n",
       "\n",
       "                                    Mugration_Branch_Minor_Confidence  \\\n",
       "Name                                                                    \n",
       "NODE0                                                        0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                                 1   \n",
       "Reference                                                           1   \n",
       "NODE1                                                        0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                                 1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                                 1   \n",
       "\n",
       "                                    Mugration_Branch_Minor_Entropy  \\\n",
       "Name                                                                 \n",
       "NODE0                                                      1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic                   -1.00009e-12   \n",
       "Reference                                             -1.00009e-12   \n",
       "NODE1                                                     0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   -9.99867e-13   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   -9.99867e-13   \n",
       "\n",
       "                                            Mugration_Country  \\\n",
       "Name                                                            \n",
       "NODE0                                                  Russia   \n",
       "GCA_009909635.1_ASM990963v1_genomic                    Russia   \n",
       "Reference                            United States of America   \n",
       "NODE1                                                   China   \n",
       "GCA_009669545.1_ASM966954v1_genomic                     China   \n",
       "GCA_009669555.1_ASM966955v1_genomic                     China   \n",
       "\n",
       "                                    Mugration_Country_Confidence  \\\n",
       "Name                                                               \n",
       "NODE0                                                   0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                            1   \n",
       "Reference                                                      1   \n",
       "NODE1                                                   0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                            1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                            1   \n",
       "\n",
       "                                    Mugration_Country_Entropy  \\\n",
       "Name                                                            \n",
       "NODE0                                                 1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic              -1.00009e-12   \n",
       "Reference                                        -1.00009e-12   \n",
       "NODE1                                                0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic              -1.00009e-12   \n",
       "GCA_009669555.1_ASM966955v1_genomic              -1.00009e-12   \n",
       "\n",
       "                                    Mugration_Country_Lat  \\\n",
       "Name                                                        \n",
       "NODE0                                             64.6863   \n",
       "GCA_009909635.1_ASM990963v1_genomic               64.6863   \n",
       "Reference                                         39.7837   \n",
       "NODE1                                             35.0001   \n",
       "GCA_009669545.1_ASM966954v1_genomic               35.0001   \n",
       "GCA_009669555.1_ASM966955v1_genomic               35.0001   \n",
       "\n",
       "                                    Mugration_Country_Lon Mugration_Province  \\\n",
       "Name                                                                           \n",
       "NODE0                                             97.7453      Rostov Oblast   \n",
       "GCA_009909635.1_ASM990963v1_genomic               97.7453      Rostov Oblast   \n",
       "Reference                                        -100.446           Colorado   \n",
       "NODE1                                                 105           Xinjiang   \n",
       "GCA_009669545.1_ASM966954v1_genomic                   105           Xinjiang   \n",
       "GCA_009669555.1_ASM966955v1_genomic                   105           Xinjiang   \n",
       "\n",
       "                                    Mugration_Province_Confidence  \\\n",
       "Name                                                                \n",
       "NODE0                                                    0.379404   \n",
       "GCA_009909635.1_ASM990963v1_genomic                             1   \n",
       "Reference                                                       1   \n",
       "NODE1                                                    0.875959   \n",
       "GCA_009669545.1_ASM966954v1_genomic                             1   \n",
       "GCA_009669555.1_ASM966955v1_genomic                             1   \n",
       "\n",
       "                                    Mugration_Province_Entropy  \\\n",
       "Name                                                             \n",
       "NODE0                                                  1.08856   \n",
       "GCA_009909635.1_ASM990963v1_genomic               -1.00009e-12   \n",
       "Reference                                         -1.00009e-12   \n",
       "NODE1                                                 0.460385   \n",
       "GCA_009669545.1_ASM966954v1_genomic               -1.00009e-12   \n",
       "GCA_009669555.1_ASM966955v1_genomic               -1.00009e-12   \n",
       "\n",
       "                                    Mugration_Province_Lat  \\\n",
       "Name                                                         \n",
       "NODE0                                              47.6222   \n",
       "GCA_009909635.1_ASM990963v1_genomic                47.6222   \n",
       "Reference                                          38.7252   \n",
       "NODE1                                              42.4805   \n",
       "GCA_009669545.1_ASM966954v1_genomic                42.4805   \n",
       "GCA_009669555.1_ASM966955v1_genomic                42.4805   \n",
       "\n",
       "                                    Mugration_Province_Lon  \n",
       "Name                                                        \n",
       "NODE0                                              40.7958  \n",
       "GCA_009909635.1_ASM990963v1_genomic                40.7958  \n",
       "Reference                                         -105.608  \n",
       "NODE1                                              85.4633  \n",
       "GCA_009669545.1_ASM966954v1_genomic                85.4633  \n",
       "GCA_009669555.1_ASM966955v1_genomic                85.4633  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for attr in ATTRIBUTE_LIST:\n",
    "    # Initialize empty values for the new mugration attribute and its confidence\n",
    "    tree_df[\"Mugration_\" + attr] = [NO_DATA_CHAR for row in range(0,len(tree_df))]\n",
    "    tree_df[\"Mugration_\" + attr + \"_Confidence\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]    \n",
    "    tree_df[\"Mugration_\" + attr + \"_Entropy\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "    \n",
    "    tree_mug = mug_dict[attr][\"tree_mug\"].tree\n",
    "    unique_states = mug_dict[attr][\"unique_states\"]\n",
    "    \n",
    "    # If this attribute is associated with lat,lon\n",
    "    if attr + \"Lat\" in tree_df.columns:\n",
    "        tree_df[\"Mugration_\" + attr + \"_Lat\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]\n",
    "        tree_df[\"Mugration_\" + attr + \"_Lon\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]       \n",
    "\n",
    "    # Iterate through the nodes in the tree\n",
    "    for c in tree_mug.find_clades():\n",
    "        # Store the name and confidence for the state with the most support\n",
    "        state_max_name = mug_dict[attr][\"letter_to_state\"][c.cseq[0]]\n",
    "        state_conf_list = c.marginal_profile[0]\n",
    "        state_max_conf = max(state_conf_list)\n",
    "        c.other = []\n",
    "\n",
    "        # Store all the states and confidence values\n",
    "        for state_name,state_conf in zip(unique_states, state_conf_list):\n",
    "            attr_other = Phylo.PhyloXML.Other(tag=state_name, value=\"{:0.4f}\".format(state_conf), namespace=attr)\n",
    "            c.other.append(attr_other)\n",
    "\n",
    "        # Store the Maximum assigned value\n",
    "        attr_other = Phylo.PhyloXML.Other(tag=state_max_name, value=\"{:0.4f}\".format(state_max_conf), namespace=attr + \"_assign\")\n",
    "        c.other.append(attr_other)\n",
    "    \n",
    "        if c.name in tree_df.index:\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr] = state_max_name\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr + \"_Confidence\"] = state_max_conf\n",
    "            \n",
    "            prob_dist = c.marginal_profile[0]\n",
    "            S = -np.sum(prob_dist * np.log(prob_dist + MUG_TINY))\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr + \"_Entropy\"] = S\n",
    "        \n",
    "        # Add mugration lat lon\n",
    "        if attr + \"Lat\" in tree_df.columns:\n",
    "            # Check for lat and lon\n",
    "            c_geo = tree_df[\"Mugration_\" + attr][c.name]\n",
    "            c_geo_match = latlon_df[(latlon_df[\"Geo\"] == attr.lower()) &\n",
    "                            (latlon_df[\"Name\"] == c_geo)\n",
    "                           ]\n",
    "            c_geo_lat = c_geo_match[\"Lat\"].values[0]\n",
    "            c_geo_lon = c_geo_match[\"Lon\"].values[0]\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr + \"_Lat\"] = c_geo_lat\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr + \"_Lon\"] = c_geo_lon\n",
    "            \n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add Metadata as Comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_to_comment(tt.tree, tree_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Plot confidence boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'for attr in ATTRIBUTE_LIST:\\n    fig, (ax1, ax2) = plt.subplots(2, \\n                               sharex=False, \\n                               gridspec_kw={\\'hspace\\': 0},\\n                               figsize=figsize, \\n                               dpi=dpi,\\n                               constrained_layout=True,\\n                              )\\n    \\n    # --------------------------\\n    # Axis 1 - Number of tips per state\\n    # Exclude samples with no attribute recorded\\n    data = tree_df[tree_df[attr] != NO_DATA_CHAR]\\n    label_order = list(data[attr].value_counts().index)\\n    label_order = [lab for lab in label_order if lab != \"nan\"] + [\"nan\"]\\n    \\n    sns.countplot(data=data, \\n                  x=attr, \\n                  #color=\"blue\", \\n                  edgecolor=\"black\",\\n                  ax=ax1, \\n                  order=label_order\\n                 )\\n    plt.setp(ax1.get_xticklabels(), visible=False)\\n    xticklabels = [item.get_text() for item in ax1.get_xticklabels()]\\n    ax1.set_xticklabels(xticklabels, rotation = 90, ha=\"right\")\\n    ax1.set_xlabel(\"\")\\n    ax1.set_ylabel(\"Number of Samples (Tips)\")\\n    ax1.set_xlim(-1,len(label_order)) \\n    ax1.set_title(attr.replace(\"_\",\" \"))\\n    \\n    # --------------------------\\n    # Axis 2 - Mugration Confidence\\n    # Exclude nodes that are terminals (ie. branch support is grey)\\n    data = tree_df[tree_df[\"Branch_Support_Color\"] != \"grey\"]\\n    \\n    # Fix typing\\n    if attr == \"Branch_Number\":\\n        label_order = [str(math.ceil(lab)) for lab in label_order if lab != \"nan\"] + [\"nan\"]\\n\\n    # Customize outlier style\\n    flierprops = dict(marker=\\'o\\', markerfacecolor=\\'white\\', markersize=2,\\n                      linestyle=\\'none\\', markeredgecolor=\\'black\\')\\n    # Create a boxplot\\n    sns.boxplot(data=data, \\n                  x=\"Mugration_\" + attr, \\n                  y=\"Mugration_\" + attr + \"_Confidence\",\\n                  #color=\"blue\", \\n                  ax=ax2, \\n                  order=label_order,\\n                  flierprops=flierprops)\\n    #plt.setp(ax2.get_xticklabels(), visible=False)\\n    xticklabels = [item.get_text() for item in ax2.get_xticklabels()]\\n    ax2.set_xticklabels(xticklabels, rotation = 90, ha=\"center\")\\n    #ax2.axhline(y=MUG_CONF_THRESH, color=THRESH_COL, linewidth=0.5, linestyle=\\'--\\')\\n    ax2.set_xlabel(\"\")\\n    ax2.set_ylabel(\"Mugration Confidence (Internal Nodes)\")\\n    ax2.set_xlim(-1,len(label_order))\\n    ax2.set_ylim(0, 1.1)\\n    \\n    fig.suptitle(\"Sampling Distribution and Mugration Analysis Confidence\")\\n    \\n    out_path = os.path.join(outdir, SCRIPT_NAME + \"_boxplot_{}.{}\".format(attr.lower(), FMT)) \\n    plt.savefig(out_path, \\n            dpi=dpi, \\n            bbox_inches = \"tight\")'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"for attr in ATTRIBUTE_LIST:\n",
    "    fig, (ax1, ax2) = plt.subplots(2, \n",
    "                               sharex=False, \n",
    "                               gridspec_kw={'hspace': 0},\n",
    "                               figsize=figsize, \n",
    "                               dpi=dpi,\n",
    "                               constrained_layout=True,\n",
    "                              )\n",
    "    \n",
    "    # --------------------------\n",
    "    # Axis 1 - Number of tips per state\n",
    "    # Exclude samples with no attribute recorded\n",
    "    data = tree_df[tree_df[attr] != NO_DATA_CHAR]\n",
    "    label_order = list(data[attr].value_counts().index)\n",
    "    label_order = [lab for lab in label_order if lab != \"nan\"] + [\"nan\"]\n",
    "    \n",
    "    sns.countplot(data=data, \n",
    "                  x=attr, \n",
    "                  #color=\"blue\", \n",
    "                  edgecolor=\"black\",\n",
    "                  ax=ax1, \n",
    "                  order=label_order\n",
    "                 )\n",
    "    plt.setp(ax1.get_xticklabels(), visible=False)\n",
    "    xticklabels = [item.get_text() for item in ax1.get_xticklabels()]\n",
    "    ax1.set_xticklabels(xticklabels, rotation = 90, ha=\"right\")\n",
    "    ax1.set_xlabel(\"\")\n",
    "    ax1.set_ylabel(\"Number of Samples (Tips)\")\n",
    "    ax1.set_xlim(-1,len(label_order)) \n",
    "    ax1.set_title(attr.replace(\"_\",\" \"))\n",
    "    \n",
    "    # --------------------------\n",
    "    # Axis 2 - Mugration Confidence\n",
    "    # Exclude nodes that are terminals (ie. branch support is grey)\n",
    "    data = tree_df[tree_df[\"Branch_Support_Color\"] != \"grey\"]\n",
    "    \n",
    "    # Fix typing\n",
    "    if attr == \"Branch_Number\":\n",
    "        label_order = [str(math.ceil(lab)) for lab in label_order if lab != \"nan\"] + [\"nan\"]\n",
    "\n",
    "    # Customize outlier style\n",
    "    flierprops = dict(marker='o', markerfacecolor='white', markersize=2,\n",
    "                      linestyle='none', markeredgecolor='black')\n",
    "    # Create a boxplot\n",
    "    sns.boxplot(data=data, \n",
    "                  x=\"Mugration_\" + attr, \n",
    "                  y=\"Mugration_\" + attr + \"_Confidence\",\n",
    "                  #color=\"blue\", \n",
    "                  ax=ax2, \n",
    "                  order=label_order,\n",
    "                  flierprops=flierprops)\n",
    "    #plt.setp(ax2.get_xticklabels(), visible=False)\n",
    "    xticklabels = [item.get_text() for item in ax2.get_xticklabels()]\n",
    "    ax2.set_xticklabels(xticklabels, rotation = 90, ha=\"center\")\n",
    "    #ax2.axhline(y=MUG_CONF_THRESH, color=THRESH_COL, linewidth=0.5, linestyle='--')\n",
    "    ax2.set_xlabel(\"\")\n",
    "    ax2.set_ylabel(\"Mugration Confidence (Internal Nodes)\")\n",
    "    ax2.set_xlim(-1,len(label_order))\n",
    "    ax2.set_ylim(0, 1.1)\n",
    "    \n",
    "    fig.suptitle(\"Sampling Distribution and Mugration Analysis Confidence\")\n",
    "    \n",
    "    out_path = os.path.join(outdir, SCRIPT_NAME + \"_boxplot_{}.{}\".format(attr.lower(), FMT)) \n",
    "    plt.savefig(out_path, \n",
    "            dpi=dpi, \n",
    "            bbox_inches = \"tight\")\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 4. Plot colored trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Color branches on mug/div tree according to state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'out_path_colors = os.path.join(outdir, SCRIPT_NAME + \"_colors.tsv\")\\nfile_colors = open(out_path_colors, \"w\")\\n\\nfor attr in ATTRIBUTE_LIST:  \\n    if attr != \"Branch_Major\" and attr != \"Branch_Number\":\\n        continue\\n    \\n    # --------------------------------------------   \\n    # Canvas\\n    fig, ax1 = plt.subplots(1, dpi=dpi, figsize=figsize, constrained_layout=True,)\\n\\n    # --------------------------------------------\\n    # Color Tree\\n    hex_dict = color_tree(tree=tree_div, \\n                          df=tree_df, \\n                          attribute=\"Mugration_\" + attr,\\n                          attribute_confidence=\"Mugration_\" + attr + \"_Confidence\",\\n                          threshold_confidence=MUG_CONF_THRESH,                          \\n                          color_pal=CONT_COLOR_PAL)\\n    \\n    # Add the hex color dict to the dict\\n    mug_dict[attr][\"hex_color\"] = hex_dict  \\n\\n    # Write to color file\\n    for state,color in hex_dict.items():\\n        file_colors.write(attr.lower() + \"\\t\" + state + \"\\t\" + color + \"\\n\")    \\n        \\n    # Draw tree\\n    Phylo.draw(tree_div,\\n               axes=ax1, \\n               show_confidence=False, \\n               label_func = lambda x:\\'\\', \\n               do_show=False)\\n    # --------------------------------------------\\n    # Draw tips\\n    colors = [mug_dict[attr][\"hex_color\"][state] for state in tree_df[\"Mugration_\" + attr]]    \\n    ax1.scatter(data=tree_df, \\n                x=\"coord_x\", \\n                y=\"coord_y\", \\n                s=0.5, \\n                c=colors,\\n               )\\n    # --------------------------------------------\\n    # Ticks\\n    x_buffer = max(tree_df[\"coord_x\"]) * 0.01\\n    y_buffer = math.ceil(len(tree_div.get_terminals()) * 0.01)   \\n    # --------------------------------------------\\n    # Limits\\n    ax1.set_xlim(0 - x_buffer, max(tree_df[\"coord_x\"]) + x_buffer)    \\n    ax1.set_ylim(len(tree_div.get_terminals()) + y_buffer, 0 - y_buffer)\\n    ax1.set_yticks([])    \\n    # --------------------------------------------\\n    # Labels\\n    ax1.set_xlabel(\"Branch Length\")\\n    ax1.set_ylabel(\"\")\\n    fig.suptitle(\"{} Mugration\".format(attr.replace(\"_\",\" \")))    \\n    # --------------------------------------------\\n    # Legend\\n    legend_elements = [patches.Patch(facecolor=value, edgecolor=value,) for value in mug_dict[attr][\"hex_color\"].values()]\\n    legend_labels = list(mug_dict[attr][\"hex_color\"].keys())\\n    ax1.legend(legend_elements, \\n              legend_labels,\\n              bbox_to_anchor=(1.0, 1.0), \\n              loc=\\'upper left\\',\\n              frameon=False,\\n             )\\n    # --------------------------------------------\\n    # Export\\n    out_path = os.path.join(outdir, SCRIPT_NAME + \"_tree_{}.{}\".format(attr.lower(), FMT))  \\n    plt.savefig(out_path, dpi=dpi, bbox_inches = \"tight\")\\n    \\n    # Save Tree!\\n    #out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_{}.phylo.obj\".format(attr.lower()))\\n    #with open(out_path_dill_tree,\"wb\") as outfile:\\n    #    dill.dump(tree_div, outfile)\\n    \\nfile_colors.close()'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"out_path_colors = os.path.join(outdir, SCRIPT_NAME + \"_colors.tsv\")\n",
    "file_colors = open(out_path_colors, \"w\")\n",
    "\n",
    "for attr in ATTRIBUTE_LIST:  \n",
    "    if attr != \"Branch_Major\" and attr != \"Branch_Number\":\n",
    "        continue\n",
    "    \n",
    "    # --------------------------------------------   \n",
    "    # Canvas\n",
    "    fig, ax1 = plt.subplots(1, dpi=dpi, figsize=figsize, constrained_layout=True,)\n",
    "\n",
    "    # --------------------------------------------\n",
    "    # Color Tree\n",
    "    hex_dict = color_tree(tree=tree_div, \n",
    "                          df=tree_df, \n",
    "                          attribute=\"Mugration_\" + attr,\n",
    "                          attribute_confidence=\"Mugration_\" + attr + \"_Confidence\",\n",
    "                          threshold_confidence=MUG_CONF_THRESH,                          \n",
    "                          color_pal=CONT_COLOR_PAL)\n",
    "    \n",
    "    # Add the hex color dict to the dict\n",
    "    mug_dict[attr][\"hex_color\"] = hex_dict  \n",
    "\n",
    "    # Write to color file\n",
    "    for state,color in hex_dict.items():\n",
    "        file_colors.write(attr.lower() + \"\\t\" + state + \"\\t\" + color + \"\\n\")    \n",
    "        \n",
    "    # Draw tree\n",
    "    Phylo.draw(tree_div,\n",
    "               axes=ax1, \n",
    "               show_confidence=False, \n",
    "               label_func = lambda x:'', \n",
    "               do_show=False)\n",
    "    # --------------------------------------------\n",
    "    # Draw tips\n",
    "    colors = [mug_dict[attr][\"hex_color\"][state] for state in tree_df[\"Mugration_\" + attr]]    \n",
    "    ax1.scatter(data=tree_df, \n",
    "                x=\"coord_x\", \n",
    "                y=\"coord_y\", \n",
    "                s=0.5, \n",
    "                c=colors,\n",
    "               )\n",
    "    # --------------------------------------------\n",
    "    # Ticks\n",
    "    x_buffer = max(tree_df[\"coord_x\"]) * 0.01\n",
    "    y_buffer = math.ceil(len(tree_div.get_terminals()) * 0.01)   \n",
    "    # --------------------------------------------\n",
    "    # Limits\n",
    "    ax1.set_xlim(0 - x_buffer, max(tree_df[\"coord_x\"]) + x_buffer)    \n",
    "    ax1.set_ylim(len(tree_div.get_terminals()) + y_buffer, 0 - y_buffer)\n",
    "    ax1.set_yticks([])    \n",
    "    # --------------------------------------------\n",
    "    # Labels\n",
    "    ax1.set_xlabel(\"Branch Length\")\n",
    "    ax1.set_ylabel(\"\")\n",
    "    fig.suptitle(\"{} Mugration\".format(attr.replace(\"_\",\" \")))    \n",
    "    # --------------------------------------------\n",
    "    # Legend\n",
    "    legend_elements = [patches.Patch(facecolor=value, edgecolor=value,) for value in mug_dict[attr][\"hex_color\"].values()]\n",
    "    legend_labels = list(mug_dict[attr][\"hex_color\"].keys())\n",
    "    ax1.legend(legend_elements, \n",
    "              legend_labels,\n",
    "              bbox_to_anchor=(1.0, 1.0), \n",
    "              loc='upper left',\n",
    "              frameon=False,\n",
    "             )\n",
    "    # --------------------------------------------\n",
    "    # Export\n",
    "    out_path = os.path.join(outdir, SCRIPT_NAME + \"_tree_{}.{}\".format(attr.lower(), FMT))  \n",
    "    plt.savefig(out_path, dpi=dpi, bbox_inches = \"tight\")\n",
    "    \n",
    "    # Save Tree!\n",
    "    #out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_{}.phylo.obj\".format(attr.lower()))\n",
    "    #with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    #    dill.dump(tree_div, outfile)\n",
    "    \n",
    "file_colors.close()\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 5. Export"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframe\n",
    "out_path_df = os.path.join(outdir, SCRIPT_NAME + \".tsv\" )\n",
    "out_path_pickle_df = os.path.join(outdir,  SCRIPT_NAME + \".df.obj\" )\n",
    "\n",
    "tree_df.to_csv(out_path_df, sep=\"\\t\")\n",
    "with open(out_path_pickle_df,\"wb\") as outfile:\n",
    "    dill.dump(tree_df, outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Timetrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tt_copy = copy.deepcopy(tt)\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \"_timetree.nexus\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_timetree.treetime.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tt_copy, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tt_copy.tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \"_timetree.nwk\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Divtrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tt_copy = copy.deepcopy(tt)\n",
    "# Convert to divtree\n",
    "for n in tt_copy.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \"_divtree.nexus\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_divtree.treetime.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tt_copy, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tt_copy.tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \"_divtree.nwk\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Augur JSON\n",
    "\n",
    "  - alignment (empty)\n",
    "  - input_tree (tree_path)\n",
    "  - nodes (node_dict)\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ufboot': 'NA', 'scf': 'NA', 'branch_length': 0.0, 'strain': 'NA', 'date': 'NA', 'datebp': 'NA', 'country': 'Russia', 'province': 'Rostov Oblast', 'biovar': 'NA', 'branch_major': '1.ORI', 'branch_minor': '1.ORI1', 'biosample': 'NA', 'biosamplecomment': 'NA', 'branch_number': '1', 'rate': 1.942140670161841e-08, 'rate_fold_change': 1.0, 'mutation_length': 2.148814885614286e-07, 'raw_date': '0855-09-28', 'num_date': 855.7409234792349, 'num_date_confidence': [855.7409234792349, 855.7409234792349], 'num_date_bar': '1165.2590765207651:1165.2590765207651', 'clock_length': 0.1, 'branch_number_confidence': {'1': 0.37940443433527377}, 'branch_number_entropy': 1.0885629792884624, 'branch_major_confidence': {'1.ORI': 0.37940443433527377}, 'branch_major_entropy': 1.0885629792884624, 'branch_minor_confidence': {'1.ORI1': 0.37940443433527377}, 'branch_minor_entropy': 1.0885629792884624, 'country_confidence': {'Russia': 0.3794044343352737}, 'country_entropy': 1.0885629792884626, 'province_confidence': {'Rostov Oblast': 0.3794044343352744}, 'province_entropy': 1.0885629792884624}\n"
     ]
    }
   ],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=out_path_nwk, \n",
    "    aln_path=aln_path, \n",
    "    tree=tt.tree,     \n",
    "    tree_df=tree_df, \n",
    "    color_keyword_exclude=[\"color\", \"coord\", \"reg\", \"lat\", \"lon\"],\n",
    "    type_convert = {\n",
    "        \"Branch_Number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"NODE0\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Mugration Model to JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "mug_models_dict = {}\n",
    "\n",
    "for attr in ATTRIBUTE_LIST:\n",
    "    mug_models_dict[attr] = {}\n",
    "    \n",
    "    # Mugration GTR model\n",
    "    gtr = mug_dict[attr][\"tree_mug\"].gtr\n",
    "    \n",
    "    # Mugration Rate\n",
    "    mug_models_dict[attr][\"rate\"] = gtr.mu\n",
    "    \n",
    "    # Mugration states\n",
    "    alphabet = mug_dict[attr][\"letter_to_state\"]\n",
    "    mug_models_dict[attr][\"alphabet\"] = [alphabet[k] for k in sorted(alphabet.keys())]\n",
    "    \n",
    "    # Mugration probabilities\n",
    "    prob =  list(mug_dict[attr][\"tree_mug\"].gtr.Pi)\n",
    "    mug_models_dict[attr][\"equilibrium_probabilities\"] = list(gtr.Pi)\n",
    "    \n",
    "    # Mugration transition matrix\n",
    "    mat = [list(x) for x in gtr.W]\n",
    "    mug_models_dict[attr][\"equilibrium_probabilities\"] = mat\n",
    "    \n",
    "augur_dict[\"models\"] = mug_models_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ufboot': 'NA', 'scf': 'NA', 'branch_length': 2.3073e-05, 'strain': 'NA', 'date': 1992.0, 'datebp': -29.0, 'country': 'United States of America', 'province': 'Colorado', 'biovar': 'Orientalis', 'branch_major': '1.ORI', 'branch_minor': '1.ORI1', 'biosample': 'NA', 'biosamplecomment': 'KEEP: Assembly Modern Reference', 'branch_number': '1', 'rate': 1.8214911688180126e-08, 'rate_fold_change': 0.9378780830876815, 'mutation_length': 2.1277164143616335e-05, 'raw_date': '1996-07-12', 'num_date': 1996.529294613252, 'num_date_confidence': [1992.0, 1992.0], 'num_date_bar': '29.0:29.0', 'clock_length': 1140.788371134017, 'branch_number_confidence': {'1': 1.0}, 'branch_number_entropy': -1.000088900581841e-12, 'branch_major_confidence': {'1.ORI': 1.0}, 'branch_major_entropy': -1.000088900581841e-12, 'branch_minor_confidence': {'1.ORI1': 1.0}, 'branch_minor_entropy': -1.000088900581841e-12, 'country_confidence': {'United States of America': 1.0}, 'country_entropy': -1.000088900581841e-12, 'province_confidence': {'Colorado': 1.0}, 'province_entropy': -1.000088900581841e-12}\n"
     ]
    }
   ],
   "source": [
    "out_path_augur_json = os.path.join(outdir, SCRIPT_NAME + \"_augur.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Auspice JSON\n",
    "\n",
    "Manual edits of https://github.com/nextstrain/augur/blob/master/augur/export_v2.py\n",
    "\n",
    "This can then be used for auspice via:\n",
    "\n",
    "```\n",
    "HOST=\"localhost\" auspice view --datasetDir .\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/mugration/all/chromosome_filter50/mugration_model_auspice.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/mugration/all/chromosome_filter50/mugration_model_auspice.json' succeeded.\n",
      "Validation successful for local JSON.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'chrom_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pcd1_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pmt1_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'ppcp1_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pla_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'chrom_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pcd1_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pmt1_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'ppcp1_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pla_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'ppcp1_pla_dep_ratio', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'num_date_bar' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_number were not specified in your provided color scale: 0, 2. Auspice will create colors for them.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "auspice_dict = auspice_export(\n",
    "    tree=tree_div, \n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=auspice_colors_path,\n",
    "    auspice_latlons_path=auspice_latlon_path, \n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(outdir, SCRIPT_NAME + \"_auspice.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "#out_path_auspice_remote_json = os.path.join(auspice_remote_dir_path, AUSPICE_PREFIX + SCRIPT_NAME.replace(\"_\",\"-\") + \".json\" )\n",
    "#utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "#export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "#print(\"Validation successful for remote JSON.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
